/* @source psiblasts application
**
** Runs PSI-BLAST given a seed alignment
** @author: Copyright (C) Ranjeeva Ranasinghe (rranasin@hgmp.mrc.ac.uk)
** @author: Copyright (C) Jon Ison (jison@hgmp.mrc.ac.uk)
** @@
**
** This program is free software; you can redistribute it and/or
** modify it under the terms of the GNU General Public License
** as published by the Free Software Foundation; either version 2
** of the License, or (at your option) any later version.
** 
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
** 
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
**
** 
** 
** 
** 
** 
** 
** Operation 
** 
** psiblasts parses a stamp alignment annotated with SCOP classification 
** records, such as those generated by the EMBOSS applications scopalign, and 
** using the alignment as a seed, calls PSIBLAST to search a sequence database. 
** It writes a file of hits (relatives to proteins of each alignment) in 
** embl-like format. 
** 
** PSIBLAST is run on each alignment in turn. The output files are parsed and 
** those hits achieving less than a specified threshold E-value are written to 
** the file of hits. 
** 
** The output file (Figure 1) uses the following records 
** (1) DE - bibliographic information. The text 'Members of scop families' is 
** always given. 
** (2) EX - experimental information. The value of the threshold E-value is 
** is given as a floating point number after 'THRESH'. The name of the residue 
** substitution matrix is given after 'MATRIX'. 
** The following four SCOP classification records are taken from the alignment 
** input file: 
** (3)  CL - Domain class.  It is identical to the text given after 'Class' in 
** the scop classification file (see documentation for the EMBOSS application 
** scope). 
** (4)  FO - Domain fold.  It is identical to the text given after 'Fold' in 
** the scop classification file (see scope documentation). 
** (5)  SF - Domain superfamily.  It is identical to the text given after 
** 'Superfamily' in the scop classification file (see scope documentation). 
** (6)  FA - Domain family. It is identical to the text given after 'Family' in 
** the scop classification file (see scope documentation). 
** (7)  NS - Number in set. The number of hits for this family. The file will 
** have a section containing an NN, AC, CL, RA and SQ records (see below) for 
** each sequence in the set. 
** (8) NN - Sequence number.  The number given in brackets after this record 
** indicates the start of the data for the relevent sequence. 
** (9) AC - Database identifier code of the hit. 
** (10) TY - Classification of hit.  Has the value PSIBLAST (for sequences retrieved 
** by psi-blast) or TRAIN (if the sequence was included in the seed alignment). 
** (Files of this type are also generated by the EMBOSS application swissparse 
** and may be be hand edited with additional sequences, in either case, the 
** value OTHER will be given for this record). 
** (11) RA - Sequence range. The numbers before START and END give the start and 
** end positions respectively of the hit relative to the full length sequence 
** in the swissprot database (a '.' may be given for swissparse output files - 
** see swissparse  documentation). 
** (12)  SQ - protein sequence. The number of residues is given before AA on the 
** first line. The protein sequence is given on subsequent lines. 
** (13) XX - used for spacing. 
** (14) // - used to delimit data for each sequence. 
** 
** 
** Figure 1  Excerpt from psiblasts output file 
** 
**  DE   Members of scop families 
**  XX 
**  EX   THRESH 0.001; MATRIX BLOSUM62; 
**  XX 
**  CL   All alpha proteins 
**  XX 
**  FO   Globin-like 
**  XX 
**  SF   Globin-like 
**  XX 
**  FA   Globins 
**  XX 
**  NS   1 
**  XX 
**  NN   [1] 
**  XX 
**  AC   HBDEX1 
**  XX 
**  CL   PSIBLAST
**  XX 
**  RA   2 START; 79 END; 
**  XX 
**  SQ   SEQUENCE   141 AA;  15127 MW;  5EC7DB1E CRC32; 
**       VLSPADKTNV KAAWGKVGAH AGEYGAEALE RMFLSFPTTK TYFPHFDLSH GSAQVKGHGK 
**       KVADALTNAV AHVDDMPNAL SALSDLHAHK LRVDPVNFKL LSHCLLVTLA AHLPAEFTPA 
**       VHASLDKFLA SVSTVLTSKY R 
**  XX 
**  NN   [2] 
**  XX 
**  AC   HBDEX2 
**  XX 
**  CL   PSIBLAST 
**  XX 
**  RA   2 START; 79 END; 
**  XX 
**  SQ   SEQUENCE   141 AA;  15127 MW;  5EC7DB1E CRC32; 
**       VLSPADKTNV KAAWGKVGAH AGEYGAEALE RMFLSFPTTK TYFPHFDLSH GSAQVKGHGK 
**       KVADALTNAV AHVDDMPNAL SALSDLHAHK LRVDPVNFKL LSHCLLVTLA AHLPAEFTPA 
**       VHASLDKFLA SVSTVLTSKY R 
**  XX 
**  // 
**  CL   All alpha proteins 
**  XX 
**  FO   Globin-like 
**  XX 
**  SF   Globin-like 
**  XX 
**  FA   Phycocyanins 
**  XX 
** 
** 
** 
**  Notes 
** 
**  Must implement better way of producing value of TRAIN for TY record - 
**  which uses file of correspondence between SCOP domains and SWISPROT 
**  sequences ?
**
**  Important Note
**
**  WHEN RUNNING PSIBLASTS AT THE HGMP IT IS ESSENTIAL THAT THE COMMAND 
**  'use blast_v2' (which runs the script /packages/menu/USE/blast_v2) IS GIVEN 
**  BEFORE IT IS RUN. 
******************************************************************************/ 
        

#include "emboss.h"

/*****************************************************************************
 **									    **
 ** This routine takes the name of an alignment file in clustal format and  **
 ** parses out the sequences and returns a AjPList containing just these    **
 ** sequences.            						    **
 **									    **
 *****************************************************************************/

/* @func SeqGet ***************************************************************
**
** This routine takes the name of an alignment file in clustal format and
** parses out the sequences and returns a AjPList containing just these
** sequences.
**
** @param [?] alignf [AjPStr] Undocumented
** @return [AjPList] Undocumented
** @@
******************************************************************************/

AjPList SeqGet(AjPStr alignf)
{

    AjPFile file        = NULL;	   /* The scopalign generated alignment file */

    AjBool  done_1st_blk= ajFalse; /* Flag for whether we've read first block of sequences */

    AjPList list_seqs   = NULL;    /* List of sequences */

    AjPStr  *arr_seqs   = NULL;    /* Array of sequences */
    AjPStr  seq         = NULL;
    AjPStr  seq1        = NULL;
    AjPStr  line        = NULL;

    ajint   cnt         = 0;       /* Counter for sequence */
    ajint   x           = 0;       /* Loop counter */
    ajint   y           = 0;

    line = ajStrNew();
    
    /* Create a new list */
    list_seqs = ajListstrNew();
    file      = ajFileNewIn(alignf);
    
    /* Start of code for reading input file */
    /*Ignore everything up to first line beginning with 'Number'*/
    while(ajFileReadLine(file,&line))
    {
	if(ajStrPrefixC(line,"Number"))
	    break;
    }

    /* Read the rest of the file */
    while(ajFileReadLine(file,&line))
    {
	/* Ignore 'Number' and 'Post_similar' lines */
	if((ajStrPrefixC(line,"Number"))||
	   (ajStrPrefixC(line,"Post_similar")))
	    continue;

	/* If we are on a blank line */
	/* ajFileReadLine will trim the tailing \n */
	if(ajStrChar(line,1)=='\0')
	{ 
	    done_1st_blk=ajTrue;
	    cnt = 0;
	    y++;

	    if(y == 1)
	    {
		x = ajListstrToArray(list_seqs, &arr_seqs); 
	    }
	    continue;
	}

	if(done_1st_blk == ajTrue)
	{
	    ajFmtScanS(line, "%*s%S", &seq1);
	    ajStrApp(&arr_seqs[cnt], seq1);
	    cnt++;
	    continue;
	}

	/* It is a sequence line from the first block */
	if(done_1st_blk==ajFalse)
	{
	    /* The first block of sequences */
	    /* Read in sequence */
	    seq = ajStrNew();		
	    ajFmtScanS(line, "%*s%S", &seq);
	    	    
	    /* Push string onto list */
	    ajListstrPushApp(list_seqs,seq);
	    continue;
	}
	
    }
    ajListstrDel(&list_seqs); 
    list_seqs = ajListstrNew();
   
    for(x=0;x<cnt;x++)
    {
	seq = ajStrNew();
	ajStrSubstituteCC(&arr_seqs[x],"-","");
	ajStrAss(&seq,arr_seqs[x]);
	ajListstrPushApp(list_seqs,seq);
    } 
        
    /* Free up the list */
    ajFileClose(&file);
    
    ajStrDel(&seq);
    ajStrDel(&seq1);
    ajStrDel(&line);
    
    return list_seqs;
    ajExit();

}

/* @prog psiblasts *******************************************************
**
** Testing
**
******************************************************************************/

int main(int argc, char **argv)
{
    AjPStr  align       = NULL;		/* Location of alignment files for input */
    AjPStr  alignextn   = NULL;		/* File extension of alignment files */
    AjPStr  database    = NULL;		/* Name of BLAST database to search */
    AjPStr  submatrix   = NULL;		/* Name of residue substitution matrix */
    AjPStr  temp        = NULL;		/* Temp string */
    AjPStr  line        = NULL;		/* Temp string */ 
    AjPStr  sub         = NULL;		/* Temp string */
    AjPStr  seq         = NULL;		/* Holds a complete sequence */
    AjPStr *seqs        = NULL;		
    AjPStr  msg         = NULL;		/* Error message */
    AjPStr  cmdline     = NULL;		/* Command line arguments to run the psi-blast program */
    AjPStr  start       = NULL;		/* Begining of sequence */
    AjPStr  end         = NULL;		/* End of sequence */
    AjPStr  sequence    = NULL;         /* Holds the hit region */ 
        
    ajint     niter     = 0;		/* Number of PSIBLAST iterations */
    ajint     nscore    = 0;		/* Number of sequece blocks per entry */
    ajint     nlines    = 0;
    ajint     hitcounter= 0;		/* counts the number of hits */
    ajint     setcounter= 0;		/* counts the total number of hits per alignment file */
    ajint     nseqs     = 0;		/* the number of sequnces in the multiple sequence alignment */
    ajint     x         = 0;
          
    float   evalue      = 0.0;		/* Threshold E-value for inclusion in family */
    
    ajlong  offset      = 0;
    ajint   fseekr      = 0;
    
    AjPList list        = NULL;		/* Used to hold list of names of files in align directory */
    AjPList scopinfo    = NULL;         /* Holds the scop information in alignment file */
    AjPList seqlist     = NULL;		/* Holds the sequences in the training set */
    
    AjPFile families    = NULL;		/* Name of families file for output */
    AjPFile alignf      = NULL;		/* Alignment file pointer */
    AjPFile logf        = NULL;		/* Log file pointer */
    AjPFile seqf        = NULL;		/* A file containing an ungaped fasta sequence from the alignment file */
    AjPFile justalign   = NULL;         /* A file containing just the alignment, wthout the Escop.dat infomation */
    
    AjPRegexp rexp      = NULL;
    AjPRegexp blanks    = NULL;
        
    
    /* Initialise strings */
    align     = ajStrNew();
    alignextn = ajStrNew();
    database  = ajStrNew();
    submatrix = ajStrNew();
    temp      = ajStrNew();
    line      = ajStrNew();
    seq       = ajStrNew();
    sub       = ajStrNew();
    msg       = ajStrNew();
    cmdline   = ajStrNew();
    start     = ajStrNew();
    end       = ajStrNew();
    sequence  = ajStrNew();
    
    /* Read data from acd */
    embInit("psiblasts",argc,argv); 
    
    align      = ajAcdGetString("align");
    alignextn  = ajAcdGetString("alignextn");
    database   = ajAcdGetString("database");
    submatrix  = ajAcdGetString("submatrix");
    niter      = ajAcdGetInt("niter");
    evalue     = ajAcdGetFloat("evalue");
    families   = ajAcdGetOutfile("families");
    logf       = ajAcdGetOutfile("logf");
    
    /* Check directories */
    if(!ajFileDir(&align))
	ajFatal("Could not open alignments directory");
    
    
    /* Create list of files in align directory */
    list = ajListNew();
    ajStrAssC(&temp, "*");	
    if((ajStrChar(alignextn, 0)=='.'))
	ajStrApp(&temp, alignextn);    
    else
    {
	ajStrAppC(&temp, ".");    
	ajStrApp(&temp, alignextn);    
    }
    ajFileScan(align, temp, &list, ajFalse, ajFalse, NULL, NULL, ajFalse, NULL); 


    /*Start of main application loop*/   
       
    ajFmtPrintF(families,"DE   Members of scop families\nXX\nEX   THRESH %.3f; MATRIX %S;\nXX\n",evalue,submatrix);
    while(ajListPop(list,(void **)&temp))
    {
	rexp      = ajRegCompC("(^[0-9][_0-9a-zA-Z/-]+)([ ]*)([A-Z-]+)");
	blanks    = ajRegCompC("(^[0-9a-zA-Z])");

	scopinfo  = ajListstrNew();
	seqlist   = ajListstrNew();
	
	/* Read alignment file*/
	if((alignf = ajFileNewIn(temp)) == NULL)
	{
	    ajFmtPrintS(&msg, "Could not open for reading %S %S", 
			temp);
	    ajWarn(ajStrStr(msg));
	    ajFmtPrintF(logf, "WARN  Could not open for reading %S\n", 
			temp);
	    ajFileClose(&alignf);
	    ajStrDel(&temp);	
	    continue;	    
	}
	ajStrAss(&sub,temp);
	
	/******************************************************************** 
	 ** Set up and Run Psi-Blast by creating a single sequence file and *
	 ** a file containing just the multiple sequence alignment.         *
	 ********************************************************************/
	
	temp = ajStrNewC("alignfile");
	justalign = ajFileNewApp(temp);
	ajFmtPrintF(justalign,"\n");

	while(ajFileReadLine(alignf,&line))
	{ 	    
	    
	    if(ajRegExec(rexp,line))
		ajFmtPrintF(justalign,"%S\n",line);
	    	    
	    else if(ajStrFindCaseC(line,"Number")>=0)
		continue;
	    	    
	    else if(ajStrFindCaseC(line,"Post_similar")>=0)
		ajFmtPrintF(justalign,"\n");	
	    	    
 	    else
	    {
		temp = ajStrNew();
		ajStrAss(&temp,line);
		ajListstrPushApp(scopinfo,temp);
	    }
	
	}
	
	ajFileClose(&justalign);
	ajFileClose(&alignf);

	seqlist = SeqGet(sub);
	nseqs = ajListstrToArray(seqlist,&seqs);
	
	line = ajStrNewC("seq.fasta");
        seqf = ajFileNewApp(line);
       	ajFmtPrintF(seqf,">SEQ\n%S\n",seqs[0]);
	ajFileClose(&seqf);


	ajFmtPrintS(&cmdline,"blastpgp -i %S -B alignfile -j %d -e %f -d %s > file\n",
		    line, niter,evalue,ajStrStr(database));

	system(ajStrStr(cmdline)); /* Run PSI-BLAST */

	

	/* write out the scop family infomation */
	while(ajListstrPop(scopinfo,&line))
	{
	    if(ajRegExec(blanks,line))
		ajFmtPrintF(families,"%S\n",line);
	}
		
	ajRegFree(&rexp);
	ajRegFree(&blanks);
	
	/* clean up the directory on the fly */
	temp = ajStrNewC("alignfile");
	ajSysUnlink(&temp);
	temp = ajStrNewC("seq.fasta");
	ajSysUnlink(&temp);
	
	
	/*********************************************************
	 ** Parse the Psi-Blast output file and put the sequences*
	 ** in an EMBL like format.                              *
	 *********************************************************/
	
	/* parse the psi-blast output */
	temp = ajStrNewC("file");
	seqf = ajFileNewIn(temp);
	
	rexp   = ajRegCompC("(^Sbjct:)([ ]+)([0-9]+)([ ]+)([A-Z-]+)([ ]+)([0-9]+)");
	blanks = ajRegCompC("(^>SW:)([A-Z0-9_]+)");
       
	/* do a first parse through the psi-blast output file to work out the number of hits */
	while(ajFileReadLine(seqf,&line))
	{
	    if(ajStrFindCaseC(line,"score = ")>=0)
		setcounter++;
	}
	fseekr = ajFileSeek(seqf,offset,SEEK_SET);
	ajFmtPrintF(families,"NS   %d\nXX\n",setcounter);

       	while(ajFileReadLine(seqf,&line))
	{
	    if(ajRegExec(blanks,line))
	    {   
		ajRegSubI(blanks,2,&seq);
				
		while(ajFileReadLine(seqf,&line))
		{

		     /* tests for multiple sequences for a given hit entry delimited by "score" */
		    if(ajStrFindCaseC(line,"score")>=0)
		    {
			nscore++;
			hitcounter++;
			
			if(nscore==1) 
			{
			    ajFmtPrintF(families,"NN   [%d]\nXX\n",hitcounter);
			    ajFmtPrintF(families,"AC   %S\nXX\n",seq);
			    nlines = 0;
			}	
			
			else
			{
			    ajFmtPrintF(families,"%S END;\nXX\n",end);
			    ajSeqWriteCdb(families,sequence);
			    ajFmtPrintF(families,"XX\n");
			    for(x = 0;x<nseqs;x++)
			    {
				if(ajStrFindCase(seqs[x],sequence)>=0)
				{
				    ajFmtPrintF(families,"TY   TRAIN\nXX\n");
				    break;
				}
				else
				{
				    ajFmtPrintF(families,"TY   PSIBLAST\nXX\n");
				    break;
				}
			    }

			    ajStrAssC(&sequence,"");
			    ajFmtPrintF(families,"NN   [%d]\nXX\nAC   %S\nXX\n",hitcounter,seq);
			    nscore = 1;
			    nlines = 0;
			}
		    }
		    
		    else if(ajRegExec(rexp,line))
		    {	
			ajRegSubI(rexp,3,&start);
			ajRegSubI(rexp,5,&sub);
			ajStrSubstituteCC(&sub,"-","");
			ajStrApp(&sequence,sub);
			ajRegSubI(rexp,7,&end);
			offset = ajFileTell(seqf);
			nlines++;
									
			if(nscore==1&&nlines==1)
			    ajFmtPrintF(families,"RA   %S START;  ",start);
		    }

		    /* test for an end of hit entry delimited by a ">" */
		    else if(ajStrFindCaseC(line,">")>=0 || ajStrFindCaseC(line,"database")>=0
			    || ajStrFindCaseC(line,"searching")>=0)
		    {
			fseekr = ajFileSeek(seqf,offset,SEEK_SET);
			if(nscore==1)
			{
			    ajFmtPrintF(families,"%S END;\nXX\n",end);		    
			    ajSeqWriteCdb(families,sequence);
			    ajFmtPrintF(families,"XX\n");
			     for(x = 0;x<nseqs;x++)
			    {
				if(ajStrFindCase(seqs[x],sequence)>=0)
				{
				    ajFmtPrintF(families,"TY   TRAIN\nXX\n");
				    break;
				}
				else
				{
				    ajFmtPrintF(families,"TY   PSIBLAST\nXX\n");
				    break;
				}
			    }
					
			    ajStrAssC(&sequence,"");
			    nscore = 0;
			}
			break;
		    }
		}
	    }
	}


	/* set all counters to zero */
	hitcounter = 0;
	setcounter = 0;
	offset 	   = 0;
	nlines     = 0;
	nscore     = 0;
	
	/* clean up */
	ajFileClose(&seqf);

	ajListDel(&seqlist);
	ajListDel(&scopinfo);

	for(x=0;x<nseqs;x++)
	    ajStrDel(&seqs[x]);
	
	ajRegFree(&rexp);
	ajRegFree(&blanks);
     
	temp = ajStrNewC("file");
	ajSysUnlink(&temp);
       
	ajStrAssC(&seq,"");	
	ajFmtPrintF(families,"//\n");
    
    }
            
    /*Tidy up */
    ajStrDel(&align);
    ajStrDel(&alignextn);
    ajStrDel(&database);
    ajStrDel(&submatrix);
    ajStrDel(&temp);
    ajStrDel(&line);
    ajStrDel(&sub);
    ajStrDel(&msg);
    ajStrDel(&cmdline);
    ajStrDel(&start);
    ajStrDel(&end);
    ajStrDel(&sequence);
        
    ajListDel(&list);
    
    ajFileClose(&families);
    ajFileClose(&logf);

    ajExit();
    return 0;
}


