<HTML>

<HEAD>
  <TITLE>
  EMBOSS: acdtrace
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="/images/emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
acdtrace
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
Trace processing of an application ACD file (for testing)

<H2>
    Description
</H2>
<p><b>acdtrace</b> reads and processes an ACD file, prompting the user as necessary in exactly the same way as if the application itself was running.  If it encounters an error it will terminate.  This allows the behaviour of the command-line interface to be tested under different conditions, even if the application itself has not yet been written.</p>
<p><b>acdtrace</b> works the same as <b>acdc</b> but reports on the processing of lines in an ACD file.  It also also prints the results of evaluating any variables and operations in the ACD file.</p>



<H2>
    Usage
</H2>
Here is a sample session with <b>acdtrace</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>acdtrace antigenic </b>
Find antigenic sites in proteins
Input protein sequence(s): <b>tsw:actb1_takru</b>
Minimum length of antigenic region [6]: <b></b>
Output report [actb1_takru.antigenic]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>


<H2>
    Command line arguments
</H2>


The first parameter is the name of the program to compile.

All other command line parameters and qualifiers are defined for that
program - see the documentation for the program of choice.

<P>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
Trace processing of an application ACD file (for testing)
Version: EMBOSS:6.6.0.0

   Standard (Mandatory) qualifiers:
  [-program]           string     Program name, followed by command line (Any
                                  string)

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers: (none)
   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-program]<br>(Parameter 1)</td>
<td>string</td>
<td>Program name, followed by command line</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>



<H2>
    Input file format
</H2>

The input is whatever the chosen application reads.
<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tsw:actb1_takru' is a sequence entry in the example protein database 'tsw'
<p>
<p><h3>Database entry: tsw:actb1_takru</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   ACTB1_TAKRU             Reviewed;         375 AA.
AC   P68142; P53484;
DT   25-OCT-2004, integrated into UniProtKB/Swiss-Prot.
DT   25-OCT-2004, sequence version 1.
DT   16-MAY-2012, entry version 49.
DE   RecName: Full=Actin, cytoplasmic 1;
DE   AltName: Full=Beta-actin A;
DE   Contains:
DE     RecName: Full=Actin, cytoplasmic 1, N-terminally processed;
GN   Name=actba;
OS   Takifugu rubripes (Japanese pufferfish) (Fugu rubripes).
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi;
OC   Actinopterygii; Neopterygii; Teleostei; Euteleostei; Neoteleostei;
OC   Acanthomorpha; Acanthopterygii; Percomorpha; Tetraodontiformes;
OC   Tetradontoidea; Tetraodontidae; Takifugu.
OX   NCBI_TaxID=31033;
RN   [1]
RP   NUCLEOTIDE SEQUENCE [GENOMIC DNA], AND TISSUE SPECIFICITY.
RX   MEDLINE=96275651; PubMed=8683572; DOI=10.1006/jmbi.1996.0347;
RA   Venkatesh B., Tay B.H., Elgar G., Brenner S.;
RT   "Isolation, characterization and evolution of nine pufferfish (Fugu
RT   rubripes) actin genes.";
RL   J. Mol. Biol. 259:655-665(1996).
CC   -!- FUNCTION: Actins are highly conserved proteins that are involved
CC       in various types of cell motility and are ubiquitously expressed
CC       in all eukaryotic cells.
CC   -!- SUBUNIT: Polymerization of globular actin (G-actin) leads to a
CC       structural filament (F-actin) in the form of a two-stranded helix.
CC       Each actin can bind to 4 others.
CC   -!- SUBCELLULAR LOCATION: Cytoplasm, cytoskeleton.
CC   -!- TISSUE SPECIFICITY: Widely distributed. Not expressed in skeletal
CC       muscle.
CC   -!- PTM: Oxidation of Met-44 by MICALs (MICAL1, MICAL2 or MICAL3) to
CC       form methionine sulfoxide promotes actin filament
CC       depolymerization. Methionine sulfoxide is produced
CC       stereospecifically, but it is not known whether the (S)-S-oxide or
CC       the (R)-S-oxide is produced (By similarity).
CC   -!- MISCELLANEOUS: In vertebrates 3 main groups of actin isoforms,
CC       alpha, beta and gamma have been identified. The alpha actins are
CC       found in muscle tissues and are a major constituent of the
CC       contractile apparatus. The beta and gamma actins coexist in most
CC       cell types as components of the cytoskeleton and as mediators of
CC       internal cell motility.
CC   -!- MISCELLANEOUS: There are three different beta-cytoplasmic actins
CC       in Fugu rubripes.
CC   -!- SIMILARITY: Belongs to the actin family.
CC   -----------------------------------------------------------------------
CC   Copyrighted by the UniProt Consortium, see http://www.uniprot.org/terms
CC   Distributed under the Creative Commons Attribution-NoDerivs License
CC   -----------------------------------------------------------------------
DR   EMBL; U37499; AAC59889.1; -; Genomic_DNA.
DR   PIR; S71124; S71124.
DR   ProteinModelPortal; P68142; -.
DR   SMR; P68142; 2-375.
DR   Ensembl; ENSTRUT00000013141; ENSTRUP00000013080; ENSTRUG00000005447.
DR   eggNOG; COG5277; -.
DR   GeneTree; ENSGT00630000089629; -.
DR   InParanoid; P68142; -.
DR   OMA; IKNLMER; -.
DR   OrthoDB; EOG41JZC9; -.
DR   GO; GO:0005737; C:cytoplasm; IEA:UniProtKB-KW.
DR   GO; GO:0005856; C:cytoskeleton; IEA:UniProtKB-SubCell.
DR   GO; GO:0005524; F:ATP binding; IEA:UniProtKB-KW.
DR   InterPro; IPR004000; Actin-like.
DR   InterPro; IPR020902; Actin/actin-like_CS.
DR   InterPro; IPR004001; Actin_CS.
DR   PANTHER; PTHR11937; Actin_like; 1.
DR   Pfam; PF00022; Actin; 1.
DR   PRINTS; PR00190; ACTIN.
DR   SMART; SM00268; ACTIN; 1.
DR   PROSITE; PS00406; ACTINS_1; 1.
DR   PROSITE; PS00432; ACTINS_2; 1.
DR   PROSITE; PS01132; ACTINS_ACT_LIKE; 1.
PE   2: Evidence at transcript level;
KW   Acetylation; ATP-binding; Complete proteome; Cytoplasm; Cytoskeleton;
KW   Methylation; Nucleotide-binding; Oxidation; Reference proteome.
FT   CHAIN         1    375       Actin, cytoplasmic 1.
FT                                /FTId=PRO_0000367094.
FT   INIT_MET      1      1       Removed; alternate (By similarity).
FT   CHAIN         2    375       Actin, cytoplasmic 1, N-terminally
FT                                processed.
FT                                /FTId=PRO_0000000809.
FT   MOD_RES       1      1       N-acetylmethionine; in Actin, cytoplasmic
FT                                1; alternate (By similarity).
FT   MOD_RES       2      2       N-acetylglutamate; in Actin, cytoplasmic
FT                                1, N-terminally processed (By
FT                                similarity).
FT   MOD_RES      44     44       Methionine sulfoxide (By similarity).
FT   MOD_RES      73     73       Tele-methylhistidine (By similarity).
SQ   SEQUENCE   375 AA;  41767 MW;  9C505616D33E9495 CRC64;
     MEDEIAALVV DNGSGMCKAG FAGDDAPRAV FPSIVGRPRH QGVMVGMGQK DSYVGDEAQS
     KRGILTLKYP IEHGIVTNWD DMEKIWHHTF YNELRVAPEE HPVLLTEAPL NPKANREKMT
     QIMFETFNTP AMYVAIQAVL SLYASGRTTG IVMDSGDGVT HTVPIYEGYA LPHAILRLDL
     AGRDLTDYLM KILTERGYSF TTTAEREIVR DIKEKLCYVA LDFEQEMGTA ASSSSLEKSY
     ELPDGQVITI GNERFRCPEA LFQPSFLGME SCGIHETTYN SIMKCDVDIR KDLYANTVLS
     GGTTMYPGIA DRMQKEITAL APSTMKIKII APPERKYSVW IGGSILASLS TFQQMWISKQ
     EYDESGPSIV HRKCF
//
</pre>
</td></tr></table><p>

<H2>
    Output file format
</H2>

<p>

The output file is empty, but any format available to the specified
program can be used.

<p>

<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: actb1_takru.antigenic</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
</pre>
</td></tr></table><p>


<H2>
    Data files
</H2>


Acdtrace will use any data files specified in the ACD file of the
specified program, but will ignore any data files that are only used
directly by the program's code.

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

<H2>
    Notes
</H2>
<p><b>acdtrace</b> does not have its own options, but takes a single parameter which is the name of the EMBOSS application that is being testsed.  It is invoked by specifying the name of the application to be tested along with any command-line options after the name of the utility application itself:</p>

<p><tt><b>acdtrace</b> ApplicationName Options</tt></p>


<p><b>acdtrace</b> will run the command line interface of any EMBOSS program by reading its ACD file and processing the command line and user responses in exactly the same way as if the true program itself were running.  The application proper is not run; it is only the ACD file that is read.  Any command-line options for the application being called can be specified. The user is prompted for any required parameter values not given on the command line and any input files are read. Output files can be specified in any available format however they will be empty (the application code is not called so there will be no output).

<p><b>acdtrace</b> will use any data files specified in the ACD file of the specified program, but will ignore any data files that are only used directly by the program's code.  EMBOSS data files are distributed with the application and stored in the standard EMBOSS data directory (see EMBOSS Users Manual for further information).</p>


<H2>
    References
</H2>


<H2>
    Warnings
</H2>

<p>The output on screen can look a little confusing but is by far the best way to see how variables and operations work in your ACD file.</p>


<H2>
    Diagnostic Error Messages
</H2>



<H2>
    Exit status
</H2>


It exits with status 0.

<H2>
    Known bugs
</H2>


None.


<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="acdc.html">acdc</a></td>
<td>Test an application ACD file</td>
</tr>

<tr>
<td><a href="acdpretty.html">acdpretty</a></td>
<td>Correctly reformat an application ACD file</td>
</tr>

<tr>
<td><a href="acdtable.html">acdtable</a></td>
<td>Generate an HTML table of parameters from an application ACD file</td>
</tr>

<tr>
<td><a href="acdvalid.html">acdvalid</a></td>
<td>Validate an application ACD file</td>
</tr>

</table>
<H2>
    Author(s)
</H2>

Peter Rice
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK

<p>
Please report all bugs to the EMBOSS bug team (emboss-bug&nbsp;&copy;&nbsp;emboss.open-bio.org) not to the original author.

<H2>
    History
</H2>


<H2>
    Target users
</H2>
This program is intended to be used by developers of applications and
interfaces. 

<H2>
    Comments
</H2>
None



</BODY>
</HTML>

