<HTML>
<HEAD>
  <TITLE>
  EMBOSS: aligncopy
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
aligncopy
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
Reads and writes alignments
<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
  -->

<b>aligncopy</b> reads a set of aligned sequences and reports the
alignment using any of the standard EMBOSS alignment formats that can
report on the size of the input alignment. An input of two aligned
sequences will allow any of the pairwise formats to be used. Other
input sizes will reject any pairwise alignment
format. Use <b>aligncopypair</b> to print these alignments one pair at
a time.

















<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

<b>Here is a sample session with aligncopy</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>aligncopy </b>
Reads and writes alignments
Input (aligned) sequence set: <b>globins.msf</b>
Output alignment [globins.aligncopy]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
Reads and writes alignments
Version: EMBOSS:6.2.0

   Standard (Mandatory) qualifiers:
  [-sequences]         seqset     File containing a sequence alignment.
  [-outfile]           align      [*.aligncopy] Output alignment file name

   Additional (Optional) qualifiers:
   -name               string     Name of the alignment (Any string)
   -comment            string     Comment on the alignment (Any string)
   -append             boolean    [N] This allows you to append the name and
                                  comments you have given on to the end of the
                                  existing alignment header.

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-sequences" associated qualifiers
   -sbegin1            integer    Start of each sequence to be used
   -send1              integer    End of each sequence to be used
   -sreverse1          boolean    Reverse (if DNA)
   -sask1              boolean    Ask for begin/end/reverse
   -snucleotide1       boolean    Sequence is nucleotide
   -sprotein1          boolean    Sequence is protein
   -slower1            boolean    Make lower case
   -supper1            boolean    Make upper case
   -sformat1           string     Input sequence format
   -sdbname1           string     Database name
   -sid1               string     Entryname
   -ufo1               string     UFO features
   -fformat1           string     Features format
   -fopenfile1         string     Features file name

   "-outfile" associated qualifiers
   -aformat2           string     Alignment format
   -aextension2        string     File name extension
   -adirectory2        string     Output directory
   -aname2             string     Base file name
   -awidth2            integer    Alignment width
   -aaccshow2          boolean    Show accession number in the header
   -adesshow2          boolean    Show description in the header
   -ausashow2          boolean    Show the full USA in the alignment
   -aglobal2           boolean    Show the full sequence in alignment

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-sequences]<br>(Parameter 1)</td>
<td>seqset</td>
<td>File containing a sequence alignment.</td>
<td>Readable set of sequences</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-outfile]<br>(Parameter 2)</td>
<td>align</td>
<td>Output alignment file name</td>
<td>Alignment output file</td>
<td><i>&lt;*&gt;</i>.aligncopy</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>-name</td>
<td>string</td>
<td>Name of the alignment</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-comment</td>
<td>string</td>
<td>Comment on the alignment</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-append</td>
<td>boolean</td>
<td>This allows you to append the name and comments you have given on to the end of the existing alignment header.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-sequences" associated seqset qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sbegin1<br>-sbegin_sequences</td>
<td>integer</td>
<td>Start of each sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -send1<br>-send_sequences</td>
<td>integer</td>
<td>End of each sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sreverse1<br>-sreverse_sequences</td>
<td>boolean</td>
<td>Reverse (if DNA)</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sask1<br>-sask_sequences</td>
<td>boolean</td>
<td>Ask for begin/end/reverse</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -snucleotide1<br>-snucleotide_sequences</td>
<td>boolean</td>
<td>Sequence is nucleotide</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sprotein1<br>-sprotein_sequences</td>
<td>boolean</td>
<td>Sequence is protein</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -slower1<br>-slower_sequences</td>
<td>boolean</td>
<td>Make lower case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -supper1<br>-supper_sequences</td>
<td>boolean</td>
<td>Make upper case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sformat1<br>-sformat_sequences</td>
<td>string</td>
<td>Input sequence format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sdbname1<br>-sdbname_sequences</td>
<td>string</td>
<td>Database name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sid1<br>-sid_sequences</td>
<td>string</td>
<td>Entryname</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ufo1<br>-ufo_sequences</td>
<td>string</td>
<td>UFO features</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fformat1<br>-fformat_sequences</td>
<td>string</td>
<td>Features format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fopenfile1<br>-fopenfile_sequences</td>
<td>string</td>
<td>Features file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-outfile" associated align qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -aformat2<br>-aformat_outfile</td>
<td>string</td>
<td>Alignment format</td>
<td>Any string</td>
<td>simple</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -aextension2<br>-aextension_outfile</td>
<td>string</td>
<td>File name extension</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -adirectory2<br>-adirectory_outfile</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -aname2<br>-aname_outfile</td>
<td>string</td>
<td>Base file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -awidth2<br>-awidth_outfile</td>
<td>integer</td>
<td>Alignment width</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -aaccshow2<br>-aaccshow_outfile</td>
<td>boolean</td>
<td>Show accession number in the header</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -adesshow2<br>-adesshow_outfile</td>
<td>boolean</td>
<td>Show description in the header</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ausashow2<br>-ausashow_outfile</td>
<td>boolean</td>
<td>Show the full USA in the alignment</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -aglobal2<br>-aglobal_outfile</td>
<td>boolean</td>
<td>Show the full sequence in alignment</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>aligncopy</b> reads any normal sequence USAs.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: globins.msf</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
!!AA_MULTIPLE_ALIGNMENT 1.0

  ../data/globins.msf MSF:  164 Type: P 25/06/01 CompCheck: 4278 ..

  Name: HBB_HUMAN Len: 164  Check: 6914 Weight: 0.61
  Name: HBB_HORSE Len: 164  Check: 6007 Weight: 0.65
  Name: HBA_HUMAN Len: 164  Check: 3921 Weight: 0.65
  Name: HBA_HORSE Len: 164  Check: 4770 Weight: 0.83
  Name: MYG_PHYCA Len: 164  Check: 7930 Weight: 1.00
  Name: GLB5_PETMA Len: 164  Check: 1857 Weight: 0.91
  Name: LGB2_LUPLU Len: 164  Check: 2879 Weight: 0.43

//

           1                                               50
HBB_HUMAN  ~~~~~~~~VHLTPEEKSAVTALWGKVN.VDEVGGEALGR.LLVVYPWTQR
HBB_HORSE  ~~~~~~~~VQLSGEEKAAVLALWDKVN.EEEVGGEALGR.LLVVYPWTQR
HBA_HUMAN  ~~~~~~~~~~~~~~VLSPADKTNVKAA.WGKVGAHAGEYGAEALERMFLS
HBA_HORSE  ~~~~~~~~~~~~~~VLSAADKTNVKAA.WSKVGGHAGEYGAEALERMFLG
MYG_PHYCA  ~~~~~~~VLSEGEWQLVLHVWAKVEAD.VAGHGQDILIR.LFKSHPETLE
GLB5_PETMA PIVDTGSVAPLSAAEKTKIRSAWAPVYSTYETSGVDILVKFFTSTPAAQE
LGB2_LUPLU ~~~~~~~~GALTESQAALVKSSWEEFNANIPKHTHRFFILVLEIAPAAKD

           51                                             100
HBB_HUMAN  FFESFGDLSTPDAVMGNPKVKAHGKKVLGAFSDGLAHLDNLKGTFATLSE
HBB_HORSE  FFDSFGDLSNPGAVMGNPKVKAHGKKVLHSFGEGVHHLDNLKGTFAALSE
HBA_HUMAN  FPTTKTYFPHFDLSHGSAQVKGHGKKVADALTNAVAHVDDMPNALSALSD
HBA_HORSE  FPTTKTYFPHFDLSHGSAQVKAHGKKVGDALTLAVGHLDDLPGALSNLSD
MYG_PHYCA  KFDRFKHLKTEAEMKASEDLKKHGVTVLTALGAILKKKGHHEAELKPLAQ
GLB5_PETMA FFPKFKGLTTADQLKKSADVRWHAERIINAVNDAVASMDDTEKMSMKLRD
LGB2_LUPLU LFSFLKGTSEVPQNNPELQAHAGKVFKLVYEAAIQLQVTGVVVTDATLKN

           101                                            150
HBB_HUMAN  LHCDKLH..VDPENFRLLGNVLVCVLAHHFGKEFTPPVQAAYQKVVAGVA
HBB_HORSE  LHCDKLH..VDPENFRLLGNVLVVVLARHFGKDFTPELQASYQKVVAGVA
HBA_HUMAN  LHAHKLR..VDPVNFKLLSHCLLVTLAAHLPAEFTPAVHASLDKFLASVS
HBA_HORSE  LHAHKLR..VDPVNFKLLSHCLLSTLAVHLPNDFTPAVHASLDKFLSSVS
MYG_PHYCA  SHATKHK..IPIKYLEFISEAIIHVLHSRHPGDFGADAQGAMNKALELFR
GLB5_PETMA LSGKHAK..SFQVDPQYFKVLAAVIADTVAAGDAGFEKLMSMICILLRSA
LGB2_LUPLU LGSVHVSKGVADAHFPVVKEAILKTIKEVVGAKWSEELNSAWTIAYDELA

           151        164
HBB_HUMAN  NALAHKYH~~~~~~
HBB_HORSE  NALAHKYH~~~~~~
HBA_HUMAN  TVLTSKYR~~~~~~
HBA_HORSE  TVLTSKYR~~~~~~
MYG_PHYCA  KDIAAKYKELGYQG
GLB5_PETMA Y~~~~~~~~~~~~~
LGB2_LUPLU IVIKKEMNDAA~~~

</pre>
</td></tr></table><p>





<H2>
    Output file format
</H2>


<p>

The output is a standard EMBOSS alignment file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-aformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  Some of the alignment formats can cope
with an unlimited number of sequences, while others are only for pairs
of sequences.  

<p>

The available multiple alignment format names are: unknown, multiple,
simple, fasta, msf, trace, srs

<p>

The available pairwise alignment format names are: pair, markx0, markx1,
markx2, markx3, markx10, srspair, score

<p>

See:
<A href="http://emboss.sf.net/docs/themes/AlignFormats.html">
http://emboss.sf.net/docs/themes/AlignFormats.html</A>
for further information on alignment formats.

<p>


<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: globins.aligncopy</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
########################################
# Program: aligncopy
# Rundate: Fri 15 Jan 2010 12:00:00
# Commandline: aligncopy
#    -sequences ../../data/globins.msf
# Align_format: simple
# Report_file: globins.aligncopy
########################################

#=======================================
#
# Aligned_sequences: 7
# 1: HBB_HUMAN
# 2: HBB_HORSE
# 3: HBA_HUMAN
# 4: HBA_HORSE
# 5: MYG_PHYCA
# 6: GLB5_PETMA
# 7: LGB2_LUPLU
# Matrix: EBLOSUM62
#
# Length: 164
# Identity:       1/164 ( 0.6%)
# Similarity:    93/164 (56.7%)
# Gaps:          31/164 (18.9%)
# 
#
#=======================================

HBB_HUMAN          1 --------VHLTPEEKSAVTALWGKVN-VDEVGGEALGR-LLVVYPWTQR     40
HBB_HORSE          1 --------VQLSGEEKAAVLALWDKVN-EEEVGGEALGR-LLVVYPWTQR     40
HBA_HUMAN          1 --------------VLSPADKTNVKAA-WGKVGAHAGEYGAEALERMFLS     35
HBA_HORSE          1 --------------VLSAADKTNVKAA-WSKVGGHAGEYGAEALERMFLG     35
MYG_PHYCA          1 -------VLSEGEWQLVLHVWAKVEAD-VAGHGQDILIR-LFKSHPETLE     41
GLB5_PETMA         1 PIVDTGSVAPLSAAEKTKIRSAWAPVYSTYETSGVDILVKFFTSTPAAQE     50
LGB2_LUPLU         1 --------GALTESQAALVKSSWEEFNANIPKHTHRFFILVLEIAPAAKD     42
                               ::  : : :.: : : :   :::: ::   :  : :  : 

HBB_HUMAN         41 FFESFGDLSTPDAVMGNPKVKAHGKKVLGAFSDGLAHLDNLKGTFATLSE     90
HBB_HORSE         41 FFDSFGDLSNPGAVMGNPKVKAHGKKVLHSFGEGVHHLDNLKGTFAALSE     90
HBA_HUMAN         36 FPTTKTYFPHFDLSHGSAQVKGHGKKVADALTNAVAHVDDMPNALSALSD     85
HBA_HORSE         36 FPTTKTYFPHFDLSHGSAQVKAHGKKVGDALTLAVGHLDDLPGALSNLSD     85
MYG_PHYCA         42 KFDRFKHLKTEAEMKASEDLKKHGVTVLTALGAILKKKGHHEAELKPLAQ     91
GLB5_PETMA        51 FFPKFKGLTTADQLKKSADVRWHAERIINAVNDAVASMDDTEKMSMKLRD    100
LGB2_LUPLU        43 LFSFLKGTSEVPQNNPELQAHAGKVFKLVYEAAIQLQVTGVVVTDATLKN     92
                     :: ::  ::  : : :: ::: :::::: :: : : ::::::   : |::

HBB_HUMAN         91 LHCDKLH--VDPENFRLLGNVLVCVLAHHFGKEFTPPVQAAYQKVVAGVA    138
HBB_HORSE         91 LHCDKLH--VDPENFRLLGNVLVVVLARHFGKDFTPELQASYQKVVAGVA    138
HBA_HUMAN         86 LHAHKLR--VDPVNFKLLSHCLLVTLAAHLPAEFTPAVHASLDKFLASVS    133
HBA_HORSE         86 LHAHKLR--VDPVNFKLLSHCLLSTLAVHLPNDFTPAVHASLDKFLSSVS    133
MYG_PHYCA         92 SHATKHK--IPIKYLEFISEAIIHVLHSRHPGDFGADAQGAMNKALELFR    139
GLB5_PETMA       101 LSGKHAK--SFQVDPQYFKVLAAVIADTVAAGDAGFEKLMSMICILLRSA    148
LGB2_LUPLU        93 LGSVHVSKGVADAHFPVVKEAILKTIKEVVGAKWSEELNSAWTIAYDELA    142
                     ::: :::  ::: ::::: : :: ::: :   :::::: ::: : :: ::

HBB_HUMAN        139 NALAHKYH------    146
HBB_HORSE        139 NALAHKYH------    146
HBA_HUMAN        134 TVLTSKYR------    141
HBA_HORSE        134 TVLTSKYR------    141
MYG_PHYCA        140 KDIAAKYKELGYQG    153
GLB5_PETMA       149 Y-------------    149
LGB2_LUPLU       143 IVIKKEMNDAA---    153
                       :  ::       


#---------------------------------------
#---------------------------------------
</pre>
</td></tr></table><p>





<H2>
    Data files
</H2>

For protein sequences EBLOSUM62 is used for the substitution matrix.
For nucleotide sequence, EDNAFULL is used.

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>




<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
<ol>

<li>

</ol>

   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="aligncopypair.html">aligncopypair</a></td>
<td>Reads and writes pairs from alignments</td>
</tr>

<tr>
<td><a href="biosed.html">biosed</a></td>
<td>Replace or delete sequence sections</td>
</tr>

<tr>
<td><a href="codcopy.html">codcopy</a></td>
<td>Copy and reformat a codon usage table</td>
</tr>

<tr>
<td><a href="cutseq.html">cutseq</a></td>
<td>Removes a section from a sequence</td>
</tr>

<tr>
<td><a href="degapseq.html">degapseq</a></td>
<td>Removes non-alphabetic (e.g. gap) characters from sequences</td>
</tr>

<tr>
<td><a href="descseq.html">descseq</a></td>
<td>Alter the name or description of a sequence</td>
</tr>

<tr>
<td><a href="entret.html">entret</a></td>
<td>Retrieves sequence entries from flatfile databases and files</td>
</tr>

<tr>
<td><a href="extractalign.html">extractalign</a></td>
<td>Extract regions from a sequence alignment</td>
</tr>

<tr>
<td><a href="extractfeat.html">extractfeat</a></td>
<td>Extract features from sequence(s)</td>
</tr>

<tr>
<td><a href="extractseq.html">extractseq</a></td>
<td>Extract regions from a sequence</td>
</tr>

<tr>
<td><a href="featcopy.html">featcopy</a></td>
<td>Reads and writes a feature table</td>
</tr>

<tr>
<td><a href="featreport.html">featreport</a></td>
<td>Reads and writes a feature table</td>
</tr>

<tr>
<td><a href="listor.html">listor</a></td>
<td>Write a list file of the logical OR of two sets of sequences</td>
</tr>

<tr>
<td><a href="makenucseq.html">makenucseq</a></td>
<td>Create random nucleotide sequences</td>
</tr>

<tr>
<td><a href="makeprotseq.html">makeprotseq</a></td>
<td>Create random protein sequences</td>
</tr>

<tr>
<td><a href="maskambignuc.html">maskambignuc</a></td>
<td>Masks all ambiguity characters in nucleotide sequences with N</td>
</tr>

<tr>
<td><a href="maskambigprot.html">maskambigprot</a></td>
<td>Masks all ambiguity characters in protein sequences with X</td>
</tr>

<tr>
<td><a href="maskfeat.html">maskfeat</a></td>
<td>Write a sequence with masked features</td>
</tr>

<tr>
<td><a href="maskseq.html">maskseq</a></td>
<td>Write a sequence with masked regions</td>
</tr>

<tr>
<td><a href="newseq.html">newseq</a></td>
<td>Create a sequence file from a typed-in sequence</td>
</tr>

<tr>
<td><a href="nohtml.html">nohtml</a></td>
<td>Remove mark-up (e.g. HTML tags) from an ASCII text file</td>
</tr>

<tr>
<td><a href="noreturn.html">noreturn</a></td>
<td>Remove carriage return from ASCII files</td>
</tr>

<tr>
<td><a href="nospace.html">nospace</a></td>
<td>Remove all whitespace from an ASCII text file</td>
</tr>

<tr>
<td><a href="notab.html">notab</a></td>
<td>Replace tabs with spaces in an ASCII text file</td>
</tr>

<tr>
<td><a href="notseq.html">notseq</a></td>
<td>Write to file a subset of an input stream of sequences</td>
</tr>

<tr>
<td><a href="nthseq.html">nthseq</a></td>
<td>Write to file a single sequence from an input stream of sequences</td>
</tr>

<tr>
<td><a href="nthseqset.html">nthseqset</a></td>
<td>Reads and writes (returns) one set of sequences from many</td>
</tr>

<tr>
<td><a href="pasteseq.html">pasteseq</a></td>
<td>Insert one sequence into another</td>
</tr>

<tr>
<td><a href="revseq.html">revseq</a></td>
<td>Reverse and complement a nucleotide sequence</td>
</tr>

<tr>
<td><a href="seqret.html">seqret</a></td>
<td>Reads and writes (returns) sequences</td>
</tr>

<tr>
<td><a href="seqretsetall.html">seqretsetall</a></td>
<td>Reads and writes (returns) many sets of sequences</td>
</tr>

<tr>
<td><a href="seqretsplit.html">seqretsplit</a></td>
<td>Reads sequences and writes them to individual files</td>
</tr>

<tr>
<td><a href="sizeseq.html">sizeseq</a></td>
<td>Sort sequences by size</td>
</tr>

<tr>
<td><a href="skipredundant.html">skipredundant</a></td>
<td>Remove redundant sequences from an input set</td>
</tr>

<tr>
<td><a href="skipseq.html">skipseq</a></td>
<td>Reads and writes (returns) sequences, skipping first few</td>
</tr>

<tr>
<td><a href="splitsource.html">splitsource</a></td>
<td>Split sequence(s) into original source sequences</td>
</tr>

<tr>
<td><a href="splitter.html">splitter</a></td>
<td>Split sequence(s) into smaller sequences</td>
</tr>

<tr>
<td><a href="trimest.html">trimest</a></td>
<td>Remove poly-A tails from nucleotide sequences</td>
</tr>

<tr>
<td><a href="trimseq.html">trimseq</a></td>
<td>Remove unwanted characters from start and end of sequence(s)</td>
</tr>

<tr>
<td><a href="trimspace.html">trimspace</a></td>
<td>Remove extra whitespace from an ASCII text file</td>
</tr>

<tr>
<td><a href="union.html">union</a></td>
<td>Concatenate multiple sequences into a single sequence</td>
</tr>

<tr>
<td><a href="vectorstrip.html">vectorstrip</a></td>
<td>Removes vectors from the ends of nucleotide sequence(s)</td>
</tr>

<tr>
<td><a href="yank.html">yank</a></td>
<td>Add a sequence reference (a full USA) to a list file</td>
</tr>

</table>
<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>

Jon Ison (jison&nbsp;&copy;&nbsp;ebi.ac.uk)
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK





<H2>
    History
</H2>
<!--
        Date written and what changes have been made go in this file.
   -->




<H2>
    Target users
</H2>
<!--
        For general users, use this text
   -->
This program is intended to be used by everyone and everything, from naive users to embedded scripts.

<H2>
    Comments
</H2>
<!--
        User/developer/other comments go in this file.
   -->
None


</BODY>
</HTML>



