<HTML>

<HEAD>
  <TITLE>
  EMBOSS: antigenic
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

antigenic
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>






<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Finds antigenic sites in proteins


<H2>
    Description
</H2>


Antigenic predicts potentially antigenic regions of a protein
sequence, using the method of Kolaskar and Tongaonkar.

<p>

Analysis of data from experimentally determined antigenic sites on
proteins has revealed that the hydrophobic residues Cys, Leu and Val, if
they occur on the surface of a protein, are more likely to be a part of
antigenic sites.  A semi-empirical method which makes use of
physicochemical properties of amino acid residues and their frequencies
of occurrence in experimentally known segmental epitopes was developed
by Kolaskar and Tongaonkar
to predict antigenic determinants on proteins.  Application of this
method to a large number of proteins has shown that their method can
predict antigenic determinants with about 75% accuracy which is better
than most of the known methods.  This method is based on a single
parameter and thus very simple to use. 

<H2>
    Usage
</H2>


Here is a sample session with antigenic.

<pre>

% antigenic
Finds antigenic sites in proteins
Input sequence: sw:act1_fugru
Minimum length [6]: 
Output file [act1_fugru.antigenic]: 

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
   -minlen             integer    Minimum length
  [-outfile]           report     Output report file name

   Optional qualifiers: (none)
   Advanced qualifiers: (none)
   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-minlen</td>
<td>Minimum length</td>
<td>Integer from 1 to 50</td>
<td>6</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output report file name</td>
<td>Report file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>


The input sequence can be one or more protein sequences.

<H2>
    Output file format
</H2>

<p>

The output is a standard EMBOSS report file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-rformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  The available format names are: embl,
genbank, gff, pir, swiss, trace, listfile, dbmotif, diffseq, excel,
feattable, motif, regions, seqtable, simple, srs, table, tagseq

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html</A>
for further information on report formats.

<p>


<p>

By default <b>antigenic</b> writes a 'motif' report file.

<p>

The output from the above example is:

<p>

<hr>
<pre>
########################################
# Program: antigenic
# Rundate: Mon Feb 11 12:01:10 2002
# Report_file: act1_fugru.antigenic
########################################

#=======================================
#
# Sequence: ACT1_FUGRU     from: 1   to: 375
# HitCount: 18
#=======================================

Max_score_pos at "*"

(1) Score 1.207 length 9 at residues 214->222
               *
 Sequence: EKLCYVALD
           |       |
         214       222

(2) Score 1.187 length 15 at residues 131->145
                 *
 Sequence: AMYVAIQAVLSLYAS
           |             |
         131             145

(3) Score 1.166 length 8 at residues 5->12
              *
 Sequence: IAALVVDN
           |      |
           5      12

(4) Score 1.164 length 12 at residues 27->38
                *
 Sequence: PRAVFPSIVGRP
           |          |
          27          38

(5) Score 1.136 length 24 at residues 160->183
                        *
 Sequence: THTVPIYEGYALPHAILRLDLAGR
           |                      |
         160                      183

(6) Score 1.135 length 6 at residues 367->372
                *
 Sequence: PSIVHR
           |    |
         367    372

(7) Score 1.116 length 16 at residues 93->108
                     *
 Sequence: ELRVAPEEHPVLLTEA
           |              |
          93              108

(8) Score 1.113 length 7 at residues 295->301
            *
 Sequence: ANTVLSG
           |     |
         295     301

(9) Score 1.110 length 11 at residues 256->266
                   *
 Sequence: RCPEALFQPSF
           |         |
         256         266

(10) Score 1.107 length 17 at residues 336->352
                      *
 Sequence: KYSVWIGGSILASLSTF
           |               |
         336               352

(11) Score 1.102 length 15 at residues 62->76        
                 *
 Sequence: RGILTLKYPIEHGIV
           |             |
          62             76

(12) Score 1.086 length 19 at residues 232->250
                        *
 Sequence: SSSSLEKSYELPDGQVITI
           |                 |
         232                 250

(13) Score 1.083 length 6 at residues 327->332
              *
 Sequence: IKIIAP
           |    |
         327    332

(14) Score 1.074 length 7 at residues 317->323
              *
 Sequence: ITALAPS
           |     |
         317     323

(15) Score 1.068 length 7 at residues 186->192
                *
 Sequence: TDYLMKI
           |     |
         186     192

(16) Score 1.066 length 7 at residues 40->46
              *
 Sequence: HQGVMVG
           |     |
          40     46

(17) Score 1.045 length 7 at residues 269->275
           *
 Sequence: MESCGIH
           |     |
         269     275

(18) Score 1.034 length 7 at residues 51->57
            *
 Sequence: DSYVGDE
           |     |
          51     57


#---------------------------------------
#---------------------------------------            
</pre>
<hr>

By using the '-rformat gff' qualifier, a GFF file of the predicted regions
can be produced. For example:
<p>

<pre>
% antigenic -rformat gff
Finds antigenic sites in proteins
Input sequence(s): sw:act1_fugru
Minimum length [6]: 
Output file [act1_fugru.antigenic]: 

% more act1_fugru.antigenic
##gff-version 2.0
##date 2002-02-11
##Type Protein ACT1_FUGRU
ACT1_FUGRU      antigenic       site    214     222     1.207   +       .	Sequence "ACT1_FUGRU.1" ; note "*pos 218"
ACT1_FUGRU      antigenic       site    131     145     1.187   +       .	Sequence "ACT1_FUGRU.2" ; note "*pos 137"
ACT1_FUGRU      antigenic       site    5       12      1.166   +       .	Sequence "ACT1_FUGRU.3" ; note "*pos 8"
ACT1_FUGRU      antigenic       site    27      38      1.164   +       .	Sequence "ACT1_FUGRU.4" ; note "*pos 32"
ACT1_FUGRU      antigenic       site    160     183     1.136   +       .	Sequence "ACT1_FUGRU.5" ; note "*pos 173"
ACT1_FUGRU      antigenic       site    367     372     1.135   +       .	Sequence "ACT1_FUGRU.6" ; note "*pos 372"
ACT1_FUGRU      antigenic       site    93      108     1.116   +       .	Sequence "ACT1_FUGRU.7" ; note "*pos 103"
ACT1_FUGRU      antigenic       site    295     301     1.113   +       .	Sequence "ACT1_FUGRU.8" ; note "*pos 296"
ACT1_FUGRU      antigenic       site    256     266     1.110   +       .	Sequence "ACT1_FUGRU.9" ; note "*pos 264"
ACT1_FUGRU      antigenic       site    336     352     1.107   +       .	Sequence "ACT1_FUGRU.10" ; note "*pos 347"
ACT1_FUGRU      antigenic       site    62      76      1.102   +       .	Sequence "ACT1_FUGRU.11" ; note "*pos 68"
ACT1_FUGRU      antigenic       site    232     250     1.086   +       .	Sequence "ACT1_FUGRU.12" ; note "*pos 245"
ACT1_FUGRU      antigenic       site    327     332     1.083   +       .	Sequence "ACT1_FUGRU.13" ; note "*pos 330"
ACT1_FUGRU      antigenic       site    317     323     1.074   +       .	Sequence "ACT1_FUGRU.14" ; note "*pos 320"
ACT1_FUGRU      antigenic       site    186     192     1.068   +       .	Sequence "ACT1_FUGRU.15" ; note "*pos 191"
ACT1_FUGRU      antigenic       site    40      46      1.066   +       .	Sequence "ACT1_FUGRU.16" ; note "*pos 43"
ACT1_FUGRU      antigenic       site    269     275     1.045   +       .	Sequence "ACT1_FUGRU.17" ; note "*pos 269"
ACT1_FUGRU      antigenic       site    51      57      1.034   +       .	Sequence "ACT1_FUGRU.18" ; note "*pos 52"
</pre>

<H2>
    Data files
</H2>


Antigenic uses a data file called <em>Eantigenic.dat</em>.

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by EMBOSS environment variable EMBOSS_DATA.

<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>

<p>
Here is the default <em>Eantigenic.dat</em> file:

<pre>

#                                               Antigenic  Surface  Antigenic
# Amino     -- Occurrence of amino acids in --   frequency frequency propensity
# Acid       Epitopes      Surface     Protein   f(Ag)    f(s)      A(p)
  A             135          328         524     0.065    0.061     1.064
  C              53           97         186     0.026    0.018     1.412
  D             118          352         414     0.057    0.066     0.866
  E             132          401         499     0.064    0.075     0.851
  F              76          180         365     0.037    0.034     1.091
  G             116          343         487     0.056    0.064     0.874
  H              59          138         191     0.029    0.026     1.105
  I              86          193         437     0.042    0.036     1.152
  K             158          439         523     0.076    0.082     0.930
  L             149          308         684     0.072    0.058     1.250
  M              23           72         152     0.011    0.013     0.826
  N              94          313         407     0.045    0.058     0.776
  P             135          328         411     0.065    0.061     1.064
  Q              99          252         332     0.048    0.047     1.015
  R             106          314         394     0.051    0.058     0.873
  S             168          429         553     0.081    0.080     1.012
  T             141          401         522     0.068    0.075     0.909
  V             128          239         515     0.062    0.045     1.383
  W              19           55         103     0.009    0.010     0.893
  Y              71          158         245     0.034    0.029     1.161
Total          2066         5340        7944

</pre>

<H2>
    Notes
</H2>


<H2>
    References
</H2>


<ol>
<li>
Kolaskar,AS and Tongaonkar,PC (1990).
A semi-empirical method for prediction of antigenic determinants
on protein antigens.
FEBS Letters 276: 172-174.

<li>
Parker,JMR, Guo,D and Hodges,RS (1986).
Biochemistry 25: 5425-5432.
</ol>

<H2>
    Warnings
</H2>


The program will warn you if the sequence is not a protein or has
ambiguity codes.


<H2>
    Diagnostic Error Messages
</H2>


<H2>
    Exit status
</H2>


It exits with status 0, unless a region is badly constructed.

<H2>
    Known bugs
</H2>


None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="digest.html">digest</a></td><td>Protein proteolytic enzyme or reagent cleavage digest</td></tr>
<tr><td><a href="fuzzpro.html">fuzzpro</a></td><td>Protein pattern search</td></tr>
<tr><td><a href="fuzztran.html">fuzztran</a></td><td>Protein pattern search after translation</td></tr>
<tr><td><a href="helixturnhelix.html">helixturnhelix</a></td><td>Report nucleic acid binding motifs</td></tr>
<tr><td><a href="oddcomp.html">oddcomp</a></td><td>Finds protein sequence regions with a biased composition</td></tr>
<tr><td><a href="patmatdb.html">patmatdb</a></td><td>Search a protein sequence with a motif</td></tr>
<tr><td><a href="patmatmotifs.html">patmatmotifs</a></td><td>Search a PROSITE motif database with a protein sequence</td></tr>
<tr><td><a href="pepcoil.html">pepcoil</a></td><td>Predicts coiled coil regions</td></tr>
<tr><td><a href="preg.html">preg</a></td><td>Regular expression search of a protein sequence</td></tr>
<tr><td><a href="pscan.html">pscan</a></td><td>Scans proteins using PRINTS</td></tr>
<tr><td><a href="sigcleave.html">sigcleave</a></td><td>Reports protein signal cleavage sites</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>

<p>
Original program "ANTIGENIC" by Peter Rice (EGCG 1991)

<H2>
    History
</H2>


Completed 9th March 1999

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>



</BODY>
</HTML>

