<HTML>

<HEAD>
  <TITLE>
  EMBOSS: backtranambig
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
backtranambig
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<H2>
    Function
</H2>
Back translate a protein sequence to ambiguous codons
<H2>
    Description
</H2>


<b>backtranambig</b> takes a protein sequence and makes the nucleic
acid sequence it could have come from.  It does this by generating
nucleotide ambiguity codes that represent all possible codons for each
amino acid.

<p>
The resulting ambiguous nucleotide sequence can be translated to the
original protein using <b>transeq</b>, which will recognise highly
redundant codons (for example "WSN" for serine) as being produced by a
program such as <b>backtranambig</b>.


<h3>Genetic code</h3>

<b>backtranambig</b> needs a genetic code to generate the ambiguous
codons.  The default genetic code is the standard ('Universal') code.


<H2>
    Usage
</H2>
<b>Here is a sample session with backtranambig</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>backtranambig </b>
Back translate a protein sequence to ambiguous codons
Input (gapped) protein sequence: <b>tsw:opsd_human</b>
(gapped) nucleotide output sequence [opsd_human.fasta]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>

<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-sequence]          sequence   (Gapped) protein sequence filename and
                                  optional format, or reference (input USA)
  [-outfile]           seqout     [<sequence>.<format>] (Gapped) nucleotide
                                  sequence filename and optional format
                                  (output USA)

   Additional (Optional) qualifiers:
   -table              menu       [0] Genetic code to use (Values: 0
                                  (Standard); 1 (Standard (with alternative
                                  initiation codons)); 2 (Vertebrate
                                  Mitochondrial); 3 (Yeast Mitochondrial); 4
                                  (Mold, Protozoan, Coelenterate Mitochondrial
                                  and Mycoplasma/Spiroplasma); 5
                                  (Invertebrate Mitochondrial); 6 (Ciliate
                                  Macronuclear and Dasycladacean); 9
                                  (Echinoderm Mitochondrial); 10 (Euplotid
                                  Nuclear); 11 (Bacterial); 12 (Alternative
                                  Yeast Nuclear); 13 (Ascidian Mitochondrial);
                                  14 (Flatworm Mitochondrial); 15
                                  (Blepharisma Macronuclear); 16
                                  (Chlorophycean Mitochondrial); 21 (Trematode
                                  Mitochondrial); 22 (Scenedesmus obliquus);
                                  23 (Thraustochytrium Mitochondrial))

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1            integer    Start of the sequence to be used
   -send1              integer    End of the sequence to be used
   -sreverse1          boolean    Reverse (if DNA)
   -sask1              boolean    Ask for begin/end/reverse
   -snucleotide1       boolean    Sequence is nucleotide
   -sprotein1          boolean    Sequence is protein
   -slower1            boolean    Make lower case
   -supper1            boolean    Make upper case
   -sformat1           string     Input sequence format
   -sdbname1           string     Database name
   -sid1               string     Entryname
   -ufo1               string     UFO features
   -fformat1           string     Features format
   -fopenfile1         string     Features file name

   "-outfile" associated qualifiers
   -osformat2          string     Output seq format
   -osextension2       string     File name extension
   -osname2            string     Base file name
   -osdirectory2       string     Output directory
   -osdbname2          string     Database name to add
   -ossingle2          boolean    Separate file for each entry
   -oufo2              string     UFO features
   -offormat2          string     Features format
   -ofname2            string     Features file name
   -ofdirectory2       string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>(Gapped) protein sequence filename and optional format, or reference (input USA)</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>(Gapped) nucleotide sequence filename and optional format (output USA)</td>
<td>Writeable sequence</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-table</td>
<td>Genetic code to use</td>
<td><table><tr><td>0</td> <td><i>(Standard)</i></td></tr><tr><td>1</td> <td><i>(Standard (with alternative initiation codons))</i></td></tr><tr><td>2</td> <td><i>(Vertebrate Mitochondrial)</i></td></tr><tr><td>3</td> <td><i>(Yeast Mitochondrial)</i></td></tr><tr><td>4</td> <td><i>(Mold, Protozoan, Coelenterate Mitochondrial and Mycoplasma/Spiroplasma)</i></td></tr><tr><td>5</td> <td><i>(Invertebrate Mitochondrial)</i></td></tr><tr><td>6</td> <td><i>(Ciliate Macronuclear and Dasycladacean)</i></td></tr><tr><td>9</td> <td><i>(Echinoderm Mitochondrial)</i></td></tr><tr><td>10</td> <td><i>(Euplotid Nuclear)</i></td></tr><tr><td>11</td> <td><i>(Bacterial)</i></td></tr><tr><td>12</td> <td><i>(Alternative Yeast Nuclear)</i></td></tr><tr><td>13</td> <td><i>(Ascidian Mitochondrial)</i></td></tr><tr><td>14</td> <td><i>(Flatworm Mitochondrial)</i></td></tr><tr><td>15</td> <td><i>(Blepharisma Macronuclear)</i></td></tr><tr><td>16</td> <td><i>(Chlorophycean Mitochondrial)</i></td></tr><tr><td>21</td> <td><i>(Trematode Mitochondrial)</i></td></tr><tr><td>22</td> <td><i>(Scenedesmus obliquus)</i></td></tr><tr><td>23</td> <td><i>(Thraustochytrium Mitochondrial)</i></td></tr></table></td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>

<H2>
    Input file format
</H2>


Any DNA sequence USA.

<p>

<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tsw:opsd_human' is a sequence entry in the example protein database 'tsw'
<p>
<p><h3>Database entry: tsw:opsd_human</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   OPSD_HUMAN     STANDARD;      PRT;   348 AA.
AC   P08100; Q16414;
DT   01-AUG-1988 (Rel. 08, Created)
DT   01-AUG-1988 (Rel. 08, Last sequence update)
DT   15-JUL-1999 (Rel. 38, Last annotation update)
DE   RHODOPSIN.
GN   RHO.
OS   Homo sapiens (Human).
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Mammalia;
OC   Eutheria; Primates; Catarrhini; Hominidae; Homo.
RN   [1]
RP   SEQUENCE FROM N.A.
RX   MEDLINE; 84272729.
RA   NATHANS J., HOGNESS D.S.;
RT   "Isolation and nucleotide sequence of the gene encoding human
RT   rhodopsin.";
RL   Proc. Natl. Acad. Sci. U.S.A. 81:4851-4855(1984).
RN   [2]
RP   SEQUENCE OF 1-120 FROM N.A.
RA   BENNETT J., BELLER B., SUN D., KARIKO K.;
RL   Submitted (NOV-1994) to the EMBL/GenBank/DDBJ databases.
RN   [3]
RP   REVIEW ON ADRP VARIANTS.
RX   MEDLINE; 94004905.
RA   AL-MAGHTHEH M., GREGORY C., INGLEHEARN C., HARDCASTLE A.,
RA   BHATTACHARYA S.;
RT   "Rhodopsin mutations in autosomal dominant retinitis pigmentosa.";
RL   Hum. Mutat. 2:249-255(1993).
RN   [4]
RP   VARIANT ADRP HIS-23.
RX   MEDLINE; 90136922.
RA   DRYJA T.P., MCGEE T.L., REICHEI E., HAHN L.B., COWLEY G.S.,
RA   YANDELL D.W., SANDBERG M.A., BERSON E.L.;
RT   "A point mutation of the rhodopsin gene in one form of retinitis
RT   pigmentosa.";
RL   Nature 343:364-366(1990).
RN   [5]
RP   VARIANTS ADRP.
RX   MEDLINE; 91051574.
RA   FARRAR G.J., KENNA P., REDMOND R., MCWILLIAM P., BRADLEY D.G.,
RA   HUMPHRIES M.M., SHARP E.M., INGLEHEARN C.F., BASHIR R., JAY M.,
RA   WATTY A., LUDWIG M., SCHINZEL A., SAMANNS C., GAL A.,
RA   BHATTACHARYA S.S., HUMPHRIES P.;
RT   "Autosomal dominant retinitis pigmentosa: absence of the rhodopsin
RT   proline--&gt;histidine substitution (codon 23) in pedigrees from
RT   Europe.";
RL   Am. J. Hum. Genet. 47:941-945(1990).
RN   [6]
RP   VARIANTS ADRP HIS-23; ARG-58; LEU-347 AND SER-347.
RX   MEDLINE; 91015273.


<font color=red>  [Part of this file has been deleted for brevity]</font>

FT                                /FTId=VAR_004816.
FT   VARIANT     209    209       V -&gt; M (EFFECT NOT KNOWN).
FT                                /FTId=VAR_004817.
FT   VARIANT     211    211       H -&gt; P (IN ADRP).
FT                                /FTId=VAR_004818.
FT   VARIANT     211    211       H -&gt; R (IN ADRP).
FT                                /FTId=VAR_004819.
FT   VARIANT     216    216       M -&gt; K (IN ADRP).
FT                                /FTId=VAR_004820.
FT   VARIANT     220    220       F -&gt; C (IN ADRP).
FT                                /FTId=VAR_004821.
FT   VARIANT     222    222       C -&gt; R (IN ADRP).
FT                                /FTId=VAR_004822.
FT   VARIANT     255    255       MISSING (IN ADRP).
FT                                /FTId=VAR_004823.
FT   VARIANT     264    264       MISSING (IN ADRP).
FT                                /FTId=VAR_004824.
FT   VARIANT     267    267       P -&gt; L (IN ADRP).
FT                                /FTId=VAR_004825.
FT   VARIANT     267    267       P -&gt; R (IN ADRP).
FT                                /FTId=VAR_004826.
FT   VARIANT     292    292       A -&gt; E (IN CSNB4).
FT                                /FTId=VAR_004827.
FT   VARIANT     296    296       K -&gt; E (IN ADRP).
FT                                /FTId=VAR_004828.
FT   VARIANT     297    297       S -&gt; R (IN ADRP).
FT                                /FTId=VAR_004829.
FT   VARIANT     342    342       T -&gt; M (IN ADRP).
FT                                /FTId=VAR_004830.
FT   VARIANT     345    345       V -&gt; L (IN ADRP).
FT                                /FTId=VAR_004831.
FT   VARIANT     345    345       V -&gt; M (IN ADRP).
FT                                /FTId=VAR_004832.
FT   VARIANT     347    347       P -&gt; A (IN ADRP).
FT                                /FTId=VAR_004833.
FT   VARIANT     347    347       P -&gt; L (IN ADRP; COMMON VARIANT).
FT                                /FTId=VAR_004834.
FT   VARIANT     347    347       P -&gt; Q (IN ADRP).
FT                                /FTId=VAR_004835.
FT   VARIANT     347    347       P -&gt; R (IN ADRP).
FT                                /FTId=VAR_004836.
FT   VARIANT     347    347       P -&gt; S (IN ADRP).
FT                                /FTId=VAR_004837.
SQ   SEQUENCE   348 AA;  38892 MW;  07443BEA CRC32;
     MNGTEGPNFY VPFSNATGVV RSPFEYPQYY LAEPWQFSML AAYMFLLIVL GFPINFLTLY
     VTVQHKKLRT PLNYILLNLA VADLFMVLGG FTSTLYTSLH GYFVFGPTGC NLEGFFATLG
     GEIALWSLVV LAIERYVVVC KPMSNFRFGE NHAIMGVAFT WVMALACAAP PLAGWSRYIP
     EGLQCSCGID YYTLKPEVNN ESFVIYMFVV HFTIPMIIIF FCYGQLVFTV KEAAAQQQES
     ATTQKAEKEV TRMVIIMVIA FLICWVPYAS VAFYIFTHQG SNFGPIFMTI PAFFAKSAAI
     YNPVIYIMMN KQFRNCMLTT ICCGKNPLGD DEASATVSKT ETSQVAPA
//
</pre>
</td></tr></table><p>

<H2>
    Output file format
</H2>


The output is a nucleotide sequence containing the most favoured back
translation of the specified protein, and using the specified
translation table (which defaults to human).

<p>

<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: opsd_human.fasta</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt;OPSD_HUMAN P08100 RHODOPSIN.
ATGAAYGGNACNGARGGNCCNAAYTTYTAYGTNCCNTTYWSNAAYGCNACNGGNGTNGTN
MGNWSNCCNTTYGARTAYCCNCARTAYTAYYTNGCNGARCCNTGGCARTTYWSNATGYTN
GCNGCNTAYATGTTYYTNYTNATHGTNYTNGGNTTYCCNATHAAYTTYYTNACNYTNTAY
GTNACNGTNCARCAYAARAARYTNMGNACNCCNYTNAAYTAYATHYTNYTNAAYYTNGCN
GTNGCNGAYYTNTTYATGGTNYTNGGNGGNTTYACNWSNACNYTNTAYACNWSNYTNCAY
GGNTAYTTYGTNTTYGGNCCNACNGGNTGYAAYYTNGARGGNTTYTTYGCNACNYTNGGN
GGNGARATHGCNYTNTGGWSNYTNGTNGTNYTNGCNATHGARMGNTAYGTNGTNGTNTGY
AARCCNATGWSNAAYTTYMGNTTYGGNGARAAYCAYGCNATHATGGGNGTNGCNTTYACN
TGGGTNATGGCNYTNGCNTGYGCNGCNCCNCCNYTNGCNGGNTGGWSNMGNTAYATHCCN
GARGGNYTNCARTGYWSNTGYGGNATHGAYTAYTAYACNYTNAARCCNGARGTNAAYAAY
GARWSNTTYGTNATHTAYATGTTYGTNGTNCAYTTYACNATHCCNATGATHATHATHTTY
TTYTGYTAYGGNCARYTNGTNTTYACNGTNAARGARGCNGCNGCNCARCARCARGARWSN
GCNACNACNCARAARGCNGARAARGARGTNACNMGNATGGTNATHATHATGGTNATHGCN
TTYYTNATHTGYTGGGTNCCNTAYGCNWSNGTNGCNTTYTAYATHTTYACNCAYCARGGN
WSNAAYTTYGGNCCNATHTTYATGACNATHCCNGCNTTYTTYGCNAARWSNGCNGCNATH
TAYAAYCCNGTNATHTAYATHATGATGAAYAARCARTTYMGNAAYTGYATGYTNACNACN
ATHTGYTGYGGNAARAAYCCNYTNGGNGAYGAYGARGCNWSNGCNACNGTNWSNAARACN
GARACNWSNCARGTNGCNCCNGCN
</pre>
</td></tr></table><p>


<H2>
    Data files
</H2>


The codon usage table is read by default from "Ehum.cut" in the 'data/CODONS'
directory of the EMBOSS distribution.  If the name of a codon usage file
is specified on the command line, then this file will first be searched
for in the current directory and then in the 'data/CODONS' directory of
the EMBOSS distribution. 

<p>
<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>


"Corrupt codon index file" - the codon usage file is incomplete or empty.

<p>
"The file 'drosoph.cut' does not exist" - the codon usage file cannot be opened.

<H2>
    Exit status
</H2>


This program always exits with a status of 0, unless the codon usage
table cannot be opened.

<H2>
    Known bugs
</H2>


None.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr>
<td><a href="backtranseq.html">backtranseq</a></td>
<td>Back translate a protein sequence</td>
</tr>

<tr>
<td><a href="charge.html">charge</a></td>
<td>Protein charge plot</td>
</tr>

<tr>
<td><a href="checktrans.html">checktrans</a></td>
<td>Reports STOP codons and ORF statistics of a protein</td>
</tr>

<tr>
<td><a href="coderet.html">coderet</a></td>
<td>Extract CDS, mRNA and translations from feature tables</td>
</tr>

<tr>
<td><a href="compseq.html">compseq</a></td>
<td>Count composition of dimer/trimer/etc words in a sequence</td>
</tr>

<tr>
<td><a href="emowse.html">emowse</a></td>
<td>Protein identification by mass spectrometry</td>
</tr>

<tr>
<td><a href="freak.html">freak</a></td>
<td>Residue/base frequency table or plot</td>
</tr>

<tr>
<td><a href="iep.html">iep</a></td>
<td>Calculates the isoelectric point of a protein</td>
</tr>

<tr>
<td><a href="mwcontam.html">mwcontam</a></td>
<td>Shows molwts that match across a set of files</td>
</tr>

<tr>
<td><a href="mwfilter.html">mwfilter</a></td>
<td>Filter noisy molwts from mass spec output</td>
</tr>

<tr>
<td><a href="octanol.html">octanol</a></td>
<td>Displays protein hydropathy</td>
</tr>

<tr>
<td><a href="pepinfo.html">pepinfo</a></td>
<td>Plots simple amino acid properties in parallel</td>
</tr>

<tr>
<td><a href="pepstats.html">pepstats</a></td>
<td>Protein statistics</td>
</tr>

<tr>
<td><a href="pepwindow.html">pepwindow</a></td>
<td>Displays protein hydropathy</td>
</tr>

<tr>
<td><a href="pepwindowall.html">pepwindowall</a></td>
<td>Displays protein hydropathy of a set of sequences</td>
</tr>

<tr>
<td><a href="plotorf.html">plotorf</a></td>
<td>Plot potential open reading frames</td>
</tr>

<tr>
<td><a href="prettyseq.html">prettyseq</a></td>
<td>Output sequence with translated ranges</td>
</tr>

<tr>
<td><a href="remap.html">remap</a></td>
<td>Display sequence with restriction sites, translation etc</td>
</tr>

<tr>
<td><a href="showorf.html">showorf</a></td>
<td>Pretty output of DNA translations</td>
</tr>

<tr>
<td><a href="showseq.html">showseq</a></td>
<td>Display a sequence with features, translation etc</td>
</tr>

<tr>
<td><a href="sixpack.html">sixpack</a></td>
<td>Display a DNA sequence with 6-frame translation and ORFs</td>
</tr>

<tr>
<td><a href="transeq.html">transeq</a></td>
<td>Translate nucleic acid sequences</td>
</tr>

</table>
<H2>
    Author(s)
</H2>
Alan Bleasby (ajb&nbsp;&copy;&nbsp;ebi.ac.uk)
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK



<H2>
    History
</H2>
None

<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.



<H2>
    Comments
</H2>
None


</BODY>
</HTML>

