<HTML>

<HEAD>
  <TITLE>
  EMBOSS: backtranseq
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: backtranseq
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program backtranseq
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Back translate a protein sequence


<H2>
    Description
</H2>


<b>backtranseq</b> takes a protein sequence and makes a best estimate of the
likely nucleic acid sequence it could have come from.  It does this by
using a codon frequency table. For each amino acid, the corresponding
most frequently occuring codon is used in the construction of the nucleic
acid sequence.

<h3>Codon usage table name</h3>

<b>backtranseq</b> reads in a data file containing the codon frequency
tables.  The default codon frequency table is 'Ehum.cut' - the human
codon frequency table.  It is important to use a codon frequency table
that is appropriate for the species that your protein comes from. 
See the <b>Data Files</b> section below for more details on these files.

<H2>
    Usage
</H2>


Here is a sample session with backtranseq.  Note that this is a human
protein and so the default (human) codon frequency file is used (i.e. 
is not specified). 
<p>

<pre>
% backtranseq
Back translate a protein sequence
Input sequence: sw:opsd_human
Output sequence [opsd_human.fasta]: 
</pre>

<p>

Here is a session using a drosophila sequence and codon table:

<p>

<pre>
% backtranseq -cfile Edrosophila.cut
Back translate a protein sequence
Input sequence: sw:ach2_drome
Output sequence [ach2_drome.fasta]: 
</pre>


<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          sequence   Sequence USA
  [-outfile]           seqout     Output sequence USA

   Optional qualifiers:
   -cfile              codon      Codon usage table name

   Advanced qualifiers: (none)
   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output sequence USA</td>
<td>Writeable sequence</td>
<td><i>&lt;sequence&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-cfile</td>
<td>Codon usage table name</td>
<td>Codon usage file in EMBOSS data path</td>
<td>Ehum.cut</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>


Any DNA sequence USA.

<H2>
    Output file format
</H2>


The output is a nucleotide sequence containing the most favoured back
translation of the specified protein, and using the specified
translation table (which defaults to human).

<p>

The output from the backtranslation of the human protein <b>sw:opsd_human</b>
follows:

<p>

<hr>
<pre>
% more opsd_human.fasta
>OPSD_HUMAN P08100 RHODOPSIN.
ATGAACGGCACCGAGGGCCCCAACTTCTACGTGCCCTTCAGCAACGCCACCGGCGTGGTG
AGAAGCCCCTTCGAGTACCCCCAGTACTACCTGGCCGAGCCCTGGCAGTTCAGCATGCTG
GCCGCCTACATGTTCCTGCTGATCGTGCTGGGCTTCCCCATCAACTTCCTGACCCTGTAC
GTGACCGTGCAGCACAAGAAGCTGAGAACCCCCCTGAACTACATCCTGCTGAACCTGGCC
GTGGCCGACCTGTTCATGGTGCTGGGCGGCTTCACCAGCACCCTGTACACCAGCCTGCAC
GGCTACTTCGTGTTCGGCCCCACCGGCTGCAACCTGGAGGGCTTCTTCGCCACCCTGGGC
GGCGAGATCGCCCTGTGGAGCCTGGTGGTGCTGGCCATCGAGAGATACGTGGTGGTGTGC
AAGCCCATGAGCAACTTCAGATTCGGCGAGAACCACGCCATCATGGGCGTGGCCTTCACC
TGGGTGATGGCCCTGGCCTGCGCCGCCCCCCCCCTGGCCGGCTGGAGCAGATACATCCCC
GAGGGCCTGCAGTGCAGCTGCGGCATCGACTACTACACCCTGAAGCCCGAGGTGAACAAC
GAGAGCTTCGTGATCTACATGTTCGTGGTGCACTTCACCATCCCCATGATCATCATCTTC
TTCTGCTACGGCCAGCTGGTGTTCACCGTGAAGGAGGCCGCCGCCCAGCAGCAGGAGAGC
GCCACCACCCAGAAGGCCGAGAAGGAGGTGACCAGAATGGTGATCATCATGGTGATCGCC
TTCCTGATCTGCTGGGTGCCCTACGCCAGCGTGGCCTTCTACATCTTCACCCACCAGGGC
AGCAACTTCGGCCCCATCTTCATGACCATCCCCGCCTTCTTCGCCAAGAGCGCCGCCATC
TACAACCCCGTGATCTACATCATGATGAACAAGCAGTTCAGAAACTGCATGCTGACCACC
ATCTGCTGCGGCAAGAACCCCCTGGGCGACGACGAGGCCAGCGCCACCGTGAGCAAGACC
GAGACCAGCCAGGTGGCCCCCGCC
</pre>
<hr>

<H2>
    Data files
</H2>


The codon usage table is read by default from "Ehum.cut" in the 'data/CODONS'
directory of the EMBOSS distribution.  If the name of a codon usage file
is specified on the command line, then this file will first be searched
for in the current directory and then in the 'data/CODONS' directory of
the EMBOSS distribution. 

<p>
To see the available EMBOSS codon usage files, run:

<pre>

% embossdata -showall

</pre>

<p>
To fetch one of the codon usage tables (for example 'Emus.cut') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Emus.cut

</pre>

<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>


"Corrupt codon index file" - the codon usage file is incomplete or empty.

<p>
"The file 'drosoph.cut' does not exist" - the codon usage file cannot be opened.

<H2>
    Exit status
</H2>


This program always exits with a status of 0, unless the codon usage
table cannot be opened.

<H2>
    Known bugs
</H2>


None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="charge.html">charge</a></td><td>Protein charge plot</td></tr>
<tr><td><a href="checktrans.html">checktrans</a></td><td>Reports STOP codons and ORF statistics of a protein sequence</td></tr>
<tr><td><a href="coderet.html">coderet</a></td><td>Extract CDS, mRNA and translations from feature tables</td></tr>
<tr><td><a href="compseq.html">compseq</a></td><td>Counts the composition of dimer/trimer/etc words in a sequence</td></tr>
<tr><td><a href="emowse.html">emowse</a></td><td>Protein identification by mass spectrometry</td></tr>
<tr><td><a href="freak.html">freak</a></td><td>Residue/base frequency table or plot</td></tr>
<tr><td><a href="iep.html">iep</a></td><td>Calculates the isoelectric point of a protein</td></tr>
<tr><td><a href="mwfilter.html">mwfilter</a></td><td>Filter noisy molwts from mass spec output</td></tr>
<tr><td><a href="octanol.html">octanol</a></td><td>Displays protein hydropathy</td></tr>
<tr><td><a href="pepinfo.html">pepinfo</a></td><td>Plots simple amino acid properties in parallel</td></tr>
<tr><td><a href="pepstats.html">pepstats</a></td><td>Protein statistics</td></tr>
<tr><td><a href="pepwindow.html">pepwindow</a></td><td>Displays protein hydropathy</td></tr>
<tr><td><a href="pepwindowall.html">pepwindowall</a></td><td>Displays protein hydropathy of a set of sequences</td></tr>
<tr><td><a href="plotorf.html">plotorf</a></td><td>Plot potential open reading frames</td></tr>
<tr><td><a href="prettyseq.html">prettyseq</a></td><td>Output sequence with translated ranges</td></tr>
<tr><td><a href="remap.html">remap</a></td><td>Display a sequence with restriction cut sites, translation etc</td></tr>
<tr><td><a href="showorf.html">showorf</a></td><td>Pretty output of DNA translations</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="transeq.html">transeq</a></td><td>Translate nucleic acid sequences</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->



<H2>
    Author(s)
</H2>


This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>


Completed 6 Oct 1999

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

