<HTML>

<HEAD>
  <TITLE>
  EMBOSS: banana
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: banana
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program banana
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Bending and curvature plot in B-DNA


<H2>
    Description
</H2>


<b>banana</b> predicts bending of a normal (B) DNA double helix, using
he method of Goodsell & Dickerson, NAR 1994 11;22(24):5497-503. 

<p>

This program calculates the magnitude of local bending and macroscopic
curvature at each point along an arbitrary B-DNA sequence, using any
desired bending model that specifies values of twist, roll and tilt as a
function of sequence. 

<p>

The data, based on the nucleosome positioning data of Satchwell et al
1986 (J.  Mol.  Biol.  191, 659-675), correctly predicts experimental
A-tract curvature as measured by gel retardation and cyclization
kinetics and successfully predicts curvature in regions containing
phased GGGCCC sequences. 

<p>

This model - showing local bending at mixed sequence DNA, strong bends
at the sequence GGC, and straight, rigid A-tracts - is the only model,
out of six models investigated in Goodsell & Dickerson paper, that is
consistent with both solution data from gel retardation and cyclization
kinetics and structural data from x-ray crystallography. 

<p>

The consensus sequence for DNA bending is 5 As and 5 non-As alternating. 
"N" is an ambiguity code for any base, and "B" is the ambiguity code for
"not A" so "BANANA" is itself a bent sequence - hence the name of this
program. 

<p>

The program outputs both a graphical display and a text file of the
results. 

<H2>
    Usage
</H2>


Here is a sample session with banana.

<p>

<pre>
% banana embl:rnu68037
Bending and curvature plot in B-DNA
Graph type [x11]: 
</pre>

<p>

<a href="banana0.gif">click here for result (1st page only)</a>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers (* if not always prompted):
  [-sequence]          sequence   Sequence USA
*  -graph              graph      Graph type

   Optional qualifiers:
   -anglesfile         datafile   angles file
   -residuesperline    integer    Number of residues to be displayed on each
                                  line
   -outfile            outfile    Output file name

   Advanced qualifiers:
   -data               bool       Output as data


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-graph</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including postscript, ps, hpgl, hp7470, hp7580, meta, colourps, cps, xwindows, x11, tektronics, tekt, tek4107t, tek, none, null, text, data, xterm</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-anglesfile</td>
<td>angles file</td>
<td>Data file</td>
<td>Eangles_tri.dat</td>
</tr>

<tr>
<td>-residuesperline</td>
<td>Number of residues to be displayed on each line</td>
<td>Any integer value</td>
<td>50</td>
</tr>

<tr>
<td>-outfile</td>
<td>Output file name</td>
<td>Output file</td>
<td>banana.profile</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-data</td>
<td>Output as data</td>
<td>Yes/No</td>
<td>No</td>
</tr>

</table>



<H2>
    Input file format
</H2>


Any DNA sequence USA.

<H2>
    Output file format
</H2>

The output is to both a graphical display and to a text file with the
default name 'banana.profile'.

<p>

The graphical display shows the sequence together with the local local
bending (solid line) and macroscopic curvature (dotted line). 

<p>

The output data file from the above example is as follows:

<p>

<hr>
<pre>
t    0.0      0.0
g   17.7      0.0
a   21.1      0.0
g   28.5      0.0
c   26.2      0.0
c   19.7      0.0
c   18.7      0.0
c   12.5      0.0
t    9.7      0.0
a   14.9      0.0
c   16.5      0.0
g   17.5      0.0
g   26.2      0.0
g   28.5      0.0
c   20.7      0.0
t   11.7      0.0
t    6.4      0.0
a    9.3      0.0
a   14.9      0.0
c   17.7      0.0
c   15.7     19.2
t   15.7     18.5
g   17.7     17.9
a   21.1     17.1
g   28.5     15.9
c   25.2     14.6
c   12.5     13.3
t    7.2     11.9
a   13.2     10.8
g   20.1     10.1
t   19.5      9.6
g   15.1      9.2
g   14.9      9.1
a   19.5      9.5
t   19.7     10.2
g   17.7     10.8
a   17.7     11.0
g   25.2     11.2
g   26.2     11.3
c   15.3     11.5
a   11.4     11.7
a   14.5     12.0
c   13.9     12.2
a   11.4     12.3
[output truncated for brevity]
</pre>
<hr>

<p>

The data file consists of three columns separated by blanks or tab
characters. 

<p>

The first column is the sequence.  
<br> 
The second column is the local bending.
<br>
The third is the curvature. 


<H2>
    Data files
</H2>

It reads in angles files for the twist, roll and tilt angles.  By
default Eangles_tri.dat is used, as in 
Goodsell & Dickerson, NAR 1994 11;22(24):5497-503 and Drew and Travers (1986) JMB 191, 659

<H2>
    Notes
</H2>

None. 

<H2>
    References
</H2>

<ol>
<li> Goodsell, D.S. & Dickerson, R.E. (1994) "Bending and Curvature Calculations in B-DNA" Nucl. Acids. Res. 22, 5497-5503.
<li>Drew and Travers (1986) JMB 191, 659
</ol>

<H2>
    Warnings
</H2>

Only ACTG allowed, if sequence contains a non ACTG character then the
    program will exit with a fatal error message.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

    0 if successfull.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="btwisted.html">btwisted</a></td><td>Calculates the twisting in a B-DNA sequence</td></tr>
<tr><td><a href="complex.html">complex</a></td><td>Find the linguistic complexity in nucleotide sequences</td></tr>
<tr><td><a href="dan.html">dan</a></td><td>Calculates DNA RNA/DNA melting temperature</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Ian Longden <a
href="mailto:il@sanger.ac.uk">(il@sanger.ac.uk)</a> Informatics
Division, The Sanger Centre, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.

<H2>
    History
</H2>

<pre>
Created 1999/06/09.
Last Updated 1999/06/14.
</pre>

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

