<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: btwisted
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">


<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

btwisted
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>





<!--END OF HEADER-->

<H2>
    Function
</H2>
Calculates the twisting in a B-DNA sequence

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
        Algorithms
  -->



   <b>btwisted</b> takes a region of a pure DNA sequence and calculates by simple 
   arithmetic the probable overall twist of the sequence and the stacking 
   energy.











<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

<b>Here is a sample session with btwisted</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>btwisted -auto tembl:ab000095 -sbegin 100 -send 120 </b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>


<!--
Here is a sample session with btwisted:

<pre>

% btwisted embl:ab000095 -sbegin 100 -send 120
Calculates the twisting in a B-DNA sequence 
Output file [ab000095.btwisted]:

</pre>
-->













<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          sequence   Sequence USA
  [-outfile]           outfile    Output file name

   Optional qualifiers: (none)
   Advanced qualifiers:
   -angledata          string     File containing base pair twist angles
   -energydata         string     File containing base pair stacking energies

   Associated qualifiers:
  "-sequence" related qualifiers
   -sbegin1             integer    First base used
   -send1               integer    Last base used, def=seq length
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sopenfile1          string     Input filename
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name
  "-outfile" related qualifiers
   -odirectory2         string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for required and optional values
   -debug               boolean    Write debug output to program.dbg
   -acdlog              boolean    Write ACD processing log to program.acdlog
   -acdpretty           boolean    Rewrite ACD file as program.acdpretty
   -acdtable            boolean    Write HTML table of options
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.btwisted</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-angledata</td>
<td>File containing base pair twist angles</td>
<td>Any string is accepted</td>
<td>Eangles.dat</td>
</tr>

<tr>
<td>-energydata</td>
<td>File containing base pair stacking energies</td>
<td>Any string is accepted</td>
<td>Eenergy.dat</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

Nucleic acid sequence USA.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:ab000095' is a sequence entry in the example nucleic acid database 'tembl'
<p>
<p><h3>Database entry: tembl:ab000095</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   AB000095   standard; RNA; HUM; 2399 BP.
XX
AC   AB000095;
XX
SV   AB000095.1
XX
DT   10-MAR-1998 (Rel. 54, Created)
DT   10-MAR-1998 (Rel. 54, Last updated, Version 1)
XX
DE   Homo sapiens mRNA for hepatocyte growth factor activator inhibitor,
DE   complete cds.
XX
KW   hepatocyte growth factor activator inhibitor.
XX
OS   Homo sapiens (human)
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia;
OC   Eutheria; Primates; Catarrhini; Hominidae; Homo.
XX
RN   [1]
RP   1-2399
RA   Denda K.;
RT   ;
RL   Submitted (24-DEC-1996) to the EMBL/GenBank/DDBJ databases.
RL   Kimitoshi Denda, Tokyo Institute of Technology, Department of Life Science;
RL   4259 Nagatsuta, Midori-ku, Yokohama, Kanagawa 227, Japan
RL   (E-mail:kdenda@bio.titech.ac.jp, Tel:45-924-5702, Fax:45-924-5771)
XX
RN   [2]
RA   Shimomura T., Denda K., Kitamura A., Kawaguchi T., Kito M., Kondo J.,
RA   Kagaya S., Qin L., Takata H., Miyazawa K., Kitamura N.;
RT   "Hepatocyte growth factor activator inhibitor, a novel Kunitz-type serine
RT   protease inhibitor";
RL   J. Biol. Chem. 272:6370-6376(1997).
XX
DR   SWISS-PROT; O43278; SPT1_HUMAN.
XX
FH   Key             Location/Qualifiers
FH
FT   source          1..2399
FT                   /db_xref="taxon:9606"
FT                   /sequenced_mol="cDNA to mRNA"
FT                   /organism="Homo sapiens"
FT   CDS             176..1717
FT                   /db_xref="SWISS-PROT:O43278"
FT                   /product="hepatocyte growth factor activator inhibitor"
FT                   /protein_id="BAA25014.1"
FT                   /translation="MAPARTMARARLAPAGIPAVALWLLCTLGLQGTQAGPPPAPPGLP
FT                   AGADCLNSFTAGVPGFVLDTNASVSNGATFLESPTVRRGWDCVRACCTTQNCNLALVEL
FT                   QPDRGEDAIAACFLINCLYEQNFVCKFAPREGFINYLTREVYRSYRQLRTQGFGGSGIP
FT                   KAWAGIDLKVQPQEPLVLKDVENTDWRLLRGDTDVRVERKDPNQVELWGLKEGTYLFQL
FT                   TVTSSDHPEDTANVTVTVLSTKQTEDYCLASNKVGRCRGSFPRWYYDPTEQICKSFVYG
FT                   GCLGNKNNYLREEECILACRGVQGPSMERRHPVCSGTCQPTQFRCSNGCCIDSFLECDD
FT                   TPNCPDASDEAACEKYTSGFDELQRIHFPSDKGHCVDLPDTGLCKESIPRWYYNPFSEH
FT                   CARFTYGGCYGNKNNFEEEQQCLESCRGISKKDVFGLRREIPIPSTGSVEMAVAVFLVI
FT                   CIVVVVAILGYCFFKNQRKDFHGHHHHPPPTPASSTVSTTEDTEHLVYNHTTRPL"
FT   polyA_signal    2379..2384
XX
SQ   Sequence 2399 BP; 490 A; 777 C; 684 G; 448 T; 0 other;
     cggccgagcc cagctctccg agcaccgggt cggaagccgc gacccgagcc gcgcaggaag        60
     ctgggaccgg aacctcggcg gacccggccc cacccaactc acctgcgcag gtcaccagca       120
     ccctcggaac ccagaggccc gcgctctgaa ggtgaccccc ctggggagga aggcgatggc       180
     ccctgcgagg acgatggccc gcgcccgcct cgccccggcc ggcatccctg ccgtcgcctt       240
     gtggcttctg tgcacgctcg gcctccaggg cacccaggcc gggccaccgc ccgcgccccc       300
     tgggctgccc gcgggagccg actgcctgaa cagctttacc gccggggtgc ctggcttcgt       360
     gctggacacc aacgcctcgg tcagcaacgg agctaccttc ctggagtccc ccaccgtgcg       420
     ccggggctgg gactgcgtgc gcgcctgctg caccacccag aactgcaact tggcgctagt       480
     ggagctgcag cccgaccgcg gggaggacgc catcgccgcc tgcttcctca tcaactgcct       540
     ctacgagcag aacttcgtgt gcaagttcgc gcccagggag ggcttcatca actacctcac       600
     gagggaagtg taccgctcct accgccagct gcggacccag ggctttggag ggtctgggat       660
     ccccaaggcc tgggcaggca tagacttgaa ggtacaaccc caggaacccc tggtgctgaa       720
     ggatgtggaa aacacagatt ggcgcctact gcggggtgac acggatgtca gggtagagag       780
     gaaagaccca aaccaggtgg aactgtgggg actcaaggaa ggcacctacc tgttccagct       840
     gacagtgact agctcagacc acccagagga cacggccaac gtcacagtca ctgtgctgtc       900
     caccaagcag acagaagact actgcctcgc atccaacaag gtgggtcgct gccggggctc       960
     tttcccacgc tggtactatg accccacgga gcagatctgc aagagtttcg tttatggagg      1020
     ctgcttgggc aacaagaaca actaccttcg ggaagaagag tgcattctag cctgtcgggg      1080
     tgtgcaaggc ccctccatgg aaaggcgcca tccagtgtgc tctggcacct gtcagcccac      1140
     ccagttccgc tgcagcaatg gctgctgcat cgacagtttc ctggagtgtg acgacacccc      1200
     caactgcccc gacgcctccg acgaggctgc ctgtgaaaaa tacacgagtg gctttgacga      1260
     gctccagcgc atccatttcc ccagtgacaa agggcactgc gtggacctgc cagacacagg      1320
     actctgcaag gagagcatcc cgcgctggta ctacaacccc ttcagcgaac actgcgcccg      1380
     ctttacctat ggtggttgtt atggcaacaa gaacaacttt gaggaagagc agcagtgcct      1440
     cgagtcttgt cgcggcatct ccaagaagga tgtgtttggc ctgaggcggg aaatccccat      1500
     tcccagcaca ggctctgtgg agatggctgt cgcagtgttc ctggtcatct gcattgtggt      1560
     ggtggtagcc atcttgggtt actgcttctt caagaaccag agaaaggact tccacggaca      1620
     ccaccaccac ccaccaccca cccctgccag ctccactgtc tccactaccg aggacacgga      1680
     gcacctggtc tataaccaca ccacccggcc cctctgagcc tgggtctcac cggctctcac      1740
     ctggccctgc ttcctgcttg ccaaggcaga ggcctgggct gggaaaaact ttggaaccag      1800
     actcttgcct gtttcccagg cccactgtgc ctcagagacc agggctccag cccctcttgg      1860
     agaagtctca gctaagctca cgtcctgaga aagctcaaag gtttggaagg agcagaaaac      1920
     ccttgggcca gaagtaccag actagatgga cctgcctgca taggagtttg gaggaagttg      1980
     gagttttgtt tcctctgttc aaagctgcct gtccctaccc catggtgcta ggaagaggag      2040
     tggggtggtg tcagaccctg gaggccccaa ccctgtcctc ccgagctcct cttccatgct      2100
     gtgcgcccag ggctgggagg aaggacttcc ctgtgtagtt tgtgctgtaa agagttgctt      2160
     tttgtttatt taatgctgtg gcatgggtga agaggagggg aagaggcctg tttggcctct      2220
     ctgtcctctc ttcctcttcc cccaagattg agctctctgc ccttgatcag ccccaccctg      2280
     gcctagacca gcagacagag ccaggagagg ctcagctgca ttccgcagcc cccaccccca      2340
     aggttctcca acatcacagc ccagcccacc cactgggtaa taaaagtggt ttgtggaaa       2399
//
</pre>
</td></tr></table><p>









<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: ab000095.btwisted</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
# Output from BTWISTED
# Twisting calculated from 100 to 120 of AB000095
Total twist (degrees): 681.1
Total turns : 1.89
Average bases per turn: 11.10
Total stacking energy : -179.34
Average stacking energy per dinucleotide: -8.97
</pre>
</td></tr></table><p>


<!--
This is the output from the example usage above:
<p>

<pre>
# Output from BTWISTED
# Twisting calculated from 100 to 120 of AB000095
Total twist (degrees): 646.6
Total turns : 1.80
Average bases per turn: 11.69
Total stacking energy : -172.77
Average stacking energy per dinucleotide: -8.64 
</pre>

-->















<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.
	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

   -->

The program reads in two data files.
<P>
The first is prompted for by:
<P>
"File containing base pair twist angles"
<P>
<br>
This is the data file 'Eangles.dat' by default.
<P>
This consists of the data:
<P>
<pre>
aa      35.6    0.1
ac      34.4    1.3
ag      27.7    1.5
at      31.5    1.1
ca      34.5    0.9
cc      33.7    0.1
cg      29.8    1.1
ct      27.7    1.5
ga      36.9    0.9
gc      40.0    1.2
gg      33.7    0.1
gt      34.4    1.3
ta      36.0    1.0
tc      36.9    0.9
tg      34.5    0.9
tt      35.6    0.1
</pre>
<p>




The second is prompted for by:
<P>
"File containing base pair stacking energies"
<br>
This is the data file 'Eenergy.dat' by default.
<P>
This consists of the data:
<P>
<pre>
#base pair stacking energy for B-DNA
aa      -5.37
ac      -10.51
ag      -6.78
at      -6.57
ca      -6.57
cc      -8.26
cg      -9.69
ct      -6.78
ga      -9.81
gc      -14.59
gg      -8.26
gt      -10.51
ta      -3.82
tc      -9.81
tg      -6.57
tt      -5.37
</pre>
<P>

  No check is made that all dinucleotides have been read for the energy and 
  twist. Attempting to run btwisted with an incomplete data file will result
  a fatal error if the sequence contains a dinucleotide not in the data files. 
 

<P>
<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>






<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->


  No check is made that all dinucleotides have been read for the energy and 
  twist. Attempting to run btwisted with an incomplete data file will result
  a fatal error if the sequence contains a dinucleotide not in the data files. 





<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

"Incomplete table" 

<BR> 

You have supplied data files for either the angles or energies that do
not contain the fulll required set of possible dinucleotides. 







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

  No check is made that all dinucleotides have been read for the energy and 
  twist. Attempting to run btwisted with an incomplete data file will result
  a fatal error if the sequence contains a dinucleotide not in the data files. 



<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="banana.html">banana</a></td><td>Bending and curvature plot in B-DNA</td></tr>
<tr><td><a href="chaos.html">chaos</a></td><td>Create a chaos game representation plot for a sequence</td></tr>
<tr><td><a href="compseq.html">compseq</a></td><td>Counts the composition of dimer/trimer/etc words in a sequence</td></tr>
<tr><td><a href="dan.html">dan</a></td><td>Calculates DNA RNA/DNA melting temperature</td></tr>
<tr><td><a href="freak.html">freak</a></td><td>Residue/base frequency table or plot</td></tr>
<tr><td><a href="isochore.html">isochore</a></td><td>Plots isochores in large DNA sequences</td></tr>
<tr><td><a href="sirna.html">sirna</a></td><td>Finds siRNA duplexes in mRNA</td></tr>
<tr><td><a href="wordcount.html">wordcount</a></td><td>Counts words of a specified size in a DNA sequence</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->



   This application was written by
David Martin (dmartin&nbsp;&copy;&nbsp;hgmp.mrc.ac.uk)
<br>

<br>
and
Alan Bleasby (ableasby&nbsp;&copy;&nbsp;hgmp.mrc.ac.uk)
<br>
HGMP-RC, Genome Campus, Hinxton, Cambridge CB10 1SB, UK

<br>
   during the first Scandinavian EMBOSS Workshop.









<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (Sept 2000) - David Martin & Alan Bleasby.

<!--
<p>
[an error occurred while processing this directive]
-->






<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->








</BODY>
</HTML>

