<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: charge
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">
<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>
<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>
</td>
<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: charge
</H2>
</font>
</td>
</tr>
</table>
</td>
</tr>
</table>
<p><hr><p>

<!--END OF HEADER-->

<H1>
Program charge
</H1>
<H2>
    Function
</H2>
Protein charge plot

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
        Algorithms
  -->

<b>charge</b> reads a protein sequence and writes a file (or plots a
graph) of the charges of the amino acids within a window of specified
length as the window is moved along the sequence. 













<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->


Here is a sample session with charge:

<p>

<pre>
% charge sw:hbb_human 
Protein charge plot
Output file [hbb_human.charge]: 
</pre>











<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers (* if not always prompted):
  [-seqall]            seqall     Sequence database USA
*  -graph              xygraph    Graph type
*  -outfile            outfile    Output file name

   Optional qualifiers:
   -window             integer    Window

   Advanced qualifiers:
   -plot               bool       Produce graphic
   -aadata             string     Amino acid property data file name

   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-seqall]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-graph</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including postscript, ps, hpgl, hp7470, hp7580, meta, colourps, cps, xwindows, x11, tektronics, tekt, tek4107t, tek, none, null, text, data, xterm, png</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr>
<td>-outfile</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.charge</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-window</td>
<td>Window</td>
<td>Integer 1 or more</td>
<td>5</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-plot</td>
<td>Produce graphic</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-aadata</td>
<td>Amino acid property data file name</td>
<td>Any string is accepted</td>
<td>Eamino.dat</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>charge</b> reads in a protein sequence.













<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->

If the option '-plot' is specified then <b>charge</b> displays a graph
of the charge along the sequence otherwise it writes out a file
containing the charges within a window starting at each position along
the sequence. 

<p>

The output file contains two columns separated by space or TAB
characters.  The first column is the position of teh start of the
window.  The second column is the charge of the amino acids within that
window. 

<p>

The output of the example above is as follows:

<p>
<hr>
<pre>
CHARGE of HBB_HUMAN from 1 to 146: window 5

Position        Charge
1               0.100
2               -0.100
3               -0.400
4               -0.200
5               -0.200
6               -0.200
7               0.000
8               0.200
9               0.000
10              0.000
11              0.000
12              0.000
13              0.200
14              0.200
15              0.200
16              0.200
17              0.000
18              -0.400
19              -0.400
20              -0.400
21              -0.400
22              -0.400
23              -0.200
24              -0.200
25              -0.200
26              0.000
27              0.200
28              0.200
29              0.200
30              0.200
31              0.000
32              0.000
33              0.000
34              0.000
35              0.000
36              0.200
37              0.200
38              0.200
39              0.000
40              0.000
41              -0.200
42              -0.200
43              -0.400
44              -0.200
45              -0.200
46              -0.200
47              -0.200
48              -0.200
49              -0.200
50              -0.200
51              -0.200
52              -0.200
53              0.000
54              0.000
55              0.200
56              0.200
57              0.400
58              0.400
59              0.500
60              0.300
61              0.500
62              0.500
63              0.500
64              0.400
65              0.400
66              0.200
67              0.000
68              0.000
69              -0.200
70              -0.200
71              -0.200
72              -0.200
73              -0.100
74              0.100
75              -0.100
76              -0.100
77              -0.100
78              0.000
79              0.000
80              0.200
81              0.200
82              0.200
83              0.000
84              0.000
85              0.000
86              -0.200
87              -0.200
88              -0.100
89              -0.100
90              -0.300
91              0.100
92              0.100
93              0.100
94              0.100
95              0.100
96              -0.100
97              -0.300
98              -0.400
99              -0.400
100             0.000
101             0.000
102             0.200
103             0.200
104             0.200
105             0.000
106             0.000
107             0.000
108             0.000
109             0.000
110             0.000
111             0.000
112             0.100
113             0.200
114             0.200
115             0.200
116             0.400
117             0.100
118             0.000
119             0.000
120             0.000
121             -0.200
122             0.000
123             0.000
124             0.000
125             0.000
126             0.000
127             0.000
128             0.200
129             0.200
130             0.200
131             0.200
132             0.200
133             0.000
134             0.000
135             0.000
136             0.000
137             0.000
138             0.000
139             0.100
140             0.300
141             0.300
142             0.400
</pre>
<hr>














<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.
	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
   -->

<b>charge</b> reads the data file 'Eamino.dat' to find the charge of the amino acids in the protein.

<p>

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>














<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="backtranseq.html">backtranseq</a></td><td>Back translate a protein sequence</td></tr>
<tr><td><a href="checktrans.html">checktrans</a></td><td>Reports STOP codons and ORF statistics of a protein sequence</td></tr>
<tr><td><a href="compseq.html">compseq</a></td><td>Counts the composition of dimer/trimer/etc words in a sequence</td></tr>
<tr><td><a href="emowse.html">emowse</a></td><td>Protein identification by mass spectrometry</td></tr>
<tr><td><a href="iep.html">iep</a></td><td>Calculates the isoelectric point of a protein</td></tr>
<tr><td><a href="octanol.html">octanol</a></td><td>Displays protein hydropathy</td></tr>
<tr><td><a href="pepinfo.html">pepinfo</a></td><td>Plots simple amino acid properties in parallel</td></tr>
<tr><td><a href="pepstats.html">pepstats</a></td><td>Protein statistics</td></tr>
<tr><td><a href="pepwindow.html">pepwindow</a></td><td>Displays protein hydropathy</td></tr>
<tr><td><a href="pepwindowall.html">pepwindowall</a></td><td>Displays protein hydropathy of a set of sequences</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>

















<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (March 2001) - Alan Bleasby.













<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->








</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

