<HTML>

<HEAD>
  <TITLE>
  EMBOSS: checktrans
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">


<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

checktrans
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>







<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Reports STOP codons and ORF statistics of a protein


<H2>
    Description
</H2>


Reads in a protein sequence containing stops, and writes a report of any
open reading frames (continuous protein sequence with no stops) that are
greater than a minimum size.  The default minimum ORF size is 100
residues. It writes out any ORF sequences.

<p>

The input sequence might typically have been produced by <b>transeq</b>.

<p>

Note that if you have only translated a nucleic sequence in one frame,
<b>checktrans</b> will miss possible ORFs in other frames.  You have to
give <b>checktrans</b> translations in all three (six?) frames in order
for it to be effective at finding all possible ORFs. 


<H2>
    Usage
</H2>

<b>Here is a sample session with checktrans</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>checktrans </b>
Reports STOP codons and ORF statistics of a protein
Input sequence(s): <b>../../data/paamir.pep</b>
Minimum ORF Length to report [100]: <b></b>
Output file [paamir_1.checktrans]: <b></b>
Output sequence [paamir_1.fasta]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>


<!--
Here is a sample session with checktrans, using the output from a transeq run.

<p>

<pre>
% transeq embl:paamir paamir.pep -auto
% checktrans
Input sequence: paamir.pep
Minimum ORF Length to report [100]: 30
Output file [paamir_1.checktrans]: 
Output sequence [paamir_1.fasta]: 
</pre>
-->

<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
   -orfml              integer    Minimum ORF Length to report
  [-report]            outfile    Output file name
   -outseq             seqoutall  Sequence file to hold output ORF sequences

   Optional qualifiers:
   -[no]addlast        boolean    An asterisk in the protein sequence
                                  indicates the position of a STOP codon.
                                  Checktrans assumes that all ORFs end in a
                                  STOP codon. Forcing the sequence to end with
                                  an asterisk, if there is not one there
                                  already, makes checktrans treat the end as a
                                  potential ORF. If an asterisk is added, it
                                  is not included in the reported count of
                                  STOPs.

   Advanced qualifiers:
   -featout            featout    File for output features

   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-orfml</td>
<td>Minimum ORF Length to report</td>
<td>Integer 1 or more</td>
<td>100</td>
</tr>

<tr>
<td>[-report]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.checktrans</td>
</tr>

<tr>
<td>-outseq</td>
<td>Sequence file to hold output ORF sequences</td>
<td>Writeable sequence(s)</td>
<td><i>&lt;sequence&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]addlast</td>
<td>An asterisk in the protein sequence indicates the position of a STOP codon. Checktrans assumes that all ORFs end in a STOP codon. Forcing the sequence to end with an asterisk, if there is not one there already, makes checktrans treat the end as a potential ORF. If an asterisk is added, it is not included in the reported count of STOPs.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-featout</td>
<td>File for output features</td>
<td>Writeable feature table</td>
<td><i>unknown.gff</i></td>
</tr>

</table>



<H2>
    Input file format
</H2>


This program reads the USA of a protein sequence with STOP codons in it.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: ../../data/paamir.pep</h3>
<table width="90%"><tr><td bgcolor="#FFDDFF">
<pre>
&gt;PAAMIR_1 Pseudomonas aeruginosa amiC and amiR gene for aliphatic amidase regulation
GTAGRASARSPPAGRRELHDLPGEPGARAGSLRTALSDSHRRGNGWDRTRSGR*SACCSP
KPASPPISSARTRMAHCSRSSN*TARAASAVARSKRCPRTPAATRTAIGCAPRTSFATGG
YGSSWAATCRTRARR*CRWSSAPTRCSATRPPTRASSIRRTSSTAVRRRTRTVRRWRRT*
FATTASGWCSSARTTSIRGKATM*CATCIASTAARCSRKSTFRCIPPTTTCSAPSSASTR
RAPTWSSPPWWAPAPPSCIAPSPVATATAGGRRSPA*PPARRRWRRWRVTWQRGRWWSRL
TSPASIRPPAGPSSRPAMVSSRRTRPSPPGPRRPTGRPCCSAAPRRPQATGGWKTCSGTC
TTSTSTRHRGRSGWSARTTTAACLRASRKSMRAACSRSAGSRPNRFAPTLMSSCITSTTG
PPAWAGDRSHERQLAARQPARVAGAGPQPAGGGQRRPGLAADPHRLFGAPVLAAAGSLRR
AGGRGLHQHFPEWPPRRDRCAARRRDSAHYPGGAGGVRKPRGALADHRAGVPRRDHPAAR
CPPGAACAGIGAAHQRGNGEAEAEDRAAPGPHRRPGPDQPGQGVADAAPWLGRARGAPAP
VAGSDEAARADPEDRSGVAGKRAVRLSDPGRPEQ*QEGYRHHAGTGSAVRWRGAVSQCRL
VAGQDQRSGGGGDQLPGRRAERLRRVLPDLFRSSRAGLAEGRSADPAIRFYLSVGGRQPV
PRX
</pre>
</td></tr></table><p>


<H2>
    Output file format
</H2>


This program writes three files: the ORF report file
(paamir_1.checktrans), the output sequence file (paamir_1.fasta) and the
feature file (paamir_1.out3) which is in GFF format by default. 

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: paamir_1.checktrans</h3>
<table width="90%"><tr><td bgcolor="#DDFFDD">
<pre>


CHECKTRANS of PAAMIR_1 from 1 to 724

	ORF#	Pos	Len	ORF Range	Sequence name

	7	635	357	278-634	PAAMIR_1_7

	Total STOPS:     7

 </pre>
</td></tr></table><p>
<p><h3>File: paamir_1.fasta</h3>
<table width="90%"><tr><td bgcolor="#DDFFDD">
<pre>
&gt;PAAMIR_1_7
PPARRRWRRWRVTWQRGRWWSRLTSPASIRPPAGPSSRPAMVSSRRTRPSPPGPRRPTGR
PCCSAAPRRPQATGGWKTCSGTCTTSTSTRHRGRSGWSARTTTAACLRASRKSMRAACSR
SAGSRPNRFAPTLMSSCITSTTGPPAWAGDRSHERQLAARQPARVAGAGPQPAGGGQRRP
GLAADPHRLFGAPVLAAAGSLRRAGGRGLHQHFPEWPPRRDRCAARRRDSAHYPGGAGGV
RKPRGALADHRAGVPRRDHPAARCPPGAACAGIGAAHQRGNGEAEAEDRAAPGPHRRPGP
DQPGQGVADAAPWLGRARGAPAPVAGSDEAARADPEDRSGVAGKRAVRLSDPGRPEQ
</pre>
</td></tr></table><p>
<p><h3>File: paamir_1.out3</h3>
<table width="90%"><tr><td bgcolor="#DDFFDD">
<pre>
##gff-version 2.0
##date 2002-11-07
##Type Protein PAAMIR_1
PAAMIR_1	checktrans	misc_feature	278	634	0.000	+	.	Sequence "PAAMIR_1.1"
</pre>
</td></tr></table><p>


<!--
<p>
The ORF report file from the above example run is:

<p>

<hr>
<pre>
CHECKTRANS of PAAMIR_1 from 1 to 723

        ORF#    Pos     Len     ORF Range       Sequence name

        1       54      53      1-53    	PAAMIR_1_1
        3       136     52      84-135  	PAAMIR_1_3
        4       180     43      137-179 	PAAMIR_1_4
        6       277     72      205-276 	PAAMIR_1_6
        7       635     357     278-634 	PAAMIR_1_7

        Total STOPS:     7
</pre>
<hr>
-->

<p>

The ORF report file gives the numeric count of the ORF, the position of
the terminating STOP codon, the length of the ORF, its start and end
positions and the name of the sequence it has been written out as. 

<p>

The name of the output sequences is constructed from the name of the
input sequence followed by an underscore and then the numeric count of
the ORF (e.g. 'PAAMIR_1_7').

<!--
<p>
The output sequence file is:

<p>

<hr>
<pre>
&gt;PAAMIR_1_1
GTAGRASARSPPAGRRELHDLPGEPGARAGSLRTALSDSHRRGNGWDRTRSGR
&gt;PAAMIR_1_3
TARAASAVARSKRCPRTPAATRTAIGCAPRTSFATGGYGSSWAATCRTRARR
&gt;PAAMIR_1_4
CRWSSAPTRCSATRPPTRASSIRRTSSTAVRRRTRTVRRWRRT
&gt;PAAMIR_1_6
CATCIASTAARCSRKSTFRCIPPTTTCSAPSSASTRRAPTWSSPPWWAPAPPSCIAPSPV
ATATAGGRRSPA
&gt;PAAMIR_1_7
PPARRRWRRWRVTWQRGRWWSRLTSPASIRPPAGPSSRPAMVSSRRTRPSPPGPRRPTGR
PCCSAAPRRPQATGGWKTCSGTCTTSTSTRHRGRSGWSARTTTAACLRASRKSMRAACSR
SAGSRPNRFAPTLMSSCITSTTGPPAWAGDRSHERQLAARQPARVAGAGPQPAGGGQRRP
GLAADPHRLFGAPVLAAAGSLRRAGGRGLHQHFPEWPPRRDRCAARRRDSAHYPGGAGGV
RKPRGALADHRAGVPRRDHPAARCPPGAACAGIGAAHQRGNGEAEAEDRAAPGPHRRPGP
DQPGQGVADAAPWLGRARGAPAPVAGSDEAARADPEDRSGVAGKRAVRLSDPGRPEQ
</pre>
<hr>
-->

<H2>
    Data files
</H2>

None.


<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>


This program always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="backtranseq.html">backtranseq</a></td><td>Back translate a protein sequence</td></tr>
<tr><td><a href="charge.html">charge</a></td><td>Protein charge plot</td></tr>
<tr><td><a href="compseq.html">compseq</a></td><td>Counts the composition of dimer/trimer/etc words in a sequence</td></tr>
<tr><td><a href="emowse.html">emowse</a></td><td>Protein identification by mass spectrometry</td></tr>
<tr><td><a href="freak.html">freak</a></td><td>Residue/base frequency table or plot</td></tr>
<tr><td><a href="iep.html">iep</a></td><td>Calculates the isoelectric point of a protein</td></tr>
<tr><td><a href="mwcontam.html">mwcontam</a></td><td>Shows molwts that match across a set of files</td></tr>
<tr><td><a href="mwfilter.html">mwfilter</a></td><td>Filter noisy molwts from mass spec output</td></tr>
<tr><td><a href="octanol.html">octanol</a></td><td>Displays protein hydropathy</td></tr>
<tr><td><a href="pepinfo.html">pepinfo</a></td><td>Plots simple amino acid properties in parallel</td></tr>
<tr><td><a href="pepstats.html">pepstats</a></td><td>Protein statistics</td></tr>
<tr><td><a href="pepwindow.html">pepwindow</a></td><td>Displays protein hydropathy</td></tr>
<tr><td><a href="pepwindowall.html">pepwindowall</a></td><td>Displays protein hydropathy of a set of sequences</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<ul>

<li>
<a href="transeq.html">transeq</a> : Translate nucleic acid sequences

<li>
<a href="getorf.html">getorf</a> : Finds and extracts open reading frames (ORFs)

<li>
<a href="plotorf.html">plotorf</a> : Plot potential open reading frames

<li>
<a href="showseq.html">showseq</a> : Display a sequence with features, translation etc..

</ul>

<H2>
    Author(s)
</H2>


This application was written by Rodrigo Lopez <a
href="mailto:rls@ebi.ac.uk">(rls@ebi.ac.uk)</a> European
Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SD, UK.
<p>
It was rewritten by Gary Williams<a
href="mailto:gwilliam@hgmp.mrc.ac.uk">(gwilliam@hgmp.mrc.ac.uk)</a>
to output
the sequence data to a single file in the conventional EMBOSS style. 

<H2>
    History
</H2>


Completed 24 Feb 2000 - Rodrigo Lopez

<p>
Rewritten 2 March 2000 - Gary Williams

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

