<HTML>

<HEAD>
  <TITLE>
  EMBOSS: chips
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: chips
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program chips
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Codon usage statistics


<H2>
    Description
</H2>


<b>chips</b> calculates Frank Wright's Nc statistic for the effective number of
codons used (ref 1).

<p>

This is a simple measure that quantifies how far the codon usage of a
gene departs from equal usage of synonymous codons.  This measure of
synonymous codon usage bias, the 'effective number of codons used in a
gene', Nc, can be easily calculated from codon usage data alone, and is
independent of gene length and amino acid (aa) composition.  Nc can take
values from 20, in the case of extreme bias where one codon is
exclusively used for each aa, to 61 when the use of alternative
synonymous codons is equally likely.  Nc thus provides an intuitively
meaningful measure of the extent of codon preference in a gene. 

<p>

The Nc statistic has problems in very short sequences (20 amino acids
or less) which are yet to be fully resolved. They are caused by the
need to consider amino acids which are missing in the sequence.

<p>
This calculation was originally in the EGCG package as "codfish"
(codon usage for fission yeast). As Frank Wright is a vegan, we looked
for a meat-free name for the EMBOSS version, "chips". The official
explanation is "Codon Heterozygosity (Inverse of) in a
Protein-coding Sequence"

<H2>
    Usage
</H2>


Here is a sample session with chips. If the sequence extends beyond
the coding region then the start and/or end positions of the CDS must
be provided because chips analyses exclusively protein coding regions.

<p>

<pre>
% chips -sbeg 135 -send 1292
Input sequence: embl:paamir
Output file [paamir.chips]: 
</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-seqall]            seqall     Sequence database USA
  [-outfile]           outfile    Output file name

   Optional qualifiers: (none)
   Advanced qualifiers:
   -cfile              codon      Codon usage file
   -window             integer    Averaging window


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-seqall]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.chips</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-cfile</td>
<td>Codon usage file</td>
<td>Codon usage file in EMBOSS data path</td>
<td>Ehum.cut</td>
</tr>

<tr>
<td>-window</td>
<td>Averaging window</td>
<td>Any integer value</td>
<td>30</td>
</tr>

</table>



<H2>
    Input file format
</H2>

A nucleic acid sequence USA.

<H2>
    Output file format
</H2>


This is the output from the example run:

<p>
<hr>
<pre>
# CHIPS codon usage statistics

Nc = 32.951
</pre>
<hr>

<p>

If all codons are used, the Nc value will be 61. If only one codon is
used for each amino acid the Nc value will be 20. Low values therefor
indicate a strong codon bias, and high values indicate a low bias and
possibly a non-coding region.

<H2>
    Data files
</H2>


<b>chips</b> reads a codon usage file but only as a template and ignores the
original data.

<p>
The codon usage table is by default the file "CODONS/Ehum.cut"
in the EMBOSS distribution directory.

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by EMBOSS environment variable EMBOSS_DATA.

<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>

<H2>
    Notes
</H2>


None.

<H2>
    References
</H2>


<ol>

<li>
Wright, F. (1990) Gene 87:23-29
"The 'effective number of codons' used in a gene."

</ol>

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.


<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="cai.html">cai</a></td><td>CAI codon adaptation index</td></tr>
<tr><td><a href="chaos.html">chaos</a></td><td>Create a chaos game representation plot for a sequence</td></tr>
<tr><td><a href="codcmp.html">codcmp</a></td><td>Codon usage table comparison</td></tr>
<tr><td><a href="compseq.html">compseq</a></td><td>Counts the composition of dimer/trimer/etc words in a sequence</td></tr>
<tr><td><a href="cusp.html">cusp</a></td><td>Create a codon usage table</td></tr>
<tr><td><a href="freak.html">freak</a></td><td>Residue/base frequency table or plot</td></tr>
<tr><td><a href="geecee.html">geecee</a></td><td>Calculates the fractional GC content of nucleic acid sequences</td></tr>
<tr><td><a href="getorf.html">getorf</a></td><td>Finds and extracts open reading frames (ORFs)</td></tr>
<tr><td><a href="isochore.html">isochore</a></td><td>Plots isochores in large DNA sequences</td></tr>
<tr><td><a href="marscan.html">marscan</a></td><td>Finds MAR/SAR sites in nucleic sequences</td></tr>
<tr><td><a href="newcpgreport.html">newcpgreport</a></td><td>Report CpG rich areas</td></tr>
<tr><td><a href="newcpgseek.html">newcpgseek</a></td><td>Reports CpG rich regions</td></tr>
<tr><td><a href="syco.html">syco</a></td><td>Synonymous codon usage Gribskov statistic plot</td></tr>
<tr><td><a href="wobble.html">wobble</a></td><td>Wobble base plot</td></tr>
<tr><td><a href="wordcount.html">wordcount</a></td><td>Counts words of a specified size in a DNA sequence</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>

1999 - Written - Alan Bleasby.

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

