<HTML>

<HEAD>
  <TITLE>
  EMBOSS: cirdna
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: cirdna
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program cirdna
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Draws circular maps of DNA constructs


<H2>
    Description
</H2>


The program 'cirdna' draws circular maps of DNA constructs.

<p>
It uses the graphical shapes: ticks, ranges, and blocks to
represent genetic markers (e.g, genes and ESTs) and places them
according to their position in a DNA fragment.  The markers can be
organized in different groups. 

<p>
The program reads in one input file in which the user specifies the
names and positions of the genetic markers.  In this file the user also
enters data for controlling the appearance of the markers. 

<p>
A refined customization of the drawing can be achieved by running the
program with '-options' on the command line and changing the values of
the desired parameters. 

<H2>
    Usage
</H2>


Here is a sample session with cirdna.

<pre>

% cirdna
Graph type [x11]: 
Input file [inputfile]: data.cirp
ticks inside or outside the circle (enter In or Out) [Out]: 
text inside or outside the blocks (enter In or Out) [In]: 

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
   -graphout           graph      Graph type
   -inputfile          infile     input file containing mapping data
   -posticks           string     ticks inside or outside the circle (enter In
                                  or Out)
   -posblocks          string     text inside or outside the blocks (enter In
                                  or Out)

   Optional qualifiers:
   -originangle        float      position of the molecule's origin on the
                                  circle (enter a number in the range 0 - 360)
   -intersymbol        string     do you want horizontal junctions between
                                  blocks (Y or N)
   -intercolor         integer    color for junctions between blocks (enter a
                                  color number)
   -interticks         string     do you want horizontal junctions between
                                  ticks (Y or N)
   -gapsize            integer    interval between ticks in the ruler (enter
                                  an integer)
   -ticklines          string     do you want vertical lines at the ruler's
                                  ticks (Y or N)
   -tickheight         float      height of ticks (enter a number to multiply
                                  the default height)
   -blockheight        float      height of blocks (enter a number to multiply
                                  the default height)
   -rangeheight        float      height of range ends (enter a number to
                                  multiply the default height)
   -gapgroup           float      space between groups (enter a number to
                                  multiply the default space)
   -postext            float      space between text and ticks, blocks, and
                                  ranges (enter a number to multiply the
                                  default space)

   Advanced qualifiers: (none)
   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-graphout</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including postscript, ps, hpgl, hp7470, hp7580, meta, colourps, cps, xwindows, x11, tektronics, tekt, tek4107t, tek, none, null, text, data, xterm, png</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr>
<td>-inputfile</td>
<td>input file containing mapping data</td>
<td>Input file</td>
<td>inputfile</td>
</tr>

<tr>
<td>-posticks</td>
<td>ticks inside or outside the circle (enter In or Out)</td>
<td>Any string is accepted</td>
<td>Out</td>
</tr>

<tr>
<td>-posblocks</td>
<td>text inside or outside the blocks (enter In or Out)</td>
<td>Any string is accepted</td>
<td>In</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-originangle</td>
<td>position of the molecule's origin on the circle (enter a number in the range 0 - 360)</td>
<td>Any integer value</td>
<td>90</td>
</tr>

<tr>
<td>-intersymbol</td>
<td>do you want horizontal junctions between blocks (Y or N)</td>
<td>Any string is accepted</td>
<td>Y</td>
</tr>

<tr>
<td>-intercolor</td>
<td>color for junctions between blocks (enter a color number)</td>
<td>Any integer value</td>
<td>1</td>
</tr>

<tr>
<td>-interticks</td>
<td>do you want horizontal junctions between ticks (Y or N)</td>
<td>Any string is accepted</td>
<td>N</td>
</tr>

<tr>
<td>-gapsize</td>
<td>interval between ticks in the ruler (enter an integer)</td>
<td>Any integer value</td>
<td>500</td>
</tr>

<tr>
<td>-ticklines</td>
<td>do you want vertical lines at the ruler's ticks (Y or N)</td>
<td>Any string is accepted</td>
<td>N</td>
</tr>

<tr>
<td>-tickheight</td>
<td>height of ticks (enter a number to multiply the default height)</td>
<td>Any integer value</td>
<td>1</td>
</tr>

<tr>
<td>-blockheight</td>
<td>height of blocks (enter a number to multiply the default height)</td>
<td>Any integer value</td>
<td>1</td>
</tr>

<tr>
<td>-rangeheight</td>
<td>height of range ends (enter a number to multiply the default height)</td>
<td>Any integer value</td>
<td>1</td>
</tr>

<tr>
<td>-gapgroup</td>
<td>space between groups (enter a number to multiply the default space)</td>
<td>Any integer value</td>
<td>1</td>
</tr>

<tr>
<td>-postext</td>
<td>space between text and ticks, blocks, and ranges (enter a number to multiply the default space)</td>
<td>Any integer value</td>
<td>1</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>


Here is the input file data.cirp for the example run: 

<p>
<pre>

----------------------------------------------------------------------
Start   1001
End     4270


group
label
Block    1011   1362    3
ex1
endlabel
label
Tick     1610   8
EcoR1
endlabel
label
Block    1647   1815    1
endlabel
label
Tick     2459   8
BamH1
endlabel
label
Block    4139   4258    3
ex2
endlabel
endgroup


group
label
Range    2541    2812    [       ]       5
Alu
endlabel
label
Range    3322    3497    &gt;       &lt;       5
MER13
endlabel
endgroup
-------------------------------------------------------------------------

</pre>

<p>
Blank lines in the input file are ignored. 

<p>
The file is organized in several fields separated by one or more space
or TAB characters.

<p>
The first non-blank line gives the start position of the DNA fragment to
be drawn after the keyword 'Start'.  The end position is on the second
line following the keyword 'End'. 

<p>
Then, one or more groups (up to a maximum of 100) of genetic markers are
defined.  Each group begins with a line consisting of the keyword
'group' and ends with a line consisting of the keyword 'endgroup'. 

<p>
If desired, a group name can be given on the very next line but
will not be displayed (only for compatibilty with lindna). 

<p>
Next, the individual markers (up to a maximum of 100) of the group are
described.  Each marker begins with a line consisting of the keyword
'label' and ends with a line consisting of the keyword 'endlabel'.

<p>
Do not start a new marker (by using a 'label' line) before you have
closed the previous marker (by using an 'endlabel' line).  You must end
a previous marker specification before starting a new one.  i.e.  do not
overlap marker specifications in the input file. 

<p>
For each marker, the line following the 'label' line describes the
marker. Three types of markers can be drawn: ticks, blocks, and ranges. 

<p>
For a tick, the descriptive line begins with the keyword
'Tick', followed by the position of the tick in the DNA fragment, and a number 
from 0 to 15 specifying the color in which the tick will be drawn (a list of 
available colors is given below). If desired, a name can be written above the 
tick. This name must be given on the next line. If more than one name are given
(one name per line), only the first one will be displayed. 

<p>
For a block, the descriptive line begins with the keyword
'Block', followed by the start and end positions of the block in the DNA 
fragment, and the number of the block's color. If desired, a name can be 
written above or inside the block. This name must be given on the next line. 
If more than one name are given (one name per line), only the first one will be
displayed.

<p>
For a range, the descriptive line begins with the keyword
'Range', followed by the start and end positions of the range in the DNA 
fragment. Next on the line are the types of boundaries that are to be drawn.
The user can choose among '&gt;', '&lt;', '[', and ']' for both start and end 
boundaries. If none of these types is wanted for a given boundary, then a point
'.' must be entered. Next is the color number. If desired, a name can be 
written above the range. This name must be given on the next line. If more than
one name are given (one name per line), only the first one will be displayed. 

<p>
For any type of markers, try to avoid long names.

<p>
The colors defined in the PLPLOT graphics library are:

<p>
0 "BLACK", 1 "RED", 2 "YELLOW", 3 "GREEN", 4 "AQUAMARINE",
5 "PINK", 6 "WHEAT", 7 "GREY", 8 "BROWN", 9 "BLUE", 10 "BLUEVIOLET",
11 "CYAN", 12 "TURQUOISE", 13 "MAGENTA", 14 "SALMON", and 15 "WHITE".

<p>
This format can also be used for the program <a
href="lindna.html">"lindna".</a>


<H2>
    Output file format
</H2>

It produces a graphical image.

<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

None.


<H2>
    References
</H2>

None.


<H2>
    Warnings
</H2>

None.


<H2>
    Diagnostic Error Messages
</H2>

None.


<H2>
    Exit status
</H2>

It always exits with a status of 0.
<H2>
    Known bugs
</H2>

None.


<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="abiview.html">abiview</a></td><td>Reads ABI file and display the trace</td></tr>
<tr><td><a href="lindna.html">lindna</a></td><td>Draws linear maps of DNA constructs</td></tr>
<tr><td><a href="pepnet.html">pepnet</a></td><td>Displays proteins as a helical net</td></tr>
<tr><td><a href="pepwheel.html">pepwheel</a></td><td>Shows protein sequences as helices</td></tr>
<tr><td><a href="prettyplot.html">prettyplot</a></td><td>Displays aligned sequences, with colouring and boxing</td></tr>
<tr><td><a href="prettyseq.html">prettyseq</a></td><td>Output sequence with translated ranges</td></tr>
<tr><td><a href="remap.html">remap</a></td><td>Display a sequence with restriction cut sites, translation etc</td></tr>
<tr><td><a href="seealso.html">seealso</a></td><td>Finds programs sharing group names</td></tr>
<tr><td><a href="showalign.html">showalign</a></td><td>Displays a multiple sequence alignment</td></tr>
<tr><td><a href="showdb.html">showdb</a></td><td>Displays information on the currently available databases</td></tr>
<tr><td><a href="showfeat.html">showfeat</a></td><td>Show features of a sequence</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="textsearch.html">textsearch</a></td><td>Search sequence documentation text. SRS and Entrez are faster!</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Nicolas Tourasse <a
href="mailto:nicolas.tourasse@biotek.uio.no">(nicolas.tourasse@biotek.uio.no)</a>
at <a href="http://www.no.embnet.org/">the Norwegian EMBnet node.</a>


<H2>
    History
</H2>

Written 1999 - Nicolas Tourasse

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 


<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

