<HTML>

<HEAD>
  <TITLE>
  EMBOSS: cirdna
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">




<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

cirdna
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>






<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Draws circular maps of DNA constructs


<H2>
    Description
</H2>


The program 'cirdna' draws circular maps of DNA constructs.

<p>
It uses the graphical shapes: ticks, ranges, and blocks to
represent genetic markers (e.g, genes and ESTs) and places them
according to their position in a DNA fragment.  The markers can be
organized in different groups. A group may contain up to 10000 markers, so the 
   program can be used to draw circular representations of genomic
sequences. 

<p>
The program reads in one input file in which the user specifies the
names and positions of the genetic markers.  In this file the user also
enters data for controlling the appearance of the markers. 

<p>
A refined customization of the drawing can be achieved by running the
program with '-options' on the command line and changing the values of
the desired parameters. 

<H2>
    Usage
</H2>


<b>Here is a sample session with cirdna</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>cirdna -graph cps </b>
Draws circular maps of DNA constructs
input file containing mapping data [inputfile]: <b>../../data/data.cirp</b>
do you want a ruler [Y]: <b></b>
      1 : Open
      2 : Filled
      3 : Outline
type of blocks [Filled]: <b></b>
      1 : In
      2 : Out
ticks inside or outside the circle (enter In or Out) [Out]: <b></b>
      1 : In
      2 : Out
text inside or outside the blocks (enter In or Out) [In]: <b></b>

Created cirdna.ps

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>


<!--
Here is a sample session with cirdna.

<pre>

% cirdna
Draws circular maps of DNA constructs
Graph type [x11]:
Input file [inputfile]: data.cirp
do you want a ruler (Y or N) [Y]:
type of blocks (enter Open, Filled, or Outline) [Filled]:
ticks inside or outside the circle (enter In or Out) [Out]:
text inside or outside the blocks (enter In or Out) [In]:  

</pre>
-->


<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
   -graphout           graph      Graph type
  [-infile]            infile     input file containing mapping data
   -[no]ruler          boolean    do you want a ruler
   -blocktype          selection  type of blocks: Open, Filled, or Outline.
                                  Option 'Outline' draws filled blocks
                                  surrounded by a black border
   -posticks           selection  ticks inside or outside the circle (enter In
                                  or Out)
   -posblocks          selection  text inside or outside the blocks (enter In
                                  or Out)

   Additional (Optional) qualifiers:
   -originangle        float      position of the molecule's origin on the
                                  circle (enter a number in the range 0 - 360)
   -[no]intersymbol    boolean    do you want horizontal junctions between
                                  blocks
   -intercolour        integer    colour of junctions between blocks (enter a
                                  colour number)
   -interticks         boolean    do you want horizontal junctions between
                                  ticks
   -gapsize            integer    interval between ticks in the ruler (enter
                                  an integer)
   -ticklines          boolean    do you want vertical lines at the ruler's
                                  ticks
   -textheight         float      height of text. Enter a number <1 or >1 to
                                  decrease or increase the size, respectively
   -textlength         float      length of text. Enter a number <1 or >1 to
                                  decrease or increase the size, respectively
   -tickheight         float      height of ticks. Enter a number <1 or >1 to
                                  decrease or increase the size, respectively
   -blockheight        float      height of blocks. Enter a number <1 or >1 to
                                  decrease or increase the size, respectively
   -rangeheight        float      height of range ends. Enter a number <1 or
                                  >1 to decrease or increase the size,
                                  respectively
   -gapgroup           float      space between groups. Enter a number <1 or
                                  >1 to decrease or increase the size,
                                  respectively
   -postext            float      space between text and ticks, blocks, and
                                  ranges. Enter a number <1 or >1 to decrease
                                  or increase the size, respectively

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-graphout" associated qualifiers
   -gprompt             boolean    Graph prompting
   -gtitle              string     Graph title
   -gsubtitle           string     Graph subtitle
   -gxtitle             string     Graph x axis title
   -gytitle             string     Graph y axis title
   -goutfile            string     Output file for non interactive displays
   -gdirectory          string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-graphout</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including postscript, ps, hpgl, hp7470, hp7580, meta, colourps, cps, xwindows, x11, tektronics, tekt, tek4107t, tek, none, null, text, data, xterm, png</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr>
<td>[-infile]<br>(Parameter 1)</td>
<td>input file containing mapping data</td>
<td>Input file</td>
<td>inputfile</td>
</tr>

<tr>
<td>-[no]ruler</td>
<td>do you want a ruler</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-blocktype</td>
<td>type of blocks: Open, Filled, or Outline. Option 'Outline' draws filled blocks surrounded by a black border</td>
<td>Choose from selection list of values</td>
<td>Filled</td>
</tr>

<tr>
<td>-posticks</td>
<td>ticks inside or outside the circle (enter In or Out)</td>
<td>Choose from selection list of values</td>
<td>Out</td>
</tr>

<tr>
<td>-posblocks</td>
<td>text inside or outside the blocks (enter In or Out)</td>
<td>Choose from selection list of values</td>
<td>In</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-originangle</td>
<td>position of the molecule's origin on the circle (enter a number in the range 0 - 360)</td>
<td>Number from 0.000 to 360.000</td>
<td>90</td>
</tr>

<tr>
<td>-[no]intersymbol</td>
<td>do you want horizontal junctions between blocks</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-intercolour</td>
<td>colour of junctions between blocks (enter a colour number)</td>
<td>Integer from 0 to 15</td>
<td>1</td>
</tr>

<tr>
<td>-interticks</td>
<td>do you want horizontal junctions between ticks</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-gapsize</td>
<td>interval between ticks in the ruler (enter an integer)</td>
<td>Integer 0 or more</td>
<td>500</td>
</tr>

<tr>
<td>-ticklines</td>
<td>do you want vertical lines at the ruler's ticks</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-textheight</td>
<td>height of text. Enter a number <1 or >1 to decrease or increase the size, respectively</td>
<td>Any numeric value</td>
<td>1</td>
</tr>

<tr>
<td>-textlength</td>
<td>length of text. Enter a number <1 or >1 to decrease or increase the size, respectively</td>
<td>Any numeric value</td>
<td>1</td>
</tr>

<tr>
<td>-tickheight</td>
<td>height of ticks. Enter a number <1 or >1 to decrease or increase the size, respectively</td>
<td>Any numeric value</td>
<td>1</td>
</tr>

<tr>
<td>-blockheight</td>
<td>height of blocks. Enter a number <1 or >1 to decrease or increase the size, respectively</td>
<td>Any numeric value</td>
<td>1</td>
</tr>

<tr>
<td>-rangeheight</td>
<td>height of range ends. Enter a number <1 or >1 to decrease or increase the size, respectively</td>
<td>Any numeric value</td>
<td>1</td>
</tr>

<tr>
<td>-gapgroup</td>
<td>space between groups. Enter a number <1 or >1 to decrease or increase the size, respectively</td>
<td>Any numeric value</td>
<td>1</td>
</tr>

<tr>
<td>-postext</td>
<td>space between text and ticks, blocks, and ranges. Enter a number <1 or >1 to decrease or increase the size, respectively</td>
<td>Any numeric value</td>
<td>1</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: ../../data/data.cirp</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
Start   1001
End     4270


group
label
Block    1011   1362    3
ex1
endlabel
label
Tick     1610   8
EcoR1
endlabel
label
Block    1647   1815    1
endlabel
label
Tick     2459   8
BamH1
endlabel
label
Block    4139   4258    3
ex2
endlabel
endgroup


group
label
Range    2541    2812    [       ]       5
Alu
endlabel
label
Range    3322    3497    &gt;       &lt;       5
MER13
endlabel
endgroup
</pre>
</td></tr></table><p>


<!--
Here is the input file data.cirp for the example run: 

<p>
<pre>

----------------------------------------------------------------------
Start   1001
End     4270


group
label
Block    1011   1362    3
ex1
endlabel
label
Tick     1610   8
EcoR1
endlabel
label
Block    1647   1815    1
endlabel
label
Tick     2459   8
BamH1
endlabel
label
Block    4139   4258    3
ex2
endlabel
endgroup


group
label
Range    2541    2812    [       ]       5
Alu
endlabel
label
Range    3322    3497    &gt;       &lt;       5
MER13
endlabel
endgroup
-------------------------------------------------------------------------

</pre>
-->


<p>

Blank lines in the input file are ignored. 

<p>

The file is organized in several fields separated by one or more space
or TAB characters.

<p>

The first non-blank line gives the start position of the DNA fragment to
be drawn after the keyword 'Start'.  The end position is on the second
line following the keyword 'End'. 

<p>

Then, one or more groups (up to a maximum of 20) of genetic markers are
defined.  Each group begins with a line consisting of the keyword
'group' and ends with a line consisting of the keyword 'endgroup'. 

<p>

If desired, a group name can be given on the very next line but
will not be displayed (only for compatibilty with lindna). 

<p>

Next, the individual markers (up to a maximum of 1000) of the group are
described.  Each marker begins with a line consisting of the keyword
'label' and ends with a line consisting of the keyword 'endlabel'.

<p>

Do not start a new marker (by using a 'label' line) before you have
closed the previous marker (by using an 'endlabel' line).  You must end
a previous marker specification before starting a new one.  i.e.  do not
overlap marker specifications in the input file. 

<p>

For each marker, the line following the 'label' line describes the
marker. Three types of markers can be drawn: ticks, blocks, and ranges. 

<p>

For a tick, the descriptive line begins with the keyword 'Tick',
followed by the position of the tick in the DNA fragment, and a number
from 0 to 15 specifying the color in which the tick will be drawn (a
list of available colors is given below).  If desired, a name can be
written above the tick.  This name must be given on the next line.  If
more than one name are given (one name per line), only the first one
will be displayed. 

<p>

For a block, the descriptive line begins with the keyword 'Block',
followed by the start and end positions of the block in the DNA
fragment, and the number of the block's color.  If desired, a name can
be written above or inside the block.  This name must be given on the
next line.  If more than one name are given (one name per line), only
the first one will be displayed. 

<p>

For a range, the descriptive line begins with the keyword 'Range',
followed by the start and end positions of the range in the DNA
fragment.  Next on the line are the types of boundaries that are to be
drawn.  The user can choose among '&gt;', '&lt;', '[', ']' and '|' for both
start and end boundaries.  If none of these types is wanted for a given
boundary, then a point '.' must be entered.  Next is the color number. 
If desired, a name can be written above the range.  This name must be
given on the next line.  If more than one name are given (one name per
line), only the first one will be displayed. 

<p>

For any type of markers, names will be written in the same color as the
marker.  Also, try to avoid long names. 

<p>

The colors defined in the PLPLOT graphics library are:

<p>

0 "BLACK", 1 "RED", 2 "YELLOW", 3 "GREEN", 4 "AQUAMARINE",
5 "PINK", 6 "WHEAT", 7 "GREY", 8 "BROWN", 9 "BLUE", 10 "BLUEVIOLET",
11 "CYAN", 12 "TURQUOISE", 13 "MAGENTA", 14 "SALMON", and 15 "WHITE".

<p>

This format can also be used for the program <a
href="lindna.html">"lindna".</a>


<H2>
    Output file format
</H2>

Output is to the specified graphics device.
   
<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>Graphics File: cirdna.ps</h3>
<p><img src="cirdna.1.cirdna.gif" alt="[cirdna results]">



<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

None.


<H2>
    References
</H2>

None.


<H2>
    Warnings
</H2>

None.


<H2>
    Diagnostic Error Messages
</H2>

None.


<H2>
    Exit status
</H2>

It always exits with a status of 0.
<H2>
    Known bugs
</H2>

None.


<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="abiview.html">abiview</a></td><td>Reads ABI file and display the trace</td></tr>
<tr><td><a href="lindna.html">lindna</a></td><td>Draws linear maps of DNA constructs</td></tr>
<tr><td><a href="pepnet.html">pepnet</a></td><td>Displays proteins as a helical net</td></tr>
<tr><td><a href="pepwheel.html">pepwheel</a></td><td>Shows protein sequences as helices</td></tr>
<tr><td><a href="prettyplot.html">prettyplot</a></td><td>Displays aligned sequences, with colouring and boxing</td></tr>
<tr><td><a href="prettyseq.html">prettyseq</a></td><td>Output sequence with translated ranges</td></tr>
<tr><td><a href="remap.html">remap</a></td><td>Display a sequence with restriction cut sites, translation etc</td></tr>
<tr><td><a href="seealso.html">seealso</a></td><td>Finds programs sharing group names</td></tr>
<tr><td><a href="showalign.html">showalign</a></td><td>Displays a multiple sequence alignment</td></tr>
<tr><td><a href="showdb.html">showdb</a></td><td>Displays information on the currently available databases</td></tr>
<tr><td><a href="showfeat.html">showfeat</a></td><td>Show features of a sequence</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="sixpack.html">sixpack</a></td><td>Display a DNA sequence with 6-frame translation and ORFs</td></tr>
<tr><td><a href="textsearch.html">textsearch</a></td><td>Search sequence documentation text. SRS and Entrez are faster!</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>

Nicolas Tourasse (nicolas.tourasse&nbsp;&copy;&nbsp;biotek.uio.no)
<br>
Biotechnology Center of Oslo



<H2>
    History
</H2>

Written 1999 - Nicolas Tourasse
<br>
Updated (2002) - Nicolas Tourasse

<!--
<p>
[an error occurred while processing this directive]
-->

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 


<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

