<HTML>

<HEAD>
  <TITLE>
  EMBOSS: codcmp
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">




<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

codcmp
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>







<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Codon usage table comparison


<H2>
    Description
</H2>


This program reads in two codon usage table files.

<p>

It counts the number of the 64 possible codons which are unused (i.e. 
has a usage fraction of 0) in either one or the other or both of the
codon usage tables.  

<p>

The usage fraction of a codon is its proportion (0 to 1) of the total of
the codons in the sequences used to construct the usage table. 

<p>

For each codon that is used in both tables, it takes the difference
between the usage fraction.  The sum of the differences and the sum of
the differences squared is reported in the output file, together with
the number of unused codons. 


<H3>
   Statistical significance
</H3>

<b>Question:</b>

<p>

How do you interpret the statistical significance of any difference
between the tables?

<p>

<b>Answer:</b>

<p>

This is a very interesting question.  I don't think that there is any
way to say if it is statistically significant just from looking at it,
as it is essentially a descriptive statistic about the difference
between two 64-mer vectors.  If you have a whole lot of sequences and
codcmp results for all the possible pairwise comparisons, then the
resulting distance matrix can be used to build a phylogenetic tree based
on codon usage. 

<p>

However, if you generate a series of random sequences, measure their codon 
usage and then do codcmp between each of your test sequences and all the 
random sequences, you could then use a z-test to see if the result between 
the two test sequences was outside of the top or bottom 5%.

<p>

This would assume that the codcmp results were normally distributed, but
you could test that too.  The simplest way is just to plot them and look 
for a bell-curve.  For more rigour, find the mean and standard deviation of 
your results from the random sequences, use the normal distribution 
equation to generate a theoretical distribution for that mean and standard 
deviation, and then perform a chi square between the random data and the 
theoretically generated normal distribution.  If you generate two sets of 
random data, each based on your two test sequences, an F-test should be 
used to establish that they have equal variances.  Then you can safely go 
ahead and perform the z-test.

<p>

You could use shuffle to base your random sequences on the test
sequences - so that would ensure the randomised background had the same
nucleotide content. 

<p>

F-tests, z-tests and chi-tests can all be done in Excel, as well as
being standard in most statistical analysis packages. 

<p>

<i>Answered by Derek Gatherer &lt;d.gatherer&nbsp;&copy;&nbsp;vir.gla.ac.uk&gt; 21 Nov 2003</i>


<H2>
    Usage
</H2>


<b>Here is a sample session with codcmp</b>
<p>
This compares the codon usage tables for Escherichia coli and Haemophilus influenzae. 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>codcmp </b>
Codon usage table comparison
Codon usage file [Ehum.cut]: <b>Eeco.cut</b>
Codon usage file [Ehum.cut]: <b>Ehin.cut</b>
Output file [outfile.codcmp]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.1">Go to the output files for this example</a><p><p>


<!--
Here is a sample session with codcmp, comparing the codon usage tables
for Escherichia coli and Haemophilus influenzae.
<P>

<pre>
% codcmp
Codon usage file [Ehum.cut]: Eeco.cut
Codon usage file [Ehum.cut]: Ehin.cut
Output file [outfile.codcmp]: 
</pre>
-->


<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-first]             codon      First codon usage file
  [-second]            codon      Second codon usage file for comparison
  [-outfile]           outfile    Output file name

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-outfile" associated qualifiers
   -odirectory3         string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-first]<br>(Parameter 1)</td>
<td>First codon usage file</td>
<td>Codon usage file in EMBOSS data path</td>
<td>Ehum.cut</td>
</tr>

<tr>
<td>[-second]<br>(Parameter 2)</td>
<td>Second codon usage file for comparison</td>
<td>Codon usage file in EMBOSS data path</td>
<td>Ehum.cut</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 3)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.codcmp</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>

It reads in the Codon Usage Tables - these are available as EMBOSS data files.
See below for details.

<p>



<H2>
    Output file format
</H2>



<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: outfile.codcmp</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
# CODCMP codon usage table comparison
# Eeco.cut vs Ehin.cut

Sum Squared Difference = 2.337
Mean Squared Difference = 0.037
Root Mean Squared Difference = 0.191
Sum Difference         = 9.840
Mean Difference         = 0.154
Codons not appearing   = 0
</pre>
</td></tr></table><p>


<!--
This is the result of the example run:
<P>

<HR>
<pre>
# CODCMP codon usage table comparison
# Eeco.cut vs Ehin.cut

Sum Squared Difference = 2.337
Mean Squared Difference = 0.037
Root Mean Squared Difference = 0.191
Sum Difference         = 9.840
Mean Difference         = 0.154
Codons not appearing   = 0
</pre>
<HR>
-->


<H2>
    Data files
</H2>


The codon usage tables are read by default from "Ehum.cut" in the data/CODONS
directory of the EMBOSS distribution.

<p>
If the name of a codon usage file is specified on the command line,
then this file will first be searched for in the current directory and
then in the 'data/CODONS' directory of the EMBOSS distribution.

<p>
To see the available EMBOSS codon usage files, run:
<pre>

% embossdata -showall

</pre>

<p>
To fetch one of the codon usage tables (for example 'Emus.cut') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Emus.cut

</pre>

<H2>
    Notes
</H2>


None.

<H2>
    References
</H2>


None.

<H2>
    Warnings
</H2>


None.

<H2>
    Diagnostic Error Messages
</H2>


None.

<H2>
    Exit status
</H2>


This program always exits with a status of 0.
 
<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="cai.html">cai</a></td><td>CAI codon adaptation index</td></tr>
<tr><td><a href="chips.html">chips</a></td><td>Codon usage statistics</td></tr>
<tr><td><a href="cusp.html">cusp</a></td><td>Create a codon usage table</td></tr>
<tr><td><a href="syco.html">syco</a></td><td>Synonymous codon usage Gribskov statistic plot</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->



<H2>
    Author(s)
</H2>


This application was written by
Alan Bleasby (ableasby&nbsp;&copy;&nbsp;hgmp.mrc.ac.uk)
<br>
HGMP-RC, Genome Campus, Hinxton, Cambridge CB10 1SB, UK


<p>

Some more statistics were added by 
David Martin (dmartin&nbsp;&copy;&nbsp;hgmp.mrc.ac.uk)
<br>



<H2>
    History
</H2>


Completed 9 Sept 1999
<br>
20 Oct 2000 - David Martin added a couple more statistics to the output.

<!--
<p>
[an error occurred while processing this directive]
-->

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

