<HTML>
<HEAD>
  <TITLE>
  EMBOSS: coderet
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

coderet
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<H2>
    Function
</H2>
Extract CDS, mRNA and translations from feature tables

<H2>
    Description
</H2>

The feature table of sequence database entries often have sections like this:
<p>

This specifies that the coding sequence for the gene is constructed by
joining several sections of code, many of which are in other entries in
this database. 

<p>
<hr>
<pre>
FT   CDS             join(U21925.1:818..987,U21926.1:258..420,
FT                   U21927.1:428..520,U21928.1:196..336,U21929.1:279..415,
FT                   U21930.1:895..1014,516..708)
</pre>
<hr>
<p>

or:
<p>

This specifies that the messenger RNA sequence for the gene is constructed by 
joining several sections of code, many of which are in other entries in
this database.


<p>
<hr>
<pre>
FT   mRNA            join(M88628.1:1006..1318,M88629.1:221..342,
FT                   M88630.1:101..223,M88631.1:46..258,M88632.1:104..172,
FT                   M88633.1:387..503,M88634.1:51..272,M88635.1:303..564,
FT                   M88635.1:849..1020,M88636.1:282..375,M88637.1:39..253,
FT                   M88638.1:91..241,M88639.1:168..377,M88640.1:627..3732,
FT                   M88641.1:158..311,M88642.1:1051..1263,M88642.1:1550..1778,
FT                   M88642.1:1986..2168,M88642.1:3904..4020,
FT                   M88642.1:4627..4698,M88643.1:39..124,M88644.1:42..197,
FT                   M88645.1:542..686,M88646.1:75..223,M88647.1:109..285,
FT                   253..2211)
</pre>
<hr>
<p>


or:
<p>

This specifies that the translation of the coding region is as follows.

<p>
<hr>
<pre>
FT                   /translation="MAQDSVDLSCDYQFWMQKLSVWDQASTLETQQDTCLHVAQFQEFL
FT                   RKMYEALKEMDSNTVIERFPTIGQLLAKACWNPFILAYDESQKILIWCLCCLINKEPQN
FT                   SGQSKLNSWIQGVLSHILSALRFDKEVALFTQGLGYAPIDYYPGLLKNMVLSLASELRE
FT                   NHLNGFNTQRRMAPERVASLSRVCVPLITLTDVDPLVEALLICHGREPQEILQPEFFEA
FT                   VNEAILLKKISLPMSAVVCLWLRHLPSLEKAMLHLFEKLISSERNCLRRIECFIKDSSL
FT                   PQAACHPAIFRVDEMFRCALLETDGALEIIATIQVFTQCFVEALEKASKQLRFALKTYF
FT                   PYTSPSLAMVLLQDPQDIPRGHWLQTLKHISELLREAVEDQTHGSCGGPFESWFLFIHF
FT                   GGWAEMVAEQLLMSAAEPPTALLWLLAFYYGPRDGRQQRAQTMVQVKAVLGHLLAMSRS
FT                   SSLSAQDLQTVAGQGTDTDLRAPAQQLIRHLLLNFLLWAPGGHTIAWDVITLMAHTAEI
FT                   THEIIGFLDQTLYRWNRLGIESPRSEKLARELLKELRTQV"
</pre>
<hr>
<p>

It is often a tedious and error-prone job to extract the several
sections of these sequences from the database entries and to join them together
to construct the indicated finished sequence. 
<p>

<b>coderet</b> does this job for you.
<p>

You specify the sequence containing the feature table and it extracts
the required CDS, mRNA and/or protein sequences specified by this
feature table.  If any sequences are in other entries of that database,
they are automatically fetched and incorporated correctly into the final
sequence. 

<p>

The translations are not made from the coding sequence, they are extracted
directly from the translation sequence held in the feature table. 



<H2>
    Usage
</H2>

<b>Here is a sample session with coderet</b>
<p>
To extract all of the CDS, mRNA and the protein translations: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>coderet </b>
Extract CDS, mRNA and translations from feature tables
Input sequence(s): <b>tembl:X03487</b>
Output sequence [hsferg1.fasta]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>
<p>
<b>Example 2</b>
<p>
To only extract the mRNA sequence: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>coderet -nocds -notranslation </b>
Extract CDS, mRNA and translations from feature tables
Input sequence(s): <b>tembl:X03487</b>
Output sequence [hsferg1.fasta]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.2">Go to the output files for this example</a><p><p>


<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-seqall]            seqall     Sequence database USA
  [-outseq]            seqout     Output sequence USA

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers:
   -[no]cds            boolean    Extract CDS sequences
   -[no]mrna           boolean    Extract mrna sequences
   -[no]translation    boolean    Extract translated sequences

   Associated qualifiers:

   "-seqall" associated qualifiers
   -sbegin1             integer    Start of each sequence to be used
   -send1               integer    End of each sequence to be used
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name

   "-outseq" associated qualifiers
   -osformat2           string     Output seq format
   -osextension2        string     File name extension
   -osname2             string     Base file name
   -osdirectory2        string     Output directory
   -osdbname2           string     Database name to add
   -ossingle2           boolean    Separate file for each entry
   -oufo2               string     UFO features
   -offormat2           string     Features format
   -ofname2             string     Features file name
   -ofdirectory2        string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-seqall]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outseq]<br>(Parameter 2)</td>
<td>Output sequence USA</td>
<td>Writeable sequence</td>
<td><i>&lt;sequence&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]cds</td>
<td>Extract CDS sequences</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]mrna</td>
<td>Extract mrna sequences</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]translation</td>
<td>Extract translated sequences</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

</table>



<H2>
    Input file format
</H2>

<b>coderet</b> reads one or more nucleic sequence USAs having CDS, mRNA
or translation headings in their feature tables. 

<p>

<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:X03487' is a sequence entry in the example nucleic acid database 'tembl'
<p>
<p><h3>Database entry: tembl:X03487</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   HSFERG1    standard; DNA; HUM; 512 BP.
XX
AC   X03487;
XX
SV   X03487.1
XX
DT   02-JUL-1986 (Rel. 09, Created)
DT   04-APR-1995 (Rel. 43, Last updated, Version 2)
XX
DE   Human apoferritin H gene exon 1
XX
KW   ferritin.
XX
OS   Homo sapiens (human)
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia;
OC   Eutheria; Primates; Catarrhini; Hominidae; Homo.
XX
RN   [1]
RP   1-512
RX   MEDLINE; 86120367.
RA   Costanzo F., Colombo M., Staempfli S., Santoro C., Marone M., Frank R.,
RA   Delius H., Cortese R.;
RT   "Structure of gene and pseudogenes of human apoferritin H";
RL   Nucleic Acids Res. 14:721-736(1986).
XX
DR   EPD; EP11112; HS_FRIH.
DR   SWISS-PROT; P02794; FRIH_HUMAN.
XX
FH   Key             Location/Qualifiers
FH
FT   source          1..512
FT                   /db_xref="taxon:9606"
FT                   /organism="Homo sapiens"
FT   misc_feature    65..70
FT                   /note="GGGCGG box"
FT   misc_feature    103..108
FT                   /note="GGGCGG box"
FT   misc_feature    126..131
FT                   /note="GGGCGG box"
FT   promoter        150..154
FT                   /note="put. TATA box"
FT   mRNA            179..500
FT                   /note="exon 1"
FT   CDS             join(387..500,X03488.1:50..196,X03488.1:453..578,
FT                   X03488.1:674..838)
FT                   /db_xref="SWISS-PROT:P02794"
FT                   /label=apoh_cds
FT                   /product="apoferritin H subunit"
FT                   /protein_id="CAA27205.1"
FT                   /translation="MTTASTSQVRQNYHQDSEAAINRQINLELYASYVYLSMSYYFDRD
FT                   DVALKNFAKYFLHQSHEEREHAEKLMKLQNQRGGRIFLQDIKKPDCDDWESGLNAMECA
FT                   LHLEKNVNQSLLELHKLATDKNDPHLCDFIETHYLNEQVKAIKELGDHVTNLRKMGAPE
FT                   SGLAEYLFDKHTLGDSDNES"
FT   intron          501..&gt;512
FT                   /note="intron I"
XX
SQ   Sequence 512 BP; 80 A; 212 C; 152 G; 64 T; 4 other;
     agnncaaacc tnagctccgc cagagcgcgc gaggcctcca gcggccgccc ctcccccaca        60
     gcaggggcgg ggntcccgcg cccaccggaa ggagcgggct cggggcgggc ggcgctgatt       120
     ggccggggcg ggcctgacgc cgacgcggct ataagagacc acaagcgacc cgcagggcca       180
     gacgttcttc gccgagagtc gtcggggttt cctgcttcaa cagtgcttgg acggaacccg       240
     gcgctcgttc cccaccccgg ccggccgccc atagccagcc ctccgtcgac ctcttcaccg       300
     caccctcgga ctgccccaag gcccccgccg ccgctccagc gccgcgcagc caccgccgcc       360
     gccgccgcct ctccttagtc gccgccatga cgaccgcgtc cacctcgcag gtgcgccaga       420
     actaccacca ggactcagag gccgccatca accgccagat caacctggag ctctacgcct       480
     cctacgttta cctgtccatg gtgagcgcgg gc                                     512
//
</pre>
</td></tr></table><p>


<H2>
    Output file format
</H2>

The output is a sequence file containing any CDS, mRNA and protein
translation sequences as specified by the feature table of the sequence(s).

<p>

One or more of CDS, mRNA, translation can be excluded from the output by
using the appropriate qualifiers to the program (i.e.  <b>-nocds</b>,
etc.)

<p>

The ID names of the output sequences are constructed from the name of
the input sequence, the type of feature being output (i.e.  <b>cds</b>,
<b>mrna</b>, <b>pro</b>) and a unique ordinal number for this type to
distinguish it from others in this sequence.  The name, type and number
of separated by underscore characters.  Thus the second CDS feature in
the sequence 'HSXYZ' would be named 'HSXYZ_cds_2'. 

<p>

The translations are not made from the coding sequence, they are extracted
directly from the translation sequence held in the feature table. 

<p>

<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: hsferg1.fasta</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt;x03487_cds_1
atgacgaccgcgtccacctcgcaggtgcgccagaactaccaccaggactcagaggccgcc
atcaaccgccagatcaacctggagctctacgcctcctacgtttacctgtccatgtcttac
tactttgaccgcgatgatgtggctttgaagaactttgccaaatactttcttcaccaatct
catgaggagagggaacatgctgagaaactgatgaagctgcagaaccaacgaggtggccga
atcttccttcaggatatcaagaaaccagactgtgatgactgggagagcgggctgaatgca
atggagtgtgcattacatttggaaaaaaatgtgaatcagtcactactggaactgcacaaa
ctggccactgacaaaaatgacccccatttgtgtgacttcattgagacacattacctgaat
gagcaggtgaaagccatcaaagaattgggtgaccacgtgaccaacttgcgcaagatggga
gcgcccgaatctggcttggcggaatatctctttgacaagcacaccctgggagacagtgat
aatgaaagctaa
&gt;x03487_mrna_1
cagacgttcttcgccgagagtcgtcggggtttcctgcttcaacagtgcttggacggaacc
cggcgctcgttccccaccccggccggccgcccatagccagccctccgtcgacctcttcac
cgcaccctcggactgccccaaggcccccgccgccgctccagcgccgcgcagccaccgccg
ccgccgccgcctctccttagtcgccgccatgacgaccgcgtccacctcgcaggtgcgcca
gaactaccaccaggactcagaggccgccatcaaccgccagatcaacctggagctctacgc
ctcctacgtttacctgtccatg
&gt;x03487_pro_1
MTTASTSQVRQNYHQDSEAAINRQINLELYASYVYLSMSYYFDRDDVALKNFAKYFLHQS
HEEREHAEKLMKLQNQRGGRIFLQDIKKPDCDDWESGLNAMECALHLEKNVNQSLLELHK
LATDKNDPHLCDFIETHYLNEQVKAIKELGDHVTNLRKMGAPESGLAEYLFDKHTLGDSD
NES
</pre>
</td></tr></table><p>

<a name="output.2"></a>
<h3>Output files for usage example 2</h3>
<p><h3>File: hsferg1.fasta</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt;x03487_mrna_1
cagacgttcttcgccgagagtcgtcggggtttcctgcttcaacagtgcttggacggaacc
cggcgctcgttccccaccccggccggccgcccatagccagccctccgtcgacctcttcac
cgcaccctcggactgccccaaggcccccgccgccgctccagcgccgcgcagccaccgccg
ccgccgccgcctctccttagtcgccgccatgacgaccgcgtccacctcgcaggtgcgcca
gaactaccaccaggactcagaggccgccatcaaccgccagatcaacctggagctctacgc
ctcctacgtttacctgtccatg
</pre>
</td></tr></table><p>


<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

The translations are not made from the coding sequence, they are extracted
directly from the translation sequence held in the feature table. 

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.


<H2>
    Exit status
</H2>

It always exits with status 0.

<H2>
    Known bugs
</H2>

None.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="backtranseq.html">backtranseq</a></td><td>Back translate a protein sequence</td></tr>
<tr><td><a href="extractfeat.html">extractfeat</a></td><td>Extract features from a sequence</td></tr>
<tr><td><a href="maskfeat.html">maskfeat</a></td><td>Mask off features of a sequence</td></tr>
<tr><td><a href="plotorf.html">plotorf</a></td><td>Plot potential open reading frames</td></tr>
<tr><td><a href="prettyseq.html">prettyseq</a></td><td>Output sequence with translated ranges</td></tr>
<tr><td><a href="remap.html">remap</a></td><td>Display a sequence with restriction cut sites, translation etc</td></tr>
<tr><td><a href="showfeat.html">showfeat</a></td><td>Show features of a sequence</td></tr>
<tr><td><a href="showorf.html">showorf</a></td><td>Pretty output of DNA translations</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="sixpack.html">sixpack</a></td><td>Display a DNA sequence with 6-frame translation and ORFs</td></tr>
<tr><td><a href="transeq.html">transeq</a></td><td>Translate nucleic acid sequences</td></tr>
<tr><td><a href="twofeat.html">twofeat</a></td><td>Finds neighbouring pairs of features in sequences</td></tr>
</table>


<H2>
    Author(s)
</H2>
Alan Bleasby (ableasby&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK




<H2>
    History
</H2>

Written (Nov 2000) - Alan Bleasby.


<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.


<H2>
    Comments
</H2>
None


</BODY>
</HTML>

