<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: coderet
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">
<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>
<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>
</td>
<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: coderet
</H2>
</font>
</td>
</tr>
</table>
</td>
</tr>
</table>
<p><hr><p>

<!--END OF HEADER-->

<H1>
Program coderet
</H1>
<H2>
    Function
</H2>
Extract CDS, mRNA and translations from feature tables

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
        Algorithms
  -->

The feature table of sequence database entries often have sections like this:
<p>

This specifies that the coding sequence for the gene is constructed by
joining several sections of code, many of which are in other entries in
this database. 

<p>
<hr>
<pre>
FT   CDS             join(U21925.1:818..987,U21926.1:258..420,
FT                   U21927.1:428..520,U21928.1:196..336,U21929.1:279..415,
FT                   U21930.1:895..1014,516..708)
</pre>
<hr>
<p>

or:
<p>

This specifies that the messenger RNA sequence for the gene is constructed by 
joining several sections of code, many of which are in other entries in
this database.


<p>
<hr>
<pre>
FT   mRNA            join(M88628.1:1006..1318,M88629.1:221..342,
FT                   M88630.1:101..223,M88631.1:46..258,M88632.1:104..172,
FT                   M88633.1:387..503,M88634.1:51..272,M88635.1:303..564,
FT                   M88635.1:849..1020,M88636.1:282..375,M88637.1:39..253,
FT                   M88638.1:91..241,M88639.1:168..377,M88640.1:627..3732,
FT                   M88641.1:158..311,M88642.1:1051..1263,M88642.1:1550..1778,
FT                   M88642.1:1986..2168,M88642.1:3904..4020,
FT                   M88642.1:4627..4698,M88643.1:39..124,M88644.1:42..197,
FT                   M88645.1:542..686,M88646.1:75..223,M88647.1:109..285,
FT                   253..2211)
</pre>
<hr>
<p>


or:
<p>

This specifies that the translation of the coding region is as follows.

<p>
<hr>
<pre>
FT                   /translation="MAQDSVDLSCDYQFWMQKLSVWDQASTLETQQDTCLHVAQFQEFL
FT                   RKMYEALKEMDSNTVIERFPTIGQLLAKACWNPFILAYDESQKILIWCLCCLINKEPQN
FT                   SGQSKLNSWIQGVLSHILSALRFDKEVALFTQGLGYAPIDYYPGLLKNMVLSLASELRE
FT                   NHLNGFNTQRRMAPERVASLSRVCVPLITLTDVDPLVEALLICHGREPQEILQPEFFEA
FT                   VNEAILLKKISLPMSAVVCLWLRHLPSLEKAMLHLFEKLISSERNCLRRIECFIKDSSL
FT                   PQAACHPAIFRVDEMFRCALLETDGALEIIATIQVFTQCFVEALEKASKQLRFALKTYF
FT                   PYTSPSLAMVLLQDPQDIPRGHWLQTLKHISELLREAVEDQTHGSCGGPFESWFLFIHF
FT                   GGWAEMVAEQLLMSAAEPPTALLWLLAFYYGPRDGRQQRAQTMVQVKAVLGHLLAMSRS
FT                   SSLSAQDLQTVAGQGTDTDLRAPAQQLIRHLLLNFLLWAPGGHTIAWDVITLMAHTAEI
FT                   THEIIGFLDQTLYRWNRLGIESPRSEKLARELLKELRTQV"
</pre>
<hr>
<p>

It is often a tedious and error-prone job to extract the several
sections of these sequences from the database entries and to join them together
to construct the indicated finished sequence. 
<p>

<b>coderet</b> does this job for you.
<p>

You specify the sequence containing the feature table and it extracts
the required CDS, mRNA and/or protein sequences specified by this
feature table.  If any sequences are in other entries of that database,
they are automatically fetched and incorporated correctly into the final
sequence. 

<p>

The translations are not made from the coding sequence, they are extracted
directly from the translation sequence held in the feature table. 



<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->


Here is a sample session with coderet:
<p>

The EMBL sequence X03487's feature table is as follows:
<p>
<hr>
<pre>
FH   Key             Location/Qualifiers
FH
FT   source          1..512
FT                   /db_xref="taxon:9606"
FT                   /organism="Homo sapiens"
FT   misc_feature    65..70
FT                   /note="GGGCGG box"
FT   misc_feature    103..108
FT                   /note="GGGCGG box"
FT   misc_feature    126..131
FT                   /note="GGGCGG box"
FT   promoter        150..154
FT                   /note="put. TATA box"
FT   mRNA            179..500
FT                   /note="exon 1"
FT   CDS             join(387..500,X03488.1:50..196,X03488.1:453..578,
FT                   X03488.1:674..838)
FT                   /db_xref="SWISS-PROT:P02794"
FT                   /label=apoh_cds
FT                   /product="apoferritin H subunit"
FT                   /protein_id="CAA27205.1"
FT                   /translation="MTTASTSQVRQNYHQDSEAAINRQINLELYASYVYLSMSYYFDRD
FT                   DVALKNFAKYFLHQSHEEREHAEKLMKLQNQRGGRIFLQDIKKPDCDDWESGLNAMECA
FT                   LHLEKNVNQSLLELHKLATDKNDPHLCDFIETHYLNEQVKAIKELGDHVTNLRKMGAPE
FT                   SGLAEYLFDKHTLGDSDNES"
FT   intron          501..>512
FT                   /note="intron I"
</pre>
<hr>
<p>

This means that the mRNA sequence is from position 179 to 500 in this sequence.
<br>
The coding sequence is formed by joining the region from 387 to 500 in this sequence together with regions from two other entries in this database: X03488 and X03488.
<br>
The translation of the coding sequence is also given.
<p>

To extract all of the CDS, mRNA and the protein translations, type:
<p>
<pre>
% coderet
Extract CDS, mRNA and translations from feature tables
Input sequence(s): embl:X03487
Output sequence [hsferg1.fasta]: 
</pre>
<p>

The result is as follows:
<p>
<hr>
<pre>
>x03487_cds_1
atgacgaccgcgtccacctcgcaggtgcgccagaactaccaccaggactcagaggccgcc
atcaaccgccagatcaacctggagctctacgcctcctacgtttacctgtccatgtcttac
tactttgaccgcgatgatgtggctttgaagaactttgccaaatactttcttcaccaatct
catgaggagagggaacatgctgagaaactgatgaagctgcagaaccaacgaggtggccga
atcttccttcaggatatcaagaaaccagactgtgatgactgggagagcgggctgaatgca
atggagtgtgcattacatttggaaaaaaatgtgaatcagtcactactggaactgcacaaa
ctggccactgacaaaaatgacccccatttgtgtgacttcattgagacacattacctgaat
gagcaggtgaaagccatcaaagaattgggtgaccacgtgaccaacttgcgcaagatggga
gcgcccgaatctggcttggcggaatatctctttgacaagcacaccctgggagacagtgat
aatgaaagctaa
>x03487_mrna_1
cagacgttcttcgccgagagtcgtcggggtttcctgcttcaacagtgcttggacggaacc
cggcgctcgttccccaccccggccggccgcccatagccagccctccgtcgacctcttcac
cgcaccctcggactgccccaaggcccccgccgccgctccagcgccgcgcagccaccgccg
ccgccgccgcctctccttagtcgccgccatgacgaccgcgtccacctcgcaggtgcgcca
gaactaccaccaggactcagaggccgccatcaaccgccagatcaacctggagctctacgc
ctcctacgtttacctgtccatg
>x03487_pro_1
MTTASTSQVRQNYHQDSEAAINRQINLELYASYVYLSMSYYFDRDDVALKNFAKYFLHQS
HEEREHAEKLMKLQNQRGGRIFLQDIKKPDCDDWESGLNAMECALHLEKNVNQSLLELHK
LATDKNDPHLCDFIETHYLNEQVKAIKELGDHVTNLRKMGAPESGLAEYLFDKHTLGDSD
NES
</pre>
<hr>  
<p> 

Note that there is a mixture of nucleic and protein  sequences in this file
<p>

To only extract the mRNA sequence, type:
<p>


<pre>
% coderet -nocds -notranslation
Extract CDS, mRNA and translations from feature tables
Input sequence(s): embl:X03487
Output sequence [hsferg1.fasta]: 
</pre>
<p>







<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-seqall]            seqall     Sequence database USA
  [-seqout]            seqout     Output sequence USA

   Optional qualifiers: (none)
   Advanced qualifiers:
   -[no]cds            bool       Extract CDS sequences
   -[no]mrna           bool       Extract mrna sequences
   -[no]translation    bool       Extract translated sequences


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-seqall]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-seqout]<br>(Parameter 2)</td>
<td>Output sequence USA</td>
<td>Writeable sequence</td>
<td><i>&lt;sequence&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]cds</td>
<td>Extract CDS sequences</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]mrna</td>
<td>Extract mrna sequences</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]translation</td>
<td>Extract translated sequences</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

USA of one or more nucleic sequences having CDS, mRNA or translation
headings in their feature tables. 












<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->

The output is a sequence file containing any CDS, mRNA and protein
translation sequences as specified by the feature table of the sequence(s).
<p>

One or more of CDS, mRNA, translation can be excluded from the output by
using the appropriate qualifiers to the program (i.e.  <b>-nocds</b>,
etc.)
<p>

The ID names of the output sequences are constructed from the name of
the input sequence, the type of feature being output (i.e.  <b>cds</b>,
<b>mrna</b>, <b>pro</b>) and a unique ordinal number for this type to
distinguish it from others in this sequence.  The name, type and number
of separated by underscore characters.  Thus the second CDS feature in
the sequence 'HSXYZ' would be named 'HSXYZ_cds_2'. 
<p>

The translations are not made from the coding sequence, they are extracted
directly from the translation sequence held in the feature table. 







<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.
	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

   -->

None.













<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

The translations are not made from the coding sequence, they are extracted
directly from the translation sequence held in the feature table. 








<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="backtranseq.html">backtranseq</a></td><td>Back translate a protein sequence</td></tr>
<tr><td><a href="maskfeat.html">maskfeat</a></td><td>Mask off features of a sequence</td></tr>
<tr><td><a href="plotorf.html">plotorf</a></td><td>Plot potential open reading frames</td></tr>
<tr><td><a href="prettyseq.html">prettyseq</a></td><td>Output sequence with translated ranges</td></tr>
<tr><td><a href="remap.html">remap</a></td><td>Display a sequence with restriction cut sites, translation etc</td></tr>
<tr><td><a href="showfeat.html">showfeat</a></td><td>Show features of a sequence</td></tr>
<tr><td><a href="showorf.html">showorf</a></td><td>Pretty output of DNA translations</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="transeq.html">transeq</a></td><td>Translate nucleic acid sequences</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>














<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (Nov 2000) - Alan Bleasby.














<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->








</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

