<HTML>

<HEAD>
  <TITLE>
  EMBOSS: complex
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
complex
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
Find the linguistic complexity in nucleotide sequences
<H2>
    Description
</H2>


<H2>
    Usage
</H2>
<b>Here is a sample session with complex</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>complex -omnia </b>
Find the linguistic complexity in nucleotide sequences
Input nucleotide sequence(s): <b>tembl:x*</b>
Window length [100]: <b></b>
Step size [5]: <b></b>
Minimum word length [4]: <b></b>
Maximum word length [6]: <b></b>
Program complex output file [x59796.complex]: <b></b>
output sequence(s) [x59796.fasta]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>

<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
Find the linguistic complexity in nucleotide sequences
Version: EMBOSS:6.3.1

   Standard (Mandatory) qualifiers (* if not always prompted):
  [-sequence]          seqall     Nucleotide sequence(s) filename and optional
                                  format, or reference (input USA)
   -lwin               integer    [100] Window length (Integer 1 or more)
   -step               integer    [5] Displacement of the window over the
                                  sequence (Integer 1 or more)
   -jmin               integer    [4] Minimum word length (Integer from 2 to
                                  20)
   -jmax               integer    [6] Maximum word length (Integer from 2 to
                                  50)
  [-outfile]           outfile    [*.complex] Program complex output file
*  -outseq             seqoutall  [<sequence>.<format>] Sequence set(s)
                                  filename and optional format (output USA)

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers:
   -omnia              toggle     [N] Calculate over a set of sequences
   -sim                integer    [0] Calculate the linguistic complexity by
                                  comparison with a number of simulations
                                  having a uniform distribution of bases (Any
                                  integer value)
   -freq               boolean    [N] Execute the simulation of a sequence
                                  based on the base frequency of the original
                                  sequence
   -print              boolean    [N] Generate a file named UjTable containing
                                  the values of Uj for each word j in the
                                  real sequence(s) and in any simulated
                                  sequences
   -ujtablefile        outfile    [complex.ujtable] Program complex temporary
                                  output file

   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1            integer    Start of each sequence to be used
   -send1              integer    End of each sequence to be used
   -sreverse1          boolean    Reverse (if DNA)
   -sask1              boolean    Ask for begin/end/reverse
   -snucleotide1       boolean    Sequence is nucleotide
   -sprotein1          boolean    Sequence is protein
   -slower1            boolean    Make lower case
   -supper1            boolean    Make upper case
   -sformat1           string     Input sequence format
   -sdbname1           string     Database name
   -sid1               string     Entryname
   -ufo1               string     UFO features
   -fformat1           string     Features format
   -fopenfile1         string     Features file name

   "-outfile" associated qualifiers
   -odirectory2        string     Output directory

   "-ujtablefile" associated qualifiers
   -odirectory         string     Output directory

   "-outseq" associated qualifiers
   -osformat           string     Output seq format
   -osextension        string     File name extension
   -osname             string     Base file name
   -osdirectory        string     Output directory
   -osdbname           string     Database name to add
   -ossingle           boolean    Separate file for each entry
   -oufo               string     UFO features
   -offormat           string     Features format
   -ofname             string     Features file name
   -ofdirectory        string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
</td></tr></table>
<p>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-sequence]<br>(Parameter 1)</td>
<td>seqall</td>
<td>Nucleotide sequence(s) filename and optional format, or reference (input USA)</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-lwin</td>
<td>integer</td>
<td>Window length</td>
<td>Integer 1 or more</td>
<td>100</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-step</td>
<td>integer</td>
<td>Displacement of the window over the sequence</td>
<td>Integer 1 or more</td>
<td>5</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-jmin</td>
<td>integer</td>
<td>Minimum word length</td>
<td>Integer from 2 to 20</td>
<td>4</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-jmax</td>
<td>integer</td>
<td>Maximum word length</td>
<td>Integer from 2 to 50</td>
<td>6</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-outfile]<br>(Parameter 2)</td>
<td>outfile</td>
<td>Program complex output file</td>
<td>Output file</td>
<td><i>&lt;*&gt;</i>.complex</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-outseq</td>
<td>seqoutall</td>
<td>Sequence set(s) filename and optional format (output USA)</td>
<td>Writeable sequence(s)</td>
<td><i>&lt;*&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>-omnia</td>
<td>toggle</td>
<td>Calculate over a set of sequences</td>
<td>Toggle value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-sim</td>
<td>integer</td>
<td>Calculate the linguistic complexity by comparison with a number of simulations having a uniform distribution of bases</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-freq</td>
<td>boolean</td>
<td>Execute the simulation of a sequence based on the base frequency of the original sequence</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-print</td>
<td>boolean</td>
<td>Generate a file named UjTable containing the values of Uj for each word j in the real sequence(s) and in any simulated sequences</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-ujtablefile</td>
<td>outfile</td>
<td>Program complex temporary output file</td>
<td>Output file</td>
<td>complex.ujtable</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-sequence" associated seqall qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sbegin1<br>-sbegin_sequence</td>
<td>integer</td>
<td>Start of each sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -send1<br>-send_sequence</td>
<td>integer</td>
<td>End of each sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sreverse1<br>-sreverse_sequence</td>
<td>boolean</td>
<td>Reverse (if DNA)</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sask1<br>-sask_sequence</td>
<td>boolean</td>
<td>Ask for begin/end/reverse</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -snucleotide1<br>-snucleotide_sequence</td>
<td>boolean</td>
<td>Sequence is nucleotide</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sprotein1<br>-sprotein_sequence</td>
<td>boolean</td>
<td>Sequence is protein</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -slower1<br>-slower_sequence</td>
<td>boolean</td>
<td>Make lower case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -supper1<br>-supper_sequence</td>
<td>boolean</td>
<td>Make upper case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sformat1<br>-sformat_sequence</td>
<td>string</td>
<td>Input sequence format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sdbname1<br>-sdbname_sequence</td>
<td>string</td>
<td>Database name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sid1<br>-sid_sequence</td>
<td>string</td>
<td>Entryname</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ufo1<br>-ufo_sequence</td>
<td>string</td>
<td>UFO features</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fformat1<br>-fformat_sequence</td>
<td>string</td>
<td>Features format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fopenfile1<br>-fopenfile_sequence</td>
<td>string</td>
<td>Features file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-outfile" associated outfile qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -odirectory2<br>-odirectory_outfile</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-ujtablefile" associated outfile qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -odirectory</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-outseq" associated seqoutall qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -osformat</td>
<td>string</td>
<td>Output seq format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -osextension</td>
<td>string</td>
<td>File name extension</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -osname</td>
<td>string</td>
<td>Base file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -osdirectory</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -osdbname</td>
<td>string</td>
<td>Database name to add</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ossingle</td>
<td>boolean</td>
<td>Separate file for each entry</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -oufo</td>
<td>string</td>
<td>UFO features</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -offormat</td>
<td>string</td>
<td>Features format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ofname</td>
<td>string</td>
<td>Features file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ofdirectory</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>

<H2>
    Input file format
</H2>

<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:x*' is a sequence entry in the example nucleic acid database 'tembl'
<p>


<H2>
    Output file format
</H2>

Sequence TEMBL:HHTETRA contains repeats and is included in the test
database for repeat analysis.
<p>

<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: complex.ujtable</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
</pre>
</td></tr></table><p>
<p><h3>File: x59796.complex</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Length of window : 100 
jmin : 4 
jmax : 6 
step : 5 
Execution without simulation 
----------------------------------------------------------------------------
|                  |                  |                  |                  |
|     number of    |      name of     |     length of    |      value of    |
|     sequence     |     sequence     |     sequence     |     complexity   |
|                  |                  |                  |                  |
----------------------------------------------------------------------------
         1                      X59796           3170             0.6921 
         2                      X65923            518             0.6739 
         3                      X65921           2016             0.7105 
         4                      X51466           3075             0.6925 
         5                      X07523           1658             0.7314 
         6                      X03487            512             0.5609 
         7                      X03488           1132             0.7217 
         8                      X07797           1675             0.6201 
         9                      X51872           1832             0.6916 
        10                      X77160           1212             0.6596 
        11                      X13776           2167             0.6562 
        12                      X77161           1130             0.6989 
</pre>
</td></tr></table><p>
<p><h3>File: x59796.fasta</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt;X59796 X59796.1 H.sapiens mRNA for cadherin-5
ctccactcacgctcagccctggacggacaggcagtccaacggaacagaaacatccctcag
cccacaggcacgatctgttcctcctgggaagatgcagaggctcatgatgctcctcgccac
atcgggcgcctgcctgggcctgctggcagtggcagcagtggcagcagcaggtgctaaccc
tgcccaacgggacacccacagcctgctgcccacccaccggcgccaaaagagagattggat
ttggaaccagatgcacattgatgaagagaaaaacacctcacttccccatcatgtaggcaa
gatcaagtcaagcgtgagtcgcaagaatgccaagtacctgctcaaaggagaatatgtggg
caaggtcttccgggtcgatgcagagacaggagacgtgttcgccattgagaggctggaccg
ggagaatatctcagagtaccacctcactgctgtcattgtggacaaggacactggcgaaaa
cctggagactccttccagcttcaccatcaaagttcatgacgtgaacgacaactggcctgt
gttcacgcatcggttgttcaatgcgtccgtgcctgagtcgtcggctgtggggacctcagt
catctctgtgacagcagtggatgcagacgaccccactgtgggagaccacgcctctgtcat
gtaccaaatcctgaaggggaaagagtattttgccatcgataattctggacgtattatcac
aataacgaaaagcttggaccgagagaagcaggccaggtatgagatcgtggtggaagcgcg
agatgcccagggcctccggggggactcgggcacggccaccgtgctggtcactctgcaaga
catcaatgacaacttccccttcttcacccagaccaagtacacatttgtcgtgcctgaaga
cacccgtgtgggcacctctgtgggctctctgtttgttgaggacccagatgagccccagaa
ccggatgaccaagtacagcatcttgcggggcgactaccaggacgctttcaccattgagac
aaaccccgcccacaacgagggcatcatcaagcccatgaagcctctggattatgaatacat
ccagcaatacagcttcatagtcgaggccacagaccccaccatcgacctccgatacatgag
ccctcccgcgggaaacagagcccaggtcattatcaacatcacagatgtggacgagccccc
cattttccagcagcctttctaccacttccagctgaaggaaaaccagaagaagcctctgat
tggcacagtgctggccatggaccctgatgcggctaggcatagcattggatactccatccg
caggaccagtgacaagggccagttcttccgagtcacaaaaaagggggacatttacaatga
gaaagaactggacagagaagtctacccctggtataacctgactgtggaggccaaagaact
ggattccactggaacccccacaggaaaagaatccattgtgcaagtccacattgaagtttt
ggatgagaatgacaatgccccggagtttgccaagccctaccagcccaaagtgtgtgagaa
cgctgtccatggccagctggtcctgcagatctccgcaatagacaaggacataacaccacg
aaacgtgaagttcaaattcatcttgaatactgagaacaactttaccctcacggataatca
cgataacacggccaacatcacagtcaagtatgggcagtttgaccgggagcataccaaggt
ccacttcctacccgtggtcatctcagacaatgggatgccaagtcgcacgggcaccagcac
gctgaccgtggccgtgtgcaagtgcaacgagcagggcgagttcaccttctgcgaggatat
ggccgcccaggtgggcgtgagcatccaggcagtggtagccatcttactctgcatcctcac
catcacagtgatcaccctgctcatcttcctgcggcggcggctccggaagcaggcccgcgc
gcacggcaagagcgtgccggagatccacgagcagctggtcacctacgacgaggagggcgg
cggcgagatggacaccaccagctacgatgtgtcggtgctcaactcggtgcgccgcggcgg
ggccaagcccccgcggcccgcgctggacgcccggccttccctctatgcgcaggtgcagaa
gccaccgaggcacgcgcctggggcacacggagggcccggggagatggcagccatgatcga
ggtgaagaaggacgaggcggaccacgacggcgacggccccccctacgacacgctgcacat
ctacggctacgagggctccgagtccatagccgagtccctcagctccctgggcaccgactc
atccgactctgacgtggattacgacttccttaacgactggggacccaggtttaagatgct
ggctgagctgtacggctcggacccccgggaggagctgctgtattaggcggccgaggtcac
tctgggcctggggacccaaaccccctgcagcccaggccagtcagactccaggcaccacag
cvncadctccaaaaatggcagtgactccccagcccagcaccccttcctcgtgggtcccag
agacctcatcagccttgggatagcaaactccaggttcctgaaatatccaggaatatatgt
cagtgatgactattctcaaatgctggcaaatccaggctggtgttctgtctgggctcagac
atccacataaccctgtcacccacagaccgccgtctaactcaaagacttcctctggctccc
caaggctgcaaagcaaaacagactgtgtttaactgctgcagggtctttttctagggtccc
tgaacgccctggtaaggctggtgaggtcctggtgcctatctgcctggaggcaaaggcctg
gacagcttgacttgtggggcaggattctctgcagcccattcccaagggagactgaccatc


<font color=red>  [Part of this file has been deleted for brevity]</font>

tacggttcctcgtgggctgctacatgtcgcacacgcgcaaggcggtgatgccggtggtcg
agcgcgccgacgcgctgctctgctacccgaccccctacgagggcttcgagtattcgccga
acatcgtctacggcggtccggcgccgaaccagaacagtgcgccgctggcggcgtacctga
ttcgccactacggcgagcgggtggtgttcatcggctcggactacatctatccgcgggaaa
gcaaccatgtgatgcgccacctgtatcgccagcacggcggcacggtgctcgaggaaatct
acattccgctgtatccctccgacgacgacttgcagcgcgccgtcgagcgcatctaccagg
cgcgcgccgacgtggtcttctccaccgtggtgggcaccggcaccgccgagctgtatcgcg
ccatcgcccgtcgctacggcgacggcaggcggccgccgatcgccagcctgaccaccagcg
aggcggaggtggcgaagatggagagtgacgtggcagaggggcaggtggtggtcgcgcctt
acttctccagcatcgatacgcccgccagccgggccttcgtccaggcctgccatggtttct
tcccggagaacgcgaccatcaccgcctgggccgaggcggcctactggcagaccttgttgc
tcggccgcgccgcgcaggccgcaggcaactggcgggtggaagacgtgcagcggcacctgt
acgacatcgacatcgacgcgccacaggggccggtccgggtggagcgccagaacaaccaca
gccgcctgtcttcgcgcatcgcggaaatcgatgcgcgcggcgtgttccaggtccgctggc
agtcgcccgaaccgattcgccccgacccttatgtcgtcgtgcataacctcgacgactggt
ccgccagcatgggcgggggaccgctcccatgagcgccaactcgctgctcggcagcctgcg
cgagttgcaggtgctggtcctcaacccgccgggggaggtcagcgacgccctggtcttgca
gctgatccgcatcggttgttcggtgcgccagtgctggccgccgccggaagccttcgacgt
gccggtggacgtggtcttcaccagcattttccagaatggccaccacgacgagatcgctgc
gctgctcgccgccgggactccgcgcactaccctggtggcgctggtggagtacgaaagccc
cgcggtgctctcgcagatcatcgagctggagtgccacggcgtgatcacccagccgctcga
tgcccaccgggtgctgcctgtgctggtatcggcgcggcgcatcagcgaggaaatggcgaa
gctgaagcagaagaccgagcagctccaggaccgcatcgccggccaggcccggatcaacca
ggccaaggtgttgctgatgcagcgccatggctgggacgagcgcgaggcgcaccagcacct
gtcgcgggaagcgatgaagcggcgcgagccgatcctgaagatcgctcaggagttgctggg
aaacgagccgtccgcctgagcgatccgggccgaccagaacaataacaagaggggtatcgt
catcatgctgggactggttctgctgtacgttggcgcggtgctgtttctcaatgccgtctg
gttgctgggcaagatcagcggtcgggaggtggcggtgatcaacttcctggtcggcgtgct
gagcgcctgcgtcgcgttctacctgatcttttccgcagcagccgggcagggctcgctgaa
ggccggagcgctgaccctgctattcgcttttacctatctgtgggtggccgccaaccagtt
cctcgag
&gt;X77161 X77161.1 Pseudomonas aeruginosa (PAC1) amiS gene.
gagccgtccgcctgagcgatccgggccgaccagaacaataacaagaggggtatcgtcatc
atgctgggactggttctgctgtacgttggcgcggtgctgtttctcaatgccgtctggttg
ctgggcaagatcagcggtcgggaggtggcggtgatcaacttcctggtcggcgtgctgagc
gcctgcgtcgcgttctacctgatcttttccgcagcagccgggcagggctcgctgaaggcc
ggagcgctgaccctgctattcgcttttacctatctgtgggtggccgccaaccagttcctc
gaggtggacggcaagggcctcggctggttctgcctgttcgtcagcctcaccgcctgcacc
gtggcgatcgagtcgttcgccggcgccagtggtccgttcggcctgtggaacgcggtcaac
tggacagtctgggcgttgctctggttctgtttcttcctgctgctggggctgtcccgcggc
atccagaagccggtggcctacctgaccctggccagcgccatattcaccgcctggttgccc
ggcctgctgctgctcggacaggtgctcaaggcatagcaggaagtcggaaagggatgacgg
cttgccgccatcccgtcccttccgaacgcctagccgagcggccagttgatcaccacgacg
gcgtcgttgtagtcgttgtcggtgccgtcttcagagccgaccagggcgaagttcagctcg
ttggtcaggattacctgtgccgagaccagatccgaggggcggccgttgacgctgacctgg
acctgtaccttgccactgctgccggagttgagcacctgggtgccgatgacggcgttattg
gtgctttgcccgctgaaggtcgcggccgtgctcgttgttgaccagcacgttcaccgtctg
ggttccggacgagttggcgaaggcggtgacgccggaacctggttgttggcgggaagggtg
aacactccttgtggttgccatggtggtatctccactgaatacctggccccttccttttca
ggcagccgtctggcgcgcggtatggcgtgtcgggagaaatccgcagtccttggcggcagg
cgatgcgcaggcaggaaggacgcatcgttcagccaatctacgccgtcgac
</pre>
</td></tr></table><p>

<H2>
    Data files
</H2>

None

<H2>
    Notes
</H2>

None


<H2>
    References
</H2>

None


<H2>
    Warnings
</H2>

None


<H2>
    Diagnostic Error Messages
</H2>

None


<H2>
    Exit status
</H2>

Always exits with status 0

<H2>
    Known bugs
</H2>

None

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="banana.html">banana</a></td>
<td>Plot bending and curvature data for B-DNA</td>
</tr>

<tr>
<td><a href="btwisted.html">btwisted</a></td>
<td>Calculate the twisting in a B-DNA sequence</td>
</tr>

<tr>
<td><a href="chaos.html">chaos</a></td>
<td>Draw a chaos game representation plot for a nucleotide sequence</td>
</tr>

<tr>
<td><a href="compseq.html">compseq</a></td>
<td>Calculate the composition of unique words in sequences</td>
</tr>

<tr>
<td><a href="dan.html">dan</a></td>
<td>Calculates nucleic acid melting temperature</td>
</tr>

<tr>
<td><a href="density.html">density</a></td>
<td>Draw a nucleic acid density plot</td>
</tr>

<tr>
<td><a href="freak.html">freak</a></td>
<td>Generate residue/base frequency table or plot</td>
</tr>

<tr>
<td><a href="isochore.html">isochore</a></td>
<td>Plots isochores in DNA sequences</td>
</tr>

<tr>
<td><a href="sirna.html">sirna</a></td>
<td>Finds siRNA duplexes in mRNA</td>
</tr>

<tr>
<td><a href="wordcount.html">wordcount</a></td>
<td>Count and extract unique words in molecular sequence(s)</td>
</tr>

</table>

<H2>
    Author(s)
</H2>



<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.

<H2>
    Comments
</H2>
None

</BODY>
</HTML>

