<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: cons
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">


<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

cons
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<!--END OF HEADER-->


<H2>
    Function
</H2>
Creates a consensus from multiple alignments

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
        Algorithms
  -->

<b>cons</b> calculates a consensus sequence from a multiple sequence
alignment.  To obtain the consensus, the sequence weights and a scoring
matrix are used to calculate a score at each position in the alignment. 
<P>

The residue (or nucleotide) <b>i</b> in an alignment column, is compared
to all other residues (<b>j</b>) in the column.  The score for <b>i</b>
is the sum over all residues <b>j</b> (not <b>i=j</b>) of the
score(<b>ij</b>)*weight(<b>j</b>) .  Where score(<b>ij</b>) is taken
from a nucleotide or protein scoring matrix (see <b>-datafile</b>
qualifier) and the "weight(<b>j</b>)" is the weighting given to the
sequence <b>j</b>, which is given in the alignment file.
<P>

The highest scoring type of residue is then found in the column.  If the
number of positive matches for this residue is greater than the
"plurality value" then this residue is the consensus.  The positive
matches for a residue <b>i</b> are calculated as being the sum of
weights of all the residues that increase the score of residue <b>i</b>
(i.e.  positive). 
<P>

Where no consensus is found at a position <b>i</b>, an 'n' or an 'x'
character is output; (depending on it being a DNA or protein sequence).
<P>

   The "plurality" qualifier allows the user to set a cut-off for the
number of positive matches below which there is no consensus.
<P>

    The "identity" qualifier provides the facility of setting the required
number of identities at a site for it to give a consensus at that
position.  Therefore, if this is set to the number of sequences in the
alignment only columns of identities contribute to the consensus. 
<P>

    The "setcase" qualifier sets the threshold for the positive matches
above which the consensus is is upper-case and below which the consensus
is in lower-case. 
<P>



<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->


<b>Here is a sample session with cons</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>cons </b>
Creates a consensus from multiple alignments
Input sequence set: <b>../../data/dna.msf</b>
Output sequence [dna.fasta]: <b>aligned.cons</b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>


<!--
Here is a sample session with cons:

<pre>

% cons
Creates a consensus from multiple alignments
Input sequence set: aligned.fasta
Output file [outfile.cons]: aligned.cons

</pre>

-->





<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-sequence]          seqset     File containing a sequence alignment.
  [-outseq]            seqout     Output sequence USA

   Additional (Optional) qualifiers:
   -datafile           matrix     This is the scoring matrix file used when
                                  comparing sequences. By default it is the
                                  file 'EBLOSUM62' (for proteins) or the file
                                  'EDNAFULL' (for nucleic sequences). These
                                  files are found in the 'data' directory of
                                  the EMBOSS installation.
   -plurality          float      Set a cut-off for the number of positive
                                  matches below which there is no consensus.
                                  The default plurality is taken as half the
                                  total weight of all the sequences in the
                                  alignment.
   -identity           integer    Provides the facility of setting the
                                  required number of identities at a site for
                                  it to give a consensus at that position.
                                  Therefore, if this is set to the number of
                                  sequences in the alignment only columns of
                                  identities contribute to the consensus.
   -setcase            float      Sets the threshold for the positive matches
                                  above which the consensus is is upper-case
                                  and below which the consensus is in
                                  lower-case.
   -name               string     Name of the consensus sequence

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1             integer    Start of each sequence to be used
   -send1               integer    End of each sequence to be used
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name

   "-outseq" associated qualifiers
   -osformat2           string     Output seq format
   -osextension2        string     File name extension
   -osname2             string     Base file name
   -osdirectory2        string     Output directory
   -osdbname2           string     Database name to add
   -ossingle2           boolean    Separate file for each entry
   -oufo2               string     UFO features
   -offormat2           string     Features format
   -ofname2             string     Features file name
   -ofdirectory2        string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>File containing a sequence alignment.</td>
<td>Readable set of sequences</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outseq]<br>(Parameter 2)</td>
<td>Output sequence USA</td>
<td>Writeable sequence</td>
<td><i>&lt;sequence&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-datafile</td>
<td>This is the scoring matrix file used when comparing sequences. By default it is the file 'EBLOSUM62' (for proteins) or the file 'EDNAFULL' (for nucleic sequences). These files are found in the 'data' directory of the EMBOSS installation.</td>
<td>Comparison matrix file in EMBOSS data path</td>
<td>EBLOSUM62 for protein<br>EDNAFULL for DNA</td>
</tr>

<tr>
<td>-plurality</td>
<td>Set a cut-off for the number of positive matches below which there is no consensus. The default plurality is taken as half the total weight of all the sequences in the alignment.</td>
<td>Any numeric value</td>
<td>Half the total sequence weighting</td>
</tr>

<tr>
<td>-identity</td>
<td>Provides the facility of setting the required number of identities at a site for it to give a consensus at that position. Therefore, if this is set to the number of sequences in the alignment only columns of identities contribute to the consensus.</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr>
<td>-setcase</td>
<td>Sets the threshold for the positive matches above which the consensus is is upper-case and below which the consensus is in lower-case.</td>
<td>Any numeric value</td>
<td>@( $(sequence.totweight) / 2)</td>
</tr>

<tr>
<td>-name</td>
<td>Name of the consensus sequence</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

The USA of a set of aligned sequences.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: ../../data/dna.msf</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
!!NA_MULTIPLE_ALIGNMENT

 dna.msf  MSF: 120  Type: N  January 01, 1776  12:00  Check: 3196 ..

 Name: MSFM1          Len:   120  Check:  8587  Weight:  1.00
 Name: MSFM2          Len:   120  Check:  6178  Weight:  1.00
 Name: MSFM3          Len:   120  Check:  8431  Weight:  1.00

//

        MSFM1  ACGTACGTAC GTACGTACGT ACGTACGTAC GTACGTACGT ACGTACGTAC
        MSFM2  ACGTACGTAC GTACGTACGT ....ACGTAC GTACGTACGT ACGTACGTAC
        MSFM3  ACGTACGTAC GTACGTACGT ACGTACGTAC GTACGTACGT CGTACGTACG

        MSFM1  GTACGTACGT ACGTACGTAC GTACGTACGT ACGTACGTAC GTACGTACGT
        MSFM2  GTACGTACGT ACGTACGTAC GTACGTACGT ACGTACGTAC GTACGTACGT
        MSFM3  TACGTACGTA CGTACGTACG TACGTACGTA ACGTACGTAC GTACGTACGT

        MSFM1  ACGTACGTAC GTACGTACGT
        MSFM2  ACGTACGTTG CAACGTACGT
        MSFM3  ACGTACGTAC GTACGTACGT

</pre>
</td></tr></table><p>












<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->

The output consists of a sequence file holding the consensus sequence. 

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: aligned.cons</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt;EMBOSS_001
ACGTACGTACGTACGTACGTnnnnACGTACGTACGTACGTnnnnnnnnnnnnnnnnnnnn
nnnnnnnnnnnnnnnnnnnnACGTACGTACGTACGTACGTACGTACGTnnnnACGTACGT
</pre>
</td></tr></table><p>


<!--
For example:

<P>
<HR>
<pre>
>EMBOSS_001
tagctgacctgacgggactgatgcgt
</pre>
<HR>
-->














<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.
	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
   -->

It uses the standard set of scoring matrix data files.
<P>

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>















<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="megamerger.html">megamerger</a></td><td>Merge two large overlapping nucleic acid sequences</td></tr>
<tr><td><a href="merger.html">merger</a></td><td>Merge two overlapping nucleic acid sequences</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

Tim Carver (tcarver&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK













<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (Oct 2000) - Tim Carver

<!--
<p>
[an error occurred while processing this directive]
-->










<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->








</BODY>
</HTML>

