<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: contacts
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">
<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>
<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>
</td>
<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: contacts
</H2>
</font>
</td>
</tr>
</table>
</td>
</tr>
</table>
<p><hr><p>

<!--END OF HEADER-->

<H1>
Program contacts
</H1>
<H2>
    Function
</H2>
Reads coordinate files and writes contact files

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
        Algorithms
  -->

<b>contacts</b> parses an embl-like clean coordinate files generated by
the <b>coorde</b> application (not currently in emboss, email Jon Ison
jison@hgmp.mrc.ac.uk) or the <b>domainer</b> application and writes, for
each file in a given directory, files of residue-residue contact data in
embl-like format.  Each of these files contains residue contact data for
each chain of every model in the coordinate file (or the single scop
domain in the case where domainer output is read). 

<p>

Contact between two residues is defined as when the van der Waals surface of 
any atom of the first residue comes within the threshold contact distance of 
the van der Waals surface of any atom of the second residue. The threshold 
contact distance is a user-defined distance with a default value of 1 
Angstrom.

<p>

The following van der Waals radii are used: 

<p>
<pre>
C:1.8 Angstrom 
O:1.4 Angstrom 
N:1.7 Angstrom 
S:2.0 Angstrom
H:1.0 Angstrom (default for other or unknown atom types)
</pre>
<p>

<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->


Here is a sample session with contacts:

<pre>

% contacts
Reads coordinate files and writes contact files
Location of coordinate files for input (embl-like format) [./]: 
Extension of coordinate files (embl-like format) [.pxyz]: 
Location of contact files for output [./]: 
Extension of contact files [.con]: 
Threshold contact distance [1.0]: 
Name of data file with van der Waals radii [Evdw.dat]: 
Name of log file for the build [contacts.log]: 
</pre>















<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-cpdb]              string     Location of coordinate files for input
                                  (embl-like format)
  [-cpdbextn]          string     Extension of coordinate files (embl-like
                                  format)
  [-con]               string     Location of contact files for output
  [-conextn]           string     Extension of contact files
  [-thresh]            float      Threshold contact distance
  [-ignore]            float      If any two atoms from two different residues
                                  are at least this distance apart then no
                                  futher inter-atomic contacts will be checked
                                  for for that residue pair . This speeds the
                                  calculation up considerably.
  [-vdwf]              string     Name of data file with van der Waals radii
  [-conerrf]           outfile    Name of log file for the build

   Optional qualifiers: (none)
   Advanced qualifiers: (none)
   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-cpdb]<br>(Parameter 1)</td>
<td>Location of coordinate files for input (embl-like format)</td>
<td>Any string is accepted</td>
<td>./</td>
</tr>

<tr>
<td>[-cpdbextn]<br>(Parameter 2)</td>
<td>Extension of coordinate files (embl-like format)</td>
<td>Any string is accepted</td>
<td>.pxyz</td>
</tr>

<tr>
<td>[-con]<br>(Parameter 3)</td>
<td>Location of contact files for output</td>
<td>Any string is accepted</td>
<td>./</td>
</tr>

<tr>
<td>[-conextn]<br>(Parameter 4)</td>
<td>Extension of contact files</td>
<td>Any string is accepted</td>
<td>.con</td>
</tr>

<tr>
<td>[-thresh]<br>(Parameter 5)</td>
<td>Threshold contact distance</td>
<td>Any integer value</td>
<td>1.0</td>
</tr>

<tr>
<td>[-ignore]<br>(Parameter 6)</td>
<td>If any two atoms from two different residues are at least this distance apart then no futher inter-atomic contacts will be checked for for that residue pair . This speeds the calculation up considerably.</td>
<td>Any integer value</td>
<td>20.0</td>
</tr>

<tr>
<td>[-vdwf]<br>(Parameter 7)</td>
<td>Name of data file with van der Waals radii</td>
<td>Any string is accepted</td>
<td>Evdw.dat</td>
</tr>

<tr>
<td>[-conerrf]<br>(Parameter 8)</td>
<td>Name of log file for the build</td>
<td>Output file</td>
<td>contacts.log</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

It reads in an embl-like clean coordinate files generated by the
<b>coorde</b> application or the <b>domainer</b> application. 

<p>

For example:

<p>

<hr>
<pre>
ID   D1HBBA_
XX
DE   Co-ordinates for SCOP domain D1HBBA_
XX
OS   See Escop.dat for domain classification
XX
EX   METHOD xray; RESO 1.90; NMOD 1; NCHA 1;
XX
CN   [1]
XX
IN   ID A; NR 141; NH 0; NW 0;
XX
SQ   SEQUENCE   141 AA;  15127 MW;  5EC7DB1E CRC32;
     VLSPADKTNV KAAWGKVGAH AGEYGAEALE RMFLSFPTTK TYFPHFDLSH GSAQVKGHGK
     KVADALTNAV AHVDDMPNAL SALSDLHAHK LRVDPVNFKL LSHCLLVTLA AHLPAEFTPA
     VHASLDKFLA SVSTVLTSKY R
XX
CO   1    1    P    1     1     V    VAL    N      7.155   17.725 4.424     1.00
    37.82
CO   1    1    P    1     1     V    VAL    CA     7.854   18.800 3.718     1.00
    35.10
CO   1    1    P    1     1     V    VAL    C      9.366   18.565 3.754     1.00
    31.92
CO   1    1    P    1     1     V    VAL    O      9.861   17.961 4.721     1.00
    35.01
CO   1    1    P    1     1     V    VAL    CB     7.529   20.168 4.360     1.00
    47.63
CO   1    1    P    1     1     V    VAL    CG1    7.806   21.300 3.369     1.00
    62.84
CO   1    1    P    1     1     V    VAL    CG2    6.136   20.244 4.936     1.00
    54.85
CO   1    1    P    2     2     L    LEU    N     10.032   19.062 2.731     1.00
    27.38
CO   1    1    P    2     2     L    LEU    CA    11.496   18.967 2.657     1.00
    23.24
CO   1    1    P    2     2     L    LEU    C     12.077   20.110 3.496     1.00
    22.99
CO   1    1    P    2     2     L    LEU    O     11.672   21.259 3.289     1.00
    25.22
</pre>
<hr>











<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->

The embl-like format used for the contact files uses the following
records:

<p>

<ol>
<li>ID - either the 4-character PDB identifier code (where clean protein 
coordinate files are used as input) or the 7-character domain identifier 
code taken from scop (where domain coordinate files were used as input; see 
documentation for the EMBOSS application scope for further info.)

<li>DE - bibliographic information.  The text "Residue-residue contact 
data" is always given.

<li>EX - experimental information. The value of the threshold contact 
distance is given as a floating point number after 'THRESH'. The number of 
models and number of polypeptide chains are given after 'NMOD' and 'NCHA' 
respectively. domain coordinate files a 1 is always given. Following the EX 
record, the file will have a section containing a CN, IN and SM records 
(see below) for each chain. The sections for each chain of a model are given 
after the MO record.

<li>MO - model number. The number given in brackets after this record 
indicates the start of a section of model-specific data. 

<li>CN - chain number.  The number given in brackets after this record 
indicates the start of a section of chain-specific data.

<li>IN - chain specific data. The character given after ID is the PDB chain 
identifier taken from the input file, (a '.` given in cases where a chain 
identifier was not specified in the original pdb file or, for domain 
coordinate files, the domain is comprised of more than one domain).  The 
number of amino acid residues comprising the chain (or the chains from 
which a domain is comprised) is given after NR. The number of residue-residue
contacts is given after NSMCON. 

<li>SM - Line of residue contact data. Pairs of amino acid identifiers and 
residue numbers are delimited by a ';'. Residue numbers are taken from the 
clean coordinate file and give a correct index into the sequence (i.e. they
are not necessarily the same as the original pdb file).

<li>XX - used for spacing.

<li>// - given on the last line of the file only.
</ol>

<p>

Note - SM records are used for contacts between either either side-chain 
or main-chain atoms as defined above.  In a future implementation, SS will
be used for side-chain only contacts, MM will be used for main-chain only 
contacts, and there will probably be several other forms of contact too.

<p>

Example contacts output file:

<p>
<hr>
<pre>
ID   D1HBBB_
XX
DE   Residue-residue side-chain contact data
XX
EX   THRESH 10.0; NMOD 1; NCHA 1;
XX
MO   [1]
XX
CN   [1]
XX
IN   ID B; NR 146; NSMCON 2514;
XX
SM   VAL 1 ; HIS 2
SM   VAL 1 ; LEU 3
SM   VAL 1 ; THR 4
SM   VAL 1 ; PRO 5
SM   VAL 1 ; GLU 6
SM   VAL 1 ; GLU 7
SM   VAL 1 ; LYS 8
SM   VAL 1 ; VAL 11
SM   VAL 1 ; PHE 71
//
</pre>
<hr>
<p>


<b>contacts</b> generates a log file an excerpt of which is shown
below.  If there is a problem in processing a coordinate file,
three lines containing the record '//', the scop domain or pdb
identifier code and an error message respectively are written.  The text
'WARN file open error filename', 'ERROR file read error filename' or
'ERROR file write error filename ' will be reported when an error was
encountered during a file open, read or write respectively.  Various
other error messages may also be given (in case of difficulty email Jon
Ison, jison@hgmp.mrc.ac.uk). 

<p>

Example log file

<p>
<hr>
<pre>
//
DS002__
WARN  Could not open for reading cpdb file s002.pxyz
//
DS003__
WARN  Could not open for reading cpdb file s003.pxyz
</pre>
<hr>
<p>
















<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.
	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
   -->

<b>contacts</b> reads in data on van der Waals radii for atoms in
proteins from the data file <b>Evdw.dat</b> (by default). 

<p>

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>














<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="scopalign.html">scopalign</a></td><td>Generate alignments for SCOP families</td></tr>
<tr><td><a href="siggen.html">siggen</a></td><td>Generates a sparse protein signature</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

This application was written by Jon Ison <a
href="mailto:jison@hgmp.mrc.ac.uk">(jison@hgmp.mrc.ac.uk)</a>
















<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (June 2001) - Jon Ison













<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->








</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

