<!-- START OF HEADER -->
<HTML><HEAD>
<TITLE> EMBASSY: CONTACTS documentation. </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

<H2> CONTACTS documentation
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<!-- END OF HEADER -->




<!-- CONTENTS
     This always includes the sections below.
     Other subsections can be added for individual applications.
-->
<H2>CONTENTS </H2>
<b> <a href="#1.0">1.0     SUMMARY                   </a></b><br>
<b> <a href="#2.0">2.0     INPUTS & OUTPUTS          </a></b><br>
<b> <a href="#3.0">3.0     INPUT FILE FORMAT         </a></b><br>
<b> <a href="#4.0">4.0     OUTPUT FILE FORMAT        </a></b><br>
<b> <a href="#5.0">5.0     DATA FILES                </a></b><br>
<b> <a href="#6.0">6.0     USAGE                     </a></b><br>    
<b> <a href="#7.0">7.0     KNOWN BUGS & WARNINGS     </a></b><br>    
<b> <a href="#8.0">8.0     NOTES                     </a></b><br>
<b> <a href="#9.0">9.0     DESCRIPTION               </a></b><br>
<b> <a href="#10.0">10.0   ALGORITHM                 </a></b><br>
<b> <a href="#11.0">11.0   RELATED APPLICATIONS      </a></b><br>
<b> <a href="#12.0">12.0   DIAGNOSTIC ERROR MESSAGES </a></b><br>
<b> <a href="#13.0">13.0   AUTHORS                   </a></b><br>
<b> <a href="#14.0">14.0   REFERENCES                </a></b><br>


<!-- SUMMARY
     Succint description of the application, particularly its inputs, outputs
     and what it does.  The same text is given at the top of the source (.c)
     file and in the <documentation> attribute of the <application definition>
     of the ACD file.
-->
<a name="1.0"></a>
<br><br><br><H2> 1.0   SUMMARY  </H2>
 Reads CCF files (clean coordinate files) and writes CON files (contact files) of intra-chain residue-residue contact data.

Generate intra-chain CON files from CCF files



<!-- INPUTS & OUTPUTS
     Short summary of the application inputs and outputs in its different 
     modes of usage (if appropriate). More detail than the summary.
-->
<a name="2.0"></a>
<br><br><br><H2> 2.0   INPUTS & OUTPUTS          </H2>
CONTACTS reads a directory of domain or protein CCF files (clean coordinate files) and writes a CON file (contacts file) of intra-chain residue-residue contact data for each file in the input directory.  Each CON file contains residue contact data for every chain of every model in a protein coordinate file, or for a single domain where a domain CCF file is read.  The user is prompted for the paths for the CCF (input) and CON (output) files and the file extensions are specified in the ACD file.  The domain identifier code or pdb identifier code is used as appropriate to name the output files.  A log file is also written.


<!-- INPUT FILE FORMAT 
     Description and example(s) of input file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.
     Use "<b>CONTACTS</b> reads any normal sequence USAs." if
     appropriate.
-->



<a name="3.0"></a>
<br><br><br><H2> 3.0   INPUT FILE FORMAT         </H2>
The format of the clean coordinate file is described in the <a href="pdbparse.html">PDBPARSE documentation.</a>






<!-- OUTPUT FILE FORMAT 
     Description and example(s) of output file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.  
     If the standard description of the avalable report formats is required, 
     use:   #include file="inc/reportformats.ihtml"
     Use "Outputs a graph to the specified graphics device."
     or  "outputs a report format file. The default format is ..."
     if appropriate.
-->
<a name="4.0"></a>
<br><br><br><H2> 4.0   OUTPUT FILE FORMAT         </H2>
The CON format used for the contact files (Figure 1) is similar to EMBL format and uses the following records:
<UL>
<LI> XX - used for spacing and comments.  The first line is bibliographic information and 
contains the text "XX Intra-chain residue-residue contact data".
<LI> TY - type of contact.  For CON files generated by CONTACTS, 'INTRA' is always given.
<LI>  EX - experimental information. The value of the threshold contact 
distance is given as a floating point number after 'THRESH'. For CON files generated by 
CONTACTS, the threshold ignore distance is given after 'THRESH'. The number of 
models and number of polypeptide chains are given after 'NMOD' and 'NCHA' 
respectively. For domain CCF files a 1 is always given. 
<LI> NE - number of entries (chains) in the file. For CON files generated by CONTACTS this is 
the number of chains, equal to the number of models multiplied by the number of unique chains. 
Following the NE record, the file has a section for each entry containing
records for entry number (EN), identifier codes (ID),  
polypeptide chain-specific data (CN), chain sequence information (S1) and 
number of contacts (NC), that together define the chain and its contacts. 
<LI> EN - entry number.  The number in brackets indicates the start of an entry (chain).
<LI> CN - polypeptide chain-specific data.  Tokens delimiting data items are as follow. 
 (1) MO: The model number (from the PDB file).  (2) CN1: Chain number. (3) CN2: Not used by 
CONTACTS, a '.' is given.  (4) ID1: PDB chain identifier (a '.` given in cases where a chain 
identifier was not specified in the original PDB file or, for domain CCF files, the domain 
is comprised of more than one domain). (5) ID2: Not used by CONTACTS, a '.' is given. 
(6) NRES1: number of amino acid residues in the chain or domain. (7) NRES2: Not used by SITES, a '.' is given. 
<LI> NC - number of contacts: (1) SM:  Number of residue-residue contacts; between the side-chain 
or main-chain atoms of different amino acid residues in the same chain (2) LI: Not used by 
CONTACTS, a '.' is given.
<LI> ID - identifier codes: (1) PDB: 4-character PDB identifier code. (2) DOM: 7-character domain identifier 
code from SCOP or CATH (where domain CCF files were read). (3) LIG: Not used by CONTACTS, a '.' 
is given. 
<LI> S1 - polypeptide chain sequence for domain or protein. The number of residues is given 
before AA on the first line.  The sequece is given on subsequent lines. 
<LI> SM - Line of residue contact data. Pairs of amino acid identifiers and 
residue numbers are delimited by a ';'. Residue numbers are taken from the 
CCF file and give a correct index into the sequence (i.e. they
are not necessarily the same as the original PDB file).  This sequence is 
given in the CON file itself (S1 record).
<LI> // - delimiter for individual entries in the file and also given on
the last line of the file.
</UL>




<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: 1cs4.con</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
XX   Intra-chain residue-residue contact data.
XX
TY   INTRA
XX
EX   THRESH 1.0; IGNORE 20.0; NMOD 1; NCHA 1
XX
NE   1
XX
EN   [1]
XX
ID   PDB 1cs4; DOM .; LIG .
XX
CN   MO 1; CN1 1; CN2 .; ID1 A; ID2 .; NRES1 52; NRES2 .
XX
S1   SEQUENCE    52 AA;   5817 MW;  47362A43 CRC32;
     ADIEGFTSLA SQCTAQELVM TLNELFARFD KLAAENHCLR IKILGDCYYC VS
XX
NC   SM 163; LI .
XX
SM   ASP 2 ; ILE 3
SM   ASP 2 ; GLU 4
SM   ASP 2 ; ASP 46
SM   ASP 2 ; CYS 47
SM   ILE 3 ; GLU 4
SM   ILE 3 ; GLY 5
SM   ILE 3 ; PHE 6
SM   ILE 3 ; LEU 9
SM   ILE 3 ; LEU 25
SM   ILE 3 ; ASP 46
SM   GLU 4 ; GLY 5
SM   GLU 4 ; PHE 6
SM   GLY 5 ; PHE 6
SM   GLY 5 ; THR 7
SM   GLY 5 ; SER 8
SM   GLY 5 ; LEU 9
SM   PHE 6 ; THR 7
SM   PHE 6 ; SER 8
SM   PHE 6 ; LEU 9
SM   PHE 6 ; ALA 10
SM   PHE 6 ; LEU 18
SM   PHE 6 ; LEU 22
SM   PHE 6 ; GLY 45
SM   PHE 6 ; ASP 46
SM   THR 7 ; SER 8
SM   THR 7 ; LEU 9
SM   THR 7 ; ALA 10
SM   THR 7 ; SER 11
SM   SER 8 ; LEU 9
SM   SER 8 ; ALA 10
SM   SER 8 ; SER 11


<font color=red>  [Part of this file has been deleted for brevity]</font>

SM   PHE 29 ; LYS 31
SM   PHE 29 ; LEU 32
SM   PHE 29 ; ALA 33
SM   ASP 30 ; LYS 31
SM   ASP 30 ; LEU 32
SM   ASP 30 ; ALA 33
SM   ASP 30 ; ALA 34
SM   ASP 30 ; ARG 40
SM   LYS 31 ; LEU 32
SM   LYS 31 ; ALA 33
SM   LYS 31 ; ALA 34
SM   LYS 31 ; GLU 35
SM   LEU 32 ; ALA 33
SM   LEU 32 ; ALA 34
SM   LEU 32 ; GLU 35
SM   LEU 32 ; ASN 36
SM   ALA 33 ; ALA 34
SM   ALA 33 ; GLU 35
SM   ALA 33 ; ASN 36
SM   ALA 33 ; HIS 37
SM   ALA 33 ; CYS 38
SM   ALA 34 ; GLU 35
SM   ALA 34 ; ASN 36
SM   ALA 34 ; HIS 37
SM   GLU 35 ; ASN 36
SM   GLU 35 ; HIS 37
SM   ASN 36 ; HIS 37
SM   ASN 36 ; CYS 38
SM   HIS 37 ; CYS 38
SM   HIS 37 ; LEU 39
SM   CYS 38 ; LEU 39
SM   CYS 38 ; ARG 40
SM   LEU 39 ; ARG 40
SM   LEU 39 ; ILE 41
SM   ARG 40 ; ILE 41
SM   ARG 40 ; LYS 42
SM   ARG 40 ; ILE 43
SM   ILE 41 ; LYS 42
SM   LYS 42 ; ILE 43
SM   LYS 42 ; LEU 44
SM   LYS 42 ; CYS 47
SM   ILE 43 ; LEU 44
SM   ILE 43 ; GLY 45
SM   ILE 43 ; CYS 47
SM   LEU 44 ; GLY 45
SM   LEU 44 ; ASP 46
SM   LEU 44 ; CYS 47
SM   GLY 45 ; ASP 46
SM   GLY 45 ; CYS 47
SM   ASP 46 ; CYS 47
//
</pre>
</td></tr></table><p>
<p><h3>File: 1ii7.con</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
XX   Intra-chain residue-residue contact data.
XX
TY   INTRA
XX
EX   THRESH 1.0; IGNORE 20.0; NMOD 1; NCHA 1
XX
NE   1
XX
EN   [1]
XX
ID   PDB 1ii7; DOM .; LIG .
XX
CN   MO 1; CN1 1; CN2 .; ID1 A; ID2 .; NRES1 65; NRES2 .
XX
S1   SEQUENCE    65 AA;   7396 MW;  0CFB92A3 CRC32;
     MKFAHLADIH LGYEQFHKPQ REEEFAEAFK NALEIAVQEN VDFILIAGDL FHSSRPSPGT
     LKKAI
XX
NC   SM 151; LI .
XX
SM   ASP 8 ; ILE 9
SM   ASP 8 ; HIS 10
SM   ASP 8 ; GLY 48
SM   ASP 8 ; ASP 49
SM   ILE 9 ; HIS 10
SM   ILE 9 ; LEU 11
SM   ILE 9 ; PHE 25
SM   ILE 9 ; PHE 29
SM   ILE 9 ; ILE 46
SM   ILE 9 ; ASP 49
SM   ILE 9 ; LEU 50
SM   HIS 10 ; LEU 11
SM   HIS 10 ; GLY 12
SM   HIS 10 ; TYR 13
SM   HIS 10 ; PHE 25
SM   HIS 10 ; ASP 49
SM   HIS 10 ; LEU 50
SM   LEU 11 ; GLY 12
SM   LEU 11 ; TYR 13
SM   LEU 11 ; ALA 26
SM   LEU 11 ; PHE 29
SM   LEU 11 ; LEU 50
SM   GLY 12 ; TYR 13
SM   GLY 12 ; GLU 14
SM   GLY 12 ; GLU 22
SM   TYR 13 ; GLU 14
SM   TYR 13 ; GLN 15
SM   TYR 13 ; GLU 22
SM   TYR 13 ; PHE 25
SM   GLU 14 ; GLN 15


<font color=red>  [Part of this file has been deleted for brevity]</font>

SM   ASN 31 ; ILE 35
SM   ALA 32 ; LEU 33
SM   ALA 32 ; GLU 34
SM   ALA 32 ; ILE 35
SM   ALA 32 ; ALA 36
SM   LEU 33 ; GLU 34
SM   LEU 33 ; ILE 35
SM   LEU 33 ; ALA 36
SM   LEU 33 ; VAL 37
SM   LEU 33 ; ILE 44
SM   GLU 34 ; ILE 35
SM   GLU 34 ; ALA 36
SM   GLU 34 ; VAL 37
SM   GLU 34 ; GLN 38
SM   ILE 35 ; ALA 36
SM   ILE 35 ; VAL 37
SM   ILE 35 ; GLN 38
SM   ILE 35 ; GLU 39
SM   ALA 36 ; VAL 37
SM   ALA 36 ; GLN 38
SM   ALA 36 ; GLU 39
SM   ALA 36 ; ASN 40
SM   ALA 36 ; VAL 41
SM   ALA 36 ; ILE 44
SM   VAL 37 ; GLN 38
SM   VAL 37 ; GLU 39
SM   VAL 37 ; ASN 40
SM   GLN 38 ; GLU 39
SM   GLN 38 ; ASN 40
SM   GLU 39 ; ASN 40
SM   GLU 39 ; VAL 41
SM   ASN 40 ; VAL 41
SM   ASN 40 ; ASP 42
SM   VAL 41 ; ASP 42
SM   VAL 41 ; PHE 43
SM   VAL 41 ; ILE 44
SM   ASP 42 ; PHE 43
SM   PHE 43 ; ILE 44
SM   PHE 43 ; LEU 45
SM   ILE 44 ; LEU 45
SM   ILE 44 ; ILE 46
SM   LEU 45 ; ILE 46
SM   LEU 45 ; ALA 47
SM   ILE 46 ; ALA 47
SM   ILE 46 ; GLY 48
SM   ILE 46 ; LEU 50
SM   ALA 47 ; GLY 48
SM   GLY 48 ; ASP 49
SM   GLY 48 ; LEU 50
SM   ASP 49 ; LEU 50
//
</pre>
</td></tr></table><p>
<p><h3>File: 2hhb.con</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
XX   Intra-chain residue-residue contact data.
XX
TY   INTRA
XX
EX   THRESH 1.0; IGNORE 20.0; NMOD 1; NCHA 4
XX
NE   4
XX
EN   [1]
XX
ID   PDB 2hhb; DOM .; LIG .
XX
CN   MO 1; CN1 1; CN2 .; ID1 A; ID2 .; NRES1 141; NRES2 .
XX
S1   SEQUENCE   141 AA;  15127 MW;  5EC7DB1E CRC32;
     VLSPADKTNV KAAWGKVGAH AGEYGAEALE RMFLSFPTTK TYFPHFDLSH GSAQVKGHGK
     KVADALTNAV AHVDDMPNAL SALSDLHAHK LRVDPVNFKL LSHCLLVTLA AHLPAEFTPA
     VHASLDKFLA SVSTVLTSKY R
XX
NC   SM 643; LI .
XX
SM   VAL 1 ; LEU 2
SM   VAL 1 ; SER 3
SM   VAL 1 ; LYS 127
SM   LEU 2 ; SER 3
SM   LEU 2 ; PRO 4
SM   LEU 2 ; ASP 6
SM   LEU 2 ; LYS 7
SM   LEU 2 ; VAL 73
SM   LEU 2 ; MET 76
SM   LEU 2 ; LYS 127
SM   LEU 2 ; PHE 128
SM   LEU 2 ; SER 131
SM   SER 3 ; PRO 4
SM   SER 3 ; ALA 5
SM   SER 3 ; ASP 6
SM   SER 3 ; LYS 7
SM   SER 3 ; LYS 127
SM   PRO 4 ; ALA 5
SM   PRO 4 ; ASP 6
SM   PRO 4 ; LYS 7
SM   PRO 4 ; THR 8
SM   ALA 5 ; ASP 6
SM   ALA 5 ; LYS 7
SM   ALA 5 ; THR 8
SM   ALA 5 ; ASN 9
SM   ASP 6 ; LYS 7
SM   ASP 6 ; THR 8
SM   ASP 6 ; ASN 9
SM   ASP 6 ; VAL 10


<font color=red>  [Part of this file has been deleted for brevity]</font>

SM   GLN 131 ; LYS 132
SM   GLN 131 ; VAL 133
SM   GLN 131 ; VAL 134
SM   GLN 131 ; ALA 135
SM   LYS 132 ; VAL 133
SM   LYS 132 ; VAL 134
SM   LYS 132 ; ALA 135
SM   LYS 132 ; GLY 136
SM   VAL 133 ; VAL 134
SM   VAL 133 ; ALA 135
SM   VAL 133 ; GLY 136
SM   VAL 133 ; VAL 137
SM   VAL 134 ; ALA 135
SM   VAL 134 ; GLY 136
SM   VAL 134 ; VAL 137
SM   VAL 134 ; ALA 138
SM   ALA 135 ; GLY 136
SM   ALA 135 ; VAL 137
SM   ALA 135 ; ALA 138
SM   ALA 135 ; ASN 139
SM   GLY 136 ; VAL 137
SM   GLY 136 ; ALA 138
SM   GLY 136 ; ASN 139
SM   GLY 136 ; ALA 140
SM   VAL 137 ; ALA 138
SM   VAL 137 ; ASN 139
SM   VAL 137 ; ALA 140
SM   VAL 137 ; LEU 141
SM   ALA 138 ; ASN 139
SM   ALA 138 ; ALA 140
SM   ALA 138 ; LEU 141
SM   ALA 138 ; ALA 142
SM   ASN 139 ; ALA 140
SM   ASN 139 ; LEU 141
SM   ASN 139 ; ALA 142
SM   ASN 139 ; HIS 143
SM   ALA 140 ; LEU 141
SM   ALA 140 ; ALA 142
SM   ALA 140 ; HIS 143
SM   LEU 141 ; ALA 142
SM   LEU 141 ; HIS 143
SM   LEU 141 ; TYR 145
SM   ALA 142 ; HIS 143
SM   ALA 142 ; LYS 144
SM   ALA 142 ; TYR 145
SM   HIS 143 ; LYS 144
SM   HIS 143 ; TYR 145
SM   LYS 144 ; TYR 145
SM   LYS 144 ; HIS 146
SM   TYR 145 ; HIS 146
//
</pre>
</td></tr></table><p>
<p><h3>File: contacts.log</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
2hhb
1cs4
1ii7
</pre>
</td></tr></table><p>




<!-- DATA FILES         
     Any data files used (e.g. translation table file, substitution matrix 
     etc.  This includes example data file formats if they are not obvious.
     For a standard description of what data files are and how embossdata can
     be used to inspect and retrieve them, use:
     #include file="inc/localfiles.ihtml" 
-->

<a name="5.0"></a>
<br><br><br><H2> 5.0   DATA FILES                </H2>

CONTACTS uses a data file containing van der Waals radii for atoms in proteins (below).  The file Evdw.dat is such a data file and is part of the EMBOSS distribution.

<!-- USAGE
     Example usage, as run from the command-line.
     Many examples illustrating different behaviours is good.
-->
<a name="6.0"></a>
<br><br><br><H2> 6.0   USAGE                     </H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" >
<tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-cpdbdir]           dirlist    This option specifies the location of CCF
                                  files (clean coordinate files) (input). A
                                  'clean cordinate file' contains protein
                                  coordinate and derived data for a single PDB
                                  file ('protein clean coordinate file') or a
                                  single domain from SCOP or CATH ('domain
                                  clean coordinate file'), in CCF format
                                  (EMBL-like). The files, generated by using
                                  PDBPARSE (PDB files) or DOMAINER (domains),
                                  contain 'cleaned-up' data that is
                                  self-consistent and error-corrected. Records
                                  for residue solvent accessibility and
                                  secondary structure are added to the file by
                                  using PDBPLUS.
   -vdwfile            datafile   This option specifies the name of the data
                                  file with van der Waals radii of atoms for
                                  different amino acid residues.
   -threshold          float      Contact between two residues is defined as
                                  when the van der Waals surface of any atom
                                  of the first residue comes within the
                                  threshold contact distance of the van der
                                  Waals surface of any atom of the second
                                  residue. The threshold contact distance is a
                                  user-defined distance with a default value
                                  of 1 Angstrom.
  [-conoutdir]         outdir     This option specifies the location of CON
                                  files (contact files) (output). A 'contact
                                  file' contains contact data for a protein or
                                  a domain from SCOP or CATH, in the CON
                                  format (EMBL-like). The contacts may be
                                  intra-chain residue-residue, inter-chain
                                  residue-residue or residue-ligand. The files
                                  are generated by using CONTACTS, INTERFACE
                                  and SITES.
   -conlogfile         outfile    The log file contains messages about any
                                  errors arising while contacts ran.

   Additional (Optional) qualifiers:
   -[no]ccfnaming      boolean    This option specifies whether to use pdbid
                                  code to name the output files. If set, the
                                  PDB identifier code (from the PDB file) is
                                  used to name the file. Otherwise, the output
                                  files have the same names as the input
                                  files.
   -skip               boolean    Whether to calculate contacts between
                                  residue adjacent in sequence.
   -ignore             float      If any two atoms from two different residues
                                  are at least this distance apart then no
                                  futher inter-atomic contacts will be checked
                                  for for that residue pair . This speeds the
                                  calculation up considerably.

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-conlogfile" associated qualifiers
   -odirectory         string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report deaths

</pre>
</pre>
</td></tr>
</table>


<H3> 6.1   COMMAND LINE ARGUMENTS </H3>
 
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-cpdbdir]<br>(Parameter 1)</td>
<td>This option specifies the location of CCF files (clean coordinate files) (input). A 'clean cordinate file' contains protein coordinate and derived data for a single PDB file ('protein clean coordinate file') or a single domain from SCOP or CATH ('domain clean coordinate file'), in CCF format (EMBL-like). The files, generated by using PDBPARSE (PDB files) or DOMAINER (domains), contain 'cleaned-up' data that is self-consistent and error-corrected. Records for residue solvent accessibility and secondary structure are added to the file by using PDBPLUS.</td>
<td>Directory with files</td>
<td>./</td>
</tr>

<tr>
<td>-vdwfile</td>
<td>This option specifies the name of the data file with van der Waals radii of atoms for different amino acid residues.</td>
<td>Data file</td>
<td>Evdw.dat</td>
</tr>

<tr>
<td>-threshold</td>
<td>Contact between two residues is defined as when the van der Waals surface of any atom of the first residue comes within the threshold contact distance of the van der Waals surface of any atom of the second residue. The threshold contact distance is a user-defined distance with a default value of 1 Angstrom.</td>
<td>Any numeric value</td>
<td>1.0</td>
</tr>

<tr>
<td>[-conoutdir]<br>(Parameter 2)</td>
<td>This option specifies the location of CON files (contact files) (output). A 'contact file' contains contact data for a protein or a domain from SCOP or CATH, in the CON format (EMBL-like). The contacts may be intra-chain residue-residue, inter-chain residue-residue or residue-ligand. The files are generated by using CONTACTS, INTERFACE and SITES.</td>
<td>Output directory</td>
<td>./</td>
</tr>

<tr>
<td>-conlogfile</td>
<td>The log file contains messages about any errors arising while contacts ran.</td>
<td>Output file</td>
<td>contacts.log</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]ccfnaming</td>
<td>This option specifies whether to use pdbid code to name the output files. If set, the PDB identifier code (from the PDB file) is used to name the file. Otherwise, the output files have the same names as the input files.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-skip</td>
<td>Whether to calculate contacts between residue adjacent in sequence.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-ignore</td>
<td>If any two atoms from two different residues are at least this distance apart then no futher inter-atomic contacts will be checked for for that residue pair . This speeds the calculation up considerably.</td>
<td>Any numeric value</td>
<td>20.0</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<H3> 6.2 EXAMPLE SESSION </H3>
An example of interactive use of CONTACTS is shown below.

<b>Here is a sample session with contacts</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>contacts </b>
Generate intra-chain CON files from CCF files.
Location of CCF files (clean coordinate files) (input) [./]: <b>../pdbplus-keep/</b>
Name of data file with van der Waals radii [Evdw.dat]: <b></b>
Threshold contact distance [1.0]: <b>1</b>
Location of CON files (contact files) (output) [./]: <b></b>
Name of log file for the build [contacts.log]: <b></b>

2hhb
1cs4
1ii7

</pre></td></tr></table><p>
<p>
<a href="#output.1">Go to the output files for this example</a><p><p>


<!-- <br><br>
All CCF files (with a file extension specified in the ACD file) were read from testdata and the contact files (e.g. d1cs4a_.con and d1ii7a_.con) were written to the same directory (the file extension .con would be set in the ACD file).  van der Waals radii were taken from the emboss data file Evdw.dat and the default contact distance of 1 Angstrom was used to determine contacts.  The log file testdata/contacts.log was written.
<br><br>


<!-- KNOWN BUGS & WARNINGS
     Bugs that have not yet been fixed, easily missued features, problems
     and caveats etc. Potentially stupid things the program will let you do.
-->
<a name="7.0"></a>
<br><br><br><H2> 7.0   KNOWN BUGS & WARNINGS     </H2>
None.








<!-- NOTES
     Important general remarks, including:
       Restrictions.
       Interesting behaviour.
       Useful things you can do with this program.
       Future plans.
       etc.
-->
<a name="8.0"></a>
<br><br><br><H2> 8.0   NOTES                     </H2>
<b> Types of contact </b>
<br><b>SM</b> records are used for contacts between either either side-chain or main-chain atoms 
as defined above.  In a future implementation, <b>SS</b> will be used for side-chain only contacts,
 <b>MM</b> will be used for main-chain only contacts, <b>LI</b> will be used for contacts to
ligands, and there will probably be several other forms of contact too.

<br><br><b>Threshold ignore distance option</b>
<br>The threshold ignore distance can be adjusted to speed-up the calculation of contacts considerably.  If any two atoms from two different residues are at least this distance apart then no futher inter-atomic contacts will be checked for for that residue pair. 

<br><br><b>Skip option</b>
<br>Two residues that are adjacent in the sequence will always be in contact through their peptide bond, therefore it is not normally necessary to calculate such contacts directly.  Furthermore, some analyses require such contacts to be omitted.  The skip option determines whether contacts between residue adjacent in sequence are calculated and written to the output file.




<H3> 8.1   GLOSSARY OF FILE TYPES    </H3>
<table BORDER CELLSPACING=5 CELLPADDING=5 BGCOLOR="#f5f5ff" >
<tr>
<td><b>FILE TYPE</b></td>
<td><b>FORMAT</b></td>
<td><b>DESCRIPTION</b></td>
<td><b>CREATED BY <b></td>
<td><b>SEE ALSO</b></td>
</tr>
<tr>
<td><b> Clean coordinate file (for protein) </b></td>
<td> CCF format (EMBL-like).</td>
<td> Protein coordinate and derived data for a single PDB file.  The data are 'cleaned-up': self-consistent and error-corrected.  </td>
<td> <a href="pdbparse.html">PDBPARSE</a> </td>
<td> Records for residue solvent accessibility and secondary structure are added to the file by using PDBPLUS. </td>
</tr>
<tr>
<td><b> Clean coordinate file (for domain)</b></td>
<td> CCF format (EMBL-like).</td>
<td> Protein coordinate and derived data for a single domain from SCOP or CATH. The data are 'cleaned-up': self-consistent and error-corrected. </td>
<td> <a href="domainer.html">DOMAINER</a></td>
<td> Records for residue solvent accessibility and secondary structure are added to the file by using PDBPLUS.</td>
</tr>
<tr>
<td><b> Contact file (intra-chain residue-residue contacts)</b></td>
<td> CON format (EMBL-like.) </td>
<td> Intra-chain residue-residue contact data for a protein or a domain from SCOP or CATH.  </td>
<td> <a href="contacts.html">CONTACTS</a> </td>
<td> N.A. </td>
</tr>
<tr>
<td><b> Contact file (inter-chain residue-residue contacts)</b></td>
<td> CON format (EMBL-like.) </td>
<td> Inter-chain residue-residue contact data for a protein or a domain from SCOP or CATH.  </td>
<td> <a href="interface.html">INTERFACE</a> </td>
<td> N.A. </td>
</tr>
<tr>
<td><b> Contact file (residue-ligand contacts)</b></td>
<td> CON format (EMBL-like.) </td>
<td> Residue-ligand contact data for a protein or a domain from SCOP or CATH. </td>
<td> <a href="sites.html">SITES</a> </td>
<td> N.A. </td>
</tr>
<tr>
<td><b>van der Waals radii  </b></td>
<td> A file of van der Waals radii for atoms in amino acid residues. Part of the emboss distribution.</td>
<td> N.A. </td>
<td> N.A. </td>
</tr>
</table>





None


<!-- DESCRIPTION
     A complete, non-technical, user-level description of the application.
-->
<a name="9.0"></a>
<br><br><br><H2> 9.0   DESCRIPTION               </H2>
Knowledge of the physical contacts that amino acid residues within a protein or domain make with one another is required for several different analyses.  CONTACTS calculates intra-chain residue-residue contact data from protein and domain CCF files (clean coordinate files).


<!-- ALGORITHM
     A technical description of algorithmic aspects, describing exactly how
     the key aspects of the application work.
-->
<a name="10.0"></a>
<br><br><br><H2> 10.0  ALGORITHM                 </H2>
Contact between two residues is defined when the van der Waals surface of any atom of the first residue comes within the threshold contact distance of the van der Waals surface of any atom of the second residue. The threshold contact distance is a user-defined distance with a default value of 1 Angstrom.



<!-- RELATED APPLICATIONS
     Other applications that either generate the input, use the output or 
     are in some other way related to the application are described here. 
-->

<a name="11.0"></a>
<br><br><br><H2> 11.0  RELATED APPLICATIONS      </H2>

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="../../apps/contactcount.html">contactcount</a></td><td>Count specific versus non-specific contacts</td></tr>
<tr><td><a href="../domalign/domainalign.html">domainalign</a></td><td>Generate alignments (DAF file) for nodes in a DCF file</td></tr>
<tr><td><a href="../domalign/domainrep.html">domainrep</a></td><td>Reorder DCF file to identify representative structures</td></tr>
<tr><td><a href="../domainatrix/domainreso.html">domainreso</a></td><td>Remove low resolution domains from a DCF file</td></tr>
<tr><td><a href="interface.html">interface</a></td><td>Generate inter-chain CON files from CCF files</td></tr>
<tr><td><a href="../signature/libgen.html">libgen</a></td><td>Generate discriminating elements from alignments</td></tr>
<tr><td><a href="../signature/matgen3d.html">matgen3d</a></td><td>Generate a 3D-1D scoring matrix from CCF files</td></tr>
<tr><td><a href="../../apps/psiphi.html">psiphi</a></td><td>Phi and psi torsion angles from protein coordinates</td></tr>
<tr><td><a href="../signature/rocon.html">rocon</a></td><td>Generates a hits file from comparing two DHF files</td></tr>
<tr><td><a href="../signature/rocplot.html">rocplot</a></td><td>Performs ROC analysis on hits files</td></tr>
<tr><td><a href="../reconstruct/scorecmapdir.html">scorecmapdir</a></td><td>Contact scores for cleaned protein chain contact files</td></tr>
<tr><td><a href="../domalign/seqalign.html">seqalign</a></td><td>Extend alignments (DAF file) with sequences (DHF file)</td></tr>
<tr><td><a href="../domsearch/seqfraggle.html">seqfraggle</a></td><td>Removes fragment sequences from DHF files</td></tr>
<tr><td><a href="../domsearch/seqsearch.html">seqsearch</a></td><td>Generate PSI-BLAST hits (DHF file) from a DAF file</td></tr>
<tr><td><a href="../domsearch/seqsort.html">seqsort</a></td><td>Remove ambiguous classified sequences from DHF files</td></tr>
<tr><td><a href="../domsearch/seqwords.html">seqwords</a></td><td>Generates DHF files from keyword search of UniProt</td></tr>
<tr><td><a href="../signature/siggen.html">siggen</a></td><td>Generates a sparse protein signature from an alignment</td></tr>
<tr><td><a href="../signature/siggenlig.html">siggenlig</a></td><td>Generate ligand-binding signatures from a CON file</td></tr>
<tr><td><a href="../signature/sigscan.html">sigscan</a></td><td>Generate hits (DHF file) from a signature search</td></tr>
<tr><td><a href="../signature/sigscanlig.html">sigscanlig</a></td><td>Search ligand-signature library & write hits (LHF file)</td></tr>
</table>



<!-- DIAGNOSTIC ERROR MESSAGES
     Description of error messages or log file, if one is written.
-->
<a name="12.0"></a>
<br><br><br><H2> 12.0  DIAGNOSTIC ERROR MESSAGES </H2>
CONTACTS generates a log file an excerpt of which is shown (Figure 2). If there is a problem in processing a CCF file, three lines containing the record '//', the domain or pdb identifier code and an error message respectively are written. The text <i>'WARN  file open error filename'</i>, <i>'ERROR file read error filename'</i> or <i>'ERROR file write error filename'</i> will be reported when an error was encountered during a file open, read or write respectively.  Various other error messages may also be given (in case of difficulty email jison@hgmp.mrc.ac.uk).



<!-- AUTHORS
     The main author first then all others. 
-->
<a name="13.0"></a>
<br><br><br><H2> 13.0  AUTHORS                   </H2>
Jon Ison <a href="mailto:jison@rfcgr.mrc.ac.uk">(jison@rfcgr.mrc.ac.uk)</a>
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK




<!-- REFERENCES
     Quote the paper where the application was first published, described, used etc. 
-->
<a name="14.0"></a>
<br><br><br><H2> 14.0  REFERENCES                </H2>

Please cite the authors and EMBOSS.
<br><br>
<i>Rice P, Longden I and Bleasby A (2000) "EMBOSS - The European
Molecular Biology Open Software Suite"  Trends in Genetics,
15:276-278.

<p>
See also <a href="http://emboss.sourceforge.net/">http://emboss.sourceforge.net/</a></i>




<H3>14.1 Other useful references  </H3>


</BODY>
</HTML>



