<HTML>

<HEAD>
  <TITLE>
  EMBOSS: cpgplot
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
cpgplot
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
    Function
</H2>
Plot CpG rich areas
<H2>
    Description
</H2>

CpG refers to a C nucleotide immediately followed by a G.  The 'p' in
'CpG' refers to the phosphate group linking the two bases. 

<p>
Detection of regions of genomic sequences that are rich in the CpG
pattern is important because such regions are resistant to methylation
and tend to be associated with genes which are frequently switched on. 
Regions rich in the CpG pattern are known as CpG islands.  

<p>
It has been estimated that about half of all mammalian genes have a
CpG-rich region around their 5' end.  It is said that all mammalian
house-keeping genes have a CpG island!

<p>
Non-mammalian vertebrates have some CpG islands that are associated with
genes, but the association gets equivocal in the farther taxonomic
groups. 

<p>
Finding a CpG island upstream of predicted exons or genes is good
contributory evidence. 

<p>

By default, this program defines a CpG island as a region where, over an
average of 10 windows, the calculated % composition is over 50% and the
calculated Obs/Exp (i.e.  Observed/Expected) ratio is over 0.6 and the
conditions hold for a minimum of 200 bases.  These conditions can be
modified by setting the values of the appropriate parameters. 

<p>


The Observed number of CpG patterns in a window is simply the count of
the number of times a 'C' is found followed immediately by a 'G'. 

<p>

The Expected number of CpG patterns is calculated for each window as the
number of CpG dinucleotides you would expect to see in that window based
on the frequency of C's and G's in that window.  Thus, the Expected
frequency of CpG's in a window is calculated as the number of 'C's in
the window multiplied by the number of 'G's in the window, divided by
the window length. 

<p>

<pre>
	Expected = (number of C's  * number of G's) / window length


</pre>

<p>

This program reads in one or more sequences and calculates the Obs/Exp
ratio, the percentage CpG over a window which is moved along the
sequence.  These calculated values can be plotted, together with the
regions which match this program's definition of a CpG island. 

<H2>
    Usage
</H2>
<b>Here is a sample session with cpgplot</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>cpgplot tembl:rnu68037 -graph cps </b>
Plot CpG rich areas
Window size [100]: <b></b>
Minimum length of an island [200]: <b></b>
Minimum observed/expected [0.6]: <b></b>
Minimum percentage [50.]: <b></b>
Output file [rnu68037.cpgplot]: <b></b>
Features output [rnu68037.gff]: <b></b>

Created cpgplot.ps

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>

<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-sequence]          seqall     Nucleotide sequence(s) filename and optional
                                  format, or reference (input USA)
   -window             integer    [100] The percentage CG content and the
                                  Observed frequency of CG is calculated
                                  within a window whose size is set by this
                                  parameter. The window is moved down the
                                  sequence and these statistics are calculated
                                  at each postition that the window is moved
                                  to. (Integer 1 or more)
   -minlen             integer    [200] This sets the minimum length that a
                                  CpG island has to be before it is reported.
                                  (Integer 1 or more)
   -minoe              float      [0.6] This sets the minimum average observed
                                  to expected ratio of C plus G to CpG in a
                                  set of 10 windows that are required before a
                                  CpG island is reported. (Number from 0.000
                                  to 10.000)
   -minpc              float      [50.] This sets the minimum average
                                  percentage of G plus C a set of 10 windows
                                  that are required before a CpG island is
                                  reported. (Number from 0.000 to 100.000)
  [-outfile]           outfile    [*.cpgplot] This sets the name of the file
                                  holding the report of the input sequence
                                  name, CpG island parameters and the output
                                  details of any CpG islands that are found.
  [-graph]             xygraph    [$EMBOSS_GRAPHICS value, or x11] Graph type
                                  (ps, hpgl, hp7470, hp7580, meta, cps, x11,
                                  tekt, tek, none, data, xterm, png)
  [-outfeat]           featout    [unknown.gff] File for output features

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers:
   -[no]obsexp         boolean    [Y] If this is set to true then the graph of
                                  the observed to expected ratio of C plus G
                                  to CpG within a window is displayed.
   -[no]cg             boolean    [Y] If this is set to true then the graph of
                                  the regions which have been determined to
                                  be CpG islands is displayed.
   -[no]pc             boolean    [Y] If this is set to true then the graph of
                                  the percentage C plus G within a window is
                                  displayed.

   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1            integer    Start of each sequence to be used
   -send1              integer    End of each sequence to be used
   -sreverse1          boolean    Reverse (if DNA)
   -sask1              boolean    Ask for begin/end/reverse
   -snucleotide1       boolean    Sequence is nucleotide
   -sprotein1          boolean    Sequence is protein
   -slower1            boolean    Make lower case
   -supper1            boolean    Make upper case
   -sformat1           string     Input sequence format
   -sdbname1           string     Database name
   -sid1               string     Entryname
   -ufo1               string     UFO features
   -fformat1           string     Features format
   -fopenfile1         string     Features file name

   "-outfile" associated qualifiers
   -odirectory2        string     Output directory

   "-graph" associated qualifiers
   -gprompt3           boolean    Graph prompting
   -gtitle3            string     Graph title
   -gsubtitle3         string     Graph subtitle
   -gxtitle3           string     Graph x axis title
   -gytitle3           string     Graph y axis title
   -goutfile3          string     Output file for non interactive displays
   -gdirectory3        string     Output directory

   "-outfeat" associated qualifiers
   -offormat4          string     Output feature format
   -ofopenfile4        string     Features file name
   -ofextension4       string     File name extension
   -ofdirectory4       string     Output directory
   -ofname4            string     Base file name
   -ofsingle4          boolean    Separate file for each entry

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Nucleotide sequence(s) filename and optional format, or reference (input USA)</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-window</td>
<td>The percentage CG content and the Observed frequency of CG is calculated within a window whose size is set by this parameter. The window is moved down the sequence and these statistics are calculated at each postition that the window is moved to.</td>
<td>Integer 1 or more</td>
<td>100</td>
</tr>

<tr>
<td>-minlen</td>
<td>This sets the minimum length that a CpG island has to be before it is reported.</td>
<td>Integer 1 or more</td>
<td>200</td>
</tr>

<tr>
<td>-minoe</td>
<td>This sets the minimum average observed to expected ratio of C plus G to CpG in a set of 10 windows that are required before a CpG island is reported.</td>
<td>Number from 0.000 to 10.000</td>
<td>0.6</td>
</tr>

<tr>
<td>-minpc</td>
<td>This sets the minimum average percentage of G plus C a set of 10 windows that are required before a CpG island is reported.</td>
<td>Number from 0.000 to 100.000</td>
<td>50.</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>This sets the name of the file holding the report of the input sequence name, CpG island parameters and the output details of any CpG islands that are found.</td>
<td>Output file</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>[-graph]<br>(Parameter 3)</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including ps, hpgl, hp7470, hp7580, meta, cps, x11, tekt, tek, none, data, xterm, png</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr>
<td>[-outfeat]<br>(Parameter 4)</td>
<td>File for output features</td>
<td>Writeable feature table</td>
<td><i>unknown.gff</i></td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]obsexp</td>
<td>If this is set to true then the graph of the observed to expected ratio of C plus G to CpG within a window is displayed.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]cg</td>
<td>If this is set to true then the graph of the regions which have been determined to be CpG islands is displayed.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]pc</td>
<td>If this is set to true then the graph of the percentage C plus G within a window is displayed.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

</table>

<H2>
    Input file format
</H2>

Any DNA sequence USA.

<p>

<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:rnu68037' is a sequence entry in the example nucleic acid database 'tembl'
<p>
<p><h3>Database entry: tembl:rnu68037</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   RNU68037   standard; RNA; ROD; 1218 BP.
XX
AC   U68037;
XX
SV   U68037.1
XX
DT   23-SEP-1996 (Rel. 49, Created)
DT   04-MAR-2000 (Rel. 63, Last updated, Version 2)
XX
DE   Rattus norvegicus EP1 prostanoid receptor mRNA, complete cds.
XX
KW   .
XX
OS   Rattus norvegicus (Norway rat)
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia;
OC   Eutheria; Rodentia; Sciurognathi; Muridae; Murinae; Rattus.
XX
RN   [1]
RP   1-1218
RA   Abramovitz M., Boie Y.;
RT   "Cloning of the rat EP1 prostanoid receptor";
RL   Unpublished.
XX
RN   [2]
RP   1-1218
RA   Abramovitz M., Boie Y.;
RT   ;
RL   Submitted (26-AUG-1996) to the EMBL/GenBank/DDBJ databases.
RL   Biochemistry &amp; Molecular Biology, Merck Frosst Center for Therapeutic
RL   Research, P. O. Box 1005, Pointe Claire - Dorval, Quebec H9R 4P8, Canada
XX
DR   SWISS-PROT; P70597; PE21_RAT.
XX
FH   Key             Location/Qualifiers
FH
FT   source          1..1218
FT                   /db_xref="taxon:10116"
FT                   /organism="Rattus norvegicus"
FT                   /strain="Sprague-Dawley"
FT   CDS             1..1218
FT                   /codon_start=1
FT                   /db_xref="SWISS-PROT:P70597"
FT                   /note="family 1 G-protein coupled receptor"
FT                   /product="EP1 prostanoid receptor"
FT                   /protein_id="AAB07735.1"
FT                   /translation="MSPYGLNLSLVDEATTCVTPRVPNTSVVLPTGGNGTSPALPIFSM
FT                   TLGAVSNVLALALLAQVAGRLRRRRSTATFLLFVASLLAIDLAGHVIPGALVLRLYTAG
FT                   RAPAGGACHFLGGCMVFFGLCPLLLGCGMAVERCVGVTQPLIHAARVSVARARLALALL
FT                   AAMALAVALLPLVHVGHYELQYPGTWCFISLGPPGGWRQALLAGLFAGLGLAALLAALV
FT                   CNTLSGLALLRARWRRRRSRRFRENAGPDDRRRWGSRGLRLASASSASSITSTTAALRS
FT                   SRGGGSARRVHAHDVEMVGQLVGIMVVSCICWSPLLVLVVLAIGGWNSNSLQRPLFLAV
FT                   RLASWNQILDPWVYILLRQAMLRQLLRLLPLRVSAKGGPTELSLTKSAWEASSLRSSRH
FT                   SGFSHL"
XX
SQ   Sequence 1218 BP; 162 A; 397 C; 387 G; 272 T; 0 other;
     atgagcccct acgggcttaa cctgagccta gtggatgagg caacaacgtg tgtaacaccc        60
     agggtcccca atacatctgt ggtgctgcca acaggcggta acggcacatc accagcgctg       120
     cctatcttct ccatgacgct gggtgctgtg tccaacgtgc tggcgctggc gctgctggcc       180
     caggttgcag gcagactgcg gcgccgccgc tcgactgcca ccttcctgtt gttcgtcgcc       240
     agcctgcttg ccatcgacct agcaggccat gtgatcccgg gcgccttggt gcttcgcctg       300
     tatactgcag gacgtgcgcc cgctggcggg gcctgtcatt tcctgggcgg ctgtatggtc       360
     ttctttggcc tgtgcccact tttgcttggc tgtggcatgg ccgtggagcg ctgcgtgggt       420
     gtcacgcagc cgctgatcca cgcggcgcgc gtgtccgtag cccgcgcacg cctggcacta       480
     gccctgctgg ccgccatggc tttggcagtg gcgctgctgc cactagtgca cgtgggtcac       540
     tacgagctac agtaccctgg cacttggtgt ttcattagcc ttgggcctcc tggaggttgg       600
     cgccaggcgt tgcttgcggg cctcttcgcc ggccttggcc tggctgcgct ccttgccgca       660
     ctagtgtgta atacgctcag cggcctggcg ctccttcgtg cccgctggag gcggcgtcgc       720
     tctcgacgtt tccgagagaa cgcaggtccc gatgatcgcc ggcgctgggg gtcccgtgga       780
     ctccgcttgg cctccgcctc gtctgcgtca tccatcactt caaccacagc tgccctccgc       840
     agctctcggg gaggcggctc cgcgcgcagg gttcacgcac acgacgtgga aatggtgggc       900
     cagctcgtgg gcatcatggt ggtgtcgtgc atctgctgga gccccctgct ggtattggtg       960
     gtgttggcca tcgggggctg gaactctaac tccctgcagc ggccgctctt tctggctgta      1020
     cgcctcgcgt cgtggaacca gatcctggac ccatgggtgt acatcctgct gcgccaggct      1080
     atgctgcgcc aacttcttcg cctcctaccc ctgagggtta gtgccaaggg tggtccaacg      1140
     gagctgagcc taaccaagag tgcctgggag gccagttcac tgcgtagctc ccggcacagt      1200
     ggcttcagcc acttgtga                                                    1218
//
</pre>
</td></tr></table><p>

<H2>
    Output file format
</H2>

<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: rnu68037.cpgplot</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>


CPGPLOT islands of unusual CG composition
RNU68037 from 1 to 1218

     Observed/Expected ratio &gt; 0.60
     Percent C + Percent G &gt; 50.00
     Length &gt; 200

 Length 406 (104..509)

 Length 329 (596..924)
</pre>
</td></tr></table><p>
<p><h3>File: rnu68037.gff</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
##gff-version 2.0
##date 2006-07-15
##Type DNA RNU68037
RNU68037	cpgplot	misc_feature	104	509	0.000	+	.	Sequence "RNU68037.1"
RNU68037	cpgplot	misc_feature	596	924	0.000	+	.	Sequence "RNU68037.2"
</pre>
</td></tr></table><p>
<p><h3>Graphics File: cpgplot.ps</h3>
<p><img src="cpgplot.1.cpgplot.gif" alt="[cpgplot results]">


<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

The original program was described in:

<ol>

<li>
Larsen F, Gundersen G, Lopez R, Prydz H
"CpG islands as gene markers in the human genome."
Genomics 1992 Aug;13(4):1095-107

</ol>

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

    0 if successful.

<H2>
    Known bugs
</H2>

None.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr>
<td><a href="cpgreport.html">cpgreport</a></td>
<td>Reports all CpG rich regions</td>
</tr>

<tr>
<td><a href="geecee.html">geecee</a></td>
<td>Calculates fractional GC content of nucleic acid sequences</td>
</tr>

<tr>
<td><a href="newcpgreport.html">newcpgreport</a></td>
<td>Report CpG rich areas</td>
</tr>

<tr>
<td><a href="newcpgseek.html">newcpgseek</a></td>
<td>Reports CpG rich regions</td>
</tr>

</table>


<p>

As there is no official definition of what is a cpg island is, and worst
where they begin and end, we have to live with 2 definitions and thus
two methods.  These are:

<p>

1.  <b>newcpgseek</b> and <b>cpgreport</b> - both declare a putative
island if the score is higher than a threshold (17 at the moment).  They
now also displaying the actual CpG count, the % CG and the
observed/expected ration in the region where the score is above the
threshold.  This scoring method based on sum/frequencies overpredicts
islands but finds the smaller ones around primary exons.
<b>newcpgseek</b> uses the same method as <b>cpgreport</b> but the
output is different and more readable.

<p>

2.  <b>newcpgreport</b> and <b>cpgplot</b> use a sliding window within
which the Obs/Exp ratio of CpG is calculated.  The important thing to
note in this method is that an island, in order to be reported, is
defined as a region that satisfies the following contraints:

<p>
<pre>
   Obs/Exp ratio > 0.6
   % C + % G > 50%
   Length > 200.
</pre>
<p>

<p>

For all practical purposes you should probably use newcpgreport.  It is
actually used to produce the human cpgisland database you can find on
the EBI's ftp server as well as on the EBI's SRS server.

<p>

<b>geecee</b> measures CG content in the entire input sequence and is
not to be used to detect CpG islands.  It can be usefull for detecting
sequences that MIGHT contain an island.
        

<H2>
    Author(s)
</H2>
Alan Bleasby (ajb&nbsp;&copy;&nbsp;ebi.ac.uk)
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK


<H2>
    History
</H2>

Completed 23rd March 1999.

<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.

<H2>
    Comments
</H2>
None

</BODY>
</HTML>

