<HTML>

<HEAD>
  <TITLE>
  EMBOSS: cpgreport
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

cpgreport
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>






<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Reports all CpG rich regions


<H2>
    Description
</H2>


cpgreport scans a nucleotide sequence for regions with higher than
expected frequencies of the dinucleotide CG. 

<p>
CpG refers to a C nucleotide immediately followed by a G.  The 'p' in
'CpG' refers to the phosphate group linking the two bases. 

<p>
Detection of regions of genomic sequences that are rich in the CpG
pattern is important because such regions are resistant to methylation
and tend to be associated with genes which are frequently switched on. 
Regions rich in the CpG pattern are known as CpG islands. 

<p>

This program does not find CpG islands as normally defined: "a region of
greater than 200 bp with a %GC of greater than 50% and observed/expected
CpG > 0.6".  <b>cpgreport</b> instead uses a running sum rather than a
window to create the score as follows: if not CpG at position i, then
decrement running-Sum counter, but if CpG then running-Sum counter is
incremented by the CPGSCORE.  Spans greater than the threshold are
searched for recursively. 

<p>

This method overpredicts islands but finds the smaller ones around
primary exons. 


<H2>
    Usage
</H2>


<b>Here is a sample session with cpgreport</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>cpgreport tembl:rnu68037 </b>
Reports all CpG rich regions
CpG score [17]: <b></b>
Output file [rnu68037.cpgreport]: <b></b>
Output features [rnu68037.gff]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>


<!--
Here is a sample session with cpgreport.

<pre>

% cpgreport embl:rnu68037
Reports CpG rich regions
CpG score [17]: 
Output file [rnu68037.cpgreport]: 

</pre>
-->


<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-sequence]          seqall     Sequence database USA
   -score              integer    This sets the score for each CG sequence
                                  found. A value of 17 is more sensitive, but
                                  28 has also been used with some success.
  [-outfile]           outfile    Output file name
  [-featout]           featout    File for output features

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1             integer    First base used
   -send1               integer    Last base used, def=seq length
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name

   "-outfile" associated qualifiers
   -odirectory2         string     Output directory

   "-featout" associated qualifiers
   -offormat3           string     Output feature format
   -ofopenfile3         string     Features file name
   -ofextension3        string     File name extension
   -ofdirectory3        string     Output directory
   -ofname3             string     Base file name
   -ofsingle3           boolean    Separate file for each entry

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-score</td>
<td>This sets the score for each CG sequence found. A value of 17 is more sensitive, but 28 has also been used with some success.</td>
<td>Integer from 1 to 200</td>
<td>17</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.cpgreport</td>
</tr>

<tr>
<td>[-featout]<br>(Parameter 3)</td>
<td>File for output features</td>
<td>Writeable feature table</td>
<td><i>unknown.gff</i></td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>

Any DNA sequence USA.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:rnu68037' is a sequence entry in the example nucleic acid database 'tembl'
<p>
<p><h3>Database entry: tembl:rnu68037</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   RNU68037   standard; RNA; ROD; 1218 BP.
XX
AC   U68037;
XX
SV   U68037.1
XX
DT   23-SEP-1996 (Rel. 49, Created)
DT   04-MAR-2000 (Rel. 63, Last updated, Version 2)
XX
DE   Rattus norvegicus EP1 prostanoid receptor mRNA, complete cds.
XX
KW   .
XX
OS   Rattus norvegicus (Norway rat)
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia;
OC   Eutheria; Rodentia; Sciurognathi; Muridae; Murinae; Rattus.
XX
RN   [1]
RP   1-1218
RA   Abramovitz M., Boie Y.;
RT   "Cloning of the rat EP1 prostanoid receptor";
RL   Unpublished.
XX
RN   [2]
RP   1-1218
RA   Abramovitz M., Boie Y.;
RT   ;
RL   Submitted (26-AUG-1996) to the EMBL/GenBank/DDBJ databases.
RL   Biochemistry &amp; Molecular Biology, Merck Frosst Center for Therapeutic
RL   Research, P. O. Box 1005, Pointe Claire - Dorval, Quebec H9R 4P8, Canada
XX
DR   SWISS-PROT; P70597; PE21_RAT.
XX
FH   Key             Location/Qualifiers
FH
FT   source          1..1218
FT                   /db_xref="taxon:10116"
FT                   /organism="Rattus norvegicus"
FT                   /strain="Sprague-Dawley"
FT   CDS             1..1218
FT                   /codon_start=1
FT                   /db_xref="SWISS-PROT:P70597"
FT                   /note="family 1 G-protein coupled receptor"
FT                   /product="EP1 prostanoid receptor"
FT                   /protein_id="AAB07735.1"
FT                   /translation="MSPYGLNLSLVDEATTCVTPRVPNTSVVLPTGGNGTSPALPIFSM
FT                   TLGAVSNVLALALLAQVAGRLRRRRSTATFLLFVASLLAIDLAGHVIPGALVLRLYTAG
FT                   RAPAGGACHFLGGCMVFFGLCPLLLGCGMAVERCVGVTQPLIHAARVSVARARLALALL
FT                   AAMALAVALLPLVHVGHYELQYPGTWCFISLGPPGGWRQALLAGLFAGLGLAALLAALV
FT                   CNTLSGLALLRARWRRRRSRRFRENAGPDDRRRWGSRGLRLASASSASSITSTTAALRS
FT                   SRGGGSARRVHAHDVEMVGQLVGIMVVSCICWSPLLVLVVLAIGGWNSNSLQRPLFLAV
FT                   RLASWNQILDPWVYILLRQAMLRQLLRLLPLRVSAKGGPTELSLTKSAWEASSLRSSRH
FT                   SGFSHL"
XX
SQ   Sequence 1218 BP; 162 A; 397 C; 387 G; 272 T; 0 other;
     atgagcccct acgggcttaa cctgagccta gtggatgagg caacaacgtg tgtaacaccc        60
     agggtcccca atacatctgt ggtgctgcca acaggcggta acggcacatc accagcgctg       120
     cctatcttct ccatgacgct gggtgctgtg tccaacgtgc tggcgctggc gctgctggcc       180
     caggttgcag gcagactgcg gcgccgccgc tcgactgcca ccttcctgtt gttcgtcgcc       240
     agcctgcttg ccatcgacct agcaggccat gtgatcccgg gcgccttggt gcttcgcctg       300
     tatactgcag gacgtgcgcc cgctggcggg gcctgtcatt tcctgggcgg ctgtatggtc       360
     ttctttggcc tgtgcccact tttgcttggc tgtggcatgg ccgtggagcg ctgcgtgggt       420
     gtcacgcagc cgctgatcca cgcggcgcgc gtgtccgtag cccgcgcacg cctggcacta       480
     gccctgctgg ccgccatggc tttggcagtg gcgctgctgc cactagtgca cgtgggtcac       540
     tacgagctac agtaccctgg cacttggtgt ttcattagcc ttgggcctcc tggaggttgg       600
     cgccaggcgt tgcttgcggg cctcttcgcc ggccttggcc tggctgcgct ccttgccgca       660
     ctagtgtgta atacgctcag cggcctggcg ctccttcgtg cccgctggag gcggcgtcgc       720
     tctcgacgtt tccgagagaa cgcaggtccc gatgatcgcc ggcgctgggg gtcccgtgga       780
     ctccgcttgg cctccgcctc gtctgcgtca tccatcactt caaccacagc tgccctccgc       840
     agctctcggg gaggcggctc cgcgcgcagg gttcacgcac acgacgtgga aatggtgggc       900
     cagctcgtgg gcatcatggt ggtgtcgtgc atctgctgga gccccctgct ggtattggtg       960
     gtgttggcca tcgggggctg gaactctaac tccctgcagc ggccgctctt tctggctgta      1020
     cgcctcgcgt cgtggaacca gatcctggac ccatgggtgt acatcctgct gcgccaggct      1080
     atgctgcgcc aacttcttcg cctcctaccc ctgagggtta gtgccaaggg tggtccaacg      1140
     gagctgagcc taaccaagag tgcctgggag gccagttcac tgcgtagctc ccggcacagt      1200
     ggcttcagcc acttgtga                                                    1218
//
</pre>
</td></tr></table><p>


<H2>
    Output file format
</H2>



<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: rnu68037.cpgreport</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>


CPGREPORT of RNU68037 from 1 to 1218

Sequence              Begin    End Score        CpG   %CG  CG/GC
RNU68037                 12     13    17          1 100.0    -
RNU68037                 47     48    17          1 100.0    -
RNU68037                 96   1032   630         87  66.1   0.65
RNU68037               1072   1100    26          3  62.1   0.00
RNU68037               1139   1140    17          1 100.0    -
RNU68037               1183   1193    26          2  72.7   2.00
</pre>
</td></tr></table><p>
<p><h3>File: rnu68037.gff</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
##gff-version 2.0
##date 2003-11-27
##Type DNA RNU68037
RNU68037	cpgreport	misc_feature	12	13	17.000	+	.	Sequence "RNU68037.1"
RNU68037	cpgreport	misc_feature	47	48	17.000	+	.	Sequence "RNU68037.2"
RNU68037	cpgreport	misc_feature	96	1032	630.000	+	.	Sequence "RNU68037.3"
RNU68037	cpgreport	misc_feature	1072	1100	26.000	+	.	Sequence "RNU68037.4"
RNU68037	cpgreport	misc_feature	1139	1140	17.000	+	.	Sequence "RNU68037.5"
RNU68037	cpgreport	misc_feature	1183	1193	26.000	+	.	Sequence "RNU68037.6"
</pre>
</td></tr></table><p>


<!--
The output file from the above example (rnu68037.cpgreport) follows: 

<pre>

CPGREPORT of RNU68037 from 1 to 1218

Sequence              Begin    End Score        CpG   %CG  CG/GC
RNU68037                 12     13    17          1 100.0    -
RNU68037                 47     48    17          1 100.0    -
RNU68037                 96   1032   630         87  66.1   0.65
RNU68037               1072   1100    26          3  62.1   0.00
RNU68037               1139   1140    17          1 100.0    -
RNU68037               1183   1193    26          2  72.7   2.00

</pre>
-->

<p>

The first non-blank line of the output file 'rnu68037.cpgreport' is the
title line giving the program name, the name of sequence being analysed
and the start and end positions of the sequence. 

<p>

The second non-blank line contains the headings of the columns.

<p>

Subsequent lines contain columns with the following information:

<p>

<ul>
<li>The name of the sequence.
<li>The begin position and the end position of the CpG-rich region.
<li>The score of the CpG-rich region.
<li>The number of CpG's in the CpG-rich region.
<li>The %(G+C) in the CpG-rich region.
<li>The ratio of CpG to GpC in the CpG-rich region.
</ul>

<p>

If the count of GpC in the region is zero, then the ratio of CG/GC is
reported as '-'. 

<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>


This program does not find CpG islands as normally defined (see cpgplot).

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

    0 if successful.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="cpgplot.html">cpgplot</a></td><td>Plot CpG rich areas</td></tr>
<tr><td><a href="geecee.html">geecee</a></td><td>Calculates the fractional GC content of nucleic acid sequences</td></tr>
<tr><td><a href="newcpgreport.html">newcpgreport</a></td><td>Report CpG rich areas</td></tr>
<tr><td><a href="newcpgseek.html">newcpgseek</a></td><td>Reports CpG rich regions</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->

<p>

<p>

As there is no official definition of what is a cpg island is, and worst
where they begin and end, we have to live with 2 definitions and thus
two methods.  These are:

<p>

1.  <b>newcpgseek</b> and <b>cpgreport</b> - both declare a putative
island if the score is higher than a threshold (17 at the moment).  They
now also displaying the actual CpG count, the % CG and the
observed/expected ration in the region where the score is above the
threshold.  This scoring method based on sum/frequencies overpredicts
islands but finds the smaller ones around primary exons.
<b>newcpgseek</b> uses the same method as <b>cpgreport</b> but the
output is different and more readable.

<p>

2.  <b>newcpgreport</b> and <b>cpgplot</b> use a sliding window within
which the Obs/Exp ratio of CpG is calculated.  The important thing to
note in this method is that an island, in order to be reported, is
defined as a region that satisfies the following contraints:

<p>
<pre>
   Obs/Exp ratio > 0.6
   % C + % G > 50%
   Length > 200.
</pre>
<p>

<p>

For all practical purposes you should probably use newcpgreport.  It is
actually used to produce the human cpgisland database you can find on
the EBI's ftp server as well as on the EBI's SRS server.

<p>

<b>geecee</b> measures CG content in the entire input sequence and is
not to be used to detect CpG islands.  It can be usefull for detecting
sequences that MIGHT contain an island.
          

<H2>
    Author(s)
</H2>


This program was originally written by 
Gos Micklem (gos&nbsp;&copy;&nbsp;ebi.ac.uk)
<br>
Informatics Division, European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK


<p>
It was modified for inclusion in EGCG under the name 'CPGSPANS' by
Rodrigo Lopez (rls&nbsp;&copy;&nbsp;ebi.ac.uk)
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK


<p>

This application was modified for inclusion in EMBOSS by 
Alan Bleasby (ableasby&nbsp;&copy;&nbsp;hgmp.mrc.ac.uk)
<br>
HGMP-RC, Genome Campus, Hinxton, Cambridge CB10 1SB, UK



<H2>
    History
</H2>

Completed 22nd March 1999.

<!--
<p>
[an error occurred while processing this directive]
-->

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

