<HTML>

<HEAD>
  <TITLE>
  EMBOSS: cusp
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: cusp
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program cusp
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Create a codon usage table


<H2>
    Description
</H2>


Reads one or more coding sequences (CDS sequence only) and calculates
a codon frequency table.

<p>
The output file can be used as a codon usage table in other applications.

<H2>
    Usage
</H2>


Here is a sample session with cusp, using just one sequence.  Normal
use would be to extract a set of coding sequences and to use these as
input.

<pre>

% cusp -sbeg 135 -send 1292
Create a codon usage table
Input sequence: embl:paamir
Output file [paamir.cusp]: 

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
  [-outfile]           outfile    Output file name

   Optional qualifiers: (none)
   Advanced qualifiers:
   -cfile              codon      Codon usage table name


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.cusp</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-cfile</td>
<td>Codon usage table name</td>
<td>Codon usage file in EMBOSS data path</td>
<td>Ehum.cut</td>
</tr>

</table>



<H2>
    Input file format
</H2>


<H2>
    Output file format
</H2>


Here is the start of the output from the example run, which used a
single CDS from Pseudomonas aeruginosa which has a very high GC
content ands a strong coding bias, as shown by the codons for Alanine
where those ending with G or C are used almost exclusively.

<pre>

# CUSP codon usage file
# Codon Amino acid      Fract   /1000   Number
GCA     A               0.077   7.772   3
GCC     A               0.462   46.632  18
GCG     A               0.462   46.632  18
GCT     A               0.000   0.000   0

.........

</pre>

<p>

The 'Fract' column gives the faction of all amino acids coded for by
this codon triplet. 
<p>

The /1000 column represents the number of codons, given the
input sequence(s), there are per 1000 bases. This will be an
extrapolation if the sequence is shorter than 1000 bases.

<p>

If multiple sequences are input then the statistics are given for all of
the sequences together, not individually. 

<H2>
    Data files
</H2>


cusp reads a codon usage file, but only as a template and does not
use any of the data so any file will give the same results.

<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

Always exits with status 0.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="cai.html">cai</a></td><td>CAI codon adaptation index</td></tr>
<tr><td><a href="chips.html">chips</a></td><td>Codon usage statistics</td></tr>
<tr><td><a href="codcmp.html">codcmp</a></td><td>Codon usage table comparison</td></tr>
<tr><td><a href="syco.html">syco</a></td><td>Synonymous codon usage Gribskov statistic plot</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>

Spring 2000 - written

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

