<HTML>
<HEAD>
  <TITLE>
  EMBOSS: dbxresource
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="/images/emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
dbxresource
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
Index a data resource catalogue using b+tree indices
<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
  -->
<b>dbxresource</b> indexes the Data Resource Catalogue, and builds
EMBOSS B+tree format index files. These indexes allow access of flat
files larger than 2Gb.

<p>
These indexes allow access of flat files larger than 2Gb.

<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

Here is a sample session with <b>dbxresource</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>dbxresource drcat </b>
Index a data resource catalogue using b+tree indices
Standard database name [drcat]: <b>drcat</b>
Resource name [drcatresource]: <b>drcatresource</b>
Database directory [.]: <b></b>
Wildcard database filename [DRCAT.dat]: <b>DRCAT.dat</b>
        id : ID
       acc : IDother
       nam : Name
       des : Description
       url : Server URL
       cat : Category name
     taxid : Taxon id
      edat : EDAM data term
      efmt : EDAM format term
       eid : EDAM data id term
      etpc : EDAM topic term
      xref : Link
      qout : Query output
      qfmt : Query output format
       qin : Query input parameters
      qurl : Query URL
      rest : Rest URL
      soap : SOAP URL
Index fields [*]: <b></b>
Compressed index files [Y]: <b></b>
General log output file [outfile.dbxresource]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
Index a data resource catalogue using b+tree indices
Version: EMBOSS:6.6.0.0

   Standard (Mandatory) qualifiers:
  [-dbname]            string     [drcat] Basename for index files (Any string
                                  from 2 to 19 characters, matching regular
                                  expression /[A-z][A-z0-9_]+/)
  [-standardname]      string     [$(dbname)] Standard database name (Any
                                  string from 2 to 19 characters, matching
                                  regular expression /[A-z][A-z0-9_]+/)
  [-dbresource]        string     [drcatresource] Resource name (Any string
                                  from 2 to 19 characters, matching regular
                                  expression /[A-z][A-z0-9_]+/)
   -directory          directory  [.] Database directory
   -filenames          string     [DRCAT.dat] Wildcard database filename (Any
                                  string)
   -fields             menu       [*] Index fields (Values: id (ID); acc
                                  (IDother); nam (Name); des (Description);
                                  url (Server URL); cat (Category name); taxid
                                  (Taxon id); edat (EDAM data term); efmt
                                  (EDAM format term); eid (EDAM data id term);
                                  etpc (EDAM topic term); xref (Link); qout
                                  (Query output); qfmt (Query output format);
                                  qin (Query input parameters); qurl (Query
                                  URL); rest (Rest URL); soap (SOAP URL))
   -[no]compressed     boolean    [Y] Compressed index files
   -outfile            outfile    [*.dbxresource] General log output file

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers:
   -release            string     [0.0] Release number (Any string up to 9
                                  characters)
   -date               string     [00/00/00] Index date (Date string dd/mm/yy)
   -exclude            string     Wildcard filename(s) to exclude (Any string)
   -indexoutdir        outdir     [.] Index file output directory

   Associated qualifiers:

   "-directory" associated qualifiers
   -extension          string     Default file extension

   "-indexoutdir" associated qualifiers
   -extension          string     Default file extension

   "-outfile" associated qualifiers
   -odirectory         string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-dbname]<br>(Parameter 1)</td>
<td>string</td>
<td>Basename for index files</td>
<td>Any string from 2 to 19 characters, matching regular expression /[A-z][A-z0-9_]+/</td>
<td>drcat</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-standardname]<br>(Parameter 2)</td>
<td>string</td>
<td>Standard database name</td>
<td>Any string from 2 to 19 characters, matching regular expression /[A-z][A-z0-9_]+/</td>
<td>$(dbname)</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-dbresource]<br>(Parameter 3)</td>
<td>string</td>
<td>Resource name</td>
<td>Any string from 2 to 19 characters, matching regular expression /[A-z][A-z0-9_]+/</td>
<td>drcatresource</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-directory</td>
<td>directory</td>
<td>Database directory</td>
<td>Directory</td>
<td>.</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-filenames</td>
<td>string</td>
<td>Wildcard database filename</td>
<td>Any string</td>
<td>DRCAT.dat</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-fields</td>
<td>list</td>
<td>Index fields</td>
<td><table><tr><td>id</td> <td><i>(ID)</i></td></tr><tr><td>acc</td> <td><i>(IDother)</i></td></tr><tr><td>nam</td> <td><i>(Name)</i></td></tr><tr><td>des</td> <td><i>(Description)</i></td></tr><tr><td>url</td> <td><i>(Server URL)</i></td></tr><tr><td>cat</td> <td><i>(Category name)</i></td></tr><tr><td>taxid</td> <td><i>(Taxon id)</i></td></tr><tr><td>edat</td> <td><i>(EDAM data term)</i></td></tr><tr><td>efmt</td> <td><i>(EDAM format term)</i></td></tr><tr><td>eid</td> <td><i>(EDAM data id term)</i></td></tr><tr><td>etpc</td> <td><i>(EDAM topic term)</i></td></tr><tr><td>xref</td> <td><i>(Link)</i></td></tr><tr><td>qout</td> <td><i>(Query output)</i></td></tr><tr><td>qfmt</td> <td><i>(Query output format)</i></td></tr><tr><td>qin</td> <td><i>(Query input parameters)</i></td></tr><tr><td>qurl</td> <td><i>(Query URL)</i></td></tr><tr><td>rest</td> <td><i>(Rest URL)</i></td></tr><tr><td>soap</td> <td><i>(SOAP URL)</i></td></tr></table></td>
<td>*</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-[no]compressed</td>
<td>boolean</td>
<td>Compressed index files</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-outfile</td>
<td>outfile</td>
<td>General log output file</td>
<td>Output file</td>
<td><i>&lt;*&gt;</i>.dbxresource</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>-release</td>
<td>string</td>
<td>Release number</td>
<td>Any string up to 9 characters</td>
<td>0.0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-date</td>
<td>string</td>
<td>Index date</td>
<td>Date string dd/mm/yy</td>
<td>00/00/00</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-exclude</td>
<td>string</td>
<td>Wildcard filename(s) to exclude</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-indexoutdir</td>
<td>outdir</td>
<td>Index file output directory</td>
<td>Output directory</td>
<td>.</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td align="left" colspan=5>"-directory" associated directory qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -extension</td>
<td>string</td>
<td>Default file extension</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td align="left" colspan=5>"-indexoutdir" associated outdir qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -extension</td>
<td>string</td>
<td>Default file extension</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td align="left" colspan=5>"-outfile" associated outfile qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -odirectory</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>dbxresource</b> reads and indexes the Data Resource Catalogue
DRCAT.dat file.


<p>






<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 

        If you wish to include the standard description of the avalable
	report formats, use:
        #include file="inc/reportformats.ihtml"
   -->

<b>dbxresource</b>  creates one summary file for the database and two files
for each field indexed.

<p>

<ul>

<li>dbalias.ent is the master file containing the names of the files
that have been indexed. It is an ASCII file. This file also contains
the database release and date information.

<li>dbalias.xid is the B+tree index file for the ID names. It is a binary
file.

<li>dbalias.pxid is an ASCII file containing information regarding the
structure of the ID name index.


<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: drcat.ent</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
# Number of files: 1
# Release: 0.0
# Date:    00/00/00
Single filename database
DRCAT.dat
</pre>
</td></tr></table><p>
<p><h3>File: drcat.pxac</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Type         Identifier
Compress     Yes
Pages        3
Secpages     0
Order        64
Fill         52
Level        0
Pagesize     2048
Cachesize    20000
Order2       22
Fill2        41
Secpagesize  512
Seccachesize 20000
Count        22
Fullcount    22
Kwlimit      18
Reffiles     0
</pre>
</td></tr></table><p>
<p><h3>File: drcat.pxcat</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Type         Secondary
Compress     Yes
Pages        3
Secpages     21
Order        27
Fill         25
Level        0
Pagesize     2048
Cachesize    20000
Order2       10
Fill2        13
Secpagesize  512
Seccachesize 20000
Count        21
Fullcount    74
Kwlimit      60
Idlimit      32
</pre>
</td></tr></table><p>
<p><h3>File: drcat.pxde</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Type         Secondary
Compress     Yes
Pages        32
Secpages     905
Order        60
Fill         49
Level        0
Pagesize     2048
Cachesize    20000
Order2       10
Fill2        13
Secpagesize  512
Seccachesize 20000
Count        829
Fullcount    2006
Kwlimit      20
Idlimit      32
</pre>
</td></tr></table><p>
<p><h3>File: drcat.pxedat</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Type         Secondary
Compress     Yes
Pages        3
Secpages     69
Order        71
Fill         56
Level        0
Pagesize     2048
Cachesize    20000
Order2       10
Fill2        13
Secpagesize  512
Seccachesize 20000
Count        61
Fullcount    152
Kwlimit      15
Idlimit      32
</pre>
</td></tr></table><p>
<p><h3>File: drcat.pxefmt</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Type         Secondary
Compress     Yes
Pages        3
Secpages     32
Order        71
Fill         56
Level        0
Pagesize     2048
Cachesize    20000
Order2       10
Fill2        13
Secpagesize  512
Seccachesize 20000
Count        11
Fullcount    135
Kwlimit      15
Idlimit      32
</pre>
</td></tr></table><p>
<p><h3>File: drcat.pxeid</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Type         Secondary
Compress     Yes
Pages        4
Secpages     110
Order        71
Fill         56
Level        0
Pagesize     2048
Cachesize    20000
Order2       10
Fill2        13
Secpagesize  512
Seccachesize 20000
Count        106
Fullcount    201
Kwlimit      15
Idlimit      32
</pre>
</td></tr></table><p>
<p><h3>File: drcat.pxetpc</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Type         Secondary
Compress     Yes
Pages        3
Secpages     71
Order        71
Fill         56
Level        0
Pagesize     2048
Cachesize    20000
Order2       10
Fill2        13
Secpagesize  512
Seccachesize 20000
Count        67
Fullcount    192
Kwlimit      15
Idlimit      32
</pre>
</td></tr></table><p>
<p><h3>File: drcat.pxid</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Type         Identifier
Compress     Yes
Pages        6
Secpages     0
Order        44
Fill         38
Level        0
Pagesize     2048
Cachesize    20000
Order2       22
Fill2        41
Secpagesize  512
Seccachesize 20000
Count        110
Fullcount    110
Kwlimit      32
Reffiles     0
</pre>
</td></tr></table><p>
<p><h3>File: drcat.pxnm</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Type         Secondary
Compress     Yes
Pages        12
Secpages     361
Order        60
Fill         49
Level        0
Pagesize     2048
Cachesize    20000
Order2       10
Fill2        13
Secpagesize  512
Seccachesize 20000
Count        335
Fullcount    659
Kwlimit      20
Idlimit      32
</pre>
</td></tr></table><p>
<p><h3>File: drcat.pxqfmt</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Type         Secondary
Compress     Yes
Pages        3
Secpages     37
Order        46
Fill         39
Level        0
Pagesize     2048
Cachesize    20000
Order2       10
Fill2        13
Secpagesize  512
Seccachesize 20000
Count        16
Fullcount    148
Kwlimit      30
Idlimit      32
</pre>
</td></tr></table><p>
<p><h3>File: drcat.pxqin</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Type         Secondary
Compress     Yes
Pages        12
Secpages     130
Order        25
Fill         23
Level        0
Pagesize     2048
Cachesize    20000
Order2       10
Fill2        13
Secpagesize  512
Seccachesize 20000
Count        126
Fullcount    206
Kwlimit      65
Idlimit      32
</pre>
</td></tr></table><p>
<p><h3>File: drcat.pxqout</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Type         Secondary
Compress     Yes
Pages        11
Secpages     124
Order        19
Fill         18
Level        0
Pagesize     2048
Cachesize    20000
Order2       10
Fill2        13
Secpagesize  512
Seccachesize 20000
Count        120
Fullcount    169
Kwlimit      90
Idlimit      32
</pre>
</td></tr></table><p>
<p><h3>File: drcat.pxqurl</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Type         Secondary
Compress     Yes
Pages        33
Secpages     900
Order        46
Fill         39
Level        0
Pagesize     2048
Cachesize    20000
Order2       10
Fill2        13
Secpagesize  512
Seccachesize 20000
Count        729
Fullcount    2147
Kwlimit      30
Idlimit      32
</pre>
</td></tr></table><p>
<p><h3>File: drcat.pxrest</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Type         Secondary
Compress     Yes
Pages        3
Secpages     12
Order        60
Fill         49
Level        0
Pagesize     2048
Cachesize    20000
Order2       10
Fill2        13
Secpagesize  512
Seccachesize 20000
Count        12
Fullcount    43
Kwlimit      20
Idlimit      32
</pre>
</td></tr></table><p>
<p><h3>File: drcat.pxrf</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Type         Secondary
Compress     Yes
Pages        3
Secpages     17
Order        71
Fill         56
Level        0
Pagesize     2048
Cachesize    20000
Order2       10
Fill2        13
Secpagesize  512
Seccachesize 20000
Count        5
Fullcount    83
Kwlimit      15
Idlimit      32
</pre>
</td></tr></table><p>
<p><h3>File: drcat.pxsoap</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Type         Secondary
Compress     Yes
Pages        3
Secpages     18
Order        71
Fill         56
Level        0
Pagesize     2048
Cachesize    20000
Order2       10
Fill2        13
Secpagesize  512
Seccachesize 20000
Count        18
Fullcount    23
Kwlimit      15
Idlimit      32
</pre>
</td></tr></table><p>
<p><h3>File: drcat.pxtaxid</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Type         Secondary
Compress     Yes
Pages        3
Secpages     40
Order        104
Fill         75
Level        0
Pagesize     2048
Cachesize    20000
Order2       10
Fill2        13
Secpagesize  512
Seccachesize 20000
Count        26
Fullcount    114
Kwlimit      6
Idlimit      32
</pre>
</td></tr></table><p>
<p><h3>File: drcat.pxurl</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Type         Secondary
Compress     Yes
Pages        11
Secpages     268
Order        52
Fill         44
Level        0
Pagesize     2048
Cachesize    20000
Order2       10
Fill2        13
Secpagesize  512
Seccachesize 20000
Count        227
Fullcount    585
Kwlimit      25
Idlimit      32
</pre>
</td></tr></table><p>
<p><h3>File: drcat.xac</h3>
<p>This file contains non-printing characters and so cannot be displayed here.
<p><h3>File: drcat.xcat</h3>
<p>This file contains non-printing characters and so cannot be displayed here.
<p><h3>File: drcat.xde</h3>
<p>This file contains non-printing characters and so cannot be displayed here.
<p><h3>File: drcat.xedat</h3>
<p>This file contains non-printing characters and so cannot be displayed here.
<p><h3>File: drcat.xefmt</h3>
<p>This file contains non-printing characters and so cannot be displayed here.
<p><h3>File: drcat.xeid</h3>
<p>This file contains non-printing characters and so cannot be displayed here.
<p><h3>File: drcat.xetpc</h3>
<p>This file contains non-printing characters and so cannot be displayed here.
<p><h3>File: drcat.xid</h3>
<p>This file contains non-printing characters and so cannot be displayed here.
<p><h3>File: drcat.xnm</h3>
<p>This file contains non-printing characters and so cannot be displayed here.
<p><h3>File: drcat.xqfmt</h3>
<p>This file contains non-printing characters and so cannot be displayed here.
<p><h3>File: drcat.xqin</h3>
<p>This file contains non-printing characters and so cannot be displayed here.
<p><h3>File: drcat.xqout</h3>
<p>This file contains non-printing characters and so cannot be displayed here.
<p><h3>File: drcat.xqurl</h3>
<p>This file contains non-printing characters and so cannot be displayed here.
<p><h3>File: drcat.xrest</h3>
<p>This file contains non-printing characters and so cannot be displayed here.
<p><h3>File: drcat.xrf</h3>
<p>This file contains non-printing characters and so cannot be displayed here.
<p><h3>File: drcat.xsoap</h3>
<p>This file contains non-printing characters and so cannot be displayed here.
<p><h3>File: drcat.xtaxid</h3>
<p>This file contains non-printing characters and so cannot be displayed here.
<p><h3>File: drcat.xurl</h3>
<p>This file contains non-printing characters and so cannot be displayed here.
<p><h3>File: outfile.dbxresource</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Processing directory: /homes/user/test/data/
Processing file: DRCAT.dat
entries: 110 (110) time: 0.0s (0.0s)
Total time: 0.0s
Entry idlen 32 OK. Maximum ID length was 28 for 'WORLD_2DPAGE_CGL14067_2DPAGE'.
Field acc acclen 18 OK. Maximum acc term length was 18 for 'REPRODUCTION2DPAGE'.
Field nam namlen 20 OK. Maximum nam term length was 16 for 'characterization'.
Field des deslen 20 OK. Maximum des term length was 17 for 'transcriptionally'.
Field url urllen 25 OK. Maximum url term length was 17 for 'compluyeast2dpage'.
Field cat catlen 60 OK. Maximum cat term length was 40 for 'Protein sequence classification database'.
Field taxid taxidlen 6 OK. Maximum taxid term length was 5 for '40674'.
Field edat edatlen 15 OK. Maximum edat term length was 4 for '2378'.
Field etpc etpclen 15 OK. Maximum etpc term length was 4 for '0166'.
Field eid eidlen 15 OK. Maximum eid term length was 4 for '1116'.
Field efmt efmtlen 15 OK. Maximum efmt term length was 4 for '2331'.
Field qout qoutlen 90 OK. Maximum qout term length was 85 for 'Protein report (transcription factor) {Documented regulators of a given set of genes}'.
Field qfmt qfmtlen 30 OK. Maximum qfmt term length was 14 for 'HTML {graphic}'.
Field qin qinlen 65 OK. Maximum qin term length was 43 for 'Sequence accession (protein) {World-2DPAGE}'.
Field qurl qurllen 30 OK. Maximum qurl term length was 29 for 'identificationAccessionNumber'.
Field xref xreflen 15 OK. Maximum xref term length was 13 for 'EMBL_explicit'.
Field rest restlen 20 OK. Maximum rest term length was 19 for 'prideMartWebService'.
Field soap soaplen 15 OK. Maximum soap term length was 12 for 'dataservices'.
</pre>
</td></tr></table><p>





<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.

	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:

        #include file="inc/localfiles.ihtml"
   -->

None.

<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
<ol>

<li>

</ol>

   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="dbiblast.html">dbiblast</a></td>
<td>Index a BLAST database</td>
</tr>

<tr>
<td><a href="dbifasta.html">dbifasta</a></td>
<td>Index a fasta file database</td>
</tr>

<tr>
<td><a href="dbiflat.html">dbiflat</a></td>
<td>Index a flat file database</td>
</tr>

<tr>
<td><a href="dbigcg.html">dbigcg</a></td>
<td>Index a GCG formatted database</td>
</tr>

<tr>
<td><a href="dbxcompress.html">dbxcompress</a></td>
<td>Compress an uncompressed dbx index</td>
</tr>

<tr>
<td><a href="dbxedam.html">dbxedam</a></td>
<td>Index the EDAM ontology using b+tree indices</td>
</tr>

<tr>
<td><a href="dbxfasta.html">dbxfasta</a></td>
<td>Index a fasta file database using b+tree indices</td>
</tr>

<tr>
<td><a href="dbxflat.html">dbxflat</a></td>
<td>Index a flat file database using b+tree indices</td>
</tr>

<tr>
<td><a href="dbxgcg.html">dbxgcg</a></td>
<td>Index a GCG formatted database using b+tree indices</td>
</tr>

<tr>
<td><a href="dbxobo.html">dbxobo</a></td>
<td>Index an obo ontology using b+tree indices</td>
</tr>

<tr>
<td><a href="dbxreport.html">dbxreport</a></td>
<td>Validate index and report internals for dbx databases</td>
</tr>

<tr>
<td><a href="dbxstat.html">dbxstat</a></td>
<td>Dump statistics for dbx databases</td>
</tr>

<tr>
<td><a href="dbxtax.html">dbxtax</a></td>
<td>Index NCBI taxonomy using b+tree indices</td>
</tr>

<tr>
<td><a href="dbxuncompress.html">dbxuncompress</a></td>
<td>Uncompress a compressed dbx index</td>
</tr>

</table>
<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>

Peter Rice
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK

<p>
Please report all bugs to the EMBOSS bug team (emboss-bug&nbsp;&copy;&nbsp;emboss.open-bio.org) not to the original author.

<H2>
    History
</H2>
<!--
        Date written and what changes have been made go in this file.
   -->




<H2>
    Target users
</H2>
<!--
        For general users, use this text
   -->
This program is intended to be used by administrators responsible
for software and database installation and maintenance. 

<H2>
    Comments
</H2>
<!--
        User/developer/other comments go in this file.
   -->
None


</BODY>
</HTML>

