<HTML>

<HEAD>
  <TITLE>
  EMBOSS: descseq
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

descseq
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>








<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Alter the name or description of a sequence


<H2>
    Description
</H2>

Most sequence formats allow, at the very minimum, a name for the
sequence and some comments to be stored in the sequence file.

<p>

Rather than using an editor to alter the name or the comments,
<b>descseq</b> allows you to simply change them and write out a new file
with the changes in and the sequence left unaltered. 

<p>

The default action is to replace the existing name or description with
your new one, but by using the qualifier '-append' what you enter is
appended to the existing name or description. 

<p>

Note that if you append to a description, no space is inserted by
default bewteen the existing description and your appended text. You
have to put in a space yourself if you require one. 

<H2>
    Usage
</H2>


<b>Here is a sample session with descseq</b>
<p>
Set the name of a sequence to "myclone23"  
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>descseq -seq ../../data/dna.text -out clone23.seq -name "myclone23" </b>
Alter the name or description of a sequence.

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>
<p>
<b>Example 2</b>
<p>
Set the description of a sequence to "This is my clone number 244"  
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>descseq -seq ../../data/dna.text -out xy24.seq -desc "This is my clone number 244" </b>
Alter the name or description of a sequence.

</pre></td></tr></table><p>
<p>
<a href="#output.2">Go to the output files for this example</a><p><p>
<p>
<b>Example 3</b>
<p>
Append some text to the description of a sequence 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>descseq -seq ../../data/dna.text -out est4.seq -desc " (submitted)" -append </b>
Alter the name or description of a sequence.

</pre></td></tr></table><p>
<p>
<a href="#output.3">Go to the output files for this example</a><p><p>




<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          sequence   Sequence USA
  [-outseq]            seqout     Output sequence USA

   Optional qualifiers:
   -name               string     Name of the sequence
   -description        string     Description of the sequence

   Advanced qualifiers:
   -append             boolean    This allows you to append the name or
                                  description you have given on to the end of
                                  the existing name or description of the
                                  sequence.

   Associated qualifiers:
  "-sequence" related qualifiers
   -sbegin1             integer    First base used
   -send1               integer    Last base used, def=seq length
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sopenfile1          string     Input filename
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name
  "-outseq" related qualifiers
   -osformat2           string     Output seq format
   -osextension2        string     File name extension
   -osname2             string     Base file name
   -osdirectory2        string     Output directory
   -osdbname2           string     Database name to add
   -ossingle2           boolean    Separate file for each entry
   -oufo2               string     UFO features
   -offormat2           string     Features format
   -ofname2             string     Features file name
   -ofdirectory2        string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for required and optional values
   -debug               boolean    Write debug output to program.dbg
   -acdlog              boolean    Write ACD processing log to program.acdlog
   -acdpretty           boolean    Rewrite ACD file as program.acdpretty
   -acdtable            boolean    Write HTML table of options
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outseq]<br>(Parameter 2)</td>
<td>Output sequence USA</td>
<td>Writeable sequence</td>
<td><i>&lt;sequence&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-name</td>
<td>Name of the sequence</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-description</td>
<td>Description of the sequence</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-append</td>
<td>This allows you to append the name or description you have given on to the end of the existing name or description of the sequence.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

</table>



<H2>
    Input file format
</H2>


<b>descseq</b> reads a normal sequence USA.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: ../../data/dna.text</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTAC
GTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGT
</pre>
</td></tr></table><p>


<H2>
    Output file format
</H2>


<b>descseq</b> writes the sequence file with a changed name or description.

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: clone23.seq</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt;myclone23
ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGT
ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGT
</pre>
</td></tr></table><p>

<a name="output.2"></a>
<h3>Output files for usage example 2</h3>
<p><h3>File: xy24.seq</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt;EMBOSS_001 This is my clone number 244
ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGT
ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGT
</pre>
</td></tr></table><p>

<a name="output.3"></a>
<h3>Output files for usage example 3</h3>
<p><h3>File: est4.seq</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt;EMBOSS_001  (submitted)
ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGT
ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGT
</pre>
</td></tr></table><p>


<H2>
    Data files
</H2>


None.
	
<H2>
    Notes
</H2>


None.
	
<H2>
    References
</H2>


None.
	
<H2>
    Warnings
</H2>


None.
	
<H2>
    Diagnostic Error Messages
</H2>


None.
	
<H2>
    Exit status
</H2>


It always exits with status 0.

<H2>
    Known bugs
</H2>


None noted.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="biosed.html">biosed</a></td><td>Replace or delete sequence sections</td></tr>
<tr><td><a href="cutseq.html">cutseq</a></td><td>Removes a specified section from a sequence</td></tr>
<tr><td><a href="degapseq.html">degapseq</a></td><td>Removes gap characters from sequences</td></tr>
<tr><td><a href="entret.html">entret</a></td><td>Reads and writes (returns) flatfile entries</td></tr>
<tr><td><a href="extractfeat.html">extractfeat</a></td><td>Extract features from a sequence</td></tr>
<tr><td><a href="extractseq.html">extractseq</a></td><td>Extract regions from a sequence</td></tr>
<tr><td><a href="listor.html">listor</a></td><td>Writes a list file of the logical OR of two sets of sequences</td></tr>
<tr><td><a href="maskfeat.html">maskfeat</a></td><td>Mask off features of a sequence</td></tr>
<tr><td><a href="maskseq.html">maskseq</a></td><td>Mask off regions of a sequence</td></tr>
<tr><td><a href="newseq.html">newseq</a></td><td>Type in a short new sequence</td></tr>
<tr><td><a href="noreturn.html">noreturn</a></td><td>Removes carriage return from ASCII files</td></tr>
<tr><td><a href="notseq.html">notseq</a></td><td>Excludes a set of sequences and writes out the remaining ones</td></tr>
<tr><td><a href="nthseq.html">nthseq</a></td><td>Writes one sequence from a multiple set of sequences</td></tr>
<tr><td><a href="pasteseq.html">pasteseq</a></td><td>Insert one sequence into another</td></tr>
<tr><td><a href="revseq.html">revseq</a></td><td>Reverse and complement a sequence</td></tr>
<tr><td><a href="seqret.html">seqret</a></td><td>Reads and writes (returns) sequences</td></tr>
<tr><td><a href="seqretsplit.html">seqretsplit</a></td><td>Reads and writes (returns) sequences in individual files</td></tr>
<tr><td><a href="skipseq.html">skipseq</a></td><td>Reads and writes (returns) sequences, skipping the first few</td></tr>
<tr><td><a href="splitter.html">splitter</a></td><td>Split a sequence into (overlapping) smaller sequences</td></tr>
<tr><td><a href="trimest.html">trimest</a></td><td>Trim poly-A tails off EST sequences</td></tr>
<tr><td><a href="trimseq.html">trimseq</a></td><td>Trim ambiguous bits off the ends of sequences</td></tr>
<tr><td><a href="union.html">union</a></td><td>Reads sequence fragments and builds one sequence</td></tr>
<tr><td><a href="vectorstrip.html">vectorstrip</a></td><td>Strips out DNA between a pair of vector sequences</td></tr>
<tr><td><a href="yank.html">yank</a></td><td>Reads a sequence range, appends the full USA to a list file</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->



<H2>
    Author(s)
</H2>


Gary Williams (gwilliam&nbsp;&copy;&nbsp;hgmp.mrc.ac.uk)
<br>
HGMP-RC, Genome Campus, Hinxton, Cambridge CB10 1SB, UK



<H2>
    History
</H2>

<!--
<p>
[an error occurred while processing this directive]
-->

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>



</BODY>
</HTML>



