<HTML>

<HEAD>
  <TITLE>
  EMBOSS
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

diffseq
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>





<!--END OF HEADER-->





<H2>
    Function
</H2>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Find differences (SNPs) between nearly identical sequences


<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.

        This is what the user looks at to find out if this program will
        do what he wants. 

        Long description.

        Why it is useful.

        When it should/shouldn't be used.

        Algorithms

        Expanded description.

        eg: Reverse engineers Wheedle's 'talpsgob' program. This does a 
        splurge of the 5' end of the sequence followed by a durm.
        The net effect is a 5' end splurt.
        etc. for as long as is required (several pages is not unreasonable)
  -->

   <b>diffseq</b> takes two overlapping, nearly identical sequences and reports the
   differences between them, together with any features that overlap with these
   regions. GFF files of the differences in each sequence are also produced.
<p>
   <b>diffseq</b> should be of value when looking for SNPs, differences between
   strains of an organism and anything else that requires the differences between
   sequences to be highlighted.
<p>  
   The sequences can be very long. The program does a match of all
   sequence words of size 10 (by default). It then reduces this to the
   minimum set of overlapping matches by sorting the matches in order of
   size (largest size first) and then for each such match it removes any
   smaller matches that overlap. The result is a set of the longest
   ungapped alignments between the two sequences that do not overlap with
   each other. The mismatched regions between these matches are reported.
<p>   
   It should be possible to find differences between sequences that are Mega bytes
   long.
<p>
     



<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->


Here is a sample session with diffseq:

<pre>

% diffseq tembl:ap000504 tembl:af129756
Find differences (SNPs) between nearly identical sequences
Word size [10]: 
Output file [ap000504.diffseq]:

</pre>


<H2>
    Command line arguments
</H2>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-asequence]         sequence   Sequence USA
  [-bsequence]         sequence   Sequence USA
   -wordsize           integer    Word size
   -outfile            report     Output report file

   Optional qualifiers:
   -afeatout           featout    File for output of first sequence's normal
                                  tab delimited gff's
   -bfeatout           featout    File for output of second sequence's normal
                                  tab delimited gff's
   -columns            boolean    The default format for the output report
                                  file is to have several lines per difference
                                  giving the sequence positions, sequences
                                  and features.
                                  If this option is set true then the output
                                  report file's format is changed to a set of
                                  columns and no feature information is given.

   Advanced qualifiers: (none)
   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-asequence]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-bsequence]<br>(Parameter 2)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-wordsize</td>
<td>Word size</td>
<td>Integer 2 or more</td>
<td>10</td>
</tr>

<tr>
<td>-outfile</td>
<td>Output report file</td>
<td>Report file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-afeatout</td>
<td>File for output of first sequence's normal tab delimited gff's</td>
<td>Writeable feature table</td>
<td>$(asequence.name).diffgff</td>
</tr>

<tr>
<td>-bfeatout</td>
<td>File for output of second sequence's normal tab delimited gff's</td>
<td>Writeable feature table</td>
<td>$(bsequence.name).diffgff</td>
</tr>

<tr>
<td>-columns</td>
<td>The default format for the output report file is to have several lines per difference giving the sequence positions, sequences and features.
If this option is set true then the output report file's format is changed to a set of columns and no feature information is given.</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        Except for sequences which can be described by 'sequence USA',
        this should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

This program reads in two nucleic acid sequence USAs or two protein sequence USAs.

<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats 
        Except for sequences which can be described by 'sequence USA',
        this should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->

A report of the differences between the two sequences is produced,
together with any features that overlap with these differing regions. 
<p>

<p>

The output is a standard EMBOSS report file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-rformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  The available format names are: embl,
genbank, gff, pir, swiss, trace, listfile, dbmotif, diffseq, excel,
feattable, motif, regions, seqtable, simple, srs, table, tagseq

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html</A>
for further information on report formats.

<p>


<p>

By default <b>marscan</b> writes a 'diffseq' report file.


<p>
<pre>
<hr>
########################################
# Program: diffseq
# Rundate: Mon Feb 11 13:16:56 2002
# Report_file: ap000504.diffseq
# Additional_files: 2
# 1: AP000504.diffgff (Feature file for first sequence)
# 2: AF129756.diffgff (Feature file for second sequence)
########################################

#=======================================
#
# Sequence: AP000504     from: 1   to: 100000
# HitCount: 119
#
# Compare: AF129756     from: 1   to: 184666
#
# AP000504 overlap starts at 1
# AF129756 overlap starts at 6036
#
# (AP000504) start end length sequence
# (AF129756) start end length sequence
#
#
#
#=======================================


AP000504 847-847 Length: 1
Sequence: a
Sequence: t
AF129756 6882-6882 Length: 1

AP000504 1795-1795 Length: 1
Sequence: g
Sequence: a
AF129756 7830-7830 Length: 1

AP000504 2273-2273 Length: 1
Sequence: t
Sequence:
Feature: repeat_region 7920-8351 rpt_family='MSTB'
AF129756 8307 Length: 0

AP000504 2466-2466 Length: 1
Sequence: g
Sequence: a
Feature: repeat_region 8391-8686 rpt_family='AluSg'
AF129756 8500-8500 Length: 1

AP000504 2655-2658 Length: 4
Sequence: tgtg
Sequence:
Feature: repeat_region 8687-8731 rpt_family='(CA)n'
AF129756 8688 Length: 0

AP000504 4914 Length: 0
Sequence:
Sequence: gtgtgtgtgtgtgtgtgt
Feature: repeat_region 10910-10972 rpt_family='(CA)n'
AF129756 10945-10962 Length: 18

AP000504 4951-4953 Length: 3
Sequence: aaa
Sequence: tat
Feature: repeat_region 10991-11020 rpt_family='AT_rich'
AF129756 10999-11001 Length: 3

AP000504 6600-6600 Length: 1
Sequence: t
Sequence:
Feature: repeat_region 12628-12930 rpt_family='AluSq'
AF129756 12647 Length: 0             


etc.


AP000504 97273-97274 Length: 2
Sequence: aa
Sequence:
Feature: repeat_region 103299-103402 rpt_family='AluSq'
AF129756 103302 Length: 0

AP000504 97716-97716 Length: 1
Sequence: a
Sequence: g
AF129756 103744-103744 Length: 1

AP000504 97827-97827 Length: 1
Sequence: c
Sequence: t
Feature: repeat_region 103784-104083 rpt_family='AluSx'
AF129756 103855-103855 Length: 1

#---------------------------------------
#
# Overlap_end: 100000 in AP000504
# Overlap_end: 106028 in AF129756
#
# SNP_count: 86
# Transitions: 58
# Transversions: 28
#
#
#---------------------------------------
<hr>
</pre>
<p>

The first line is the title giving the names of the sequences used.
<p>
The next two non-blank lines state the positions in each sequence where the detected
overlap between them starts.
<p>
There then follows a set of reports of the mismatches between the sequences.
<br>
Each report consists of 4 or more lines.
<br>
<ul>
<li>The first line has the name of the first sequence followed by the start and end
positions of the mismatched region in that sequence, followed by the length of
the mismatched region. If the mismatched region is of zero length in this sequence,
then only the position of the last matching base before the mismatch is given.</li>

<li>If a feature of the first sequence overlaps with this mismatch region, then one or more
lines starting with 'Feature:' comes next with the type, position and tag field
of the feature.</li>

<li>Next is a line starting "Sequence:" giving the sequence of the mismatch in the
first sequence.</li>
</ul>
<p>

This is followed by the equivalent information for the second sequence,
but in the reverse order, namely 'Sequence:' line, 'Feature:' lines and line
giving the position of the mismatch in the second sequence.
<p>

At the end of the report are two non-blank lines giving the positions in each
sequence where the detected
overlap between them ends.
<p>

The last three lines of the report gives the counts of SNPs (defined as a change of one nucleotide to
one other nucleotide, no deletions or insertions are counted, no multi-base changes are counted).
<p>

The counts of transitions (Pyrimide to Pyrimidine or Purine to Purine) and transversions (Pyrimidine to
Purine) are also given.
<p>

It should be noted that not all features are reported.
<p>

The 'source' feature found in all EMBL/Genbank feature table entries is not reported
as this covers all of the sequence and so overlaps with any difference found in that
sequence and so is uninformative and irritating. It has therefore been removed from
the output report.
<p>

The translation information of CDS features is often extremely long and does not add useful
information to the report. It has therefore been removed from the output report.
<p>

If no regions of alignment are found, the following output is given:
<p>

<pre>
<hr>
########################################
# Program: diffseq
# Rundate: Mon Feb 11 13:21:20 2002
# Report_file: ap000504.diffseq
# Additional_files: 2
# 1: AP000504.diffgff (Feature file for first sequence)
# 2: fred.diffgff (Feature file for second sequence)
########################################

#=======================================
#
# Sequence: AP000504     from: 1   to: 100000
# HitCount: 0
#=======================================


#---------------------------------------
#
# No regions of alignment found.
#
#
#---------------------------------------              
<hr>
</pre>
<p>

If the <b>-rformat table</b> qualifier is given then the output is given
in a columnar format. 

<p>

The columns are separated by one or more spaces or TAB characters in the
order:

<p>

<ul>
<li>sequence 1 name
<li>sequence 1 start
<li>sequence 1 end
<li>score (always 0.000)
<li>sequence 2 start
<li>sequence 2 end
<li>sequence 2 name
<li>sequence difference length
<li>sequence 
<li>first feature
<li>second feature
</ul>
<p>

For example:
<p>

<pre>
<hr>
########################################
# Program: diffseq
# Rundate: Mon Feb 11 13:28:25 2002
# Report_file: ap000504.diffseq
# Additional_files: 2
# 1: AP000504.diffgff (Feature file for first sequence)
# 2: AF129756.diffgff (Feature file for second sequence)
########################################

#=======================================
#
# Sequence: AP000504     from: 1   to: 100000
# HitCount: 119
#
# Compare: AF129756     from: 1   to: 184666
#
# AP000504 overlap starts at 1
# AF129756 overlap starts at 6036
#
# (AP000504) start end length sequence
# (AF129756) start end length sequence
#
#
#
#=======================================

USA               Start     End   Score start  end    length name   sequence first_feature second_feature
AP000504            847     847   0.000 6882   6882   1      AF129756 t        .             .
AP000504           1795    1795   0.000 7830   7830   1      AF129756 a        .             .
AP000504           2273    2273   0.000 8307   8306   .      AF129756 .        .             repeat_region 7920-8351 rpt_family='MSTB'
AP000504           2466    2466   0.000 8500   8500   1      AF129756 a        .             repeat_region 8391-8686 rpt_family='AluSg'
AP000504           2655    2658   0.000 8688   8687   .      AF129756 .        .             repeat_region 8687-8731 rpt_family='(CA)n'
AP000504           4914    4913   0.000 10945  10962  18     AF129756 gtgtgtgtgtgtgtgtgt .             repeat_region 10910-10972 rpt_family='(CA)n'


etc.

AP000504          93860   93860   0.000 99890  99890  1      AF129756 g        .             .
AP000504          95451   95451   0.000 101481 101481 1      AF129756 t        .             .
AP000504          96650   96650   0.000 102680 102680 1      AF129756 t        .             .
AP000504          97273   97274   0.000 103302 103301 .      AF129756 .        .             repeat_region 103299-103402 rpt_family='AluSq'
AP000504          97716   97716   0.000 103744 103744 1      AF129756 g        .             .
AP000504          97827   97827   0.000 103855 103855 1      AF129756 t        .             repeat_region 103784-104083 rpt_family='AluSx'

#---------------------------------------
#
# Overlap_end: 100000 in AP000504
# Overlap_end: 106028 in AF129756
#
# SNP_count: 86
# Transitions: 58
# Transversions: 28
#
#
#---------------------------------------            
<hr>
</pre>
<p>


If no regions of alignment are found, the following output is given:
<p>

<pre>
<hr>
########################################
# Program: diffseq
# Rundate: Mon Feb 11 13:34:34 2002
# Report_file: ap000504.diffseq
# Additional_files: 2
# 1: AP000504.diffgff (Feature file for first sequence)
# 2: fred.diffgff (Feature file for second sequence)
########################################

#=======================================
#
# Sequence: AP000504     from: 1   to: 100000
# HitCount: 0
#=======================================

USA               Start     End   Score start  end    length name   sequence first_feature second_feature

#---------------------------------------
#
# No regions of alignment found.
#
#
#---------------------------------------    
<hr>
</pre>
<p>



<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.

        eg:
        data file format:

	blank lines and comment lines are ignored.
        comment lines start with '#'
        There are three columns separated by one or more space or TAB
	characters for X and Y values of durm, followed by energy
        all coefficients are given as floating point values
        An example is:

        # Durm coefficient enery file
        # X     Y       Energy
        0.1     0.2     0.331
        0.2     0.4     0.332
        0.3     0.6     0.333
        0.3     0.8     0.334



If you wish to include the standard description of what data files
are and how to use embossdata to inspect and retrieve them, use:

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>


   -->



<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

It should be noted that not all features are reported.

<p>

The 'source' feature found in all EMBL/Genbank feature table entries is
not reported as this covers all of the sequence and so overlaps with any
difference found in that sequence and so is uninformative and
irritating.  It has therefore been removed from the output report. 

<p>

The translation information of CDS features is often extremely long and
does not add useful information to the report.  It has therefore been
removed from the output report. 

<p>

If you run out of memory, use a larger word size.

<p>

Using a larger word size increases the length between mismatches that
will be reported as one event.  Thus a word size of 50 will report two
SNP that are with 50 bases of each other as one mismatch. 

   
<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.

<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
        eg:
        It doesn't check to see if the sequence is a protein sequence or not.
   -->

None.

<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program.
        eg:

        "Threshold value greater than 1." - means that the threshold has
        been set to be greater than 1 but a burgle exception has
        occurred making this invalid - you should reduce the wobble
        parameter and try again. 

   -->
None.

<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
        eg:
        "0"
   -->
It always exits with status 0.

<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->
<p>
   A graphical dotplot of the matches used in this program can be displayed
   using the program <b>dotpath</b>.



<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
e.g.

This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>
   -->

This application was written by Gary Williams <a
href="mailto:gwilliam@hgmp.mrc.ac.uk">(gwilliam@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written 15th Aug 2000 - Gary Williams.
<br>
18th Aug 2000 - Added writing out GFF files of the mismatched regions

<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:

   -->
This program is intended to be used by everyone and everything, from naive users to embedded scripts.



<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->



</BODY>
</HTML>

