<HTML>

<HEAD>
  <TITLE>
  EMBOSS: digest
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

digest
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>







<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Protein proteolytic enzyme or reagent cleavage digest


<H2>
    Description
</H2>


<b>digest</b> finds the positions where a specified proteolytic enzyme
or reagent might cut a peptide sequence. 

<p>

This programs allows you to input a protein sequence and to specify one
proteolytic agent from a list.  It will then output a file containing
the positions where the agent cuts, together with the peptides produced. 

<p>

Trypsin will not normally cut after a K if it is followed by (e.g.)
another K or a P.  Specifying the qualifier <b>-unfavoured</b> shows
those cuts.  as well as the favoured ones. 

<p>

If you wish to emulate a partial digestion, then using the
<b>-overlap</b> qualifier will display the results from a digest in
which all cut sites are used and in which one site at a time is not cut. 
Thus the resulting peptide fragments from the cut sites numbered 1, 2,
3, 4 etc.  are shown, plus the fragments produced by cutting at the
sites 1..3, 2..4, etc. 

<p>

i.e. fragments containing one potential cut site are also shown.

<p>

If you wish to emulate a <b>very</b> partial digestion (!) then using
the <b>-allpartials</b> qualifier will do what <b>-overlap</b> did, but
also show all possible fragments from using all possible combinations of
cut sites.  For example, fragments from the cut sites numbered 1, 2, 3,
4, 5 and from the fragments produced by cutting between sites 1..3,
1..4, 1..5, 2..4, 2..5, 3..5, etc. 

<p>

i.e. fragments containing one or more potential cut site are also shown.


<H2>
    Usage
</H2>


Here is a sample session with digest.

<pre>

% digest
Input sequence: sw:opsd_human
Enzymes and Reagents
         1 : Trypsin
         2 : Lys-C
         3 : Arg-C
         4 : Asp-N
         5 : V8-bicarb
         6 : V8-phosph
         7 : Chymotrypsin
         8 : CNBr
Select number [1]: 
Output file [opsd_human.digest]: 

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequencea]         sequence   Sequence USA
   -menu               menu       Select number
  [-outfile]           report     Output report file name

   Optional qualifiers: (none)
   Advanced qualifiers:
   -unfavoured         boolean    Trypsin will not normally cut after a K if
                                  it is followed by (e.g.) another K or a P.
                                  Specifying this shows those cuts. as well as
                                  the favoured ones.
   -aadata             string     Molecular weight data for amino acids
   -overlap            boolean    Used for partial digestion. Shows all cuts
                                  from favoured cut sites plus 1..3, 2..4,
                                  3..5 etc but not (e.g.) 2..5. Overlaps are
                                  therefore fragments with exactly one
                                  potential cut site within it.
   -allpartials        boolean    As for overlap but fragments containing more
                                  than one potential cut site are included.

   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequencea]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-menu</td>
<td>Select number</td>
<td><table><tr><td>1</td> <td><i>(Trypsin)</i></td></tr><tr><td>2</td> <td><i>(Lys-C)</i></td></tr><tr><td>3</td> <td><i>(Arg-C)</i></td></tr><tr><td>4</td> <td><i>(Asp-N)</i></td></tr><tr><td>5</td> <td><i>(V8-bicarb)</i></td></tr><tr><td>6</td> <td><i>(V8-phosph)</i></td></tr><tr><td>7</td> <td><i>(Chymotrypsin)</i></td></tr><tr><td>8</td> <td><i>(CNBr)</i></td></tr></table></td>
<td>1</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output report file name</td>
<td>Report file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-unfavoured</td>
<td>Trypsin will not normally cut after a K if it is followed by (e.g.) another K or a P. Specifying this shows those cuts. as well as the favoured ones.</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-aadata</td>
<td>Molecular weight data for amino acids</td>
<td>Any string is accepted</td>
<td>Eamino.dat</td>
</tr>

<tr>
<td>-overlap</td>
<td>Used for partial digestion. Shows all cuts from favoured cut sites plus 1..3, 2..4, 3..5 etc but not (e.g.) 2..5. Overlaps are therefore fragments with exactly one potential cut site within it.</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-allpartials</td>
<td>As for overlap but fragments containing more than one potential cut site are included.</td>
<td>Yes/No</td>
<td>No</td>
</tr>

</table>



<H2>
    Input file format
</H2>


Any protein sequence.

<H2>
    Output file format
</H2>


<p>

The output is a standard EMBOSS report file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-rformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  The available format names are: embl,
genbank, gff, pir, swiss, trace, listfile, dbmotif, diffseq, excel,
feattable, motif, regions, seqtable, simple, srs, table, tagseq

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html</A>
for further information on report formats.

<p>


<p>

By default <b>digest</b> writes a 'seqtable' report file.


<p>

<hr>
<pre>
########################################
# Program: digest
# Rundate: Mon Feb 11 13:36:16 2002
# Report_file: opsd_human.digest
########################################

#=======================================
#
# Sequence: OPSD_HUMAN     from: 1   to: 348
# HitCount: 14
#
# Complete digestion with Trypsin yields 14 fragments
#
#=======================================

  Start     End Mol_Weight cterm  nterm  Sequence
     70     135 7129.319   R      Y      TPLNYILLNLAVADLFMVLGGFTSTLYTSLHGYFVFGPTGCNLEGFFATLGGEIALWSLVVLAIER
    178     231 6335.495   R      E      YIPEGLQCSCGIDYYTLKPEVNNESFVIYMFVVHFTIPMIIIFFCYGQLVFTVK
     22      69 5788.873   R      T      SPFEYPQYYLAEPWQFSMLAAYMFLLIVLGFPINFLTLYVTVQHKKLR
    253     296 5004.085   R      S      MVIIMVIAFLICWVPYASVAFYIFTHQGSNFGPIFMTIPAFFAK
    136     177 4600.460   R      Y      YVVVCKPMSNFRFGENHAIMGVAFTWVMALACAAPPLAGWSR
      1      21 2257.495   .      S      MNGTEGPNFYVPFSNATGVVR
    297     311 1728.089   K      Q      SAAIYNPVIYIMMNK
    232     245 1490.538   K      A      EAAAQQQESATTQK
    326     339 1403.457   K      T      NPLGDDEASATVSK
    315     325 1186.476   R      N      NCMLTTICCGK
    340     348 902.950    K      .      TETSQVAPA
    249     252 503.550    K      M      EVTR
    312     314 449.504    K      N      QFR
    246     248 346.378    K      E      AEK

#---------------------------------------
#---------------------------------------   
</pre>
<hr>

<p>

The header information contains the propgram name, date of run, name of
the reagent used to digest the protein and the number of fragments
reported.  The header will report if complete or partial digestion was
chosen.  

<p>
The rest of the file consists of columns holding the following data:

<ul>
<li> The start position of the fragment
<li> The end position of the fragment
<li> The molecular weight of the fragment
<li> The residue before the cut site ('.' if start of sequence)
<li> The residue after the second cut site ('.' if end of sequence)
<li> The sequence of the fragment. 
</ul>

<H2>
    Data files
</H2>

None.


<H2>
    Notes
</H2>

None.


<H2>
    References
</H2>

None.


<H2>
    Warnings
</H2>

None.


<H2>
    Diagnostic Error Messages
</H2>

None.


<H2>
    Exit status
</H2>

ALways exist with a status of 0.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="antigenic.html">antigenic</a></td><td>Finds antigenic sites in proteins</td></tr>
<tr><td><a href="fuzzpro.html">fuzzpro</a></td><td>Protein pattern search</td></tr>
<tr><td><a href="fuzztran.html">fuzztran</a></td><td>Protein pattern search after translation</td></tr>
<tr><td><a href="helixturnhelix.html">helixturnhelix</a></td><td>Report nucleic acid binding motifs</td></tr>
<tr><td><a href="oddcomp.html">oddcomp</a></td><td>Finds protein sequence regions with a biased composition</td></tr>
<tr><td><a href="patmatdb.html">patmatdb</a></td><td>Search a protein sequence with a motif</td></tr>
<tr><td><a href="patmatmotifs.html">patmatmotifs</a></td><td>Search a PROSITE motif database with a protein sequence</td></tr>
<tr><td><a href="pepcoil.html">pepcoil</a></td><td>Predicts coiled coil regions</td></tr>
<tr><td><a href="preg.html">preg</a></td><td>Regular expression search of a protein sequence</td></tr>
<tr><td><a href="pscan.html">pscan</a></td><td>Scans proteins using PRINTS</td></tr>
<tr><td><a href="sigcleave.html">sigcleave</a></td><td>Reports protein signal cleavage sites</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>

Written (1999) - Alan Bleasby

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

