<HTML>

<HEAD>
  <TITLE>
  EMBOSS: digest
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: digest
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program digest
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Protein proteolytic enzyme or reagent cleavage digest


<H2>
    Description
</H2>


This finds the positions where a specified proteolytic enzyme or reagent
might cut a peptide sequence.

<H2>
    Usage
</H2>


Here is a sample session with digest.

<pre>

% digest
Input sequence: sw:opsd_human
Enzymes and Reagents
         1 : Trypsin
         2 : Lys-C
         3 : Arg-C
         4 : Asp-N
         5 : V8-bicarb
         6 : V8-phosph
         7 : Chymotrypsin
         8 : CNBr
Select number [1]: 
Output file [opsd_human.digest]: 

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequencea]         sequence   Sequence USA
   -menu               list       Select number
  [-outfile]           outfile    Output file name

   Optional qualifiers: (none)
   Advanced qualifiers:
   -unfavoured         bool       Trypsin will not normally cut after a K if
                                  it is followed by (e.g.) another K or a P.
                                  Specifying this shows those cuts. as well as
                                  the favoured ones.
   -overlap            bool       Used for partial digestion. Shows all cuts
                                  from favoured cut sites plus 1..3, 2..4,
                                  3..5 etc but not (e.g.) 2..5. Overlaps are
                                  therefore fragments with exactly one
                                  potential cut site within it.
   -allpartials        bool       As for overlap but fragments containing more
                                  than one potential cut site are included.


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequencea]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-menu</td>
<td>Select number</td>
<td><table><tr><td>1</td> <td><i>(Trypsin)</i></td></tr><tr><td>2</td> <td><i>(Lys-C)</i></td></tr><tr><td>3</td> <td><i>(Arg-C)</i></td></tr><tr><td>4</td> <td><i>(Asp-N)</i></td></tr><tr><td>5</td> <td><i>(V8-bicarb)</i></td></tr><tr><td>6</td> <td><i>(V8-phosph)</i></td></tr><tr><td>7</td> <td><i>(Chymotrypsin)</i></td></tr><tr><td>8</td> <td><i>(CNBr)</i></td></tr></table></td>
<td>1</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.digest</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-unfavoured</td>
<td>Trypsin will not normally cut after a K if it is followed by (e.g.) another K or a P. Specifying this shows those cuts. as well as the favoured ones.</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-overlap</td>
<td>Used for partial digestion. Shows all cuts from favoured cut sites plus 1..3, 2..4, 3..5 etc but not (e.g.) 2..5. Overlaps are therefore fragments with exactly one potential cut site within it.</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-allpartials</td>
<td>As for overlap but fragments containing more than one potential cut site are included.</td>
<td>Yes/No</td>
<td>No</td>
</tr>

</table>



<H2>
    Input file format
</H2>


Any protein sequence.

<H2>
    Output file format
</H2>


Here is the output from the example:

<pre>

DIGEST of OPSD_HUMAN from 1 to 348 Molwt= 38892.536

Complete digestion with Trypsin yields 14 fragments:
Start   End     Molwt      Sequence (up to 38 residues)
70      135     7129.324   (R) TPLNYILLNLAVADLFMVLGGFTSTLYTSLHGYFVFGP (Y) ...
178     231     6335.500   (R) YIPEGLQCSCGIDYYTLKPEVNNESFVIYMFVVHFTIP (E) ...
22      69      5788.878   (R) SPFEYPQYYLAEPWQFSMLAAYMFLLIVLGFPINFLTL (T) ...
253     296     5004.090   (R) MVIIMVIAFLICWVPYASVAFYIFTHQGSNFGPIFMTI (S) ...
136     177     4600.464   (R) YVVVCKPMSNFRFGENHAIMGVAFTWVMALACAAPPLA (Y) ...
1       21      2257.500    () MNGTEGPNFYVPFSNATGVVR (S) 
297     311     1728.093   (K) SAAIYNPVIYIMMNK (Q) 
232     245     1490.543   (K) EAAAQQQESATTQK (A) 
326     339     1403.462   (K) NPLGDDEASATVSK (T) 
315     325     1186.480   (R) NCMLTTICCGK (N) 
340     348     902.954    (K) TETSQVAPA  () 
249     252     503.555    (K) EVTR (M) 
312     314     449.509    (K) QFR (N) 
246     248     346.383    (K) AEK (E) 

</pre>

<p>
The first non-blank line is the title containing the name of the program
and the name o fthe input sequence with the range considered and the
calculated molecular weight of the protein. 

<p>
The next non-blank line gives the name of the reagent used to digest the
protein and the number of fragments reported.  The line will report is
complete or partial digestion was chosen.  If the protein is not
digested at all by the reagent chosen the following is reported: 

<pre>
"Is not proteolytically digested using (reagent name)"
</pre>

<p>
If '-overlap' was specified, the next line will be:

<pre>
"Only overlapping partials shown:"
</pre>

<p>
If '-allpartials' was specified, the next line will be:

<pre>
"All partials shown:"
</pre>


<p>
The next non-blank line gives the headings of the columns of data.

<p>
The rest of the file consists of columns holding the following data:

<ul>
<li> The start position of the fragment
<li> The end position of the fragment
<li> The residue before the cut site (in brackets)
<li> The sequence of the fragment.  If the fragment is longer than 38
	residues then the sequence reported is truncated to 38 residues and
	'...' is in the last column
<li> The residue after the second cut site (in brackets)
<li> If the fragment is longer than 38 then '...' is in the last column,
	else this is blank. 
</ul>

<H2>
    Data files
</H2>


<H2>
    Notes
</H2>


<H2>
    References
</H2>


<H2>
    Warnings
</H2>


<H2>
    Diagnostic Error Messages
</H2>


<H2>
    Exit status
</H2>


<H2>
    Known bugs
</H2>


<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="checktrans.html">checktrans</a></td><td>Reports STOP codons and ORF statistics of a protein sequence</td></tr>
<tr><td><a href="iep.html">iep</a></td><td>Calculates the isoelectric point of a protein</td></tr>
<tr><td><a href="octanol.html">octanol</a></td><td>Displays protein hydropathy</td></tr>
<tr><td><a href="pepinfo.html">pepinfo</a></td><td>Plots simple amino acid properties in parallel</td></tr>
<tr><td><a href="pepnet.html">pepnet</a></td><td>Displays proteins as a helical net</td></tr>
<tr><td><a href="pepstats.html">pepstats</a></td><td>Protein statistics</td></tr>
<tr><td><a href="pepwheel.html">pepwheel</a></td><td>Shows protein sequences as helices</td></tr>
<tr><td><a href="pepwindow.html">pepwindow</a></td><td>Displays protein hydropathy</td></tr>
<tr><td><a href="pepwindowall.html">pepwindowall</a></td><td>Displays protein hydropathy of a set of sequences</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>


<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

