<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: distmat
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">
<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>
<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>
</td>
<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: distmat
</H2>
</font>
</td>
</tr>
</table>
</td>
</tr>
</table>
<p><hr><p>

<!--END OF HEADER-->

<H1>
Program distmat
</H1>
<H2>
    Function
</H2>
Creates a distance matrix from multiple alignments

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
        Algorithms
  -->

<b>distmat</b> calculates the evolutionary distances between every pair of
sequences in a multiple alignment. The sequences need to be aligned
before running this program. The quality of the alignment is of
paramount importance in obtaining meaningful information from this
analysis. This application calculates a distance matrix for the set of
sequences in the alignment. The distances are expressed in terms of the
number of substitutions per 100 bases or amino acids.

<p>

   As sequence diverge so does the probability of there being multiple
substitutions at any one site in the alignment increase. The distance
will then be an underestimate of the true evolutionary distance between
the sequences. Therefore, there are a number of methods for correcting
the observed substitution rate for the occurence of multiple
substutions.

<p>

  For nucleotides, the "-position" flag allows the user to choose base
positions to analyse in each codon, i.e. 123 (all bases), 12 (the first
two bases), 1, 2, or 3 individual bases.

<h3>Uncorrected distances</h3>

This method does not make any corrections for multiple substitutions.
Therefore, the score will be an underestimate of the distance between
the sequences. This will not be less significant for highly similar sets
of sequences.

<p>

<pre>
S = m/(npos + gaps*gap_penalty)                                  (1)

m 	    - score of matches (1 for an exact match, a fraction for partial
	      matches and 0 for no match)
npos	    - number of positions included in m
gaps        - number of gaps in the sequences
gap_penalty - the score given to a gapped position
</pre>

<p>

<pre>
D = uncorrected distance = p-distance = 1-S          (2)
</pre>

<p>

The score of match includes all exact matches. For nucleotides, if the
flag "-ambiguous" is used then partial matches are included in the
score. For example, a match of M (A or C) with A will increment m by 0.5
(0.5*1.0). Gaps are not included in the calculation unless a non zero
value is given with "-gapweight". It should be noted that end gaps and
internal gaps will be weighted by the same amount. So it is recommended
that this be used with "-sbegin"and "-send" to specify the start and end
of the region to calculate the distance from.


<h2>Multiple Substitution correction algorithms</h2>


<h3>Jukes-Cantor</h3>


This can be used for nucleotide and protein sequences.

<p>

<pre>
distance = -b ln (1-D/b)

D - uncorrected distance
b - constant. b= 3/4 for nucleotides and 19/20 for proteins.
</pre>

<p>

Partial matches and gap  positions can be taken into account in the
calculation of D, by setting the "-ambiguous" and "-gapweight" flags
(see "uncorrected distance" method).

<p>


Reference: 
<br>
"Phylogenetic Inference", Swofford, Olsen, Waddell, and
Hillis, in Molecular Systematics, 2nd ed., Sinauer Ass., Inc., 1996, Ch. 11. 


<h3>Tajima-Nei</h3>

This method is only for nucleotide sequences. It uses the same equation
as Jukes-Cantor, but the b-parameter is not constant. Also, only exact
matches are considered in the calculation of the match score and gap
positions are ignored.

<p>

<pre>
A = 1, T = 2, C = 3, G = 4

b = 0.5(1.- Sum(i=A,G)(fraction[i]^2  + D^2/h)

h = Sum(i=A,C)Sum(k=T,G) (0.5 * pair_frequency[i,k]^2/(fraction[i]*fraction[k]))

distance = -b ln(1.-D/b)

pair_frequency[i,k]   - frequency of the i and k base pair at sites in
			the alignement of the pair of sequences.
fraction[i]           - average content of the base i in both sequences
</pre>

<p>

Reference: 
<br>
F. Tajima and M. Nei, Mol. Biol. Evol. 1984, 1, 269.


<h3>Kimura Two-Parameter distance</h3>

This method is only for nucleotide sequences. This uses the principle
that transition substitutions (purine-purine and pyrimidine-purine) are
more likely than transversion substitutions (purine-pyprimidine). Purine
being the nucleic acid constituent of A and G, and pyrimidine being the
nucleic acid derivative of the bases C, T and U. Gaps are ignored and
abiguous symbols other than R (purine) and Y (pyrimidine) are ingnored.

<p>

<pre>
P = transitions/npos
Q = transversions/npos

npos - number of positions scored

distance = -0.5 ln[ (1-2P-Q)*sqrt(1-2Q)]
</pre>

<p>

Reference: 
<br>
M. kimura, J. Mol. Evol. 1980, 16, 111.


<h3>Tamura</h3>


This method is only for nucleotide sequences.  This method uses
transition and transversion rates and takes into account the deviation
of GC content from the expected value of 50 %. Gap and ambiguous
positions are ignored.

<p>

<pre>
P = transitions/npos
Q = transversions/npos

npos - number of positions scored

GC1 = GC fraction in sequence 1
GC2 = GC fraction in sequence 2
C = GC1 + GC2 - 2*GC1*GC2

distance = -C ln(1-P/C-Q) - 0.5(1-C) ln(1-2Q)
</pre>

<p>

Reference:
<br>
K. Tamura, Mol. Biol. Evol. 1992, 9, 678.


<h3>Jin-Nei Gamma distance</h3>

This method applies to nucleotides only. This again uses transition and
transversion rates. As with the Kimura two parameter method, gaps and
ambiguous symbols other than R and Y are not oncluded in the score. The
shape parameter, i.e. "a", is the square of the inverse of the
coefficient of variation of the average substitution,

<p>

<pre>
L = average substituition = transition_rate + 2 * transversion_rate
a = (average L)^2/(variance of L)

P = transitions/npos
Q = transversions/npos

npos - number of positions scored

distance = 0.5 * a ((1-2P-Q)^(-1/a) + 0.5 (1-2Q)^(-1/a) -3/2)
</pre>

<p>

It is suggested [Jin et al.], in general, that the distance be
calculated with an a-value of 1. However, the user can specify their own
value, using the "-parametera" option, or calculate for each pair of
sequence, using "-calculatea".

<p>

Reference:
<br>
L. Jin and M. Nei, Mol. Biol. Evol. 1990, 7, 82.


<h3>Kimura Protein distance</h3>

This method is used for proteins only. Gaps are ignored and only exact
matches and ambiguity codes contribute to the match score.

<p>

<pre>
S = m/npos

m  - exact match
npos - number of positions scored

D = 1-S
distance = -ln(1 - D - 0.2D^2)
</pre>

<p>

Reference:
<br>
M. Kimura, The Neutral Theory of Molecular Evolution, Camb. Uni. Press,
Camb., 1983.















<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->


Here is a sample session with distmat:

<p>

<pre>
% distmat pax.align
Creates a distance matrix from multiple alignments
Multiple substitution correction methods for proteins
         0 : Uncorrected
         1 : Jukes-Cantor
         2 : Kimura Protein
Method to use [0]: 2
Output file [outfile.distmat]: 
</pre>












<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers (* if not always prompted):
  [-msf]               seqset     File containing a sequence alignment.
*  -nucmethod          list       Multiple substitution correction methods for
                                  nucleotides.
*  -protmethod         list       Multiple substitution correction methods for
                                  proteins.
  [-outf]              outfile    Enter a name for the distance matrix

   Optional qualifiers (* if not always prompted):
*  -ambiguous          bool       Option to use the abiguous codes in the
                                  calculation of the Jukes-Cantor method or if
                                  the sequences are proteins.
*  -gapweight          float      Option to weight gaps in the uncorrected
                                  (nucleotide) and Jukes-Cantor distance
                                  methods.
*  -position           integer    Choose base positions to analyse in each
                                  codon i.e. 123 (all bases), 12 (the first
                                  two bases), 1, 2, or 3 individual bases.
*  -calculatea         bool       This will force the calculation of the
                                  a-parameter in the Jin-Nei Gamma distance
                                  calculation, otherwise the default is 1.0
                                  (see -parametera option).
*  -parametera         float      User defined a parameter to be use in the
                                  Jin-Nei Gamma distance calculation. The
                                  suggested value to be used is 1.0 [Jin et
                                  al.] and this is the default.

   Advanced qualifiers: (none)

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-msf]<br>(Parameter 1)</td>
<td>File containing a sequence alignment.</td>
<td>Readable sequences</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-nucmethod</td>
<td>Multiple substitution correction methods for nucleotides.</td>
<td><table><tr><td>0</td> <td><i>(Uncorrected)</i></td></tr><tr><td>1</td> <td><i>(Jukes-Cantor)</i></td></tr><tr><td>2</td> <td><i>(Kimura)</i></td></tr><tr><td>3</td> <td><i>(Tamura)</i></td></tr><tr><td>4</td> <td><i>(Tajima-Nei)</i></td></tr><tr><td>5</td> <td><i>(Jin-Nei Gamma)</i></td></tr></table></td>
<td>0</td>
</tr>

<tr>
<td>-protmethod</td>
<td>Multiple substitution correction methods for proteins.</td>
<td><table><tr><td>0</td> <td><i>(Uncorrected)</i></td></tr><tr><td>1</td> <td><i>(Jukes-Cantor)</i></td></tr><tr><td>2</td> <td><i>(Kimura Protein)</i></td></tr></table></td>
<td>0</td>
</tr>

<tr>
<td>[-outf]<br>(Parameter 2)</td>
<td>Enter a name for the distance matrix</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.distmat</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-ambiguous</td>
<td>Option to use the abiguous codes in the calculation of the Jukes-Cantor method or if the sequences are proteins.</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-gapweight</td>
<td>Option to weight gaps in the uncorrected (nucleotide) and Jukes-Cantor distance methods.</td>
<td>Any integer value</td>
<td>0.</td>
</tr>

<tr>
<td>-position</td>
<td>Choose base positions to analyse in each codon i.e. 123 (all bases), 12 (the first two bases), 1, 2, or 3 individual bases.</td>
<td>Any integer value</td>
<td>123</td>
</tr>

<tr>
<td>-calculatea</td>
<td>This will force the calculation of the a-parameter in the Jin-Nei Gamma distance calculation, otherwise the default is 1.0 (see -parametera option).</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-parametera</td>
<td>User defined a parameter to be use in the Jin-Nei Gamma distance calculation. The suggested value to be used is 1.0 [Jin et al.] and this is the default.</td>
<td>Any integer value</td>
<td>1.0</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

It reads in a normal multiple sequence alignment file.

<p>

The quality of the alignment is of paramount importance in obtaining
meaningful information from this analysis. 











<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->

The output from the program is a file containing a matrix of the
calculated distances between each of the input aligned sequences.  The
distances are expressed in terms of the number of substitutions per 100
bases or amino acids. 

<p>

The output from the example usage above is:

<p>

<hr>
<pre>
Distance Matrix
---------------

Using the Kimura correction method
Gap weighting is 0.000000

            1       2       3       4       5       6       7       8       9       10
          0.00   96.10  137.41  128.65  161.04  160.26  157.46  154.14  164.22  152.59          PAX4_HUMAN 1
                  0.00  111.86  109.96  156.25  149.70  143.75  135.71  150.60  146.87          PAX6_HUMAN 2
                          0.00   26.21  131.54  143.54  162.95  151.39  163.56  159.78          PAX3_HUMAN 3
                                  0.00  145.45  138.76  158.79  149.96  167.26  161.82          PAX7_HUMAN 4
                                          0.00   44.29  120.84  123.00  131.69  130.22          PAX1_HUMAN 5
                                                  0.00  123.56  130.21  131.64  130.17          PAX9_HUMAN 6
                                                          0.00   36.43   53.12   64.32          PAX2_HUMAN 7
                                                                  0.00   60.88   73.82          PAX5_HUMAN 8
                                                                          0.00   20.37          PX8A_HUMAN 9
</pre>
<hr>
















<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.
	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

   -->

None.




<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

See the following for details of the methods used:

<p>

<ol>
<li>"Phylogenetic Inference", Swofford, Olsen, Waddell, and Hillis, in Molecular Systematics, 2nd ed., Sinauer Ass., Inc., 1996, Ch. 11.
<li>F. Tajima and M. Nei, Mol. Biol. Evol. 1984, 1, 269.
<li>M. Kimura, J. Mol. Evol. 1980, 16, 111.
<li>K. Tamura, Mol. Biol. Evol. 1992, 9, 678.
<li>L. Jin and M. Nei, Mol. Biol. Evol. 1990, 7, 82.
<li>M. Kimura, The Neutral Theory of Molecular Evolution, Camb. Uni. Press, Camb., 1983.
</ol>







<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

The quality of the alignment is of paramount importance in obtaining
meaningful information from this analysis. 







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

This application was written by Tim Carver <a
href="mailto:tcarver@hgmp.mrc.ac.uk">(tcarver@hgmp.mrc.ac.uk)</a>





<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (March 2001) - Tim Carver










<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->








</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

