<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: domainer
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">
<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>
<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>
</td>
<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: domainer
</H2>
</font>
</td>
</tr>
</table>
</td>
</tr>
</table>
<p><hr><p>

<!--END OF HEADER-->

<H1>
Program domainer
</H1>
<H2>
    Function
</H2>
Build domain coordinate files

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
        Algorithms
  -->

Nearly all proteins have structural similarities with other proteins
and, in some of these cases, share a common evolutionary origin.  A
knowledge of these relationships is crucial to our understanding of the
evolution of proteins and of development.  It will also play an
important role in the analysis of the sequence data that is being
produced by worldwide genome projects. 

<p>

The 
<A HREF="http://scop.mrc-lmb.cam.ac.uk/scop/">SCOP</A> 
database aims to provide a detailed and comprehensive
description of the structural and evolutionary relationships between all
proteins whose structure is known, including all entries in the Protein
Data Bank 
(<A HREF="http://pdb-browsers.ebi.ac.uk//">PDB</A>).  

<p>

<b>domainer</b> reads in an EMBL-like format SCOP classification file
generated by the EMBOSS applications 
<a href="scope.html"><b>scope</b></a>
or 
<a href="nrscope.html"><b>nrscope</b></a>, 
and EMBL-like format clean protein coordinate files generated by the
<a href="coorde.html"><b>coorde</b></a> application. 

(not currently in EMBOSS, email Jon Ison 
<a href="mailto:jison@hgmp.mrc.ac.uk">jison@hgmp.mrc.ac.uk</a>) 

For each domain in the scop classification file <b>domainer</b> writes clean
domain coordinate files in EMBL-like and PDB formats.  
Each of these output files contains coordinates for a single SCOP domain.  In
cases where multiple models were determined, the data in the domain
files correspond to the first model.  
In the rare cases where a domain
is comprised of more than one chain, the data will be presented as
belonging to a single chain (i.e.  a single sequence, chain identifier
etc will be given). 













<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->


Here is a sample session with domainer:
<p>

<pre>

% domainer
Build domain coordinate files
Name of scop file for input (embl-like format) [Escop.dat]: /data/scop/Escop.dat
Location of coordinate files for input (embl-like format) [./]: /data/cpdb/
Location of coordinate files for output (embl-like format) [./]:
Extension of coordinate files (embl-like format) [.pxyz]:
Location of coordinate files for output (pdb format) [./]:
Extension of coordinate files (pdb format) [.ent]:
Name of log file for the embl-like format build [domainer.log1]: log.1
Name of log file for the pdb format build [domainer.log2]: log.2
D3SDHA_
D3SDHB_
D3HBIA_
D3HBIB_
D4SDHA_
D4SDHB_
D4HBIA_
D4HBIB_
D5HBIA_
D5HBIB_
D7HBIA_
D7HBIB_

</pre>












<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-scop]              infile     Name of scop file for input (embl-like
                                  format)
  [-cpdb]              string     Location of coordinate files for input
                                  (embl-like format)
  [-cpdbscop]          string     Location of coordinate files for output
                                  (embl-like format)
  [-cpdbextn]          string     Extension of coordinate files (embl-like
                                  format)
  [-pdbscop]           string     Location of coordinate files for output (pdb
                                  format)
  [-pdbextn]           string     Extension of coordinate files (pdb format)
  [-cpdberrf]          outfile    Name of log file for the embl-like format
                                  build
  [-pdberrf]           outfile    Name of log file for the pdb format build

   Optional qualifiers: (none)
   Advanced qualifiers: (none)
   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-scop]<br>(Parameter 1)</td>
<td>Name of scop file for input (embl-like format)</td>
<td>Input file</td>
<td>Escop.dat</td>
</tr>

<tr>
<td>[-cpdb]<br>(Parameter 2)</td>
<td>Location of coordinate files for input (embl-like format)</td>
<td>Any string is accepted</td>
<td>./</td>
</tr>

<tr>
<td>[-cpdbscop]<br>(Parameter 3)</td>
<td>Location of coordinate files for output (embl-like format)</td>
<td>Any string is accepted</td>
<td>./</td>
</tr>

<tr>
<td>[-cpdbextn]<br>(Parameter 4)</td>
<td>Extension of coordinate files (embl-like format)</td>
<td>Any string is accepted</td>
<td>.pxyz</td>
</tr>

<tr>
<td>[-pdbscop]<br>(Parameter 5)</td>
<td>Location of coordinate files for output (pdb format)</td>
<td>Any string is accepted</td>
<td>./</td>
</tr>

<tr>
<td>[-pdbextn]<br>(Parameter 6)</td>
<td>Extension of coordinate files (pdb format)</td>
<td>Any string is accepted</td>
<td>.ent</td>
</tr>

<tr>
<td>[-cpdberrf]<br>(Parameter 7)</td>
<td>Name of log file for the embl-like format build</td>
<td>Output file</td>
<td>domainer.log1</td>
</tr>

<tr>
<td>[-pdberrf]<br>(Parameter 8)</td>
<td>Name of log file for the pdb format build</td>
<td>Output file</td>
<td>domainer.log2</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->


The EMBL-like format used for the input clean protein data (and the output
domain format) uses the following records:

<p>

(1)  ID - Either the 4-character PDB identifier code (for clean protein 
coordinate files) or the 7-character domain identifier code taken from scop
(for domain coordinate files; see documentation for the EMBOSS application 
<a href="scope.html"><b>scope</b></a> 
for further info.)
<p>

(2)  DE - compound information.  Text from the COMPND records from the 
original pdb file are given.
<p>

(3)  OS - protein source information.  Text from the SOURCE records from the 
original pdb file are given.
<p>

(4)  EX - experimental information. The text 'nmr_or_model' (for nuclear  
magnetic resonance and model structures) or 'xray' (for structures 
determined by X-ray crystallography) appears as appropriate after the text
'METHOD'.  The resolution of X-ray structures, or '0' for structures of type
'nmr_or_model', is given after 'RESO'.  The number of models and number of 
polypeptide chains are given after 'NMOD' and 'NCHA' respectively. For 
domain coordinate files a 1 is always given. Following the EX record, the 
file will have a section containing a CN, IN and SQ records (see below) for
each chain.
<p>

(5)  CN - chain number.  The number given in brackets after this record 
indicates the start of a section of chain-specific data.
<p>

(6)  IN - chain specific data. The character given after ID is the PDB chain 
identifier (a '.' is given in cases where a chain identifier was not 
specified in the pdb file or, for domain coordinate files, the domain is 
comprised of more than one domain).  The number of amino acid residues 
comprising the chain (or the chains from which a domain is comprised) is 
given after NR. The number of atoms in heterogens and water molecules are 
given after NH and NW respectively. Domain coordinate files do not include
coordinates for these groups so a value of 0 is always given.
<p>

(7)  SQ - protein sequence. The number of residues is given before AA on the
first line. The protein sequence is given on subsequent lines. 
<p>

(8)  CO - coordinate data.  The columns of the records are as follows.
<OL>
<LI>CO is always given.
<LI>Model number (always 1 for domain coordinate files).
<LI>Chain number (always 1 for domain coordinate files).
<LI>Either P (a protein atom), H (a heterogen atom) or W (an atom in a water molecule).
<LI>Position of the residue in the protein sequence given in the SQ record (for protein atoms) or a sequential count of the atoms (for heterogens and water).
<LI>Residue number according to the original pdb file, or or a sequential count of the atoms (for heterogens and water).
<LI>Single character amino acid code or a '.' (for heterogens and water).
<LI>3-character residue identifier code.
<LI>Atom type.
<LI>The x orthogonal coordinate.
<LI>The y orthogonal coordinate.
<LI>The z orthogonal coordinate.
<LI>Occupancy.
<LI>Temperature factor.
</OL>
<p>

(9)  XX - Used for spacing.
<p>

(10) // - Given on the last line of the file only.
<p>








<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->

The PDB format is explained at:
<br>
<A HREF="http://www.rcsb.org/pdb/docs/format/pdbguide2.2/guide2.2_frame.html">
http://www.rcsb.org/pdb/docs/format/pdbguide2.2/guide2.2_frame.html</A>
<br>

<b>domainer</b> writes the following records in PDB format:
<p>

(1)  HEADER - bibliographic information.  The text 'CLEANED-UP PDB FILE FOR 
SCOP DOMAIN XXXXXXX' is always given (where XXXXXXX is a 7-character domain 
identifier code).
<p>

(2)  TITLE - bibliographic information. The text ' THIS FILE IS MISSING 
MOST RECORDS FROM THE ORIGINAL PDB FILE' is always given.
<p>

(3)  COMPND - compound information.  The COMPND records from the original  
pdb file are given.
<p>

(4)  SOURCE - protein source information.  The SOURCE records from the 
original PDB file are given.
<p>

(5)  REMARK - remark records. Remark records are used for spacing. One 
REMARK line containing the protein resolution is always given.
<p>

(6)  SEQRES - protein sequence.  
<p>

(7)  ATOM - atomic coordinates.
<p>

(8)  TER - indicates the end of a chain.

<p>

The following is an example of an excerpt from an output clean domain
coordinate file (PDB format):

<p>

<hr>
<pre>
HEADER     CLEANED-UP PDB FILE FOR SCOP DOMAIN D1HBBA_
TITLE      THIS FILE IS MISSING MOST RECORDS FROM THE ORIGINAL PDB FILE
COMPND     HEMOGLOBIN A (DEOXY, LOW SALT, 100MM CL)
SOURCE     HUMAN (HOMO SAPIENS)
REMARK
REMARK     RESOLUTION. 1.90  ANGSTROMS.
REMARK
SEQRES   1 A  141  VAL LEU SER PRO ALA ASP LYS THR ASN VAL LYS ALA ALA
SEQRES   2 A  141  TRP GLY LYS VAL GLY ALA HIS ALA GLY GLU TYR GLY ALA
SEQRES   3 A  141  GLU ALA LEU GLU ARG MET PHE LEU SER PHE PRO THR THR
SEQRES   4 A  141  LYS THR TYR PHE PRO HIS PHE ASP LEU SER HIS GLY SER
SEQRES   5 A  141  ALA GLN VAL LYS GLY HIS GLY LYS LYS VAL ALA ASP ALA
SEQRES   6 A  141  LEU THR ASN ALA VAL ALA HIS VAL ASP ASP MET PRO ASN
SEQRES   7 A  141  ALA LEU SER ALA LEU SER ASP LEU HIS ALA HIS LYS LEU
SEQRES   8 A  141  ARG VAL ASP PRO VAL ASN PHE LYS LEU LEU SER HIS CYS
SEQRES   9 A  141  LEU LEU VAL THR LEU ALA ALA HIS LEU PRO ALA GLU PHE
SEQRES  10 A  141  THR PRO ALA VAL HIS ALA SER LEU ASP LYS PHE LEU ALA
SEQRES  11 A  141  SER VAL SER THR VAL LEU THR SER LYS TYR ARG
ATOM      1  N   VAL A   1       7.155  17.725   4.424  1.00 37.82           N
ATOM      2  CA  VAL A   1       7.854  18.800   3.718  1.00 35.10           C
ATOM      3  C   VAL A   1       9.366  18.565   3.754  1.00 31.92           C
ATOM      4  O   VAL A   1       9.861  17.961   4.721  1.00 35.01           O
ATOM      5  CB  VAL A   1       7.529  20.168   4.360  1.00 47.63           C
ATOM      6  CG1 VAL A   1       7.806  21.300   3.369  1.00 62.84           C
ATOM      7  CG2 VAL A   1       6.136  20.244   4.936  1.00 54.85           C
ATOM      8  N   LEU A   2      10.032  19.062   2.731  1.00 27.38           N
ATOM      9  CA  LEU A   2      11.496  18.967   2.657  1.00 23.24           C
ATOM     10  C   LEU A   2      12.077  20.110   3.496  1.00 22.99           C
ATOM     11  O   LEU A   2      11.672  21.259   3.289  1.00 25.22           O
ATOM     12  CB  LEU A   2      11.924  19.005   1.204  1.00 18.04           C
ATOM     13  CG  LEU A   2      11.563  17.855   0.286  1.00 17.80           C
ATOM     14  CD1 LEU A   2      12.166  18.109  -1.097  1.00 20.08           C
ATOM     15  CD2 LEU A   2      12.116  16.542   0.839  1.00 13.84           C
ATOM     16  N   SER A   3      12.979  19.784   4.391  1.00 22.22           N
ATOM     17  CA  SER A   3      13.652  20.792   5.257  1.00 20.53           C
ATOM     18  C   SER A   3      14.871  21.318   4.505  1.00 18.31           C
ATOM     19  O   SER A   3      15.273  20.709   3.496  1.00 17.73           O
ATOM     20  CB  SER A   3      14.084  20.042   6.534  1.00 17.61           C
</pre>
<hr>

<p>

<b>domainer</b> also writes out the clean domain coordinate files in
EMBL-like format.  The format for this EMBL-like data is described in
the <b>Input File format</b> section of this document as it used the
same format for the input clean protein EMBL-like data and the output
clean domain EMBL-like data. 

<p>

The following is an example of an excerpt from an output clean domain
coordinate file (EMBL-like format):

<p>

<hr>
<pre>
ID   D1HBBA_
XX
DE   Co-ordinates for SCOP domain D1HBBA_
XX
OS   See Escop.dat for domain classification
XX
EX   METHOD xray; RESO 1.90; NMOD 1; NCHA 1;
XX
CN   [1]
XX
IN   ID A; NR 141; NH 0; NW 0;
XX
SQ   SEQUENCE   141 AA;  15127 MW;  5EC7DB1E CRC32;
     VLSPADKTNV KAAWGKVGAH AGEYGAEALE RMFLSFPTTK TYFPHFDLSH GSAQVKGHGK
     KVADALTNAV AHVDDMPNAL SALSDLHAHK LRVDPVNFKL LSHCLLVTLA AHLPAEFTPA
     VHASLDKFLA SVSTVLTSKY R
XX
CO   1    1    P    1     1     V    VAL    N      7.155   17.725 4.424     1.00    37.82
CO   1    1    P    1     1     V    VAL    CA     7.854   18.800 3.718     1.00    35.10
CO   1    1    P    1     1     V    VAL    C      9.366   18.565 3.754     1.00    31.92
CO   1    1    P    1     1     V    VAL    O      9.861   17.961 4.721     1.00    35.01
CO   1    1    P    1     1     V    VAL    CB     7.529   20.168 4.360     1.00    47.63
CO   1    1    P    1     1     V    VAL    CG1    7.806   21.300 3.369     1.00    62.84
CO   1    1    P    1     1     V    VAL    CG2    6.136   20.244 4.936     1.00    54.85
CO   1    1    P    2     2     L    LEU    N     10.032   19.062 2.731     1.00    27.38
CO   1    1    P    2     2     L    LEU    CA    11.496   18.967 2.657     1.00    23.24
CO   1    1    P    2     2     L    LEU    C     12.077   20.110 3.496     1.00    22.99
CO   1    1    P    2     2     L    LEU    O     11.672   21.259 3.289     1.00    25.22
</pre>
<hr>

<p>

<b>domainer</b> generates a log file, an excerpt of which is shown below. 
If there is a problem in processing a domain, three lines containing the
record '//', the domain identifier code and an error message
respectively are written.  The text 'WARN filename not found' is given
in cases where a clean coordinate file could not be found.  'ERROR
filename file read error' or 'ERROR filename file write error' will be
reported when an error was encountered during a file read or write
respectively.  Various other error messages may also be given (in case
of difficulty email Jon Ison, 
<a href="mailto:jison@hgmp.mrc.ac.uk">jison@hgmp.mrc.ac.uk</a>). 

<p>
<hr>
<pre>
//
DS002__
WARN  Could not open for reading cpdb file s002.pxyz
//
DS003__
WARN  Could not open for reading cpdb file s003.pxyz
</pre>
<hr>
<p>








<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.
	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

   -->

The ready-made input and output data for <b>domainer</b> may be
downloaded from the HGMP:

<p>

<a href="ftp://ftp.hgmp.mrc.ac.uk/pub/databases/cpdb/">EMBL-like format clean protein coordinate files</a>
<br>
<a href="ftp://ftp.hgmp.mrc.ac.uk/pub/databases/cpdbscop/">EMBL-like format clean domain coordinate files</a>
<br>
<a href="ftp://ftp.hgmp.mrc.ac.uk/pub/databases/pdbscop/">PDB-format clean domain coordinate files</a>















<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="cutgextract.html">cutgextract</a></td><td>Extract data from CUTG</td></tr>
<tr><td><a href="nrscope.html">nrscope</a></td><td>Converts redundant EMBL-format SCOP file to non-redundant one</td></tr>
<tr><td><a href="pdbtosp.html">pdbtosp</a></td><td>Convert raw swissprot:pdb equivalence file to embl-like format</td></tr>
<tr><td><a href="printsextract.html">printsextract</a></td><td>Extract data from PRINTS</td></tr>
<tr><td><a href="prosextract.html">prosextract</a></td><td>Builds the PROSITE motif database for patmatmotifs to search</td></tr>
<tr><td><a href="rebaseextract.html">rebaseextract</a></td><td>Extract data from REBASE</td></tr>
<tr><td><a href="scope.html">scope</a></td><td>Convert raw scop classification file to embl-like format</td></tr>
<tr><td><a href="scopparse.html">scopparse</a></td><td>Reads raw scop classifications file and writes embl-like format scop classification file</td></tr>
<tr><td><a href="seqnr.html">seqnr</a></td><td>Converts redundant database results to a non-redundant set of hits</td></tr>
<tr><td><a href="tfextract.html">tfextract</a></td><td>Extract data from TRANSFAC</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

This application was written by Jon Ison <a
href="mailto:jison@hgmp.mrc.ac.uk">(jison@hgmp.mrc.ac.uk)</a>

















<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (Jan 2001) - Jon Ison.















<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
This program is intended to be used by everyone and everything, from naive users to embedded scripts.

   -->

This program is intended to be run by EMBOSS site maintainers or those
responsible for setting up and maintaining protein 3D structural data
for use by others.











<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->








</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

