<!-- START OF HEADER -->
<HTML><HEAD>
<TITLE> EMBASSY: DOMAINSEQS documentation. </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

<H2> DOMAINSEQS documentation
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<!-- END OF HEADER -->





<!-- CONTENTS
     This always includes the sections below.
     Other subsections can be added for individual applications.
-->
<br><H2>CONTENTS </H2>
<b> <a href="#1.0">1.0     SUMMARY                   </a></b><br>
<b> <a href="#2.0">2.0     INPUTS & OUTPUTS          </a></b><br>
<b> <a href="#3.0">3.0     INPUT FILE FORMAT         </a></b><br>
<b> <a href="#4.0">4.0     OUTPUT FILE FORMAT        </a></b><br>
<b> <a href="#5.0">5.0     DATA FILES                </a></b><br>
<b> <a href="#6.0">6.0     USAGE                     </a></b><br>    
<b> <a href="#7.0">7.0     KNOWN BUGS & WARNINGS     </a></b><br>    
<b> <a href="#8.0">8.0     NOTES                     </a></b><br>
<b> <a href="#9.0">9.0     DESCRIPTION               </a></b><br>
<b> <a href="#10.0">10.0   ALGORITHM                 </a></b><br>
<b> <a href="#11.0">11.0   RELATED APPLICATIONS      </a></b><br>
<b> <a href="#12.0">12.0   DIAGNOSTIC ERROR MESSAGES </a></b><br>
<b> <a href="#13.0">13.0   AUTHORS                   </a></b><br>
<b> <a href="#14.0">14.0   REFERENCES                </a></b><br>

<!-- SUMMARY
     Succint description of the application, particularly its inputs, outputs
     and what it does.  The same text is given at the top of the source (.c)
     file and in the <documentation> attribute of the <application definition>
     of the ACD file.
-->
<a name="1.0"></a>
<br><br><br><H2> 1.0   SUMMARY  </H2>



Adds sequence records to a DCF file



<!-- INPUTS & OUTPUTS
     Short summary of the application inputs and outputs in its different 
     modes of usage (if appropriate). More detail than the summary.
-->
<a name="2.0"></a>
<br><br><br><H2> 2.0   INPUTS & OUTPUTS          </H2>
DOMAINSEQS parses a DCF file (domain classification file) and writes a file 
containing the same data, except that domain sequence 
information derived from structural and, optionally, sequence databases are
added.  
Domain sequences are taken from domain CCF files (clean coordinate files) 
and (optionally) the swissprot database.  

If the swissprot sequence is used, DOMAINSEQS requires a swissprot:PDB 
equivalence file that gives the accession number of each PDB file 
corresponding to the domains in the DCF file.  

The path for the CCF files (input) and names of DCF files (input and output) are specified by the user.  A log file is also written.






<!-- INPUT FILE FORMAT 
     Description and example(s) of input file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.

     Cannot use the test data files because they might be empty or need 
     hand-editing 

     Use "<b>DOMAINSEQS</b> reads any normal sequence USAs." if
     appropriate.
-->
<a name="3.0"></a>
<br><br><br><H2> 3.0   INPUT FILE FORMAT         </H2>
The format of the DCF file is described in the <a href="scopparse.html">SCOPPARSE documentation.</a>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: ../scopparse-keep/all.scop</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   D1CS4A_
XX
EN   1CS4
XX
TY   SCOP
XX
SI   53931 CL; 54861 FO; 55073 SF; 55074 FA; 55077 DO; 55078 SO; 39418 DD;
XX
CL   Alpha and beta proteins (a+b)
XX
FO   Ferredoxin-like
XX
SF   Adenylyl and guanylyl cyclase catalytic domain
XX
FA   Adenylyl and guanylyl cyclase catalytic domain
XX
DO   Adenylyl cyclase VC1, domain C1a
XX
OS   Dog (Canis familiaris)
XX
NC   1
XX
CN   [1]
XX
CH   A CHAIN; . START; . END;
//
ID   D1II7A_
XX
EN   1II7
XX
TY   SCOP
XX
SI   53931 CL; 56299 FO; 56300 SF; 64427 FA; 64428 DO; 64429 SO; 62415 DD;
XX
CL   Alpha and beta proteins (a+b)
XX
FO   Metallo-dependent phosphatases
XX
SF   Metallo-dependent phosphatases
XX
FA   DNA double-strand break repair nuclease
XX
DO   Mre11
XX
OS   Archaeon Pyrococcus furiosus
XX
NC   1
XX
CN   [1]
XX
CH   A CHAIN; . START; . END;
//
</pre>
</td></tr></table><p>
<p><h3>File: ../pdbtosp-keep/Epdbtosp.dat</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
EN   101M
XX
NE   1
XX
IN   MYG_PHYCA ID; P02185 ACC;
XX
//
EN   102L
XX
NE   1
XX
IN   LYCV_BPT4 ID; P00720 ACC;
XX
//
EN   102M
XX
NE   1
XX
IN   MYG_PHYCA ID; P02185 ACC;
XX
//
EN   103L
XX
NE   1
XX
IN   LYCV_BPT4 ID; P00720 ACC;
XX
//
EN   103M
XX
NE   1
XX
IN   MYG_PHYCA ID; P02185 ACC;
XX
//
EN   9XIA
XX
NE   1
XX
IN   XYLA_STRRU ID; P24300 ACC;
XX
//
EN   9XIM
XX
NE   1
XX
IN   XYLA_ACTMI ID; P12851 ACC;
XX
//
</pre>
</td></tr></table><p>


<!-- OUTPUT FILE FORMAT 
     Description and example(s) of output file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.  
     If the standard description of the avalable report formats is required, 
     use:   #include file="inc/reportformats.ihtml"
     Use "Outputs a graph to the specified graphics device."
     or  "outputs a report format file. The default format is ..."
     if appropriate.
-->
<a name="4.0"></a>
<br><br><br><H2> 4.0   OUTPUT FILE FORMAT         </H2>
The format of the DCF file is described in the <a href="scopparse.html">SCOPPARSE documentation.</a>
<br>
DOMAINSEQS may add the following records:
<UL>
<LI>(1) AC - Accession number of the domain sequence.  This record will only 
be present if the DCF file has been processed using 
DOMAINSEQS and if an accession number for the PDB file corresponding to the 
domain is given in the swissprot:PDB equivalence file (generated by 
PDBTOSP) that DOMAINSEQS makes use of.
<LI>(2) SP - Swissprot code of the domain sequence.  This record will only 
be present if the domain classification file has been processed using 
DOMAINSEQS and if an swissprot code for the PDB file corresponding to the 
domain is given in the swissprot:PDB equivalence file (generated by 
PDBTOSP) that DOMAINSEQS makes use of.
<LI>(3) RA - Position of domain in swissprot sequence.  The integers 
preceeding START and END give the start and end points respectively of the
domain sequence relative to the full-length swissprot sequence.
<LI>(4) SQ - Sequence of the domain according to swissprot.  This sequence
is taken from the swissprot database.  The SQ record will only be 
present if the SCOP classification file has been processed using 
DOMAINSEQS and if an accession number for the PDB file corresponding to the 
domain is given in the swissprot:PDB equivalence file (generated by 
PDBTOSP) that DOMAINSEQS makes use of.
</UL>



<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: all_s.scop</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
ID   D1CS4A_
XX
EN   1CS4
XX
TY   SCOP
XX
SI   53931 CL; 54861 FO; 55073 SF; 55074 FA; 55077 DO; 55078 SO; 39418 DD;
XX
CL   Alpha and beta proteins (a+b)
XX
FO   Ferredoxin-like
XX
SF   Adenylyl and guanylyl cyclase catalytic domain
XX
FA   Adenylyl and guanylyl cyclase catalytic domain
XX
DO   Adenylyl cyclase VC1, domain C1a
XX
OS   Dog (Canis familiaris)
XX
DS   SEQUENCE    52 AA;   5817 MW;  47362A43 CRC32;
     ADIEGFTSLA SQCTAQELVM TLNELFARFD KLAAENHCLR IKILGDCYYC VS
XX
NC   1
XX
CN   [1]
XX
CH   A CHAIN; . START; . END;
//
ID   D1II7A_
XX
EN   1II7
XX
TY   SCOP
XX
SI   53931 CL; 56299 FO; 56300 SF; 64427 FA; 64428 DO; 64429 SO; 62415 DD;
XX
CL   Alpha and beta proteins (a+b)
XX
FO   Metallo-dependent phosphatases
XX
SF   Metallo-dependent phosphatases
XX
FA   DNA double-strand break repair nuclease
XX
DO   Mre11
XX
OS   Archaeon Pyrococcus furiosus
XX
DS   SEQUENCE    65 AA;   7396 MW;  0CFB92A3 CRC32;
     MKFAHLADIH LGYEQFHKPQ REEEFAEAFK NALEIAVQEN VDFILIAGDL FHSSRPSPGT
     LKKAI
XX
NC   1
XX
CN   [1]
XX
CH   A CHAIN; . START; . END;
//
</pre>
</td></tr></table><p>
<p><h3>File: domainseqs.log</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
//
D1CS4A_        
NO_ACCESSION_NUMBER
//
D1II7A_        
NO_ACCESSION_NUMBER
</pre>
</td></tr></table><p>


<!-- DATA FILES         
     Any data files used (e.g. translation table file, substitution matrix 
     etc.  This includes example data file formats if they are not obvious.
     For a standard description of what data files are and how embossdata can
     be used to inspect and retrieve them, use:
     #include file="inc/localfiles.ihtml" 
-->

<a name="5.0"></a>
<br><br><br><H2> 5.0   DATA FILES                </H2>

If the user specified retrieval of sequences from a sequence database, 
DOMAINSEQS uses a swissprot:PDB equivalence file is generated by using 
<a href="pdbtosp.html">PDBTOSP.</a>







<!-- USAGE
     Example usage, as run from the command-line.
     Many examples illustrating different behaviours is good.
-->
<a name="6.0"></a>
<br><br><br><H2> 6.0   USAGE                     </H2>
<H3> 6.1   COMMAND LINE ARGUMENTS </H3>
 
<pre>
   Standard (Mandatory) qualifiers (* if not always prompted):
  [-dcfinfile]         infile     This option specifies the name of DCF file
                                  (domain classification file) (input). A
                                  'domain classification file' contains
                                  classification and other data for domains
                                  from SCOP or CATH, in DCF format
                                  (EMBL-like). The files are generated by
                                  using SCOPPARSE and CATHPARSE. Domain
                                  sequence information can be added to the
                                  file by using DOMAINSEQS.
  [-dpdbdir]           directory  This option specifies the location of domain
                                  CCF file (clean coordinate files) (input).
                                  A 'clean cordinate file' contains coordinate
                                  and other data for a single PDB file or a
                                  single domain from SCOP or CATH, in CCF
                                  format (EMBL-like). The files, generated by
                                  using PDBPARSE (PDB files) or DOMAINER
                                  (domains), contain 'cleaned-up' data that is
                                  self-consistent and error-corrected.
                                  Records for residue solvent accessibility
                                  and secondary structure are added to the
                                  file by using PDBPLUS.
   -getswiss           toggle     Retrieve swissprot sequence.
*  -pdbtospfile        infile     This option specifies the name of the
                                  pdbcodes to swissprot indexing file. The
                                  swissprot:PDB equivalence file is generated
                                  by PDBTOSP
  [-dcfoutfile]        outfile    This option specifies the name of DCF file
                                  (domain classification file) (output). A
                                  'domain classification file' contains
                                  classification and other data for domains
                                  from SCOP or CATH, in DCF format
                                  (EMBL-like). The files are generated by
                                  using SCOPPARSE and CATHPARSE. Domain
                                  sequence information can be added to the
                                  file by using DOMAINSEQS.
   -logfile            outfile    This option specifies the name of log file
                                  for the build. The log file contains
                                  messages about any errors arising while
                                  domainseqs ran.

   Additional (Optional) qualifiers (* if not always prompted):
*  -datafile           matrixf    This option specifies the residue
                                  substitution matrix, which is used for
                                  sequence comparison.
*  -gapopen            float      This option specifies the gap insertion
                                  penalty. This is the score taken away when a
                                  gap is created. The best value depends on
                                  the choice of comparison matrix. The default
                                  value assumes you are using the EBLOSUM62
                                  matrix for protein sequences, and the
                                  EDNAFULL matrix for nucleotide sequences.
*  -gapextend          float      This option specifies the gap extension
                                  penalty. This is added to the standard gap
                                  penalty for each base or residue in the gap.
                                  This is how long gaps are penalized.
                                  Usually you will expect a few long gaps
                                  rather than many short gaps, so the gap
                                  extension penalty should be lower than the
                                  gap penalty.

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-dcfoutfile" associated qualifiers
   -odirectory3        string     Output directory

   "-logfile" associated qualifiers
   -odirectory         string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report deaths


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-dcfinfile]<br>(Parameter 1)</td>
<td>This option specifies the name of DCF file (domain classification file) (input). A 'domain classification file' contains classification and other data for domains from SCOP or CATH, in DCF format (EMBL-like). The files are generated by using SCOPPARSE and CATHPARSE. Domain sequence information can be added to the file by using DOMAINSEQS.</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-dpdbdir]<br>(Parameter 2)</td>
<td>This option specifies the location of domain CCF file (clean coordinate files) (input). A 'clean cordinate file' contains coordinate and other data for a single PDB file or a single domain from SCOP or CATH, in CCF format (EMBL-like). The files, generated by using PDBPARSE (PDB files) or DOMAINER (domains), contain 'cleaned-up' data that is self-consistent and error-corrected. Records for residue solvent accessibility and secondary structure are added to the file by using PDBPLUS.</td>
<td>Directory</td>
<td>./</td>
</tr>

<tr>
<td>-getswiss</td>
<td>Retrieve swissprot sequence.</td>
<td>Toggle value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-pdbtospfile</td>
<td>This option specifies the name of the pdbcodes to swissprot indexing file. The swissprot:PDB equivalence file is generated by PDBTOSP</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-dcfoutfile]<br>(Parameter 3)</td>
<td>This option specifies the name of DCF file (domain classification file) (output). A 'domain classification file' contains classification and other data for domains from SCOP or CATH, in DCF format (EMBL-like). The files are generated by using SCOPPARSE and CATHPARSE. Domain sequence information can be added to the file by using DOMAINSEQS.</td>
<td>Output file</td>
<td>domainseqs.out</td>
</tr>

<tr>
<td>-logfile</td>
<td>This option specifies the name of log file for the build. The log file contains messages about any errors arising while domainseqs ran.</td>
<td>Output file</td>
<td>domainseqs.log</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-datafile</td>
<td>This option specifies the residue substitution matrix, which is used for sequence comparison.</td>
<td>Comparison matrix file in EMBOSS data path</td>
<td>EBLOSUM62</td>
</tr>

<tr>
<td>-gapopen</td>
<td>This option specifies the gap insertion penalty. This is the score taken away when a gap is created. The best value depends on the choice of comparison matrix. The default value assumes you are using the EBLOSUM62 matrix for protein sequences, and the EDNAFULL matrix for nucleotide sequences.</td>
<td>Floating point number from 1.0 to 100.0</td>
<td>10.0 for any sequence</td>
</tr>

<tr>
<td>-gapextend</td>
<td>This option specifies the gap extension penalty. This is added to the standard gap penalty for each base or residue in the gap. This is how long gaps are penalized. Usually you will expect a few long gaps rather than many short gaps, so the gap extension penalty should be lower than the gap penalty.</td>
<td>Floating point number from 0.0 to 10.0</td>
<td>0.5 for any sequence</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>

 


<H3> 6.2 EXAMPLE SESSION </H3>
An example of interactive use of DOMAINSEQS is shown below.

<b>Here is a sample session with domainseqs</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>domainseqs </b>
Adds sequence records to a DCF file.
Name of DCF file (domain classification file) (input).: <b>../scopparse-keep/all.scop</b>
Location of domain CCF file (clean coordinate files) (input). [./]: <b>../domainer-keep</b>
Retrieve swissprot sequence. [N]: <b>Y</b>
Name of the pdbcodes to swissprot indexing file.: <b>../pdbtosp-keep/Epdbtosp.dat</b>
Name of DCF file (domain classification file) (output). [domainseqs.out]: <b>all_s.scop</b>
Name of log file for the build. [domainseqs.log]: <b></b>

//
D1CS4A_        
//
D1II7A_        

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>


<br><br>
<!-- The DCF file (domain classification file) lacking sequence information called /test_data/all.scop was parsed and a DCF file with sequence information called /test_data/domainseqs/all_s.scop was written.  The file Epdbtosp.dat was queried for accession numbers of PDB files corresponding to the domains in all.scop, but no accession numbers were found for either of the domains d1ii7a_ or d1cs4a_.  A log file called /test_data/domainseqs/domainseqs.log was written. -->



<!-- KNOWN BUGS & WARNINGS
     Bugs that have not yet been fixed, easily missued features, problems
     and caveats etc. Potentially stupid things the program will let you do.
-->
<a name="7.0"></a>
<br><br><br><H2> 7.0   KNOWN BUGS & WARNINGS     </H2>

The swissprot:PDB equivalence file available as part of the EMBOSS
distribution does not contain accession numbers for all PDB files and covers
only a relatively small proportion of domains in SCOP.




<!-- NOTES
     Important general remarks, including:
       Restrictions.
       Interesting behaviour.
       Useful things you can do with this program.
       Future plans.
       etc.
-->
<a name="8.0"></a>
<br><br><br><H2> 8.0   NOTES                     </H2>


DOMAINSEQS will not attempt to retrieve a swissprot sequence for any domains 
comprised of more than a single segment, i.e. for domains whose NC record 
in the DCF file has a value other than 1 
(see <a href="scopparse.html">SCOPPARSE documentation</a>).
<br>
The region of sequence to retrieve is identified by alignment of the sequence from the
CCF file (structural database) to the full length sequences in swissprot.   If this were to be done
for segmented domains, the start and end point for the retrieved sequence 
(relative to the full length sequence) might actually include a completely 
different domain.
<br><br>
Consider :
<br> (A) represents a sequence from the coordinate file for a segmented domain D1
<br> (B) represents the full-length swissprot sequence, it includes D1, but D1 is
 split by a second domain D2
<pre> 
                 D1
  (A)  XXXXXXXXXXXXXXXXXXXXX
  
          D1         D2        D1
  (B) XXXXXXXXXX-----------XXXXXXXXXXX
</pre>
  
 It should be clear that for an alignment retrieving B, the start and end 
 points for the retrieved sequence would be misleading.

<br><br>
The user should be aware that sequences from the domain CCF file
for domains comprised of more than a single segment are not biologically 
significant, as the sequences are derived from different segments of one or
more chains.  However, such sequences might be acceptable for redundancy 
calculations (e.g. by using <a href="domainnr.html">DOMAINNR</a>) because 
two redundant domains made of similar fragments will have similar sequences, 
so the redundancy should be detectable.

<H3> 8.1   GLOSSARY OF FILE TYPES    </H3>
<table BORDER CELLSPACING=5 CELLPADDING=5 BGCOLOR="#f5f5ff" >
<tr>
<td><b>FILE TYPE</b></td>
<td><b>FORMAT</b></td>
<td><b>DESCRIPTION</b></td>
<td><b>CREATED BY <b></td>
<td><b>SEE ALSO</b></td>
</tr>
<tr>
<td><b> Domain classification file (for SCOP)</b></td>
<td> DCF format (EMBL-like format for domain classification data). </td>
<td> Classification and other data for domains from SCOP.  </td>
<td> <a href="scopparse.html">SCOPPARSE</a> </td>
<td> Domain sequence information can be added to the file by using DOMAINSEQS.  </td>
</tr>
<tr>
<td><b> Domain classification file (for CATH)</b></td>
<td> DCF format (EMBL-like format for domain classification data). </td>
<td> Classification and other data for domains from CATH.  </td>
<td> <a href="cathparse.html">CATHPARSE</a> </td>
<td> Domain sequence information can be added to the file by using DOMAINSEQS.  </td>
</tr>
<tr>
<td><b> Clean coordinate file (for protein) </b></td>
<td> CCF format (EMBL-like format for protein coordinate and derived data).</td>
<td> Coordinate and other data for a single PDB file.  The data are 'cleaned-up': self-consistent and error-corrected.  </td>
<td> <a href="pdbparse.html">PDBPARSE</a> </td>
<td> Records for residue solvent accessibility and secondary structure are added 
    to the file by using PDBPLUS. </td>
</tr>
<tr>
<td><b> Clean coordinate file (for domain)</b></td>
<td> CCF format (EMBL-like format for protein coordinate and derived data).</td>
<td> Coordinate and other data for a single domain from SCOP or CATH. The data are 'cleaned-up': self-consistent and error-corrected. </td>
<td> <a href="domainer.html">DOMAINER</a> </td>
<td> Records for residue solvent accessibility and secondary structure are added 
to the file by using PDBPLUS.</td>
</tr>
<tr>
<td><b> swissprot:PDB equivalence file</b></td>
<td> EMBL-like format. </td>
<td> A file containing swissprot identifiers for PDB codes. </td>
<td> Included in the EMBOSS distribution</td>
<td> N.A. </td>
</tr>
</table>







None



<!-- DESCRIPTION
     A complete, non-technical, user-level description of the application.
-->
<a name="9.0"></a>
<br><br><br><H2> 9.0   DESCRIPTION               </H2>

Domain sequences are not given in the raw SCOP or CATH parsable files, but are
required for many analyses and for convenience should, ideally, be provided
along with the classification itself.  DOMAINSEQS reads a DCF file (domain 
classificaiton file) that lacks sequence information and writes one 
containing sequence information.





<!-- ALGORITHM
     A technical description of algorithmic aspects, describing exactly how
     the key aspects of the application work.
-->
<a name="10.0"></a>
<br><br><br><H2> 10.0  ALGORITHM                 </H2>

In order to find the start and end of a domain in a sequence from swissprot, 
the domain sequence from the domain CCF file is aligned to the 
full length protein sequence from swissprot.  Alignment is performed first
by string handling and if that fails, by using the EMBOSS implementation of
the Needleman and Wunsch global alignment algorithm.  Gap insertion and 
extension penalties used in the alignments are user-specified.




<!-- RELATED APPLICATIONS
     Other applications that either generate the input, use the output or 
     are in some other way related to the application are described here. 
     (Take this from "Sister applications" in the old documentation)
-->

<a name="11.0"></a>
<br><br><br><H2> 11.0  RELATED APPLICATIONS      </H2>

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="../../apps/aaindexextract.html">aaindexextract</a></td><td>Extract data from AAINDEX</td></tr>
<tr><td><a href="../domalign/allversusall.html">allversusall</a></td><td>Sequence similarity data from all-versus-all comparison</td></tr>
<tr><td><a href="cathparse.html">cathparse</a></td><td>Generates DCF file from raw CATH files</td></tr>
<tr><td><a href="../../apps/cutgextract.html">cutgextract</a></td><td>Extract data from CUTG</td></tr>
<tr><td><a href="../structure/domainer.html">domainer</a></td><td>Generates domain CCF files from protein CCF files</td></tr>
<tr><td><a href="domainnr.html">domainnr</a></td><td>Removes redundant domains from a DCF file</td></tr>
<tr><td><a href="domainsse.html">domainsse</a></td><td>Add secondary structure records to a DCF file</td></tr>
<tr><td><a href="../structure/hetparse.html">hetparse</a></td><td>Converts heterogen group dictionary to EMBL-like format</td></tr>
<tr><td><a href="../structure/pdbparse.html">pdbparse</a></td><td>Parses PDB files and writes protein CCF files</td></tr>
<tr><td><a href="../structure/pdbplus.html">pdbplus</a></td><td>Add accessibility & secondary structure to a CCF file</td></tr>
<tr><td><a href="../structure/pdbtosp.html">pdbtosp</a></td><td>Convert swissprot:PDB codes file to EMBL-like format</td></tr>
<tr><td><a href="../../apps/printsextract.html">printsextract</a></td><td>Extract data from PRINTS</td></tr>
<tr><td><a href="../../apps/prosextract.html">prosextract</a></td><td>Build the PROSITE motif database for use by patmatmotifs</td></tr>
<tr><td><a href="../../apps/rebaseextract.html">rebaseextract</a></td><td>Extract data from REBASE</td></tr>
<tr><td><a href="scopparse.html">scopparse</a></td><td>Generate DCF file from raw SCOP files</td></tr>
<tr><td><a href="../domsearch/seqnr.html">seqnr</a></td><td>Removes redundancy from DHF files</td></tr>
<tr><td><a href="../structure/sites.html">sites</a></td><td>Generate residue-ligand CON files from CCF files</td></tr>
<tr><td><a href="ssematch.html">ssematch</a></td><td>Search a DCF file for secondary structure matches</td></tr>
<tr><td><a href="../../apps/tfextract.html">tfextract</a></td><td>Extract data from TRANSFAC</td></tr>
</table>



<!-- DIAGNOSTIC ERROR MESSAGES
     Description of error messages or log file, if one is written.
-->
<a name="12.0"></a>
<br><br><br><H2> 12.0  DIAGNOSTIC ERROR MESSAGES </H2>
None.






<!-- AUTHORS -->
<a name="13.0"><br><br><br><H2> 13.0  AUTHORS                   </H2></a>
 Ranjeeva Ranasinghe  <a href="mailto:rranasin@hgmp.mrc.ac.uk">(rranasin@hgmp.mrc.ac.uk)</a>
<br><br>Jon Ison <a href="mailto:jison@rfcgr.mrc.ac.uk">(jison@rfcgr.mrc.ac.uk)</a>
<br>MRC Rosalind Franklin Centre for Genomics Research, Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK





<!-- REFERENCES
     Quote the paper where the application was first published, described, used etc. 
-->
<a name="14.0"></a>
<br><br><br><H2> 14.0  REFERENCES                </H2>

Please cite the authors and EMBOSS.
<br><br>
<i>Rice P, Longden I and Bleasby A (2000) "EMBOSS - The European
Molecular Biology Open Software Suite"  Trends in Genetics,
15:276-278.

<p>
See also <a href="http://emboss.sourceforge.net/">http://emboss.sourceforge.net/</a></i>



<H3>14.1 Other useful references  </H3>

<br>

</BODY>
</HTML>
