<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: domainer
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">
<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>
<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>
</td>
<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: domainer
</H2>
</font>
</td>
</tr>
</table>
</td>
</tr>
</table>
<p><hr><p>

<!--END OF HEADER-->

<H1>
Program domainer
</H1>
<H2>
    Function
</H2>
Build domain coordinate files

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
        Algorithms
  -->

Nearly all proteins have structural similarities with other proteins
and, in some of these cases, share a common evolutionary origin.  A
knowledge of these relationships is crucial to our understanding of the
evolution of proteins and of development.  It will also play an
important role in the analysis of the sequence data that is being
produced by worldwide genome projects. 

<p>

The 
<A HREF="http://scop.mrc-lmb.cam.ac.uk/scop/">SCOP</A> 
database aims to provide a detailed and comprehensive
description of the structural and evolutionary relationships between all
proteins whose structure is known, including all entries in the Protein
Data Bank 
(<A HREF="http://pdb-browsers.ebi.ac.uk//">PDB</A>).  

<p>

<b>domainer</b> reads in an EMBL-like format SCOP classification file
generated by the EMBOSS applications <b>scope</b> or <b>nrscope</b>, 
and clean protein coordinate files generated by the <b>coorde</b> application.

(not currently in EMBOSS, email Jon Ison jison@hgmp.mrc.ac.uk) 

For each domain in the scop classification file <b>domainer</b> writes clean
domain coordinate files in EMBL-like and PDB formats.  

<p>

Each of these output files contains coordinates for a single SCOP domain.  In
cases where multiple models were determined, the data in the domain
files correspond to the first model.  

<p>

In the rare cases where a domain
is comprised of more than one chain, the data will be presented as
belonging to a single chain (i.e.  a single sequence, chain identifier
etc will be given). 













<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->


Here is a sample session with domainer:
<p>

<pre>

% domainer
Build domain coordinate files
Name of scop file for input (embl-like format) [Escop.dat]: /data/scop/Escop.dat
Location of coordinate files for input (embl-like format) [./]: /data/cpdb/
Location of coordinate files for output (embl-like format) [./]:
Extension of coordinate files (embl-like format) [.pxyz]:
Location of coordinate files for output (pdb format) [./]:
Extension of coordinate files (pdb format) [.ent]:
Name of log file for the embl-like format build [domainer.log1]: log.1
Name of log file for the pdb format build [domainer.log2]: log.2
D3SDHA_
D3SDHB_
D3HBIA_
D3HBIB_
D4SDHA_
D4SDHB_
D4HBIA_
D4HBIB_
D5HBIA_
D5HBIB_
D7HBIA_
D7HBIB_

</pre>












<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-scop]              infile     Name of scop file for input (embl-like
                                  format)
  [-cpdb]              string     Location of coordinate files for input
                                  (embl-like format)
  [-cpdbscop]          string     Location of coordinate files for output
                                  (embl-like format)
  [-cpdbextn]          string     Extension of coordinate files (embl-like
                                  format)
  [-pdbscop]           string     Location of coordinate files for output (pdb
                                  format)
  [-pdbextn]           string     Extension of coordinate files (pdb format)
  [-cpdberrf]          outfile    Name of log file for the embl-like format
                                  build
  [-pdberrf]           outfile    Name of log file for the pdb format build

   Optional qualifiers: (none)
   Advanced qualifiers: (none)

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-scop]<br>(Parameter 1)</td>
<td>Name of scop file for input (embl-like format)</td>
<td>Input file</td>
<td>Escop.dat</td>
</tr>

<tr>
<td>[-cpdb]<br>(Parameter 2)</td>
<td>Location of coordinate files for input (embl-like format)</td>
<td>Any string is accepted</td>
<td>./</td>
</tr>

<tr>
<td>[-cpdbscop]<br>(Parameter 3)</td>
<td>Location of coordinate files for output (embl-like format)</td>
<td>Any string is accepted</td>
<td>./</td>
</tr>

<tr>
<td>[-cpdbextn]<br>(Parameter 4)</td>
<td>Extension of coordinate files (embl-like format)</td>
<td>Any string is accepted</td>
<td>.pxyz</td>
</tr>

<tr>
<td>[-pdbscop]<br>(Parameter 5)</td>
<td>Location of coordinate files for output (pdb format)</td>
<td>Any string is accepted</td>
<td>./</td>
</tr>

<tr>
<td>[-pdbextn]<br>(Parameter 6)</td>
<td>Extension of coordinate files (pdb format)</td>
<td>Any string is accepted</td>
<td>.ent</td>
</tr>

<tr>
<td>[-cpdberrf]<br>(Parameter 7)</td>
<td>Name of log file for the embl-like format build</td>
<td>Output file</td>
<td>domainer.log1</td>
</tr>

<tr>
<td>[-pdberrf]<br>(Parameter 8)</td>
<td>Name of log file for the pdb format build</td>
<td>Output file</td>
<td>domainer.log2</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->


The EMBL-like format used for clean coordinate files uses the following 
records:
<p>

(1)  ID - Either the 4-character PDB identifier code (for clean protein 
coordinate files) or the 7-character domain identifier code taken from scop
(for domain coordinate files; see documentation for the EMBOSS application 
scope for further info.)
<p>

(2)  DE - compound information.  Text from the COMPND records from the 
original pdb file are given.
<p>

(3)  OS - protein source information.  Text from the SOURCE records from the 
original pdb file are given.
<p>

(4)  EX - experimental information. The text 'nmr_or_model' (for nuclear  
magnetic resonance and model structures) or 'xray' (for structures 
determined by X-ray crystallography) appears as appropriate after the text
'METHOD'.  The resolution of X-ray structures, or '0' for structures of type
'nmr_or_model', is given after 'RESO'.  The number of models and number of 
polypeptide chains are given after 'NMOD' and 'NCHA' respectively. For 
domain coordinate files a 1 is always given. Following the EX record, the 
file will have a section containing a CN, IN and SQ records (see below) for
each chain.
<p>

(5)  CN - chain number.  The number given in brackets after this record 
indicates the start of a section of chain-specific data.
<p>

(6)  IN - chain specific data. The character given after ID is the PDB chain 
identifier (a '.' is given in cases where a chain identifier was not 
specified in the pdb file or, for domain coordinate files, the domain is 
comprised of more than one domain).  The number of amino acid residues 
comprising the chain (or the chains from which a domain is comprised) is 
given after NR. The number of atoms in heterogens and water molecules are 
given after NH and NW respectively. Domain coordinate files do not include
coordinates for these groups so a value of 0 is always given.
<p>

(7)  SQ - protein sequence. The number of residues is given before AA on the
first line. The protein sequence is given on subsequent lines. 
<p>

(8)  CO - coordinate data.  The columns of the records are as follows.
<OL>
<LI>CO is always given.
<LI>Model number (always 1 for domain coordinate files).
<LI>Chain number (always 1 for domain coordinate files).
<LI>Either P (a protein atom), H (a heterogen atom) or W (an atom in a water molecule).
<LI>Position of the residue in the protein sequence given in the SQ record (for protein atoms) or a sequential count of the atoms (for heterogens and water).
<LI>Residue number according to the original pdb file, or or a sequential count of the atoms (for heterogens and water).
<LI>Single character amino acid code or a '.' (for heterogens and water).
<LI>3-character residue identifier code.
<LI>Atom type.
<LI>The x orthogonal coordinate.
<LI>The y orthogonal coordinate.
<LI>The z orthogonal coordinate.
<LI>Occupancy.
<LI>Temperature factor.
</OL>
<p>

(9)  XX - Used for spacing.
<p>

(10) // - Given on the last line of the file only.
<p>













<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->

The PDB format is explained at:
<br>
<A HREF="http://www.rcsb.org/pdb/docs/format/pdbguide2.2/guide2.2_frame.html">
http://www.rcsb.org/pdb/docs/format/pdbguide2.2/guide2.2_frame.html</A>
<br>

<b>domainer</b> writes the following records in PDB format:
<p>

(1)  HEADER - bibliographic information.  The text 'CLEANED-UP PDB FILE FOR 
SCOP DOMAIN XXXXXXX' is always given (where XXXXXXX is a 7-character domain 
identifier code).
<p>

(2)  TITLE - bibliographic information. The text ' THIS FILE IS MISSING 
MOST RECORDS FROM THE ORIGINAL PDB FILE' is always given.
<p>

(3)  COMPND - compound information.  The COMPND records from the original  
pdb file are given.
<p>

(4)  SOURCE - protein source information.  The SOURCE records from the 
original PDB file are given.
<p>

(5)  REMARK - remark records. Remark records are used for spacing. One 
REMARK line containing the protein resolution is always given.
<p>

(6)  SEQRES - protein sequence.  
<p>

(7)  ATOM - atomic coordinates.
<p>

(8)  TER - indicates the end of a chain.

<p>

domainer generates a log file an excerpt of which is shown below. If 
there is a problem in processing a domain, three lines containing the 
record '//', the domain identifier code and an error message respectively
are written. The text 'WARN  filename not found' is given in cases where a 
clean coordinate file could not be found. 'ERROR filename file read error' 
or 'ERROR filename file write error' will be reported when an error was 
encountered during a file read or write respectively.  Various other error
messages may also be given (in case of difficulty email Jon Ison, 
jison@hgmp.mrc.ac.uk).

<p>
<hr>
<pre>
//
DS002__
WARN  Could not open for reading cpdb file s002.pxyz
//
DS003__
WARN  Could not open for reading cpdb file s003.pxyz
</pre>
<hr>
<p>








<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.
	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

   -->

None.













<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="nrscope.html">nrscope</a></td><td>Converts redundant EMBL-format SCOP file to non-redundant one</td></tr>
<tr><td><a href="scope.html">scope</a></td><td>Convert raw scop classification file to embl-like format</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

This application was written by Jon Ison <a
href="mailto:jison@hgmp.mrc.ac.uk">(jison@hgmp.mrc.ac.uk)</a>

















<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (Jan 2001) - Jon Ison.















<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->








</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

