<HTML>
<HEAD>
  <TITLE>
  EMBOSS: drfindid
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
drfindid
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
Find public databases by identifier
<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>



<b>drfindid</b> searches the Data Resource Catalogue to find entries
with EDAM data identifier terms matching a query string.





<H2>
    Algorithm
</H2>


The first search is of the EDAM ontology data namespace, using the
term names and their synonynms. All child terms are automatically
included in the set of matches inless the <tt>-nosubclasses</tt>
qualifier is used.

<p>
The <tt>-sensitive</tt> qualifier also searches the definition strings.

<p>
The set of EDAM terms are then compared to entries in the Data
Resource Catalogue, searching the 'eid' EDAM identifier index.










<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

Here is a sample session with <b>drfindid</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>drfindid protein </b>
Find public databases by identifier
Data resource output file [drfindid.drcat]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
Find public databases by identifier
Version: EMBOSS:6.4.0.0

   Standard (Mandatory) qualifiers:
  [-query]             string     List of EDAM data keywords (Any string)
  [-outfile]           outresource [*.drfindid] Output data resource file name

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers:
   -sensitive          boolean    [N] By default, the query keywords are
                                  matched against the EDAM term names (and
                                  synonyms) only. This option also matches the
                                  keywords against the EDAM term definitions
                                  and will therefore (typically) report more
                                  matches.
   -[no]subclasses     boolean    [Y] Extend the query matches to include all
                                  terms which are specialisations (EDAM
                                  sub-classes) of the matched type.

   Associated qualifiers:

   "-outfile" associated qualifiers
   -odirectory2        string     Output directory
   -oformat2           string     Data resource output format

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-query]<br>(Parameter 1)</td>
<td>string</td>
<td>List of EDAM data keywords</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-outfile]<br>(Parameter 2)</td>
<td>outresource</td>
<td>Output data resource file name</td>
<td>Data resource entry</td>
<td><i>&lt;*&gt;</i>.drfindid</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>-sensitive</td>
<td>boolean</td>
<td>By default, the query keywords are matched against the EDAM term names (and synonyms) only. This option also matches the keywords against the EDAM term definitions and will therefore (typically) report more matches.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-[no]subclasses</td>
<td>boolean</td>
<td>Extend the query matches to include all terms which are specialisations (EDAM sub-classes) of the matched type.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-outfile" associated outresource qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -odirectory2<br>-odirectory_outfile</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -oformat2<br>-oformat_outfile</td>
<td>string</td>
<td>Data resource output format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

None.

<p>






<H2>
    Output file format
</H2>

<p>

The output is a standard EMBOSS resource file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <tt>-oformat xxx</tt>, where 'xxx' is replaced by
the name of the required format.  The available format names are: 
drcat, basic, wsbasic, list.
<p>

See:
<A href="http://emboss.sf.net/docs/themes/ResourceFormats.html">
http://emboss.sf.net/docs/themes/ResourceFormats.html</A>
for further information on resource formats.

<p>

<p>

<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: drfindid.drcat</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
ID      CuticleDB
Name    Structural proteins of the arthropod cuticle (CuticleDB)
Desc    A relational database containing all structural proteins of Arthropod cuticle identified to date. Many come from direct sequencing of proteins isolated from cuticle and from sequences from cDNAs that share common features with these authentic cuticular proteins. It also includes proteins from the Drosophila melanogaster and the Anopheles gambiae genomes, that have been predicted to be cuticular proteins, based on a Pfam motif responsible for chitin binding in Arthropod cuticle.
URL     http://biophysics.biol.uoa.gr/cuticleDB/
Taxon   6656 | Arthropoda
EDAMtpc 0000635 | Specific protein
EDAMdat 0000896 | Protein report
EDAMid  0002715 | Protein ID (CuticleDB)
EDAMfmt 0002331 | HTML
Query    Protein report | HTML | Protein ID (CuticleDB) | http://bioinformatics.biol.uoa.gr/cuticleDB/protview.jsp?id=%s
Example Protein ID (CuticleDB) | 639

ID      PIRSF
Acc     DB-0079
Name    PIRSF whole-protein classification database
Desc    Non-overlapping clustering of UniProtKB sequences into a hierarchical order to reflect their evolutionary relationships. The PIRSF classification system is based on whole proteins rather than on the component domains therefore, it allows annotation of generic biochemical and specific biological functions, as well as classification of proteins without well-defined domains.
URL     http://pir.georgetown.edu/pirsf/
Cat     Family and domain databases
Taxon   1 | all
EDAMtpc 0000164 | Sequence clustering
EDAMtpc 0000724 | Protein families
EDAMdat 0000907 | Protein family annotation
EDAMid  0001136 | PIRSF ID
EDAMfmt 0002331 | HTML
Xref    SP_explicit | PIRSF ID
Query    Protein family annotation {PIRSF entry} | HTML | PIRSF ID | http://pir.georgetown.edu/cgi-bin/ipcSF?id=%s
Example PIRSF ID | SF000186

ID      Pfam
Acc     DB-0073
Name    Pfam protein domain database
Desc    Large collection of protein families, each represented by multiple sequence alignments and hidden Markov models (HMMs).
URL     http://pfam.sanger.ac.uk/
Cat     Family and domain databases
Taxon   1 | all
EDAMtpc 0000741 | Protein sequence alignment
EDAMtpc 0000188 | Sequence profiles and HMMs
EDAMtpc 0000724 | Protein families
EDAMtpc 0000736 | Protein domains
EDAMdat 0000907 | Protein family annotation
EDAMid  0001096 | Sequence accession (protein)
EDAMid  0001138 | Pfam accession number
EDAMid  0001179 | NCBI taxonomy ID
EDAMid  0002758 | Pfam clan ID
EDAMfmt 0002331 | HTML
Xref    SP_explicit | Pfam accession number
Xref    SP_FT | Pfam accession number
Query    Protein family annotation {Pfam entry} | HTML | Sequence accession (protein) | http://pfam.sanger.ac.uk/protein/acc=%s
Query    Protein family annotation {Pfam entry} | HTML | Pfam accession number | http://pfam.sanger.ac.uk/family?acc=%s
Query    Protein family annotation {Pfam entry} | HTML | Pfam clan ID | http://pfam.sanger.ac.uk/clan?acc=%s


<font color=red>  [Part of this file has been deleted for brevity]</font>


ID      SCOWL
Name    Structural characterization of water, ligands and proteins (SCOWL)
Desc    PDB interface interactions at atom, residue and domain level. The database contains protein interfaces and residue-residue interactions formed by structural units and interacting solvent.
URL     http://www.scowlp.org
Taxon   1 | all
EDAMtpc 0000128 | Protein interactions
EDAMdat 0002359 | Domain-domain interaction
EDAMid  0001127 | PDB ID
EDAMid  0001042 | SCOP sunid
EDAMfmt 0002331 | HTML
Query    Domain-domain interaction | HTML | PDB ID | http://www.scowlp.org/scowlp/PDB.action?show=&amp;PDBId=%s
Query    Domain-domain interaction | HTML | SCOP sunid {SCOP family id} | http://www.scowlp.org/scowlp/Search.action?childs=&amp;currentChoice=SCOP_FAMILY%s
Example PDB ID | 1fxk
Example SCOP sunid | 2381635

ID      SCOP
Name    Structural classification of proteins (SCOP) database
Desc    The SCOP database, created by manual inspection and abetted by a battery of automated methods, aims to provide a detailed and comprehensive description of the structural and evolutionary relationships between all proteins whose structure is known. As such, it provides a broad survey of all known protein folds, detailed information about the close relatives of any particular protein, and a framework for future research and classification.
URL     http://scop.mrc-lmb.cam.ac.uk/scop
Taxon   1 | all
EDAMtpc 0000736 | Protein domains
EDAMdat 0001554 | SCOP node
EDAMdat 0002093 | Data reference
EDAMid  0001042 | SCOP sunid
EDAMid  0001127 | PDB ID
EDAMid  0000842 | Identifier
EDAMfmt 0002331 | HTML
Query    SCOP node | HTML | SCOP sunid | http://scop.mrc-lmb.cam.ac.uk/scop/search.cgi?sunid=%s
Query    Data reference {PDB Entry search} | HTML | PDB ID | http://scop.mrc-lmb.cam.ac.uk/scop/search.cgi?PDB=%s
Query    Data reference | HTML | Identifier {Keyword} | http://scop.mrc-lmb.cam.ac.uk/scop/search.cgi?key=%s
Example SCOP sunid | 47718
Example PDB ID | 1djh
Example Identifier {Keyword} | immunoglobulin

ID      BIPA
Name    BIPA database for interaction between protein and nucleic acid
Desc    Protein-nucleic acid interactions, with various features of protein-nuleic acid interfaces.
URL     http://www-cryst.bioc.cam.ac.uk/bipa
Taxon   1 | all
EDAMtpc 0000149 | Protein-nucleic acid interactions
EDAMdat 0001567 | Protein-nucleic acid interaction
EDAMdat 0002358 | Domain-nucleic acid interaction
EDAMid  0001042 | SCOP sunid
EDAMid  0001127 | PDB ID
EDAMfmt 0002331 | HTML
Query    Protein-nucleic acid interaction | HTML | PDB ID | http://mordred.bioc.cam.ac.uk/bipa/structures/%s
Query    Domain-nucleic acid interaction | HTML | SCOP sunid | http://mordred.bioc.cam.ac.uk/bipa/scops/%s
Example PDB ID | 10MH
Example SCOP sunid | 120412

</pre>
</td></tr></table><p>





<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.

	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:

        #include file="inc/localfiles.ihtml"
   -->


The Data Resource Catalogue is included in EMBOSS as local
database <b>drcat</b>.


The EDAM Ontology is included in EMBOSS as local
database <b>edam</b>.





<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
<ol>

<li>

</ol>

   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="drfinddata.html">drfinddata</a></td>
<td>Find public databases by data type</td>
</tr>

<tr>
<td><a href="drfindformat.html">drfindformat</a></td>
<td>Find public databases by format</td>
</tr>

<tr>
<td><a href="drfindresource.html">drfindresource</a></td>
<td>Find public databases by resource</td>
</tr>

<tr>
<td><a href="drget.html">drget</a></td>
<td>Get data resource entries</td>
</tr>

<tr>
<td><a href="drtext.html">drtext</a></td>
<td>Get data resource entries complete text</td>
</tr>

<tr>
<td><a href="edamdef.html">edamdef</a></td>
<td>Find EDAM ontology terms by definition</td>
</tr>

<tr>
<td><a href="edamhasinput.html">edamhasinput</a></td>
<td>Find EDAM ontology terms by has_input relation</td>
</tr>

<tr>
<td><a href="edamhasoutput.html">edamhasoutput</a></td>
<td>Find EDAM ontology terms by has_output relation</td>
</tr>

<tr>
<td><a href="edamisformat.html">edamisformat</a></td>
<td>Find EDAM ontology terms by is_format_of relation</td>
</tr>

<tr>
<td><a href="edamisid.html">edamisid</a></td>
<td>Find EDAM ontology terms by is_identifier_of relation</td>
</tr>

<tr>
<td><a href="edamname.html">edamname</a></td>
<td>Find EDAM ontology terms by name</td>
</tr>

<tr>
<td><a href="wossdata.html">wossdata</a></td>
<td>Finds programs by EDAM data</td>
</tr>

<tr>
<td><a href="wossinput.html">wossinput</a></td>
<td>Finds programs by EDAM input data</td>
</tr>

<tr>
<td><a href="wossoperation.html">wossoperation</a></td>
<td>Finds programs by EDAM operation</td>
</tr>

<tr>
<td><a href="wossoutput.html">wossoutput</a></td>
<td>Finds programs by EDAM output data</td>
</tr>

<tr>
<td><a href="wossparam.html">wossparam</a></td>
<td>Finds programs by EDAM parameter</td>
</tr>

<tr>
<td><a href="wosstopic.html">wosstopic</a></td>
<td>Finds programs by EDAM topic</td>
</tr>

</table>
<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>

Peter Rice
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK

<p>
Please report all bugs to the EMBOSS bug team (emboss-bug&nbsp;&copy;&nbsp;emboss.open-bio.org) not to the original author.

<H2>
    History
</H2>
<!--
        Date written and what changes have been made go in this file.
   -->




<H2>
    Target users
</H2>
<!--
        For general users, use this text
   -->
This program is intended to be used by everyone and everything, from naive users to embedded scripts.

<H2>
    Comments
</H2>
<!--
        User/developer/other comments go in this file.
   -->
None


</BODY>
</HTML>

