<HTML>
<HEAD>
  <TITLE>
  EMBOSS: drtext
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
drtext
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
Get data resource entries complete text
<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>


<p><b>drtext</b> reads one or more complete data resource entries from a database or a file and writes them to a text file.  Optionally, the first data resourcefrom the input stream only can be retrieved.  The complete entry, including heading annotation, is retrieved and written and the data is not altered or reformatted in any way.</p>



<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

Here is a sample session with <b>drtext</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>drtext drcat:uniprot </b>
Get data resource entries complete text
Full text output file [uniprot.drtext]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
Get data resource entries complete text
Version: EMBOSS:6.4.0.0

   Standard (Mandatory) qualifiers:
  [-resources]         resource   Data resource filename and optional format,
                                  or reference (input query)
  [-outfile]           outfile    [*.drtext] Full text output file

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-resources" associated qualifiers
   -iformat1           string     Input resource format
   -idbname1           string     User-provided database name

   "-outfile" associated qualifiers
   -odirectory2        string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-resources]<br>(Parameter 1)</td>
<td>resource</td>
<td>Data resource filename and optional format, or reference (input query)</td>
<td>Data resource catalogue entry(s)</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-outfile]<br>(Parameter 2)</td>
<td>outfile</td>
<td>Full text output file</td>
<td>Output file</td>
<td><i>&lt;*&gt;</i>.drtext</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-resources" associated resource qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -iformat1<br>-iformat_resources</td>
<td>string</td>
<td>Input resource format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -idbname1<br>-idbname_resources</td>
<td>string</td>
<td>User-provided database name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-outfile" associated outfile qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -odirectory2<br>-odirectory_outfile</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<p>

The input is a standard EMBOSS resource query.

<p>

The expected source of data resource information is the Data
Resource Catalogue database "drcat" which is available as a standard
database in all EMBOSS installations.

<p>

Data can also be read from resource output in "drcat" format written by an
EMBOSS application.


<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>Database entry: drcat:uniprot</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID      UniProt
IDalt   UniProtKB
Name    Universal protein resource
Desc    A comprehensive, high-quality and freely accessible resource of protein sequence and functional information.
URL     http://www.uniprot.org/
Taxon   1 | all
EDAMtpc 0000639 | Protein sequences
EDAMdat 0002201 | Sequence record full
EDAMid  0003021 | UniProt accession
EDAMfmt 0001929 | FASTA format
EDAMfmt 0002376 | RDF
EDAMfmt 0002188 | uniprot
EDAMfmt 0002331 | HTML
EDAMfmt 0002332 | XML
Xref    SP_FT | None
Query    Sequence record full | HTML | UniProt accession | http://www.uniprot.org/uniprot/%s
Query    Sequence record full | uniprot | UniProt accession | http://www.uniprot.org/uniprot/%s.txt
Query    Sequence record full | XML | UniProt accession | http://www.uniprot.org/uniprot/%s.xml
Query    Sequence record full | RDF | UniProt accession | http://www.uniprot.org/uniprot/%s.rdf
Query    Sequence record full | FASTA format | UniProt accession | http://www.uniprot.org/uniprot/%s.fasta
Example UniProt accession | P12345

</pre>
</td></tr></table><p>





<H2>
    Output file format
</H2>

<b>drtext</b> reports the full text of the input resources.



<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: uniprot.drtext</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
ID      UniProt
IDalt   UniProtKB
Name    Universal protein resource
Desc    A comprehensive, high-quality and freely accessible resource of protein sequence and functional information.
URL     http://www.uniprot.org/
Taxon   1 | all
EDAMtpc 0000639 | Protein sequences
EDAMdat 0002201 | Sequence record full
EDAMid  0003021 | UniProt accession
EDAMfmt 0001929 | FASTA format
EDAMfmt 0002376 | RDF
EDAMfmt 0002188 | uniprot
EDAMfmt 0002331 | HTML
EDAMfmt 0002332 | XML
Xref    SP_FT | None
Query   Sequence record full | HTML | UniProt accession | http://www.uniprot.org/uniprot/%s
Query   Sequence record full | uniprot | UniProt accession | http://www.uniprot.org/uniprot/%s.txt
Query   Sequence record full | XML | UniProt accession | http://www.uniprot.org/uniprot/%s.xml
Query   Sequence record full | RDF | UniProt accession | http://www.uniprot.org/uniprot/%s.rdf
Query   Sequence record full | FASTA format | UniProt accession | http://www.uniprot.org/uniprot/%s.fasta
Example UniProt accession | P12345
</pre>
</td></tr></table><p>





<H2>
    Data files
</H2>

The Data Resource Catalogue is included in EMBOSS as local
database <b>drcat</b>.




<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
<ol>

<li>

</ol>

   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="drfinddata.html">drfinddata</a></td>
<td>Find public databases by data type</td>
</tr>

<tr>
<td><a href="drfindformat.html">drfindformat</a></td>
<td>Find public databases by format</td>
</tr>

<tr>
<td><a href="drfindid.html">drfindid</a></td>
<td>Find public databases by identifier</td>
</tr>

<tr>
<td><a href="drfindresource.html">drfindresource</a></td>
<td>Find public databases by resource</td>
</tr>

<tr>
<td><a href="drget.html">drget</a></td>
<td>Get data resource entries</td>
</tr>

<tr>
<td><a href="entret.html">entret</a></td>
<td>Retrieves sequence entries from flatfile databases and files</td>
</tr>

<tr>
<td><a href="ontotext.html">ontotext</a></td>
<td>Get ontology term(s) original full text</td>
</tr>

<tr>
<td><a href="textget.html">textget</a></td>
<td>Get text data entries</td>
</tr>

<tr>
<td><a href="textsearch.html">textsearch</a></td>
<td>Search the textual description of sequence(s)</td>
</tr>

</table>
<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>

Peter Rice
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK

<p>
Please report all bugs to the EMBOSS bug team (emboss-bug&nbsp;&copy;&nbsp;emboss.open-bio.org) not to the original author.
<H2>
    History
</H2>
<!--
        Date written and what changes have been made go in this file.
   -->




<H2>
    Target users
</H2>
<!--
        For general users, use this text
   -->
This program is intended to be used by everyone and everything, from naive users to embedded scripts.

<H2>
    Comments
</H2>
<!--
        User/developer/other comments go in this file.
   -->
None


</BODY>
</HTML>

