<HTML>

<HEAD>
  <TITLE>
  EMBOSS: einverted
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: einverted
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program einverted
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Finds DNA inverted repeats


<H2>
    Description
</H2>


<b>einverted</b> looks for inverted repeats (stem loops) in a nucleotide
sequence. 

<p>

It will find inverted repeats that include a proprtion of mismatches and
gaps (bulges in the stem loop). 

<p>

It works by finding alignments between the sequence and its reverse
complement that exceed a threshold score.  Gaps and Mismatches are
assigned a penalty (negative) score.  Matches are assigned a positive
score.  The score is calculated by summing the values of each match, the
penalties of each mismatch and the large penalties of any gaps.  Any
region whose score exceeds the threshold is reported. 

<p>

<b>einverted</b>  uses dynamic programming and thus is guaranteed to find
the optimal alignment, but is slower than, for example, a self-by-self BLAST.
It can find multiple inverted repeats in a sequence.  

<p>

Secondary structures like inverted repeats in genomic sequences may be
implicated in initiation of DNA replication.

<H2>
    Usage
</H2>


Here is a sample session with einverted.

<pre>

% einverted
Input sequence: embl:hsts1
Output file [hsts1.inv]: 
Gap penalty [12]: 
Minimum score threshold [50]: 
Match score [3]: 
Mismatch score [-4]: 

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          sequence   Sequence USA
   -gap                integer    Gap penalty
   -threshold          integer    Minimum score threshold
   -match              integer    Match score
   -mismatch           integer    Mismatch score
  [-outfile]           outfile    Output file name

   Optional qualifiers:
   -maxrepeat          integer    Maximum separation between the start of
                                  repeat and the end of the inverted repeat
                                  (the default is 4000 bases).

   Advanced qualifiers: (none)
   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-gap</td>
<td>Gap penalty</td>
<td>Any integer value</td>
<td>12</td>
</tr>

<tr>
<td>-threshold</td>
<td>Minimum score threshold</td>
<td>Any integer value</td>
<td>50</td>
</tr>

<tr>
<td>-match</td>
<td>Match score</td>
<td>Any integer value</td>
<td>3</td>
</tr>

<tr>
<td>-mismatch</td>
<td>Mismatch score</td>
<td>Any integer value</td>
<td>-4</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.einverted</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-maxrepeat</td>
<td>Maximum separation between the start of repeat and the end of the inverted repeat (the default is 4000 bases).</td>
<td>Any integer value</td>
<td>4000</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>


The input for <b>einverted</b> is a nucleotide sequence

<H2>
    Output file format
</H2>


Here is the output form the example run. The first two hits are
removed to avoid the output lines being too wide.

<pre>

......................

Score 80: 44/51 ( 86%) matches, 2 gaps
   12246 ctcctgcctcag-cctccaagtagctgggattaca-gcatgtgccaccatgcc 12296   
         |||||| ||||| | |||||   |||||||||||| ||||| |||||||| ||
   13938 gaggacagagtcagaaggtttcacgaccctaatgtccgtactcggtggtatgg 13886   

Score 99: 53/65 ( 81%) matches, 1 gaps
   13884 tgggtatggtggctcatgcctgtaatcccagcactttggaagactgagacaggagcaattgcttga 13949   
         ||||| |||||||   ||||||||||||||||    ||| || ||||| ||| || ||||||||||
   14692 acccacaccaccgtacacggacattagggtcgatggaccctccgactccgtcttc-ttaacgaact 14628   

</pre>

<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

Sometimes you can find repeats using the program <b>palindrome</b> that
you can't find with <b>einverted</b> using the default parameters. 

<p>

This is not due to a problem with either program.  It is simply because
some of the shortest repeats that you find with <b>palindrome</b>'s
default parameter values are below <b>einverted</b>'s default cutoff
score - you should decrease the 'Minimum score threshold' to see them. 

<p>

For example, when <b>palindrome</b> is run with 'em:hsfau1', 
it finds the repeat:

<p>
<pre>
64    aaaactaaggc    74
      |||||||||||
98    ttttgattccg    88
</pre>
<p>

<b>einverted</b> will not report this as its score is 33 (11 bases
scoring 3 each, no mismatches or gaps) with is below the default score
cutoff of 50. 

<p>

If  <b>einverted</b> is run as:
<p>

% einverted em:hsfau1  -threshold 33 
<p>

then it will find it:
<p>

<pre>
Score 33: 11/11 (100%) matches, 0 gaps
      64 aaaactaaggc 74      
         |||||||||||
      98 ttttgattccg 88      
</pre>
<p>

(Anything can be considered to be a repeat if you set the score
threshold low enough!)

<H2>
    References
</H2>


Some assorted references on inverted repeats:

<p>

<ol>

<li>
Pearson CE, Zorbas H, Price GB, Zannis-Hadjopoulos M
Inverted repeats, stem-loops, and cruciforms: significance for initiation of DNA replication. 
J Cell Biochem 1996 Oct;63(1):1-22

<li>
Waldman AS, Tran H, Goldsmith EC, Resnick MA.	q
Long inverted repeats are an at-risk motif for recombination in mammalian cells.
Genetics. 1999 Dec;153(4):1873-83.
PMID: 10581292; UI: 20050682

<li>
Jacobsen SE 
Gene silencing: Maintaining methylation patterns. 
Curr Biol 1999 Aug 26;9(16):R617-9

<li>
Lewis S, Akgun E, Jasin M. 
Palindromic DNA and genome stability. Further studies.
Ann N Y Acad Sci. 1999 May 18;870:45-57.
PMID: 10415472; UI: 99343961

<li>
Dai X, Greizerstein MB, Nadas-Chinni K, Rothman-Denes LB 
Supercoil-induced extrusion of a regulatory DNA hairpin. 
Proc Natl Acad Sci U S A 1997 Mar 18;94(6):2174-9

</ol>

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="equicktandem.html">equicktandem</a></td><td>Finds tandem repeats</td></tr>
<tr><td><a href="etandem.html">etandem</a></td><td>Looks for tandem repeats in a nucleotide sequence</td></tr>
<tr><td><a href="palindrome.html">palindrome</a></td><td>Looks for inverted repeats in a nucleotide sequence</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->

<p>

<b>palindrome</b> also looks for inverted repeats but is much faster
and less sensitive, as it looks for near-perfect repeats.

<H2>
    Author(s)
</H2>


This program was originally written by Richard Durbin at the Sanger Centre.
<p>

This application was modified for inclusion in EMBOSS by Peter Rice <a
href="mailto:pmr@sanger.ac.uk">(pmr@sanger.ac.uk)</a> Informatics
Division, The Sanger Centre, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.

<H2>
    History
</H2>

Written (1999) - Peter Rice

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

