<HTML>

<HEAD>
  <TITLE>
  EMBOSS: einverted
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

einverted
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>






<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Finds DNA inverted repeats


<H2>
    Description
</H2>


<b>einverted</b> looks for inverted repeats (stem loops) in a nucleotide
sequence. 

<p>

It will find inverted repeats that include a proprtion of mismatches and
gaps (bulges in the stem loop). 

<p>

It works by finding alignments between the sequence and its reverse
complement that exceed a threshold score.  Gaps and Mismatches are
assigned a penalty (negative) score.  Matches are assigned a positive
score.  The score is calculated by summing the values of each match, the
penalties of each mismatch and the large penalties of any gaps.  Any
region whose score exceeds the threshold is reported. 

<p>

<b>einverted</b>  uses dynamic programming and thus is guaranteed to find
the optimal alignment, but is slower than, for example, a self-by-self BLAST.
It can find multiple inverted repeats in a sequence.  

<p>

Secondary structures like inverted repeats in genomic sequences may be
implicated in initiation of DNA replication.

<H2>
    Usage
</H2>


<b>Here is a sample session with einverted</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>einverted tembl:hsts1 </b>
Finds DNA inverted repeats
Gap penalty [12]: <b></b>
Minimum score threshold [50]: <b></b>
Match score [3]: <b></b>
Mismatch score [-4]: <b></b>
Output file [hsts1.inv]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>


<!--
Here is a sample session with einverted.

<pre>

% einverted
Input sequence: embl:hsts1
Output file [hsts1.inv]: 
Gap penalty [12]: 
Minimum score threshold [50]: 
Match score [3]: 
Mismatch score [-4]: 

</pre>
-->


<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          sequence   Sequence USA
   -gap                integer    Gap penalty
   -threshold          integer    Minimum score threshold
   -match              integer    Match score
   -mismatch           integer    Mismatch score
  [-outfile]           outfile    Output file name

   Optional qualifiers:
   -maxrepeat          integer    Maximum separation between the start of
                                  repeat and the end of the inverted repeat
                                  (the default is 4000 bases).

   Advanced qualifiers: (none)
   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-gap</td>
<td>Gap penalty</td>
<td>Any integer value</td>
<td>12</td>
</tr>

<tr>
<td>-threshold</td>
<td>Minimum score threshold</td>
<td>Any integer value</td>
<td>50</td>
</tr>

<tr>
<td>-match</td>
<td>Match score</td>
<td>Any integer value</td>
<td>3</td>
</tr>

<tr>
<td>-mismatch</td>
<td>Mismatch score</td>
<td>Any integer value</td>
<td>-4</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.einverted</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-maxrepeat</td>
<td>Maximum separation between the start of repeat and the end of the inverted repeat (the default is 4000 bases).</td>
<td>Any integer value</td>
<td>4000</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>


The input for <b>einverted</b> is a nucleotide sequence

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:hsts1' is a sequence entry in the example nucleic acid database 'tembl'
<p>
<p><h3>Database entry: tembl:hsts1</h3>
<table width="90%"><tr><td bgcolor="#FFDDFF">
<pre>
ID   HSTS1      standard; DNA; HUM; 18596 BP.
XX
AC   D00596;
XX
SV   D00596.1
XX
DT   17-JUL-1991 (Rel. 28, Created)
DT   27-OCT-1998 (Rel. 57, Last updated, Version 2)
XX
DE   Homo sapiens gene for thymidylate synthase, exons 1, 2, 3, 4, 5, 6, 7,
DE   complete cds.
XX
KW   thymidylate syntase.
XX
OS   Homo sapiens (human)
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia;
OC   Eutheria; Primates; Catarrhini; Hominidae; Homo.
XX
RN   [1]
RP   1-18596
RX   MEDLINE; 91056070.
RA   Kaneda S., Nalbantoglu J., Takeishi K., Shimizu K., Gotoh O., Seno T.,
RA   Ayusawa D.;
RT   "Structural and Functional Analysis of the Human Thymidylate Synthase
RT   Gene";
RL   J. Biol. Chem. 265:20277-20284(1990).
XX
DR   SWISS-PROT; P04818; TYSY_HUMAN.
XX
CC   These data kindly submitted in computer readable form by:
CC   Sumiko Kaneda
CC   National Institute of Genetics
CC   1111 Yata
CC   Mishima 411
CC   Japan
CC   Phone:  +81-559-72-2732
CC   Fax:    +81-559-71-3651
XX
FH   Key             Location/Qualifiers
FH
FT   source          1..18596
FT                   /chromosome="18"
FT                   /db_xref="taxon:9606"
FT                   /sequenced_mol="DNA"
FT                   /organism="Homo sapiens"
FT                   /clone="lambdaHTS-1 and lambdaHTS-3"
FT                   /map="18p11.32"
FT   repeat_unit     1..148
FT                   /note="Alu sequence"
FT   repeat_unit     202..477


<font color=red>  [Part of this file has been deleted for brevity]</font>

     ttttgttttt agcttcagcg agaacccaga cctttcccaa agctcaggat tcttcgaaaa     15660
     gttgagaaaa ttgatgactt caaagctgaa gactttcaga ttgaagggta caatccgcat     15720
     ccaactatta aaatggaaat ggctgtttag ggtgctttca aaggagctcg aaggatattg     15780
     tcagtcttta ggggttgggc tggatgccga ggtaaaagtt ctttttgctc taaaagaaaa     15840
     aggaactagg tcaaaaatct gtccgtgacc tatcagttat taatttttaa ggatgttgcc     15900
     actggcaaat gtaactgtgc cagttctttc cataataaaa ggctttgagt taactcactg     15960
     agggtatctg acaatgctga ggttatgaac aaagtgagga gaatgaaatg tatgtgctct     16020
     tagcaaaaac atgtatgtgc atttcaatcc cacgtactta taaagaaggt tggtgaattt     16080
     cacaagctat ttttggaata tttttagaat attttaagaa tttcacaagc tattccctca     16140
     aatctgaggg agctgagtaa caccatcgat catgatgtag agtgtggtta tgaactttaa     16200
     agttatagtt gttttatatg ttgctataat aaagaagtgt tctgcattcg tccacgcttt     16260
     gttcattctg tactgccact tatctgctca gttccttcct aaaatagatt aaagaactct     16320
     ccttaagtaa acatgtgctg tattctggtt tggatgctac ttaaaagagt atattttaga     16380
     aataatagtg aatatatttt gccctatttt tctcatttta actgcatctt atcctcaaaa     16440
     tataatgacc atttaggata gagttttttt tttttttttt taaactttta taaccttaaa     16500
     gggttatttt aaaataatct atggactacc attttgccct cattagcttc agcatggtgt     16560
     gacttctcta ataatatgct tagattaagc aaggaaaaga tgcaaaacca cttcggggtt     16620
     aatcagtgaa atatttttcc cttcgttgca taccagatac ccccggtgtt gcacgactat     16680
     ttttattctg ctaatttatg acaagtgtta aacagaacaa ggaattattc caacaagtta     16740
     tgcaacatgt tgcttatttt caaattacag tttaatgtct aggtgccagc ccttgatata     16800
     gctatttttg taagaacatc ctcctggact ttgggttagt taaatctaaa cttatttaag     16860
     gattaagtag gataacgtgc attgatttgc taaaagaatc aagtaataat tacttagctg     16920
     attcctgagg gtggtatgac ttctagctga actcatcttg atcggtagga ttttttaaat     16980
     ccatttttgt aaaactattt ccaagaaatt ttaagccctt tcacttcaga aagaaaaaag     17040
     ttgttggggc tgagcactta attttcttga gcaggaagga gtttcttcca aacttcacca     17100
     tctggagact ggtgtttctt tacagattcc tccttcattt ctgttgagta gccgggatcc     17160
     tatcaaagac caaaaaaatg agtcctgtta acaaccacct ggaacaaaaa cagattttat     17220
     gcatttatgc tgctccaaga aatgctttta cgtctaagcc agaggcaatt aattaatttt     17280
     tttttttttg acatggagtc actgtccgtt gcccaggctg cagtgcagtg gcgcaatctt     17340
     ggctcactgc aacctccacc tcccaggttc aagtgattct cctgcctcag cctcccatgt     17400
     agctgggatc acaggcacct gccaccatgc ccggctaatt ttttgtattt tttgtagaga     17460
     cagggtttca ccatgttggc caggctggtc tcaaacacct gacctcaaat gatccacctg     17520
     cctcagcctc ccaaagtgtt gggattacag gcgtaagcca ccatgcccag ccctgaatta     17580
     atatttttaa aataagtttg gagactgttg gaaataatag ggcagaggaa catattttac     17640
     tggctacttg ccagagttag ttaactcatc aaactctttg ataatagttt gacctctgtt     17700
     ggtgaaaatg agccatgatc tcttgaacat gatcagaata aatgccccag ccacacaatt     17760
     gtagtccaaa ctttttaggt cactaacttg ctagatggtg ccaggttttt ttgcacaagg     17820
     agtgcaaatg ttaagatctc cactagtgag gaaaggctag tattacagaa gccttgtcag     17880
     aggcaattga acctccaagc cctggccctc aggcctgagg attttgatac agacaaactg     17940
     aagaaccgtt tgttagtgga tattgcaaac aaacaggagt caaagcttgg tgctccacag     18000
     tctagttcac gagacaggcg tggcagtggc tggcagcatc tcttctcaca ggggccctca     18060
     ggcacagctt accttgggag gcatgtagga agcccgctgg atcatcacgg gatacttgaa     18120
     atgctcatgc aggtggtcaa catactcaca caccctagga ggagggaatc agatcggggc     18180
     aatgatgcct gaagtcagat tattcacgtg gtgctaactt aaagcagaag gagcgagtac     18240
     cactcaattg acagtgttgg ccaaggctta gctgtgttac catgcgtttc taggcaagtc     18300
     cctaaacctc tgtgcctcag gtccttttct tctaaaatat agcaatgtga ggtggggact     18360
     ttgatgacat gaacacacga agtccctctg agaggttttg tggtgccctt taaaagggat     18420
     caattcagac tctgtaaata tccagaatta tttgggttcc tctggtcaaa agtcagatga     18480
     atagattaaa atcaccacat tttgtgatct atttttcaag aagcgtttgt attttttcat     18540
     atggctgcag cagctgccag gggcttgggg tttttttggc aggtagggtt gggagg         18596
//
</pre>
</td></tr></table><p>


<H2>
    Output file format
</H2>



<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: hsts1.inv</h3>
<table width="90%"><tr><td bgcolor="#DDFFDD">
<pre>

Score 236: 108/130 ( 83%) matches, 0 gaps
      13 gctacgcgagaggctgaggcagcagaattacttgaacccaggaggcggaggttgcagtgagccgagatcgcgccactgcactccagcctgggtgagagagcgagactctgtctcaaaaaaaaaaaaaaaa 142     
         |||||  | ||||||||||||| |||||| ||||||||  |||||| |||||||||||||||| |||||   ||| || ||||||||||||| || ||||| ||||| |  |  ||||||||||||||||
     328 cgatgaaccctccgactccgtcttcttaacgaacttggaccctccgtctccaacgtcactcggttctaggctggtaacatgaggtcggacccgctgtctcgttctgacagggtttttttttttttttttt 199     

Score 228: 108/132 ( 81%) matches, 0 gaps
      11 cagctacgcgagaggctgaggcagcagaattacttgaacccaggaggcggaggttgcagtgagccgagatcgcgccactgcactccagcctgggtgagagagcgagactctgtctcaaaaaaaaaaaaaaaa 142     
           |||||  | ||||||||||||| |||||| ||||||||  |||||| |||||||||||||||| |||||   ||| || ||||||||||||| || ||||| ||||| |  |  ||||||||||||||||
     330 accgatgaaccctccgactccgtcttcttaacgaacttggaccctccgtctccaacgtcactcggttctaggctggtaacatgaggtcggacccgctgtctcgttctgacagggtttttttttttttttttt 199     

Score 226: 110/136 ( 80%) matches, 0 gaps
       7 gtcccagctacgcgagaggctgaggcagcagaattacttgaacccaggaggcggaggttgcagtgagccgagatcgcgccactgcactccagcctgggtgagagagcgagactctgtctcaaaaaaaaaaaaaaaa 142     
         ||    |||||  | ||||||||||||| |||||| ||||||||  |||||| |||||||||||||||| |||||   ||| || ||||||||||||| || ||||| ||||| |  |  ||||||||||||||||
     334 caccaccgatgaaccctccgactccgtcttcttaacgaacttggaccctccgtctccaacgtcactcggttctaggctggtaacatgaggtcggacccgctgtctcgttctgacagggtttttttttttttttttt 199     

Score 221: 111/139 ( 79%) matches, 0 gaps
       4 gtagtcccagctacgcgagaggctgaggcagcagaattacttgaacccaggaggcggaggttgcagtgagccgagatcgcgccactgcactccagcctgggtgagagagcgagactctgtctcaaaaaaaaaaaaaaaa 142     
         |  ||    |||||  | ||||||||||||| |||||| ||||||||  |||||| |||||||||||||||| |||||   ||| || ||||||||||||| || ||||| ||||| |  |  ||||||||||||||||
     337 ccgcaccaccgatgaaccctccgactccgtcttcttaacgaacttggaccctccgtctccaacgtcactcggttctaggctggtaacatgaggtcggacccgctgtctcgttctgacagggtttttttttttttttttt 199     

Score 216: 112/142 ( 78%) matches, 0 gaps
       1 cctgtagtcccagctacgcgagaggctgaggcagcagaattacttgaacccaggaggcggaggttgcagtgagccgagatcgcgccactgcactccagcctgggtgagagagcgagactctgtctcaaaaaaaaaaaaaaaa 142     
         |  |  ||    |||||  | ||||||||||||| |||||| ||||||||  |||||| |||||||||||||||| |||||   ||| || ||||||||||||| || ||||| ||||| |  |  ||||||||||||||||
     340 gacccgcaccaccgatgaaccctccgactccgtcttcttaacgaacttggaccctccgtctccaacgtcactcggttctaggctggtaacatgaggtcggacccgctgtctcgttctgacagggtttttttttttttttttt 199     

Score 199: 54/170 ( 31%) matches, 1 gaps
       1 -cctgtagtcccagctacgcgagaggctgaggcagcagaattacttgaacccaggaggcggaggttgcagtgagccgagatcgcgccactgcactccagcctgggtgagagagcgagactctgtctcaaaaaaaaaaaaaaaagaccgccagggctcaaacaaaaaacctc 170     
           |||        |        |  |     |    |  | |   |  | || ||  |     | |         | |     || |  |      |   |  ||          |        |  |||||||||||||||| |     |     |   ||       ||
     342 tcgacccgcaccaccgatgaaccctccgactccgtcttcttaacgaacttggaccctccgtctccaacgtcactcggttctaggctggtaacatgaggtcggacccgctgtctcgttctgacagggttttttttttttttttttttttttttttctggcggtcccgaaaag 172     

Score 175: 56/171 ( 32%) matches, 1 gaps
       1 -cctgtagtcccagctacgcgagaggctgaggcagcagaattacttgaacccaggaggcggaggttgcagtgagccgagatcgcgccactgcactccagcctgggtgagagagcgagactctgtctcaaaaaaaaaaaaaaaagaccgccagggctcaaacaaaaaacctcg 171     
              | | |   |   |      ||  |  |  ||||||  | |           || ||  |  |  |      |  |     | ||     | |  |      |          |        | ||||||||||||||| |     |      |     |   | |  
     344 aatcgacccgcaccaccgatgaaccctccgactccgtcttcttaacgaacttggaccctccgtctccaacgtcactcggttctaggctggtaacatgaggtcggacccgctgtctcgttctgacagggttttttttttttttttttttttttttttctggcggtcccgaaaa 173     

Score 172: 100/120 ( 83%) matches, 4 gaps
     345 tttttgtacttttagtagagacgggggtttcaccatgttgtccaggctggtcttgaactcctgacctcaggtgatccacccgcctcggccccccaaagtactaggattacaggcgtgagccacc 468     
         |||||||  | ||||||||||| ||||| || |||||||| |||||||||||   || |||| ||||||||||||    | | ||| ||| |||||||   | |||||||||||||||||||||
    3177 aaaaacacaataatcatctctgaccccacagcggtacaaccggtccgaccagggtttaaggattggagtccacta----gactgagtcggagggtttcgccaccctaatgtccgcactcggtgg 3058    

Score 164: 128/174 ( 73%) matches, 3 gaps
   12128 agaggatttttttttttttttttttttttgagacagagttttgctctgttgcccaggctggaatgcaacggcgtgatcttggctcactgtaacctctgcctcc-tgggttcgagtgattctcctgcctcagcctc-caagtagctgggattaca-gcatgtgccaccatgcctggct 12301   
         ||||  || || || || || |||||    |  ||| || |  |||||||||||||| |||| ||||| ||||||||| || ||||||  | ||||    ||| ||||||| | |||| |||  ||||  |||||   ||| | ||| |||||| |  || |||||||  |||||||
   12749 tctcaaaacaacaacaacaacaaaaatttataagtcccagagtgagacaacgggtcctacctcacgttaccgcactagtactgagtgatgtcggagtttaaggcacccaagtttactaggagagcggaaccggagacctcaacaaccttaatgtccacactcggtggtgtggaccga 12573   

Score 80: 44/51 ( 86%) matches, 2 gaps
   12246 ctcctgcctcag-cctccaagtagctgggattaca-gcatgtgccaccatgcc 12296   
         |||||| ||||| | |||||   |||||||||||| ||||| |||||||| ||
   13938 gaggacagagtcagaaggtttcacgaccctaatgtccgtactcggtggtatgg 13886   

Score 130: 74/94 ( 78%) matches, 1 gaps
   13890 tggtggctcatgcctgtaatcccagcactttggaagactgagacaggagcaattgcttgaggtctggagttcaataccagcctgggcaacataac 13984   
         |||||||| ||||||||||| |||| ||||| | || ||||| |||| |  |   ||||| ||| ||||||  | | |||||||| |||||| ||
   14822 accaccgaatacggacattaaggtcttgaaagcctccgactccgtccaccat-gtgaacttcagtcctcaaactctagtcggaccggttgtactg 14729   
</pre>
</td></tr></table><p>


<!--
Here is the output form the example run. The first two hits are
removed to avoid the output lines being too wide.

<pre>

......................

Score 80: 44/51 ( 86%) matches, 2 gaps
   12246 ctcctgcctcag-cctccaagtagctgggattaca-gcatgtgccaccatgcc 12296   
         |||||| ||||| | |||||   |||||||||||| ||||| |||||||| ||
   13938 gaggacagagtcagaaggtttcacgaccctaatgtccgtactcggtggtatgg 13886   

Score 99: 53/65 ( 81%) matches, 1 gaps
   13884 tgggtatggtggctcatgcctgtaatcccagcactttggaagactgagacaggagcaattgcttga 13949   
         ||||| |||||||   ||||||||||||||||    ||| || ||||| ||| || ||||||||||
   14692 acccacaccaccgtacacggacattagggtcgatggaccctccgactccgtcttc-ttaacgaact 14628   

</pre>
-->

<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

Sometimes you can find repeats using the program <b>palindrome</b> that
you can't find with <b>einverted</b> using the default parameters. 

<p>

This is not due to a problem with either program.  It is simply because
some of the shortest repeats that you find with <b>palindrome</b>'s
default parameter values are below <b>einverted</b>'s default cutoff
score - you should decrease the 'Minimum score threshold' to see them. 

<p>

For example, when <b>palindrome</b> is run with 'em:hsfau1', 
it finds the repeat:

<p>
<pre>
64    aaaactaaggc    74
      |||||||||||
98    ttttgattccg    88
</pre>
<p>

<b>einverted</b> will not report this as its score is 33 (11 bases
scoring 3 each, no mismatches or gaps) with is below the default score
cutoff of 50. 

<p>

If  <b>einverted</b> is run as:
<p>

% einverted em:hsfau1  -threshold 33 
<p>

then it will find it:
<p>

<pre>
Score 33: 11/11 (100%) matches, 0 gaps
      64 aaaactaaggc 74      
         |||||||||||
      98 ttttgattccg 88      
</pre>
<p>

(Anything can be considered to be a repeat if you set the score
threshold low enough!)

<H2>
    References
</H2>

None.

<!--
Some assorted references on inverted repeats:

<p>

<ol>

<li>
Pearson CE, Zorbas H, Price GB, Zannis-Hadjopoulos M
Inverted repeats, stem-loops, and cruciforms: significance for initiation of DNA replication. 
J Cell Biochem 1996 Oct;63(1):1-22

<li>
Waldman AS, Tran H, Goldsmith EC, Resnick MA.	q
Long inverted repeats are an at-risk motif for recombination in mammalian cells.
Genetics. 1999 Dec;153(4):1873-83.
PMID: 10581292; UI: 20050682

<li>
Jacobsen SE 
Gene silencing: Maintaining methylation patterns. 
Curr Biol 1999 Aug 26;9(16):R617-9

<li>
Lewis S, Akgun E, Jasin M. 
Palindromic DNA and genome stability. Further studies.
Ann N Y Acad Sci. 1999 May 18;870:45-57.
PMID: 10415472; UI: 99343961

<li>
Dai X, Greizerstein MB, Nadas-Chinni K, Rothman-Denes LB 
Supercoil-induced extrusion of a regulatory DNA hairpin. 
Proc Natl Acad Sci U S A 1997 Mar 18;94(6):2174-9

</ol>
-->

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="equicktandem.html">equicktandem</a></td><td>Finds tandem repeats</td></tr>
<tr><td><a href="etandem.html">etandem</a></td><td>Looks for tandem repeats in a nucleotide sequence</td></tr>
<tr><td><a href="palindrome.html">palindrome</a></td><td>Looks for inverted repeats in a nucleotide sequence</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->

<p>

<b>palindrome</b> also looks for inverted repeats but is much faster
and less sensitive, as it looks for near-perfect repeats.

<H2>
    Author(s)
</H2>


This program was originally written by Richard Durbin at the Sanger Centre.
<p>

This application was modified for inclusion in EMBOSS by Peter Rice <a
href="mailto:pmr@sanger.ac.uk">(pmr@sanger.ac.uk)</a> Informatics
Division, The Sanger Centre, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.

<H2>
    History
</H2>

Written (1999) - Peter Rice

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

