<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: eprimer3
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">
<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>
<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>
</td>
<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: eprimer3
</H2>
</font>
</td>
</tr>
</table>
</td>
</tr>
</table>
<p><hr><p>

<!--END OF HEADER-->

<H1>
Program eprimer3
</H1>
<H2>
    Function
</H2>
Picks PCR primers and hybridization oligos

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
        Algorithms
  -->

<b>eprimer3</b> is an interface to the 'primer3' program from the
Whitehead Institute. 

<p>

The Whitehead program must be set up and on the path in order for
<b>eprimer3</b> to find and run it. 

<p>

Primer3 picks primers for PCR reactions, considering as criteria:
<p>

<UL>
<LI>oligonucleotide melting temperature, size, GC content,
  and primer-dimer possibilities,

<LI>PCR product size,

<LI>positional constraints within the source sequence, and

<LI>miscellaneous other constraints.
</UL>

<p>

All of these criteria are user-specifiable as constraints.

<p>

<b>eprimer3</b> can also pick hybridisation oligos that are internal to
the product. 


<H3>ADVICE FOR PICKING PRIMERS</H3>

We suggest referring to: Wojciech Rychlik, "Selection of Primers
for Polymerase Chain Reaction" in BA White, Ed., "Methods in
Molecular Biology, Vol. 15: PCR Protocols: Current Methods and
Applications", 1993, pp 31-40, Humana Press, Totowa NJ

<H4>Cautions</H4>

Some of the most important issues in primer picking can be
addressed only before using eprimer3.  These are sequence quality
(including making sure the sequence is not vector and not
chimeric) and avoiding repetitive elements.

<p>

Techniques for avoiding problems include a thorough understanding
of possible vector contaminants and cloning artifacts coupled
with database searches using blast, fasta, or other similarity
searching program to screen for vector contaminants and possible
repeats.  Repbase (J. Jurka, A.F.A. Smit, C. Pethiyagoda, and
others, 1995-1996, ftp://ncbi.nlm.nih.gov/repository/repbase)
is an excellent source of repeat sequences and pointers to the
literature.  eprimer3 now allows you to screen candidate oligos
against a Mispriming Library (or a Mishyb Library in the case
of internal oligos).

<p>

Sequence quality can be controlled by manual trace viewing and
quality clipping or automatic quality clipping programs.  Low-
quality bases should be changed to N's or can be made part of
Excluded Regions. The beginning of a sequencing read is often
problematic because of primer peaks, and the end of the read
often contains many low-quality or even meaningless called bases.
Therefore when picking primers from single-pass sequence it is
often best to use the INCLUDED_REGION parameter to ensure that
eprimer3 chooses primers in the high quality region of the read.

<p>

In addition, eprimer3 takes as input a Sequence Quality list for
use with those base calling programs 

<p>

(e.g. Phred, Bass/Grace, Trout) that output this information.

<H4>What to do if eprimer3 cannot find a primers?</H4>

Try relaxing various parameters, including the
self-complementarity parameters and max and min oligo melting
temperatures.  For example, for very A-T-rich regions you might
have to increase maximum primer size or decrease minimum melting
temperature.  It is usually unwise to reduce the minimum primer
size if your template is complex (e.g. a mammalian genome), since
small primers are more likely to be non-specific.  Make sure that
there are adequate stretches of non-Ns in the regions in which
you wish to pick primers.  If necessary you can also allow an N
in your primer and use an oligo mixture containing all four bases
at that position.

<p>

Try setting the '-explain' option.






<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->


Here is a sample session with eprimer3:

<pre>

% eprimer3 em:hsfau1 hsfau.eprimer3 -explain

</pre>














<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     The sequence from which to choose primers.
                                  The sequence must be presented 5' to 3'
  [-outfile]           outfile    Output file name

   Optional qualifiers (* if not always prompted):
   -task               menu       Tell EPrimer3 what task to perform. Legal
                                  values are 0: 'Pick PCR primers', 1: 'Pick
                                  PCR primers and hybridization probe', 2:
                                  'Pick forward primer only', 3: 'Pick reverse
                                  primer only', 4: 'Pick hybridization probe
                                  only'.
                                  The tasks should be self explanatory.
                                  Briefly, an 'internal oligo' is intended to
                                  be used as a hybridization probe (hyb probe)
                                  to detect the PCR product after
                                  amplification.
   -numreturn          integer    The maximum number of primer pairs to
                                  return. Primer pairs returned are sorted by
                                  their 'quality', in other words by the value
                                  of the objective function (where a lower
                                  number indicates a better primer pair).
                                  Caution: setting this parameter to a large
                                  value will increase running time.
   -includedregion     range      A sub-region of the given sequence in which
                                  to pick primers. For example, often the
                                  first dozen or so bases of a sequence are
                                  vector, and should be excluded from
                                  consideration. The value for this parameter
                                  has the form
                                  (start),(end)
                                  where (start) is the index of the first base
                                  to consider, and (end) is the last in the
                                  primer-picking region.
   -target             range      If one or more Targets is specified then a
                                  legal primer pair must flank at least one of
                                  them. A Target might be a simple sequence
                                  repeat site (for example a CA repeat) or a
                                  single-base-pair polymorphism. The value
                                  should be a space-separated list of
                                  (start),(end)
                                  pairs where (start) is the index of the
                                  first base of a Target, and (end) is the
                                  last
                                  E.g. 50,51 requires primers to surround the
                                  2 bases at positions 50 and 51.
   -excludedregion     range      Primer oligos may not overlap any region
                                  specified in this tag. The associated value
                                  must be a space-separated list of
                                  (start),(end)
                                  pairs where (start) is the index of the
                                  first base of the excluded region, and and
                                  (end) is the last. This tag is useful for
                                  tasks such as excluding regions of low
                                  sequence quality or for excluding regions
                                  containing repetitive elements such as ALUs
                                  or LINEs.
                                  E.g. 401,407 68,70 forbids selection of
                                  primers in the 7 bases starting at 401 and
                                  the 3 bases at 68.
   -forwardinput       string     The sequence of a forward primer to check
                                  and around which to design reverse primers
                                  and optional internal oligos. Must be a
                                  substring of SEQUENCE.
   -reverseinput       string     The sequence of a reverse primer to check
                                  and around which to design forward primers
                                  and optional internal oligos. Must be a
                                  substring of the reverse strand of SEQUENCE.
*  -gcclamp            integer    Require the specified number of consecutive
                                  Gs and Cs at the 3' end of both the forward
                                  and reverse primer. (This parameter has no
                                  effect on the internal oligo if one is
                                  requested.)
*  -osize              integer    Optimum length (in bases) of a primer oligo.
                                  EPrimer3 will attempt to pick primers close
                                  to this length.
*  -minsize            integer    Minimum acceptable length of a primer. Must
                                  be greater than 0 and less than or equal to
                                  MAX-SIZE.
*  -maxsize            integer    Maximum acceptable length (in bases) of a
                                  primer. Currently this parameter cannot be
                                  larger than 35. This limit is governed by
                                  the maximum oligo size for which EPrimer3's
                                  melting-temperature is valid.
*  -otm                float      Optimum melting temperature(Celsius) for a
                                  primer oligo. EPrimer3 will try to pick
                                  primers with melting temperatures are close
                                  to this temperature. The oligo melting
                                  temperature formula in EPrimer3 is that
                                  given in Rychlik, Spencer and Rhoads,
                                  Nucleic Acids Research, vol 18, num 12, pp
                                  6409-6412 and Breslauer, Frank, Bloeker and
                                  Marky, Proc. Natl. Acad. Sci. USA, vol 83,
                                  pp 3746-3750. Please refer to the former
                                  paper for background discussion.
*  -mintm              float      Minimum acceptable melting
                                  temperature(Celsius) for a primer oligo.
*  -maxtm              float      Maximum acceptable melting
                                  temperature(Celsius) for a primer oligo.
*  -maxdifftm          float      Maximum acceptable (unsigned) difference
                                  between the melting temperatures of the
                                  forward and reverse primers.
*  -ogcpercent         float      Primer optimum GC percent.
*  -mingc              float      Minimum allowable percentage of Gs and Cs in
                                  any primer.
*  -maxgc              float      Maximum allowable percentage of Gs and Cs in
                                  any primer generated by Primer.
*  -saltconc           float      The millimolar concentration of salt
                                  (usually KCl) in the PCR. EPrimer3 uses this
                                  argument to calculate oligo melting
                                  temperatures.
*  -dnaconc            float      The nanomolar concentration of annealing
                                  oligos in the PCR. EPrimer3 uses this
                                  argument to calculate oligo melting
                                  temperatures. The default (50nM) works well
                                  with the standard protocol used at the
                                  Whitehead/MIT Center for Genome
                                  Research--0.5 microliters of 20 micromolar
                                  concentration for each primer oligo in a 20
                                  microliter reaction with 10 nanograms
                                  template, 0.025 units/microliter Taq
                                  polymerase in 0.1 mM each dNTP, 1.5mM MgCl2,
                                  50mM KCl, 10mM Tris-HCL (pH 9.3) using 35
                                  cycles with an annealing temperature of 56
                                  degrees Celsius. This parameter corresponds
                                  to 'c' in Rychlik, Spencer and Rhoads'
                                  equation (ii) (Nucleic Acids Research, vol
                                  18, num 12) where a suitable value (for a
                                  lower initial concentration of template) is
                                  'empirically determined'. The value of this
                                  parameter is less than the actual
                                  concentration of oligos in the reaction
                                  because it is the concentration of annealing
                                  oligos, which in turn depends on the amount
                                  of template (including PCR product) in a
                                  given cycle. This concentration increases a
                                  great deal during a PCR; fortunately PCR
                                  seems quite robust for a variety of oligo
                                  melting temperatures.
                                  See ADVICE FOR PICKING PRIMERS.
*  -maxpolyx           integer    The maximum allowable length of a
                                  mononucleotide repeat in a primer, for
                                  example AAAAAA.
*  -productosize       integer    The optimum size for the PCR product. 0
                                  indicates that there is no optimum product
                                  size.
*  -productsizerange   range      The associated values specify the lengths of
                                  the product that the user wants the primers
                                  to create, and is a space separated list of
                                  elements of the form
                                  (x)-(y)
                                  where an (x)-(y) pair is a legal range of
                                  lengths for the product. For example, if one
                                  wants PCR products to be between 100 to 150
                                  bases (inclusive) then one would set this
                                  parameter to 100-150. If one desires PCR
                                  products in either the range from 100 to 150
                                  bases or in the range from 200 to 250 bases
                                  then one would set this parameter to
                                  100-150 200-250.
                                  EPrimer3 favors ranges to the left side of
                                  the parameter string. EPrimer3 will return
                                  legal primers pairs in the first range
                                  regardless the value of the objective
                                  function for these pairs. Only if there are
                                  an insufficient number of primers in the
                                  first range will EPrimer3 return primers in
                                  a subsequent range.
*  -productotm         float      The optimum melting temperature for the PCR
                                  product. 0 indicates that there is no
                                  optimum temperature.
*  -productmintm       float      The minimum allowed melting temperature of
                                  the amplicon. Please see the documentation
                                  on the maximum melting temperature of the
                                  product for details.
*  -productmaxtm       float      The maximum allowed melting temperature of
                                  the amplicon. Product Tm is calculated using
                                  the formula from Bolton and McCarthy, PNAS
                                  84:1390 (1962) as presented in Sambrook,
                                  Fritsch and Maniatis, Molecular Cloning, p
                                  11.46 (1989, CSHL Press).
                                  Tm = 81.5 + 16.6(log10([Na+])) + .41*(%GC) -
                                  600/length
                                  Where [Na+} is the molar sodium
                                  concentration, (%GC) is the percent of Gs
                                  and Cs in the sequence, and length is the
                                  length of the sequence.
                                  A similar formula is used by the prime
                                  primer selection program in GCG
                                  http://www.gcg.com), which instead uses
                                  675.0/length in the last term (after F.
                                  Baldino, Jr, M.-F. Chesselet, and M.E.
                                  Lewis, Methods in Enzymology 168:766 (1989)
                                  eqn (1) on page 766 without the mismatch and
                                  formamide terms). The formulas here and in
                                  Baldino et al. assume Na+ rather than K+.
                                  According to J.G. Wetmur, Critical Reviews
                                  in BioChem. and Mol. Bio. 26:227 (1991) 50
                                  mM K+ should be equivalent in these formulae
                                  to .2 M Na+. EPrimer3 uses the same salt
                                  concentration value for calculating both the
                                  primer melting temperature and the oligo
                                  melting temperature. If you are planning to
                                  use the PCR product for hybridization later
                                  this behavior will not give you the Tm under
                                  hybridization conditions.
*  -oligoexcludedregion range      Middle oligos may not overlap any region
                                  specified by this tag. The associated value
                                  must be a space-separated list of
                                  (start),(end)
                                  pairs, where (start) is the index of the
                                  first base of an excluded region, and (end)
                                  is the last. Often one would make Target
                                  regions excluded regions for internal
                                  oligos.
*  -oligoinput         string     The sequence of an internal oligo to check
                                  and around which to design forward and
                                  reverse primers. Must be a substring of
                                  SEQUENCE.
*  -oligoosize         integer    Optimum length (in bases) of an internal
                                  oligo. EPrimer3 will attempt to pick primers
                                  close to this length.
*  -oligominsize       integer    Minimum acceptable length of an internal
                                  oligo. Must be greater than 0 and less than
                                  or equal to INTERNAL-OLIGO-MAX-SIZE.
*  -oligomaxsize       integer    Maximum acceptable length (in bases) of an
                                  internal oligo. Currently this parameter
                                  cannot be larger than 35. This limit is
                                  governed by maximum oligo size for which
                                  EPrimer3's melting-temperature is valid.
*  -oligootm           float      Optimum melting temperature (Celsius) for an
                                  internal oligo. EPrimer3 will try to pick
                                  oligos with melting temperatures that are
                                  close to this temperature. The oligo melting
                                  temperature formula in EPrimer3 is that
                                  given in Rychlik, Spencer and Rhoads,
                                  Nucleic Acids Research, vol 18, num 12, pp
                                  6409-6412 and Breslauer, Frank, Bloeker and
                                  Marky, Proc. Natl. Acad. Sci. USA, vol 83,
                                  pp 3746-3750. Please refer to the former
                                  paper for background discussion.
*  -oligomintm         float      Minimum acceptable melting
                                  temperature(Celsius) for an internal oligo.
*  -oligomaxtm         float      Maximum acceptable melting temperature
                                  (Celsius) for an internal oligo.
*  -oligoogcpercent    float      Internal oligo optimum GC percent.
*  -oligomingc         float      Minimum allowable percentage of Gs and Cs in
                                  an internal oligo.
*  -oligomaxgc         float      Maximum allowable percentage of Gs and Cs in
                                  any internal oligo generated by Primer.
*  -oligosaltconc      float      The millimolar concentration of salt
                                  (usually KCl) in the hybridization. EPrimer3
                                  uses this argument to calculate internal
                                  oligo melting temperatures.
*  -oligodnaconc       float      The nanomolar concentration of annealing
                                  internal oligo in the hybridization.
*  -oligoselfany       float      The maximum allowable local alignment score
                                  when testing an internal oligo for (local)
                                  self-complementarity. Local
                                  self-complementarity is taken to predict the
                                  tendency of oligos to anneal to themselves
                                  The scoring system gives 1.00 for
                                  complementary bases, -0.25 for a match of
                                  any base (or N) with an N, -1.00 for a
                                  mismatch, and -2.00 for a gap. Only
                                  single-base-pair gaps are allowed. For
                                  example, the alignment
                                  5' ATCGNA 3'
                                  || | |
                                  3' TA-CGT 5'
                                  is allowed (and yields a score of 1.75), but
                                  the alignment
                                  5' ATCCGNA 3'
                                  || | |
                                  3' TA--CGT 5'
                                  is not considered. Scores are non-negative,
                                  and a score of 0.00 indicates that there is
                                  no reasonable local alignment between two
                                  oligos.
*  -oligoselfend       float      The maximum allowable 3'-anchored global
                                  alignment score when testing a single oligo
                                  for self-complementarity.
                                  The scoring system is as for the Maximum
                                  Complementarity argument. In the examples
                                  above the scores are 7.00 and 6.00
                                  respectively. Scores are non-negative, and a
                                  score of 0.00 indicates that there is no
                                  reasonable 3'-anchored global alignment
                                  between two oligos. In order to estimate
                                  3'-anchored global alignments for candidate
                                  oligos, Primer assumes that the sequence
                                  from which to choose oligos is presented 5'
                                  to 3'.
                                  INTERNAL-OLIGO-SELF-END is meaningless when
                                  applied to internal oligos used for
                                  hybridization-based detection, since
                                  primer-dimer will not occur. We recommend
                                  that INTERNAL-OLIGO-SELF-END be set at least
                                  as high as INTERNAL-OLIGO-SELF-ANY.
*  -oligomaxpolyx      integer    The maximum allowable length of an internal
                                  oligo mononucleotide repeat, for example
                                  AAAAAA.

   Advanced qualifiers:
   -explainflag        bool       If this flag is non-0, produce LEFT-EXPLAIN,
                                  RIGHT-EXPLAIN, and INTERNAL-OLIGO-EXPLAIN
                                  output tags, which are intended to provide
                                  information on the number of oligos and
                                  primer pairs that EPrimer3 examined, and
                                  statistics on the number discarded for
                                  various reasons.
   -fileflag           bool       If the associated value is non-0, then
                                  EPrimer3 creates two output files for each
                                  input SEQUENCE. File (sequence-id).for lists
                                  all acceptable forward primers for
                                  (sequence-id), and (sequence-id).rev lists
                                  all acceptable reverse primers for
                                  (sequence-id), where (sequence-id) is the
                                  value of the SEQUENCE-ID tag (which must be
                                  supplied). In addition, if the input tag
                                  TASK is 1 or 4, EPrimer3 produces a file
                                  (sequence-id).int, which lists all
                                  acceptable internal oligos.
   -firstbaseindex     integer    This parameter is the index of the first
                                  base in the input sequence. For input and
                                  output using 1-based indexing (such as that
                                  used in GenBank and to which many users are
                                  accustomed) set this parameter to 1. For
                                  input and output using 0-based indexing set
                                  this parameter to 0. (This parameter also
                                  affects the indexes in the contents of the
                                  files produced when the primer file flag is
                                  set.)
   -pickanyway         bool       If true pick a primer pair even if
                                  LEFT-INPUT, RIGHT-INPUT, or
                                  INTERNAL-OLIGO-INPUT violates specific
                                  constraints.
   -mispriminglibrary  infile     The name of a file containing a nucleotide
                                  sequence library of sequences to avoid
                                  amplifying (for example repetitive
                                  sequences, or possibly the sequences of
                                  genes in a gene family that should not be
                                  amplified.) The file must be in (a slightly
                                  restricted) FASTA format (W. B. Pearson and
                                  D.J. Lipman, PNAS 85:8 pp 2444-2448 [1988]);
                                  we briefly discuss the organization of this
                                  file below. If this parameter is specified
                                  then EPrimer3 locally aligns each candidate
                                  primer against each library sequence and
                                  rejects those primers for which the local
                                  alignment score times a specified weight
                                  (see below) exceeds MAX-MISPRIMING. (The
                                  maximum value of the weight is arbitrarily
                                  set to 100.0.)
                                  Each sequence entry in the FASTA-format file
                                  must begin with an 'id line' that starts
                                  with '>'. The contents of the id line is
                                  'slightly restricted' in that EPrimer3
                                  parses everything after any optional
                                  asterisk ('*') as a floating point number to
                                  use as the weight mentioned above. If the
                                  id line contains no asterisk then the weight
                                  defaults to 1.0. The alignment scoring
                                  system used is the same as for calculating
                                  complementarity among oligos (e.g.
                                  SELF-ANY). The remainder of an entry
                                  contains the sequence as lines following the
                                  id line up until a line starting with '>'
                                  or the end of the file. Whitespace and
                                  newlines are ignored. Characters 'A', 'T',
                                  'G', 'C', 'a', 't', 'g', 'c' are retained
                                  and any other character is converted to 'N'
                                  (with the consequence that any IUB / IUPAC
                                  codes for ambiguous bases are converted to
                                  'N'). There are no restrictions on line
                                  length.
                                  An empty value for this parameter indicates
                                  that no repeat library should be used.
   -maxmispriming      float      The maximum allowed weighted similarity with
                                  any sequence in MISPRIMING-LIBRARY.
   -pairmaxmispriming  float      The maximum allowed sum of weighted
                                  similarities of a primer pair (one
                                  similarity for each primer) with any single
                                  sequence in MISPRIMING-LIBRARY.
   -numnsaccepted      integer    Maximum number of unknown bases (N)
                                  allowable in any primer.
   -selfany            float      The maximum allowable local alignment score
                                  when testing a single primer for (local)
                                  self-complementarity and the maximum
                                  allowable local alignment score when testing
                                  for complementarity between forward and
                                  reverse primers. Local self-complementarity
                                  is taken to predict the tendency of primers
                                  to anneal to each other without necessarily
                                  causing self-priming in the PCR. The scoring
                                  system gives 1.00 for complementary bases,
                                  -0.25 for a match of any base (or N) with an
                                  N, -1.00 for a mismatch, and -2.00 for a
                                  gap. Only single-base-pair gaps are allowed.
                                  For example, the alignment
                                  5' ATCGNA 3'
                                  ...|| | |
                                  3' TA-CGT 5'
                                  is allowed (and yields a score of 1.75), but
                                  the alignment
                                  5' ATCCGNA 3'
                                  ...|| | |
                                  3' TA--CGT 5'
                                  is not considered. Scores are non-negative,
                                  and a score of 0.00 indicates that there is
                                  no reasonable local alignment between two
                                  oligos.
   -selfend            float      The maximum allowable 3'-anchored global
                                  alignment score when testing a single primer
                                  for self-complementarity, and the maximum
                                  allowable 3'-anchored global alignment score
                                  when testing for complementarity between
                                  forward and reverse primers. The 3'-anchored
                                  global alignment score is taken to predict
                                  the likelihood of PCR-priming primer-dimers,
                                  for example
                                  5' ATGCCCTAGCTTCCGGATG 3'
                                  .............||| |||||
                                  ..........3' AAGTCCTACATTTAGCCTAGT 5'
                                  or
                                  5' AGGCTATGGGCCTCGCGA 3'
                                  ...............||||||
                                  ............3' AGCGCTCCGGGTATCGGA 5'
                                  The scoring system is as for the Maximum
                                  Complementarity argument. In the examples
                                  above the scores are 7.00 and 6.00
                                  respectively. Scores are non-negative, and a
                                  score of 0.00 indicates that there is no
                                  reasonable 3'-anchored global alignment
                                  between two oligos. In order to estimate
                                  3'-anchored global alignments for candidate
                                  primers and primer pairs, Primer assumes
                                  that the sequence from which to choose
                                  primers is presented 5' to 3'. It is
                                  nonsensical to provide a larger value for
                                  this parameter than for the Maximum (local)
                                  Complementarity parameter because the score
                                  of a local alignment will always be at least
                                  as great as the score of a global
                                  alignment.
   -maxendstability    float      The maximum stability for the five 3' bases
                                  of a forward or reverse primer. Bigger
                                  numbers mean more stable 3' ends. The value
                                  is the maximum delta G for duplex disruption
                                  for the five 3' bases as calculated using
                                  the nearest neighbor parameters published in
                                  Breslauer, Frank, Bloeker and Marky, Proc.
                                  Natl. Acad. Sci. USA, vol 83, pp 3746-3750.
                                  EPrimer3 uses a completely permissive
                                  default value for backward compatibility
                                  (which we may change in the next release).
                                  Rychlik recommends a maximum value of 9
                                  (Wojciech Rychlik, 'Selection of Primers for
                                  Polymerase Chain Reaction' in BA White,
                                  Ed., 'Methods in Molecular Biology, Vol. 15:
                                  PCR Protocols: Current Methods and
                                  Applications', 1993, pp 31-40, Humana Press,
                                  Totowa NJ).
   -oligomishyblibrary infile     Similar to MISPRIMING-LIBRARY, except that
                                  the event we seek to avoid is hybridization
                                  of the internal oligo to sequences in this
                                  library rather than priming from them.
                                  The file must be in (a slightly restricted)
                                  FASTA format (W. B. Pearson and D.J. Lipman,
                                  PNAS 85:8 pp 2444-2448 [1988]); we briefly
                                  discuss the organization of this file below.
                                  If this parameter is specified then
                                  EPrimer3 locally aligns each candidate oligo
                                  against each library sequence and rejects
                                  those primers for which the local alignment
                                  score times a specified weight (see below)
                                  exceeds INTERNAL-OLIGO-MAX-MISHYB. (The
                                  maximum value of the weight is arbitrarily
                                  set to 12.0.)
                                  Each sequence entry in the FASTA-format file
                                  must begin with an 'id line' that starts
                                  with '>'. The contents of the id line is
                                  'slightly restricted' in that EPrimer3
                                  parses everything after any optional
                                  asterisk ('*') as a floating point number to
                                  use as the weight mentioned above. If the
                                  id line contains no asterisk then the weight
                                  defaults to 1.0. The alignment scoring
                                  system used is the same as for calculating
                                  complementarity among oligos (e.g.
                                  SELF-ANY). The remainder of an entry
                                  contains the sequence as lines following the
                                  id line up until a line starting with '>'
                                  or the end of the file. Whitespace and
                                  newlines are ignored. Characters 'A', 'T',
                                  'G', 'C', 'a', 't', 'g', 'c' are retained
                                  and any other character is converted to 'N'
                                  (with the consequence that any IUB / IUPAC
                                  codes for ambiguous bases are converted to
                                  'N'). There are no restrictions on line
                                  length.
                                  An empty value for this parameter indicates
                                  that no library should be used.
   -oligomaxmishyb     float      Similar to MAX-MISPRIMING except that this
                                  parameter applies to the similarity of
                                  candidate internal oligos to the library
                                  specified in INTERNAL-OLIGO-MISHYB-LIBRARY.

   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>The sequence from which to choose primers. The sequence must be presented 5' to 3'</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.eprimer3</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-task</td>
<td>Tell EPrimer3 what task to perform. Legal values are 0: 'Pick PCR primers', 1: 'Pick PCR primers and hybridization probe', 2: 'Pick forward primer only', 3: 'Pick reverse primer only', 4: 'Pick hybridization probe only'.
The tasks should be self explanatory.
Briefly, an 'internal oligo' is intended to be used as a hybridization probe (hyb probe) to detect the PCR product after amplification.</td>
<td><table><tr><td>0</td> <td><i>(Pick PCR primers)</i></td></tr><tr><td>1</td> <td><i>(Pick PCR primers and hybridization probe)</i></td></tr><tr><td>2</td> <td><i>(Pick forward primer only)</i></td></tr><tr><td>3</td> <td><i>(Pick reverse primer only)</i></td></tr><tr><td>4</td> <td><i>(Pick hybridization probe only)</i></td></tr></table></td>
<td>0</td>
</tr>

<tr>
<td>-numreturn</td>
<td>The maximum number of primer pairs to return. Primer pairs returned are sorted by their 'quality', in other words by the value of the objective function (where a lower number indicates a better primer pair). Caution: setting this parameter to a large value will increase running time.</td>
<td>Integer 0 or more</td>
<td>5</td>
</tr>

<tr>
<td>-includedregion</td>
<td>A sub-region of the given sequence in which to pick primers. For example, often the first dozen or so bases of a sequence are vector, and should be excluded from consideration. The value for this parameter has the form
(start),(end)
where (start) is the index of the first base to consider, and (end) is the last in the primer-picking region.</td>
<td>Sequence range</td>
<td><i>full sequence</i></td>
</tr>

<tr>
<td>-target</td>
<td>If one or more Targets is specified then a legal primer pair must flank at least one of them. A Target might be a simple sequence repeat site (for example a CA repeat) or a single-base-pair polymorphism. The value should be a space-separated list of
(start),(end)
pairs where (start) is the index of the first base of a Target, and (end) is the last
E.g. 50,51 requires primers to surround the 2 bases at positions 50 and 51.</td>
<td>Sequence range</td>
<td><i>full sequence</i></td>
</tr>

<tr>
<td>-excludedregion</td>
<td>Primer oligos may not overlap any region specified in this tag. The associated value must be a space-separated list of
(start),(end)
pairs where (start) is the index of the first base of the excluded region, and and (end) is the last. This tag is useful for tasks such as excluding regions of low sequence quality or for excluding regions containing repetitive elements such as ALUs or LINEs.
E.g. 401,407 68,70 forbids selection of primers in the 7 bases starting at 401 and the 3 bases at 68.</td>
<td>Sequence range</td>
<td><i>full sequence</i></td>
</tr>

<tr>
<td>-forwardinput</td>
<td>The sequence of a forward primer to check and around which to design reverse primers and optional internal oligos. Must be a substring of SEQUENCE.</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-reverseinput</td>
<td>The sequence of a reverse primer to check and around which to design forward primers and optional internal oligos. Must be a substring of the reverse strand of SEQUENCE.</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-gcclamp</td>
<td>Require the specified number of consecutive Gs and Cs at the 3' end of both the forward and reverse primer. (This parameter has no effect on the internal oligo if one is requested.)</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr>
<td>-osize</td>
<td>Optimum length (in bases) of a primer oligo. EPrimer3 will attempt to pick primers close to this length.</td>
<td>Integer 1 or more</td>
<td>20</td>
</tr>

<tr>
<td>-minsize</td>
<td>Minimum acceptable length of a primer. Must be greater than 0 and less than or equal to MAX-SIZE.</td>
<td>Integer 1 or more</td>
<td>18</td>
</tr>

<tr>
<td>-maxsize</td>
<td>Maximum acceptable length (in bases) of a primer. Currently this parameter cannot be larger than 35. This limit is governed by the maximum oligo size for which EPrimer3's melting-temperature is valid.</td>
<td>Integer up to 35</td>
<td>27</td>
</tr>

<tr>
<td>-otm</td>
<td>Optimum melting temperature(Celsius) for a primer oligo. EPrimer3 will try to pick primers with melting temperatures are close to this temperature. The oligo melting temperature formula in EPrimer3 is that given in Rychlik, Spencer and Rhoads, Nucleic Acids Research, vol 18, num 12, pp 6409-6412 and Breslauer, Frank, Bloeker and Marky, Proc. Natl. Acad. Sci. USA, vol 83, pp 3746-3750. Please refer to the former paper for background discussion.</td>
<td>Any integer value</td>
<td>60.0</td>
</tr>

<tr>
<td>-mintm</td>
<td>Minimum acceptable melting temperature(Celsius) for a primer oligo.</td>
<td>Any integer value</td>
<td>57.0</td>
</tr>

<tr>
<td>-maxtm</td>
<td>Maximum acceptable melting temperature(Celsius) for a primer oligo.</td>
<td>Any integer value</td>
<td>63.0</td>
</tr>

<tr>
<td>-maxdifftm</td>
<td>Maximum acceptable (unsigned) difference between the melting temperatures of the forward and reverse primers.</td>
<td>Any integer value</td>
<td>100.0</td>
</tr>

<tr>
<td>-ogcpercent</td>
<td>Primer optimum GC percent.</td>
<td>Any integer value</td>
<td>50.0</td>
</tr>

<tr>
<td>-mingc</td>
<td>Minimum allowable percentage of Gs and Cs in any primer.</td>
<td>Any integer value</td>
<td>20.0</td>
</tr>

<tr>
<td>-maxgc</td>
<td>Maximum allowable percentage of Gs and Cs in any primer generated by Primer.</td>
<td>Any integer value</td>
<td>80.0</td>
</tr>

<tr>
<td>-saltconc</td>
<td>The millimolar concentration of salt (usually KCl) in the PCR. EPrimer3 uses this argument to calculate oligo melting temperatures.</td>
<td>Any integer value</td>
<td>50.0</td>
</tr>

<tr>
<td>-dnaconc</td>
<td>The nanomolar concentration of annealing oligos in the PCR. EPrimer3 uses this argument to calculate oligo melting temperatures. The default (50nM) works well with the standard protocol used at the Whitehead/MIT Center for Genome Research--0.5 microliters of 20 micromolar concentration for each primer oligo in a 20 microliter reaction with 10 nanograms template, 0.025 units/microliter Taq polymerase in 0.1 mM each dNTP, 1.5mM MgCl2, 50mM KCl, 10mM Tris-HCL (pH 9.3) using 35 cycles with an annealing temperature of 56 degrees Celsius. This parameter corresponds to 'c' in Rychlik, Spencer and Rhoads' equation (ii) (Nucleic Acids Research, vol 18, num 12) where a suitable value (for a lower initial concentration of template) is 'empirically determined'. The value of this parameter is less than the actual concentration of oligos in the reaction because it is the concentration of annealing oligos, which in turn depends on the amount of template (including PCR product) in a given cycle. This concentration increases a great deal during a PCR; fortunately PCR seems quite robust for a variety of oligo melting temperatures.
See ADVICE FOR PICKING PRIMERS.</td>
<td>Any integer value</td>
<td>50.0</td>
</tr>

<tr>
<td>-maxpolyx</td>
<td>The maximum allowable length of a mononucleotide repeat in a primer, for example AAAAAA.</td>
<td>Integer 0 or more</td>
<td>5</td>
</tr>

<tr>
<td>-productosize</td>
<td>The optimum size for the PCR product. 0 indicates that there is no optimum product size.</td>
<td>Integer 0 or more</td>
<td>200</td>
</tr>

<tr>
<td>-productsizerange</td>
<td>The associated values specify the lengths of the product that the user wants the primers to create, and is a space separated list of elements of the form
(x)-(y)
where an (x)-(y) pair is a legal range of lengths for the product. For example, if one wants PCR products to be between 100 to 150 bases (inclusive) then one would set this parameter to 100-150. If one desires PCR products in either the range from 100 to 150 bases or in the range from 200 to 250 bases then one would set this parameter to 100-150 200-250.
EPrimer3 favors ranges to the left side of the parameter string. EPrimer3 will return legal primers pairs in the first range regardless the value of the objective function for these pairs. Only if there are an insufficient number of primers in the first range will EPrimer3 return primers in a subsequent range.</td>
<td>Sequence range</td>
<td>100-300</td>
</tr>

<tr>
<td>-productotm</td>
<td>The optimum melting temperature for the PCR product. 0 indicates that there is no optimum temperature.</td>
<td>Any integer value</td>
<td>0.0</td>
</tr>

<tr>
<td>-productmintm</td>
<td>The minimum allowed melting temperature of the amplicon. Please see the documentation on the maximum melting temperature of the product for details.</td>
<td>Any integer value</td>
<td>-1000000.0</td>
</tr>

<tr>
<td>-productmaxtm</td>
<td>The maximum allowed melting temperature of the amplicon. Product Tm is calculated using the formula from Bolton and McCarthy, PNAS 84:1390 (1962) as presented in Sambrook, Fritsch and Maniatis, Molecular Cloning, p 11.46 (1989, CSHL Press).
Tm = 81.5 + 16.6(log10([Na+])) + .41*(%GC) - 600/length
Where [Na+} is the molar sodium concentration, (%GC) is the percent of Gs and Cs in the sequence, and length is the length of the sequence.
A similar formula is used by the prime primer selection program in GCG http://www.gcg.com), which instead uses 675.0/length in the last term (after F. Baldino, Jr, M.-F. Chesselet, and M.E. Lewis, Methods in Enzymology 168:766 (1989) eqn (1) on page 766 without the mismatch and formamide terms). The formulas here and in Baldino et al. assume Na+ rather than K+. According to J.G. Wetmur, Critical Reviews in BioChem. and Mol. Bio. 26:227 (1991) 50 mM K+ should be equivalent in these formulae to .2 M Na+. EPrimer3 uses the same salt concentration value for calculating both the primer melting temperature and the oligo melting temperature. If you are planning to use the PCR product for hybridization later this behavior will not give you the Tm under hybridization conditions.</td>
<td>Any integer value</td>
<td>1000000.0</td>
</tr>

<tr>
<td>-oligoexcludedregion</td>
<td>Middle oligos may not overlap any region specified by this tag. The associated value must be a space-separated list of
(start),(end)
pairs, where (start) is the index of the first base of an excluded region, and (end) is the last. Often one would make Target regions excluded regions for internal oligos.</td>
<td>Sequence range</td>
<td><i>full sequence</i></td>
</tr>

<tr>
<td>-oligoinput</td>
<td>The sequence of an internal oligo to check and around which to design forward and reverse primers. Must be a substring of SEQUENCE.</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-oligoosize</td>
<td>Optimum length (in bases) of an internal oligo. EPrimer3 will attempt to pick primers close to this length.</td>
<td>Integer 0 or more</td>
<td>20</td>
</tr>

<tr>
<td>-oligominsize</td>
<td>Minimum acceptable length of an internal oligo. Must be greater than 0 and less than or equal to INTERNAL-OLIGO-MAX-SIZE.</td>
<td>Integer 0 or more</td>
<td>18</td>
</tr>

<tr>
<td>-oligomaxsize</td>
<td>Maximum acceptable length (in bases) of an internal oligo. Currently this parameter cannot be larger than 35. This limit is governed by maximum oligo size for which EPrimer3's melting-temperature is valid.</td>
<td>Integer up to 35</td>
<td>27</td>
</tr>

<tr>
<td>-oligootm</td>
<td>Optimum melting temperature (Celsius) for an internal oligo. EPrimer3 will try to pick oligos with melting temperatures that are close to this temperature. The oligo melting temperature formula in EPrimer3 is that given in Rychlik, Spencer and Rhoads, Nucleic Acids Research, vol 18, num 12, pp 6409-6412 and Breslauer, Frank, Bloeker and Marky, Proc. Natl. Acad. Sci. USA, vol 83, pp 3746-3750. Please refer to the former paper for background discussion.</td>
<td>Any integer value</td>
<td>60.0</td>
</tr>

<tr>
<td>-oligomintm</td>
<td>Minimum acceptable melting temperature(Celsius) for an internal oligo.</td>
<td>Any integer value</td>
<td>57.0</td>
</tr>

<tr>
<td>-oligomaxtm</td>
<td>Maximum acceptable melting temperature (Celsius) for an internal oligo.</td>
<td>Any integer value</td>
<td>63.0</td>
</tr>

<tr>
<td>-oligoogcpercent</td>
<td>Internal oligo optimum GC percent.</td>
<td>Any integer value</td>
<td>50.0</td>
</tr>

<tr>
<td>-oligomingc</td>
<td>Minimum allowable percentage of Gs and Cs in an internal oligo.</td>
<td>Any integer value</td>
<td>20.0</td>
</tr>

<tr>
<td>-oligomaxgc</td>
<td>Maximum allowable percentage of Gs and Cs in any internal oligo generated by Primer.</td>
<td>Any integer value</td>
<td>80.0</td>
</tr>

<tr>
<td>-oligosaltconc</td>
<td>The millimolar concentration of salt (usually KCl) in the hybridization. EPrimer3 uses this argument to calculate internal oligo melting temperatures.</td>
<td>Any integer value</td>
<td>50.0</td>
</tr>

<tr>
<td>-oligodnaconc</td>
<td>The nanomolar concentration of annealing internal oligo in the hybridization.</td>
<td>Any integer value</td>
<td>50.0</td>
</tr>

<tr>
<td>-oligoselfany</td>
<td>The maximum allowable local alignment score when testing an internal oligo for (local) self-complementarity. Local self-complementarity is taken to predict the tendency of oligos to anneal to themselves The scoring system gives 1.00 for complementary bases, -0.25 for a match of any base (or N) with an N, -1.00 for a mismatch, and -2.00 for a gap. Only single-base-pair gaps are allowed. For example, the alignment
5' ATCGNA 3'
|| | |
3' TA-CGT 5'
is allowed (and yields a score of 1.75), but the alignment
5' ATCCGNA 3'
|| | |
3' TA--CGT 5'
is not considered. Scores are non-negative, and a score of 0.00 indicates that there is no reasonable local alignment between two oligos.</td>
<td>Number up to 9999.990</td>
<td>12.00</td>
</tr>

<tr>
<td>-oligoselfend</td>
<td>The maximum allowable 3'-anchored global alignment score when testing a single oligo for self-complementarity.
The scoring system is as for the Maximum Complementarity argument. In the examples above the scores are 7.00 and 6.00 respectively. Scores are non-negative, and a score of 0.00 indicates that there is no reasonable 3'-anchored global alignment between two oligos. In order to estimate 3'-anchored global alignments for candidate oligos, Primer assumes that the sequence from which to choose oligos is presented 5' to 3'.
INTERNAL-OLIGO-SELF-END is meaningless when applied to internal oligos used for hybridization-based detection, since primer-dimer will not occur. We recommend that INTERNAL-OLIGO-SELF-END be set at least as high as INTERNAL-OLIGO-SELF-ANY.</td>
<td>Number up to 9999.990</td>
<td>12.00</td>
</tr>

<tr>
<td>-oligomaxpolyx</td>
<td>The maximum allowable length of an internal oligo mononucleotide repeat, for example AAAAAA.</td>
<td>Integer 0 or more</td>
<td>5</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-explainflag</td>
<td>If this flag is non-0, produce LEFT-EXPLAIN, RIGHT-EXPLAIN, and INTERNAL-OLIGO-EXPLAIN output tags, which are intended to provide information on the number of oligos and primer pairs that EPrimer3 examined, and statistics on the number discarded for various reasons.</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-fileflag</td>
<td>If the associated value is non-0, then EPrimer3 creates two output files for each input SEQUENCE. File (sequence-id).for lists all acceptable forward primers for (sequence-id), and (sequence-id).rev lists all acceptable reverse primers for (sequence-id), where (sequence-id) is the value of the SEQUENCE-ID tag (which must be supplied). In addition, if the input tag TASK is 1 or 4, EPrimer3 produces a file (sequence-id).int, which lists all acceptable internal oligos.</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-firstbaseindex</td>
<td>This parameter is the index of the first base in the input sequence. For input and output using 1-based indexing (such as that used in GenBank and to which many users are accustomed) set this parameter to 1. For input and output using 0-based indexing set this parameter to 0. (This parameter also affects the indexes in the contents of the files produced when the primer file flag is set.)</td>
<td>Any integer value</td>
<td>1</td>
</tr>

<tr>
<td>-pickanyway</td>
<td>If true pick a primer pair even if LEFT-INPUT, RIGHT-INPUT, or INTERNAL-OLIGO-INPUT violates specific constraints.</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-mispriminglibrary</td>
<td>The name of a file containing a nucleotide sequence library of sequences to avoid amplifying (for example repetitive sequences, or possibly the sequences of genes in a gene family that should not be amplified.) The file must be in (a slightly restricted) FASTA format (W. B. Pearson and D.J. Lipman, PNAS 85:8 pp 2444-2448 [1988]); we briefly discuss the organization of this file below. If this parameter is specified then EPrimer3 locally aligns each candidate primer against each library sequence and rejects those primers for which the local alignment score times a specified weight (see below) exceeds MAX-MISPRIMING. (The maximum value of the weight is arbitrarily set to 100.0.)
Each sequence entry in the FASTA-format file must begin with an 'id line' that starts with '>'. The contents of the id line is 'slightly restricted' in that EPrimer3 parses everything after any optional asterisk ('*') as a floating point number to use as the weight mentioned above. If the id line contains no asterisk then the weight defaults to 1.0. The alignment scoring system used is the same as for calculating complementarity among oligos (e.g. SELF-ANY). The remainder of an entry contains the sequence as lines following the id line up until a line starting with '>' or the end of the file. Whitespace and newlines are ignored. Characters 'A', 'T', 'G', 'C', 'a', 't', 'g', 'c' are retained and any other character is converted to 'N' (with the consequence that any IUB / IUPAC codes for ambiguous bases are converted to 'N'). There are no restrictions on line length.
An empty value for this parameter indicates that no repeat library should be used.</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-maxmispriming</td>
<td>The maximum allowed weighted similarity with any sequence in MISPRIMING-LIBRARY.</td>
<td>Number up to 9999.990</td>
<td>12.00</td>
</tr>

<tr>
<td>-pairmaxmispriming</td>
<td>The maximum allowed sum of weighted similarities of a primer pair (one similarity for each primer) with any single sequence in MISPRIMING-LIBRARY.</td>
<td>Number up to 9999.990</td>
<td>24.00</td>
</tr>

<tr>
<td>-numnsaccepted</td>
<td>Maximum number of unknown bases (N) allowable in any primer.</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr>
<td>-selfany</td>
<td>The maximum allowable local alignment score when testing a single primer for (local) self-complementarity and the maximum allowable local alignment score when testing for complementarity between forward and reverse primers. Local self-complementarity is taken to predict the tendency of primers to anneal to each other without necessarily causing self-priming in the PCR. The scoring system gives 1.00 for complementary bases, -0.25 for a match of any base (or N) with an N, -1.00 for a mismatch, and -2.00 for a gap. Only single-base-pair gaps are allowed. For example, the alignment
5' ATCGNA 3'
...|| | |
3' TA-CGT 5'
is allowed (and yields a score of 1.75), but the alignment
5' ATCCGNA 3'
...|| | |
3' TA--CGT 5'
is not considered. Scores are non-negative, and a score of 0.00 indicates that there is no reasonable local alignment between two oligos.</td>
<td>Number from 0.000 to 9999.990</td>
<td>8.00</td>
</tr>

<tr>
<td>-selfend</td>
<td>The maximum allowable 3'-anchored global alignment score when testing a single primer for self-complementarity, and the maximum allowable 3'-anchored global alignment score when testing for complementarity between forward and reverse primers. The 3'-anchored global alignment score is taken to predict the likelihood of PCR-priming primer-dimers, for example
5' ATGCCCTAGCTTCCGGATG 3'
.............||| |||||
..........3' AAGTCCTACATTTAGCCTAGT 5'
or
5' AGGCTATGGGCCTCGCGA 3'
...............||||||
............3' AGCGCTCCGGGTATCGGA 5'
The scoring system is as for the Maximum Complementarity argument. In the examples above the scores are 7.00 and 6.00 respectively. Scores are non-negative, and a score of 0.00 indicates that there is no reasonable 3'-anchored global alignment between two oligos. In order to estimate 3'-anchored global alignments for candidate primers and primer pairs, Primer assumes that the sequence from which to choose primers is presented 5' to 3'. It is nonsensical to provide a larger value for this parameter than for the Maximum (local) Complementarity parameter because the score of a local alignment will always be at least as great as the score of a global alignment.</td>
<td>Number 0.000 or more</td>
<td>3.00</td>
</tr>

<tr>
<td>-maxendstability</td>
<td>The maximum stability for the five 3' bases of a forward or reverse primer. Bigger numbers mean more stable 3' ends. The value is the maximum delta G for duplex disruption for the five 3' bases as calculated using the nearest neighbor parameters published in Breslauer, Frank, Bloeker and Marky, Proc. Natl. Acad. Sci. USA, vol 83, pp 3746-3750. EPrimer3 uses a completely permissive default value for backward compatibility (which we may change in the next release). Rychlik recommends a maximum value of 9 (Wojciech Rychlik, 'Selection of Primers for Polymerase Chain Reaction' in BA White, Ed., 'Methods in Molecular Biology, Vol. 15: PCR Protocols: Current Methods and Applications', 1993, pp 31-40, Humana Press, Totowa NJ).</td>
<td>Number up to 1000.000</td>
<td>9.0</td>
</tr>

<tr>
<td>-oligomishyblibrary</td>
<td>Similar to MISPRIMING-LIBRARY, except that the event we seek to avoid is hybridization of the internal oligo to sequences in this library rather than priming from them.
The file must be in (a slightly restricted) FASTA format (W. B. Pearson and D.J. Lipman, PNAS 85:8 pp 2444-2448 [1988]); we briefly discuss the organization of this file below. If this parameter is specified then EPrimer3 locally aligns each candidate oligo against each library sequence and rejects those primers for which the local alignment score times a specified weight (see below) exceeds INTERNAL-OLIGO-MAX-MISHYB. (The maximum value of the weight is arbitrarily set to 12.0.)
Each sequence entry in the FASTA-format file must begin with an 'id line' that starts with '>'. The contents of the id line is 'slightly restricted' in that EPrimer3 parses everything after any optional asterisk ('*') as a floating point number to use as the weight mentioned above. If the id line contains no asterisk then the weight defaults to 1.0. The alignment scoring system used is the same as for calculating complementarity among oligos (e.g. SELF-ANY). The remainder of an entry contains the sequence as lines following the id line up until a line starting with '>' or the end of the file. Whitespace and newlines are ignored. Characters 'A', 'T', 'G', 'C', 'a', 't', 'g', 'c' are retained and any other character is converted to 'N' (with the consequence that any IUB / IUPAC codes for ambiguous bases are converted to 'N'). There are no restrictions on line length.
An empty value for this parameter indicates that no library should be used.</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-oligomaxmishyb</td>
<td>Similar to MAX-MISPRIMING except that this parameter applies to the similarity of candidate internal oligos to the library specified in INTERNAL-OLIGO-MISHYB-LIBRARY.</td>
<td>Number up to 9999.990</td>
<td>12.0</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

It reads a normal nucleic acid sequence












<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->

An example output file follows:

<p>
<hr>
<pre>
# EPRIMER3 RESULTS FOR HSFAU1

# FORWARD PRIMER STATISTICS:
# considered 18855
# GC content failed 156
# low tm 3616
# high tm 11295
# high any compl 1
# high end compl 1
# long poly-x seq 45
# ok 3741

# REVERSE PRIMER STATISTICS:
# considered 18706
# GC content failed 161
# low tm 3580
# high tm 11241
# long poly-x seq 66
# ok 3658

# PRIMER PAIR STATISTICS:
# considered 188
# unacceptable product size 169
# high end compl 4
# ok 15

#                      Start  Len   Tm     GC%   Sequence

   1 PRODUCT SIZE: 289
     FORWARD PRIMER    1725   20  59.96  55.00  AGGGAAGGGATGCTAGGTGT

     REVERSE PRIMER    1994   20  59.99  55.00  AGAAGCACACCTCTCCCTGA


   2 PRODUCT SIZE: 290
     FORWARD PRIMER    1725   20  59.96  55.00  AGGGAAGGGATGCTAGGTGT

     REVERSE PRIMER    1995   20  59.99  60.00  GAGAAGCACACCTCTCCCTG


   3 PRODUCT SIZE: 288
     FORWARD PRIMER    1725   20  59.96  55.00  AGGGAAGGGATGCTAGGTGT

     REVERSE PRIMER    1993   20  59.99  60.00  GAAGCACACCTCTCCCTGAG


   4 PRODUCT SIZE: 286
     FORWARD PRIMER    1728   20  60.07  55.00  GAAGGGATGCTAGGTGTGGA

     REVERSE PRIMER    1994   20  59.99  55.00  AGAAGCACACCTCTCCCTGA


   5 PRODUCT SIZE: 284
     FORWARD PRIMER    1731   20  60.07  55.00  GGGATGCTAGGTGTGGAAGA

     REVERSE PRIMER    1995   20  59.99  60.00  GAGAAGCACACCTCTCCCTG



</pre>
<hr>
<p>

If the '-explain' flag has been used, as in the example, then statistics
are output describing the number of primers that were considered and
rejected for various reasons. 

<p>

Headers describing the input file name and the names of the output
columns are displayed.

<p>

The best 5 primer pairs are displayed with the product size.

<p>

The reverse sequence is displayed as the reverse complement to the input
forward sense sequence. 

<p>

The 'Start' positions are given counting from the start of the sequence
for both the forward and reverse primer (and for the internal oligos). 

<p>

Note that if you compare the results to the output from the public
Primer3 web site you will see a difference in the reverse primer
positions - this is because the original Primer3 program reports the
reverse primer positions as couted from the 3' end.  The convention in
EMBOSS is to report both forward and reverse featyres as counted from
the 5' end, so the reverse primer positions are given counted from the
5' start of the sequence. 









<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.
	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

   -->

None.













<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

The Whitehead Institute program that is run by this program is available from:
http://www-genome.wi.mit.edu/genome_software/other/primer3.html
<br>
(Then see the link 'Get release 0.9')
<p>

The version that is run by this program is 3.0.9
currently available from:
http://www-genome.wi.mit.edu/ftp/distribution/software/primer3_0_9_test.tar.gz






<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

The message: "EMBOSS An error in eprimer3.c at line 315: The program
'primer3_core' must be on the path.  It is part of the 'primer3'
package, available from the Whitehead Institute.  See:
http://www-genome.wi.mit.edu/" is output if you do not have the
Whitehead Institute's primer3 program set up and on your path. 






<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="primersearch.html">primersearch</a></td><td>Searches DNA sequences for matches with primer pairs</td></tr>
<tr><td><a href="stssearch.html">stssearch</a></td><td>Searches a DNA database for matches with a set of STS primers</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

This application was written by Gary Williams <a
href="mailto:gwilliam@hgmp.mrc.ac.uk">(gwilliam@hgmp.mrc.ac.uk)</a>

<p>

The Whitehead Institute's primer3 program is not part of this program,
but it must be set up on your system and on your path. 

<p>
The Whitehead Institute's primer3 program is:
<br>
 Copyright (c) 1996,1997,1998
<br>
        Whitehead Institute for Biomedical Research. All rights reserved.

<p>

The Whitehead Institute requests that use of this software be cited in
publications as:

<p>

Steve Rozen, Helen J. Skaletsky (1996,1997,1998)
<br>
   Primer3. Code available at
<br>
   http://www-genome.wi.mit.edu/genome_software/other/primer3.html












<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (Dec 2001) - Gary Williams













<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->








</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

