<HTML>

<HEAD>
  <TITLE>
  EMBOSS: equicktandem
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

equicktandem
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>




<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Finds tandem repeats


<H2>
    Description
</H2>


<b>equicktandem</b> scans a sequence for potential tandem repeats up to a
specified size.  The results can be used to run <b>etandem</b> on the
candidate repeat lengths to identify genuine tandem repeats.


<p>

<b>equicktandem</b> is a simple program that looks for segments in which
each base tends to match the base 'repeatsize' back, i.e.  with an
autocorrelation peak at 'repeatsize'.  This can allow drift in the
repeating sequence, i.e.  it does not look for a consensus sequence for
the whole repeat block (that is what <b>etandem</b> does).  But it is
much quicker than etandem.  It does not account for gaps.

<p>

The score is +1 for a match to the corresponding base
'repeatsize' back and -1 for a mismatch.

<H2>
    Usage
</H2>


Here is a sample session with equicktandem. The input sequence is the
human herpesvirus tandem repeat.

<pre>

% equicktandem
Input sequence: embl:hhtetra
Output file [hhtetra.qtan]: 
Maximum repeat size [600]: 
Threshold score [20]: 

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          sequence   Sequence USA
   -maxrepeat          integer    Maximum repeat size
   -threshold          integer    Threshold score
  [-outfile]           report     Output report file name

   Optional qualifiers: (none)
   Advanced qualifiers:
   -origfile           outfile    Output file name

   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-maxrepeat</td>
<td>Maximum repeat size</td>
<td>Any integer value</td>
<td>600</td>
</tr>

<tr>
<td>-threshold</td>
<td>Threshold score</td>
<td>Any integer value</td>
<td>20</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output report file name</td>
<td>Report file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-origfile</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.equicktandem</td>
</tr>

</table>



<H2>
    Input file format
</H2>


The input for equicktandem is a nucleotide sequence

<H2>
    Output file format
</H2>


<p>

The output is a standard EMBOSS report file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-rformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  The available format names are: embl,
genbank, gff, pir, swiss, trace, listfile, dbmotif, diffseq, excel,
feattable, motif, regions, seqtable, simple, srs, table, tagseq

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html</A>
for further information on report formats.

<p>


<p>

By default <b>equicktandem</b> writes a 'table' report file.

<p>

The output from the above example is:

<p>

<pre>
########################################
# Program: equicktandem
# Rundate: Thu Apr 11 13:28:55 2002
# Report_file: stdout
########################################

#=======================================
#
# Sequence: HHTETRA     from: 1   to: 1272
# HitCount: 1
#
# Threshold: 20
# Maxrepeat: 600
#
#=======================================

  Start     End   Score   Size  Count
    191     935     339      6    124

#---------------------------------------
#--------------------------------------- 
</pre>

<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="einverted.html">einverted</a></td><td>Finds DNA inverted repeats</td></tr>
<tr><td><a href="etandem.html">etandem</a></td><td>Looks for tandem repeats in a nucleotide sequence</td></tr>
<tr><td><a href="palindrome.html">palindrome</a></td><td>Looks for inverted repeats in a nucleotide sequence</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->

<p>

<b>equicktandem</b> identifies regions wich are likely to contain tanden
repeats.  <b>tandem</b> should then be run on those regions to confirm
them and to get an accurate specification of the repeats.  <b>tandem</b>
runs slowly. 

<H2>
    Author(s)
</H2>


This program was originally written by Richard Durbin at the Sanger Centre.
<p>

This application was modified for inclusion in EMBOSS by Peter Rice <a
href="mailto:pmr@sanger.ac.uk">(pmr@sanger.ac.uk)</a> Informatics
Division, The Sanger Centre, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.

<H2>
    History
</H2>


Completed 25 May 1999

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

