<HTML>

<HEAD>
  <TITLE>
  EMBOSS: eseqboot
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: eseqboot
</H2>
</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program eseqboot
</H1>
<H2>
    Function
</H2>
Conversion of PHYLIP's seqboot

<H2>
    Description
</H2>         ESEQBOOT -- Bootstrap, Jackknife, or Permutation Resampling of
    Molecular Sequence, Restriction Site, Gene Frequency or Character
    Data<p>
<a href="seqboot.html">Phylip seqboot documentation.</a><p>

<H2>
    Usage
</H2>
Here is a sample session with eseqboot using the data below
and replicates are set to 10 and seed to 4331<p>

<pre>
    5    6
Alpha     AACAAC
Beta      AACCCC
Gamma     ACCAAC
Delta     CCACCA
Epsilon   CCAAAC
</pre>

<pre>

% eseqboot test

Bootstrapped sequences algorithm
Output file [eseqboot.outfile]: stdout
Random number seed (must be odd) [3]: 4331
Method
       Seq : Molecular sequences
       Mor : Discrete Morphology
       Res : Restriction Sites
       Gen : Gene Frequencies
Choose the method [Seq]: 
test
      Boot : Bootstrap
      Jack : Delete-half Jackknife
      Perm : Permute species for each character
Choose test [Boot]: 
How many replicates [100]: 10
    5     6
Alpha        ACCCAC
Beta         ACCCCC
Gamma        CCCCAC
Delta        CAAACA
Epsilon      CAAAAC
    5     6
Alpha        AAAACC
Beta         AACCCC
Gamma        ACAACC
Delta        CCCCAA
Epsilon      CCAACC
    5     6
Alpha        AAAAAC
Beta         AACCCC
Gamma        CCAAAC
Delta        CCCCCA
Epsilon      CCAAAC
    5     6
Alpha        AAAAAA
Beta         AACCCC
Gamma        ACAAAA
Delta        CCCCCC
Epsilon      CCAAAA
    5     6
Alpha        ACCCAA
Beta         ACCCCC
Gamma        CCCCAA
Delta        CAAACC
Epsilon      CAAAAA
    5     6
Alpha        AAACCC
Beta         AAACCC
Gamma        AACCCC
Delta        CCCAAA
Epsilon      CCCACC
    5     6
Alpha        AACAAC
Beta         AACCCC
Gamma        ACCAAC
Delta        CCACCA
Epsilon      CCAAAC
    5     6
Alpha        ACCCAA
Beta         ACCCCC
Gamma        ACCCAA
Delta        CAAACC
Epsilon      CAAAAA
    5     6
Alpha        AACACC
Beta         AACCCC
Gamma        ACCACC
Delta        CCACAA
Epsilon      CCAACC
    5     6
Alpha        AAAACA
Beta         AAAACC
Gamma        AAAACA
Delta        CCCCAC
Epsilon      CCCCAA


</pre>


<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-datafile]          infile     Input file
  [-outfile]           outfile    Output file name
   -randseed           integer    Random number seed (must be odd)
   -method             list       Choose the method
   -test               list       Choose test
   -reps               integer    How many replicates

   Optional qualifiers (* if not always prompted):
   -inter              bool       Interleaved input
*  -enzymes            bool       Present in input file
*  -all                bool       All alleles present at each locus
   -printdata          bool       Print out the data at start of run
   -progress           bool       Print indications of progress of run

   Advanced qualifiers: (none)
</pre>
</td></tr></table>

<p>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-datafile]<br>(Parameter 1)</td>
<td>Input file</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td>eseqboot.outfile</td>
</tr>

<tr>
<td>-randseed</td>
<td>Random number seed (must be odd)</td>
<td>Any integer value</td>
<td>3</td>
</tr>

<tr>
<td>-method</td>
<td>Choose the method</td>
<td><table><tr><td>Seq</td> <td><i>(Molecular sequences)</i></td></tr><tr><td>Mor</td> <td><i>(Discrete Morphology)</i></td></tr><tr><td>Res</td> <td><i>(Restriction Sites)</i></td></tr><tr><td>Gen</td> <td><i>(Gene Frequencies)</i></td></tr></table></td>
<td>Seq</td>
</tr>

<tr>
<td>-test</td>
<td>Choose test</td>
<td><table><tr><td>Boot</td> <td><i>(Bootstrap)</i></td></tr><tr><td>Jack</td> <td><i>(Delete-half Jackknife)</i></td></tr><tr><td>Perm</td> <td><i>(Permute species for each character)</i></td></tr></table></td>
<td>Boot</td>
</tr>

<tr>
<td>-reps</td>
<td>How many replicates</td>
<td>Integer 1 or more</td>
<td>100</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-inter</td>
<td>Interleaved input</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-enzymes</td>
<td>Present in input file</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-all</td>
<td>All alleles present at each locus</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-printdata</td>
<td>Print out the data at start of run</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-progress</td>
<td>Print indications of progress of run</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<H2>
    Data files
</H2>
<H2>
    Notes
</H2>
<H2>
    References
</H2>
<H2>
    Warnings
</H2>
<H2>
    Diagnostics
</H2>
<H2>
    Exit status
</H2>
<H2>
    Known bugs
</H2>
<H2>
    See also
</H2>
<H2>
    Author(s)
</H2>(c)  Copyright  1991-1993  by  the  University  of  Washington  and  by  Joseph
Felsenstein.   Written  by  Joseph  Felsenstein.  Permission is granted to copy
this document provided that no fee is charged for it and  that  this  copyright
notice is not removed.<p>


This application was modified for inclusion in EMBOSS by Ian Longden <a
href="mailto:il@sanger.ac.uk">(il@sanger.ac.uk)</a> Informatics
Division, The Sanger Centre, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.

<H2>
    Priority
</H2>
<H2>
    Target users
</H2>
<H2>
    Comments
</H2>
<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

