<HTML>

<HEAD>
  <TITLE>
  EMBOSS: etandem
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

etandem
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>





<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Looks for tandem repeats in a nucleotide sequence


<H2>
Description
</H2>


<b>etandem</b> looks for tandem repeats in a sequence. It is normally used after
equicktandem has been run to identify potential repeat sizes.
It calculates a consensus for the repeat region and gives a score for
how many matches there are to the consensus - the number of mismatches. 

<p>

Input sequences are converted into ACGT or N (so ambiguity codes are ignored).
<br>
The score is +1 for a match, -1 for a mismatch.
<br>
The first copy of a repeat is ignored.
<br>
The highest score is kept for each start position and repeat size.

<p>

The lowest score to be reported is set by the threshold score.  The
threshold score can be set on the command-line using the
<b>-threshold</b> qualifier, the default is 20.  For perfect repeats, the
score is the length of the repeat (except for the first copy).  Reduce
the threshold score a little if you wish to to allow mismatches.  Each
mismatch scores -1 instead of +1 so it scores 2 less than a perfect
match of the same number of bases. 

<p>

Running with a wide range of repeat sizes is inefficient. That is why
equicktandem was written - to give a rapid estimate of the major repeat
sizes.



<H2>
Usage
</H2>


Here is a sample session with etandem. The input sequence is the
human herpesvirus tandem repeat.

<p>

<pre>
% etandem
Input sequence: embl:hhtetra
Output file [hhtetra.tan]: 
Minimum repeat size [10]: 6
Maximum repeat size [6]: 
</pre>


<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          sequence   Sequence USA
   -minrepeat          integer    Minimum repeat size
   -maxrepeat          integer    Maximum repeat size
  [-outfile]           report     Output report file name

   Optional qualifiers: (none)
   Advanced qualifiers:
   -threshold          integer    Threshold score
   -mismatch           boolean    Allow N as a mismatch
   -uniform            boolean    Allow uniform consensus
   -origfile           outfile    Output file name

   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-minrepeat</td>
<td>Minimum repeat size</td>
<td>Integer, 2 or higher</td>
<td>10</td>
</tr>

<tr>
<td>-maxrepeat</td>
<td>Maximum repeat size</td>
<td>Integer, same as -minrepeat or higher</td>
<td>Same as -minrepeat</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output report file name</td>
<td>Report file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-threshold</td>
<td>Threshold score</td>
<td>Any integer value</td>
<td>20</td>
</tr>

<tr>
<td>-mismatch</td>
<td>Allow N as a mismatch</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-uniform</td>
<td>Allow uniform consensus</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-origfile</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.etandem</td>
</tr>

</table>



<H2>
    Input file format
</H2>


The input for etandem is a nucleotide sequence.

<H2>
    Output file format
</H2>

<p>

The output is a standard EMBOSS report file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-rformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  The available format names are: embl,
genbank, gff, pir, swiss, trace, listfile, dbmotif, diffseq, excel,
feattable, motif, regions, seqtable, simple, srs, table, tagseq

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html</A>
for further information on report formats.

<p>


<p>

By default <b>etandem</b> writes a 'table' report file.

<p>
    
The output from the above example is:

<p>   

<hr>
<pre>
########################################
# Program: etandem
# Rundate: Thu Apr 11 13:31:10 2002
# Report_file: stdout
########################################

#=======================================
#
# Sequence: HHTETRA     from: 1   to: 1272
# HitCount: 5
#
# Threshold: 20
# Minrepeat: 6
# Maxrepeat: 6
# Mismatch: No
# Uniform: No
#
#=======================================

  Start     End   Score   Size  Count Identity Consensus
    793     936     120      6     24     93.8 acccta
    283     420      90      6     23     84.8 taaccc
    432     485      38      6      9     90.7 ccctaa
    494     529      26      6      6     94.4 ccctaa
    568     597      24      6      5    100.0 aaccct

#---------------------------------------
#---------------------------------------
</pre>
<hr>


<H2>
Data files
</H2>


<H2>
Notes
</H2>

Running with a wide range of repeat sizes is inefficient. That is why
equicktandem was written - to give a rapid estimate of the major repeat
sizes.

<H2>
References
</H2>

None.

<H2>
Warnings
</H2>

None.


<H2>
Diagnostics
</H2>

None.


<H2>
Exit status
</H2>

None.

<H2>
Known bugs
</H2>

None.


<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="einverted.html">einverted</a></td><td>Finds DNA inverted repeats</td></tr>
<tr><td><a href="equicktandem.html">equicktandem</a></td><td>Finds tandem repeats</td></tr>
<tr><td><a href="palindrome.html">palindrome</a></td><td>Looks for inverted repeats in a nucleotide sequence</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->

<p>

Running with a wide range of repeat sizes is inefficient. That is why
equicktandem was written - to give a rapid estimate of the major repeat
sizes.


<H2>
Authors
</H2>

This program was originally written by Richard Durbin at the Sanger Centre.
<p>

This application was modified for inclusion in EMBOSS by Peter Rice <a
href="mailto:pmr@sanger.ac.uk">(pmr@sanger.ac.uk)</a> Informatics
Division, The Sanger Centre, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.

<H2>
Priority
</H2>

Completed 25 May 1999

<H2>
Target
</H2>

etandem is aimed at automated repeat identification in genomic sequnece
but can also be used by general users.

<H2>
Comments
</H2>

<pre>
</pre>




</BODY>
</HTML>


