S<HTML>

<HEAD>
  <TITLE>
  EMBOSS: etandem
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

etandem
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>





<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Looks for tandem repeats in a nucleotide sequence


<H2>
Description
</H2>


<b>etandem</b> looks for tandem repeats in a sequence. It is normally used after
equicktandem has been run to identify potential repeat sizes.
It calculates a consensus for the repeat region and gives a score for
how many matches there are to the consensus - the number of mismatches. 

<p>

Input sequences are converted into ACGT or N (so ambiguity codes are ignored).
<br>
The score is +1 for a match, -1 for a mismatch.
<br>
The first copy of a repeat is ignored.
<br>
The highest score is kept for each start position and repeat size.

<p>

The lowest score to be reported is set by the threshold score.  The
threshold score can be set on the command-line using the
<b>-threshold</b> qualifier, the default is 20.  For perfect repeats, the
score is the length of the repeat (except for the first copy).  Reduce
the threshold score a little if you wish to to allow mismatches.  Each
mismatch scores -1 instead of +1 so it scores 2 less than a perfect
match of the same number of bases. 

<p>

Running with a wide range of repeat sizes is inefficient. That is why
equicktandem was written - to give a rapid estimate of the major repeat
sizes.



<H2>
Usage
</H2>


<b>Here is a sample session with etandem</b>
<p>
The input sequence is the human herpesvirus tandem repeat.  
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>etandem -noorigfile </b>
Looks for tandem repeats in a nucleotide sequence
Input sequence: <b>tembl:hhtetra</b>
Minimum repeat size [10]: <b>6</b>
Maximum repeat size [6]: <b></b>
Output report [hhtetra.tan]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>




<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-sequence]          sequence   Sequence USA
   -minrepeat          integer    Minimum repeat size
   -maxrepeat          integer    Maximum repeat size
  [-outfile]           report     Output report file name

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers:
   -threshold          integer    Threshold score
   -mismatch           boolean    Allow N as a mismatch
   -uniform            boolean    Allow uniform consensus
   -origfile           outfile    Output file name

   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1             integer    Start of the sequence to be used
   -send1               integer    End of the sequence to be used
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name

   "-outfile" associated qualifiers
   -rformat2            string     Report format
   -rname2              string     Base file name
   -rextension2         string     File name extension
   -rdirectory2         string     Output directory
   -raccshow2           boolean    Show accession number in the report
   -rdesshow2           boolean    Show description in the report
   -rscoreshow2         boolean    Show the score in the report
   -rusashow2           boolean    Show the full USA in the report

   "-origfile" associated qualifiers
   -odirectory          string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-minrepeat</td>
<td>Minimum repeat size</td>
<td>Integer, 2 or higher</td>
<td>10</td>
</tr>

<tr>
<td>-maxrepeat</td>
<td>Maximum repeat size</td>
<td>Integer, same as -minrepeat or higher</td>
<td>Same as -minrepeat</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output report file name</td>
<td>Report output file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-threshold</td>
<td>Threshold score</td>
<td>Any integer value</td>
<td>20</td>
</tr>

<tr>
<td>-mismatch</td>
<td>Allow N as a mismatch</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-uniform</td>
<td>Allow uniform consensus</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-origfile</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.etandem</td>
</tr>

</table>



<H2>
    Input file format
</H2>


The input for etandem is a nucleotide sequence USA.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:hhtetra' is a sequence entry in the example nucleic acid database 'tembl'
<p>
<p><h3>Database entry: tembl:hhtetra</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   HHTETRA    standard; DNA; VRL; 1272 BP.
XX
AC   L46634; L46689;
XX
SV   L46634.1
XX
DT   06-NOV-1995 (Rel. 45, Created)
DT   04-MAR-2000 (Rel. 63, Last updated, Version 3)
XX
DE   Human herpesvirus 7 (clone ED132'1.2) telomeric repeat region.
XX
KW   telomeric repeat.
XX
OS   Human herpesvirus 7
OC   Viruses; dsDNA viruses, no RNA stage; Herpesviridae; Betaherpesvirinae.
XX
RN   [1]
RP   1-1272
RX   MEDLINE; 96079055.
RA   Secchiero P., Nicholas J., Deng H., Xiaopeng T., van Loon N., Ruvolo V.R.,
RA   Berneman Z.N., Reitz M.S. Jr., Dewhurst S.;
RT   "Identification of human telomeric repeat motifs at the genome termini of
RT   human herpesvirus 7: structural analysis and heterogeneity";
RL   J. Virol. 69(12):8041-8045(1995).
XX
FH   Key             Location/Qualifiers
FH
FT   source          1..1272
FT                   /db_xref="taxon:10372"
FT                   /organism="Human herpesvirus 7"
FT                   /strain="JI"
FT                   /clone="ED132'1.2"
FT   repeat_region   207..928
FT                   /note="long and complex repeat region composed of various
FT                   direct repeats, including TAACCC (TRS), degenerate copies
FT                   of TRS motifs and a 14-bp repeat, TAGGGCTGCGGCCC"
FT   misc_signal     938..998
FT                   /note="pac2 motif"
FT   misc_feature    1009
FT                   /note="right genome terminus (...ACA)"
XX
SQ   Sequence 1272 BP; 346 A; 455 C; 222 G; 249 T; 0 other;
     aagcttaaac tgaggtcaca cacgacttta attacggcaa cgcaacagct gtaagctgca        60
     ggaaagatac gatcgtaagc aaatgtagtc ctacaatcaa gcgaggttgt agacgttacc       120
     tacaatgaac tacacctcta agcataacct gtcgggcaca gtgagacacg cagccgtaaa       180
     ttcaaaactc aacccaaacc gaagtctaag tctcacccta atcgtaacag taaccctaca       240
     actctaatcc tagtccgtaa ccgtaacccc aatcctagcc cttagcccta accctagccc       300
     taaccctagc tctaacctta gctctaactc tgaccctagg cctaacccta agcctaaccc       360
     taaccgtagc tctaagttta accctaaccc taaccctaac catgaccctg accctaaccc       420
     tagggctgcg gccctaaccc tagccctaac cctaacccta atcctaatcc tagccctaac       480
     cctagggctg cggccctaac cctagcccta accctaaccc taaccctagg gctgcggccc       540
     taaccctaac cctagggctg cggcccgaac cctaacccta accctaaccc taaccctagg       600
     gctgcggccc taaccctaac cctagggctg cggccctaac cctaacccta gggctgcggc       660
     ccgaacccta accctaaccc taaccctagg gctgcggccc taaccctaac cctagggctg       720
     cggccctaac cctaacccta actctagggc tgcggcccta accctaaccc taaccctaac       780
     cctagggctg cggcccgaac cctagcccta accctaaccc tgaccctgac cctaacccta       840
     accctaaccc taaccctaac cctaacccta accctaaccc taaccctaac cctaacccta       900
     accctaaccc taaccctaac cctaaccccg cccccactgg cagccaatgt cttgtaatgc       960
     cttcaaggca ctttttctgc gagccgcgcg cagcactcag tgaaaaacaa gtttgtgcac      1020
     gagaaagacg ctgccaaacc gcagctgcag catgaaggct gagtgcacaa ttttggcttt      1080
     agtcccataa aggcgcggct tcccgtagag tagaaaaccg cagcgcggcg cacagagcga      1140
     aggcagcggc tttcagactg tttgccaagc gcagtctgca tcttaccaat gatgatcgca      1200
     agcaagaaaa atgttctttc ttagcatatg cgtggttaat cctgttgtgg tcatcactaa      1260
     gttttcaagc tt                                                          1272
//
</pre>
</td></tr></table><p>


<H2>
    Output file format
</H2>

<p>

The output is a standard EMBOSS report file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-rformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  The available format names are: embl,
genbank, gff, pir, swiss, trace, listfile, dbmotif, diffseq, excel,
feattable, motif, regions, seqtable, simple, srs, table, tagseq

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html</A>
for further information on report formats.

<p>


<p>

By default <b>etandem</b> writes a 'table' report file.

<p>
    

<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: hhtetra.tan</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
########################################
# Program: etandem
# Rundate: Tue Apr 06 14:29:36 2004
# Report_format: table
# Report_file: hhtetra.tan
########################################

#=======================================
#
# Sequence: HHTETRA     from: 1   to: 1272
# HitCount: 5
#
# Threshold: 20
# Minrepeat: 6
# Maxrepeat: 6
# Mismatch: No
# Uniform: No
#
#=======================================

  Start     End   Score   Size  Count Identity Consensus
    793     936     120      6     24     93.8 acccta   
    283     420      90      6     23     84.8 taaccc   
    432     485      38      6      9     90.7 ccctaa   
    494     529      26      6      6     94.4 ccctaa   
    568     597      24      6      5    100.0 aaccct   

#---------------------------------------
#---------------------------------------
</pre>
</td></tr></table><p>




<H2>
Data files
</H2>


<H2>
Notes
</H2>

Running with a wide range of repeat sizes is inefficient. That is why
equicktandem was written - to give a rapid estimate of the major repeat
sizes.

<H2>
References
</H2>

None.

<H2>
Warnings
</H2>

None.


<H2>
Diagnostics
</H2>

None.


<H2>
Exit status
</H2>

None.

<H2>
Known bugs
</H2>

None.


<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="einverted.html">einverted</a></td><td>Finds DNA inverted repeats</td></tr>
<tr><td><a href="equicktandem.html">equicktandem</a></td><td>Finds tandem repeats</td></tr>
<tr><td><a href="palindrome.html">palindrome</a></td><td>Looks for inverted repeats in a nucleotide sequence</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->

<p>

Running with a wide range of repeat sizes is inefficient. That is why
equicktandem was written - to give a rapid estimate of the major repeat
sizes.


<H2>
Authors
</H2>

This program was originally written by 
Richard Durbin (rd&nbsp;&copy;&nbsp;sanger.ac.uk)
<br>
Sanger Institute, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.                      


<p>

This application was modified for inclusion in EMBOSS by 
Peter Rice (pmr&nbsp;&copy;&nbsp;ebi.ac.uk)
<br>
Informatics Division, European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK



<H2>
History
</H2>

Completed 25 May 1999

<!--
<p>
[an error occurred while processing this directive]
-->

<H2>
Target
</H2>

etandem is aimed at automated repeat identification in genomic sequnece
but can also be used by general users.

<H2>
Comments
</H2>

<pre>
</pre>




</BODY>
</HTML>


