<HTML>

<HEAD>
  <TITLE>
  EMBOSS: extractseq
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

extractseq
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>





<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Extract regions from a sequence


<H2>
    Description
</H2>


<b>extractseq</b> allows you to specify one or more regions of a sequence to
extract sub-sequences from to build up a contiguous resulting sequence. 

<p>

This is modelled on the cell's process of splicing out exons from
mRNA, but the program is generally applicable to any cutting and
splicing or editing operation on a single sequence. 

<p>

<b>extractseq</b> reads in a sequence and a set of regions of that
sequence as specified by pairs of start and end positions (either on the
command-line or contained in a file) and writes out the specified
regions of the input sequence in the order in which they have been
specified.  Thus, if the sequence "AAAGGGTTT" has been input and the
regions: "7-9, 3-4" have been specified, then the output sequence will
be: "TTTAG". 

<H2>
    Usage
</H2>


Extract the region from position 10 to 20.
<pre>
% extractseq main.seq result.seq -regions '10-20'
</pre>

<p>
Extract the regions 10 to 20, 30 to 45, 533 to 537 
<pre>
% extractseq main.seq result2.seq -regions '10-20, 30-45, 533-537'
</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          sequence   Sequence USA
   -regions            range      Regions to extract.
                                  A set of regions is specified by a set of
                                  pairs of positions.
                                  The positions are integers.
                                  They are separated by any non-digit,
                                  non-alpha character.
                                  Examples of region specifications are:
                                  24-45, 56-78
                                  1:45, 67=99;765..888
                                  1,5,8,10,23,45,57,99
  [-outseq]            seqoutall  Output sequence(s) USA

   Optional qualifiers:
   -separate           boolean    If this is set true then each specified
                                  region is written out as a separate
                                  sequence. The name of the sequence is
                                  created from the name of the original
                                  sequence with the start and end positions of
                                  the range appended with underscore
                                  characters between them, eg: XYZ region 2 to
                                  34 is written as: XYZ_2_34

   Advanced qualifiers: (none)
   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-regions</td>
<td>Regions to extract.
A set of regions is specified by a set of pairs of positions.
The positions are integers.
They are separated by any non-digit, non-alpha character.
Examples of region specifications are:
24-45, 56-78
1:45, 67=99;765..888
1,5,8,10,23,45,57,99</td>
<td>Sequence range</td>
<td>Whole sequence</td>
</tr>

<tr>
<td>[-outseq]<br>(Parameter 2)</td>
<td>Output sequence(s) USA</td>
<td>Writeable sequence(s)</td>
<td><i>&lt;sequence&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-separate</td>
<td>If this is set true then each specified region is written out as a separate sequence. The name of the sequence is created from the name of the original sequence with the start and end positions of the range appended with underscore characters between them, eg: XYZ region 2 to 34 is written as: XYZ_2_34</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>


Normal sequence.

<p>
You can specifiy a file of ranges to extract by giving the '-regions'
qualifier the value '@' followed by the name of the file containing the
ranges. (eg: '-regions @myfile').

<p>
The format of the range file is:

<ul>
<li>Comment lines start with '#' in the first column.
<li>Comment lines and blank lines are ignored.
<li>The line may start with white-space.
<li>There are two positive (integer) numbers per line separated by one or
        more space or TAB characters. 
<li>The second number must be greater or equal to the first number.
<li>There can be optional text after the two numbers to annotate the line. 
<li>White-space before or after the text is removed.
</ul>

<p>
An example range file is:

<pre>

# this is my set of ranges
12   23
 4   5       this is like 12-23, but smaller
67   10348   interesting region

</pre>

<H2>
    Output file format
</H2>


The output is a normal sequence file.

<p>

For example, the coding regions of em:hsfau1 are joined as:

<p>
<pre>
% extractseq em:hsfau1 -reg "782..856,951..1095,1557..1612,1787..1912" stdout

>HSFAU X65923 H.sapiens fau mRNA
atgcagctctttgtccgcgcccaggagctacacaccttcgaggtgaccggccaggaaacg
gtcgcccagatcaaggctcatgtagcctcactggagggcattgccccggaagatcaagtc
gtgctcctggcaggcgcgcccctggaggatgaggccactctgggccagtgcggggtggag
gccctgactaccctggaagtagcaggccgcatgcttggaggtaaagttcatggttccctg
gcccgtgctggaaaagtgagaggtcagactcctaaggtggccaaacaggagaagaagaag
aagaagacaggtcgggctaagcggcggatgcagtacaaccggcgctttgtcaacgttgtg
cccacctttggcaagaagaagggccccaatgccaactcttaa
</pre>

<p>

If the option '-separate' is used then each specified region is written
to the output file as a separate sequence.  The name of the sequence is
created from the name of the original sequence with the start and end
positions of the range appended with underscore characters between them,

<p>

For example: "XYZ region 2 to 34" is written as: "XYZ_2_34"

<p>

To output each of the exons in em:hsfau1 to a separate entry:

<p>

<pre>
% extractseq em:hsfau1 -reg "782..856,951..1095,1557..1612,1787..1912" stdout -separate

>HSFAU1_782_856 H.sapiens fau 1 gene
atgcagctctttgtccgcgcccaggagctacacaccttcgaggtgaccggccaggaaacg
gtcgcccagatcaag
>HSFAU1_951_1095 H.sapiens fau 1 gene
gctcatgtagcctcactggagggcattgccccggaagatcaagtcgtgctcctggcaggc
gcgcccctggaggatgaggccactctgggccagtgcggggtggaggccctgactaccctg
gaagtagcaggccgcatgcttggag
>HSFAU1_1557_1612 H.sapiens fau 1 gene
gtaaagtccatggttccctggcccgtgctggaaaagtgagaggtcagactcctaag
>HSFAU1_1787_1912 H.sapiens fau 1 gene
gtggccaaacaggagaagaagaagaagaagacaggtcgggctaagcggcggatgcagtac
aaccggcgctttgtcaacgttgtgcccacctttggcaagaagaagggccccaatgccaac
tcttaa
</pre>



<H2>
    Data files
</H2>


None.
	
<H2>
    Notes
</H2>


None.
	
<H2>
    References
</H2>


None.
	
<H2>
    Warnings
</H2>


None.
	
<H2>
    Diagnostic Error Messages
</H2>


Several warning messages about malformed region specifications: 

<ul>
<li>       Non-digit found in region ... 
<li>       Unpaired start of a region found in ... 
<li>       Non-digit found in region ... 
<li>       The start of a pair of region positions must be smaller than the
	end in ... 
</ul>
	
<H2>
    Exit status
</H2>


It exits with status 0, unless a region is badly constructed. 

<H2>
    Known bugs
</H2>


None noted.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="biosed.html">biosed</a></td><td>Replace or delete sequence sections</td></tr>
<tr><td><a href="cutseq.html">cutseq</a></td><td>Removes a specified section from a sequence</td></tr>
<tr><td><a href="degapseq.html">degapseq</a></td><td>Removes gap characters from sequences</td></tr>
<tr><td><a href="descseq.html">descseq</a></td><td>Alter the name or description of a sequence</td></tr>
<tr><td><a href="entret.html">entret</a></td><td>Reads and writes (returns) flatfile entries</td></tr>
<tr><td><a href="extractfeat.html">extractfeat</a></td><td>Extract features from a sequence</td></tr>
<tr><td><a href="listor.html">listor</a></td><td>Writes a list file of the logical OR of two sets of sequences</td></tr>
<tr><td><a href="maskfeat.html">maskfeat</a></td><td>Mask off features of a sequence</td></tr>
<tr><td><a href="maskseq.html">maskseq</a></td><td>Mask off regions of a sequence</td></tr>
<tr><td><a href="newseq.html">newseq</a></td><td>Type in a short new sequence</td></tr>
<tr><td><a href="noreturn.html">noreturn</a></td><td>Removes carriage return from ASCII files</td></tr>
<tr><td><a href="notseq.html">notseq</a></td><td>Excludes a set of sequences and writes out the remaining ones</td></tr>
<tr><td><a href="nthseq.html">nthseq</a></td><td>Writes one sequence from a multiple set of sequences</td></tr>
<tr><td><a href="pasteseq.html">pasteseq</a></td><td>Insert one sequence into another</td></tr>
<tr><td><a href="revseq.html">revseq</a></td><td>Reverse and complement a sequence</td></tr>
<tr><td><a href="seqret.html">seqret</a></td><td>Reads and writes (returns) sequences</td></tr>
<tr><td><a href="seqretsplit.html">seqretsplit</a></td><td>Reads and writes (returns) sequences in individual files</td></tr>
<tr><td><a href="skipseq.html">skipseq</a></td><td>Reads and writes (returns) sequences, skipping the first few</td></tr>
<tr><td><a href="splitter.html">splitter</a></td><td>Split a sequence into (overlapping) smaller sequences</td></tr>
<tr><td><a href="swissparse.html">swissparse</a></td><td>Retrieves sequences from swissprot using keyword search</td></tr>
<tr><td><a href="trimest.html">trimest</a></td><td>Trim poly-A tails off EST sequences</td></tr>
<tr><td><a href="trimseq.html">trimseq</a></td><td>Trim ambiguous bits off the ends of sequences</td></tr>
<tr><td><a href="union.html">union</a></td><td>Reads sequence fragments and builds one sequence</td></tr>
<tr><td><a href="vectorstrip.html">vectorstrip</a></td><td>Strips out DNA between a pair of vector sequences</td></tr>
<tr><td><a href="yank.html">yank</a></td><td>Reads a sequence range, appends the full USA to a list file</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->




<H2>
    Author(s)
</H2>


This application was written by Gary Williams <a
href="mailto:gwilliam@hgmp.mrc.ac.uk">(gwilliam@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>


Written (2000) - Gary Williams

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

