<HTML>
<HEAD>
  <TITLE>
  EMBOSS: fdiscboot
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

fdiscboot
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<H2>
    Function
</H2>
Bootstrapped discrete sites algorithm

<H2>
    Description
</H2>

Reads in a data set, and produces multiple data sets from it by
bootstrap resampling. Since most programs in the current version of
the package allow processing of multiple data sets, this can be used
together with the consensus tree program CONSENSE to do bootstrap (or
delete-half-jackknife) analyses with most of the methods in this
package. This program also allows the Archie/Faith technique of
permutation of species within characters. It can also rewrite a data
set to convert it from between the PHYLIP Interleaved and Sequential
forms, and into a preliminary version of a new XML sequence alignment
format which is under development


<H2>
    Algorithm
</H2>

SEQBOOT is a general bootstrapping and data set translation tool. It
is intended to allow you to generate multiple data sets that are
resampled versions of the input data set. Since almost all programs in
the package can analyze these multiple data sets, this allows almost
anything in this package to be bootstrapped, jackknifed, or
permuted. SEQBOOT can handle molecular sequences, binary characters,
restriction sites, or gene frequencies. It can also convert data sets
between Sequential and Interleaved format, and into the NEXUS format
or into a new XML sequence alignment format.

<p>
To carry out a bootstrap (or jackknife, or permutation test) with some
method in the package, you may need to use three programs. First, you
need to run SEQBOOT to take the original data set and produce a large
number of bootstrapped or jackknifed data sets (somewhere between 100
and 1000 is usually adequate). Then you need to find the phylogeny
estimate for each of these, using the particular method of
interest. For example, if you were using DNAPARS you would first run
SEQBOOT and make a file with 100 bootstrapped data sets. Then you
would give this file the proper name to have it be the input file for
DNAPARS. Running DNAPARS with the M (Multiple Data Sets) menu choice
and informing it to expect 100 data sets, you would generate a big
output file as well as a treefile with the trees from the 100 data
sets. This treefile could be renamed so that it would serve as the
input for CONSENSE. When CONSENSE is run the majority rule consensus
tree will result, showing the outcome of the analysis.

<p>
This may sound tedious, but the run of CONSENSE is fast, and that of
SEQBOOT is fairly fast, so that it will not actually take any longer
than a run of a single bootstrap program with the same original data
and the same number of replicates. This is not very hard and allows
bootstrapping or jackknifing on many of the methods in this
package. The same steps are necessary with all of them. Doing things
this way some of the intermediate files (the tree file from the
DNAPARS run, for example) can be used to summarize the results of the
bootstrap in other ways than the majority rule consensus method does.

<p>
If you are using the Distance Matrix programs, you will have to add
one extra step to this, calculating distance matrices from each of the
replicate data sets, using DNADIST or GENDIST. So (for example) you
would run SEQBOOT, then run DNADIST using the output of SEQBOOT as its
input, then run (say) NEIGHBOR using the output of DNADIST as its
input, and then run CONSENSE using the tree file from NEIGHBOR as its
input.

<p>
The resampling methods available are: 

<ul>

<li>
The bootstrap. Bootstrapping was invented by Bradley Efron in 1979,
and its use in phylogeny estimation was introduced by me (Felsenstein,
1985b; see also Penny and Hendy, 1985). It involves creating a new
data set by sampling N characters randomly with replacement, so that
the resulting data set has the same size as the original, but some
characters have been left out and others are duplicated. The random
variation of the results from analyzing these bootstrapped data sets
can be shown statistically to be typical of the variation that you
would get from collecting new data sets. The method assumes that the
characters evolve independently, an assumption that may not be
realistic for many kinds of data.

<li>
The partial bootstrap.. This is the bootstrap where fewer than the
full number of characters are sampled. The user is asked for the
fraction of characters to be sampled. It is primarily useful in
carrying out Zharkikh and Li's (1995) Complete And Partial Bootstrap
method, and Shimodaira's (2002) AU method, both of which correct the
bias of bootstrap P values.

<li>
Block-bootstrapping. One pattern of departure from indeopendence of
character evolution is correlation of evolution in adjacent
characters. When this is thought to have occurred, we can correct for
it by samopling, not individual characters, but blocks of adjacent
characters. This is called a block bootstrap and was introduced by
Künsch (1989). If the correlations are believed to extend over some
number of characters, you choose a block size, B, that is larger than
this, and choose N/B blocks of size B. In its implementation here the
block bootstrap "wraps around" at the end of the characters (so that
if a block starts in the last B-1 characters, it continues by wrapping
around to the first character after it reaches the last
character). Note also that if you have a DNA sequence data set of an
exon of a coding region, you can ensure that equal numbers of first,
second, and third coding positions are sampled by using the block
bootstrap with B = 3.

<li>
Partial block-bootstrapping. Similar to partial bootstrapping except
sampling blocks rather than single characters.

<li>
Delete-half-jackknifing.. This alternative to the bootstrap involves
sampling a random half of the characters, and including them in the
data but dropping the others. The resulting data sets are half the
size of the original, and no characters are duplicated. The random
variation from doing this should be very similar to that obtained from
the bootstrap. The method is advocated by Wu (1986). It was mentioned
by me in my bootstrapping paper (Felsenstein, 1985b), and has been
available for many years in this program as an option. Note that, for
the present, block-jackknifing is not available, because I cannot
figure out how to do it straightforwardly when the block size is not a
divisor of the number of characters.

<li>
Delete-fraction jackknifing. Jackknifing is advocated by Farris
et. al. (1996) but as deleting a fraction 1/e (1/2.71828). This
retains too many characters and will lead to overconfidence in the
resulting groups when there are conflicting characters. However it is
made available here as an option, with the user asked to supply the
fraction of characters that are to be retained.

<li>
Permuting species within characters. This method of resampling (well,
OK, it may not be best to call it resampling) was introduced by Archie
(1989) and Faith (1990; see also Faith and Cranston, 1991). It
involves permuting the columns of the data matrix separately. This
produces data matrices that have the same number and kinds of
characters but no taxonomic structure. It is used for different
purposes than the bootstrap, as it tests not the variation around an
estimated tree but the hypothesis that there is no taxonomic structure
in the data: if a statistic such as number of steps is significantly
smaller in the actual data than it is in replicates that are permuted,
then we can argue that there is some taxonomic structure in the data
(though perhaps it might be just the presence of aa pair of sibling
species).

<li>
Permuting characters. This simply permutes the order of the
characters, the same reordering being applied to all species. For many
methods of tree inference this will make no difference to the outcome
(unless one has rates of evolution correlated among adjacent
sites). It is included as a possible step in carrying out a
permutation test of homogeneity of characters (such as the
Incongruence Length Difference test).

<li>
Permuting characters separately for each species. This is a method
introduced by Steel, Lockhart, and Penny (1993) to permute data so as
to destroy all phylogenetic structure, while keeping the base
composition of each species the same as before. It shuffles the
character order separately for each species.

<li>
Rewriting. This is not a resampling or permutation method: it simply
rewrites the data set into a different format. That format can be the
PHYLIP format. For molecular sequences and discrete morphological
character it can also be the NEXUS format. For molecular sequences one
other format is available, a new (and nonstandard) XML format of our
own devising. When the PHYLIP format is chosen the data set is
coverted between Interleaved and Sequential format. If it was read in
as Interleaved sequences, it will be written out as Sequential format,
and vice versa. The NEXUS format for molecular sequences is always
written as interleaved sequences. The XML format is different from
(though similar to) a number of other XML sequence alignment
formats. An example will be found below. Here is a table to links to
those other XML alignment formats:

<table>
<tr><td>
Andrew Rambaut's
BEAST XML format  </td>
<td>http://evolve.zoo.ox.ac.uk/beast/introXML.html
and http://evolve.zoo.ox.ac.uk/beast/referenindex.html</td>
<td>A format for alignments. There
is also a format for phylogenies
described there.</td></tr>

<tr> 
<td>MSAML  M</td>
<td>http://xml.coverpages.org/msaml-desc-dec.html </td>
<td> Defined by Paul Gordon of
University of Calgary. See his
big list of molecular biology
XML projects.</td></tr>
<tr>
<td>BSML </td>
<td> http://www.bsml.org/resources/default.asp  </td>
<td>Bioinformatic Sequence
Markup Language
includes a multiple sequence
alignment XML format  </td></tr>
</table>

</ul>


<H2>
    Usage
</H2>




<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-infile]            discretestates (no help text) discretestates value
  [-outfile]           outfile    Output file name
  [-outancfile]        outfile    Out ancestor file name
  [-outmixfile]        outfile    Out mix file name
  [-outfactfile]       outfile    Out fact file name

   Additional (Optional) qualifiers (* if not always prompted):
   -mixfile            properties File of mixtures
   -ancfile            properties File of ancestors
   -weights            properties Weights file
   -factorfile         properties Factors file
   -test               menu       Choose test
*  -regular            toggle     Altered sampling fraction
*  -fracsample         float      Samples as percentage of sites
*  -morphseqtype       menu       Output format
*  -blocksize          integer    Block size for bootstraping
*  -reps               integer    How many replicates
*  -justweights        menu       Write out datasets or just weights
*  -seed               integer    Random number seed between 1 and 32767 (must
                                  be odd)
   -printdata          boolean    Print out the data at start of run
*  -[no]dotdiff        boolean    Use dot-differencing
   -[no]progress       boolean    Print indications of progress of run

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-outfile" associated qualifiers
   -odirectory2        string     Output directory

   "-outancfile" associated qualifiers
   -odirectory3        string     Output directory

   "-outmixfile" associated qualifiers
   -odirectory4        string     Output directory

   "-outfactfile" associated qualifiers
   -odirectory5        string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report deaths


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-infile]<br>(Parameter 1)</td>
<td>(no help text) discretestates value</td>
<td>Discrete states file</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.fdiscboot</td>
</tr>

<tr>
<td>[-outancfile]<br>(Parameter 3)</td>
<td>Out ancestor file name</td>
<td>Output file</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>[-outmixfile]<br>(Parameter 4)</td>
<td>Out mix file name</td>
<td>Output file</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>[-outfactfile]<br>(Parameter 5)</td>
<td>Out fact file name</td>
<td>Output file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-mixfile</td>
<td>File of mixtures</td>
<td>Property value(s)</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>-ancfile</td>
<td>File of ancestors</td>
<td>Property value(s)</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>-weights</td>
<td>Weights file</td>
<td>Property value(s)</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>-factorfile</td>
<td>Factors file</td>
<td>Property value(s)</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>-test</td>
<td>Choose test</td>
<td><table><tr><td>b</td> <td><i>(Bootstrap)</i></td></tr><tr><td>j</td> <td><i>(Jackknife)</i></td></tr><tr><td>c</td> <td><i>(Permute species for each character)</i></td></tr><tr><td>o</td> <td><i>(Permute character order)</i></td></tr><tr><td>s</td> <td><i>(Permute within species)</i></td></tr><tr><td>r</td> <td><i>(Rewrite data)</i></td></tr></table></td>
<td>b</td>
</tr>

<tr>
<td>-regular</td>
<td>Altered sampling fraction</td>
<td>Toggle value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-fracsample</td>
<td>Samples as percentage of sites</td>
<td>Number from 0.100 to 100.000</td>
<td>100.0</td>
</tr>

<tr>
<td>-morphseqtype</td>
<td>Output format</td>
<td><table><tr><td>p</td> <td><i>(PHYLIP)</i></td></tr><tr><td>n</td> <td><i>(NEXUS)</i></td></tr></table></td>
<td>p</td>
</tr>

<tr>
<td>-blocksize</td>
<td>Block size for bootstraping</td>
<td>Integer 1 or more</td>
<td>1</td>
</tr>

<tr>
<td>-reps</td>
<td>How many replicates</td>
<td>Integer 1 or more</td>
<td>100</td>
</tr>

<tr>
<td>-justweights</td>
<td>Write out datasets or just weights</td>
<td><table><tr><td>d</td> <td><i>(Datasets)</i></td></tr><tr><td>w</td> <td><i>(Weights)</i></td></tr></table></td>
<td>d</td>
</tr>

<tr>
<td>-seed</td>
<td>Random number seed between 1 and 32767 (must be odd)</td>
<td>Integer from 1 to 32767</td>
<td>1</td>
</tr>

<tr>
<td>-printdata</td>
<td>Print out the data at start of run</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]dotdiff</td>
<td>Use dot-differencing</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]progress</td>
<td>Print indications of progress of run</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<H2>
    Input file format
</H2>

<b>fdiscboot</b> reads discrete character data

<p>







<H2>
    Output file format
</H2>

<b>fdiscboot</b> 
writes a bootstrap multiple set of discrete character data

<p>




<H2>
    Data files
</H2>

None

<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.
<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with status 0.

<H2>
    Known bugs
</H2>

None.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="../phylip/ednacomp.html">ednacomp</a></td><td>DNA compatibility algorithm</td></tr>
<tr><td><a href="../phylip/ednadist.html">ednadist</a></td><td>Nucleic acid sequence Distance Matrix program</td></tr>
<tr><td><a href="../phylip/ednainvar.html">ednainvar</a></td><td>Nucleic acid sequence Invariants method</td></tr>
<tr><td><a href="../phylip/ednaml.html">ednaml</a></td><td>Estimates phylogenies from nucleic acid sequence Maximum Likelihood</td></tr>
<tr><td><a href="../phylip/ednamlk.html">ednamlk</a></td><td>Estimates phylogenies from nucleic acid sequence Maximum Likelihood with molecular clock</td></tr>
<tr><td><a href="../phylip/ednapars.html">ednapars</a></td><td>DNA parsimony algorithm</td></tr>
<tr><td><a href="../phylip/ednapenny.html">ednapenny</a></td><td>Penny algorithm for DNA</td></tr>
<tr><td><a href="../phylip/eprotdist.html">eprotdist</a></td><td>Protein distance algorithm</td></tr>
<tr><td><a href="../phylip/eprotpars.html">eprotpars</a></td><td>Protein parsimony algorithm</td></tr>
<tr><td><a href="../phylip/erestml.html">erestml</a></td><td>Restriction site Maximum Likelihood method</td></tr>
<tr><td><a href="../phylip/eseqboot.html">eseqboot</a></td><td>Bootstrapped sequences algorithm</td></tr>
<tr><td><a href="fdnacomp.html">fdnacomp</a></td><td>DNA compatibility algorithm</td></tr>
<tr><td><a href="fdnadist.html">fdnadist</a></td><td>Nucleic acid sequence Distance Matrix program</td></tr>
<tr><td><a href="fdnainvar.html">fdnainvar</a></td><td>Nucleic acid sequence Invariants method</td></tr>
<tr><td><a href="fdnaml.html">fdnaml</a></td><td>Estimates nucleotide phylogeny by maximum likelihood</td></tr>
<tr><td><a href="fdnamlk.html">fdnamlk</a></td><td>Estimates nucleotide phylogeny by maximum likelihood</td></tr>
<tr><td><a href="fdnamove.html">fdnamove</a></td><td>Interactive DNA parsimony</td></tr>
<tr><td><a href="fdnapars.html">fdnapars</a></td><td>DNA parsimony algorithm</td></tr>
<tr><td><a href="fdnapenny.html">fdnapenny</a></td><td>Penny algorithm for DNA</td></tr>
<tr><td><a href="fdolmove.html">fdolmove</a></td><td>Interactive Dollo or Polymorphism Parsimony</td></tr>
<tr><td><a href="ffreqboot.html">ffreqboot</a></td><td>Bootstrapped genetic frequencies algorithm</td></tr>
<tr><td><a href="fproml.html">fproml</a></td><td>Protein phylogeny by maximum likelihood</td></tr>
<tr><td><a href="fpromlk.html">fpromlk</a></td><td>Protein phylogeny by maximum likelihood</td></tr>
<tr><td><a href="fprotdist.html">fprotdist</a></td><td>Protein distance algorithm</td></tr>
<tr><td><a href="fprotpars.html">fprotpars</a></td><td>Protein pasimony algorithm</td></tr>
<tr><td><a href="frestboot.html">frestboot</a></td><td>Bootstrapped restriction sites algorithm</td></tr>
<tr><td><a href="frestdist.html">frestdist</a></td><td>Computes distance matrix from restriction sites or fragments</td></tr>
<tr><td><a href="frestml.html">frestml</a></td><td>Restriction site maximum Likelihood method</td></tr>
<tr><td><a href="fseqboot.html">fseqboot</a></td><td>Bootstrapped sequences algorithm</td></tr>
<tr><td><a href="fseqbootall.html">fseqbootall</a></td><td>Bootstrapped sequences algorithm</td></tr>
</table>


<H2>
    Author(s)
</H2>

This program is an EMBOSS conversion of a program written by Joe
Felsenstein as part of his PHYLIP package.

<p>
Although we take every care to ensure that the results of the EMBOSS
version are identical to those from the original package, we recommend
that you check your inputs give the same results in both versions
before publication.

<p>
Please report all bugs in the EMBOSS version to the EMBOSS bug team,
not to the original author.


<H2>
    History
</H2>






<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.



<H2>
    Comments
</H2>

None


</BODY>
</HTML>

