<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: fdnacomp
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

fdnacomp
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<!--END OF HEADER-->






<H2>
    Function
</H2>
DNA compatibility algorithm

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

Estimates phylogenies from nucleic acid sequence data using the
compatibility criterion, which searches for the largest number of
sites which could have all states (nucleotides) uniquely evolved on
the same tree. Compatibility is particularly appropriate when sites
vary greatly in their rates of evolution, but we do not know in
advance which are the less reliable ones.



<H2>
    Algorithm
</H2>


<p>
This program implements the compatibility method for DNA sequence
data. For a four-state character without a character-state tree, as in
DNA sequences, the usual clique theorems cannot be applied. The
approach taken in this program is to directly evaluate each tree
topology by counting how many substitutions are needed in each site,
comparing this to the minimum number that might be needed (one less
than the number of bases observed at that site), and then evaluating
the number of sites which achieve the minimum number. This is the
evaluation of the tree (the number of compatible sites), and the
topology is chosen so as to maximize that number.

<p>
Compatibility methods originated with Le Quesne's (1969) suggestion
that one ought to look for trees supported by the largest number of
perfectly fitting (compatible) characters. Fitch (1975) showed by
counterexample that one could not use the pairwise compatibility
methods used in CLIQUE to discover the largest clique of jointly
compatible characters.

<p>
The assumptions of this method are similar to those of CLIQUE. In a
paper in the Biological Journal of the Linnean Society (1981b) I
discuss this matter extensively. In effect, the assumptions are that:

<ol>

<li>
Each character evolves independently. 

<li>
Different lineages evolve independently. 

<li>
The ancestral base at each site is unknown. 

<li>
The rates of change in most sites over the time spans involved in the the divergence of the group are very small. 

<li>
A few of the sites have very high rates of change. 

<li>
We do not know in advance which are the high and which the low rate sites. 

</ol>

<p>
That these are the assumptions of compatibility methods has been
documented in a series of papers of mine: (1973a, 1978b, 1979, 1981b,
1983b, 1988b). For an opposing view arguing that arguments such as
mine are invalid and that parsimony (and perhaps compatibility)
methods make no substantive assumptions such as these, see the papers
by Farris (1983) and Sober (1983a, 1983b, 1988), but also read the
exchange between Felsenstein and Sober (1986).

<p>
There is, however, some reason to believe that the present criterion
is not the proper way to correct for the presence of some sites with
high rates of change in nucleotide sequence data. It can be argued
that sites showing more than two nucleotide states, even if those are
compatible with the other sites, are also candidates for sites with
high rates of change. It might then be more proper to use DNAPARS with
the Threshold option with a threshold value of 2.

<p>
Change from an occupied site to a gap is counted as one
change. Reversion from a gap to an occupied site is allowed and is
also counted as one change. Note that this in effect assumes that a
gap N bases long is N separate events. This may be an
overcorrection. When we have nonoverlapping gaps, we could instead
code a gap as a single event by changing all but the first "-" in the
gap into "?" characters. In this way only the first base of the gap
causes the program to infer a change.

<p>
If the U (User Tree) option is used and more than one tree is
supplied, the program also performs a statistical test of each of
these trees against the one with highest likelihood. If there are two
user trees, the test done is one which is due to Kishino and Hasegawa
(1989), a version of a test originally introduced by Templeton
(1983). In this implementation it uses the mean and variance of
weighted compatibility differences between trees, taken across
sites. If the two trees compatibilities are more than 1.96 standard
deviations different then the trees are declared significantly
different.

<p>
If there are more than two trees, the test done is an extension of the
KHT test, due to Shimodaira and Hasegawa (1999). They pointed out that
a correction for the number of trees was necessary, and they
introduced a resampling method to make this correction. In the version
used here the variances and covariances of the sum of weighted
compatibilities of sites are computed for all pairs of trees. To test
whether the difference between each tree and the best one is larger
than could have been expected if they all had the same expected
compatibility, compatibilities for all trees are sampled with these
covariances and equal means (Shimodaira and Hasegawa's "least
favorable hypothesis"), and a P value is computed from the fraction of
times the difference between the tree's value and the highest
compatibility exceeds that actually observed. Note that this sampling
needs random numbers, and so the program will prompt the user for a
random number seed if one has not already been supplied. With the
two-tree KHT test no random numbers are used.

<p>
In either the KHT or the SH test the program prints out a table of the
compatibility of each tree, the differences of each from the highest
one, the variance of that quantity as determined by the compatibility
differences at individual sites, and a conclusion as to whether that
tree is or is not significantly worse than the best one.

<p>
The algorithm is a straightforward modification of DNAPARS, but with
some extra machinery added to calculate, as each species is added, how
many base changes are the minimum which could be required at that
site. The program runs fairly quickly.

<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

<b>Here is a sample session with fdnacomp</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>fdnacomp -ancseq -stepbox -printdata </b>
DNA compatibility algorithm
Input sequence: <b>dnacomp.dat</b>
Input tree file: <b></b>
Weights file: <b></b>
Output file [dnacomp.fdnacomp]: <b></b>

Adding species:
   1. Alpha     
   2. Beta      
   3. Gamma     
   4. Delta     
   5. Epsilon   

Doing global rearrangements
  !---------!
   .........

Output written to file "dnacomp.fdnacomp"

Trees also written onto file "dnacomp.treefile"


</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>
<p>
<b>Example 2</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>fdnacomp </b>
DNA compatibility algorithm
Input sequence: <b>dnacomp.dat</b>
Input tree file: <b>dnacomptree.dat</b>
Weights file: <b></b>
Output file [dnacomp.fdnacomp]: <b></b>

Output written to file "dnacomp.fdnacomp"

Trees also written onto file "dnacomp.treefile"


</pre></td></tr></table><p>
<p>
<a href="#input.2">Go to the input files for this example</a><br><a href="#output.2">Go to the output files for this example</a><p><p>




<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-sequence]          seqsetall  File containing one or more sequence
                                  alignments
  [-intreefile]        tree       (no help text) tree value
   -weights            properties Weights file
  [-outfile]           outfile    Output file name

   Additional (Optional) qualifiers (* if not always prompted):
   -outgrno            integer    Species number to use as outgroup
*  -njumble            integer    Number of times to randomise
*  -seed               integer    Random number seed between 1 and 32767 (must
                                  be odd)
   -[no]trout          toggle     Write out trees to tree file
*  -outtreefile        outfile    Tree file name
   -printdata          boolean    Print data at start of run
   -[no]progress       boolean    Print indications of progress of run
   -[no]treeprint      boolean    Print out tree
   -stepbox            boolean    Print steps & compatibility at sites
   -ancseq             boolean    Print sequences at all nodes of tree

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1            integer    Start of each sequence to be used
   -send1              integer    End of each sequence to be used
   -sreverse1          boolean    Reverse (if DNA)
   -sask1              boolean    Ask for begin/end/reverse
   -snucleotide1       boolean    Sequence is nucleotide
   -sprotein1          boolean    Sequence is protein
   -slower1            boolean    Make lower case
   -supper1            boolean    Make upper case
   -sformat1           string     Input sequence format
   -sdbname1           string     Database name
   -sid1               string     Entryname
   -ufo1               string     UFO features
   -fformat1           string     Features format
   -fopenfile1         string     Features file name

   "-outfile" associated qualifiers
   -odirectory3        string     Output directory

   "-outtreefile" associated qualifiers
   -odirectory         string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report deaths


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>File containing one or more sequence alignments</td>
<td>Readable sets of sequences</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>[-intreefile]<br>(Parameter 2)</td>
<td>(no help text) tree value</td>
<td>Phylogenetic tree</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>-weights</td>
<td>Weights file</td>
<td>Property value(s)</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 3)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.fdnacomp</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-outgrno</td>
<td>Species number to use as outgroup</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr>
<td>-njumble</td>
<td>Number of times to randomise</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr>
<td>-seed</td>
<td>Random number seed between 1 and 32767 (must be odd)</td>
<td>Integer from 1 to 32767</td>
<td>1</td>
</tr>

<tr>
<td>-[no]trout</td>
<td>Write out trees to tree file</td>
<td>Toggle value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-outtreefile</td>
<td>Tree file name</td>
<td>Output file</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>-printdata</td>
<td>Print data at start of run</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]progress</td>
<td>Print indications of progress of run</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]treeprint</td>
<td>Print out tree</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-stepbox</td>
<td>Print steps & compatibility at sites</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-ancseq</td>
<td>Print sequences at all nodes of tree</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>fdnacomp</b> reads any normal sequence USAs.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: dnacomp.dat</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
    5   13
Alpha     AACGUGGCCAAAU
Beta      AAGGUCGCCAAAC
Gamma     CAUUUCGUCACAA
Delta     GGUAUUUCGGCCU
Epsilon   GGGAUCUCGGCCC
</pre>
</td></tr></table><p>

<a name="input.2"></a>
<h3>Input files for usage example 2</h3>
<p><h3>File: dnacomptree.dat</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
((((Epsilon,Delta),Gamma),Beta),Alpha);
</pre>
</td></tr></table><p>






<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 

        If you wish to include the standard description of the avalable
	report formats, use:
[an error occurred while processing this directive]
   -->

<b>fdnacomp</b> output is standard: if option 1 is toggled on, the
data is printed out, with the convention that "." means "the same as
in the first species". Then comes a list of equally parsimonious
trees, and (if option 2 is toggled on) a table of the number of
changes of state required in each character. If option 5 is toggled
on, a table is printed out after each tree, showing for each branch
whether there are known to be changes in the branch, and what the
states are inferred to have been at the top end of the branch. If the
inferred state is a "?" or one of the IUB ambiguity symbols, there
will be multiple equally-parsimonious assignments of states; the user
must work these out for themselves by hand. A "?" in the reconstructed
states means that in addition to one or more bases, a gap may or may
not be present. If option 6 is left in its default state the trees
found will be written to a tree file, so that they are available to be
used in other programs. If the program finds multiple trees tied for
best, all of these are written out onto the output tree file. Each is
followed by a numerical weight in square brackets (such as
[0.25000]). This is needed when we use the trees to make a consensus
tree of the results of bootstrapping or jackknifing, to avoid
overrepresenting replicates that find many tied trees.


<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: dnacomp.fdnacomp</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>

DNA compatibility algorithm, version 3.6b

 5 species,  13  sites

Name            Sequences
----            ---------

Alpha        AACGTGGCCA AAT
Beta         AAGGTCGCCA AAC
Gamma        CATTTCGTCA CAA
Delta        GGTATTTCGG CCT
Epsilon      GGGATCTCGG CCC



One most parsimonious tree found:




           +--Epsilon   
        +--4  
     +--3  +--Delta     
     !  !  
  +--2  +-----Gamma     
  !  !  
  1  +--------Beta      
  !  
  +-----------Alpha     

  remember: this is an unrooted tree!


total number of compatible sites is       11.0

steps in each site:
         0   1   2   3   4   5   6   7   8   9
     *-----------------------------------------
    0|       2   1   3   2   0   2   1   1   1
   10|   1   1   1   3                        

 compatibility (Y or N) of each site with this tree:

      0123456789
     *----------
   0 ! YYNYYYYYY
  10 !YYYN      

From    To     Any Steps?    State at upper node
                            
          1                AABGTSGCCA AAY
   1      2        maybe   AABGTCGCCA AAY
   2      3         yes    VAKDTCGCCA CAY
   3      4         yes    GGKATCTCGG CCY
   4   Epsilon     maybe   GGGATCTCGG CCC
   4   Delta        yes    GGTATTTCGG CCT
   3   Gamma        yes    CATTTCGTCA CAA
   2   Beta        maybe   AAGGTCGCCA AAC
   1   Alpha       maybe   AACGTGGCCA AAT


</pre>
</td></tr></table><p>
<p><h3>File: dnacomp.treefile</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
((((Epsilon,Delta),Gamma),Beta),Alpha);
</pre>
</td></tr></table><p>

<a name="output.2"></a>
<h3>Output files for usage example 2</h3>
<p><h3>File: dnacomp.fdnacomp</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>

DNA compatibility algorithm, version 3.6b

User-defined tree:



           +--Epsilon   
        +--4  
     +--3  +--Delta     
     !  !  
  +--2  +-----Gamma     
  !  !  
  1  +--------Beta      
  !  
  +-----------Alpha     

  remember: this is an unrooted tree!


total number of compatible sites is       11.0


</pre>
</td></tr></table><p>






<H2>
    Data files
</H2>

None


<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="../phylip/ednacomp.html">ednacomp</a></td><td>DNA compatibility algorithm</td></tr>
<tr><td><a href="../phylip/ednadist.html">ednadist</a></td><td>Nucleic acid sequence Distance Matrix program</td></tr>
<tr><td><a href="../phylip/ednainvar.html">ednainvar</a></td><td>Nucleic acid sequence Invariants method</td></tr>
<tr><td><a href="../phylip/ednaml.html">ednaml</a></td><td>Estimates phylogenies from nucleic acid sequence Maximum Likelihood</td></tr>
<tr><td><a href="../phylip/ednamlk.html">ednamlk</a></td><td>Estimates phylogenies from nucleic acid sequence Maximum Likelihood with molecular clock</td></tr>
<tr><td><a href="../phylip/ednapars.html">ednapars</a></td><td>DNA parsimony algorithm</td></tr>
<tr><td><a href="../phylip/ednapenny.html">ednapenny</a></td><td>Penny algorithm for DNA</td></tr>
<tr><td><a href="../phylip/eprotdist.html">eprotdist</a></td><td>Protein distance algorithm</td></tr>
<tr><td><a href="../phylip/eprotpars.html">eprotpars</a></td><td>Protein parsimony algorithm</td></tr>
<tr><td><a href="../phylip/erestml.html">erestml</a></td><td>Restriction site Maximum Likelihood method</td></tr>
<tr><td><a href="../phylip/eseqboot.html">eseqboot</a></td><td>Bootstrapped sequences algorithm</td></tr>
<tr><td><a href="fdiscboot.html">fdiscboot</a></td><td>Bootstrapped discrete sites algorithm</td></tr>
<tr><td><a href="fdnadist.html">fdnadist</a></td><td>Nucleic acid sequence Distance Matrix program</td></tr>
<tr><td><a href="fdnainvar.html">fdnainvar</a></td><td>Nucleic acid sequence Invariants method</td></tr>
<tr><td><a href="fdnaml.html">fdnaml</a></td><td>Estimates nucleotide phylogeny by maximum likelihood</td></tr>
<tr><td><a href="fdnamlk.html">fdnamlk</a></td><td>Estimates nucleotide phylogeny by maximum likelihood</td></tr>
<tr><td><a href="fdnamove.html">fdnamove</a></td><td>Interactive DNA parsimony</td></tr>
<tr><td><a href="fdnapars.html">fdnapars</a></td><td>DNA parsimony algorithm</td></tr>
<tr><td><a href="fdnapenny.html">fdnapenny</a></td><td>Penny algorithm for DNA</td></tr>
<tr><td><a href="fdolmove.html">fdolmove</a></td><td>Interactive Dollo or Polymorphism Parsimony</td></tr>
<tr><td><a href="ffreqboot.html">ffreqboot</a></td><td>Bootstrapped genetic frequencies algorithm</td></tr>
<tr><td><a href="fproml.html">fproml</a></td><td>Protein phylogeny by maximum likelihood</td></tr>
<tr><td><a href="fpromlk.html">fpromlk</a></td><td>Protein phylogeny by maximum likelihood</td></tr>
<tr><td><a href="fprotdist.html">fprotdist</a></td><td>Protein distance algorithm</td></tr>
<tr><td><a href="fprotpars.html">fprotpars</a></td><td>Protein pasimony algorithm</td></tr>
<tr><td><a href="frestboot.html">frestboot</a></td><td>Bootstrapped restriction sites algorithm</td></tr>
<tr><td><a href="frestdist.html">frestdist</a></td><td>Computes distance matrix from restriction sites or fragments</td></tr>
<tr><td><a href="frestml.html">frestml</a></td><td>Restriction site maximum Likelihood method</td></tr>
<tr><td><a href="fseqboot.html">fseqboot</a></td><td>Bootstrapped sequences algorithm</td></tr>
<tr><td><a href="fseqbootall.html">fseqbootall</a></td><td>Bootstrapped sequences algorithm</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>

This program is an EMBOSS conversion of a program written by Joe
Felsenstein as part of his PHYLIP package.

<p>
Although we take every care to ensure that the results of the EMBOSS
version are identical to those from the original package, we recommend
that you check your inputs give the same results in both versions
before publication.

<p>
Please report all bugs in the EMBOSS version to the EMBOSS bug team,
not to the original author.










<H2>
    History
</H2>

Written (2004) - Joe Felsenstein, University of Washington.
<p>
Converted (August 2004) to an EMBASSY program by the EMBOSS team.




<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












</BODY>
</HTML>

