<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: fdolmove
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

fdolmove
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<!--END OF HEADER-->






<H2>
    Function
</H2>
Interactive Dollo or Polymorphism Parsimony

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

Interactive construction of phylogenies from discrete character data
with two states (0 and 1) using the Dollo or polymorphism parsimony
criteria. Evaluates parsimony and compatibility criteria for those
phylogenies and displays reconstructed states throughout the
tree. This can be used to find parsimony or compatibility estimates by
hand. 



<H2>
    Algorithm
</H2>


DOLMOVE is an interactive parsimony program which uses the Dollo and
Polymorphism parsimony criteria. It is inspired on Wayne Maddison and
David Maddison's marvellous program MacClade, which is written for
Apple MacIntosh computers. DOLMOVE reads in a data set which is
prepared in almost the same format as one for the Dollo and
polymorhism parsimony program DOLLOP. It allows the user to choose an
initial tree, and displays this tree on the screen. The user can look
at different characters and the way their states are distributed on
that tree, given the most parsimonious reconstruction of state changes
for that particular tree. The user then can specify how the tree is to
be rearraranged, rerooted or written out to a file. By looking at
different rearrangements of the tree the user can manually search for
the most parsimonious tree, and can get a feel for how different
characters are affected by changes in the tree topology.

<p>
This program is compatible with fewer computer systems than the other
programs in PHYLIP. It can be adapted to PCDOS systems or to any
system whose screen or terminals emulate DEC VT100 terminals (such as
Telnet programs for logging in to remote computers over a TCP/IP
network, VT100-compatible windows in the X windowing system, and any
terminal compatible with ANSI standard terminals). For any other
screen types, there is a generic option which does not make use of
screen graphics characters to display the character states. This will
be less effective, as the states will be less easy to see when
displayed.


<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

<b>Here is a sample session with fdolmove</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>fdolmove </b>
Interactive Dollo or Polymorphism Parsimony
Input file: <b>dolmove.dat</b>
Input tree file: <b></b>


Interactive Dollo or polymorphism parsimony, version 3.6b

 5 species,   6 characters


Computing steps needed for compatibility in sites ...


(unrooted)                           5.0 Steps             4 chars compatible
Dollo               
  ,-----------5:Epsilon   
--9  
  !  ,--------4:Delta     
  `--8  
     !  ,-----3:Gamma     
     `--7  
        !  ,--2:Beta      
        `--6  
           `--1:Alpha     

NEXT? (Options: R # + - S . T U W O F H J K L C ? X Q) (? for Help) NEXT? (Options: R # + - S . T U W O F H J K L C ? X Q) (? for Help) Do you want to write out the tree to a file? (Y or N) 
Tree written to file "dolmove.treefile"


</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>




<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-infile]            discretestates File containing data set
  [-intreefile]        tree       (no help text) tree value

   Additional (Optional) qualifiers (* if not always prompted):
   -weights            properties Weights file
   -ancfile            properties Ancestral states file
   -factorfile         properties Factors file
   -method             menu       Parsimony method
   -thresh             toggle     Use threshold parsimony
*  -threshold          float      Threshold value
   -initialtree        menu       Initial tree
   -screenwidth        integer    Width of terminal screen in characters
   -screenlines        integer    Number of lines on screen
   -outtreefile        outfile    Output file name

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-outtreefile" associated qualifiers
   -odirectory         string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report deaths


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-infile]<br>(Parameter 1)</td>
<td>File containing data set</td>
<td>Discrete states file</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>[-intreefile]<br>(Parameter 2)</td>
<td>(no help text) tree value</td>
<td>Phylogenetic tree</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-weights</td>
<td>Weights file</td>
<td>Property value(s)</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>-ancfile</td>
<td>Ancestral states file</td>
<td>Property value(s)</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>-factorfile</td>
<td>Factors file</td>
<td>Property value(s)</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>-method</td>
<td>Parsimony method</td>
<td><table><tr><td>d</td> <td><i>(Dollo)</i></td></tr><tr><td>p</td> <td><i>(Polymorphism)</i></td></tr></table></td>
<td>d</td>
</tr>

<tr>
<td>-thresh</td>
<td>Use threshold parsimony</td>
<td>Toggle value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-threshold</td>
<td>Threshold value</td>
<td>Number 0.000 or more</td>
<td>1</td>
</tr>

<tr>
<td>-initialtree</td>
<td>Initial tree</td>
<td><table><tr><td>a</td> <td><i>(Arbitary)</i></td></tr><tr><td>u</td> <td><i>(User)</i></td></tr><tr><td>s</td> <td><i>(Specify)</i></td></tr></table></td>
<td>Arbitary</td>
</tr>

<tr>
<td>-screenwidth</td>
<td>Width of terminal screen in characters</td>
<td>Any integer value</td>
<td>80</td>
</tr>

<tr>
<td>-screenlines</td>
<td>Number of lines on screen</td>
<td>Any integer value</td>
<td>24</td>
</tr>

<tr>
<td>-outtreefile</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.fdolmove</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>fdolmove</b>  reads discrete character data with "?", "P", "B" states
allowed. .


<h3>(0,1) Discrete character data</h3>

<p>
These programs are intended for the use of morphological
systematists who are dealing with discrete characters, or by molecular
evolutionists dealing with presence-absence data on restriction
sites. One of the programs (PARS) allows multistate characters, with
up to 8 states, plus the unknown state symbol "?". For the others, the
characters are assumed to be coded into a series of (0,1) two-state
characters. For most of the programs there are two other states
possible, "P", which stands for the state of Polymorphism for both
states (0 and 1), and "?", which stands for the state of ignorance: it
is the state "unknown", or "does not apply". The state "P" can also be
denoted by "B", for "both".

<p> There is a method invented by Sokal and Sneath (1963) for linear
sequences of character states, and fully developed for branching
sequences of character states by Kluge and Farris (1969) for recoding
a multistate character into a series of two-state (0,1)
characters. Suppose we had a character with four states whose
character-state tree had the rooted form:

<p>
<pre>
               1 ---> 0 ---> 2
                      |
                      |
                      V
                      3

</pre>

<p>
so that 1 is the ancestral state and 0, 2 and 3 derived states. We
can represent this as three two-state characters:


<p>
<pre>
                Old State           New States
                --- -----           --- ------
                    0                  001
                    1                  000
                    2                  011
                    3                  101

</pre>

<p>
The three new states correspond to the three arrows in the above
character state tree. Possession of one of the new states corresponds
to whether or not the old state had that arrow in its ancestry. Thus
the first new state corresponds to the bottommost arrow, which only
state 3 has in its ancestry, the second state to the rightmost of the
top arrows, and the third state to the leftmost top arrow. This coding
will guarantee that the number of times that states arise on the tree
(in programs MIX, MOVE, PENNY and BOOT) or the number of polymorphic
states in a tree segment (in the Polymorphism option of DOLLOP,
DOLMOVE, DOLPENNY and DOLBOOT) will correctly correspond to what would
have been the case had our programs been able to take multistate
characters into account. Although I have shown the above character
state tree as rooted, the recoding method works equally well on
unrooted multistate characters as long as the connections between the
states are known and contain no loops.

<p>
However, in the default option of programs DOLLOP, DOLMOVE,
DOLPENNY and DOLBOOT the multistate recoding does not necessarily work
properly, as it may lead the program to reconstruct nonexistent state
combinations such as 010. An example of this problem is given in my
paper on alternative phylogenetic methods (1979).

<p>
If you have multistate character data where the states are connected
in a branching "character state tree" you may want to do the binary
recoding yourself. Thanks to Christopher Meacham, the package contains
a program, FACTOR, which will do the recoding itself. For details see
the documentation file for FACTOR.

<p>
We now also have the program PARS, which can do parsimony for
unordered character states.



<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: dolmove.dat</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
     5    6
Alpha     110110
Beta      110000
Gamma     100110
Delta     001001
Epsilon   001110
</pre>
</td></tr></table><p>






<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 

        If you wish to include the standard description of the avalable
	report formats, use:
[an error occurred while processing this directive]
   -->

<b>fdolmove</b> output:

<p>

If the A option is used, then the program will infer, for any
character whose ancestral state is unknown ("?") whether the ancestral
state 0 or 1 will give the fewest changes (according to the criterion
in use). If these are tied, then it may not be possible for the
program to infer the state in the internal nodes, and many of these
will be shown as "?". If the A option is not used, then the program
will assume 0 as the ancestral state.

<p>
When reconstructing the placement of forward changes and reversions
under the Dollo method, keep in mind that each polymorphic state in
the input data will require one "last minute" reversion. This is
included in the counts. Thus if we have both states 0 and 1 at a tip
of the tree the program will assume that the lineage had state 1 up to
the last minute, and then state 0 arose in that population by
reversion, without loss of state 1.

<p>
When DOLMOVE calculates the number of characters compatible with the
tree, it will take the F option into account and count the multistate
characters as units, counting a character as compatible with the tree
only when all of the binary characters corresponding to it are
compatible with the tree.


<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: dolmove.treefile</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
(Epsilon,(Delta,(Gamma,(Beta,Alpha))));
</pre>
</td></tr></table><p>






<H2>
    Data files
</H2>

None

<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="../phylip/ednacomp.html">ednacomp</a></td><td>DNA compatibility algorithm</td></tr>
<tr><td><a href="../phylip/ednadist.html">ednadist</a></td><td>Nucleic acid sequence Distance Matrix program</td></tr>
<tr><td><a href="../phylip/ednainvar.html">ednainvar</a></td><td>Nucleic acid sequence Invariants method</td></tr>
<tr><td><a href="../phylip/ednaml.html">ednaml</a></td><td>Estimates phylogenies from nucleic acid sequence Maximum Likelihood</td></tr>
<tr><td><a href="../phylip/ednamlk.html">ednamlk</a></td><td>Estimates phylogenies from nucleic acid sequence Maximum Likelihood with molecular clock</td></tr>
<tr><td><a href="../phylip/ednapars.html">ednapars</a></td><td>DNA parsimony algorithm</td></tr>
<tr><td><a href="../phylip/ednapenny.html">ednapenny</a></td><td>Penny algorithm for DNA</td></tr>
<tr><td><a href="../phylip/eprotdist.html">eprotdist</a></td><td>Protein distance algorithm</td></tr>
<tr><td><a href="../phylip/eprotpars.html">eprotpars</a></td><td>Protein parsimony algorithm</td></tr>
<tr><td><a href="../phylip/erestml.html">erestml</a></td><td>Restriction site Maximum Likelihood method</td></tr>
<tr><td><a href="../phylip/eseqboot.html">eseqboot</a></td><td>Bootstrapped sequences algorithm</td></tr>
<tr><td><a href="fdiscboot.html">fdiscboot</a></td><td>Bootstrapped discrete sites algorithm</td></tr>
<tr><td><a href="fdnacomp.html">fdnacomp</a></td><td>DNA compatibility algorithm</td></tr>
<tr><td><a href="fdnadist.html">fdnadist</a></td><td>Nucleic acid sequence Distance Matrix program</td></tr>
<tr><td><a href="fdnainvar.html">fdnainvar</a></td><td>Nucleic acid sequence Invariants method</td></tr>
<tr><td><a href="fdnaml.html">fdnaml</a></td><td>Estimates nucleotide phylogeny by maximum likelihood</td></tr>
<tr><td><a href="fdnamlk.html">fdnamlk</a></td><td>Estimates nucleotide phylogeny by maximum likelihood</td></tr>
<tr><td><a href="fdnamove.html">fdnamove</a></td><td>Interactive DNA parsimony</td></tr>
<tr><td><a href="fdnapars.html">fdnapars</a></td><td>DNA parsimony algorithm</td></tr>
<tr><td><a href="fdnapenny.html">fdnapenny</a></td><td>Penny algorithm for DNA</td></tr>
<tr><td><a href="ffreqboot.html">ffreqboot</a></td><td>Bootstrapped genetic frequencies algorithm</td></tr>
<tr><td><a href="fproml.html">fproml</a></td><td>Protein phylogeny by maximum likelihood</td></tr>
<tr><td><a href="fpromlk.html">fpromlk</a></td><td>Protein phylogeny by maximum likelihood</td></tr>
<tr><td><a href="fprotdist.html">fprotdist</a></td><td>Protein distance algorithm</td></tr>
<tr><td><a href="fprotpars.html">fprotpars</a></td><td>Protein pasimony algorithm</td></tr>
<tr><td><a href="frestboot.html">frestboot</a></td><td>Bootstrapped restriction sites algorithm</td></tr>
<tr><td><a href="frestdist.html">frestdist</a></td><td>Computes distance matrix from restriction sites or fragments</td></tr>
<tr><td><a href="frestml.html">frestml</a></td><td>Restriction site maximum Likelihood method</td></tr>
<tr><td><a href="fseqboot.html">fseqboot</a></td><td>Bootstrapped sequences algorithm</td></tr>
<tr><td><a href="fseqbootall.html">fseqbootall</a></td><td>Bootstrapped sequences algorithm</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>

This program is an EMBOSS conversion of a program written by Joe
Felsenstein as part of his PHYLIP package.

<p>
Although we take every care to ensure that the results of the EMBOSS
version are identical to those from the original package, we recommend
that you check your inputs give the same results in both versions
before publication.

<p>
Please report all bugs in the EMBOSS version to the EMBOSS bug team,
not to the original author.


<H2>
    History
</H2>

Written (2004) - Joe Felsenstein, University of Washington.
<p>
Converted (August 2004) to an EMBASSY program by the EMBOSS team.




<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












</BODY>
</HTML>

