<HTML>
<HEAD>
  <TITLE>
  EMBOSS: featmerge
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="/images/emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
featmerge
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
Merge two overlapping feature tables
<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<b>featmerge</b> reads two feature tables for the same sequence
and combines them into a single table.










<H2>
    Algorithm
</H2>


The second feature table is appended to the first, retaining the
original order.







<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

This run merges the GFF format output of one of the <b>antigenic</b>
examples with the original feature table read from SwissProt.

Here is a sample session with <b>featmerge</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>featmerge </b>
Merge two overlapping feature tables
Input feature table: <b>tsw:actb1_takru</b>
Second input feature table: <b>../antigenic-keep/actb1_takru.antigenic</b>
Features output [actb1_takru.gff]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
Merge two overlapping feature tables
Version: EMBOSS:6.6.0.0

   Standard (Mandatory) qualifiers:
  [-afeatures]         features   (no help text) features value
  [-bfeatures]         features   (no help text) features value
  [-outfeat]           featout    [unknown.gff] Output features UFO

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-afeatures" associated qualifiers
   -fformat1           string     Features format
   -iquery1            string     Input query fields or ID list
   -ioffset1           integer    Input start position offset
   -fopenfile1         string     Features file name
   -fask1              boolean    Prompt for begin/end/reverse
   -fbegin1            integer    Start of the features to be used
   -fend1              integer    End of the features to be used
   -freverse1          boolean    Reverse (if DNA)
   -fcircular1         boolean    Circular sequence features

   "-bfeatures" associated qualifiers
   -fformat2           string     Features format
   -iquery2            string     Input query fields or ID list
   -ioffset2           integer    Input start position offset
   -fopenfile2         string     Features file name
   -fask2              boolean    Prompt for begin/end/reverse
   -fbegin2            integer    Start of the features to be used
   -fend2              integer    End of the features to be used
   -freverse2          boolean    Reverse (if DNA)
   -fcircular2         boolean    Circular sequence features

   "-outfeat" associated qualifiers
   -offormat3          string     Output feature format
   -ofopenfile3        string     Features file name
   -ofextension3       string     File name extension
   -ofdirectory3       string     Output directory
   -ofname3            string     Base file name
   -ofsingle3          boolean    Separate file for each entry

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-afeatures]<br>(Parameter 1)</td>
<td>features</td>
<td>(no help text) features value</td>
<td>Readable feature table</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-bfeatures]<br>(Parameter 2)</td>
<td>features</td>
<td>(no help text) features value</td>
<td>Readable feature table</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-outfeat]<br>(Parameter 3)</td>
<td>featout</td>
<td>Output features UFO</td>
<td>Writeable feature table</td>
<td><i>unknown.gff</i></td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td align="left" colspan=5>"-afeatures" associated features qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fformat1<br>-fformat_afeatures</td>
<td>string</td>
<td>Features format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -iquery1<br>-iquery_afeatures</td>
<td>string</td>
<td>Input query fields or ID list</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ioffset1<br>-ioffset_afeatures</td>
<td>integer</td>
<td>Input start position offset</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fopenfile1<br>-fopenfile_afeatures</td>
<td>string</td>
<td>Features file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fask1<br>-fask_afeatures</td>
<td>boolean</td>
<td>Prompt for begin/end/reverse</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fbegin1<br>-fbegin_afeatures</td>
<td>integer</td>
<td>Start of the features to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fend1<br>-fend_afeatures</td>
<td>integer</td>
<td>End of the features to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -freverse1<br>-freverse_afeatures</td>
<td>boolean</td>
<td>Reverse (if DNA)</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fcircular1<br>-fcircular_afeatures</td>
<td>boolean</td>
<td>Circular sequence features</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td align="left" colspan=5>"-bfeatures" associated features qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fformat2<br>-fformat_bfeatures</td>
<td>string</td>
<td>Features format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -iquery2<br>-iquery_bfeatures</td>
<td>string</td>
<td>Input query fields or ID list</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ioffset2<br>-ioffset_bfeatures</td>
<td>integer</td>
<td>Input start position offset</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fopenfile2<br>-fopenfile_bfeatures</td>
<td>string</td>
<td>Features file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fask2<br>-fask_bfeatures</td>
<td>boolean</td>
<td>Prompt for begin/end/reverse</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fbegin2<br>-fbegin_bfeatures</td>
<td>integer</td>
<td>Start of the features to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fend2<br>-fend_bfeatures</td>
<td>integer</td>
<td>End of the features to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -freverse2<br>-freverse_bfeatures</td>
<td>boolean</td>
<td>Reverse (if DNA)</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fcircular2<br>-fcircular_bfeatures</td>
<td>boolean</td>
<td>Circular sequence features</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td align="left" colspan=5>"-outfeat" associated featout qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -offormat3<br>-offormat_outfeat</td>
<td>string</td>
<td>Output feature format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ofopenfile3<br>-ofopenfile_outfeat</td>
<td>string</td>
<td>Features file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ofextension3<br>-ofextension_outfeat</td>
<td>string</td>
<td>File name extension</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ofdirectory3<br>-ofdirectory_outfeat</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ofname3<br>-ofname_outfeat</td>
<td>string</td>
<td>Base file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ofsingle3<br>-ofsingle_outfeat</td>
<td>boolean</td>
<td>Separate file for each entry</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<p>

The input is a standard EMBOSS feaure query.

<p>

<p>

Data can also be read from feaure output in any supported format
format written by an EMBOSS application.

<b>featmerge</b> reads two feature UFOs.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tsw:actb1_takru' is a sequence entry in the example protein database 'tsw'
<p>
<p><h3>File: ../antigenic-keep/actb1_takru.antigenic</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
##gff-version 3
##sequence-region ACTB1_TAKRU 1 375
#!Date 2013-07-15
#!Type Protein
#!Source-version EMBOSS 6.6.0.0
ACTB1_TAKRU	antigenic	epitope	214	222	1.207	.	.	ID=ACTB1_TAKRU.1;note=*pos 218
ACTB1_TAKRU	antigenic	epitope	131	145	1.187	.	.	ID=ACTB1_TAKRU.2;note=*pos 137
ACTB1_TAKRU	antigenic	epitope	5	12	1.166	.	.	ID=ACTB1_TAKRU.3;note=*pos 8
ACTB1_TAKRU	antigenic	epitope	27	38	1.164	.	.	ID=ACTB1_TAKRU.4;note=*pos 32
ACTB1_TAKRU	antigenic	epitope	160	183	1.136	.	.	ID=ACTB1_TAKRU.5;note=*pos 173
ACTB1_TAKRU	antigenic	epitope	367	372	1.135	.	.	ID=ACTB1_TAKRU.6;note=*pos 372
ACTB1_TAKRU	antigenic	epitope	93	108	1.116	.	.	ID=ACTB1_TAKRU.7;note=*pos 103
ACTB1_TAKRU	antigenic	epitope	295	301	1.113	.	.	ID=ACTB1_TAKRU.8;note=*pos 296
ACTB1_TAKRU	antigenic	epitope	256	266	1.11	.	.	ID=ACTB1_TAKRU.9;note=*pos 264
ACTB1_TAKRU	antigenic	epitope	336	352	1.107	.	.	ID=ACTB1_TAKRU.10;note=*pos 347
ACTB1_TAKRU	antigenic	epitope	62	76	1.102	.	.	ID=ACTB1_TAKRU.11;note=*pos 68
ACTB1_TAKRU	antigenic	epitope	232	250	1.086	.	.	ID=ACTB1_TAKRU.12;note=*pos 245
ACTB1_TAKRU	antigenic	epitope	327	332	1.083	.	.	ID=ACTB1_TAKRU.13;note=*pos 330
ACTB1_TAKRU	antigenic	epitope	317	323	1.074	.	.	ID=ACTB1_TAKRU.14;note=*pos 320
ACTB1_TAKRU	antigenic	epitope	186	192	1.068	.	.	ID=ACTB1_TAKRU.15;note=*pos 191
ACTB1_TAKRU	antigenic	epitope	40	46	1.066	.	.	ID=ACTB1_TAKRU.16;note=*pos 43
ACTB1_TAKRU	antigenic	epitope	269	275	1.045	.	.	ID=ACTB1_TAKRU.17;note=*pos 269
ACTB1_TAKRU	antigenic	epitope	51	57	1.034	.	.	ID=ACTB1_TAKRU.18;note=*pos 52
</pre>
</td></tr></table><p>
<p><h3>Database entry: tsw:actb1_takru</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   ACTB1_TAKRU             Reviewed;         375 AA.
AC   P68142; P53484;
DT   25-OCT-2004, integrated into UniProtKB/Swiss-Prot.
DT   25-OCT-2004, sequence version 1.
DT   16-MAY-2012, entry version 49.
DE   RecName: Full=Actin, cytoplasmic 1;
DE   AltName: Full=Beta-actin A;
DE   Contains:
DE     RecName: Full=Actin, cytoplasmic 1, N-terminally processed;
GN   Name=actba;
OS   Takifugu rubripes (Japanese pufferfish) (Fugu rubripes).
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi;
OC   Actinopterygii; Neopterygii; Teleostei; Euteleostei; Neoteleostei;
OC   Acanthomorpha; Acanthopterygii; Percomorpha; Tetraodontiformes;
OC   Tetradontoidea; Tetraodontidae; Takifugu.
OX   NCBI_TaxID=31033;
RN   [1]
RP   NUCLEOTIDE SEQUENCE [GENOMIC DNA], AND TISSUE SPECIFICITY.
RX   MEDLINE=96275651; PubMed=8683572; DOI=10.1006/jmbi.1996.0347;
RA   Venkatesh B., Tay B.H., Elgar G., Brenner S.;
RT   "Isolation, characterization and evolution of nine pufferfish (Fugu
RT   rubripes) actin genes.";
RL   J. Mol. Biol. 259:655-665(1996).
CC   -!- FUNCTION: Actins are highly conserved proteins that are involved
CC       in various types of cell motility and are ubiquitously expressed
CC       in all eukaryotic cells.
CC   -!- SUBUNIT: Polymerization of globular actin (G-actin) leads to a
CC       structural filament (F-actin) in the form of a two-stranded helix.
CC       Each actin can bind to 4 others.
CC   -!- SUBCELLULAR LOCATION: Cytoplasm, cytoskeleton.
CC   -!- TISSUE SPECIFICITY: Widely distributed. Not expressed in skeletal
CC       muscle.
CC   -!- PTM: Oxidation of Met-44 by MICALs (MICAL1, MICAL2 or MICAL3) to
CC       form methionine sulfoxide promotes actin filament
CC       depolymerization. Methionine sulfoxide is produced
CC       stereospecifically, but it is not known whether the (S)-S-oxide or
CC       the (R)-S-oxide is produced (By similarity).
CC   -!- MISCELLANEOUS: In vertebrates 3 main groups of actin isoforms,
CC       alpha, beta and gamma have been identified. The alpha actins are
CC       found in muscle tissues and are a major constituent of the
CC       contractile apparatus. The beta and gamma actins coexist in most
CC       cell types as components of the cytoskeleton and as mediators of
CC       internal cell motility.
CC   -!- MISCELLANEOUS: There are three different beta-cytoplasmic actins
CC       in Fugu rubripes.
CC   -!- SIMILARITY: Belongs to the actin family.
CC   -----------------------------------------------------------------------
CC   Copyrighted by the UniProt Consortium, see http://www.uniprot.org/terms
CC   Distributed under the Creative Commons Attribution-NoDerivs License
CC   -----------------------------------------------------------------------
DR   EMBL; U37499; AAC59889.1; -; Genomic_DNA.
DR   PIR; S71124; S71124.
DR   ProteinModelPortal; P68142; -.
DR   SMR; P68142; 2-375.
DR   Ensembl; ENSTRUT00000013141; ENSTRUP00000013080; ENSTRUG00000005447.
DR   eggNOG; COG5277; -.
DR   GeneTree; ENSGT00630000089629; -.
DR   InParanoid; P68142; -.
DR   OMA; IKNLMER; -.
DR   OrthoDB; EOG41JZC9; -.
DR   GO; GO:0005737; C:cytoplasm; IEA:UniProtKB-KW.
DR   GO; GO:0005856; C:cytoskeleton; IEA:UniProtKB-SubCell.
DR   GO; GO:0005524; F:ATP binding; IEA:UniProtKB-KW.
DR   InterPro; IPR004000; Actin-like.
DR   InterPro; IPR020902; Actin/actin-like_CS.
DR   InterPro; IPR004001; Actin_CS.
DR   PANTHER; PTHR11937; Actin_like; 1.
DR   Pfam; PF00022; Actin; 1.
DR   PRINTS; PR00190; ACTIN.
DR   SMART; SM00268; ACTIN; 1.
DR   PROSITE; PS00406; ACTINS_1; 1.
DR   PROSITE; PS00432; ACTINS_2; 1.
DR   PROSITE; PS01132; ACTINS_ACT_LIKE; 1.
PE   2: Evidence at transcript level;
KW   Acetylation; ATP-binding; Complete proteome; Cytoplasm; Cytoskeleton;
KW   Methylation; Nucleotide-binding; Oxidation; Reference proteome.
FT   CHAIN         1    375       Actin, cytoplasmic 1.
FT                                /FTId=PRO_0000367094.
FT   INIT_MET      1      1       Removed; alternate (By similarity).
FT   CHAIN         2    375       Actin, cytoplasmic 1, N-terminally
FT                                processed.
FT                                /FTId=PRO_0000000809.
FT   MOD_RES       1      1       N-acetylmethionine; in Actin, cytoplasmic
FT                                1; alternate (By similarity).
FT   MOD_RES       2      2       N-acetylglutamate; in Actin, cytoplasmic
FT                                1, N-terminally processed (By
FT                                similarity).
FT   MOD_RES      44     44       Methionine sulfoxide (By similarity).
FT   MOD_RES      73     73       Tele-methylhistidine (By similarity).
SQ   SEQUENCE   375 AA;  41767 MW;  9C505616D33E9495 CRC64;
     MEDEIAALVV DNGSGMCKAG FAGDDAPRAV FPSIVGRPRH QGVMVGMGQK DSYVGDEAQS
     KRGILTLKYP IEHGIVTNWD DMEKIWHHTF YNELRVAPEE HPVLLTEAPL NPKANREKMT
     QIMFETFNTP AMYVAIQAVL SLYASGRTTG IVMDSGDGVT HTVPIYEGYA LPHAILRLDL
     AGRDLTDYLM KILTERGYSF TTTAEREIVR DIKEKLCYVA LDFEQEMGTA ASSSSLEKSY
     ELPDGQVITI GNERFRCPEA LFQPSFLGME SCGIHETTYN SIMKCDVDIR KDLYANTVLS
     GGTTMYPGIA DRMQKEITAL APSTMKIKII APPERKYSVW IGGSILASLS TFQQMWISKQ
     EYDESGPSIV HRKCF
//
</pre>
</td></tr></table><p>





<H2>
    Output file format
</H2>

<p>

The output is a standard EMBOSS feature file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <tt>-offormat xxx</tt>, where 'xxx' is replaced
by the name of the required format.  The available format names are:
gff (gff3), gff2, embl (em), genbank (gb, refseq), ddbj, refseqp, pir
(nbrf), swissprot (swiss, sw), dasgff and debug.

<p>

See:
<A href="http://emboss.sf.net/docs/themes/FeatureFormats.html">
http://emboss.sf.net/docs/themes/FeatureFormats.html</A>
for further information on feature formats.

<p>

<b>featmerge</b> writes a single feature table. By default the output
is in 'gff' feature format.


<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: actb1_takru.gff</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
##gff-version 3
##sequence-region ACTB1_TAKRU 1 375
#!Date 2013-07-15
#!Type Protein
#!Source-version EMBOSS 6.6.0.0
ACTB1_TAKRU	SWISSPROT	mature_protein_region	1	375	.	.	.	ID=ACTB1_TAKRU.1;note=Actin%2C cytoplasmic 1;ftid=PRO_0000367094
ACTB1_TAKRU	SWISSPROT	cleaved_initiator_methionine	1	1	.	.	.	ID=ACTB1_TAKRU.2;note=Removed%3B alternate;comment=By similarity
ACTB1_TAKRU	SWISSPROT	mature_protein_region	2	375	.	.	.	ID=ACTB1_TAKRU.3;note=Actin%2C cytoplasmic 1%2C N-terminally processed;ftid=PRO_0000000809
ACTB1_TAKRU	SWISSPROT	post_translationally_modified_region	1	1	.	.	.	ID=ACTB1_TAKRU.4;note=N-acetylmethionine%3B in Actin%2C cytoplasmic 1%3B alternate;comment=By similarity
ACTB1_TAKRU	SWISSPROT	post_translationally_modified_region	2	2	.	.	.	ID=ACTB1_TAKRU.5;note=N-acetylglutamate%3B in Actin%2C cytoplasmic 1%2C N-terminally processed;comment=By similarity
ACTB1_TAKRU	SWISSPROT	post_translationally_modified_region	44	44	.	.	.	ID=ACTB1_TAKRU.6;note=Methionine sulfoxide;comment=By similarity
ACTB1_TAKRU	SWISSPROT	post_translationally_modified_region	73	73	.	.	.	ID=ACTB1_TAKRU.7;note=Tele-methylhistidine;comment=By similarity
ACTB1_TAKRU	antigenic	epitope	214	222	1.207	.	.	ID=ACTB1_TAKRU.1;note=*pos 218
ACTB1_TAKRU	antigenic	epitope	131	145	1.187	.	.	ID=ACTB1_TAKRU.2;note=*pos 137
ACTB1_TAKRU	antigenic	epitope	5	12	1.166	.	.	ID=ACTB1_TAKRU.3;note=*pos 8
ACTB1_TAKRU	antigenic	epitope	27	38	1.164	.	.	ID=ACTB1_TAKRU.4;note=*pos 32
ACTB1_TAKRU	antigenic	epitope	160	183	1.136	.	.	ID=ACTB1_TAKRU.5;note=*pos 173
ACTB1_TAKRU	antigenic	epitope	367	372	1.135	.	.	ID=ACTB1_TAKRU.6;note=*pos 372
ACTB1_TAKRU	antigenic	epitope	93	108	1.116	.	.	ID=ACTB1_TAKRU.7;note=*pos 103
ACTB1_TAKRU	antigenic	epitope	295	301	1.113	.	.	ID=ACTB1_TAKRU.8;note=*pos 296
ACTB1_TAKRU	antigenic	epitope	256	266	1.11	.	.	ID=ACTB1_TAKRU.9;note=*pos 264
ACTB1_TAKRU	antigenic	epitope	336	352	1.107	.	.	ID=ACTB1_TAKRU.10;note=*pos 347
ACTB1_TAKRU	antigenic	epitope	62	76	1.102	.	.	ID=ACTB1_TAKRU.11;note=*pos 68
ACTB1_TAKRU	antigenic	epitope	232	250	1.086	.	.	ID=ACTB1_TAKRU.12;note=*pos 245
ACTB1_TAKRU	antigenic	epitope	327	332	1.083	.	.	ID=ACTB1_TAKRU.13;note=*pos 330
ACTB1_TAKRU	antigenic	epitope	317	323	1.074	.	.	ID=ACTB1_TAKRU.14;note=*pos 320
ACTB1_TAKRU	antigenic	epitope	186	192	1.068	.	.	ID=ACTB1_TAKRU.15;note=*pos 191
ACTB1_TAKRU	antigenic	epitope	40	46	1.066	.	.	ID=ACTB1_TAKRU.16;note=*pos 43
ACTB1_TAKRU	antigenic	epitope	269	275	1.045	.	.	ID=ACTB1_TAKRU.17;note=*pos 269
ACTB1_TAKRU	antigenic	epitope	51	57	1.034	.	.	ID=ACTB1_TAKRU.18;note=*pos 52
</pre>
</td></tr></table><p>





<H2>
    Data files
</H2>

None.


<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
<ol>

<li>

</ol>

   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="aligncopy.html">aligncopy</a></td>
<td>Read and write alignments</td>
</tr>

<tr>
<td><a href="aligncopypair.html">aligncopypair</a></td>
<td>Read and write pairs from alignments</td>
</tr>

<tr>
<td><a href="biosed.html">biosed</a></td>
<td>Replace or delete sequence sections</td>
</tr>

<tr>
<td><a href="codcopy.html">codcopy</a></td>
<td>Copy and reformat a codon usage table</td>
</tr>

<tr>
<td><a href="cutseq.html">cutseq</a></td>
<td>Remove a section from a sequence</td>
</tr>

<tr>
<td><a href="degapseq.html">degapseq</a></td>
<td>Remove non-alphabetic (e.g. gap) characters from sequences</td>
</tr>

<tr>
<td><a href="descseq.html">descseq</a></td>
<td>Alter the name or description of a sequence</td>
</tr>

<tr>
<td><a href="entret.html">entret</a></td>
<td>Retrieve sequence entries from flatfile databases and files</td>
</tr>

<tr>
<td><a href="extractalign.html">extractalign</a></td>
<td>Extract regions from a sequence alignment</td>
</tr>

<tr>
<td><a href="extractfeat.html">extractfeat</a></td>
<td>Extract features from sequence(s)</td>
</tr>

<tr>
<td><a href="extractseq.html">extractseq</a></td>
<td>Extract regions from a sequence</td>
</tr>

<tr>
<td><a href="featcopy.html">featcopy</a></td>
<td>Read and write a feature table</td>
</tr>

<tr>
<td><a href="featreport.html">featreport</a></td>
<td>Read and write a feature table</td>
</tr>

<tr>
<td><a href="feattext.html">feattext</a></td>
<td>Return a feature table original text</td>
</tr>

<tr>
<td><a href="listor.html">listor</a></td>
<td>Write a list file of the logical OR of two sets of sequences</td>
</tr>

<tr>
<td><a href="makenucseq.html">makenucseq</a></td>
<td>Create random nucleotide sequences</td>
</tr>

<tr>
<td><a href="makeprotseq.html">makeprotseq</a></td>
<td>Create random protein sequences</td>
</tr>

<tr>
<td><a href="maskambignuc.html">maskambignuc</a></td>
<td>Mask all ambiguity characters in nucleotide sequences with N</td>
</tr>

<tr>
<td><a href="maskambigprot.html">maskambigprot</a></td>
<td>Mask all ambiguity characters in protein sequences with X</td>
</tr>

<tr>
<td><a href="maskfeat.html">maskfeat</a></td>
<td>Write a sequence with masked features</td>
</tr>

<tr>
<td><a href="maskseq.html">maskseq</a></td>
<td>Write a sequence with masked regions</td>
</tr>

<tr>
<td><a href="newseq.html">newseq</a></td>
<td>Create a sequence file from a typed-in sequence</td>
</tr>

<tr>
<td><a href="nohtml.html">nohtml</a></td>
<td>Remove mark-up (e.g. HTML tags) from an ASCII text file</td>
</tr>

<tr>
<td><a href="noreturn.html">noreturn</a></td>
<td>Remove carriage return from ASCII files</td>
</tr>

<tr>
<td><a href="nospace.html">nospace</a></td>
<td>Remove whitespace from an ASCII text file</td>
</tr>

<tr>
<td><a href="notab.html">notab</a></td>
<td>Replace tabs with spaces in an ASCII text file</td>
</tr>

<tr>
<td><a href="notseq.html">notseq</a></td>
<td>Write to file a subset of an input stream of sequences</td>
</tr>

<tr>
<td><a href="nthseq.html">nthseq</a></td>
<td>Write to file a single sequence from an input stream of sequences</td>
</tr>

<tr>
<td><a href="nthseqset.html">nthseqset</a></td>
<td>Read and write (return) one set of sequences from many</td>
</tr>

<tr>
<td><a href="pasteseq.html">pasteseq</a></td>
<td>Insert one sequence into another</td>
</tr>

<tr>
<td><a href="revseq.html">revseq</a></td>
<td>Reverse and complement a nucleotide sequence</td>
</tr>

<tr>
<td><a href="seqcount.html">seqcount</a></td>
<td>Read and count sequences</td>
</tr>

<tr>
<td><a href="seqret.html">seqret</a></td>
<td>Read and write (return) sequences</td>
</tr>

<tr>
<td><a href="seqretsetall.html">seqretsetall</a></td>
<td>Read and write (return) many sets of sequences</td>
</tr>

<tr>
<td><a href="seqretsplit.html">seqretsplit</a></td>
<td>Read sequences and write them to individual files</td>
</tr>

<tr>
<td><a href="sizeseq.html">sizeseq</a></td>
<td>Sort sequences by size</td>
</tr>

<tr>
<td><a href="skipredundant.html">skipredundant</a></td>
<td>Remove redundant sequences from an input set</td>
</tr>

<tr>
<td><a href="skipseq.html">skipseq</a></td>
<td>Read and write (return) sequences, skipping first few</td>
</tr>

<tr>
<td><a href="splitsource.html">splitsource</a></td>
<td>Split sequence(s) into original source sequences</td>
</tr>

<tr>
<td><a href="splitter.html">splitter</a></td>
<td>Split sequence(s) into smaller sequences</td>
</tr>

<tr>
<td><a href="trimest.html">trimest</a></td>
<td>Remove poly-A tails from nucleotide sequences</td>
</tr>

<tr>
<td><a href="trimseq.html">trimseq</a></td>
<td>Remove unwanted characters from start and end of sequence(s)</td>
</tr>

<tr>
<td><a href="trimspace.html">trimspace</a></td>
<td>Remove extra whitespace from an ASCII text file</td>
</tr>

<tr>
<td><a href="union.html">union</a></td>
<td>Concatenate multiple sequences into a single sequence</td>
</tr>

<tr>
<td><a href="vectorstrip.html">vectorstrip</a></td>
<td>Remove vectors from the ends of nucleotide sequence(s)</td>
</tr>

<tr>
<td><a href="yank.html">yank</a></td>
<td>Add a sequence reference (a full USA) to a list file</td>
</tr>

</table>
<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>

Peter Rice
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK

<p>
Please report all bugs to the EMBOSS bug team (emboss-bug&nbsp;&copy;&nbsp;emboss.open-bio.org) not to the original author.



<H2>
    History
</H2>
<!--
        Date written and what changes have been made go in this file.
   -->




<H2>
    Target users
</H2>
<!--
        For general users, use this text
   -->
This program is intended to be used by everyone and everything, from naive users to embedded scripts.

<H2>
    Comments
</H2>
<!--
        User/developer/other comments go in this file.
   -->
None


</BODY>
</HTML>

