<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: fneighbor
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

fneighbor
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<!--END OF HEADER-->






<H2>
    Function
</H2>
Phylogenies from distance matrix by N-J or UPGMA method

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

An implementation by Mary Kuhner and John Yamato of Saitou and Nei's
"Neighbor Joining Method," and of the UPGMA (Average Linkage
clustering) method. Neighbor Joining is a distance matrix method
producing an unrooted tree without the assumption of a clock. UPGMA
does assume a clock. The branch lengths are not optimized by the least
squares criterion but the methods are very fast and thus can handle
much larger data sets.


<H2>
    Algorithm
</H2>

This program implements the Neighbor-Joining method of Saitou and Nei
(1987) and the UPGMA method of clustering. The program was written by
Mary Kuhner and Jon Yamato, using some code from program FITCH. An
important part of the code was translated from FORTRAN code from the
neighbor-joining program written by Naruya Saitou and by Li Jin, and
is used with the kind permission of Drs. Saitou and Jin.

<p>
NEIGHBOR constructs a tree by successive clustering of lineages,
setting branch lengths as the lineages join. The tree is not
rearranged thereafter. The tree does not assume an evolutionary clock,
so that it is in effect an unrooted tree. It should be somewhat
similar to the tree obtained by FITCH. The program cannot evaluate a
User tree, nor can it prevent branch lengths from becoming
negative. However the algorithm is far faster than FITCH or
KITSCH. This will make it particularly effective in their place for
large studies or for bootstrap or jackknife resampling studies which
require runs on multiple data sets.

<p>
The UPGMA option constructs a tree by successive (agglomerative)
clustering using an average-linkage method of clustering. It has some
relationship to KITSCH, in that when the tree topology turns out the
same, the branch lengths with UPGMA will turn out to be the same as
with the P = 0 option of KITSCH.

<p>
The programs FITCH, KITSCH, and NEIGHBOR are for dealing with data
which comes in the form of a matrix of pairwise distances between all
pairs of taxa, such as distances based on molecular sequence data,
gene frequency genetic distances, amounts of DNA hybridization, or
immunological distances. In analyzing these data, distance matrix
programs implicitly assume that:

<ul>

<li>
Each distance is measured independently from the others: no item of
data contributes to more than one distance.

<li>
The distance between each pair of taxa is drawn from a distribution
with an expectation which is the sum of values (in effect amounts of
evolution) along the tree from one tip to the other. The variance of
the distribution is proportional to a power p of the expectation.

</ul>

<p>
These assumptions can be traced in the least squares methods of
programs FITCH and KITSCH but it is not quite so easy to see them in
operation in the Neighbor-Joining method of NEIGHBOR, where the
independence assumptions is less obvious.

<p>
THESE TWO ASSUMPTIONS ARE DUBIOUS IN MOST CASES: independence will not
be expected to be true in most kinds of data, such as genetic
distances from gene frequency data. For genetic distance data in which
pure genetic drift without mutation can be assumed to be the mechanism
of change CONTML may be more appropriate. However, FITCH, KITSCH, and
NEIGHBOR will not give positively misleading results (they will not
make a statistically inconsistent estimate) provided that additivity
holds, which it will if the distance is computed from the original
data by a method which corrects for reversals and parallelisms in
evolution. If additivity is not expected to hold, problems are more
severe. A short discussion of these matters will be found in a review
article of mine (1984a). For detailed, if sometimes irrelevant,
controversy see the papers by Farris (1981, 1985, 1986) and myself
(1986, 1988b).

<p>
For genetic distances from gene frequencies, FITCH, KITSCH, and
NEIGHBOR may be appropriate if a neutral mutation model can be assumed
and Nei's genetic distance is used, or if pure drift can be assumed
and either Cavalli-Sforza's chord measure or Reynolds, Weir, and
Cockerham's (1983) genetic distance is used. However, in the latter
case (pure drift) CONTML should be better.

<p>
Restriction site and restriction fragment data can be treated by
distance matrix methods if a distance such as that of Nei and Li
(1979) is used. Distances of this sort can be computed in PHYLIp by
the program RESTDIST.

<p>
For nucleic acid sequences, the distances computed in DNADIST allow
correction for multiple hits (in different ways) and should allow one
to analyse the data under the presumption of additivity. In all of
these cases independence will not be expected to hold. DNA
hybridization and immunological distances may be additive and
independent if transformed properly and if (and only if) the standards
against which each value is measured are independent. (This is rarely
exactly true).

<p>
FITCH and the Neighbor-Joining option of NEIGHBOR fit a tree which has
the branch lengths unconstrained. KITSCH and the UPGMA option of
NEIGHBOR, by contrast, assume that an "evolutionary clock" is valid,
according to which the true branch lengths from the root of the tree
to each tip are the same: the expected amount of evolution in any
lineage is proportional to elapsed time.


<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

<b>Here is a sample session with fneighbor</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>fneighbor </b>
Phylogenies from distance matrix by N-J or UPGMA method
Input file: <b>neighbor.dat</b>
Output file [neighbor.fneighbor]: <b></b>



Cycle   4: species 1 (   0.91769) joins species 2 (   0.76891)
Cycle   3: node 1 (   0.42027) joins species 3 (   0.35793)
Cycle   2: species 6 (   0.15168) joins species 7 (   0.11752)
Cycle   1: node 1 (   0.04648) joins species 4 (   0.28469)
last cycle:
 node 1  (   0.02696) joins species 5  (   0.15393) joins node 6  (   0.03982)
coordinates lengthsum: 0.000000
Calling coordinates q->v 0.768910
coordinates lengthsum: 0.768910
Calling coordinates q->v 0.420269
coordinates lengthsum: 0.420269
Calling coordinates q->v 0.357931
coordinates lengthsum: 0.778200
Calling coordinates q->v 0.046481
coordinates lengthsum: 0.466750
Calling coordinates q->v 0.284694
coordinates lengthsum: 0.751444
Calling coordinates q->v 0.026956
coordinates lengthsum: 0.493706
Calling coordinates q->v 0.153931
coordinates lengthsum: 0.647637
Calling coordinates q->v 0.039819
coordinates lengthsum: 0.533525
Calling coordinates q->v 0.151675
coordinates lengthsum: 0.685200
Calling coordinates q->v 0.117525
coordinates lengthsum: 0.651050
Calling coordinates q->v 0.917690
coordinates lengthsum: 0.917690

Output written on file "neighbor.fneighbor"

Tree written on file "neighbor.treefile"

Done.


</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>




<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-datafile]          distances  (no help text) distances value
  [-outfile]           outfile    Output file name

   Additional (Optional) qualifiers (* if not always prompted):
   -matrixtype         menu       Type of data matrix
   -treetype           menu       Neighbor-joining or UPGMA tree
*  -outgrno            integer    Species number to use as outgroup
   -jumble             toggle     Randomise input order of species
*  -seed               integer    Random number seed between 1 and 32767 (must
                                  be odd)
   -replicates         boolean    Subreplicates
   -[no]trout          toggle     Write out trees to tree file
*  -outtreefile        outfile    Tree file name
   -printdata          boolean    Print data at start of run
   -[no]progress       boolean    Print indications of progress of run
   -[no]treeprint      boolean    Print out tree

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-outfile" associated qualifiers
   -odirectory2        string     Output directory

   "-outtreefile" associated qualifiers
   -odirectory         string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report deaths


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-datafile]<br>(Parameter 1)</td>
<td>(no help text) distances value</td>
<td>Distance matrix</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.fneighbor</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-matrixtype</td>
<td>Type of data matrix</td>
<td><table><tr><td>s</td> <td><i>(Square)</i></td></tr><tr><td>u</td> <td><i>(Upper triangular)</i></td></tr><tr><td>l</td> <td><i>(Lower triangular)</i></td></tr></table></td>
<td>s</td>
</tr>

<tr>
<td>-treetype</td>
<td>Neighbor-joining or UPGMA tree</td>
<td><table><tr><td>n</td> <td><i>(Neighbor-joining)</i></td></tr><tr><td>u</td> <td><i>(UPGMA)</i></td></tr></table></td>
<td>n</td>
</tr>

<tr>
<td>-outgrno</td>
<td>Species number to use as outgroup</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr>
<td>-jumble</td>
<td>Randomise input order of species</td>
<td>Toggle value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-seed</td>
<td>Random number seed between 1 and 32767 (must be odd)</td>
<td>Integer from 1 to 32767</td>
<td>1</td>
</tr>

<tr>
<td>-replicates</td>
<td>Subreplicates</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]trout</td>
<td>Write out trees to tree file</td>
<td>Toggle value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-outtreefile</td>
<td>Tree file name</td>
<td>Output file</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>-printdata</td>
<td>Print data at start of run</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]progress</td>
<td>Print indications of progress of run</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]treeprint</td>
<td>Print out tree</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>fneighbor</b> reads any normal sequence USAs.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: neighbor.dat</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
    7
Bovine      0.0000  1.6866  1.7198  1.6606  1.5243  1.6043  1.5905
Mouse       1.6866  0.0000  1.5232  1.4841  1.4465  1.4389  1.4629
Gibbon      1.7198  1.5232  0.0000  0.7115  0.5958  0.6179  0.5583
Orang       1.6606  1.4841  0.7115  0.0000  0.4631  0.5061  0.4710
Gorilla     1.5243  1.4465  0.5958  0.4631  0.0000  0.3484  0.3083
Chimp       1.6043  1.4389  0.6179  0.5061  0.3484  0.0000  0.2692
Human       1.5905  1.4629  0.5583  0.4710  0.3083  0.2692  0.0000
</pre>
</td></tr></table><p>






<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 

        If you wish to include the standard description of the avalable
	report formats, use:
[an error occurred while processing this directive]
   -->

<b>fneighbor</b> output consists of an tree (rooted if UPGMA, unrooted
if Neighbor-Joining) and the lengths of the interior segments. The
Average Percent Standard Deviation is not computed or printed out. If
the tree found by Neighbor is fed into FITCH as a User Tree, it will
compute this quantity if one also selects the N option of FITCH to
ensure that none of the branch lengths is re-estimated.

<p>
As NEIGHBOR runs it prints out an account of the successive clustering
levels, if you allow it to. This is mostly for reassurance and can be
suppressed using menu option 2. In this printout of cluster levels the
word "OTU" refers to a tip species, and the word "NODE" to an interior
node of the resulting tree.



<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: neighbor.fneighbor</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>

Neighbor-Joining/UPGMA method version 3.6b


   7 Populations

 Neighbor-joining method

 Negative branch lengths allowed


  +---------------------------------------------Mouse     
  ! 
  !                        +---------------------Gibbon    
  1------------------------2 
  !                        !  +----------------Orang     
  !                        +--4 
  !                           ! +--------Gorilla   
  !                           +-5 
  !                             ! +--------Chimp     
  !                             +-3 
  !                               +------Human     
  ! 
  +------------------------------------------------------Bovine    


remember: this is an unrooted tree!

Between        And            Length
-------        ---            ------
   1          Mouse           0.76891
   1             2            0.42027
   2          Gibbon          0.35793
   2             4            0.04648
   4          Orang           0.28469
   4             5            0.02696
   5          Gorilla         0.15393
   5             3            0.03982
   3          Chimp           0.15168
   3          Human           0.11752
   1          Bovine          0.91769


</pre>
</td></tr></table><p>
<p><h3>File: neighbor.treefile</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
(Mouse:0.76891,(Gibbon:0.35793,(Orang:0.28469,(Gorilla:0.15393,
(Chimp:0.15168,Human:0.11752):0.03982):0.02696):0.04648):0.42027,Bovine:0.91769);
</pre>
</td></tr></table><p>






<H2>
    Data files
</H2>

None

<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="../../apps/distmat.html">distmat</a></td><td>Creates a distance matrix from multiple alignments</td></tr>
<tr><td><a href="../phylip/efitch.html">efitch</a></td><td>Fitch-Margoliash and Least-Squares Distance Methods</td></tr>
<tr><td><a href="../phylip/ekitsch.html">ekitsch</a></td><td>Fitch-Margoliash method with contemporary tips</td></tr>
<tr><td><a href="../phylip/eneighbor.html">eneighbor</a></td><td>Phylogenies from distance matrix by N-J or UPGMA method</td></tr>
<tr><td><a href="ffitch.html">ffitch</a></td><td>Fitch-Margoliash and Least-Squares Distance Methods</td></tr>
<tr><td><a href="fkitsch.html">fkitsch</a></td><td>Fitch-Margoliash method with contemporary tips</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>

This program is an EMBOSS conversion of a program written by Joe
Felsenstein as part of his PHYLIP package.

<p>
Although we take every care to ensure that the results of the EMBOSS
version are identical to those from the original package, we recommend
that you check your inputs give the same results in both versions
before publication.

<p>
Please report all bugs in the EMBOSS version to the EMBOSS bug team,
not to the original author.


<H2>
    History
</H2>

Written (2004) - Joe Felsenstein, University of Washington.
<p>
Converted (August 2004) to an EMBASSY program by the EMBOSS team.




<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












</BODY>
</HTML>

