<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: fpromlk
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

fpromlk
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<!--END OF HEADER-->






<H2>
    Function
</H2>
Protein phylogeny by maximum likelihood

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

Same as PROML but assumes a molecular clock. The use of the two
programs together permits a likelihood ratio test of the molecular
clock hypothesis to be made.
<p>
Estimates phylogenies from protein amino acid sequences by maximum
likelihood. The PAM, JTT, or PMB models can be employed, and also use
of a Hidden Markov model of rates, with the program inferring which
sites have which rates. This also allows gamma-distribution and
gamma-plus-invariant sites distributions of rates across sites. It
also allows different rates of change at known sites.


<H2>
    Algorithm
</H2>


This program implements the maximum likelihood method for protein
amino acid sequences under the constraint that the trees estimated
must be consistent with a molecular clock. The molecular clock is the
assumption that the tips of the tree are all equidistant, in branch
length, from its root. This program is indirectly related to PROML. It
uses the Dayhoff probability model of change between amino acids. Its
algorithmic details are not yet published, but many of them are
similar to DNAMLK.

<p>
The assumptions of the model are: 

<ol>

<li>
Each position in the sequence evolves independently. 

<li>
Different lineages evolve independently. 

<li>
Each position undergoes substitution at an expected rate which is
chosen from a series of rates (each with a probability of occurrence)
which we specify.

<li>
All relevant positions are included in the sequence, not just those
that have changed or those that are "phylogenetically informative".

<li>
The probabilities of change between amino acids are given by the model
of Jones,

<li>
 Taylor, and Thornton (1992), the PMB model of Veerassamy, Smith and
 Tillier (2004), or the PAM model of Dayhoff (Dayhoff and Eck, 1968;
 Dayhoff et. al., 1979).

</ol>

<p>
Note the assumption that we are looking at all positions, including
those that have not changed at all. It is important not to restrict
attention to some positions based on whether or not they have changed;
doing that would bias branch lengths by making them too long, and that
in turn would cause the method to misinterpret the meaning of those
positions that had changed.

<p>
This program uses a Hidden Markov Model (HMM) method of inferring
different rates of evolution at different amino acid positions. This
was described in a paper by me and Gary Churchill (1996). It allows us
to specify to the program that there will be a number of different
possible evolutionary rates, what the prior probabilities of
occurrence of each is, and what the average length of a patch of
positions all having the same rate. The rates can also be chosen by
the program to approximate a Gamma distribution of rates, or a Gamma
distribution plus a class of invariant positions. The program computes
the likelihood by summing it over all possible assignments of rates to
positions, weighting each by its prior probability of occurrence.

<p>
For example, if we have used the C and A options (described below) to
specify that there are three possible rates of evolution, 1.0, 2.4,
and 0.0, that the prior probabilities of a position having these rates
are 0.4, 0.3, and 0.3, and that the average patch length (number of
consecutive positions with the same rate) is 2.0, the program will sum
the likelihood over all possibilities, but giving less weight to those
that (say) assign all positions to rate 2.4, or that fail to have
consecutive positions that have the same rate.

<p>
The Hidden Markov Model framework for rate variation among positions
was independently developed by Yang (1993, 1994, 1995). We have
implemented a general scheme for a Hidden Markov Model of rates; we
allow the rates and their prior probabilities to be specified
arbitrarily by the user, or by a discrete approximation to a Gamma
distribution of rates (Yang, 1995), or by a mixture of a Gamma
distribution and a class of invariant positions.

<p>
This feature effectively removes the artificial assumption that all
positions have the same rate, and also means that we need not know in
advance the identities of the positions that have a particular rate of
evolution.

<p>
Another layer of rate variation also is available. The user can assign
categories of rates to each positions (for example, we might want
amino acid positions in the active site of a protein to change more
slowly than other positions. This is done with the categories input
file and the C option. We then specify (using the menu) the relative
rates of evolution of amino acid positions in the different
categories. For example, we might specify that positions in the active
site evolve at relative rates of 0.2 compared to 1.0 at other
positions. If we are assuming that a particular position maintains a
cysteine bridge to another, we may want to put it in a category of
positions (including perhaps the initial position of the protein
sequence which maintains methionine) which changes at a rate of 0.0.

<p>
If both user-assigned rate categories and Hidden Markov Model rates
are allowed, the program assumes that the actual rate at a position is
the product of the user-assigned category rate and the Hidden Markov
Model regional rate. (This may not always make perfect biological
sense: it would be more natural to assume some upper bound to the
rate, as we have discussed in the Felsenstein and Churchill
paper). Nevertheless you may want to use both types of rate variation.


<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

<b>Here is a sample session with fpromlk</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>fpromlk </b>
Protein phylogeny by maximum likelihood
Input sequence: <b>promlk.dat</b>
Input tree file: <b></b>
Output file [promlk.fpromlk]: <b></b>


Amino acid sequence
Maximum Likelihood method with molecular clock, version 3.6b


Adding species:
   1. Alpha     
   2. Beta      
   3. Gamma     
   4. Delta     
   5. Epsilon   


Output written to file "promlk.fpromlk"

Tree also written onto file "promlk.treefile"

Done.


</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>




<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-sequence]          seqsetall  File containing one or more sequence
                                  alignments
  [-intreefile]        tree       (no help text) tree value
  [-outfile]           outfile    Output file name

   Additional (Optional) qualifiers (* if not always prompted):
   -ncategories        integer    Number of substitution rate categories
*  -rate               array      Rate for each category
*  -categories         properties File of substitution rate categories
   -weights            properties Weights file
*  -lengths            boolean    Use branch lengths from user trees
   -model              menu       Probability model for amino acid change
   -gamma              menu       Rate variation among sites
*  -gammacoefficient   float      Coefficient of variation of substitution
                                  rate among sites
*  -ngammacat          integer    Number of categories (1-9)
*  -invarcoefficient   float      Coefficient of variation of substitution
                                  rate among sites
*  -ninvarcat          integer    Number of categories (1-9) including one for
                                  invariant sites
*  -invarfrac          float      Fraction of invariant sites
*  -nhmmcategories     integer    Number of HMM rate categories
*  -hmmrates           array      HMM category rates
*  -hmmprobabilities   array      Probability for each HMM category
*  -adjsite            boolean    Rates at adjacent sites correlated
*  -lambda             float      Mean block length of sites having the same
                                  rate
*  -njumble            integer    Number of times to randomise
*  -seed               integer    Random number seed between 1 and 32767 (must
                                  be odd)
*  -global             boolean    Global rearrangements
   -outgrno            integer    Species number to use as outgroup
   -[no]trout          toggle     Write out trees to tree file
*  -outtreefile        outfile    Tree file name
   -printdata          boolean    Print data at start of run
   -[no]progress       boolean    Print indications of progress of run
   -[no]treeprint      boolean    Print out tree
   -hypstate           boolean    Reconstruct hypothetical sequence

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1            integer    Start of each sequence to be used
   -send1              integer    End of each sequence to be used
   -sreverse1          boolean    Reverse (if DNA)
   -sask1              boolean    Ask for begin/end/reverse
   -snucleotide1       boolean    Sequence is nucleotide
   -sprotein1          boolean    Sequence is protein
   -slower1            boolean    Make lower case
   -supper1            boolean    Make upper case
   -sformat1           string     Input sequence format
   -sdbname1           string     Database name
   -sid1               string     Entryname
   -ufo1               string     UFO features
   -fformat1           string     Features format
   -fopenfile1         string     Features file name

   "-outfile" associated qualifiers
   -odirectory3        string     Output directory

   "-outtreefile" associated qualifiers
   -odirectory         string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report deaths


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>File containing one or more sequence alignments</td>
<td>Readable sets of sequences</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>[-intreefile]<br>(Parameter 2)</td>
<td>(no help text) tree value</td>
<td>Phylogenetic tree</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 3)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.fpromlk</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-ncategories</td>
<td>Number of substitution rate categories</td>
<td>Integer from 1 to 9</td>
<td>1</td>
</tr>

<tr>
<td>-rate</td>
<td>Rate for each category</td>
<td>List of floating point numbers</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>-categories</td>
<td>File of substitution rate categories</td>
<td>Property value(s)</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>-weights</td>
<td>Weights file</td>
<td>Property value(s)</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>-lengths</td>
<td>Use branch lengths from user trees</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-model</td>
<td>Probability model for amino acid change</td>
<td><table><tr><td>j</td> <td><i>(Jones-Taylor-Thornton)</i></td></tr><tr><td>h</td> <td><i>(Henikoff/Tillier PMBs)</i></td></tr><tr><td>d</td> <td><i>(Dayhoff PAM)</i></td></tr></table></td>
<td>Jones-Taylor-Thornton</td>
</tr>

<tr>
<td>-gamma</td>
<td>Rate variation among sites</td>
<td><table><tr><td>g</td> <td><i>(Gamma distributed rates)</i></td></tr><tr><td>i</td> <td><i>(Gamma+invariant sites)</i></td></tr><tr><td>h</td> <td><i>(User defined HMM of rates)</i></td></tr><tr><td>n</td> <td><i>(Constant rate)</i></td></tr></table></td>
<td>n</td>
</tr>

<tr>
<td>-gammacoefficient</td>
<td>Coefficient of variation of substitution rate among sites</td>
<td>Number 0.001 or more</td>
<td>1</td>
</tr>

<tr>
<td>-ngammacat</td>
<td>Number of categories (1-9)</td>
<td>Integer from 1 to 9</td>
<td>1</td>
</tr>

<tr>
<td>-invarcoefficient</td>
<td>Coefficient of variation of substitution rate among sites</td>
<td>Number 0.001 or more</td>
<td>1</td>
</tr>

<tr>
<td>-ninvarcat</td>
<td>Number of categories (1-9) including one for invariant sites</td>
<td>Integer from 1 to 9</td>
<td>1</td>
</tr>

<tr>
<td>-invarfrac</td>
<td>Fraction of invariant sites</td>
<td>Number from 0.000 to 1.000</td>
<td>0.0</td>
</tr>

<tr>
<td>-nhmmcategories</td>
<td>Number of HMM rate categories</td>
<td>Integer from 1 to 9</td>
<td>1</td>
</tr>

<tr>
<td>-hmmrates</td>
<td>HMM category rates</td>
<td>List of floating point numbers</td>
<td>1.0</td>
</tr>

<tr>
<td>-hmmprobabilities</td>
<td>Probability for each HMM category</td>
<td>List of floating point numbers</td>
<td>1.0</td>
</tr>

<tr>
<td>-adjsite</td>
<td>Rates at adjacent sites correlated</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-lambda</td>
<td>Mean block length of sites having the same rate</td>
<td>Number 1.000 or more</td>
<td>1.0</td>
</tr>

<tr>
<td>-njumble</td>
<td>Number of times to randomise</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr>
<td>-seed</td>
<td>Random number seed between 1 and 32767 (must be odd)</td>
<td>Integer 1 or more</td>
<td>0</td>
</tr>

<tr>
<td>-global</td>
<td>Global rearrangements</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-outgrno</td>
<td>Species number to use as outgroup</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr>
<td>-[no]trout</td>
<td>Write out trees to tree file</td>
<td>Toggle value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-outtreefile</td>
<td>Tree file name</td>
<td>Output file</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>-printdata</td>
<td>Print data at start of run</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]progress</td>
<td>Print indications of progress of run</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]treeprint</td>
<td>Print out tree</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-hypstate</td>
<td>Reconstruct hypothetical sequence</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>fpromlk</b> reads any normal sequence USAs.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: promlk.dat</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
   5   13
Alpha     AACGTGGCCAAAT
Beta      AAGGTCGCCAAAC
Gamma     CATTTCGTCACAA
Delta     GGTATTTCGGCCT
Epsilon   GGGATCTCGGCCC
</pre>
</td></tr></table><p>






<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 

        If you wish to include the standard description of the avalable
	report formats, use:
[an error occurred while processing this directive]
   -->

<b>fpromlk</b> output starts by giving the number of species, the
number of amino acid positions.

<p>
If the R (HMM rates) option is used a table of the relative rates of
expected substitution at each category of positions is printed, as
well as the probabilities of each of those rates.

<p>
There then follow the data sequences, if the user has selected the
menu option to print them out, with the base sequences printed in
groups of ten amino acids. The trees found are printed as a rooted
tree topology. The internal nodes are numbered arbitrarily for the
sake of identification. The number of trees evaluated so far and the
log likelihood of the tree are also given. The branch lengths in the
diagram are roughly proportional to the estimated branch lengths,
except that very short branches are printed out at least three
characters in length so that the connections can be seen. The unit of
branch length is the expected fraction of amino acids changed (so that
1.0 is 100 PAMs).

<p>
A table is printed showing the length of each tree segment, and the
time (in units of expected amino acid substitutions per position) of
each fork in the tree, measured from the root of the tree. I have not
attempted in include code for approximate confidence limits on branch
points, as I have done for branch lengths in PROML, both because of
the extreme crudeness of that test, and because the variation of times
for different forks would be highly correlated.

<p>
The log likelihood printed out with the final tree can be used to
perform various likelihood ratio tests. One can, for example, compare
runs with different values of the relative rate of change in the
active site and in the rest of the protein to determine which value is
the maximum likelihood estimate, and what is the allowable range of
values (using a likelihood ratio test, which you will find described
in mathematical statistics books). One could also estimate the base
frequencies in the same way. Both of these, particularly the latter,
require multiple runs of the program to evaluate different possibl
values, and this might get expensive.

<p>
This program makes possible a (reasonably) legitimate statistical test
of the molecular clock. To do such a test, run PROML and PROMLK on the
same data. If the trees obtained are of the same topology (when
considered as unrooted), it is legitimate to compare their likelihoods
by the likelihood ratio test. In PROML the likelihood has been
computed by estimating 2n-3 branch lengths, if their are n tips on the
tree. In PROMLK it has been computed by estimating n-1 branching times
(in effect, n-1 branch lengths). The difference in the number of
parameters is (2n-3)-(n-1) = n-2. To perform the test take the
difference in log likelihoods between the two runs (PROML should be
the higher of the two, barring numerical iteration difficulties) and
double it. Look this up on a chi-square distribution with n-2 degrees
of freedom. If the result is significant, the log likelihood has been
significantly increased by allowing all 2n-3 branch lengths to be
estimated instead of just n-1, and molecular clock may be rejected.

<p>
If the U (User Tree) option is used and more than one tree is
supplied, and the program is not told to assume autocorrelation
between the rates at different amino acid positions, the program also
performs a statistical test of each of these trees against the one
with highest likelihood. If there are two user trees, the test done is
one which is due to Kishino and Hasegawa (1989), a version of a test
originally introduced by Templeton (1983). In this implementation it
uses the mean and variance of log-likelihood differences between
trees, taken across amino acid positions. If the two trees' means are
more than 1.96 standard deviations different then the trees are
declared significantly different. This use of the empirical variance
of log-likelihood differences is more robust and nonparametric than
the classical likelihood ratio test, and may to some extent compensate
for the any lack of realism in the model underlying this program.

<p>
If there are more than two trees, the test done is an extension of the
KHT test, due to Shimodaira and Hasegawa (1999). They pointed out that
a correction for the number of trees was necessary, and they
introduced a resampling method to make this correction. In the version
used here the variances and covariances of the sum of log likelihoods
across amino acid positions are computed for all pairs of trees. To
test whether the difference between each tree and the best one is
larger than could have been expected if they all had the same expected
log-likelihood, log-likelihoods for all trees are sampled with these
covariances and equal means (Shimodaira and Hasegawa's "least
favorable hypothesis"), and a P value is computed from the fraction of
times the difference between the tree's value and the highest
log-likelihood exceeds that actually observed. Note that this sampling
needs random numbers, and so the program will prompt the user for a
random number seed if one has not already been supplied. With the
two-tree KHT test no random numbers are used.

<p>
In either the KHT or the SH test the program prints out a table of the
log-likelihoods of each tree, the differences of each from the highest
one, the variance of that quantity as determined by the log-likelihood
differences at individual sites, and a conclusion as to whether that
tree is or is not significantly worse than the best one. However the
test is not available if we assume that there is autocorrelation of
rates at neighboring positions (option A) and is not done in those
cases.

<p>
The branch lengths printed out are scaled in terms of 100 times the
expected numbers of amino acid substitutions, scaled so that the
average rate of change, averaged over all the positions analyzed, is
set to 100.0, if there are multiple categories of positions. This
means that whether or not there are multiple categories of positions,
the expected percentage of change for very small branches is equal to
the branch length. Of course, when a branch is twice as long this does
not mean that there will be twice as much net change expected along
it, since some of the changes occur in the same position and overlie
or even reverse each other. underlying numbers of changes. That means
that a branch of length 26 is 26 times as long as one which would show
a 1% difference between the amino acid sequences at the beginning and
end of the branch, but we would not expect the sequences at the
beginning and end of the branch to be 26% different, as there would be
some overlaying of changes.

<p>
Because of limitations of the numerical algorithm, branch length
estimates of zero will often print out as small numbers such as
0.00001. If you see a branch length that small, it is really estimated
to be of zero length.

<p>
Another possible source of confusion is the existence of negative
values for the log likelihood. This is not really a problem; the log
likelihood is not a probability but the logarithm of a
probability. When it is negative it simply means that the
corresponding probability is less than one (since we are seeing its
logarithm). The log likelihood is maximized by being made more
positive: -30.23 is worse than -29.14.

<p>
At the end of the output, if the R option is in effect with multiple
HMM rates, the program will print a list of what amino acid position
categories contributed the most to the final likelihood. This
combination of HMM rate categories need not have contributed a
majority of the likelihood, just a plurality. Still, it will be
helpful as a view of where the program infers that the higher and
lower rates are. Note that the use in this calculations of the prior
probabilities of different rates, and the average patch length, gives
this inference a "smoothed" appearance: some other combination of
rates might make a greater contribution to the likelihood, but be
discounted because it conflicts with this prior information. See the
example output below to see what this printout of rate categories
looks like. A second list will also be printed out, showing for each
position which rate accounted for the highest fraction of the
likelihood. If the fraction of the likelihood accounted for is less
than 95%, a dot is printed instead.

<p>
Option 3 in the menu controls whether the tree is printed out into the
output file. This is on by default, and usually you will want to leave
it this way. However for runs with multiple data sets such as
bootstrapping runs, you will primarily be interested in the trees
which are written onto the output tree file, rather than the trees
printed on the output file. To keep the output file from becoming too
large, it may be wisest to use option 3 to prevent trees being printed
onto the output file.

<p>
Option 4 in the menu controls whether the tree estimated by the
program is written onto a tree file. The default name of this output
tree file is "outtree". If the U option is in effect, all the
user-defined trees are written to the output tree file.

<p>
Option 5 in the menu controls whether ancestral states are estimated
at each node in the tree. If it is in effect, a table of ancestral
sequences is printed out (including the sequences in the tip species
which are the input sequences). The symbol printed out is for the
amino acid which accounts for the largest fraction of the likelihood
at that position. In that table, if a position has an amino acid which
accounts for more than 95% of the likelihood, its symbol printed in
capital letters (W rather than w). One limitation of the current
version of the program is that when there are multiple HMM rates
(option R) the reconstructed amino acids are based on only the single
assignment of rates to positions which accounts for the largest amount
of the likelihood. Thus the assessment of 95% of the likelihood, in
tabulating the ancestral states, refers to 95% of the likelihood that
is accounted for by that particular combination of rates.

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: promlk.fpromlk</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>

Amino acid sequence
   Maximum Likelihood method with molecular clock, version 3.6b

Jones-Taylor-Thornton model of amino acid change





                                         +-----------Epsilon   
  +--------------------------------------4  
  !                                      +-----------Delta     
--3  
  !                     +----------------------------Gamma     
  +---------------------2  
                        !                   +--------Beta      
                        +-------------------1  
                                            +--------Alpha     


Ln Likelihood =  -134.75201

 Ancestor      Node      Node Height     Length
 --------      ----      ---- ------     ------
 root            3      
   3             4          0.66099      0.66099
   4          Epsilon       0.85359      0.19260
   4          Delta         0.85359      0.19260
   3             2          0.36919      0.36919
   2          Gamma         0.85359      0.48440
   2             1          0.69651      0.32732
   1          Beta          0.85359      0.15708
   1          Alpha         0.85359      0.15708




</pre>
</td></tr></table><p>
<p><h3>File: promlk.treefile</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
((Epsilon:0.19260,Delta:0.19260):0.66099,(Gamma:0.48440,
(Beta:0.15708,Alpha:0.15708):0.32732):0.36919);
</pre>
</td></tr></table><p>






<H2>
    Data files
</H2>

None


<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="../phylip/ednacomp.html">ednacomp</a></td><td>DNA compatibility algorithm</td></tr>
<tr><td><a href="../phylip/ednadist.html">ednadist</a></td><td>Nucleic acid sequence Distance Matrix program</td></tr>
<tr><td><a href="../phylip/ednainvar.html">ednainvar</a></td><td>Nucleic acid sequence Invariants method</td></tr>
<tr><td><a href="../phylip/ednaml.html">ednaml</a></td><td>Estimates phylogenies from nucleic acid sequence Maximum Likelihood</td></tr>
<tr><td><a href="../phylip/ednamlk.html">ednamlk</a></td><td>Estimates phylogenies from nucleic acid sequence Maximum Likelihood with molecular clock</td></tr>
<tr><td><a href="../phylip/ednapars.html">ednapars</a></td><td>DNA parsimony algorithm</td></tr>
<tr><td><a href="../phylip/ednapenny.html">ednapenny</a></td><td>Penny algorithm for DNA</td></tr>
<tr><td><a href="../phylip/eprotdist.html">eprotdist</a></td><td>Protein distance algorithm</td></tr>
<tr><td><a href="../phylip/eprotpars.html">eprotpars</a></td><td>Protein parsimony algorithm</td></tr>
<tr><td><a href="../phylip/erestml.html">erestml</a></td><td>Restriction site Maximum Likelihood method</td></tr>
<tr><td><a href="../phylip/eseqboot.html">eseqboot</a></td><td>Bootstrapped sequences algorithm</td></tr>
<tr><td><a href="fdiscboot.html">fdiscboot</a></td><td>Bootstrapped discrete sites algorithm</td></tr>
<tr><td><a href="fdnacomp.html">fdnacomp</a></td><td>DNA compatibility algorithm</td></tr>
<tr><td><a href="fdnadist.html">fdnadist</a></td><td>Nucleic acid sequence Distance Matrix program</td></tr>
<tr><td><a href="fdnainvar.html">fdnainvar</a></td><td>Nucleic acid sequence Invariants method</td></tr>
<tr><td><a href="fdnaml.html">fdnaml</a></td><td>Estimates nucleotide phylogeny by maximum likelihood</td></tr>
<tr><td><a href="fdnamlk.html">fdnamlk</a></td><td>Estimates nucleotide phylogeny by maximum likelihood</td></tr>
<tr><td><a href="fdnamove.html">fdnamove</a></td><td>Interactive DNA parsimony</td></tr>
<tr><td><a href="fdnapars.html">fdnapars</a></td><td>DNA parsimony algorithm</td></tr>
<tr><td><a href="fdnapenny.html">fdnapenny</a></td><td>Penny algorithm for DNA</td></tr>
<tr><td><a href="fdolmove.html">fdolmove</a></td><td>Interactive Dollo or Polymorphism Parsimony</td></tr>
<tr><td><a href="ffreqboot.html">ffreqboot</a></td><td>Bootstrapped genetic frequencies algorithm</td></tr>
<tr><td><a href="fproml.html">fproml</a></td><td>Protein phylogeny by maximum likelihood</td></tr>
<tr><td><a href="fprotdist.html">fprotdist</a></td><td>Protein distance algorithm</td></tr>
<tr><td><a href="fprotpars.html">fprotpars</a></td><td>Protein pasimony algorithm</td></tr>
<tr><td><a href="frestboot.html">frestboot</a></td><td>Bootstrapped restriction sites algorithm</td></tr>
<tr><td><a href="frestdist.html">frestdist</a></td><td>Computes distance matrix from restriction sites or fragments</td></tr>
<tr><td><a href="frestml.html">frestml</a></td><td>Restriction site maximum Likelihood method</td></tr>
<tr><td><a href="fseqboot.html">fseqboot</a></td><td>Bootstrapped sequences algorithm</td></tr>
<tr><td><a href="fseqbootall.html">fseqbootall</a></td><td>Bootstrapped sequences algorithm</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>

This program is an EMBOSS conversion of a program written by Joe
Felsenstein as part of his PHYLIP package.

<p>
Although we take every care to ensure that the results of the EMBOSS
version are identical to those from the original package, we recommend
that you check your inputs give the same results in both versions
before publication.

<p>
Please report all bugs in the EMBOSS version to the EMBOSS bug team,
not to the original author.


<H2>
    History
</H2>

Written (2004) - Joe Felsenstein, University of Washington.
<p>
Converted (August 2004) to an EMBASSY program by the EMBOSS team.




<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












</BODY>
</HTML>

