<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: frestdist
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

frestdist
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<!--END OF HEADER-->






<H2>
    Function
</H2>
Computes distance matrix from restriction sites or fragments

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

Distances calculated from restriction sites data or restriction
fragments data. The restriction sites option is the one to use to also
make distances for RAPDs or AFLPs.

<H2>
    Algorithm
</H2>


Restdist reads the same restriction sites format as RESTML and
computes a restriction sites distance. It can also compute a
restriction fragments distance. The original restriction fragments and
restriction sites distance methods were introduced by Nei and Li
(1979). Their original method for restriction fragments is also
available in this program, although its default methods are my
modifications of the original Nei and Li methods.

<p>
These two distances assume that the restriction sites are accidental
byproducts of random change of nucleotide sequences. For my
restriction sites distance the DNA sequences are assumed to be
changing according to the Kimura 2-parameter model of DNA change
(Kimura, 1980). The user can set the transition/transversion rate for
the model. For my restriction fragments distance there is there is an
implicit assumption of a Jukes-Cantor (1969) model of change, The user
can also set the parameter of a correction for unequal rates of
evolution between sites in the DNA sequences, using a Gamma
distribution of rates among sites. The Jukes-Cantor model is also
implicit in the restriction fragments distance of Nei and Li(1979). It
does not allow us to correct for a Gamma distribution of rates among
sites.


<h3>Restriction Sites Distance</h3>

<p>
The restriction sites distances use data coded for the presence of
absence of individual restriction sites (usually as + and - or 0 and
1). My distance is based on the proportion, out of all sites observed
in one species or the other, which are present in both species. This
is done to correct for the ascertainment of sites, for the fact that
we are not aware of many sites because they do not appear in any
species.

<p>
My distance starts by computing from the particular pair of species
the fraction

<p>
<pre>
                 n++
   f =  ---------------------
         n++ + 1/2 (n+- + n-+)

</pre>
<p>
where n++ is the number of sites contained in both species, n+- is the
number of sites contained in the first of the two species but not in
the second, and n-+ is the number of sites contained in the second of
the two species but not in the first. This is the fraction of sites
that are present in one species which are present in both. Since the
number of sites present in the two species will often differ, the
denominator is the average of the number of sites found in the two
species.

<p>
If each restriction site is s nucleotides long, the probability that a
restriction site is present in the other species, given that it is
present in a species, is

<p>
<pre>
      Qs,
</pre>

<p>
`where Q is the probability that a nucleotide has no net change as one
goes from the one species to the other. It may have changed in
between; we are interested in the probability that that nucleotide
site is in the same base in both species, irrespective of what has
happened in between.  The distance is then computed by finding the
branch length of a two-species tree (connecting these two species with
a single branch) such that Q equals the s-th root of f. For this the
program computes Q for various values of branch length, iterating them
by a Newton-Raphson algorithm until the two quantities are equal.

<p>
The resulting distance should be numerically close to the original
restriction sites distance of Nei and Li (1979) when divergence is
small. Theirs computes the probability of retention of a site in a way
that assumes that the site is present in the common ancestor of the
two species. Ours does not make this assumption. It is inspired by
theirs, but differs in this detail. Their distance also assumes a
Jukes-Cantor (1969) model of base change, and does not allow for
transitions being more frequent than transversions. In this sense mine
generalizes theres somewhat. Their distance does include, as mine does
as well, a correction for Gamma distribution of rate of change among
nucleotide sites.

<p>
I have made their original distance available here 


<h3>Restriction Fragments Distance</h3>

<p>
For restriction fragments data we use a different distance. If we
average over all restriction fragment lengths, each at its own
expected frequency, the probability that the fragment will still be in
existence after a certain amount of branch length, we must take into
account the probability that the two restriction sites at the ends of
the fragment do not mutate, and the probability that no new
restriction site occurs within the fragment in that amount of branch
length. The result for a restriction site length of s is:

<p>
<pre>
                Q2s
          f = --------
               2 - Qs

</pre>
<p>
(The details of the derivation will be given in my forthcoming book
Inferring Phylogenies (to be published by Sinauer Associates in
2001). Given the observed fraction of restriction sites retained, f,
we can solve a quadratic equation from the above expression for
Qs. That makes it easy to obtain a value of Q, and the branch length
can then be estimated by adjusting it so the probability of a base not
changing is equal to that value.  Alternatively, if we use the Nei and
Li (1979) restriction fragments distance, this involves solving for g
in the nonlinear equation

<p>
<pre>
       g  =  [ f (3 - 2g) ]1/4
</pre>
<p>
and then the distance is given by 

<p>
<pre>
       d  =  - (2/r) loge(g)
</pre>
<p>
where r is the length of the restriction site. 

<p>
Comparing these two restriction fragments distances in a case where
their underlying DNA model is the same (which is when the
transition/transversion ratio of the modified model is set to 0.5),
you will find that they are very close to each other, differing very
little at small distances, with the modified distance become smaller
than the Nei/Li distance at larger distances. It will therefore matter
very little which one you use.


<h3>A Comment About RAPDs and AFLPs</h3>

<p>
Although these distances are designed for restriction sites and
restriction fragments data, they can be applied to RAPD and AFLP data
as well. RAPD (Randomly Amplified Polymorphic DNA) and AFLP (Amplified
Fragment Length Polymorphism) data consist of presence or absence of
individual bands on a gel. The bands are segments of DNA with PCR
primers at each end. These primers are defined sequences of known
length (often about 10 nucleotides each). For AFLPs the reolevant
length is the primer length, plus three nucleotides. Mutation in these
sequences makes them no longer be primers, just as in the case of
restriction sites. Thus a pair of 10-nucleotide primers will behave
much the same as a 20-nucleotide restriction site, for RAPDs (26 for
AFLPs). You can use the restriction sites distance as the distance
between RAPD or AFLP patterns if you set the proper value for the
total length of the site to the total length of the primers (plus 6 in
the case of AFLPs). Of course there are many possible sources of noise
in these data, including confusing fragments of similar length for
each other and having primers near each other in the genome, and these
are not taken into account in the statistical model used here.

<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

<b>Here is a sample session with frestdist</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>frestdist </b>
Computes distance matrix from restriction sites or fragments
Input file: <b>restdist.dat</b>
Output file [restdist.frestdist]: <b></b>


Restriction site or fragment distances, version 3.6b

Distances calculated for species
    Alpha        ....
    Beta         ...
    Gamma        ..
    Delta        .
    Epsilon   

Distances written to file "restdist.frestdist"

Done.


</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>




<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-data]              discretestates File containing one or more sets of
                                  restriction data
  [-outfile]           outfile    Output file name

   Additional (Optional) qualifiers (* if not always prompted):
   -[no]restsites      boolean    Restriction sites (put N if you want
                                  restriction fragments)
   -neili              boolean    Use original Nei/Li model (default uses
                                  modified Nei/Li model)
*  -gamma              boolean    Gama distributed rates among sites
*  -gammacoefficient   float      Coefficient of variation of substitution
                                  rate among sites
   -ttratio            float      Transition/transversion ratio
   -sitelength         integer    Site length
   -lower              boolean    Lower triangular distance matrix
   -printdata          boolean    Print data at start of run
   -[no]progress       boolean    Print indications of progress of run

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-outfile" associated qualifiers
   -odirectory2        string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report deaths


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-data]<br>(Parameter 1)</td>
<td>File containing one or more sets of restriction data</td>
<td>Discrete states file</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.frestdist</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]restsites</td>
<td>Restriction sites (put N if you want restriction fragments)</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-neili</td>
<td>Use original Nei/Li model (default uses modified Nei/Li model)</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-gamma</td>
<td>Gama distributed rates among sites</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-gammacoefficient</td>
<td>Coefficient of variation of substitution rate among sites</td>
<td>Number 0.001 or more</td>
<td>1</td>
</tr>

<tr>
<td>-ttratio</td>
<td>Transition/transversion ratio</td>
<td>Number 0.001 or more</td>
<td>2.0</td>
</tr>

<tr>
<td>-sitelength</td>
<td>Site length</td>
<td>Integer 1 or more</td>
<td>6</td>
</tr>

<tr>
<td>-lower</td>
<td>Lower triangular distance matrix</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-printdata</td>
<td>Print data at start of run</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]progress</td>
<td>Print indications of progress of run</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>frestdist</b> input is fairly standard, with one addition. As usual
the first line of the file gives the number of species and the number
of sites, but there is also a third number, which is the number of
different restriction enzymes that were used to detect the restriction
sites. Thus a data set with 10 species and 35 different sites,
representing digestion with 4 different enzymes, would have the first
line of the data file look like this:

<p>

<pre>
   10   35    4

</pre>

<p>
The site data are in standard form. Each species starts with a species
name whose maximum length is given by the constant "nmlngth" (whose
value in the program as distributed is 10 characters). The name
should, as usual, be padded out to that length with blanks if
necessary. The sites data then follows, one character per site (any
blanks will be skipped and ignored). Like the DNA and protein sequence
data, the restriction sites data may be either in the "interleaved"
form or the "sequential" form. Note that if you are analyzing
restriction sites data with the programs DOLLOP or MIX or other
discrete character programs, at the moment those programs do not use
the "aligned" or "interleaved" data format. Therefore you may want to
avoid that format when you have restriction sites data that you will
want to feed into those programs.

<p>
The presence of a site is indicated by a "+" and the absence by a
"-". I have also allowed the use of "1" and "0" as synonyms for "+"
and "-", for compatibility with MIX and DOLLOP which do not allow "+"
and "-". If the presence of the site is unknown (for example, if the
DNA containing it has been deleted so that one does not know whether
it would have contained the site) then the state "?" can be used to
indicate that the state of this site is unknown.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: restdist.dat</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
   5   13   2
Alpha     ++-+-++--+++-
Beta      ++++--+--+++-
Gamma     -+--+-++-+-++
Delta     ++-+----++---
Epsilon   ++++----++---
</pre>
</td></tr></table><p>






<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 

        If you wish to include the standard description of the avalable
	report formats, use:
[an error occurred while processing this directive]
   -->

<b>frestdist</b> output contains on its first line the number of
species. The distance matrix is then printed in standard form, with
each species starting on a new line with the species name, followed by
the distances to the species in order. These continue onto a new line
after every nine distances. If the L option is used, the matrix or
distances is in lower triangular form, so that only the distances to
the other species that precede each species are printed. Otherwise the
distance matrix is square with zero distances on the diagonal. In
general the format of the distance matrix is such that it can serve as
input to any of the distance matrix programs.

<p>
If the option to print out the data is selected, the output file will
precede the data by more complete information on the input and the
menu selections. The output file begins by giving the number of
species and the number of characters.

<p>
The distances printed out are scaled in terms of expected numbers of
substitutions per DNA site, counting both transitions and
transversions but not replacements of a base by itself, and scaled so
that the average rate of change, averaged over all sites analyzed, is
set to 1.0. Thus when the G option is used, the rate of change at one
site may be higher than at another, but their mean is expected to be
1.

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: restdist.frestdist</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
    5
Alpha       0.000000  0.022368  0.107681  0.068767  0.082639
Beta        0.022368  0.000000  0.107681  0.068767  0.044216
Gamma       0.107681  0.107681  0.000000  0.176484  0.192466
Delta       0.068767  0.068767  0.176484  0.000000  0.019724
Epsilon     0.082639  0.044216  0.192466  0.019724  0.000000
</pre>
</td></tr></table><p>






<H2>
    Data files
</H2>

None


<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="../phylip/ednacomp.html">ednacomp</a></td><td>DNA compatibility algorithm</td></tr>
<tr><td><a href="../phylip/ednadist.html">ednadist</a></td><td>Nucleic acid sequence Distance Matrix program</td></tr>
<tr><td><a href="../phylip/ednainvar.html">ednainvar</a></td><td>Nucleic acid sequence Invariants method</td></tr>
<tr><td><a href="../phylip/ednaml.html">ednaml</a></td><td>Estimates phylogenies from nucleic acid sequence Maximum Likelihood</td></tr>
<tr><td><a href="../phylip/ednamlk.html">ednamlk</a></td><td>Estimates phylogenies from nucleic acid sequence Maximum Likelihood with molecular clock</td></tr>
<tr><td><a href="../phylip/ednapars.html">ednapars</a></td><td>DNA parsimony algorithm</td></tr>
<tr><td><a href="../phylip/ednapenny.html">ednapenny</a></td><td>Penny algorithm for DNA</td></tr>
<tr><td><a href="../phylip/eprotdist.html">eprotdist</a></td><td>Protein distance algorithm</td></tr>
<tr><td><a href="../phylip/eprotpars.html">eprotpars</a></td><td>Protein parsimony algorithm</td></tr>
<tr><td><a href="../phylip/erestml.html">erestml</a></td><td>Restriction site Maximum Likelihood method</td></tr>
<tr><td><a href="../phylip/eseqboot.html">eseqboot</a></td><td>Bootstrapped sequences algorithm</td></tr>
<tr><td><a href="fdiscboot.html">fdiscboot</a></td><td>Bootstrapped discrete sites algorithm</td></tr>
<tr><td><a href="fdnacomp.html">fdnacomp</a></td><td>DNA compatibility algorithm</td></tr>
<tr><td><a href="fdnadist.html">fdnadist</a></td><td>Nucleic acid sequence Distance Matrix program</td></tr>
<tr><td><a href="fdnainvar.html">fdnainvar</a></td><td>Nucleic acid sequence Invariants method</td></tr>
<tr><td><a href="fdnaml.html">fdnaml</a></td><td>Estimates nucleotide phylogeny by maximum likelihood</td></tr>
<tr><td><a href="fdnamlk.html">fdnamlk</a></td><td>Estimates nucleotide phylogeny by maximum likelihood</td></tr>
<tr><td><a href="fdnamove.html">fdnamove</a></td><td>Interactive DNA parsimony</td></tr>
<tr><td><a href="fdnapars.html">fdnapars</a></td><td>DNA parsimony algorithm</td></tr>
<tr><td><a href="fdnapenny.html">fdnapenny</a></td><td>Penny algorithm for DNA</td></tr>
<tr><td><a href="fdolmove.html">fdolmove</a></td><td>Interactive Dollo or Polymorphism Parsimony</td></tr>
<tr><td><a href="ffreqboot.html">ffreqboot</a></td><td>Bootstrapped genetic frequencies algorithm</td></tr>
<tr><td><a href="fproml.html">fproml</a></td><td>Protein phylogeny by maximum likelihood</td></tr>
<tr><td><a href="fpromlk.html">fpromlk</a></td><td>Protein phylogeny by maximum likelihood</td></tr>
<tr><td><a href="fprotdist.html">fprotdist</a></td><td>Protein distance algorithm</td></tr>
<tr><td><a href="fprotpars.html">fprotpars</a></td><td>Protein pasimony algorithm</td></tr>
<tr><td><a href="frestboot.html">frestboot</a></td><td>Bootstrapped restriction sites algorithm</td></tr>
<tr><td><a href="frestml.html">frestml</a></td><td>Restriction site maximum Likelihood method</td></tr>
<tr><td><a href="fseqboot.html">fseqboot</a></td><td>Bootstrapped sequences algorithm</td></tr>
<tr><td><a href="fseqbootall.html">fseqbootall</a></td><td>Bootstrapped sequences algorithm</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>

This program is an EMBOSS conversion of a program written by Joe
Felsenstein as part of his PHYLIP package.

<p>
Although we take every care to ensure that the results of the EMBOSS
version are identical to those from the original package, we recommend
that you check your inputs give the same results in both versions
before publication.

<p>
Please report all bugs in the EMBOSS version to the EMBOSS bug team,
not to the original author.


<H2>
    History
</H2>

Written (2004) - Joe Felsenstein, University of Washington.
<p>
Converted (August 2004) to an EMBASSY program by the EMBOSS team.




<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












</BODY>
</HTML>

