<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: fretree
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

fretree
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<!--END OF HEADER-->






<H2>
    Function
</H2>
Interactive tree rearrangement

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

Reads in a tree (with branch lengths if necessary) and allows you to
reroot the tree, to flip branches, to change species names and branch
lengths, and then write the result out. Can be used to convert between
rooted and unrooted trees, and to write the tree into a preliminary
version of a new XML tree file format which is under development and
which is described in the RETREE documentation web page.


<H2>
    Algorithm
</H2>


RETREE is a tree editor. It reads in a tree, or allows the user to
construct one, and displays this tree on the screen. The user then can
specify how the tree is to be rearranged, rerooted or written out to a
file.

<p>
The input trees are in one file (with default file name intree), the
output trees are written into another (outtree). The user can reroot,
flip branches, change names of species, change or remove branch
lengths, and move around to look at various parts of the tree if it is
too large to fit on the screen. The trees can be multifurcating at any
level, although the user is warned that many PHYLIP programs still
cannot handle multifurcations above the root, or even at the root.

<p>
A major use for this program will be to change rootedness of trees so
that a rooted tree derived from one program can be fed in as an
unrooted tree to another (you are asked about this when you give the
command to write out the tree onto the tree output file). It will also
be useful for specifying the length of a branch in a tree where you
want a program like DNAML, DNAMLK, FITCH, or CONTML to hold that
branch length constant (see the L suboption of the User Tree option in
those programs. It will also be useful for changing the order of
species for purely cosmetic reasons for DRAWGRAM and DRAWTREE,
including using the Midpoint method of rooting the tree. It can also
be used to write out a tree file in the Nexus format used by Paup and
MacClade or in our XML tree file format.

<p>
This program uses graphic characters that show the tree to best
advantage on some computer systems. Its graphic characters will work
best on MSDOS systems or MSDOS windows in Windows, and to any system
whose screen or terminals emulate ANSI standard terminals such as old
Digitial VT100 terminals, Telnet programs, or VT100-compatible windows
in the X windowing system. For any other screen types, (such as
Macintosh windows) there is a generic option which does not make use
of screen graphics characters. The program will work well in those
cases, but the tree it displays will look a bit uglier.



<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

<b>Here is a sample session with fretree</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>fretree </b>
Interactive tree rearrangement
Number of species [0]: <b>10</b>
Input tree file: <b>retree.dat</b>


Reading tree file ...



                                      ,>>1:Human
                                   ,>22  
                                ,>21  `>>2:Chimp
                                !  !  
                             ,>20  `>>>>>3:Gorilla
                             !  !  
                 ,>>>>>>>>>>19  `>>>>>>>>4:Orang
                 !           !  
              ,>18           `>>>>>>>>>>>5:Gibbon
              !  !  
              !  !              ,>>>>>>>>6:Barbary Ma
              !  `>>>>>>>>>>>>>23  
              !                 !  ,>>>>>7:Crab-e. Ma
     ,>>>>>>>17                 `>24  
     !        !                    !  ,>>8:Rhesus Mac
     !        !                    `>25  
     !        !                       `>>9:Jpn Macaq
  ,>16        !  
  !  !        `>>>>>>>>>>>>>>>>>>>>>>>>>10:Squir. Mon
  !  !  
  !  !                                ,>11:Tarsier
** 7 lines below screen **

NEXT? (Options: R . U W O T F D B N H J K L C + ? X Q) (? for Help) Do you want to write out the tree to a file? (Y or N) Enter R if the tree is to be rooted
OR enter U if the tree is to be unrooted: 
Tree written to file "retree.treefile"


</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>




<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-spp]               integer    Number of species
  [-intreefile]        tree       (no help text) tree value

   Additional (Optional) qualifiers:
   -initialtree        menu       Initial tree
   -format             menu       Format to write trees
   -screenwidth        integer    Width of terminal screen in characters
   -vscreenwidth       integer    Width of plotting area in characters
   -screenlines        integer    Number of lines on screen

   Advanced (Unprompted) qualifiers:
   -outtreefile        outfile    Output file name

   Associated qualifiers:

   "-outtreefile" associated qualifiers
   -odirectory         string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report deaths


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-spp]<br>(Parameter 1)</td>
<td>Number of species</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr>
<td>[-intreefile]<br>(Parameter 2)</td>
<td>(no help text) tree value</td>
<td>Phylogenetic tree</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-initialtree</td>
<td>Initial tree</td>
<td><table><tr><td>a</td> <td><i>(Arbitary)</i></td></tr><tr><td>u</td> <td><i>(User)</i></td></tr><tr><td>s</td> <td><i>(Specify)</i></td></tr></table></td>
<td>Arbitary</td>
</tr>

<tr>
<td>-format</td>
<td>Format to write trees</td>
<td><table><tr><td>p</td> <td><i>(PHYLIP)</i></td></tr><tr><td>n</td> <td><i>(NEXUS)</i></td></tr><tr><td>x</td> <td><i>(XML)</i></td></tr></table></td>
<td>p</td>
</tr>

<tr>
<td>-screenwidth</td>
<td>Width of terminal screen in characters</td>
<td>Any integer value</td>
<td>80</td>
</tr>

<tr>
<td>-vscreenwidth</td>
<td>Width of plotting area in characters</td>
<td>Any integer value</td>
<td>80</td>
</tr>

<tr>
<td>-screenlines</td>
<td>Number of lines on screen</td>
<td>Any integer value</td>
<td>24</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-outtreefile</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.fretree</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>fretree</b> reads any normal sequence USAs.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: retree.dat</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
((((((((Human,Chimp),Gorilla),Orang),Gibbon),(Barbary_Ma,(Crab-e._Ma,
(Rhesus_Mac,Jpn_Macaq)))),Squir._Mon),((Tarsier,Lemur),Bovine)),Mouse);
</pre>
</td></tr></table><p>






<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 

        If you wish to include the standard description of the avalable
	report formats, use:
[an error occurred while processing this directive]
   -->

The N (output file format) option allows the user to
specify that the tree files that are written by the program will be in
one of three formats:


<ol>


<li> 
The PHYLIP default file format (the Newick standard) used by the
programs in this package.

<li> The Nexus format defined by David Swofford and by Wayne Maddison
and David Maddison for their programs PAUP and MacClade. A tree file
written in Nexus format should be directly readable by those programs
(They also have options to read a regular PHYLIP tree file as well).

<li> 
An XML tree file format which we have defined. 

</ol> 

<p>
The XML tree file format is fairly simple. The tree file, which may
have multiple trees, is enclosed in a pair of &lt;PHYLOGENIES&gt;
... &lt;/PHYLOGENIES&gt; tags. Each tree is included in tags &lt;PHYLOGENY&gt;
... &lt;/PHYLOGENY&gt;. Each branch of the tree is enclosed in a pair of
tags &lt;CLADE&gt; ... &lt;/CLADE&gt;, which enclose the branch and all its
descendants. If the branch has a length, this is given by the LENGTH
attribute of the CLADE tag, so that the pair of tags looks like this:


&lt;CLADE LENGTH="0.09362"&gt; ... &lt;/CLADE&gt;

<p>
A tip of the tree is at the end of a branch (and hence that branch is enclosed in a pair of &lt;CLADE&gt; ... &lt;/CLADE&gt; tags). Its name is enclosed by &lt;NAME&gt; ... &lt;/NAME&gt; tags. Here is an XML tree: 

<p>
<pre>
&lt;phylogenies&gt;
  &lt;phylogeny&gt;
    &lt;clade&gt;
      &lt;clade length="0.87231"&gt;&lt;name&gt;Mouse&lt;/name&gt;&lt;/clade&gt;
      &lt;clade length="0.49807"&gt;&lt;name&gt;Bovine&lt;/name&gt;&lt;/clade&gt;
      &lt;clade length="0.39538"&gt;
        &lt;clade length="0.25930"&gt;&lt;name&gt;Gibbon&lt;/name&gt;&lt;/clade&gt;
        &lt;clade length="0.10815"&gt;
          &lt;clade length="0.24166"&gt;&lt;name&gt;Orang&lt;/name&gt;&lt;/clade&gt;
          &lt;clade length="0.04405"&gt;
            &lt;clade length="0.12322"&gt;&lt;name&gt;Gorilla&lt;/name&gt;&lt;/clade&gt;
            &lt;clade length="0.06026"&gt;
              &lt;clade length="0.13846"&gt;&lt;name&gt;Chimp&lt;/name&gt;&lt;/clade&gt;
              &lt;clade length="0.0857"&gt;&lt;name&gt;Human&lt;/name&gt;&lt;/clade&gt;
            &lt;/clade&gt;
          &lt;/clade&gt;
        &lt;/clade&gt;
      &lt;/clade&gt;
    &lt;/clade&gt;
  &lt;/phylogeny&gt;
&lt;/phylogenies&gt;
  
</pre>


<p>
The indentation is for readability but is not part of the XML tree
standard, which ignores that kind of white space.

<p>
What programs can read an XML tree? None right now, not even PHYLIP
programs! But soon our lab's LAMARC package will have programs that
can read an XML tree. XML is rapidly becoming the standard for
representing and interchanging complex data -- it is time to have an
XML tree standard. Certain extensions are obvious (to represent the
bootstrap proportion for a branch, use BOOTP=0.83 in the CLADE tag,
for example).

<p>
There are other proposals for an XML tree standard. They have many
similarities to this one, but are not identical to it. At the moment
there is no mechanism in place for deciding between them other than
seeing which get widely used. Here are links to other proposals:


<table>

<tr>
<td>Taxonomic Markup Language  </td>
<td>http://www.albany.edu/~gilmr/pubxml/.
and preprint at
xml.coverpages.org/gilmour-TML.pdf  </td>
<td>published in the paper by
Ron Gilmour (2000).  </td>
</tr>

<tr>
<td>Andrew Rambaut's
BEAST XML phylogeny format  </td>
<td>See page 9 of PDF of BEAST manual at
http://evolve.zoo.ox.ac.uk/beast/  </td>
<td>An XML format for phylogenies is sketchly described there.</td>
</tr>
<tr>  
<td>treeml  </td>
<td>http://www.nomencurator.org/InfoVis2003/download/treeml.dtd
(see also example: )
http://www.cs.umd.edu/hcil/iv03contest/datasets/treeml-sample.xml</td>
<td>  Jean-Daniel Fekete's DTD
for a tree XML file </td>
</tr>
</table>


<p>
The W (screen and window Width) option specifies the width in
characters of the area which the trees will be plotted to fit
into. This is by default 80 characters so that they will fit on a
normal width terminal. The actual width of the display on the terminal
(normally 80 characters) will be regarded as a window displaying part
of the tree. Thus you could set the "plotting area" to 132 characters,
and inform the program that the screen width is 80 characters. Then
the program will display only part of the tree at any one time.


<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: retree.treefile</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
(((((((Human,Chimp),Gorilla),Orang),Gibbon),(Barbary_Ma,(Crab-e._Ma,
(Rhesus_Mac,Jpn_Macaq)))),Squir._Mon),((Tarsier,Lemur),Bovine),Mouse);
</pre>
</td></tr></table><p>






<H2>
    Data files
</H2>

None


<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="fdrawgram.html">fdrawgram</a></td><td>Plots a cladogram- or phenogram-like rooted tree diagram</td></tr>
<tr><td><a href="fdrawtree.html">fdrawtree</a></td><td>Plots an unrooted tree diagram</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>

This program is an EMBOSS conversion of a program written by Joe
Felsenstein as part of his PHYLIP package.

<p>
Although we take every care to ensure that the results of the EMBOSS
version are identical to those from the original package, we recommend
that you check your inputs give the same results in both versions
before publication.

<p>
Please report all bugs in the EMBOSS version to the EMBOSS bug team,
not to the original author.


<H2>
    History
</H2>

Written (2004) - Joe Felsenstein, University of Washington.
<p>
Converted (August 2004) to an EMBASSY program by the EMBOSS team.




<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












</BODY>
</HTML>

