<HTML>

<HEAD>
  <TITLE>
  EMBOSS: fuzzpro
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: fuzzpro
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program fuzzpro
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Protein pattern search


<H2>
    Description
</H2>


fuzzpro uses PROSITE style patterns to search protein sequences. 

<p>
Patterns are specifications of a (typically short) length of sequence to
be found.  They can specify a search for an exact sequence or they can
allow various ambiguities, matches to variable lengths of sequence and
repeated subsections of the sequence. 

<p>
fuzzpro intelligently selects the optimum searching algorithm to use,
depending on the complexity of the search pattern specified. 

<H2>
    Usage
</H2>


Here is a sample session with fuzzpro.

<pre>

% fuzzpro
Input sequence: sw:*
Search pattern: [FY]-[LIV]-G-[DE]-E-A-Q-x-[RKQ](2)-G
Number of mismatches [0]: 
Output file [5h1d_fugru.fuzzpro]:

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
   -pattern            string     Search pattern
   -mismatch           integer    Number of mismatches
  [-outf]              outfile    Output file name

   Optional qualifiers: (none)
   Advanced qualifiers:
   -mmshow             bool       Show mismatches
   -accshow            bool       Show accession numbers
   -usashow            bool       Showing the USA (Uniform Sequence Address)
                                  of the matching sequences will turn your
                                  output file into a 'list' file that can then
                                  be read in by many other EMBOSS programs by
                                  specifying it with a '@' in front of the
                                  filename.
   -descshow           bool       Show descriptions

   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-pattern</td>
<td>Search pattern</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-mismatch</td>
<td>Number of mismatches</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr>
<td>[-outf]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.fuzzpro</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-mmshow</td>
<td>Show mismatches</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-accshow</td>
<td>Show accession numbers</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-usashow</td>
<td>Showing the USA (Uniform Sequence Address) of the matching sequences will turn your output file into a 'list' file that can then be read in by many other EMBOSS programs by specifying it with a '@' in front of the filename.</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-descshow</td>
<td>Show descriptions</td>
<td>Yes/No</td>
<td>No</td>
</tr>

</table>



<H2>
    Input file format
</H2>


<p>
Patterns for fuzzpro are based on the format of pattern used in the
PROSITE database, with the difference that the terminating dot '.' and
the hyphens, '-', between the characters are optional. 

<p>
The PROSITE pattern definition from the PROSITE documentation follows. 

<p>
<ul>

<li>The standard IUPAC one-letter codes for the amino acids are used.

<li>The symbol `x' is used for a position where any amino acid is accepted.

<li>Ambiguities are indicated by listing the acceptable amino acids for a
given position, between square parentheses `[ ]'.  For example: [ALT]
stands for Ala or Leu or Thr. 

<li>Ambiguities are also indicated by listing between a pair of curly
brackets `{ }' the amino acids that are not accepted at a given
position.  For example: {AM} stands for any amino acid except Ala and
Met. 

<li>Each element in a pattern is separated from its neighbor by a `-'.
(Optional in fuzzpro). 

<li>Repetition of an element of the pattern can be indicated by following
that element with a numerical value or a numerical range between
parenthesis.  Examples: x(3) corresponds to x-x-x, x(2,4) corresponds to
x-x or x-x-x or x-x-x-x. 

<li>When a pattern is restricted to either the N- or C-terminal of a
sequence, that pattern either starts with a `&lt;' symbol or respectively
ends with a `&gt;' symbol. 

<li>A period ends the pattern.  (Optional in fuzzpro). 

</ul>

<p>
For example, in SWISSPROT entry 100K_RAT you can look for the pattern: 

<pre>
[DE](2)HS{P}X(2)PX(2,4)C
</pre>

<p>
This means: Two Asps or Glus in any order followed by His, Ser, any
residue other then Pro, then two of any residue followed by Pro followed
by two to four of any residue followed by Cys. 

<p>
The search is case-independent, so 'AAA' matches 'aaa'. 

<H2>
    Output file format
</H2>


Here is the output from the example search:

<p>
<pre>

     ACT1_FUGRU    53 YVGDEAQSKRG
     ACT2_FUGRU    53 YVGDEAQSKRG
     ACT3_FUGRU    53 YVGDEAQSKRG
     ACTC_FUGRU    55 YVGDEAQSKRG
     ACTS_FUGRU    55 YVGDEAQSKRG
     ACTT_FUGRU    55 YVGDEAQSKRG

</pre>

<p>
It is composed of three columns of data.

<p>
<ul>

<li>The first column is the name of the sequence being searched.
<li>The second is the start position of the match found.
<li>The third is the actual sequence that was matched by the search pattern. 

</ul>

<p>
If the option '-mmshow' is used, then an extra column of data is output 
indicating how many mismatches there are:

<p>

<pre>
% fuzzpro -mmshow
Protein pattern search
Input sequence(s): sw:100k_rat
Search pattern: RARLX(3)R
Number of mismatches [0]: 1
Output file [100k_rat.fuzzpro]: stdout

       100K_RAT   613     1 EARLNCFRN
</pre>

<p>

If the option '-desc' is used then the description of the sequence is
displayed before each line showing the match details.  For example:

<p>

<pre>
% fuzzpro 'sw:*_HUMAN' -desc
Protein pattern search
Search pattern: [FY]-[LIV]-G-[DE]-E-A-Q-x-[RKQ](2)-G
Number of mismatches [0]: 
Output file [143b_human.fuzzpro]: stdout

ACTIN, AORTIC SMOOTH MUSCLE (ALPHA-ACTIN 2).
     ACTA_HUMAN    55 YVGDEAQSKRG
ACTIN, CYTOPLASMIC 1 (BETA-ACTIN).
     ACTB_HUMAN    53 YVGDEAQSKRG
ACTIN, ALPHA CARDIAC.
     ACTC_HUMAN    55 YVGDEAQSKRG
ACTIN, CYTOPLASMIC 2 (GAMMA-ACTIN).
     ACTG_HUMAN    53 YVGDEAQSKRG
ACTIN, GAMMA-ENTERIC SMOOTH MUSCLE (ALPHA-ACTIN 3).
     ACTH_HUMAN    54 YVGDEAQSKRG
ACTIN, ALPHA SKELETAL MUSCLE (ALPHA-ACTIN 1).
     ACTS_HUMAN    55 YVGDEAQSKRG
</pre>

<p>

If the option '-acc' is also used then the accession number of the sequence is
displayed before each line showing the match details.  For example:

<p>

<pre>
% fuzzpro 'sw:*_HUMAN' -desc -acc
Protein pattern search
Search pattern: [FY]-[LIV]-G-[DE]-E-A-Q-x-[RKQ](2)-G
Number of mismatches [0]: 
Output file [143b_human.fuzzpro]: stdout

P03996 ACTIN, AORTIC SMOOTH MUSCLE (ALPHA-ACTIN 2).
     ACTA_HUMAN    55 YVGDEAQSKRG
P02570 ACTIN, CYTOPLASMIC 1 (BETA-ACTIN).
     ACTB_HUMAN    53 YVGDEAQSKRG
P04270 ACTIN, ALPHA CARDIAC.
     ACTC_HUMAN    55 YVGDEAQSKRG
P02571 ACTIN, CYTOPLASMIC 2 (GAMMA-ACTIN).
     ACTG_HUMAN    53 YVGDEAQSKRG
P12718 ACTIN, GAMMA-ENTERIC SMOOTH MUSCLE (ALPHA-ACTIN 3).
     ACTH_HUMAN    54 YVGDEAQSKRG
P02568 ACTIN, ALPHA SKELETAL MUSCLE (ALPHA-ACTIN 1).
     ACTS_HUMAN    55 YVGDEAQSKRG
</pre>

<p>

If the option '-usa' is used then the Uniform Sequence Address is output
at the start of each line of match details.  For example:

<p>

<pre>
% fuzzpro 'sw:*_HUMAN' -usa
Protein pattern search
Search pattern: [FY]-[LIV]-G-[DE]-E-A-Q-x-[RKQ](2)-G
Number of mismatches [0]: 
Output file [143b_human.fuzzpro]: stdout
sw-id:ACTA_HUMAN             ACTA_HUMAN    55 YVGDEAQSKRG
sw-id:ACTB_HUMAN             ACTB_HUMAN    53 YVGDEAQSKRG
sw-id:ACTC_HUMAN             ACTC_HUMAN    55 YVGDEAQSKRG
sw-id:ACTG_HUMAN             ACTG_HUMAN    53 YVGDEAQSKRG
sw-id:ACTH_HUMAN             ACTH_HUMAN    54 YVGDEAQSKRG
sw-id:ACTS_HUMAN             ACTS_HUMAN    55 YVGDEAQSKRG</pre>

<p>

This is useful because it turns the output into a 'list' file of
sequence names that can then be read in by other EMBOSS programs when a
'@' is put at the start of the file. 



<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="antigenic.html">antigenic</a></td><td>Finds antigenic sites in proteins</td></tr>
<tr><td><a href="digest.html">digest</a></td><td>Protein proteolytic enzyme or reagent cleavage digest</td></tr>
<tr><td><a href="fuzztran.html">fuzztran</a></td><td>Protein pattern search after translation</td></tr>
<tr><td><a href="helixturnhelix.html">helixturnhelix</a></td><td>Report nucleic acid binding motifs</td></tr>
<tr><td><a href="oddcomp.html">oddcomp</a></td><td>Finds protein sequence regions with a biased composition</td></tr>
<tr><td><a href="patmatdb.html">patmatdb</a></td><td>Search a protein sequence with a motif</td></tr>
<tr><td><a href="patmatmotifs.html">patmatmotifs</a></td><td>Search a PROSITE motif database with a protein sequence</td></tr>
<tr><td><a href="pepcoil.html">pepcoil</a></td><td>Predicts coiled coil regions</td></tr>
<tr><td><a href="preg.html">preg</a></td><td>Regular expression search of a protein sequence</td></tr>
<tr><td><a href="pscan.html">pscan</a></td><td>Scans proteins using PRINTS</td></tr>
<tr><td><a href="sigcleave.html">sigcleave</a></td><td>Reports protein signal cleavage sites</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<ul>

<li><a href="fuzznuc.html">fuzznuc</a> - Nucleic acid  pattern search

<li><a href="fuzztran.html">fuzztran</a> - Protein pattern search
after translation

</ul>

<p>
Other EMBOSS programs allow you to search for regular expression
patterns but may be less easy for the user who has never used regular
expressions before:

<p>

<ul>

<li><a href="dreg.html">dreg</a> - Regular expression search of a
nucleotide sequence

<li><a href="preg.html">preg</a> - Regular expression search of a protein
sequence

</ul>

<H2>
    Author(s)
</H2>


This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>

Written (2000) - Alan Bleasby
<br>
'-usa' added (13 March 2001) - Gary Williams

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

