<HTML>

<HEAD>
  <TITLE>
  EMBOSS: fuzznuc
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

fuzznuc
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>





<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Nucleic acid pattern search


<H2>
    Description
</H2>


fuzznuc uses PROSITE style patterns to search nucleotide sequences. 

<p>
Patterns are specifications of a (typically short) length of sequence to
be found.  They can specify a search for an exact sequence or they can
allow various ambiguities, matches to variable lengths of sequence and
repeated subsections of the sequence. 

<p>
fuzznuc intelligently selects the optimum searching algorithm to use,
depending on the complexity of the search pattern specified. 

<H2>
    Usage
</H2>


Here is a sample session with fuzznuc.

<pre>

% fuzznuc
Input sequence: embl:hhtetra
Search pattern: AAGCTT
Number of mismatches [0]: 
Output file [hhtetra.fuzznuc]:

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
   -pattern            string     The standard IUPAC one-letter codes for the
                                  nucleotides are used.
                                  The symbol `n' is used for a position where
                                  any nucleotide is accepted.
                                  Ambiguities are indicated by listing the
                                  acceptable nucleotides for a given position,
                                  between square parentheses `[ ]'. For
                                  example: [ACG] stands for A or C or G.
                                  Ambiguities are also indicated by listing
                                  between a pair of curly brackets `{ }' the
                                  nucleotides that are not accepted at a given
                                  position. For example: {AG} stands for any
                                  nucleotides except A and G.
                                  Each element in a pattern is separated from
                                  its neighbor by a `-'. (Optional in
                                  fuzznuc).
                                  Repetition of an element of the pattern can
                                  be indicated by following that element with
                                  a numerical value or a numerical range
                                  between parenthesis. Examples: N(3)
                                  corresponds to N-N-N, N(2,4) corresponds to
                                  N-N or N-N-N or N-N-N-N.
                                  When a pattern is restricted to either the
                                  5' or 3' end of a sequence, that pattern
                                  either starts with a `<' symbol or
                                  respectively ends with a `>' symbol.
                                  A period ends the pattern. (Optional in
                                  fuzznuc).
                                  For example, [CG](5)TG{A}N(1,5)C
   -mismatch           integer    Number of mismatches
  [-outfile]           report     Output report file name

   Optional qualifiers: (none)
   Advanced qualifiers:
   -complement         boolean    Search complementary strand

   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-pattern</td>
<td>The standard IUPAC one-letter codes for the nucleotides are used.
The symbol `n' is used for a position where any nucleotide is accepted.
Ambiguities are indicated by listing the acceptable nucleotides for a given position, between square parentheses `[ ]'. For example: [ACG] stands for A or C or G.
Ambiguities are also indicated by listing between a pair of curly brackets `{ }' the nucleotides that are not accepted at a given position. For example: {AG} stands for any nucleotides except A and G.
Each element in a pattern is separated from its neighbor by a `-'. (Optional in fuzznuc).
Repetition of an element of the pattern can be indicated by following that element with a numerical value or a numerical range between parenthesis. Examples: N(3) corresponds to N-N-N, N(2,4) corresponds to N-N or N-N-N or N-N-N-N.
When a pattern is restricted to either the 5' or 3' end of a sequence, that pattern either starts with a `<' symbol or respectively ends with a `>' symbol.
A period ends the pattern. (Optional in fuzznuc).
For example, [CG](5)TG{A}N(1,5)C</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-mismatch</td>
<td>Number of mismatches</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output report file name</td>
<td>Report file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-complement</td>
<td>Search complementary strand</td>
<td>Yes/No</td>
<td>No</td>
</tr>

</table>



<H2>
    Input file format
</H2>


Patterns for fuzznuc are based on the format of pattern used in the
PROSITE database, with the difference that the terminating dot '.' and
the hyphens, '-', between the characters are optional. 

<p>
The PROSITE pattern definition from the PROSITE documentation (amended
to refer to nucleic acid sequences, not proteins) follows. 

<ul>

<li>The standard IUPAC one-letter codes for the nucleotides are used. 

<li>The symbol `n' is used for a position where any nucleotide is
accepted. 

<li>Ambiguities are indicated by listing the acceptable nucleotides for a
given position, between square parentheses `[ ]'.  For example: [ACG]
stands for A or C or G. 

<li>Ambiguities are also indicated by listing between a pair of curly
brackets `{ }' the nucleotides that are not accepted at a given
position.  For example: {AG} stands for any nucleotides except A and G. 

<li>Each element in a pattern is separated from its neighbor by a `-'. 
(Optional in fuzznuc). 

<li>Repetition of an element of the pattern can be indicated by following
that element with a numerical value or a numerical range between
parenthesis.  Examples: N(3) corresponds to N-N-N, N(2,4) corresponds to
N-N or N-N-N or N-N-N-N. 

<li>When a pattern is restricted to either the 5' or 3' end of a
sequence, that pattern either starts with a `&lt;' symbol or respectively
ends with a `&gt;' symbol. 

<li>A period ends the pattern.  (Optional in fuzznuc). 

</ul>

<p>
For example, in the EMBL entry ECLAC you can look for the pattern:

<pre>

[CG](5)TG{A}N(1,5)C

</pre>

<p>
This searches for "C or G" 5 times, followed by T and G, then anything
except A, then any base (1 to 5 times) before a C. 

<p>
You can use ambiguity codes for nucleic acid searches but not within []
or {} as they expand to bracketed counterparts.  For example, "s" is
expanded to "[GC]" therefore [S] would be expanded to [[GC]] which is
illegal. 

<p>
Note the use of X is reserved for proteins.  You must use N for nucleic
acids to refer to any base. 

<p>
The search is case-independent, so 'AAA' matches 'aaa'. 

<H2>
    Output file format
</H2>


<p>

The output is a standard EMBOSS report file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-rformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  The available format names are: embl,
genbank, gff, pir, swiss, trace, listfile, dbmotif, diffseq, excel,
feattable, motif, regions, seqtable, simple, srs, table, tagseq

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html</A>
for further information on report formats.

<p>


<p>

By default <b>fuzznuc</b> writes a 'seqtable' report file.

<p>
    
The output from the above example is:

<p>

<pre>
########################################
# Program: fuzznuc
# Rundate: Thu Apr 11 13:34:06 2002
# Report_file: stdout
########################################

#=======================================
#
# Sequence: HHTETRA     from: 1   to: 1272
# HitCount: 2
#
# Pattern: aagctt
# Mismatch: 0
# Complement: No
#
#=======================================

  Start     End Mismatch Sequence
      1       6        . aagctt
   1267    1272        . aagctt

#---------------------------------------
#---------------------------------------
</pre>


	
<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

None.


<H2>
    References
</H2>

None.


<H2>
    Warnings
</H2>

None.


<H2>
    Diagnostic Error Messages
</H2>

None.


<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.


<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="dreg.html">dreg</a></td><td>regular expression search of a nucleotide sequence</td></tr>
<tr><td><a href="fuzztran.html">fuzztran</a></td><td>Protein pattern search after translation</td></tr>
<tr><td><a href="marscan.html">marscan</a></td><td>Finds MAR/SAR sites in nucleic sequences</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->



<p>
Other EMBOSS programs allow you to search for regular expression
patterns but may be less easy for the user who has never used regular
expressions before:

<p>

<ul>

<li><a href="dreg.html">dreg</a> - Regular expression search of a
nucleotide sequence

<li><a href="preg.html">preg</a> - Regular expression search of a protein
sequence

</ul>

<H2>
    Author(s)
</H2>


This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>

Written (2000) - Alan Bleasby
<br>
'-usa' added (13 March 2001) - Gary Williams

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

