<HTML>

<HEAD>
  <TITLE>
  EMBOSS: fuzztran
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

fuzztran
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<H2>
    Function
</H2>
Protein pattern search after translation

<H2>
    Description
</H2>

fuzztran uses PROSITE style protein patterns to search nucleic acid
sequences translated in the specified frame(s). 

<p>
Patterns are specifications of a (typically short) length of sequence to
be found.  They can specify a search for an exact sequence or they can
allow various ambiguities, matches to variable lengths of sequence and
repeated subsections of the sequence. 

<p>
fuzztran intelligently selects the optimum searching algorithm to use,
depending on the complexity of the search pattern specified. 

<H2>
    Usage
</H2>
<b>Here is a sample session with fuzztran</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>fuzztran -opt </b>
Protein pattern search after translation
Input sequence(s): <b>tembl:rnops</b>
Search pattern: <b>RA</b>
Number of mismatches [0]: <b></b>
Translation frames
         1 : 1
         2 : 2
         3 : 3
         F : Forward three frames
        -1 : -1
        -2 : -2
        -3 : -3
         R : Reverse three frames
         6 : All six frames
Frame(s) to translate [1]: <b>f</b>
Genetic codes
         0 : Standard
         1 : Standard (with alternative initiation codons)
         2 : Vertebrate Mitochondrial
         3 : Yeast Mitochondrial
         4 : Mold, Protozoan, Coelenterate Mitochondrial and Mycoplasma/Spiroplasma
         5 : Invertebrate Mitochondrial
         6 : Ciliate Macronuclear and Dasycladacean
         9 : Echinoderm Mitochondrial
        10 : Euplotid Nuclear
        11 : Bacterial
        12 : Alternative Yeast Nuclear
        13 : Ascidian Mitochondrial
        14 : Flatworm Mitochondrial
        15 : Blepharisma Macronuclear
        16 : Chlorophycean Mitochondrial
        21 : Trematode Mitochondrial
        22 : Scenedesmus obliquus
        23 : Thraustochytrium Mitochondrial
Code to use [0]: <b></b>
Output report [rnops.fuzztran]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-sequence]          seqall     Sequence database USA
   -pattern            string     The standard IUPAC one-letter codes for the
                                  amino acids are used.
                                  The symbol `x' is used for a position where
                                  any amino acid is accepted.
                                  Ambiguities are indicated by listing the
                                  acceptable amino acids for a given position,
                                  between square parentheses `[ ]'. For
                                  example: [ALT] stands for Ala or Leu or Thr.
                                  Ambiguities are also indicated by listing
                                  between a pair of curly brackets `{ }' the
                                  amino acids that are not accepted at a gven
                                  position. For example: {AM} stands for any
                                  amino acid except Ala and Met.
                                  Each element in a pattern is separated from
                                  its neighbor by a `-'. (Optional in
                                  fuzztran)
                                  Repetition of an element of the pattern can
                                  be indicated by following that element with
                                  a numerical value or a numerical range
                                  between parenthesis. Examples: x(3)
                                  corresponds to x-x-x, x(2,4) corresponds to
                                  x-x or x-x-x or x-x-x-x.
                                  When a pattern is restricted to either the
                                  N- or C-terminal of a sequence, that pattern
                                  either starts with a `<' symbol or
                                  respectively ends with a `>' symbol.
                                  A period ends the pattern. (Optional in
                                  fuzztran).
                                  For example, [DE](2)HS{P}X(2)PX(2,4)C
   -mismatch           integer    Number of mismatches
  [-outfile]           report     Output report file name

   Additional (Optional) qualifiers:
   -frame              menu       Frame(s) to translate
   -table              menu       Code to use

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1             integer    Start of each sequence to be used
   -send1               integer    End of each sequence to be used
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name

   "-outfile" associated qualifiers
   -rformat2            string     Report format
   -rname2              string     Base file name
   -rextension2         string     File name extension
   -rdirectory2         string     Output directory
   -raccshow2           boolean    Show accession number in the report
   -rdesshow2           boolean    Show description in the report
   -rscoreshow2         boolean    Show the score in the report
   -rusashow2           boolean    Show the full USA in the report

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-pattern</td>
<td>The standard IUPAC one-letter codes for the amino acids are used.
The symbol `x' is used for a position where any amino acid is accepted.
Ambiguities are indicated by listing the acceptable amino acids for a given position, between square parentheses `[ ]'. For example: [ALT] stands for Ala or Leu or Thr.
Ambiguities are also indicated by listing between a pair of curly brackets `{ }' the amino acids that are not accepted at a gven position. For example: {AM} stands for any amino acid except Ala and Met.
Each element in a pattern is separated from its neighbor by a `-'. (Optional in fuzztran)
Repetition of an element of the pattern can be indicated by following that element with a numerical value or a numerical range between parenthesis. Examples: x(3) corresponds to x-x-x, x(2,4) corresponds to x-x or x-x-x or x-x-x-x.
When a pattern is restricted to either the N- or C-terminal of a sequence, that pattern either starts with a `<' symbol or respectively ends with a `>' symbol.
A period ends the pattern. (Optional in fuzztran).
For example, [DE](2)HS{P}X(2)PX(2,4)C</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-mismatch</td>
<td>Number of mismatches</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output report file name</td>
<td>Report output file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-frame</td>
<td>Frame(s) to translate</td>
<td><table><tr><td>1</td> <td><i>(1)</i></td></tr><tr><td>2</td> <td><i>(2)</i></td></tr><tr><td>3</td> <td><i>(3)</i></td></tr><tr><td>F</td> <td><i>(Forward three frames)</i></td></tr><tr><td>-1</td> <td><i>(-1)</i></td></tr><tr><td>-2</td> <td><i>(-2)</i></td></tr><tr><td>-3</td> <td><i>(-3)</i></td></tr><tr><td>R</td> <td><i>(Reverse three frames)</i></td></tr><tr><td>6</td> <td><i>(All six frames)</i></td></tr></table></td>
<td>1</td>
</tr>

<tr>
<td>-table</td>
<td>Code to use</td>
<td><table><tr><td>0</td> <td><i>(Standard)</i></td></tr><tr><td>1</td> <td><i>(Standard (with alternative initiation codons))</i></td></tr><tr><td>2</td> <td><i>(Vertebrate Mitochondrial)</i></td></tr><tr><td>3</td> <td><i>(Yeast Mitochondrial)</i></td></tr><tr><td>4</td> <td><i>(Mold, Protozoan, Coelenterate Mitochondrial and Mycoplasma/Spiroplasma)</i></td></tr><tr><td>5</td> <td><i>(Invertebrate Mitochondrial)</i></td></tr><tr><td>6</td> <td><i>(Ciliate Macronuclear and Dasycladacean)</i></td></tr><tr><td>9</td> <td><i>(Echinoderm Mitochondrial)</i></td></tr><tr><td>10</td> <td><i>(Euplotid Nuclear)</i></td></tr><tr><td>11</td> <td><i>(Bacterial)</i></td></tr><tr><td>12</td> <td><i>(Alternative Yeast Nuclear)</i></td></tr><tr><td>13</td> <td><i>(Ascidian Mitochondrial)</i></td></tr><tr><td>14</td> <td><i>(Flatworm Mitochondrial)</i></td></tr><tr><td>15</td> <td><i>(Blepharisma Macronuclear)</i></td></tr><tr><td>16</td> <td><i>(Chlorophycean Mitochondrial)</i></td></tr><tr><td>21</td> <td><i>(Trematode Mitochondrial)</i></td></tr><tr><td>22</td> <td><i>(Scenedesmus obliquus)</i></td></tr><tr><td>23</td> <td><i>(Thraustochytrium Mitochondrial)</i></td></tr></table></td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<H2>
    Input file format
</H2>

<b>fuzztran</b> reads in normal nucleic acid sequence USAs. 

<p>

<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:rnops' is a sequence entry in the example nucleic acid database 'tembl'
<p>
<p><h3>Database entry: tembl:rnops</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   RNOPS      standard; RNA; ROD; 1493 BP.
XX
AC   Z46957;
XX
SV   Z46957.1
XX
DT   20-DEC-1994 (Rel. 42, Created)
DT   27-AUG-1996 (Rel. 49, Last updated, Version 8)
XX
DE   R.norvegicus mRNA for rhodopsin.
XX
KW   rhodopsin.
XX
OS   Rattus norvegicus (Norway rat)
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia;
OC   Eutheria; Rodentia; Sciurognathi; Muridae; Murinae; Rattus.
XX
RN   [1]
RP   1-1493
RA   Huber A., Sander P.H., Paulsen R.;
RT   "Phosphorylation of the InaD gene product, a photoreceptor membrane protein
RT   required for recovery of visual excitation";
RL   J. Biol. Chem. 271:11710-11717(1996).
XX
RN   [2]
RP   1-1493
RA   Huber A.;
RT   ;
RL   Submitted (20-DEC-1994) to the EMBL/GenBank/DDBJ databases.
RL   Huber A., Universitaet Karlsruhe, Zoologie I, Kornblumenstr. 13, 76128
RL   Karlsruhe, Germany
XX
DR   SWISS-PROT; P51489; OPSD_RAT.
XX
FH   Key             Location/Qualifiers
FH
FT   source          1..1493
FT                   /db_xref="taxon:10116"
FT                   /organism="Rattus norvegicus"
FT                   /strain="Sprague-Dawley"
FT                   /clone="pRO4"
FT                   /dev_stage="adult"
FT                   /tissue_type="retina"
FT                   /cell_type="rod"
FT                   /clone_lib="rat retinal library"
FT   5'UTR           1..83
FT   CDS             84..1130
FT                   /citation=[1]
FT                   /db_xref="SWISS-PROT:P51489"
FT                   /function="phototransduction"
FT                   /product="rhodopsin"
FT                   /protein_id="CAA87081.1"
FT                   /translation="MNGTEGPNFYVPFSNITGVVRSPFEQPQYYLAEPWQFSMLAAYMF
FT                   LLIVLGFPINFLTLYVTVQHKKLRTPLNYILLNLAVADLFMVFGGFTTTLYTSLHGYFV
FT                   FGPTGCNLEGFFATLGGEIGLWSLVVLAIERYVVVCKPMSNFRFGENHAIMGVAFTWVM
FT                   ALACAAPPLVGWSRYIPEGMQCSCGIDYYTLKPEVNNESFVIYMFVVHFTIPMIVIFFC
FT                   YGQLVFTVKEAAAQQQESATTQKAEKEVTRMVIIMVIFFLICWLPYASVAMYIFTHQGS
FT                   NFGPIFMTLPAFFAKTASIYNPIIYIMMNKQFRNCMLTTLCCGKNPLGDDEASATASKT
FT                   ETSQVAPA"
FT   3'UTR           1128..1493
XX
SQ   Sequence 1493 BP; 309 A; 475 C; 365 G; 344 T; 0 other;
     ggagccgtag gtagctgagc tcgccaggca gccttggtct ctgtctacga acagcccgtg        60
     gggcagcctc aagggccgca gccatgaacg gcacagaggg ccccaatttt tatgtgccct       120
     tctccaacat cacgggcgtg gtgcgcagcc cctttgagca gccgcagtac tacctggcgg       180
     agccatggca gttctccatg ctggcagcct acatgttcct gctcatcgtg ctgggcttcc       240
     ccatcaactt cctcacgctc tacgtcaccg tacagcacaa gaagctgcgc acaccactca       300
     actacatcct gctcaacttg gctgtggctg acctcttcat ggtcttcgga ggattcacca       360
     ccaccctcta cacctcactg catggctact ttgtctttgg gcccacaggc tgcaaccttg       420
     agggcttctt tgccaccctt ggaggtgaaa tcggcctgtg gtccctggta gtcctggcca       480
     ttgagcgcta cgtggtggtc tgcaagccca tgagcaactt ccgctttggg gagaatcatg       540
     ccattatggg tgtggccttc acctgggtca tggcgttggc ctgtgctgct cccccactgg       600
     ttggctggtc caggtacatc cccgagggca tgcagtgttc atgtgggatt gactactata       660
     cactcaagcc tgaggtcaac aatgagtcct tcgtcatcta catgttcgtg gtccacttca       720
     ccatccccat gatcgtcatc ttcttctgct acgggcagct ggtcttcacc gtcaaggagg       780
     ccgccgccca gcaacaggag tcggctacca ctcagaaggc agagaaggaa gtcacgcgca       840
     tggtcatcat catggtcatc ttcttcctga tctgctggct tccctatgcc agtgtggcca       900
     tgtacatctt tacccaccag ggctccaact tcggccccat cttcatgacc cttcccgctt       960
     tctttgctaa gaccgcctcc atctacaacc caatcatcta catcatgatg aacaagcagt      1020
     tccggaactg catgctcacc acgctctgct gcggcaagaa tccactggga gatgatgagg      1080
     cctctgccac tgcctccaag acggagacca gccaggtggc tccagcctaa gcctggccag      1140
     agactgtggc tgactgtagg agtctcctgt ccccactcac cccagccaca gcccccacca      1200
     ggagcagcac ccgttggaat gaggtcatgc aggctccctc agtgttcttt tctttgtttt      1260
     taatgaattc atgaaagcaa aatgaggctc cccactcaac agggacagcc tgacaaagga      1320
     catccatcca ccaagacccc cagcctggag tccccaattc ccgggggcca gcgggatctg      1380
     tacccctccc ctcagcttgt gtctcaggaa catgacaagt gtcccggctt acggctaagt      1440
     gtctaggaca gaatggaaca catagtagct gattaataaa tgctacctgg atg             1493
//
</pre>
</td></tr></table><p>



<h3>Pattern specification</h3>

Patterns for fuzztran are based on the format of pattern used in the PROSITE
database, with the difference that the terminating dot '.' and the hyphens, '-',
between the characters are optional. 

<p>
The PROSITE pattern definition from the PROSITE documentation follows. 

<ul>

<li>
      The standard IUPAC one-letter codes for the amino acids are
      used.

<li>
      The symbol `x' is used for a position where any amino acid is accepted. 

<li>
      Ambiguities are indicated by listing the acceptable amino acids
      for a given position, between square parentheses `[ ]'. For
      example: [ALT] stands for Ala or Leu or Thr.

<li>
      Ambiguities are also indicated by listing between a pair of
      curly brackets `{ }' the amino acids that are not accepted at a
      given position. For example: {AM} stands for any amino acid
      except Ala and Met.

<li>
      Each element in a pattern is separated from its neighbor by a
      `-'.  (Optional in fuzztran).

<li>
      Repetition of an element of the pattern can be indicated by
      following that element with a numerical value or a numerical
      range between parenthesis. Examples: x(3) corresponds to x-x-x,
      x(2,4) corresponds to x-x or x-x-x or x-x-x-x.

<li>
      When a pattern is restricted to either the N- or C-terminal of a
      sequence, that pattern either starts with a `&lt;' symbol or
      respectively ends with a `&gt;' symbol.

<li>
      A period ends the pattern. (Optional in fuzztran).

<li>All other characters, including spaces are not allowed.

</ul>

<p>
For example, you can look for the pattern: 

<pre>

[DE](2)HS{P}X(2)PX(2,4)C

</pre>

<p>
This means: Two Asps or Glus in any order followed by His, Ser, any
residue other then Pro, then two of any residue followed by Pro
followed by two to four of any residue followed by Cys.

<p>
The search is case-independent, so 'AAA' matches 'aaa'. 

<H2>
    Output file format
</H2>


<p>

The output is a standard EMBOSS report file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-rformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  The available format names are: embl,
genbank, gff, pir, swiss, trace, listfile, dbmotif, diffseq, excel,
feattable, motif, regions, seqtable, simple, srs, table, tagseq

<p>

See:
<A href="http://emboss.sf.net/docs/themes/ReportFormats.html">
http://emboss.sf.net/docs/themes/ReportFormats.html</A>
for further information on report formats.

<p>

<p>

By default <b>fuzztran</b> writes a 'table' report file.

<p>
 
<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: rnops.fuzztran</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
########################################
# Program: fuzztran
# Rundate: Thu Jul 15 2004 12:00:00
# Report_format: table
# Report_file: rnops.fuzztran
########################################

#=======================================
#
# Sequence: RNOPS     from: 1   to: 1493
# HitCount: 9
#
# Pattern: RA
# Mismatch: 0
# TransTable: 0
# Frames: F
#
#=======================================

  Start     End   Score Mismatch  Frame PStart   PEnd Translation
     97     102       2        .      1     33     34 RA         
    133     138       2        .      1     45     46 RA         
    421     426       2        .      1    141    142 RA         
    625     630       2        .      1    209    210 RA         
    835     840       2        .      1    279    280 RA         
    919     924       2        .      1    307    308 RA         
    227     232       2        .      2     76     77 RA         
    752     757       2        .      2    251    252 RA         
     72      77       2        .      3     24     25 RA         

#---------------------------------------
#---------------------------------------
</pre>
</td></tr></table><p>

   
<p>

The columns of data are as follows:

<p>

<ol>

<li>Start - the start position of the pattern in the nucleic acids sequence.
<li>End - the end position of the pattern in the nucleic acids sequence.
<li>Score - the score of the match.
<li>Mismatch - the number of mismatches .
<li>Frame - the translation frame that the pattern match occurs in.
<li>PStart - the start position of the match in the resulting protein sequence.
<li>PEnd - the end position of the match in the resulting protein sequence.
<li>Translation - the protein sequence that is matched.

</ol>


<H2>
    Data files
</H2>


<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

<p>
The Genetic Code data files are based on the NCBI genetic code tables.
Their names and descriptions are:

<dl>
<dt>EGC.0 </dt><dd>
      Standard (Differs from GC.1 in that it only has initiation site 'AUG')
<dt>EGC.1 </dt><dd>
      Standard
<dt>EGC.2 </dt><dd>
      Vertebrate Mitochodrial
<dt>EGC.3 </dt><dd>
      Yeast Mitochondrial
<dt>EGC.4 </dt><dd>
      Mold, Protozoan, Coelenterate Mitochondrial and Mycoplasma/Spiroplasma
<dt>EGC.5 </dt><dd>
      Invertebrate Mitochondrial
<dt>EGC.6 </dt><dd>
      Ciliate Macronuclear and Dasycladacean
<dt>EGC.9 </dt><dd>
      Echinoderm Mitochondrial
<dt>EGC.10 </dt><dd>
      Euplotid Nuclear
<dt>EGC.11 </dt><dd>
      Bacterial
<dt>EGC.12 </dt><dd>
      Alternative Yeast Nuclear
<dt>EGC.13 </dt><dd>
      Ascidian Mitochondrial
<dt>EGC.14 </dt><dd>
      Flatworm Mitochondrial
<dt>EGC.15</dt><dd>
      Blepharisma Macronuclear
<dt>EGC.16</dt><dd>
      Chlorophycean Mitochondrial
<dt>EGC.21</dt><dd>
      Trematode Mitochondrial
<dt>EGC.22</dt><dd>
      Scenedesmus obliquus
<dt>EGC.23</dt><dd>
      Thraustochytrium Mitochondrial
</dl>

<p>

The format of these files is very simple.

<p>   

It consists of several lines of optional comments, each starting with
a '#' character.

<p>

These are followed the line: 'Genetic Code [n]', where 'n' is the
number of the genetic code file.

<p> 

This is followed by the description of the code and then by four lines
giving the IUPAC one-letter code of the translated amino acid, the
start codons (indicdated by an 'M') and the three bases of the codon,
lined up one on top of the other.
   
<p>

For example:

<pre>
   
------------------------------------------------------------------------------
# Genetic Code Table
#
# Obtained from: http://www.ncbi.nlm.nih.gov/collab/FT/genetic_codes.html
# and: http://www3.ncbi.nlm.nih.gov/htbin-post/Taxonomy/wprintgc?mode=c
#
# Differs from Genetic Code [1] only in that the initiation sites have been
# changed to only 'AUG'

Genetic Code [0]
Standard
   
AAs  =   FFLLSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG
Starts = -----------------------------------M----------------------------
Base1  = TTTTTTTTTTTTTTTTCCCCCCCCCCCCCCCCAAAAAAAAAAAAAAAAGGGGGGGGGGGGGGGG
Base2  = TTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGG
Base3  = TCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAG
------------------------------------------------------------------------------

</pre>


<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>


When translating using non-standard genetic code table, always check the
table carefully for deviations from your particular organism's code. 


<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>


It always exits with status 0.

<H2>
    Known bugs
</H2>

None.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="antigenic.html">antigenic</a></td><td>Finds antigenic sites in proteins</td></tr>
<tr><td><a href="digest.html">digest</a></td><td>Protein proteolytic enzyme or reagent cleavage digest</td></tr>
<tr><td><a href="dreg.html">dreg</a></td><td>Regular expression search of a nucleotide sequence</td></tr>
<tr><td><a href="epestfind.html">epestfind</a></td><td>Finds PEST motifs as potential proteolytic cleavage sites</td></tr>
<tr><td><a href="fuzznuc.html">fuzznuc</a></td><td>Nucleic acid pattern search</td></tr>
<tr><td><a href="fuzzpro.html">fuzzpro</a></td><td>Protein pattern search</td></tr>
<tr><td><a href="helixturnhelix.html">helixturnhelix</a></td><td>Report nucleic acid binding motifs</td></tr>
<tr><td><a href="marscan.html">marscan</a></td><td>Finds MAR/SAR sites in nucleic sequences</td></tr>
<tr><td><a href="oddcomp.html">oddcomp</a></td><td>Finds protein sequence regions with a biased composition</td></tr>
<tr><td><a href="patmatdb.html">patmatdb</a></td><td>Search a protein sequence with a motif</td></tr>
<tr><td><a href="patmatmotifs.html">patmatmotifs</a></td><td>Search a PROSITE motif database with a protein sequence</td></tr>
<tr><td><a href="pepcoil.html">pepcoil</a></td><td>Predicts coiled coil regions</td></tr>
<tr><td><a href="preg.html">preg</a></td><td>Regular expression search of a protein sequence</td></tr>
<tr><td><a href="pscan.html">pscan</a></td><td>Scans proteins using PRINTS</td></tr>
<tr><td><a href="sigcleave.html">sigcleave</a></td><td>Reports protein signal cleavage sites</td></tr>
</table>

<p>
Other EMBOSS programs allow you to search for regular expression
patterns but may be less easy for the user who has never used regular
expressions before:

<p>

<ul>

<li><a href="dreg.html">dreg</a> - Regular expression search of a
nucleotide sequence

<li><a href="preg.html">preg</a> - Regular expression search of a protein
sequence

</ul>

<H2>
    Author(s)
</H2>
Alan Bleasby (ableasby&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK



<H2>
    History
</H2>

Written (2000) - Alan Bleasby
<br>
'-usa' added (13 March 2001) - Gary Williams


<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.


<H2>
    Comments
</H2>
None


</BODY>
</HTML>

