<HTML>

<HEAD>
  <TITLE>
  EMBOSS: garnier
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

garnier
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>





<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Predicts protein secondary structure


<H2>
    Description
</H2>


This is an implementation of the original Garnier Osguthorpe Robson
algorithm (GOR I) for predicting protein secondary structure.

<p>

Secondary structure prediction is notoriously difficult to do accurately.
The GOR I alogorithm is one of the first semi-successful methods.

<p>

The Garnier method is not regarded as the most accurate prediction, but
is simple to calculate on most workstations.

<p>

The accuracy of any secondary structure prediction program is not much
better than 70% to 80% at best.  This is an early algorithm and will
probably not predict with much better than about 65% accuracy. 

<p>

The Web servers for PHD, DSC, and others are generally preferred.

<p>

<b>Do not rely on this (or any other) program alone to make your
predictions with.  Use several programs and take a consensus of the
results.</b>

<H2>
    Usage
</H2>


Here is a sample session with garnier.

<pre>

% garnier
Input sequence: sw:amic_pseae
Output file [amic_pseae.garnier]: 

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequencea]         seqall     Sequence database USA
  [-outfile]           report     Output report file name

   Optional qualifiers: (none)
   Advanced qualifiers:
   -idc                integer    idc param

   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequencea]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output report file name</td>
<td>Report file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-idc</td>
<td>idc param</td>
<td>Integer from 0 to 6</td>
<td>0</td>
</tr>

</table>



<p>

The meaning and use of the parameter 'idc' is currently being
investigated.  The original author, Bill Pearson writes:

<p>
<i>"In their paper, GOR mention that if you know something about the
secondary structure content of the protein you are analyzing, you can do
better in prediction.  "idc" is an index into a set of arrays, dharr[]
and dsarr[], which provide "decision constants" (dch, dcs), which are
offsets that are applied to the weights for the helix and sheet (extend)
terms.  So, idc=0 says don't use the decision constant offsets, and
idc=1 to 6 indicates that various combinations of dch,dcs offsets should
be used.  I don't remember what they are, but I must have gotten the
values from their paper."</i>

<H2>
    Input file format
</H2>


Any protein sequence.


<H2>
    Output file format
</H2>


<p>

The output is a standard EMBOSS report file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-rformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  The available format names are: embl,
genbank, gff, pir, swiss, trace, listfile, dbmotif, diffseq, excel,
feattable, motif, regions, seqtable, simple, srs, table, tagseq

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html</A>
for further information on report formats.

<p>


<p>

By default <b>garnier</b> writes a 'tagseq' report file.

<p>

Here is the output from the example run.

<p>

<pre>
<hr>
########################################
# Program: garnier
# Rundate: Mon Feb 11 13:42:25 2002
# Report_file: amic_pseae.garnier
########################################

#=======================================
#
# Sequence: AMIC_PSEAE     from: 1   to: 384
# HitCount: 113
#
# DCH = 0, DCS = 0
#
#  Please cite:
#  Garnier, Osguthorpe and Robson (1978) J. Mol. Biol. 120:97-120
#
#
#
#=======================================


          .   10    .   20    .   30    .   40    .   50
      GSHQERPLIGLLFSETGVTADIERSQRYGALLAVEQLNREGGVGGRPIET
helix                    HHHHH        HHHHH
sheet      EE EEEEE                 EE              EEEE
turns        T                TTTT          TTTT
 coil CCCCC        CCCCCC         CC       C    CCCC

          .   60    .   70    .   80    .   90    .  100
      LSQDPGGDPDRYRLCAEDFIRNRGVRFLVGCYMSHTRKAVMPVVERADAL
helix               HHHHHH            HHHH H     HHHHHH
sheet E         EEEE           EEEE          EEEE      E
turns  TT TT   T          TTTTT    TTT    T T
 coil    C  CCC

          .  110    .  120    .  130    .  140    .  150
      LCYPTPYEGFEYSPNIVYGGPAPNQNSAPLAAYLIRHYGERVVFIGSDYI
helix                              HHH
sheet EEE    E       EE           E   EEEE    EEEEE
turns       T TTT  TT  T     TT           TT T     TTTT
 coil    CCC     CC     CCCCC  CCC          C          C

          .  160    .  170    .  180    .  190    .  200
      YPRESNHVMRHLYRQHGGTVLEEIYIPLYPSDDDVQRAVERIYQARADVV
helix       HHHH                       HHHHHHHHHHHHH
sheet           EEE       EEEEEEE                   EEEE
turns   TTT        TTT             TTTT
 coil CC   C          CCCC       CC

          .  210    .  220    .  230    .  240    .  250
      FSTVVGTGTAELYRAIARRYGDGRRPPIASLTTSEAEVAKMESDVAEGQV
helix          HHHHHHH                HHHHHHHHHHHHHHHHH
sheet EEEE            EE         EEE                   E
turns                   TTTTTT
 coil     CCCCC               CCC   CC

          .  260    .  270    .  280    .  290    .  300
      VVAPYFSSIDTAASRAFVQACHGFFPENATITAWAEAAYWQTLLLGRAAQ
helix            HHHHHHH           HHHHHHHHHHHHH    HHHH
sheet EEEE  EEE         EE                      E
turns     TT              TTT   TT
 coil          CC            CCC  C              CCC

          .  310    .  320    .  330    .  340    .  350
      AAGSWRVEDVQRHLYDICIDAPQGPVRVERQNNHSRLSSRIAEIDARGVF
helix H     HHHH                                HHH
sheet                 EEEE     EEEEE         EEE      EE
turns           TTTTTT     T        TT   T         TTT
 coil  CCCCC              C CCC       CCC CCC

          .  360    .  370    .  380
      QVRWQSPEPIRPDPYVVVHNLDDWSASMGGGALP
helix
sheet EE           EEEEEEE     E       E
turns   TT    TT           TTT   TTT
<hr>
</pre>




<H2>
    Data files
</H2>

None.


<H2>
    Notes
</H2>


The Garnier method is not regarded as the most accurate prediction, but
is simple to calculate on most workstations.

<p>

The Web servers for PHD, DSC, and others are generally preferred.

<p>

<b>Do not rely on this (or any other) program alone to make your
predictions with.  Use several programs and take a consensus of the
results.</b>

<p>

The 3D structure for the example sequence is known, although the
2D structure elements were not in the SwissProt feature table for
release 38 when the test data was extracted.

<p>
DSSP shows:

<pre>
 From     To   Structure
    9     13   E beta sheet
   21     39   H alpha helix
   50     54   E beta sheet
   60     72   H alpha helix
   78     81   E beta sheet
   85     97   H alpha helix
  101    104   E beta sheet
  117    119   E beta sheet
  128    136   H alpha helix
  142    148   E beta sheet
  151    166   H alpha helix
  170    177   E beta sheet
  183    196   H alpha helix
  200    204   E beta sheet
  208    221   H alpha helix
  229    231   E beta sheet
  236    239   H alpha helix
  244    247   H alpha helix
  251    254   E beta sheet
  263    273   H alpha helix
  284    303   H alpha helix
  308    315   H alpha helix
  320    322   E beta sheet
  325    329   E beta sheet
  336    337   E beta sheet
  341    345   E beta sheet
  351    356   E beta sheet

</pre>

<H2>
    References
</H2>


Garnier J, Osguthorpe DJ, Robson B
Analysis of the accuracy and implications of simple methods for
predicting the secondary structure of globular proteins. 
J Mol Biol 1978 Mar 25;120(1):97-120

<H2>
    Warnings
</H2>


The accuracy of any secondary structure prediction program is not much
better than 70% to 80% at best.  This is an early algorithm and will
probably not predict with much better than about 65% accuracy. 

<p>
You are advised to use several of the latest Web-based prediction sites
and combine them to make a consensus prediction.


<H2>
    Diagnostic Error Messages
</H2>


None. 

<H2>
    Exit status
</H2>

It always exist with a status of 0.

<H2>
    Known bugs
</H2>

None. 


<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="helixturnhelix.html">helixturnhelix</a></td><td>Report nucleic acid binding motifs</td></tr>
<tr><td><a href="hmoment.html">hmoment</a></td><td>Hydrophobic moment calculation</td></tr>
<tr><td><a href="pepcoil.html">pepcoil</a></td><td>Predicts coiled coil regions</td></tr>
<tr><td><a href="pepnet.html">pepnet</a></td><td>Displays proteins as a helical net</td></tr>
<tr><td><a href="pepwheel.html">pepwheel</a></td><td>Shows protein sequences as helices</td></tr>
<tr><td><a href="tmap.html">tmap</a></td><td>Displays membrane spanning regions</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This program ('GARNIER') was originally written by William Pearson
(wrp@virginia.edu) and released as part of his FASTA package. 

<p>
This application was modified for inclusion in EMBOSS by Rodrigo Lopez <a
href="mailto:rls@ebi.ac.uk">(rls@ebi.ac.uk)</a> European
Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SD, UK.


<H2>
    History
</H2>

None. 


<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

