<HTML>

<HEAD>
  <TITLE>
  EMBOSS: garnier
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: garnier
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program garnier
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Predicts protein secondary structure


<H2>
    Description
</H2>


This is an implementation of the original Garnier Osguthorpe Robson
algorithm (GOR I) for predicting protein secondary structure.

<H2>
    Usage
</H2>


Here is a sample session with garnier.

<pre>

% garnier
Input sequence: sw:amic_pseae
Output file [amic_pseae.garnier]: 

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequencea]         seqall     Sequence database USA
  [-outfile]           outfile    Output file name

   Optional qualifiers: (none)
   Advanced qualifiers:
   -idc                integer    idc param


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequencea]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.garnier</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-idc</td>
<td>idc param</td>
<td>Integer from 0 to 6</td>
<td>0</td>
</tr>

</table>



<H2>
    Input file format
</H2>


Any protein sequence.


<H2>
    Output file format
</H2>


Here is the output from the example run.

<pre>

 GARNIER plot of AMIC_PSEAE, 384 aa; DCH = 0, DCS = 0

 Please cite:
 Garnier, Osguthorpe and Robson (1978) J. Mol. Biol. 120:97-120

           .   10    .   20    .   30    .   40    .   50    .   60
       GSHQERPLIGLLFSETGVTADIERSHAYGALLAVEQLNREGGVGGRPIETLSQDPGGDPD
 helix                   HHHHHHHHHHHHHHHHHHH                       
 sheet      EE EEEEE                                 EEEEE         
 turns        T                              TTTT         TT TT   T
 coil  CCCCC        CCCCC                   C    CCCC       C  CCC 

           .   70    .   80    .   90    .  100    .  110    .  120
       RYRLCAEDFIRNRGVRFLVGCYMSHTRKAVMPVVERADALLCYPTPYEGFEYSPNIVYGG
 helix     HHHHHH            HHHH H     HHHHHH                     
 sheet EEEE           EEEE          EEEE      EEEE    E       EE   
 turns           TTTTT    TTT    T T                 T TTT  TT  T  
 coil                                             CCC     CC     CC

           .  130    .  140    .  150    .  160    .  170    .  180
       PAPNQNSAPLAAYLIRHYGERVVFIGSDYIYPRESNHVMRHLYRQHGGTVLEEIYIPLYP
 helix          HHH                        HHHH                    
 sheet         E   EEEE    EEEEE               EEE       EEEEEEE   
 turns    TT           TT T     TTTT   TTT        TTT             T
 coil  CCC  CCC          C          CCC   C          CCCC       CC 

           .  190    .  200    .  210    .  220    .  230    .  240
       SDDDLQRAVERIYQARADVVFSTVVGTGTAELYRAIARRYGDGRRPPIASLTTSEAEVAK
 helix    HHHHHHHHHHHHH             HHHHHHH                HHHHHHHH
 sheet                 EEEEEEEE            EE         EEE          
 turns TTT                                   TTTTTT                
 coil                          CCCCC               CCC   CC        

           .  250    .  260    .  270    .  280    .  290    .  300
       MESDVAEGQVVVAPYFSSIDTPASRAFVQACHGFFPENATITAWAEAAYWQTLLLGRAAQ
 helix HHHHHHHHH               HHHH           HHHHHHHHHHHHH    HHHH
 sheet          EEEEE   E          EE                      E       
 turns               TTT T   T       TTT   TT                      
 coil                     CCC C         CCC  C              CCC    

           .  310    .  320    .  330    .  340    .  350    .  360
       AAGNWRVEDVQRHLYDIDIDAPQGPVRVERQNNHSRLSSRIAEIDARGVFQVRWQSPEPI
 helix       HHHHHHH                             HHH               
 sheet              E  EEEE     EEEEE         EEE      EEEE        
 turns               TT     T        TT   T         TTT    TT    TT
 coil  CCCCCC              C CCC       CCC CCC               CCCC  

           .  370    .  380
       RPDPYVVVHNLDDWSASMGGGPLP
 helix                         
 sheet    EEEEEEE     E        
 turns            TTT  TTT     
 coil  CCC       C   C    CCCCC

 Residue totals: H:  0   E:  0   T:  0   C:  1
        percent: H:  0.0 E:  0.0 T:  0.0 C:  0.3
--------------------------------------------------------------------

</pre>

<H2>
    Data files
</H2>


<H2>
    Notes
</H2>


The Garnier method is not regarded as the most accurate prediction, but
is simple to calculate on most workstations.

<p>
The Web servers for PHD, DSC, and others are generally preferred.

<p>
The 3D structure for the example sequence is known, although the
2D structure elements were not in the SwissProt feature table for
release 38 when the test data was extracted.

<p>
DSSP shows:

<pre>
 From     To   Structure
    9     13   E beta sheet
   21     39   H alpha helix
   50     54   E beta sheet
   60     72   H alpha helix
   78     81   E beta sheet
   85     97   H alpha helix
  101    104   E beta sheet
  117    119   E beta sheet
  128    136   H alpha helix
  142    148   E beta sheet
  151    166   H alpha helix
  170    177   E beta sheet
  183    196   H alpha helix
  200    204   E beta sheet
  208    221   H alpha helix
  229    231   E beta sheet
  236    239   H alpha helix
  244    247   H alpha helix
  251    254   E beta sheet
  263    273   H alpha helix
  284    303   H alpha helix
  308    315   H alpha helix
  320    322   E beta sheet
  325    329   E beta sheet
  336    337   E beta sheet
  341    345   E beta sheet
  351    356   E beta sheet

</pre>

<p>
The meaning and use of the parameter 'idc' is currently being
investigated.  The original author, Bill Pearson writes:

<p>
<i>"In their paper, GOR mention that if you know something about the
secondary structure content of the protein you are analyzing, you can do
better in prediction.  "idc" is an index into a set of arrays, dharr[]
and dsarr[], which provide "decision constants" (dch, dcs), which are
offsets that are applied to the weights for the helix and sheet (extend)
terms.  So, idc=0 says don't use the decision constant offsets, and
idc=1 to 6 indicates that various combinations of dch,dcs offsets should
be used.  I don't remember what they are, but I must have gotten the
values from their paper."</i>

<H2>
    References
</H2>


Garnier J, Osguthorpe DJ, Robson B
Analysis of the accuracy and implications of simple methods for
predicting the secondary structure of globular proteins. 
J Mol Biol 1978 Mar 25;120(1):97-120

<H2>
    Warnings
</H2>


The accuracy of any secondary structure prediction program is not much
better than 70% to 80% at best.  This is an early algorithm and will
probably not predict with much better than about 65% accuracy. 

<p>
You are advised to use several of the latest Web-based prediction sites
and combine them to make a consensus prediction.


<H2>
    Diagnostic Error Messages
</H2>


<H2>
    Exit status
</H2>


<H2>
    Known bugs
</H2>


<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="antigenic.html">antigenic</a></td><td>Finds antigenic sites in proteins</td></tr>
<tr><td><a href="diffseq.html">diffseq</a></td><td>Find differences (SNPs) between nearly identical sequences</td></tr>
<tr><td><a href="dotmatcher.html">dotmatcher</a></td><td>Displays a thresholded dotplot of two sequences</td></tr>
<tr><td><a href="dotpath.html">dotpath</a></td><td>Displays a non-overlapping wordmatch dotplot of two sequences</td></tr>
<tr><td><a href="dottup.html">dottup</a></td><td>Displays a wordmatch dotplot of two sequences</td></tr>
<tr><td><a href="helixturnhelix.html">helixturnhelix</a></td><td>Report nucleic acid binding motifs</td></tr>
<tr><td><a href="oddcomp.html">oddcomp</a></td><td>Finds protein sequence regions with a biased composition</td></tr>
<tr><td><a href="pepcoil.html">pepcoil</a></td><td>Predicts coiled coil regions</td></tr>
<tr><td><a href="pepnet.html">pepnet</a></td><td>Displays proteins as a helical net</td></tr>
<tr><td><a href="pepwheel.html">pepwheel</a></td><td>Shows protein sequences as helices</td></tr>
<tr><td><a href="polydot.html">polydot</a></td><td>Displays all-against-all dotplots of a set of sequences</td></tr>
<tr><td><a href="pscan.html">pscan</a></td><td>Scans proteins using PRINTS</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="sigcleave.html">sigcleave</a></td><td>Reports protein signal cleavage sites</td></tr>
<tr><td><a href="tmap.html">tmap</a></td><td>Displays membrane spanning regions</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This program ('GARNIER') was originally written by William Pearson
(wrp@virginia.edu) and released as part of his FASTA package. 

<p>
This application was modified for inclusion in EMBOSS by Rodrigo Lopez <a
href="mailto:rls@ebi.ac.uk">(rls@ebi.ac.uk)</a> European
Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SD, UK.


<H2>
    History
</H2>


<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

