<HTML>

<HEAD>
  <TITLE>
  EMBOSS: getorf
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: getorf
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program getorf
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Finds and extracts open reading frames (ORFs)


<H2>
    Description
</H2>


This program finds and outputs the sequences of open reading frames
(ORFs).

<p>
The ORFs can be defined as regions of a specified minimum size between
STOP codons or between START and STOP codons.  

<p>
The ORFs can be output as the nucleotide sequence or as the translation. 

<p>
The program can also output the region around the START or the initial
STOP codon or the ending STOP codons of an ORF for those doing analysis
of the properties of these regions. 

<p>
The START and STOP codons are defined in the Genetic Code tables.
A suitable Genetic Code table can be selected for the organism you are
investigating. 

<H2>
    Usage
</H2>


Here is a sample session with getorf.

<pre>

% getorf -minsize 300
Input sequence: embl:eclaci
Output sequence [eclaci.orf]: 

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
  [-outseq]            seqoutall  Output sequence(s) USA

   Optional qualifiers:
   -table              menu       Code to use
   -minsize            integer    Minimum nucleotide size of ORF to report
   -find               menu       This is a small menu of possible output
                                  options. The first four options are to
                                  select either the protein translation or the
                                  original nucleic acid sequence of the open
                                  reading frame. There are two possible
                                  definitions of an open reading frame: it can
                                  either be a region that is free of STOP
                                  codons or a region that begins with a START
                                  codon and ends with a STOP codon. The last
                                  three options are probably only of interest
                                  to people who wish to investigate the
                                  statistical properties of the regions around
                                  potential START or STOP codons. The last
                                  option assumes that ORF lengths are
                                  calculated between two STOP codons.

   Advanced qualifiers:
   -[no]methionine     bool       START codons at the beginning of protein
                                  products will usually code for Methionine,
                                  despite what the codon will code for when it
                                  is internal to a protein. This qualifier
                                  sets all such START codons to code for
                                  Methionine by default.
   -circular           bool       Is the sequence circular
   -[no]reverse        bool       Set this to be false if you do not wish to
                                  find ORFs in the reverse complement of the
                                  sequence.
   -flanking           integer    If you have chosen one of the options of the
                                  type of sequence to find that gives the
                                  flanking sequence around a STOP or START
                                  codon, this allows you to set the number of
                                  nucleotides either side of that codon to
                                  output. If the region of flanking
                                  nucleotides crosses the start or end of the
                                  sequence, no output is given for this codon.

   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outseq]<br>(Parameter 2)</td>
<td>Output sequence(s) USA</td>
<td>Writeable sequence(s)</td>
<td><i>&lt;sequence&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-table</td>
<td>Code to use</td>
<td><table><tr><td>0</td> <td><i>(Standard)</i></td></tr><tr><td>1</td> <td><i>(Standard (with alternative initiation codons))</i></td></tr><tr><td>2</td> <td><i>(Vertebrate Mitochondrial)</i></td></tr><tr><td>3</td> <td><i>(Yeast Mitochondrial)</i></td></tr><tr><td>4</td> <td><i>(Mold, Protozoan, Coelenterate Mitochondrial and Mycoplasma/Spiroplasma)</i></td></tr><tr><td>5</td> <td><i>(Invertebrate Mitochondrial)</i></td></tr><tr><td>6</td> <td><i>(Ciliate Macronuclear and Dasycladacean)</i></td></tr><tr><td>9</td> <td><i>(Echinoderm Mitochondrial)</i></td></tr><tr><td>10</td> <td><i>(Euplotid Nuclear)</i></td></tr><tr><td>11</td> <td><i>(Bacterial)</i></td></tr><tr><td>12</td> <td><i>(Alternative Yeast Nuclear)</i></td></tr><tr><td>13</td> <td><i>(Ascidian Mitochondrial)</i></td></tr><tr><td>14</td> <td><i>(Flatworm Mitochondrial)</i></td></tr><tr><td>15</td> <td><i>(Blepharisma Macronuclear)</i></td></tr><tr><td>16</td> <td><i>(Chlorophycean Mitochondrial)</i></td></tr><tr><td>21</td> <td><i>(Trematode Mitochondrial)</i></td></tr><tr><td>22</td> <td><i>(Scenedesmus obliquus)</i></td></tr><tr><td>23</td> <td><i>(Thraustochytrium Mitochondrial)</i></td></tr></table></td>
<td>0</td>
</tr>

<tr>
<td>-minsize</td>
<td>Minimum nucleotide size of ORF to report</td>
<td>Any integer value</td>
<td>30</td>
</tr>

<tr>
<td>-find</td>
<td>This is a small menu of possible output options. The first four options are to select either the protein translation or the original nucleic acid sequence of the open reading frame. There are two possible definitions of an open reading frame: it can either be a region that is free of STOP codons or a region that begins with a START codon and ends with a STOP codon. The last three options are probably only of interest to people who wish to investigate the statistical properties of the regions around potential START or STOP codons. The last option assumes that ORF lengths are calculated between two STOP codons.</td>
<td><table><tr><td>0</td> <td><i>(Translation of regions between STOP codons)</i></td></tr><tr><td>1</td> <td><i>(Translation of regions between START and STOP codons)</i></td></tr><tr><td>2</td> <td><i>(Nucleic sequences between STOP codons)</i></td></tr><tr><td>3</td> <td><i>(Nucleic sequences between START and STOP codons)</i></td></tr><tr><td>4</td> <td><i>(Nucleotides flanking START codons)</i></td></tr><tr><td>5</td> <td><i>(Nucleotides flanking initial STOP codons)</i></td></tr><tr><td>6</td> <td><i>(Nucleotides flanking ending STOP codons)</i></td></tr></table></td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]methionine</td>
<td>START codons at the beginning of protein products will usually code for Methionine, despite what the codon will code for when it is internal to a protein. This qualifier sets all such START codons to code for Methionine by default.</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-circular</td>
<td>Is the sequence circular</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]reverse</td>
<td>Set this to be false if you do not wish to find ORFs in the reverse complement of the sequence.</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-flanking</td>
<td>If you have chosen one of the options of the type of sequence to find that gives the flanking sequence around a STOP or START codon, this allows you to set the number of nucleotides either side of that codon to output. If the region of flanking nucleotides crosses the start or end of the sequence, no output is given for this codon.</td>
<td>Any integer value</td>
<td>100</td>
</tr>

</table>



<H2>
    Input file format
</H2>


Any nucleic acid sequence USA.

<H2>
    Output file format
</H2>


The output is a sequence file containing predicted open reading frames
longer than the minimum size, which defaults to 30 bases or 10 amino acids.

<p>
The results from the example run are:

<p>
<pre>
&gt;ECLACI_1 [735 - 1112] E. coli laci gene (codes for the lac repressor).
GHRSHCDAGCQRSDGAGRNARHYRVRAARWCGYLGSGIRRYRRQLMLYPAVNHHQTGFSP
AGANQRGPLAATLSGPGGEGQSAVARLTGEKKNHPGAQYANRLSPRVGRFINAAGTTGFP
TGKRAV
&gt;ECLACI_2 [1 - 1110] E. coli laci gene (codes for the lac repressor).
PEESQFRVVNVKPVTLYDVAEYAGVSYQTVSRVVNQASHVSAKTREKVEAAMAELNYIPN
RVAQQLAGKQSLLIGVATSSLALHAPSQIVAAIKSRADQLGASVVVSMVERSGVEACKAA
VHNLLAQRVSGLIINYPLDDQDAIAVEAACTNVPALFLDVSDQTPINSIIFSHEDGTRLG
VEHLVALGHQQIALLAGPLSSVSARLRLAGWHKYLTRNQIQPIAEREGDWSAMSGFQQTM
QMLNEGIVPTAMLVANDQMALGAMRAITESGLRVGADISVVGYDDTEDSSCYIPPSTTIK
QDFRLLGQTSVDRLLQLSQGQAVKGNQLLPVSLVKRKTTLAPNTQTASPRALADSLMQLA
RQVSRLESGQ*
&gt;ECLACI_3 [465 - 49] E. coli laci gene (codes for the lac repressor).
RRNISAGSFHSNGILVIQRIVNDQPTDALREKIVHRRFTGFDAASFYHRHHHAGTQLIGA
RFNRRDNLRRRVQGQTGGGNANQQRLFARQLLCHAVGNVIQLRHRRFHFFPRFRRNVAGL
VHHAGNGLIRDTGILCDIV
</pre>
<p>

All output ORF sequences are written to the specified outut file.

<p> 

The name of the ORF sequences is constructed from the name of the
input sequence with an underscore character ('_') and a unique ordinal number
of the ORF found appended.  The description of the output ORF sequence
is constructed from the description of the input sequence with the start
and end positions of the ORF prepended.  

<p>

The unique number appended to the name is simply used to create new
unique sequence names, it does not imply any further information
indicating any order, positioning or sense-strand of the ORFs. 

<p>

If the ORF has been found in the reverse sense, then the start position
will be smaller than the end position.  The numbering uses the
forward-sense positions, but read in the reverse sense.  For example,
<b>&gt;ECLACI_3 [465 - 49]</b> in the output above is a reverse-sense ORF
running from position 465 to 49. 

<H2>
    Data files
</H2>

The START and STOP codons used by <b>getorf</b> are defined in the
Genetic Code data files. By default, Genetic Code file <b>EGC.0</b>
is used.
<p>

The default file <b>EGC.0</b> is the 'Standard Code' with the rarely
used alternate START codons omitted, it only has the normal 'AUG' START
codon.  The 'Standard Code' with the rarely used alternate START codons
included is Genetic Code file <b>EGC.1</b>. 

<p>

It is expected that user will sometimes wish to customise a Genetic Code
file. To do this, use the program <b>embossdata</b>.

<p>

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>


<p>

The Genetic Code data files are based on the NCBI genetic code tables.
Their names and descriptions are:

<dl>
<dt>EGC.0 </dt><dd>
      Standard (Differs from GC.1 in that it only has initiation site 'AUG')
<dt>EGC.1 </dt><dd>
      Standard
<dt>EGC.2 </dt><dd>
      Vertebrate Mitochodrial
<dt>EGC.3 </dt><dd>
      Yeast Mitochondrial
<dt>EGC.4 </dt><dd>
      Mold, Protozoan, Coelenterate Mitochondrial and Mycoplasma/Spiroplasma
<dt>EGC.5 </dt><dd>
      Invertebrate Mitochondrial
<dt>EGC.6 </dt><dd>
      Ciliate Macronuclear and Dasycladacean
<dt>EGC.9 </dt><dd>
      Echinoderm Mitochondrial
<dt>EGC.10 </dt><dd>
      Euplotid Nuclear
<dt>EGC.11 </dt><dd>
      Bacterial
<dt>EGC.12 </dt><dd>
      Alternative Yeast Nuclear
<dt>EGC.13 </dt><dd>
      Ascidian Mitochondrial
<dt>EGC.14 </dt><dd>
      Flatworm Mitochondrial
<dt>EGC.15</dt><dd>
      Blepharisma Macronuclear
<dt>EGC.16</dt><dd>
      Chlorophycean Mitochondrial
<dt>EGC.21</dt><dd>
      Trematode Mitochondrial
<dt>EGC.22</dt><dd>
      Scenedesmus obliquus
<dt>EGC.23</dt><dd>
      Thraustochytrium Mitochondrial
</dl>

<p>

The format of these files is very simple.

<p>   

It consists of several lines of optional comments, each starting with
a '#' character.

<p>

These are followed the line: 'Genetic Code [n]', where 'n' is the
number of the genetic code file.

<p> 

This is followed by the description of the code and then by four lines
giving the IUPAC one-letter code of the translated amino acid, the
start codons (indicdated by an 'M') and the three bases of the codon,
lined up one on top of the other.
   
<p>

For example:

<pre>
   
------------------------------------------------------------------------------
# Genetic Code Table
#
# Obtained from: http://www.ncbi.nlm.nih.gov/collab/FT/genetic_codes.html
# and: http://www3.ncbi.nlm.nih.gov/htbin-post/Taxonomy/wprintgc?mode=c
#
# Differs from Genetic Code [1] only in that the initiation sites have been
# changed to only 'AUG'

Genetic Code [0]
Standard
   
AAs  =   FFLLSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG
Starts = -----------------------------------M----------------------------
Base1  = TTTTTTTTTTTTTTTTCCCCCCCCCCCCCCCCAAAAAAAAAAAAAAAAGGGGGGGGGGGGGGGG
Base2  = TTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGG
Base3  = TCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAG
------------------------------------------------------------------------------

</pre>


<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>


It always exits with status 0. 

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="marscan.html">marscan</a></td><td>Finds MAR/SAR sites in nucleic sequences</td></tr>
<tr><td><a href="plotorf.html">plotorf</a></td><td>Plot potential open reading frames</td></tr>
<tr><td><a href="showorf.html">showorf</a></td><td>Pretty output of DNA translations</td></tr>
<tr><td><a href="wobble.html">wobble</a></td><td>Wobble base plot</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->

<P>
<ul>

<li><a href="checktrans.html">checktrans</a> - Reports STOP codons
and ORF statistics of a protein sequence

</ul>

<H2>
    Author(s)
</H2>


This application was written by Gary Williams <a
href="mailto:gwilliam@hgmp.mrc.ac.uk">(gwilliam@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>

2000 - written - Gary Williams

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

