<HTML>
<HEAD>
  <TITLE>
  EMBOSS: godef
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="/images/emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
godef
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
Find GO ontology terms by definition
<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>


<b>godef</b> searches the definition of gene ontology terms and returns matching terms.
The input is read from the installed GO database. The ontology term output can be written to screen, to file, or passed to another program. A wide range of standard ontology term formats may be specified for input and output.

<p>
Optionally the search can be restricted to specified GO namespaces.




<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

Here is a sample session with <b>godef</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>godef glycoprotein </b>
Find GO ontology terms by definition
Obo output file [godef.obo]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
Find GO ontology terms by definition
Version: EMBOSS:6.6.0.0

   Standard (Mandatory) qualifiers:
  [-query]             string     Definition word(s) to search for in ontology
                                  (Any string)
  [-outfile]           outobo     [*.godef] Output ontology term file name

   Additional (Optional) qualifiers:
   -namespace          menu       [*] By default all terms are returned.
                                  Searches can be limited to one or a few
                                  namespaces. (Values: biological_process
                                  (Biological process); cellular_component
                                  (Cellular component); molecular_function
                                  (Molecular function))

   Advanced (Unprompted) qualifiers:
   -subclasses         boolean    [N] Extend the query matches to include all
                                  terms which are specialisations (GO
                                  sub-classes) of the matched type.
   -obsolete           boolean    [N] The default behaviour is to not use or
                                  return obsolete terms. This option if set
                                  will include all terms.

   Associated qualifiers:

   "-outfile" associated qualifiers
   -odirectory2        string     Output directory
   -oformat2           string     Ontology term output format

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-query]<br>(Parameter 1)</td>
<td>string</td>
<td>Definition word(s) to search for in ontology</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-outfile]<br>(Parameter 2)</td>
<td>outobo</td>
<td>Output ontology term file name</td>
<td>OBO ontology term(s)</td>
<td><i>&lt;*&gt;</i>.godef</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>-namespace</td>
<td>list</td>
<td>By default all terms are returned. Searches can be limited to one or a few namespaces.</td>
<td><table><tr><td>biological_process</td> <td><i>(Biological process)</i></td></tr><tr><td>cellular_component</td> <td><i>(Cellular component)</i></td></tr><tr><td>molecular_function</td> <td><i>(Molecular function)</i></td></tr></table></td>
<td>*</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>-subclasses</td>
<td>boolean</td>
<td>Extend the query matches to include all terms which are specialisations (GO sub-classes) of the matched type.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-obsolete</td>
<td>boolean</td>
<td>The default behaviour is to not use or return obsolete terms. This option if set will include all terms.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td align="left" colspan=5>"-outfile" associated outobo qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -odirectory2<br>-odirectory_outfile</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -oformat2<br>-oformat_outfile</td>
<td>string</td>
<td>Ontology term output format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>godef</b> queries the gene ontology.

<p>






<H2>
    Output file format
</H2>

<p>

The output is a standard EMBOSS ontology term file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <tt>-oformat xxx</tt>, where 'xxx' is replaced by
the name of the required format.  The available format names are: obo,
brief, list, html, xml, json, excel.
<p>

See:
<A href="http://emboss.sf.net/docs/themes/OntologyFormats.html">
http://emboss.sf.net/docs/themes/OntologyFormats.html</A>
for further information on ontology formats.

<p>

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: godef.obo</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
[Term]
id: GO:0043256
name: laminin complex
namespace: cellular_component
def: A large, extracellular glycoprotein complex composed of three different polypeptide chains, alpha, beta and gamma. Provides an integral part of the structural scaffolding of basement membranes.
subset: goslim_pir
is_a: GO:0043234 ! protein complex
is_a: GO:0044420 ! extracellular matrix part
is_a: GO:0044421 ! extracellular region part
relationship: part_of GO:0005605 ! basal lamina

[Term]
id: GO:0038026
name: reelin-mediated signaling pathway
namespace: biological_process
def: A series of molecular signals initiated by the binding of reelin (a secreted glycoprotein) to a receptor on the surface of a target cell, and ending with regulation of a downstream cellular process, e.g. transcription.
synonym: "reeler-mediated signaling pathway" EXACT [GOC:bf, PR:000013879]
synonym: "reelin-mediated signal transduction pathway" EXACT [GOC:bf]
synonym: "reelin-mediated signalling pathway" EXACT [GOC:bf]
is_a: GO:0007166 ! cell surface receptor signaling pathway
created_by: rfoulger
creation_date: 2011-08-02T02:06:18Z

[Term]
id: GO:0016011
name: dystroglycan complex
namespace: cellular_component
def: A protein complex that includes alpha- and beta-dystroglycan, which are alternative products of the same gene; the laminin-binding component of the dystrophin-associated glycoprotein complex, providing a link between the subsarcolemmal cytoskeleton (in muscle cells) and the extracellular matrix. Alpha-dystroglycan is an extracellular protein binding to alpha-laminin and to beta-dystroglycan; beta-dystroglycan is a transmembrane protein which binds alpha-dystroglycan and dystrophin.
is_a: GO:0043234 ! protein complex
is_a: GO:0044459 ! plasma membrane part
relationship: part_of GO:0016010 ! dystrophin-associated glycoprotein complex

[Term]
id: GO:0010405
name: arabinogalactan protein metabolic process
namespace: biological_process
def: The chemical reactions and pathways involving a cell wall arabinogalactan II glycoprotein, which is composed of a group of core protein of highly varying length and domain complexity. These are O-glycosylated at one or more hydroxyproline residues by arabinogalactan (AG) type II groups, which consist of (1-&gt;3)-beta-galactan and (1-&gt;6)-beta-linked galactan chains connected to each other by (1-&gt;3,1-&gt;6)-linked branch points, O-3 and O-6 positions substituted with terminal arabinosyl residues. Also, rhamnose, fucose, glucuronic and galacturonic acid can be present in the glycan structures.
synonym: "arabinogalactan protein metabolism" EXACT []
is_a: GO:0010404 ! cell wall hydroxyproline-rich glycoprotein metabolic process

[Term]
id: GO:0010404
name: cell wall hydroxyproline-rich glycoprotein metabolic process
namespace: biological_process
def: The chemical reactions and pathways involving a cell wall hydroxyproline-rich glycoprotein that consist of a core-protein backbone O-glycosylated by one or more complex carbohydrates.
synonym: "cell wall hydroxyproline-rich glycoprotein metabolism" EXACT []
is_a: GO:0010384 ! cell wall proteoglycan metabolic process

[Term]
id: GO:0004569


<font color=red>  [Part of this file has been deleted for brevity]</font>

def: The chemical reactions and pathways involving chondroitin sulfate proteoglycan, any glycoprotein whose glycosaminoglycan units are chondroitin sulfate. Chondroitin sulfates are a group of 10-60 kDa glycosaminoglycans, widely distributed in cartilage and other mammalian connective tissues; the repeat units consist of beta-(1,4)-linked D-glucuronyl beta-(1,3)-N-acetyl-D-galactosamine sulfate.
subset: gosubset_prok
synonym: "chondroitin sulfate proteoglycan metabolism" EXACT []
synonym: "chondroitin sulphate proteoglycan metabolic process" EXACT []
synonym: "chondroitin sulphate proteoglycan metabolism" EXACT []
is_a: GO:0006029 ! proteoglycan metabolic process
is_a: GO:0006790 ! sulfur compound metabolic process
is_a: GO:0030203 ! glycosaminoglycan metabolic process
is_a: GO:0043436 ! oxoacid metabolic process

[Term]
id: GO:0046884
name: follicle-stimulating hormone secretion
namespace: biological_process
def: The regulated release of follicle-stimulating hormone, a gonadotropic glycoprotein hormone secreted by the anterior pituitary.
synonym: "follicle stimulating hormone secretion" EXACT []
synonym: "follitropin secretion" EXACT []
synonym: "FSH secretion" EXACT []
is_a: GO:0032274 ! gonadotropin secretion

[Term]
id: GO:0018405
name: protein-keratan sulfate linkage via keratan sulfate D-glucuronyl-D-galactosyl-D-galactosyl-D-xylosyl-L-threonine
namespace: biological_process
def: Keratan sulfate components are covalently linked to a core glycoprotein via O-glycosidic linkages between xylose and threonine residues.
subset: gosubset_prok
synonym: "protein-keratan sulphate linkage via keratan sulphate D-glucuronyl-D-galactosyl-D-galactosyl-D-xylosyl-L-threonine" EXACT []
xref: RESID:AA0247
is_a: GO:0018210 ! peptidyl-threonine modification

[Term]
id: GO:0010408
name: fasciclin-like arabinogalactan protein metabolic process
namespace: biological_process
def: The chemical reactions and pathways involving the cell wall arabinogalactan II glycoprotein variant which contains both an arabinogalactan protein (AGP) motif and a fasciclin domain.
synonym: "fasciclin-like arabinogalactan protein metabolism" EXACT []
is_a: GO:0010405 ! arabinogalactan protein metabolic process

[Term]
id: GO:0006045
name: N-acetylglucosamine biosynthetic process
namespace: biological_process
def: The chemical reactions and pathways resulting in the formation of N-acetylglucosamine. The D isomer is a common structural unit of glycoproteins in plants, bacteria and animals; it is often the terminal sugar of an oligosaccharide group of a glycoprotein.
subset: gosubset_prok
synonym: "N-acetylglucosamine anabolism" EXACT []
synonym: "N-acetylglucosamine biosynthesis" EXACT []
synonym: "N-acetylglucosamine formation" EXACT []
synonym: "N-acetylglucosamine synthesis" EXACT []
is_a: GO:0006044 ! N-acetylglucosamine metabolic process
is_a: GO:1901073 ! glucosamine-containing compound biosynthetic process

</pre>
</td></tr></table><p>





<H2>
    Data files
</H2>

The gene Ontology is included in EMBOSS as local
database <b>go</b>.





<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
<ol>

<li>

</ol>

   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="edamdef.html">edamdef</a></td>
<td>Find EDAM ontology terms by definition</td>
</tr>

<tr>
<td><a href="edamhasinput.html">edamhasinput</a></td>
<td>Find EDAM ontology terms by has_input relation</td>
</tr>

<tr>
<td><a href="edamhasoutput.html">edamhasoutput</a></td>
<td>Find EDAM ontology terms by has_output relation</td>
</tr>

<tr>
<td><a href="edamisformat.html">edamisformat</a></td>
<td>Find EDAM ontology terms by is_format_of relation</td>
</tr>

<tr>
<td><a href="edamisid.html">edamisid</a></td>
<td>Find EDAM ontology terms by is_identifier_of relation</td>
</tr>

<tr>
<td><a href="edamname.html">edamname</a></td>
<td>Find EDAM ontology terms by name</td>
</tr>

<tr>
<td><a href="goname.html">goname</a></td>
<td>Find GO ontology terms by name</td>
</tr>

<tr>
<td><a href="ontoget.html">ontoget</a></td>
<td>Get ontology term(s)</td>
</tr>

<tr>
<td><a href="ontogetcommon.html">ontogetcommon</a></td>
<td>Get common ancestor for terms</td>
</tr>

<tr>
<td><a href="ontogetdown.html">ontogetdown</a></td>
<td>Get ontology term(s) by parent id</td>
</tr>

<tr>
<td><a href="ontogetobsolete.html">ontogetobsolete</a></td>
<td>Get ontology ontology terms</td>
</tr>

<tr>
<td><a href="ontogetroot.html">ontogetroot</a></td>
<td>Get ontology root terms by child identifier</td>
</tr>

<tr>
<td><a href="ontogetsibs.html">ontogetsibs</a></td>
<td>Get ontology term(s) by id with common parent</td>
</tr>

<tr>
<td><a href="ontogetup.html">ontogetup</a></td>
<td>Get ontology term(s) by id of child</td>
</tr>

<tr>
<td><a href="ontoisobsolete.html">ontoisobsolete</a></td>
<td>Report whether an ontology term id is obsolete</td>
</tr>

<tr>
<td><a href="ontotext.html">ontotext</a></td>
<td>Get ontology term(s) original full text</td>
</tr>

</table>
<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
Peter Rice
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK

<p>
Please report all bugs to the EMBOSS bug team (emboss-bug&nbsp;&copy;&nbsp;emboss.open-bio.org) not to the original author.

<H2>
    History
</H2>
<!--
        Date written and what changes have been made go in this file.
   -->




<H2>
    Target users
</H2>
<!--
        For general users, use this text
   -->
This program is intended to be used by everyone and everything, from naive users to embedded scripts.

<H2>
    Comments
</H2>
<!--
        User/developer/other comments go in this file.
   -->
None


</BODY>
</HTML>

