<HTML>
<HEAD>
  <TITLE>
  EMBOSS: godef
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
godef
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
Find GO ontology terms by definition
<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>


<b>godef</b> searches the definition of gene ontology terms and returns matching terms.
The input is read from the installed GO database. The ontology term output can be written to screen, to file, or passed to another program. A wide range of standard ontology term formats may be specified for input and output.

<p>
Optionally the search can be restricted to specified GO namespaces.




<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

Here is a sample session with <b>godef</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>godef glycoprotein </b>
Find GO ontology terms by definition
Obo output file [godef.obo]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
Find GO ontology terms by definition
Version: EMBOSS:6.4.0.0

   Standard (Mandatory) qualifiers:
  [-query]             string     Definition word(s) to search for in ontology
                                  (Any string)
  [-outfile]           outobo     [*.godef] Output ontology term file name

   Additional (Optional) qualifiers:
   -namespace          menu       [*] By default all terms are returned.
                                  Searches can be limited to one or a few
                                  namespaces. (Values: biological_process
                                  (Biological process); cellular_component
                                  (Cellular component); molecular_function
                                  (Molecular function))

   Advanced (Unprompted) qualifiers:
   -subclasses         boolean    [N] Extend the query matches to include all
                                  terms which are specialisations (GO
                                  sub-classes) of the matched type.
   -obsolete           boolean    [N] The default behaviour is to not use or
                                  return obsolete terms. This option if set
                                  will include all terms.

   Associated qualifiers:

   "-outfile" associated qualifiers
   -odirectory2        string     Output directory
   -oformat2           string     Ontology term output format

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-query]<br>(Parameter 1)</td>
<td>string</td>
<td>Definition word(s) to search for in ontology</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-outfile]<br>(Parameter 2)</td>
<td>outobo</td>
<td>Output ontology term file name</td>
<td>OBO ontology term(s)</td>
<td><i>&lt;*&gt;</i>.godef</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>-namespace</td>
<td>list</td>
<td>By default all terms are returned. Searches can be limited to one or a few namespaces.</td>
<td><table><tr><td>biological_process</td> <td><i>(Biological process)</i></td></tr><tr><td>cellular_component</td> <td><i>(Cellular component)</i></td></tr><tr><td>molecular_function</td> <td><i>(Molecular function)</i></td></tr></table></td>
<td>*</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>-subclasses</td>
<td>boolean</td>
<td>Extend the query matches to include all terms which are specialisations (GO sub-classes) of the matched type.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-obsolete</td>
<td>boolean</td>
<td>The default behaviour is to not use or return obsolete terms. This option if set will include all terms.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-outfile" associated outobo qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -odirectory2<br>-odirectory_outfile</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -oformat2<br>-oformat_outfile</td>
<td>string</td>
<td>Ontology term output format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>godef</b> queries the gene ontology.

<p>






<H2>
    Output file format
</H2>

<p>

The output is a standard EMBOSS ontology term file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <tt>-oformat xxx</tt>, where 'xxx' is replaced by
the name of the required format.  The available format names are: obo,
brief, list, html, xml, json, excel.
<p>

See:
<A href="http://emboss.sf.net/docs/themes/OntologyFormats.html">
http://emboss.sf.net/docs/themes/OntologyFormats.html</A>
for further information on ontology formats.

<p>

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: godef.obo</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
[Term]
id: GO:0004579
name: dolichyl-diphosphooligosaccharide-protein glycotransferase activity
namespace: molecular_function
def: Catalysis of the reaction: dolichyl diphosphooligosaccharide + protein L-asparagine = dolichyl diphosphate + a glycoprotein with the oligosaccharide chain attached by glycosylamine linkage to protein L-asparagine.
synonym: "asparagine N-glycosyltransferase activity" EXACT [EC:2.4.1.119]
synonym: "dolichyl-diphosphooligosaccharide-protein glycosyltransferase activity" EXACT []
synonym: "dolichyl-diphosphooligosaccharide:protein-L-asparagine oligopolysaccharidotransferase activity" EXACT [EC:2.4.1.119]
synonym: "dolichyldiphosphooligosaccharide-protein glycosyltransferase activity" EXACT [EC:2.4.1.119]
synonym: "dolichyldiphosphooligosaccharide-protein oligosaccharyltransferase activity" EXACT [EC:2.4.1.119]
synonym: "dolichyldiphosphoryloligosaccharide-protein oligosaccharyltransferase activity" EXACT [EC:2.4.1.119]
synonym: "dolichylpyrophosphodiacetylchitobiose-protein glycosyltransferase activity" EXACT [EC:2.4.1.119]
synonym: "oligomannosyltransferase activity" RELATED [EC:2.4.1.119]
xref: EC:2.4.1.119
xref: MetaCyc:2.4.1.119-RXN
xref: Reactome:12500 "dolichyl-diphosphooligosaccharide-protein glycotransferase activity"
is_a: GO:0004576 ! oligosaccharyl transferase activity

[Term]
id: GO:0015572
name: N-acetylglucosamine transmembrane transporter activity
namespace: molecular_function
def: Catalysis of the transfer of N-acetylglucosamine from one side of the membrane to the other. The D isomer of N-acetylglucosamine is a common structural unit of glycoproteins in plants, bacteria and animals; it is often the terminal sugar of an oligosaccharide group of a glycoprotein.
alt_id: GO:0015580
subset: gosubset_prok
synonym: "D-GlcNAc transmembrane transporter activity" EXACT []
synonym: "N-Acetyl-D-glucosamine permease" RELATED []
synonym: "N-acetyl-D-glucosamine transmembrane transporter activity" EXACT []
synonym: "N-acetylchitosamine transmembrane transporter activity" EXACT []
synonym: "N-acetylglucosamine permease activity" RELATED []
is_a: GO:0015149 ! hexose transmembrane transporter activity

[Term]
id: GO:0002162
name: dystroglycan binding
namespace: molecular_function
def: Interacting selectively and non-covalently with dystroglycan. Dystroglycan is glycoprotein found in non-muscle tissues as well as in muscle tissues, often in association with dystrophin. The native dystroglycan cleaved into two non-covalently associated subunits, alpha (N-terminal) and beta (C-terminal).
is_a: GO:0001948 ! glycoprotein binding

[Term]
id: GO:0031150
name: sorocarp stalk development
namespace: biological_process
def: The process whose specific outcome is the progression of the sorocarp stalk over time, from its formation to the mature structure. The sorocarp stalk is a tubular structure that consists of cellulose-covered cells stacked on top of each other and surrounded by an acellular stalk tube composed of cellulose and glycoprotein. An example of this process is found in Dictyostelium discoideum.
synonym: "sorophore development" EXACT []
synonym: "stalk development" BROAD []
synonym: "stalk formation" BROAD []
is_a: GO:0003006 ! developmental process involved in reproduction
is_a: GO:0048856 ! anatomical structure development
relationship: part_of GO:0031154 ! culmination involved in sorocarp development


<font color=red>  [Part of this file has been deleted for brevity]</font>

synonym: "positive regulation of glycoprotein biosynthetic process of formation of immunological synapse" EXACT [GOC:obol]
synonym: "positive regulation of glycoprotein biosynthetic process of immunological synapse formation" EXACT [GOC:obol]
is_a: GO:0010560 ! positive regulation of glycoprotein biosynthetic process
relationship: part_of GO:0001771 ! immunological synapse formation

[Term]
id: GO:0016914
name: follicle-stimulating hormone complex
namespace: cellular_component
def: A gonadotrophic glycoprotein hormone secreted, in mammals, by the anterior pituitary gland; consists of alpha and beta subunits, the latter of which confers hormonal specificity.
synonym: "follicle stimulating hormone complex" EXACT []
synonym: "FSH complex" EXACT []
is_a: GO:0043234 ! protein complex
is_a: GO:0044421 ! extracellular region part

[Term]
id: GO:0043256
name: laminin complex
namespace: cellular_component
def: A large, extracellular glycoprotein complex composed of three different polypeptide chains, alpha, beta and gamma. Provides an integral part of the structural scaffolding of basement membranes.
subset: goslim_pir
is_a: GO:0043234 ! protein complex
is_a: GO:0044420 ! extracellular matrix part
relationship: part_of GO:0005605 ! basal lamina

[Term]
id: GO:0048829
name: root cap development
namespace: biological_process
def: The process whose specific outcome is the progression of the root cap over time, from its formation to the mature structure. The root cap protects the root meristem from friction as the root grows through the soil. The cap is made up of a group of parenchyma cells which secrete a glycoprotein mucilage as a lubricant.
is_a: GO:0048856 ! anatomical structure development
relationship: part_of GO:0048364 ! root development

[Term]
id: GO:0051541
name: elastin metabolic process
namespace: biological_process
def: The chemical reactions and pathways involving elastin, a glycoprotein which is randomly coiled and crosslinked to form elastic fibers that are found in connective tissue.
synonym: "elastin metabolism" EXACT []
is_a: GO:0009100 ! glycoprotein metabolic process

[Term]
id: GO:0033947
name: mannosylglycoprotein endo-beta-mannosidase activity
namespace: molecular_function
def: Catalysis of the hydrolysis of the alpha-D-mannosyl-(1-&gt;6)-beta-D-mannosyl-(1-&gt;4)-beta-D-N-acetylglucosaminyl-(1-&gt;4)-beta-D-N-acetylglucosaminyl sequence of glycoprotein to alpha-D-mannosyl-(1-&gt;6)-D-mannose and beta-D-N-acetylglucosaminyl-(1-&gt;4)-beta-D-N-acetylglucosaminyl sequences.
synonym: "endo-beta-mannosidase activity" EXACT [EC:3.2.1.152]
xref: EC:3.2.1.152
xref: MetaCyc:3.2.1.152-RXN
is_a: GO:0015923 ! mannosidase activity

</pre>
</td></tr></table><p>





<H2>
    Data files
</H2>

The gene Ontology is included in EMBOSS as local
database <b>go</b>.





<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
<ol>

<li>

</ol>

   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="edamdef.html">edamdef</a></td>
<td>Find EDAM ontology terms by definition</td>
</tr>

<tr>
<td><a href="edamhasinput.html">edamhasinput</a></td>
<td>Find EDAM ontology terms by has_input relation</td>
</tr>

<tr>
<td><a href="edamhasoutput.html">edamhasoutput</a></td>
<td>Find EDAM ontology terms by has_output relation</td>
</tr>

<tr>
<td><a href="edamisformat.html">edamisformat</a></td>
<td>Find EDAM ontology terms by is_format_of relation</td>
</tr>

<tr>
<td><a href="edamisid.html">edamisid</a></td>
<td>Find EDAM ontology terms by is_identifier_of relation</td>
</tr>

<tr>
<td><a href="edamname.html">edamname</a></td>
<td>Find EDAM ontology terms by name</td>
</tr>

<tr>
<td><a href="goname.html">goname</a></td>
<td>Find GO ontology terms by name</td>
</tr>

<tr>
<td><a href="ontoget.html">ontoget</a></td>
<td>Get ontology term(s)</td>
</tr>

<tr>
<td><a href="ontogetcommon.html">ontogetcommon</a></td>
<td>Get common ancestor for terms</td>
</tr>

<tr>
<td><a href="ontogetdown.html">ontogetdown</a></td>
<td>Get ontology term(s) by parent id</td>
</tr>

<tr>
<td><a href="ontogetobsolete.html">ontogetobsolete</a></td>
<td>Get ontology ontology terms</td>
</tr>

<tr>
<td><a href="ontogetroot.html">ontogetroot</a></td>
<td>Get ontology root terms by child identifier</td>
</tr>

<tr>
<td><a href="ontogetsibs.html">ontogetsibs</a></td>
<td>Get ontology term(s) by id with common parent</td>
</tr>

<tr>
<td><a href="ontogetup.html">ontogetup</a></td>
<td>Get ontology term(s) by id of child</td>
</tr>

<tr>
<td><a href="ontoisobsolete.html">ontoisobsolete</a></td>
<td>Report whether an ontology term id is obsolete</td>
</tr>

<tr>
<td><a href="ontotext.html">ontotext</a></td>
<td>Get ontology term(s) original full text</td>
</tr>

</table>
<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
Peter Rice
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK

<p>
Please report all bugs to the EMBOSS bug team (emboss-bug&nbsp;&copy;&nbsp;emboss.open-bio.org) not to the original author.

<H2>
    History
</H2>
<!--
        Date written and what changes have been made go in this file.
   -->




<H2>
    Target users
</H2>
<!--
        For general users, use this text
   -->
This program is intended to be used by everyone and everything, from naive users to embedded scripts.

<H2>
    Comments
</H2>
<!--
        User/developer/other comments go in this file.
   -->
None


</BODY>
</HTML>

