<!-- START OF HEADER -->
<HTML><HEAD>
<TITLE> EMBASSY: HETPARSE documentation. </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

<H2> HETPARSE documentation
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<!-- END OF HEADER -->




<!-- CONTENTS
     This always includes the sections below.
     Other subsections can be added for individual applications.
-->
<H2>CONTENTS </H2>
<b> <a href="#1.0">1.0     SUMMARY                   </a></b><br>
<b> <a href="#2.0">2.0     INPUTS & OUTPUTS          </a></b><br>
<b> <a href="#3.0">3.0     INPUT FILE FORMAT         </a></b><br>
<b> <a href="#4.0">4.0     OUTPUT FILE FORMAT        </a></b><br>
<b> <a href="#5.0">5.0     DATA FILES                </a></b><br>
<b> <a href="#6.0">6.0     USAGE                     </a></b><br>    
<b> <a href="#7.0">7.0     KNOWN BUGS & WARNINGS     </a></b><br>    
<b> <a href="#8.0">8.0     NOTES                     </a></b><br>
<b> <a href="#9.0">9.0     DESCRIPTION               </a></b><br>
<b> <a href="#10.0">10.0   ALGORITHM                 </a></b><br>
<b> <a href="#11.0">11.0   RELATED APPLICATIONS      </a></b><br>
<b> <a href="#12.0">12.0   DIAGNOSTIC ERROR MESSAGES </a></b><br>
<b> <a href="#13.0">13.0   AUTHORS                   </a></b><br>
<b> <a href="#14.0">14.0   REFERENCES                </a></b><br>


<!-- SUMMARY
     Succint description of the application, particularly its inputs, outputs
     and what it does.  The same text is given at the top of the source (.c)
     file and in the <documentation> attribute of the <application definition>
     of the ACD file.
-->
<a name="1.0"></a>
<br><br><br><H2> 1.0   SUMMARY  </H2>
Converts raw dictionary of heterogen groups EMBL-like format.  
Converts heterogen group dictionary to EMBL-like format



<!-- INPUTS & OUTPUTS
     Short summary of the application inputs and outputs in its different 
     modes of usage (if appropriate). More detail than the summary.
-->
<a name="2.0"></a>
<br><br><br><H2> 2.0   INPUTS & OUTPUTS          </H2>
HETPARSE parse the dictionary of heterogen groups available at 
<a href="http://pdb.rutgers.edu/het_dictionary.txt">http://pdb.rutgers.edu/het_dictionary.txt </a> and writes a file containing the
group names, synonyms and 3-letter codes in EMBL-like format. 
Optionally, HETPARSE will search a directory of PDB files and will count 
the number of files that each heterogen appears in.  The path and extension 
of the PDB files and the names of the input and output files are user-
specified (file extension is set in the ACD file).


<!-- INPUT FILE FORMAT 
     Description and example(s) of input file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.
     Use "<b>HETPARSE</b> reads any normal sequence USAs." if
     appropriate.
-->



<a name="3.0"></a>
<br><br><br><H2> 3.0   INPUT FILE FORMAT         </H2>

An excerpt from the raw dictionary of heterogen group is shown (Figure 1).


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: het.txt</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
RESIDUE   061     58
CONECT      N1     2 N2   C5  
CONECT      N2     2 N1   N3  
CONECT      N3     2 N2   N4  
CONECT      N4     3 N3   C5   HN4 
CONECT      C5     3 N1   N4   C6  
CONECT      C6     3 C5   C7   C11 
CONECT      C7     3 C6   C8   C12 
CONECT      C8     3 C7   C9   H8  
CONECT      C9     3 C8   C10  H9  
CONECT      C10    3 C9   C11  H10 
CONECT      C11    3 C6   C10  H11 
CONECT      C12    3 C7   C13  C17 
CONECT      C13    3 C12  C14  H13 
CONECT      C14    3 C13  C15  H14 
CONECT      C15    3 C14  C16  C18 
CONECT      C16    3 C15  C17  H16 
CONECT      C17    3 C12  C16  H17 
CONECT      C18    4 C15  N19 1H18 2H18 
CONECT      N19    3 C18  C20  C33 
CONECT      C20    3 N19  C21  N25 
CONECT      C21    4 C20  C22 1H21 2H21 
CONECT      C22    4 C21  C23 1H22 2H22 
CONECT      C23    4 C22  C24 1H23 2H23 
CONECT      C24    4 C23 1H24 2H24 3H24 
CONECT      N25    2 C20  C26 
CONECT      C26    3 N25  C27  C32 
CONECT      C27    3 C26  C28  H27 
CONECT      C28    3 C27  C29  H28 
CONECT      C29    3 C28  O30  C31 
CONECT      O30    2 C29  HOU 
CONECT      C31    3 C29  C32  H31 
CONECT      C32    3 C26  C31  C33 
CONECT      C33    3 N19  C32  O34 
CONECT      O34    1 C33 
CONECT      HN4    1 N4  
CONECT      H8     1 C8  
CONECT      H9     1 C9  
CONECT      H10    1 C10 
CONECT      H11    1 C11 
CONECT      H13    1 C13 
CONECT      H14    1 C14 
CONECT      H16    1 C16 
CONECT      H17    1 C17 
CONECT     1H18    1 C18 
CONECT     2H18    1 C18 
CONECT     1H21    1 C21 
CONECT     2H21    1 C21 
CONECT     1H22    1 C22 
CONECT     2H22    1 C22 


<font color=red>  [Part of this file has been deleted for brevity]</font>

CONECT     2H6     1 C6  
CONECT     1H8     1 C8  
CONECT     2H8     1 C8  
CONECT     1H9     1 C9  
CONECT     2H9     1 C9  
END
HET    104             28
HETSYN     104 TRIENTINE
HETNAM     104 N,N'-BIS(2-AMINOETHYL)-1,2-ETHANEDIAMINE
FORMUL      104    C6 H18 N4

RESIDUE   105     32
CONECT      B      3 O1   O2   C3  
CONECT      O1     2 B    H1  
CONECT      O2     2 B    H2  
CONECT      C3     4 B    N4  1H3  2H3  
CONECT      N4     3 C3   C5   H4  
CONECT      C5     3 N4   O6   C7  
CONECT      O6     1 C5  
CONECT      C7     3 C5   C8   C12 
CONECT      N11    2 O10  C12 
CONECT      O10    2 N11  C8  
CONECT      C8     3 C7   O10  C9  
CONECT      C12    3 C7   N11  C13 
CONECT      C9     4 C8  1H9  2H9  3H9  
CONECT      C13    3 C12  C14  C18 
CONECT      C14    3 C13  C15 CL1  
CONECT     CL1     1 C14 
CONECT      C15    3 C14  C16  H15 
CONECT      C16    3 C15  C17  H16 
CONECT      C17    3 C16  C18  H17 
CONECT      C18    3 C13  C17  H18 
CONECT      H1     1 O1  
CONECT      H2     1 O2  
CONECT     1H3     1 C3  
CONECT     2H3     1 C3  
CONECT      H4     1 N4  
CONECT     1H9     1 C9  
CONECT     2H9     1 C9  
CONECT     3H9     1 C9  
CONECT      H15    1 C15 
CONECT      H16    1 C16 
CONECT      H17    1 C17 
CONECT      H18    1 C18 
END
HET    105             32
HETSYN     105 CLOXACILLIN DERIVATIVE
HETNAM     105 N-[5-METHYL-3-O-TOLYL-ISOXAZOLE-4-CARBOXYLIC ACID
HETNAM   2 105 AMIDE] BORONIC ACID
FORMUL      105    C12 H12 N2 O4 B1 CL1

</pre>
</td></tr></table><p>







<!-- OUTPUT FILE FORMAT 
     Description and example(s) of output file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.  
     If the standard description of the avalable report formats is required, 
     use:   #include file="inc/reportformats.ihtml"
     Use "Outputs a graph to the specified graphics device."
     or  "outputs a report format file. The default format is ..."
     if appropriate.
-->
<a name="4.0"></a>
<br><br><br><H2> 4.0   OUTPUT FILE FORMAT         </H2>

The records used in the output file (Figure 2) are as follows:
<UL>
<LI> ID - 3-character abbreviation of heterogen
<LI> DE - full description
<LI> SY - synonym
<LI> NN - number of files which this heterogen appears in
</UL>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: Ehet.dat</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
ID   105
DE   N-[5-METHYL-3-O-TOLYL-ISOXAZOLE-4-CARBOXYLIC ACIDAMIDE] BORONIC ACID
SY   CLOXACILLIN DERIVATIVE
NN   0
//
ID   104
DE   N,N'-BIS(2-AMINOETHYL)-1,2-ETHANEDIAMINE
SY   TRIENTINE
NN   0
//
ID   103
DE   2',5'-DIDEOXY-ADENOSINE 3'-MONOPHOSPHATE
SY   .
NN   0
//
ID   102
DE   GAMMA-DEOXY-GAMMA-SULFO-GUANOSINE-5'-TRIPHOSPHATE
SY   .
NN   0
//
ID   101
DE   2'-DEOXY-ADENOSINE 3'-MONOPHOSPHATE
SY   .
NN   0
//
ID   100
DE   1-(5-CHLOROINDOL-3-YL)-3-HYDROXY-3-(2H-TETRAZOL-5-YL)-PROPENONE
SY   .
NN   0
//
ID   074
DE   [PROPYLAMINO-3-HYDROXY-BUTAN-1,4-DIONYL]-ISOLEUCYL-PROLINE
SY   CA-074;
SY   [N-(L-3-TRANS-PROPYLCARBAMOYL-OXIRANE-2-CARBONYL)-L-ISOLEUCYL-L-PROLINE]
NN   0
//
ID   072
DE   
DE   (+/-)(2S,5S)-3-(4-(4-CARBOXYPHENYL)BUTYL)-2-HEPTYL-4-OXO-5-THIAZOLIDINE
SY   THIAZOLIDINONE; GW0072
NN   0
//
ID   061
DE   
DE   2-BUTYL-6-HYDROXY-3-[2'-(1H-TETRAZOL-5-YL)-BIPHENYL-4-YLMETHYL]-3H-QUINAZOLIN-4-ONE
SY   L-159,061
NN   0
//
</pre>
</td></tr></table><p>




<!-- DATA FILES         
     Any data files used (e.g. translation table file, substitution matrix 
     etc.  This includes example data file formats if they are not obvious.
     For a standard description of what data files are and how embossdata can
     be used to inspect and retrieve them, use:
     #include file="inc/localfiles.ihtml" 
-->

<a name="5.0"></a>
<br><br><br><H2> 5.0   DATA FILES                </H2>
HETPARSE does not use a data file.





<!-- USAGE
     Example usage, as run from the command-line.
     Many examples illustrating different behaviours is good.
-->
<a name="6.0"></a>
<br><br><br><H2> 6.0   USAGE                     </H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" >
<tr><td>
<pre>
   Standard (Mandatory) qualifiers (* if not always prompted):
  [-infile]            infile     This option specifies the name of input file
                                  (raw dictionary of heterogen groups) to
                                  parse, which should be of the format
                                  specified at
                                  http://pdb.rutgers.edu/het_dictionary.txt
   -dogrep             toggle     This option specifies whether to search a
                                  directory of files (typically PDB files)
                                  with keywords. If set, HETPARSE will search
                                  the directory and will count the number of
                                  files that each heterogen appears in.
*  -dirlistpath        dirlist    This option specifies the directory to
                                  search with keywords.
  [-outfile]           outfile    This option specifies the name of EMBL-like
                                  format dictionary of heterogen groups.

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-outfile" associated qualifiers
   -odirectory2        string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report deaths

</pre>
</pre>
</td></tr>
</table>


<H3> 6.1   COMMAND LINE ARGUMENTS </H3>
 
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-infile]<br>(Parameter 1)</td>
<td>This option specifies the name of input file (raw dictionary of heterogen groups) to parse, which should be of the format specified at http://pdb.rutgers.edu/het_dictionary.txt</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-dogrep</td>
<td>This option specifies whether to search a directory of files (typically PDB files) with keywords. If set, HETPARSE will search the directory and will count the number of files that each heterogen appears in.</td>
<td>Toggle value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-dirlistpath</td>
<td>This option specifies the directory to search with keywords.</td>
<td>Directory with files</td>
<td>./</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>This option specifies the name of EMBL-like format dictionary of heterogen groups.</td>
<td>Output file</td>
<td>Ehet.dat</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<H3> 6.2 EXAMPLE SESSION </H3>
An example of interactive use of HETPARSE is shown below.

<b>Here is a sample session with hetparse</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>hetparse </b>
Converts heterogen group dictionary to EMBL-like format.
Name of input file (raw dictionary of heterogen groups): <b>het.txt</b>
Search a directory of PDB files with keywords? [N]: <b>Y</b>
Directory to search with keywords [./]: <b></b>
Name of EMBL-like format dictionary of heterogen groups. [Ehet.dat]: <b>Ehet.dat</b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>


<!-- <br><br>
A file containing the raw (unprocessed) dictionary of heterogen groups
called testdata/het.txt was read and a reformatted 
dictionary file called testdata/hetparse/Ehet.dat was written.
HETPARSE searched in the directory testdata/hetparse for PDB files
and made counts of the number of files that
each heterogen in the dictionary file appeared in.  Two PDB files 
(pdb1cs4.ent and pdb1ii7.ent) were read.
<br><br>


<!-- KNOWN BUGS & WARNINGS
     Bugs that have not yet been fixed, easily missued features, problems
     and caveats etc. Potentially stupid things the program will let you do.
-->
<a name="7.0"></a>
<br><br><br><H2> 7.0   KNOWN BUGS & WARNINGS     </H2>
None.








<!-- NOTES
     Important general remarks, including:
       Restrictions.
       Interesting behaviour.
       Useful things you can do with this program.
       Future plans.
       etc.
-->
<a name="8.0"></a>
<br><br><br><H2> 8.0   NOTES                     </H2>
HETPARSE is used to create the EMBOSS data file Ehet.dat that is 
included in the EMBOSS distribution.


<H3> 8.1   GLOSSARY OF FILE TYPES    </H3>
<table BORDER CELLSPACING=5 CELLPADDING=5 BGCOLOR="#f5f5ff" >
<tr>
<td><b>FILE TYPE</b></td>
<td><b>FORMAT</b></td>
<td><b>DESCRIPTION</b></td>
<td><b>CREATED BY <b></td>
<td><b>SEE ALSO</b></td>
</tr>
<tr>
<td><b>Dictionary of heterogen groups  </b></td>
<td> A file of the dictionary of heterogen groups in PDB. </td>
<td> <a href="hetparse.html">HETPARSE</a> </td>
<td> N.A. </td>
</tr>
</table>




None


<!-- DESCRIPTION
     A complete, non-technical, user-level description of the application.
-->
<a name="9.0"></a>
<br><br><br><H2> 9.0   DESCRIPTION               </H2>
Some research applications require knowledge of the types of small molecules or
'heterogens' (non-protein groups) that are represented in PDB files.  A dictionary
of such groups containing various data for all of the heterogens
found in PDB is available, but is not in a convenient format.
HETPARSE parses the dictionary in its raw format and converts it to an
EMBL-like format.




<!-- ALGORITHM
     A technical description of algorithmic aspects, describing exactly how
     the key aspects of the application work.
-->
<a name="10.0"></a>
<br><br><br><H2> 10.0  ALGORITHM                 </H2>

None.



<!-- RELATED APPLICATIONS
     Other applications that either generate the input, use the output or 
     are in some other way related to the application are described here. 
-->

<a name="11.0"></a>
<br><br><br><H2> 11.0  RELATED APPLICATIONS      </H2>


<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="../../apps/aaindexextract.html">aaindexextract</a></td><td>Extract data from AAINDEX</td></tr>
<tr><td><a href="../domalign/allversusall.html">allversusall</a></td><td>Sequence similarity data from all-versus-all comparison</td></tr>
<tr><td><a href="../domainatrix/cathparse.html">cathparse</a></td><td>Generates DCF file from raw CATH files</td></tr>
<tr><td><a href="../../apps/cutgextract.html">cutgextract</a></td><td>Extract data from CUTG</td></tr>
<tr><td><a href="domainer.html">domainer</a></td><td>Generates domain CCF files from protein CCF files</td></tr>
<tr><td><a href="../domainatrix/domainnr.html">domainnr</a></td><td>Removes redundant domains from a DCF file</td></tr>
<tr><td><a href="../domainatrix/domainseqs.html">domainseqs</a></td><td>Adds sequence records to a DCF file</td></tr>
<tr><td><a href="../domainatrix/domainsse.html">domainsse</a></td><td>Add secondary structure records to a DCF file</td></tr>
<tr><td><a href="pdbparse.html">pdbparse</a></td><td>Parses PDB files and writes protein CCF files</td></tr>
<tr><td><a href="pdbplus.html">pdbplus</a></td><td>Add accessibility & secondary structure to a CCF file</td></tr>
<tr><td><a href="pdbtosp.html">pdbtosp</a></td><td>Convert swissprot:PDB codes file to EMBL-like format</td></tr>
<tr><td><a href="../../apps/printsextract.html">printsextract</a></td><td>Extract data from PRINTS</td></tr>
<tr><td><a href="../../apps/prosextract.html">prosextract</a></td><td>Build the PROSITE motif database for use by patmatmotifs</td></tr>
<tr><td><a href="../../apps/rebaseextract.html">rebaseextract</a></td><td>Extract data from REBASE</td></tr>
<tr><td><a href="../domainatrix/scopparse.html">scopparse</a></td><td>Generate DCF file from raw SCOP files</td></tr>
<tr><td><a href="../domsearch/seqnr.html">seqnr</a></td><td>Removes redundancy from DHF files</td></tr>
<tr><td><a href="sites.html">sites</a></td><td>Generate residue-ligand CON files from CCF files</td></tr>
<tr><td><a href="../domainatrix/ssematch.html">ssematch</a></td><td>Search a DCF file for secondary structure matches</td></tr>
<tr><td><a href="../../apps/tfextract.html">tfextract</a></td><td>Extract data from TRANSFAC</td></tr>
</table>




<!-- DIAGNOSTIC ERROR MESSAGES
     Description of error messages or log file, if one is written.
-->
<a name="12.0"></a>
<br><br><br><H2> 12.0  DIAGNOSTIC ERROR MESSAGES </H2>
None.



<!-- AUTHORS
     The main author first then all others. 
-->
<a name="13.0"></a>
<br><br><br><H2> 13.0  AUTHORS                   </H2>
Jon Ison <a href="mailto:jison@rfcgr.mrc.ac.uk">(jison@rfcgr.mrc.ac.uk)</a>
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK




<!-- REFERENCES
     Quote the paper where the application was first published, described, used etc. 
-->
<a name="14.0"></a>
<br><br><br><H2> 14.0  REFERENCES                </H2>

Please cite the authors and EMBOSS.
<br><br>
<i>Rice P, Longden I and Bleasby A (2000) "EMBOSS - The European
Molecular Biology Open Software Suite"  Trends in Genetics,
15:276-278.

<p>
See also <a href="http://emboss.sourceforge.net/">http://emboss.sourceforge.net/</a></i>




<H3>14.1 Other useful references  </H3>


</BODY>
</HTML>
