<HTML>
<HEAD>
  <TITLE>
  EMBOSS: hmoment
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

hmoment
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<H2>
    Function
</H2>
Hydrophobic moment calculation

<H2>
    Description
</H2>

<b>hmoment</b> plots or writes out the hydrophobic moment.  Hydrophic
moment is the hydrophobicity of a peptide measured for a specified angle
of rotation per residue. 

<p>

Periodicities in the polar/apolar character of the amino acid sequence
of a protein can be examined by assigning to each residue a numerical
hydrophobicity and searching for periodicity in the resulting
one-dimensional function.  The strength of each periodic component is
the quantity that has been termed the hydrophobic moment. 

<p>

When proteins of known three-dimensional structure are examined, it is
found that sequences that form alpha helices tend to have, on average, a
strong periodicity in the hydrophobicity of af 3.6 residues, the period
of the alpha helix. The angle of rotation per residue in alpha helices is 
100 degrees.

Similarly, many sequences that form strands of beta sheets tend to have
a periodicity in their hydrophobicity of about 2.3 residues, the period
typical of beta structure.  The angle of rotation per residue in beta
sheets is 160 degrees. 

This means that many protein sequences tend to form the periodic
structure that maximizes their amphiphilicity. 

<p>

The hydrophobic moment is measured within a moving window using the method
of Eisenberg et al.  The default angle of 100 degrees is used for the
alpha-helix results and the default of 160 degrees is used for the
beta-sheet results.  These angles can be changed if required using the
appropriate options. 

<p>

<b>hmoment</b> can plot two graphs when the option '-double' is given,
one for the alpha helix moment and one for the beta sheet moment. 
Otherwise it just plots the alpha helix moment.

<H2>
    Usage
</H2>
<b>Here is a sample session with hmoment</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>hmoment tsw:hbb_human </b>
Hydrophobic moment calculation
Output file [hbb_human.hmoment]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>




<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers (* if not always prompted):
  [-seqall]            seqall     Sequence database USA
*  -graph              xygraph    Graph type
*  -outfile            outfile    Output file name

   Additional (Optional) qualifiers:
   -window             integer    Window
   -aangle             integer    Alpha helix angle (degrees)
   -bangle             integer    Beta sheet angle (degrees)

   Advanced (Unprompted) qualifiers:
   -baseline           float      Graph marker line
   -plot               toggle     Produce graphic
   -double             boolean    Plot two graphs

   Associated qualifiers:

   "-seqall" associated qualifiers
   -sbegin1             integer    Start of each sequence to be used
   -send1               integer    End of each sequence to be used
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name

   "-graph" associated qualifiers
   -gprompt             boolean    Graph prompting
   -gtitle              string     Graph title
   -gsubtitle           string     Graph subtitle
   -gxtitle             string     Graph x axis title
   -gytitle             string     Graph y axis title
   -goutfile            string     Output file for non interactive displays
   -gdirectory          string     Output directory

   "-outfile" associated qualifiers
   -odirectory          string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-seqall]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-graph</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including postscript, ps, hpgl, hp7470, hp7580, meta, colourps, cps, xwindows, x11, tektronics, tekt, tek4107t, tek, none, null, text, data, xterm, png, xml</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr>
<td>-outfile</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.hmoment</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-window</td>
<td>Window</td>
<td>Any integer value</td>
<td>10</td>
</tr>

<tr>
<td>-aangle</td>
<td>Alpha helix angle (degrees)</td>
<td>Any integer value</td>
<td>100</td>
</tr>

<tr>
<td>-bangle</td>
<td>Beta sheet angle (degrees)</td>
<td>Any integer value</td>
<td>160</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-baseline</td>
<td>Graph marker line</td>
<td>Any numeric value</td>
<td>0.35</td>
</tr>

<tr>
<td>-plot</td>
<td>Produce graphic</td>
<td>Toggle value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-double</td>
<td>Plot two graphs</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

</table>


<H2>
    Input file format
</H2>

<b>hmoment</b> reads in a protein sequence USA.

<p>

<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tsw:hbb_human' is a sequence entry in the example protein database 'tsw'
<p>
<p><h3>Database entry: tsw:hbb_human</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   HBB_HUMAN      STANDARD;      PRT;   146 AA.
AC   P02023;
DT   21-JUL-1986 (Rel. 01, Created)
DT   21-JUL-1986 (Rel. 01, Last sequence update)
DT   15-JUL-1999 (Rel. 38, Last annotation update)
DE   HEMOGLOBIN BETA CHAIN.
GN   HBB.
OS   Homo sapiens (Human), Pan troglodytes (Chimpanzee), and
OS   Pan paniscus (Pygmy chimpanzee) (Bonobo).
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Mammalia;
OC   Eutheria; Primates; Catarrhini; Hominidae; Homo.
RN   [1]
RP   SEQUENCE.
RC   SPECIES=HUMAN;
RA   BRAUNITZER G., GEHRING-MULLER R., HILSCHMANN N., HILSE K., HOBOM G.,
RA   RUDLOFF V., WITTMANN-LIEBOLD B.;
RT   "The constitution of normal adult human haemoglobin.";
RL   Hoppe-Seyler's Z. Physiol. Chem. 325:283-286(1961).
RN   [2]
RP   SEQUENCE FROM N.A.
RC   SPECIES=HUMAN;
RX   MEDLINE; 81064667.
RA   LAWN R.M., EFSTRATIADIS A., O'CONNELL C., MANIATIS T.;
RT   "The nucleotide sequence of the human beta-globin gene.";
RL   Cell 21:647-651(1980).
RN   [3]
RP   SEQUENCE OF 121-146 FROM N.A.
RC   SPECIES=HUMAN;
RX   MEDLINE; 85205333.
RA   LANG K.M., SPRITZ R.A.;
RT   "Cloning specific complete polyadenylylated 3'-terminal cDNA
RT   segments.";
RL   Gene 33:191-196(1985).
RN   [4]
RP   X-RAY CRYSTALLOGRAPHY (2.5 ANGSTROMS) OF DEOXYHEMOGLOBIN.
RC   SPECIES=HUMAN;
RX   MEDLINE; 76027820.
RA   FERMI G.;
RT   "Three-dimensional fourier synthesis of human deoxyhaemoglobin at
RT   2.5-A resolution: refinement of the atomic model.";
RL   J. Mol. Biol. 97:237-256(1975).
RN   [5]
RP   SEQUENCE.
RC   SPECIES=P.TROGLODYTES;
RX   MEDLINE; 66071496.
RA   RIFKIN D.B., KONIGSBERG W.;
RT   "The characterization of the tryptic peptides from the hemoglobin of
RT   the chimpanzee (Pan troglodytes).";
RL   Biochim. Biophys. Acta 104:457-461(1965).
RN   [6]


<font color=red>  [Part of this file has been deleted for brevity]</font>

FT   VARIANT     140    140       A -&gt; T (IN ST JACQUES: O2 AFFINITY UP).
FT                                /FTId=VAR_003081.
FT   VARIANT     140    140       A -&gt; V (IN PUTTELANGE; POLYCYTHEMIA;
FT                                O2 AFFINITY UP).
FT                                /FTId=VAR_003082.
FT   VARIANT     141    141       L -&gt; R (IN OLMSTED; UNSTABLE).
FT                                /FTId=VAR_003083.
FT   VARIANT     142    142       A -&gt; D (IN OHIO; O2 AFFINITY UP).
FT                                /FTId=VAR_003084.
FT   VARIANT     143    143       H -&gt; D (IN RANCHO MIRAGE).
FT                                /FTId=VAR_003085.
FT   VARIANT     143    143       H -&gt; Q (IN LITTLE ROCK; O2 AFFINITY UP).
FT                                /FTId=VAR_003086.
FT   VARIANT     143    143       H -&gt; P (IN SYRACUSE; O2 AFFINITY UP).
FT                                /FTId=VAR_003087.
FT   VARIANT     143    143       H -&gt; R (IN ABRUZZO; O2 AFFINITY UP).
FT                                /FTId=VAR_003088.
FT   VARIANT     144    144       K -&gt; E (IN MITO; O2 AFFINITY UP).
FT                                /FTId=VAR_003089.
FT   VARIANT     145    145       Y -&gt; C (IN RAINIER; O2 AFFINITY UP).
FT                                /FTId=VAR_003090.
FT   VARIANT     145    145       Y -&gt; H (IN BETHESDA; O2 AFFINITY UP).
FT                                /FTId=VAR_003091.
FT   VARIANT     146    146       H -&gt; D (IN HIROSHIMA; O2 AFFINITY UP).
FT                                /FTId=VAR_003092.
FT   VARIANT     146    146       H -&gt; L (IN COWTOWN; O2 AFFINITY UP).
FT                                /FTId=VAR_003093.
FT   VARIANT     146    146       H -&gt; P (IN YORK; O2 AFFINITY UP).
FT                                /FTId=VAR_003094.
FT   VARIANT     146    146       H -&gt; Q (IN KODAIRA; O2 AFFINITY UP).
FT                                /FTId=VAR_003095.
FT   HELIX         5     15
FT   TURN         16     17
FT   HELIX        20     34
FT   HELIX        36     41
FT   HELIX        43     45
FT   HELIX        51     55
FT   TURN         56     56
FT   HELIX        58     75
FT   TURN         76     77
FT   HELIX        78     94
FT   TURN         95     96
FT   TURN        100    100
FT   HELIX       101    121
FT   HELIX       124    142
FT   TURN        143    144
SQ   SEQUENCE   146 AA;  15867 MW;  EC9744C9 CRC32;
     VHLTPEEKSA VTALWGKVNV DEVGGEALGR LLVVYPWTQR FFESFGDLST PDAVMGNPKV
     KAHGKKVLGA FSDGLAHLDN LKGTFATLSE LHCDKLHVDP ENFRLLGNVL VCVLAHHFGK
     EFTPPVQAAY QKVVAGVANA LAHKYH
//
</pre>
</td></tr></table><p>


<H2>
    Output file format
</H2>

With the '-plot' option <b>hmoment</b> displays a graphical plot of the
hydrophobic moment of the specified alpha helix angle. With the '-plot -double'
options it also displays the beta-sheet hydrophobic moment.

<p>

Otherwise it writes out a files containing the two columns separated by
space or TAB characters.  The first column is the position of the start
of the window that the hydrophobic moment was calculated in.  The second is
the hydrophobic moment ('uH').  (If the option '-double' is given then the
beta-sheet angle hydrophobicity is given as a third column.)

<p>

<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: hbb_human.hmoment</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
HMOMENT of HBB_HUMAN from 1 to 146

Window: 10 Angle: 100 Max uH: 0.714
Position	uH
1		0.091
2		0.216
3		0.208
4		0.123
5		0.211
6		0.194
7		0.185
8		0.169
9		0.312
10		0.292
11		0.185
12		0.092
13		0.050
14		0.164
15		0.245
16		0.187
17		0.130
18		0.262
19		0.396
20		0.317
21		0.342
22		0.492
23		0.508
24		0.517
25		0.418
26		0.416
27		0.350
28		0.292
29		0.196
30		0.102
31		0.288
32		0.314
33		0.442
34		0.560
35		0.464
36		0.577
37		0.584
38		0.676
39		0.714
40		0.670
41		0.462
42		0.369
43		0.221
44		0.176
45		0.073
46		0.058


<font color=red>  [Part of this file has been deleted for brevity]</font>

87		0.179
88		0.201
89		0.065
90		0.041
91		0.110
92		0.181
93		0.261
94		0.312
95		0.200
96		0.263
97		0.369
98		0.310
99		0.362
100		0.495
101		0.585
102		0.450
103		0.488
104		0.547
105		0.315
106		0.311
107		0.252
108		0.187
109		0.215
110		0.295
111		0.425
112		0.303
113		0.421
114		0.528
115		0.487
116		0.436
117		0.501
118		0.532
119		0.409
120		0.447
121		0.308
122		0.383
123		0.252
124		0.361
125		0.374
126		0.314
127		0.219
128		0.310
129		0.361
130		0.439
131		0.433
132		0.507
133		0.336
134		0.344
135		0.207
136		0.234
137		0.299
</pre>
</td></tr></table><p>




<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

<ol>

<li>Eisenberg D, Weiss RM, Terwilliger TC 
"The hydrophobic moment detects periodicity in protein hydrophobicity."
Proc Natl Acad Sci U S A 1984 Jan;81(1):140-4

</ol>

<H2>
    Warnings
</H2>


None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with status 0.

<H2>
    Known bugs
</H2>

None.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="garnier.html">garnier</a></td><td>Predicts protein secondary structure</td></tr>
<tr><td><a href="helixturnhelix.html">helixturnhelix</a></td><td>Report nucleic acid binding motifs</td></tr>
<tr><td><a href="pepcoil.html">pepcoil</a></td><td>Predicts coiled coil regions</td></tr>
<tr><td><a href="pepnet.html">pepnet</a></td><td>Displays proteins as a helical net</td></tr>
<tr><td><a href="pepwheel.html">pepwheel</a></td><td>Shows protein sequences as helices</td></tr>
<tr><td><a href="tmap.html">tmap</a></td><td>Displays membrane spanning regions</td></tr>
</table>


<H2>
    Author(s)
</H2>

Alan Bleasby (ableasby&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK



<H2>
    History
</H2>

Written (March 2001) - Alan Bleasby.


<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.


<H2>
    Comments
</H2>
None

</BODY>
</HTML>

