<HTML>

<HEAD>
  <TITLE>
  EMBOSS: iep
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">


<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
iep
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<H2>
    Function
</H2>
Calculates the isoelectric point of a protein

<H2>
    Description
</H2>


This calculates the isoelectric point of a protein from its amino acid
composition assuming that no electrostatic interactions change the
propensity for ionization. 

<p>

Adjusting the pH of an aqueous protein solution to the point where the
numbers of positive and negative charges on the protein are equal brings
the protein to its isoelectric point.  This is often the point of lowest
solubility, presumably because it is the point at which there are fewest
intermolecular repulsions, so that the molecules tend to form
aggregates. 

<p>

The application can make a plot of the ionization curve with respect to
pH and can write an output file of the data. 

<H2>
    Usage
</H2>

<b>Here is a sample session with iep</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>iep tsw:laci_ecoli </b>
Calculates the isoelectric point of a protein
Output file [laci_ecoli.iep]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>
<p>
<b>Example 2</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>iep tsw:ifna2_human -disulphide 2 -lysinemodified 2  </b>
Calculates the isoelectric point of a protein
Output file [ifna2_human.iep]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.2">Go to the input files for this example</a><br><a href="#output.2">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers (* if not always prompted):
  [-sequence]          seqall     Protein sequence(s) filename and optional
                                  format, or reference (input USA)
*  -graph              xygraph    [$EMBOSS_GRAPHICS value, or x11] Graph type
                                  (ps, hpgl, hp7470, hp7580, meta, cps, x11,
                                  tekt, tek, none, data, xterm, png)
*  -outfile            outfile    [*.iep] Output file name

   Additional (Optional) qualifiers:
   -amino              integer    [1] Number of N-termini (Integer 0 or more)
   -[no]termini        boolean    [Y] Include charge at N and C terminus
   -lysinemodified     integer    [0] Number of modified lysines (Integer 0 or
                                  more)
   -disulphides        integer    [0] Number of disulphide bridges (Integer 0
                                  or more)

   Advanced (Unprompted) qualifiers:
   -step               float      [.5] Step value for pH (Number from 0.010 to
                                  1.000)
   -plot               toggle     [N] Plot charge vs pH
   -[no]report         toggle     [Y] Write results to a file

   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1            integer    Start of each sequence to be used
   -send1              integer    End of each sequence to be used
   -sreverse1          boolean    Reverse (if DNA)
   -sask1              boolean    Ask for begin/end/reverse
   -snucleotide1       boolean    Sequence is nucleotide
   -sprotein1          boolean    Sequence is protein
   -slower1            boolean    Make lower case
   -supper1            boolean    Make upper case
   -sformat1           string     Input sequence format
   -sdbname1           string     Database name
   -sid1               string     Entryname
   -ufo1               string     UFO features
   -fformat1           string     Features format
   -fopenfile1         string     Features file name

   "-graph" associated qualifiers
   -gprompt            boolean    Graph prompting
   -gtitle             string     Graph title
   -gsubtitle          string     Graph subtitle
   -gxtitle            string     Graph x axis title
   -gytitle            string     Graph y axis title
   -goutfile           string     Output file for non interactive displays
   -gdirectory         string     Output directory

   "-outfile" associated qualifiers
   -odirectory         string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Protein sequence(s) filename and optional format, or reference (input USA)</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-graph</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including ps, hpgl, hp7470, hp7580, meta, cps, x11, tekt, tek, none, data, xterm, png</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr>
<td>-outfile</td>
<td>Output file name</td>
<td>Output file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-amino</td>
<td>Number of N-termini</td>
<td>Integer 0 or more</td>
<td>1</td>
</tr>

<tr>
<td>-[no]termini</td>
<td>Include charge at N and C terminus</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-lysinemodified</td>
<td>Number of modified lysines</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr>
<td>-disulphides</td>
<td>Number of disulphide bridges</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-step</td>
<td>Step value for pH</td>
<td>Number from 0.010 to 1.000</td>
<td>.5</td>
</tr>

<tr>
<td>-plot</td>
<td>Plot charge vs pH</td>
<td>Toggle value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]report</td>
<td>Write results to a file</td>
<td>Toggle value Yes/No</td>
<td>Yes</td>
</tr>

</table>




<H2>
    Input file format
</H2>

<p>iep</p> reads in any protein sequence USA.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tsw:laci_ecoli' is a sequence entry in the example protein database 'tsw'
<p>
<p><h3>Database entry: tsw:laci_ecoli</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   LACI_ECOLI     STANDARD;      PRT;   360 AA.
AC   P03023; P71309; Q47338; O09196;
DT   21-JUL-1986 (Rel. 01, Created)
DT   01-NOV-1997 (Rel. 35, Last sequence update)
DT   15-DEC-1998 (Rel. 37, Last annotation update)
DE   LACTOSE OPERON REPRESSOR.
GN   LACI.
OS   Escherichia coli.
OC   Bacteria; Proteobacteria; gamma subdivision; Enterobacteriaceae;
OC   Escherichia.
RN   [1]
RP   SEQUENCE FROM N.A.
RX   MEDLINE; 78246991.
RA   FARABAUGH P.J.;
RT   "Sequence of the lacI gene.";
RL   Nature 274:765-769(1978).
RN   [2]
RP   SEQUENCE FROM N.A.
RC   STRAIN=K12 / MG1655;
RX   MEDLINE; 97426617.
RA   BLATTNER F.R., PLUNKETT G. III, BLOCH C.A., PERNA N.T., BURLAND V.,
RA   RILEY M., COLLADO-VIDES J., GLASNER F.D., RODE C.K., MAYHEW G.F.,
RA   GREGOR J., DAVIS N.W., KIRKPATRICK H.A., GOEDEN M.A., ROSE D.J.,
RA   MAU B., SHAO Y.;
RT   "The complete genome sequence of Escherichia coli K-12.";
RL   Science 277:1453-1474(1997).
RN   [3]
RP   SEQUENCE FROM N.A.
RC   STRAIN=K12 / MG1655;
RA   DUNCAN M., ALLEN E., ARAUJO R., APARICIO A.M., CHUNG E., DAVIS K.,
RA   FEDERSPIEL N., HYMAN R., KALMAN S., KOMP C., KURDI O., LEW H.,
RA   LIN D., NAMATH A., OEFNER P., ROBERTS D., SCHRAMM S., DAVIS R.W.;
RL   Submitted (NOV-1996) to the EMBL/GenBank/DDBJ databases.
RN   [4]
RP   SEQUENCE FROM N.A.
RA   CHEN J., MATTHEWS K.K.S.M.;
RL   Submitted (MAY-1991) to the EMBL/GenBank/DDBJ databases.
RN   [5]
RP   SEQUENCE FROM N.A.
RA   MARSH S.;
RL   Submitted (JAN-1997) to the EMBL/GenBank/DDBJ databases.
RN   [6]
RP   SEQUENCE OF 1-147; 159-230 AND 233-360.
RX   MEDLINE; 76091932.
RA   BEYREUTHER K., ADLER K., FANNING E., MURRAY C., KLEMM A., GEISLER N.;
RT   "Amino-acid sequence of lac repressor from Escherichia coli.
RT   Isolation, sequence analysis and sequence assembly of tryptic
RT   peptides and cyanogen-bromide fragments.";
RL   Eur. J. Biochem. 59:491-509(1975).
RN   [7]


<font color=red>  [Part of this file has been deleted for brevity]</font>

CC   between  the Swiss Institute of Bioinformatics  and the  EMBL outstation -
CC   the European Bioinformatics Institute.  There are no  restrictions on  its
CC   use  by  non-profit  institutions as long  as its content  is  in  no  way
CC   modified and this statement is not removed.  Usage  by  and for commercial
CC   entities requires a license agreement (See http://www.isb-sib.ch/announce/
CC   or send an email to license@isb-sib.ch).
CC   --------------------------------------------------------------------------
DR   EMBL; V00294; CAA23569.1; -.
DR   EMBL; J01636; AAA24052.1; -.
DR   EMBL; AE000141; AAC73448.1; -.
DR   EMBL; U73857; AAB18069.1; ALT_INIT.
DR   EMBL; X58469; CAA41383.1; -.
DR   EMBL; U86347; AAB47270.1; ALT_INIT.
DR   EMBL; U72488; AAB36549.1; -.
DR   EMBL; U78872; AAB37348.1; -.
DR   EMBL; U78873; AAB37351.1; -.
DR   EMBL; U78874; AAB37354.1; -.
DR   PIR; A03558; RPECL.
DR   PIR; S02540; S02540.
DR   PDB; 1LCC; 31-JAN-94.
DR   PDB; 1LCD; 31-JAN-94.
DR   PDB; 1LTP; 31-OCT-93.
DR   PDB; 1TLF; 31-JUL-95.
DR   PDB; 1LBG; 11-JUL-96.
DR   PDB; 1LBH; 11-JUL-96.
DR   PDB; 1LBI; 11-JUL-96.
DR   PDB; 1LQC; 12-FEB-97.
DR   ECO2DBASE; H039.0; 6TH EDITION.
DR   ECOGENE; EG10525; LACI.
DR   PFAM; PF00356; lacI; 1.
DR   PFAM; PF00532; Peripla_BP_like; 1.
DR   PROSITE; PS00356; HTH_LACI_FAMILY; 1.
KW   Transcription regulation; DNA-binding; Repressor; 3D-structure.
FT   DNA_BIND      6     25       H-T-H MOTIF.
FT   MUTAGEN      17     17       Y-&gt;H: BROADENING OF SPECIFICITY.
FT   MUTAGEN      22     22       R-&gt;N: RECOGNIZE AN OPERATOR VARIANT.
FT   VARIANT     282    282       Y -&gt; D (IN T41 MUTANT).
FT   CONFLICT    286    286       S -&gt; L (IN AAA24052, REF. 2, 4 AND 5).
FT   HELIX         6     13
FT   TURN         14     14
FT   HELIX        17     24
FT   HELIX        32     44
FT   TURN         49     50
SQ   SEQUENCE   360 AA;  38564 MW;  4CA5A1D6 CRC32;
     MKPVTLYDVA EYAGVSYQTV SRVVNQASHV SAKTREKVEA AMAELNYIPN RVAQQLAGKQ
     SLLIGVATSS LALHAPSQIV AAIKSRADQL GASVVVSMVE RSGVEACKAA VHNLLAQRVS
     GLIINYPLDD QDAIAVEAAC TNVPALFLDV SDQTPINSII FSHEDGTRLG VEHLVALGHQ
     QIALLAGPLS SVSARLRLAG WHKYLTRNQI QPIAEREGDW SAMSGFQQTM QMLNEGIVPT
     AMLVANDQMA LGAMRAITES GLRVGADISV VGYDDTEDSS CYIPPSTTIK QDFRLLGQTS
     VDRLLQLSQG QAVKGNQLLP VSLVKRKTTL APNTQTASPR ALADSLMQLA RQVSRLESGQ
//
</pre>
</td></tr></table><p>

<a name="input.2"></a>
<h3>Input files for usage example 2</h3>
<p><h3>Database entry: tsw:ifna2_human</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   IFNA2_HUMAN    STANDARD;      PRT;   188 AA.
AC   P01563; P01564; Q14606; Q96KI6;
DT   21-JUL-1986, integrated into UniProtKB/Swiss-Prot.
DT   21-JUL-1986, sequence version 1.
DT   27-JUN-2006, entry version 72.
DE   Interferon alpha-2 precursor (Interferon alpha-A) (LeIF A).
GN   Name=IFNA2;
OS   Homo sapiens (Human).
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi;
OC   Mammalia; Eutheria; Euarchontoglires; Primates; Haplorrhini;
OC   Catarrhini; Hominidae; Homo.
OX   NCBI_TaxID=9606;
RN   [1]
RP   NUCLEOTIDE SEQUENCE.
RX   MEDLINE=81052322; PubMed=6159538; DOI=10.1038/287411a0;
RA   Goeddel D.V., Yelverton E., Ullrich A., Heyneker H.L., Miozzari G.,
RA   Holmes W., Seeburg P.H., Dull T.J., May L., Stebbing N., Crea R.,
RA   Maeda S., McCandliss R., Sloma A., Tabor J.M., Gross M.,
RA   Familletti P.C., Pestka S.;
RT   "Human leukocyte interferon produced by E. coli is biologically
RT   active.";
RL   Nature 287:411-416(1980).
RN   [2]
RP   NUCLEOTIDE SEQUENCE.
RX   MEDLINE=81148795; PubMed=6163083; DOI=10.1038/290020a0;
RA   Goeddel D.V., Leung D.W., Dull T.J., Gross M., Lawn R.M.,
RA   McCandliss R., Seeburg P.H., Ullrich A., Yelverton E., Gray P.W.;
RT   "The structure of eight distinct cloned human leukocyte interferon
RT   cDNAs.";
RL   Nature 290:20-26(1981).
RN   [3]
RP   NUCLEOTIDE SEQUENCE.
RX   MEDLINE=82060261; PubMed=6170983;
RA   Lawn R.M., Gross M., Houck C.M., Franke A.E., Gray P.V., Goeddel D.V.;
RT   "DNA sequence of a major human leukocyte interferon gene.";
RL   Proc. Natl. Acad. Sci. U.S.A. 78:5435-5439(1981).
RN   [4]
RP   NUCLEOTIDE SEQUENCE.
RC   TISSUE=Bone marrow tumor;
RX   MEDLINE=86069501; PubMed=3906813;
RA   Oliver G., Balbas P., Valle F., Soberon X., Bolivar F.;
RT   "Cloning of human leukocyte interferon cDNA and a strategy for its
RT   production in E. coli.";
RL   Rev. Latinoam. Microbiol. 27:141-150(1985).
RN   [5]
RP   NUCLEOTIDE SEQUENCE.
RC   TISSUE=Placenta;
RX   MEDLINE=98357449; PubMed=9694076;
RA   Austruy E., Bagnis C., Carbuccia N., Maroc C., Birg F., Dubreuil P.,
RA   Mannoni P., Chabannon C.;


<font color=red>  [Part of this file has been deleted for brevity]</font>

DR   Ensembl; ENSG00000188379; Homo sapiens.
DR   HGNC; HGNC:5423; IFNA2.
DR   MIM; 147562; gene.
DR   LinkHub; P01563; -.
DR   RZPD-ProtExp; A0813; -.
DR   RZPD-ProtExp; IOH35221; -.
DR   RZPD-ProtExp; RZPDo834E0933; -.
DR   GO; GO:0005132; F:interferon-alpha/beta receptor binding; TAS.
DR   GO; GO:0007166; P:cell surface receptor linked signal transdu...; TAS.
DR   GO; GO:0007267; P:cell-cell signaling; TAS.
DR   GO; GO:0006917; P:induction of apoptosis; TAS.
DR   GO; GO:0006954; P:inflammatory response; TAS.
DR   InterPro; IPR009079; 4_helix_cytokine.
DR   InterPro; IPR000471; Interferon_abd.
DR   PANTHER; PTHR11691; Interferon_abd; 1.
DR   Pfam; PF00143; Interferon; 1.
DR   PRINTS; PR00266; INTERFERONAB.
DR   ProDom; PD000550; Interferon_abd; 1.
DR   SMART; SM00076; IFabd; 1.
DR   PROSITE; PS00252; INTERFERON_A_B_D; 1.
KW   3D-structure; Antiviral defense; Cytokine; Direct protein sequencing;
KW   Glycoprotein; Pharmaceutical; Polymorphism; Signal.
FT   SIGNAL        1     23
FT   CHAIN        24    188       Interferon alpha-2.
FT                                /FTId=PRO_0000016360.
FT   CARBOHYD    129    129       O-linked (GalNAc...).
FT                                /FTId=CAR_000049.
FT   DISULFID     24    121
FT   DISULFID     52    161
FT   VARIANT      46     46       K -&gt; R (in alpha-2B and alpha-2C).
FT                                /FTId=VAR_004012.
FT   VARIANT      57     57       H -&gt; R (in alpha-2C).
FT                                /FTId=VAR_013001.
FT   HELIX        33     44
FT   TURN         49     54
FT   HELIX        63     66
FT   HELIX        76     91
FT   HELIX        93     98
FT   HELIX       101    123
FT   TURN        126    127
FT   TURN        133    133
FT   HELIX       134    155
FT   TURN        156    157
FT   HELIX       160    178
FT   TURN        179    182
SQ   SEQUENCE   188 AA;  21550 MW;  101DD21D394CBF97 CRC64;
     MALTFALLVA LLVLSCKSSC SVGCDLPQTH SLGSRRTLML LAQMRKISLF SCLKDRHDFG
     FPQEEFGNQF QKAETIPVLH EMIQQIFNLF STKDSSAAWD ETLLDKFYTE LYQQLNDLEA
     CVIQGVGVTE TPLMKEDSIL AVRKYFQRIT LYLKEKKYSP CAWEVVRAEI MRSFSLSTNL
     QESLRSKE
//
</pre>
</td></tr></table><p>



<H2>
    Output file format
</H2>




<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: laci_ecoli.iep</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
IEP of LACI_ECOLI from 1 to 360
Isoelectric Point = 6.8820

   pH     Bound    Charge
  1.00    81.96    37.96
  1.50    81.89    37.89
  2.00    81.65    37.65
  2.50    80.91    36.91
  3.00    78.79    34.79
  3.50    73.70    29.70
  4.00    65.15    21.15
  4.50    56.73    12.73
  5.00    51.75     7.75
  5.50    49.36     5.36
  6.00    47.63     3.63
  6.50    45.56     1.56
  7.00    43.59    -0.41
  7.50    42.27    -1.73
  8.00    41.22    -2.78
  8.50    39.87    -4.13
  9.00    38.26    -5.74
  9.50    36.24    -7.76
 10.00    33.03   -10.97
 10.50    28.46   -15.54
 11.00    23.58   -20.42
 11.50    19.41   -24.59
 12.00    15.19   -28.81
 12.50     9.75   -34.25
 13.00     4.64   -39.36
 13.50     1.75   -42.25
 14.00     0.59   -43.41
</pre>
</td></tr></table><p>

<a name="output.2"></a>
<h3>Output files for usage example 2</h3>
<p><h3>File: ifna2_human.iep</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
IEP of IFNA2_HUMAN from 1 to 188
Isoelectric Point = 5.7322

   pH     Bound    Charge
  1.00    52.98    22.98
  1.50    52.93    22.93
  2.00    52.77    22.77
  2.50    52.28    22.28
  3.00    50.87    20.87
  3.50    47.47    17.47
  4.00    41.62    11.62
  4.50    35.67     5.67
  5.00    32.10     2.10
  5.50    30.47     0.47
  6.00    29.51    -0.49
  6.50    28.55    -1.45
  7.00    27.65    -2.35
  7.50    27.01    -2.99
  8.00    26.36    -3.64
  8.50    25.41    -4.59
  9.00    24.25    -5.75
  9.50    22.81    -7.19
 10.00    20.49    -9.51
 10.50    17.03   -12.97
 11.00    13.16   -16.84
 11.50    10.04   -19.96
 12.00     7.49   -22.51
 12.50     4.72   -25.28
 13.00     2.23   -27.77
 13.50     0.84   -29.16
 14.00     0.28   -29.72
</pre>
</td></tr></table><p>
<p>

For each pH point it gives the number of bound electrons and the charge.


<H2>
    Data files
</H2>

<b>iep</b> reads in local data file <em>Epk.dat</em> which contains amino acid
pK values used to calculate the isoelectric point of the whole protein

<p>

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

<p>
Here is the default <em>Epk.dat</em> file:

<pre>
# pK values for amino acids
# O=Ornithine J=Hydroxyproline
#
# Amino acid	pK
Amino		8.6
Carboxyl	3.6

C		8.5
D		3.9
E		4.1
H		6.5
K		10.8
R		12.5
Y		10.1
</pre>


<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

0 if successful.

<H2>
    Known bugs
</H2>

None.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr>
<td><a href="backtranambig.html">backtranambig</a></td>
<td>Back translate a protein sequence to ambiguous codons</td>
</tr>

<tr>
<td><a href="backtranseq.html">backtranseq</a></td>
<td>Back translate a protein sequence</td>
</tr>

<tr>
<td><a href="charge.html">charge</a></td>
<td>Protein charge plot</td>
</tr>

<tr>
<td><a href="checktrans.html">checktrans</a></td>
<td>Reports STOP codons and ORF statistics of a protein</td>
</tr>

<tr>
<td><a href="compseq.html">compseq</a></td>
<td>Count composition of dimer/trimer/etc words in a sequence</td>
</tr>

<tr>
<td><a href="emowse.html">emowse</a></td>
<td>Protein identification by mass spectrometry</td>
</tr>

<tr>
<td><a href="freak.html">freak</a></td>
<td>Residue/base frequency table or plot</td>
</tr>

<tr>
<td><a href="mwcontam.html">mwcontam</a></td>
<td>Shows molwts that match across a set of files</td>
</tr>

<tr>
<td><a href="mwfilter.html">mwfilter</a></td>
<td>Filter noisy molwts from mass spec output</td>
</tr>

<tr>
<td><a href="octanol.html">octanol</a></td>
<td>Displays protein hydropathy</td>
</tr>

<tr>
<td><a href="pepinfo.html">pepinfo</a></td>
<td>Plots simple amino acid properties in parallel</td>
</tr>

<tr>
<td><a href="pepstats.html">pepstats</a></td>
<td>Protein statistics</td>
</tr>

<tr>
<td><a href="pepwindow.html">pepwindow</a></td>
<td>Displays protein hydropathy</td>
</tr>

<tr>
<td><a href="pepwindowall.html">pepwindowall</a></td>
<td>Displays protein hydropathy of a set of sequences</td>
</tr>

</table>


<H2>
    Author(s)
</H2>
Alan Bleasby (ajb&nbsp;&copy;&nbsp;ebi.ac.uk)
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK


<H2>
    History
</H2>



Completed 1st August 1999.

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything, from naive users to embedded scripts.

<H2>
    Comments
</H2>
None

</BODY>
</HTML>

