<HTML>

<HEAD>
  <TITLE>
  EMBOSS: The Applications (programs)
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="../index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="../emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: The Applications (programs)
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>

The programs are listed in alphabetical order, divided into four
sections.  Look at the individual applications or go to the <a
href="groups.html">GROUPS</a> page.

<ul>
<li><a href="#current">Current release</a>
<li><a href="#embassy">EMBASSY</a>
<li><a href="#developing">Under development</a>
<li><a href="#proposed">Proposed</a>
</ul>
<p>
There are also some applications included only for <a href="check.html">temporary testing</a> which are documented separately.
<p>

<h3><A NAME="current">Applications</A> in the <a
href="ftp://ftp.uk.embnet.org/pub/EMBOSS/">current release</a></h3>

<table border cellpadding=4 bgcolor="#FFFFF0">

<tr><th>Program name</th><th>Author(s)</th><th>
Description
</th></tr>

<tr><td><a href="acdc.html">acdc</a></td><td>Sanger</td><td>
Tests definition files for any EMBOSS application.
</td></tr>

<tr><td><a href="antigenic.html">antigenic</a></td><td>HGMP</td><td>
Finds antigenic sites in proteins
</td></tr>

<tr><td><a href="backtranseq.html">backtranseq</a></td><td>HGMP</td><td>
Back translate a protein sequence
</td></tr>

<tr><td><a href="banana.html">banana</a></td><td>Sanger</td><td>
Bending and Curvature Plot in B-DNA
</td></tr>

<tr><td><a href="chaos.html">chaos</a></td><td>Sanger</td><td>
Create a chaos plot for a sequence.
</td></tr>

<tr><td><a href="checktrans.html">checktrans</a></td><td>EBI</td><td>
ORF property statistics
</td></tr>

<tr><td><a href="chips.html">chips</a></td><td>HGMP</td><td>
Codon usage statistics
</td></tr>

<tr><td><a href="cirdna.html">cirdna</a></td><td>Norway</td><td>
Draws circular maps of DNA constructs
</td></tr>

<tr><td><a href="cluster.html">cluster</a></td><td>Sanger</td><td>
DNA sequence clustering
</td></tr>

<tr><td><a href="codcmp.html">codcmp</a></td><td>HGMP</td><td>
Codon usage table comparison
</td></tr>

<tr><td><a href="complex.html">complex</a></td><td>EMBnet Italy</td><td>
Find the linguistic complexity in nucleotide sequences
</td></tr>

<tr><td><a href="compseq.html">compseq</a></td><td>HGMP</td><td>
Counts the composition of dimer/trimer/etc words in a sequence
</td></tr>

<tr><td><a href="cpgplot.html">cpgplot</a></td><td>HGMP</td><td>
Plot CpG rich areas
</td></tr>

<tr><td><a href="cpgreport.html">cpgreport</a></td><td>HGMP</td><td>
Reports CpG rich regions
</td></tr>

<tr><td><a href="cusp.html">cusp</a></td><td>HGMP</td><td>
Create a codon usage table
</td></tr>

<tr><td><a href="cutseq.html">cutseq</a></td><td>HGMP</td><td>
Removes a specified section from a sequence.
</td></tr>

<tr><td><a href="dan.html">dan</a></td><td>HGMP</td><td>
Plot melting temperatures for DNA.
</td></tr>

<tr><td><a href="dbiblast.html">dbiblast</a></td><td>Sanger</td><td>
Database indexing for BLAST 1 and 2 indexed databases
</td></tr>

<tr><td><a href="dbifasta.html">dbifasta</a></td><td>HGMP</td><td>
Index a fasta database
</td></tr>

<tr><td><a href="dbiflat.html">dbiflat</a></td><td>Sanger</td><td>
Database indexing for flat file databases
</td></tr>

<tr><td><a href="dbigcg.html">dbigcg</a></td><td>Sanger</td><td>
Database indexing for GCG formatted databases
</td></tr>

<tr><td><a href="descseq.html">descseq</a></td><td>HGMP</td><td>
Alter the name or description of a sequence.
</td></tr>

<tr><td><a href="diffseq.html">diffseq</a></td><td>HGMP</td><td>
Find differences (SNPs) between nearly identical sequences
</td></tr>

<tr><td><a href="digest.html">digest</a></td><td>HGMP</td><td>
Protein proteolytic enzyme or reagent cleavage digest
</td></tr>

<tr><td><a href="dotmatcher.html">dotmatcher</a></td><td>Sanger</td><td>
Produces a dotplot of two sequences.
</td></tr>

<tr><td><a href="dotpath.html">dotpath</a></td><td>HGMP</td><td>
Displays a non-overlapping wordmatch dotplot of two sequences
</td></tr>

<tr><td><a href="dottup.html">dottup</a></td><td>Sanger</td><td>
DNA sequence dot plot
</td></tr>

<tr><td><a href="dreg.html">dreg</a></td><td>Sanger</td><td>
Regular expression search of a nucleotide sequence
</td></tr>

<tr><td><a href="einverted.html">einverted</a></td><td>Sanger</td><td>
Finds DNA inverted repeats
</td></tr>

<tr><td><a href="embossdata.html">embossdata</a></td><td>HGMP</td><td>
Finds or fetches the data files read in by the EMBOSS programs
</td></tr>

<tr><td><a href="emma.html">emma</a></td><td>HGMP</td><td>
Multiple alignment program
</td></tr>

<tr><td><a href="equicktandem.html">equicktandem</a></td><td>Sanger</td><td>
Finds tandem repeats
</td></tr>

<tr><td><a href="est2genome.html">est2genome</a></td><td>Sanger</td><td>
Align EST and genomic DNA sequences
</td></tr>

<tr><td><a href="etandem.html">etandem</a></td><td>Sanger</td><td>
Looks for tandem repeats in a nucleotide sequence.
</td></tr>

<tr><td><a href="extractseq.html">extractseq</a></td><td>HGMP</td><td>
Extract regions from a sequence.
</td></tr>

<tr><td><a href="findkm.html">findkm</a></td><td>HGMP</td><td>
Calculates Km and Vmax for an enzyme reaction
</td></tr>

<tr><td><a href="freak.html">freak</a></td><td>HGMP</td><td>
Residue/base frequency table or plot
</td></tr>

<tr><td><a href="fuzznuc.html">fuzznuc</a></td><td>HGMP</td><td>
Nucleic acid pattern search
</td></tr>

<tr><td><a href="fuzzpro.html">fuzzpro</a></td><td>HGMP</td><td>
Protein pattern search
</td></tr>

<tr><td><a href="fuzztran.html">fuzztran</a></td><td>HGMP</td><td>
Protein pattern search after translation
</td></tr>

<tr><td><a href="garnier.html">garnier</a></td><td>EBI</td><td>
Predicts protein secondary structure
</td></tr>

<tr><td><a href="geecee.html">geecee</a></td><td>Sanger</td><td>
Calculates the fractional GC content of nucleic acid sequences
</td></tr>

<tr><td><a href="getorf.html">getorf</a></td><td>HGMP</td><td>
Finds and extracts open reading frames (ORFs)
</td></tr>

<tr><td><a href="helixturnhelix.html">helixturnhelix</a></td><td>HGMP</td><td>
Finds nucleic acid binding domains.
</td></tr>

<tr><td><a href="iep.html">iep</a></td><td>HGMP</td><td>
Calculates the isoelectric point of a protein
</td></tr>

<tr><td><a href="infoseq.html">infoseq</a></td><td>HGMP</td><td>
Displays some simple information about sequences
</td></tr>

<tr><td><a href="isochore.html">isochore</a></td><td>Sanger</td><td>
Plots isochores in large DNA sequences
</td></tr>

<tr><td><a href="lindna.html">lindna</a></td><td>Norway</td><td>
Draws linear maps of DNA constructs
</td></tr>

<tr><td><a href="maskseq.html">maskseq</a></td><td>HGMP</td><td>
Mask off regions of a sequence.
</td></tr>

<tr><td><a href="matcher.html">matcher</a></td><td>Sanger</td><td>
Local alignment of two sequences
</td></tr>

<tr><td><a href="megamerger.html">megamerger</a></td><td>HGMP</td><td>
Merge two large overlapping nucleic acid sequences
</td></tr>

<tr><td><a href="merger.html">merger</a></td><td>HGMP</td><td>
Merge two overlapping sequences
</td></tr>

<tr><td><a href="msbar.html">msbar</a></td><td>HGMP</td><td>
Mutate sequence beyond all recognition
</td></tr>

<tr><td><a href="needle.html">needle</a></td><td>HGMP</td><td>
Needleman-Wunsch global alignment.
</td></tr>

<tr><td><a href="newcpgreport.html">newcpgreport</a></td><td>EBI</td><td>
Report CpG rich areas
</td></tr>

<tr><td><a href="newcpgseek.html">newcpgseek</a></td><td>EBI</td><td>
Reports CpG rich regions
</td></tr>

<tr><td><a href="newseq.html">newseq</a></td><td>HGMP</td><td>
Type in a short new sequence.
</td></tr>

<tr><td><a href="noreturn.html">noreturn</a></td><td>HGMP</td><td>
Removes carriage return from ASCII files
</td></tr>

<tr><td><a href="nthseq.html">nthseq</a></td><td>HGMP</td><td>
Writes one sequence from a multiple set of sequences
</td></tr>

<tr><td><a href="octanol.html">octanol</a></td><td>Sanger</td><td>
Displays protein hydropathy
</td></tr>

<tr><td><a href="oddcomp.html">oddcomp</a></td><td>Norway</td><td>
Finds protein sequence regions with a biased composition.
</td></tr>

<tr><td><a href="palindrome.html">palindrome</a></td><td>HGMP</td><td>
Looks for inverted repeats in a nucleotide sequence.
</td></tr>

<tr><td><a href="pasteseq.html">pasteseq</a></td><td>HGMP</td><td>
Insert one sequence into another.
</td></tr>

<tr><td><a href="patmatdb.html">patmatdb</a></td><td>HGMP</td><td>
Matching a Prosite motif against a Protein Sequence Database.
</td></tr>

<tr><td><a href="patmatmotifs.html">patmatmotifs</a></td><td>HGMP</td><td>
Compares a protein sequence to the PROSITE motif database.
</td></tr>

<tr><td><a href="pepcoil.html">pepcoil</a></td><td>HGMP</td><td>
Predicts coiled coil regions
</td></tr>

<tr><td><a href="pepinfo.html">pepinfo</a></td><td>HGMP</td><td>
Plots simple amino acid properties in parallel
</td></tr>

<tr><td><a href="pepnet.html">pepnet</a></td><td>HGMP</td><td>
Protein helical net plot
</td></tr>

<tr><td><a href="pepstats.html">pepstats</a></td><td>HGMP</td><td>
Protein statistics
</td></tr>

<tr><td><a href="pepwheel.html">pepwheel</a></td><td>HGMP</td><td>
Shows protein sequences as helices
</td></tr>

<tr><td><a href="pepwindow.html">pepwindow</a></td><td>Sanger</td><td>
Displays protein hydropathy
</td></tr>

<tr><td><a href="pepwindowall.html">pepwindowall</a></td><td>Sanger</td><td>
Displays protein hydropathy of a set of sequences
</td></tr>

<tr><td><a href="plotorf.html">plotorf</a></td><td>HGMP</td><td>
Plot potential open reading frames
</td></tr>

<tr><td><a href="polydot.html">polydot</a></td><td>Sanger</td><td>
Multiple dotplot
</td></tr>

<tr><td><a href="preg.html">preg</a></td><td>Sanger</td><td>
Regular expression search of a protein sequence
</td></tr>

<tr><td><a href="prettyplot.html">prettyplot</a></td><td>Sanger</td><td>
Displays aligned sequences, with colouring and boxing.
</td></tr>

<tr><td><a href="prettyseq.html">prettyseq</a></td><td>HGMP</td><td>
Output sequence with translated ranges
</td></tr>

<tr><td><a href="prima.html">prima</a></td><td>HGMP</td><td>
Selecting primers for PCR and DNA amplification.
</td></tr>

<tr><td><a href="primersearch.html">primersearch</a></td><td>HGMP</td><td>
Searches DNA sequences for matches with primer pairs
</td></tr>

<tr><td><a href="printsextract.html">printsextract</a></td><td>HGMP</td><td>
Preprocesses the PRINTS database for use with the program PSCAN
</td></tr>

<tr><td><a href="profit.html">profit</a></td><td>HGMP</td><td>
Scan a sequence or database with a matrix or profile
</td></tr>

<tr><td><a href="prophecy.html">prophecy</a></td><td>HGMP</td><td>
Creates matrices/profiles from multiple alignments
</td></tr>

<tr><td><a href="prophet.html">prophet</a></td><td>HGMP</td><td>
Gapped alignment for profiles
</td></tr>

<tr><td><a href="prosextract.html">prosextract</a></td><td>HGMP</td><td>
Extracts ID, AC, and PA lines from the PROSITE motif database.
</td></tr>

<tr><td><a href="pscan.html">pscan</a></td><td>HGMP</td><td>
Locates fingerprints (multiple motif features) in a protein sequence.
</td></tr>

<tr><td><a href="rebaseextract.html">rebaseextract</a></td><td>HGMP</td><td>
Extract data from REBASE
</td></tr>

<tr><td><a href="redata.html">redata</a></td><td>HGMP</td><td>
Isoschizomers, references and Suppliers for Restriction Enzymes
</td></tr>

<tr><td><a href="remap.html">remap</a></td><td>HGMP</td><td>
Display a sequence with restriction cut sites, translation etc..
</td></tr>

<tr><td><a href="restrict.html">restrict</a></td><td>HGMP</td><td>
Finds Restriction Enzyme Cleavage Sites
</td></tr>

<tr><td><a href="revseq.html">revseq</a></td><td>HGMP</td><td>
Reverse and complement a sequence.
</td></tr>

<tr><td><a href="seealso.html">seealso</a></td><td>HGMP</td><td>
Finds programs sharing group names
</td></tr>

<tr><td><a href="seqmatchall.html">seqmatchall</a></td><td>Sanger</td><td>
Sequence comparison
</td></tr>

<tr><td><a href="seqret.html">seqret</a></td><td>Sanger</td><td>
Reads and writes (returns) a sequence.
</td></tr>

<tr><td><a href="seqretall.html">seqretall</a></td><td>Sanger</td><td>
Reads and writes (returns) a set of sequences one at a time.
</td></tr>

<tr><td><a href="seqretset.html">seqretset</a></td><td>Sanger</td><td>
Reads and writes (returns) a set of sequences all at once.
</td></tr>

<tr><td><a href="showdb.html">showdb</a></td><td>HGMP</td><td>
Displays information on the currently available databases
</td></tr>

<tr><td><a href="showfeat.html">showfeat</a></td><td>HGMP</td><td>
Show features of a sequence.
</td></tr>

<tr><td><a href="showorf.html">showorf</a></td><td>HGMP</td><td>
Pretty output of DNA translations
</td></tr>

<tr><td><a href="showseq.html">showseq</a></td><td>HGMP</td><td>
Display a sequence with features, translation etc
</td></tr>

<tr><td><a href="shuffleseq.html">shuffleseq</a></td><td>HGMP</td><td>
Shuffles a set of sequences maintaining composition
</td></tr>

<tr><td><a href="sigcleave.html">sigcleave</a></td><td>HGMP</td><td>
Predicts signal peptide cleavage sites.
</td></tr>

<tr><td><a href="splitter.html">splitter</a></td><td>HGMP</td><td>
Split a sequence into (overlapping) smaller sequences.
</td></tr>

<tr><td><a href="stretcher.html">stretcher</a></td><td>Sanger</td><td>
Global alignment of two sequences.
</td></tr>

<tr><td><a href="stssearch.html">stssearch</a></td><td>Sanger</td><td>
Searches a DNA database for matches with a set of STS primers
</td></tr>

<tr><td><a href="supermatcher.html">supermatcher</a></td><td>Sanger</td><td>
Finds a match of a large sequence against one or more sequences
</td></tr>

<tr><td><a href="syco.html">syco</a></td><td>HGMP</td><td>
Synonymous codon usage Gribskov statistic plot
</td></tr>

<tr><td><a href="textsearch.html">textsearch</a></td><td>HGMP</td><td>
Search sequence documentation text. SRS and Entrez are faster!
</td></tr>

<tr><td><a href="tfextract.html">tfextract</a></td><td>HGMP</td><td>
Extract data from TRANSFAC
</td></tr>

<tr><td><a href="tfm.html">tfm</a></td><td>HGMP</td><td>
Displays a program's help documentation manual
</td></tr>

<tr><td><a href="tfscan.html">tfscan</a></td><td>HGMP</td><td>
Scans DNA sequences for transcription factors.
</td></tr>

<tr><td><a href="tmap.html">tmap</a></td><td>Sanger</td><td>
Predict transmembrane proteins
</td></tr>

<tr><td><a href="transeq.html">transeq</a></td><td>HGMP</td><td>
Translates nucleic acid sequences.
</td></tr>

<tr><td><a href="trimseq.html">trimseq</a></td><td>HGMP</td><td>
Trim ambiguous bits off the ends of sequences
</td></tr>

<tr><td><a href="vectorstrip.html">vectorstrip</a></td><td>HGMP</td><td>
Strips out DNA between a pair of vector sequences
</td></tr>

<tr><td><a href="water.html">water</a></td><td>HGMP</td><td>
Smith-Waterman local alignment.
</td></tr>

<tr><td><a href="wobble.html">wobble</a></td><td>HGMP</td><td>
Wobble base plot
</td></tr>

<tr><td><a href="wordcount.html">wordcount</a></td><td>Sanger</td><td>
Counts words of a specified size in a DNA sequence.
</td></tr>

<tr><td><a href="wordmatch.html">wordmatch</a></td><td>Sanger</td><td>
Finds all exact matches of a given size between 2 sequences
</td></tr>

<tr><td><a href="wossname.html">wossname</a></td><td>HGMP</td><td>
Finds programs by keywords in their one-line documentation.
</td></tr>

</table>

<h3><a name=embassy>EMBASSY</a></h3>
<table border cellpadding=4 bgcolor="#FFFFF0">

<tr><th>Program name</th><th>Author(s)</th><th>
Description
</th></tr>

<tr><td><a href="../EMBASSY/PHYLIP/econsense.html">econsense</a></td><td>Sanger</td><td>
Conversion of PHYLIP's consense
</td></tr>

<tr><td><a href="../EMBASSY/PHYLIP/ednadist.html">ednadist</a></td><td>Sanger</td><td>
Conversion of PHYLIP's dnadist
</td></tr>

<tr><td><a href="../EMBASSY/PHYLIP/ednaml.html">ednaml</a></td><td>Sanger</td><td>
Conversion of PHYLIP's dnaml
</td></tr>

<tr><td><a href="../EMBASSY/PHYLIP/ednamlk.html">ednamlk</a></td><td>Sanger</td><td>
Conversion of PHYLIP's dnamlk
</td></tr>

<tr><td><a href="../EMBASSY/PHYLIP/ednapars.html">ednapars</a></td><td>Sanger</td><td>
Conversion of PHYLIP's dnapars
</td></tr>

<tr><td><a href="../EMBASSY/PHYLIP/efitch.html">efitch</a></td><td>Sanger</td><td>
Conversion of PHYLIP's fitch
</td></tr>

<tr><td><a href="../EMBASSY/PHYLIP/ekitsch.html">ekitsch</a></td><td>Sanger</td><td>
Conversion of PHYLIP's kitsch
</td></tr>

<tr><td><a href="../EMBASSY/PHYLIP/eneighbor.html">eneighbor</a></td><td>Sanger</td><td>
Conversion of PHYLIP's neighbor
</td></tr>

<tr><td><a href="../EMBASSY/PHYLIP/eprotdist.html">eprotdist</a></td><td>Sanger</td><td>
Conversion of PHYLIP's protdist
</td></tr>

<tr><td><a href="../EMBASSY/PHYLIP/eprotpars.html">eprotpars</a></td><td>Sanger</td><td>
Conversion of PHYLIP's protpars
</td></tr>

<tr><td><a href="../EMBASSY/PHYLIP/eseqboot.html">eseqboot</a></td><td>Sanger</td><td>
Conversion of PHYLIP's seqboot
</td></tr>

<tr><td><a href="../EMBASSY/MSE/mse.html">mse</a></td><td>Sanger</td><td>
Conversion of Will Gilbert's MSE editor
</td></tr>

<tr><td><a href="../EMBASSY/TOPO/topo.html">topo</a></td><td>Sanger</td><td>
Conversion of Susan Jean Johns' TOPO
</td></tr>

</table>

<h3><a name=developing>Under development</a></h3>
<table border cellpadding=4 bgcolor="#FFFFF0">

<tr><th>Program name</th><th>Author(s)</th><th>
Description
</th></tr>

</table>

<h3><a name=proposed>Proposed</a></h3>
<table border cellpadding=4 bgcolor="#FFFFF0">

<tr><th>Program name</th><th>Proposer(s)</th><th>
Description
</th></tr>

<tr><td><a href="alignutils.html">alignutils</a></td><td>HGMP</td><td>
Sequence alignment utilities, to replace EGCG sortconsensus.
</td></tr>

<tr><td><a href="dodayhoffstat.html">dodayhoffstat</a></td><td>Sanger</td><td>
Replacement for EGCG's dodayhoffstat
</td></tr>

<tr><td><a href="dottie.html">dottie</a></td><td>Sanger</td><td>
A general interactive dot plot application.
Possibly using Erik Sonnhammer's dotter.
</td></tr>

<tr><td><a href="fasta.html">fasta</a></td><td>Sanger</td><td>
Integration of Bill Pearson's FASTA, TFASTA, etc.
</td></tr>

<tr><td><a href="fastacheck.html">fastacheck</a></td><td>Sanger</td><td>
Replacement for EGCG's fastacheck
</td></tr>

<tr><td><a href="gapframe.html">gapframe</a></td><td>HGMP</td><td>
Adjust gap positions to be only at codon boundaries
in a DNA sequence with known CDS position(s).
</td></tr>

<tr><td><a href="genetrans.html">genetrans</a></td><td>Sanger</td><td>
Replacement for EGCG's genetrans
</td></tr>

<tr><td><a href="hitmatch.html">hitmatch</a></td><td>Sanger</td><td>
Replacement for EGCG's equickmatch, using blastn output
</td></tr>

<tr><td><a href="homologies.html">homologies</a></td><td>CAOS/CAMM</td><td>
Table of the pairwise distances of aligned sequences.
Assigned to the original authors for comment.
</td></tr>

<tr><td><a href="rnafolding.html">rnafolding</a></td><td>???</td><td>
Integration of external RNA folding applications.
</td></tr>

<tr><td><a href="HMMs.html">HMMs</a></td><td>???</td><td>
Inclusion of Sean Eddys(?) HMM library routines and applications.
</td></tr>

<tr><td><a href="neural-nets.html">neural-nets</a></td><td>???</td><td>
Neural net routines and applications.
</td></tr>

<tr><td><a href="GAs.html">GAs</a></td><td>???</td><td>
Genetic Algorithm routines and applications.
</td></tr>

</table>




</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>







