<HTML>
<HEAD>
  <TITLE>
  EMBOSS: infoassembly
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="/images/emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
infoassembly
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
Display information about assemblies
<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<b>infoassembly</b> writes statistics for a sequence assembly.

<p>
The initial release is a basic version. More output can be
added. Please contact the authors with suggestions.





<H2>
    Algorithm
</H2>

None.


<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

Here is a sample session with <b>infoassembly</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>infoassembly sam::samspec1.4example.sam stdout -auto </b>


Read distribution:
------------------

READ_CATEGORY	#READS	Av_Read_Length
all reads	6	10.33
first of pair 	1	9.00
second of pair	1	17.00
pair          	2	13.00
unpaired      	4	9.00

Contig stats:
-------------

CONTIG	LENGTH	Mean_QUAL	#READS	Max DEPTH	Mean_DEPTH	%GC
ref	45	0.00	6	0	0.00	0.00

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><p>
<p>
<b>Example 2</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>infoassembly sam::xxx.sam stdout -auto -qual qualsdist.txt </b>

Phred encoding: Sanger / Illumina 1.9

Read distribution:
------------------

READ_CATEGORY	#READS	Av_Read_Length
all reads	2	35.00
first of pair 	1	35.00
second of pair	1	35.00
pair          	2	35.00
unpaired      	0

Contig stats:
-------------

CONTIG	LENGTH	Mean_QUAL	#READS	Max DEPTH	Mean_DEPTH	%GC
chr20	62435964	25.36	2	2	1.97	0.00

</pre></td></tr></table><p>
<p>
<a href="#input.2">Go to the input files for this example</a><br><a href="#output.2">Go to the output files for this example</a><p><p>
<p>
<b>Example 3</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>infoassembly sam::../data/samspec1.4example.ref.fasta </b>


Read distribution:
------------------

READ_CATEGORY	#READS	Av_Read_Length
all reads	6	10.33
first of pair 	1	9.00
second of pair	1	17.00
pair          	2	13.00
unpaired      	4	9.00

Contig stats:
-------------

CONTIG	LENGTH	Mean_QUAL	#READS	Max DEPTH	Mean_DEPTH	%GC
ref	45	0.00	6	3	1.82	46.67

</pre></td></tr></table><p>
<p>
<a href="#input.3">Go to the input files for this example</a><br><p>
<p>
<b>Example 4</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>infoassembly sam::../data/samspec1.4example.ref.fasta -gc gcbias.txt </b>


Read distribution:
------------------

READ_CATEGORY	#READS	Av_Read_Length
all reads	6	10.33
first of pair 	1	9.00
second of pair	1	17.00
pair          	2	13.00
unpaired      	4	9.00

Contig stats:
-------------

CONTIG	LENGTH	Mean_QUAL	#READS	Max DEPTH	Mean_DEPTH	%GC
ref	45	0.00	6	3	1.82	46.67

</pre></td></tr></table><p>
<p>
<a href="#output.4">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
Display information about assemblies
Version: EMBOSS:6.6.0.0

   Standard (Mandatory) qualifiers:
  [-assembly]          assembly   (no help text) assembly value
  [-outassembly]       outassembly (no help text) outassembly value
  [-gcbiasmetricsoutfile] outfile    [*.infoassembly] GC bias metrics
  [-qualvaluesdistoutfile] outfile    [*.infoassembly] Distribution of quality
                                  values for the reads

   Additional (Optional) qualifiers:
   -refsequence        seqset     Reference sequences in the assembly
   -windowsize         integer    [100] The size of windows on the genome that
                                  are used to bin reads. (Any integer value)
   -bisulfite          boolean    [N] If this is true, it is assumed that the
                                  reads were bisulfite treated

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-assembly" associated qualifiers
   -cbegin1            integer    Start of the contig/consensus sequences
   -cend1              integer    End of the contig/consensus sequences
   -iformat1           string     Input assembly format
   -iquery1            string     Input query fields or ID list
   -ioffset1           integer    Input start position offset
   -idbname1           string     User-provided database name

   "-refsequence" associated qualifiers
   -sbegin             integer    Start of each sequence to be used
   -send               integer    End of each sequence to be used
   -sreverse           boolean    Reverse (if DNA)
   -sask               boolean    Ask for begin/end/reverse
   -snucleotide        boolean    Sequence is nucleotide
   -sprotein           boolean    Sequence is protein
   -slower             boolean    Make lower case
   -supper             boolean    Make upper case
   -scircular          boolean    Sequence is circular
   -squick             boolean    Read id and sequence only
   -sformat            string     Input sequence format
   -iquery             string     Input query fields or ID list
   -ioffset            integer    Input start position offset
   -sdbname            string     Database name
   -sid                string     Entryname
   -ufo                string     UFO features
   -fformat            string     Features format
   -fopenfile          string     Features file name

   "-outassembly" associated qualifiers
   -odirectory2        string     Output directory
   -oformat2           string     Assembly output format

   "-gcbiasmetricsoutfile" associated qualifiers
   -odirectory3        string     Output directory

   "-qualvaluesdistoutfile" associated qualifiers
   -odirectory4        string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-assembly]<br>(Parameter 1)</td>
<td>assembly</td>
<td>(no help text) assembly value</td>
<td>Assembly of sequence reads</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-outassembly]<br>(Parameter 2)</td>
<td>outassembly</td>
<td>(no help text) outassembly value</td>
<td>Assembly of sequence reads</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-gcbiasmetricsoutfile]<br>(Parameter 3)</td>
<td>outfile</td>
<td>GC bias metrics</td>
<td>Output file</td>
<td><i>&lt;*&gt;</i>.infoassembly</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-qualvaluesdistoutfile]<br>(Parameter 4)</td>
<td>outfile</td>
<td>Distribution of quality values for the reads</td>
<td>Output file</td>
<td><i>&lt;*&gt;</i>.infoassembly</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>-refsequence</td>
<td>seqset</td>
<td>Reference sequences in the assembly</td>
<td>Readable set of sequences</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-windowsize</td>
<td>integer</td>
<td>The size of windows on the genome that are used to bin reads.</td>
<td>Any integer value</td>
<td>100</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-bisulfite</td>
<td>boolean</td>
<td>If this is true, it is assumed that the reads were bisulfite treated</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td align="left" colspan=5>"-assembly" associated assembly qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -cbegin1<br>-cbegin_assembly</td>
<td>integer</td>
<td>Start of the contig/consensus sequences</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -cend1<br>-cend_assembly</td>
<td>integer</td>
<td>End of the contig/consensus sequences</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -iformat1<br>-iformat_assembly</td>
<td>string</td>
<td>Input assembly format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -iquery1<br>-iquery_assembly</td>
<td>string</td>
<td>Input query fields or ID list</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ioffset1<br>-ioffset_assembly</td>
<td>integer</td>
<td>Input start position offset</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -idbname1<br>-idbname_assembly</td>
<td>string</td>
<td>User-provided database name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td align="left" colspan=5>"-refsequence" associated seqset qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sbegin</td>
<td>integer</td>
<td>Start of each sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -send</td>
<td>integer</td>
<td>End of each sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sreverse</td>
<td>boolean</td>
<td>Reverse (if DNA)</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sask</td>
<td>boolean</td>
<td>Ask for begin/end/reverse</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -snucleotide</td>
<td>boolean</td>
<td>Sequence is nucleotide</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sprotein</td>
<td>boolean</td>
<td>Sequence is protein</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -slower</td>
<td>boolean</td>
<td>Make lower case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -supper</td>
<td>boolean</td>
<td>Make upper case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -scircular</td>
<td>boolean</td>
<td>Sequence is circular</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -squick</td>
<td>boolean</td>
<td>Read id and sequence only</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sformat</td>
<td>string</td>
<td>Input sequence format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -iquery</td>
<td>string</td>
<td>Input query fields or ID list</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ioffset</td>
<td>integer</td>
<td>Input start position offset</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sdbname</td>
<td>string</td>
<td>Database name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sid</td>
<td>string</td>
<td>Entryname</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ufo</td>
<td>string</td>
<td>UFO features</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fformat</td>
<td>string</td>
<td>Features format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fopenfile</td>
<td>string</td>
<td>Features file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td align="left" colspan=5>"-outassembly" associated outassembly qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -odirectory2<br>-odirectory_outassembly</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -oformat2<br>-oformat_outassembly</td>
<td>string</td>
<td>Assembly output format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td align="left" colspan=5>"-gcbiasmetricsoutfile" associated outfile qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -odirectory3<br>-odirectory_gcbiasmetricsoutfile</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td align="left" colspan=5>"-qualvaluesdistoutfile" associated outfile qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -odirectory4<br>-odirectory_qualvaluesdistoutfile</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<p>

The input is a standard EMBOSS assembly query.

<p> The major assembly sources are files in SAM and BAM format.


<b>infoassembly</b> reads a sequnece assembly.


<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: samspec1.4example.sam</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
@CO	example from SAM format specification v1.4
@HD	VN:1.3	SO:coordinate
@SQ	SN:ref	LN:45
r001	163	ref	7	30	8M2I4M1D3M	=	37	39	TTAGATAAAGGATACTG	*
r002	0	ref	9	30	3S6M1P1I4M	*	0	0	AAAAGATAAGGATA	*
r003	0	ref	9	30	5H6M	*	0	0	AGCTAA	*	NM:i:1
r004	0	ref	16	30	6M14N5M	*	0	0	ATAGCTTCAGC	*
r003	16	ref	29	30	6H5M	*	0	0	TAGGC	*	NM:i:0
r001	83	ref	37	30	9M	=	7	-39	CAGCGCCAT	*
</pre>
</td></tr></table><p>

<a name="input.2"></a>
<h3>Input files for usage example 2</h3>
<p><h3>File: xxx.sam</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
@HD	VN:1.0
@SQ	SN:chr20	LN:62435964
@RG	ID:L1	PU:SC_1_10	LB:SC_1	SM:NA12891
@RG	ID:L2	PU:SC_2_12	LB:SC_2	SM:NA12891
read_28833_29006_6945	99	chr20	28833	20	10M1D25M	=	28993	195	AGCTTAGCTAGCTACCTATATCTTGGTCTTGGCCG	&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;:&lt;9/,&amp;,22;;&lt;&lt;&lt;	NM:i:1	RG:Z:L1
read_28701_28881_323b	147	chr20	28834	30	35M	=	28701	-168	ACCTATATCTTGGCCTTGGCCGATGCGGCCTTGCA	&lt;&lt;&lt;&lt;&lt;;&lt;&lt;&lt;&lt;7;:&lt;&lt;&lt;6;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;7&lt;&lt;&lt;&lt;	MF:i:18	RG:Z:L2
</pre>
</td></tr></table><p>

<a name="input.3"></a>
<h3>Input files for usage example 3</h3>
<p><h3>File: samspec1.4example.ref.fasta</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
&gt;ref
AGCATGTTAGATAAGATAGCTGTGCTAGTAGGCAGTCAGCGCCAT
</pre>
</td></tr></table><p>





<H2>
    Output file format
</H2>


<b>infoassembly</b> writes a simple text report.


<p>


<a name="output.2"></a>
<h3>Output files for usage example 2</h3>
<p><h3>File: qualsdist.txt</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
#QUALITY: Phred quality score
#BASES: Number of bases with the quality score

QUALITY	#BASES
5	1
11	2
14	1
17	2
21	1
22	2
24	1
25	2
26	5
27	53
</pre>
</td></tr></table><p>

<a name="output.4"></a>
<h3>Output files for usage example 4</h3>
<p><h3>File: gcbias.txt</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
# GC bias metrics as described in GcBiasDetailMetrics class
# in picard  project (picard.sf.net).
# The mean base quality are determined via the error rate
# of all bases of all reads that are assigned to windows of a GC.
GC	WINDOWS	READ_STARTS	MEAN_BASE_QUALITY	NORMALIZED_COVERAGE
20	5	1	8	1.13333
30	5	2	10	2.26667
40	4	0	0	0
50	11	1	0	0.515152
60	4	1	0	1.41667
70	5	0	0	0
</pre>
</td></tr></table><p>





<H2>
    Data files
</H2>

None.



<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
<ol>

<li>

</ol>

   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="assemblyget.html">assemblyget</a></td>
<td>Get assembly of sequence reads</td>
</tr>

</table>
<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>

Mahmut Uludag
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK

<p>
Please report all bugs to the EMBOSS bug team (emboss-bug&nbsp;&copy;&nbsp;emboss.open-bio.org) not to the original author.








<H2>
    History
</H2>
<!--
        Date written and what changes have been made go in this file.
   -->




<H2>
    Target users
</H2>
<!--
        For general users, use this text
   -->
This program is intended to be used by everyone and everything, from naive users to embedded scripts.

<H2>
    Comments
</H2>
<!--
        User/developer/other comments go in this file.
   -->
None


</BODY>
</HTML>

