<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: infoalign
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">
<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>
<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>
</td>
<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: infoalign
</H2>
</font>
</td>
</tr>
</table>
</td>
</tr>
</table>
<p><hr><p>

<!--END OF HEADER-->

<H1>
Program infoalign
</H1>
<H2>
    Function
</H2>
Information on a multiple sequence alignment

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
        Algorithms
  -->

<b>infoalign</b> is small utility to list some simple properties of
sequences in an alignment. 

<p>

It will write a table containing one line per sequence.  The information
is written out in columns separated by space or TAB characters.  The
columns of data are: the sequences' USA, name, two measures of length,
counts of gaps, and numbers of identical, similar and different residues
or bases in this sequence when compared to a reference sequence,
together with a simple statistic of the % change between the reference
sequence and this sequence. 

<p>

The reference sequence can be either the calculated consensus sequence (the default) or it can be one of
the set of aligned sequences, specified by either the ordinal number of that sequence in the input file, or
by its name. 

<p>

Any combination of these types of information can be easily selected or
unselected. 

<p>

By default, the output file starts each line with the USA of the
sequence being described, so the output file is a list file that can be
manually edited and read in by any other EMBOSS program that can read in
one or more sequence to be analysed. 


<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->


Here is a sample session with infoalign:
<br>
<pre>
% infoalign globin.seq
</pre>

<p>

Don't display the USA of a sequence 
<br>
<pre>
% infoalign globin.seq -nousa
</pre>

<p>

Display only the name and sequence length of a sequence 
<br>
<pre>
% infoalign globin.seq -only -name -seqlength
</pre>

<p>

Display only the name, number of gap characters and differences to the consensus sequence
<br>
<pre>
% infoalign globin.seq -only -name -gapcount -diffcount
</pre>

<p>

Display the name and  number of gaps within a sequence
<br>
<pre>
% infoalign globin.seq -only -name -gaps
</pre>

<p>

Display information formatted with HTML 
<br>
<pre>
% infoalign globin.seq -html
</pre>


<p>

Use the first sequence as the reference sequence to compare to
<br>
<pre>
% infoalign globin.seq -refseq 1
</pre>



<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqset     The sequence alignment to be displayed.

   Optional qualifiers:
   -outfile            outfile    If you enter the name of a file here then
                                  this program will write the sequence details
                                  into that file.
   -refseq             string     If you give the number in the alignment or
                                  the name of a sequence, it will be taken to
                                  be the reference sequence. The reference
                                  sequence is the one against which all the
                                  other sequences are compared. If this is set
                                  to 0 then the consensus sequence will be
                                  used as the reference sequence. By default
                                  the consensus sequence is used as the
                                  reference sequence.
   -matrix             matrix     Similarity scoring Matrix file
   -html               bool       Format output as an HTML table

   Advanced qualifiers:
   -plurality          float      Set a cut-off for the % of positive scoring
                                  matches below which there is no consensus.
                                  The default plurality is taken as 50% of the
                                  total weight of all the sequences in the
                                  alignment.
   -identity           float      Provides the facility of setting the
                                  required number of identities at a position
                                  for it to give a consensus. Therefore, if
                                  this is set to 100% only columns of
                                  identities contribute to the consensus.
   -only               bool       This is a way of shortening the command line
                                  if you only want a few things to be
                                  displayed. Instead of specifying:
                                  '-nohead -nousa -noname -noalign -nogaps
                                  -nogapcount -nosimcount -noidcount
                                  -nodiffcount'
                                  to get only the sequence length output, you
                                  can specify
                                  '-only -seqlength'
   -heading            bool       Display column headings
   -usa                bool       Display the USA of the sequence
   -name               bool       Display 'name' column
   -seqlength          bool       Display 'seqlength' column
   -alignlength        bool       Display 'alignlength' column
   -gaps               bool       Display number of gaps
   -gapcount           bool       Display number of gap positions
   -idcount            bool       Display number of identical positions
   -simcount           bool       Display number of similar positions
   -diffcount          bool       Display number of different positions
   -change             bool       Display % number of changed positions
   -description        bool       Display 'description' column

   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>The sequence alignment to be displayed.</td>
<td>Readable sequences</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-outfile</td>
<td>If you enter the name of a file here then this program will write the sequence details into that file.</td>
<td>Output file</td>
<td>stdout</td>
</tr>

<tr>
<td>-refseq</td>
<td>If you give the number in the alignment or the name of a sequence, it will be taken to be the reference sequence. The reference sequence is the one against which all the other sequences are compared. If this is set to 0 then the consensus sequence will be used as the reference sequence. By default the consensus sequence is used as the reference sequence.</td>
<td>Any string is accepted</td>
<td>0</td>
</tr>

<tr>
<td>-matrix</td>
<td>Similarity scoring Matrix file</td>
<td>Comparison matrix file in EMBOSS data path</td>
<td>EBLOSUM62 for protein<br>EDNAFULL for DNA</td>
</tr>

<tr>
<td>-html</td>
<td>Format output as an HTML table</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-plurality</td>
<td>Set a cut-off for the % of positive scoring matches below which there is no consensus. The default plurality is taken as 50% of the total weight of all the sequences in the alignment.</td>
<td>Number from 0.000 to 100.000</td>
<td>50.0</td>
</tr>

<tr>
<td>-identity</td>
<td>Provides the facility of setting the required number of identities at a position for it to give a consensus. Therefore, if this is set to 100% only columns of identities contribute to the consensus.</td>
<td>Number from 0.000 to 100.000</td>
<td>0.0</td>
</tr>

<tr>
<td>-only</td>
<td>This is a way of shortening the command line if you only want a few things to be displayed. Instead of specifying:
'-nohead -nousa -noname -noalign -nogaps -nogapcount -nosimcount -noidcount -nodiffcount'
to get only the sequence length output, you can specify
'-only -seqlength'</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-heading</td>
<td>Display column headings</td>
<td>Yes/No</td>
<td>@(!$(only))</td>
</tr>

<tr>
<td>-usa</td>
<td>Display the USA of the sequence</td>
<td>Yes/No</td>
<td>@(!$(only))</td>
</tr>

<tr>
<td>-name</td>
<td>Display 'name' column</td>
<td>Yes/No</td>
<td>@(!$(only))</td>
</tr>

<tr>
<td>-seqlength</td>
<td>Display 'seqlength' column</td>
<td>Yes/No</td>
<td>@(!$(only))</td>
</tr>

<tr>
<td>-alignlength</td>
<td>Display 'alignlength' column</td>
<td>Yes/No</td>
<td>@(!$(only))</td>
</tr>

<tr>
<td>-gaps</td>
<td>Display number of gaps</td>
<td>Yes/No</td>
<td>@(!$(only))</td>
</tr>

<tr>
<td>-gapcount</td>
<td>Display number of gap positions</td>
<td>Yes/No</td>
<td>@(!$(only))</td>
</tr>

<tr>
<td>-idcount</td>
<td>Display number of identical positions</td>
<td>Yes/No</td>
<td>@(!$(only))</td>
</tr>

<tr>
<td>-simcount</td>
<td>Display number of similar positions</td>
<td>Yes/No</td>
<td>@(!$(only))</td>
</tr>

<tr>
<td>-diffcount</td>
<td>Display number of different positions</td>
<td>Yes/No</td>
<td>@(!$(only))</td>
</tr>

<tr>
<td>-change</td>
<td>Display % number of changed positions</td>
<td>Yes/No</td>
<td>@(!$(only))</td>
</tr>

<tr>
<td>-description</td>
<td>Display 'description' column</td>
<td>Yes/No</td>
<td>@(!$(only))</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

A normal multiple sequence alignment file, as produced by an alignment
program is read in. 




<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->

The output is displayed on the screen (stdout) by default. 

A typical output file is: 

<hr>
<pre>
# USA             Name        SeqLen    AlignLen        Gaps    GapLen  Ident  Similar  Differ  % Change        Description
fasta::globin.seq:HBB_MACFU     HBB_MACFU     147       148     1       1      134      3       10      8.843537
fasta::globin.seq:HBD_HUMAN     HBD_HUMAN     147       148     1       1      128      7       12      12.925170
fasta::globin.seq:HBB_RABIT     HBB_RABIT     147       148     1       1      131      7       9       10.884354
fasta::globin.seq:HBB_BOVIN     HBB_BOVIN     146       147     1       1      121      10      15      17.123287
fasta::globin.seq:HBB_LAMGL     HBB_LAMGL     151       152     1       1      122      11      18      19.205297
fasta::globin.seq:HBE_HUMAN     HBE_HUMAN     143       150     2       7      104      14      25      27.272728
fasta::globin.seq:HBE_RABIT     HBE_RABIT     147       148     1       1      113      16      18      23.129251
</pre>
<hr>


The first non-blank line is the heading.  This is followed by one line
per sequence containing the following columns of data separated by one
of more space or TAB characters:

<p>

<UL>
<LI> The USA (Uniform Sequence Address) that EMBOSS can use to read in the sequence.
<LI> Name - name of the sequence.
<LI> SeqLen - length of the sequence when all gap characters are removed.
<LI> AlignLen - length of the sequence including internal gap characters i.e. gaps at the start or the end are not included.
<LI> Gaps - number of gaps e.g. 'AAA---AAA' is 1 gap (and 3 gap characters long (see GapLen)).
<LI> GapLen - total number of internal gap characters, see the 3 gap characters above. This is the sum total of all of the internal gap characters in this sequence.
<LI> Ident - number of characters that are identical to the specified reference sequence (uppercase 'A' is identical to lowercase 'a').
<LI> Similar - number of characters which are non-identical - which score > 0 in the comparison matrix when compared to the reference sequence, but which are not identical.
<LI> Different - number of characters which score <= 0 in the comparison matrix when compared to the reference sequence.
<LI> %Change - a simple measure of the percentage change as compared to the reference sequence: (GapLen - Ident) * 100 / GapLen.
<LI> Description - the description annotation of teh sequence (if any).
</UL>

<p>

If qualifiers to inhibit various columns of information are used, then
the remaining columns of information are output in the same order as
shown above, so if '-noseqlength' is used, the order of output is: usa,
name, alignlength, gaps, gapcount, idcount, simcount, diffcount, change,
description. 

<p>

When the -html qualifier is specified, then the output will be wrapped
in HTML tags, ready for inclusion in a Web page.  Note that tags such as
<HTML> and <BODY> are not output by this program as the table of
databases is expected to form only part of the contents of a web page -
the rest of the web page must be supplier by the user. 

<p>

The lines of output information are guaranteed not to have trailing
white-space at the end. 


<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.
	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
   -->

<b>infoalign</b> reads in scoring matrices to determine the consensus
sequence and to determine which matches are similar or not. 

<p>

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>






<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

By default, the output file starts each line with the USA of the
sequence being described, so the output file is a list file that can be
manually edited and read in by other EMBOSS programs using the list-file
specification of '@filename'. 






<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="emma.html">emma</a></td><td>Multiple alignment program - interface to ClustalW program</td></tr>
<tr><td><a href="infoseq.html">infoseq</a></td><td>Displays some simple information about sequences</td></tr>
<tr><td><a href="plotcon.html">plotcon</a></td><td>Plots the quality of conservation of a sequence alignment</td></tr>
<tr><td><a href="prettyplot.html">prettyplot</a></td><td>Displays aligned sequences, with colouring and boxing</td></tr>
<tr><td><a href="seealso.html">seealso</a></td><td>Finds programs sharing group names</td></tr>
<tr><td><a href="showalign.html">showalign</a></td><td>Displays a multiple sequence alignment</td></tr>
<tr><td><a href="showdb.html">showdb</a></td><td>Displays information on the currently available databases</td></tr>
<tr><td><a href="textsearch.html">textsearch</a></td><td>Search sequence documentation text. SRS and Entrez are faster!</td></tr>
<tr><td><a href="tfm.html">tfm</a></td><td>Displays a program's help documentation manual</td></tr>
<tr><td><a href="wossname.html">wossname</a></td><td>Finds programs by keywords in their one-line documentation</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

This application was written by Gary Williams <a
href="mailto:gwilliam@hgmp.mrc.ac.uk">(gwilliam@hgmp.mrc.ac.uk)</a>
















<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (June 2001) - Gary Williams














<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->








</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

