<!-- START OF HEADER -->
<HTML><HEAD>
<TITLE> EMBASSY: INTERFACE documentation. </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

<H2> INTERFACE documentation
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<!-- END OF HEADER -->




<!-- CONTENTS
     This always includes the sections below.
     Other subsections can be added for individual applications.
-->
<H2>CONTENTS </H2>
<b> <a href="#1.0">1.0     SUMMARY                   </a></b><br>
<b> <a href="#2.0">2.0     INPUTS & OUTPUTS          </a></b><br>
<b> <a href="#3.0">3.0     INPUT FILE FORMAT         </a></b><br>
<b> <a href="#4.0">4.0     OUTPUT FILE FORMAT        </a></b><br>
<b> <a href="#5.0">5.0     DATA FILES                </a></b><br>
<b> <a href="#6.0">6.0     USAGE                     </a></b><br>    
<b> <a href="#7.0">7.0     KNOWN BUGS & WARNINGS     </a></b><br>    
<b> <a href="#8.0">8.0     NOTES                     </a></b><br>
<b> <a href="#9.0">9.0     DESCRIPTION               </a></b><br>
<b> <a href="#10.0">10.0   ALGORITHM                 </a></b><br>
<b> <a href="#11.0">11.0   RELATED APPLICATIONS      </a></b><br>
<b> <a href="#12.0">12.0   DIAGNOSTIC ERROR MESSAGES </a></b><br>
<b> <a href="#13.0">13.0   AUTHORS                   </a></b><br>
<b> <a href="#14.0">14.0   REFERENCES                </a></b><br>


<!-- SUMMARY
     Succint description of the application, particularly its inputs, outputs
     and what it does.  The same text is given at the top of the source (.c)
     file and in the <documentation> attribute of the <application definition>
     of the ACD file.
-->
<a name="1.0"></a>
<br><br><br><H2> 1.0   SUMMARY  </H2>
Reads CCF files (clean coordinate files) and writes CON files (contact files) of inter-chain residue-residue contact data.

Generate inter-chain CON files from CCF files



<!-- INPUTS & OUTPUTS
     Short summary of the application inputs and outputs in its different 
     modes of usage (if appropriate). More detail than the summary.
-->
<a name="2.0"></a>
<br><br><br><H2> 2.0   INPUTS & OUTPUTS          </H2>
INTERFACE reads a protein CCF file (clean coordinate file) and writes a CON file (contacts file) of inter-chain residue-residue contact data.  The file contains residue contact data for all pairs of chains in each model in the CCF file.  The input and output files are specified by the user.  A log file is also written.


<!-- INPUT FILE FORMAT 
     Description and example(s) of input file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.
     Use "<b>INTERFACE</b> reads any normal sequence USAs." if
     appropriate.
-->



<a name="3.0"></a>
<br><br><br><H2> 3.0   INPUT FILE FORMAT         </H2>
The format of the protein CCF file is described in the <a href="pdbparse.html">PDBPARSE documentation.</a>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: ../pdbplus-keep/2hhb.ccf</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   2hhb
XX
DE   HEMOGLOBIN (DEOXY)
XX
OS   HUMAN (HOMO SAPIENS)
XX
EX   METHOD xray; RESO 1.74; NMOD 1; NCHN 4; NGRP 0;
XX
CN   [1]
XX
IN   ID A; NR 141; NL 1; NH 7; NE 0;
XX
SQ   SEQUENCE   141 AA;  15127 MW;  5EC7DB1E CRC32;
     VLSPADKTNV KAAWGKVGAH AGEYGAEALE RMFLSFPTTK TYFPHFDLSH GSAQVKGHGK
     KVADALTNAV AHVDDMPNAL SALSDLHAHK LRVDPVNFKL LSHCLLVTLA AHLPAEFTPA
     VHASLDKFLA SVSTVLTSKY R
XX
CN   [2]
XX
IN   ID B; NR 146; NL 1; NH 8; NE 0;
XX
SQ   SEQUENCE   146 AA;  15868 MW;  EC9744C9 CRC32;
     VHLTPEEKSA VTALWGKVNV DEVGGEALGR LLVVYPWTQR FFESFGDLST PDAVMGNPKV
     KAHGKKVLGA FSDGLAHLDN LKGTFATLSE LHCDKLHVDP ENFRLLGNVL VCVLAHHFGK
     EFTPPVQAAY QKVVAGVANA LAHKYH
XX
CN   [3]
XX
IN   ID C; NR 141; NL 1; NH 7; NE 0;
XX
SQ   SEQUENCE   141 AA;  15127 MW;  5EC7DB1E CRC32;
     VLSPADKTNV KAAWGKVGAH AGEYGAEALE RMFLSFPTTK TYFPHFDLSH GSAQVKGHGK
     KVADALTNAV AHVDDMPNAL SALSDLHAHK LRVDPVNFKL LSHCLLVTLA AHLPAEFTPA
     VHASLDKFLA SVSTVLTSKY R
XX
CN   [4]
XX
IN   ID D; NR 146; NL 2; NH 8; NE 0;
XX
SQ   SEQUENCE   146 AA;  15868 MW;  EC9744C9 CRC32;
     VHLTPEEKSA VTALWGKVNV DEVGGEALGR LLVVYPWTQR FFESFGDLST PDAVMGNPKV
     KAHGKKVLGA FSDGLAHLDN LKGTFATLSE LHCDKLHVDP ENFRLLGNVL VCVLAHHFGK
     EFTPPVQAAY QKVVAGVANA LAHKYH
XX
RE   1    1    1    1     V VAL   .    .    .    .    .    C      360.00  130.57  137.00  142.86   94.30  113.45   99.30   29.40   79.10  113.56   98.30   29.29   81.40
RE   1    1    2    2     L LEU   .    .    .    .    .    C      -80.50  121.94   16.80   18.20   10.20    7.73    5.50   10.48   27.90    7.73    5.40   10.48   28.80
RE   1    1    3    3     S SER   1    AA   H    1    .    C      -71.32  164.78   44.40   44.59   38.30   44.42   56.90    0.17    0.40   39.26   80.90    5.34    7.90
RE   1    1    4    4     P PRO   1    AA   H    1    1    H      -62.31  -35.05   99.60  105.25   77.30  103.02   85.90    2.23   13.70  103.65   85.70    1.60   10.50
RE   1    1    5    5     A ALA   1    AA   H    1    1    H      -65.00  -41.52   59.80   65.59   60.80   64.35   92.70    1.24    3.20   64.35   90.20    1.24    3.40
RE   1    1    6    6     D ASP   1    AA   H    1    1    H      -56.29  -42.84   13.60   13.83    9.90   13.80   13.40    0.03    0.10    0.44    0.90   13.39   14.70


<font color=red>  [Part of this file has been deleted for brevity]</font>

AT   1    .    .    .    174   . HOH   W O        -4.764   -6.228    5.515    8.00   40.89
AT   1    .    .    .    175   . HOH   W O        23.809   19.925    1.758    8.00   39.37
AT   1    .    .    .    176   . HOH   W O        -7.871   -9.078    2.406    8.00   43.37
AT   1    .    .    .    177   . HOH   W O         4.693   12.083    7.558    8.00   40.24
AT   1    .    .    .    178   . HOH   W O         8.775  -23.438   16.055    8.00   42.33
AT   1    .    .    .    179   . HOH   W O        -7.480  -10.898   17.998    8.00   38.06
AT   1    .    .    .    180   . HOH   W O        -4.731   16.453    2.295    8.00   36.37
AT   1    .    .    .    181   . HOH   W O        -1.055   11.866   -0.448    8.00   43.19
AT   1    .    .    .    182   . HOH   W O       -27.610  -10.991    5.353    8.00   43.46
AT   1    .    .    .    183   . HOH   W O        26.015   11.766    5.159    8.00   40.95
AT   1    .    .    .    184   . HOH   W O       -18.517   -8.355   15.267    8.00   35.55
AT   1    .    .    .    185   . HOH   W O       -14.034    2.806  -30.367    8.00   41.77
AT   1    .    .    .    186   . HOH   W O       -32.905   -9.033    0.480    8.00   43.68
AT   1    .    .    .    187   . HOH   W O       -28.749  -13.315    1.938    8.00   45.36
AT   1    .    .    .    188   . HOH   W O         0.516   -8.074  -26.354    8.00   41.53
AT   1    .    .    .    189   . HOH   W O       -20.080   -9.873  -22.862    8.00   36.25
AT   1    .    .    .    190   . HOH   W O       -13.442    9.778  -13.572    8.00   39.70
AT   1    .    .    .    191   . HOH   W O       -24.804   -2.608  -15.488    8.00   37.79
AT   1    .    .    .    192   . HOH   W O         6.547    9.706   16.296    8.00   41.86
AT   1    .    .    .    193   . HOH   W O         0.029   22.606   14.164    8.00   43.02
AT   1    .    .    .    194   . HOH   W O       -11.367    0.306   28.463    8.00   44.30
AT   1    .    .    .    195   . HOH   W O       -19.950  -10.635   14.301    8.00   40.17
AT   1    .    .    .    196   . HOH   W O        -7.047   -6.324   20.098    8.00   36.98
AT   1    .    .    .    197   . HOH   W O       -23.876    1.108   14.102    8.00   33.31
AT   1    .    .    .    198   . HOH   W O       -34.199    8.033   11.037    8.00   40.72
AT   1    .    .    .    199   . HOH   W O       -14.173   13.393   -8.778    8.00   43.21
AT   1    .    .    .    200   . HOH   W O        11.388  -11.044   24.763    8.00   39.34
AT   1    .    .    .    201   . HOH   W O         3.735   -3.643    2.734    8.00   42.17
AT   1    .    .    .    202   . HOH   W O         3.149   -0.692    2.083    8.00   41.40
AT   1    .    .    .    203   . HOH   W O         4.511  -25.886   13.006    8.00   39.83
AT   1    .    .    .    204   . HOH   W O         8.712  -21.655    3.577    8.00   43.08
AT   1    .    .    .    205   . HOH   W O        22.926   -4.304   24.079    8.00   38.10
AT   1    .    .    .    206   . HOH   W O        11.435    9.654   20.618    8.00   40.23
AT   1    .    .    .    207   . HOH   W O        18.099    5.542   27.744    8.00   39.03
AT   1    .    .    .    208   . HOH   W O        12.174    9.951    9.804    8.00   44.34
AT   1    .    .    .    209   . HOH   W O        24.745   -2.501   15.270    8.00   39.78
AT   1    .    .    .    210   . HOH   W O        24.231    0.100   14.764    8.00   42.94
AT   1    .    .    .    211   . HOH   W O        23.324  -18.136   10.981    8.00   53.60
AT   1    .    .    .    212   . HOH   W O        25.576  -22.211    6.309    8.00   45.18
AT   1    .    .    .    213   . HOH   W O        14.639   24.823   -4.300    8.00   41.35
AT   1    .    .    .    214   . HOH   W O        14.903    5.393  -23.047    8.00   37.45
AT   1    .    .    .    215   . HOH   W O        16.650   -5.137  -16.717    8.00   39.12
AT   1    .    .    .    216   . HOH   W O         7.424   -6.700  -20.085    8.00   38.62
AT   1    .    .    .    217   . HOH   W O        -1.263   -2.837  -21.251    8.00   45.10
AT   1    .    .    .    218   . HOH   W O        23.120   -3.118  -12.992    8.00   37.05
AT   1    .    .    .    219   . HOH   W O        23.664    0.968  -14.389    8.00   36.25
AT   1    .    .    .    220   . HOH   W O        25.698    7.981  -15.362    8.00   35.85
AT   1    .    .    .    221   . HOH   W O        30.009   16.347   -6.794    8.00   37.62
AT   1    .    .    .    222   . HOH   W O        27.728   16.677   -1.376    8.00   42.54
AT   1    .    .    .    223   . HOH   W O         8.142   18.836    1.041    8.00   39.90
//
</pre>
</td></tr></table><p>







<!-- OUTPUT FILE FORMAT 
     Description and example(s) of output file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.  
     If the standard description of the avalable report formats is required, 
     use:   #include file="inc/reportformats.ihtml"
     Use "Outputs a graph to the specified graphics device."
     or  "outputs a report format file. The default format is ..."
     if appropriate.
-->
<a name="4.0"></a>
<br><br><br><H2> 4.0   OUTPUT FILE FORMAT         </H2>

The CON format used for the contact files (Figure 1) is similar to EMBL format and is described in the <a href="contacts.html">CONTACTS documentation.</a>  A few of the records differ in the INTERFACE output compared to the CONTACTS output, however, so for the sake of clarity all records are described below.

<UL>
<LI> XX - used for spacing and comments. The first line is bibliographic information and contains
 the text "Inter-chain residue-residue contact data".
<LI> TY - type of contact.  For CON files generated by CONTACTS, 'INTER' is always given.
<LI>  EX - experimental information. The value of the threshold contact 
distance is given as a floating point number after 'THRESH'. For CON files generated by 
INTERFACE, a '.' is given after 'THRESH' (this record is used by the CONTACTS application and
can be disregarded). The number of models and number of polypeptide chains are given after 
'NMOD' and 'NCHA' respectively. For domain CCF files a 1 is always given.
<LI> NE - number of entries (interfaces) in the file. For CON files generated by INTERFACE this is 
the number of interfaces (pairs of chains), equal to the number of models multiplied by
the number of unique interfaces.
Following the NE record, the file has a section for each entry containing
records for entry number (EN), identifier codes (ID),  
polypeptide chain-specific data (CN), chain sequence information (S1 & S2) and 
number of contacts (NC), that together define the interface and its contacts. 
<LI> EN - entry number.  The number in brackets indicates the start of an entry (interface).
<LI> CN - polypeptide chain-specific data.  Tokens delimiting data items are as follow. 
 (1) MO: The model number (from the PDB file).  (2) CN1: Chain number (first chain). 
(3) CN2: Chain number (second chain).  (4) ID1: PDB chain identifier (first chain)
(a '.` given in cases where a chain identifier was not specified in the original PDB file or, for domain CCF files, the domain 
is comprised of more than one domain). (5) ID2: Chain identifier (second chain). (6) NRES1: number of residues (first chain). 
(7) NRES2: number of residues (second chain).
<LI> NC - number of contacts: (1) SM: Number of residue-residue contacts; between side-chain or 
main-chain atoms of an amino acid residue in one chain and the side- or main-chain atoms of a 
residue in a different chain. (2) LI: Number of residue-ligand contacts; between side-chain or 
main-chain atoms of an amino acid residue and a ligand.
<LI> ID - identifier codes: (1) PDB: 4-character PDB identifier code. (2) DOM: SCOP or CATH domain identifier
code. (3) LIG: Not used by INTERFACE, a '.' is given. 
<LI> S1 - polypeptide chain sequence (first chain). The number of residues is given 
before AA on the first line.  The sequece is given on subsequent lines. 
<LI> S2 - polypeptide chain sequence (second chain).
<LI> SM - Line of residue contact data. Pairs of amino acid identifiers and 
residue numbers are delimited by a ';'. The left-hand term refers to the first chain and 
the right-hand term the second chain.  Residue numbers are taken from the 
CCF files and give a correct index into the sequences (i.e. they
are not necessarily the same as the original PDB file).  These sequence is 
given in the CON file itself (S1 & S2 records).
<LI> // - delimiter for individual entries in the file and also given on
the last line of the file.
</UL> 



<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: 2hhb.con</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
XX   Inter-chain residue-residue contact data.
XX
TY   INTER
XX
EX   THRESH 1.0; IGNORE 20.0; NMOD 1; NCHA 4
XX
NE   6
XX
EN   [1]
XX
ID   PDB 2hhb; DOM .; LIG .
XX
CN   MO 1; CN1 1; CN2 2; ID1 A; ID2 B; NRES1 141; NRES2 146
XX
S1   SEQUENCE   141 AA;  15127 MW;  5EC7DB1E CRC32;
     VLSPADKTNV KAAWGKVGAH AGEYGAEALE RMFLSFPTTK TYFPHFDLSH GSAQVKGHGK
     KVADALTNAV AHVDDMPNAL SALSDLHAHK LRVDPVNFKL LSHCLLVTLA AHLPAEFTPA
     VHASLDKFLA SVSTVLTSKY R
XX
S2   SEQUENCE   146 AA;  15868 MW;  EC9744C9 CRC32;
     VHLTPEEKSA VTALWGKVNV DEVGGEALGR LLVVYPWTQR FFESFGDLST PDAVMGNPKV
     KAHGKKVLGA FSDGLAHLDN LKGTFATLSE LHCDKLHVDP ENFRLLGNVL VCVLAHHFGK
     EFTPPVQAAY QKVVAGVANA LAHKYH
XX
NC   SM 48; LI .
XX
SM   GLU 30 ; PRO 124
SM   ARG 31 ; PHE 122
SM   ARG 31 ; THR 123
SM   ARG 31 ; PRO 124
SM   ARG 31 ; GLN 127
SM   LEU 34 ; PRO 124
SM   LEU 34 ; PRO 125
SM   LEU 34 ; ALA 128
SM   SER 35 ; GLN 127
SM   SER 35 ; ALA 128
SM   SER 35 ; GLN 131
SM   PHE 36 ; GLN 131
SM   HIS 103 ; ASN 108
SM   HIS 103 ; VAL 111
SM   HIS 103 ; CYS 112
SM   HIS 103 ; GLN 127
SM   HIS 103 ; GLN 131
SM   CYS 104 ; GLN 127
SM   LEU 106 ; CYS 112
SM   VAL 107 ; VAL 111
SM   VAL 107 ; CYS 112
SM   VAL 107 ; ALA 115
SM   VAL 107 ; PHE 122
SM   VAL 107 ; GLN 127


<font color=red>  [Part of this file has been deleted for brevity]</font>

SM   GLU 27 ; PRO 124
SM   GLU 30 ; PRO 124
SM   ARG 31 ; PHE 122
SM   ARG 31 ; THR 123
SM   ARG 31 ; PRO 124
SM   ARG 31 ; GLN 127
SM   LEU 34 ; PRO 124
SM   LEU 34 ; PRO 125
SM   LEU 34 ; ALA 128
SM   SER 35 ; GLN 127
SM   SER 35 ; ALA 128
SM   SER 35 ; GLN 131
SM   PHE 36 ; GLN 131
SM   HIS 103 ; ASN 108
SM   HIS 103 ; VAL 111
SM   HIS 103 ; CYS 112
SM   HIS 103 ; GLN 127
SM   HIS 103 ; GLN 131
SM   CYS 104 ; GLN 127
SM   LEU 106 ; CYS 112
SM   VAL 107 ; VAL 111
SM   VAL 107 ; CYS 112
SM   VAL 107 ; ALA 115
SM   VAL 107 ; PHE 122
SM   VAL 107 ; GLN 127
SM   ALA 110 ; CYS 112
SM   ALA 110 ; ALA 115
SM   ALA 110 ; HIS 116
SM   ALA 111 ; ALA 115
SM   ALA 111 ; GLY 119
SM   LEU 113 ; HIS 116
SM   PRO 114 ; HIS 116
SM   PHE 117 ; ARG 30
SM   PHE 117 ; CYS 112
SM   PHE 117 ; HIS 116
SM   THR 118 ; ARG 30
SM   PRO 119 ; ARG 30
SM   PRO 119 ; VAL 33
SM   PRO 119 ; VAL 34
SM   PRO 119 ; MET 55
SM   ALA 120 ; VAL 33
SM   ALA 120 ; PRO 51
SM   HIS 122 ; ARG 30
SM   HIS 122 ; VAL 34
SM   HIS 122 ; VAL 109
SM   HIS 122 ; CYS 112
SM   ALA 123 ; VAL 33
SM   ALA 123 ; VAL 34
SM   ASP 126 ; VAL 34
SM   ASP 126 ; TYR 35
//
</pre>
</td></tr></table><p>
<p><h3>File: interface.log</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
2hhb
</pre>
</td></tr></table><p>




<!-- DATA FILES         
     Any data files used (e.g. translation table file, substitution matrix 
     etc.  This includes example data file formats if they are not obvious.
     For a standard description of what data files are and how embossdata can
     be used to inspect and retrieve them, use:
     #include file="inc/localfiles.ihtml" 
-->

<a name="5.0"></a>
<br><br><br><H2> 5.0   DATA FILES                </H2>

INTERFACE uses a data file containing van der Waals radii for atoms in 
proteins (see <a href="contacts.html">CONTACTS documentation.</a>)
The file Evdw.dat is such a data file and is part of the EMBOSS distribution.
<!-- USAGE
     Example usage, as run from the command-line.
     Many examples illustrating different behaviours is good.
-->
<a name="6.0"></a>
<br><br><br><H2> 6.0   USAGE                     </H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" >
<tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-infile]            infile     This option specifies the name of the
                                  protein CCF file (clean coordinate file)
                                  (input). A 'clean cordinate file' contains
                                  protein coordinate and derived data for a
                                  single PDB file ('protein clean coordinate
                                  file') or a single domain from SCOP or CATH
                                  ('domain clean coordinate file'), in CCF
                                  format (EMBL-like). The files, generated by
                                  using PDBPARSE (PDB files) or DOMAINER
                                  (domains), contain 'cleaned-up' data that is
                                  self-consistent and error-corrected.
                                  Records for residue solvent accessibility
                                  and secondary structure are added to the
                                  file by using PDBPLUS.
   -thresh             float      This option specifies the threshold contact
                                  distance. Contact between two residues is
                                  defined as when the van der Waals surface of
                                  any atom of the first residue comes within
                                  the threshold contact distance of the van
                                  der Waals surface of any atom of the second
                                  residue. The threshold contact distance is a
                                  user-defined distance with a default value
                                  of 1 Angstrom.
  [-outfile]           outfile    This option specifies the name of CON file
                                  (contact file) (output). A 'contact file'
                                  contains contact data for a protein or a
                                  domain from SCOP or CATH, in the CON format
                                  (EMBL-like). The contacts may be intra-chain
                                  residue-residue, inter-chain
                                  residue-residue or residue-ligand. The files
                                  are generated by using CONTACTS, INTERFACE
                                  and SITES.
   -conlogfile         outfile    This option specifies the name of the log
                                  file for the build. The log file contains
                                  messages about any errors arising while
                                  INTERFACE ran.

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers:
   -vdwfile            datafile   This option specifies the name of data file
                                  with van der Waals radii. The file of van
                                  der Waals radii for atoms in amino acid
                                  residues is part of the emboss distribution.
   -ignore             float      This option specifies the threshold ignore
                                  distance. If any two atoms from two
                                  different residues are at least this
                                  distance apart then no futher inter-atomic
                                  contacts will be checked for for that
                                  residue pair . This speeds the calculation
                                  up considerably.

   Associated qualifiers:

   "-outfile" associated qualifiers
   -odirectory2        string     Output directory

   "-conlogfile" associated qualifiers
   -odirectory         string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report deaths

</pre>
</pre>
</td></tr>
</table>


<H3> 6.1   COMMAND LINE ARGUMENTS </H3>
 
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-infile]<br>(Parameter 1)</td>
<td>This option specifies the name of the protein CCF file (clean coordinate file) (input). A 'clean cordinate file' contains protein coordinate and derived data for a single PDB file ('protein clean coordinate file') or a single domain from SCOP or CATH ('domain clean coordinate file'), in CCF format (EMBL-like). The files, generated by using PDBPARSE (PDB files) or DOMAINER (domains), contain 'cleaned-up' data that is self-consistent and error-corrected. Records for residue solvent accessibility and secondary structure are added to the file by using PDBPLUS.</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-thresh</td>
<td>This option specifies the threshold contact distance. Contact between two residues is defined as when the van der Waals surface of any atom of the first residue comes within the threshold contact distance of the van der Waals surface of any atom of the second residue. The threshold contact distance is a user-defined distance with a default value of 1 Angstrom.</td>
<td>Any numeric value</td>
<td>1.0</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>This option specifies the name of CON file (contact file) (output). A 'contact file' contains contact data for a protein or a domain from SCOP or CATH, in the CON format (EMBL-like). The contacts may be intra-chain residue-residue, inter-chain residue-residue or residue-ligand. The files are generated by using CONTACTS, INTERFACE and SITES.</td>
<td>Output file</td>
<td>test.con</td>
</tr>

<tr>
<td>-conlogfile</td>
<td>This option specifies the name of the log file for the build. The log file contains messages about any errors arising while INTERFACE ran.</td>
<td>Output file</td>
<td>interface.log</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-vdwfile</td>
<td>This option specifies the name of data file with van der Waals radii. The file of van der Waals radii for atoms in amino acid residues is part of the emboss distribution.</td>
<td>Data file</td>
<td>Evdw.dat</td>
</tr>

<tr>
<td>-ignore</td>
<td>This option specifies the threshold ignore distance. If any two atoms from two different residues are at least this distance apart then no futher inter-atomic contacts will be checked for for that residue pair . This speeds the calculation up considerably.</td>
<td>Any numeric value</td>
<td>20.0</td>
</tr>

</table>


<H3> 6.2 EXAMPLE SESSION </H3>
An example of interactive use of INTERFACE is shown below.

<b>Here is a sample session with interface</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>interface </b>
Generate inter-chain CON files from CCF files.
Name of protein CCF file (clean coordinate file) (input): <b>../pdbplus-keep/2hhb.ccf</b>
Threshold contact distance [1.0]: <b>1</b>
Name of CON file (contact file) (output) [test.con]: <b>2hhb.con</b>
Name of log file for the build [interface.log]: <b></b>

2hhb

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>


<!-- <br><br>
The protein CCF file /test_data/interface/2hhb.ccf was read and the CON file /test_data/interface/2hhb.con was written.  van der Waals radii were taken from the emboss data file Evdw.dat which is specified as a default value in interface.acd and the default contact distance of 1 Angstrom was used to determine contacts.  The log file test_data/interface/interface.log was written.
<br><br>


<!-- KNOWN BUGS & WARNINGS
     Bugs that have not yet been fixed, easily missued features, problems
     and caveats etc. Potentially stupid things the program will let you do.
-->
<a name="7.0"></a>
<br><br><br><H2> 7.0   KNOWN BUGS & WARNINGS     </H2>
None.








<!-- NOTES
     Important general remarks, including:
       Restrictions.
       Interesting behaviour.
       Useful things you can do with this program.
       Future plans.
       etc.
-->
<a name="8.0"></a>
<br><br><br><H2> 8.0   NOTES                     </H2>

<b> Types of contact </b>
<br><b>SM</b> records are used for contacts between either either side-chain or main-chain atoms 
as defined above.  In a future implementation, <b>SS</b> will be used for side-chain only contacts,
 <b>MM</b> will be used for main-chain only contacts, <b>LI</b> will be used for contacts to
ligands, and there will probably be several other forms of contact too.

<br><br><b>Threshold ignore distance option</b>
<br>The threshold ignore distance can be adjusted to speed-up the calculation of contacts considerably.  If any two atoms from two different residues are at least this distance apart then no futher inter-atomic contacts will be checked for for that residue pair. 


<H3> 8.1   GLOSSARY OF FILE TYPES    </H3>
<table BORDER CELLSPACING=5 CELLPADDING=5 BGCOLOR="#f5f5ff" >
<tr>
<td><b>FILE TYPE</b></td>
<td><b>FORMAT</b></td>
<td><b>DESCRIPTION</b></td>
<td><b>CREATED BY <b></td>
<td><b>SEE ALSO</b></td>
</tr>
<tr>
<td><b> Clean coordinate file (for protein) </b></td>
<td> CCF format (EMBL-like).</td>
<td> Protein coordinate and derived data for a single PDB file.  The data are 'cleaned-up': self-consistent and error-corrected.  </td>
<td> <a href="pdbparse.html">PDBPARSE</a> </td>
<td> Records for residue solvent accessibility and secondary structure are added to the file by using PDBPLUS. </td>
</tr>
<tr>
<td><b> Clean coordinate file (for domain)</b></td>
<td> CCF format (EMBL-like).</td>
<td> Protein coordinate and derived data for a single domain from SCOP or CATH. The data are 'cleaned-up': self-consistent and error-corrected. </td>
<td> <a href="domainer.html">DOMAINER</a></td>
<td> Records for residue solvent accessibility and secondary structure are added to the file by using PDBPLUS.</td>
</tr>
<tr>
<td><b> Contact file (intra-chain residue-residue contacts)</b></td>
<td> CON format (EMBL-like.) </td>
<td> Intra-chain residue-residue contact data for a protein or a domain from SCOP or CATH.  </td>
<td> <a href="contacts.html">CONTACTS</a> </td>
<td> N.A. </td>
</tr>
<tr>
<td><b> Contact file (inter-chain residue-residue contacts)</b></td>
<td> CON format (EMBL-like.) </td>
<td> Inter-chain residue-residue contact data for a protein or a domain from SCOP or CATH.  </td>
<td> <a href="interface.html">INTERFACE</a> </td>
<td> N.A. </td>
</tr>
<tr>
<td><b> Contact file (residue-ligand contacts)</b></td>
<td> CON format (EMBL-like.) </td>
<td> Residue-ligand contact data for a protein or a domain from SCOP or CATH. </td>
<td> <a href="sites.html">SITES</a> </td>
<td> N.A. </td>
</tr>
<tr>
<td><b>van der Waals radii  </b></td>
<td> A file of van der Waals radii for atoms in amino acid residues. Part of the emboss distribution.</td>
<td> N.A. </td>
<td> N.A. </td>
</tr>
</table>




None


<!-- DESCRIPTION
     A complete, non-technical, user-level description of the application.
-->
<a name="9.0"></a>
<br><br><br><H2> 9.0   DESCRIPTION               </H2>
Knowledge of the physical contacts that amino acid residues in two different polypeptide chains of a protein make with one another is required for several different analyses.  INTERFACE calculates inter-chain residue-residue contact data from protein CCF files (clean coordinate files).


<!-- ALGORITHM
     A technical description of algorithmic aspects, describing exactly how
     the key aspects of the application work.
-->
<a name="10.0"></a>
<br><br><br><H2> 10.0  ALGORITHM                 </H2>
Contact between two residues is defined as when the van der Waals surface of any atom of the first residue comes within the threshold contact distance of the van der Waals surface of any atom of the second residue. The threshold contact distance is a user-defined distance with a default value of 1 Angstrom.



<!-- RELATED APPLICATIONS
     Other applications that either generate the input, use the output or 
     are in some other way related to the application are described here. 
-->

<a name="11.0"></a>
<br><br><br><H2> 11.0  RELATED APPLICATIONS      </H2>

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="../../apps/aaindexextract.html">aaindexextract</a></td><td>Extract data from AAINDEX</td></tr>
<tr><td><a href="../domalign/allversusall.html">allversusall</a></td><td>Sequence similarity data from all-versus-all comparison</td></tr>
<tr><td><a href="../domainatrix/cathparse.html">cathparse</a></td><td>Generates DCF file from raw CATH files</td></tr>
<tr><td><a href="../../apps/cutgextract.html">cutgextract</a></td><td>Extract data from CUTG</td></tr>
<tr><td><a href="domainer.html">domainer</a></td><td>Generates domain CCF files from protein CCF files</td></tr>
<tr><td><a href="../domainatrix/domainnr.html">domainnr</a></td><td>Removes redundant domains from a DCF file</td></tr>
<tr><td><a href="../domainatrix/domainseqs.html">domainseqs</a></td><td>Adds sequence records to a DCF file</td></tr>
<tr><td><a href="../domainatrix/domainsse.html">domainsse</a></td><td>Add secondary structure records to a DCF file</td></tr>
<tr><td><a href="hetparse.html">hetparse</a></td><td>Converts heterogen group dictionary to EMBL-like format</td></tr>
<tr><td><a href="pdbplus.html">pdbplus</a></td><td>Add accessibility & secondary structure to a CCF file</td></tr>
<tr><td><a href="pdbtosp.html">pdbtosp</a></td><td>Convert swissprot:PDB codes file to EMBL-like format</td></tr>
<tr><td><a href="../../apps/printsextract.html">printsextract</a></td><td>Extract data from PRINTS</td></tr>
<tr><td><a href="../../apps/prosextract.html">prosextract</a></td><td>Build the PROSITE motif database for use by patmatmotifs</td></tr>
<tr><td><a href="../../apps/rebaseextract.html">rebaseextract</a></td><td>Extract data from REBASE</td></tr>
<tr><td><a href="../domainatrix/scopparse.html">scopparse</a></td><td>Generate DCF file from raw SCOP files</td></tr>
<tr><td><a href="../domsearch/seqnr.html">seqnr</a></td><td>Removes redundancy from DHF files</td></tr>
<tr><td><a href="sites.html">sites</a></td><td>Generate residue-ligand CON files from CCF files</td></tr>
<tr><td><a href="../domainatrix/ssematch.html">ssematch</a></td><td>Search a DCF file for secondary structure matches</td></tr>
<tr><td><a href="../../apps/tfextract.html">tfextract</a></td><td>Extract data from TRANSFAC</td></tr>
</table>




<!-- DIAGNOSTIC ERROR MESSAGES
     Description of error messages or log file, if one is written.
-->
<a name="12.0"></a>
<br><br><br><H2> 12.0  DIAGNOSTIC ERROR MESSAGES </H2>
INTERFACE generates a log file an excerpt of which is shown below. If there is a problem in processing a CCF file, three lines containing the record '//', the PDB identifier code and an error message respectively are written. The text <i>'WARN  file open error filename' </i>,  <i>'ERROR file read error filename' </i> or  <i>'ERROR file write error filename ' </i> will be reported when an error was encountered during a file open, read or write respectively.  Various other error messages may also be given (in case of difficulty email Jon Ison, jison@hgmp.mrc.ac.uk).



<!-- AUTHORS
     The main author first then all others. 
-->
<a name="13.0"></a>
<br><br><br><H2> 13.0  AUTHORS                   </H2>
Jon Ison <a href="mailto:jison@rfcgr.mrc.ac.uk">(jison@rfcgr.mrc.ac.uk)</a>
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK




<!-- REFERENCES
     Quote the paper where the application was first published, described, used etc. 
-->
<a name="14.0"></a>
<br><br><br><H2> 14.0  REFERENCES                </H2>

Please cite the authors and EMBOSS.
<br><br>
<i>Rice P, Longden I and Bleasby A (2000) "EMBOSS - The European
Molecular Biology Open Software Suite"  Trends in Genetics,
15:276-278.

<p>
See also <a href="http://emboss.sourceforge.net/">http://emboss.sourceforge.net/</a></i>




<H3>14.1 Other useful references  </H3>


</BODY>
</HTML>



