<HTML>

<HEAD>
  <TITLE>
  EMBOSS: isochore
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: isochore
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program isochore
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Plots isochores in large DNA sequences


<H2>
    Description
</H2>


The nuclear genomes of vertebrates are mosaics of isochores, very long
stretches (&gt;300kb) of DNA that are homogeneous in base composition and
are compositionally correlated with the coding sequences that they
embed.  Isochores can be partitioned in a small number of families that
cover a range of GC levels (GC is the molar ratio of guanine+cytosine in
DNA), which is narrow in cold-blooded vertebrates, but broad in
warm-blooded vertebrates. 

<p>
This application plots GC content over a sequence.  It is intended for
large sequences such as complete chromosomes or large genomic contigs,
although interesting results can also be obtained from shorter
sequences. 

<H2>
    Usage
</H2>


Here is a sample session with isochore. This reads yeast chromosome 3
from a remote web server so the sequence reading may take a little time.

<pre>

% isochore
Input sequence: tgb:scchriii
Output file [scchriii.iso]: 
</pre>

<a href="./isochore0.gif">click here for result</a>


<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          sequence   Sequence USA
  [-out]               outfile    Output file name
   -graph              xygraph    Graph type

   Optional qualifiers:
   -window             integer    Window size
   -shift              integer    Shift increment

   Advanced qualifiers: (none)

</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-out]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.isochore</td>
</tr>

<tr>
<td>-graph</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including postscript, ps, hpgl, hp7470, hp7580, meta, colourps, cps, xwindows, x11, tektronics, tekt, tek4107t, tek, none, null, text, data, xterm</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-window</td>
<td>Window size</td>
<td>Integer 1 or more</td>
<td>1000</td>
</tr>

<tr>
<td>-shift</td>
<td>Shift increment</td>
<td>Integer 1 or more</td>
<td>100</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>


<H2>
    Output file format
</H2>


The graph from the example run 
<H2>
    Data files
</H2>


<H2>
    Notes
</H2>


<H2>
    References
</H2>


<ol>

<li>
Bernardi G 
Isochores and the evolutionary genomics of vertebrates. 
Gene 2000 Jan 4;241(1):3-17

<li>
Pesole G, Bernardi G, Saccone C 
Isochore specificity of AUG initiator context of human genes. 
FEBS Lett 1999 Dec 24;464(1-2):60-2
<li>

Bernardi G 
The human genome: organization and evolutionary history.
Annu Rev Genet 1995;29:445-76

</ol>

<H2>
    Warnings
</H2>


<H2>
    Diagnostic Error Messages
</H2>


<H2>
    Exit status
</H2>


<H2>
    Known bugs
</H2>


<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="cai.html">cai</a></td><td>CAI codon adaptation index</td></tr>
<tr><td><a href="chaos.html">chaos</a></td><td>Create a chaos game representation plot for a sequence</td></tr>
<tr><td><a href="chips.html">chips</a></td><td>Codon usage statistics</td></tr>
<tr><td><a href="codcmp.html">codcmp</a></td><td>Codon usage table comparison</td></tr>
<tr><td><a href="compseq.html">compseq</a></td><td>Counts the composition of dimer/trimer/etc words in a sequence</td></tr>
<tr><td><a href="cpgplot.html">cpgplot</a></td><td>Plot CpG rich areas</td></tr>
<tr><td><a href="cpgreport.html">cpgreport</a></td><td>Reports all CpG rich regions</td></tr>
<tr><td><a href="cusp.html">cusp</a></td><td>Create a codon usage table</td></tr>
<tr><td><a href="diffseq.html">diffseq</a></td><td>Find differences (SNPs) between nearly identical sequences</td></tr>
<tr><td><a href="dotmatcher.html">dotmatcher</a></td><td>Displays a thresholded dotplot of two sequences</td></tr>
<tr><td><a href="dotpath.html">dotpath</a></td><td>Displays a non-overlapping wordmatch dotplot of two sequences</td></tr>
<tr><td><a href="dottup.html">dottup</a></td><td>Displays a wordmatch dotplot of two sequences</td></tr>
<tr><td><a href="einverted.html">einverted</a></td><td>Finds DNA inverted repeats</td></tr>
<tr><td><a href="equicktandem.html">equicktandem</a></td><td>Finds tandem repeats</td></tr>
<tr><td><a href="etandem.html">etandem</a></td><td>Looks for tandem repeats in a nucleotide sequence</td></tr>
<tr><td><a href="freak.html">freak</a></td><td>Residue/base frequency table or plot</td></tr>
<tr><td><a href="geecee.html">geecee</a></td><td>Calculates the fractional GC content of nucleic acid sequences</td></tr>
<tr><td><a href="marscan.html">marscan</a></td><td>Finds MAR/SAR sites in nucleic sequences</td></tr>
<tr><td><a href="newcpgreport.html">newcpgreport</a></td><td>Report CpG rich areas</td></tr>
<tr><td><a href="newcpgseek.html">newcpgseek</a></td><td>Reports CpG rich regions</td></tr>
<tr><td><a href="palindrome.html">palindrome</a></td><td>Looks for inverted repeats in a nucleotide sequence</td></tr>
<tr><td><a href="polydot.html">polydot</a></td><td>Displays all-against-all dotplots of a set of sequences</td></tr>
<tr><td><a href="recode.html">recode</a></td><td>Remove restriction sites but maintain the same translation</td></tr>
<tr><td><a href="redata.html">redata</a></td><td>Search REBASE for enzyme name, references, suppliers etc</td></tr>
<tr><td><a href="restover.html">restover</a></td><td>Finds restriction enzymes that produce a specific overhang</td></tr>
<tr><td><a href="restrict.html">restrict</a></td><td>Finds restriction enzyme cleavage sites</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="silent.html">silent</a></td><td>Silent mutation restriction enzyme scan</td></tr>
<tr><td><a href="tfscan.html">tfscan</a></td><td>Scans DNA sequences for transcription factors</td></tr>
<tr><td><a href="wobble.html">wobble</a></td><td>Wobble base plot</td></tr>
<tr><td><a href="wordcount.html">wordcount</a></td><td>Counts words of a specified size in a DNA sequence</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Peter Rice <a
href="mailto:pmr@sanger.ac.uk">(pmr@sanger.ac.uk)</a> Informatics
Division, The Sanger Centre, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.

<H2>
    History
</H2>


<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

