<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: libscan
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="/images/emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

libscan
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<!--END OF HEADER-->






<H2>
    Function
</H2>
Scans a database with a set of models that represents a family

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
  -->

Scans each signature, profile or HMM in a directory against a sequence
database and writes a signature hits file for each one. Or scans sequences
against such a library of discriminating elements and writes a library scan
file for each one. 

<p>

This program is part of a suite of EMBOSS applications that directly or
indirectly make use of the protein structure databases pdb and scop.
This program is part of an experimental analysis pipeline described in an
accompanying document.  We provide the software in the hope that it will
be useful.  The applications were designed for specific research purposes
and may not be useful or reliable in contexts other than the described
pipeline.  The development of the suite was coordinated by Jon Ison to
whom enquiries and bug reports should be sent (email jison@hgmp.mrc.ac.uk).

<p>

The advantage of screening a relatively small library of discriminating
elements with a sequence is that it is sufficient for the sequence to
detect its true family (discriminator) in the first rank for an effective
prediction.  This is in contrast to searching a larger sequence database to
identify homology, where biologically significant hits may achieve
statistically insignificant scores and therefore be missed.  A library
might help the detection of such proteins because they may still score
their true discriminator higher than the others in the library regardless
of statistical estimates.  Further improvements to predictions are gained
when multiple sources of evidence, in this case the different types of
discriminating element, are considered.

<p>

The libscan application allows a protein sequence or sequences to be
screened against the library of discriminating elements.  The results of a
screen are returned to the user in a library scan file containing a list
of top-scoring SCOP domains rank-ordered by p-value for each individual
type of discriminator, and also for all of the discriminator types in
combination (combined prediction).  For the combined prediction, the
p-value is derived from an empirically derived distribution of the product
of the p-values of the individual methods.    

<p>

libscan runs in one of two modes either (i) database search mode or (ii)
library screen mode. In database search mode libscan reads one or more
directories each containing a single type of discriminating element, the
permitted types are sparse sequence signature, Gribskov profile, Henikoff
profile or hidden Markov model.  Each directory must contain a
discriminating element for each of one or more scop families; files for
indivdual families should have the same name (possibly with different file
extensions) in the different directories, e.g. the scop family with the
identifier 46445 might be represented by files 46445.hmm, 46445.sig,
46445.gribs and 46445.henik in their respective directories.  libscan scans
each discriminating element against a sequence set that must also be
provided by the user and generates a signature hits file (of scored hits
to sequences in the set) for each individual discriminating element.  In
mode (i), the user must also provide a scop validation file so that a
classification of hits can be given in the signature hits files.  The user
specifies a maximum number of false hits that are written to the signature
hits files.

<p>

In library screen mode, libscan
reads a sequence set, screens each sequence against the library
(directories of discriminating elements) and writes a library scan file
(of top-scoring families) for each sequence.  In either mode, a scop
classification file is required as a source of family classification data.
The paths and extensions of the sequence set and scop classification files
(input) different types of discriminating elements (input), and paths and
extension of signature hits files or library scan file (output) are
specified by the user.       












<H2>
    Algorithm
</H2>

<!-- 
        Algorithms
	  Pseudocode/Vague description/Handwaving/web links to useful pages
  -->



The approach for calculating p-values for individual discriminator-sequence
matches is the same irrespective of which mode libscan is run in and is
summarised as follows: (i) Each discriminating element is scanned (scored)
against each sequence.  (ii) The ditribution of scores is generated for
each different type of element.  (iii) Using these distributions, a p-value
is calculated for each discriminator-sequence match.
The approach for calculating the 'combined p-value' for each individual
sequence is as follows: (i) p-values of the matches for the different types
of discriminator versus a sequence are multiplied to yield the 'product of
p-values'.  (ii) A distribution of 'product of p-values' is generated for
each individual sequence.  (iii) Using these distributions, a so-called
'combined p-value', reflecting the overall significance of the matches of
the sequence to the different types of discriminators, is generated.
Importantly, two (or more) p-values are only ever multiplied together if
corresponding hit is to the same region of the protein.  'Same' is defined
as two hits overlapping by a user-defined number of residues. 










<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->





<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers (* if not always prompted):
   -mode               menu       Libscan runs in one of two modes either (i)
                                  database search mode or (ii) library screen
                                  mode. In database search mode libscan reads
                                  one or more directories each containing a
                                  single type of discriminating element, the
                                  permitted types are sparse sequence
                                  signature, Gribskov profile, Henikoff
                                  profile or hidden Markov model. In library
                                  screen mode, libscan reads a sequence set,
                                  screens each sequence against the library
                                  (directories of discriminating elements) and
                                  writes a library scan file (of top-scoring
                                  families) for each one.
   -db                 seqset     In database search mode libscan scans each
                                  discriminating element against a sequence
                                  set. In library screen mode, libscan reads a
                                  sequence set and screens each sequence
                                  against the library (directories of
                                  disciminating elements)
   -[no]grib           boolean    Use Gribskov profile
   -[no]henik          boolean    Use Henikoff profile
   -[no]hmm            boolean    Use HMM profile
   -[no]sig            boolean    Use signature
*  -hmmpath            string     Location of HMM profile files (input)
*  -hmmextn            string     Extension of HMM profile files (input)
*  -hmmoutpath         string     Location of signature hits/library scan
                                  files for hmm searches (output)
*  -hmmoutextn         string     Extension of signature hits/library scan
                                  files for hmm searches (output)
*  -gbvpath            string     Location of Gribskov profile files (input)
*  -gbvextn            string     Extension of Gribskov profile files (input)
*  -gbvgapo            float      Gap insertion penalty
*  -gbvgape            float      Gap extension penalty
*  -gbvoutpath         string     Location of signature hits/library scan
                                  files for Gribskov profile searches (output)
*  -gbvoutextn         string     Extension of signature hits/library scan
                                  files for Gribskov profile searches (output)
*  -hnfpath            string     Location of Hennikoff profile files (input)
*  -hnfextn            string     Extension of Hennikoff profile files (input)
*  -hnfgapo            float      Gap insertion penalty
*  -hnfgape            float      Gap extension penalty
*  -hnfoutpath         string     Location of signature hits/library scan
                                  files for Henikoff profile searches (output)
*  -hnfoutextn         string     Extension of signature hits/library scan
                                  files for Henikoff profile searches (output)
*  -sigpath            string     Location of signature files (input)
*  -sigextn            string     Extension of signature files (input)
*  -nterm              menu       Select number
*  -sub                matrixf    Residue substitution matrix
*  -siggapo            float      Gap insertion penalty
*  -siggape            float      Gap extension penalty
*  -sigoutpath         string     Location of signature hits/library scan
                                  files for signature searches (output)
*  -sigoutextn         string     Extension for signature hits/library scan
                                  files for signature searches (output)
   -scopf              infile     In either mode, a 'scop classification file'
                                  is required as a source of family
                                  classification data. A scop classification
                                  file contains classification and other data
                                  for domains from the scop database. The file
                                  is in embl-like format and is generated by
                                  scopparse. Domain sequence information can
                                  be added to the file by using scopseqs.
*  -targetf            infile     A 'scop validation file' contains sequence
                                  relatives (hits) for each of a number of
                                  different scop families, superfamilies and
                                  folds. The file contains the collated
                                  results from psiblast searches of a sequence
                                  database for the indvidual scop families;
                                  hits of unambiguous family assignment are
                                  listed under their respective family,
                                  otherwise a hit is assigned as relatives to
                                  a scop superfamily or fold instead. The scop
                                  validation file is generated by seqnr and
                                  is in embl-like format.
*  -maxhits            integer    The maximum number of hits to be written to
                                  the signature hits file.
*  -overlap            integer    When estimating the statistical significance
                                  of scored matches of different types of
                                  discriminating element to a sequence, two
                                  (or more) p-values are only ever multiplied
                                  together if corresponding hit is to the same
                                  region of the protein. 'Same' is defined as
                                  two hits overlapping by a user-defined
                                  number of residues.
   -mrgoutpath         string     Location of signature hits / library scan
                                  files for merged results (output)
   -mrgoutextn         string     Extension for signature hits / library scan
                                  files for merged results (output)

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-db" associated qualifiers
   -sbegin              integer    Start of each sequence to be used
   -send                integer    End of each sequence to be used
   -sreverse            boolean    Reverse (if DNA)
   -sask                boolean    Ask for begin/end/reverse
   -snucleotide         boolean    Sequence is nucleotide
   -sprotein            boolean    Sequence is protein
   -slower              boolean    Make lower case
   -supper              boolean    Make upper case
   -sformat             string     Input sequence format
   -sdbname             string     Database name
   -sid                 string     Entryname
   -ufo                 string     UFO features
   -fformat             string     Features format
   -fopenfile           string     Features file name

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-mode</td>
<td>Libscan runs in one of two modes either (i) database search mode or (ii) library screen mode. In database search mode libscan reads one or more directories each containing a single type of discriminating element, the permitted types are sparse sequence signature, Gribskov profile, Henikoff profile or hidden Markov model. In library screen mode, libscan reads a sequence set, screens each sequence against the library (directories of discriminating elements) and writes a library scan file (of top-scoring families) for each one.</td>
<td><table><tr><td>1</td> <td><i>(Database search mode)</i></td></tr><tr><td>2</td> <td><i>(Library screen mode)</i></td></tr></table></td>
<td>1</td>
</tr>

<tr>
<td>-db</td>
<td>In database search mode libscan scans each discriminating element against a sequence set. In library screen mode, libscan reads a sequence set and screens each sequence against the library (directories of disciminating elements)</td>
<td>Readable set of sequences</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-[no]grib</td>
<td>Use Gribskov profile</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]henik</td>
<td>Use Henikoff profile</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]hmm</td>
<td>Use HMM profile</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]sig</td>
<td>Use signature</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-hmmpath</td>
<td>Location of HMM profile files (input)</td>
<td>Any string is accepted</td>
<td>./</td>
</tr>

<tr>
<td>-hmmextn</td>
<td>Extension of HMM profile files (input)</td>
<td>Any string is accepted</td>
<td>.hmm</td>
</tr>

<tr>
<td>-hmmoutpath</td>
<td>Location of signature hits/library scan files for hmm searches (output)</td>
<td>Any string is accepted</td>
<td>./</td>
</tr>

<tr>
<td>-hmmoutextn</td>
<td>Extension of signature hits/library scan files for hmm searches (output)</td>
<td>Any string is accepted</td>
<td>.hmmout</td>
</tr>

<tr>
<td>-gbvpath</td>
<td>Location of Gribskov profile files (input)</td>
<td>Any string is accepted</td>
<td>./</td>
</tr>

<tr>
<td>-gbvextn</td>
<td>Extension of Gribskov profile files (input)</td>
<td>Any string is accepted</td>
<td>.gribs</td>
</tr>

<tr>
<td>-gbvgapo</td>
<td>Gap insertion penalty</td>
<td>Any numeric value</td>
<td>1.0</td>
</tr>

<tr>
<td>-gbvgape</td>
<td>Gap extension penalty</td>
<td>Any numeric value</td>
<td>1.0</td>
</tr>

<tr>
<td>-gbvoutpath</td>
<td>Location of signature hits/library scan files for Gribskov profile searches (output)</td>
<td>Any string is accepted</td>
<td>./</td>
</tr>

<tr>
<td>-gbvoutextn</td>
<td>Extension of signature hits/library scan files for Gribskov profile searches (output)</td>
<td>Any string is accepted</td>
<td>.gribout</td>
</tr>

<tr>
<td>-hnfpath</td>
<td>Location of Hennikoff profile files (input)</td>
<td>Any string is accepted</td>
<td>./</td>
</tr>

<tr>
<td>-hnfextn</td>
<td>Extension of Hennikoff profile files (input)</td>
<td>Any string is accepted</td>
<td>.henik</td>
</tr>

<tr>
<td>-hnfgapo</td>
<td>Gap insertion penalty</td>
<td>Any numeric value</td>
<td>1.0</td>
</tr>

<tr>
<td>-hnfgape</td>
<td>Gap extension penalty</td>
<td>Any numeric value</td>
<td>1.0</td>
</tr>

<tr>
<td>-hnfoutpath</td>
<td>Location of signature hits/library scan files for Henikoff profile searches (output)</td>
<td>Any string is accepted</td>
<td>./</td>
</tr>

<tr>
<td>-hnfoutextn</td>
<td>Extension of signature hits/library scan files for Henikoff profile searches (output)</td>
<td>Any string is accepted</td>
<td>.henikout</td>
</tr>

<tr>
<td>-sigpath</td>
<td>Location of signature files (input)</td>
<td>Any string is accepted</td>
<td>./</td>
</tr>

<tr>
<td>-sigextn</td>
<td>Extension of signature files (input)</td>
<td>Any string is accepted</td>
<td>.sig</td>
</tr>

<tr>
<td>-nterm</td>
<td>Select number</td>
<td><table><tr><td>1</td> <td><i>(Align anywhere and allow only complete signature-sequence fit)</i></td></tr><tr><td>2</td> <td><i>(Align anywhere and allow partial signature-sequence fit)</i></td></tr><tr><td>3</td> <td><i>(Use empirical gaps only)</i></td></tr></table></td>
<td>1</td>
</tr>

<tr>
<td>-sub</td>
<td>Residue substitution matrix</td>
<td>Comparison matrix file in EMBOSS data path</td>
<td>EBLOSUM62</td>
</tr>

<tr>
<td>-siggapo</td>
<td>Gap insertion penalty</td>
<td>Any numeric value</td>
<td>10.0</td>
</tr>

<tr>
<td>-siggape</td>
<td>Gap extension penalty</td>
<td>Any numeric value</td>
<td>0.5</td>
</tr>

<tr>
<td>-sigoutpath</td>
<td>Location of signature hits/library scan files for signature searches (output)</td>
<td>Any string is accepted</td>
<td>./</td>
</tr>

<tr>
<td>-sigoutextn</td>
<td>Extension for signature hits/library scan files for signature searches (output)</td>
<td>Any string is accepted</td>
<td>.sigout</td>
</tr>

<tr>
<td>-scopf</td>
<td>In either mode, a 'scop classification file' is required as a source of family classification data. A scop classification file contains classification and other data for domains from the scop database. The file is in embl-like format and is generated by scopparse. Domain sequence information can be added to the file by using scopseqs.</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-targetf</td>
<td>A 'scop validation file' contains sequence relatives (hits) for each of a number of different scop families, superfamilies and folds. The file contains the collated results from psiblast searches of a sequence database for the indvidual scop families; hits of unambiguous family assignment are listed under their respective family, otherwise a hit is assigned as relatives to a scop superfamily or fold instead. The scop validation file is generated by seqnr and is in embl-like format.</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-maxhits</td>
<td>The maximum number of hits to be written to the signature hits file.</td>
<td>Any integer value</td>
<td>50</td>
</tr>

<tr>
<td>-overlap</td>
<td>When estimating the statistical significance of scored matches of different types of discriminating element to a sequence, two (or more) p-values are only ever multiplied together if corresponding hit is to the same region of the protein. 'Same' is defined as two hits overlapping by a user-defined number of residues.</td>
<td>Any integer value</td>
<td>10</td>
</tr>

<tr>
<td>-mrgoutpath</td>
<td>Location of signature hits / library scan files for merged results (output)</td>
<td>Any string is accepted</td>
<td>./</td>
</tr>

<tr>
<td>-mrgoutextn</td>
<td>Extension for signature hits / library scan files for merged results (output)</td>
<td>Any string is accepted</td>
<td>.mrgout</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>libscan</b> reads any normal sequence USAs.

<p>



<p>

The format of the scop classification file is described in scopparse

<p>

The format of the signature files is explained in siggen

<p>

The format of the profile and hmm files is explained in documentation for
the propechy application and the hmmer package respectively. 



<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 

        If you wish to include the standard description of the avalable
	report formats, use:
<p>

The output is a standard EMBOSS report file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-rformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  The available format names are: embl,
genbank, gff, pir, swiss, trace, listfile, dbmotif, diffseq, excel,
feattable, motif, regions, seqtable, simple, srs, table, tagseq

<p>

See:
<A href="http://emboss.sf.net/docs/themes/ReportFormats.html">
http://emboss.sf.net/docs/themes/ReportFormats.html</A>
for further information on report formats.

<p>

   -->



<p>



<p>

The format of the signature hits file is described in sigscan



<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.

	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
   -->

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>


<p>

libscan requires a residue substitution matrix.













<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

The score distributions and therefore p-values are calculated only from the
data that libscan is provided.  Therefore, meaningful p-values are only
guaranteed if the sequence set (database search mode) or number of families
(library screeen mode) is sufficiently large. 







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="contacts.html">contacts</a></td><td>Reads coordinate files and writes files of intra-chain residue-residue contact data</td></tr>
<tr><td><a href="fraggle.html">fraggle</a></td><td>Removes fragment sequences from files of hits for scop families</td></tr>
<tr><td><a href="hmmgen.html">hmmgen</a></td><td>Generates a hidden Markov model for each alignment in a directory by using the HMMER package</td></tr>
<tr><td><a href="interface.html">interface</a></td><td>Reads coordinate files and writes files of inter-chain residue-residue contact data</td></tr>
<tr><td><a href="profgen.html">profgen</a></td><td>Generates various profiles for each alignment in a directory</td></tr>
<tr><td><a href="rocplot.html">rocplot</a></td><td>Provides interpretation and graphical display of the performance of discriminating elements (e.g. profiles for protein families). rocplot reads file(s) of hits from discriminator-database search(es), performs ROC analysis on the hits, and writes graphs illustrating the diagnostic performance of the discriminating elements</td></tr>
<tr><td><a href="scopalign.html">scopalign</a></td><td>Generate alignments for families in a scop classification file by using STAMP</td></tr>
<tr><td><a href="scoprep.html">scoprep</a></td><td>Reorder scop classificaiton file so that the representative structure of each family is given first</td></tr>
<tr><td><a href="scopreso.html">scopreso</a></td><td>Removes low resolution domains from a scop classification file</td></tr>
<tr><td><a href="seqalign.html">seqalign</a></td><td>Generate extended alignments for families in a scop families file by using CLUSTALW with seed alignments</td></tr>
<tr><td><a href="seqsearch.html">seqsearch</a></td><td>Generate files of hits for families in a scop classification file by using PSI-BLAST with seed alignments</td></tr>
<tr><td><a href="seqsort.html">seqsort</a></td><td>Reads multiple files of hits and writes (i) a scop families file and (ii) a scop ambiguities file</td></tr>
<tr><td><a href="seqwords.html">seqwords</a></td><td>Generate file of hits for scop families by searching swissprot with keywords</td></tr>
<tr><td><a href="siggen.html">siggen</a></td><td>Generates a sparse protein signature from an alignment and residue contact data</td></tr>
<tr><td><a href="sigplot.html">sigplot</a></td><td>Generates data files of signature performance</td></tr>
<tr><td><a href="sigscan.html">sigscan</a></td><td>Scans a signature against swissprot and writes a signature hits file</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->


<p>

Hidden Markov models can be generated for scop families by using hmmgen
which uses the HMMER package.

<p>

Simple frequency matrices, Gribskov profiles and Hennikoff profiles can be
generated for scop families by using profgen.

<p>

A 'signature file' contains a sparse sequence signature.  The files are
generated by siggen.

<p>

A 'signature hits file' contains the results of a search of a
discriminating element (one of protein signature, hidden Markov model,
simple frequency matrix, Gribskov profile or Hennikoff profile) against a
sequence database.  The files are generated by sigscan and modelscan.

<p>

A 'library scan file' holds the results of a search of a sequence against a
library of discriminating elements for scop families (one directory for
each type of element, e.g. sparse sequence signatures, profiles or HMMs).

<p>

The library scan file contains, for each invividual type of discriminating
element and for the combination of the elements, classification and other
data for each of a user-defined number of top-scoring families
(discriminators) from the scop database.

<p>

A 'scop classification file' contains classification and other data for
domains from the scop database.  The file is in embl-like format and is
generated by scopparse.  Domain sequence information can be added to the
file by using scopseqs.

<p>

A 'scop validation file' contains sequence relatives (hits) for each of a
number of different scop families, superfamilies and folds.  The file
contains the collated results from psiblast searches of a sequence database
for the indvidual scop families; hits of unambiguous family assignment are
listed under their respective family, otherwise a hit is assigned as
relatives to a scop superfamily or fold instead.  The scop validation file
is generated by seqnr and is in embl-like format. 







<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
e.g. one of:
Alan Bleasby (ableasby&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK


Bernd Jagla (bernd&nbsp;&copy;&nbsp;golgi.ski.mskcc.org)
<br>
Cellular Biochemistry and Biophysics Program, Rockefeller
Research Laboratories, Memorial Sloan-Kettering Cancer Center, 1275 York
Avenue, Box 251,New York, NY 10021.

David Martin (dmartin&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>

Gos Micklem (gos&nbsp;&copy;&nbsp;ebi.ac.uk)
<br>
Informatics Division, European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK

Gary Williams (gwilliam&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK


Ian Longden (il&nbsp;&copy;&nbsp;sanger.ac.uk)
<br>
Sanger Institute, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.                      

Mark Faller (current e-mail address unknown)
<br>
while he was with:
<br>
HGMP-RC, Genome Campus, Hinxton, Cambridge CB10 1SB, UK

Michael K. Schuster and Martin Grabner (martin.grabner&nbsp;&copy;&nbsp;univie.ac.at)
<br>
from the Austrian National EMBnet node.


Michael Schmitz (mschmitz&nbsp;&copy;&nbsp;lbl.gov)
<br>
Lawrence Berkeley Labs, USA

Nicolas Tourasse (nicolas.tourasse&nbsp;&copy;&nbsp;biotek.uio.no)
<br>
Biotechnology Center of Oslo

Peter Rice (pmr&nbsp;&copy;&nbsp;ebi.ac.uk)
<br>
Informatics Division, European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK

Richard Durbin (rd&nbsp;&copy;&nbsp;sanger.ac.uk)
<br>
Sanger Institute, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.                      

Richard Bruskiewich (r.bruskiewich@cgiar.org)
<br>
while he was at:
<br>
Sanger Institute, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.                      

Rodrigo Lopez (rls&nbsp;&copy;&nbsp;ebi.ac.uk)
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK

Sinead O'Leary (current e-mail address unknown)
<br>
while she was at:
<br>
HGMP-RC, Genome Campus, Hinxton, Cambridge CB10 1SB, UK


Tim Carver (tcarver&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK

Thomas Laurent (thomas.laurent&nbsp;&copy;&nbsp;uk.lionbioscience.com)
<br>
Lion Bioscience Ltd,
     Compass House,
     80-82 Newmarket Road,
     Cambridge,
     CB5 8DZ,
     UK

Val Curwen (vac&nbsp;&copy;&nbsp;sanger.ac.uk)
<br>
Sanger Institute, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK. 


   -->


Jon Ison (jison&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK



<p>

Ranjeeva Ranasinghe (rranasin&nbsp;&copy;&nbsp;hgmp.mrc.ac.uk)
<br>
HGMP-RC, Genome Campus, Hinxton, Cambridge CB10 1SB, UK











<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (2003) - Jon Ison & Ranjeeva Ranasinghe

<!--
<p>


-->












<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












</BODY>
</HTML>

