<HTML>

<HEAD>
  <TITLE>
  EMBOSS: lindna
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
lindna
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
Draws linear maps of DNA constructs

<H2>
    Description
</H2>

<p><b>lindna</b> draws linear maps of DNA constructs. It uses the graphical shapes ticks, ranges, and blocks to represent genetic markers (e.g, genes and ESTs) and places them according to their position in a DNA fragment. The markers can be organized in different groups. It reads an input file containing the names and positions of the genetic markers, and data for controlling the appearance of the markers (drawing commands).  The output is a plot with with the linear DNA map.  </p>

<H2>
    Usage
</H2>
<b>Here is a sample session with lindna</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>lindna -graph cps </b>
Draws linear maps of DNA constructs
Commands to the lindna drawing program file [inputfile]: <b>data.linp</b>
Draw a ruler [Y]: <b></b>
      Open : Open blocks
    Filled : Filled blocks
   Outline : Black border
Type of blocks [Filled]: <b></b>

Created lindna.ps

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>

<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
Draws linear maps of DNA constructs
Version: EMBOSS:6.2.0

   Standard (Mandatory) qualifiers:
  [-infile]            infile     [inputfile] Commands to the lindna drawing
                                  program file
   -[no]ruler          boolean    [Y] Draw a ruler
   -blocktype          menu       [Filled] Type of blocks (Values: Open (Open
                                  blocks); Filled (Filled blocks); Outline
                                  (Black border))
   -graphout           graph      [$EMBOSS_GRAPHICS value, or x11] Graph type
                                  (ps, hpgl, hp7470, hp7580, meta, cps, x11,
                                  tek, tekt, none, data, xterm, png, gif)

   Additional (Optional) qualifiers:
   -maxgroups          integer    [20] Maximum number of groups (Integer 1 or
                                  more)
   -maxlabels          integer    [10000] Maximum number of labels (Integer 1
                                  or more)
   -intersymbol        selection  [Straight] Type of junctions between blocks
   -intercolour        integer    [1] Colour of junctions between blocks
                                  (enter a colour number) (Integer from 0 to
                                  15)
   -interticks         boolean    [N] Horizontal junctions between ticks
   -gapsize            integer    [500] Interval between ticks in the ruler
                                  (Integer 0 or more)
   -ticklines          boolean    [N] Vertical lines at the ruler's ticks
   -textheight         float      [1.0] Height of text. Enter a number <1.0 or
                                  >1.0 to decrease or increase the size,
                                  respectively (Number 0.000 or more)
   -textlength         float      [1.0] Length of text. Enter a number <1.0 or
                                  >1.0 to decrease or increase the size,
                                  respectively (Number 0.000 or more)
   -margin             float      [1.0] Width of left margin. This is the
                                  region left to the groups where the names of
                                  the groups are displayed. Enter a number
                                  <1.0 or >1.0 to decrease or increase the
                                  size, respectively (Number 0.000 or more)
   -tickheight         float      [1.0] Height of ticks. Enter a number <1.0
                                  or >1.0 to decrease or increase the size,
                                  respectively (Number 0.000 or more)
   -blockheight        float      [1] Height of blocks. Enter a number <1.0 or
                                  >1.0 to decrease or increase the size,
                                  respectively (Number 0.000 or more)
   -rangeheight        float      [1.0] Height of range ends. Enter a number
                                  <1.0 or >1.0 to decrease or increase the
                                  size, respectively (Number 0.000 or more)
   -gapgroup           float      [1.0] Space between groups. Enter a number
                                  <1.0 or >1.0 to decrease or increase the
                                  size, respectively (Number 0.000 or more)
   -postext            float      [1.0] Space between text and ticks, blocks,
                                  and ranges. Enter a number <1.0 or >1.0 to
                                  decrease or increase the size, respectively
                                  (Number 0.000 or more)

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-graphout" associated qualifiers
   -gprompt            boolean    Graph prompting
   -gdesc              string     Graph description
   -gtitle             string     Graph title
   -gsubtitle          string     Graph subtitle
   -gxtitle            string     Graph x axis title
   -gytitle            string     Graph y axis title
   -goutfile           string     Output file for non interactive displays
   -gdirectory         string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-infile]<br>(Parameter 1)</td>
<td>infile</td>
<td>Commands to the lindna drawing program file</td>
<td>Input file</td>
<td>inputfile</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-[no]ruler</td>
<td>boolean</td>
<td>Draw a ruler</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-blocktype</td>
<td>list</td>
<td>Type of blocks</td>
<td><table><tr><td>Open</td> <td><i>(Open blocks)</i></td></tr><tr><td>Filled</td> <td><i>(Filled blocks)</i></td></tr><tr><td>Outline</td> <td><i>(Black border)</i></td></tr></table></td>
<td>Filled</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-graphout</td>
<td>graph</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including ps, hpgl, hp7470, hp7580, meta, cps, x11, tek, tekt, none, data, xterm, png, gif</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>-maxgroups</td>
<td>integer</td>
<td>Maximum number of groups</td>
<td>Integer 1 or more</td>
<td>20</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-maxlabels</td>
<td>integer</td>
<td>Maximum number of labels</td>
<td>Integer 1 or more</td>
<td>10000</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-intersymbol</td>
<td>selection</td>
<td>Type of junctions between blocks</td>
<td>Choose from selection list of values</td>
<td>Straight</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-intercolour</td>
<td>integer</td>
<td>Colour of junctions between blocks (enter a colour number)</td>
<td>Integer from 0 to 15</td>
<td>1</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-interticks</td>
<td>boolean</td>
<td>Horizontal junctions between ticks</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-gapsize</td>
<td>integer</td>
<td>Interval between ticks in the ruler</td>
<td>Integer 0 or more</td>
<td>500</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-ticklines</td>
<td>boolean</td>
<td>Vertical lines at the ruler's ticks</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-textheight</td>
<td>float</td>
<td>Height of text. Enter a number &lt;1.0 or &gt;1.0 to decrease or increase the size, respectively</td>
<td>Number 0.000 or more</td>
<td>1.0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-textlength</td>
<td>float</td>
<td>Length of text. Enter a number &lt;1.0 or &gt;1.0 to decrease or increase the size, respectively</td>
<td>Number 0.000 or more</td>
<td>1.0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-margin</td>
<td>float</td>
<td>Width of left margin. This is the region left to the groups where the names of the groups are displayed. Enter a number &lt;1.0 or &gt;1.0 to decrease or increase the size, respectively</td>
<td>Number 0.000 or more</td>
<td>1.0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-tickheight</td>
<td>float</td>
<td>Height of ticks. Enter a number &lt;1.0 or &gt;1.0 to decrease or increase the size, respectively</td>
<td>Number 0.000 or more</td>
<td>1.0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-blockheight</td>
<td>float</td>
<td>Height of blocks. Enter a number &lt;1.0 or &gt;1.0 to decrease or increase the size, respectively</td>
<td>Number 0.000 or more</td>
<td>1</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-rangeheight</td>
<td>float</td>
<td>Height of range ends. Enter a number &lt;1.0 or &gt;1.0 to decrease or increase the size, respectively</td>
<td>Number 0.000 or more</td>
<td>1.0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-gapgroup</td>
<td>float</td>
<td>Space between groups. Enter a number &lt;1.0 or &gt;1.0 to decrease or increase the size, respectively</td>
<td>Number 0.000 or more</td>
<td>1.0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-postext</td>
<td>float</td>
<td>Space between text and ticks, blocks, and ranges. Enter a number &lt;1.0 or &gt;1.0 to decrease or increase the size, respectively</td>
<td>Number 0.000 or more</td>
<td>1.0</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-graphout" associated graph qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -gprompt</td>
<td>boolean</td>
<td>Graph prompting</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -gdesc</td>
<td>string</td>
<td>Graph description</td>
<td>Any string</td>
<td>Linear DNA map</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -gtitle</td>
<td>string</td>
<td>Graph title</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -gsubtitle</td>
<td>string</td>
<td>Graph subtitle</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -gxtitle</td>
<td>string</td>
<td>Graph x axis title</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -gytitle</td>
<td>string</td>
<td>Graph y axis title</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -goutfile</td>
<td>string</td>
<td>Output file for non interactive displays</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -gdirectory</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>


<H2>
    Input file format
</H2>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: data.linp</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
Start	1001
End	4270

group
exons
label
Block	1011	1362	3	H
ex1
endlabel
label
Tick	1610	8	H
EcoR1
endlabel
label
Block	1647	1815	1	H
endlabel
label
Tick	2459	8	H
BamH1
endlabel
label
Block	4139	4258	3	H
ex2
endlabel
endgroup

group
repeat-regions
label
Range	2541	2812	[	]	5	V
Alu
endlabel
label
Range	3322	3497	&gt;	&lt;	5	V
MER13
endlabel
</pre>
</td></tr></table><p>


<p>

Blank lines in the input file are ignored. 

<p>

The file is organized in several fields separated by one or more space
or TAB characters.

<p>

The first non-blank line gives the start position of the DNA fragment to
be drawn after the keyword 'Start'.  The end position is on the second
line following the keyword 'End'. 

<p>

Then, one or more groups (up to a maximum of 20) of genetic markers are
defined.  Each group begins with a line consisting of the keyword
'group' and ends with a line consisting of the keyword 'endgroup'. 

<p>

If desired, a group name can be given on the very next line. Spaces 
and tabulations are allowed but only the first 20 characters of that name 
(including spaces) will be displayed. 

<p>

Next, the individual markers (up to a maximum of 1000) of the group are
described.  Each marker begins with a line consisting of the keyword
'label' and ends with a line consisting of the keyword 'endlabel'.

<p>

Do not start a new marker (by using a 'label' line) before you have
closed the previous marker (by using an 'endlabel' line).  You must end
a previous marker specification before starting a new one.  i.e.  do not
overlap marker specifications in the input file. 

<p>

For each marker, the line following the 'label' line describes the
marker. Three types of markers can be drawn: ticks, blocks, and ranges. 

<p>

For a tick, the descriptive line begins with the keyword 'Tick',
followed by the position of the tick in the DNA fragment, and a number
from 0 to 15 specifying the color in which the tick will be drawn (a
list of available colors is given below).  If desired, one or more names
can be written above the tick.  In this case an optional letter H or V
should be added to the line.  H or V indicate whether the text will be
horizontal or vertical, respectively.  If no letter is specified, H is
assumed by default.  A list of one or more names can be given on the
next lines with one name per line. 

<p>

For a block, the descriptive line begins with the keyword 'Block',
followed by the start and end positions of the block in the DNA
fragment, and the number of the block's color.  If desired, one or more
names can be written above the block.  Names are given on the following
lines.  An optional letter H or V can be added to the end of the
descriptive line to specify the orientation of the name(s). 

<p>

For a range, the descriptive line begins with the keyword 'Range',
followed by the start and end positions of the range in the DNA
fragment.  Next on the line are the types of boundaries that are to be
drawn.  The user can choose among '&gt;', '&lt;', '[', ']' and '|' for
both start and end boundaries.  If none of these types is wanted for a
given boundary, then a point '.' must be entered.  Next is the color
number.  If desired, one or more names can be written above the range. 
Names are given on the following lines.  An optional letter H or V can
be added to the end of the descriptive line to specify the orientation
of the name(s). 

<p>

For any type of markers, names will be written in the same color as the
marker.  Also, try to avoid long names. 

<p>

The colors defined in the PLPLOT graphics library are:

<p>

0 "BLACK", 1 "RED", 2 "YELLOW", 3 "GREEN", 4 "AQUAMARINE",
5 "PINK", 6 "WHEAT", 7 "GREY", 8 "BROWN", 9 "BLUE", 10 "BLUEVIOLET",
11 "CYAN", 12 "TURQUOISE", 13 "MAGENTA", 14 "SALMON", and 15 "WHITE".

<p>

This format can also be used for the program <a
href="cirdna.html">"cirdna". </a>




<H2>
    Output file format
</H2>

Output is to the specified graphics device.

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>Graphics File: lindna.ps</h3>
<p><img src="lindna.1.lindna.gif" alt="[lindna results]">



<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

<p>Whether to draw a rule and the type of block-fill style must be specified.  There are many other (optional) qualifiers to control the appearance of the plot which will be prompted for if the program is run with <tt>-options</tt> on the command line and.</p>


<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="abiview.html">abiview</a></td>
<td>Display the trace in an ABI sequencer file</td>
</tr>

<tr>
<td><a href="cirdna.html">cirdna</a></td>
<td>Draws circular maps of DNA constructs</td>
</tr>

<tr>
<td><a href="pepnet.html">pepnet</a></td>
<td>Draw a helical net for a protein sequence</td>
</tr>

<tr>
<td><a href="pepwheel.html">pepwheel</a></td>
<td>Draw a helical wheel diagram for a protein sequence</td>
</tr>

<tr>
<td><a href="prettyplot.html">prettyplot</a></td>
<td>Draw a sequence alignment with pretty formatting</td>
</tr>

<tr>
<td><a href="prettyseq.html">prettyseq</a></td>
<td>Write a nucleotide sequence and its translation to file</td>
</tr>

<tr>
<td><a href="remap.html">remap</a></td>
<td>Display restriction enzyme binding sites in a nucleotide sequence</td>
</tr>

<tr>
<td><a href="seealso.html">seealso</a></td>
<td>Finds programs with similar function to a specified program</td>
</tr>

<tr>
<td><a href="showalign.html">showalign</a></td>
<td>Display a multiple sequence alignment in pretty format</td>
</tr>

<tr>
<td><a href="showdb.html">showdb</a></td>
<td>Displays information on configured databases</td>
</tr>

<tr>
<td><a href="showfeat.html">showfeat</a></td>
<td>Display features of a sequence in pretty format</td>
</tr>

<tr>
<td><a href="showpep.html">showpep</a></td>
<td>Displays protein sequences with features in pretty format</td>
</tr>

<tr>
<td><a href="showseq.html">showseq</a></td>
<td>Displays sequences with features in pretty format</td>
</tr>

<tr>
<td><a href="sixpack.html">sixpack</a></td>
<td>Display a DNA sequence with 6-frame translation and ORFs</td>
</tr>

<tr>
<td><a href="textsearch.html">textsearch</a></td>
<td>Search the textual description of sequence(s)</td>
</tr>

</table>

<H2>
    Author(s)
</H2>


Nicolas Tourasse (nicolas.tourasse&nbsp;&copy;&nbsp;biotek.uio.no)
<br>
Biotechnology Center of Oslo

<H2>
    History
</H2>



Written  (2000) - Nicolas Tourasse
<br>
Updated (2002) - Nicolas Tourasse


<H2>
    Target users
</H2>

This program is intended to be used by everyone and everything, from naive users to embedded scripts.


<H2>
    Comments
</H2>
None

</BODY>
</HTML>

