<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: listor
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

listor
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>






<!--END OF HEADER-->

<H2>
    Function
</H2>
Writes a list file of the logical OR of two sets of sequences

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
        Algorithms
  -->

<b>listor</b> reads in two sets of sequences and writes out a list file
(file of file names) that result from the logical union of these two
sets of sequences.  It is a simple way of manipulating and editing lists
or sets of sequences to produce a list file.

<p>

When comparing sequences to see if they are the same between two sets of
sequences, no use is made of the ID name or accession number of the
sequences.  Only the sequences themselves are compared.  The comparison
of the sequences is case-independent. 

<p>


The logical union is an OR operation by default.  Other available
operations are: AND, XOR and NOT. 

<p>

The (default) logical OR of the two sets of sequences is simply the
result of merging the two sets of sequences, (without listing any
shared sequences twice). 

<p>

A logical AND simply lists those sequences that occur in both sets of
sequences. 

<p>

A logical XOR lists those sequences that ONLY occur in the first set or
only occur in the second set - sequences occuring in both sets are
ignored (the opposite of an AND). 

<p>

A logical NOT lists all those sequences in the first set except for
those that also occur in the second set. 



<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

An example of the four types of logical operation performed with the two
sets of sequences comprising these sequences follows.  The two sets of
sequences are in the files named <b>file1</b> and <b>file2</b>.  There
are 2 sequences which are duplicated between these two files ('two' and
'three') and the sequences 'one' and 'four' are unique to the files
'file1' and 'file2' respectively. 

<p>


<b>Here is a sample session with listor</b>
<p>
Write the logical OR of two lists: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>listor ../../data/file1 ../../data/file2 </b>
Writes a list file of the logical OR of two sets of sequences
Output file [file1.list]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>
<p>
<b>Example 2</b>
<p>
Write the logical AND of two lists: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>listor ../../data/file1 ../../data/file2 -operator and </b>
Writes a list file of the logical OR of two sets of sequences
Output file [file1.list]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.2">Go to the output files for this example</a><p><p>
<p>
<b>Example 3</b>
<p>
Write the logical XOR of two lists: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>listor ../../data/file1 ../../data/file2 -operator xor </b>
Writes a list file of the logical OR of two sets of sequences
Output file [file1.list]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.3">Go to the output files for this example</a><p><p>
<p>
<b>Example 4</b>
<p>
Write the logical NOT of two lists: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>listor ../../data/file1 ../../data/file2 -operator not </b>
Writes a list file of the logical OR of two sets of sequences
Output file [file1.list]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.4">Go to the output files for this example</a><p><p>





<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-firstset]          seqset     Sequence set USA
  [-secondset]         seqset     Sequence set USA
  [-outfile]           outfile    The list of sequence names will be written
                                  to this list file

   Additional (Optional) qualifiers:
   -operator           menu       The following logical operators combine the
                                  sequences in the following ways:
                                  OR - gives all that occur in one set or the
                                  other
                                  AND - gives only those which occur in both
                                  sets
                                  XOR - gives those which only occur in one
                                  set or the other, but not in both
                                  NOT - gives those which occur in the first
                                  set except for those that also occur in the
                                  second

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-firstset" associated qualifiers
   -sbegin1             integer    Start of each sequence to be used
   -send1               integer    End of each sequence to be used
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name

   "-secondset" associated qualifiers
   -sbegin2             integer    Start of each sequence to be used
   -send2               integer    End of each sequence to be used
   -sreverse2           boolean    Reverse (if DNA)
   -sask2               boolean    Ask for begin/end/reverse
   -snucleotide2        boolean    Sequence is nucleotide
   -sprotein2           boolean    Sequence is protein
   -slower2             boolean    Make lower case
   -supper2             boolean    Make upper case
   -sformat2            string     Input sequence format
   -sdbname2            string     Database name
   -sid2                string     Entryname
   -ufo2                string     UFO features
   -fformat2            string     Features format
   -fopenfile2          string     Features file name

   "-outfile" associated qualifiers
   -odirectory3         string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-firstset]<br>(Parameter 1)</td>
<td>Sequence set USA</td>
<td>Readable set of sequences</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-secondset]<br>(Parameter 2)</td>
<td>Sequence set USA</td>
<td>Readable set of sequences</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 3)</td>
<td>The list of sequence names will be written to this list file</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.listor</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-operator</td>
<td>The following logical operators combine the sequences in the following ways:
OR - gives all that occur in one set or the other
AND - gives only those which occur in both sets
XOR - gives those which only occur in one set or the other, but not in both
NOT - gives those which occur in the first set except for those that also occur in the second</td>
<td><table><tr><td>O</td> <td><i>(OR - merger of both sets)</i></td></tr><tr><td>A</td> <td><i>(AND - only those in both sets)</i></td></tr><tr><td>X</td> <td><i>(XOR - only those not in both sets)</i></td></tr><tr><td>N</td> <td><i>(NOT - those of the first set that are not in the second)</i></td></tr></table></td>
<td>OR</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

The input sets of sequences can be of any valid USAs.  The program was
written to perform logical operations on list files, but in practice,
wildcarded database entries and file names are also perfectly legal
specifications of the input sequences. 

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: ../../data/file1</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
&gt;one
tagctagcg
&gt;two
tagctagcggctacgt
&gt;three
tagctattttatgctacgtcagtgac
</pre>
</td></tr></table><p>
<p><h3>File: ../../data/file2</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
&gt;two
tagctagcggctacgt
&gt;three
tagctattttatgctacgtcagtgac
&gt;four
gcgcggcgcgcgtgcgtcgttgctggggccc
</pre>
</td></tr></table><p>













<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->

The ouput is simply a list of the USAs (format and sequence
specification) resulting from the required logical union of the two sets
of input sequence. 

<p>

The order that the USAs are written out is not necessarily the same as
the order of either of the input sets of sequences. 

<p>

The results of the four types of logical union follows.  Note that the
duplicated sequences in these two files have been given the same name. 
This is not necessary for the operation of <b>listor</b> as it compares
the sequences themselves, not the ID names of the sequences. 

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: file1.list</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
fasta::../../data/file1:one
fasta::../../data/file1:two
fasta::../../data/file1:three
fasta::../../data/file2:four
</pre>
</td></tr></table><p>

<a name="output.2"></a>
<h3>Output files for usage example 2</h3>
<p><h3>File: file1.list</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
fasta::../../data/file1:two
fasta::../../data/file1:three
</pre>
</td></tr></table><p>

<a name="output.3"></a>
<h3>Output files for usage example 3</h3>
<p><h3>File: file1.list</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
fasta::../../data/file1:one
fasta::../../data/file2:four
</pre>
</td></tr></table><p>

<a name="output.4"></a>
<h3>Output files for usage example 4</h3>
<p><h3>File: file1.list</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
fasta::../../data/file1:one
</pre>
</td></tr></table><p>



<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.
	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

   -->

None.













<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

The program stores all of the input sequences in memory while it is
working out the logical unions of the two sets of sequences.  This means
that it is restricted by the available memory.  Doing logical unions
involving all of the sequences in large databases, such as EMBL, is
probably impractical unless you are lucky enough to have extraordinary
amounts of memory on your machine. 






<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

If you try to do a logical union with all of the sequences in EMBL and
another sequence set, this program will attempt to read all of the EMBL
sequences into memory at once. This will probably not succeed.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="biosed.html">biosed</a></td><td>Replace or delete sequence sections</td></tr>
<tr><td><a href="cutseq.html">cutseq</a></td><td>Removes a specified section from a sequence</td></tr>
<tr><td><a href="degapseq.html">degapseq</a></td><td>Removes gap characters from sequences</td></tr>
<tr><td><a href="descseq.html">descseq</a></td><td>Alter the name or description of a sequence</td></tr>
<tr><td><a href="entret.html">entret</a></td><td>Reads and writes (returns) flatfile entries</td></tr>
<tr><td><a href="extractfeat.html">extractfeat</a></td><td>Extract features from a sequence</td></tr>
<tr><td><a href="extractseq.html">extractseq</a></td><td>Extract regions from a sequence</td></tr>
<tr><td><a href="maskfeat.html">maskfeat</a></td><td>Mask off features of a sequence</td></tr>
<tr><td><a href="maskseq.html">maskseq</a></td><td>Mask off regions of a sequence</td></tr>
<tr><td><a href="newseq.html">newseq</a></td><td>Type in a short new sequence</td></tr>
<tr><td><a href="noreturn.html">noreturn</a></td><td>Removes carriage return from ASCII files</td></tr>
<tr><td><a href="notseq.html">notseq</a></td><td>Excludes a set of sequences and writes out the remaining ones</td></tr>
<tr><td><a href="nthseq.html">nthseq</a></td><td>Writes one sequence from a multiple set of sequences</td></tr>
<tr><td><a href="pasteseq.html">pasteseq</a></td><td>Insert one sequence into another</td></tr>
<tr><td><a href="revseq.html">revseq</a></td><td>Reverse and complement a sequence</td></tr>
<tr><td><a href="seqret.html">seqret</a></td><td>Reads and writes (returns) sequences</td></tr>
<tr><td><a href="seqretsplit.html">seqretsplit</a></td><td>Reads and writes (returns) sequences in individual files</td></tr>
<tr><td><a href="skipseq.html">skipseq</a></td><td>Reads and writes (returns) sequences, skipping the first few</td></tr>
<tr><td><a href="splitter.html">splitter</a></td><td>Split a sequence into (overlapping) smaller sequences</td></tr>
<tr><td><a href="trimest.html">trimest</a></td><td>Trim poly-A tails off EST sequences</td></tr>
<tr><td><a href="trimseq.html">trimseq</a></td><td>Trim ambiguous bits off the ends of sequences</td></tr>
<tr><td><a href="union.html">union</a></td><td>Reads sequence fragments and builds one sequence</td></tr>
<tr><td><a href="vectorstrip.html">vectorstrip</a></td><td>Strips out DNA between a pair of vector sequences</td></tr>
<tr><td><a href="yank.html">yank</a></td><td>Reads a sequence range, appends the full USA to a list file</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

Gary Williams (gwilliam&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK







<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (1 Aug 2001) - Gary Williams

<!--
<p>
[an error occurred while processing this directive]
-->










<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->








</BODY>
</HTML>

