<HTML>
<HEAD>
  <TITLE>
  EMBOSS: listor
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
listor
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
    Function
</H2>
Write a list file of the logical OR of two sets of sequences

<H2>
    Description
</H2>

<b>listor</b> reads in two sets of sequences and writes out a list file
(file of file names) that result from the logical union of these two
sets of sequences.  It is a simple way of manipulating and editing lists
or sets of sequences to produce a list file.

<p>

When comparing sequences to see if they are the same between two sets of
sequences, no use is made of the ID name or accession number of the
sequences.  Only the sequences themselves are compared.  The comparison
of the sequences is case-independent. 

<p>


The logical union is an OR operation by default.  Other available
operations are: AND, XOR and NOT. 

<p>

The (default) logical OR of the two sets of sequences is simply the
result of merging the two sets of sequences, (without listing any
shared sequences twice). 

<p>

A logical AND simply lists those sequences that occur in both sets of
sequences. 

<p>

A logical XOR lists those sequences that ONLY occur in the first set or
only occur in the second set - sequences occuring in both sets are
ignored (the opposite of an AND). 

<p>

A logical NOT lists all those sequences in the first set except for
those that also occur in the second set. 



<H2>
    Usage
</H2>
<b>Here is a sample session with listor</b>
<p>
Write the logical OR of two lists: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>listor ../data/file2 </b>
Write a list file of the logical OR of two sets of sequences
List of USAs output file [file1.list]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>
<p>
<b>Example 2</b>
<p>
Write the logical AND of two lists: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>listor ../data/file2 -operator and </b>
Write a list file of the logical OR of two sets of sequences
List of USAs output file [file1.list]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.2">Go to the output files for this example</a><p><p>
<p>
<b>Example 3</b>
<p>
Write the logical XOR of two lists: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>listor ../data/file2 -operator xor </b>
Write a list file of the logical OR of two sets of sequences
List of USAs output file [file1.list]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.3">Go to the output files for this example</a><p><p>
<p>
<b>Example 4</b>
<p>
Write the logical NOT of two lists: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>listor ../data/file2 -operator not </b>
Write a list file of the logical OR of two sets of sequences
List of USAs output file [file1.list]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.4">Go to the output files for this example</a><p><p>

<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-firstsequences]    seqset     Sequence set filename and optional format,
                                  or reference (input USA)
  [-secondsequences]   seqset     Sequence set filename and optional format,
                                  or reference (input USA)
  [-outfile]           outfile    [*.listor] The list of sequence names will
                                  be written to this list file

   Additional (Optional) qualifiers:
   -operator           menu       [OR] The following logical operators combine
                                  the sequences in the following ways:
                                  OR - gives all that occur in one set or the
                                  other
                                  AND - gives only those which occur in both
                                  sets
                                  XOR - gives those which only occur in one
                                  set or the other, but not in both
                                  NOT - gives those which occur in the first
                                  set except for those that also occur in the
                                  second (Values: O (OR - merger of both
                                  sets); A (AND - only those in both sets); X
                                  (XOR - only those not in both sets); N (NOT
                                  - those of the first set that are not in the
                                  second))

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-firstsequences" associated qualifiers
   -sbegin1            integer    Start of each sequence to be used
   -send1              integer    End of each sequence to be used
   -sreverse1          boolean    Reverse (if DNA)
   -sask1              boolean    Ask for begin/end/reverse
   -snucleotide1       boolean    Sequence is nucleotide
   -sprotein1          boolean    Sequence is protein
   -slower1            boolean    Make lower case
   -supper1            boolean    Make upper case
   -sformat1           string     Input sequence format
   -sdbname1           string     Database name
   -sid1               string     Entryname
   -ufo1               string     UFO features
   -fformat1           string     Features format
   -fopenfile1         string     Features file name

   "-secondsequences" associated qualifiers
   -sbegin2            integer    Start of each sequence to be used
   -send2              integer    End of each sequence to be used
   -sreverse2          boolean    Reverse (if DNA)
   -sask2              boolean    Ask for begin/end/reverse
   -snucleotide2       boolean    Sequence is nucleotide
   -sprotein2          boolean    Sequence is protein
   -slower2            boolean    Make lower case
   -supper2            boolean    Make upper case
   -sformat2           string     Input sequence format
   -sdbname2           string     Database name
   -sid2               string     Entryname
   -ufo2               string     UFO features
   -fformat2           string     Features format
   -fopenfile2         string     Features file name

   "-outfile" associated qualifiers
   -odirectory3        string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-firstsequences]<br>(Parameter 1)</td>
<td>Sequence set filename and optional format, or reference (input USA)</td>
<td>Readable set of sequences</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-secondsequences]<br>(Parameter 2)</td>
<td>Sequence set filename and optional format, or reference (input USA)</td>
<td>Readable set of sequences</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 3)</td>
<td>The list of sequence names will be written to this list file</td>
<td>Output file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-operator</td>
<td>The following logical operators combine the sequences in the following ways:
OR - gives all that occur in one set or the other
AND - gives only those which occur in both sets
XOR - gives those which only occur in one set or the other, but not in both
NOT - gives those which occur in the first set except for those that also occur in the second</td>
<td><table><tr><td>O</td> <td><i>(OR - merger of both sets)</i></td></tr><tr><td>A</td> <td><i>(AND - only those in both sets)</i></td></tr><tr><td>X</td> <td><i>(XOR - only those not in both sets)</i></td></tr><tr><td>N</td> <td><i>(NOT - those of the first set that are not in the second)</i></td></tr></table></td>
<td>OR</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>

The input sets of sequences can be of any valid USAs.  The program was
written to perform logical operations on list files, but in practice,
wildcarded database entries and file names are also perfectly legal
specifications of the input sequences. 

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: file1</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
&gt;one
tagctagcg
&gt;two
tagctagcggctacgt
&gt;three
tagctattttatgctacgtcagtgac
</pre>
</td></tr></table><p>
<p><h3>File: file2</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
&gt;two
tagctagcggctacgt
&gt;three
tagctattttatgctacgtcagtgac
&gt;four
gcgcggcgcgcgtgcgtcgttgctggggccc
</pre>
</td></tr></table><p>

<H2>
    Output file format
</H2>

The ouput is simply a list of the USAs (format and sequence
specification) resulting from the required logical union of the two sets
of input sequence. 

<p>

The order that the USAs are written out is not necessarily the same as
the order of either of the input sets of sequences. 

<p>

The results of the four types of logical union follows.  Note that the
duplicated sequences in these two files have been given the same name. 
This is not necessary for the operation of <b>listor</b> as it compares
the sequences themselves, not the ID names of the sequences. 

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: file1.list</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
fasta::../../data/file1:one
fasta::../../data/file1:two
fasta::../../data/file1:three
fasta::../../data/file2:four
</pre>
</td></tr></table><p>

<a name="output.2"></a>
<h3>Output files for usage example 2</h3>
<p><h3>File: file1.list</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
fasta::../../data/file1:two
fasta::../../data/file1:three
</pre>
</td></tr></table><p>

<a name="output.3"></a>
<h3>Output files for usage example 3</h3>
<p><h3>File: file1.list</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
fasta::../../data/file1:one
fasta::../../data/file2:four
</pre>
</td></tr></table><p>

<a name="output.4"></a>
<h3>Output files for usage example 4</h3>
<p><h3>File: file1.list</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
fasta::../../data/file1:one
</pre>
</td></tr></table><p>


<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

The program stores all of the input sequences in memory while it is
working out the logical unions of the two sets of sequences.  This means
that it is restricted by the available memory.  Doing logical unions
involving all of the sequences in large databases, such as EMBL, is
probably impractical unless you are lucky enough to have extraordinary
amounts of memory on your machine. 

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

If you try to do a logical union with all of the sequences in EMBL and
another sequence set, this program will attempt to read all of the EMBL
sequences into memory at once. This will probably not succeed.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with status 0.

<H2>
    Known bugs
</H2>

None.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr>
<td><a href="biosed.html">biosed</a></td>
<td>Replace or delete sequence sections</td>
</tr>

<tr>
<td><a href="codcopy.html">codcopy</a></td>
<td>Reads and writes a codon usage table</td>
</tr>

<tr>
<td><a href="cutseq.html">cutseq</a></td>
<td>Removes a specified section from a sequence</td>
</tr>

<tr>
<td><a href="degapseq.html">degapseq</a></td>
<td>Removes gap characters from sequences</td>
</tr>

<tr>
<td><a href="descseq.html">descseq</a></td>
<td>Alter the name or description of a sequence</td>
</tr>

<tr>
<td><a href="entret.html">entret</a></td>
<td>Reads and writes (returns) flatfile entries</td>
</tr>

<tr>
<td><a href="extractfeat.html">extractfeat</a></td>
<td>Extract features from a sequence</td>
</tr>

<tr>
<td><a href="extractseq.html">extractseq</a></td>
<td>Extract regions from a sequence</td>
</tr>

<tr>
<td><a href="makenucseq.html">makenucseq</a></td>
<td>Creates random nucleotide sequences</td>
</tr>

<tr>
<td><a href="makeprotseq.html">makeprotseq</a></td>
<td>Creates random protein sequences</td>
</tr>

<tr>
<td><a href="maskfeat.html">maskfeat</a></td>
<td>Mask off features of a sequence</td>
</tr>

<tr>
<td><a href="maskseq.html">maskseq</a></td>
<td>Mask off regions of a sequence</td>
</tr>

<tr>
<td><a href="newseq.html">newseq</a></td>
<td>Type in a short new sequence</td>
</tr>

<tr>
<td><a href="noreturn.html">noreturn</a></td>
<td>Removes carriage return from ASCII files</td>
</tr>

<tr>
<td><a href="notseq.html">notseq</a></td>
<td>Exclude a set of sequences and write out the remaining ones</td>
</tr>

<tr>
<td><a href="nthseq.html">nthseq</a></td>
<td>Writes one sequence from a multiple set of sequences</td>
</tr>

<tr>
<td><a href="pasteseq.html">pasteseq</a></td>
<td>Insert one sequence into another</td>
</tr>

<tr>
<td><a href="revseq.html">revseq</a></td>
<td>Reverse and complement a sequence</td>
</tr>

<tr>
<td><a href="seqret.html">seqret</a></td>
<td>Reads and writes (returns) sequences</td>
</tr>

<tr>
<td><a href="seqretsplit.html">seqretsplit</a></td>
<td>Reads and writes (returns) sequences in individual files</td>
</tr>

<tr>
<td><a href="skipseq.html">skipseq</a></td>
<td>Reads and writes (returns) sequences, skipping first few</td>
</tr>

<tr>
<td><a href="splitter.html">splitter</a></td>
<td>Split a sequence into (overlapping) smaller sequences</td>
</tr>

<tr>
<td><a href="trimest.html">trimest</a></td>
<td>Trim poly-A tails off EST sequences</td>
</tr>

<tr>
<td><a href="trimseq.html">trimseq</a></td>
<td>Trim ambiguous bits off the ends of sequences</td>
</tr>

<tr>
<td><a href="union.html">union</a></td>
<td>Reads sequence fragments and builds one sequence</td>
</tr>

<tr>
<td><a href="vectorstrip.html">vectorstrip</a></td>
<td>Strips out DNA between a pair of vector sequences</td>
</tr>

<tr>
<td><a href="yank.html">yank</a></td>
<td>Reads a sequence range, appends the full USA to a list file</td>
</tr>

</table>


<H2>
    Author(s)
</H2>

Gary Williams (gwilliam&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK


<H2>
    History
</H2>


Written (1 Aug 2001) - Gary Williams


<H2>
    Target users
</H2>

This program is intended to be used by everyone and everything, from naive users to embedded scripts.

<H2>
    Comments
</H2>


None

</BODY>
</HTML>

