<HTML>
<HEAD>
  <TITLE>
  EMBOSS: makeprotseq
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
makeprotseq
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
    Function
</H2>
Creates random protein sequences
<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
  -->

Makes a set of random sequences. The sequence composition can be
defined by reading an optional <a href="pepstats.html">pepstats</a>
output file of protein composition.











<H2>
    Algorithm
</H2>

<!-- 
        Algorithms
	  Pseudocode/Vague description/Handwaving/web links to useful pages
  -->











<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

<b>Here is a sample session with makeprotseq</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>makeprotseq </b>
Creates random protein sequences
Pepstats program output file (optional): <b></b>
Number of sequences created [100]: <b></b>
Length of each sequence [100]: <b></b>
protein output sequence(s) [makeseq.fasta]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.1">Go to the output files for this example</a><p><p>
<p>
<b>Example 2</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>makeprotseq </b>
Creates random protein sequences
Pepstats program output file (optional): <b>../pepstats-keep/laci_ecoli.pepstats</b>
Number of sequences created [100]: <b></b>
Length of each sequence [100]: <b></b>
protein output sequence(s) [makeseq.fasta]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.2">Go to the input files for this example</a><br><a href="#output.2">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers (* if not always prompted):
   -pepstatsfile       infile     This file should be a pepstats output file.
                                  Protein sequences will be created with the
                                  composition in the pepstats output file.
   -amount             integer    [100] Number of sequences created (Integer 1
                                  or more)
   -length             integer    [100] Length of each sequence (Integer 1 or
                                  more)
*  -insert             string     String that is inserted into sequence (Any
                                  string is accepted)
*  -start              integer    [1] Start point of inserted sequence
                                  (Integer 1 or more)
  [-outseq]            seqoutall  [<sequence>.<format>] Protein sequence
                                  set(s) filename and optional format (output
                                  USA)

   Additional (Optional) qualifiers:
   -useinsert          toggle     [N] Do you want to make an insert

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-outseq" associated qualifiers
   -osformat1          string     Output seq format
   -osextension1       string     File name extension
   -osname1            string     Base file name
   -osdirectory1       string     Output directory
   -osdbname1          string     Database name to add
   -ossingle1          boolean    Separate file for each entry
   -oufo1              string     UFO features
   -offormat1          string     Features format
   -ofname1            string     Features file name
   -ofdirectory1       string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-pepstatsfile</td>
<td>This file should be a pepstats output file. Protein sequences will be created with the composition in the pepstats output file.</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-amount</td>
<td>Number of sequences created</td>
<td>Integer 1 or more</td>
<td>100</td>
</tr>

<tr>
<td>-length</td>
<td>Length of each sequence</td>
<td>Integer 1 or more</td>
<td>100</td>
</tr>

<tr>
<td>-insert</td>
<td>String that is inserted into sequence</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-start</td>
<td>Start point of inserted sequence</td>
<td>Integer 1 or more</td>
<td>1</td>
</tr>

<tr>
<td>[-outseq]<br>(Parameter 1)</td>
<td>Protein sequence set(s) filename and optional format (output USA)</td>
<td>Writeable sequence(s)</td>
<td><i>&lt;*&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-useinsert</td>
<td>Do you want to make an insert</td>
<td>Toggle value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>makeprotseq</b> reads any normal sequence USAs.

<p>


<a name="input.2"></a>
<h3>Input files for usage example 2</h3>
<p><h3>File: ../pepstats-keep/laci_ecoli.pepstats</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
PEPSTATS of LACI_ECOLI from 1 to 360

Molecular weight = 38590.05  		Residues = 360   
Average Residue Weight  = 107.195 	Charge   = 1.5   
Isoelectric Point = 6.8820
A280 Molar Extinction Coefficient  = 21620   
A280 Extinction Coefficient 1mg/ml = 0.56      
Improbability of expression in inclusion bodies = 0.660

Residue		Number		Mole%		DayhoffStat
A = Ala		44		12.222 		1.421  	
B = Asx		0		0.000  		0.000  	
C = Cys		3		0.833  		0.287  	
D = Asp		17		4.722  		0.859  	
E = Glu		15		4.167  		0.694  	
F = Phe		4		1.111  		0.309  	
G = Gly		22		6.111  		0.728  	
H = His		7		1.944  		0.972  	
I = Ile		18		5.000  		1.111  	
J = ---		0		0.000  		0.000  	
K = Lys		11		3.056  		0.463  	
L = Leu		41		11.389 		1.539  	
M = Met		10		2.778  		1.634  	
N = Asn		12		3.333  		0.775  	
O = ---		0		0.000  		0.000  	
P = Pro		14		3.889  		0.748  	
Q = Gln		28		7.778  		1.994  	
R = Arg		19		5.278  		1.077  	
S = Ser		32		8.889  		1.270  	
T = Thr		19		5.278  		0.865  	
U = ---		0		0.000  		0.000  	
V = Val		34		9.444  		1.431  	
W = Trp		2		0.556  		0.427  	
X = Xaa		0		0.000  		0.000  	
Y = Tyr		8		2.222  		0.654  	
Z = Glx		0		0.000  		0.000  	

Property	Residues		Number		Mole%
Tiny		(A+C+G+S+T)		120		33.333
Small		(A+B+C+D+G+N+P+S+T+V)	197		54.722
Aliphatic	(A+I+L+V)		137		38.056
Aromatic	(F+H+W+Y)		21		 5.833
Non-polar	(A+C+F+G+I+L+M+P+V+W+Y)	200		55.556
Polar		(D+E+H+K+N+Q+R+S+T+Z)	160		44.444
Charged		(B+D+E+H+K+R+Z)		69		19.167
Basic		(H+K+R)			37		10.278
Acidic		(B+D+E+Z)		32		 8.889

</pre>
</td></tr></table><p>





<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 

        If you wish to include the standard description of the avalable
	report formats, use:
        #include file="inc/reportformats.ihtml"
   -->

<b>makeprotseq</b> 
outputs a graph to the specified graphics device. 
outputs a report format file. The default format is ...


<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: makeseq.fasta</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt;EMBOSS_001
nrvlhpepnprtdniytpawirllygvwvwnrqachnkeerkryppklmmydsqfwcdfe
wadccspkqgwhgnlvkvnrteemfgmqflpqvhpgkkvd
&gt;EMBOSS_002
vtvkddwhkdwwcrpamdylhywlkqrnhytdlslyyttstprwarmadtflapegndcv
qtmywrwvndgdivclecqvcgrfdiymvqdsgqidkghs
&gt;EMBOSS_003
celpniypyweragingdwhetvtvrmhcnnddilwyqmnykppsshavhyivwrrnwcw
nfidqgdgdnrncmnytsnapeqksqlkyghkrqftvvvr
&gt;EMBOSS_004
cshpdepancgridtykhvaydmtdtkaeyhgsspelqslrqkfsnqvwhnraviwwehp
iqdcrlkhselrchskhlseikmpvevtmsdwlmytgyfm
&gt;EMBOSS_005
cpcqytiqygsdlfldsqmpkckkisvelvclvynaqsnlsyfiheaafmvfhpfsllci
meecinwincriaiwppkfvqleidkmiwkvklqcknvcw
&gt;EMBOSS_006
ifawkitieywnktydldkmrklakdfgfppfdpwpihvgccnisnwfmepkfwaqmkcw
mstltiedndwlmlnttefgeqllfywmhwmpcqdewqph
&gt;EMBOSS_007
eiviqqfmvshealkqlgnkwnsqqmhvndriydvkhlvdasnfihhplnkryfrenvns
tacccvhtwsipclfqtidhivnldgaygpwyrvkyshsp
&gt;EMBOSS_008
hmcmmmshfyvgycffvsvrdqrqtceyphvlmhnilftqgralwvrskqpqcadnhqpk
ghwwvawrlqsymkgpqykpqkdwwqgkkffghiwemvrc
&gt;EMBOSS_009
rwydtkfimsgkfaysarqyprqikegeatalrsgpqicpaewiaanypgasfekrqfmd
mqwicgyeprehrwsekymshesvkkgyrhglkngveyqt
&gt;EMBOSS_010
yatmyyygtgmtmkwgepiyvaqflirneqepkvkhahghdascrpkirldlfleerpnp
yksvsfnrfyaggkliigiitydttchkihahdrkeekar
&gt;EMBOSS_011
arakqhrkhmlvknsanwayqswdgkskllvtfghvmenmfkhwrkrsmncvrpinrhfe
gpvmigvkadcqghgqidniqcawpnfedmhamrtvqqvm
&gt;EMBOSS_012
dkihtlrhnhypshmtmtewetvvgvfmipcsmariscpvwgnwphmqglcyppwsgtpn
esqgctnnitfwmnwvspywlfpdlpnfatfmtlgnqrfh
&gt;EMBOSS_013
fkghwqakfphwlsydlkitkyftrqfhmnfiarmngasnfhgrsriawmahlwnqhara
qaflmrlstheyewyfrnqapldqlfnecvlpvmsawhmw
&gt;EMBOSS_014
shhsqgkrnnectscqdqagfdcadnfttqvmekhwtheifnhivgisaittthyvytmt
wcqsfsnmnnlwsragchwevdiagvrmmdicfsvyercf
&gt;EMBOSS_015
hknwdqldarqikalervvclpcenqvidtmvvglifkkkdlfmintwqtwkgisvysci
hqfligfinkgwfpdaysvgvlmfdqdaienadahgdhhl
&gt;EMBOSS_016
gwylwhlgntdaqfeghstgnvhhedkathldfyhedwgchnrtqppvfgmanrwdvakk
seygwmvgfhqcddtlgyfemfhawywgypgcdfhfnrpp
&gt;EMBOSS_017
lfqtfqvwnaymgcgryaclewysaisgcsvmqfgdfkdpifhfptdrlwggiesgdkev


<font color=red>  [Part of this file has been deleted for brevity]</font>

&gt;EMBOSS_084
yscilavytfadkdtptkkapetdlnpdhermykmhdhsrtghwtrnnigcksssyisqv
tciplripvnfrlvvawrcwmdlqddwkphmnmffmrray
&gt;EMBOSS_085
iivhagtkvksgdpaiglirectikwcemwpdsicdtfkivkyfiqskttmqsyyvnilr
lsravaskdsqetdtcgdsmdmshvcpkchqnwhslgsdh
&gt;EMBOSS_086
frqhcqqnsdvfqyrhwktfpfryakchgihvwpkptmeccwlprhqwkktnrvnpvlnf
ytdhhkviceqlelvvsykhskdrvthqmprrnvgegqdw
&gt;EMBOSS_087
ctedmshteihecsmfddlrdwhmkrdcsipnagrkkegwipkldmmrycvylqipktti
fhdvsatcvasfmvhkpndfmlckrciyhykgdtwwrpwt
&gt;EMBOSS_088
ygscslkqtqlgsggkehqqqyleqayewavmisgspnciqsepwmmsmgkpkfwivryn
yekcvmdgyacfqkwegpdhwvsvvhrkmydrggffylge
&gt;EMBOSS_089
hhqkiqscaitwtgiinqqwfmmnrrgwakkqdhqasafvitdnkrswtreyfnkmvism
mkmhcvqtnvhcsaqrrsmhtppyntlitvprmgcpyfka
&gt;EMBOSS_090
akrdfsklilitwvlhficlrfgafaffgmemailrpsdqiywsnmqhnamfkptdhmlv
pelycargfavhpadqgghwwlderteiphyrikdrshde
&gt;EMBOSS_091
pyrehgtmcpckvwqcplydgtadsrgagywpsfhneppqgvseiefeccmaciqqctka
ffqgdamkkkylqnqevptpypdhildnmlqsvyqggqsq
&gt;EMBOSS_092
alhntshnerdmetlwgagcfivemkahfvcyelrkgcrcrrsfnvkfhahaqdwwffmd
snptmgncgeqltgpkqhadilfidinqmmctmchchqye
&gt;EMBOSS_093
aflkktaalkrfrmmtvtmvnqlwcagatifmhilkdlsrsmilgrhcegmwillcfasy
ralvkvdsgtkrwwhkdflelywwdeayrhysetqyasav
&gt;EMBOSS_094
rshvqnvdhpssidmkripdwtcqglqmqinlwllrragirravmykiwdkgemnvelah
iwayhgfrsfikfgqpiqkmrvmyerilptdllyiedcnh
&gt;EMBOSS_095
yenmtynnavemahpeirnwlwrsgfdpmgvhchladnppvscatlvmklwitnfwaaqy
pvwnqfltcyqmprqleflrglvgnchhdrwwqdalitrt
&gt;EMBOSS_096
rdhgrspnwmkftvgvdmtalmycwrdrlewliacfhftmkrgkciewmselicppptnf
vqisnrcmsirgwmymradkeaqcnirwyiyahlkhkwgm
&gt;EMBOSS_097
aygfwfqqhhhfeptkqpeylawqtwtcvkkrcqsytachfeqvthvnkcpsqpiapced
dpsptitsssscnierktgigltcysnhgalmritpkyci
&gt;EMBOSS_098
vaaiydmiaqwvviqtwfvttddqerclwwesrhqyydeyeffftistpvwcyprrrlia
ghaqvikywawvdeslipsefqqapeipdpigpgkvqpeq
&gt;EMBOSS_099
vlmpqdsprpefivnvfntadpscecsckfqlfmcekffncikangvnvdmpchipkaam
tecytlqiawppqnkltypgvhtksfgeewldhmpieqgg
&gt;EMBOSS_100
seffvlpnicymnwnramyvfshwwycfpcsymkdklfwdtrypqihpapklnmedacpg
ytctrpqryqilarqtwyarkhygwvnwtdiascntmdgq
</pre>
</td></tr></table><p>

<a name="output.2"></a>
<h3>Output files for usage example 2</h3>
<p><h3>File: makeseq.fasta</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt;EMBOSS_001
eqngievefrcplriaaeyahqsgatanpaedqllifgvlqsqaerggfgavpdkallvv
nllvlargqkaikegpsaecpvlgttgqtgrdntekssnv
&gt;EMBOSS_002
npnsvlatsvmavqeyslmgtmagsqdriaavgagmapaaprcmydsllavglrvkevln
qaranqnaelilialsvlqnltqvvsarpqlqtqilhtta
&gt;EMBOSS_003
llgersmqmmvdyksrivatvaapndgtveevligamdgemhrraqiyataipnqqrmln
rlhldiltlrdevgrmpaeyevqgadghlkihcdkpappq
&gt;EMBOSS_004
laielvelevkdhvaasinlmlgavagylmttqaevgqqsqqglprdaairqyasmavtr
sdlvcgsiqvgqvtasisplsssravapgqvassmatakr
&gt;EMBOSS_005
vevqaphdmtavsksvqdfrsvhssqavgalsalryqaegqmksivyllfaktrvpssli
flvlsrmseldsysaeqsvaqglhvsginsagsdwhrnla
&gt;EMBOSS_006
ivynsspslmnfgamlglsgdhsygvktleelvrnrhtailvrsqeavsvrhvnldfgva
fqagpsvlelngrgeaalvtlqsgkmmginsevqllndei
&gt;EMBOSS_007
vsaiqdkgaqillssqstrgaraqqftarvqhlvagtsnlyqrvhttegegdaldvepea
allvvaipnqievgvqahvthargltlatealqpglatqr
&gt;EMBOSS_008
iglrssattmniawklaaaqlqqdpllmqingfifiglaqidysnnqasqrdvlleidqh
kimmnyadsqalshkrqmsedglanqtghvkttsnvgnql
&gt;EMBOSS_009
qallaskifqtsklaqyqdaqqdssltvlpygcqieqslrlvaiyyrldtlpvvgddvsv
rqasvtavdqliqaalgaraivqasgtmditggriplaqa
&gt;EMBOSS_010
myafmmakatgpssatvqianydkgsceldlegasiytkilypvqrsidslgvsvvqqre
asaaavrqkmlktssihkisamvpavtssilivdgllslq
&gt;EMBOSS_011
lclgdiqsiggngfqyraymdaalksagggppvtiafleglstndsdagfvpqrifqikv
idpssiaslllqtitdslridlyarrvvlrtygdpnddas
&gt;EMBOSS_012
lssiagdiriarqtspgavmlanpkpvgirlasyqsaleaatraeirdtglarraatpqe
vqdtvaeesavasranpemasvqvgeeklavfpgirdqkt
&gt;EMBOSS_013
ksitaqyhvetngpmlghsasmvpqdktgrviydrrtyarkitdqdslagytgmrqiyql
qwvggdsqatvalaatcrdlqgldskrvvagdpgqyaisn
&gt;EMBOSS_014
atiadkgdrrvlppvqldlkkllylrkaadagvstnptvsleiiaksqlspaptanapsp
aldqkqrgffgaaqyivtnlplswkaqgsvsvvqpmvdvl
&gt;EMBOSS_015
igeavdgllddhslgldaplselvrqnhlasnntgskhshlsksheaadaagthaaacls
tqksstkseskakrlyaqatagflvqlysvrlvlttltts
&gt;EMBOSS_016
inagntgtealydkvtiaatrattvvsypiglkaivlmivirqpqrqavtryednlawsg
avmiasaikidlvvpsiaklrvtlmmakaekvlkitrdre
&gt;EMBOSS_017
gvqpvqaarlasiliqalvsvnmqysqtvqarqktlkhvesltlqplqgaktslptvgln


<font color=red>  [Part of this file has been deleted for brevity]</font>

&gt;EMBOSS_084
mqvsgynaavlvslpepsgyevpvsreltvdsasrivtaqaitaadresivgppqmipda
alsrgqhenrlqsaalmdwnsvsqvlasrifrgvkfddla
&gt;EMBOSS_085
hsailtasngatvelhtsiqlvpssallgaevqslvpvhhasakhqchppsqammarssd
gqqyayphvadlplalklqrlsaiavqgliqfaiasiqli
&gt;EMBOSS_086
kqdtwdqrqlakqmqtasakrkqmygvtthinmehrpslvvasectqashprdpernsrv
maliisnslvdglsaaqmhiasldaatdsecqeptvkqva
&gt;EMBOSS_087
valvgqtavitvlprkvvgdvatssqllqsqeltcssvtniegglrrqmvnagqhqsapi
livnalplalavsptserllsgvsdvsmiastvpnnqeaa
&gt;EMBOSS_088
atalqgsqaqskattsvtqddagvdlavalashpkaerlsdaleargagthrslnsaqae
mlgvpfviallldgmvievtanapntdhgmlditvvastv
&gt;EMBOSS_089
ttqsidalyhpmaiiirdqnvgfeqdknlgsdviqlqyvnsplegcqnpqlmkrgrasqr
ssriwpdaraivqydqqagtpedmrpssparqftlellgy
&gt;EMBOSS_090
yhdltqsgissanngtkiwgqvklvylktglfyssdeqvqsmnqrgdielrvgepvtssn
qvsmlldiklptrlvqktimasvvdalhriaqhhvqqilv
&gt;EMBOSS_091
eacvtkaglrvgpnqlegmltpwvadklkaarqvielerdiaavhlvlvlryvsqqvahy
vkqtllggsgmgdrdvneprmdvisgvegsdqnlqtkqqd
&gt;EMBOSS_092
lgieaqtrvdvglpsniltlvsnvgsytkalmvgqgilqlqqpkraskiliyqvmntvgl
cedagielilqsatqsqtylhsvilsrqrflpgvtliqav
&gt;EMBOSS_093
lvghhalyshdvqsgpnpsafqsnlytypivgihssvspqqssskdtlvtsmssswvyaa
dygpgalptpsqantsltglsaamvlymdiaalaqayala
&gt;EMBOSS_094
qqindeavteaqhvrhqselapvqksdrqsrsnssqqlticdyarmsinvhivreavslt
inymitvqakisvkdqsqssdagmvqigrpvggmsvvvft
&gt;EMBOSS_095
mlrgpaerypvgyievscrngaqqikvrrinilisyvreeaqvyagarsgasaelayydl
eanrqkspvadrqqqslvgqkgairvtildnaqvlsspqp
&gt;EMBOSS_096
qvtiqqeensgvpakavgplggavaqlcslmgillvtkasgdtslivmsplghledrprv
nqsardwgqsqkasarqlvgvldlricamhmytsgignts
&gt;EMBOSS_097
lmtkavqqtitkvqagdrlasyndamalaghqvdqaayvtkvdppinrhleaqqsyelvv
leqrpspqaqalrsvdsptstgpvmariilggqspesals
&gt;EMBOSS_098
nlyhmvrswqaaasqpmvapalvqvcvganvqdtdmmvvmvtkkpsqpenavaddqqssl
tiyqasslmlnnllagiqavlddlqvsdvrhtrtsndrld
&gt;EMBOSS_099
ngredlqedrvvsnrakrawveplvvqwhvdsvgllgvlelhglrkneavgrvihegyls
pllaagdslnqrqrgsamekatpsqtklvnglifeildtt
&gt;EMBOSS_100
plklpgqfslafrnedlgmnkqtanmlkevqmsslggkavpdaeditelegseflvwldt
mpvaqqddadhgldqanaydhiminpraplilqleasltq
</pre>
</td></tr></table><p>





<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.

	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:

        #include file="inc/localfiles.ihtml"
   -->


<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
<ol>

<li>

</ol>

   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr>
<td><a href="biosed.html">biosed</a></td>
<td>Replace or delete sequence sections</td>
</tr>

<tr>
<td><a href="codcopy.html">codcopy</a></td>
<td>Reads and writes a codon usage table</td>
</tr>

<tr>
<td><a href="cutseq.html">cutseq</a></td>
<td>Removes a specified section from a sequence</td>
</tr>

<tr>
<td><a href="degapseq.html">degapseq</a></td>
<td>Removes gap characters from sequences</td>
</tr>

<tr>
<td><a href="descseq.html">descseq</a></td>
<td>Alter the name or description of a sequence</td>
</tr>

<tr>
<td><a href="entret.html">entret</a></td>
<td>Reads and writes (returns) flatfile entries</td>
</tr>

<tr>
<td><a href="extractalign.html">extractalign</a></td>
<td>Extract regions from a sequence alignment</td>
</tr>

<tr>
<td><a href="extractfeat.html">extractfeat</a></td>
<td>Extract features from a sequence</td>
</tr>

<tr>
<td><a href="extractseq.html">extractseq</a></td>
<td>Extract regions from a sequence</td>
</tr>

<tr>
<td><a href="listor.html">listor</a></td>
<td>Write a list file of the logical OR of two sets of sequences</td>
</tr>

<tr>
<td><a href="makenucseq.html">makenucseq</a></td>
<td>Creates random nucleotide sequences</td>
</tr>

<tr>
<td><a href="maskfeat.html">maskfeat</a></td>
<td>Mask off features of a sequence</td>
</tr>

<tr>
<td><a href="maskseq.html">maskseq</a></td>
<td>Mask off regions of a sequence</td>
</tr>

<tr>
<td><a href="newseq.html">newseq</a></td>
<td>Type in a short new sequence</td>
</tr>

<tr>
<td><a href="noreturn.html">noreturn</a></td>
<td>Removes carriage return from ASCII files</td>
</tr>

<tr>
<td><a href="notseq.html">notseq</a></td>
<td>Exclude a set of sequences and write out the remaining ones</td>
</tr>

<tr>
<td><a href="nthseq.html">nthseq</a></td>
<td>Writes one sequence from a multiple set of sequences</td>
</tr>

<tr>
<td><a href="pasteseq.html">pasteseq</a></td>
<td>Insert one sequence into another</td>
</tr>

<tr>
<td><a href="revseq.html">revseq</a></td>
<td>Reverse and complement a sequence</td>
</tr>

<tr>
<td><a href="seqret.html">seqret</a></td>
<td>Reads and writes (returns) sequences</td>
</tr>

<tr>
<td><a href="seqretsplit.html">seqretsplit</a></td>
<td>Reads and writes (returns) sequences in individual files</td>
</tr>

<tr>
<td><a href="skipseq.html">skipseq</a></td>
<td>Reads and writes (returns) sequences, skipping first few</td>
</tr>

<tr>
<td><a href="splitter.html">splitter</a></td>
<td>Split a sequence into (overlapping) smaller sequences</td>
</tr>

<tr>
<td><a href="trimest.html">trimest</a></td>
<td>Trim poly-A tails off EST sequences</td>
</tr>

<tr>
<td><a href="trimseq.html">trimseq</a></td>
<td>Trim ambiguous bits off the ends of sequences</td>
</tr>

<tr>
<td><a href="union.html">union</a></td>
<td>Reads sequence fragments and builds one sequence</td>
</tr>

<tr>
<td><a href="vectorstrip.html">vectorstrip</a></td>
<td>Strips out DNA between a pair of vector sequences</td>
</tr>

<tr>
<td><a href="yank.html">yank</a></td>
<td>Reads a sequence range, appends the full USA to a list file</td>
</tr>

</table>
<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
This application was contributed by 
Henrikki Almusa, Medicel, Helsinki, Finland


<H2>
    History
</H2>
<!--
        Date written and what changes have been made go in this file.
   -->




<H2>
    Target users
</H2>
<!--
        For general users, use this text
   -->
This program is intended to be used by everyone and everything, from naive users to embedded scripts.

<H2>
    Comments
</H2>
<!--
        User/developer/other comments go in this file.
   -->
None


</BODY>
</HTML>

