<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: marscan
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

marscan
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>





<!--END OF HEADER-->

<H2>
    Function
</H2>
Finds MAR/SAR sites in nucleic sequences

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
        Algorithms
  -->


Matrix/scaffold attachment regions (MARs/SARs) are genomic elements
thought to delineate the structural and functional organisation of the
eukaryotic genome.  Originally, MARs and SARs were identified through
their ability to bind to the nuclear matrix or scaffold.  Binding cannot
be assigned to a unique sequence element, but is dispersed over a region
of several hundred base pairs.  These elements are found flanking a gene
or a small cluster of genes and are located often in the vicinity of
cis-regulatory sequences.  This has led to the suggestion that they
contribute to higher order regulation of transcription by defining
boundaries of independently controlled chromatin domains.  There is
indirect evidence to support this notion.  In transgenic experiments
MARs/SARs dampen position effects by shielding the transgene from the
effects of the chromatin structure at the site of integration. 
Furthermore, they may act as boundary elements for enhancers,
restricting their long range effect to only the promoters that are
located in the same chromatin domain. 

<p>

<b>marscan</b> finds a bipartite sequence element that is unique for a
large group of eukaryotic MARs/SARs.  This MAR/SAR recognition signature
(<b>MRS</b>) comprises two individual sequence elements that are <200 bp
apart and may be aligned on positioned nucleosomes in MARs.  The MRS can
be used to correctly predict the position of MARs/SARs in plants and
animals, based on genomic DNA sequence information alone.  Experimental
evidence from the analysis of >300 kb of sequence data from several
eukaryotic organisms show that wherever a MRS is observed in the DNA
sequence, the corresponding genomic fragment is a biochemically
identifiable SAR. 

<p>

The MRS is a bipartite sequence element that consists of two individual
sequences of 8 (AATAAYAA) and 16 bp (AWWRTAANNWWGNNNC) within a 200 bp
distance from each other.  One mismatch is allowed in the 16 bp pattern. 
The patterns can occur on either strand of the DNA with respect to each
other. The 8 bp and the 16 bp sites can overlap.

<p>

Where there are many possible MRS sites caused by many 8 bp and/or 16 bp
pattern sites located within 200 bp of each other, then only the 8 bp
site and the 16 bp site that occur closest to each other are reported. 

<p>

Once a MRS has been reported, no more sites will be looked for within
200 bp of that site.  This reduces (but maybe will not totally
eliminate) over-reporting of the clusters of MRS's that tend to occur
within a MAR/SAR. 


<p>

Not all SARs contain a MRS.  Analysis of >300 kb of genomic sequence
from a variety of eukaryotic organisms shows that the MRS faithfully
predicts 80% of MARs and SARs, suggesting that at least one other type
of MAR/SAR may exist which does not contain a MRS. 

<p>





<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->


Here is a sample session with marscan:

<pre>

% marscan
marscan
Finds MAR/SAR sites in nucleic sequences
Input sequence(s): EMBL:HSHBB
Output file [hshbb.marscan]:

</pre>














<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
  [-outfile]           report     File for output of MAR/SAR recognition
                                  signature (MRS) regions. This contains
                                  details of the MRS in normal GFF format. The
                                  MRS consists of two recognition sites, one
                                  of 8 bp and one of 16 bp on either sense
                                  strand of the genomic DNA, within 200 bp of
                                  each other.

   Optional qualifiers: (none)
   Advanced qualifiers: (none)
   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>File for output of MAR/SAR recognition signature (MRS) regions. This contains details of the MRS in normal GFF format. The MRS consists of two recognition sites, one of 8 bp and one of 16 bp on either sense strand of the genomic DNA, within 200 bp of each other.</td>
<td>Report file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

Normal genomic DNA USA.













<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->

<p>

The output is a standard EMBOSS report file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-rformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  The available format names are: embl,
genbank, gff, pir, swiss, trace, listfile, dbmotif, diffseq, excel,
feattable, motif, regions, seqtable, simple, srs, table, tagseq

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html</A>
for further information on report formats.

<p>


<p>

By default <b>marscan</b> writes a GFF (Gene Feature Format) report file.

<p>

The output from the above example is:

<p>

<hr>
<pre>
##gff-version 2.0
##date 2002-04-04
##Type DNA HSHBB
HSHBB	marscan	misc_signal	2242	2458	0.000	+	.	Sequence "HSHBB.1" ; note "*type MAR/SAR recognition site (MRS)" ; note "*start8bp 2451" ; note "*end8bp 2458" ; note "*start16bp 2242" ; note "*end16bp 2257"
HSHBB	marscan	misc_signal	17654	17730	0.000	+	.	Sequence "HSHBB.2" ; note "*type MAR/SAR recognition site (MRS)" ; note "*start8bp 17723" ; note "*end8bp 17730" ; note "*start16bp 17654" ; note "*end16bp 17669"
HSHBB	marscan	misc_signal	40956	41123	0.000	+	.	Sequence "HSHBB.3" ; note "*type MAR/SAR recognition site (MRS)" ; note "*start8bp 40956" ; note "*end8bp 40963" ; note "*start16bp 41108" ; note "*end16bp 41123"
HSHBB	marscan	misc_signal	42232	42248	0.000	+	.	Sequence "HSHBB.4" ; note "*type MAR/SAR recognition site (MRS)" ; note "*start8bp 42232" ; note "*end8bp 42239" ; note "*start16bp 42233" ; note "*end16bp 42248"
HSHBB	marscan	misc_signal	47834	47966	0.000	+	.	Sequence "HSHBB.5" ; note "*type MAR/SAR recognition site (MRS)" ; note "*start8bp 47959" ; note "*end8bp 47966" ; note "*start16bp 47834" ; note "*end16bp 47849"
HSHBB	marscan	misc_signal	65112	65146	0.000	+	.	Sequence "HSHBB.6" ; note "*type MAR/SAR recognition site (MRS)" ; note "*start8bp 65139" ; note "*end8bp 65146" ; note "*start16bp 65112" ; note "*end16bp 65127"
HSHBB	marscan	misc_signal	65947	65963	0.000	+	.	Sequence "HSHBB.7" ; note "*type MAR/SAR recognition site (MRS)" ; note "*start8bp 65947" ; note "*end8bp 65954" ; note "*start16bp 65948" ; note "*end16bp 65963"
</pre>
<hr>


<p>






<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.
	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

   -->

None.














<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

The method for finding the MAR/SAR sites is described in:
<p>

van Drunen CM., Sewalt RGAB., Oosterling RW., Weisbeek PJ., Smeekens
SCM. and van Driel R.  "A bipartite sequence element associated with
matrix/scaffold attachment regions" Nucleic Acids Research.  1999.  Vol
27, No.  14, pp.  2924-2930

<p>

The original paper on MASs/SARs is:
<p>

Mirkovitch J., Mirault M-E. and Laemmli UK. Cell. 1984. Vol. 39 pp. 223-232.







<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

It does not check that the DNA input seequnce is genomic or not.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="dreg.html">dreg</a></td><td>regular expression search of a nucleotide sequence</td></tr>
<tr><td><a href="fuzznuc.html">fuzznuc</a></td><td>Nucleic acid pattern search</td></tr>
<tr><td><a href="fuzztran.html">fuzztran</a></td><td>Protein pattern search after translation</td></tr>
<tr><td><a href="getorf.html">getorf</a></td><td>Finds and extracts open reading frames (ORFs)</td></tr>
<tr><td><a href="plotorf.html">plotorf</a></td><td>Plot potential open reading frames</td></tr>
<tr><td><a href="showorf.html">showorf</a></td><td>Pretty output of DNA translations</td></tr>
<tr><td><a href="wobble.html">wobble</a></td><td>Wobble base plot</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

This application was written by Gary Williams <a
href="mailto:gwilliam@hgmp.mrc.ac.uk">(gwilliam@hgmp.mrc.ac.uk)</a>













<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (Jan 2001) - Gary Williams.

<p>

Changed output file to standard EMBOSS report format (April 2002) - Peter Rice











<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->








</BODY>
</HTML>

