<HTML>
<HEAD>
  <TITLE>
  EMBOSS: marscan
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
marscan
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
Finds matrix/scaffold recognition (MRS) signatures in DNA sequences

<H2>
    Description
</H2>

<p><b>marscan</b> finds MRS recognition signatures in DNA sequences.  The MRS signature is associated with matrix/scaffold attachment regions (MARs/SARs) which are genomic elements thought to delineate the structural and functional organisation of the eukaryotic genome. The MRS is a bipartite sequence element that consists of an 8bp motif (<tt>AATAAYAA</tt>) and a 16 bp motif (<tt>AWWRTAANNWWGNNNC</tt>) within a 200 bp distance from each other, on either sense strand of the genomic DNA.  <b>marscan</b> reads a DNA sequence and writes a standard EMBOSS report file with details of the MRS signatures identified.</p>



<H2>
Algorithm
</H2>

<p><b>marscan</b> searches for an MRS signature, that being the 8bp sequence (<tt>AATAAYAA</tt>) and the 16 bp sequence (<tt>AWWRTAANNWWGNNNC</tt>) within a 200 bp distance from each other. One mismatch is allowed in the 16 bp pattern. The patterns may occur on the same or different strands and can overlap.</p>
<p>Where there are many suitable 8 bp and/or 16 bp pattern sites located within 200 bp of each other, then only the closest pair of 8 bp / 16 bp sites are reported.</p>
<p>Once an MRS has been reported, no more sites will be looked for within 200 bp of that site. This reduces (but not eliminates entirely) over-reporting of the clusters of MRS's that tend to occur within a MAR/SAR.</par>


<H2>
    Usage
</H2>
<b>Here is a sample session with marscan</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>marscan </b>
Finds matrix/scaffold recognition (MRS) signatures in DNA sequences
Input nucleotide sequence(s): <b>tembl:u01317</b>
Output report [u01317.marscan]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>

<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
Finds matrix/scaffold recognition (MRS) signatures in DNA sequences
Version: EMBOSS:6.3.0

   Standard (Mandatory) qualifiers:
  [-sequence]          seqall     Nucleotide sequence(s) filename and optional
                                  format, or reference (input USA)
  [-outfile]           report     [*.marscan] File for output of MAR/SAR
                                  recognition signature (MRS) regions. This
                                  contains details of the MRS in normal GFF
                                  format. The MRS consists of two recognition
                                  sites, one of 8 bp and one of 16 bp on
                                  either sense strand of the genomic DNA,
                                  within 200 bp of each other.

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1            integer    Start of each sequence to be used
   -send1              integer    End of each sequence to be used
   -sreverse1          boolean    Reverse (if DNA)
   -sask1              boolean    Ask for begin/end/reverse
   -snucleotide1       boolean    Sequence is nucleotide
   -sprotein1          boolean    Sequence is protein
   -slower1            boolean    Make lower case
   -supper1            boolean    Make upper case
   -sformat1           string     Input sequence format
   -sdbname1           string     Database name
   -sid1               string     Entryname
   -ufo1               string     UFO features
   -fformat1           string     Features format
   -fopenfile1         string     Features file name

   "-outfile" associated qualifiers
   -rformat2           string     Report format
   -rname2             string     Base file name
   -rextension2        string     File name extension
   -rdirectory2        string     Output directory
   -raccshow2          boolean    Show accession number in the report
   -rdesshow2          boolean    Show description in the report
   -rscoreshow2        boolean    Show the score in the report
   -rstrandshow2       boolean    Show the nucleotide strand in the report
   -rusashow2          boolean    Show the full USA in the report
   -rmaxall2           integer    Maximum total hits to report
   -rmaxseq2           integer    Maximum hits to report for one sequence

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-sequence]<br>(Parameter 1)</td>
<td>seqall</td>
<td>Nucleotide sequence(s) filename and optional format, or reference (input USA)</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-outfile]<br>(Parameter 2)</td>
<td>report</td>
<td>File for output of MAR/SAR recognition signature (MRS) regions. This contains details of the MRS in normal GFF format. The MRS consists of two recognition sites, one of 8 bp and one of 16 bp on either sense strand of the genomic DNA, within 200 bp of each other.</td>
<td>Report output file</td>
<td><i>&lt;*&gt;</i>.marscan</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-sequence" associated seqall qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sbegin1<br>-sbegin_sequence</td>
<td>integer</td>
<td>Start of each sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -send1<br>-send_sequence</td>
<td>integer</td>
<td>End of each sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sreverse1<br>-sreverse_sequence</td>
<td>boolean</td>
<td>Reverse (if DNA)</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sask1<br>-sask_sequence</td>
<td>boolean</td>
<td>Ask for begin/end/reverse</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -snucleotide1<br>-snucleotide_sequence</td>
<td>boolean</td>
<td>Sequence is nucleotide</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sprotein1<br>-sprotein_sequence</td>
<td>boolean</td>
<td>Sequence is protein</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -slower1<br>-slower_sequence</td>
<td>boolean</td>
<td>Make lower case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -supper1<br>-supper_sequence</td>
<td>boolean</td>
<td>Make upper case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sformat1<br>-sformat_sequence</td>
<td>string</td>
<td>Input sequence format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sdbname1<br>-sdbname_sequence</td>
<td>string</td>
<td>Database name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sid1<br>-sid_sequence</td>
<td>string</td>
<td>Entryname</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ufo1<br>-ufo_sequence</td>
<td>string</td>
<td>UFO features</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fformat1<br>-fformat_sequence</td>
<td>string</td>
<td>Features format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fopenfile1<br>-fopenfile_sequence</td>
<td>string</td>
<td>Features file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-outfile" associated report qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rformat2<br>-rformat_outfile</td>
<td>string</td>
<td>Report format</td>
<td>Any string</td>
<td>gff</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rname2<br>-rname_outfile</td>
<td>string</td>
<td>Base file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rextension2<br>-rextension_outfile</td>
<td>string</td>
<td>File name extension</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rdirectory2<br>-rdirectory_outfile</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -raccshow2<br>-raccshow_outfile</td>
<td>boolean</td>
<td>Show accession number in the report</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rdesshow2<br>-rdesshow_outfile</td>
<td>boolean</td>
<td>Show description in the report</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rscoreshow2<br>-rscoreshow_outfile</td>
<td>boolean</td>
<td>Show the score in the report</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rstrandshow2<br>-rstrandshow_outfile</td>
<td>boolean</td>
<td>Show the nucleotide strand in the report</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rusashow2<br>-rusashow_outfile</td>
<td>boolean</td>
<td>Show the full USA in the report</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rmaxall2<br>-rmaxall_outfile</td>
<td>integer</td>
<td>Maximum total hits to report</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rmaxseq2<br>-rmaxseq_outfile</td>
<td>integer</td>
<td>Maximum hits to report for one sequence</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>


<H2>
    Input file format
</H2>

<b>marscan</b> reads a normal genomic DNA USA.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:u01317' is a sequence entry in the example nucleic acid database 'tembl'
<p>
<p><h3>Database entry: tembl:u01317</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   U01317; SV 1; linear; genomic DNA; STD; HUM; 73308 BP.
XX
AC   U01317; J00093-J00094; J00096; J00158-J00175; J00177-J00179; K01239;
AC   K01890; K02544; M18047; M19067; M24868; M24886;
XX
DT   19-MAR-1994 (Rel. 39, Created)
DT   08-NOV-2008 (Rel. 97, Last updated, Version 36)
XX
DE   Human beta globin region on chromosome 11.
XX
KW   allelic variation; alternate cap site; Alu repeat; beta-1 pseudogene;
KW   beta-globin; delta-globin; epsilon-globin; gamma-globin; gene duplication;
KW   globin; HPFH; KpnI repetitive sequence; polymorphism; promoter mutation;
KW   pseudogene; repetitive sequence; RNA polymerase III; thalassemia.
XX
OS   Homo sapiens (human)
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia;
OC   Eutheria; Euarchontoglires; Primates; Haplorrhini; Catarrhini; Hominidae;
OC   Homo.
XX
RN   [1]
RP   62409-62631, 63482-63610
RX   DOI; 10.1073/pnas.71.6.2300.
RX   PUBMED; 4135409.
RA   Marotta C.A., Forget B.G., Weissman S.M., Verma I.M., McCaffrey R.P.,
RA   Baltimore D.;
RT   "Nucleotide sequences of human globin messenger RNA";
RL   Proc. Natl. Acad. Sci. U.S.A. 71(6):2300-2304(1974).
XX
RN   [2]
RP   63602-63646
RX   DOI; 10.1073/pnas.72.9.3614.
RX   PUBMED; 1059150.
RA   Forget B.G., Marotta C.A., Weissman S.M., Cohen-Solal M.;
RT   "Nucleotide sequences of the 3'-terminal untranslated region of messenger
RT   RNA for human beta globin chain";
RL   Proc. Natl. Acad. Sci. U.S.A. 72(9):3614-3618(1975).
XX
RN   [3]
RP   63593-63626
RX   PUBMED; 788834.
RA   Proudfoot N.J., Brownlee G.G.;
RT   "Nucleotide sequences of globin messenger RNA";
RL   Br. Med. Bull. 32(3):251-256(1976).
XX
RN   [4]
RP   63673-63743
RX   DOI; 10.1016/0092-8674(76)90137-9.
RX   PUBMED; 1035137.
RA   Proudfoot N.J., Longley J.I.;


<font color=red>  [Part of this file has been deleted for brevity]</font>

     aaaggggaga agaatcaaat agacgcaata aaaaatgaca cggggtatca ccactgatcc     70380
     cacagaaata caaactaccg tcagagaata ctataaacac ctctacgcaa ataaactaga     70440
     aaatctagaa gaaatggata aattcctcga cacatacact ctgccaagac taaaccagga     70500
     agaagttgta tctctgaata gaccaataac aggctctgaa attgaggcaa taattaatag     70560
     cttatcaacc aaaaaaagtc cgggaccagt aggattcata gccgaattct accagaggta     70620
     caaggaggag ctggtaccat tccttctgaa actattccaa tcaatagaaa aagagggaat     70680
     cctccctaac tcattttatg aggccagcat catcctgata ccaaagcctg acagagacac     70740
     aacaaaaaaa gagaatgtta caccaatatc cttgatgaac atcgatgcaa aaatcctcaa     70800
     taaaatactg gcaaactgaa tccagcagca catcaaaaag cttatcctcc atgatcaagt     70860
     gggcttcatc cctgccatgc aaggctggtt caacatacga aatcaataaa cataatccag     70920
     catataaaca gaaccaaaga cacaaaccat atgattatct caatagatgc agaaaaggcc     70980
     tttgacaaaa ttcaacaatg cttcatgcta aaaactctca ataaattagg tattgatggg     71040
     acatatctca aaataataag agctatctat gacaaaccca cagccaatat catactgagt     71100
     ggacaaaaac tggaagcatt ccctttgaaa actggcacaa ggcagggatg ccctctctca     71160
     ccactcctat tcaacatagt gttggaagtt ctggccaggg caatcaggca ggagaaggaa     71220
     ataaagggca ttcaattagg aaaagaggaa ggtgaaattg tccctgtttg cagatgacat     71280
     gattgtatat ctagaaaacc ccattgtctc agcccaaaat ctccttaagc tgataagcaa     71340
     cttcagcaaa gtctcaggat ataaaatcag tgtgcaaaaa tcacaagtat tcctatgcac     71400
     caataacaga caaacagaga gccaaatcat gagtgaactc ccattcacaa ttgcttcaaa     71460
     gagaataaaa tacctaggaa tccaacttac aagggatgtg aaggacctct tcaaggagaa     71520
     ctacaaacca ctgctcaatg aaataaaaga ggatacaaac aaatggaaga acattccatg     71580
     cttatgggta ggaagaatca tatcgtgaaa atggtcatac tgcccaaggt aatttataga     71640
     ttcaatgcca tccccatcaa gctaccaatg actttcttca cagaactgga aaaaactact     71700
     ttaaagttca tatggaatca aaaaagagcc cacatcacca aggcaatcct aagccaaaag     71760
     aacaaagctg gaggcatcac gctacctgac ttcaaactat actacaatgc tacggtaacc     71820
     aaaacagcat ggtactggta ccaaaacaga gatctagacc aatggaacag aacagagccc     71880
     tcagaaataa tgccgcatat ctacaactat ccgatctttg acaaacctga gagaaacaag     71940
     caatggggaa aggattccct atttaataaa tggtgctggg aaaactggct agccatatgt     72000
     agaaagctga aactggatcc ttccttacac cttatacaaa aattaattca agatggatta     72060
     aagacttaaa cattagacct aaaaccataa aaaccctaga aaaaaaccta ggcaatacca     72120
     ttcaggacat aggcatgggc aaggacttca tgtctaaaac accaaaacga atggcaacaa     72180
     aagacaaaat ggacaaacgg gatctaatta aactaaagag cttctgcaca gctaaagaaa     72240
     ctaccatcag agtgaacagg caacctacaa aatgggagaa aatttttgca atctactcat     72300
     ctgacaaagg gctaatatcc agaatctaca atgaactcaa acaaatttac aagaaaaaac     72360
     aaacaacccc atcaaaaagt gggcaaagga tatgaacaga cacttctcaa aagaagacat     72420
     ttatgtaatc aaaaaacaca tgaaaaaatg ctcatcatca ctagccatca gagaaatgca     72480
     aatcaaaacc acaatgagat accatctcac accagttaga atggcgatca ttaaaaagtc     72540
     aggaaacaac aggtgctgga gaggatgtgg agaaacagga acaactttta cactgttggt     72600
     gggactgtaa actagttcaa ccattgcgga agtcagtgtg gcaattcctc aggaatctag     72660
     aactagaaat accatttgac ccagccatcc cattactggg tagataccca aaggattata     72720
     aatcatgctg ctataaagac acatgcacac gtatgtttat tgcagcacta ttcacaatag     72780
     caaagacttg gaaccaaccc aaatgtccaa caacgataga ttggattaag aaaatgtggc     72840
     acatatacac catggaatac tatgcagcca taaaaaatga tgagttcatg tcctttgtag     72900
     ggacatggat gaagctggaa actatcattc tcagcaaact atcacaagga caataaacca     72960
     aacaccgcat gttctcactc ataggtggga attgaacaat gagaacacat ggacacatga     73020
     agaggaacat cacactctgg ggactgttat ggggtggggg gcaggggcag ggatagcact     73080
     aggagatata cctaatgcta aatgacgagt taatgggtgc agcacaccaa catggcacat     73140
     gtatacatat ataacaaacc tgccgttgtg cacatgtacc ctaaaacttg aagtataata     73200
     ataaaaaaaa gttatcctat taaaactgat ctcacacatc cgtagagcca ttatcaagtc     73260
     tttctctttg aaacagacag aaatttagtg ttttctcagt cagttaac                  73308
//
</pre>
</td></tr></table><p>

<H2>
    Output file format
</H2>

<p>

The output is a standard EMBOSS report file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-rformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  The available format names are: embl,
genbank, gff, pir, swiss, debug, listfile, dbmotif, diffseq, excel,
feattable, motif, regions, seqtable, simple, srs, table, tagseq

<p>

See:
<A href="http://emboss.sf.net/docs/themes/ReportFormats.html">
http://emboss.sf.net/docs/themes/ReportFormats.html</A>
for further information on report formats.

<p>

<p>

By default <b>marscan</b> writes a GFF (Gene Feature Format) report file.

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: u01317.marscan </h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
##gff-version 3
##sequence-region U01317 1 65963
#!Date 2010-01-15
#!Type DNA
#!Source-version EMBOSS 6.3.0
U01317	marscan	regulatory_region	2242	2458	0.000	+	.	ID="U01317.1";note="*type MAR/SAR recognition site (MRS)";note="*start8bp 2451";note="*end8bp 2458";note="*start16bp 2242";note="*end16bp 2257"
U01317	marscan	regulatory_region	17654	17730	0.000	+	.	ID="U01317.2";note="*type MAR/SAR recognition site (MRS)";note="*start8bp 17723";note="*end8bp 17730";note="*start16bp 17654";note="*end16bp 17669"
U01317	marscan	regulatory_region	40956	41123	0.000	+	.	ID="U01317.3";note="*type MAR/SAR recognition site (MRS)";note="*start8bp 40956";note="*end8bp 40963";note="*start16bp 41108";note="*end16bp 41123"
U01317	marscan	regulatory_region	42232	42248	0.000	+	.	ID="U01317.4";note="*type MAR/SAR recognition site (MRS)";note="*start8bp 42232";note="*end8bp 42239";note="*start16bp 42233";note="*end16bp 42248"
U01317	marscan	regulatory_region	47834	47966	0.000	+	.	ID="U01317.5";note="*type MAR/SAR recognition site (MRS)";note="*start8bp 47959";note="*end8bp 47966";note="*start16bp 47834";note="*end16bp 47849"
U01317	marscan	regulatory_region	65112	65146	0.000	+	.	ID="U01317.6";note="*type MAR/SAR recognition site (MRS)";note="*start8bp 65139";note="*end8bp 65146";note="*start16bp 65112";note="*end16bp 65127"
U01317	marscan	regulatory_region	65947	65963	0.000	+	.	ID="U01317.7";note="*type MAR/SAR recognition site (MRS)";note="*start8bp 65947";note="*end8bp 65954";note="*start16bp 65948";note="*end16bp 65963"
</pre>
</td></tr></table><p>



<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

<p>Matrix/scaffold attachment regions (MARs/SARs) are genomic elements thought to delineate the structural and functional organisation of the eukaryotic genome. Originally, MARs and SARs were identified through their ability to bind to the nuclear matrix or scaffold. Binding cannot be assigned to a unique sequence element, but is dispersed over a region of several hundred base pairs. These elements are found flanking a gene or a small cluster of genes and are located often in the vicinity of cis-regulatory sequences. This has led to the suggestion that they contribute to higher order regulation of transcription by defining boundaries of independently controlled chromatin domains. There is indirect evidence to support this notion. In transgenic experiments MARs/SARs dampen position effects by shielding the transgene from the effects of the chromatin structure at the site of integration. Furthermore, they may act as boundary elements for enhancers, restricting their long range effect to only the promoters that are located in the same chromatin domain.</p>

<p><b>marscan</b> finds a bipartite sequence element that is unique for a large group of eukaryotic MARs/SARs. This MAR/SAR recognition signature (MRS) comprises two individual sequence elements (AATAAYAA and AWWRTAANNWWGNNNC) that are <200 bp apart and may be aligned on positioned nucleosomes in MARs. The MRS signature can be used to correctly predict the position of MARs/SARs in plants and animals, based on genomic DNA sequence information alone. Experimental evidence from the analysis of >300 kb of sequence data from several eukaryotic organisms show that wherever an MRS signature is observed in the DNA sequence, the corresponding genomic fragment is a biochemically identifiable SAR.</p>

<p>It it still not at all clear whether MAR/SARs are real biological phenomena or just experimental artefacts and the problem of how to define and find MARs is still being actively invetsigated. For a recent evaluation of this method and others, see reference 3.  Not all SARs contain a MRS. Analysis of >300 kb of genomic sequence from a variety of eukaryotic organisms shows that the MRS faithfully predicts 80% of MARs and SARs, suggesting that at least one other type of MAR/SAR may exist which does not contain a MRS.</p>

<H2>
    References
</H2>

<ol>

<li>
The method for finding the MAR/SAR sites is described in:
<p>

van Drunen CM., Sewalt RGAB., Oosterling RW., Weisbeek PJ., Smeekens
SCM. and van Driel R.  "A bipartite sequence element associated with
matrix/scaffold attachment regions" Nucleic Acids Research.  1999.  Vol
27, No.  14, pp.  2924-2930

<li>
The original paper on MARs/SARs is:

<p>

Mirkovitch J., Mirault M-E. and Laemmli UK. Cell. 1984. Vol. 39 pp. 223-232.


<li>

A recent evaluation of methods to find MARs/SARs:

<p>

I. Liebich, J. Bode, I. Reuter and E. Wingender
"Evaluation of sequence motifs found in scaffold/matrix-attached regions (S/MARs)"
Nucleic Acids Research 2002, Vol. 30, No. 15 3433-3442

</ol>

<H2>
    Warnings
</H2>

<p><b>marscan</b> does not check whether the DNA input sequence is genomic or not.</p>

<H2>
    Diagnostic Error Messages
</H2>

None.


<H2>
    Exit status
</H2>

It always exits with status 0.

<H2>
    Known bugs
</H2>

None.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="dreg.html">dreg</a></td>
<td>Regular expression search of nucleotide sequence(s)</td>
</tr>

<tr>
<td><a href="fuzznuc.html">fuzznuc</a></td>
<td>Search for patterns in nucleotide sequences</td>
</tr>

<tr>
<td><a href="fuzztran.html">fuzztran</a></td>
<td>Search for patterns in protein sequences (translated)</td>
</tr>

<tr>
<td><a href="getorf.html">getorf</a></td>
<td>Finds and extracts open reading frames (ORFs)</td>
</tr>

<tr>
<td><a href="plotorf.html">plotorf</a></td>
<td>Plot potential open reading frames in a nucleotide sequence</td>
</tr>

<tr>
<td><a href="showorf.html">showorf</a></td>
<td>Display a nucleotide sequence and translation in pretty format</td>
</tr>

<tr>
<td><a href="sixpack.html">sixpack</a></td>
<td>Display a DNA sequence with 6-frame translation and ORFs</td>
</tr>

<tr>
<td><a href="syco.html">syco</a></td>
<td>Draw synonymous codon usage statistic plot for a nucleotide sequence</td>
</tr>

<tr>
<td><a href="tcode.html">tcode</a></td>
<td>Identify protein-coding regions using Fickett TESTCODE statistic</td>
</tr>

<tr>
<td><a href="wobble.html">wobble</a></td>
<td>Plot third base position variability in a nucleotide sequence</td>
</tr>

</table>


<H2>
    Author(s)
</H2>
Gary Williams formerly at:
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK

<p>
Please report all bugs to the EMBOSS bug team (emboss-bug&nbsp;&copy;&nbsp;emboss.open-bio.org) not to the original author.

<H2>
    History
</H2>

Written (Jan 2001) - Gary Williams.

<p>

Changed output file to standard EMBOSS report format (April 2002) - Peter Rice


<H2>
    Target users
</H2>

This program is intended to be used by everyone and everything, from naive users to embedded scripts.

<H2>
    Comments
</H2>

None

</BODY>
</HTML>

