<HTML>

<HEAD>
  <TITLE>
  EMBOSS: maskfeat
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: maskfeat
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program maskfeat
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Mask off features of a sequence


<H2>
    Description
</H2>

<b>maskfeat</b> reads in a sequence with its associated features.  The
features can be found in the annotation of the sequence if it is in a
format such as EMBL or SWISSPROT which includes features in the
annotation, or they may be supplied explicitly in a GFF file by using
the command-lin option '-gff filename'.

<p>

The feature table is then searched for features whose type matches the
specified feature type to be masked.  By default, the type is 'repeat*'
(i.e.  any type whose name starts with 'repeat').  You can specify the
name of any other type of feature, or features that you wish to mask. 
If you wish to specify more than one type of feature, separate their
names with spaces or commas.  The names of the types of feature to be
found may be wild-carded with asterisks '*' to find gruops of feature
types sharing a common part of their names. 

<p>

If you are unsure of the names of feature types in use, please consult
<A HREF="http://www3.ebi.ac.uk/Services/WebFeat/">
http://www3.ebi.ac.uk/Services/WebFeat/</A>
for a list of the EMBL feature types and see Appendix A of the Swissprot
user manual in 
<A HREF="http://www.expasy.ch/txt/userman.txt">
http://www.expasy.ch/txt/userman.txt</A>
for a list of the Swissprot feature types. 

<p>

If any features matching the specified names of feature types are found,
then those regions of the sequence will be masked out by replacing that
part of the sequence by masking characters.  The default masking
characters are 'X' for a protein sequence and 'N' for a nucleic acid
sequence, although you can specify your own masking character, if
required. 

<H2>
    Usage
</H2>


Here is a sample session with maskfeat.
It masks out a feature whose type is "repeat_region" from position
2331 to 2356:

<p>

<pre>
% maskfeat em:AB000360
Mask off features of a sequence.
Output sequence [ab000360.fasta]:    
</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
  [-outseq]            seqout     Output sequence USA

   Optional qualifiers:
   -type               string     By default any feature in the feature table
                                  with a type starting 'repeat' is masked. You
                                  can set this to be any feature type you
                                  wish to mask.
                                  See http://www3.ebi.ac.uk/Services/WebFeat/
                                  for a list of the EMBL feature types and see
                                  Appendix A of the Swissprot user manual in
                                  http://www.expasy.ch/txt/userman.txt for a
                                  list of the Swissprot feature types.
                                  The type may be wildcarded by using '*'.
                                  If you wish to mask more than one type,
                                  separate their names with spaces or commas,
                                  eg:
                                  *UTR repeat*
   -maskchar           string     Character to use when masking.
                                  Default is 'X' for protein sequences, 'N'
                                  for nucleic sequences.

   Advanced qualifiers: (none)

</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outseq]<br>(Parameter 2)</td>
<td>Output sequence USA</td>
<td>Writeable sequence</td>
<td><i>&lt;sequence&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-type</td>
<td>By default any feature in the feature table with a type starting 'repeat' is masked. You can set this to be any feature type you wish to mask.
See http://www3.ebi.ac.uk/Services/WebFeat/ for a list of the EMBL feature types and see Appendix A of the Swissprot user manual in http://www.expasy.ch/txt/userman.txt for a list of the Swissprot feature types.
The type may be wildcarded by using '*'.
If you wish to mask more than one type, separate their names with spaces or commas, eg:
*UTR repeat*</td>
<td>Any string is accepted</td>
<td>repeat*</td>
</tr>

<tr>
<td>-maskchar</td>
<td>Character to use when masking.
Default is 'X' for protein sequences, 'N' for nucleic sequences.</td>
<td>Any string is accepted</td>
<td>'X' for protein, 'N' for nucleic</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>

It reads in a sequence USA.  You can optionally specify a GFF feature
file to use with the command-line option '-gff filename'. 

<H2>
    Output file format
</H2>

A normal, masked, sequence file.

<p>

The output from the above example is:

<p>

<hr>
<pre>
>AB000360 AB000360 Homo sapiens PIGC gene, complete cds.
ggatccctgctgcagagggggtaacggtgtctggcttgccaagcaatatttgttgtggtc
tatcatggaagaaataaagtcgggcaatatgaattttttttttctcaaatttgccggatg
gctgtggtgtttctgactcttagttttctcattgtgaaaaaggaatgattatcttcttcg
atcctctcaagagtttccttgttttgagtagattgatagctctttaaaggatgctaagct
cagctaatggaagaagagtctagtttctttgaggctttgattttggttaaactatagagc
tcatacctttctgtatggtgcagcttactattgtctttggattggtaacttaaaaaatac
aaataacatgcctttgagaaccaataaaaactatggatattatccctataaatttacaca
aatccagatataagcatgcaatgtgatatacctaagggatatgtgaaccactgagttaag
aactgctttagagggagatacaatgtgagacacaggctttgggataagactttggtttga
atcctggctctgctctgttaccttagggcaaagttacttaagcatcttgaatctcagctt
ttttaccaaagcaggactaatactaacttacaaggtggtgaggattaagtgaaagaagat
acataaggcacttagcacatagtaggtactcaataagcgatagctaacagatgtctatta
ttattcaaggaattataattttcaaatctgaaatgcagttttaatgtcccataaggtgac
taccacatacatttttctcagacttttagtaaactgagttgatttgactttatctcagta
ctactcttgacctttcacaactttcgtaggttcacagtctctctttttctaggaacttgg
ctgtgttgtcctgcctcagagacaaattcatctattgtaggcctagcccctgcctttgaa
aacaaggaaaggttggtagaacatcaacacagcatggaatttccagggaggtctcatttc
aaaacttcataaagaacaagaaccacctggacttctgtgagggcgatgattaaactggcc
tgagtttgaatgaaaggataatgtatgctcaacctgtgactaacaccaaggaggtcaagt
ggcagaaggtcttgtatgagcgacagccctttcctgataactatgtggaccggcgattcc
tggaagagctccggaaaaacatccatgctcggaaataccaatattgggctgtggtatttg
agtccagtgtggtgatccagcagctgtgcagtgtttgtgtttttgtggttatctggtggt
atatggatgagggtcttctggccccccattggcttttagggactggcctggcttcttcac
tgattgggtatgttttgtttgatctcattgatggaggtgaagggcggaagaagagtgggc
agacccggtgggctgacctgaagagtgccctagtcttcattactttcacttatgggtttt
caccagtgctgaagacccttacagagtctgtcagcactgacaccatctatgccatgtcag
tcttcatgctgttaggccatctcatcttttttgactatggtgccaatgctgccattgtat
ccagcacactatccttgaacatggccatctttgcttctgtatgcttggcatcacgtcttc
cccggtccctgcatgccttcatcatggtgacatttgccattcagatttttgccctgtggc
ccatgttgcagaagaaactaaaggcatgtactccccggagctatgtgggggtcacactgc
tttttgcattttcagccgtgggaggcctactgtccattagtgctgtgggagccgtactct
ttgcccttctgctgatgtctatctcatgtctgtgttcattctacctcattcgcttgcagc
tttttaaagaaaacattcatgggccttgggatgaagctgaaatcaaggaagacttgtcca
ggttcctcagttaaattaggacatccattacattattaaagcaagctgatagattagcct
cctaactagtatagaacttaaagacagagttccattctggaagcagcatgtcattgtggt
aagagaatagagatcaaaaccaaaaaaaatgaaccaaaggcttgggtggtgagggtgctt
atcctttctgttattttgtagatgaaaaaactttctggggacctcttgaattacatgctg
taacatatgaagtgatgtggtttctattaaaaaaataacacatccatcaagttgtctcat
gatttttccataaacaggaggcagacagaggggcatgaagagtgaagtaaNNNNNNNNNN
NNNNNNNNNNNNNNNNaaagtcacttctttctacccttttcaatgtgctaatgctctttt
atttatctagggctcaaatcttagaacacagggtgctatgctcagttttgttgcccaaga
tcacagaattggttacttaaccttgactcagagtttctaccttgttcttagggaagcata
tcacaactaattgcaaagcagagtgtgatgtgtcacaataagcagaatgctagggggaat
tc
</pre>
<hr>


<H2>
    Data files
</H2>

None.


<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="coderet.html">coderet</a></td><td>Extract CDS, mRNA and translations from feature tables</td></tr>
<tr><td><a href="cutseq.html">cutseq</a></td><td>Removes a specified section from a sequence</td></tr>
<tr><td><a href="degapseq.html">degapseq</a></td><td>Removes gap characters from sequences</td></tr>
<tr><td><a href="descseq.html">descseq</a></td><td>Alter the name or description of a sequence</td></tr>
<tr><td><a href="entret.html">entret</a></td><td>Reads and writes (returns) flatfile entries</td></tr>
<tr><td><a href="extractseq.html">extractseq</a></td><td>Extract regions from a sequence</td></tr>
<tr><td><a href="infoseq.html">infoseq</a></td><td>Displays some simple information about sequences</td></tr>
<tr><td><a href="maskseq.html">maskseq</a></td><td>Mask off regions of a sequence</td></tr>
<tr><td><a href="newseq.html">newseq</a></td><td>Type in a short new sequence</td></tr>
<tr><td><a href="noreturn.html">noreturn</a></td><td>Removes carriage return from ASCII files</td></tr>
<tr><td><a href="notseq.html">notseq</a></td><td>Excludes a set of sequences and writes out the remaining ones</td></tr>
<tr><td><a href="nthseq.html">nthseq</a></td><td>Writes one sequence from a multiple set of sequences</td></tr>
<tr><td><a href="pasteseq.html">pasteseq</a></td><td>Insert one sequence into another</td></tr>
<tr><td><a href="revseq.html">revseq</a></td><td>Reverse and complement a sequence</td></tr>
<tr><td><a href="seqret.html">seqret</a></td><td>Reads and writes (returns) sequences</td></tr>
<tr><td><a href="seqretall.html">seqretall</a></td><td>Reads and writes (returns) a set of sequences one at a time</td></tr>
<tr><td><a href="seqretset.html">seqretset</a></td><td>Reads and writes (returns) a set of sequences all at once</td></tr>
<tr><td><a href="seqretsplit.html">seqretsplit</a></td><td>Reads and writes (returns) sequences in individual files</td></tr>
<tr><td><a href="showfeat.html">showfeat</a></td><td>Show features of a sequence</td></tr>
<tr><td><a href="splitter.html">splitter</a></td><td>Split a sequence into (overlapping) smaller sequences</td></tr>
<tr><td><a href="trimseq.html">trimseq</a></td><td>Trim ambiguous bits off the ends of sequences</td></tr>
<tr><td><a href="vectorstrip.html">vectorstrip</a></td><td>Strips out DNA between a pair of vector sequences</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->

<p>

<b>maskseq</b> simply masks a user-specified set of regions, without
using annotated features. 

<H2>
    Author(s)
</H2>


This application was written by Gary Williams <a
href="mailto:gwilliam@hgmp.mrc.ac.uk">(gwilliam@hgmp.mrc.ac.uk)</a>


<H2>
    History
</H2>

Written (2000) - Gary Williams

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

