<HTML>

<HEAD>
  <TITLE>
  EMBOSS: merger
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: merger
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program merger
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Merge two overlapping nucleic acid sequences


<H2>
    Description
</H2>


This joins two overlapping nucleic acid sequences into one merged
sequence. 

<p>
It uses a global alignment algorithm (Needleman & Wunsch) to optimally
align the sequences and then it creates the merged sequence from the
alignment.  When there is a mismatch in the alignment between the two
sequences, the correct base to include in the resulting sequence is
chosen by using the base from the sequence which has the best local
sequence quality score.  The following heuristic is used to find the
sequence quality score:

<p>
If one of the bases is a 'N', then the other sequence's base is used,
else:

<p>
A window size around the disputed base is used to find the local quality
score.  This window size is increased from 5, to 10 to 20 bases or until
there is a clear decision on the best choice.  If there is no best
choice after using a window of 20, then the base in the first sequence
is used. 

<p>
To calculate the quality of a window of a sequence around a base:

<ul>
<li>quality = sequence value/length under window either side of the base
<li>sequence value = sum of points in that window
<li>unambiguous bases (ACGTU) score 2 points
<li>ambiguous bases (MRWSYKVHDB) score 1 point
<li>Ns score 0 points
<li>off end of the sequence scores 0 points
</ul>

<p>
N.B.  This heavily discriminates against the iffy bits at the end of
sequence reads. 

<p>
This program was originally written to aid in the reconstruction of mRNA
sequences which had been sequenced from both ends as a 5' and 3' EST
(cDNA). eg. joining two reads produced by primer walking sequencing.

<p>
Care should be taken to reverse one of the sequences (e.g.  using the
qualifier '-sreverse2') if this is required to get them both in the
correct orientation. 

<p>
Because it uses a Needleman & Wunsch alignment the required memory may
be greater than the available memory when attempting to merge large
(cosmid-sized or greater) sequences. 

<p>
The gap open and gap extension penalties have been set at a higher level
than is usual (50 and 5).  This was experimentally determined to give
the best results with a set of poor quality EST test sequences. 

<H2>
    Usage
</H2>


Here is a sample session with merger.


<pre>

% merger
Input sequence: embl:eclacy
Second sequence: embl:eclaca
Output sequence [eclacy.fasta]: 
Output file [stdout]: 

Global: ECLACY vs ECLACA
Score: 795.00

ECLACY          1     ttccagctgagcgccggtcgctaccattaccagttggtctggtgt 45   
                                                                   
ECLACA                                                              


.................... until ......................


ECLACY          1306  agcggccccggcccgctttccctgctgcgtcgtcaggtgaatgaa 1350 
                                                          |||||||||
ECLACA          1                                         gtgaatgaa 9    

ECLACY          1351  gtcgcttaagcaatcaatgtcggatgcggcgcgacgcttatccga 1395 
                      |||||||||||||||||||||||||||||||||||||||||||||
ECLACA          10    gtcgcttaagcaatcaatgtcggatgcggcgcgacgcttatccga 54   

ECLACY          1396  ccaacatatcataacggagtgatcgcattgaacatgccaatgacc 1440 
                      |||||||||||||||||||||||||||||||||||||||||||||
ECLACA          55    ccaacatatcataacggagtgatcgcattgaacatgccaatgacc 99   

ECLACY          1441  gaaagaataagagcaggcaagctatttaccgatatgtgcgaaggc 1485 
                      |||||||||||||||||||||||||||||||||||||||||||||
ECLACA          100   gaaagaataagagcaggcaagctatttaccgatatgtgcgaaggc 144  

ECLACY          1486  ttaccggaaaaaaga                               1500 
                      |||||||||||||||                              
ECLACA          145   ttaccggaaaaaagacttcgtgggaaaacgttaatgtatgagttt 189  

ECLACY                                                              
                                                                   
ECLACA          190   aatcactcgcatccatcagaagttgaaaaaagagaaagcctgatt 234  

ECLACY                                                              
                                                                   
</pre>

<p>
Typically, one of the sequences will need to be reverse-complemented to put it
into the correct orientation to make it join. For example:
<p>
<pre>
% merger file1.seq file2.seq -sreverse2 -out merged.seq
</pre>

<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-seqa]              sequence   Sequence USA
  [-seqb]              sequence   Sequence USA
  [-outseq]            seqout     Output sequence USA
   -report             outfile    Output alignment and explanation

   Optional qualifiers:
   -datafile           matrixf    Matrix file
   -gapopen            float      Gap opening penalty
   -gapextend          float      Gap extension penalty

   Advanced qualifiers: (none)

</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-seqa]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-seqb]<br>(Parameter 2)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outseq]<br>(Parameter 3)</td>
<td>Output sequence USA</td>
<td>Writeable sequence</td>
<td><i>&lt;sequence&gt;</i>.<i>format</i></td>
</tr>

<tr>
<td>-report</td>
<td>Output alignment and explanation</td>
<td>Output file</td>
<td>stdout</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-datafile</td>
<td>Matrix file</td>
<td>Comparison matrix file in EMBOSS data path</td>
<td>EBLOSUM62 for protein<br>EDNAMAT for DNA</td>
</tr>

<tr>
<td>-gapopen</td>
<td>Gap opening penalty</td>
<td>Number from 1.000 to 100.000</td>
<td>50.0</td>
</tr>

<tr>
<td>-gapextend</td>
<td>Gap extension penalty</td>
<td>Number from 0.100 to 10.000</td>
<td>5</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>


<H2>
    Output file format
</H2>


The output <i>sequence file</i> contains the joined sequence, by default in
FASTA format.  Where there is a mismatch in the alignment, the chosen
base is written to the output sequence in uppercase. 

<p>
The output <i>report file</i> contains descriptions of the positions where
there is a mismatch in the alignment and shows the alignment.  Where
there is a mismatch in the alignment, the chosen base is written in
uppercase. 

<p>
An example report file showing mismatches follows:

<pre>

# j1 position base           j2 position base        Using
        12      'G'             1       'a'             'G'
        13      'C'             2       'a'             'C'
        14      'G'             3       'a'             'G'
        16      'T'             5       'a'             'T'
        20      'T'             9       'g'             'T'
        23      'G'             12      'c'             'G'
        24      'C'             13      'g'             'C'
        41      'G'             30      't'             'G'
        57      't'             46      'C'             'C'
Global: j1 vs j2
Score: 188.00

j1              1        gtatggtcgatGCGaTgcgTatGCtGacgttAgcggcggcGatat 45      
                                       | ||| ||  | ||||| |||||||| ||||
j2              1                   aaaaagcggatcgtnacgttngcggcggctatat 34      

j1              46       attgcgagctatgatgctnatcgtngc                   72      
                         ||||||||||| |||||| ||||| ||                  
j2              35       attgcgagctaCgatgctGatcgtAgcgtacgttgaaagctacta 79      

j1                                                                     
                                                 
j2              80       ctatgctgtgctagctgacgtagc                      103     

</pre>

<H2>
    Data files
</H2>


<H2>
    Notes
</H2>


<H2>
    References
</H2>


<H2>
    Warnings
</H2>


<H2>
    Diagnostic Error Messages
</H2>


<H2>
    Exit status
</H2>


<H2>
    Known bugs
</H2>


<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="cutseq.html">cutseq</a></td><td>Removes a specified section from a sequence</td></tr>
<tr><td><a href="descseq.html">descseq</a></td><td>Alter the name or description of a sequence</td></tr>
<tr><td><a href="est2genome.html">est2genome</a></td><td>Align EST and genomic DNA sequences</td></tr>
<tr><td><a href="extractseq.html">extractseq</a></td><td>Extract regions from a sequence</td></tr>
<tr><td><a href="maskfeat.html">maskfeat</a></td><td>Mask off features of a sequence</td></tr>
<tr><td><a href="maskseq.html">maskseq</a></td><td>Mask off regions of a sequence</td></tr>
<tr><td><a href="megamerger.html">megamerger</a></td><td>Merge two large overlapping nucleic acid sequences</td></tr>
<tr><td><a href="newseq.html">newseq</a></td><td>Type in a short new sequence</td></tr>
<tr><td><a href="noreturn.html">noreturn</a></td><td>Removes carriage return from ASCII files</td></tr>
<tr><td><a href="nthseq.html">nthseq</a></td><td>Writes one sequence from a multiple set of sequences</td></tr>
<tr><td><a href="pasteseq.html">pasteseq</a></td><td>Insert one sequence into another</td></tr>
<tr><td><a href="revseq.html">revseq</a></td><td>Reverse and complement a sequence</td></tr>
<tr><td><a href="splitter.html">splitter</a></td><td>Split a sequence into (overlapping) smaller sequences</td></tr>
<tr><td><a href="trimseq.html">trimseq</a></td><td>Trim ambiguous bits off the ends of sequences</td></tr>
<tr><td><a href="vectorstrip.html">vectorstrip</a></td><td>Strips out DNA between a pair of vector sequences</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Gary Williams <a
href="mailto:gwilliam@hgmp.mrc.ac.uk">(gwilliam@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>


<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

