<HTML>

<HEAD>
  <TITLE>
  EMBOSS: merger
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">




<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

merger
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>







<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Merge two overlapping nucleic acid sequences


<H2>
    Description
</H2>


This joins two overlapping nucleic acid sequences into one merged
sequence. 

<p>
It uses a global alignment algorithm (Needleman & Wunsch) to optimally
align the sequences and then it creates the merged sequence from the
alignment.  When there is a mismatch in the alignment between the two
sequences, the correct base to include in the resulting sequence is
chosen by using the base from the sequence which has the best local
sequence quality score.  The following heuristic is used to find the
sequence quality score:

<p>
If one of the bases is a 'N', then the other sequence's base is used,
else:

<p>
A window size around the disputed base is used to find the local quality
score.  This window size is increased from 5, to 10 to 20 bases or until
there is a clear decision on the best choice.  If there is no best
choice after using a window of 20, then the base in the first sequence
is used. 

<p>
To calculate the quality of a window of a sequence around a base:

<ul>
<li>quality = sequence value/length under window either side of the base
<li>sequence value = sum of points in that window
<li>unambiguous bases (ACGTU) score 2 points
<li>ambiguous bases (MRWSYKVHDB) score 1 point
<li>Ns score 0 points
<li>off end of the sequence scores 0 points
</ul>

<p>
N.B.  This heavily discriminates against the iffy bits at the end of
sequence reads. 

<p>
This program was originally written to aid in the reconstruction of mRNA
sequences which had been sequenced from both ends as a 5' and 3' EST
(cDNA). eg. joining two reads produced by primer walking sequencing.

<p>
Care should be taken to reverse one of the sequences (e.g.  using the
qualifier '-sreverse2') if this is required to get them both in the
correct orientation. 

<p>
Because it uses a Needleman & Wunsch alignment the required memory may
be greater than the available memory when attempting to merge large
(cosmid-sized or greater) sequences. 

<p>
The gap open and gap extension penalties have been set at a higher level
than is usual (50 and 5).  This was experimentally determined to give
the best results with a set of poor quality EST test sequences. 

<H2>
    Usage
</H2>


Here is a sample session with merger.


<pre>

% merger
Merge two overlapping nucleic acid sequences
Input sequence: tembl:eclacy
Second sequence: tembl:eclaca
Output sequence [eclacy.fasta]: 
Output alignment [eclacy.out2]:
                                                                   
</pre>

<p>

Typically, one of the sequences will need to be reverse-complemented to put it
into the correct orientation to make it join. For example:

<p>

<pre>
% merger file1.seq file2.seq -sreverse2 -outseq merged.seq -outfile stdout
</pre>

<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-seqa]              sequence   Sequence USA
  [-seqb]              sequence   Sequence USA
  [-outseq]            seqout     Output sequence USA
  [-outfile]           align      Output alignment and explanation

   Optional qualifiers:
   -datafile           matrixf    This is the scoring matrix file used when
                                  comparing sequences. By default it is the
                                  file 'EBLOSUM62' (for proteins) or the file
                                  'EDNAFULL' (for nucleic sequences). These
                                  files are found in the 'data' directory of
                                  the EMBOSS installation.
   -gapopen            float      Gap opening penalty
   -gapextend          float      Gap extension penalty

   Advanced qualifiers: (none)
   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-seqa]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-seqb]<br>(Parameter 2)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outseq]<br>(Parameter 3)</td>
<td>Output sequence USA</td>
<td>Writeable sequence</td>
<td><i>&lt;sequence&gt;</i>.<i>format</i></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 4)</td>
<td>Output alignment and explanation</td>
<td>Alignment file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-datafile</td>
<td>This is the scoring matrix file used when comparing sequences. By default it is the file 'EBLOSUM62' (for proteins) or the file 'EDNAFULL' (for nucleic sequences). These files are found in the 'data' directory of the EMBOSS installation.</td>
<td>Comparison matrix file in EMBOSS data path</td>
<td>EBLOSUM62 for protein<br>EDNAFULL for DNA</td>
</tr>

<tr>
<td>-gapopen</td>
<td>Gap opening penalty</td>
<td>Number from 1.000 to 100.000</td>
<td>50.0</td>
</tr>

<tr>
<td>-gapextend</td>
<td>Gap extension penalty</td>
<td>Number from 0.100 to 10.000</td>
<td>5</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>


<H2>
    Output file format
</H2>


The output <i>sequence file</i> contains the joined sequence, by default in
FASTA format.  Where there is a mismatch in the alignment, the chosen
base is written to the output sequence in uppercase. 

<p>

<p>

The output is a standard EMBOSS alignment file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-aformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  Some of the alignment formats can cope
with an unlimited number of sequences, while others are only for pairs
of sequences.  

<p>

The available multiple alignment format names are: unknown, multiple,
simple, fasta, msf, trace, srs

<p>

The available pairwise alignment format names are: pair, markx0, markx1,
markx2, markx3, markx10, srspair, score

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/AlignFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/AlignFormats.html</A>
for further information on alignment formats.

<p>

 
<p>

The output <i>report file</i> contains descriptions of the positions where
there is a mismatch in the alignment and shows the alignment.  Where
there is a mismatch in the alignment, the chosen base is written in
uppercase. 


<p>

<pre>
########################################
# Program:  merger
# Rundate:  Mon May 20 16:17:43 2002
# Report_file: stdout
########################################
#=======================================
#
# Aligned_sequences: 2
# 1: ECLACY
# 2: ECLACA
# Matrix: EDNAFULL
# Gap_penalty: 50.0
# Extend_penalty: 5.0
#
# Length: 3173
# Identity:     159/3173 ( 5.0%)
# Similarity:   159/3173 ( 5.0%)
# Gaps:        3014/3173 (95.0%)
# Score: 795.0
#
#
#=======================================

ECLACY             1 ttccagctgagcgccggtcgctaccattaccagttggtctggtgtcaaaa     50

ECLACA             1                                                         0
 

.................... until ......................


ECLACY          1301 cgcttagcggccccggcccgctttccctgctgcgtcgtcaggtgaatgaa   1350
                                                              |||||||||
ECLACA             1                                          gtgaatgaa      9

ECLACY          1351 gtcgcttaagcaatcaatgtcggatgcggcgcgacgcttatccgaccaac   1400
                     ||||||||||||||||||||||||||||||||||||||||||||||||||
ECLACA            10 gtcgcttaagcaatcaatgtcggatgcggcgcgacgcttatccgaccaac     59

ECLACY          1401 atatcataacggagtgatcgcattgaacatgccaatgaccgaaagaataa   1450
                     ||||||||||||||||||||||||||||||||||||||||||||||||||
ECLACA            60 atatcataacggagtgatcgcattgaacatgccaatgaccgaaagaataa    109

ECLACY          1451 gagcaggcaagctatttaccgatatgtgcgaaggcttaccggaaaaaaga   1500
                     ||||||||||||||||||||||||||||||||||||||||||||||||||
ECLACA           110 gagcaggcaagctatttaccgatatgtgcgaaggcttaccggaaaaaaga    159
 

.................... until ......................

 
#---------------------------------------
#
# ECLACY position base          ECLACA position base    Using
#
#
#---------------------------------------         

</pre>

<H2>
    Data files
</H2>



<H2>
    Notes
</H2>


None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It exits with a status of 0

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="cons.html">cons</a></td><td>Creates a consensus from multiple alignments</td></tr>
<tr><td><a href="megamerger.html">megamerger</a></td><td>Merge two large overlapping nucleic acid sequences</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Gary Williams <a
href="mailto:gwilliam@hgmp.mrc.ac.uk">(gwilliam@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>

Written (Gary Williams) 1999

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

