<HTML>

<HEAD>
  <TITLE>
  EMBOSS: merger
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
merger
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
Merge two overlapping sequences

<H2>
    Description
</H2>
<p><b>merger</b> reads two overlapping input sequences of the same
type (typically nucleotide) and uses a global alignment algorithm
(Needleman & Wunsch) to optimally align the sequences.  A merged
sequence is generated from the alignment and writen to the output
file. The sequence alignment is also written.</p>



<H2>
Algorithm
</H2>

<p>It uses a global alignment algorithm (Needleman & Wunsch) to
optimally align the sequences and then creates a merged sequence from
the alignment. When there is a mismatch in the alignment between the
two sequences, the base included in the merged sequence is the base
from the sequence which has the best local sequence quality score.</p>

<p>The following heuristic is used to find the sequence quality score.
If one of the bases is a 'N', then the other sequence's base is used.
Otherwise, a window size around the disputed base is used to find the
local quality score. This window size is increased from 5, to 10 to 20
bases or until there is a clear decision on the best choice. If there
is no best choice after using a window of 20, then the base in the
first sequence is used.</p>

<p>To calculate the quality of a window of a sequence around a base:</p>

<pre>
    * quality = sequence value/length under window either side of the base
    * sequence value = sum of points in that window
    * unambiguous bases (ACGTU) score 2 points
    * ambiguous bases (MRWSYKVHDB) score 1 point
    * Ns score 0 points
    * off end of the sequence scores 0 points 
</pre>
<p>This heavily discriminates against the iffy bits at the end of
sequence reads.</p>



<H2>
    Usage
</H2>
<b>Here is a sample session with merger</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>merger </b>
Merge two overlapping sequences
Input sequence: <b>tembl:v00295</b>
Second sequence: <b>tembl:x51872</b>
Output alignment [v00295.merger]: <b></b>
output sequence [v00295.fasta]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>


<p>

Typically, one of the sequences will need to be reverse-complemented
to put it into the correct orientation to make it join. For example:

<p>

<pre>
% merger file1.seq file2.seq -sreverse2 -outseq merged.seq
</pre>

<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
Merge two overlapping sequences
Version: EMBOSS:6.2.0

   Standard (Mandatory) qualifiers:
  [-asequence]         sequence   Sequence filename and optional format, or
                                  reference (input USA)
  [-bsequence]         sequence   Sequence filename and optional format, or
                                  reference (input USA)
  [-outfile]           align      [*.merger] Output alignment file name
  [-outseq]            seqout     [<sequence>.<format>] Sequence filename and
                                  optional format (output USA)

   Additional (Optional) qualifiers:
   -datafile           matrixf    [EBLOSUM62 for protein, EDNAFULL for DNA]
                                  This is the scoring matrix file used when
                                  comparing sequences. By default it is the
                                  file 'EBLOSUM62' (for proteins) or the file
                                  'EDNAFULL' (for nucleic sequences). These
                                  files are found in the 'data' directory of
                                  the EMBOSS installation.
   -gapopen            float      [@($(acdprotein)? 50.0 : 50.0 )] Gap opening
                                  penalty (Number from 0.000 to 100.000)
   -gapextend          float      [@($(acdprotein)? 5.0 : 5.0)] Gap extension
                                  penalty (Number from 0.000 to 10.000)

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-asequence" associated qualifiers
   -sbegin1            integer    Start of the sequence to be used
   -send1              integer    End of the sequence to be used
   -sreverse1          boolean    Reverse (if DNA)
   -sask1              boolean    Ask for begin/end/reverse
   -snucleotide1       boolean    Sequence is nucleotide
   -sprotein1          boolean    Sequence is protein
   -slower1            boolean    Make lower case
   -supper1            boolean    Make upper case
   -sformat1           string     Input sequence format
   -sdbname1           string     Database name
   -sid1               string     Entryname
   -ufo1               string     UFO features
   -fformat1           string     Features format
   -fopenfile1         string     Features file name

   "-bsequence" associated qualifiers
   -sbegin2            integer    Start of the sequence to be used
   -send2              integer    End of the sequence to be used
   -sreverse2          boolean    Reverse (if DNA)
   -sask2              boolean    Ask for begin/end/reverse
   -snucleotide2       boolean    Sequence is nucleotide
   -sprotein2          boolean    Sequence is protein
   -slower2            boolean    Make lower case
   -supper2            boolean    Make upper case
   -sformat2           string     Input sequence format
   -sdbname2           string     Database name
   -sid2               string     Entryname
   -ufo2               string     UFO features
   -fformat2           string     Features format
   -fopenfile2         string     Features file name

   "-outfile" associated qualifiers
   -aformat3           string     Alignment format
   -aextension3        string     File name extension
   -adirectory3        string     Output directory
   -aname3             string     Base file name
   -awidth3            integer    Alignment width
   -aaccshow3          boolean    Show accession number in the header
   -adesshow3          boolean    Show description in the header
   -ausashow3          boolean    Show the full USA in the alignment
   -aglobal3           boolean    Show the full sequence in alignment

   "-outseq" associated qualifiers
   -osformat4          string     Output seq format
   -osextension4       string     File name extension
   -osname4            string     Base file name
   -osdirectory4       string     Output directory
   -osdbname4          string     Database name to add
   -ossingle4          boolean    Separate file for each entry
   -oufo4              string     UFO features
   -offormat4          string     Features format
   -ofname4            string     Features file name
   -ofdirectory4       string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-asequence]<br>(Parameter 1)</td>
<td>sequence</td>
<td>Sequence filename and optional format, or reference (input USA)</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-bsequence]<br>(Parameter 2)</td>
<td>sequence</td>
<td>Sequence filename and optional format, or reference (input USA)</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-outfile]<br>(Parameter 3)</td>
<td>align</td>
<td>Output alignment file name</td>
<td>Alignment output file</td>
<td><i>&lt;*&gt;</i>.merger</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-outseq]<br>(Parameter 4)</td>
<td>seqout</td>
<td>Sequence filename and optional format (output USA)</td>
<td>Writeable sequence</td>
<td><i>&lt;*&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>-datafile</td>
<td>matrixf</td>
<td>This is the scoring matrix file used when comparing sequences. By default it is the file 'EBLOSUM62' (for proteins) or the file 'EDNAFULL' (for nucleic sequences). These files are found in the 'data' directory of the EMBOSS installation.</td>
<td>Comparison matrix file in EMBOSS data path</td>
<td>EBLOSUM62 for protein<br>EDNAFULL for DNA</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-gapopen</td>
<td>float</td>
<td>Gap opening penalty</td>
<td>Number from 0.000 to 100.000</td>
<td>@($(acdprotein)? 50.0 : 50.0 )</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-gapextend</td>
<td>float</td>
<td>Gap extension penalty</td>
<td>Number from 0.000 to 10.000</td>
<td>@($(acdprotein)? 5.0 : 5.0)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-asequence" associated sequence qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sbegin1<br>-sbegin_asequence</td>
<td>integer</td>
<td>Start of the sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -send1<br>-send_asequence</td>
<td>integer</td>
<td>End of the sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sreverse1<br>-sreverse_asequence</td>
<td>boolean</td>
<td>Reverse (if DNA)</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sask1<br>-sask_asequence</td>
<td>boolean</td>
<td>Ask for begin/end/reverse</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -snucleotide1<br>-snucleotide_asequence</td>
<td>boolean</td>
<td>Sequence is nucleotide</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sprotein1<br>-sprotein_asequence</td>
<td>boolean</td>
<td>Sequence is protein</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -slower1<br>-slower_asequence</td>
<td>boolean</td>
<td>Make lower case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -supper1<br>-supper_asequence</td>
<td>boolean</td>
<td>Make upper case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sformat1<br>-sformat_asequence</td>
<td>string</td>
<td>Input sequence format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sdbname1<br>-sdbname_asequence</td>
<td>string</td>
<td>Database name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sid1<br>-sid_asequence</td>
<td>string</td>
<td>Entryname</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ufo1<br>-ufo_asequence</td>
<td>string</td>
<td>UFO features</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fformat1<br>-fformat_asequence</td>
<td>string</td>
<td>Features format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fopenfile1<br>-fopenfile_asequence</td>
<td>string</td>
<td>Features file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-bsequence" associated sequence qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sbegin2<br>-sbegin_bsequence</td>
<td>integer</td>
<td>Start of the sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -send2<br>-send_bsequence</td>
<td>integer</td>
<td>End of the sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sreverse2<br>-sreverse_bsequence</td>
<td>boolean</td>
<td>Reverse (if DNA)</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sask2<br>-sask_bsequence</td>
<td>boolean</td>
<td>Ask for begin/end/reverse</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -snucleotide2<br>-snucleotide_bsequence</td>
<td>boolean</td>
<td>Sequence is nucleotide</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sprotein2<br>-sprotein_bsequence</td>
<td>boolean</td>
<td>Sequence is protein</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -slower2<br>-slower_bsequence</td>
<td>boolean</td>
<td>Make lower case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -supper2<br>-supper_bsequence</td>
<td>boolean</td>
<td>Make upper case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sformat2<br>-sformat_bsequence</td>
<td>string</td>
<td>Input sequence format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sdbname2<br>-sdbname_bsequence</td>
<td>string</td>
<td>Database name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sid2<br>-sid_bsequence</td>
<td>string</td>
<td>Entryname</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ufo2<br>-ufo_bsequence</td>
<td>string</td>
<td>UFO features</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fformat2<br>-fformat_bsequence</td>
<td>string</td>
<td>Features format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fopenfile2<br>-fopenfile_bsequence</td>
<td>string</td>
<td>Features file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-outfile" associated align qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -aformat3<br>-aformat_outfile</td>
<td>string</td>
<td>Alignment format</td>
<td>Any string</td>
<td>simple</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -aextension3<br>-aextension_outfile</td>
<td>string</td>
<td>File name extension</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -adirectory3<br>-adirectory_outfile</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -aname3<br>-aname_outfile</td>
<td>string</td>
<td>Base file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -awidth3<br>-awidth_outfile</td>
<td>integer</td>
<td>Alignment width</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -aaccshow3<br>-aaccshow_outfile</td>
<td>boolean</td>
<td>Show accession number in the header</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -adesshow3<br>-adesshow_outfile</td>
<td>boolean</td>
<td>Show description in the header</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ausashow3<br>-ausashow_outfile</td>
<td>boolean</td>
<td>Show the full USA in the alignment</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -aglobal3<br>-aglobal_outfile</td>
<td>boolean</td>
<td>Show the full sequence in alignment</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-outseq" associated seqout qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -osformat4<br>-osformat_outseq</td>
<td>string</td>
<td>Output seq format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -osextension4<br>-osextension_outseq</td>
<td>string</td>
<td>File name extension</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -osname4<br>-osname_outseq</td>
<td>string</td>
<td>Base file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -osdirectory4<br>-osdirectory_outseq</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -osdbname4<br>-osdbname_outseq</td>
<td>string</td>
<td>Database name to add</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ossingle4<br>-ossingle_outseq</td>
<td>boolean</td>
<td>Separate file for each entry</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -oufo4<br>-oufo_outseq</td>
<td>string</td>
<td>UFO features</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -offormat4<br>-offormat_outseq</td>
<td>string</td>
<td>Features format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ofname4<br>-ofname_outseq</td>
<td>string</td>
<td>Features file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ofdirectory4<br>-ofdirectory_outseq</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>


<H2>
    Input file format
</H2>

<b>merger</b> reads any two sequence USAs of the same type (protein or
nucleic acid.)

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:v00295' is a sequence entry in the example nucleic acid database 'tembl'
<p>
<p><h3>Database entry: tembl:v00295</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   V00295; SV 1; linear; genomic DNA; STD; PRO; 1500 BP.
XX
AC   V00295;
XX
DT   09-JUN-1982 (Rel. 01, Created)
DT   07-JUL-1995 (Rel. 44, Last updated, Version 4)
XX
DE   E. coli lacY gene (codes for lactose permease).
XX
KW   membrane protein.
XX
OS   Escherichia coli
OC   Bacteria; Proteobacteria; Gammaproteobacteria; Enterobacteriales;
OC   Enterobacteriaceae; Escherichia.
XX
RN   [1]
RP   1-1500
RX   DOI; 10.1038/283541a0.
RX   PUBMED; 6444453.
RA   Buechel D.E., Gronenborn B., Mueller-Hill B.;
RT   "Sequence of the lactose permease gene";
RL   Nature 283(5747):541-545(1980).
XX
CC   lacZ is a beta-galactosidase and lacA is transacetylase.
CC   KST ECO.LACY
XX
FH   Key             Location/Qualifiers
FH
FT   source          1..1500
FT                   /organism="Escherichia coli"
FT                   /mol_type="genomic DNA"
FT                   /db_xref="taxon:562"
FT   CDS             &lt;1..54
FT                   /codon_start=1
FT                   /transl_table=11
FT                   /note="reading frame (lacZ)"
FT                   /db_xref="GOA:P00722"
FT                   /db_xref="InterPro:IPR014718"
FT                   /db_xref="PDB:3DYP"
FT                   /db_xref="UniProtKB/Swiss-Prot:P00722"
FT                   /protein_id="CAA23570.1"
FT                   /translation="FQLSAGRYHYQLVWCQK"
FT   CDS             106..1359
FT                   /transl_table=11
FT                   /note="reading frame (lacY)"
FT                   /db_xref="GOA:P02920"
FT                   /db_xref="InterPro:IPR018457"
FT                   /db_xref="PDB:2V8N"
FT                   /db_xref="UniProtKB/Swiss-Prot:P02920"
FT                   /protein_id="CAA23571.1"
FT                   /translation="MYYLKNTNFWMFGLFFFFYFFIMGAYFPFFPIWLHDINHISKSDT
FT                   GIIFAAISLFSLLFQPLFGLLSDKLGLRKYLLWIITGMLVMFAPFFIFIFGPLLQYNIL
FT                   VGSIVGGIYLGFCFNAGAPAVEAFIEKVSRRSNFEFGRARMFGCVGWALCASIVGIMFT
FT                   INNQFVFWLGSGCALILAVLLFFAKTDAPSSATVANAVGANHSAFSLKLALELFRQPKL
FT                   WFLSLYVIGVSCTYDVFDQQFANFFTSFFATGEQGTRVFGYVTTMGELLNASIMFFAPL
FT                   IINRIGGKNALLLAGTIMSVRIIGSSFATSALEVVILKTLHMFEVPFLLVGCFKYITSQ
FT                   FEVRFSATIYLVCFCFFKQLAMIFMSVLAGNMYESIGFQGAYLVLGLVALGFTLISVFT
FT                   LSGPGPLSLLRRQVNEVA"
FT   CDS             1423..&gt;1500
FT                   /transl_table=11
FT                   /note="reading frame (lacA)"
FT                   /db_xref="GOA:P07464"
FT                   /db_xref="InterPro:IPR018357"
FT                   /db_xref="PDB:1KRV"
FT                   /db_xref="UniProtKB/Swiss-Prot:P07464"
FT                   /protein_id="CAA23572.1"
FT                   /translation="MNMPMTERIRAGKLFTDMCEGLPEKR"
XX
SQ   Sequence 1500 BP; 315 A; 342 C; 357 G; 486 T; 0 other;
     ttccagctga gcgccggtcg ctaccattac cagttggtct ggtgtcaaaa ataataataa        60
     ccgggcaggc catgtctgcc cgtatttcgc gtaaggaaat ccattatgta ctatttaaaa       120
     aacacaaact tttggatgtt cggtttattc tttttctttt acttttttat catgggagcc       180
     tacttcccgt ttttcccgat ttggctacat gacatcaacc atatcagcaa aagtgatacg       240
     ggtattattt ttgccgctat ttctctgttc tcgctattat tccaaccgct gtttggtctg       300
     ctttctgaca aactcgggct gcgcaaatac ctgctgtgga ttattaccgg catgttagtg       360
     atgtttgcgc cgttctttat ttttatcttc gggccactgt tacaatacaa cattttagta       420
     ggatcgattg ttggtggtat ttatctaggc ttttgtttta acgccggtgc gccagcagta       480
     gaggcattta ttgagaaagt cagccgtcgc agtaatttcg aatttggtcg cgcgcggatg       540
     tttggctgtg ttggctgggc gctgtgtgcc tcgattgtcg gcatcatgtt caccatcaat       600
     aatcagtttg ttttctggct gggctctggc tgtgcactca tcctcgccgt tttactcttt       660
     ttcgccaaaa cggatgcgcc ctcttctgcc acggttgcca atgcggtagg tgccaaccat       720
     tcggcattta gccttaagct ggcactggaa ctgttcagac agccaaaact gtggtttttg       780
     tcactgtatg ttattggcgt ttcctgcacc tacgatgttt ttgaccaaca gtttgctaat       840
     ttctttactt cgttctttgc taccggtgaa cagggtacgc gggtatttgg ctacgtaacg       900
     acaatgggcg aattacttaa cgcctcgatt atgttctttg cgccactgat cattaatcgc       960
     atcggtggga aaaacgccct gctgctggct ggcactatta tgtctgtacg tattattggc      1020
     tcatcgttcg ccacctcagc gctggaagtg gttattctga aaacgctgca tatgtttgaa      1080
     gtaccgttcc tgctggtggg ctgctttaaa tatattacca gccagtttga agtgcgtttt      1140
     tcagcgacga tttatctggt ctgtttctgc ttctttaagc aactggcgat gatttttatg      1200
     tctgtactgg cgggcaatat gtatgaaagc atcggtttcc agggcgctta tctggtgctg      1260
     ggtctggtgg cgctgggctt caccttaatt tccgtgttca cgcttagcgg ccccggcccg      1320
     ctttccctgc tgcgtcgtca ggtgaatgaa gtcgcttaag caatcaatgt cggatgcggc      1380
     gcgacgctta tccgaccaac atatcataac ggagtgatcg cattgaacat gccaatgacc      1440
     gaaagaataa gagcaggcaa gctatttacc gatatgtgcg aaggcttacc ggaaaaaaga      1500
//
</pre>
</td></tr></table><p>
<p><h3>Database entry: tembl:x51872</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   X51872; SV 1; linear; genomic DNA; STD; PRO; 1832 BP.
XX
AC   X51872;
XX
DT   17-APR-1990 (Rel. 23, Created)
DT   05-JUL-1999 (Rel. 60, Last updated, Version 5)
XX
DE   Escherichia coli lacA gene for thiogalactoside transacetylase
XX
KW   lac operon; lacA gene; lacY gene; thiogalactoside transacetylase.
XX
OS   Escherichia coli
OC   Bacteria; Proteobacteria; Gammaproteobacteria; Enterobacteriales;
OC   Enterobacteriaceae; Escherichia.
XX
RN   [1]
RC   (1-1832)
RP   1-1832
RX   DOI; 10.1073/pnas.82.19.6414.
RX   PUBMED; 3901000.
RA   Hediger M.A., Johnson D.F., Nierlich D.P., Zabin I.;
RT   "DNA sequence of the lactose operon: the lacA gene and the transcriptional
RT   termination region";
RL   Proc. Natl. Acad. Sci. U.S.A. 82(19):6414-6418(1985).
XX
FH   Key             Location/Qualifiers
FH
FT   source          1..1832
FT                   /organism="Escherichia coli"
FT                   /mol_type="genomic DNA"
FT                   /db_xref="taxon:562"
FT   CDS             &lt;1..18
FT                   /codon_start=1
FT                   /transl_table=11
FT                   /product="lacY gene product"
FT                   /protein_id="CAA36161.1"
FT                   /translation="VNEVA"
FT   CDS             82..693
FT                   /transl_table=11
FT                   /gene="lacA"
FT                   /product="thiogalactoside transacetylase"
FT                   /db_xref="GOA:P07464"
FT                   /db_xref="InterPro:IPR018357"
FT                   /db_xref="PDB:1KRV"
FT                   /db_xref="UniProtKB/Swiss-Prot:P07464"
FT                   /protein_id="CAA36162.1"
FT                   /translation="MNMPMTERIRAGKLFTDMCEGLPEKRLRGKTLMYEFNHSHPSEVE
FT                   KRESLIKEMFATVGENAWVEPPVYFSYGSNIHIGRNFYANFNLTIVDDYTVTIGDNVLI
FT                   APNVTLSVTGHPVHHELRKNGEMYSFPITIGNNVWIGSHVVINPGVTIGDNSVIGAGSI
FT                   VTKDIPPNVVAAGVPCRVIREINDRDKHYYFKDYKVESSV"
XX
SQ   Sequence 1832 BP; 519 A; 510 C; 450 G; 353 T; 0 other;
     gtgaatgaag tcgcttaagc aatcaatgtc ggatgcggcg cgacgcttat ccgaccaaca        60
     tatcataacg gagtgatcgc attgaacatg ccaatgaccg aaagaataag agcaggcaag       120
     ctatttaccg atatgtgcga aggcttaccg gaaaaaagac ttcgtgggaa aacgttaatg       180
     tatgagttta atcactcgca tccatcagaa gttgaaaaaa gagaaagcct gattaaagaa       240
     atgtttgcca cggtagggga aaacgcctgg gtagaaccgc ctgtctattt ctcttacggt       300
     tccaacatcc atataggccg caatttttat gcaaatttca atttaaccat tgtcgatgac       360
     tacacggtaa caatcggtga taacgtactg attgcaccca acgttactct ttccgttacg       420
     ggacaccctg tacaccatga attgagaaaa aacggcgaga tgtactcttt tccgataacg       480
     attggcaata acgtctggat cggaagtcat gtggttatta atccaggcgt caccatcggg       540
     gataattctg ttattggcgc gggtagtatc gtcacaaaag acattccacc aaacgtcgtg       600
     gcggctggcg ttccttgtcg ggttattcgc gaaataaacg accgggataa gcactattat       660
     ttcaaagatt ataaagttga atcgtcagtt taaattataa aaattgcctg atacgctgcg       720
     cttatcaggc ctacaagttc agcgatctac attagccgca tccggcatga acaaagcgca       780
     ggaacaagcg tcgcatcatg cctctttgac ccacagctgc ggaaaacgta ctggtgcaaa       840
     acgcagggtt atgatcatca gcccaacgac gcacagcgca tgaaatgccc agtccatcag       900
     gtaattgccg ctgatactac gcagcacgcc agaaaaccac ggggcaagcc cggcgatgat       960
     aaaaccgatt ccctgcataa acgccaccag cttgccagca atagccggtt gcacagagtg      1020
     atcgagcgcc agcagcaaac agagcggaaa cgcgccgccc agacctaacc cacacaccat      1080
     cgcccacaat accggcaatt gcatcggcag ccagataaag ccgcagaacc ccaccagttg      1140
     taacaccagc gccagcatta acagtttgcg ccgatcctga tggcgagcca tagcaggcat      1200
     cagcaaagct cctgcggctt gcccaagcgt catcaatgcc agtaaggaac cgctgtactg      1260
     cgcgctggca ccaatctcaa tatagaaagc gggtaaccag gcaatcaggc tggcgtaacc      1320
     gccgttaatc agaccgaagt aaacacccag cgtccacgcg cggggagtga ataccacgcg      1380
     aaccggagtg gttgttgtct tgtgggaaga ggcgacctcg cgggcgcttt gccaccacca      1440
     ggcaaagagc gcaacaacgg caggcagcgc caccaggcga gtgtttgata ccaggtttcg      1500
     ctatgttgaa ctaaccaggg cgttatggcg gcaccaagcc caccgccgcc catcagagcc      1560
     gcggaccaca gccccatcac cagtggcgtg cgctgctgaa accgccgttt aatcaccgaa      1620
     gcatcaccgc ctgaatgatg ccgatcccca ccccaccaag cagtgcgctg ctaagcagca      1680
     gcgcactttg cgggtaaagc tcacgcatca atgcaccgac ggcaatcagc aacagactga      1740
     tggcgacact gcgacgttcg ctgacatgct gatgaagcca gcttccggcc agcgccagcc      1800
     cgcccatggt aaccaccggc agagcggtcg ac                                    1832
//
</pre>
</td></tr></table><p>

<H2>
    Output file format
</H2>


The output <i>sequence file</i> contains the joined sequence, by default in
FASTA format.  Where there is a mismatch in the alignment, the chosen
base is written to the output sequence in uppercase. 

<p>

<p>

The output is a standard EMBOSS alignment file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-aformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  Some of the alignment formats can cope
with an unlimited number of sequences, while others are only for pairs
of sequences.  

<p>

The available multiple alignment format names are: unknown, multiple,
simple, fasta, msf, trace, srs

<p>

The available pairwise alignment format names are: pair, markx0, markx1,
markx2, markx3, markx10, srspair, score

<p>

See:
<A href="http://emboss.sf.net/docs/themes/AlignFormats.html">
http://emboss.sf.net/docs/themes/AlignFormats.html</A>
for further information on alignment formats.

<p>
 
<p>

The output <i>report file</i> contains descriptions of the positions where
there is a mismatch in the alignment and shows the alignment.  Where
there is a mismatch in the alignment, the chosen base is written in
uppercase. 


<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: v00295.merger</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
########################################
# Program: merger
# Rundate: Fri 15 Jan 2010 12:00:00
# Commandline: merger
#    -asequence tembl:v00295
#    -bsequence tembl:x51872
# Align_format: simple
# Report_file: v00295.merger
########################################

#=======================================
#
# Aligned_sequences: 2
# 1: V00295
# 2: X51872
# Matrix: EDNAFULL
# Gap_penalty: 50.0
# Extend_penalty: 5.0
#
# Length: 3173
# Identity:     159/3173 ( 5.0%)
# Similarity:   159/3173 ( 5.0%)
# Gaps:        3014/3173 (95.0%)
# Score: 795.0
# 
#
#=======================================

V00295             1 ttccagctgagcgccggtcgctaccattaccagttggtctggtgtcaaaa     50
                                                                       
X51872             1 --------------------------------------------------      0

V00295            51 ataataataaccgggcaggccatgtctgcccgtatttcgcgtaaggaaat    100
                                                                       
X51872             1 --------------------------------------------------      0

V00295           101 ccattatgtactatttaaaaaacacaaacttttggatgttcggtttattc    150
                                                                       
X51872             1 --------------------------------------------------      0

V00295           151 tttttcttttacttttttatcatgggagcctacttcccgtttttcccgat    200
                                                                       
X51872             1 --------------------------------------------------      0

V00295           201 ttggctacatgacatcaaccatatcagcaaaagtgatacgggtattattt    250
                                                                       
X51872             1 --------------------------------------------------      0

V00295           251 ttgccgctatttctctgttctcgctattattccaaccgctgtttggtctg    300
                                                                       


<font color=red>  [Part of this file has been deleted for brevity]</font>

                                                                       
X51872          1310 ctggcgtaaccgccgttaatcagaccgaagtaaacacccagcgtccacgc   1359

V00295          1501 --------------------------------------------------   1500
                                                                       
X51872          1360 gcggggagtgaataccacgcgaaccggagtggttgttgtcttgtgggaag   1409

V00295          1501 --------------------------------------------------   1500
                                                                       
X51872          1410 aggcgacctcgcgggcgctttgccaccaccaggcaaagagcgcaacaacg   1459

V00295          1501 --------------------------------------------------   1500
                                                                       
X51872          1460 gcaggcagcgccaccaggcgagtgtttgataccaggtttcgctatgttga   1509

V00295          1501 --------------------------------------------------   1500
                                                                       
X51872          1510 actaaccagggcgttatggcggcaccaagcccaccgccgcccatcagagc   1559

V00295          1501 --------------------------------------------------   1500
                                                                       
X51872          1560 cgcggaccacagccccatcaccagtggcgtgcgctgctgaaaccgccgtt   1609

V00295          1501 --------------------------------------------------   1500
                                                                       
X51872          1610 taatcaccgaagcatcaccgcctgaatgatgccgatccccaccccaccaa   1659

V00295          1501 --------------------------------------------------   1500
                                                                       
X51872          1660 gcagtgcgctgctaagcagcagcgcactttgcgggtaaagctcacgcatc   1709

V00295          1501 --------------------------------------------------   1500
                                                                       
X51872          1710 aatgcaccgacggcaatcagcaacagactgatggcgacactgcgacgttc   1759

V00295          1501 --------------------------------------------------   1500
                                                                       
X51872          1760 gctgacatgctgatgaagccagcttccggccagcgccagcccgcccatgg   1809

V00295          1501 -----------------------   1500
                                            
X51872          1810 taaccaccggcagagcggtcgac   1832


#---------------------------------------
#
# Conflicts:          V00295          X51872
#              position base   position base Using
# 
#
#---------------------------------------
</pre>
</td></tr></table><p>
<p><h3>File: v00295.fasta</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt;V00295 V00295.1 E. coli lacY gene (codes for lactose permease).
ttccagctgagcgccggtcgctaccattaccagttggtctggtgtcaaaaataataataa
ccgggcaggccatgtctgcccgtatttcgcgtaaggaaatccattatgtactatttaaaa
aacacaaacttttggatgttcggtttattctttttcttttacttttttatcatgggagcc
tacttcccgtttttcccgatttggctacatgacatcaaccatatcagcaaaagtgatacg
ggtattatttttgccgctatttctctgttctcgctattattccaaccgctgtttggtctg
ctttctgacaaactcgggctgcgcaaatacctgctgtggattattaccggcatgttagtg
atgtttgcgccgttctttatttttatcttcgggccactgttacaatacaacattttagta
ggatcgattgttggtggtatttatctaggcttttgttttaacgccggtgcgccagcagta
gaggcatttattgagaaagtcagccgtcgcagtaatttcgaatttggtcgcgcgcggatg
tttggctgtgttggctgggcgctgtgtgcctcgattgtcggcatcatgttcaccatcaat
aatcagtttgttttctggctgggctctggctgtgcactcatcctcgccgttttactcttt
ttcgccaaaacggatgcgccctcttctgccacggttgccaatgcggtaggtgccaaccat
tcggcatttagccttaagctggcactggaactgttcagacagccaaaactgtggtttttg
tcactgtatgttattggcgtttcctgcacctacgatgtttttgaccaacagtttgctaat
ttctttacttcgttctttgctaccggtgaacagggtacgcgggtatttggctacgtaacg
acaatgggcgaattacttaacgcctcgattatgttctttgcgccactgatcattaatcgc
atcggtgggaaaaacgccctgctgctggctggcactattatgtctgtacgtattattggc
tcatcgttcgccacctcagcgctggaagtggttattctgaaaacgctgcatatgtttgaa
gtaccgttcctgctggtgggctgctttaaatatattaccagccagtttgaagtgcgtttt
tcagcgacgatttatctggtctgtttctgcttctttaagcaactggcgatgatttttatg
tctgtactggcgggcaatatgtatgaaagcatcggtttccagggcgcttatctggtgctg
ggtctggtggcgctgggcttcaccttaatttccgtgttcacgcttagcggccccggcccg
ctttccctgctgcgtcgtcaggtgaatgaagtcgcttaagcaatcaatgtcggatgcggc
gcgacgcttatccgaccaacatatcataacggagtgatcgcattgaacatgccaatgacc
gaaagaataagagcaggcaagctatttaccgatatgtgcgaaggcttaccggaaaaaaga
cttcgtgggaaaacgttaatgtatgagtttaatcactcgcatccatcagaagttgaaaaa
agagaaagcctgattaaagaaatgtttgccacggtaggggaaaacgcctgggtagaaccg
cctgtctatttctcttacggttccaacatccatataggccgcaatttttatgcaaatttc
aatttaaccattgtcgatgactacacggtaacaatcggtgataacgtactgattgcaccc
aacgttactctttccgttacgggacaccctgtacaccatgaattgagaaaaaacggcgag
atgtactcttttccgataacgattggcaataacgtctggatcggaagtcatgtggttatt
aatccaggcgtcaccatcggggataattctgttattggcgcgggtagtatcgtcacaaaa
gacattccaccaaacgtcgtggcggctggcgttccttgtcgggttattcgcgaaataaac
gaccgggataagcactattatttcaaagattataaagttgaatcgtcagtttaaattata
aaaattgcctgatacgctgcgcttatcaggcctacaagttcagcgatctacattagccgc
atccggcatgaacaaagcgcaggaacaagcgtcgcatcatgcctctttgacccacagctg
cggaaaacgtactggtgcaaaacgcagggttatgatcatcagcccaacgacgcacagcgc
atgaaatgcccagtccatcaggtaattgccgctgatactacgcagcacgccagaaaacca
cggggcaagcccggcgatgataaaaccgattccctgcataaacgccaccagcttgccagc
aatagccggttgcacagagtgatcgagcgccagcagcaaacagagcggaaacgcgccgcc
cagacctaacccacacaccatcgcccacaataccggcaattgcatcggcagccagataaa
gccgcagaaccccaccagttgtaacaccagcgccagcattaacagtttgcgccgatcctg
atggcgagccatagcaggcatcagcaaagctcctgcggcttgcccaagcgtcatcaatgc
cagtaaggaaccgctgtactgcgcgctggcaccaatctcaatatagaaagcgggtaacca
ggcaatcaggctggcgtaaccgccgttaatcagaccgaagtaaacacccagcgtccacgc
gcggggagtgaataccacgcgaaccggagtggttgttgtcttgtgggaagaggcgacctc
gcgggcgctttgccaccaccaggcaaagagcgcaacaacggcaggcagcgccaccaggcg
agtgtttgataccaggtttcgctatgttgaactaaccagggcgttatggcggcaccaagc
ccaccgccgcccatcagagccgcggaccacagccccatcaccagtggcgtgcgctgctga
aaccgccgtttaatcaccgaagcatcaccgcctgaatgatgccgatccccaccccaccaa
gcagtgcgctgctaagcagcagcgcactttgcgggtaaagctcacgcatcaatgcaccga
cggcaatcagcaacagactgatggcgacactgcgacgttcgctgacatgctgatgaagcc
agcttccggccagcgccagcccgcccatggtaaccaccggcagagcggtcgac
</pre>
</td></tr></table><p>


<H2>
    Data files
</H2>

It reads the scoring matrix for the alignment from the standard EMBOSS
'data' directory.  By default it is the file 'EBLOSUM62' (for proteins)
or the file 'EDNAFULL' (for nucleic sequences). 

<H2>
    Notes
</H2>

<p>This program was originally written to aid in the reconstruction of
mRNA sequences which had been sequenced from both ends as a 5' and 3'
EST (cDNA). eg. joining two reads produced by primer walking
sequencing.</p>

<p>The gap open and gap extension penalties have been set at a higher
level than is usual (50 and 5). This was experimentally determined to
give the best results with a set of poor quality EST test
sequences.</p>

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

<p>Care should be taken to reverse one of the sequences (e.g. using the qualifier <tt>-sreverse2</tt>) if this is required to get them both in the correct orientation..</p>
<p><b>merger</b> uses the memory-hungry Needleman & Wunsch alignment.  The required memory may be greater than the available memory when attempting to merge large (cosmid-sized or greater) sequences.</p>


<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It exits with a status of 0

<H2>
    Known bugs
</H2>

None.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="cons.html">cons</a></td>
<td>Create a consensus sequence from a multiple alignment</td>
</tr>

<tr>
<td><a href="consambig.html">consambig</a></td>
<td>Create an ambiguous consensus sequence from a multiple alignment</td>
</tr>

<tr>
<td><a href="megamerger.html">megamerger</a></td>
<td>Merge two large overlapping DNA sequences</td>
</tr>

</table>

<H2>
    Author(s)
</H2>


Gary Williams (gwilliam&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK



<H2>
    History
</H2>

Written (Gary Williams) 1999

<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.


<H2>
    Comments
</H2>
None

</BODY>
</HTML>

