<HTML>

<HEAD>
  <TITLE>
  EMBOSS: mse
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>
<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: mse
</H2>
</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>

<H1>
Program mse
</H1>
<H2>
    Function
</H2>

Conversion of William A. Gilbert's ( Whitehead Insitutute ) MSE.

<H2>
    Description
</H2>

 A Multiple Sequence Screen Editor for Biological Sequences.


<H2>
    Usage
</H2>

    This is an interactive program and 
it is similar to many other multiple sequence editors, in that you
run it, using a file of existing sequences, as:
<p>

% mse file.seq
<p>

...you will then see a screen looking like this:
<p>

<hr>
<pre>
MSE V3.0                                                                    WIBR



     

     tgactgatcgttgctgattgatgctgactg
     tgactgatcgt-gctgactgatgctgactg
     tgactgatcgtagctgactgatgctgactg
....|.........|.........|.........|.........|.........|.........|.........|....
    0        10        20        30        40        50        60        70







      |......|......|......|......|......|......|......|......|......|......|
      0     10     20     30     40     50     60     70     80     90     100

003  M +-----#------------>
002    +------------------>
001    +------------------>
Press "?" for help
</pre>
<hr>
<p>

Pressing the cursor keys will move the cursor around, left and right, up
and down. Moving the cursor in the sequences will also move the '#' character
in the "bird's eye view" in the lower screen - this is an overview of where
you currently are in the full-length set of sequences.
<p>

If you press a key that is a valid nucleotide or residue code, then the
appropriate sequence in the "bird's eye view" will be marked with a 'M'
indicating that it has been modified.
<p>

Pressing '?' at any time will display the help:
<p>

<hr>
<pre>
                          Screen Mode Commands,
             ([n] is an optional numeric (keyboard) parameter)

[n]s            - Inserts symbol "s" [n] times
[n]&lt;delete&gt;     - Delete [n] sequence character(s)
[n]&lt;space bar&gt;  - Move the sequence(s) to the right [n] spaces
[n]&lt;backspace&gt;  - Move the sequence(s) to the left [n] spaces, VT200's
                   use F12 key or Control-H

[n]&lt;arrow&gt;      - Go forward, backward, up or down [n] positions or lines
[n]&lt;return&gt;     - Go to position [n], also use keypad ENTER key
   *            - Go to end of current sequence (Also Control-E)
[n]&lt;            - Go [n]x50 back, VT200's also use PREV SCREEN key
[n]&gt;            - Go [n]x50 ahead, VT200's also use NEXT SCREEN key

   /string&lt;rtn&gt; - Find "string", VT200's also use FIND key
[n]?            - Display help starting at page [n], VT200's also use HELP key
   !            - Display status of current sequence, same as "SEQ" command
   Control-W    - Redraw the screen (also Control-R)
   Control-Z    - Switch to Command Mode



     [Help page 1 of 10.  Press RETURN for more help. SPACE to quit.] 
</pre>
<hr>
<p>

See below for the full display of these commands.
<p>

As you can see in the help-screen above, you type in <b>Control-Z</b>
to switch from moving around the sequences and editing the sequences to
the mode in which you can type in text commands.
<p>

When you have typed <b>Control-Z</b>, the prompt <b>Command:</b> appears
at the bottom of the screen waiting for you to either type in a text
command, or to simply press the RETURN key to return to editing the
sequences. 
<p>

To get help when in command-mode, type <b>help</b> and press RETURN.
<p>


To exit from the editor and write out your changed sequences to file,
give the text command <b>exit</b>.


<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-msf]               seqset     File containing a sequence alignment

   Optional qualifiers: (none)
   Advanced qualifiers:
   -tmpdir             string     temp directory

</pre>
</td></tr></table>

<p>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-msf]<br>(Parameter 1)</td>
<td>File containing a sequence alignment</td>
<td>Readable sequences</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-tmpdir</td>
<td>temp directory</td>
<td>Any string is accepted</td>
<td>/tmp</td>
</tr>

</table>


<H2>
    Commands
</H2><pre>
                          Screen Mode Commands
             ([n] is an optional numeric (keyboard) parameter)

   Control-Z    - Switch to Command Mode


[n]s            - Inserts symbol "s" [n] times
[n]&lt;delete&gt;     - Delete [n] sequence character(s)
[n]&lt;space bar&gt;  - Move the sequence(s) to the right [n] spaces
[n]&lt;backspace&gt;  - Move the sequence(s) to the left [n] spaces, VT200's
                   use F12 key or Control-H

[n]&lt;arrow&gt;      - Go forward, backward, up or down [n] positions or lines
[n]&lt;return&gt;     - Go to position [n], also use keypad ENTER key
   *            - Go to end of current sequence (Also Control-E)
[n]&lt;            - Go [n]x50 back, VT200's also use PREV SCREEN key
[n]&gt;            - Go [n]x50 ahead, VT200's also use NEXT SCREEN key

   /string&lt;rtn&gt; - Find "string", VT200's also use FIND key
[n]?            - Display help starting at page [n], VT200's also use HELP key
   !            - Display status of current sequence, same as "SEQ" command
   Control-W    - Redraw the screen (also Control-R)


Commands end with &lt;ret&gt;, [ ] indicates an optional parameter.
"s" and "f" are line numbers for to and from or start and finish.
Only the capitalized part of the command is necessary


      EDIt [SeqSpec(s)] - edit another sequence(s) or indirect FOSS file

      FInd string       - find the next occurrence of "string"

      [s,f] DELete      - delete a range of symbols or answer prompts
                          for a range, default will be: current to *

      [n]               - go to position [n], same as raw mode [n]

      [s] HELp          - display help starting at page [s]

      REDraw            - redraw the screen

      MATches [attr]    - Display matches with plurality sequence

      DIFferences [attr]- Display differences with plurality sequence

      NEIther           - Turn off plurality sequence calculation
                           Saves CPU time!!

      ALNED [filename]  - Read in an alignment file from the NBRF/PIR program
                           ALNED (A multiple sequence alignment program).

      [s,f] MOVe        - Move from line [s] to line [f].
                           Default is from current to next available line.

      [s,f] ELIminate   - Delete lines [s] to [f]. Default is current line.

      [s,f] ANChor      - Force lines [s] to [f] to move as one unit.
                           Default is to add the current strand to the anchored
                           group. PF4 key toggles anchoring.

      NOAnchor          - Turn off anchoring. Default is the current strand.
                           PF4 toggles anchoring

      [s,f] SORt [dir]  - Sort lines [s] to [f] according to their "Offsets".
                           Default is from current to highest numbered line.
                           If "dir" is DEScending, the sort is reversed.

      [s,f] DEGap       - Remove gaps from lines [s] to [f].




      Open              - Open up a new line by moving all the line above the current
                           line up by one

      [s,f] Lock        - Lock lines "s" to "f" such that you cannot add or delete
                           sequence symbols. You may still shift and/or reverse lines.
                           You may also add/delete gap characters "-".  Default is the
                           current line.

      [s,f] Unlock      - Unlocks locked lines.

      [s] Offset        - Set the Offset of the current line to position [s].
                           Default is set the Offset of the current line to the
                           current line position. Also works with anchored groups.

      [s] REVerse       - Reverse/complement line "s".  The current line is the defualt
                           If the reversed line is part of an anchored group then the
                           entire group is reversed.



      SEQ               - Show current position. Same as "!" in raw mode.

   NAMe [code=filename] - Change the code and/or filename.

   TItle [New Title]    - Change the sequence title.
   TYpe [type]          - Change the sequence type, values for TYPE:

                             DNA     PROTEIN     RNA
                            rRNA    tRNA        uRNA        mRNA

   FORmat [New format]  - Change the sequence format, for FORMAT:

                          GCG PIR FASTA SWISS NCBI etc..



   [s,f] SELect         - Mark a range [s] to [f] to be selected, or start
                           selection at current cursor position.
                           VT200's also use SELECT key.

    REMove              - Remove a selected, (Reverse video), region. Does not
                           include current cursor position.
                           VT200's also use REMOVE key.

    [s] INSert          - Insert cut region at [s] or current cursor position.
                           VT200's also use INSERT HERE key.

      CANcel            - Cancel SELECT function. Clears PASTE buffer.




      GelAlign          - Will take all of the strands currently loaded into MSE
                           and produce the best alignment based on exact matches.
                           GelAlign does not add gaps but it will get the strands
                           started for you

   [s,f] Meld [seqname] - Merges strands "s" to "f" to create a new entry
                           named "seqname".  If the current strand is part of an
                           anchored group then that group will be melded.  You
                           will be prompted to continue if undetermined postitions
                           are found in the melded sequence.

[s,f] HARdcopy [filename] - write [a part of] aligned sequences to a file.

[s,f] WRIte [seqname]     - write [a part of] the sequence to a file. You
                             may rename it to "seqname" in the process

      EXIt  [FOSSname]    - write any modified sequences, a new FOSS file,
                             and then quit.

      QUIt                - quit the editor, do not write sequences.




</pre>

<H2>
    Data files
</H2>
<H2>
    Notes
</H2>Multiple sequences sets not implemented yet!.

<H2>
    References
</H2>
<H2>
    Warnings
</H2>
<H2>
    Diagnostics
</H2>
<H2>
    Exit status
</H2>
<H2>
    Known bugs
</H2>
<H2>
    See also
</H2>
<H2>
    Author(s)
</H2>
 William A. Gilbert, Whitehead Insitutute.<p>
This application was modified for inclusion in EMBOSS by Ian Longden <a
href="mailto:il@sanger.ac.uk">(il@sanger.ac.uk)</a> Informatics
Division, The Sanger Centre, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.


<H2>
    Priority
</H2>
<H2>
    Target users
</H2>
<H2>
    Comments
</H2>
<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

