<HTML>

<HEAD>
  <TITLE>
  EMBOSS: needle
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">




<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

needle
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>





<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Needleman-Wunsch global alignment


<H2>
    Description
</H2>


This program uses the Needleman-Wunsch global alignment algorithm to
find the optimum alignment (including gaps) of two sequences when
considering their entire length. 

<p>
The Needleman-Wunsch algorithm is a member of the class of algorithms
that can calculate the best score and alignment in the order of mn
steps, (where 'n' and 'm' are the lengths of the two sequences).  These
dynamic programming algorithms were first developed for protein sequence
comparison by Needleman and Wunsch, though similar methods were
independently devised during the late 1960's and early 1970's for use in
the fields of speech processing and computer science.

<p>
What is the optimal alignment? Dynamic programming methods ensure the
optimal global alignment by exploring all possible alignments and
choosing the best.  It does this by reading in a scoring matrix that
contains values for every possible residue or nucleotide match.  Needle
finds an alignment with the maximum possible score where the score of an
alignment is equal to the sum of the matches taken from the scoring
matrix. 

<p>
An important problem is the treatment of gaps, i.e., spaces inserted to
optimise the alignment score.  A penalty is subtracted from the score
for each gap opened (the 'gap open' penalty) and a penalty is subtracted
from the score for the total number of gap spaces multiplied by a cost
(the 'gap extension' penalty).

<p>
Typically, the cost of extending a gap is set to be 5-10 times lower
than the cost for opening a gap. 

<H2>
    Usage
</H2>


Here is a sample session with needle.

<pre>

% needle sw:hba_human sw:hbb_human

Gap opening penalty [10.0]: 
Gap extension penalty [0.5]: 
Output file [hba_human.needle]: 

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequencea]         sequence   Sequence USA
  [-seqall]            seqall     Sequence database USA
   -gapopen            float      The gap open penalty is the score taken away
                                  when a gap is created. The best value
                                  depends on the choice of comparison matrix.
                                  The default value assumes you are using the
                                  EBLOSUM62 matrix for protein sequences, and
                                  the EDNAFULL matrix for nucleotide
                                  sequences.
   -gapextend          float      The gap extension, penalty is added to the
                                  standard gap penalty for each base or
                                  residue in the gap. This is how long gaps
                                  are penalized. Usually you will expect a few
                                  long gaps rather than many short gaps, so
                                  the gap extension penalty should be lower
                                  than the gap penalty. An exception is where
                                  one or both sequences are single reads with
                                  possible sequencing errors in which case you
                                  would expect many single base gaps. You can
                                  get this result by setting the gap open
                                  penalty to zero (or very low) and using the
                                  gap extension penalty to control gap
                                  scoring.
  [-outfile]           align      Output alignment file name

   Optional qualifiers:
   -datafile           matrixf    This is the scoring matrix file used when
                                  comparing sequences. By default it is the
                                  file 'EBLOSUM62' (for proteins) or the file
                                  'EDNAFULL' (for nucleic sequences). These
                                  files are found in the 'data' directory of
                                  the EMBOSS installation.

   Advanced qualifiers:
   -[no]similarity     boolean    Display percent identity and similarity

   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequencea]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-seqall]<br>(Parameter 2)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-gapopen</td>
<td>The gap open penalty is the score taken away when a gap is created. The best value depends on the choice of comparison matrix. The default value assumes you are using the EBLOSUM62 matrix for protein sequences, and the EDNAFULL matrix for nucleotide sequences.</td>
<td>Floating point number from 1.0 to 100.0</td>
<td>10.0 for any sequence</td>
</tr>

<tr>
<td>-gapextend</td>
<td>The gap extension, penalty is added to the standard gap penalty for each base or residue in the gap. This is how long gaps are penalized. Usually you will expect a few long gaps rather than many short gaps, so the gap extension penalty should be lower than the gap penalty. An exception is where one or both sequences are single reads with possible sequencing errors in which case you would expect many single base gaps. You can get this result by setting the gap open penalty to zero (or very low) and using the gap extension penalty to control gap scoring.</td>
<td>Floating point number from 0.0 to 10.0</td>
<td>0.5 for any sequence</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 3)</td>
<td>Output alignment file name</td>
<td>Alignment file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-datafile</td>
<td>This is the scoring matrix file used when comparing sequences. By default it is the file 'EBLOSUM62' (for proteins) or the file 'EDNAFULL' (for nucleic sequences). These files are found in the 'data' directory of the EMBOSS installation.</td>
<td>Comparison matrix file in EMBOSS data path</td>
<td>EBLOSUM62 for protein<br>EDNAFULL for DNA</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]similarity</td>
<td>Display percent identity and similarity</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

</table>



<H2>
    Input file format
</H2>

    Any 2 sequence USAs of the same type (DNA or protein).

<H2>
    Output file format
</H2>

<p>

The output is a standard EMBOSS alignment file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-aformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  Some of the alignment formats can cope
with an unlimited number of sequences, while others are only for pairs
of sequences.  

<p>

The available multiple alignment format names are: unknown, multiple,
simple, fasta, msf, trace, srs

<p>

The available pairwise alignment format names are: pair, markx0, markx1,
markx2, markx3, markx10, srspair, score

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/AlignFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/AlignFormats.html</A>
for further information on alignment formats.

<p>


<p>


Here is the output file from the example run:
<P>

<HR>
<pre>
########################################
# Program:  needle
# Rundate:  Mon May 20 16:22:49 2002
# Report_file: hba_human.needle
########################################
#=======================================
#
# Aligned_sequences: 2
# 1: HBA_HUMAN
# 2: HBB_HUMAN
# Matrix: EBLOSUM62
# Gap_penalty: 10.0
# Extend_penalty: 0.5
#
# Length: 148
# Identity:      63/148 (42.6%)
# Similarity:    88/148 (59.5%)
# Gaps:           9/148 ( 6.1%)
# Score: 290.5
#
#
#=======================================

HBA_HUMAN          1  VLSPADKTNVKAAWGKVGAHAGEYGAEALERMFLSFPTTKTYFPHF-DL     48
                      .|:|.:|:.|.|.||||  :..|.|.|||.|:.:.:|.|:.:|..| ||
HBB_HUMAN          1 VHLTPEEKSAVTALWGKV--NVDEVGGEALGRLLVVYPWTQRFFESFGDL     48

HBA_HUMAN         49 S-----HGSAQVKGHGKKVADALTNAVAHVDDMPNALSALSDLHAHKLRV     93
                     |     .|:.:||.|||||..|.::.:||:|::....:.||:||..||.|
HBB_HUMAN         49 STPDAVMGNPKVKAHGKKVLGAFSDGLAHLDNLKGTFATLSELHCDKLHV     98

HBA_HUMAN         94 DPVNFKLLSHCLLVTLAAHLPAEFTPAVHASLDKFLASVSTVLTSKYR    141
                     ||.||:||.:.|:..||.|...||||.|.|:..|.:|.|:..|..||.
HBB_HUMAN         99 DPENFRLLGNVLVCVLAHHFGKEFTPPVQAAYQKVVAGVANALAHKYH    146


#---------------------------------------
#---------------------------------------  </pre>
<HR>
<P>

The <b>%id</b> is the percentage of identical matches between the two
sequences over the reported aligned region.
<P>

The <b>%similarity</b> is the percentage of matches between the two
sequences over the reported aligned region where the scoring matrix
value is greater or equal to 0.0.
<P>

The <b>Overall %id</b> and <b>Overall %similarity</b> are calculated in
a similar manner for the number of matches over the length of the
longest of the two sequences.


<H2>
    Data files
</H2>


For protein sequences EBLOSUM62 is used for the substitution
matrix. For nucleotide sequence, EDNAFULL is used. Others can be
specified.

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by EMBOSS environment variable EMBOSS_DATA.

<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>

<H2>
    Notes
</H2>

<b>needle</b> is a true imple,entation of the Needleman-Wunsch algorithm 
and so produces a full path
matrix. It therefore cannot be used with genome sized
sequences unless you've a lot of memory and a lot of time.

<H2>
    References
</H2>


<ol>

<li>
Needleman, S.  B.  and Wunsch, C.  D.  (1970) J.  Mol.  Biol.  48,
443-453. 

<li>
Kruskal, J.  B.  (1983) An overview of squence comparison In D.  Sankoff
and J.  B.  Kruskal, (ed.), Time warps, string edits and macromolecules:
the theory and practice of sequence comparison, pp.  1-44 Addison
Wesley. 
</ol>

<H2>
    Warnings
</H2>


needle is for aligning two sequences over their entire length.  This
works best with closely related sequences.  If you use needle to align
very distantly-related sequences, it will produce a result but much of
the alignment may have little or no biological significance. 

<p>

A true Needleman Wunsch implementation like <b>needle</b> needs memory
proportional to the product of the sequence lengths.  For two sequences
of length 10,000,000 and 1,000 it therefore needs memory proportional to
10,000,000,000 characters.  Two arrays of this size are produced, one of
ints and one of floats so multiply that figure by 8 to get the memory
usage in bytes.  That doesn't include other overheads.  Therefore only
use water and needle for accurate alignment of reasonably short
sequences.

<p>

If you run out of memory, try using 
<b>stretcher</b> instead.

<H2>
    Diagnostic Error Messages
</H2>

<pre>
Uncaught exception
 Assertion failed
 raised at ajmem.c:xxx
</pre>

<p>

Probably means you have run out of memory. Try using 
<b>stretcher</b> if this happens. 

<H2>
    Exit status
</H2>

  0 upon successful completion.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="alignwrap.html">alignwrap</a></td><td>Aligns a set of sequences to a seed alignment</td></tr>
<tr><td><a href="est2genome.html">est2genome</a></td><td>Align EST and genomic DNA sequences</td></tr>
<tr><td><a href="stretcher.html">stretcher</a></td><td>Finds the best global alignment between two sequences</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->

<P>
When you want an alignment that covers the whole length of both
sequences, use <a href="needle.html">needle. </A> 

<p>
When you are trying to find the best region of similarity between two
sequences, use <a href="water.html">water.</a>
<p>

<a href="stretcher.html">stretcher</a> is a more suitable program to use to
find global alignments of very long sequences.


<H2>
    Author(s)
</H2>


This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>

<pre>
Completed 8th July 1999.
Modified 26th July 1999 - scoring tweaked.
Modified 22 Oct 2000 - %ID and %Similaritty scores added.
</pre>
<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

