<HTML>

<HEAD>
  <TITLE>
  EMBOSS: newcpgreport
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
newcpgreport
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>

<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
Identify CpG islands in nucleotide sequence(s)

<H2>
    Description
</H2>

<p><b>newcpgreport</b> identifies CpG islands in one or more
nucleotide sequences. The ratio of observered to expected number of GC
dinucleotides patterns is calculated over a window (sequence region)
which is moved along the sequence. The calculated ratios are used to
identify regions which match the program's definition of a "CpG
island" (a CG dinucleotide rich area).  A report file is written
giving the input sequence name, CpG island parameters and data on any
CpG islands that are found.</p>


<H2>
Algorithm
</H2>

<p>The ratio of observered to expected number of GC dinucleotides
patterns is calculated over a window of user-specified size
(<tt>-window</tt> parameter).  The window is slid along the sequence
and the ratio recalculated until the end of the sequence is
reached.</p>

<p>By default, <b>cpgplot</b> defines a CpG island as a region where,
over an average of 10 windows and not less than 200 bases, the
calculated (%G + %C) content is over 50% and the calculated
Observed/Expected ratio is over 0.6. These conditions can be modified
by setting the values of the appropriate parameters.</p>

<p>The Observed number of CpG patterns in a window is simply the
number of times a <tt>C</tt> is found followed immediately by
a <tt>G</tt>.</p>

<p>The Expected number of CpG patterns is calculated for each window
as the number of CpG dinucleotides you would expect to see in that
window based on the frequency of <tt>C</tt>'s and <tt>G</tt>'s in that
window. Thus, the Expected frequency of CpG's in a window is
calculated as the number of <tt>C</tt>s in the window multiplied by
the number of <tt>G</tt>s in the window, divided by the window length.

<pre>
	Expected = (number of C's  * number of G's) / window length
</pre>
</p>



<H2>
    Usage
</H2>
<b>Here is a sample session with newcpgreport</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>newcpgreport </b>
Identify CpG islands in nucleotide sequence(s)
Input nucleotide sequence(s): <b>tembl:u68037</b>
Window size [100]: <b></b>
Shift increment [1]: <b></b>
Minimum Length [200]: <b></b>
Minimum observed/expected [0.6]: <b></b>
Minimum percentage [50.]: <b></b>
Output file [u68037.newcpgreport]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>


<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
Identify CpG islands in nucleotide sequence(s)
Version: EMBOSS:6.2.0

   Standard (Mandatory) qualifiers:
  [-sequence]          seqall     Nucleotide sequence(s) filename and optional
                                  format, or reference (input USA)
   -window             integer    [100] Window size (Integer 1 or more)
   -shift              integer    [1] Shift increment (Integer 1 or more)
   -minlen             integer    [200] Minimum Length (Integer 1 or more)
   -minoe              float      [0.6] Minimum observed/expected (Number from
                                  0.000 to 10.000)
   -minpc              float      [50.] Minimum percentage (Number from 0.000
                                  to 100.000)
  [-outfile]           outfile    [*.newcpgreport] Output file name

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1            integer    Start of each sequence to be used
   -send1              integer    End of each sequence to be used
   -sreverse1          boolean    Reverse (if DNA)
   -sask1              boolean    Ask for begin/end/reverse
   -snucleotide1       boolean    Sequence is nucleotide
   -sprotein1          boolean    Sequence is protein
   -slower1            boolean    Make lower case
   -supper1            boolean    Make upper case
   -sformat1           string     Input sequence format
   -sdbname1           string     Database name
   -sid1               string     Entryname
   -ufo1               string     UFO features
   -fformat1           string     Features format
   -fopenfile1         string     Features file name

   "-outfile" associated qualifiers
   -odirectory2        string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-sequence]<br>(Parameter 1)</td>
<td>seqall</td>
<td>Nucleotide sequence(s) filename and optional format, or reference (input USA)</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-window</td>
<td>integer</td>
<td>Window size</td>
<td>Integer 1 or more</td>
<td>100</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-shift</td>
<td>integer</td>
<td>Shift increment</td>
<td>Integer 1 or more</td>
<td>1</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-minlen</td>
<td>integer</td>
<td>Minimum Length</td>
<td>Integer 1 or more</td>
<td>200</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-minoe</td>
<td>float</td>
<td>Minimum observed/expected</td>
<td>Number from 0.000 to 10.000</td>
<td>0.6</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-minpc</td>
<td>float</td>
<td>Minimum percentage</td>
<td>Number from 0.000 to 100.000</td>
<td>50.</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-outfile]<br>(Parameter 2)</td>
<td>outfile</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;*&gt;</i>.newcpgreport</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-sequence" associated seqall qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sbegin1<br>-sbegin_sequence</td>
<td>integer</td>
<td>Start of each sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -send1<br>-send_sequence</td>
<td>integer</td>
<td>End of each sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sreverse1<br>-sreverse_sequence</td>
<td>boolean</td>
<td>Reverse (if DNA)</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sask1<br>-sask_sequence</td>
<td>boolean</td>
<td>Ask for begin/end/reverse</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -snucleotide1<br>-snucleotide_sequence</td>
<td>boolean</td>
<td>Sequence is nucleotide</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sprotein1<br>-sprotein_sequence</td>
<td>boolean</td>
<td>Sequence is protein</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -slower1<br>-slower_sequence</td>
<td>boolean</td>
<td>Make lower case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -supper1<br>-supper_sequence</td>
<td>boolean</td>
<td>Make upper case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sformat1<br>-sformat_sequence</td>
<td>string</td>
<td>Input sequence format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sdbname1<br>-sdbname_sequence</td>
<td>string</td>
<td>Database name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sid1<br>-sid_sequence</td>
<td>string</td>
<td>Entryname</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ufo1<br>-ufo_sequence</td>
<td>string</td>
<td>UFO features</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fformat1<br>-fformat_sequence</td>
<td>string</td>
<td>Features format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fopenfile1<br>-fopenfile_sequence</td>
<td>string</td>
<td>Features file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-outfile" associated outfile qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -odirectory2<br>-odirectory_outfile</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>


<H2>
    Input file format
</H2>


<b>newcpgreport</b> reads one or more nucleic acid sequences.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:u68037' is a sequence entry in the example nucleic acid database 'tembl'
<p>
<p><h3>Database entry: tembl:u68037</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   U68037; SV 1; linear; mRNA; STD; ROD; 1218 BP.
XX
AC   U68037;
XX
DT   23-SEP-1996 (Rel. 49, Created)
DT   04-MAR-2000 (Rel. 63, Last updated, Version 2)
XX
DE   Rattus norvegicus EP1 prostanoid receptor mRNA, complete cds.
XX
KW   .
XX
OS   Rattus norvegicus (Norway rat)
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia;
OC   Eutheria; Euarchontoglires; Glires; Rodentia; Sciurognathi; Muroidea;
OC   Muridae; Murinae; Rattus.
XX
RN   [1]
RP   1-1218
RA   Abramovitz M., Boie Y.;
RT   "Cloning of the rat EP1 prostanoid receptor";
RL   Unpublished.
XX
RN   [2]
RP   1-1218
RA   Abramovitz M., Boie Y.;
RT   ;
RL   Submitted (26-AUG-1996) to the EMBL/GenBank/DDBJ databases.
RL   Biochemistry &amp; Molecular Biology, Merck Frosst Center for Therapeutic
RL   Research, P. O. Box 1005, Pointe Claire - Dorval, Quebec H9R 4P8, Canada
XX
DR   ASTD; TRAN00000010928.
DR   Ensembl-Gn; ENSRNOG00000004094; Rattus_norvegicus.
DR   Ensembl-Tr; ENSRNOT00000005470; Rattus_norvegicus.
XX
FH   Key             Location/Qualifiers
FH
FT   source          1..1218
FT                   /organism="Rattus norvegicus"
FT                   /strain="Sprague-Dawley"
FT                   /mol_type="mRNA"
FT                   /db_xref="taxon:10116"
FT   CDS             1..1218
FT                   /codon_start=1
FT                   /product="EP1 prostanoid receptor"
FT                   /note="family 1 G-protein coupled receptor"
FT                   /db_xref="GOA:P70597"
FT                   /db_xref="InterPro:IPR017452"
FT                   /db_xref="UniProtKB/Swiss-Prot:P70597"
FT                   /protein_id="AAB07735.1"
FT                   /translation="MSPYGLNLSLVDEATTCVTPRVPNTSVVLPTGGNGTSPALPIFSM
FT                   TLGAVSNVLALALLAQVAGRLRRRRSTATFLLFVASLLAIDLAGHVIPGALVLRLYTAG
FT                   RAPAGGACHFLGGCMVFFGLCPLLLGCGMAVERCVGVTQPLIHAARVSVARARLALALL
FT                   AAMALAVALLPLVHVGHYELQYPGTWCFISLGPPGGWRQALLAGLFAGLGLAALLAALV
FT                   CNTLSGLALLRARWRRRRSRRFRENAGPDDRRRWGSRGLRLASASSASSITSTTAALRS
FT                   SRGGGSARRVHAHDVEMVGQLVGIMVVSCICWSPLLVLVVLAIGGWNSNSLQRPLFLAV
FT                   RLASWNQILDPWVYILLRQAMLRQLLRLLPLRVSAKGGPTELSLTKSAWEASSLRSSRH
FT                   SGFSHL"
XX
SQ   Sequence 1218 BP; 162 A; 397 C; 387 G; 272 T; 0 other;
     atgagcccct acgggcttaa cctgagccta gtggatgagg caacaacgtg tgtaacaccc        60
     agggtcccca atacatctgt ggtgctgcca acaggcggta acggcacatc accagcgctg       120
     cctatcttct ccatgacgct gggtgctgtg tccaacgtgc tggcgctggc gctgctggcc       180
     caggttgcag gcagactgcg gcgccgccgc tcgactgcca ccttcctgtt gttcgtcgcc       240
     agcctgcttg ccatcgacct agcaggccat gtgatcccgg gcgccttggt gcttcgcctg       300
     tatactgcag gacgtgcgcc cgctggcggg gcctgtcatt tcctgggcgg ctgtatggtc       360
     ttctttggcc tgtgcccact tttgcttggc tgtggcatgg ccgtggagcg ctgcgtgggt       420
     gtcacgcagc cgctgatcca cgcggcgcgc gtgtccgtag cccgcgcacg cctggcacta       480
     gccctgctgg ccgccatggc tttggcagtg gcgctgctgc cactagtgca cgtgggtcac       540
     tacgagctac agtaccctgg cacttggtgt ttcattagcc ttgggcctcc tggaggttgg       600
     cgccaggcgt tgcttgcggg cctcttcgcc ggccttggcc tggctgcgct ccttgccgca       660
     ctagtgtgta atacgctcag cggcctggcg ctccttcgtg cccgctggag gcggcgtcgc       720
     tctcgacgtt tccgagagaa cgcaggtccc gatgatcgcc ggcgctgggg gtcccgtgga       780
     ctccgcttgg cctccgcctc gtctgcgtca tccatcactt caaccacagc tgccctccgc       840
     agctctcggg gaggcggctc cgcgcgcagg gttcacgcac acgacgtgga aatggtgggc       900
     cagctcgtgg gcatcatggt ggtgtcgtgc atctgctgga gccccctgct ggtattggtg       960
     gtgttggcca tcgggggctg gaactctaac tccctgcagc ggccgctctt tctggctgta      1020
     cgcctcgcgt cgtggaacca gatcctggac ccatgggtgt acatcctgct gcgccaggct      1080
     atgctgcgcc aacttcttcg cctcctaccc ctgagggtta gtgccaaggg tggtccaacg      1140
     gagctgagcc taaccaagag tgcctgggag gccagttcac tgcgtagctc ccggcacagt      1200
     ggcttcagcc acttgtga                                                    1218
//
</pre>
</td></tr></table><p>

<H2>
    Output file format
</H2>



<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: u68037.newcpgreport</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
ID   U68037  1218 BP.
XX
DE   CpG Island report.
XX
CC   Obs/Exp ratio &gt; 0.60.
CC   % C + % G &gt; 50.00.
CC   Length &gt; 200.
XX
FH   Key              Location/Qualifiers
FT   CpG island       104..509
FT                    /size=406
FT                    /Sum C+G=269
FT                    /Percent CG=66.26
FT                    /ObsExp=0.81
FT   CpG island       596..924
FT                    /size=329
FT                    /Sum C+G=223
FT                    /Percent CG=67.78
FT                    /ObsExp=1.01
FT   numislands       2
//
</pre>
</td></tr></table><p>


<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

<p>"CpG" refers to a <tt>C</tt> nucleotide immediately followed by
a <tt>G</tt>. The 'p' in 'CpG' refers to the phosphate group linking
the two bases. Regions of genomic sequences rich in the CpG pattern or
"CpG islands" are resistant to methylation and tend to be associated
with genes which are frequently switched on. It's been estimated that
about half of all mammalian genes, and, possibly all mammalian
house-keeping genes, have a CpG-rich region around their 5' end.
Non-mammalian vertebrates have some CpG islands that are associated
with genes, but the association gets equivocal in the farther
taxonomic groups. The detection of CpG island upstream of predicted
exons or genes is evidence in support of a highly expressed gene.</p>

<p><b>newcpgreport</b> is used in the production of the CpG Island
database 'CPGISLE'. It produces a report for potential CpG islands in
CPGISLE database entry format. See the FTP
site: <a href="ftp://ftp.ebi.ac.uk/pub/databases/cpgisle/">
ftp://ftp.ebi.ac.uk/pub/databases/cpgisle/</a>
for the finished database.</p>



<p>As there is no official definition of what is a CpG island is or
how to identify where they begin and end, we work with two definitions
and thus two methods. These are:</p>

<p>1. <b>cpgplot</b> and <b>newcpgreport</b> use a sliding window
within which the Observed/Expected ratio of CpG is calculated. For a
sequence region to reported as a CpG island, it must satisfy the
following contraints:

<pre>  Observed/Expected ratio > 0.6
   % C + % G > 50%
   Sequence Length > 200</pre>
</p>

<p>2. <b>newcpgseek</b> and <b>cpgreport</b> use a running sum
calculated from all positions in a sequence rather than a window to
produce a score.  If there is not a <tt>CG</tt> dinucleotide at a
position, the score is decremented, if there is one, the score is
incremented by a constant (user-defined) value. If the score for a
region in the sequence is higher than a threshold (17 at the moment)
then a putative island is declared.  Sequence regions scoring above
the threshold are searched for recursively.</p>

<p>This method overpredicts islands but finds the smaller ones around
primary exons. <b>newcpgseek</b> uses the same method
as <b>cpgreport</b> but the output is different and more readable. For
most purposes you should probably use <b>newcpgreport</b> rather
than <b>cpgreport</b>. It is used to produce the human cpgisland
database you can find on the EBI's ftp server as well as on the EBI's
SRS server.</p>

<p><b>newcpgseek</b> and <b>cpgreport</b> both now display the actual
CpG count, the (%C + %G) and the Observed/Expected ratio in the region
where the score is above the threshold.</p>

<p>The <b>geecee</b> program measures CG content in the entire input
sequence and is not to be used to detect CpG islands. It can be useful
for detecting sequences that MIGHT contain an island.</p>





<H2>
    References
</H2>


<ol>

<li>
Larsen F., Gundersen, G., Lopez L., Prydz H.
"CpG island as Gene Markers in the Human Genome"
Genomics  13:1095-1107 (1992)

</ol>

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="cpgplot.html">cpgplot</a></td>
<td>Identify and plot CpG islands in nucleotide sequence(s)</td>
</tr>

<tr>
<td><a href="cpgreport.html">cpgreport</a></td>
<td>Identify and report CpG-rich regions in nucleotide sequence(s)</td>
</tr>

<tr>
<td><a href="geecee.html">geecee</a></td>
<td>Calculate fractional GC content of nucleic acid sequences</td>
</tr>

<tr>
<td><a href="newcpgseek.html">newcpgseek</a></td>
<td>Identify and report CpG-rich regions in nucleotide sequence(s)</td>
</tr>

</table>

<H2>
    Author(s)
</H2>


Rodrigo Lopez (rls&nbsp;&copy;&nbsp;ebi.ac.uk)
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK


<H2>
    History
</H2>

Written (1999) - Rodrigo Lopez.

<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.


<H2>
    Comments
</H2>
None

</BODY>
</HTML>

