<HTML>

<HEAD>
  <TITLE>
  EMBOSS: newcpgreport
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: newcpgreport
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program newcpgreport
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Report CpG rich areas


<H2>
    Description
</H2>


This application is used in the production of the CpG Island database
'CPGISLE'.  It produces CPGISLE database entry format reports for a
potential CpG island. 

<p>
See the FTP site: <a
href="ftp://ftp.ebi.ac.uk/pub/databases/cpgisle/">ftp://ftp.ebi.ac.uk/pub/databases/cpgisle/</a>
for the finished database.

<H2>
    Usage
</H2>


Here is a sample session with newcpgreport.

<pre>

% newcpgreport
Input sequence: embl:rnu68037
Window size [100]: 
Shift increment [1]: 
Minimum Length [200]: 
Minimum observed/expected [0.6]: 
Minimum percentage [50.]: 
Output file [rnu68037.newcpgreport]: 

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
   -window             integer    Window size
   -shift              integer    Shift increment
   -minlen             integer    Minimum Length
   -minoe              float      Minimum observed/expected
   -minpc              float      Minimum percentage
  [-outfile]           outfile    Output file name

   Optional qualifiers: (none)
   Advanced qualifiers:
   -[no]obsexp         bool       Show observed/expected threshold line
   -[no]cg             bool       Show CpG rich regions
   -[no]pc             bool       Show percentage line


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-window</td>
<td>Window size</td>
<td>Integer 1 or more</td>
<td>100</td>
</tr>

<tr>
<td>-shift</td>
<td>Shift increment</td>
<td>Integer 1 or more</td>
<td>1</td>
</tr>

<tr>
<td>-minlen</td>
<td>Minimum Length</td>
<td>Integer 1 or more</td>
<td>200</td>
</tr>

<tr>
<td>-minoe</td>
<td>Minimum observed/expected</td>
<td>Number from 0.000 to 10.000</td>
<td>0.6</td>
</tr>

<tr>
<td>-minpc</td>
<td>Minimum percentage</td>
<td>Number from 0.000 to 100.000</td>
<td>50.</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.newcpgreport</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]obsexp</td>
<td>Show observed/expected threshold line</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]cg</td>
<td>Show CpG rich regions</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]pc</td>
<td>Show percentage line</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

</table>



<H2>
    Input file format
</H2>


<H2>
    Output file format
</H2>


Here is the output file from the example run:

<pre>

ID   RNU68037  1118 BP.
XX
DE   CpG Island report.
XX
CC   Obs/Exp ratio &gt; 0.60.
CC   % C + % G &gt; 50.00.
CC   Length &gt; 200.
XX
FH   Key              Location/Qualifiers
FT   CpG island       157..389
FT                    /size=232
FT                    /Sum C+G=152
FT                    /Percent CG=65.24
FT                    /ObsExp=0.73
FT   CpG island       654..963
FT                    /size=309
FT                    /Sum C+G=206
FT                    /Percent CG=66.45
FT                    /ObsExp=0.96
FT   numislands       2
//

</pre>

<H2>
    Data files
</H2>


<H2>
    Notes
</H2>


<H2>
    References
</H2>


<ol>

<li>
Larsen F., Gundersen, G., Lopez L., Prydz H.
CpG island as Gene Markers in the Human Genome
Genomics  13:1095-1107 (1992)

</ol>

<H2>
    Warnings
</H2>


<H2>
    Diagnostic Error Messages
</H2>


<H2>
    Exit status
</H2>


<H2>
    Known bugs
</H2>


<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="chaos.html">chaos</a></td><td>Create a chaos game representation plot for a sequence</td></tr>
<tr><td><a href="chips.html">chips</a></td><td>Codon usage statistics</td></tr>
<tr><td><a href="codcmp.html">codcmp</a></td><td>Codon usage table comparison</td></tr>
<tr><td><a href="compseq.html">compseq</a></td><td>Counts the composition of dimer/trimer/etc words in a sequence</td></tr>
<tr><td><a href="cpgplot.html">cpgplot</a></td><td>Plot CpG rich areas</td></tr>
<tr><td><a href="cpgreport.html">cpgreport</a></td><td>Reports all CpG rich regions</td></tr>
<tr><td><a href="cusp.html">cusp</a></td><td>Create a codon usage table</td></tr>
<tr><td><a href="diffseq.html">diffseq</a></td><td>Find differences (SNPs) between nearly identical sequences</td></tr>
<tr><td><a href="dotmatcher.html">dotmatcher</a></td><td>Displays a thresholded dotplot of two sequences</td></tr>
<tr><td><a href="dotpath.html">dotpath</a></td><td>Displays a non-overlapping wordmatch dotplot of two sequences</td></tr>
<tr><td><a href="dottup.html">dottup</a></td><td>Displays a wordmatch dotplot of two sequences</td></tr>
<tr><td><a href="einverted.html">einverted</a></td><td>Finds DNA inverted repeats</td></tr>
<tr><td><a href="equicktandem.html">equicktandem</a></td><td>Finds tandem repeats</td></tr>
<tr><td><a href="etandem.html">etandem</a></td><td>Looks for tandem repeats in a nucleotide sequence</td></tr>
<tr><td><a href="freak.html">freak</a></td><td>Residue/base frequency table or plot</td></tr>
<tr><td><a href="geecee.html">geecee</a></td><td>Calculates the fractional GC content of nucleic acid sequences</td></tr>
<tr><td><a href="isochore.html">isochore</a></td><td>Plots isochores in large DNA sequences</td></tr>
<tr><td><a href="newcpgseek.html">newcpgseek</a></td><td>Reports CpG rich regions</td></tr>
<tr><td><a href="palindrome.html">palindrome</a></td><td>Looks for inverted repeats in a nucleotide sequence</td></tr>
<tr><td><a href="polydot.html">polydot</a></td><td>Displays all-against-all dotplots of a set of sequences</td></tr>
<tr><td><a href="redata.html">redata</a></td><td>Search REBASE for enzyme name, references, suppliers etc</td></tr>
<tr><td><a href="restrict.html">restrict</a></td><td>Finds restriction enzyme cleavage sites</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="silent.html">silent</a></td><td>Silent mutation restriction enzyme scan</td></tr>
<tr><td><a href="tfscan.html">tfscan</a></td><td>Scans DNA sequences for transcription factors</td></tr>
<tr><td><a href="wobble.html">wobble</a></td><td>Wobble base plot</td></tr>
<tr><td><a href="wordcount.html">wordcount</a></td><td>Counts words of a specified size in a DNA sequence</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->




<H2>
    Author(s)
</H2>


This application was written by Rodrigo Lopez <a
href="mailto:rls@ebi.ac.uk">(rls@ebi.ac.uk)</a> European
Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SD, UK.

<H2>
    History
</H2>


<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

