<HTML>

<HEAD>
  <TITLE>
  EMBOSS: newcpgreport
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">




<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

newcpgreport
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>




<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Report CpG rich areas


<H2>
    Description
</H2>


This application is used in the production of the CpG Island database
'CPGISLE'.  It produces CPGISLE database entry format reports for a
potential CpG island. 

<p>
See the FTP site: <a
href="ftp://ftp.ebi.ac.uk/pub/databases/cpgisle/">ftp://ftp.ebi.ac.uk/pub/databases/cpgisle/</a>
for the finished database.

<H2>
    Usage
</H2>


Here is a sample session with newcpgreport.

<pre>

% newcpgreport
Input sequence: embl:rnu68037
Window size [100]: 
Shift increment [1]: 
Minimum Length [200]: 
Minimum observed/expected [0.6]: 
Minimum percentage [50.]: 
Output file [rnu68037.newcpgreport]: 

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
   -window             integer    Window size
   -shift              integer    Shift increment
   -minlen             integer    Minimum Length
   -minoe              float      Minimum observed/expected
   -minpc              float      Minimum percentage
  [-outfile]           outfile    Output file name

   Optional qualifiers: (none)
   Advanced qualifiers: (none)
   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-window</td>
<td>Window size</td>
<td>Integer 1 or more</td>
<td>100</td>
</tr>

<tr>
<td>-shift</td>
<td>Shift increment</td>
<td>Integer 1 or more</td>
<td>1</td>
</tr>

<tr>
<td>-minlen</td>
<td>Minimum Length</td>
<td>Integer 1 or more</td>
<td>200</td>
</tr>

<tr>
<td>-minoe</td>
<td>Minimum observed/expected</td>
<td>Number from 0.000 to 10.000</td>
<td>0.6</td>
</tr>

<tr>
<td>-minpc</td>
<td>Minimum percentage</td>
<td>Number from 0.000 to 100.000</td>
<td>50.</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.newcpgreport</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>


One or more nucleic acid sequences.

<H2>
    Output file format
</H2>


Here is the output file from the example run:

<p>

<hr>
<pre>
ID   RNU68037  1118 BP.
XX
DE   CpG Island report.
XX
CC   Obs/Exp ratio &gt; 0.60.
CC   % C + % G &gt; 50.00.
CC   Length &gt; 200.
XX
FH   Key              Location/Qualifiers
FT   CpG island       157..389
FT                    /size=232
FT                    /Sum C+G=152
FT                    /Percent CG=65.24
FT                    /ObsExp=0.73
FT   CpG island       654..963
FT                    /size=309
FT                    /Sum C+G=206
FT                    /Percent CG=66.45
FT                    /ObsExp=0.96
FT   numislands       2
//
</pre>
<hr>

<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>


<ol>

<li>
Larsen F., Gundersen, G., Lopez L., Prydz H.
CpG island as Gene Markers in the Human Genome
Genomics  13:1095-1107 (1992)

</ol>

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="cpgplot.html">cpgplot</a></td><td>Plot CpG rich areas</td></tr>
<tr><td><a href="cpgreport.html">cpgreport</a></td><td>Reports all CpG rich regions</td></tr>
<tr><td><a href="geecee.html">geecee</a></td><td>Calculates the fractional GC content of nucleic acid sequences</td></tr>
<tr><td><a href="newcpgseek.html">newcpgseek</a></td><td>Reports CpG rich regions</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->




<H2>
    Author(s)
</H2>


This application was written by Rodrigo Lopez <a
href="mailto:rls@ebi.ac.uk">(rls@ebi.ac.uk)</a> European
Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SD, UK.

<H2>
    History
</H2>

Written (1999) - Rodrigo Lopez.

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

