<HTML>

<HEAD>
  <TITLE>
  EMBOSS: newcpgreport
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
newcpgreport
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>

<H2>
    Function
</H2>
Report CpG rich areas

<H2>
    Description
</H2>


This application is used in the production of the CpG Island database
'CPGISLE'.  It produces CPGISLE database entry format reports for a
potential CpG island. 

<p>
See the FTP site: <a
href="ftp://ftp.ebi.ac.uk/pub/databases/cpgisle/">ftp://ftp.ebi.ac.uk/pub/databases/cpgisle/</a>
for the finished database.

<p>

CpG refers to a C nucleotide immediately followed by a G.  The 'p' in
'CpG' refers to the phosphate group linking the two bases.

<p>
Detection of regions of genomic sequences that are rich in the CpG
pattern is important because such regions are resistant to methylation
and tend to be associated with genes which are frequently switched on.
Regions rich in the CpG pattern are known as CpG islands.

<p>
It has been estimated that about half of all mammalian genes have a
CpG-rich region around their 5' end.  It is said that all mammalian
house-keeping genes have a CpG island!

<p>
Non-mammalian vertebrates have some CpG islands that are associated with
genes, but the association gets equivocal in the farther taxonomic
groups.

<p>
Finding a CpG island upstream of predicted exons or genes is good
contributory evidence for that gene's existance.

<p>
By default, this program defines a CpG island as a region where, over an
average of 10 windows, the calculated % composition is over 50% and the
calculated Obs/Exp ratio is over 0.6 and the conditions hold for a
minimum of 200 bases.  These conditions can be modified by setting the
values of the appropriate parameters.

<p>
The Expected number of CpG patterns in a window is calculated as the
number of 'C's in the window multiplied by the number of 'G's in the
window, divided by the window length.
                                          
<p>

This program reads in one or more sequences and finds regions where
there is a high absolute frequency of CpG dimers as well as a high
proportion of CpG compared to GpC.  

<H2>
    Usage
</H2>
<b>Here is a sample session with newcpgreport</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>newcpgreport </b>
Report CpG rich areas
Input nucleotide sequence(s): <b>tembl:rnu68037</b>
Window size [100]: <b></b>
Shift increment [1]: <b></b>
Minimum Length [200]: <b></b>
Minimum observed/expected [0.6]: <b></b>
Minimum percentage [50.]: <b></b>
Output file [rnu68037.newcpgreport]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>


<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-sequence]          seqall     Nucleotide sequence(s) filename and optional
                                  format, or reference (input USA)
   -window             integer    [100] Window size (Integer 1 or more)
   -shift              integer    [1] Shift increment (Integer 1 or more)
   -minlen             integer    [200] Minimum Length (Integer 1 or more)
   -minoe              float      [0.6] Minimum observed/expected (Number from
                                  0.000 to 10.000)
   -minpc              float      [50.] Minimum percentage (Number from 0.000
                                  to 100.000)
  [-outfile]           outfile    [*.newcpgreport] Output file name

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1            integer    Start of each sequence to be used
   -send1              integer    End of each sequence to be used
   -sreverse1          boolean    Reverse (if DNA)
   -sask1              boolean    Ask for begin/end/reverse
   -snucleotide1       boolean    Sequence is nucleotide
   -sprotein1          boolean    Sequence is protein
   -slower1            boolean    Make lower case
   -supper1            boolean    Make upper case
   -sformat1           string     Input sequence format
   -sdbname1           string     Database name
   -sid1               string     Entryname
   -ufo1               string     UFO features
   -fformat1           string     Features format
   -fopenfile1         string     Features file name

   "-outfile" associated qualifiers
   -odirectory2        string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages

</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Nucleotide sequence(s) filename and optional format, or reference (input USA)</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-window</td>
<td>Window size</td>
<td>Integer 1 or more</td>
<td>100</td>
</tr>

<tr>
<td>-shift</td>
<td>Shift increment</td>
<td>Integer 1 or more</td>
<td>1</td>
</tr>

<tr>
<td>-minlen</td>
<td>Minimum Length</td>
<td>Integer 1 or more</td>
<td>200</td>
</tr>

<tr>
<td>-minoe</td>
<td>Minimum observed/expected</td>
<td>Number from 0.000 to 10.000</td>
<td>0.6</td>
</tr>

<tr>
<td>-minpc</td>
<td>Minimum percentage</td>
<td>Number from 0.000 to 100.000</td>
<td>50.</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<H2>
    Input file format
</H2>


<b>newcpgreport</b> reads one or more nucleic acid sequences.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:rnu68037' is a sequence entry in the example nucleic acid database 'tembl'
<p>
<p><h3>Database entry: tembl:rnu68037</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   RNU68037   standard; RNA; ROD; 1218 BP.
XX
AC   U68037;
XX
SV   U68037.1
XX
DT   23-SEP-1996 (Rel. 49, Created)
DT   04-MAR-2000 (Rel. 63, Last updated, Version 2)
XX
DE   Rattus norvegicus EP1 prostanoid receptor mRNA, complete cds.
XX
KW   .
XX
OS   Rattus norvegicus (Norway rat)
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia;
OC   Eutheria; Rodentia; Sciurognathi; Muridae; Murinae; Rattus.
XX
RN   [1]
RP   1-1218
RA   Abramovitz M., Boie Y.;
RT   "Cloning of the rat EP1 prostanoid receptor";
RL   Unpublished.
XX
RN   [2]
RP   1-1218
RA   Abramovitz M., Boie Y.;
RT   ;
RL   Submitted (26-AUG-1996) to the EMBL/GenBank/DDBJ databases.
RL   Biochemistry &amp; Molecular Biology, Merck Frosst Center for Therapeutic
RL   Research, P. O. Box 1005, Pointe Claire - Dorval, Quebec H9R 4P8, Canada
XX
DR   SWISS-PROT; P70597; PE21_RAT.
XX
FH   Key             Location/Qualifiers
FH
FT   source          1..1218
FT                   /db_xref="taxon:10116"
FT                   /organism="Rattus norvegicus"
FT                   /strain="Sprague-Dawley"
FT   CDS             1..1218
FT                   /codon_start=1
FT                   /db_xref="SWISS-PROT:P70597"
FT                   /note="family 1 G-protein coupled receptor"
FT                   /product="EP1 prostanoid receptor"
FT                   /protein_id="AAB07735.1"
FT                   /translation="MSPYGLNLSLVDEATTCVTPRVPNTSVVLPTGGNGTSPALPIFSM
FT                   TLGAVSNVLALALLAQVAGRLRRRRSTATFLLFVASLLAIDLAGHVIPGALVLRLYTAG
FT                   RAPAGGACHFLGGCMVFFGLCPLLLGCGMAVERCVGVTQPLIHAARVSVARARLALALL
FT                   AAMALAVALLPLVHVGHYELQYPGTWCFISLGPPGGWRQALLAGLFAGLGLAALLAALV
FT                   CNTLSGLALLRARWRRRRSRRFRENAGPDDRRRWGSRGLRLASASSASSITSTTAALRS
FT                   SRGGGSARRVHAHDVEMVGQLVGIMVVSCICWSPLLVLVVLAIGGWNSNSLQRPLFLAV
FT                   RLASWNQILDPWVYILLRQAMLRQLLRLLPLRVSAKGGPTELSLTKSAWEASSLRSSRH
FT                   SGFSHL"
XX
SQ   Sequence 1218 BP; 162 A; 397 C; 387 G; 272 T; 0 other;
     atgagcccct acgggcttaa cctgagccta gtggatgagg caacaacgtg tgtaacaccc        60
     agggtcccca atacatctgt ggtgctgcca acaggcggta acggcacatc accagcgctg       120
     cctatcttct ccatgacgct gggtgctgtg tccaacgtgc tggcgctggc gctgctggcc       180
     caggttgcag gcagactgcg gcgccgccgc tcgactgcca ccttcctgtt gttcgtcgcc       240
     agcctgcttg ccatcgacct agcaggccat gtgatcccgg gcgccttggt gcttcgcctg       300
     tatactgcag gacgtgcgcc cgctggcggg gcctgtcatt tcctgggcgg ctgtatggtc       360
     ttctttggcc tgtgcccact tttgcttggc tgtggcatgg ccgtggagcg ctgcgtgggt       420
     gtcacgcagc cgctgatcca cgcggcgcgc gtgtccgtag cccgcgcacg cctggcacta       480
     gccctgctgg ccgccatggc tttggcagtg gcgctgctgc cactagtgca cgtgggtcac       540
     tacgagctac agtaccctgg cacttggtgt ttcattagcc ttgggcctcc tggaggttgg       600
     cgccaggcgt tgcttgcggg cctcttcgcc ggccttggcc tggctgcgct ccttgccgca       660
     ctagtgtgta atacgctcag cggcctggcg ctccttcgtg cccgctggag gcggcgtcgc       720
     tctcgacgtt tccgagagaa cgcaggtccc gatgatcgcc ggcgctgggg gtcccgtgga       780
     ctccgcttgg cctccgcctc gtctgcgtca tccatcactt caaccacagc tgccctccgc       840
     agctctcggg gaggcggctc cgcgcgcagg gttcacgcac acgacgtgga aatggtgggc       900
     cagctcgtgg gcatcatggt ggtgtcgtgc atctgctgga gccccctgct ggtattggtg       960
     gtgttggcca tcgggggctg gaactctaac tccctgcagc ggccgctctt tctggctgta      1020
     cgcctcgcgt cgtggaacca gatcctggac ccatgggtgt acatcctgct gcgccaggct      1080
     atgctgcgcc aacttcttcg cctcctaccc ctgagggtta gtgccaaggg tggtccaacg      1140
     gagctgagcc taaccaagag tgcctgggag gccagttcac tgcgtagctc ccggcacagt      1200
     ggcttcagcc acttgtga                                                    1218
//
</pre>
</td></tr></table><p>

<H2>
    Output file format
</H2>



<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: rnu68037.newcpgreport</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
ID   RNU68037  1218 BP.
XX
DE   CpG Island report.
XX
CC   Obs/Exp ratio &gt; 0.60.
CC   % C + % G &gt; 50.00.
CC   Length &gt; 200.
XX
FH   Key              Location/Qualifiers
FT   CpG island       104..509
FT                    /size=406
FT                    /Sum C+G=269
FT                    /Percent CG=66.26
FT                    /ObsExp=0.81
FT   CpG island       596..924
FT                    /size=329
FT                    /Sum C+G=223
FT                    /Percent CG=67.78
FT                    /ObsExp=1.01
FT   numislands       2
//
</pre>
</td></tr></table><p>


<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>


<ol>

<li>
Larsen F., Gundersen, G., Lopez L., Prydz H.
"CpG island as Gene Markers in the Human Genome"
Genomics  13:1095-1107 (1992)

</ol>

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr>
<td><a href="cpgplot.html">cpgplot</a></td>
<td>Plot CpG rich areas</td>
</tr>

<tr>
<td><a href="cpgreport.html">cpgreport</a></td>
<td>Reports all CpG rich regions</td>
</tr>

<tr>
<td><a href="geecee.html">geecee</a></td>
<td>Calculates fractional GC content of nucleic acid sequences</td>
</tr>

<tr>
<td><a href="newcpgseek.html">newcpgseek</a></td>
<td>Reports CpG rich regions</td>
</tr>

</table>

<p>

As there is no official definition of what is a cpg island is, and worst
where they begin and end, we have to live with 2 definitions and thus
two methods.  These are:

<p>

1.  <b>newcpgseek</b> and <b>cpgreport</b> - both declare a putative
island if the score is higher than a threshold (17 at the moment).  They
now also display the actual CpG count, the % CG and the
observed/expected ration in the region where the score is above the
threshold.  This scoring method based on sum/frequencies overpredicts
islands but finds the smaller ones around primary exons. 
<b>newcpgseek</b> uses the same method as <b>cpgreport</b> but the
output is different and more readable. 

<p>

2.  <b>newcpgreport</b> and <b>cpgplot</b> use the method which
mentioned in the Description section above.  The important thing to note
in this method is that an island, in order to be reported, is defined as
a region that satisfies the following contraints:

<p>
<pre>
   Obs/Exp ratio > 0.6
   % C + % G > 50%
   Length > 200.
</pre>
<p>

<p>

For all practical purposes you should probably use newcpgreport.  It is
actually used to produce the human cpgisland database you can find on
the EBI's ftp server as well as on the EBI's SRS server. 

<p>

<b>geecee</b> measures CG content in the entire input sequence and is
not to be used to detect CpG islands.  It can be useful for detecting
sequences that MIGHT contain an island. 

<H2>
    Author(s)
</H2>


Rodrigo Lopez (rls&nbsp;&copy;&nbsp;ebi.ac.uk)
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK


<H2>
    History
</H2>

Written (1999) - Rodrigo Lopez.

<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.


<H2>
    Comments
</H2>
None

</BODY>
</HTML>

