<HTML>

<HEAD>
  <TITLE>
  EMBOSS: newcpgseek
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: newcpgseek
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program newcpgseek
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Reports CpG rich regions


<H2>
    Description
</H2>


<b>newcpgseek</b> reports CpG rich regions of a sequence as candidate
CpG islands. 

<p>

CpG refers to a C nucleotide immediately followed by a G.  The 'p' in
'CpG' refers to the phosphate group linking the two bases. 

<p>
Detection of regions of genomic sequences that are rich in the CpG
pattern is important because such regions are resistant to methylation
and tend to be associated with genes which are frequently switched on. 
Regions rich in the CpG pattern are known as CpG islands.  

<p>
It has been estimated that about half of all mammalian genes have a
CpG-rich region around their 5' end.  It is said that all mammalian
house-keeping genes have a CpG island!

<p>
Non-mammalian vertebrates have some CpG islands that are associated with
genes, but the association gets equivocal in the farther taxonomic
groups. 

<p>
Finding a CpG island upstream of predicted exons or genes is good
contributory evidence. 

<p>
By default, this program defines a CpG island as a region where, over an
average of 10 windows, the calculated % composition is over 50% and the
calculated Obs/Exp ratio is over 0.6 and the conditions hold for a
minimum of 200 bases.  These conditions can be modified by setting the
values of the appropriate parameters. 

<p>
The Expected number of CpG patterns in a window is calculated as the
number of 'C's in the window multiplied by the number of 'G's in the
window, divided by the window length. 

<p>

This program reads in one or more sequences and finds regions where
there is a high absolute frequency of CpG dimers as well as a high
proportion of CpG compared to GpC. 



<H2>
    Usage
</H2>


Here is a sample session with newcpgseek.

<pre>

% newcpgseek
Input sequence: embl:rnu68037
CpG score [17]: 
Output file [rnu68037.newcpgseek]: 

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
   -score              integer    CpG score
  [-outfile]           outfile    Output file name

   Optional qualifiers: (none)
   Advanced qualifiers: (none)

</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-score</td>
<td>CpG score</td>
<td>Integer from 1 to 200</td>
<td>17</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.newcpgseek</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>

A nucleic acid sequence.
	
<H2>
    Output file format
</H2>


Here is the output from the example run:

<p>

<hr>
<pre>
NEWCPGSEEK of RNU68037 from 1 to 1218
with score &gt; 17 

 Begin    End  Score        CpG  %CG  CG/GC
*    96   1032   630         87  66.1   0.65
  1072   1100    26          3  62.1   0.00
  1183   1193    26          2  72.7   2.00
-------------------------------------------
</pre>
<hr>

<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="cpgplot.html">cpgplot</a></td><td>Plot CpG rich areas</td></tr>
<tr><td><a href="cpgreport.html">cpgreport</a></td><td>Reports all CpG rich regions</td></tr>
<tr><td><a href="geecee.html">geecee</a></td><td>Calculates the fractional GC content of nucleic acid sequences</td></tr>
<tr><td><a href="newcpgreport.html">newcpgreport</a></td><td>Report CpG rich areas</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->




<H2>
    Author(s)
</H2>


This application was written by Rodrigo Lopez <a
href="mailto:rls@ebi.ac.uk">(rls@ebi.ac.uk)</a> European
Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SD, UK.

<H2>
    History
</H2>

Written (1999) - Rodrigo Lopez

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

