<HTML>

<HEAD>
  <TITLE>
  EMBOSS: newseq
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">




<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

newseq
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>





<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Type in a short new sequence


<H2>
    Description
</H2>


This allows you to type a sequence into a file in a quick and
easy manner.

<p>
The length of the sequence you can type in is restricted to a
fairly short length (typically less than 255 characters).  This
length restriction is not a property of the EMBOSS package, but of the
computer system you are using.  This is because, as you type in
response to a prompt from this program, what you type is stored in the
computer operating system before being handed over to the program. 
There is often a limit of less than 255 characters on the length of a
response that a computer system will allow you to give. 

<p>
Despite this restriction, it is expected that this program will
be a useful and easy way of constructing new sequence files. 

<p>
(You wouldn't want to type a long sequence in by hand, anyway,
would you?)

<H2>
    Usage
</H2>


<b>Here is a sample session with newseq</b>
<p>
Type in a short sequence to the file 'mycc.pep' in SWISSPROT format: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>newseq </b>
Type in a short new sequence.
Name of the sequence: <b>cytoc</b>
Description of the sequence: <b>fragment of cytochrome c</b>
Type of sequence
         N : Nucleic
         P : Protein
Type of sequence [N]: <b>p</b>
Output sequence [outfile.fasta]: <b>swiss::mycc.pep</b>
Enter the sequence: <b>KKKEERADLIAY</b>

</pre></td></tr></table><p>
<p>
<a href="#output.1">Go to the output files for this example</a><p><p>

	

<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-name]              string     The name of of the sequence should be a
                                  single word that you will use to identify
                                  the sequence. It should have no (or few)
                                  punctuation characters in it.
  [-description]       string     Enter any description of the sequence that
                                  you require.
  [-type]              menu       Type of sequence
  [-outseq]            seqout     Output sequence USA
  [-sequence]          string     The sequence itself.
                                  Because of the limitation of the operating
                                  system, you will only be able to type in a
                                  short sequence of (typically) 250
                                  characters, or so.
                                  The keyboard will beep at you when you have
                                  reached this limit and you will not be able
                                  to press the RETURN/ENTER key until you have
                                  deleted a few characters.

   Optional qualifiers: (none)
   Advanced qualifiers: (none)
   Associated qualifiers:
  "-outseq" related qualifiers
   -osformat4           string     Output seq format
   -osextension4        string     File name extension
   -osname4             string     Base file name
   -osdirectory4        string     Output directory
   -osdbname4           string     Database name to add
   -ossingle4           boolean    Separate file for each entry
   -oufo4               string     UFO features
   -offormat4           string     Features format
   -ofname4             string     Features file name
   -ofdirectory4        string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for required and optional values
   -debug               boolean    Write debug output to program.dbg
   -acdlog              boolean    Write ACD processing log to program.acdlog
   -acdpretty           boolean    Rewrite ACD file as program.acdpretty
   -acdtable            boolean    Write HTML table of options
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-name]<br>(Parameter 1)</td>
<td>The name of of the sequence should be a single word that you will use to identify the sequence. It should have no (or few) punctuation characters in it.</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>[-description]<br>(Parameter 2)</td>
<td>Enter any description of the sequence that you require.</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>[-type]<br>(Parameter 3)</td>
<td>Type of sequence</td>
<td><table><tr><td>N</td> <td><i>(Nucleic)</i></td></tr><tr><td>P</td> <td><i>(Protein)</i></td></tr></table></td>
<td>N</td>
</tr>

<tr>
<td>[-outseq]<br>(Parameter 4)</td>
<td>Output sequence USA</td>
<td>Writeable sequence</td>
<td><i>&lt;sequence&gt;</i>.<i>format</i></td>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 5)</td>
<td>The sequence itself.
Because of the limitation of the operating system, you will only be able to type in a short sequence of (typically) 250 characters, or so.
The keyboard will beep at you when you have reached this limit and you will not be able to press the RETURN/ENTER key until you have deleted a few characters.</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>




<H2>
    Output file format
</H2>


<b>newseq</b> writes a normal sequence file.

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: mycc.pep</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
ID   cytoc          STANDARD;      PRT;    12 AA.
DE   fragment of cytochrome c
SQ   SEQUENCE    12 AA;   1464 MW;  0000000000000000 CRC64;
     KKKEERADLI AY
//
</pre>
</td></tr></table><p>


<H2>
    Data files
</H2>


None.	

<H2>
    Notes
</H2>


The length of the sequence you can type in is restricted to a
fairly short length (typically less than 255 characters). 
If your computer beeps at you, you will typically have to delete
the last character you typed and press the RETURN key.

<H2>
    References
</H2>


None.

<H2>
    Warnings
</H2>


None.

<H2>
    Diagnostic Error Messages
</H2>


None.

<H2>
    Exit status
</H2>


It always exits with status 0.

<H2>
    Known bugs
</H2>


None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="biosed.html">biosed</a></td><td>Replace or delete sequence sections</td></tr>
<tr><td><a href="cutseq.html">cutseq</a></td><td>Removes a specified section from a sequence</td></tr>
<tr><td><a href="degapseq.html">degapseq</a></td><td>Removes gap characters from sequences</td></tr>
<tr><td><a href="descseq.html">descseq</a></td><td>Alter the name or description of a sequence</td></tr>
<tr><td><a href="entret.html">entret</a></td><td>Reads and writes (returns) flatfile entries</td></tr>
<tr><td><a href="extractfeat.html">extractfeat</a></td><td>Extract features from a sequence</td></tr>
<tr><td><a href="extractseq.html">extractseq</a></td><td>Extract regions from a sequence</td></tr>
<tr><td><a href="listor.html">listor</a></td><td>Writes a list file of the logical OR of two sets of sequences</td></tr>
<tr><td><a href="maskfeat.html">maskfeat</a></td><td>Mask off features of a sequence</td></tr>
<tr><td><a href="maskseq.html">maskseq</a></td><td>Mask off regions of a sequence</td></tr>
<tr><td><a href="noreturn.html">noreturn</a></td><td>Removes carriage return from ASCII files</td></tr>
<tr><td><a href="notseq.html">notseq</a></td><td>Excludes a set of sequences and writes out the remaining ones</td></tr>
<tr><td><a href="nthseq.html">nthseq</a></td><td>Writes one sequence from a multiple set of sequences</td></tr>
<tr><td><a href="pasteseq.html">pasteseq</a></td><td>Insert one sequence into another</td></tr>
<tr><td><a href="revseq.html">revseq</a></td><td>Reverse and complement a sequence</td></tr>
<tr><td><a href="seqret.html">seqret</a></td><td>Reads and writes (returns) sequences</td></tr>
<tr><td><a href="seqretsplit.html">seqretsplit</a></td><td>Reads and writes (returns) sequences in individual files</td></tr>
<tr><td><a href="skipseq.html">skipseq</a></td><td>Reads and writes (returns) sequences, skipping the first few</td></tr>
<tr><td><a href="splitter.html">splitter</a></td><td>Split a sequence into (overlapping) smaller sequences</td></tr>
<tr><td><a href="trimest.html">trimest</a></td><td>Trim poly-A tails off EST sequences</td></tr>
<tr><td><a href="trimseq.html">trimseq</a></td><td>Trim ambiguous bits off the ends of sequences</td></tr>
<tr><td><a href="union.html">union</a></td><td>Reads sequence fragments and builds one sequence</td></tr>
<tr><td><a href="vectorstrip.html">vectorstrip</a></td><td>Strips out DNA between a pair of vector sequences</td></tr>
<tr><td><a href="yank.html">yank</a></td><td>Reads a sequence range, appends the full USA to a list file</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->





<H2>
    Author(s)
</H2>


Gary Williams (gwilliam&nbsp;&copy;&nbsp;hgmp.mrc.ac.uk)
<br>
HGMP-RC, Genome Campus, Hinxton, Cambridge CB10 1SB, UK



<H2>
    History
</H2>


Written (1999) - Gary Williams

<!--
<p>
[an error occurred while processing this directive]
-->

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

