<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: nrscope
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">
<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>
<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>
</td>
<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: nrscope
</H2>
</font>
</td>
</tr>
</table>
</td>
</tr>
</table>
<p><hr><p>

<!--END OF HEADER-->

<H1>
Program nrscope
</H1>
<H2>
    Function
</H2>
Converts redundant EMBL-format SCOP file to non-redundant one

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
        Algorithms
  -->

Nearly all proteins have structural similarities with other proteins
and, in some of these cases, share a common evolutionary origin.  A
knowledge of these relationships is crucial to our understanding of the
evolution of proteins and of development.  It will also play an
important role in the analysis of the sequence data that is being
produced by worldwide genome projects. 

<p>

The 
<A HREF="http://scop.mrc-lmb.cam.ac.uk/scop/">SCOP</A> 
database aims to provide a detailed and comprehensive
description of the structural and evolutionary relationships between all
proteins whose structure is known, including all entries in the Protein
Data Bank 
(<A HREF="http://pdb-browsers.ebi.ac.uk//">PDB</A>).  

<p>

<b>nrscope</b> reads in the EMBL-like format SCOP classification file
generated by the EMBOSS application 
<a href="scope.html"><b>scope</b></a>, 
and writes a file of non-redundant domains in the same format.  Domain
sequences are extracted from the clean domain coordinate files generated
by the EMBOSS application
<a href="domainer.html"><b>domainer</b></a>. 

<p>


The current version of <b>nrscope</b> removes redundancy at the level of
the SCOP family, i.e.  entries belonging to the same family will be
non-redundant.  All permutations of pair-wise sequence alignments are
calculated for each SCOP family in turn, using the EMBOSS implementation
of the Needleman and Wunsch global alignment algorithm.  If a pair of
proteins achieve greater than a threshold percentage sequence identity
(specified by the user) the shortest sequence is discarded.  The user
must specify gap insertion and extension penalties and a residue
substitution matrix for use in the alignments.  

<p>







<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->


Here is a sample session with nrscope:

<p>

<pre>
% nrscope
Converts redundant EMBL-format SCOP file to non-redundant one
Name of scop file for input (embl-like format) [Escop.dat]: /data/scop/Escop.dat
Name of non-redundant scop file for output (embl-like format) [EscopNR.dat]: EscopNR.test
Location of clean domain coordinate files for input (embl-like format) [./]: /data/cpdbscop/
File extension of clean domain coordinate files [.pxyz]:
The % sequence identity redundancy threshold [95]: 95
Residue substitution file [EBLOSUM62]:
Gap insertion penalty [10]: 20
Gap extension penalty [0.5]: 1
Name of log file for the build [nrscope.log]: EscopNR.log
D3SDHA_
D3SDHB_
D3HBIA_
D3HBIB_
D4SDHA_
D4SDHB_
D4HBIA_
D4HBIB_
D5HBIA_
D5HBIB_
</pre>














<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-scopin]            infile     Name of scop file for input (embl-like
                                  format)
  [-scopout]           outfile    Name of non-redundant scop file for output
                                  (embl-like format)
  [-dpdb]              string     Location of clean domain coordinate files
                                  for input (embl-like format)
  [-extn]              string     File extension of clean domain coordinate
                                  files
  [-thresh]            float      The % sequence identity redundancy threshold
  [-datafile]          matrixf    Residue substitution matrix
  [-gapopen]           float      The gap insertion penalty is the score taken
                                  away when a gap is created. The best value
                                  depends on the choice of comparison matrix.
                                  The default value assumes you are using the
                                  EBLOSUM62 matrix for protein sequences, and
                                  the EDNAFULL matrix for nucleotide
                                  sequences.
  [-gapextend]         float      The gap extension, penalty is added to the
                                  standard gap penalty for each base or
                                  residue in the gap. This is how long gaps
                                  are penalized. Usually you will expect a few
                                  long gaps rather than many short gaps, so
                                  the gap extension penalty should be lower
                                  than the gap penalty. An exception is where
                                  one or both sequences are single reads with
                                  possible sequencing errors in which case you
                                  would expect many single base gaps. You can
                                  get this result by setting the gap open
                                  penalty to zero (or very low) and using the
                                  gap extension penalty to control gap
                                  scoring.
  [-errf]              outfile    Name of log file for the build

   Optional qualifiers: (none)
   Advanced qualifiers: (none)
   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-scopin]<br>(Parameter 1)</td>
<td>Name of scop file for input (embl-like format)</td>
<td>Input file</td>
<td>Escop.dat</td>
</tr>

<tr>
<td>[-scopout]<br>(Parameter 2)</td>
<td>Name of non-redundant scop file for output (embl-like format)</td>
<td>Output file</td>
<td>EscopNR.dat</td>
</tr>

<tr>
<td>[-dpdb]<br>(Parameter 3)</td>
<td>Location of clean domain coordinate files for input (embl-like format)</td>
<td>Any string is accepted</td>
<td>./</td>
</tr>

<tr>
<td>[-extn]<br>(Parameter 4)</td>
<td>File extension of clean domain coordinate files</td>
<td>Any string is accepted</td>
<td>.pxyz</td>
</tr>

<tr>
<td>[-thresh]<br>(Parameter 5)</td>
<td>The % sequence identity redundancy threshold</td>
<td>Any integer value</td>
<td>95.0</td>
</tr>

<tr>
<td>[-datafile]<br>(Parameter 6)</td>
<td>Residue substitution matrix</td>
<td>Comparison matrix file in EMBOSS data path</td>
<td>EBLOSUM62</td>
</tr>

<tr>
<td>[-gapopen]<br>(Parameter 7)</td>
<td>The gap insertion penalty is the score taken away when a gap is created. The best value depends on the choice of comparison matrix. The default value assumes you are using the EBLOSUM62 matrix for protein sequences, and the EDNAFULL matrix for nucleotide sequences.</td>
<td>Floating point number from 1.0 to 100.0</td>
<td>10.0 for any sequence</td>
</tr>

<tr>
<td>[-gapextend]<br>(Parameter 8)</td>
<td>The gap extension, penalty is added to the standard gap penalty for each base or residue in the gap. This is how long gaps are penalized. Usually you will expect a few long gaps rather than many short gaps, so the gap extension penalty should be lower than the gap penalty. An exception is where one or both sequences are single reads with possible sequencing errors in which case you would expect many single base gaps. You can get this result by setting the gap open penalty to zero (or very low) and using the gap extension penalty to control gap scoring.</td>
<td>Floating point number from 0.0 to 10.0</td>
<td>0.5 for any sequence</td>
</tr>

<tr>
<td>[-errf]<br>(Parameter 9)</td>
<td>Name of log file for the build</td>
<td>Output file</td>
<td>nrscope.log</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

The EMBL-like format SCOP classification file generated by the EMBOSS
application 
<a href="scope.html"><b>scope</b></a>
is as follows:

<p>

Records (4) to (8) 
are used to describe the position of the domain in the scop hierarchy.
<p>

<ol>

<LI>ID - Domain identifier code.  This is a 7-character code that uniquely
identifies the domain in scop.  It is identical to the first 7 characters 
of a line in the scop classification file.  The first character is always 
'D', the next four characters are the PDB identifier code, the fifth 
character is the PDB chain identifier to which the domain belongs (a '.' is
given in cases where the domain is composed of multiple chains, a '_' is 
given where a chain identifier was not specified in the PDB file) and the 
final character is the number of the domain in the chain (for chains 
comprising more than one domain) or '_' (the chain comprises a single 
domain only).

<LI>EN - PDB identifier code.  This is the 4-character PDB identifier code
of the PDB entry containing the domain.

<LI>OS - Source of the protein.  It is identical to the text given after 
'Species' in the scop classification file.

<LI>CL - Domain class.  It is identical to the text given after 'Class' in 
the scop classification file.

<LI>FO - Domain fold.  It is identical to the text given after 'Fold' in 
the scop classification file.

<LI>SF - Domain superfamily.  It is identical to the text given after 
'Superfamily' in the scop classification file.

<LI>FA - Domain family. It is identical to the text given after 'Family' in 
the scop classification file.

<LI>DO - Domain name. It is identical to the text given after 'Protein' in 
the scop classification file.

<LI>NC - Number of chains comprising the domain (usually 1).  If the number
of chains is greater than 1, then the domain entry will have a section  
containing a CN and a CH record (see below) for each chain.

<LI>CN - Chain number.  The number given in brackets after this record 
indicates the start of the data for the relevent chain.

<LI>CH - Domain definition.  The character given before CHAIN is the PDB 
chain identifier (a '.' is given in cases where a chain identifier was not 
specified in the scop classification file), the strings before START and 
END give the start and end positions respectively of the domain in the PDB 
file (a '.' is given in cases where a position was not specified).  Note 
that the start and end positions refer to residue numbering given in the 
original pdb file and therefore must be treated as strings.

</ol>












<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->

The output file has the same format as the input EMBL-like format SCOP
classification file generated by the EMBOSS application 
<a href="scope.html"><b>scope</b></a>
except that it contains non-redundant domains, as explained in the
<b>Description</b> section. 

<p>

<b>nrscope</b> generates a log file, an excerpt of which is shown below. 
The first two lines describe the level in the SCOP hierarchy at which
redundancy was removed (always 'FAMILIES' for the current
implementation) and the value of the redundancy threshold.  The file
then contains a section for each SCOP family.  Each section contains a
line with the record '//' immediately followed by the name of the SCOP
family, and two lines containing 'Retained' and 'Rejected' respectively. 
Domain identifier codes of domains that were discarded by <b>nrscope</b>
are listed under 'Rejected' while those that appear in the output file
are listed under 'Retained'.  The text 'WARN filename not found' is
given in cases where a clean domain coordinate file could not be found
and 'WARN Empty family' where no files for an entire family could be
found.  'ERROR filename file read error' will be given when an error was
encountered during a file read. 

<p>

<hr>
<pre>
FAMILIES are non-redundant
95% redundancy threshold
// Homeodomain
Retained
D2HDDA_
D1AKHA_
D1MNMC_
Rejected
D2HDDB_
D1ENH__
D3HDDA_
WARN  d3hdda_.pxyz not found
// Di-haem cytohrome c peroxidase
WARN  ds005__.pxyz not found
WARN  Empty family
// Nuclear receptor coactivator Src-1
Retained
D2PRGC_
Rejected
</pre>
<hr>












<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.
	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

   -->

None.














<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="cutgextract.html">cutgextract</a></td><td>Extract data from CUTG</td></tr>
<tr><td><a href="domainer.html">domainer</a></td><td>Build domain coordinate files</td></tr>
<tr><td><a href="printsextract.html">printsextract</a></td><td>Extract data from PRINTS</td></tr>
<tr><td><a href="prosextract.html">prosextract</a></td><td>Builds the PROSITE motif database for patmatmotifs to search</td></tr>
<tr><td><a href="rebaseextract.html">rebaseextract</a></td><td>Extract data from REBASE</td></tr>
<tr><td><a href="scope.html">scope</a></td><td>Convert raw scop classification file to embl-like format</td></tr>
<tr><td><a href="tfextract.html">tfextract</a></td><td>Extract data from TRANSFAC</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

This application was written by Jon Ison <a
href="mailto:jison@hgmp.mrc.ac.uk">(jison@hgmp.mrc.ac.uk)</a>
















<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (Jan 2001) - Jon Ison.












<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
This program is intended to be used by everyone and everything, from naive users to embedded scripts.

   -->

This program is intended to be run by EMBOSS site maintainers or those
responsible for setting up and maintaining protein 3D structural data
for use by others. 










<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->








</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

