<HTML>

<HEAD>
  <TITLE>
  EMBOSS: octanol
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: octanol
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program octanol
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Displays protein hydropathy


<H2>
    Description
</H2>


Protein sequences that form transmembrane regions are assumed to have
a thermodynamic preference for a hydrophobic environment (inside the
membrane lipid bilayer), rather than an aqueous environment in water.

<p>
The free energy change for each amino acid residue between a lipid and
a water environment can be measured experimentally, and the values for
peptides can be shown to be additive (White and Wimley 1999).

<p>
The octanol program calculates two free energy differences.

<p>
The first is the free energy difference between
solution in water and association with the interface (glycerol group) of a
POPC (palmitoyloleoylphosphocholine) bilayer.

<p>
The second is the free energy difference between water and octanol,
equivalent to the environment inside a lipid bilayer.

<p>
Residues which can be buried inside a lipid bilayer must be in a
region of the peptide where most residues show a free energy
difference in favour of being in an octanol environment or at least being
in the lipid/water interface region.

<p>
White and Wimley (1999) showed that a sliding window of either free
energy difference will indicate the location of probably transmembrane
regions, but that the best indicator is the difference between the two
values, which is the free energy difference between the interface and
octanol environments.

<p> The free energies are calculated over a sliding window of 19
residues, about the size of a membrane spanning alphahelix. The energy
values for each residue are added over the window.


<H2>
    Usage
</H2>


Here is a sample session with octanol.

<pre>

% octanol
Input sequence: sw:opsd_human
Graph type [x11]: 

</pre>

<a href="octanol0.gif">Click here to see the results.</a> 



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequencea]         sequence   Sequence USA
  [-graph]             xygraph    Graph type

   Optional qualifiers:
   -datafile           datafile   White-Wimley data file (Ewhite-wimley.dat)
   -width              integer    window size
   -octanolplot        bool       Display the octanol plot
   -interfaceplot      bool       Display the interface plot
   -[no]differenceplot bool       Do not display the difference plot

   Advanced qualifiers: (none)

</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequencea]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-graph]<br>(Parameter 2)</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including postscript, ps, hpgl, hp7470, hp7580, meta, colourps, cps, xwindows, x11, tektronics, tekt, tek4107t, tek, none, null, text, data, xterm</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-datafile</td>
<td>White-Wimley data file (Ewhite-wimley.dat)</td>
<td>Data file</td>
<td>Ewhite-wimley.dat</td>
</tr>

<tr>
<td>-width</td>
<td>window size</td>
<td>Integer from 1 to 200</td>
<td>19</td>
</tr>

<tr>
<td>-octanolplot</td>
<td>Display the octanol plot</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-interfaceplot</td>
<td>Display the interface plot</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]differenceplot</td>
<td>Do not display the difference plot</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>

Any protein sequence.

<H2>
    Output file format
</H2>

octanol draws a graph showing the
free energy calcuated over a sliding window.

<p>
The line on the default plot is the difference between the interface
and octanol free energy calculations. Command line options allow the
display of the interface and octanol values, or hiding the difference
values.

<p>
In the example, the human opsin protein has 7 transmembrane
regions: 37-61, 74-98, 114-133, 153-176, 203-230, 253-276 and 285-309.
Each is about 20 residues in length, which is also the gap between tick
marks on the sequence axis. All have energetic preferences for being
in the lipid (octanol) enviroment - shown as being above the zero
line - or have at least no clear preference.

<p>
Running octanol with all three plots:

<pre>

% octanol -interface -octanol
Input sequence: tsw:opsd_human
   Graph type [x11]: 

</pre>

gives a graph with the water-interface and water-octanol plots (link
to the output).

For those regions where the diference plot is close to zero, both the
other two plots are above the line, showing a preference for either
the octanol or the interface membrane environments rather than water.



<H2>
    Data files
</H2>

File <b>Ewhite-wimley.dat</b> contains the experimental free energy values
for the water-interface and water-octanol transitions.

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>


<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

<ol>

<li>
White S.H. and Wimley W.C. (1999)
"Membrane protein folding and stability: physical principles"
Ann. Rev.Biophys. Biomol. Struct. 28:319-365.

</ol>

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with status 0.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="backtranseq.html">backtranseq</a></td><td>Back translate a protein sequence</td></tr>
<tr><td><a href="charge.html">charge</a></td><td>Protein charge plot</td></tr>
<tr><td><a href="checktrans.html">checktrans</a></td><td>Reports STOP codons and ORF statistics of a protein sequence</td></tr>
<tr><td><a href="compseq.html">compseq</a></td><td>Counts the composition of dimer/trimer/etc words in a sequence</td></tr>
<tr><td><a href="emowse.html">emowse</a></td><td>Protein identification by mass spectrometry</td></tr>
<tr><td><a href="iep.html">iep</a></td><td>Calculates the isoelectric point of a protein</td></tr>
<tr><td><a href="pepinfo.html">pepinfo</a></td><td>Plots simple amino acid properties in parallel</td></tr>
<tr><td><a href="pepstats.html">pepstats</a></td><td>Protein statistics</td></tr>
<tr><td><a href="pepwindow.html">pepwindow</a></td><td>Displays protein hydropathy</td></tr>
<tr><td><a href="pepwindowall.html">pepwindowall</a></td><td>Displays protein hydropathy of a set of sequences</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Ian Longden <a
href="mailto:il@sanger.ac.uk">(il@sanger.ac.uk)</a> Informatics
Division, The Sanger Centre, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.

<H2>
    History
</H2>


<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

