<HTML>

<HEAD>
  <TITLE>
  EMBOSS: octanol
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">




<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

octanol
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>





<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Displays protein hydropathy


<H2>
    Description
</H2>


Protein sequences that form transmembrane regions are assumed to have
a thermodynamic preference for a hydrophobic environment (inside the
membrane lipid bilayer), rather than an aqueous environment in water.

<p>
The free energy change for each amino acid residue between a lipid and
a water environment can be measured experimentally, and the values for
peptides can be shown to be additive (White and Wimley 1999).

<p>
The octanol program calculates two free energy differences.

<p>
The first is the free energy difference between
solution in water and association with the interface (glycerol group) of a
POPC (palmitoyloleoylphosphocholine) bilayer.

<p>
The second is the free energy difference between water and octanol,
equivalent to the environment inside a lipid bilayer.

<p>
Residues which can be buried inside a lipid bilayer must be in a
region of the peptide where most residues show a free energy
difference in favour of being in an octanol environment or at least being
in the lipid/water interface region.

<p>
White and Wimley (1999) showed that a sliding window of either free
energy difference will indicate the location of probably transmembrane
regions, but that the best indicator is the difference between the two
values, which is the free energy difference between the interface and
octanol environments.

<p> The free energies are calculated over a sliding window of 19
residues, about the size of a membrane spanning alphahelix. The energy
values for each residue are added over the window.


<H2>
    Usage
</H2>


<b>Here is a sample session with octanol</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>octanol </b>
Displays protein hydropathy
Input sequence: <b>tsw:opsd_human</b>
Graph type [x11]: <b>ps</b>

Created octanol.ps

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequencea]         sequence   Sequence USA
  [-graph]             xygraph    Graph type

   Optional qualifiers:
   -datafile           datafile   White-Wimley data file (Ewhite-wimley.dat)
   -width              integer    window size
   -octanolplot        boolean    Display the octanol plot
   -interfaceplot      boolean    Display the interface plot
   -[no]differenceplot boolean    Display the difference plot

   Advanced qualifiers: (none)
   Associated qualifiers:
  "-sequencea" related qualifiers
   -sbegin1             integer    First base used
   -send1               integer    Last base used, def=seq length
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sopenfile1          string     Input filename
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name
  "-graph" related qualifiers
   -gprompt2            boolean    Graph prompting
   -gtitle2             string     Graph title
   -gsubtitle2          string     Graph subtitle
   -gxtitle2            string     Graph x axis title
   -gytitle2            string     Graph y axis title
   -goutfile2           string     Output file for non interactive displays
   -gdirectory2         string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for required and optional values
   -debug               boolean    Write debug output to program.dbg
   -acdlog              boolean    Write ACD processing log to program.acdlog
   -acdpretty           boolean    Rewrite ACD file as program.acdpretty
   -acdtable            boolean    Write HTML table of options
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequencea]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-graph]<br>(Parameter 2)</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including postscript, ps, hpgl, hp7470, hp7580, meta, colourps, cps, xwindows, x11, tektronics, tekt, tek4107t, tek, none, null, text, data, xterm, png</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-datafile</td>
<td>White-Wimley data file (Ewhite-wimley.dat)</td>
<td>Data file</td>
<td>Ewhite-wimley.dat</td>
</tr>

<tr>
<td>-width</td>
<td>window size</td>
<td>Integer from 1 to 200</td>
<td>19</td>
</tr>

<tr>
<td>-octanolplot</td>
<td>Display the octanol plot</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-interfaceplot</td>
<td>Display the interface plot</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]differenceplot</td>
<td>Display the difference plot</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>

<b>octanol</b> reads any protein sequence USA.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tsw:opsd_human' is a sequence entry in the example protein database 'tsw'
<p>
<p><h3>Database entry: tsw:opsd_human</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   OPSD_HUMAN     STANDARD;      PRT;   348 AA.
AC   P08100; Q16414;
DT   01-AUG-1988 (Rel. 08, Created)
DT   01-AUG-1988 (Rel. 08, Last sequence update)
DT   15-JUL-1999 (Rel. 38, Last annotation update)
DE   RHODOPSIN.
GN   RHO.
OS   Homo sapiens (Human).
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Mammalia;
OC   Eutheria; Primates; Catarrhini; Hominidae; Homo.
RN   [1]
RP   SEQUENCE FROM N.A.
RX   MEDLINE; 84272729.
RA   NATHANS J., HOGNESS D.S.;
RT   "Isolation and nucleotide sequence of the gene encoding human
RT   rhodopsin.";
RL   Proc. Natl. Acad. Sci. U.S.A. 81:4851-4855(1984).
RN   [2]
RP   SEQUENCE OF 1-120 FROM N.A.
RA   BENNETT J., BELLER B., SUN D., KARIKO K.;
RL   Submitted (NOV-1994) to the EMBL/GenBank/DDBJ databases.
RN   [3]
RP   REVIEW ON ADRP VARIANTS.
RX   MEDLINE; 94004905.
RA   AL-MAGHTHEH M., GREGORY C., INGLEHEARN C., HARDCASTLE A.,
RA   BHATTACHARYA S.;
RT   "Rhodopsin mutations in autosomal dominant retinitis pigmentosa.";
RL   Hum. Mutat. 2:249-255(1993).
RN   [4]
RP   VARIANT ADRP HIS-23.
RX   MEDLINE; 90136922.
RA   DRYJA T.P., MCGEE T.L., REICHEI E., HAHN L.B., COWLEY G.S.,
RA   YANDELL D.W., SANDBERG M.A., BERSON E.L.;
RT   "A point mutation of the rhodopsin gene in one form of retinitis
RT   pigmentosa.";
RL   Nature 343:364-366(1990).
RN   [5]
RP   VARIANTS ADRP.
RX   MEDLINE; 91051574.
RA   FARRAR G.J., KENNA P., REDMOND R., MCWILLIAM P., BRADLEY D.G.,
RA   HUMPHRIES M.M., SHARP E.M., INGLEHEARN C.F., BASHIR R., JAY M.,
RA   WATTY A., LUDWIG M., SCHINZEL A., SAMANNS C., GAL A.,
RA   BHATTACHARYA S.S., HUMPHRIES P.;
RT   "Autosomal dominant retinitis pigmentosa: absence of the rhodopsin
RT   proline--&gt;histidine substitution (codon 23) in pedigrees from
RT   Europe.";
RL   Am. J. Hum. Genet. 47:941-945(1990).
RN   [6]
RP   VARIANTS ADRP HIS-23; ARG-58; LEU-347 AND SER-347.
RX   MEDLINE; 91015273.


<font color=red>  [Part of this file has been deleted for brevity]</font>

FT                                /FTId=VAR_004816.
FT   VARIANT     209    209       V -&gt; M (EFFECT NOT KNOWN).
FT                                /FTId=VAR_004817.
FT   VARIANT     211    211       H -&gt; P (IN ADRP).
FT                                /FTId=VAR_004818.
FT   VARIANT     211    211       H -&gt; R (IN ADRP).
FT                                /FTId=VAR_004819.
FT   VARIANT     216    216       M -&gt; K (IN ADRP).
FT                                /FTId=VAR_004820.
FT   VARIANT     220    220       F -&gt; C (IN ADRP).
FT                                /FTId=VAR_004821.
FT   VARIANT     222    222       C -&gt; R (IN ADRP).
FT                                /FTId=VAR_004822.
FT   VARIANT     255    255       MISSING (IN ADRP).
FT                                /FTId=VAR_004823.
FT   VARIANT     264    264       MISSING (IN ADRP).
FT                                /FTId=VAR_004824.
FT   VARIANT     267    267       P -&gt; L (IN ADRP).
FT                                /FTId=VAR_004825.
FT   VARIANT     267    267       P -&gt; R (IN ADRP).
FT                                /FTId=VAR_004826.
FT   VARIANT     292    292       A -&gt; E (IN CSNB4).
FT                                /FTId=VAR_004827.
FT   VARIANT     296    296       K -&gt; E (IN ADRP).
FT                                /FTId=VAR_004828.
FT   VARIANT     297    297       S -&gt; R (IN ADRP).
FT                                /FTId=VAR_004829.
FT   VARIANT     342    342       T -&gt; M (IN ADRP).
FT                                /FTId=VAR_004830.
FT   VARIANT     345    345       V -&gt; L (IN ADRP).
FT                                /FTId=VAR_004831.
FT   VARIANT     345    345       V -&gt; M (IN ADRP).
FT                                /FTId=VAR_004832.
FT   VARIANT     347    347       P -&gt; A (IN ADRP).
FT                                /FTId=VAR_004833.
FT   VARIANT     347    347       P -&gt; L (IN ADRP; COMMON VARIANT).
FT                                /FTId=VAR_004834.
FT   VARIANT     347    347       P -&gt; Q (IN ADRP).
FT                                /FTId=VAR_004835.
FT   VARIANT     347    347       P -&gt; R (IN ADRP).
FT                                /FTId=VAR_004836.
FT   VARIANT     347    347       P -&gt; S (IN ADRP).
FT                                /FTId=VAR_004837.
SQ   SEQUENCE   348 AA;  38892 MW;  07443BEA CRC32;
     MNGTEGPNFY VPFSNATGVV RSPFEYPQYY LAEPWQFSML AAYMFLLIVL GFPINFLTLY
     VTVQHKKLRT PLNYILLNLA VADLFMVLGG FTSTLYTSLH GYFVFGPTGC NLEGFFATLG
     GEIALWSLVV LAIERYVVVC KPMSNFRFGE NHAIMGVAFT WVMALACAAP PLAGWSRYIP
     EGLQCSCGID YYTLKPEVNN ESFVIYMFVV HFTIPMIIIF FCYGQLVFTV KEAAAQQQES
     ATTQKAEKEV TRMVIIMVIA FLICWVPYAS VAFYIFTHQG SNFGPIFMTI PAFFAKSAAI
     YNPVIYIMMN KQFRNCMLTT ICCGKNPLGD DEASATVSKT ETSQVAPA
//
</pre>
</td></tr></table><p>


<H2>
    Output file format
</H2>

<b>octanol</b> draws a graph showing the
free energy calcuated over a sliding window.

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>Graphics File: octanol.ps</h3>
<p><img src="octanol.1.octanol.gif" alt="[octanol results]">


<p>

The line on the default plot is the difference between the interface
and octanol free energy calculations. Command line options allow the
display of the interface and octanol values, or hiding the difference
values.

<p>

In the example, the human opsin protein has 7 transmembrane
regions: 37-61, 74-98, 114-133, 153-176, 203-230, 253-276 and 285-309.
Each is about 20 residues in length, which is also the gap between tick
marks on the sequence axis. All have energetic preferences for being
in the lipid (octanol) enviroment - shown as being above the zero
line - or have at least no clear preference.

<p>

Running octanol with all three plots:

<pre>

% octanol -interface -octanol
Input sequence: tsw:opsd_human
   Graph type [x11]: 

</pre>

gives a graph with the water-interface and water-octanol plots.

<p>

For those regions where the diference plot is close to zero, both the
other two plots are above the line, showing a preference for either
the octanol or the interface membrane environments rather than water.



<H2>
    Data files
</H2>

File <b>Ewhite-wimley.dat</b> contains the experimental free energy values
for the water-interface and water-octanol transitions.

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>


<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

<ol>

<li>
White S.H. and Wimley W.C. (1999)
"Membrane protein folding and stability: physical principles"
Ann. Rev.Biophys. Biomol. Struct. 28:319-365.

</ol>

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with status 0.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="backtranseq.html">backtranseq</a></td><td>Back translate a protein sequence</td></tr>
<tr><td><a href="charge.html">charge</a></td><td>Protein charge plot</td></tr>
<tr><td><a href="checktrans.html">checktrans</a></td><td>Reports STOP codons and ORF statistics of a protein</td></tr>
<tr><td><a href="compseq.html">compseq</a></td><td>Counts the composition of dimer/trimer/etc words in a sequence</td></tr>
<tr><td><a href="emowse.html">emowse</a></td><td>Protein identification by mass spectrometry</td></tr>
<tr><td><a href="freak.html">freak</a></td><td>Residue/base frequency table or plot</td></tr>
<tr><td><a href="iep.html">iep</a></td><td>Calculates the isoelectric point of a protein</td></tr>
<tr><td><a href="mwcontam.html">mwcontam</a></td><td>Shows molwts that match across a set of files</td></tr>
<tr><td><a href="mwfilter.html">mwfilter</a></td><td>Filter noisy molwts from mass spec output</td></tr>
<tr><td><a href="pepinfo.html">pepinfo</a></td><td>Plots simple amino acid properties in parallel</td></tr>
<tr><td><a href="pepstats.html">pepstats</a></td><td>Protein statistics</td></tr>
<tr><td><a href="pepwindow.html">pepwindow</a></td><td>Displays protein hydropathy</td></tr>
<tr><td><a href="pepwindowall.html">pepwindowall</a></td><td>Displays protein hydropathy of a set of sequences</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


Ian Longden (il&nbsp;&copy;&nbsp;sanger.ac.uk)
<br>
Sanger Institute, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.                      



<H2>
    History
</H2>

<!--
<p>
[an error occurred while processing this directive]
-->

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

