<HTML>
<HEAD>
  <TITLE>
  EMBOSS: ontotext
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="/images/emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
ontotext
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
Get ontology term(s) original full text
<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<p><b>ontotext</b> reads one or more complete ontology term entries from a database or a file and writes them to a text file.  Optionally, the first ontology term from the input stream only can be retrieved.  The complete entry, including heading annotation, is retrieved and written and the data is not altered or reformatted in any way.</p>


<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

Here is a sample session with <b>ontotext</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>ontotext edam:1234 </b>
Get ontology term(s) original full text
Full text output file [1234.ontotext]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>
<p>
<b>Example 2</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>ontotext edam:0850 -subclasses </b>
Get ontology term(s) original full text
Full text output file [0850.ontotext]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.2">Go to the input files for this example</a><br><a href="#output.2">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
Get ontology term(s) original full text
Version: EMBOSS:6.6.0.0

   Standard (Mandatory) qualifiers:
  [-oboterms]          obo        Obo term filename and optional format, or
                                  reference (input query)
  [-outfile]           outfile    [*.ontotext] Full text output file

   Additional (Optional) qualifiers:
   -subclasses         boolean    [N] Extend the query matches to include all
                                  terms which are specialisations
                                  (sub-classes) of the matched terms.
   -obsolete           boolean    [N] The default behaviour is to not use or
                                  return obsolete terms. This option if set
                                  will include all terms.

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-oboterms" associated qualifiers
   -iformat1           string     Input obo format
   -iquery1            string     Input query fields or ID list
   -ioffset1           integer    Input start position offset
   -idbname1           string     User-provided database name

   "-outfile" associated qualifiers
   -odirectory2        string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-oboterms]<br>(Parameter 1)</td>
<td>obo</td>
<td>Obo term filename and optional format, or reference (input query)</td>
<td>OBO bio-ontology term(s)</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-outfile]<br>(Parameter 2)</td>
<td>outfile</td>
<td>Full text output file</td>
<td>Output file</td>
<td><i>&lt;*&gt;</i>.ontotext</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>-subclasses</td>
<td>boolean</td>
<td>Extend the query matches to include all terms which are specialisations (sub-classes) of the matched terms.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-obsolete</td>
<td>boolean</td>
<td>The default behaviour is to not use or return obsolete terms. This option if set will include all terms.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td align="left" colspan=5>"-oboterms" associated obo qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -iformat1<br>-iformat_oboterms</td>
<td>string</td>
<td>Input obo format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -iquery1<br>-iquery_oboterms</td>
<td>string</td>
<td>Input query fields or ID list</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ioffset1<br>-ioffset_oboterms</td>
<td>integer</td>
<td>Input start position offset</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -idbname1<br>-idbname_oboterms</td>
<td>string</td>
<td>User-provided database name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td align="left" colspan=5>"-outfile" associated outfile qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -odirectory2<br>-odirectory_outfile</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<p>

The input is a standard EMBOSS ontology query.

<p> The major ontology sources defined as standard in EMBOSS
installations are EDAM, GO (gene Ontology) and SO (Sequence Ontology).

<p>

Data can also be read from ontology output in "obo" format written by an
EMBOSS application.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>Database entry: edam:1234</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
[Term]
id: EDAM_data:1234
name: Sequence set (nucleic acid)
namespace: data
def: Any collection of multiple nucleotide sequences and associated metadata that do not (typically) correspond to common sequence database records or database entries.
subset: bioinformatics
subset: data
subset: edam
created_in: "beta12orEarlier"
is_a: EDAM_data:0850 ! Sequence set
is_a: EDAM_data:2977 ! Nucleic acid sequence

</pre>
</td></tr></table><p>

<a name="input.2"></a>
<h3>Input files for usage example 2</h3>
<p><h3>Database entry: edam:0850</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
[Term]
id: EDAM_data:0850
name: Sequence set
namespace: data
def: A collection of multiple molecular sequences and associated metadata that do not (typically) correspond to molecular sequence database records or entries and which (typically) are derived from some analytical method.
comment: This concept may be used for arbitrary sequence sets and associated data arising from processing.
subset: bioinformatics
subset: data
subset: edam
synonym: "SO:0001260" RELATED []
created_in: "beta12orEarlier"
is_a: EDAM_data:2955 ! Sequence report

</pre>
</td></tr></table><p>





<H2>
    Output file format
</H2>


<b>ontotext</b> reports the full text of the input ontology terms.

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: 1234.ontotext</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
[Term]
id: EDAM_data:1234
name: Sequence set (nucleic acid)
subset: bioinformatics
subset: data
subset: edam
created_in: "beta12orEarlier"
def: "Any collection of multiple nucleotide sequences and associated metadata that do not (typically) correspond to common sequence database records or database entries." [http://edamontology.org]
namespace: data
is_a: EDAM_data:0850 ! Sequence set
is_a: EDAM_data:2977 ! Nucleic acid sequence

</pre>
</td></tr></table><p>

<a name="output.2"></a>
<h3>Output files for usage example 2</h3>
<p><h3>File: 0850.ontotext</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
[Term]
id: EDAM_data:0850
name: Sequence set
comment: This concept may be used for arbitrary sequence sets and associated data arising from processing.
subset: bioinformatics
subset: data
subset: edam
synonym: "SO:0001260" RELATED []
created_in: "beta12orEarlier"
def: "A collection of multiple molecular sequences and associated metadata that do not (typically) correspond to molecular sequence database records or entries and which (typically) are derived from some analytical method." [http://edamontology.org]
namespace: data
is_a: EDAM_data:2955 ! Sequence report

[Term]
id: EDAM_data:1233
name: Sequence set (protein)
subset: bioinformatics
subset: data
subset: edam
created_in: "beta12orEarlier"
def: "Any collection of multiple protein sequences and associated metadata that do not (typically) correspond to common sequence database records or database entries." [http://edamontology.org]
namespace: data
is_a: EDAM_data:0850 ! Sequence set

[Term]
id: EDAM_data:1245
name: Sequence cluster (protein)
comment: The sequences are typically related, for example a family of sequences.
subset: bioinformatics
subset: data
subset: edam
synonym: "Protein sequence cluster" EXACT []
created_in: "beta12orEarlier"
def: "A cluster of protein sequences." [http://edamontology.org]
namespace: data
is_a: EDAM_data:1233 ! Sequence set (protein)
is_a: EDAM_data:1235 ! Sequence cluster

[Term]
id: EDAM_data:1262
name: Peptide molecular weight hits
subset: bioinformatics
subset: data
subset: edam
created_in: "beta12orEarlier"
def: "A report on peptide fragments of certain molecular weight(s) in one or more protein sequences." [http://edamontology.org]
namespace: data
is_a: EDAM_data:1233 ! Sequence set (protein)

[Term]


<font color=red>  [Part of this file has been deleted for brevity]</font>

id: EDAM_data:1246
name: Sequence cluster (nucleic acid)
comment: The sequences are typically related, for example a family of sequences.
subset: bioinformatics
subset: data
subset: edam
synonym: "Nucleotide sequence cluster" EXACT []
created_in: "beta12orEarlier"
def: "A cluster of nucleotide sequences." [http://edamontology.org]
namespace: data
is_a: EDAM_data:1234 ! Sequence set (nucleic acid)
is_a: EDAM_data:1235 ! Sequence cluster

[Term]
id: EDAM_data:1239
name: Restriction digest
subset: bioinformatics
subset: data
subset: edam
xref: SO:0000412
created_in: "beta12orEarlier"
def: "Restriction digest fragments from digesting a nucleotide sequence with restriction sites using a restriction endonuclease." [http://edamontology.org]
namespace: data
is_a: EDAM_data:1234 ! Sequence set (nucleic acid)

[Term]
id: EDAM_data:2245
name: Sequence set (bootstrapped)
comment: Bootstrapping is often performed in phylogenetic analysis.
subset: bioinformatics
subset: data
subset: edam
created_in: "beta12orEarlier"
def: "A collection of sequences output from a bootstrapping (resampling) procedure." [http://edamontology.org]
namespace: data
is_a: EDAM_data:0850 ! Sequence set

[Term]
id: EDAM_data:1235
name: Sequence cluster
comment: The cluster might include sequences identifiers, short descriptions, alignment and summary information.
subset: bioinformatics
subset: data
subset: edam
created_in: "beta12orEarlier"
def: "A set of sequences that have been clustered or otherwise classified as belonging to a group including (typically) sequence cluster information." [http://edamontology.org]
namespace: data
is_a: EDAM_data:0850 ! Sequence set
relationship: has_topic EDAM_topic:0724 ! Protein families
relationship: has_topic EDAM_topic:3052 ! Sequence clusters and classification

</pre>
</td></tr></table><p>

<H2>
    Data files
</H2>

The EDAM Ontology is included in EMBOSS as local
database <b>edam</b>.


<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
<ol>

<li>

</ol>

   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="drtext.html">drtext</a></td>
<td>Get data resource entries complete text</td>
</tr>

<tr>
<td><a href="edamdef.html">edamdef</a></td>
<td>Find EDAM ontology terms by definition</td>
</tr>

<tr>
<td><a href="edamhasinput.html">edamhasinput</a></td>
<td>Find EDAM ontology terms by has_input relation</td>
</tr>

<tr>
<td><a href="edamhasoutput.html">edamhasoutput</a></td>
<td>Find EDAM ontology terms by has_output relation</td>
</tr>

<tr>
<td><a href="edamisformat.html">edamisformat</a></td>
<td>Find EDAM ontology terms by is_format_of relation</td>
</tr>

<tr>
<td><a href="edamisid.html">edamisid</a></td>
<td>Find EDAM ontology terms by is_identifier_of relation</td>
</tr>

<tr>
<td><a href="edamname.html">edamname</a></td>
<td>Find EDAM ontology terms by name</td>
</tr>

<tr>
<td><a href="entret.html">entret</a></td>
<td>Retrieve sequence entries from flatfile databases and files</td>
</tr>

<tr>
<td><a href="godef.html">godef</a></td>
<td>Find GO ontology terms by definition</td>
</tr>

<tr>
<td><a href="goname.html">goname</a></td>
<td>Find GO ontology terms by name</td>
</tr>

<tr>
<td><a href="ontocount.html">ontocount</a></td>
<td>Count ontology term(s)</td>
</tr>

<tr>
<td><a href="ontoget.html">ontoget</a></td>
<td>Get ontology term(s)</td>
</tr>

<tr>
<td><a href="ontogetcommon.html">ontogetcommon</a></td>
<td>Get common ancestor for terms</td>
</tr>

<tr>
<td><a href="ontogetdown.html">ontogetdown</a></td>
<td>Get ontology term(s) by parent id</td>
</tr>

<tr>
<td><a href="ontogetobsolete.html">ontogetobsolete</a></td>
<td>Get ontology ontology terms</td>
</tr>

<tr>
<td><a href="ontogetroot.html">ontogetroot</a></td>
<td>Get ontology root terms by child identifier</td>
</tr>

<tr>
<td><a href="ontogetsibs.html">ontogetsibs</a></td>
<td>Get ontology term(s) by id with common parent</td>
</tr>

<tr>
<td><a href="ontogetup.html">ontogetup</a></td>
<td>Get ontology term(s) by id of child</td>
</tr>

<tr>
<td><a href="ontoisobsolete.html">ontoisobsolete</a></td>
<td>Report whether an ontology term id is obsolete</td>
</tr>

<tr>
<td><a href="textget.html">textget</a></td>
<td>Get text data entries</td>
</tr>

<tr>
<td><a href="textsearch.html">textsearch</a></td>
<td>Search the textual description of sequence(s)</td>
</tr>

<tr>
<td><a href="xmltext.html">xmltext</a></td>
<td>Get XML document original full text</td>
</tr>

</table>
<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
Peter Rice
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK

<p>
Please report all bugs to the EMBOSS bug team (emboss-bug&nbsp;&copy;&nbsp;emboss.open-bio.org) not to the original author.
<H2>
    History
</H2>
<!--
        Date written and what changes have been made go in this file.
   -->




<H2>
    Target users
</H2>
<!--
        For general users, use this text
   -->
This program is intended to be used by everyone and everything, from naive users to embedded scripts.

<H2>
    Comments
</H2>
<!--
        User/developer/other comments go in this file.
   -->
None


</BODY>
</HTML>

