<HTML>

<HEAD>
  <TITLE>
  EMBOSS: patmatdb
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: patmatdb
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program patmatdb
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Search a protein sequence with a motif


<H2>
    Description
</H2>


Takes a protein motif and compares it to a set of protein sequences. 

<p>
The motif is converted to a regular expression, and matches are
completed using Henry Spencer's regular expression library of functions. 

<p>
It returns the number of matches there were between the motif and each
matched sequence, length of match, start and end positions of match, and
writes out an alignment. 

<H2>
    Usage
</H2>


Here is a sample session with patmatdb.

<pre>

% patmatdb
Search a protein sequence with a motif
Input sequence(s): sw:*
Protein motif to search for: st[ty]s
Output file [5h1d_fugru.patmatdb]:

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
  [-motif]             string     Patterns for patmatdb are based on the
                                  format of pattern used in the PROSITE
                                  database. For example:
                                  '[DE](2)HS{P}X(2)PX(2,4)C' means two Asps or
                                  Glus in any order followed by His, Ser, any
                                  residue other then Pro, then two of any
                                  residue followed by Pro followed by two to
                                  four of any residue followed by Cys. The
                                  search is case-independent, so 'AAA' matches
                                  'aaa'.
  [-outfile]           outfile    Output file name

   Optional qualifiers: (none)
   Advanced qualifiers: (none)

</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-motif]<br>(Parameter 2)</td>
<td>Patterns for patmatdb are based on the format of pattern used in the PROSITE database. For example: '[DE](2)HS{P}X(2)PX(2,4)C' means two Asps or Glus in any order followed by His, Ser, any residue other then Pro, then two of any residue followed by Pro followed by two to four of any residue followed by Cys. The search is case-independent, so 'AAA' matches 'aaa'.</td>
<td>A string of at least 2 characters</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 3)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.patmatdb</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>


Patterns for patmatdb are based on the format of pattern used in the
PROSITE database, with the difference that the terminating dot '.' and
the hyphens, '-', between the characters are optional.

<p>
The PROSITE pattern definition from the PROSITE documentation follows. 

<ul>
<li> The standard IUPAC one-letter codes for the amino acids are used.

<li> The symbol `x' is used for a position where any amino acid is
accepted.

<li> Ambiguities are indicated by listing the acceptable amino acids
      for a given position, between square parentheses `[ ]'. For
      example: [ALT] stands for Ala or Leu or Thr.

<li> Ambiguities are also indicated by listing between a pair of curly
      brackets `{ }' the amino acids that are not accepted at a given
      position. For example: {AM} stands for any amino acid except Ala
      and Met.

<li> Each element in a pattern is separated from its neighbor by a
      `-'.  (Optional in patmatdb).

<li> Repetition of an element of the pattern can be indicated by
      following that element with a numerical value or a numerical
      range between parenthesis. Examples: x(3) corresponds to x-x-x,
      x(2,4) corresponds to x-x or x-x-x or x-x-x-x.

<li> When a pattern is restricted to either the N- or C-terminal of a
      sequence, that pattern either starts with a `&lt;' symbol or
      respectively ends with a `&gt;' symbol.

<li> A period ends the pattern. (Optional in patmatdb).

</ul>

<p>
For example, in SWISSPROT entry 100K_RAT you can look for the pattern: 

<pre>

[DE](2)HS{P}X(2)PX(2,4)C

</pre>

<p>
This means: Two Asps or Glus in any order followed by His, Ser, any residue
other then Pro, then two of any residue followed by Pro followed by two to four
of any residue followed by Cys. 

<p>
The search is case-independent, so 'AAA' matches 'aaa'. 

<H2>
    Output file format
</H2>


The output from the above example follows:

<pre>

Number of matches found in HD_FUGRU = 1
Length of the sequence = 3148 basepairs
Length of match = 4
Start of match = position 1037 of sequence
End of match = position 1040 of sequence

patmatDB of HD_FUGRU from 1037 to 1040 using pattern st[ty]s

STPASSTTSSAVDP
     |  |
  1037  1040



Number of matches found in GCN4_YEAST = 1
Length of the sequence = 281 basepairs
Length of match = 4
Start of match = position 157 of sequence
End of match = position 160 of sequence

patmatDB of GCN4_YEAST from 157 to 160 using pattern st[ty]s

SNLEVSTTSFLPTP
     |  |
   157  160

</pre>

<H2>
    Data files
</H2>


See outfile file format above.

<H2>
    Notes
</H2>


<H2>
    References
</H2>


<ol>
<li>Bairoch, A., Bucher P. (1994) PROSITE: recent developments. Nucleic 
Acids Research, Vol 22, No.17 3583-3589.

<li>Bairoch, A., (1992) PROSITE: a dictionary of sites and patterns in 
proteins. Nucleic Acids Research, Vol 20, Supplement, 2013-2018.

<li>Peek, J., O'Reilly, T., Loukides, M., (1997) Unix Power Tools, 2nd 
Edition.

<li>Gusfield, D., (1997) Algorithms on strings, Trees and Sequences.

<li>Sedgewick, R,. (1990) Algorithms in C.
</ol>

<H2>
    Warnings
</H2>


The program will warn user if input protein sequence file is 
incorrectly formatted.
 
<H2>
    Diagnostic Error Messages
</H2>


As in warnings.
        
<H2>
    Exit status
</H2>


It always exits with status 0

<H2>
    Known bugs
</H2>


<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="dreg.html">dreg</a></td><td>regular expression search of a nucleotide sequence</td></tr>
<tr><td><a href="fuzznuc.html">fuzznuc</a></td><td>Nucleic acid pattern search</td></tr>
<tr><td><a href="fuzzpro.html">fuzzpro</a></td><td>Protein pattern search</td></tr>
<tr><td><a href="fuzztran.html">fuzztran</a></td><td>Protein pattern search after translation</td></tr>
<tr><td><a href="helixturnhelix.html">helixturnhelix</a></td><td>Report nucleic acid binding motifs</td></tr>
<tr><td><a href="patmatmotifs.html">patmatmotifs</a></td><td>Search a PROSITE motif database with a protein sequence</td></tr>
<tr><td><a href="preg.html">preg</a></td><td>regular expression search of a protein sequence</td></tr>
<tr><td><a href="printsextract.html">printsextract</a></td><td>Extract data from PRINTS</td></tr>
<tr><td><a href="prosextract.html">prosextract</a></td><td>Builds the PROSITE motif database for patmatmotifs to search</td></tr>
<tr><td><a href="pscan.html">pscan</a></td><td>Scans proteins using PRINTS</td></tr>
<tr><td><a href="tfscan.html">tfscan</a></td><td>Scans DNA sequences for transcription factors</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<ul>

<li><a href="fuzzpro.html">fuzzpro</a> - Protein pattern search This
does much the same is patmatdb, but has a slightly different style of
output and allows mismatches.

<li><a href="patmatmotifs.html">patmatmotifs</a> - Compares a protein
sequence to the PROSITE motif database.  This does the opposite to
patmatdb, comparing a database of motifs to one sequence.

</ul>

<H2>
    Author(s)
</H2>


This application was written by Sinead O'Leary <a
href="mailto:soleary@hgmp.mrc.ac.uk">(soleary@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>


Completed April 30 1999.

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

