<!-- START OF HEADER -->
<HTML><HEAD>
<TITLE> EMBOSS: PDBPLUS documentation. </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">
<!-- #include file="header1.inc" -->
<H2> PDBPLUS documentation </H2>
<!-- #include file="header2.inc" -->
<!-- END OF HEADER -->





<!-- CONTENTS
     This always includes the sections below.
     Other subsections can be added for individual applications.
-->
<H2>CONTENTS </H2>
<b> <a href="#1.0">1.0     SUMMARY                   </a></b><br>
<b> <a href="#2.0">2.0     INPUTS & OUTPUTS          </a></b><br>
<b> <a href="#3.0">3.0     INPUT FILE FORMAT         </a></b><br>
<b> <a href="#4.0">4.0     OUTPUT FILE FORMAT        </a></b><br>
<b> <a href="#5.0">5.0     DATA FILES                </a></b><br>
<b> <a href="#6.0">6.0     USAGE                     </a></b><br>    
<b> <a href="#7.0">7.0     KNOWN BUGS & WARNINGS     </a></b><br>    
<b> <a href="#8.0">8.0     NOTES                     </a></b><br>
<b> <a href="#9.0">9.0     DESCRIPTION               </a></b><br>
<b> <a href="#10.0">10.0   ALGORITHM                 </a></b><br>
<b> <a href="#11.0">11.0   RELATED APPLICATIONS      </a></b><br>
<b> <a href="#12.0">12.0   DIAGNOSTIC ERROR MESSAGES </a></b><br>
<b> <a href="#13.0">13.0   AUTHORS                   </a></b><br>
<b> <a href="#14.0">14.0   REFERENCES                </a></b><br>

<!-- SUMMARY
     Succint description of the application, particularly its inputs, outputs
     and what it does.  The same text is given at the top of the source (.c)
     file and in the <documentation> attribute of the <application definition>
     of the ACD file.
-->
<a name="1.0"></a>
<H2> 1.0   SUMMARY  </H2>

<!-- #include file="inc/pdbplus.ione" -->



<!-- INPUTS & OUTPUTS
     Short summary of the application inputs and outputs in its different 
     modes of usage (if appropriate). More detail than the summary.
-->
<a name="2.0"></a>
<H2> 2.0   INPUTS & OUTPUTS          </H2>
PDBPLUS reads a directory of domain OR protein CCF files (clean coordinate files) and writes
a directory containing the same files as the input directory but in which 
values for records for secondary structure and residue accessibility are
given.  These data are calculated by using the STRIDE and NACCESS programs
and these must be installed on the system that is running PDBPLUS (see Section 8.0
below).  PDBPLUS also requires a directory containing files in PDB format
for each coordinate file in the input directory: either the original PDB file 
(for proteins) or domain PDB files generated by using DOMAINER (for domains).  The paths 
and extensions of the pdb (input) and coordinate (input and output) files are 
specified by the user (file extensions are given in the ACD file).  The names of the 
output CCF files will be identical to those that are read.

<br><br>PDBPLUS generates 3 log files containing diagnostic messages for various problems encountered by PDBPARSE, STRIDE or NACCESS respectively (see Section 12.0 below).  


<!-- INPUT FILE FORMAT 
     Description and example(s) of input file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.
     Use "<b>PDBPLUS</b> reads any normal sequence USAs." if
     appropriate.
-->
<a name="3.0"></a>
<H2> 3.0   INPUT FILE FORMAT         </H2>
The format of the CCF file is described in the <a href="pdbparse.html">PDBPARSE documentation.</a>




<!-- OUTPUT FILE FORMAT 
     Description and example(s) of output file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.  
     If the standard description of the avalable report formats is required, 
     use:   #include file="inc/reportformats.ihtml"
     Use "Outputs a graph to the specified graphics device."
     or  "outputs a report format file. The default format is ..."
     if appropriate.
-->
<a name="4.0"></a>
<H2> 4.0   OUTPUT FILE FORMAT         </H2>
The format of the CCF file is described in the <a href="pdbparse.html">PDBPARSE documentation.</a>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: pdbplus.log</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
stride output for: /ebi/services/idata/pmr/hgmp/test/data/structure/1cs4.ent
//
stride data for: /ebi/services/idata/pmr/hgmp/test/data/structure/1cs4.ent
//
naccess output for: /ebi/services/idata/pmr/hgmp/test/data/structure/1cs4.ent
//
naccess data for: /ebi/services/idata/pmr/hgmp/test/data/structure/1cs4.ent
//
stride output for: /ebi/services/idata/pmr/hgmp/test/data/structure/1ii7.ent
//
stride data for: /ebi/services/idata/pmr/hgmp/test/data/structure/1ii7.ent
//
naccess output for: /ebi/services/idata/pmr/hgmp/test/data/structure/1ii7.ent
//
naccess data for: /ebi/services/idata/pmr/hgmp/test/data/structure/1ii7.ent
//


STRIDE  failures: 0
NACCESS failures: 0
</pre>
</td></tr></table><p>
<p><h3>File: stride.log</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
</pre>
</td></tr></table><p>
<p><h3>File: naccess.log</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
naccess: using defualt vdw.radii
naccess: using default STD FILE
naccess: using defualt vdw.radii
naccess: using default STD FILE
</pre>
</td></tr></table><p>
<p><h3>File: 1cs4.ccf</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
ID   1cs4
XX
DE   MOL_ID: 1; MOLECULE: TYPE V ADENYLATE CYCLASE;
XX
OS   MOL_ID: 1; ORGANISM_SCIENTIFIC: CANIS FAMILIARIS;
XX
EX   METHOD xray; RESO 2.50; NMOD 1; NCHN 1; NGRP 0;
XX
CN   [1]
XX
IN   ID A; NR 52; NL 7; NH 2; NE 0;
XX
SQ   SEQUENCE    52 AA;   5817 MW;  47362A43 CRC32;
     ADIEGFTSLA SQCTAQELVM TLNELFARFD KLAAENHCLR IKILGDCYYC VS
XX
CO   1    1    .    P    2     396   .    .    .    .    C    .    D    ASP    N     51.711  -11.782   62.798    1.00   51.17  360.00  138.76  146.70  139.38   99.30   65.52   63.80   73.85  195.90   37.10   75.30  102.28  112.20
CO   1    1    .    P    2     396   .    .    .    .    C    .    D    ASP    CA    52.810  -11.644   61.848    1.00   54.45  360.00  138.76  146.70  139.38   99.30   65.52   63.80   73.85  195.90   37.10   75.30  102.28  112.20
CO   1    1    .    P    2     396   .    .    .    .    C    .    D    ASP    C     54.137  -11.314   62.530    1.00   55.11  360.00  138.76  146.70  139.38   99.30   65.52   63.80   73.85  195.90   37.10   75.30  102.28  112.20
CO   1    1    .    P    2     396   .    .    .    .    C    .    D    ASP    O     54.175  -10.524   63.469    1.00   55.34  360.00  138.76  146.70  139.38   99.30   65.52   63.80   73.85  195.90   37.10   75.30  102.28  112.20
CO   1    1    .    P    2     396   .    .    .    .    C    .    D    ASP    CB    52.437  -10.555   60.831    1.00   57.50  360.00  138.76  146.70  139.38   99.30   65.52   63.80   73.85  195.90   37.10   75.30  102.28  112.20
CO   1    1    .    P    2     396   .    .    .    .    C    .    D    ASP    CG    53.460  -10.391   59.729    1.00   61.38  360.00  138.76  146.70  139.38   99.30   65.52   63.80   73.85  195.90   37.10   75.30  102.28  112.20
CO   1    1    .    P    2     396   .    .    .    .    C    .    D    ASP    OD1   54.316   -9.485   59.841    1.00   65.55  360.00  138.76  146.70  139.38   99.30   65.52   63.80   73.85  195.90   37.10   75.30  102.28  112.20
CO   1    1    .    P    2     396   .    .    .    .    C    .    D    ASP    OD2   53.390  -11.146   58.736    1.00   63.68  360.00  138.76  146.70  139.38   99.30   65.52   63.80   73.85  195.90   37.10   75.30  102.28  112.20
CO   1    1    .    P    3     397   .    .    .    .    T    .    I    ILE    N     55.216  -11.941   62.066    1.00   57.14  -70.66  130.71   41.60   42.25   24.10   39.03   28.30    3.22    8.70   39.03   28.10    3.22    9.00
CO   1    1    .    P    3     397   .    .    .    .    T    .    I    ILE    CA    56.546  -11.705   62.624    1.00   59.46  -70.66  130.71   41.60   42.25   24.10   39.03   28.30    3.22    8.70   39.03   28.10    3.22    9.00
CO   1    1    .    P    3     397   .    .    .    .    T    .    I    ILE    C     57.020  -10.305   62.230    1.00   60.12  -70.66  130.71   41.60   42.25   24.10   39.03   28.30    3.22    8.70   39.03   28.10    3.22    9.00
CO   1    1    .    P    3     397   .    .    .    .    T    .    I    ILE    O     56.963   -9.927   61.060    1.00   59.12  -70.66  130.71   41.60   42.25   24.10   39.03   28.30    3.22    8.70   39.03   28.10    3.22    9.00
CO   1    1    .    P    3     397   .    .    .    .    T    .    I    ILE    CB    57.583  -12.722   62.094    1.00   60.84  -70.66  130.71   41.60   42.25   24.10   39.03   28.30    3.22    8.70   39.03   28.10    3.22    9.00
CO   1    1    .    P    3     397   .    .    .    .    T    .    I    ILE    CG1   57.184  -14.163   62.447    1.00   63.12  -70.66  130.71   41.60   42.25   24.10   39.03   28.30    3.22    8.70   39.03   28.10    3.22    9.00
CO   1    1    .    P    3     397   .    .    .    .    T    .    I    ILE    CG2   58.975  -12.384   62.632    1.00   61.24  -70.66  130.71   41.60   42.25   24.10   39.03   28.30    3.22    8.70   39.03   28.10    3.22    9.00
CO   1    1    .    P    3     397   .    .    .    .    T    .    I    ILE    CD1   57.408  -14.554   63.895    1.00   63.92  -70.66  130.71   41.60   42.25   24.10   39.03   28.30    3.22    8.70   39.03   28.10    3.22    9.00
CO   1    1    .    P    4     398   .    .    .    .    T    .    E    GLU    N     57.492   -9.548   63.212    1.00   60.23  -80.63  128.14  189.80  172.42  100.10  152.25  113.00   20.18   53.80   71.11  118.00  101.31   90.50
CO   1    1    .    P    4     398   .    .    .    .    T    .    E    GLU    CA    57.975   -8.198   62.971    1.00   62.14  -80.63  128.14  189.80  172.42  100.10  152.25  113.00   20.18   53.80   71.11  118.00  101.31   90.50
CO   1    1    .    P    4     398   .    .    .    .    T    .    E    GLU    C     59.401   -8.277   62.424    1.00   60.59  -80.63  128.14  189.80  172.42  100.10  152.25  113.00   20.18   53.80   71.11  118.00  101.31   90.50
CO   1    1    .    P    4     398   .    .    .    .    T    .    E    GLU    O     60.244   -8.972   62.987    1.00   61.84  -80.63  128.14  189.80  172.42  100.10  152.25  113.00   20.18   53.80   71.11  118.00  101.31   90.50
CO   1    1    .    P    4     398   .    .    .    .    T    .    E    GLU    CB    57.917   -7.386   64.272    1.00   65.47  -80.63  128.14  189.80  172.42  100.10  152.25  113.00   20.18   53.80   71.11  118.00  101.31   90.50
CO   1    1    .    P    4     398   .    .    .    .    T    .    E    GLU    CG    58.037   -5.874   64.091    1.00   70.25  -80.63  128.14  189.80  172.42  100.10  152.25  113.00   20.18   53.80   71.11  118.00  101.31   90.50
CO   1    1    .    P    4     398   .    .    .    .    T    .    E    GLU    CD    57.588   -5.089   65.324    1.00   72.94  -80.63  128.14  189.80  172.42  100.10  152.25  113.00   20.18   53.80   71.11  118.00  101.31   90.50
CO   1    1    .    P    4     398   .    .    .    .    T    .    E    GLU    OE1   58.262   -5.175   66.377    1.00   70.76  -80.63  128.14  189.80  172.42  100.10  152.25  113.00   20.18   53.80   71.11  118.00  101.31   90.50
CO   1    1    .    P    4     398   .    .    .    .    T    .    E    GLU    OE2   56.555   -4.380   65.232    1.00   74.36  -80.63  128.14  189.80  172.42  100.10  152.25  113.00   20.18   53.80   71.11  118.00  101.31   90.50
CO   1    1    .    P    5     399   H    1    1    1    T    .    G    GLY    N     59.642   -7.608   61.298    1.00   57.58   60.62   44.06   47.30   46.32   57.80   33.67  104.20   12.64   26.50   33.67   89.70   12.64   29.70
CO   1    1    .    P    5     399   H    1    1    1    T    .    G    GLY    CA    60.956   -7.615   60.681    1.00   56.14   60.62   44.06   47.30   46.32   57.80   33.67  104.20   12.64   26.50   33.67   89.70   12.64   29.70
CO   1    1    .    P    5     399   H    1    1    1    T    .    G    GLY    C     61.452   -8.993   60.265    1.00   58.03   60.62   44.06   47.30   46.32   57.80   33.67  104.20   12.64   26.50   33.67   89.70   12.64   29.70
CO   1    1    .    P    5     399   H    1    1    1    T    .    G    GLY    O     62.620   -9.322   60.480    1.00   57.47   60.62   44.06   47.30   46.32   57.80   33.67  104.20   12.64   26.50   33.67   89.70   12.64   29.70
CO   1    1    .    P    6     400   H    1    1    1    H    1    F    PHE    N     60.576   -9.789   59.649    1.00   58.00  -63.71  -43.06   38.10   42.33   21.20   38.88   23.70    3.46    9.80   38.88   23.50    3.46   10.10
CO   1    1    .    P    6     400   H    1    1    1    H    1    F    PHE    CA    60.914  -11.143   59.200    1.00   58.07  -63.71  -43.06   38.10   42.33   21.20   38.88   23.70    3.46    9.80   38.88   23.50    3.46   10.10
CO   1    1    .    P    6     400   H    1    1    1    H    1    F    PHE    C     61.995  -11.219   58.117    1.00   58.57  -63.71  -43.06   38.10   42.33   21.20   38.88   23.70    3.46    9.80   38.88   23.50    3.46   10.10
CO   1    1    .    P    6     400   H    1    1    1    H    1    F    PHE    O     62.862  -12.091   58.161    1.00   59.44  -63.71  -43.06   38.10   42.33   21.20   38.88   23.70    3.46    9.80   38.88   23.50    3.46   10.10
CO   1    1    .    P    6     400   H    1    1    1    H    1    F    PHE    CB    59.657  -11.881   58.734    1.00   59.18  -63.71  -43.06   38.10   42.33   21.20   38.88   23.70    3.46    9.80   38.88   23.50    3.46   10.10
CO   1    1    .    P    6     400   H    1    1    1    H    1    F    PHE    CG    59.900  -13.316   58.322    1.00   58.53  -63.71  -43.06   38.10   42.33   21.20   38.88   23.70    3.46    9.80   38.88   23.50    3.46   10.10


<font color=red>  [Part of this file has been deleted for brevity]</font>

CO   1    1    5    H    .     1002  .    .    .    .    .    .    .    FOK    C9    42.200  -11.309   50.489    1.00   41.39    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     1002  .    .    .    .    .    .    .    FOK    O6    42.275  -12.455   49.593    1.00   43.23    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     1002  .    .    .    .    .    .    .    FOK    C10   43.008  -11.601   51.811    1.00   39.11    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     1002  .    .    .    .    .    .    .    FOK    C11   40.680  -11.078   50.616    1.00   44.36    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     1002  .    .    .    .    .    .    .    FOK    O7    40.106  -10.945   51.688    1.00   48.77    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     1002  .    .    .    .    .    .    .    FOK    C12   39.943  -11.046   49.301    1.00   40.67    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     1002  .    .    .    .    .    .    .    FOK    C13   40.595  -10.085   48.292    1.00   41.47    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     1002  .    .    .    .    .    .    .    FOK    C14   40.276  -10.620   46.930    1.00   46.69    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     1002  .    .    .    .    .    .    .    FOK    C15   39.971  -11.751   46.590    1.00   53.22    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     1002  .    .    .    .    .    .    .    FOK    C16   40.047   -8.685   48.426    1.00   42.42    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     1002  .    .    .    .    .    .    .    FOK    C17   42.671   -8.737   50.253    1.00   39.67    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     1002  .    .    .    .    .    .    .    FOK    C18   46.732  -13.026   51.827    1.00   35.74    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     1002  .    .    .    .    .    .    .    FOK    C19   45.859  -11.483   53.586    1.00   34.48    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     1002  .    .    .    .    .    .    .    FOK    C20   42.913  -10.426   52.807    1.00   39.44    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     1002  .    .    .    .    .    .    .    FOK    C21   45.883   -9.553   47.821    1.00   42.15    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     1002  .    .    .    .    .    .    .    FOK    O5    46.157  -10.520   47.166    1.00   40.91    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     1002  .    .    .    .    .    .    .    FOK    C22   46.769   -8.315   48.006    1.00   37.08    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1003  .    .    .    .    .    .    .    MES    O1    45.676    7.326   49.092    1.00   77.86    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1003  .    .    .    .    .    .    .    MES    C2    44.367    6.816   48.900    1.00   75.17    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1003  .    .    .    .    .    .    .    MES    C3    44.349    5.317   48.923    1.00   74.42    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1003  .    .    .    .    .    .    .    MES    N4    44.832    4.804   50.196    1.00   72.45    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1003  .    .    .    .    .    .    .    MES    C5    46.234    5.425   50.473    1.00   73.23    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1003  .    .    .    .    .    .    .    MES    C6    46.176    6.914   50.355    1.00   75.06    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1003  .    .    .    .    .    .    .    MES    C7    44.806    3.336   50.302    1.00   73.39    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1003  .    .    .    .    .    .    .    MES    C8    44.672    2.791   51.713    1.00   76.85    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1003  .    .    .    .    .    .    .    MES    S     45.724    1.379   51.967    1.00   78.26    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1003  .    .    .    .    .    .    .    MES    O1S   47.062    1.828   51.737    1.00   79.39    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1003  .    .    .    .    .    .    .    MES    O2S   45.303    0.380   51.016    1.00   81.58    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1003  .    .    .    .    .    .    .    MES    O3S   45.523    0.961   53.326    1.00   80.59    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1004  .    .    .    .    .    .    .    MES    O1    59.246   -5.152   27.381    1.00   99.99    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1004  .    .    .    .    .    .    .    MES    C2    60.067   -4.021   27.127    1.00   99.99    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1004  .    .    .    .    .    .    .    MES    C3    60.447   -3.301   28.378    1.00   99.78    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1004  .    .    .    .    .    .    .    MES    N4    61.180   -4.156   29.270    1.00   96.33    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1004  .    .    .    .    .    .    .    MES    C5    60.358   -5.461   29.506    1.00   97.90    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1004  .    .    .    .    .    .    .    MES    C6    59.965   -6.072   28.203    1.00   99.68    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1004  .    .    .    .    .    .    .    MES    C7    61.596   -3.484   30.507    1.00   93.33    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1004  .    .    .    .    .    .    .    MES    C8    61.931   -2.010   30.442    1.00   90.74    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1004  .    .    .    .    .    .    .    MES    S     60.763   -0.978   31.301    0.50   90.72    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1004  .    .    .    .    .    .    .    MES    O1S   59.476   -1.170   30.680    0.50   91.60    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1004  .    .    .    .    .    .    .    MES    O2S   61.249    0.383   31.164    0.50   91.20    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    6    H    .     1004  .    .    .    .    .    .    .    MES    O3S   60.776   -1.430   32.647    0.50   90.05    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    7    H    .     1005  .    .    .    .    .    .    .    POP    P1    58.812   -7.766   57.091    1.00   57.40    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    7    H    .     1005  .    .    .    .    .    .    .    POP    O1    60.254   -7.589   56.745    1.00   54.93    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    7    H    .     1005  .    .    .    .    .    .    .    POP    O2    58.618   -8.839   58.095    1.00   55.36    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    7    H    .     1005  .    .    .    .    .    .    .    POP    O3    57.949   -8.024   55.908    1.00   55.10    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    7    H    .     1005  .    .    .    .    .    .    .    POP    O     58.295   -6.370   57.759    1.00   57.30    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    7    H    .     1005  .    .    .    .    .    .    .    POP    P2    56.998   -5.955   58.661    1.00   59.66    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    7    H    .     1005  .    .    .    .    .    .    .    POP    O4    57.491   -5.746   60.070    1.00   54.95    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    7    H    .     1005  .    .    .    .    .    .    .    POP    O5    56.004   -7.075   58.550    1.00   56.24    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    7    H    .     1005  .    .    .    .    .    .    .    POP    O6    56.427   -4.710   58.044    1.00   56.50    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
//
</pre>
</td></tr></table><p>
<p><h3>File: 1ii7.ccf</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
ID   1ii7
XX
DE   MOL_ID: 1; MOLECULE: MRE11 NUCLEASE;
XX
OS   MOL_ID: 1; ORGANISM_SCIENTIFIC: PYROCOCCUS FURIOSUS;
XX
EX   METHOD xray; RESO 2.20; NMOD 1; NCHN 1; NGRP 0;
XX
CN   [1]
XX
IN   ID A; NR 65; NL 5; NH 1; NE 0;
XX
SQ   SEQUENCE    65 AA;   7396 MW;  0CFB92A3 CRC32;
     MKFAHLADIH LGYEQFHKPQ REEEFAEAFK NALEIAVQEN VDFILIAGDL FHSSRPSPGT
     LKKAI
XX
CO   1    1    .    P    8     8     .    .    .    .    C    .    D    ASP    N      7.977   32.254   19.055    1.00   26.65  360.00   53.69  128.30  117.77   83.90   84.46   82.20   33.31   88.40   34.39   69.80   83.38   91.50
CO   1    1    .    P    8     8     .    .    .    .    C    .    D    ASP    CA     8.882   31.149   19.290    1.00   28.34  360.00   53.69  128.30  117.77   83.90   84.46   82.20   33.31   88.40   34.39   69.80   83.38   91.50
CO   1    1    .    P    8     8     .    .    .    .    C    .    D    ASP    C     10.273   31.305   18.659    1.00   27.71  360.00   53.69  128.30  117.77   83.90   84.46   82.20   33.31   88.40   34.39   69.80   83.38   91.50
CO   1    1    .    P    8     8     .    .    .    .    C    .    D    ASP    O     10.740   30.453   17.907    1.00   29.27  360.00   53.69  128.30  117.77   83.90   84.46   82.20   33.31   88.40   34.39   69.80   83.38   91.50
CO   1    1    .    P    8     8     .    .    .    .    C    .    D    ASP    CB     8.186   29.886   18.777    1.00   30.25  360.00   53.69  128.30  117.77   83.90   84.46   82.20   33.31   88.40   34.39   69.80   83.38   91.50
CO   1    1    .    P    8     8     .    .    .    .    C    .    D    ASP    CG     6.918   29.584   19.554    1.00   34.25  360.00   53.69  128.30  117.77   83.90   84.46   82.20   33.31   88.40   34.39   69.80   83.38   91.50
CO   1    1    .    P    8     8     .    .    .    .    C    .    D    ASP    OD1    7.041   29.177   20.724    1.00   33.37  360.00   53.69  128.30  117.77   83.90   84.46   82.20   33.31   88.40   34.39   69.80   83.38   91.50
CO   1    1    .    P    8     8     .    .    .    .    C    .    D    ASP    OD2    5.801   29.770   19.010    1.00   35.77  360.00   53.69  128.30  117.77   83.90   84.46   82.20   33.31   88.40   34.39   69.80   83.38   91.50
CO   1    1    .    P    9     9     .    .    .    .    C    .    I    ILE    N     10.913   32.423   18.967    1.00   26.84  -72.50  -37.00   42.90   42.38   24.20   41.07   29.80    1.31    3.50   41.07   29.50    1.31    3.60
CO   1    1    .    P    9     9     .    .    .    .    C    .    I    ILE    CA    12.236   32.739   18.477    1.00   26.52  -72.50  -37.00   42.90   42.38   24.20   41.07   29.80    1.31    3.50   41.07   29.50    1.31    3.60
CO   1    1    .    P    9     9     .    .    .    .    C    .    I    ILE    C     13.285   31.864   19.173    1.00   28.22  -72.50  -37.00   42.90   42.38   24.20   41.07   29.80    1.31    3.50   41.07   29.50    1.31    3.60
CO   1    1    .    P    9     9     .    .    .    .    C    .    I    ILE    O     14.257   31.426   18.550    1.00   28.01  -72.50  -37.00   42.90   42.38   24.20   41.07   29.80    1.31    3.50   41.07   29.50    1.31    3.60
CO   1    1    .    P    9     9     .    .    .    .    C    .    I    ILE    CB    12.520   34.214   18.759    1.00   27.25  -72.50  -37.00   42.90   42.38   24.20   41.07   29.80    1.31    3.50   41.07   29.50    1.31    3.60
CO   1    1    .    P    9     9     .    .    .    .    C    .    I    ILE    CG1   11.338   35.051   18.249    1.00   26.35  -72.50  -37.00   42.90   42.38   24.20   41.07   29.80    1.31    3.50   41.07   29.50    1.31    3.60
CO   1    1    .    P    9     9     .    .    .    .    C    .    I    ILE    CG2   13.823   34.640   18.094    1.00   26.40  -72.50  -37.00   42.90   42.38   24.20   41.07   29.80    1.31    3.50   41.07   29.50    1.31    3.60
CO   1    1    .    P    9     9     .    .    .    .    C    .    I    ILE    CD1   11.011   34.845   16.725    1.00   25.85  -72.50  -37.00   42.90   42.38   24.20   41.07   29.80    1.31    3.50   41.07   29.50    1.31    3.60
CO   1    1    .    P    10    10    .    .    .    .    T    .    H    HIS    N     13.072   31.599   20.462    1.00   26.69   50.63   45.81   76.20   75.98   41.50   75.67   51.40    0.31    0.90   48.67   50.10   27.31   31.90
CO   1    1    .    P    10    10    .    .    .    .    T    .    H    HIS    CA    14.009   30.787   21.254    1.00   28.02   50.63   45.81   76.20   75.98   41.50   75.67   51.40    0.31    0.90   48.67   50.10   27.31   31.90
CO   1    1    .    P    10    10    .    .    .    .    T    .    H    HIS    C     15.483   31.203   21.181    1.00   27.57   50.63   45.81   76.20   75.98   41.50   75.67   51.40    0.31    0.90   48.67   50.10   27.31   31.90
CO   1    1    .    P    10    10    .    .    .    .    T    .    H    HIS    O     16.357   30.359   21.019    1.00   23.99   50.63   45.81   76.20   75.98   41.50   75.67   51.40    0.31    0.90   48.67   50.10   27.31   31.90
CO   1    1    .    P    10    10    .    .    .    .    T    .    H    HIS    CB    13.867   29.284   20.899    1.00   27.15   50.63   45.81   76.20   75.98   41.50   75.67   51.40    0.31    0.90   48.67   50.10   27.31   31.90
CO   1    1    .    P    10    10    .    .    .    .    T    .    H    HIS    CG    12.599   28.681   21.415    1.00   28.59   50.63   45.81   76.20   75.98   41.50   75.67   51.40    0.31    0.90   48.67   50.10   27.31   31.90
CO   1    1    .    P    10    10    .    .    .    .    T    .    H    HIS    ND1   12.536   27.980   22.603    1.00   30.06   50.63   45.81   76.20   75.98   41.50   75.67   51.40    0.31    0.90   48.67   50.10   27.31   31.90
CO   1    1    .    P    10    10    .    .    .    .    T    .    H    HIS    CD2   11.319   28.794   20.978    1.00   26.35   50.63   45.81   76.20   75.98   41.50   75.67   51.40    0.31    0.90   48.67   50.10   27.31   31.90
CO   1    1    .    P    10    10    .    .    .    .    T    .    H    HIS    CE1   11.276   27.691   22.875    1.00   27.36   50.63   45.81   76.20   75.98   41.50   75.67   51.40    0.31    0.90   48.67   50.10   27.31   31.90
CO   1    1    .    P    10    10    .    .    .    .    T    .    H    HIS    NE2   10.519   28.180   21.909    1.00   28.44   50.63   45.81   76.20   75.98   41.50   75.67   51.40    0.31    0.90   48.67   50.10   27.31   31.90
CO   1    1    .    P    11    11    .    .    .    .    T    .    L    LEU    N     15.758   32.503   21.303    1.00   26.85  -57.60  139.21   78.50   79.58   44.50   55.49   39.30   24.09   64.20   56.43   39.70   23.15   63.70
CO   1    1    .    P    11    11    .    .    .    .    T    .    L    LEU    CA    17.152   32.972   21.286    1.00   27.57  -57.60  139.21   78.50   79.58   44.50   55.49   39.30   24.09   64.20   56.43   39.70   23.15   63.70
CO   1    1    .    P    11    11    .    .    .    .    T    .    L    LEU    C     17.879   32.257   22.409    1.00   28.11  -57.60  139.21   78.50   79.58   44.50   55.49   39.30   24.09   64.20   56.43   39.70   23.15   63.70
CO   1    1    .    P    11    11    .    .    .    .    T    .    L    LEU    O     17.323   32.093   23.505    1.00   29.47  -57.60  139.21   78.50   79.58   44.50   55.49   39.30   24.09   64.20   56.43   39.70   23.15   63.70
CO   1    1    .    P    11    11    .    .    .    .    T    .    L    LEU    CB    17.225   34.494   21.530    1.00   25.89  -57.60  139.21   78.50   79.58   44.50   55.49   39.30   24.09   64.20   56.43   39.70   23.15   63.70
CO   1    1    .    P    11    11    .    .    .    .    T    .    L    LEU    CG    16.526   35.348   20.465    1.00   23.73  -57.60  139.21   78.50   79.58   44.50   55.49   39.30   24.09   64.20   56.43   39.70   23.15   63.70
CO   1    1    .    P    11    11    .    .    .    .    T    .    L    LEU    CD1   16.604   36.805   20.823    1.00   24.72  -57.60  139.21   78.50   79.58   44.50   55.49   39.30   24.09   64.20   56.43   39.70   23.15   63.70
CO   1    1    .    P    11    11    .    .    .    .    T    .    L    LEU    CD2   17.179   35.102   19.081    1.00   22.96  -57.60  139.21   78.50   79.58   44.50   55.49   39.30   24.09   64.20   56.43   39.70   23.15   63.70


<font color=red>  [Part of this file has been deleted for brevity]</font>

CO   1    1    .    P    50    50    .    .    .    .    C    .    L    LEU    CD2   12.425   39.035   22.798    1.00   23.77  -69.71  360.00   98.90  101.26   56.70   78.17   55.40   23.09   61.60   96.01   67.50    5.25   14.50
CO   1    1    1    H    .     402   .    .    .    .    .    .    .    PO4    P     34.178   32.996   46.387    1.00   60.84    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    1    H    .     402   .    .    .    .    .    .    .    PO4    O1    35.146   33.243   45.291    1.00   57.95    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    1    H    .     402   .    .    .    .    .    .    .    PO4    O2    34.912   32.751   47.670    1.00   59.15    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    1    H    .     402   .    .    .    .    .    .    .    PO4    O3    33.291   34.184   46.538    1.00   58.92    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    1    H    .     402   .    .    .    .    .    .    .    PO4    O4    33.352   31.796   46.060    1.00   61.86    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    2    H    .     403   .    .    .    .    .    .    .     MN    MN     8.130   27.788   21.899    1.00   36.09    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    2    H    .     404   .    .    .    .    .    .    .     MN    MN     5.801   27.935   24.271    1.00   39.57    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    3    H    .     405   .    .    .    .    .    .    .     MN    MN    36.023   34.916   44.253    1.00   39.52    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    3    H    .     406   .    .    .    .    .    .    .     MN    MN    33.658   36.365   46.296    1.00   33.69    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    4    H    .     501   .    .    .    .    .    .    .    SO4    S     17.175   28.112   32.476    1.00  100.80    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    4    H    .     501   .    .    .    .    .    .    .    SO4    O1    18.136   28.230   31.357    1.00  100.18    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    4    H    .     501   .    .    .    .    .    .    .    SO4    O2    17.097   26.692   32.887    1.00  100.80    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    4    H    .     501   .    .    .    .    .    .    .    SO4    O3    17.633   28.926   33.626    1.00  100.14    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    4    H    .     501   .    .    .    .    .    .    .    SO4    O4    15.834   28.575   32.045    1.00  100.56    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    4    H    .     502   .    .    .    .    .    .    .    SO4    S      0.566   29.512   36.007    1.00   86.73    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    4    H    .     502   .    .    .    .    .    .    .    SO4    O1     1.690   28.556   35.971    1.00   87.27    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    4    H    .     502   .    .    .    .    .    .    .    SO4    O2    -0.620   28.803   36.523    1.00   87.87    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    4    H    .     502   .    .    .    .    .    .    .    SO4    O3     0.896   30.642   36.905    1.00   86.58    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    4    H    .     502   .    .    .    .    .    .    .    SO4    O4     0.287   30.037   34.658    1.00   86.51    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    4    H    .     503   .    .    .    .    .    .    .    SO4    S    -13.586   39.644   36.031    1.00  100.28    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    4    H    .     503   .    .    .    .    .    .    .    SO4    O1   -12.340   39.512   35.250    1.00  100.72    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    4    H    .     503   .    .    .    .    .    .    .    SO4    O2   -14.638   38.811   35.421    1.00  100.46    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    4    H    .     503   .    .    .    .    .    .    .    SO4    O3   -13.347   39.201   37.420    1.00   99.66    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    4    H    .     503   .    .    .    .    .    .    .    SO4    O4   -14.020   41.056   36.015    1.00   99.97    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     401   .    .    .    .    .    .    .    101    P      7.599   25.305   23.994    1.00   56.33    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     401   .    .    .    .    .    .    .    101    O1P    8.249   24.467   25.030    1.00   56.70    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     401   .    .    .    .    .    .    .    101    O2P    6.700   26.285   24.649    1.00   54.49    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     401   .    .    .    .    .    .    .    101    O3P    8.637   26.026   23.216    1.00   53.97    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     401   .    .    .    .    .    .    .    101    O5*    7.095   23.970   23.128    1.00   59.20    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     401   .    .    .    .    .    .    .    101    C5*    7.073   23.961   21.762    1.00   66.74    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     401   .    .    .    .    .    .    .    101    C4*    6.041   23.013   21.296    1.00   71.22    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     401   .    .    .    .    .    .    .    101    O4*    6.029   21.855   22.189    1.00   73.78    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     401   .    .    .    .    .    .    .    101    C3*    4.736   23.676   21.350    1.00   73.80    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     401   .    .    .    .    .    .    .    101    O3*    4.355   23.874   19.995    1.00   76.51    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     401   .    .    .    .    .    .    .    101    C2*    3.864   22.749   22.165    1.00   74.04    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     401   .    .    .    .    .    .    .    101    C1*    4.682   21.474   22.506    1.00   74.70    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     401   .    .    .    .    .    .    .    101    N9     4.578   21.123   23.969    1.00   76.71    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     401   .    .    .    .    .    .    .    101    C8     3.630   21.533   24.876    1.00   76.87    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     401   .    .    .    .    .    .    .    101    N7     3.758   21.069   26.081    1.00   77.50    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     401   .    .    .    .    .    .    .    101    C5     4.896   20.300   25.989    1.00   77.78    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     401   .    .    .    .    .    .    .    101    C6     5.570   19.479   26.941    1.00   78.16    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     401   .    .    .    .    .    .    .    101    N6     5.155   19.409   28.200    1.00   78.77    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     401   .    .    .    .    .    .    .    101    N1     6.682   18.805   26.554    1.00   78.32    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     401   .    .    .    .    .    .    .    101    C2     7.090   18.888   25.277    1.00   78.14    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     401   .    .    .    .    .    .    .    101    N3     6.541   19.611   24.271    1.00   78.05    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    1    5    H    .     401   .    .    .    .    .    .    .    101    C4     5.403   20.288   24.700    1.00   78.10    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    .    .    W    .     407   .    .    .    .    .    .    .    HOH    O      5.997   27.242   22.189    1.00   38.84    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    .    .    W    .     408   .    .    .    .    .    .    .    HOH    O     35.697   35.756   46.350    1.00   41.39    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
CO   1    .    .    W    .     600   .    .    .    .    .    .    .    HOH    O     20.825   31.690   27.031    1.00   20.90    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
//
</pre>
</td></tr></table><p>



<!-- DATA FILES         
     Any data files used (e.g. translation table file, substitution matrix 
     etc.  This includes example data file formats if they are not obvious.
     For a standard description of what data files are and how embossdata can
     be used to inspect and retrieve them, use:
     #include file="inc/localfiles.ihtml" 
-->

<a name="5.0"></a>
<H2> 5.0   DATA FILES                </H2>
PDBPLUS does not use a data file.







<!-- USAGE
     Example usage, as run from the command-line.
     Many examples illustrating different behaviours is good.
-->
<a name="6.0"></a>
<H2> 6.0   USAGE                     </H2>

<pre>
   Standard (Mandatory) qualifiers (* if not always prompted):
  [-ccfinpath]         dirlist    This option specifies the location of CCF
                                  files (clean coordinate files) (input). A
                                  'clean cordinate file' contains protein
                                  coordinate and derived data for a single PDB
                                  file ('protein clean coordinate file') or a
                                  single domain from SCOP or CATH ('domain
                                  clean coordinate file'), in CCF format
                                  (EMBL-like). The files, generated by using
                                  PDBPARSE (PDB files) or DOMAINER (domains),
                                  contain 'cleaned-up' data that is
                                  self-consistent and error-corrected. Records
                                  for residue solvent accessibility and
                                  secondary structure are added to the file by
                                  using PDBPLUS.
  [-pdbindir]          directory  This option specifies the location of PDB
                                  files (input). A PDB file contains protein
                                  coordinate and other data. A detailed
                                  explanation of the PDB file format is
                                  available on the PDB web site
                                  http://www.rcsb.org/pdb/info.html
   -pdbprefix          string     This option specifies the file prefix of PDB
                                  files (if any)
   -mode               menu       This option specifies the mode of operation.
*  -thresholdsize      integer    This option specifies the threshold size for
                                  SSEs (secondary structure elements)
  [-ccfoutdir]         outdir     This option specifies the location of CCF
                                  files (clean coordinate files) (output). A
                                  'clean cordinate file' contains protein
                                  coordinate and derived data for a single PDB
                                  file ('protein clean coordinate file') or a
                                  single domain from SCOP or CATH ('domain
                                  clean coordinate file'), in CCF format
                                  (EMBL-like). The files, generated by using
                                  PDBPARSE (PDB files) or DOMAINER (domains),
                                  contain 'cleaned-up' data that is
                                  self-consistent and error-corrected. Records
                                  for residue solvent accessibility and
                                  secondary structure are added to the file by
                                  using PDBPLUS.
   -errfile            outfile    This option specifies the name of the log
                                  file for PDBPLUS
*  -serrfile           outfile    This option specifies the name of the log
                                  file for STRIDE
*  -nerrfile           outfile    This option specifies the name of the log
                                  file for NACCESS

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-errfile" associated qualifiers
   -odirectory         string     Output directory

   "-serrfile" associated qualifiers
   -odirectory         string     Output directory

   "-nerrfile" associated qualifiers
   -odirectory         string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report deaths


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-ccfinpath]<br>(Parameter 1)</td>
<td>This option specifies the location of CCF files (clean coordinate files) (input). A 'clean cordinate file' contains protein coordinate and derived data for a single PDB file ('protein clean coordinate file') or a single domain from SCOP or CATH ('domain clean coordinate file'), in CCF format (EMBL-like). The files, generated by using PDBPARSE (PDB files) or DOMAINER (domains), contain 'cleaned-up' data that is self-consistent and error-corrected. Records for residue solvent accessibility and secondary structure are added to the file by using PDBPLUS.</td>
<td>Directory with files</td>
<td>./</td>
</tr>

<tr>
<td>[-pdbindir]<br>(Parameter 2)</td>
<td>This option specifies the location of PDB files (input). A PDB file contains protein coordinate and other data. A detailed explanation of the PDB file format is available on the PDB web site http://www.rcsb.org/pdb/info.html</td>
<td>Directory</td>
<td>./</td>
</tr>

<tr>
<td>-pdbprefix</td>
<td>This option specifies the file prefix of PDB files (if any)</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-mode</td>
<td>This option specifies the mode of operation.</td>
<td><table><tr><td>1</td> <td><i>(Use STRIDE only)</i></td></tr><tr><td>2</td> <td><i>(Use NACCESS only)</i></td></tr><tr><td>3</td> <td><i>(Use STRIDE and NACCESS)</i></td></tr></table></td>
<td>1</td>
</tr>

<tr>
<td>-thresholdsize</td>
<td>This option specifies the threshold size for SSEs (secondary structure elements)</td>
<td>Any integer value</td>
<td>4</td>
</tr>

<tr>
<td>[-ccfoutdir]<br>(Parameter 3)</td>
<td>This option specifies the location of CCF files (clean coordinate files) (output). A 'clean cordinate file' contains protein coordinate and derived data for a single PDB file ('protein clean coordinate file') or a single domain from SCOP or CATH ('domain clean coordinate file'), in CCF format (EMBL-like). The files, generated by using PDBPARSE (PDB files) or DOMAINER (domains), contain 'cleaned-up' data that is self-consistent and error-corrected. Records for residue solvent accessibility and secondary structure are added to the file by using PDBPLUS.</td>
<td>Output directory</td>
<td>./</td>
</tr>

<tr>
<td>-errfile</td>
<td>This option specifies the name of the log file for PDBPLUS</td>
<td>Output file</td>
<td>pdbplus.log</td>
</tr>

<tr>
<td>-serrfile</td>
<td>This option specifies the name of the log file for STRIDE</td>
<td>Output file</td>
<td>stride.log</td>
</tr>

<tr>
<td>-nerrfile</td>
<td>This option specifies the name of the log file for NACCESS</td>
<td>Output file</td>
<td>naccess.log</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H3> 6.2 EXAMPLE SESSION </H3>
An example of interactive use of PDBPLUS is shown below.

<b>Here is a sample session with pdbplus</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>pdbplus </b>
Add residue solvent accessibility and secondary structure data to a CCF
file (clean coordinate file) for a protein or domain.
Location of CCF files (clean coordinate files) (input) [./]: <b>pdbplus</b>
Location of PDB files (input) [./]: <b></b>
File prefix of PDB files (if any): <b></b>
Mode of operation
         1 : Use STRIDE only
         2 : Use NACCESS only
         3 : Use STRIDE and NACCESS
Select mode [1]: <b>3</b>
Threshold size for SSEs (secondary structure elements) [4]: <b>4</b>
Location of CCF files (clean coordinate files) (output) [./]: <b></b>
Name of log file for PDBPLUS [pdbplus.log]: <b></b>
Name of log file for STRIDE [stride.log]: <b></b>
Name of log file for NACCESS [naccess.log]: <b></b>

Processing /ebi/services/idata/pmr/hgmp/test/data/structure/pdbplus/1cs4.ccf
stride /ebi/services/idata/pmr/hgmp/test/data/structure/1cs4.ent -f./pdbplus-1108721730.13957 >> stride.log 2>&1
naccess /ebi/services/idata/pmr/hgmp/test/data/structure/1cs4.ent  >> naccess.log 2>&1
Processing /ebi/services/idata/pmr/hgmp/test/data/structure/pdbplus/1ii7.ccf
stride /ebi/services/idata/pmr/hgmp/test/data/structure/1ii7.ent -f./pdbplus-1108721730.13957 >> stride.log 2>&1
naccess /ebi/services/idata/pmr/hgmp/test/data/structure/1ii7.ent  >> naccess.log 2>&1
STRIDE  failures: 0
NACCESS failures: 0

</pre></td></tr></table><p>
<p>
<a href="#output.1">Go to the output files for this example</a><p><p>


All CCF files were read from testdata/pdbplus/ccfin and CCF files
(annotated with values for secondary structure and solvent
accessibility) were written to testdata/pdbplus/ccfout.  The STRIDE
and NACCESS programs ran on PDB files within testdata/. Three log
files (pdbplus.log, naccess.log and stride.log) were written. The
threshold size for secondary structure elements was 4 residues.  File
extension of all input and output files would be set in the ACD file.



<!-- KNOWN BUGS & WARNINGS
     Bugs that have not yet been fixed, easily missued features, problems
     and caveats etc. Potentially stupid things the program will let you do.
-->
<a name="7.0"></a>
<H2> 7.0   KNOWN BUGS & WARNINGS     </H2>
NACCESS and STRIDE do not always work as advertised owing, e.g. to arbitrary array limits in their code.  If the derived data appears to be missing from the PDDBPARSE output this is probably why.





<!-- NOTES
     Important general remarks, including:
       Restrictions.
       Interesting behaviour.
       Useful things you can do with this program.
       Future plans.
       etc.
-->
<a name="8.0"></a>
<H2> 8.0   NOTES                     </H2>
PDBPLUS requires a working version of NACCESS and STRIDE.
When running PDBPLUS at the RFCGR it is essential that the commands 'use naccess' or 
'use stride' (which run the scripts /packages/menu/USE/naccess and /packages/menu/USE/stride respectively) 
are given before it is run.  
<br>
<br>


<H3> 8.1   GLOSSARY OF FILE TYPES    </H3>
<table BORDER CELLSPACING=5 CELLPADDING=5 BGCOLOR="#f5f5ff" >
<tr>
<td><b>FILE TYPE</b></td>
<td><b>FORMAT</b></td>
<td><b>DESCRIPTION</b></td>
<td><b>CREATED BY <b></td>
<td><b>SEE ALSO</b></td>
</tr>
<tr>
<td><b> Clean coordinate file (for protein) </b></td>
<td> CCF format (EMBL-like).</td>
<td> Protein coordinate and derived data for a single PDB file.  The data are 'cleaned-up': self-consistent and error-corrected.  </td>
<td> <a href="pdbparse.html">PDBPARSE</a> </td>
<td> Records for residue solvent accessibility and secondary structure are added to the file by using PDBPLUS. </td>
</tr>
<tr>
<td><b> Clean coordinate file (for domain)</b></td>
<td> CCF format (EMBL-like).</td>
<td> Protein coordinate and derived data for a single domain from SCOP or CATH. The data are 'cleaned-up': self-consistent and error-corrected. </td>
<td> <a href="domainer.html">DOMAINER</a></td>
<td> Records for residue solvent accessibility and secondary structure are added to the file by using PDBPLUS.</td>
</tr>
<tr>
<td><b>PDB file </b></td>
<td> PDB format. </td>
<td> Protein coordinate data in PDB format. </td>
<td> N.A. </td>
<td> N.A. </td>
</tr>
<tr>
<td><b>Domain PDB file </b></td>
<td> PDB format. </td>
<td> Protein coordinate data for a single domain from SCOP or CATH. </td>
<td> <a href="domainer.html">DOMAINER</a> </td>
<td> N.A. </td>
</tr>
</table>



None



<a name="9.0"></a>
<H2> 9.0   DESCRIPTION               </H2>
PDBPLUS is a wrapper to the STRIDE [1] and NACCESS [2] programs that adds derived data (i - iv below) to a CCF file.  The values of the derived data for a given residue will of course depend on whether they were calculated from the coordinates for the entire structure or just the domain, i.e. whether PDBPLUS is provided with protein or domain CCF files.  Thus PDBPLUS provides useful flexibility in generating the derived data.
<br>
<UL>
<LI>  Absolute (Abs.) and relative (Rel.) accessible surface area of residues according to NACCESS.  Abs. is the summed accessible surface area of the atoms, whereas Rel. is expressed as a percentage relative to the accessibility of the atoms in an extended ALA-x-ALA tri-peptide conformation.  The NACCESS authors treat alpha carbons as side-chain atoms so that glycine can have a value for side-chain accessible surface area (see 'Database format' below).  They are therefore not included in the main-chain.  
<LI>  Phi and Psi angle and solvent accessible surface area of residues as calculated by using STRIDE.  
<LI>  Secondary structure assignment according to STRIDE, one of 'H' (alpha helix), 'G' (3-10 helix), 'I' (Pi-helix), 'E' (extended conformation), 'B' or 'b' (isolated bridge), 'T' (turn) or 'C' (coil, i.e. none of the above).  
<LI>  The number of helical or beta-strand secondary structure elements (SSEs) in the chain or domain.  An 'element' is defined as a run of a user-defined number (typically 4) of residues in the 'H', 'G' or 'I' conformation (helices) or the 'E' conformation (beta-strands).  
</UL>



<!-- ALGORITHM
     A technical description of algorithmic aspects, describing exactly how
     the key aspects of the application work.
-->
<a name="10.0"></a>
<H2> 10.0  ALGORITHM                 </H2>
Values in the CCF file for the number of helices (NH) or beta-strands (NE) in a chain and columns 12-13 and 22-34 of the coordinate line record (CO) are given null values ('.' or 0) by PDBPARSE (see Section 4.0).  PDBPLUS is used to assign values to these these records.



<!-- RELATED APPLICATIONS
     Other applications that either generate the input, use the output or 
     are in some other way related to the application are described here. 
-->
<a name="11.0"></a>
<br><H2> 11.0  RELATED APPLICATIONS      </H2>


<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="../../apps/aaindexextract.html">aaindexextract</a></td><td>Extract data from AAINDEX</td></tr>
<tr><td><a href="allversusall.html">allversusall</a></td><td>Does an all-versus-all global alignment for each set of sequences in an input directory and writes files of sequence similarity values</td></tr>
<tr><td><a href="cathparse.html">cathparse</a></td><td>Reads raw CATH classification files and writes DCF file (domain classification file)</td></tr>
<tr><td><a href="../../apps/cutgextract.html">cutgextract</a></td><td>Extract data from CUTG</td></tr>
<tr><td><a href="domainer.html">domainer</a></td><td>Reads CCF files (clean coordinate files) for proteins and writes CCF files for domains, taken from a DCF file (domain classification file)</td></tr>
<tr><td><a href="domainnr.html">domainnr</a></td><td>Removes redundant domains from a DCF file (domain classification file). The file must contain domain sequence information, which can be added by using DOMAINSEQS</td></tr>
<tr><td><a href="domainseqs.html">domainseqs</a></td><td>Adds sequence records to a DCF file (domain classification file)</td></tr>
<tr><td><a href="domainsse.html">domainsse</a></td><td>Adds secondary structure records to a DCF file (domain classification file)</td></tr>
<tr><td><a href="hetparse.html">hetparse</a></td><td>Converts raw dictionary of heterogen groups to a file in EMBL-like format</td></tr>
<tr><td><a href="pdbparse.html">pdbparse</a></td><td>Parses PDB files and writes CCF files (clean coordinate files) for proteins</td></tr>
<tr><td><a href="pdbtosp.html">pdbtosp</a></td><td>Convert raw swissprot:PDB equivalence file to EMBL-like format</td></tr>
<tr><td><a href="../../apps/printsextract.html">printsextract</a></td><td>Extract data from PRINTS</td></tr>
<tr><td><a href="../../apps/prosextract.html">prosextract</a></td><td>Builds the PROSITE motif database for patmatmotifs to search</td></tr>
<tr><td><a href="../../apps/rebaseextract.html">rebaseextract</a></td><td>Extract data from REBASE</td></tr>
<tr><td><a href="scopparse.html">scopparse</a></td><td>Reads raw SCOP classification files and writes a DCF file (domain classification file)</td></tr>
<tr><td><a href="seqnr.html">seqnr</a></td><td>Removes redundancy from DHF files (domain hits files) or other files of sequences</td></tr>
<tr><td><a href="sites.html">sites</a></td><td>Reads CCF files (clean coordinate files) and writes CON files (contact files) of residue-ligand contact data for domains in a DCF file (domain classification file)</td></tr>
<tr><td><a href="ssematch.html">ssematch</a></td><td>Searches a DCF file (domain classification file) for secondary structure matches</td></tr>
<tr><td><a href="../../apps/tfextract.html">tfextract</a></td><td>Extract data from TRANSFAC</td></tr>
</table>



<!-- DIAGNOSTIC ERROR MESSAGES
     Description of error messages or log file, if one is written.
-->
<a name="12.0"></a>
<H2> 12.0  DIAGNOSTIC ERROR MESSAGES </H2>
The following messages (in "quotes") may appear in the PDBPARSE log file (an explanation is given in parenethesis). 

<UL>
<LI>"clean coordinate file not found my.file" (a clean coordinate input file was not found for input)
<LI>"ERROR Clean coordinate file read error: my.file" (error in reading clean coordinate file)
<LI>"no stride output for: my.file" (No STRIDE output was generated for my.file)
<LI>"stride output for: my.file" (STRIDE output *was* generated for my.file)
<LI>"no naccess output for:" (No NACCESS output was generated for my.file)
<LI>"naccess output for:" (NACCESS output *was* generated for my.file)
<LI>"PDB file not found: my.file" (a pdb file was not found for input)
<LI>"Could not write results file: my.file" (output file could not be written)
<LI>"Could not convert chain id A to chain number in pdb file my.file" (chain identifier was not found in the clean coordinate file)
<LI>"STRIDE  failures: %12\nNACCESS failures: 123" (STRIDE failed for 12 pdb files and NACCESS failed for 123 pdb files.  The clean coordinate files will lack the appropriate data (see Section 10.0).
</UL>

<!-- AUTHORS
     The main author first then all others. 
-->
<a name="13.0"></a>
<H2> 13.0  AUTHORS                   </H2>
Jon Ison <a href="mailto:jison@rfcgr.mrc.ac.uk">(jison@rfcgr.mrc.ac.uk)</a>
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK



<!-- REFERENCES
     Quote the paper where the application was first published, described, used etc. 
-->
<a name="14.0"></a>
<H2> 14.0  REFERENCES                </H2>
Please cite the authors and EMBOSS.


<H3>14.1 Other useful references  </H3>
<pre>
1. Frishman, D. & Argos, P. (1995).  Knowledge-Based Secondary Structure Assignment.  Proteins 23:566-579.

2. Hubbard,S.J. & Thornton, J.M. (1993), 'NACCESS', Computer Program, Department of Biochemistry and Molecular Biology, University College London.
</pre>
</BODY>
</HTML>
