<!-- START OF HEADER -->
<HTML><HEAD>
<TITLE> EMBASSY: PDBPLUS documentation. </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

<H2> PDBPLUS documentation
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<!-- END OF HEADER -->






<!-- CONTENTS
     This always includes the sections below.
     Other subsections can be added for individual applications.
-->
<H2>CONTENTS </H2>
<b> <a href="#1.0">1.0     SUMMARY                   </a></b><br>
<b> <a href="#2.0">2.0     INPUTS & OUTPUTS          </a></b><br>
<b> <a href="#3.0">3.0     INPUT FILE FORMAT         </a></b><br>
<b> <a href="#4.0">4.0     OUTPUT FILE FORMAT        </a></b><br>
<b> <a href="#5.0">5.0     DATA FILES                </a></b><br>
<b> <a href="#6.0">6.0     USAGE                     </a></b><br>    
<b> <a href="#7.0">7.0     KNOWN BUGS & WARNINGS     </a></b><br>    
<b> <a href="#8.0">8.0     NOTES                     </a></b><br>
<b> <a href="#9.0">9.0     DESCRIPTION               </a></b><br>
<b> <a href="#10.0">10.0   ALGORITHM                 </a></b><br>
<b> <a href="#11.0">11.0   RELATED APPLICATIONS      </a></b><br>
<b> <a href="#12.0">12.0   DIAGNOSTIC ERROR MESSAGES </a></b><br>
<b> <a href="#13.0">13.0   AUTHORS                   </a></b><br>
<b> <a href="#14.0">14.0   REFERENCES                </a></b><br>

<!-- SUMMARY
     Succint description of the application, particularly its inputs, outputs
     and what it does.  The same text is given at the top of the source (.c)
     file and in the <documentation> attribute of the <application definition>
     of the ACD file.
-->
<a name="1.0"></a>
<br><br><br><H2> 1.0   SUMMARY  </H2>

<!-- #include file="inc/pdbplus.ione" -->



<!-- INPUTS & OUTPUTS
     Short summary of the application inputs and outputs in its different 
     modes of usage (if appropriate). More detail than the summary.
-->
<a name="2.0"></a>
<br><br><br><H2> 2.0   INPUTS & OUTPUTS          </H2>
PDBPLUS reads a directory of domain OR protein CCF files (clean coordinate files) and writes
a directory containing the same files as the input directory but in which 
values for records for secondary structure and residue accessibility are
given.  These data are calculated by using the STRIDE and NACCESS programs
and these must be installed on the system that is running PDBPLUS (see Section 8.0
below).  PDBPLUS also requires a directory containing files in PDB format
for each coordinate file in the input directory: either the original PDB file 
(for proteins) or domain PDB files generated by using DOMAINER (for domains).  The paths 
and extensions of the pdb (input) and coordinate (input and output) files are 
specified by the user (file extensions are given in the ACD file).  The names of the 
output CCF files will be identical to those that are read.

<br><br>PDBPLUS generates 3 log files containing diagnostic messages for various problems encountered by PDBPARSE, STRIDE or NACCESS respectively (see Section 12.0 below).  


<!-- INPUT FILE FORMAT 
     Description and example(s) of input file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.
     Use "<b>PDBPLUS</b> reads any normal sequence USAs." if
     appropriate.
-->
<a name="3.0"></a>
<br><br><br><H2> 3.0   INPUT FILE FORMAT         </H2>
The format of the CCF file is described in the <a href="pdbparse.html">PDBPARSE documentation.</a>




<!-- OUTPUT FILE FORMAT 
     Description and example(s) of output file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.  
     If the standard description of the avalable report formats is required, 
     use:   #include file="inc/reportformats.ihtml"
     Use "Outputs a graph to the specified graphics device."
     or  "outputs a report format file. The default format is ..."
     if appropriate.
-->
<a name="4.0"></a>
<br><br><br><H2> 4.0   OUTPUT FILE FORMAT         </H2>
The format of the CCF file is described in the <a href="pdbparse.html">PDBPARSE documentation.</a>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: pdbplus.log</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
stride output for: /ebi/services/idata/pmr/hgmp/test/data/structure/2hhb.ent
//
stride data for: /ebi/services/idata/pmr/hgmp/test/data/structure/2hhb.ent
//
naccess output for: /ebi/services/idata/pmr/hgmp/test/data/structure/2hhb.ent
//
naccess data for: /ebi/services/idata/pmr/hgmp/test/data/structure/2hhb.ent
//
stride output for: /ebi/services/idata/pmr/hgmp/test/data/structure/1cs4.ent
//
stride data for: /ebi/services/idata/pmr/hgmp/test/data/structure/1cs4.ent
//
naccess output for: /ebi/services/idata/pmr/hgmp/test/data/structure/1cs4.ent
//
naccess data for: /ebi/services/idata/pmr/hgmp/test/data/structure/1cs4.ent
//
stride output for: /ebi/services/idata/pmr/hgmp/test/data/structure/1ii7.ent
//
stride data for: /ebi/services/idata/pmr/hgmp/test/data/structure/1ii7.ent
//
naccess output for: /ebi/services/idata/pmr/hgmp/test/data/structure/1ii7.ent
//
naccess data for: /ebi/services/idata/pmr/hgmp/test/data/structure/1ii7.ent
//


STRIDE  failures: 0
NACCESS failures: 0
</pre>
</td></tr></table><p>
<p><h3>File: stride.log</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
</pre>
</td></tr></table><p>
<p><h3>File: naccess.log</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
naccess: using defualt vdw.radii
naccess: using default STD FILE
naccess: using defualt vdw.radii
naccess: using default STD FILE
naccess: using defualt vdw.radii
naccess: using default STD FILE
</pre>
</td></tr></table><p>
<p><h3>File: 1cs4.ccf</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
ID   1cs4
XX
DE   MOL_ID: 1; MOLECULE: TYPE V ADENYLATE CYCLASE;
XX
OS   MOL_ID: 1; ORGANISM_SCIENTIFIC: CANIS FAMILIARIS;
XX
EX   METHOD xray; RESO 2.50; NMOD 1; NCHN 1; NGRP 0;
XX
CN   [1]
XX
IN   ID A; NR 52; NL 7; NH 2; NE 0;
XX
SQ   SEQUENCE    52 AA;   5817 MW;  47362A43 CRC32;
     ADIEGFTSLA SQCTAQELVM TLNELFARFD KLAAENHCLR IKILGDCYYC VS
XX
RE   1    1    2    396   D ASP   .    .    .    .    .    C      360.00  138.76  146.70  139.38   99.30   65.52   63.80   73.85  195.90   37.10   75.30  102.28  112.20
RE   1    1    3    397   I ILE   .    .    .    .    .    T      -70.66  130.71   41.60   42.25   24.10   39.03   28.30    3.22    8.70   39.03   28.10    3.22    9.00
RE   1    1    4    398   E GLU   .    .    .    .    .    T      -80.63  128.14  189.80  172.42  100.10  152.25  113.00   20.18   53.80   71.11  118.00  101.31   90.50
RE   1    1    5    399   G GLY   1    1    H    1    .    T       60.62   44.06   47.30   46.32   57.80   33.67  104.20   12.64   26.50   33.67   89.70   12.64   29.70
RE   1    1    6    400   F PHE   1    1    H    1    1    H      -63.71  -43.06   38.10   42.33   21.20   38.88   23.70    3.46    9.80   38.88   23.50    3.46   10.10
RE   1    1    7    401   T THR   1    1    H    1    1    H      -63.13  -33.45  122.50  117.00   84.00  113.46  111.60    3.54    9.40   81.54  107.70   35.46   55.80
RE   1    1    8    402   S SER   1    1    H    1    1    H      -69.61  -44.64   76.80   69.25   59.40   66.11   84.60    3.14    8.20   36.50   75.20   32.75   48.20
RE   1    1    9    403   L LEU   1    1    H    1    1    H      -63.95  -53.38   81.30   81.72   45.70   81.25   57.60    0.47    1.30   81.25   57.10    0.47    1.30
RE   1    1    10   404   A ALA   1    1    H    1    1    H      -67.72  -13.25   35.90   35.58   33.00   16.63   24.00   18.95   49.20   18.20   25.50   17.38   47.50
RE   1    1    11   405   S SER   1    1    H    1    .    T      -82.48  -15.19  100.90   90.08   77.30   61.12   78.20   28.96   75.40   25.90   53.40   64.18   94.40
RE   1    1    12   406   Q GLN   .    .    .    .    .    T      -95.87  -61.03  151.90  142.66   79.90  109.97   78.00   32.69   87.10   49.16   94.10   93.49   74.00
RE   1    1    13   407   C CYS   .    .    .    .    .    T      -79.87  165.96   32.00   36.35   27.10   27.88   28.80    8.47   22.60   29.04   29.70    7.31   20.10
RE   1    1    14   408   T THR   2    2    H    1    .    C      -67.21  155.23   79.50   78.09   56.10   73.27   72.00    4.82   12.80   53.17   70.20   24.92   39.20
RE   1    1    15   409   A ALA   2    2    H    1    2    H      -61.46  -26.61   74.80   80.21   74.30   71.49  103.00    8.72   22.60   72.91  102.10    7.31   20.00
RE   1    1    16   410   Q GLN   2    2    H    1    2    H      -69.67  -54.28  138.50  127.16   71.20  124.25   88.10    2.91    7.80   46.24   88.60   80.92   64.10
RE   1    1    17   411   E GLU   2    2    H    1    2    H      -57.11  -42.09  104.70   89.88   52.20   89.35   66.30    0.53    1.40   31.77   52.70   58.11   51.90
RE   1    1    18   412   L LEU   2    2    H    1    2    H      -57.70  -47.91   25.70   25.00   14.00   25.00   17.70    0.00    0.00   25.00   17.60    0.00    0.00
RE   1    1    19   413   V VAL   2    2    H    1    2    H      -66.05  -30.11   88.20   90.90   60.00   90.80   79.50    0.10    0.30   90.90   78.70    0.00    0.00
RE   1    1    20   414   M MET   2    2    H    1    2    H      -69.27  -42.21  124.20  127.92   65.90  122.84   78.40    5.08   13.50  124.37   78.80    3.54    9.80
RE   1    1    21   415   T THR   2    2    H    1    2    H      -65.51  -37.49   81.70   78.15   56.10   76.18   74.90    1.96    5.20   56.58   74.70   21.56   33.90
RE   1    1    22   416   L LEU   2    2    H    1    2    H      -74.87  -41.05   26.30   23.57   13.20   23.31   16.50    0.26    0.70   23.57   16.60    0.00    0.00
RE   1    1    23   417   N ASN   2    2    H    1    2    H      -67.86  -35.26  113.60  101.30   70.40   99.87   94.00    1.43    3.80   18.39   39.80   82.91   84.80
RE   1    1    24   418   E GLU   2    2    H    1    2    H      -65.83  -40.82  113.30  100.53   58.40   99.74   74.00    0.79    2.10   38.76   64.30   61.77   55.20
RE   1    1    25   419   L LEU   2    2    H    1    2    H      -72.73  -52.39   57.40   57.47   32.20   57.33   40.60    0.14    0.40   57.33   40.30    0.14    0.40
RE   1    1    26   420   F PHE   2    2    H    1    2    H      -70.80  -21.35   27.00   28.81   14.40   27.76   16.90    1.06    3.00   28.81   17.40    0.00    0.00
RE   1    1    27   421   A ALA   2    2    H    1    2    H      -69.34  -45.28   29.90   35.82   33.20   34.83   50.20    1.00    2.60   34.83   48.80    1.00    2.70
RE   1    1    28   422   R ARG   2    2    H    1    2    H      -61.82  -45.79  136.10  134.66   56.40  134.62   66.90    0.03    0.10   47.36   60.90   87.30   54.20
RE   1    1    29   423   F PHE   2    2    H    1    2    H      -57.44  -43.57  101.50   99.54   49.90   95.45   58.20    4.09   11.60   96.26   58.30    3.28    9.60
RE   1    1    30   424   D ASP   2    2    H    1    2    H      -62.18  -31.31   60.50   54.34   38.70   42.49   41.40   11.84   31.40   25.02   50.80   29.31   32.20
RE   1    1    31   425   K LYS   2    2    H    1    2    H      -78.40  -41.15  135.60  131.47   65.50  129.94   79.60    1.53    4.10   94.19   80.80   37.29   44.30
RE   1    1    32   426   L LEU   2    2    H    1    2    H      -64.72  -26.22   81.20   86.96   48.70   86.30   61.20    0.66    1.80   86.96   61.10    0.00    0.00
RE   1    1    33   427   A ALA   2    2    H    1    2    H      -69.69  -37.04   17.80   18.88   17.50   18.70   26.90    0.18    0.50   18.70   26.20    0.18    0.50
RE   1    1    34   428   A ALA   2    2    H    1    2    H      -70.19  -38.47   77.30   79.34   73.50   60.25   86.80   19.09   49.50   61.54   86.20   17.79   48.60
RE   1    1    35   429   E GLU   2    2    H    1    2    H      -66.66  -43.34  133.10  123.38   71.60  108.84   80.80   14.54   38.80   51.79   85.90   71.59   63.90
RE   1    1    36   430   N ASN   2    2    H    1    2    H      -85.82   10.44  105.20   96.15   66.80   83.39   78.50   12.75   33.80   21.86   47.30   74.28   76.00


<font color=red>  [Part of this file has been deleted for brevity]</font>

AT   1    1    5    .    1002  . FOK   H C9       42.200  -11.309   50.489    1.00   41.39
AT   1    1    5    .    1002  . FOK   H O6       42.275  -12.455   49.593    1.00   43.23
AT   1    1    5    .    1002  . FOK   H C10      43.008  -11.601   51.811    1.00   39.11
AT   1    1    5    .    1002  . FOK   H C11      40.680  -11.078   50.616    1.00   44.36
AT   1    1    5    .    1002  . FOK   H O7       40.106  -10.945   51.688    1.00   48.77
AT   1    1    5    .    1002  . FOK   H C12      39.943  -11.046   49.301    1.00   40.67
AT   1    1    5    .    1002  . FOK   H C13      40.595  -10.085   48.292    1.00   41.47
AT   1    1    5    .    1002  . FOK   H C14      40.276  -10.620   46.930    1.00   46.69
AT   1    1    5    .    1002  . FOK   H C15      39.971  -11.751   46.590    1.00   53.22
AT   1    1    5    .    1002  . FOK   H C16      40.047   -8.685   48.426    1.00   42.42
AT   1    1    5    .    1002  . FOK   H C17      42.671   -8.737   50.253    1.00   39.67
AT   1    1    5    .    1002  . FOK   H C18      46.732  -13.026   51.827    1.00   35.74
AT   1    1    5    .    1002  . FOK   H C19      45.859  -11.483   53.586    1.00   34.48
AT   1    1    5    .    1002  . FOK   H C20      42.913  -10.426   52.807    1.00   39.44
AT   1    1    5    .    1002  . FOK   H C21      45.883   -9.553   47.821    1.00   42.15
AT   1    1    5    .    1002  . FOK   H O5       46.157  -10.520   47.166    1.00   40.91
AT   1    1    5    .    1002  . FOK   H C22      46.769   -8.315   48.006    1.00   37.08
AT   1    1    6    .    1003  . MES   H O1       45.676    7.326   49.092    1.00   77.86
AT   1    1    6    .    1003  . MES   H C2       44.367    6.816   48.900    1.00   75.17
AT   1    1    6    .    1003  . MES   H C3       44.349    5.317   48.923    1.00   74.42
AT   1    1    6    .    1003  . MES   H N4       44.832    4.804   50.196    1.00   72.45
AT   1    1    6    .    1003  . MES   H C5       46.234    5.425   50.473    1.00   73.23
AT   1    1    6    .    1003  . MES   H C6       46.176    6.914   50.355    1.00   75.06
AT   1    1    6    .    1003  . MES   H C7       44.806    3.336   50.302    1.00   73.39
AT   1    1    6    .    1003  . MES   H C8       44.672    2.791   51.713    1.00   76.85
AT   1    1    6    .    1003  . MES   H S        45.724    1.379   51.967    1.00   78.26
AT   1    1    6    .    1003  . MES   H O1S      47.062    1.828   51.737    1.00   79.39
AT   1    1    6    .    1003  . MES   H O2S      45.303    0.380   51.016    1.00   81.58
AT   1    1    6    .    1003  . MES   H O3S      45.523    0.961   53.326    1.00   80.59
AT   1    1    6    .    1004  . MES   H O1       59.246   -5.152   27.381    1.00   99.99
AT   1    1    6    .    1004  . MES   H C2       60.067   -4.021   27.127    1.00   99.99
AT   1    1    6    .    1004  . MES   H C3       60.447   -3.301   28.378    1.00   99.78
AT   1    1    6    .    1004  . MES   H N4       61.180   -4.156   29.270    1.00   96.33
AT   1    1    6    .    1004  . MES   H C5       60.358   -5.461   29.506    1.00   97.90
AT   1    1    6    .    1004  . MES   H C6       59.965   -6.072   28.203    1.00   99.68
AT   1    1    6    .    1004  . MES   H C7       61.596   -3.484   30.507    1.00   93.33
AT   1    1    6    .    1004  . MES   H C8       61.931   -2.010   30.442    1.00   90.74
AT   1    1    6    .    1004  . MES   H S        60.763   -0.978   31.301    0.50   90.72
AT   1    1    6    .    1004  . MES   H O1S      59.476   -1.170   30.680    0.50   91.60
AT   1    1    6    .    1004  . MES   H O2S      61.249    0.383   31.164    0.50   91.20
AT   1    1    6    .    1004  . MES   H O3S      60.776   -1.430   32.647    0.50   90.05
AT   1    1    7    .    1005  . POP   H P1       58.812   -7.766   57.091    1.00   57.40
AT   1    1    7    .    1005  . POP   H O1       60.254   -7.589   56.745    1.00   54.93
AT   1    1    7    .    1005  . POP   H O2       58.618   -8.839   58.095    1.00   55.36
AT   1    1    7    .    1005  . POP   H O3       57.949   -8.024   55.908    1.00   55.10
AT   1    1    7    .    1005  . POP   H O        58.295   -6.370   57.759    1.00   57.30
AT   1    1    7    .    1005  . POP   H P2       56.998   -5.955   58.661    1.00   59.66
AT   1    1    7    .    1005  . POP   H O4       57.491   -5.746   60.070    1.00   54.95
AT   1    1    7    .    1005  . POP   H O5       56.004   -7.075   58.550    1.00   56.24
AT   1    1    7    .    1005  . POP   H O6       56.427   -4.710   58.044    1.00   56.50
//
</pre>
</td></tr></table><p>
<p><h3>File: 1ii7.ccf</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
ID   1ii7
XX
DE   MOL_ID: 1; MOLECULE: MRE11 NUCLEASE;
XX
OS   MOL_ID: 1; ORGANISM_SCIENTIFIC: PYROCOCCUS FURIOSUS;
XX
EX   METHOD xray; RESO 2.20; NMOD 1; NCHN 1; NGRP 0;
XX
CN   [1]
XX
IN   ID A; NR 65; NL 6; NH 1; NE 0;
XX
SQ   SEQUENCE    65 AA;   7396 MW;  0CFB92A3 CRC32;
     MKFAHLADIH LGYEQFHKPQ REEEFAEAFK NALEIAVQEN VDFILIAGDL FHSSRPSPGT
     LKKAI
XX
RE   1    1    8    8     D ASP   .    .    .    .    .    C      360.00   53.69  128.30  117.77   83.90   84.46   82.20   33.31   88.40   34.39   69.80   83.38   91.50
RE   1    1    9    9     I ILE   .    .    .    .    .    C      -72.50  -37.00   42.90   42.38   24.20   41.07   29.80    1.31    3.50   41.07   29.50    1.31    3.60
RE   1    1    10   10    H HIS   .    .    .    .    .    T       50.63   45.81   76.20   75.98   41.50   75.67   51.40    0.31    0.90   48.67   50.10   27.31   31.90
RE   1    1    11   11    L LEU   .    .    .    .    .    T      -57.60  139.21   78.50   79.58   44.50   55.49   39.30   24.09   64.20   56.43   39.70   23.15   63.70
RE   1    1    12   12    G GLY   .    .    .    .    .    T       91.01   -4.77   47.80   51.11   63.80   23.22   71.80   27.90   58.40   27.09   72.10   24.02   56.50
RE   1    1    13   13    Y TYR   .    .    .    .    .    T      -74.99  115.31  107.80  100.62   47.30  100.09   56.40    0.52    1.50   66.13   48.50   34.48   45.20
RE   1    1    14   14    E GLU   1    1    H    5    .    C      -96.17   74.41   82.00   71.84   41.70   61.45   45.60   10.39   27.70   17.84   29.60   54.00   48.20
RE   1    1    15   15    Q GLN   1    1    H    5    .    G      -56.45  130.81   34.70   33.39   18.70   33.39   23.70    0.00    0.00    0.31    0.60   33.08   26.20
RE   1    1    16   16    F PHE   1    1    H    5    .    G       57.56   29.90  146.10  150.86   75.60  136.18   83.00   14.68   41.50  136.40   82.50   14.46   42.20
RE   1    1    17   17    H HIS   1    1    H    5    .    G       55.61   29.81  160.00  158.76   86.80  143.03   97.20   15.72   43.90   93.02   95.70   65.74   76.70
RE   1    1    18   18    K LYS   2    A    E    .    .    C     -111.48  108.83  115.50  109.57   54.60  109.57   67.10    0.00    0.00   67.34   57.80   42.23   50.10
RE   1    1    19   19    P PRO   2    A    E    .    1    H      -54.76  -17.53   74.50   83.79   61.60   82.99   69.20    0.80    4.90   82.99   68.60    0.80    5.30
RE   1    1    20   20    Q GLN   2    A    E    .    1    H      -73.35  -35.75  120.10  112.90   63.30  111.79   79.30    1.11    3.00   28.62   54.80   84.28   66.70
RE   1    1    21   21    R ARG   2    A    E    .    1    H      -68.53  -38.89   82.60   84.22   35.30   84.22   41.80    0.00    0.00   22.91   29.40   61.31   38.10
RE   1    1    22   22    E GLU   2    A    E    .    1    H      -56.60  -42.19   71.00   59.66   34.60   59.41   44.10    0.25    0.70   17.46   29.00   42.20   37.70
RE   1    1    23   23    E GLU   2    A    E    .    1    H      -66.34  -36.55  123.60  108.67   63.10  104.55   77.60    4.12   11.00   54.20   89.90   54.47   48.60
RE   1    1    24   24    E GLU   2    A    E    .    1    H      -71.24  -35.34   80.40   71.99   41.80   67.32   50.00    4.67   12.40   41.76   69.30   30.23   27.00
RE   1    1    25   25    F PHE   2    A    E    .    1    H      -63.59  -40.62   41.80   41.19   20.60   39.81   24.30    1.38    3.90   39.81   24.10    1.38    4.00
RE   1    1    26   26    A ALA   2    A    E    .    1    H      -66.25  -43.69   46.10   46.40   43.00   41.37   59.60    5.03   13.10   42.55   59.60    3.85   10.50
RE   1    1    27   27    E GLU   2    A    E    .    1    H      -60.95  -39.89   91.00   84.33   49.00   83.96   62.30    0.36    1.00   41.68   69.10   42.65   38.10
RE   1    1    28   28    A ALA   2    A    E    .    1    H      -58.03  -47.16   55.00   58.08   53.80   52.98   76.30    5.10   13.20   54.38   76.20    3.70   10.10
RE   1    1    29   29    F PHE   2    A    E    .    1    H      -61.12  -42.56   38.50   33.87   17.00   33.36   20.30    0.51    1.40   33.36   20.20    0.51    1.50
RE   1    1    30   30    K LYS   2    A    E    .    1    H      -63.44  -42.40  104.70  105.98   52.80  105.45   64.60    0.53    1.40   58.41   50.10   47.56   56.50
RE   1    1    31   31    N ASN   2    A    E    .    1    H      -60.26  -47.05   77.00   73.22   50.90   72.53   68.30    0.70    1.80   23.06   49.90   50.16   51.30
RE   1    1    32   32    A ALA   2    A    E    .    1    H      -58.51  -48.57   56.20   56.15   52.00   49.87   71.80    6.28   16.30   51.18   71.70    4.97   13.60
RE   1    1    33   33    L LEU   2    A    E    .    1    H      -61.69  -37.59   59.70   59.93   33.60   59.56   42.20    0.38    1.00   59.93   42.10    0.00    0.00
RE   1    1    34   34    E GLU   2    A    E    .    1    H      -68.18  -35.02   87.40   76.92   44.70   76.19   56.50    0.74    2.00   33.77   56.00   43.15   38.50
RE   1    1    35   35    I ILE   2    A    E    .    1    H      -69.89  -35.85   89.80   96.84   55.30   96.84   70.20    0.00    0.00   96.84   69.60    0.00    0.00
RE   1    1    36   36    A ALA   2    A    E    .    1    H      -61.30  -45.01   19.20   19.50   18.10   19.22   27.70    0.27    0.70   19.22   26.90    0.27    0.70
RE   1    1    37   37    V VAL   2    A    E    .    1    H      -64.92  -44.30  109.50  111.45   73.60  102.58   89.80    8.87   23.90  102.58   88.80    8.87   24.70
RE   1    1    38   38    Q GLN   2    A    E    .    1    H      -59.93  -32.77  145.80  140.48   78.70  113.02   80.20   27.46   73.20   49.34   94.50   91.14   72.20
RE   1    1    39   39    E GLU   2    A    E    .    1    H      -83.49    5.97  133.40  118.49   68.80   99.06   73.50   19.43   51.80   31.08   51.60   87.41   78.10
RE   1    1    40   40    N ASN   .    .    .    .    .    C       54.97   43.05  136.50  126.77   88.10  116.79  109.90    9.98   26.50   31.92   69.00   94.86   97.10
RE   1    1    41   41    V VAL   .    .    .    .    .    C      -70.50  156.23   65.90   69.23   45.70   50.82   44.50   18.40   49.50   50.86   44.00   18.36   51.10


<font color=red>  [Part of this file has been deleted for brevity]</font>

AT   1    1    .    50   50    L LEU   P CD2      12.425   39.035   22.798    1.00   23.77
AT   1    1    1    .    402   . PO4   H P        34.178   32.996   46.387    1.00   60.84
AT   1    1    1    .    402   . PO4   H O1       35.146   33.243   45.291    1.00   57.95
AT   1    1    1    .    402   . PO4   H O2       34.912   32.751   47.670    1.00   59.15
AT   1    1    1    .    402   . PO4   H O3       33.291   34.184   46.538    1.00   58.92
AT   1    1    1    .    402   . PO4   H O4       33.352   31.796   46.060    1.00   61.86
AT   1    1    2    .    403   . MN    H MN        8.130   27.788   21.899    1.00   36.09
AT   1    1    2    .    404   . MN    H MN        5.801   27.935   24.271    1.00   39.57
AT   1    1    3    .    405   . MN    H MN       36.023   34.916   44.253    1.00   39.52
AT   1    1    3    .    406   . MN    H MN       33.658   36.365   46.296    1.00   33.69
AT   1    1    5    .    501   . SO4   H S        17.175   28.112   32.476    1.00  100.80
AT   1    1    5    .    501   . SO4   H O1       18.136   28.230   31.357    1.00  100.18
AT   1    1    5    .    501   . SO4   H O2       17.097   26.692   32.887    1.00  100.80
AT   1    1    5    .    501   . SO4   H O3       17.633   28.926   33.626    1.00  100.14
AT   1    1    5    .    501   . SO4   H O4       15.834   28.575   32.045    1.00  100.56
AT   1    1    5    .    502   . SO4   H S         0.566   29.512   36.007    1.00   86.73
AT   1    1    5    .    502   . SO4   H O1        1.690   28.556   35.971    1.00   87.27
AT   1    1    5    .    502   . SO4   H O2       -0.620   28.803   36.523    1.00   87.87
AT   1    1    5    .    502   . SO4   H O3        0.896   30.642   36.905    1.00   86.58
AT   1    1    5    .    502   . SO4   H O4        0.287   30.037   34.658    1.00   86.51
AT   1    1    5    .    503   . SO4   H S       -13.586   39.644   36.031    1.00  100.28
AT   1    1    5    .    503   . SO4   H O1      -12.340   39.512   35.250    1.00  100.72
AT   1    1    5    .    503   . SO4   H O2      -14.638   38.811   35.421    1.00  100.46
AT   1    1    5    .    503   . SO4   H O3      -13.347   39.201   37.420    1.00   99.66
AT   1    1    5    .    503   . SO4   H O4      -14.020   41.056   36.015    1.00   99.97
AT   1    1    6    .    401   . 101   H P         7.599   25.305   23.994    1.00   56.33
AT   1    1    6    .    401   . 101   H O1P       8.249   24.467   25.030    1.00   56.70
AT   1    1    6    .    401   . 101   H O2P       6.700   26.285   24.649    1.00   54.49
AT   1    1    6    .    401   . 101   H O3P       8.637   26.026   23.216    1.00   53.97
AT   1    1    6    .    401   . 101   H O5*       7.095   23.970   23.128    1.00   59.20
AT   1    1    6    .    401   . 101   H C5*       7.073   23.961   21.762    1.00   66.74
AT   1    1    6    .    401   . 101   H C4*       6.041   23.013   21.296    1.00   71.22
AT   1    1    6    .    401   . 101   H O4*       6.029   21.855   22.189    1.00   73.78
AT   1    1    6    .    401   . 101   H C3*       4.736   23.676   21.350    1.00   73.80
AT   1    1    6    .    401   . 101   H O3*       4.355   23.874   19.995    1.00   76.51
AT   1    1    6    .    401   . 101   H C2*       3.864   22.749   22.165    1.00   74.04
AT   1    1    6    .    401   . 101   H C1*       4.682   21.474   22.506    1.00   74.70
AT   1    1    6    .    401   . 101   H N9        4.578   21.123   23.969    1.00   76.71
AT   1    1    6    .    401   . 101   H C8        3.630   21.533   24.876    1.00   76.87
AT   1    1    6    .    401   . 101   H N7        3.758   21.069   26.081    1.00   77.50
AT   1    1    6    .    401   . 101   H C5        4.896   20.300   25.989    1.00   77.78
AT   1    1    6    .    401   . 101   H C6        5.570   19.479   26.941    1.00   78.16
AT   1    1    6    .    401   . 101   H N6        5.155   19.409   28.200    1.00   78.77
AT   1    1    6    .    401   . 101   H N1        6.682   18.805   26.554    1.00   78.32
AT   1    1    6    .    401   . 101   H C2        7.090   18.888   25.277    1.00   78.14
AT   1    1    6    .    401   . 101   H N3        6.541   19.611   24.271    1.00   78.05
AT   1    1    6    .    401   . 101   H C4        5.403   20.288   24.700    1.00   78.10
AT   1    .    .    .    407   . HOH   W O         5.997   27.242   22.189    1.00   38.84
AT   1    .    .    .    408   . HOH   W O        35.697   35.756   46.350    1.00   41.39
AT   1    .    .    .    600   . HOH   W O        20.825   31.690   27.031    1.00   20.90
//
</pre>
</td></tr></table><p>
<p><h3>File: 2hhb.ccf</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
ID   2hhb
XX
DE   HEMOGLOBIN (DEOXY)
XX
OS   HUMAN (HOMO SAPIENS)
XX
EX   METHOD xray; RESO 1.74; NMOD 1; NCHN 4; NGRP 0;
XX
CN   [1]
XX
IN   ID A; NR 141; NL 1; NH 7; NE 0;
XX
SQ   SEQUENCE   141 AA;  15127 MW;  5EC7DB1E CRC32;
     VLSPADKTNV KAAWGKVGAH AGEYGAEALE RMFLSFPTTK TYFPHFDLSH GSAQVKGHGK
     KVADALTNAV AHVDDMPNAL SALSDLHAHK LRVDPVNFKL LSHCLLVTLA AHLPAEFTPA
     VHASLDKFLA SVSTVLTSKY R
XX
CN   [2]
XX
IN   ID B; NR 146; NL 1; NH 8; NE 0;
XX
SQ   SEQUENCE   146 AA;  15868 MW;  EC9744C9 CRC32;
     VHLTPEEKSA VTALWGKVNV DEVGGEALGR LLVVYPWTQR FFESFGDLST PDAVMGNPKV
     KAHGKKVLGA FSDGLAHLDN LKGTFATLSE LHCDKLHVDP ENFRLLGNVL VCVLAHHFGK
     EFTPPVQAAY QKVVAGVANA LAHKYH
XX
CN   [3]
XX
IN   ID C; NR 141; NL 1; NH 7; NE 0;
XX
SQ   SEQUENCE   141 AA;  15127 MW;  5EC7DB1E CRC32;
     VLSPADKTNV KAAWGKVGAH AGEYGAEALE RMFLSFPTTK TYFPHFDLSH GSAQVKGHGK
     KVADALTNAV AHVDDMPNAL SALSDLHAHK LRVDPVNFKL LSHCLLVTLA AHLPAEFTPA
     VHASLDKFLA SVSTVLTSKY R
XX
CN   [4]
XX
IN   ID D; NR 146; NL 2; NH 8; NE 0;
XX
SQ   SEQUENCE   146 AA;  15868 MW;  EC9744C9 CRC32;
     VHLTPEEKSA VTALWGKVNV DEVGGEALGR LLVVYPWTQR FFESFGDLST PDAVMGNPKV
     KAHGKKVLGA FSDGLAHLDN LKGTFATLSE LHCDKLHVDP ENFRLLGNVL VCVLAHHFGK
     EFTPPVQAAY QKVVAGVANA LAHKYH
XX
RE   1    1    1    1     V VAL   .    .    .    .    .    C      360.00  130.57  137.00  142.86   94.30  113.45   99.30   29.40   79.10  113.56   98.30   29.29   81.40
RE   1    1    2    2     L LEU   .    .    .    .    .    C      -80.50  121.94   16.80   18.20   10.20    7.73    5.50   10.48   27.90    7.73    5.40   10.48   28.80
RE   1    1    3    3     S SER   1    AA   H    1    .    C      -71.32  164.78   44.40   44.59   38.30   44.42   56.90    0.17    0.40   39.26   80.90    5.34    7.90
RE   1    1    4    4     P PRO   1    AA   H    1    1    H      -62.31  -35.05   99.60  105.25   77.30  103.02   85.90    2.23   13.70  103.65   85.70    1.60   10.50
RE   1    1    5    5     A ALA   1    AA   H    1    1    H      -65.00  -41.52   59.80   65.59   60.80   64.35   92.70    1.24    3.20   64.35   90.20    1.24    3.40
RE   1    1    6    6     D ASP   1    AA   H    1    1    H      -56.29  -42.84   13.60   13.83    9.90   13.80   13.40    0.03    0.10    0.44    0.90   13.39   14.70


<font color=red>  [Part of this file has been deleted for brevity]</font>

AT   1    .    .    .    174   . HOH   W O        -4.764   -6.228    5.515    8.00   40.89
AT   1    .    .    .    175   . HOH   W O        23.809   19.925    1.758    8.00   39.37
AT   1    .    .    .    176   . HOH   W O        -7.871   -9.078    2.406    8.00   43.37
AT   1    .    .    .    177   . HOH   W O         4.693   12.083    7.558    8.00   40.24
AT   1    .    .    .    178   . HOH   W O         8.775  -23.438   16.055    8.00   42.33
AT   1    .    .    .    179   . HOH   W O        -7.480  -10.898   17.998    8.00   38.06
AT   1    .    .    .    180   . HOH   W O        -4.731   16.453    2.295    8.00   36.37
AT   1    .    .    .    181   . HOH   W O        -1.055   11.866   -0.448    8.00   43.19
AT   1    .    .    .    182   . HOH   W O       -27.610  -10.991    5.353    8.00   43.46
AT   1    .    .    .    183   . HOH   W O        26.015   11.766    5.159    8.00   40.95
AT   1    .    .    .    184   . HOH   W O       -18.517   -8.355   15.267    8.00   35.55
AT   1    .    .    .    185   . HOH   W O       -14.034    2.806  -30.367    8.00   41.77
AT   1    .    .    .    186   . HOH   W O       -32.905   -9.033    0.480    8.00   43.68
AT   1    .    .    .    187   . HOH   W O       -28.749  -13.315    1.938    8.00   45.36
AT   1    .    .    .    188   . HOH   W O         0.516   -8.074  -26.354    8.00   41.53
AT   1    .    .    .    189   . HOH   W O       -20.080   -9.873  -22.862    8.00   36.25
AT   1    .    .    .    190   . HOH   W O       -13.442    9.778  -13.572    8.00   39.70
AT   1    .    .    .    191   . HOH   W O       -24.804   -2.608  -15.488    8.00   37.79
AT   1    .    .    .    192   . HOH   W O         6.547    9.706   16.296    8.00   41.86
AT   1    .    .    .    193   . HOH   W O         0.029   22.606   14.164    8.00   43.02
AT   1    .    .    .    194   . HOH   W O       -11.367    0.306   28.463    8.00   44.30
AT   1    .    .    .    195   . HOH   W O       -19.950  -10.635   14.301    8.00   40.17
AT   1    .    .    .    196   . HOH   W O        -7.047   -6.324   20.098    8.00   36.98
AT   1    .    .    .    197   . HOH   W O       -23.876    1.108   14.102    8.00   33.31
AT   1    .    .    .    198   . HOH   W O       -34.199    8.033   11.037    8.00   40.72
AT   1    .    .    .    199   . HOH   W O       -14.173   13.393   -8.778    8.00   43.21
AT   1    .    .    .    200   . HOH   W O        11.388  -11.044   24.763    8.00   39.34
AT   1    .    .    .    201   . HOH   W O         3.735   -3.643    2.734    8.00   42.17
AT   1    .    .    .    202   . HOH   W O         3.149   -0.692    2.083    8.00   41.40
AT   1    .    .    .    203   . HOH   W O         4.511  -25.886   13.006    8.00   39.83
AT   1    .    .    .    204   . HOH   W O         8.712  -21.655    3.577    8.00   43.08
AT   1    .    .    .    205   . HOH   W O        22.926   -4.304   24.079    8.00   38.10
AT   1    .    .    .    206   . HOH   W O        11.435    9.654   20.618    8.00   40.23
AT   1    .    .    .    207   . HOH   W O        18.099    5.542   27.744    8.00   39.03
AT   1    .    .    .    208   . HOH   W O        12.174    9.951    9.804    8.00   44.34
AT   1    .    .    .    209   . HOH   W O        24.745   -2.501   15.270    8.00   39.78
AT   1    .    .    .    210   . HOH   W O        24.231    0.100   14.764    8.00   42.94
AT   1    .    .    .    211   . HOH   W O        23.324  -18.136   10.981    8.00   53.60
AT   1    .    .    .    212   . HOH   W O        25.576  -22.211    6.309    8.00   45.18
AT   1    .    .    .    213   . HOH   W O        14.639   24.823   -4.300    8.00   41.35
AT   1    .    .    .    214   . HOH   W O        14.903    5.393  -23.047    8.00   37.45
AT   1    .    .    .    215   . HOH   W O        16.650   -5.137  -16.717    8.00   39.12
AT   1    .    .    .    216   . HOH   W O         7.424   -6.700  -20.085    8.00   38.62
AT   1    .    .    .    217   . HOH   W O        -1.263   -2.837  -21.251    8.00   45.10
AT   1    .    .    .    218   . HOH   W O        23.120   -3.118  -12.992    8.00   37.05
AT   1    .    .    .    219   . HOH   W O        23.664    0.968  -14.389    8.00   36.25
AT   1    .    .    .    220   . HOH   W O        25.698    7.981  -15.362    8.00   35.85
AT   1    .    .    .    221   . HOH   W O        30.009   16.347   -6.794    8.00   37.62
AT   1    .    .    .    222   . HOH   W O        27.728   16.677   -1.376    8.00   42.54
AT   1    .    .    .    223   . HOH   W O         8.142   18.836    1.041    8.00   39.90
//
</pre>
</td></tr></table><p>



<!-- DATA FILES         
     Any data files used (e.g. translation table file, substitution matrix 
     etc.  This includes example data file formats if they are not obvious.
     For a standard description of what data files are and how embossdata can
     be used to inspect and retrieve them, use:
     #include file="inc/localfiles.ihtml" 
-->

<a name="5.0"></a>
<br><br><br><H2> 5.0   DATA FILES                </H2>
PDBPLUS does not use a data file.







<!-- USAGE
     Example usage, as run from the command-line.
     Many examples illustrating different behaviours is good.
-->
<a name="6.0"></a>
<br><br><br><H2> 6.0   USAGE                     </H2>

<pre>
   Standard (Mandatory) qualifiers (* if not always prompted):
  [-ccfinpath]         dirlist    This option specifies the location of CCF
                                  files (clean coordinate files) (input). A
                                  'clean cordinate file' contains protein
                                  coordinate and derived data for a single PDB
                                  file ('protein clean coordinate file') or a
                                  single domain from SCOP or CATH ('domain
                                  clean coordinate file'), in CCF format
                                  (EMBL-like). The files, generated by using
                                  PDBPARSE (PDB files) or DOMAINER (domains),
                                  contain 'cleaned-up' data that is
                                  self-consistent and error-corrected. Records
                                  for residue solvent accessibility and
                                  secondary structure are added to the file by
                                  using PDBPLUS.
  [-pdbindir]          directory  This option specifies the location of PDB
                                  files (input). A PDB file contains protein
                                  coordinate and other data. A detailed
                                  explanation of the PDB file format is
                                  available on the PDB web site
                                  http://www.rcsb.org/pdb/info.html
   -pdbprefix          string     This option specifies the file prefix of PDB
                                  files (if any)
   -mode               menu       This option specifies the mode of operation.
*  -thresholdsize      integer    This option specifies the threshold size for
                                  SSEs (secondary structure elements)
  [-ccfoutdir]         outdir     This option specifies the location of CCF
                                  files (clean coordinate files) (output). A
                                  'clean cordinate file' contains protein
                                  coordinate and derived data for a single PDB
                                  file ('protein clean coordinate file') or a
                                  single domain from SCOP or CATH ('domain
                                  clean coordinate file'), in CCF format
                                  (EMBL-like). The files, generated by using
                                  PDBPARSE (PDB files) or DOMAINER (domains),
                                  contain 'cleaned-up' data that is
                                  self-consistent and error-corrected. Records
                                  for residue solvent accessibility and
                                  secondary structure are added to the file by
                                  using PDBPLUS.
   -logfile            outfile    This option specifies the name of the log
                                  file for PDBPLUS
*  -slogfile           outfile    This option specifies the name of the log
                                  file for STRIDE
*  -nlogfile           outfile    This option specifies the name of the log
                                  file for NACCESS

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-logfile" associated qualifiers
   -odirectory         string     Output directory

   "-slogfile" associated qualifiers
   -odirectory         string     Output directory

   "-nlogfile" associated qualifiers
   -odirectory         string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report deaths


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-ccfinpath]<br>(Parameter 1)</td>
<td>This option specifies the location of CCF files (clean coordinate files) (input). A 'clean cordinate file' contains protein coordinate and derived data for a single PDB file ('protein clean coordinate file') or a single domain from SCOP or CATH ('domain clean coordinate file'), in CCF format (EMBL-like). The files, generated by using PDBPARSE (PDB files) or DOMAINER (domains), contain 'cleaned-up' data that is self-consistent and error-corrected. Records for residue solvent accessibility and secondary structure are added to the file by using PDBPLUS.</td>
<td>Directory with files</td>
<td>./</td>
</tr>

<tr>
<td>[-pdbindir]<br>(Parameter 2)</td>
<td>This option specifies the location of PDB files (input). A PDB file contains protein coordinate and other data. A detailed explanation of the PDB file format is available on the PDB web site http://www.rcsb.org/pdb/info.html</td>
<td>Directory</td>
<td>./</td>
</tr>

<tr>
<td>-pdbprefix</td>
<td>This option specifies the file prefix of PDB files (if any)</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-mode</td>
<td>This option specifies the mode of operation.</td>
<td><table><tr><td>1</td> <td><i>(Use STRIDE only)</i></td></tr><tr><td>2</td> <td><i>(Use NACCESS only)</i></td></tr><tr><td>3</td> <td><i>(Use STRIDE and NACCESS)</i></td></tr></table></td>
<td>1</td>
</tr>

<tr>
<td>-thresholdsize</td>
<td>This option specifies the threshold size for SSEs (secondary structure elements)</td>
<td>Any integer value</td>
<td>4</td>
</tr>

<tr>
<td>[-ccfoutdir]<br>(Parameter 3)</td>
<td>This option specifies the location of CCF files (clean coordinate files) (output). A 'clean cordinate file' contains protein coordinate and derived data for a single PDB file ('protein clean coordinate file') or a single domain from SCOP or CATH ('domain clean coordinate file'), in CCF format (EMBL-like). The files, generated by using PDBPARSE (PDB files) or DOMAINER (domains), contain 'cleaned-up' data that is self-consistent and error-corrected. Records for residue solvent accessibility and secondary structure are added to the file by using PDBPLUS.</td>
<td>Output directory</td>
<td>./</td>
</tr>

<tr>
<td>-logfile</td>
<td>This option specifies the name of the log file for PDBPLUS</td>
<td>Output file</td>
<td>pdbplus.log</td>
</tr>

<tr>
<td>-slogfile</td>
<td>This option specifies the name of the log file for STRIDE</td>
<td>Output file</td>
<td>stride.log</td>
</tr>

<tr>
<td>-nlogfile</td>
<td>This option specifies the name of the log file for NACCESS</td>
<td>Output file</td>
<td>naccess.log</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H3> 6.2 EXAMPLE SESSION </H3>
An example of interactive use of PDBPLUS is shown below.

<b>Here is a sample session with pdbplus</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>pdbplus </b>
Add accessibility & secondary structure to a CCF file.
Location of CCF files (clean coordinate files) (input) [./]: <b>../pdbparse-keep/</b>
Location of PDB files (input) [./]: <b></b>
File prefix of PDB files (if any): <b></b>
Mode of operation
         1 : Use STRIDE only
         2 : Use NACCESS only
         3 : Use STRIDE and NACCESS
Select mode [1]: <b>3</b>
Threshold size for SSEs (secondary structure elements) [4]: <b>4</b>
Location of CCF files (clean coordinate files) (output) [./]: <b></b>
Name of log file for PDBPLUS [pdbplus.log]: <b></b>
Name of log file for STRIDE [stride.log]: <b></b>
Name of log file for NACCESS [naccess.log]: <b></b>

Processing /ebi/services/idata/pmr/hgmp/test/qa/pdbparse-keep/2hhb.ccf
stride /ebi/services/idata/pmr/hgmp/test/data/structure/2hhb.ent -f./pdbplus-1121267205.25929 >> stride.log 2>&1
naccess /ebi/services/idata/pmr/hgmp/test/data/structure/2hhb.ent  >> naccess.log 2>&1
Processing /ebi/services/idata/pmr/hgmp/test/qa/pdbparse-keep/1cs4.ccf
stride /ebi/services/idata/pmr/hgmp/test/data/structure/1cs4.ent -f./pdbplus-1121267205.25929 >> stride.log 2>&1
naccess /ebi/services/idata/pmr/hgmp/test/data/structure/1cs4.ent  >> naccess.log 2>&1
Processing /ebi/services/idata/pmr/hgmp/test/qa/pdbparse-keep/1ii7.ccf
stride /ebi/services/idata/pmr/hgmp/test/data/structure/1ii7.ent -f./pdbplus-1121267205.25929 >> stride.log 2>&1
naccess /ebi/services/idata/pmr/hgmp/test/data/structure/1ii7.ent  >> naccess.log 2>&1
STRIDE  failures: 0
NACCESS failures: 0

</pre></td></tr></table><p>
<p>
<a href="#output.1">Go to the output files for this example</a><p><p>

<!--
All CCF files were read from testdata/pdbplus/ccfin and CCF files (annotated with values for secondary structure and solvent accessibility) were written to testdata/pdbplus/ccfout.  The STRIDE and NACCESS programs ran on PDB files within testdata/. Three log files (pdbplus.log, naccess.log and stride.log) were written. The threshold size for secondary structure elements was 4 residues.  File extension of all input and output files would be set in the ACD file. -->



<!-- KNOWN BUGS & WARNINGS
     Bugs that have not yet been fixed, easily missued features, problems
     and caveats etc. Potentially stupid things the program will let you do.
-->
<a name="7.0"></a>
<br><br><br><H2> 7.0   KNOWN BUGS & WARNINGS     </H2>
NACCESS and STRIDE do not always work as advertised owing, e.g. to arbitrary array limits in their code.  If the derived data appears to be missing from the PDDBPARSE output this is probably why.





<!-- NOTES
     Important general remarks, including:
       Restrictions.
       Interesting behaviour.
       Useful things you can do with this program.
       Future plans.
       etc.
-->
<a name="8.0"></a>
<br><br><br><H2> 8.0   NOTES                     </H2>
PDBPLUS requires a working version of NACCESS and STRIDE.
When running PDBPLUS at the RFCGR it is essential that the commands 'use naccess' or 
'use stride' (which run the scripts /packages/menu/USE/naccess and /packages/menu/USE/stride respectively) 
are given before it is run.  
<br>
<br>

<H3> 8.1   GLOSSARY OF FILE TYPES    </H3>
<table BORDER CELLSPACING=5 CELLPADDING=5 BGCOLOR="#f5f5ff" >
<tr>
<td><b>FILE TYPE</b></td>
<td><b>FORMAT</b></td>
<td><b>DESCRIPTION</b></td>
<td><b>CREATED BY <b></td>
<td><b>SEE ALSO</b></td>
</tr>
<tr>
<td><b> Clean coordinate file (for protein) </b></td>
<td> CCF format (EMBL-like).</td>
<td> Protein coordinate and derived data for a single PDB file.  The data are 'cleaned-up': self-consistent and error-corrected.  </td>
<td> <a href="pdbparse.html">PDBPARSE</a> </td>
<td> Records for residue solvent accessibility and secondary structure are added to the file by using PDBPLUS. </td>
</tr>
<tr>
<td><b> Clean coordinate file (for domain)</b></td>
<td> CCF format (EMBL-like).</td>
<td> Protein coordinate and derived data for a single domain from SCOP or CATH. The data are 'cleaned-up': self-consistent and error-corrected. </td>
<td> <a href="domainer.html">DOMAINER</a></td>
<td> Records for residue solvent accessibility and secondary structure are added to the file by using PDBPLUS.</td>
</tr>
<tr>
<td><b>PDB file </b></td>
<td> PDB format. </td>
<td> Protein coordinate data in PDB format. </td>
<td> N.A. </td>
<td> N.A. </td>
</tr>
<tr>
<td><b>Domain PDB file </b></td>
<td> PDB format. </td>
<td> Protein coordinate data for a single domain from SCOP or CATH. </td>
<td> <a href="domainer.html">DOMAINER</a> </td>
<td> N.A. </td>
</tr>
</table>






None



<a name="9.0"></a>
<br><br><br><H2> 9.0   DESCRIPTION               </H2>
PDBPLUS is a wrapper to the STRIDE [1] and NACCESS [2] programs that adds derived data (i - iv below) to a CCF file.  The values of the derived data for a given residue will of course depend on whether they were calculated from the coordinates for the entire structure or just the domain, i.e. whether PDBPLUS is provided with protein or domain CCF files.  Thus PDBPLUS provides useful flexibility in generating the derived data.
<br>
<UL>
<LI>  Absolute (Abs.) and relative (Rel.) accessible surface area of residues according to NACCESS.  Abs. is the summed accessible surface area of the atoms, whereas Rel. is expressed as a percentage relative to the accessibility of the atoms in an extended ALA-x-ALA tri-peptide conformation.  The NACCESS authors treat alpha carbons as side-chain atoms so that glycine can have a value for side-chain accessible surface area (see 'Database format' below).  They are therefore not included in the main-chain.  
<LI>  Phi and Psi angle and solvent accessible surface area of residues as calculated by using STRIDE.  
<LI>  Secondary structure assignment according to STRIDE, one of 'H' (alpha helix), 'G' (3-10 helix), 'I' (Pi-helix), 'E' (extended conformation), 'B' or 'b' (isolated bridge), 'T' (turn) or 'C' (coil, i.e. none of the above).  
<LI>  The number of helical or beta-strand secondary structure elements (SSEs) in the chain or domain.  An 'element' is defined as a run of a user-defined number (typically 4) of residues in the 'H', 'G' or 'I' conformation (helices) or the 'E' conformation (beta-strands).  
</UL>



<!-- ALGORITHM
     A technical description of algorithmic aspects, describing exactly how
     the key aspects of the application work.
-->
<a name="10.0"></a>
<br><br><br><H2> 10.0  ALGORITHM                 </H2>
Values in the CCF file for the number of helices (NH) or beta-strands (NE) in a chain and columns 12-13 and 22-34 of the coordinate line record (CO) are given null values ('.' or 0) by PDBPARSE (see Section 4.0).  PDBPLUS is used to assign values to these these records.



<!-- RELATED APPLICATIONS
     Other applications that either generate the input, use the output or 
     are in some other way related to the application are described here. 
-->

<a name="11.0"></a>
<br><br><br><br><H2> 11.0  RELATED APPLICATIONS      </H2>

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="../../apps/aaindexextract.html">aaindexextract</a></td><td>Extract data from AAINDEX</td></tr>
<tr><td><a href="../domalign/allversusall.html">allversusall</a></td><td>Sequence similarity data from all-versus-all comparison</td></tr>
<tr><td><a href="../domainatrix/cathparse.html">cathparse</a></td><td>Generates DCF file from raw CATH files</td></tr>
<tr><td><a href="../../apps/cutgextract.html">cutgextract</a></td><td>Extract data from CUTG</td></tr>
<tr><td><a href="domainer.html">domainer</a></td><td>Generates domain CCF files from protein CCF files</td></tr>
<tr><td><a href="../domainatrix/domainnr.html">domainnr</a></td><td>Removes redundant domains from a DCF file</td></tr>
<tr><td><a href="../domainatrix/domainseqs.html">domainseqs</a></td><td>Adds sequence records to a DCF file</td></tr>
<tr><td><a href="../domainatrix/domainsse.html">domainsse</a></td><td>Add secondary structure records to a DCF file</td></tr>
<tr><td><a href="hetparse.html">hetparse</a></td><td>Converts heterogen group dictionary to EMBL-like format</td></tr>
<tr><td><a href="pdbparse.html">pdbparse</a></td><td>Parses PDB files and writes protein CCF files</td></tr>
<tr><td><a href="pdbtosp.html">pdbtosp</a></td><td>Convert swissprot:PDB codes file to EMBL-like format</td></tr>
<tr><td><a href="../../apps/printsextract.html">printsextract</a></td><td>Extract data from PRINTS</td></tr>
<tr><td><a href="../../apps/prosextract.html">prosextract</a></td><td>Build the PROSITE motif database for use by patmatmotifs</td></tr>
<tr><td><a href="../../apps/rebaseextract.html">rebaseextract</a></td><td>Extract data from REBASE</td></tr>
<tr><td><a href="../domainatrix/scopparse.html">scopparse</a></td><td>Generate DCF file from raw SCOP files</td></tr>
<tr><td><a href="../domsearch/seqnr.html">seqnr</a></td><td>Removes redundancy from DHF files</td></tr>
<tr><td><a href="sites.html">sites</a></td><td>Generate residue-ligand CON files from CCF files</td></tr>
<tr><td><a href="../domainatrix/ssematch.html">ssematch</a></td><td>Search a DCF file for secondary structure matches</td></tr>
<tr><td><a href="../../apps/tfextract.html">tfextract</a></td><td>Extract data from TRANSFAC</td></tr>
</table>



<!-- DIAGNOSTIC ERROR MESSAGES
     Description of error messages or log file, if one is written.
-->
<a name="12.0"></a>
<br><br><br><H2> 12.0  DIAGNOSTIC ERROR MESSAGES </H2>
The following messages (in "quotes") may appear in the PDBPARSE log file (an explanation is given in parenethesis). 

<UL>
<LI>"clean coordinate file not found my.file" (a clean coordinate input file was not found for input)
<LI>"ERROR Clean coordinate file read error: my.file" (error in reading clean coordinate file)
<LI>"no stride output for: my.file" (No STRIDE output was generated for my.file)
<LI>"stride output for: my.file" (STRIDE output *was* generated for my.file)
<LI>"no naccess output for:" (No NACCESS output was generated for my.file)
<LI>"naccess output for:" (NACCESS output *was* generated for my.file)
<LI>"PDB file not found: my.file" (a pdb file was not found for input)
<LI>"Could not write results file: my.file" (output file could not be written)
<LI>"Could not convert chain id A to chain number in pdb file my.file" (chain identifier was not found in the clean coordinate file)
<LI>"STRIDE  failures: %12\nNACCESS failures: 123" (STRIDE failed for 12 pdb files and NACCESS failed for 123 pdb files.  The clean coordinate files will lack the appropriate data (see Section 10.0).
</UL>

<!-- AUTHORS
     The main author first then all others. 
-->
<a name="13.0"></a>
<br><br><br><H2> 13.0  AUTHORS                   </H2>
Jon Ison <a href="mailto:jison@rfcgr.mrc.ac.uk">(jison@rfcgr.mrc.ac.uk)</a>
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK



<!-- REFERENCES
     Quote the paper where the application was first published, described, used etc. 
-->
<a name="14.0"></a>
<br><br><br><H2> 14.0  REFERENCES                </H2>
Please cite the authors and EMBOSS.
Please cite the authors and EMBOSS.
<br><br>
<i>Rice P, Longden I and Bleasby A (2000) "EMBOSS - The European
Molecular Biology Open Software Suite"  Trends in Genetics,
15:276-278.

<p>
See also <a href="http://emboss.sourceforge.net/">http://emboss.sourceforge.net/</a></i>




<H3>14.1 Other useful references  </H3>
<pre>
1. Frishman, D. & Argos, P. (1995).  Knowledge-Based Secondary Structure Assignment.  Proteins 23:566-579.

2. Hubbard,S.J. & Thornton, J.M. (1993), 'NACCESS', Computer Program, Department of Biochemistry and Molecular Biology, University College London.
</pre>
</BODY>
</HTML>
