<!-- START OF HEADER -->
<HTML><HEAD>
<TITLE> EMBASSY: PDBTOSP documentation. </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

<H2> PDBTOSP documentation
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<!-- END OF HEADER -->




<!-- CONTENTS
     This always includes the sections below.
     Other subsections can be added for individual applications.
-->
<H2>CONTENTS </H2>
<b> <a href="#1.0">1.0     SUMMARY                   </a></b><br>
<b> <a href="#2.0">2.0     INPUTS & OUTPUTS          </a></b><br>
<b> <a href="#3.0">3.0     INPUT FILE FORMAT         </a></b><br>
<b> <a href="#4.0">4.0     OUTPUT FILE FORMAT        </a></b><br>
<b> <a href="#5.0">5.0     DATA FILES                </a></b><br>
<b> <a href="#6.0">6.0     USAGE                     </a></b><br>    
<b> <a href="#7.0">7.0     KNOWN BUGS & WARNINGS     </a></b><br>    
<b> <a href="#8.0">8.0     NOTES                     </a></b><br>
<b> <a href="#9.0">9.0     DESCRIPTION               </a></b><br>
<b> <a href="#10.0">10.0   ALGORITHM                 </a></b><br>
<b> <a href="#11.0">11.0   RELATED APPLICATIONS      </a></b><br>
<b> <a href="#12.0">12.0   DIAGNOSTIC ERROR MESSAGES </a></b><br>
<b> <a href="#13.0">13.0   AUTHORS                   </a></b><br>
<b> <a href="#14.0">14.0   REFERENCES                </a></b><br>


<!-- SUMMARY
     Succint description of the application, particularly its inputs, outputs
     and what it does.  The same text is given at the top of the source (.c)
     file and in the <documentation> attribute of the <application definition>
     of the ACD file.
-->
<a name="1.0"></a>
<br><br><br><H2> 1.0   SUMMARY  </H2>
Convert raw swissprot:PDB equivalence file to EMBL-like format.
Convert swissprot:PDB codes file to EMBL-like format



<!-- INPUTS & OUTPUTS
     Short summary of the application inputs and outputs in its different 
     modes of usage (if appropriate). More detail than the summary.
-->
<a name="2.0"></a>
<br><br><br><H2> 2.0   INPUTS & OUTPUTS          </H2>

PDBTOSP parses the swissprot:PDB equivalence table available at <a href ="http://www.expasy.ch/cgi-bin/lists?pdbtosp.txt">http://www.expasy.ch/cgi-bin/lists?pdbtosp.txt</a> and writes the data out in EMBL-like format. The raw (input) file can be obtained by doing a "save as ... text format" from the web page. No changes are made to the data other than changing the format in which it is held.  The input and output files are specified by the user.

<!-- INPUT FILE FORMAT 
     Description and example(s) of input file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.
     Use "<b>PDBTOSP</b> reads any normal sequence USAs." if
     appropriate.
-->



<a name="3.0"></a>
<br><br><br><H2> 3.0   INPUT FILE FORMAT         </H2>
An excerpt from the raw swissprot:PDB equivalence file is shown (Figure 1).


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: pdbtosp.txt</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
  ------------------------------------------------------------------------
   ExPASy Home page   Site Map    Search ExPASy   Contact us    SWISS-PROT

 Hosted by SIB       Mirror                                        USA[new]
 Switzerland         sites:      AustraliaCanada China Korea Taiwan
  ------------------------------------------------------------------------

----------------------------------------------------------------------------
        SWISS-PROT Protein Knowledgebase
        Swiss Institute of Bioinformatics (SIB); Geneva, Switzerland
        European Bioinformatics Institute (EBI); Hinxton, United Kingdom
----------------------------------------------------------------------------

Description: Index of Protein Data Bank (PDB) entries referenced in
             SWISS-PROT
Name:        PDBTOSP.TXT
Release:     40.9 of 31-Jan-2002

----------------------------------------------------------------------------

The PDB database is available at the following URL:

USA:  http://www.rcsb.org/pdb/
EBI:  http://www2.ebi.ac.uk/pdb/

 - Number of PDB entries referenced in SWISS-PROT: 9901
 - Number of SWISS-PROT entries with one or more pointers to PDB: 3260

PDB   Last revision
code  date           SWISS-PROT entry name(s)
____  ___________    __________________________________________
101M  (08-APR-98)  : MYG_PHYCA   (P02185)
102L  (31-OCT-93)  : LYCV_BPT4   (P00720)
102M  (08-APR-98)  : MYG_PHYCA   (P02185)
103L  (31-OCT-93)  : LYCV_BPT4   (P00720)
103M  (08-APR-98)  : MYG_PHYCA   (P02185)
9XIA  (15-JUL-92)  : XYLA_STRRU  (P24300)
9XIM  (15-JUL-93)  : XYLA_ACTMI  (P12851)

----------------------------------------------------------------------------
SWISS-PROT is copyright.  It is produced through a collaboration between the
Swiss Institute  of  Bioinformatics   and the EMBL Outstation - the European
Bioinformatics Institute. There are no restrictions on its use by non-profit
institutions as long as its  content is in no way modified. Usage by and for
commercial entities requires a license agreement.  For information about the
licensing  scheme  see: http://www.isb-sib.ch/announce/ or send  an email to
license@isb-sib.ch.
----------------------------------------------------------------------------

  ------------------------------------------------------------------------
   ExPASy Home page   Site Map    Search ExPASy   Contact us    SWISS-PROT

 Hosted by SIB       Mirror                                        USA[new]
 Switzerland         sites:      AustraliaCanada China Korea Taiwan
  ------------------------------------------------------------------------
</pre>
</td></tr></table><p>







<!-- OUTPUT FILE FORMAT 
     Description and example(s) of output file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.  
     If the standard description of the avalable report formats is required, 
     use:   #include file="inc/reportformats.ihtml"
     Use "Outputs a graph to the specified graphics device."
     or  "outputs a report format file. The default format is ..."
     if appropriate.
-->
<a name="4.0"></a>
<br><br><br><H2> 4.0   OUTPUT FILE FORMAT         </H2>
An excerpt from the swissprot:PDB equivalence file in EMBL-like format is shown (Figure 2). 


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: Epdbtosp.dat</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
EN   101M
XX
NE   1
XX
IN   MYG_PHYCA ID; P02185 ACC;
XX
//
EN   102L
XX
NE   1
XX
IN   LYCV_BPT4 ID; P00720 ACC;
XX
//
EN   102M
XX
NE   1
XX
IN   MYG_PHYCA ID; P02185 ACC;
XX
//
EN   103L
XX
NE   1
XX
IN   LYCV_BPT4 ID; P00720 ACC;
XX
//
EN   103M
XX
NE   1
XX
IN   MYG_PHYCA ID; P02185 ACC;
XX
//
EN   9XIA
XX
NE   1
XX
IN   XYLA_STRRU ID; P24300 ACC;
XX
//
EN   9XIM
XX
NE   1
XX
IN   XYLA_ACTMI ID; P12851 ACC;
XX
//
</pre>
</td></tr></table><p>




<!-- DATA FILES         
     Any data files used (e.g. translation table file, substitution matrix 
     etc.  This includes example data file formats if they are not obvious.
     For a standard description of what data files are and how embossdata can
     be used to inspect and retrieve them, use:
     #include file="inc/localfiles.ihtml" 
-->

<a name="5.0"></a>
<br><br><br><H2> 5.0   DATA FILES                </H2>
No data files are used by PDBTOSP.

<!-- USAGE
     Example usage, as run from the command-line.
     Many examples illustrating different behaviours is good.
-->
<a name="6.0"></a>
<br><br><br><H2> 6.0   USAGE                     </H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" >
<tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-infile]            infile     This option specifies the name of raw
                                  swissprot:PDB equivalence file (input).
                                  HETPARSE parses this file, which is
                                  available at URL
                                  http://www.expasy.ch/cgi-bin/lists?pdbtosp.txt
  [-outfile]           outfile    This option specifies the name of
                                  swissprot:PDB equivalence file (EMBL-like
                                  format). This is the PDBTOSP output file.

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-outfile" associated qualifiers
   -odirectory2        string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report deaths

</pre>
</pre>
</td></tr>
</table>


<H3> 6.1   COMMAND LINE ARGUMENTS </H3>
 
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-infile]<br>(Parameter 1)</td>
<td>This option specifies the name of raw swissprot:PDB equivalence file (input). HETPARSE parses this file, which is available at URL http://www.expasy.ch/cgi-bin/lists?pdbtosp.txt</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>This option specifies the name of swissprot:PDB equivalence file (EMBL-like format). This is the PDBTOSP output file.</td>
<td>Output file</td>
<td>Epdbtosp.dat</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<H3> 6.2 EXAMPLE SESSION </H3>
An example of interactive use of PDBTOSP is shown below.

<b>Here is a sample session with pdbtosp</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>pdbtosp </b>
Convert swissprot:PDB codes file to EMBL-like format.
Name of raw swissprot:PDB equivalence file (input): <b>pdbtosp.txt</b>
Name of swissprot:PDB equivalence file (EMBL-like format output) [Epdbtosp.dat]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>


<!-- <br><br>
The raw swissprot:PDB equivalence file /testdata/pdbtosp.txt was parsed and a new file in EMBL-like format called /testdata/pdbtosp/Epdbtosp.dat was written.
<br><br>


<!-- KNOWN BUGS & WARNINGS
     Bugs that have not yet been fixed, easily missued features, problems
     and caveats etc. Potentially stupid things the program will let you do.
-->
<a name="7.0"></a>
<br><br><br><H2> 7.0   KNOWN BUGS & WARNINGS     </H2>
None.








<!-- NOTES
     Important general remarks, including:
       Restrictions.
       Interesting behaviour.
       Useful things you can do with this program.
       Future plans.
       etc.
-->
<a name="8.0"></a>
<br><br><br><H2> 8.0   NOTES                     </H2>
PDBTOSP is used to create the EMBOSS data file Epdbtosp.dat that is included in the EMBOSS distribution.


<H3> 8.1   GLOSSARY OF FILE TYPES    </H3>
<table BORDER CELLSPACING=5 CELLPADDING=5 BGCOLOR="#f5f5ff" >
<tr>
<td><b>FILE TYPE</b></td>
<td><b>FORMAT</b></td>
<td><b>DESCRIPTION</b></td>
<td><b>CREATED BY <b></td>
<td><b>SEE ALSO</b></td>
</tr>
<tr>
<td><b> swissprot:PDB equivalence file</b></td>
<td> EMBL-like format. </td>
<td> A file containing swissprot identifiers for PDB codes. </td>
<td> Included in the EMBOSS distribution</td>
<td> N.A. </td>
</tr>
</table>



None


<!-- DESCRIPTION
     A complete, non-technical, user-level description of the application.
-->
<a name="9.0"></a>
<br><br><br><H2> 9.0   DESCRIPTION               </H2>

Some research applications require knowledge of the database sequence(s) that corresponds to the sequence(s) given in a PDB file.  A 'swissprot:PDB equivalence' file listing accession numbers and swissprot database identifiers for certain PDB codes is available, but is not in a format that is consistent with flat file formats used for protein structural data in emboss.  PDBTOSP parses the swissprot:PDB equivalence in its raw format and converts it to an EMBL-like format.

<!-- ALGORITHM
     A technical description of algorithmic aspects, describing exactly how
     the key aspects of the application work.
-->
<a name="10.0"></a>
<br><br><br><H2> 10.0  ALGORITHM                 </H2>
PDBTOSP relies on finding a line beginning with '____  _' in the input file (all lines up to and including this one are ignored). Lines of code data are then parsed, up until the first blank line.



<!-- RELATED APPLICATIONS
     Other applications that either generate the input, use the output or 
     are in some other way related to the application are described here. 
-->

<a name="11.0"></a>
<br><br><br><H2> 11.0  RELATED APPLICATIONS      </H2>

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="../../apps/aaindexextract.html">aaindexextract</a></td><td>Extract data from AAINDEX</td></tr>
<tr><td><a href="../domalign/allversusall.html">allversusall</a></td><td>Sequence similarity data from all-versus-all comparison</td></tr>
<tr><td><a href="../domainatrix/cathparse.html">cathparse</a></td><td>Generates DCF file from raw CATH files</td></tr>
<tr><td><a href="../../apps/cutgextract.html">cutgextract</a></td><td>Extract data from CUTG</td></tr>
<tr><td><a href="domainer.html">domainer</a></td><td>Generates domain CCF files from protein CCF files</td></tr>
<tr><td><a href="../domainatrix/domainnr.html">domainnr</a></td><td>Removes redundant domains from a DCF file</td></tr>
<tr><td><a href="../domainatrix/domainseqs.html">domainseqs</a></td><td>Adds sequence records to a DCF file</td></tr>
<tr><td><a href="../domainatrix/domainsse.html">domainsse</a></td><td>Add secondary structure records to a DCF file</td></tr>
<tr><td><a href="hetparse.html">hetparse</a></td><td>Converts heterogen group dictionary to EMBL-like format</td></tr>
<tr><td><a href="pdbplus.html">pdbplus</a></td><td>Add accessibility & secondary structure to a CCF file</td></tr>
<tr><td><a href="pdbtosp.html">pdbtosp</a></td><td>Convert swissprot:PDB codes file to EMBL-like format</td></tr>
<tr><td><a href="../../apps/printsextract.html">printsextract</a></td><td>Extract data from PRINTS</td></tr>
<tr><td><a href="../../apps/prosextract.html">prosextract</a></td><td>Build the PROSITE motif database for use by patmatmotifs</td></tr>
<tr><td><a href="../../apps/rebaseextract.html">rebaseextract</a></td><td>Extract data from REBASE</td></tr>
<tr><td><a href="../domainatrix/scopparse.html">scopparse</a></td><td>Generate DCF file from raw SCOP files</td></tr>
<tr><td><a href="../domsearch/seqnr.html">seqnr</a></td><td>Removes redundancy from DHF files</td></tr>
<tr><td><a href="sites.html">sites</a></td><td>Generate residue-ligand CON files from CCF files</td></tr>
<tr><td><a href="../domainatrix/ssematch.html">ssematch</a></td><td>Search a DCF file for secondary structure matches</td></tr>
<tr><td><a href="../../apps/tfextract.html">tfextract</a></td><td>Extract data from TRANSFAC</td></tr>
</table>




<!-- DIAGNOSTIC ERROR MESSAGES
     Description of error messages or log file, if one is written.
-->
<a name="12.0"></a>
<br><br><br><H2> 12.0  DIAGNOSTIC ERROR MESSAGES </H2>
None.



<!-- AUTHORS
     The main author first then all others. 
-->
<a name="13.0"></a>
<br><br><br><H2> 13.0  AUTHORS                   </H2>
Jon Ison <a href="mailto:jison@rfcgr.mrc.ac.uk">(jison@rfcgr.mrc.ac.uk)</a>
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK




<!-- REFERENCES
     Quote the paper where the application was first published, described, used etc. 
-->
<a name="14.0"></a>
<br><br><br><H2> 14.0  REFERENCES                </H2>

Please cite the authors and EMBOSS.
<br><br>
<i>Rice P, Longden I and Bleasby A (2000) "EMBOSS - The European
Molecular Biology Open Software Suite"  Trends in Genetics,
15:276-278.

<p>
See also <a href="http://emboss.sourceforge.net/">http://emboss.sourceforge.net/</a></i>




<H3>14.1 Other useful references  </H3>


</BODY>
</HTML>



