<HTML>

<HEAD>
  <TITLE>
  EMBOSS: pepcoil
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: pepcoil
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program pepcoil
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Predicts coiled coil regions


<H2>
    Description
</H2>


Coiled coils are formed by two or three alpha helices in parallel and in
register that cross at an angle of approximately 20 degrees, are
strongly amphipathic and display a pattern of hydrophilic and
hydrophobic residues that is repeated every seven residues.  The seven
positions of the heptad repeat are designated a through g, a and d being
generally hydrophobic, while the others are hydrophilic. 

<p>
The parallel two-stranded alpha-helical coiled coil is the most
frequently encountered subunit-oligomerization motif in proteins. 

<p>
pepcoil calculates the probability of a coiled-coil structure for
windows of 28 residues through a protein sequence using the method of
Lupas A, van Dyke M & Stock J (1991); Science 252:1162-4

<H2>
    Usage
</H2>


Here is a sample session with pepcoil.

<pre>

% pepcoil
Input sequence: sw:gcn4_yeast
Window size [28]: 
Output file [gcn4_yeast.pepcoil]: 

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
   -window             integer    Window size
  [-outfile]           outfile    Output file name

   Optional qualifiers: (none)
   Advanced qualifiers:
   -[no]coil           bool       Report coiled coil regions
   -frame              bool       Show coil frameshifts
   -[no]other          bool       Report non coiled coil regions

   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-window</td>
<td>Window size</td>
<td>Integer from 7 to 28</td>
<td>28</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.pepcoil</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]coil</td>
<td>Report coiled coil regions</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-frame</td>
<td>Show coil frameshifts</td>
<td>Yes/No</td>
<td>Yes if -coil is true</td>
</tr>

<tr>
<td>-[no]other</td>
<td>Report non coiled coil regions</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

</table>



<H2>
    Input file format
</H2>

It reads in a protein sequence.

<H2>
    Output file format
</H2>


Here is the output file from the example run. The SwissProt annotation
marks the true leucine zipper motif as from 253 to 274. The leucine zipper
is a special case of a coiled-coil region.

<p>

<hr>
<pre>
PEPCOIL of GCN4_YEAST
   using a window of 28 residues


Other structures from 1 to 232 (232 residues)
   Max score: 1.283 (probability 0.21)

Prediction starts at 233
Probable coiled-coil from 233 to 281 (49 residues)
   Max score: 1.910 (probability 1.00)
</pre>
<hr>

<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>


<ol>

<li>
Lupas A, van Dyke M & Stock J; 
Predicting Coiled Coils from Protein Sequences.
Science 252:1162-4 (1991)

</ol>

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="antigenic.html">antigenic</a></td><td>Finds antigenic sites in proteins</td></tr>
<tr><td><a href="digest.html">digest</a></td><td>Protein proteolytic enzyme or reagent cleavage digest</td></tr>
<tr><td><a href="fuzzpro.html">fuzzpro</a></td><td>Protein pattern search</td></tr>
<tr><td><a href="fuzztran.html">fuzztran</a></td><td>Protein pattern search after translation</td></tr>
<tr><td><a href="garnier.html">garnier</a></td><td>Predicts protein secondary structure</td></tr>
<tr><td><a href="helixturnhelix.html">helixturnhelix</a></td><td>Report nucleic acid binding motifs</td></tr>
<tr><td><a href="hmoment.html">hmoment</a></td><td>Hydrophobic moment calculation</td></tr>
<tr><td><a href="oddcomp.html">oddcomp</a></td><td>Finds protein sequence regions with a biased composition</td></tr>
<tr><td><a href="patmatdb.html">patmatdb</a></td><td>Search a protein sequence with a motif</td></tr>
<tr><td><a href="patmatmotifs.html">patmatmotifs</a></td><td>Search a PROSITE motif database with a protein sequence</td></tr>
<tr><td><a href="pepnet.html">pepnet</a></td><td>Displays proteins as a helical net</td></tr>
<tr><td><a href="pepwheel.html">pepwheel</a></td><td>Shows protein sequences as helices</td></tr>
<tr><td><a href="preg.html">preg</a></td><td>Regular expression search of a protein sequence</td></tr>
<tr><td><a href="pscan.html">pscan</a></td><td>Scans proteins using PRINTS</td></tr>
<tr><td><a href="sigcleave.html">sigcleave</a></td><td>Reports protein signal cleavage sites</td></tr>
<tr><td><a href="tmap.html">tmap</a></td><td>Displays membrane spanning regions</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>

<p>
Original program "PEPCOIL" by Peter Rice (EGCG 1991)

<H2>
    History
</H2>

Written (1999) - Alan Bleasby

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

