<HTML>

<HEAD>
  <TITLE>
  EMBOSS: pepcoil
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">




<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

pepcoil
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>






<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Predicts coiled coil regions


<H2>
    Description
</H2>


Coiled coils are formed by two or three alpha helices in parallel and in
register that cross at an angle of approximately 20 degrees, are
strongly amphipathic and display a pattern of hydrophilic and
hydrophobic residues that is repeated every seven residues.  The seven
positions of the heptad repeat are designated a through g, a and d being
generally hydrophobic, while the others are hydrophilic. 

<p>
The parallel two-stranded alpha-helical coiled coil is the most
frequently encountered subunit-oligomerization motif in proteins. 

<p>
pepcoil calculates the probability of a coiled-coil structure for
windows of 28 residues through a protein sequence using the method of
Lupas A, van Dyke M & Stock J (1991); Science 252:1162-4

<H2>
    Usage
</H2>


<b>Here is a sample session with pepcoil</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>pepcoil </b>
Predicts coiled coil regions
Input sequence(s): <b>tsw:gcn4_yeast</b>
Window size [28]: <b></b>
Output file [gcn4_yeast.pepcoil]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>




<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-sequence]          seqall     Sequence database USA
   -window             integer    Window size
  [-outfile]           outfile    Output file name

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers:
   -[no]coil           boolean    Report coiled coil regions
   -frame              boolean    Show coil frameshifts
   -[no]other          boolean    Report non coiled coil regions

   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1             integer    First base used
   -send1               integer    Last base used, def=seq length
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name

   "-outfile" associated qualifiers
   -odirectory2         string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-window</td>
<td>Window size</td>
<td>Integer from 7 to 28</td>
<td>28</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.pepcoil</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]coil</td>
<td>Report coiled coil regions</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-frame</td>
<td>Show coil frameshifts</td>
<td>Boolean value Yes/No</td>
<td>Yes if -coil is true</td>
</tr>

<tr>
<td>-[no]other</td>
<td>Report non coiled coil regions</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

</table>



<H2>
    Input file format
</H2>

<b>pepcoil</b> reads in a protein sequence USA.


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tsw:gcn4_yeast' is a sequence entry in the example protein database 'tsw'
<p>
<p><h3>Database entry: tsw:gcn4_yeast</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   GCN4_YEAST     STANDARD;      PRT;   281 AA.
AC   P03069; P03068;
DT   21-JUL-1986 (Rel. 01, Created)
DT   21-JUL-1986 (Rel. 01, Last sequence update)
DT   15-JUL-1999 (Rel. 38, Last annotation update)
DE   GENERAL CONTROL PROTEIN GCN4 (AMINO ACID BIOSYNTHESIS REGULATORY
DE   PROTEIN).
GN   GCN4 OR ARG9 OR AAS3 OR YEL009C.
OS   Saccharomyces cerevisiae (Baker's yeast).
OC   Eukaryota; Fungi; Ascomycota; Hemiascomycetes; Saccharomycetales;
OC   Saccharomycetaceae; Saccharomyces.
RN   [1]
RP   SEQUENCE FROM N.A.
RX   MEDLINE; 85038531.
RA   HINNEBUSCH A.G.;
RT   "Evidence for translational regulation of the activator of general
RT   amino acid control in yeast.";
RL   Proc. Natl. Acad. Sci. U.S.A. 81:6442-6446(1984).
RN   [2]
RP   SEQUENCE FROM N.A.
RX   MEDLINE; 84298088.
RA   THIREOS G., PENN M.D., GREER H.;
RT   "5' untranslated sequences are required for the translational control
RT   of a yeast regulatory gene.";
RL   Proc. Natl. Acad. Sci. U.S.A. 81:5096-5100(1984).
RN   [3]
RP   SEQUENCE FROM N.A.
RC   STRAIN=S288C / AB972;
RA   DIETRICH F.S., MULLIGAN J.T., HENNESSEY K.M., ALLEN E., ARAUJO R.,
RA   AVILES E., BERNO A., BRENNAN T., CARPENTER J., CHEN E., CHERRY J.M.,
RA   CHUNG E., DUNCAN M., GUZMAN E., HARTZELL G., HUNICKE-SMITH S.,
RA   HYMAN R., KAYSER A., KOMP C., LASHKARI D., LEW H., LIN D.,
RA   MOSEDALE D., NAKAHARA K., NAMATH A., NORGREN R., OEFNER P., OH C.,
RA   PETEL F.X., ROBERTS D., SEHL P., SCHRAMM S., SHOGREN T., SMITH V.,
RA   TAYLOR P., WEI Y., YELTON M., BOTSTEIN D., DAVIS R.W.;
RL   Submitted (DEC-1994) to the EMBL/GenBank/DDBJ databases.
RN   [4]
RP   DOMAINS.
RX   MEDLINE; 87002456.
RA   HOPE I.A., STRUHL K.;
RT   "Functional dissection of a eukaryotic transcriptional activator
RT   protein, GCN4 of yeast.";
RL   Cell 46:885-894(1986).
RN   [5]
RP   X-RAY CRYSTALLOGRAPHY (1.8 ANGSTROMS) OF 250-281.
RX   MEDLINE; 92054531.
RA   O'SHEA E.K., KLEMM J.D., KIM P.S., ALBER T.;
RT   "X-ray structure of the GCN4 leucine zipper, a two-stranded, parallel
RT   coiled coil.";
RL   Science 254:539-544(1991).


<font color=red>  [Part of this file has been deleted for brevity]</font>

CC   -!- SUBCELLULAR LOCATION: NUCLEAR.
CC   -!- SIMILARITY: TO OTHER BZIP PROTEINS.
CC   --------------------------------------------------------------------------
CC   This SWISS-PROT entry is copyright. It is produced through a collaboration
CC   between  the Swiss Institute of Bioinformatics  and the  EMBL outstation -
CC   the European Bioinformatics Institute.  There are no  restrictions on  its
CC   use  by  non-profit  institutions as long  as its content  is  in  no  way
CC   modified and this statement is not removed.  Usage  by  and for commercial
CC   entities requires a license agreement (See http://www.isb-sib.ch/announce/
CC   or send an email to license@isb-sib.ch).
CC   --------------------------------------------------------------------------
DR   EMBL; K02205; AAA34640.1; -.
DR   EMBL; K02649; AAA65521.1; -.
DR   EMBL; U18530; AAB64486.1; -.
DR   PIR; A03604; RGBYA1.
DR   PIR; A03605; RGBYA2.
DR   PDB; 1ZTA; 15-APR-93.
DR   PDB; 2ZTA; 15-OCT-92.
DR   PDB; 1YSA; 31-OCT-93.
DR   PDB; 1DGC; 22-JUN-94.
DR   PDB; 2DGC; 08-MAR-96.
DR   PDB; 1GCL; 03-JUN-95.
DR   PDB; 1GCM; 29-JAN-96.
DR   PDB; 1SWI; 08-NOV-96.
DR   PDB; 1ZII; 07-JUL-97.
DR   PDB; 1ZIJ; 07-JUL-97.
DR   PDB; 1ZIK; 07-JUL-97.
DR   PDB; 1ZIL; 07-JUL-97.
DR   PDB; 1ZIM; 07-JUL-97.
DR   PDB; 1PIQ; 30-SEP-98.
DR   PDB; 1CE9; 25-MAR-99.
DR   TRANSFAC; T00321; -.
DR   SGD; L0000683; GCN4.
DR   PFAM; PF00170; bZIP; 1.
DR   PROSITE; PS00036; BZIP_BASIC; 1.
KW   Amino-acid biosynthesis; Transcription regulation; Activator;
KW   DNA-binding; Nuclear protein; 3D-structure.
FT   DOMAIN      106    125       REQUIRED FOR TRANSCRIPTIONAL ACTIVATION.
FT   DNA_BIND    231    249       BASIC MOTIF.
FT   DOMAIN      253    274       LEUCINE-ZIPPER.
FT   CONFLICT    239    281       ARRSRARKLQRMKQLEDKVEELLSKNYHLENEVARLKKLVG
FT                                ER -&gt; PGVLVRESCKE (IN REF. 2).
FT   HELIX       227    276
FT   TURN        277    277
SQ   SEQUENCE   281 AA;  31310 MW;  A2155238 CRC32;
     MSEYQPSLFA LNPMGFSPLD GSKSTNENVS ASTSTAKPMV GQLIFDKFIK TEEDPIIKQD
     TPSNLDFDFA LPQTATAPDA KTVLPIPELD DAVVESFFSS STDSTPMFEY ENLEDNSKEW
     TSLFDNDIPV TTDDVSLADK AIESTEEVSL VPSNLEVSTT SFLPTPVLED AKLTQTRKVK
     KPNSVVKKSH HVGKDDESRL DHLGVVAYNR KQRSIPLSPI VPESSDPAAL KRARNTEAAR
     RSRARKLQRM KQLEDKVEEL LSKNYHLENE VARLKKLVGE R
//
</pre>
</td></tr></table><p>


<H2>
    Output file format
</H2>



<a name="output.1"></a>
<h3>Output files for usage example </h3>
The SwissProt annotation marks the true leucine zipper motif as from 253 to 274. The leucine zipper is a special case of a coiled-coil region. <p><h3>File: gcn4_yeast.pepcoil</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
PEPCOIL of GCN4_YEAST
   using a window of 28 residues


Other structures from 1 to 232 (232 residues)
   Max score: 1.283 (probability 0.21)

Prediction starts at 233
Probable coiled-coil from 233 to 281 (49 residues)
   Max score: 1.910 (probability 1.00)
</pre>
</td></tr></table><p>



<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>


<ol>

<li>
Lupas A, van Dyke M & Stock J; 
Predicting Coiled Coils from Protein Sequences.
Science 252:1162-4 (1991)

</ol>

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="antigenic.html">antigenic</a></td><td>Finds antigenic sites in proteins</td></tr>
<tr><td><a href="digest.html">digest</a></td><td>Protein proteolytic enzyme or reagent cleavage digest</td></tr>
<tr><td><a href="epestfind.html">epestfind</a></td><td>Finds PEST motifs as potential proteolytic cleavage sites</td></tr>
<tr><td><a href="fuzzpro.html">fuzzpro</a></td><td>Protein pattern search</td></tr>
<tr><td><a href="fuzztran.html">fuzztran</a></td><td>Protein pattern search after translation</td></tr>
<tr><td><a href="garnier.html">garnier</a></td><td>Predicts protein secondary structure</td></tr>
<tr><td><a href="helixturnhelix.html">helixturnhelix</a></td><td>Report nucleic acid binding motifs</td></tr>
<tr><td><a href="hmoment.html">hmoment</a></td><td>Hydrophobic moment calculation</td></tr>
<tr><td><a href="oddcomp.html">oddcomp</a></td><td>Finds protein sequence regions with a biased composition</td></tr>
<tr><td><a href="patmatdb.html">patmatdb</a></td><td>Search a protein sequence with a motif</td></tr>
<tr><td><a href="patmatmotifs.html">patmatmotifs</a></td><td>Search a PROSITE motif database with a protein sequence</td></tr>
<tr><td><a href="pepnet.html">pepnet</a></td><td>Displays proteins as a helical net</td></tr>
<tr><td><a href="pepwheel.html">pepwheel</a></td><td>Shows protein sequences as helices</td></tr>
<tr><td><a href="preg.html">preg</a></td><td>Regular expression search of a protein sequence</td></tr>
<tr><td><a href="pscan.html">pscan</a></td><td>Scans proteins using PRINTS</td></tr>
<tr><td><a href="sigcleave.html">sigcleave</a></td><td>Reports protein signal cleavage sites</td></tr>
<tr><td><a href="tmap.html">tmap</a></td><td>Displays membrane spanning regions</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


Alan Bleasby (ableasby&nbsp;&copy;&nbsp;hgmp.mrc.ac.uk)
<br>
HGMP-RC, Genome Campus, Hinxton, Cambridge CB10 1SB, UK



<p>
Original program "PEPCOIL" by Peter Rice (EGCG 1991)

<H2>
    History
</H2>

Written (1999) - Alan Bleasby

<!--
<p>
[an error occurred while processing this directive]
-->

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

