<HTML>

<HEAD>
  <TITLE>
  EMBOSS: pepinfo
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: pepinfo
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program pepinfo
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Plots simple amino acid properties in parallel


<H2>
    Description
</H2>


pepinfo detects and displays various useful metrics about a protein
sequence. 

<p>
It can plot and display the following properties:

<ul>

<li>plots of hydrophobicity (using the method of Kyte & Doolittle), of
OHM (Sweet & Eisenberg), or of consensus parameters (Eisenberg et al))

<li>a histogram of the presence of residues with the physico-chemical
properties: Tiny, Small, Aliphatic, Aromatic, Non-polar, Polar,
Charged, Positive, Negative.

</ul>

<p>
The data are also written out to a data file.

<H2>
    Usage
</H2>


Here is a sample session with pepinfo.

<pre>

% pepinfo
Input sequence : sw:opsd_human
Output file [pepinfo.out]: 
Which graphical output format would you like
      1 : xwin
      2 : postscript
      3 : none
Graphical output format [1]: 
Which plots would you like
      1 : general properties
      2 : hydropathy
Please select one or more options [1, 2]: 

</pre>

<p>
<a href="pepinfo1.gif">Click here to see the histogram</a>

<p>
<a href="pepinfo2.gif">Click here to see the hydrophobicity plots</a>


<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-inseq]             sequence   Sequence USA
   -graph              xygraph    Graph type
  [-outfile]           outfile    Output file name

   Optional qualifiers:
   -hwindow            integer    Window size for hydropathy averaging
   -aaproperties       string     User defined amino acid properties
   -aahydropathy       string     User defined hydropathy data

   Advanced qualifiers:
   -[no]generalplot    bool       plot histogram of general properties
   -[no]hydropathyplot bool       plot graphs of hydropathy


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-inseq]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-graph</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including postscript, ps, hpgl, hp7470, hp7580, meta, colourps, cps, xwindows, x11, tektronics, tekt, tek4107t, tek, none, null, text, data, xterm</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td>pepinfo.out</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-hwindow</td>
<td>Window size for hydropathy averaging</td>
<td>Integer 1 or more</td>
<td>9</td>
</tr>

<tr>
<td>-aaproperties</td>
<td>User defined amino acid properties</td>
<td>Any string is accepted</td>
<td>Eaa_properties.dat</td>
</tr>

<tr>
<td>-aahydropathy</td>
<td>User defined hydropathy data</td>
<td>Any string is accepted</td>
<td>Eaa_hydropathy.dat</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]generalplot</td>
<td>plot histogram of general properties</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]hydropathyplot</td>
<td>plot graphs of hydropathy</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

</table>



<H2>
    Input file format
</H2>


<H2>
    Output file format
</H2>


The output file contains the coordinates from the graphs.

<p>
For the first set of graphs, 9 sets of true/false values are written out.
For the second set of graphs, 3 sets of hydrophobicity values are written.
<pre>

Printing out Tiny residues in OPSD_HUMAN from position 1 to 348

Position  Residue                       Result
       1       M                               0
       2       N                               0
       3       G                               1
       4       T                               1
       5       E                               0
       6       G                               1
       7       P                               0
       8       N                               0
       9       F                               0
      10       Y                               0

....................

     341       E                               0
     342       T                               1
     343       S                               1
     344       Q                               0
     345       V                               0
     346       A                               1
     347       P                               0
     348       A                               1



Printing out Small residues in OPSD_HUMAN from position 1 to 348

Position  Residue                       Result
       1       M                               0
       2       N                               1
       3       G                               1
       4       T                               1
       5       E                               0
       6       G                               1
       7       P                               1
       8       N                               1
       9       F                               0
      10       Y                               0
</pre>

<H2>
    Data files
</H2>


The physico-chemical properties of the residues are read from the EMBOSS
data file 'Eaa_properties.dat'.  This file can be copied into your
current directory and inspected ot altered by using the application
'embossdata -fetch'.  Another file can be specified using the qualifier
'-aaproperties'.

<p>
The hydropathy data of the residues are read from the EMBOSS data file
'Eaa_hydropathy.dat'.  This file can be copied into your current
directory and inspected ot altered by using the application 'embossdata
-fetch'.  Another file can be specified using the qualifier
'-aahydropathy'. 

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by EMBOSS environment variable EMBOSS_DATA.

<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>

<H2>
    Notes
</H2>


<H2>
    References
</H2>


<ol>

<li>
Kyte J, Doolittle RF 
A simple method for displaying the hydropathic character of a protein. 
J Mol Biol 1982 May 5;157(1):105-132

<li>
Sweet RM, Eisenberg D 
Correlation of sequence hydrophobicities measures similarity in
three-dimensional protein structure. 
J Mol Biol 1983 Dec 25;171(4):479-488

</ol>

<H2>
    Warnings
</H2>


<H2>
    Diagnostic Error Messages
</H2>


<H2>
    Exit status
</H2>


<H2>
    Known bugs
</H2>


<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="checktrans.html">checktrans</a></td><td>Reports STOP codons and ORF statistics of a protein sequence</td></tr>
<tr><td><a href="digest.html">digest</a></td><td>Protein proteolytic enzyme or reagent cleavage digest</td></tr>
<tr><td><a href="iep.html">iep</a></td><td>Calculates the isoelectric point of a protein</td></tr>
<tr><td><a href="octanol.html">octanol</a></td><td>Displays protein hydropathy</td></tr>
<tr><td><a href="pepnet.html">pepnet</a></td><td>Displays proteins as a helical net</td></tr>
<tr><td><a href="pepstats.html">pepstats</a></td><td>Protein statistics</td></tr>
<tr><td><a href="pepwheel.html">pepwheel</a></td><td>Shows protein sequences as helices</td></tr>
<tr><td><a href="pepwindow.html">pepwindow</a></td><td>Displays protein hydropathy</td></tr>
<tr><td><a href="pepwindowall.html">pepwindowall</a></td><td>Displays protein hydropathy of a set of sequences</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Mark Faller <a
href="mailto:mfaller@hgmp.mrc.ac.uk">(mfaller@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>


<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

