<HTML>

<HEAD>
  <TITLE>
  EMBOSS: pepnet
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">




<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

pepnet
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>







<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Displays proteins as a helical net


<H2>
    Description
</H2>


This is a method of displaying the residues of a protein in a simple
3,4,3,4 repeating pattern that emulates at a simple level the
arrangement of residues around an alpha helix. 

<p>
It is therefore easy to see patterns of amphipathicity that you may wish
to investigate in more detail by using displays such as pepwheel. 

<p>
You can specify which residues to mark up in squares, diamonds and
octagons. 

<H2>
    Usage
</H2>


<b>Here is a sample session with pepnet</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>pepnet -sask </b>
Displays proteins as a helical net
Input sequence: <b>tsw:gcn4_yeast</b>
     Begin at position [start]: <b>253</b>
       End at position [end]: <b>274</b>
Graph type [x11]: <b>cps</b>

Created pepnet.ps

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-sequence]          sequence   Sequence USA
   -graph              graph      Graph type

   Additional (Optional) qualifiers (* if not always prompted):
*  -squares            string     By default the aliphatic residues ILVM are
                                  marked with squares.
*  -diamonds           string     By default the residues DENQST are marked
                                  with diamonds.
*  -octags             string     By default the positively charged residues
                                  HKR are marked with octagons.

   Advanced (Unprompted) qualifiers:
   -amphipathic        toggle     If this is true then the residues ACFGILMVWY
                                  are marked as squares and all other
                                  residues are unmarked. This overrides any
                                  other markup that you may have specified
                                  using the qualifiers '-squares', '-diamonds'
                                  and '-octags'.

   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1             integer    Start of the sequence to be used
   -send1               integer    End of the sequence to be used
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name

   "-graph" associated qualifiers
   -gprompt             boolean    Graph prompting
   -gtitle              string     Graph title
   -gsubtitle           string     Graph subtitle
   -gxtitle             string     Graph x axis title
   -gytitle             string     Graph y axis title
   -goutfile            string     Output file for non interactive displays
   -gdirectory          string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-graph</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including postscript, ps, hpgl, hp7470, hp7580, meta, colourps, cps, xwindows, x11, tektronics, tekt, tek4107t, tek, none, null, text, data, xterm, png</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-squares</td>
<td>By default the aliphatic residues ILVM are marked with squares.</td>
<td>Any string is accepted</td>
<td>ILVM</td>
</tr>

<tr>
<td>-diamonds</td>
<td>By default the residues DENQST are marked with diamonds.</td>
<td>Any string is accepted</td>
<td>DENQST</td>
</tr>

<tr>
<td>-octags</td>
<td>By default the positively charged residues HKR are marked with octagons.</td>
<td>Any string is accepted</td>
<td>HKR</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-amphipathic</td>
<td>If this is true then the residues ACFGILMVWY are marked as squares and all other residues are unmarked. This overrides any other markup that you may have specified using the qualifiers '-squares', '-diamonds' and '-octags'.</td>
<td>Toggle value Yes/No</td>
<td>&nbsp;</td>
</tr>

</table>



<H2>
    Input file format
</H2>

<b>pepnet</b> reads any protein sequence USA.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tsw:gcn4_yeast' is a sequence entry in the example protein database 'tsw'
<p>
<p><h3>Database entry: tsw:gcn4_yeast</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   GCN4_YEAST     STANDARD;      PRT;   281 AA.
AC   P03069; P03068;
DT   21-JUL-1986 (Rel. 01, Created)
DT   21-JUL-1986 (Rel. 01, Last sequence update)
DT   15-JUL-1999 (Rel. 38, Last annotation update)
DE   GENERAL CONTROL PROTEIN GCN4 (AMINO ACID BIOSYNTHESIS REGULATORY
DE   PROTEIN).
GN   GCN4 OR ARG9 OR AAS3 OR YEL009C.
OS   Saccharomyces cerevisiae (Baker's yeast).
OC   Eukaryota; Fungi; Ascomycota; Hemiascomycetes; Saccharomycetales;
OC   Saccharomycetaceae; Saccharomyces.
RN   [1]
RP   SEQUENCE FROM N.A.
RX   MEDLINE; 85038531.
RA   HINNEBUSCH A.G.;
RT   "Evidence for translational regulation of the activator of general
RT   amino acid control in yeast.";
RL   Proc. Natl. Acad. Sci. U.S.A. 81:6442-6446(1984).
RN   [2]
RP   SEQUENCE FROM N.A.
RX   MEDLINE; 84298088.
RA   THIREOS G., PENN M.D., GREER H.;
RT   "5' untranslated sequences are required for the translational control
RT   of a yeast regulatory gene.";
RL   Proc. Natl. Acad. Sci. U.S.A. 81:5096-5100(1984).
RN   [3]
RP   SEQUENCE FROM N.A.
RC   STRAIN=S288C / AB972;
RA   DIETRICH F.S., MULLIGAN J.T., HENNESSEY K.M., ALLEN E., ARAUJO R.,
RA   AVILES E., BERNO A., BRENNAN T., CARPENTER J., CHEN E., CHERRY J.M.,
RA   CHUNG E., DUNCAN M., GUZMAN E., HARTZELL G., HUNICKE-SMITH S.,
RA   HYMAN R., KAYSER A., KOMP C., LASHKARI D., LEW H., LIN D.,
RA   MOSEDALE D., NAKAHARA K., NAMATH A., NORGREN R., OEFNER P., OH C.,
RA   PETEL F.X., ROBERTS D., SEHL P., SCHRAMM S., SHOGREN T., SMITH V.,
RA   TAYLOR P., WEI Y., YELTON M., BOTSTEIN D., DAVIS R.W.;
RL   Submitted (DEC-1994) to the EMBL/GenBank/DDBJ databases.
RN   [4]
RP   DOMAINS.
RX   MEDLINE; 87002456.
RA   HOPE I.A., STRUHL K.;
RT   "Functional dissection of a eukaryotic transcriptional activator
RT   protein, GCN4 of yeast.";
RL   Cell 46:885-894(1986).
RN   [5]
RP   X-RAY CRYSTALLOGRAPHY (1.8 ANGSTROMS) OF 250-281.
RX   MEDLINE; 92054531.
RA   O'SHEA E.K., KLEMM J.D., KIM P.S., ALBER T.;
RT   "X-ray structure of the GCN4 leucine zipper, a two-stranded, parallel
RT   coiled coil.";
RL   Science 254:539-544(1991).


<font color=red>  [Part of this file has been deleted for brevity]</font>

CC   -!- SUBCELLULAR LOCATION: NUCLEAR.
CC   -!- SIMILARITY: TO OTHER BZIP PROTEINS.
CC   --------------------------------------------------------------------------
CC   This SWISS-PROT entry is copyright. It is produced through a collaboration
CC   between  the Swiss Institute of Bioinformatics  and the  EMBL outstation -
CC   the European Bioinformatics Institute.  There are no  restrictions on  its
CC   use  by  non-profit  institutions as long  as its content  is  in  no  way
CC   modified and this statement is not removed.  Usage  by  and for commercial
CC   entities requires a license agreement (See http://www.isb-sib.ch/announce/
CC   or send an email to license@isb-sib.ch).
CC   --------------------------------------------------------------------------
DR   EMBL; K02205; AAA34640.1; -.
DR   EMBL; K02649; AAA65521.1; -.
DR   EMBL; U18530; AAB64486.1; -.
DR   PIR; A03604; RGBYA1.
DR   PIR; A03605; RGBYA2.
DR   PDB; 1ZTA; 15-APR-93.
DR   PDB; 2ZTA; 15-OCT-92.
DR   PDB; 1YSA; 31-OCT-93.
DR   PDB; 1DGC; 22-JUN-94.
DR   PDB; 2DGC; 08-MAR-96.
DR   PDB; 1GCL; 03-JUN-95.
DR   PDB; 1GCM; 29-JAN-96.
DR   PDB; 1SWI; 08-NOV-96.
DR   PDB; 1ZII; 07-JUL-97.
DR   PDB; 1ZIJ; 07-JUL-97.
DR   PDB; 1ZIK; 07-JUL-97.
DR   PDB; 1ZIL; 07-JUL-97.
DR   PDB; 1ZIM; 07-JUL-97.
DR   PDB; 1PIQ; 30-SEP-98.
DR   PDB; 1CE9; 25-MAR-99.
DR   TRANSFAC; T00321; -.
DR   SGD; L0000683; GCN4.
DR   PFAM; PF00170; bZIP; 1.
DR   PROSITE; PS00036; BZIP_BASIC; 1.
KW   Amino-acid biosynthesis; Transcription regulation; Activator;
KW   DNA-binding; Nuclear protein; 3D-structure.
FT   DOMAIN      106    125       REQUIRED FOR TRANSCRIPTIONAL ACTIVATION.
FT   DNA_BIND    231    249       BASIC MOTIF.
FT   DOMAIN      253    274       LEUCINE-ZIPPER.
FT   CONFLICT    239    281       ARRSRARKLQRMKQLEDKVEELLSKNYHLENEVARLKKLVG
FT                                ER -&gt; PGVLVRESCKE (IN REF. 2).
FT   HELIX       227    276
FT   TURN        277    277
SQ   SEQUENCE   281 AA;  31310 MW;  A2155238 CRC32;
     MSEYQPSLFA LNPMGFSPLD GSKSTNENVS ASTSTAKPMV GQLIFDKFIK TEEDPIIKQD
     TPSNLDFDFA LPQTATAPDA KTVLPIPELD DAVVESFFSS STDSTPMFEY ENLEDNSKEW
     TSLFDNDIPV TTDDVSLADK AIESTEEVSL VPSNLEVSTT SFLPTPVLED AKLTQTRKVK
     KPNSVVKKSH HVGKDDESRL DHLGVVAYNR KQRSIPLSPI VPESSDPAAL KRARNTEAAR
     RSRARKLQRM KQLEDKVEEL LSKNYHLENE VARLKKLVGE R
//
</pre>
</td></tr></table><p>


<H2>
    Output file format
</H2>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>Graphics File: pepnet.ps</h3>
<p><img src="pepnet.1.pepnet.gif" alt="[pepnet results]">


<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It exits with a status of 0.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="abiview.html">abiview</a></td><td>Reads ABI file and display the trace</td></tr>
<tr><td><a href="cirdna.html">cirdna</a></td><td>Draws circular maps of DNA constructs</td></tr>
<tr><td><a href="garnier.html">garnier</a></td><td>Predicts protein secondary structure</td></tr>
<tr><td><a href="helixturnhelix.html">helixturnhelix</a></td><td>Report nucleic acid binding motifs</td></tr>
<tr><td><a href="hmoment.html">hmoment</a></td><td>Hydrophobic moment calculation</td></tr>
<tr><td><a href="lindna.html">lindna</a></td><td>Draws linear maps of DNA constructs</td></tr>
<tr><td><a href="pepcoil.html">pepcoil</a></td><td>Predicts coiled coil regions</td></tr>
<tr><td><a href="pepwheel.html">pepwheel</a></td><td>Shows protein sequences as helices</td></tr>
<tr><td><a href="prettyplot.html">prettyplot</a></td><td>Displays aligned sequences, with colouring and boxing</td></tr>
<tr><td><a href="prettyseq.html">prettyseq</a></td><td>Output sequence with translated ranges</td></tr>
<tr><td><a href="remap.html">remap</a></td><td>Display a sequence with restriction cut sites, translation etc</td></tr>
<tr><td><a href="seealso.html">seealso</a></td><td>Finds programs sharing group names</td></tr>
<tr><td><a href="showalign.html">showalign</a></td><td>Displays a multiple sequence alignment</td></tr>
<tr><td><a href="showdb.html">showdb</a></td><td>Displays information on the currently available databases</td></tr>
<tr><td><a href="showfeat.html">showfeat</a></td><td>Show features of a sequence</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="sixpack.html">sixpack</a></td><td>Display a DNA sequence with 6-frame translation and ORFs</td></tr>
<tr><td><a href="textsearch.html">textsearch</a></td><td>Search sequence documentation text. SRS and Entrez are faster!</td></tr>
<tr><td><a href="tmap.html">tmap</a></td><td>Displays membrane spanning regions</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->




<H2>
    Author(s)
</H2>


Alan Bleasby (ableasby&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK




<H2>
    History
</H2>

Written (2000) - Alan Bleasby

<!--
<p>
[an error occurred while processing this directive]
-->

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

