<HTML>

<HEAD>
  <TITLE>
  EMBOSS: pepstats
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: pepstats
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program pepstats
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Protein statistics


<H2>
    Description
</H2>


<b>pepstats</b> outputs a report of simple protein sequence information including:

<p>
<ul>

<li>molecular weight
<li>number of residues
<li>average residue weight
<li>charge
<li>isoelectric point
<li>for each type of amino acid: number, molar percent, DayhoffStat
<li>for each physico-chemical class of amino acid: number, molar percent
</ul>

<p>
DayhoffStat is the amino acid's Dayhoff statistic divided by the molar
percent.  The Dayhoff statistic is the amino acid's relative occurence
per 1000 aa normalised to 100 by rls@ebi.ac.uk (original work from 1993)

<H2>
    Usage
</H2>


Here is a sample session with pepstats.

<p>

<pre>
% pepstats
Protein statistics
Input sequence: sw:laci_ecoli
Output file [laci_ecoli.pepstats]:
</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequencea]         sequence   Sequence USA
   -outfile            outfile    Output file name

   Optional qualifiers: (none)
   Advanced qualifiers:
   -[no]termini        bool       Include charge at N and C terminus


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequencea]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-outfile</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.pepstats</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]termini</td>
<td>Include charge at N and C terminus</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

</table>



<H2>
    Input file format
</H2>

Normal protein sequence USA.


<H2>
    Output file format
</H2>


Here is the output from the example run:

<p>

<hr>
<pre>
PEPSTATS of LACI_ECOLI from 1 to 360

Molecular weight = 38563.98             Residues = 360   
Average Residue Weight  = 107.122       Charge   = 1.5   
Isoelectric Point = 6.8820

Residue         Number          Mole%           DayhoffStat
A = Ala         44              12.222          1.421  
B = Asx         0               0.000           0.000  
C = Cys         3               0.833           0.287  
D = Asp         17              4.722           0.859  
E = Glu         15              4.167           0.694  
F = Phe         4               1.111           0.309  
G = Gly         22              6.111           0.728  
H = His         7               1.944           0.972  
I = Ile         18              5.000           1.111  
K = Lys         11              3.056           0.463  
L = Leu         40              11.111          1.502  
M = Met         10              2.778           1.634  
N = Asn         12              3.333           0.775  
P = Pro         14              3.889           0.748  
Q = Gln         28              7.778           1.994  
R = Arg         19              5.278           1.077  
S = Ser         33              9.167           1.310  
T = Thr         19              5.278           0.865  
V = Val         34              9.444           1.431  
W = Trp         2               0.556           0.427  
X = Xxx         0               0.000           0.000  
Y = Tyr         8               2.222           0.654  
Z = Glx         0               0.000           0.000  

Property        Residues                Number          Mole%
Tiny            (A+C+G+S+T)             121             33.611
Small           (A+B+C+D+G+N+P+S+T+V)   198             55.000
Aliphatic       (I+L+V)                 92              25.556
Aromatic        (F+H+W+Y)               21               5.833
Non-polar       (A+C+F+G+I+L+M+P+V+W+Y) 199             55.278
Polar           (D+E+H+K+N+Q+R+S+T+Z)   161             44.722
Charged         (B+D+E+H+K+R+Z)         69              19.167
Basic           (H+K+R)                 37              10.278
Acidic          (B+D+E+Z)               32               8.889
</pre>
<hr>

<H2>
    Data files
</H2>


The Dayhoff statistic is read from the EMBOSS data file 'Edayhoff.freq'. 
You can inspect and modify this file by copying it into your current
directory with the command: 'embossdata -fetch'. 

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by EMBOSS environment variable EMBOSS_DATA.

<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>

<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="backtranseq.html">backtranseq</a></td><td>Back translate a protein sequence</td></tr>
<tr><td><a href="charge.html">charge</a></td><td>Protein charge plot</td></tr>
<tr><td><a href="checktrans.html">checktrans</a></td><td>Reports STOP codons and ORF statistics of a protein sequence</td></tr>
<tr><td><a href="compseq.html">compseq</a></td><td>Counts the composition of dimer/trimer/etc words in a sequence</td></tr>
<tr><td><a href="emowse.html">emowse</a></td><td>Protein identification by mass spectrometry</td></tr>
<tr><td><a href="iep.html">iep</a></td><td>Calculates the isoelectric point of a protein</td></tr>
<tr><td><a href="octanol.html">octanol</a></td><td>Displays protein hydropathy</td></tr>
<tr><td><a href="pepinfo.html">pepinfo</a></td><td>Plots simple amino acid properties in parallel</td></tr>
<tr><td><a href="pepwindow.html">pepwindow</a></td><td>Displays protein hydropathy</td></tr>
<tr><td><a href="pepwindowall.html">pepwindowall</a></td><td>Displays protein hydropathy of a set of sequences</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->





<H2>
    Author(s)
</H2>


This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>

Written (1999) - Alan Bleasby

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

