<HTML>

<HEAD>
  <TITLE>
  EMBOSS: pepstats
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">




<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

pepstats
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>




<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Protein statistics


<H2>
    Description
</H2>


<b>pepstats</b> outputs a report of simple protein sequence information including:

<p>
<ul>

<li>Molecular weight
<li>Number of residues
<li>Average residue weight
<li>Charge
<li>Isoelectric point
<li>For each type of amino acid: number, molar percent, DayhoffStat
<li>For each physico-chemical class of amino acid: number, molar percent
<li>Probability of protein expression in E. coli inclusion bodies
<li>Molar extinction coefficient (A280)
<li>Extinction coefficient at 1 mg/ml (A280)
</ul>

<p>

DayhoffStat is the amino acid's Dayhoff statistic divided by the molar
percent.  The Dayhoff statistic is the amino acid's relative occurence
per 1000 aa normalised to 100 by rls@ebi.ac.uk (original work from 1993)

<p>

The probability of expression in inclusion bodies is sometimes
referred to as a type of solubility measure. If, however, a
recombinant protein is expressed in Escherichia coli, it can be
expressed as soluble in the cytosol or insoluble in inclusion bodies. If
the Harrison model predicts a given protein to be probably expressed
in includion bodies, this doesn't mean that it is not possible to get
it soluble in the cytosol. One example: Thermatoga maritima cell
divison protein FtsA with a C-terminal His-Tag has a 58% Harrison
probability of being expressed in inclusion bodies. However, there was
plenty of soluble protein in the E. coli cytosol (F. van den Ent and
J.  Lowe, EMBO J. 19, 5300-5307 2000). If the protein is expressed in
inclusion bodies or not is not only dependent on the sequence, but
also on many other factors, such as E. coli strain, incubation
temperature, type of expression vector, strength of promoter and
medium.



<H2>
    Usage
</H2>


<b>Here is a sample session with pepstats</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>pepstats </b>
Protein statistics
Input sequence(s): <b>tsw:laci_ecoli</b>
Output file [laci_ecoli.pepstats]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-sequence]          seqall     Sequence database USA
  [-outfile]           outfile    Output file name

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers:
   -[no]termini        boolean    Include charge at N and C terminus
   -aadata             datafile   Molecular weight data for amino acids

   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1             integer    Start of each sequence to be used
   -send1               integer    End of each sequence to be used
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name

   "-outfile" associated qualifiers
   -odirectory2         string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.pepstats</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]termini</td>
<td>Include charge at N and C terminus</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-aadata</td>
<td>Molecular weight data for amino acids</td>
<td>Data file</td>
<td>Eamino.dat</td>
</tr>

</table>



<H2>
    Input file format
</H2>

<b>pepstats</b> reads a normal protein sequence USA.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tsw:laci_ecoli' is a sequence entry in the example protein database 'tsw'
<p>
<p><h3>Database entry: tsw:laci_ecoli</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   LACI_ECOLI     STANDARD;      PRT;   360 AA.
AC   P03023; P71309; Q47338; O09196;
DT   21-JUL-1986 (Rel. 01, Created)
DT   01-NOV-1997 (Rel. 35, Last sequence update)
DT   15-DEC-1998 (Rel. 37, Last annotation update)
DE   LACTOSE OPERON REPRESSOR.
GN   LACI.
OS   Escherichia coli.
OC   Bacteria; Proteobacteria; gamma subdivision; Enterobacteriaceae;
OC   Escherichia.
RN   [1]
RP   SEQUENCE FROM N.A.
RX   MEDLINE; 78246991.
RA   FARABAUGH P.J.;
RT   "Sequence of the lacI gene.";
RL   Nature 274:765-769(1978).
RN   [2]
RP   SEQUENCE FROM N.A.
RC   STRAIN=K12 / MG1655;
RX   MEDLINE; 97426617.
RA   BLATTNER F.R., PLUNKETT G. III, BLOCH C.A., PERNA N.T., BURLAND V.,
RA   RILEY M., COLLADO-VIDES J., GLASNER F.D., RODE C.K., MAYHEW G.F.,
RA   GREGOR J., DAVIS N.W., KIRKPATRICK H.A., GOEDEN M.A., ROSE D.J.,
RA   MAU B., SHAO Y.;
RT   "The complete genome sequence of Escherichia coli K-12.";
RL   Science 277:1453-1474(1997).
RN   [3]
RP   SEQUENCE FROM N.A.
RC   STRAIN=K12 / MG1655;
RA   DUNCAN M., ALLEN E., ARAUJO R., APARICIO A.M., CHUNG E., DAVIS K.,
RA   FEDERSPIEL N., HYMAN R., KALMAN S., KOMP C., KURDI O., LEW H.,
RA   LIN D., NAMATH A., OEFNER P., ROBERTS D., SCHRAMM S., DAVIS R.W.;
RL   Submitted (NOV-1996) to the EMBL/GenBank/DDBJ databases.
RN   [4]
RP   SEQUENCE FROM N.A.
RA   CHEN J., MATTHEWS K.K.S.M.;
RL   Submitted (MAY-1991) to the EMBL/GenBank/DDBJ databases.
RN   [5]
RP   SEQUENCE FROM N.A.
RA   MARSH S.;
RL   Submitted (JAN-1997) to the EMBL/GenBank/DDBJ databases.
RN   [6]
RP   SEQUENCE OF 1-147; 159-230 AND 233-360.
RX   MEDLINE; 76091932.
RA   BEYREUTHER K., ADLER K., FANNING E., MURRAY C., KLEMM A., GEISLER N.;
RT   "Amino-acid sequence of lac repressor from Escherichia coli.
RT   Isolation, sequence analysis and sequence assembly of tryptic
RT   peptides and cyanogen-bromide fragments.";
RL   Eur. J. Biochem. 59:491-509(1975).
RN   [7]


<font color=red>  [Part of this file has been deleted for brevity]</font>

CC   between  the Swiss Institute of Bioinformatics  and the  EMBL outstation -
CC   the European Bioinformatics Institute.  There are no  restrictions on  its
CC   use  by  non-profit  institutions as long  as its content  is  in  no  way
CC   modified and this statement is not removed.  Usage  by  and for commercial
CC   entities requires a license agreement (See http://www.isb-sib.ch/announce/
CC   or send an email to license@isb-sib.ch).
CC   --------------------------------------------------------------------------
DR   EMBL; V00294; CAA23569.1; -.
DR   EMBL; J01636; AAA24052.1; -.
DR   EMBL; AE000141; AAC73448.1; -.
DR   EMBL; U73857; AAB18069.1; ALT_INIT.
DR   EMBL; X58469; CAA41383.1; -.
DR   EMBL; U86347; AAB47270.1; ALT_INIT.
DR   EMBL; U72488; AAB36549.1; -.
DR   EMBL; U78872; AAB37348.1; -.
DR   EMBL; U78873; AAB37351.1; -.
DR   EMBL; U78874; AAB37354.1; -.
DR   PIR; A03558; RPECL.
DR   PIR; S02540; S02540.
DR   PDB; 1LCC; 31-JAN-94.
DR   PDB; 1LCD; 31-JAN-94.
DR   PDB; 1LTP; 31-OCT-93.
DR   PDB; 1TLF; 31-JUL-95.
DR   PDB; 1LBG; 11-JUL-96.
DR   PDB; 1LBH; 11-JUL-96.
DR   PDB; 1LBI; 11-JUL-96.
DR   PDB; 1LQC; 12-FEB-97.
DR   ECO2DBASE; H039.0; 6TH EDITION.
DR   ECOGENE; EG10525; LACI.
DR   PFAM; PF00356; lacI; 1.
DR   PFAM; PF00532; Peripla_BP_like; 1.
DR   PROSITE; PS00356; HTH_LACI_FAMILY; 1.
KW   Transcription regulation; DNA-binding; Repressor; 3D-structure.
FT   DNA_BIND      6     25       H-T-H MOTIF.
FT   MUTAGEN      17     17       Y-&gt;H: BROADENING OF SPECIFICITY.
FT   MUTAGEN      22     22       R-&gt;N: RECOGNIZE AN OPERATOR VARIANT.
FT   VARIANT     282    282       Y -&gt; D (IN T41 MUTANT).
FT   CONFLICT    286    286       S -&gt; L (IN AAA24052, REF. 2, 4 AND 5).
FT   HELIX         6     13
FT   TURN         14     14
FT   HELIX        17     24
FT   HELIX        32     44
FT   TURN         49     50
SQ   SEQUENCE   360 AA;  38564 MW;  4CA5A1D6 CRC32;
     MKPVTLYDVA EYAGVSYQTV SRVVNQASHV SAKTREKVEA AMAELNYIPN RVAQQLAGKQ
     SLLIGVATSS LALHAPSQIV AAIKSRADQL GASVVVSMVE RSGVEACKAA VHNLLAQRVS
     GLIINYPLDD QDAIAVEAAC TNVPALFLDV SDQTPINSII FSHEDGTRLG VEHLVALGHQ
     QIALLAGPLS SVSARLRLAG WHKYLTRNQI QPIAEREGDW SAMSGFQQTM QMLNEGIVPT
     AMLVANDQMA LGAMRAITES GLRVGADISV VGYDDTEDSS CYIPPSTTIK QDFRLLGQTS
     VDRLLQLSQG QAVKGNQLLP VSLVKRKTTL APNTQTASPR ALADSLMQLA RQVSRLESGQ
//
</pre>
</td></tr></table><p>


<H2>
    Output file format
</H2>



<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: laci_ecoli.pepstats</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
PEPSTATS of LACI_ECOLI from 1 to 360

Molecular weight = 38563.97  		Residues = 360   
Average Residue Weight  = 107.122 	Charge   = 1.5   
Isoelectric Point = 6.8820
A280 Molar Extinction Coefficient  = 21620   
A280 Extinction Coefficient 1mg/ml = 0.56      
Improbability of expression in inclusion bodies = 0.670

Residue		Number		Mole%		DayhoffStat
A = Ala		44		12.222 		1.421  	
B = Asx		0		0.000  		0.000  	
C = Cys		3		0.833  		0.287  	
D = Asp		17		4.722  		0.859  	
E = Glu		15		4.167  		0.694  	
F = Phe		4		1.111  		0.309  	
G = Gly		22		6.111  		0.728  	
H = His		7		1.944  		0.972  	
I = Ile		18		5.000  		1.111  	
K = Lys		11		3.056  		0.463  	
L = Leu		40		11.111 		1.502  	
M = Met		10		2.778  		1.634  	
N = Asn		12		3.333  		0.775  	
P = Pro		14		3.889  		0.748  	
Q = Gln		28		7.778  		1.994  	
R = Arg		19		5.278  		1.077  	
S = Ser		33		9.167  		1.310  	
T = Thr		19		5.278  		0.865  	
V = Val		34		9.444  		1.431  	
W = Trp		2		0.556  		0.427  	
X = Xaa		0		0.000  		0.000  	
Y = Tyr		8		2.222  		0.654  	
Z = Glx		0		0.000  		0.000  	

Property	Residues		Number		Mole%
Tiny		(A+C+G+S+T)		121		33.611
Small		(A+B+C+D+G+N+P+S+T+V)	198		55.000
Aliphatic	(I+L+V)			92		25.556
Aromatic	(F+H+W+Y)		21		 5.833
Non-polar	(A+C+F+G+I+L+M+P+V+W+Y)	199		55.278
Polar		(D+E+H+K+N+Q+R+S+T+Z)	161		44.722
Charged		(B+D+E+H+K+R+Z)		69		19.167
Basic		(H+K+R)			37		10.278
Acidic		(B+D+E+Z)		32		 8.889

</pre>
</td></tr></table><p>



<H2>
    Data files
</H2>


The Dayhoff statistic is read from the EMBOSS data file 'Edayhoff.freq'. 
You can inspect and modify this file by copying it into your current
directory with the command: 'embossdata -fetch'. 

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by EMBOSS environment variable EMBOSS_DATA.

<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>

<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

<ol>
<li>Roger G. Harrison "Expression of soluble heterologous proteins via fusion with NusA protein" in inNovations 11, June 2000, p 4 - 7.
</ol>

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="backtranseq.html">backtranseq</a></td><td>Back translate a protein sequence</td></tr>
<tr><td><a href="charge.html">charge</a></td><td>Protein charge plot</td></tr>
<tr><td><a href="checktrans.html">checktrans</a></td><td>Reports STOP codons and ORF statistics of a protein</td></tr>
<tr><td><a href="compseq.html">compseq</a></td><td>Counts the composition of dimer/trimer/etc words in a sequence</td></tr>
<tr><td><a href="emowse.html">emowse</a></td><td>Protein identification by mass spectrometry</td></tr>
<tr><td><a href="freak.html">freak</a></td><td>Residue/base frequency table or plot</td></tr>
<tr><td><a href="iep.html">iep</a></td><td>Calculates the isoelectric point of a protein</td></tr>
<tr><td><a href="mwcontam.html">mwcontam</a></td><td>Shows molwts that match across a set of files</td></tr>
<tr><td><a href="mwfilter.html">mwfilter</a></td><td>Filter noisy molwts from mass spec output</td></tr>
<tr><td><a href="octanol.html">octanol</a></td><td>Displays protein hydropathy</td></tr>
<tr><td><a href="pepinfo.html">pepinfo</a></td><td>Plots simple amino acid properties in parallel</td></tr>
<tr><td><a href="pepwindow.html">pepwindow</a></td><td>Displays protein hydropathy</td></tr>
<tr><td><a href="pepwindowall.html">pepwindowall</a></td><td>Displays protein hydropathy of a set of sequences</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->





<H2>
    Author(s)
</H2>


Alan Bleasby (ableasby&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK



<H2>
    History
</H2>

Written (1999) - Alan Bleasby

<!--
<p>
[an error occurred while processing this directive]
-->

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

