<HTML>

<HEAD>
  <TITLE>
  EMBOSS: pepwheel
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
pepwheel
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
    Function
</H2>
Shows protein sequences as helices

<H2>
    Description
</H2>

<b>pepwheel</b> displays peptide sequences in a helical representation.

<p>

This gives a view of a helix from a protein sequence looking down the
axis of the helix. 

<p>

It is useful for highlighting amphipathicity and other properties of
residues around a helix. 

<H2>
    Usage
</H2>
<b>Here is a sample session with pepwheel</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>pepwheel tsw:hbb_human -send 30 </b>
Shows protein sequences as helices
Graph type [x11]: <b>cps</b>

Created pepwheel.ps

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>


<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-sequence]          sequence   Protein sequence filename and optional
                                  format, or reference (input USA)
   -graph              graph      [$EMBOSS_GRAPHICS value, or x11] Graph type
                                  (ps, hpgl, hp7470, hp7580, meta, cps, x11,
                                  tekt, tek, none, data, xterm, png)

   Additional (Optional) qualifiers (* if not always prompted):
   -steps              integer    [18] The number of residues plotted per turn
                                  is this value divided by the 'turns' value.
                                  (Integer from 2 to 100)
   -turns              integer    [5] The number of residues plotted per turn
                                  is the 'steps' value divided by this value.
                                  (Integer from 1 to 100)
*  -squares            string     [ILVM] By default the aliphatic residues
                                  ILVM are marked with squares. (Any string is
                                  accepted)
*  -diamonds           string     [DENQST] By default the residues DENQST are
                                  marked with diamonds. (Any string is
                                  accepted)
*  -octags             string     [HKR] By default the positively charged
                                  residues HKR are marked with octagons. (Any
                                  string is accepted)

   Advanced (Unprompted) qualifiers:
   -[no]wheel          boolean    [Y] Plot the wheel
   -amphipathic        toggle     If this is true then the residues ACFGILMVWY
                                  are marked as squares and all other
                                  residues are unmarked. This overrides any
                                  other markup that you may have specified
                                  using the qualifiers '-squares', '-diamonds'
                                  and '-octags'.

   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1            integer    Start of the sequence to be used
   -send1              integer    End of the sequence to be used
   -sreverse1          boolean    Reverse (if DNA)
   -sask1              boolean    Ask for begin/end/reverse
   -snucleotide1       boolean    Sequence is nucleotide
   -sprotein1          boolean    Sequence is protein
   -slower1            boolean    Make lower case
   -supper1            boolean    Make upper case
   -sformat1           string     Input sequence format
   -sdbname1           string     Database name
   -sid1               string     Entryname
   -ufo1               string     UFO features
   -fformat1           string     Features format
   -fopenfile1         string     Features file name

   "-graph" associated qualifiers
   -gprompt            boolean    Graph prompting
   -gdesc              string     Graph description
   -gtitle             string     Graph title
   -gsubtitle          string     Graph subtitle
   -gxtitle            string     Graph x axis title
   -gytitle            string     Graph y axis title
   -goutfile           string     Output file for non interactive displays
   -gdirectory         string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages

</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Protein sequence filename and optional format, or reference (input USA)</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-graph</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including ps, hpgl, hp7470, hp7580, meta, cps, x11, tekt, tek, none, data, xterm, png</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-steps</td>
<td>The number of residues plotted per turn is this value divided by the 'turns' value.</td>
<td>Integer from 2 to 100</td>
<td>18</td>
</tr>

<tr>
<td>-turns</td>
<td>The number of residues plotted per turn is the 'steps' value divided by this value.</td>
<td>Integer from 1 to 100</td>
<td>5</td>
</tr>

<tr>
<td>-squares</td>
<td>By default the aliphatic residues ILVM are marked with squares.</td>
<td>Any string is accepted</td>
<td>ILVM</td>
</tr>

<tr>
<td>-diamonds</td>
<td>By default the residues DENQST are marked with diamonds.</td>
<td>Any string is accepted</td>
<td>DENQST</td>
</tr>

<tr>
<td>-octags</td>
<td>By default the positively charged residues HKR are marked with octagons.</td>
<td>Any string is accepted</td>
<td>HKR</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]wheel</td>
<td>Plot the wheel</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-amphipathic</td>
<td>If this is true then the residues ACFGILMVWY are marked as squares and all other residues are unmarked. This overrides any other markup that you may have specified using the qualifiers '-squares', '-diamonds' and '-octags'.</td>
<td>Toggle value Yes/No</td>
<td>No</td>
</tr>

</table>


<H2>
    Input file format
</H2>

<b>pepwheel</b> reads any protein sequence USA.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tsw:hbb_human' is a sequence entry in the example protein database 'tsw'
<p>
<p><h3>Database entry: tsw:hbb_human</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   HBB_HUMAN      STANDARD;      PRT;   146 AA.
AC   P02023;
DT   21-JUL-1986 (Rel. 01, Created)
DT   21-JUL-1986 (Rel. 01, Last sequence update)
DT   15-JUL-1999 (Rel. 38, Last annotation update)
DE   HEMOGLOBIN BETA CHAIN.
GN   HBB.
OS   Homo sapiens (Human), Pan troglodytes (Chimpanzee), and
OS   Pan paniscus (Pygmy chimpanzee) (Bonobo).
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Mammalia;
OC   Eutheria; Primates; Catarrhini; Hominidae; Homo.
RN   [1]
RP   SEQUENCE.
RC   SPECIES=HUMAN;
RA   BRAUNITZER G., GEHRING-MULLER R., HILSCHMANN N., HILSE K., HOBOM G.,
RA   RUDLOFF V., WITTMANN-LIEBOLD B.;
RT   "The constitution of normal adult human haemoglobin.";
RL   Hoppe-Seyler's Z. Physiol. Chem. 325:283-286(1961).
RN   [2]
RP   SEQUENCE FROM N.A.
RC   SPECIES=HUMAN;
RX   MEDLINE; 81064667.
RA   LAWN R.M., EFSTRATIADIS A., O'CONNELL C., MANIATIS T.;
RT   "The nucleotide sequence of the human beta-globin gene.";
RL   Cell 21:647-651(1980).
RN   [3]
RP   SEQUENCE OF 121-146 FROM N.A.
RC   SPECIES=HUMAN;
RX   MEDLINE; 85205333.
RA   LANG K.M., SPRITZ R.A.;
RT   "Cloning specific complete polyadenylylated 3'-terminal cDNA
RT   segments.";
RL   Gene 33:191-196(1985).
RN   [4]
RP   X-RAY CRYSTALLOGRAPHY (2.5 ANGSTROMS) OF DEOXYHEMOGLOBIN.
RC   SPECIES=HUMAN;
RX   MEDLINE; 76027820.
RA   FERMI G.;
RT   "Three-dimensional fourier synthesis of human deoxyhaemoglobin at
RT   2.5-A resolution: refinement of the atomic model.";
RL   J. Mol. Biol. 97:237-256(1975).
RN   [5]
RP   SEQUENCE.
RC   SPECIES=P.TROGLODYTES;
RX   MEDLINE; 66071496.
RA   RIFKIN D.B., KONIGSBERG W.;
RT   "The characterization of the tryptic peptides from the hemoglobin of
RT   the chimpanzee (Pan troglodytes).";
RL   Biochim. Biophys. Acta 104:457-461(1965).
RN   [6]


<font color=red>  [Part of this file has been deleted for brevity]</font>

FT   VARIANT     140    140       A -&gt; T (IN ST JACQUES: O2 AFFINITY UP).
FT                                /FTId=VAR_003081.
FT   VARIANT     140    140       A -&gt; V (IN PUTTELANGE; POLYCYTHEMIA;
FT                                O2 AFFINITY UP).
FT                                /FTId=VAR_003082.
FT   VARIANT     141    141       L -&gt; R (IN OLMSTED; UNSTABLE).
FT                                /FTId=VAR_003083.
FT   VARIANT     142    142       A -&gt; D (IN OHIO; O2 AFFINITY UP).
FT                                /FTId=VAR_003084.
FT   VARIANT     143    143       H -&gt; D (IN RANCHO MIRAGE).
FT                                /FTId=VAR_003085.
FT   VARIANT     143    143       H -&gt; Q (IN LITTLE ROCK; O2 AFFINITY UP).
FT                                /FTId=VAR_003086.
FT   VARIANT     143    143       H -&gt; P (IN SYRACUSE; O2 AFFINITY UP).
FT                                /FTId=VAR_003087.
FT   VARIANT     143    143       H -&gt; R (IN ABRUZZO; O2 AFFINITY UP).
FT                                /FTId=VAR_003088.
FT   VARIANT     144    144       K -&gt; E (IN MITO; O2 AFFINITY UP).
FT                                /FTId=VAR_003089.
FT   VARIANT     145    145       Y -&gt; C (IN RAINIER; O2 AFFINITY UP).
FT                                /FTId=VAR_003090.
FT   VARIANT     145    145       Y -&gt; H (IN BETHESDA; O2 AFFINITY UP).
FT                                /FTId=VAR_003091.
FT   VARIANT     146    146       H -&gt; D (IN HIROSHIMA; O2 AFFINITY UP).
FT                                /FTId=VAR_003092.
FT   VARIANT     146    146       H -&gt; L (IN COWTOWN; O2 AFFINITY UP).
FT                                /FTId=VAR_003093.
FT   VARIANT     146    146       H -&gt; P (IN YORK; O2 AFFINITY UP).
FT                                /FTId=VAR_003094.
FT   VARIANT     146    146       H -&gt; Q (IN KODAIRA; O2 AFFINITY UP).
FT                                /FTId=VAR_003095.
FT   HELIX         5     15
FT   TURN         16     17
FT   HELIX        20     34
FT   HELIX        36     41
FT   HELIX        43     45
FT   HELIX        51     55
FT   TURN         56     56
FT   HELIX        58     75
FT   TURN         76     77
FT   HELIX        78     94
FT   TURN         95     96
FT   TURN        100    100
FT   HELIX       101    121
FT   HELIX       124    142
FT   TURN        143    144
SQ   SEQUENCE   146 AA;  15867 MW;  EC9744C9 CRC32;
     VHLTPEEKSA VTALWGKVNV DEVGGEALGR LLVVYPWTQR FFESFGDLST PDAVMGNPKV
     KAHGKKVLGA FSDGLAHLDN LKGTFATLSE LHCDKLHVDP ENFRLLGNVL VCVLAHHFGK
     EFTPPVQAAY QKVVAGVANA LAHKYH
//
</pre>
</td></tr></table><p>

<H2>
    Output file format
</H2>

An image is output to the specified graphics device.

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>Graphics File: pepwheel.ps</h3>
<p><img src="pepwheel.1.pepwheel.gif" alt="[pepwheel results]">

<H2>
    Data files
</H2>

None.


<H2>
    Notes
</H2>


The default settings for turn (5) and steps (18) apply to alpha
helices. For other possible secondary structures, see the following
table:

<p>

<pre>
helix       phi   psi  omega  res/turn  transl.  turns  steps
alpha       -57   -47    180       3.6     1.50      5     18
3-10        -49   -26    180       3.0     2.00      1      3
pi          -57   -70    180       4.4     1.15      5     22
PP I        -83   158      0       3.33    1.9       3     10
PP II       -78   149    180       3.0     3.12      1      3
PG II       -80   150    180       3.0     3.1       1      3
anti Beta  -139   135   -178       2.0     3.4       4      9
para Beta  -119   113    180       2.0     3.2       4      9

PP is polyproline
PG II is polyGlycine.
</pre>

<H2>
    References
</H2>


<ol>

<li>
Rachamandran and Sasisekharan, Adv. Protein Chem.
23:283-437, 1968

<li>
IUPAC-IUB Commission on biochemical nomenclature,
Biochemistry 9:3471-3479, 1970

</ol>

<H2>
    Warnings
</H2>


When you specify the number of turns and steps, be aware that you
should remove common factors from these two numbers.  For example, if
you (for some improbable reason) wished to plot a wheel with 8 turns
in 18 steps (2.25 residues per turn) you should remove the common
factor (i.e.  2) and specify 4 turns in 9 steps.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

0 upon successful completion.

<H2>
    Known bugs
</H2>

None.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr>
<td><a href="abiview.html">abiview</a></td>
<td>Reads ABI file and display the trace</td>
</tr>

<tr>
<td><a href="cirdna.html">cirdna</a></td>
<td>Draws circular maps of DNA constructs</td>
</tr>

<tr>
<td><a href="garnier.html">garnier</a></td>
<td>Predicts protein secondary structure</td>
</tr>

<tr>
<td><a href="helixturnhelix.html">helixturnhelix</a></td>
<td>Report nucleic acid binding motifs</td>
</tr>

<tr>
<td><a href="hmoment.html">hmoment</a></td>
<td>Hydrophobic moment calculation</td>
</tr>

<tr>
<td><a href="lindna.html">lindna</a></td>
<td>Draws linear maps of DNA constructs</td>
</tr>

<tr>
<td><a href="pepcoil.html">pepcoil</a></td>
<td>Predicts coiled coil regions</td>
</tr>

<tr>
<td><a href="pepnet.html">pepnet</a></td>
<td>Displays proteins as a helical net</td>
</tr>

<tr>
<td><a href="prettyplot.html">prettyplot</a></td>
<td>Displays aligned sequences, with colouring and boxing</td>
</tr>

<tr>
<td><a href="prettyseq.html">prettyseq</a></td>
<td>Output sequence with translated ranges</td>
</tr>

<tr>
<td><a href="remap.html">remap</a></td>
<td>Display sequence with restriction sites, translation etc</td>
</tr>

<tr>
<td><a href="seealso.html">seealso</a></td>
<td>Finds programs sharing group names</td>
</tr>

<tr>
<td><a href="showalign.html">showalign</a></td>
<td>Displays a multiple sequence alignment</td>
</tr>

<tr>
<td><a href="showdb.html">showdb</a></td>
<td>Displays information on the currently available databases</td>
</tr>

<tr>
<td><a href="showfeat.html">showfeat</a></td>
<td>Show features of a sequence</td>
</tr>

<tr>
<td><a href="showseq.html">showseq</a></td>
<td>Display a sequence with features, translation etc</td>
</tr>

<tr>
<td><a href="sixpack.html">sixpack</a></td>
<td>Display a DNA sequence with 6-frame translation and ORFs</td>
</tr>

<tr>
<td><a href="textsearch.html">textsearch</a></td>
<td>Search sequence documentation. Slow, use SRS and Entrez!</td>
</tr>

<tr>
<td><a href="tmap.html">tmap</a></td>
<td>Displays membrane spanning regions</td>
</tr>

</table>


<H2>
    Author(s)
</H2>


Alan Bleasby (ajb&nbsp;&copy;&nbsp;ebi.ac.uk)
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK



<H2>
    History
</H2>

Written (March 1999) - Alan Bleasby

<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.



<H2>
    Comments
</H2>
None


</BODY>
</HTML>

