<HTML>

<HEAD>
  <TITLE>
  EMBOSS: pepwindow
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">





<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

pepwindow
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>





<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Displays protein hydropathy


<H2>
    Description
</H2>

<b>pepwindow</b> reads in a protein sequence and displays a graph of the
classic Kyte &amp; Doolittle hydropathy plot of that protein. 




<H2>
    Usage
</H2>


<b>Here is a sample session with pepwindow</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>pepwindow tsw:hba_human </b>
Displays protein hydropathy
Graph type [x11]: <b>cps</b>

Created pepwindow.ps

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-sequence]          sequence   Sequence USA
   -graph              xygraph    Graph type

   Additional (Optional) qualifiers:
   -datafile           datafile   Aaindex database file
   -length             integer    Window size

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1             integer    Start of the sequence to be used
   -send1               integer    End of the sequence to be used
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name

   "-graph" associated qualifiers
   -gprompt             boolean    Graph prompting
   -gtitle              string     Graph title
   -gsubtitle           string     Graph subtitle
   -gxtitle             string     Graph x axis title
   -gytitle             string     Graph y axis title
   -goutfile            string     Output file for non interactive displays
   -gdirectory          string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-graph</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including postscript, ps, hpgl, hp7470, hp7580, meta, colourps, cps, xwindows, x11, tektronics, tekt, tek4107t, tek, none, null, text, data, xterm, png</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-datafile</td>
<td>Aaindex database file</td>
<td>Data file</td>
<td>Enakai.dat</td>
</tr>

<tr>
<td>-length</td>
<td>Window size</td>
<td>Integer from 1 to 200</td>
<td>7</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>


<b>pepwindow</b> reads any protein sequence USA.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tsw:hba_human' is a sequence entry in the example protein database 'tsw'
<p>
<p><h3>Database entry: tsw:hba_human</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   HBA_HUMAN      STANDARD;      PRT;   141 AA.
AC   P01922;
DT   21-JUL-1986 (Rel. 01, Created)
DT   21-JUL-1986 (Rel. 01, Last sequence update)
DT   15-JUL-1999 (Rel. 38, Last annotation update)
DE   HEMOGLOBIN ALPHA CHAIN.
GN   HBA1 AND HBA2.
OS   Homo sapiens (Human), Pan troglodytes (Chimpanzee), and
OS   Pan paniscus (Pygmy chimpanzee) (Bonobo).
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Mammalia;
OC   Eutheria; Primates; Catarrhini; Hominidae; Homo.
RN   [1]
RP   SEQUENCE FROM N.A. (ALPHA-1).
RX   MEDLINE; 81088339.
RA   MICHELSON A.M., ORKIN S.H.;
RT   "The 3' untranslated regions of the duplicated human alpha-globin
RT   genes are unexpectedly divergent.";
RL   Cell 22:371-377(1980).
RN   [2]
RP   SEQUENCE FROM N.A. (ALPHA-2).
RX   MEDLINE; 81175088.
RA   LIEBHABER S.A., GOOSSENS M.J., KAN Y.W.;
RT   "Cloning and complete nucleotide sequence of human 5'-alpha-globin
RT   gene.";
RL   Proc. Natl. Acad. Sci. U.S.A. 77:7054-7058(1980).
RN   [3]
RP   SEQUENCE FROM N.A. (ALPHA-2).
RX   MEDLINE; 80137531.
RA   WILSON J.T., WILSON L.B., REDDY V.B., CAVALLESCO C., GHOSH P.K.,
RA   DERIEL J.K., FORGET B.G., WEISSMAN S.M.;
RT   "Nucleotide sequence of the coding portion of human alpha globin
RT   messenger RNA.";
RL   J. Biol. Chem. 255:2807-2815(1980).
RN   [4]
RP   SEQUENCE FROM N.A. (ALPHA-1 AND ALPHA-2).
RA   FLINT J., HIGGS D.R.;
RL   Submitted (JAN-1997) to the EMBL/GenBank/DDBJ databases.
RN   [5]
RP   SEQUENCE.
RA   BRAUNITZER G., GEHRING-MULLER R., HILSCHMANN N., HILSE K., HOBOM G.,
RA   RUDLOFF V., WITTMANN-LIEBOLD B.;
RT   "The constitution of normal adult human haemoglobin.";
RL   Hoppe-Seyler's Z. Physiol. Chem. 325:283-286(1961).
RN   [6]
RP   SEQUENCE.
RA   HILL R.J., KONIGSBERG W.;
RT   "The structure of human hemoglobin: IV. The chymotryptic digestion of
RT   the alpha chain of human hemoglobin.";
RL   J. Biol. Chem. 237:3151-3156(1962).
RN   [7]


<font color=red>  [Part of this file has been deleted for brevity]</font>

FT                                /FTId=VAR_002841.
FT   VARIANT     130    130       A -&gt; D (IN YUDA; O2 AFFINITY DOWN).
FT                                /FTId=VAR_002842.
FT   VARIANT     131    131       S -&gt; P (IN QUESTEMBERT; HIGHLY UNSTABLE;
FT                                CAUSES ALPHA-THALASSEMIA).
FT                                /FTId=VAR_002843.
FT   VARIANT     133    133       S -&gt; R (IN VAL DE MARNE; O2 AFFINITY UP).
FT                                /FTId=VAR_002844.
FT   VARIANT     135    135       V -&gt; E (IN PAVIE).
FT                                /FTId=VAR_002845.
FT   VARIANT     136    136       L -&gt; M (IN CHICAGO).
FT                                /FTId=VAR_002846.
FT   VARIANT     136    136       L -&gt; P (IN BIBBA; UNSTABLE;
FT                                CAUSES ALPHA-THALASSEMIA).
FT                                /FTId=VAR_002847.
FT   VARIANT     138    138       S -&gt; P (IN ATTLEBORO; O2 AFFINITY UP).
FT                                /FTId=VAR_002848.
FT   VARIANT     139    139       K -&gt; E (IN HANAKAMI; O2 AFFINITY UP).
FT                                /FTId=VAR_002849.
FT   VARIANT     139    139       K -&gt; T (IN TOKONAME; O2 AFFINITY UP).
FT                                /FTId=VAR_002850.
FT   VARIANT     140    140       Y -&gt; H (IN ROUEN; O2 AFFINITY UP).
FT                                /FTId=VAR_002851.
FT   VARIANT     141    141       R -&gt; C (IN NUNOBIKI; O2 AFFINITY UP).
FT                                /FTId=VAR_002852.
FT   VARIANT     141    141       R -&gt; L (IN LEGNANO; O2 AFFINITY UP).
FT                                /FTId=VAR_002853.
FT   VARIANT     141    141       R -&gt; H (IN SURESNES; O2 AFFINITY UP).
FT                                /FTId=VAR_002854.
FT   VARIANT     141    141       R -&gt; P (IN SINGAPORE).
FT                                /FTId=VAR_002855.
FT   HELIX         4     35
FT   HELIX        37     42
FT   TURN         44     45
FT   TURN         50     51
FT   HELIX        53     71
FT   TURN         72     74
FT   HELIX        76     79
FT   TURN         80     80
FT   HELIX        81     89
FT   TURN         90     91
FT   TURN         95     95
FT   HELIX        96    112
FT   TURN        114    116
FT   HELIX       119    136
FT   TURN        137    139
SQ   SEQUENCE   141 AA;  15126 MW;  5EC7DB1E CRC32;
     VLSPADKTNV KAAWGKVGAH AGEYGAEALE RMFLSFPTTK TYFPHFDLSH GSAQVKGHGK
     KVADALTNAV AHVDDMPNAL SALSDLHAHK LRVDPVNFKL LSHCLLVTLA AHLPAEFTPA
     VHASLDKFLA SVSTVLTSKY R
//
</pre>
</td></tr></table><p>


<H2>
    Output file format
</H2>

An image is displayed on the specified graphics device.

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>Graphics File: pepwindow.ps</h3>
<p><img src="pepwindow.1.pepwindow.gif" alt="[pepwindow results]">


<H2>
    Data files
</H2>

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>


<p>

<b>pepwindow</b> reads the Kyte-Doolittle hydropathy data from the file
'Enakai.dat'

<p>

The EMBOSS data file 'Enakai.dat' contains :-
<p>
<pre>
D Hydropathy index (Kyte-Doolittle, 1982)
R 0807099
A Kyte, J. and Doolittle, R.F.
T A simple method for displaying the hydropathic character of a protein
J J. Mol. Biol. 157, 105-132 (1982)
C CHOC760103    0.964  JANJ780102    0.922  DESM900102    0.898
  EISD860103    0.897  CHOC760104    0.889  WOLR810101    0.885
  RADA880101    0.884  MANP780101    0.881  EISD840101    0.878
  PONP800103    0.870  NAKH920108    0.868  JANJ790101    0.867
  JANJ790102    0.866  PONP800102    0.861  MEIH800103    0.856
  PONP800101    0.851  PONP800108    0.850  WARP780101    0.845
  RADA880108    0.842  ROSG850102    0.841  DESM900101    0.837
  BIOV880101    0.829  RADA880107    0.828  LIFS790102    0.824
  KANM800104    0.824  CIDH920104    0.824  MIYS850101    0.821
  RADA880104    0.819  NAKH900111    0.817  NISK800101    0.812
  FAUJ830101    0.811  ARGP820103    0.806  NAKH920105    0.803
  ARGP820102    0.803  KRIW790101   -0.805  CHOC760102   -0.838
  GUYH850101   -0.843  RACS770102   -0.844  JANJ780103   -0.845
  ROSM880101   -0.845  PRAM900101   -0.850  JANJ780101   -0.852
  GRAR740102   -0.859  MEIH800102   -0.871  ROSM880102   -0.878
  OOBM770101   -0.899
I   A/L    R/K    N/M    D/F    C/P    Q/S    E/T    G/W    H/Y    I/V
    1.8   -4.5   -3.5   -3.5    2.5   -3.5   -3.5   -0.4   -3.2    4.5
    3.8   -3.9    1.9    2.8   -1.6   -0.8   -0.7   -0.9   -1.3    4.2
//
</pre>

<p>

<b>pepwindow</b> can use any of the "Nakai et al." database of amino 
acid parameters - these used to be in a database called "NAKAI" but are 
now in one called "AAINDEX". EMBOSS has a program <b>aaindexextract</b> that 
takes data from this database and makes it available for <b>pepwindow</b>.

<p>

1. FTP the AAINDEX database from Japan:

<br>

<A HREF="ftp://ftp.genome.ad.jp/pub/db/genomenet/aaindex/aaindex1">ftp://ftp.genome.ad.jp/pub/db/genomenet/aaindex/aaindex1</A>

<p>

2. Run <b>aaindexextract</b> with the aaindex1 file as input (or ask whoever 
installs EMBOSS to run it)

<p>

3.  Run <b>pepwindow</b> with -datafile specifying the name of whatever
"AAINDEX" datafile you wish to use.  (Use <b>embossdata -showall</b> to
see your available "AAINDEX" data file names.)

<p>




<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

<pre>
Kyte, J. and Doolittle, R.F.
A simple method for displaying the hydropathic character of a protein
J. Mol. Biol. 157, 105-132 (1982)

</pre> 
<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

0 upon successful completion.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="backtranseq.html">backtranseq</a></td><td>Back translate a protein sequence</td></tr>
<tr><td><a href="charge.html">charge</a></td><td>Protein charge plot</td></tr>
<tr><td><a href="checktrans.html">checktrans</a></td><td>Reports STOP codons and ORF statistics of a protein</td></tr>
<tr><td><a href="compseq.html">compseq</a></td><td>Counts the composition of dimer/trimer/etc words in a sequence</td></tr>
<tr><td><a href="emowse.html">emowse</a></td><td>Protein identification by mass spectrometry</td></tr>
<tr><td><a href="freak.html">freak</a></td><td>Residue/base frequency table or plot</td></tr>
<tr><td><a href="iep.html">iep</a></td><td>Calculates the isoelectric point of a protein</td></tr>
<tr><td><a href="mwcontam.html">mwcontam</a></td><td>Shows molwts that match across a set of files</td></tr>
<tr><td><a href="mwfilter.html">mwfilter</a></td><td>Filter noisy molwts from mass spec output</td></tr>
<tr><td><a href="octanol.html">octanol</a></td><td>Displays protein hydropathy</td></tr>
<tr><td><a href="pepinfo.html">pepinfo</a></td><td>Plots simple amino acid properties in parallel</td></tr>
<tr><td><a href="pepstats.html">pepstats</a></td><td>Protein statistics</td></tr>
<tr><td><a href="pepwindowall.html">pepwindowall</a></td><td>Displays protein hydropathy of a set of sequences</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->

<p>

<b>pepwindow</b> can use any of the "Nakai et al." database of amino
acid parameters - these are in a database called "AAINDEX".  The program
<b>aaindexextract</b> takes data from this database and makes it
available for <b>pepwindow</b>. 
 

<H2>
    Author(s)
</H2>

Ian Longden (il&nbsp;&copy;&nbsp;sanger.ac.uk)
<br>
Sanger Institute, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.                      

<p>


Based on the original program by Jack Kyte and Russell F. Doolittle.


<H2>
    History
</H2>

<pre>
Completed 27th May 1999.
</pre>

<!--
<p>
[an error occurred while processing this directive]
-->

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

