<HTML>

<HEAD>
  <TITLE>
  EMBOSS: pepwheel
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: pepwheel
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program pepwheel
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Shows protein sequences as helices


<H2>
    Description
</H2>


<p>pepwheel</p> displays peptide sequences in a helical representation.

<p>

This gives a view of a helix from a protein sequence looking down the
axis of the helix. 

<p>

It is useful for highlighting amphipathicity and other properties of
residues around a helix. 

<H2>
    Usage
</H2>


Here is a sample session with pepwheel.

<p>

<pre>
% pepwheel sw:hbb_human -send 30
Shows protein sequences as helices
Graph type [x11]: 
</pre>

<p>

<a href="./pepwheel0.gif">click here for result</a>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers (* if not always prompted):
  [-sequence]          sequence   Sequence USA
*  -graph              graph      Graph type
*  -outfile            outfile    Output file name

   Optional qualifiers (* if not always prompted):
   -steps              integer    The number of residues plotted per turn is
                                  this value divided by the 'turns' value.
   -turns              integer    The number of residues plotted per turn is
                                  the 'steps' value divided by this value.
*  -squares            string     By default the aliphatic residues ILVM are
                                  marked with squares.
*  -diamonds           string     By default the residues DENQST are marked
                                  with diamonds.
*  -octags             string     By default the positively charged residues
                                  HKR are marked with octagons.

   Advanced qualifiers:
   -amphipathic        bool       If this is true then the residues ACFGILMVWY
                                  are marked as squares and all other
                                  residues are unmarked. This overrides any
                                  other markup that you may have specified
                                  using the qualifiers '-squares', '-diamonds'
                                  and '-octags'.
   -[no]wheel          bool       Plot the wheel
   -data               bool       Output the match data to a file instead of
                                  plotting it


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-graph</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including postscript, ps, hpgl, hp7470, hp7580, meta, colourps, cps, xwindows, x11, tektronics, tekt, tek4107t, tek, none, null, text, data, xterm</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr>
<td>-outfile</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.pepwheel</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-steps</td>
<td>The number of residues plotted per turn is this value divided by the 'turns' value.</td>
<td>Integer from 2 to 100</td>
<td>18</td>
</tr>

<tr>
<td>-turns</td>
<td>The number of residues plotted per turn is the 'steps' value divided by this value.</td>
<td>Integer from 1 to 100</td>
<td>5</td>
</tr>

<tr>
<td>-squares</td>
<td>By default the aliphatic residues ILVM are marked with squares.</td>
<td>Any string is accepted</td>
<td>ILVM</td>
</tr>

<tr>
<td>-diamonds</td>
<td>By default the residues DENQST are marked with diamonds.</td>
<td>Any string is accepted</td>
<td>DENQST</td>
</tr>

<tr>
<td>-octags</td>
<td>By default the positively charged residues HKR are marked with octagons.</td>
<td>Any string is accepted</td>
<td>HKR</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-amphipathic</td>
<td>If this is true then the residues ACFGILMVWY are marked as squares and all other residues are unmarked. This overrides any other markup that you may have specified using the qualifiers '-squares', '-diamonds' and '-octags'.</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]wheel</td>
<td>Plot the wheel</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-data</td>
<td>Output the match data to a file instead of plotting it</td>
<td>Yes/No</td>
<td>No</td>
</tr>

</table>



<H2>
    Input file format
</H2>

Any protein sequence USA.

<H2>
    Output file format
</H2>

An image is output.


<H2>
    Data files
</H2>

None.


<H2>
    Notes
</H2>


The default settings for turn (5) and steps (18) apply to alpha
helices. For other possible secondary structures, see the following
table:

<p>

<pre>
helix       phi   psi  omega  res/turn  transl.  turns  steps
alpha       -57   -47    180       3.6     1.50      5     18
3-10        -49   -26    180       3.0     2.00      1      3
pi          -57   -70    180       4.4     1.15      5     22
PP I        -83   158      0       3.33    1.9       3     10
PP II       -78   149    180       3.0     3.12      1      3
PG II       -80   150    180       3.0     3.1       1      3
anti Beta  -139   135   -178       2.0     3.4       4      9
para Beta  -119   113    180       2.0     3.2       4      9

PP is polyproline
PG II is polyGlycine.
</pre>

<H2>
    References
</H2>


<ol>

<li>
Rachamandran and Sasisekharan, Adv. Protein Chem.
23:283-437, 1968

<li>
IUPAC-IUB Commission on biochemical nomenclature,
Biochemistry 9:3471-3479, 1970

</ol>

<H2>
    Warnings
</H2>


When you specify the number of turns and steps, be aware that you
should remove common factors from these two numbers.  For example, if
you (for some improbable reason) wished to plot a wheel with 8 turns
in 18 steps (2.25 residues per turn) you should remove the common
factor (i.e.  2) and specify 4 turns in 9 steps.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

0 upon successful completion.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="charge.html">charge</a></td><td>Protein charge plot</td></tr>
<tr><td><a href="checktrans.html">checktrans</a></td><td>Reports STOP codons and ORF statistics of a protein sequence</td></tr>
<tr><td><a href="digest.html">digest</a></td><td>Protein proteolytic enzyme or reagent cleavage digest</td></tr>
<tr><td><a href="garnier.html">garnier</a></td><td>Predicts protein secondary structure</td></tr>
<tr><td><a href="helixturnhelix.html">helixturnhelix</a></td><td>Report nucleic acid binding motifs</td></tr>
<tr><td><a href="hmoment.html">hmoment</a></td><td>Hydrophobic moment calculation</td></tr>
<tr><td><a href="iep.html">iep</a></td><td>Calculates the isoelectric point of a protein</td></tr>
<tr><td><a href="octanol.html">octanol</a></td><td>Displays protein hydropathy</td></tr>
<tr><td><a href="pepcoil.html">pepcoil</a></td><td>Predicts coiled coil regions</td></tr>
<tr><td><a href="pepinfo.html">pepinfo</a></td><td>Plots simple amino acid properties in parallel</td></tr>
<tr><td><a href="pepnet.html">pepnet</a></td><td>Displays proteins as a helical net</td></tr>
<tr><td><a href="pepstats.html">pepstats</a></td><td>Protein statistics</td></tr>
<tr><td><a href="pepwindow.html">pepwindow</a></td><td>Displays protein hydropathy</td></tr>
<tr><td><a href="pepwindowall.html">pepwindowall</a></td><td>Displays protein hydropathy of a set of sequences</td></tr>
<tr><td><a href="sigcleave.html">sigcleave</a></td><td>Reports protein signal cleavage sites</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->



<H2>
    Author(s)
</H2>


This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>

Writte (March 1999) - Alan Bleasby

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

