<HTML>

<HEAD>
  <TITLE>
  EMBOSS: plotorf
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">





<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

plotorf
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>





<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Plot potential open reading frames


<H2>
    Description
</H2>


Plot potential open reading frames.

<p>

A graphical representation of where the open reading frames are in all 6
reading frames is shown.  The ORFs are displayed as blue boxes.  

<p>

ORFs in this program are defined as being regions between START and STOP
codons. 

<p>

Note that this definition of an ORF would miss those exons in eukaryotic
genomic sequences which do not contain a START codon.  <b>plotorf</b> is
only really useful when dealing with prokaryotic or mRNA eukaryotic
sequences. 

<p>

The default START codon is: "ATG".
<br>
The default STOP codons are: "TAA,TAG,TGA".

<p>

You can specify your own set of start and stop codons using the
<b>-start</b> and <b>-stop</b> qualifiers. 


<H2>
    Usage
</H2>


Here is a sample session with plotorf.

<p>

<pre>
% plotorf
Plot potential open reading frames
Input sequence: embl:paamir
Graph type [x11]: 
</pre>

<p>

<a href="plotorf0.gif">Click here to see the graph</a>

<p>

An example of specifying your own START and STOP codons with a
mitochondrial sequence would be:

<p>

<pre>
% plotorf -start ATT,ATC,ATA,ATG,GTG -stop TAA,TAG,AGA,AGG 
Plot potential open reading frames
Input sequence: mito.seq
Graph type [x11]: 
</pre>

<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          sequence   Sequence USA
   -graph              xygraph    Graph type

   Optional qualifiers: (none)
   Advanced qualifiers:
   -start              string     Start codons
   -stop               string     Stop codons

   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-graph</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including postscript, ps, hpgl, hp7470, hp7580, meta, colourps, cps, xwindows, x11, tektronics, tekt, tek4107t, tek, none, null, text, data, xterm, png</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-start</td>
<td>Start codons</td>
<td>Any string is accepted</td>
<td>ATG</td>
</tr>

<tr>
<td>-stop</td>
<td>Stop codons</td>
<td>Any string is accepted</td>
<td>TAA,TAG,TGA</td>
</tr>

</table>



<H2>
    Input file format
</H2>

Any nucleic acid sequence.

<H2>
    Output file format
</H2>

Graphical output.

<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

ORFs in this program are defined as being regions between START and STOP
codons. 

<p>

Note that this definition of an ORF would miss those exons in eukaryotic
genomic sequences which do not contain a START codon.  <b>plotorf</b> is
only really useful when dealing with prokaryotic or mRNA eukaryotic
sequences. 


<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

See the Notes section.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="backtranseq.html">backtranseq</a></td><td>Back translate a protein sequence</td></tr>
<tr><td><a href="coderet.html">coderet</a></td><td>Extract CDS, mRNA and translations from feature tables</td></tr>
<tr><td><a href="getorf.html">getorf</a></td><td>Finds and extracts open reading frames (ORFs)</td></tr>
<tr><td><a href="marscan.html">marscan</a></td><td>Finds MAR/SAR sites in nucleic sequences</td></tr>
<tr><td><a href="prettyseq.html">prettyseq</a></td><td>Output sequence with translated ranges</td></tr>
<tr><td><a href="remap.html">remap</a></td><td>Display a sequence with restriction cut sites, translation etc</td></tr>
<tr><td><a href="showorf.html">showorf</a></td><td>Pretty output of DNA translations</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="transeq.html">transeq</a></td><td>Translate nucleic acid sequences</td></tr>
<tr><td><a href="wobble.html">wobble</a></td><td>Wobble base plot</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>

Written (1999) - Alan Bleasby

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

