<HTML>

<HEAD>
  <TITLE>
  EMBOSS: preg
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">





<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

preg
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>





<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Regular expression search of a protein sequence


<H2>
    Description
</H2>


This searches for matches of a regular expression to a protein sequence.

<p>
A regular expression is a way of specifying an ambiguous pattern to
search for.  Regular expressions are commonly used in some computer
programming languages and may be more familiar to some users than to
others. 

<p>
The following is a short guide to regular expressions in EMBOSS:


<dl>
<dt>^</dt><dd>
use this at the start of a pattern to insist that the pattern can only
match at the start of a sequence.  (eg.  '^M' matches a methionine at
the start of the sequence)


<dt>$</dt><dd>
use this at the end of a pattern to insist that the pattern can only
match at the end of a sequence (eg.  'R$' matches an arginine at
the end of the sequence)

<dt>()</dt><dd>
groups a pattern.  This is commonly used with '|' (eg.  '(ACD)|(VWY)'
matches either the first 'ACD' or the second 'VWY' pattern )

<dt>|</dt><dd>
This is the OR operator to enable a match to be made to either one
pattern OR another.  There is no AND operator in this version of regular
expressions. 
</dl>

<p>

The following quantifier characters specify the number of time that
the character before (in this case 'x') matches:

<dl>
<dt>x?</dt><dd>
matches 0 or 1 times (ie, '' or 'x')
<dt>x*</dt><dd>
matches 0 or more times (ie, '' or 'x' or 'xx' or 'xxx', etc)
<dt>x+</dt><dd>
matches 1 or more times (ie, 'x' or 'xx' or 'xxx', etc)

</dl>

<p>
Quantifiers can follow any of the following types of character specification:

<dl>
<dt>x</dt><dd>
any character (ie 'A')

<dt>\x</dt><dd>
the character after the backslash is used instead of its normal
regular expression meaning.  This is commonly used to turn off the
special meaning of the characters '^$()|?*+[]-.'.  It may be especially
useful when searching for gap characters in a sequence (eg '\.' matches
only a dot character '.')

<dt>[xy]</dt><dd>
match one of the characters 'x' or 'y'.  You may have one or more
characters in this set. 

<dt>[x-z]</dt><dd>
match any one of the set of characters starting with 'x' and
ending in 'y' in ASCII order (eg '[A-G]' matches any one of: 'A', 'B',
'C', 'D', 'E', 'F', 'G')

<dt>[^x-z]</dt><dd>
matches anything except any one of the group of characters in
ASCII order (eg '[^A-G]' matches anything EXCEPT any one of: 'A', 'B',
'C', 'D', 'E', 'F', 'G')

<dt>.</dt><dd>
the dot character matches any other character (eg: 'A.G' matches
'AAG', 'AaG', 'AZG', 'A-G' 'A G', etc.)
</dl>

<p>
Combining some of these features gives these examples from the PROSITE
patterns database: 

<pre>
'[STAGCN][RKH][LIVMAFY]$'
</pre>

<p>
which is the 'Microbodies C-terminal targeting
signal'. 

<pre>
'LP.TG[STGAVDE]'
</pre>

<p>
which is the 'Gram-positive cocci surface proteins
anchoring hexapeptide'. 

<p>
Regular expressions are case-sensitive.
The pattern 'AAAA' will not match the sequence 'aaaa'.

<H2>
    Usage
</H2>


Here is a sample session with preg.

<p>

<pre>
% preg
regular expression search of a protein sequence
Input sequence(s): sw:*_rat
Output file [100k_rat.preg]: stdout
Regular expression pattern: IA[QWF]A
</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
  [-pattern]           regexp     Regular expression pattern
  [-outfile]           outfile    Output file name

   Optional qualifiers: (none)
   Advanced qualifiers: (none)
   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-pattern]<br>(Parameter 2)</td>
<td>Regular expression pattern</td>
<td>Any regular epression pattern is accepted</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 3)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.preg</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>


Any protein sequence.

<H2>
    Output file format
</H2>


Here is the output from the example run:

<hr>
<pre>
preg search of sw:*_rat with pattern IA[QWF]A
Matches in 100K_RAT
       100K_RAT   390 IAQA
Matches in 5H6_RAT
        5H6_RAT   289 IAQA
Matches in ACDS_RAT
       ACDS_RAT   282 IAQA
Matches in ANX2_RAT
       ANX2_RAT    70 IAFA
Matches in APB3_RAT
       APB3_RAT   336 IAQA
Matches in AQP9_RAT
       AQP9_RAT    44 IAQA
Matches in ATHA_RAT
       ATHA_RAT   122 IAFA
Matches in CD14_RAT
       CD14_RAT   178 IAQA
Matches in CIKE_RAT
       CIKE_RAT   231 IAFA
Matches in CLCB_RAT
       CLCB_RAT    90 IAQA
Matches in CTR1_RAT
       CTR1_RAT   590 IAFA
Matches in CYGF_RAT
       CYGF_RAT   359 IAQA
Matches in DPY2_RAT
       DPY2_RAT   264 IAQA
Matches in ENOB_RAT
       ENOB_RAT   327 IAQA
Matches in ERBP_RAT
       ERBP_RAT    40 IAFA
Matches in GLPK_RAT
       GLPK_RAT   392 IAFA
Matches in GPV_RAT
        GPV_RAT   529 IAQA
Matches in IRKB_RAT
       IRKB_RAT    93 IAFA
Matches in KGP2_RAT
       KGP2_RAT   477 IAFA
Matches in NPX1_RAT
       NPX1_RAT   407 IAWA
Matches in NTDO_RAT
       NTDO_RAT   160 IAWA
Matches in NTSE_RAT
       NTSE_RAT   180 IAWA
Matches in PAX8_RAT
       PAX8_RAT   188 IAQA
Matches in SRA4_RAT
       SRA4_RAT   491 IAWA
Matches in SYNP_RAT
       SYNP_RAT    43 IAFA
Matches in TGN3_RAT
       TGN3_RAT   330 IAFA
Matches in TGR3_RAT
       TGR3_RAT   792 IAFA
Matches in UDB2_RAT
       UDB2_RAT   325 IAWA
Matches in UDB3_RAT
       UDB3_RAT   325 IAWA
Matches in UDB6_RAT
       UDB6_RAT   325 IAWA
Matches in UDBC_RAT
       UDBC_RAT   325 IAWA
Matches in VMT2_RAT
       VMT2_RAT   462 IAFA
</pre>
<hr>

<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>


Regular expressions are case-sensitive.
The pattern 'AAAA' will not match the sequence 'aaaa'.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with a status of 0.
Always returns 0.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="antigenic.html">antigenic</a></td><td>Finds antigenic sites in proteins</td></tr>
<tr><td><a href="digest.html">digest</a></td><td>Protein proteolytic enzyme or reagent cleavage digest</td></tr>
<tr><td><a href="fuzzpro.html">fuzzpro</a></td><td>Protein pattern search</td></tr>
<tr><td><a href="fuzztran.html">fuzztran</a></td><td>Protein pattern search after translation</td></tr>
<tr><td><a href="helixturnhelix.html">helixturnhelix</a></td><td>Report nucleic acid binding motifs</td></tr>
<tr><td><a href="oddcomp.html">oddcomp</a></td><td>Finds protein sequence regions with a biased composition</td></tr>
<tr><td><a href="patmatdb.html">patmatdb</a></td><td>Search a protein sequence with a motif</td></tr>
<tr><td><a href="patmatmotifs.html">patmatmotifs</a></td><td>Search a PROSITE motif database with a protein sequence</td></tr>
<tr><td><a href="pepcoil.html">pepcoil</a></td><td>Predicts coiled coil regions</td></tr>
<tr><td><a href="pscan.html">pscan</a></td><td>Scans proteins using PRINTS</td></tr>
<tr><td><a href="sigcleave.html">sigcleave</a></td><td>Reports protein signal cleavage sites</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->



<p>

Other EMBOSS programs allow you to search for simple patterns and may be
easier for the user who has never used regular expressions before:

<p>
<ul>
<li><a href="fuzznuc.html">fuzznuc</a>  - Nucleic acid pattern search
<li><a href="fuzzpro.html">fuzzpro</a>  - Protein pattern search
<li><a href="fuzztran.html">fuzztran</a> - Protein pattern search after translation
</ul>

<H2>
    Author(s)
</H2>


This application was written by Peter Rice <a
href="mailto:pmr@sanger.ac.uk">(pmr@sanger.ac.uk)</a> Informatics
Division, The Sanger Centre, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.


<H2>
    History
</H2>

Written (1999) - Peter Rice

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

