<HTML>

<HEAD>
  <TITLE>
  EMBOSS: prettyplot
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">





<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

prettyplot
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>







<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Displays aligned sequences, with colouring and boxing


<H2>
    Description
</H2>


<b>prettyplot</b> reads in a set of aligned DNA or protein sequences. 
It displays them graphically, with conserved regions highlighted in
various ways. 

<H2>
    Usage
</H2>


<b>Here is a sample session with prettyplot</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>prettyplot -resbreak=10 -boxcol -consensus -plurality=3 </b>
Displays aligned sequences, with colouring and boxing
Input sequence set: <b>../../data/globins.msf</b>
Graph type [x11]: <b>cps</b>

Created prettyplot.ps

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>
<p>
<b>Example 2</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>prettyplot ../../data/globins.msf -plurality=3 -docolour </b>
Displays aligned sequences, with colouring and boxing
Graph type [x11]: <b>cps</b>

Created prettyplot.ps

</pre></td></tr></table><p>
<p>
<a href="#output.2">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers (* if not always prompted):
  [-msf]               seqset     File containing a sequence alignment
*  -graph              graph      Graph type

   Optional qualifiers:
   -residuesperline    integer    The number of residues to be displayed on
                                  each line
   -resbreak           integer    Residues before a space
   -[no]ccolours       boolean    Colour residues by their consensus value.
   -cidentity          string     Colour to display identical residues (RED)
   -csimilarity        string     Colour to display similar residues (GREEN)
   -cother             string     Colour to display other residues (BLACK)
   -docolour           boolean    Colour residues by table oily, amide etc.
   -[no]title          boolean    Do not display the title
   -shade              string     Set to BPLW for normal shading
                                  so for pair = 1.5,1.0,0.5 and shade = BPLW
                                  Residues score Colour
                                  1.5 or over....... BLACK (B)
                                  1.0 to 1.5 ....... BROWN (P)
                                  0.5 to 1.0 ....... WHEAT (L)
                                  under 0.5 ....... WHITE (W)
                                  The only four letters allowed are BPLW, in
                                  any order.
   -pair               string     Values to represent identical similar
                                  related
   -identity           integer    Only match those which are identical in all
                                  sequences.
   -[no]box            boolean    Display prettyboxes
   -boxcol             boolean    Colour the background in the boxes
   -boxcolval          string     Colour to be used for background. (GREY)
   -[no]name           boolean    Display the sequence names
   -maxnamelen         integer    Margin size for the sequence name.
   -[no]number         boolean    Display the residue number
   -[no]listoptions    boolean    Display the date and options used
   -plurality          float      Plurality check value (totweight/2)
   -consensus          boolean    Display the consensus
   -[no]collision      boolean    Allow collisions in calculating consensus
   -alternative        integer    Use alternative collisions routine
                                  0) Normal collision check. (default)
                                  1) checks identical scores with the max
                                  score found. So if any other residue matches
                                  the identical score then a collision has
                                  occurred.
                                  2) If another residue has a greater than or
                                  equal to matching score and these do not
                                  match then a collision has occurred.
                                  3) Checks all those not in the current
                                  consensus.If any of these give a top score
                                  for matching or identical scores then a
                                  collision has occured.
   -matrixfile         matrix     This is the scoring matrix file used when
                                  comparing sequences. By default it is the
                                  file 'EBLOSUM62' (for proteins) or the file
                                  'EDNAFULL' (for nucleic sequences). These
                                  files are found in the 'data' directory of
                                  the EMBOSS installation.
   -showscore          integer    Print residue scores
   -portrait           boolean    Set page to Portrait

   Advanced qualifiers:
   -data               boolean    (no help text) boolean value

   Associated qualifiers:
  "-msf" related qualifiers
   -sbegin1             integer    First base used
   -send1               integer    Last base used, def=seq length
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sopenfile1          string     Input filename
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name
  "-graph" related qualifiers
   -gprompt             boolean    Graph prompting
   -gtitle              string     Graph title
   -gsubtitle           string     Graph subtitle
   -gxtitle             string     Graph x axis title
   -gytitle             string     Graph y axis title
   -goutfile            string     Output file for non interactive displays
   -gdirectory          string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for required and optional values
   -debug               boolean    Write debug output to program.dbg
   -acdlog              boolean    Write ACD processing log to program.acdlog
   -acdpretty           boolean    Rewrite ACD file as program.acdpretty
   -acdtable            boolean    Write HTML table of options
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-msf]<br>(Parameter 1)</td>
<td>File containing a sequence alignment</td>
<td>Readable sequences</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-graph</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including postscript, ps, hpgl, hp7470, hp7580, meta, colourps, cps, xwindows, x11, tektronics, tekt, tek4107t, tek, none, null, text, data, xterm, png</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-residuesperline</td>
<td>The number of residues to be displayed on each line</td>
<td>Any integer value</td>
<td>50</td>
</tr>

<tr>
<td>-resbreak</td>
<td>Residues before a space</td>
<td>Integer 1 or more</td>
<td>Same as -residuesperline to give no breaks</td>
</tr>

<tr>
<td>-[no]ccolours</td>
<td>Colour residues by their consensus value.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-cidentity</td>
<td>Colour to display identical residues (RED)</td>
<td>Any string is accepted</td>
<td>RED</td>
</tr>

<tr>
<td>-csimilarity</td>
<td>Colour to display similar residues (GREEN)</td>
<td>Any string is accepted</td>
<td>GREEN</td>
</tr>

<tr>
<td>-cother</td>
<td>Colour to display other residues (BLACK)</td>
<td>Any string is accepted</td>
<td>BLACK</td>
</tr>

<tr>
<td>-docolour</td>
<td>Colour residues by table oily, amide etc.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]title</td>
<td>Do not display the title</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-shade</td>
<td>Set to BPLW for normal shading
so for pair = 1.5,1.0,0.5 and shade = BPLW
Residues score Colour
1.5 or over....... BLACK (B)
1.0 to 1.5 ....... BROWN (P)
0.5 to 1.0 ....... WHEAT (L)
under 0.5 ....... WHITE (W)
The only four letters allowed are BPLW, in any order.</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-pair</td>
<td>Values to represent identical similar related</td>
<td>Any string is accepted</td>
<td>1.5,1.0,0.5</td>
</tr>

<tr>
<td>-identity</td>
<td>Only match those which are identical in all sequences.</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr>
<td>-[no]box</td>
<td>Display prettyboxes</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-boxcol</td>
<td>Colour the background in the boxes</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-boxcolval</td>
<td>Colour to be used for background. (GREY)</td>
<td>Any string is accepted</td>
<td>GREY</td>
</tr>

<tr>
<td>-[no]name</td>
<td>Display the sequence names</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-maxnamelen</td>
<td>Margin size for the sequence name.</td>
<td>Any integer value</td>
<td>10</td>
</tr>

<tr>
<td>-[no]number</td>
<td>Display the residue number</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]listoptions</td>
<td>Display the date and options used</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-plurality</td>
<td>Plurality check value (totweight/2)</td>
<td>Any numeric value</td>
<td>Half the total sequence weighting</td>
</tr>

<tr>
<td>-consensus</td>
<td>Display the consensus</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]collision</td>
<td>Allow collisions in calculating consensus</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-alternative</td>
<td>Use alternative collisions routine
0) Normal collision check. (default)
1) checks identical scores with the max score found. So if any other residue matches the identical score then a collision has occurred.
2) If another residue has a greater than or equal to matching score and these do not match then a collision has occurred.
3) Checks all those not in the current consensus.If any of these give a top score for matching or identical scores then a collision has occured.</td>
<td>Integer from 0 to 3</td>
<td>0</td>
</tr>

<tr>
<td>-matrixfile</td>
<td>This is the scoring matrix file used when comparing sequences. By default it is the file 'EBLOSUM62' (for proteins) or the file 'EDNAFULL' (for nucleic sequences). These files are found in the 'data' directory of the EMBOSS installation.</td>
<td>Comparison matrix file in EMBOSS data path</td>
<td>EBLOSUM62 for protein<br>EDNAFULL for DNA</td>
</tr>

<tr>
<td>-showscore</td>
<td>Print residue scores</td>
<td>Any integer value</td>
<td>-1</td>
</tr>

<tr>
<td>-portrait</td>
<td>Set page to Portrait</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-data</td>
<td>(no help text) boolean value</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

</table>



<H2>
    Input file format
</H2>

<b>prettyplot</b> reads any sequence USA.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: ../../data/globins.msf</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
!!AA_MULTIPLE_ALIGNMENT 1.0

  ../data/globins.msf MSF:  164 Type: P 25/06/01 CompCheck: 4278 ..

  Name: HBB_HUMAN Len: 164  Check: 6914 Weight: 0.14
  Name: HBB_HORSE Len: 164  Check: 6007 Weight: 0.15
  Name: HBA_HUMAN Len: 164  Check: 3921 Weight: 0.15
  Name: HBA_HORSE Len: 164  Check: 4770 Weight: 0.19
  Name: MYG_PHYCA Len: 164  Check: 7930 Weight: 0.23
  Name: GLB5_PETMA Len: 164  Check: 1857 Weight: 0.21
  Name: LGB2_LUPLU Len: 164  Check: 2879 Weight: 0.10

//

           1                                               50
HBB_HUMAN  ~~~~~~~~VHLTPEEKSAVTALWGKVN.VDEVGGEALGR.LLVVYPWTQR
HBB_HORSE  ~~~~~~~~VQLSGEEKAAVLALWDKVN.EEEVGGEALGR.LLVVYPWTQR
HBA_HUMAN  ~~~~~~~~~~~~~~VLSPADKTNVKAA.WGKVGAHAGEYGAEALERMFLS
HBA_HORSE  ~~~~~~~~~~~~~~VLSAADKTNVKAA.WSKVGGHAGEYGAEALERMFLG
MYG_PHYCA  ~~~~~~~VLSEGEWQLVLHVWAKVEAD.VAGHGQDILIR.LFKSHPETLE
GLB5_PETMA PIVDTGSVAPLSAAEKTKIRSAWAPVYSTYETSGVDILVKFFTSTPAAQE
LGB2_LUPLU ~~~~~~~~GALTESQAALVKSSWEEFNANIPKHTHRFFILVLEIAPAAKD

           51                                             100
HBB_HUMAN  FFESFGDLSTPDAVMGNPKVKAHGKKVLGAFSDGLAHLDNLKGTFATLSE
HBB_HORSE  FFDSFGDLSNPGAVMGNPKVKAHGKKVLHSFGEGVHHLDNLKGTFAALSE
HBA_HUMAN  FPTTKTYFPHFDLSHGSAQVKGHGKKVADALTNAVAHVDDMPNALSALSD
HBA_HORSE  FPTTKTYFPHFDLSHGSAQVKAHGKKVGDALTLAVGHLDDLPGALSNLSD
MYG_PHYCA  KFDRFKHLKTEAEMKASEDLKKHGVTVLTALGAILKKKGHHEAELKPLAQ
GLB5_PETMA FFPKFKGLTTADQLKKSADVRWHAERIINAVNDAVASMDDTEKMSMKLRD
LGB2_LUPLU LFSFLKGTSEVPQNNPELQAHAGKVFKLVYEAAIQLQVTGVVVTDATLKN

           101                                            150
HBB_HUMAN  LHCDKLH..VDPENFRLLGNVLVCVLAHHFGKEFTPPVQAAYQKVVAGVA
HBB_HORSE  LHCDKLH..VDPENFRLLGNVLVVVLARHFGKDFTPELQASYQKVVAGVA
HBA_HUMAN  LHAHKLR..VDPVNFKLLSHCLLVTLAAHLPAEFTPAVHASLDKFLASVS
HBA_HORSE  LHAHKLR..VDPVNFKLLSHCLLSTLAVHLPNDFTPAVHASLDKFLSSVS
MYG_PHYCA  SHATKHK..IPIKYLEFISEAIIHVLHSRHPGDFGADAQGAMNKALELFR
GLB5_PETMA LSGKHAK..SFQVDPQYFKVLAAVIADTVAAGDAGFEKLMSMICILLRSA
LGB2_LUPLU LGSVHVSKGVADAHFPVVKEAILKTIKEVVGAKWSEELNSAWTIAYDELA

           151        164
HBB_HUMAN  NALAHKYH~~~~~~
HBB_HORSE  NALAHKYH~~~~~~
HBA_HUMAN  TVLTSKYR~~~~~~
HBA_HORSE  TVLTSKYR~~~~~~
MYG_PHYCA  KDIAAKYKELGYQG
GLB5_PETMA Y~~~~~~~~~~~~~
LGB2_LUPLU IVIKKEMNDAA~~~

</pre>
</td></tr></table><p>


<H2>
    Output file format
</H2>

An image of the alignment is displayed on the specified graphics device.

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>Graphics File: prettyplot.ps</h3>
<p><img src="prettyplot.1.prettyplot.gif" alt="[prettyplot results]">

<a name="output.2"></a>
<h3>Output files for usage example 2</h3>
<p><h3>Graphics File: prettyplot.ps</h3>
<p><img src="prettyplot.2.prettyplot.gif" alt="[prettyplot results]">


<H2>
    Data files
</H2>


Prettyplot uses a comparison matrix file to calculate similarity to
the consensus.

<p>
For protein sequences EBLOSUM62 is used for the substitution matrix.
For nucleotide sequence, EDNAFULL is used.

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by EMBOSS environment variable EMBOSS_DATA.

<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>


<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It exits with status 0 unless an error is reported.

<H2>
    Known bugs
</H2>


Portrait mode does not cover the whole page! This is a "feature" in
plplot.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="abiview.html">abiview</a></td><td>Reads ABI file and display the trace</td></tr>
<tr><td><a href="cirdna.html">cirdna</a></td><td>Draws circular maps of DNA constructs</td></tr>
<tr><td><a href="emma.html">emma</a></td><td>Multiple alignment program - interface to ClustalW program</td></tr>
<tr><td><a href="infoalign.html">infoalign</a></td><td>Information on a multiple sequence alignment</td></tr>
<tr><td><a href="lindna.html">lindna</a></td><td>Draws linear maps of DNA constructs</td></tr>
<tr><td><a href="pepnet.html">pepnet</a></td><td>Displays proteins as a helical net</td></tr>
<tr><td><a href="pepwheel.html">pepwheel</a></td><td>Shows protein sequences as helices</td></tr>
<tr><td><a href="plotcon.html">plotcon</a></td><td>Plots the quality of conservation of a sequence alignment</td></tr>
<tr><td><a href="prettyseq.html">prettyseq</a></td><td>Output sequence with translated ranges</td></tr>
<tr><td><a href="remap.html">remap</a></td><td>Display a sequence with restriction cut sites, translation etc</td></tr>
<tr><td><a href="seealso.html">seealso</a></td><td>Finds programs sharing group names</td></tr>
<tr><td><a href="showalign.html">showalign</a></td><td>Displays a multiple sequence alignment</td></tr>
<tr><td><a href="showdb.html">showdb</a></td><td>Displays information on the currently available databases</td></tr>
<tr><td><a href="showfeat.html">showfeat</a></td><td>Show features of a sequence</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="sixpack.html">sixpack</a></td><td>Display a DNA sequence with 6-frame translation and ORFs</td></tr>
<tr><td><a href="textsearch.html">textsearch</a></td><td>Search sequence documentation text. SRS and Entrez are faster!</td></tr>
<tr><td><a href="tranalign.html">tranalign</a></td><td>Align nucleic coding regions given the aligned proteins</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


Ian Longden (il&nbsp;&copy;&nbsp;sanger.ac.uk)
<br>
Sanger Institute, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.                      



<p>

Many features were first implemented in the EGCG program "prettyplot" by
Peter Rice (pmr&nbsp;&copy;&nbsp;ebi.ac.uk)
<br>
Informatics Division, European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK



<p> 

The original suggestions for the PrettyPlot program were from Denis
Duboule and Sigfried Labeit at EMBL. Gert Vriend added the star
marking. Rita Grandori suggested the -NOCOLLISION option. 

<H2>
    History
</H2>


Completed 5th May 1999.

<!--
<p>
[an error occurred while processing this directive]
-->

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

