<HTML>

<HEAD>
  <TITLE>
  EMBOSS: prettyplot
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: prettyplot
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program prettyplot
</H1>


<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Displays aligned sequences, with colouring and boxing


<H2>
    Description
</H2>


<b>prettyplot</b> reads in a set of aligned DNA or protein sequences. 
It displays them graphically, with conserved regions highlighted in
various ways. 

<H2>
    Usage
</H2>


Here is a sample session with prettyplot.

<p>

<pre>
% prettyplot -resbreak=10 -boxcol -consensus -plurality=3
Displays aligned sequences, with colouring and boxing
Input sequence set: globin.msf
Graph type [x11]:
</pre>
<a href="./prettyplot3.gif">click here for result</a>
<pre>
$ prettyplot globin.msf -plurality=3 -docolour
Displays aligned sequences, with colouring and boxing
Graph type [x11]: 
</pre>
<a href="./prettyplot4.gif">click here for result</a>


<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers (* if not always prompted):
  [-msf]               seqset     File containing a sequence alignment
*  -graph              graph      Graph type

   Optional qualifiers:
   -residuesperline    integer    The number of residues to be displayed on
                                  each line
   -resbreak           integer    Residues before a space
   -[no]ccolours       bool       Colour residues by their consensus value.
   -cidentity          string     Colour to display identical residues (RED)
   -csimilarity        string     Colour to display similar residues (GREEN)
   -cother             string     Colour to display other residues (BLACK)
   -docolour           bool       Colour residues by table oily, amide etc.
   -[no]title          bool       Do not display the title
   -shade              string     Set to BPLW for normal shading
                                  so for pair = 1.5,1.0,0.5 and shade = BPLW
                                  Residues score Colour
                                  1.5 or over....... BLACK (B)
                                  1.0 to 1.5 ....... BROWN (P)
                                  0.5 to 1.0 ....... WHEAT (L)
                                  under 0.5 ....... WHITE (W)
                                  The only four letters allowed are BPLW, in
                                  any order.
   -pair               string     Values to represent identical similar
                                  related
   -identity           integer    Only match those which are identical in all
                                  sequences.
   -[no]box            bool       Display prettyboxes
   -boxcol             bool       Colour the background in the boxes
   -boxcolval          string     Colour to be used for background. (GREY)
   -consensus          bool       Display the consensus
   -[no]name           bool       Display the sequence names
   -maxnamelen         integer    Margin size for the sequence name.
   -[no]number         bool       Display the residue number
   -[no]listoptions    bool       Display the date and options used
   -plurality          float      Plurality check value (totweight/2)
   -[no]collision      bool       Allow collisions in calculating consensus
   -alternative        integer    Use alternative collisions routine
                                  0) Normal collision check. (default)
                                  1) checks identical scores with the max
                                  score found. So if any other residue matches
                                  the identical score then a collision has
                                  occurred.
                                  2) If another residue has a greater than or
                                  equal to matching score and these do not
                                  match then a collision has occurred.
                                  3) Checks all those not in the current
                                  consensus.If any of these give a top score
                                  for matching or identical scores then a
                                  collision has occured.
   -portrait           bool       Set page to Portrait
   -matrixfile         matrix     Matrix file
   -showscore          integer    Print residue scores

   Advanced qualifiers:
   -data               bool       (no help text) bool value

   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-msf]<br>(Parameter 1)</td>
<td>File containing a sequence alignment</td>
<td>Readable sequences</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-graph</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including postscript, ps, hpgl, hp7470, hp7580, meta, colourps, cps, xwindows, x11, tektronics, tekt, tek4107t, tek, none, null, text, data, xterm, png</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-residuesperline</td>
<td>The number of residues to be displayed on each line</td>
<td>Any integer value</td>
<td>50</td>
</tr>

<tr>
<td>-resbreak</td>
<td>Residues before a space</td>
<td>Integer 1 or more</td>
<td>Same as -residuesperline to give no breaks</td>
</tr>

<tr>
<td>-[no]ccolours</td>
<td>Colour residues by their consensus value.</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-cidentity</td>
<td>Colour to display identical residues (RED)</td>
<td>Any string is accepted</td>
<td>RED</td>
</tr>

<tr>
<td>-csimilarity</td>
<td>Colour to display similar residues (GREEN)</td>
<td>Any string is accepted</td>
<td>GREEN</td>
</tr>

<tr>
<td>-cother</td>
<td>Colour to display other residues (BLACK)</td>
<td>Any string is accepted</td>
<td>BLACK</td>
</tr>

<tr>
<td>-docolour</td>
<td>Colour residues by table oily, amide etc.</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]title</td>
<td>Do not display the title</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-shade</td>
<td>Set to BPLW for normal shading
so for pair = 1.5,1.0,0.5 and shade = BPLW
Residues score Colour
1.5 or over....... BLACK (B)
1.0 to 1.5 ....... BROWN (P)
0.5 to 1.0 ....... WHEAT (L)
under 0.5 ....... WHITE (W)
The only four letters allowed are BPLW, in any order.</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-pair</td>
<td>Values to represent identical similar related</td>
<td>Any string is accepted</td>
<td>1.5,1.0,0.5</td>
</tr>

<tr>
<td>-identity</td>
<td>Only match those which are identical in all sequences.</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr>
<td>-[no]box</td>
<td>Display prettyboxes</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-boxcol</td>
<td>Colour the background in the boxes</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-boxcolval</td>
<td>Colour to be used for background. (GREY)</td>
<td>Any string is accepted</td>
<td>GREY</td>
</tr>

<tr>
<td>-consensus</td>
<td>Display the consensus</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]name</td>
<td>Display the sequence names</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-maxnamelen</td>
<td>Margin size for the sequence name.</td>
<td>Any integer value</td>
<td>10</td>
</tr>

<tr>
<td>-[no]number</td>
<td>Display the residue number</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]listoptions</td>
<td>Display the date and options used</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-plurality</td>
<td>Plurality check value (totweight/2)</td>
<td>Any integer value</td>
<td>Half the total sequence weighting</td>
</tr>

<tr>
<td>-[no]collision</td>
<td>Allow collisions in calculating consensus</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-alternative</td>
<td>Use alternative collisions routine
0) Normal collision check. (default)
1) checks identical scores with the max score found. So if any other residue matches the identical score then a collision has occurred.
2) If another residue has a greater than or equal to matching score and these do not match then a collision has occurred.
3) Checks all those not in the current consensus.If any of these give a top score for matching or identical scores then a collision has occured.</td>
<td>Integer from 0 to 3</td>
<td>0</td>
</tr>

<tr>
<td>-portrait</td>
<td>Set page to Portrait</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-matrixfile</td>
<td>Matrix file</td>
<td>Comparison matrix file in EMBOSS data path</td>
<td>EBLOSUM62 for protein<br>EDNAFULL for DNA</td>
</tr>

<tr>
<td>-showscore</td>
<td>Print residue scores</td>
<td>Any integer value</td>
<td>-1</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-data</td>
<td>(no help text) bool value</td>
<td>Yes/No</td>
<td>No</td>
</tr>

</table>



<H2>
    Input file format
</H2>

Any sequence USA.

<H2>
    Output file format
</H2>

An image of the alignment is displayed.

<H2>
    Data files
</H2>


Prettyplot uses a comparison matrix file to calculate similarity to
the consensus.

<p>
For protein sequences EBLOSUM62 is used for the substitution matrix.
For nucleotide sequence, EDNAFULL is used.

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by EMBOSS environment variable EMBOSS_DATA.

<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>


<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It exits with status 0 unless an error is reported.

<H2>
    Known bugs
</H2>


Portrait mode does not cover the whole page! This is a "feature" in
plplot.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="abiview.html">abiview</a></td><td>Reads ABI file and display the trace</td></tr>
<tr><td><a href="cirdna.html">cirdna</a></td><td>Draws circular maps of DNA constructs</td></tr>
<tr><td><a href="emma.html">emma</a></td><td>Multiple alignment program - interface to ClustalW program</td></tr>
<tr><td><a href="infoalign.html">infoalign</a></td><td>Information on a multiple sequence alignment</td></tr>
<tr><td><a href="lindna.html">lindna</a></td><td>Draws linear maps of DNA constructs</td></tr>
<tr><td><a href="pepnet.html">pepnet</a></td><td>Displays proteins as a helical net</td></tr>
<tr><td><a href="pepwheel.html">pepwheel</a></td><td>Shows protein sequences as helices</td></tr>
<tr><td><a href="plotcon.html">plotcon</a></td><td>Plots the quality of conservation of a sequence alignment</td></tr>
<tr><td><a href="prettyseq.html">prettyseq</a></td><td>Output sequence with translated ranges</td></tr>
<tr><td><a href="remap.html">remap</a></td><td>Display a sequence with restriction cut sites, translation etc</td></tr>
<tr><td><a href="seealso.html">seealso</a></td><td>Finds programs sharing group names</td></tr>
<tr><td><a href="showalign.html">showalign</a></td><td>Displays a multiple sequence alignment</td></tr>
<tr><td><a href="showdb.html">showdb</a></td><td>Displays information on the currently available databases</td></tr>
<tr><td><a href="showfeat.html">showfeat</a></td><td>Show features of a sequence</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="textsearch.html">textsearch</a></td><td>Search sequence documentation text. SRS and Entrez are faster!</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Ian Longden <a
href="mailto:il@sanger.ac.uk">(il@sanger.ac.uk)</a> Informatics
Division, The Sanger Centre, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.

<p>

Many features were first implemented in the EGCG program "prettyplot" by
Peter Rice. 

<p> 

The original suggestions for the PrettyPlot program were from Denis
Duboule and Sigfried Labeit at EMBL. Gert Vriend added the star
marking. Rita Grandori suggested the -NOCOLLISION option. 

<H2>
    History
</H2>


Completed 5th May 1999.

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

