<HTML>

<HEAD>
  <TITLE>
  EMBOSS: prima
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: prima
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program prima
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Selects primers for PCR and DNA amplification


<H2>
    Description
</H2>

<b>prima</b> analyzes a template DNA sequence and chooses primer pairs for the
polymerase chain reaction (PCR) and primers for DNA sequencing.  For PCR
primer pair selection, you can choose a target range of the template
sequence to be amplified.  For DNA sequencing primers, you can specify
positions on the template that must be included in the sequencing.  You
can allow <b>prima</b> to choose primers from the whole template or limit the
choices to a particular set of primers listed in a file. 
<p>

In selecting appropriate primers, <b>prima</b> considers a variety of
constraints on the primer and amplified product sequences.  You either
can use the program's default constraint values or modify those values
to customize the analysis.  You can specify upper and lower limits for
primer and product melting temperatures and for primer and product GC
contents.  For primers, you can specify a range of acceptable primer
sizes, any required bases at the 3' end of the primer (3' clamp), and a
maximum difference in primer melting temperatures for PCR primer pairs. 
For PCR products, you can specify a range of acceptable product sizes. 
<p>


For efficient priming, you should avoid primers with extensive
self-complementarity in order to minimize primer secondary structure and
primer dimer formation.  Additionally, in PCR experiments, primer pairs
with extensive complementarity between the two primers should be avoided
in order to minimize primer dimer formation.  <b>prima</b> uses the annealing
test described in the ALGORITHM topic to check individual primers for
self-complementarity and to check the two primers in a PCR primer pair
for complementarity to each other.  Using this same annealing test,
<b>prima</b> optionally can screen against non-specific primer binding on the
template sequence and on any repeated sequences you specify. 
<p>


The terms forward primer and reverse primer are used in the remainder of
this document and in the program output.  Forward primers are
complementary to sequences on the reverse template strand and create
copies of the forward strand by primer extension.  Conversely, reverse
primers are complementary to sequences on the forward template strand
and create copies of the reverse strand by primer extension. 
<p>


The name <b>prima</b> is a truly European one.  In searching for yet another
primer program name, we discovered that not only can "<b>prima</b>" be
pronounced in English as "Primer", but it has additional meanings in
other languages.  In Italian it means "first", in German it means
"super", and in Spanish it means "cousin". 
<p>


<H3>Thermodynamic Calculations</H3>

<b>prima</b> determines primer melting temperatures by a calculation using the
nearest-neighbor model of Borer, et al.  (J.  Mol.  Biol.  86; 843-853
(1974)) as modified slightly by Rychlik, et al.  (Nucleic Acids Res. 
18; 6409-6412 (1990)) and the thermodynamic parameters for DNA
nearest-neighbor interactions determined by Breslauer, et al.  (Proc. 
Natl.  Acad.  Sci.  USA.  83; 3746-3750 (1986)):
<p>

T(m)primer) = delta H / (delta S + R x ln(c/4) - 273.15 + 16.6 x log[K(+)]
<p>

where delta H is the enthalpy of helix formation, delta S is the entropy
of helix formation (including helix initiation), R is the molar gas
constant (1.987 cal/degree Celsius/mol), and c is the primer
concentration. 
<p>

<b>prima</b> determines PCR product melting temperatures using the formula of
Baldino, et al.  (in Methods Enzymol.  168; 761-777 (1989)) as modified
slightly by Rychlik, et al.  (Nucleic Acids Res.  18; 6409-6412 (1990)). 
<p>

T(m)product) = 0.41 x (% G+C) + 16.6 x log[K(+)] - 675 / len + 81.5
<p>

where len is the length of the product. 
<p>

If you are selecting PCR primer pairs, the output includes a proposed
annealing temperature for each listed primer pair.  The annealing
temperature is calculated using the formula of Rychlik, et al.  (Nucleic
Acids Res.  18; 6409-6412 (1990)). 
<p>

T(a) = 0.3 x T(m)primer) + 0.7 x T(m)product) - 14.9


<H3>Annealing Tests</H3>

<b>prima</b> uses an annealing test described by Hillier and Green (PCR
Methods and Applications.  1; 124-128 (1991)), with slight modification,
to check individual primers for self-complementarity and to check the
two primers in a PCR primer pair for complementarity to each other.  For
tests of self-complementarity, a primer sequence in the 5' to 3'
orientation is compared with the same sequence in the 3' to 5'
orientation.  For tests of complementarity between two different
primers, one of the primer sequences in the 5' to 3' orientation is
compared to the other sequence in the 3' to 5' orientation.  The
sequences are compared in every register of comparison, using a scoring
matrix containing values of complementarity for every pair of nucleotide
symbols.  For each register of comparison, the score of each base pair
comparison is determined.  The scores of contiguous base pairs with
positive comparison values are summed.  The maximum score of all such
contiguous segments, taken over all registers of comparison between the
sequences, determines the total primer-primer annealing score. 
Complementarity at the 3' ends of the primer sequences has a
particularly large influence on primer-dimer formation.  Therefore, the
maximum score of all contiguous segments that include the 3' position of
either primer sequence, taken over all registers of comparison, is
separately determined as the 3' primer-primer annealing score.  
<p>

The same annealing test is used to determine complementarity between the
primer and any non-specific binding sites on the template sequences.  In
this case, the primer in the 5' to 3' orientation is compared over all
registers of comparison with the template sequence in the 3' to 5'
orientation to determine a total primer-template annealing score.  Since
complementarity at the 3' end of the primer sequence has a particularly
large effect on non-specific primer binding, the 3' primer-template
annealing score is also determined.  If you screen against non-specific
primer binding on any specified repeated sequences, then total
primer-repeat and 3' primer-repeat annealing scores, taken over all
registers of comparison in all repeated sequences, are also determined. 
<p>

Total and 3' annealing scores are saved in tests of primer
self-complementarity (to check for secondary structure and primer dimer
formation) and in tests of complementarity between the two primers in
PCR primer pairs (to check for primer dimer formation).  Total and 3'
annealing scores are also saved when you screen against non-specific
primer binding on the template sequence and when you screen against
non-specific primer binding on any specified repeated sequences. 
Primers are rejected that exceed the maximum score you specify for any
of these tests.  For those primers that are accepted, the program uses
the sum of all annealing scores to determine the order of primers or PCR
primer pairs in the output list.  You can specify weights for each of
these scores to adjust their relative contributions in determining the
output order.  By default, 3' annealing scores have twice the weight of
total annealing scores in determining the output order. 
<p>

<H3>The Polymerase Chain Reaction</H3>

The Polymerase Chain Reaction (PCR) process for amplifying nucleic acids
is covered by U.S.  Patent Nos.  4,683,195 and 4,683,202 owned by
Hoffmann La Roche.  A license for research may be obtained through the
purchase and use of authorized reagents and thermocyclers from
Perkin-Elmer Corp., or by otherwise negotiating a license with
Perkin-Elmer.  No license to use PCR is granted by the purchase or use
of the EMBOSS package.
<p>


<H2>
    Usage
</H2>


Here is a sample session with <b>prima</b>.

<pre>

% prima
Input sequence: embl:eclaci
Specify a Target Range? [N]: 
Minimum Primer Tm (deg Celsius) [53]: 
Maximum Primer Tm (deg Celsius) [58]: 
Minimum product length [100]: 
Maximum product length [300]: 
Output file [eclaci.prima]: 

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers (* if not always prompted):
  [-sequence]          sequence   Sequence USA
  [-targetrange]       bool       Specify a Target Range?
*  -targetstart        integer    Target start position.
*  -targetend          integer    Target end position.
   -minprimertm        float      Minimum Primer Tm (deg Celsius)
   -maxprimertm        float      Maximum Primer Tm (deg Celsius)
*  -minprodlen         integer    Minimum product length
*  -maxprodlen         integer    Maximum product length
  [-outf]              outfile    Output file name

   Optional qualifiers: (none)
   Advanced qualifiers:
   -overlap            integer    Minimum overlap of sequences
   -minprimerlen       integer    Minimum primer length
   -maxprimerlen       integer    Minimum primer length
   -minpmgccont        float      Minimum primer GC fraction
   -maxpmgccont        float      Maximum primer GC fraction
   -minprodgccont      float      Minimum product GC fraction
   -maxprodgccont      float      Maximum product GC fraction
   -saltconc           float      Salt concentration (mM)
   -dnaconc            float      DNA concentration (mM)
   -list               bool       Force list-style output


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-targetrange]<br>(Parameter 2)</td>
<td>Specify a Target Range?</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-targetstart</td>
<td>Target start position.</td>
<td>Any integer value</td>
<td>Start of sequence</td>
</tr>

<tr>
<td>-targetend</td>
<td>Target end position.</td>
<td>Any integer value</td>
<td>End of sequence</td>
</tr>

<tr>
<td>-minprimertm</td>
<td>Minimum Primer Tm (deg Celsius)</td>
<td>Any integer value</td>
<td>53</td>
</tr>

<tr>
<td>-maxprimertm</td>
<td>Maximum Primer Tm (deg Celsius)</td>
<td>Any integer value</td>
<td>58</td>
</tr>

<tr>
<td>-minprodlen</td>
<td>Minimum product length</td>
<td>Any integer value</td>
<td>100</td>
</tr>

<tr>
<td>-maxprodlen</td>
<td>Maximum product length</td>
<td>Any integer value</td>
<td>300</td>
</tr>

<tr>
<td>[-outf]<br>(Parameter 3)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.prima</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-overlap</td>
<td>Minimum overlap of sequences</td>
<td>Any integer value</td>
<td>50</td>
</tr>

<tr>
<td>-minprimerlen</td>
<td>Minimum primer length</td>
<td>Any integer value</td>
<td>18</td>
</tr>

<tr>
<td>-maxprimerlen</td>
<td>Minimum primer length</td>
<td>Any integer value</td>
<td>22</td>
</tr>

<tr>
<td>-minpmgccont</td>
<td>Minimum primer GC fraction</td>
<td>Number from 0.300 to 0.700</td>
<td>.40</td>
</tr>

<tr>
<td>-maxpmgccont</td>
<td>Maximum primer GC fraction</td>
<td>Number from 0.300 to 0.700</td>
<td>.55</td>
</tr>

<tr>
<td>-minprodgccont</td>
<td>Minimum product GC fraction</td>
<td>Number from 0.300 to 0.700</td>
<td>.40</td>
</tr>

<tr>
<td>-maxprodgccont</td>
<td>Maximum product GC fraction</td>
<td>Number from 0.300 to 0.700</td>
<td>.55</td>
</tr>

<tr>
<td>-saltconc</td>
<td>Salt concentration (mM)</td>
<td>Number from 1.000 to 100.000</td>
<td>50</td>
</tr>

<tr>
<td>-dnaconc</td>
<td>DNA concentration (mM)</td>
<td>Number from 1.000 to 100.000</td>
<td>50</td>
</tr>

<tr>
<td>-list</td>
<td>Force list-style output</td>
<td>Yes/No</td>
<td>No</td>
</tr>

</table>



<H2>
    Input file format
</H2>

Nucleic acid sequence USA.

<H2>
    Output file format
</H2>


Here is the output from the example run:

<p>

<pre>



INPUT SUMMARY
*************

Prima of ECLACI
PRIMER CONSTRAINTS:
PRIMA DOES NOT ALLOW PRIMER SEQUENCE AMBIGUITY OR DUPLICATE PRIMER ENDPOINTS
Primer size range is 18-22
Primer GC content range is 0.40-0.55
Primer melting Temp range is 53.00 - 58.00 C
PRODUCT CONSTRAINTS:
Product GC content range is 0.40-0.55
Salt concentration is 50.00 (mM)
DNA concentration is 50.00 (nM)
Considering all suitable Primer pairs with Product length ranges 100 to 300




PRIMER/PRODUCT PAIR CALCULATIONS & OUTPUT
*****************************************

3 pairs found


                Forward                                 Reverse

[1]
    10 AGTCAATTCAGGGTGGTGAA      29        154 ATGTAATTCAGCTCCGCCAT      173
       Tm  56.24 C  (GC 50.00%)                Tm  57.14 C  (GC 50.00%)
             Length: 20                              Length: 20
             Tma:    40.56 C                         Tma:    40.83 C


       Product GC: 53.23%
       Product Tm: 55.12 C
       Length:     124


[2]
   266 TTGTCGCGGCGATTAAATCTC     286       510 GTACCGTCTTCATGGGAGAAA     530
       Tm  57.88 C  (GC 42.86%)                Tm  55.78 C  (GC 42.86%)
             Length: 21                              Length: 21
             Tma:    42.45 C                         Tma:    41.82 C


       Product GC: 54.71%
       Product Tm: 57.13 C
       Length:     223


[3]
   477 TGTCTCTGACCAGACACCC       495       728 GGAACGATGCCCTCATTCA       746
       Tm  54.99 C  (GC 52.63%)                Tm  55.20 C  (GC 52.63%)
             Length: 19                              Length: 19
             Tma:    42.28 C                         Tma:    42.34 C


       Product GC: 53.02%
       Product Tm: 58.12 C
       Length:     232


</pre>

The output file begins with a summary listing all of the constraints
used by the program to select appropriate primers or PCR primer pairs. 
Most of these constraints can be modified by adjusting prompted and
optional program parameters. 
<p>

Following these summaries is an ordered listing of the most appropriate
primers or PCR primer pairs selected by <b>prima</b>.  The list is ordered by
total annealing score so that those primers or PCR primer pairs with the
least amount of complementarity to sequences other than the appropriate
primer binding sites are listed first.  Each output primer or PCR primer
pair is designated by a number that corresponds to a line number in the
plot of primer sites.  While the text output file lists the location of
the primer binding site along with each primer sequence, the plot
provides a convenient way to review the primer binding sites of many of
the selected primers at once. 
<p>




<H2>
    Data files
</H2>


None.

<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>


Borer, P.N., Dengler, B., and Tinoco, I., Jr. (1974). "Stability of Ribonucleic Acid and
Double-stranded Helices." Journal of Molecular Biology 86, 843-853. 
<p>

Rychlik, W. and Rhoads, R.E. (1990). "Optimization of the Annealing Temperature for DNA
Amplification in vitro." Nucleic Acids Research 18, 6409-6412. 
<p>

Breslauer, K.J., Frank, R., Blocker, H., and Marky, L.A. (1986). "Predicting DNA Duplex Stability
from the Base Sequence." Proceedings of the National Academy of Sciences USA 83, 3746-3750.
<p>

Baldino, M., Jr. (1989). "High Resolution In Situ Hybridization Histochemistry." In Methods in
Enzymology, (P.M. Conn, ed.), 168, 761-777, Academic Press, San Diego, California, USA. 
<p>

Hillier, L. and Green, P. (1991). "OSP: A Computer Program for Choosing PCR and DNA
Sequencing Primers." PCR Methods and Applications 1, 124-128. 
<p>

Slightom et al.  (1994) "Nucleotide sequencing double-stranded plasmids
with primers selected from a nonamer library." Biotechniques 17(3),
536-7, 540-4. 
<p>

<H2>
    Warnings
</H2>


The template sequence may contain ambiguous bases, but <b>prima</b> will not
select primers complementary to any ambiguous sites on the template
sequence. 
<p>

When several acceptable PCR primer pairs have the same 3' ends for both
primers, <b>prima</b> outputs only the PCR primer pair with the shortest primer
sequences.  By not allowing duplicate primer endpoints, <b>prima</b> increases
the diversity among the PCR primer pairs in the output list. 
<p>

<b>prima</b> only determines melting temperatures for DNA primers.  We do not
know of any appropriate nearest-neighbor thermodynamic parameters for
RNA-DNA hybrids, so we haven't attempted to calculate melting
temperatures for RNA primers.  While thermodynamic parameters for RNA
duplexes involving mismatches have been described, we do not know of any
similar results for DNA duplexes.  Therefore, we have not attempted to
calculate melting temperatures or other thermodynamic properties for DNA
duplexes involving mismatches. 
<p>

<b>prima</b> does not currently consider formamide concentration in determining
primer melting temperatures. 

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="primersearch.html">primersearch</a></td><td>Searches DNA sequences for matches with primer pairs</td></tr>
<tr><td><a href="stssearch.html">stssearch</a></td><td>Searches a DNA database for matches with a set of STS primers</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Sinead O'Leary <a
href="mailto:soleary@hgmp.mrc.ac.uk">(soleary@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>

Finished 2000

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

