<!--
Edit this file to replace 'primersearch' with the name of the program.
-->

<HTML>

<HEAD>
  <TITLE>
  EMBOSS
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">





<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

primersearch
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>





<!--END OF HEADER-->





<H1>
Program primersearch
</H1>
<H2>
    Function
</H2>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Searches DNA sequences for matches with primer pairs


<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.

        This is what the user looks at to find out if this program will
        do what he wants. 

        Long description.

        Why it is useful.

        When it should/shouldn't be used.

        Algorithms

        Expanded description.

        eg: Reverse engineers Wheedle's 'talpsgob' program. This does a 
        splurge of the 5' end of the sequence followed by a durm.
        The net effect is a 5' end splurt.
        etc. for as long as is required (several pages is not unreasonable)
  -->

primersearch reads in primer pairs from an input file and searches
them against sequence(s) specified by the user. Each of the primers in
a pair is searched against the sequence and potential amplimers are
reported.
<p>

The user can specify a maximum percent mismatch level; for example,
10% mismatch on a primer of length 20bp means that the program will
classify a primer as matching a sequence if 18 of the 20 base pairs
matches.
<p>

It will only report matches if both primers in the pair have a match in
opposite orientations. 

<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->


<b>Here is a sample session with primersearch</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>primersearch tembl:z52466 </b>
Searches DNA sequences for matches with primer pairs
Primer file: <b>../../data/primers</b>
Allowed percent mismatch [0]: <b></b>
Output file [hsa203yc1.primersearch]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>
<p>
<b>Example 2</b>
<p>
Here we run the same example but allowing 20% mismatch between the primers and the sequence: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>primersearch tembl:z52466 </b>
Searches DNA sequences for matches with primer pairs
Primer file: <b>../../data/primers</b>
Allowed percent mismatch [0]: <b>20</b>
Output file [hsa203yc1.primersearch]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.2">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-seqall]            seqall     Sequence database USA
  [-primersfile]       infile     Primer file
  [-mismatchpercent]   integer    Allowed percent mismatch
  [-outfile]           outfile    Output file name

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-seqall" associated qualifiers
   -sbegin1             integer    First base used
   -send1               integer    Last base used, def=seq length
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name

   "-outfile" associated qualifiers
   -odirectory4         string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-seqall]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-primersfile]<br>(Parameter 2)</td>
<td>Primer file</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-mismatchpercent]<br>(Parameter 3)</td>
<td>Allowed percent mismatch</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 4)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.primersearch</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        Except for sequences which can be described by 'sequence USA',
        this should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>primersearch</b> reads in any nucleic acid sequence USAs.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:z52466' is a sequence entry in the example nucleic acid database 'tembl'
<p>
<p><h3>File: ../../data/primers</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
# This is my primer file
D1S243  cacacaggctcacatgcc      gctccagcgtcatggact
D1S468 aattaaccgttttggtcct     gcgacacacacttccc 
D1S2845 ccaaagggtgcttctc        gtggcattccaacctc
D1S1608 gatggcttttggggactatt    cactgagccaagtgacacag
D1S2893 aaaacatcaactctcccctg    ctcaaaccccaataagcctt
D1S2660 cacacatgcacatgcac       agtgacaccagcaggg
</pre>
</td></tr></table><p>
<p><h3>Database entry: tembl:z52466</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   HSA203YC1  standard; DNA; STS; 389 BP.
XX
AC   Z52466;
XX
SV   Z52466.1
XX
DT   18-MAR-1996 (Rel. 47, Created)
DT   23-MAR-1996 (Rel. 47, Last updated, Version 3)
XX
DE   H.sapiens (D1S2660) DNA segment containing (CA) repeat; clone AFMa203yc1;
DE   single read.
XX
KW   CA repeat; dinucleotide repeat; GT repeat; microsatellite DNA;
KW   microsatellite marker; repeat polymorphism; STS.
XX
OS   Homo sapiens (human)
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia;
OC   Eutheria; Primates; Catarrhini; Hominidae; Homo.
XX
RN   [1]
RP   1-389
RA   Weissenbach J.;
RT   ;
RL   Submitted (01-SEP-1995) to the EMBL/GenBank/DDBJ databases.
RL   Genethon, B.P. 60, 91002 Evry Cedex France. E-mail:
RL   Jean.Weissenbach@genethon.fr
XX
RN   [2]
RP   1-389
RX   MEDLINE; 96176476.
RA   Dib C., Faure S., Fizames C., Samson D., Drouot N., Vignal A.,
RA   Millasseau P., Marc S., Hazan J., Seboun E., Lathrop M., Gyapay G.,
RA   Morissette J., Weissenbach J.;
RT   "A comprehensive genetic map of the human genome based on 5,264
RT   microsatellites";
RL   Nature 380:152-154(1996).
XX
CC   full automatic;
XX
FH   Key             Location/Qualifiers
FH
FT   source          1..389
FT                   /chromosome="1"
FT                   /db_xref="taxon:9606"
FT                   /note="cloning vector is M13mp18"
FT                   /organism="Homo sapiens"
FT                   /cell_line="CEPH 134702"
FT                   /clone_lib="genomic DNA"
XX
SQ   Sequence 389 BP; 118 A; 124 C; 86 G; 57 T; 4 other;
     agctgtgtgc acacaacatg anggggcaca catgcacatg cacacatgcc cacatgcata        60
     tgcacacaca cacacacaca cacacacaca ttcatgccca agcacgccca ccctcatgtc       120
     tcaccatgtg cacataacac acagtcacat ataccctggc acacatgccc acatgcagac       180
     acgaaacaca ggcccacgnt tncatgcaca caggtatggg cacacatacc atgcacacat       240
     aangacaaat accaggccag acatgatttg cccctgctgg tgtcactgtt aagtgtgaca       300
     gacaagcaga ggacacacac ccacctggga cgcggggctt caggagagag gcagacctaa       360
     tagggcccgg attcggggct ggggaggct                                         389
//
</pre>
</td></tr></table><p>


<p>

The input primer file has the following format:

<p>
Comment lines start with a '#'
<br>
Lines with primer information have three fields separated by spaces or TAB characters. The columns contain:
<br>
<ol>
<li>The name of the primer pair - this is reported in the output.
<li>The sequence of the first primer.
<li>The sequence of the second primer.
</ol>

<p>

Empty files will cause primersearch to note that no primers have 
been found, and to exit.


<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats 
        Except for sequences which can be described by 'sequence USA',
        this should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: hsa203yc1.primersearch</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>

Primer name D1S243

Primer name D1S468

Primer name D1S2845

Primer name D1S1608

Primer name D1S2893

Primer name D1S2660
Amplimer 1
	Sequence: HSA203YC1 Z52466 
	H.sapiens (D1S2660) DNA segment containing (CA) repeat; clone AFMa203yc1; single read.
	CACACATGCACATGCAC hits forward strand at 27 with 0 mismatches
	AGTGACACCAGCAGGG hits reverse strand at [103] with 0 mismatches
	Amplimer length: 261 bp
</pre>
</td></tr></table><p>

<a name="output.2"></a>
<h3>Output files for usage example 2</h3>
<p><h3>File: hsa203yc1.primersearch</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>

Primer name D1S243

Primer name D1S468

Primer name D1S2845

Primer name D1S1608

Primer name D1S2893

Primer name D1S2660
Amplimer 1
	Sequence: HSA203YC1 Z52466 
	H.sapiens (D1S2660) DNA segment containing (CA) repeat; clone AFMa203yc1; single read.
	CACACATGCACATGCAC hits forward strand at 49 with 2 mismatches
	AGTGACACCAGCAGGG hits reverse strand at [103] with 0 mismatches
	Amplimer length: 239 bp
Amplimer 2
	Sequence: HSA203YC1 Z52466 
	H.sapiens (D1S2660) DNA segment containing (CA) repeat; clone AFMa203yc1; single read.
	CACACATGCACATGCAC hits forward strand at 27 with 0 mismatches
	AGTGACACCAGCAGGG hits reverse strand at [103] with 0 mismatches
	Amplimer length: 261 bp
</pre>
</td></tr></table><p>


<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.

        eg:
        data file format:

	blank lines and comment lines are ignored.
        comment lines start with '#'
        There are three columns separated by one or more space or TAB
	characters for X and Y values of durm, followed by energy
        all coefficients are given as floating point values
        An example is:

        # Durm coefficient enery file
        # X     Y       Energy
        0.1     0.2     0.331
        0.2     0.4     0.332
        0.3     0.6     0.333
        0.3     0.8     0.334

   -->

None.

<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

Every potential amplimer will be reported; if one primer matches the
forward strand twice and the other matches the reverse strand only
once, two potential amplimers are reported. If the reverse primer
matches twice, four potential amplimers are reported.
<p>

Each primer pair consists of a name, followed by two primer sequences,
primerA and primerB. The program first compares primerA to the forward
strand and if it matches, primerB is compared to the reverse
strand. The approach is then reversed, with the primerB being compared
to the forward strand and primerA to the reverse. In this way all
possible amplimers are reported.

<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.

<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
        eg:
        It doesn't check to see if the sequence is a protein sequence or not.
   -->

This program is slow.
<br>
Don't try searching all of EMBL!
<br>
It might not be sensible to search all of the human section of EMBL - in 2000, this would take a day to do with one primer pair.
<br>
It will take longer with more primer pairs.
<br>
It will take longer if you allow mismatches.


<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program.
        eg:

        "Threshold value greater than 1." - means that the threshold has
        been set to be greater than 1 but a burgle exception has
        occurred making this invalid - you should reduce the wobble
        parameter and try again. 

   --> 

"No suitable primers found - exiting" means that either the primers
file was empty or there were no compilable primer pairs contained in it.


<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
        eg:
        "It always exits with status 0"
   -->
It always exits with status 0

<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="eprimer3.html">eprimer3</a></td><td>Picks PCR primers and hybridization oligos</td></tr>
<tr><td><a href="stssearch.html">stssearch</a></td><td>Searches a DNA database for matches with a set of STS primers</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->


<a href="stssearch.html">stssearch</a> does something similar, but doesn't
allow you to find mismatches and will report any match in any
orientation and doesn't require you to have both primers matching. 


<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

Val Curwen (vac&nbsp;&copy;&nbsp;sanger.ac.uk)
<br>
Sanger Institute, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK. 



<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written Aug 2000 - Val Curwen

<!--
<p>
[an error occurred while processing this directive]
-->

<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:

   -->
This program is intended to be used by everyone and everything, from naive users to embedded scripts.



<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->



</BODY>
</HTML>

