<HTML>

<HEAD>
  <TITLE>
  EMBOSS: prophecy
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

prophecy
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<H2>
    Function
</H2>
Creates matrices/profiles from multiple alignments


<H2>
    Description
</H2>

<b>prophecy</b> produces a simple frequency matrix for use by
<b>profit</b> or a position specific weighted profile using either the
Gribskov (1) or Henikoff (2) method for use by <b>prophet</b>. 

<p>

Profile analysis is a method for detecting distantly related proteins by
sequence comparison.  The basis for comparison is not only the customary
Dayhoff mutational-distance matrix but also the results of structural
studies and information implicit in the alignments of the sequences of
families of similar proteins.  This information is expressed in a
position-specific scoring table (profile), which is created from a group
of sequences previously aligned by structural or sequence similarity. 
The similarity of any other target sequence to the group of aligned
probe sequences can be tested by comparing the target to the profile
using dynamic programming algorithms.  The profile method differs in two
major respects from methods of sequence comparison in common use: (i)
Any number of known sequences can be used to construct the profile,
allowing more information to be used in the testing of the target than
is possible with pairwise alignment methods.  (ii) The profile includes
the penalties for insertion or deletion at each position, which allow
one to include the probe secondary structure in the testing scheme. 


<h3>Algorithm</h3>

For Gribskov the scoring scheme is based on a notion of distance between
a sequence and an ancestral or generalized sequence.

<p>


For Henikoff it is based on weights of the diversity observed at each
position in the alignment, rather than on a sequence distance measure. 


<H2>
    Usage
</H2>
<b>Here is a sample session with prophecy</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>prophecy </b>
Creates matrices/profiles from multiple alignments
Input sequence set: <b>globins.msf</b>
Profile type
         F : Frequency
         G : Gribskov
         H : Henikoff
Select type [F]: <b></b>
Enter a name for the profile [mymatrix]: <b>globins</b>
Enter threshold reporting percentage [75]: <b></b>
Output file [globins.prophecy]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>
<p>
<b>Example 2</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>prophecy </b>
Creates matrices/profiles from multiple alignments
Input sequence set: <b>globins.msf</b>
Profile type
         F : Frequency
         G : Gribskov
         H : Henikoff
Select type [F]: <b>g</b>
Scoring matrix [Epprofile]: <b></b>
Enter a name for the profile [mymatrix]: <b>globins</b>
Gap opening penalty [3.0]: <b></b>
Gap extension penalty [0.3]: <b></b>
Output file [globins.prophecy]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.2">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers (* if not always prompted):
  [-sequence]          seqset     Sequence set USA
   -type               menu       Select type
*  -datafile           matrixf    Scoring matrix
   -name               string     Enter a name for the profile
*  -threshold          integer    Enter threshold reporting percentage
*  -open               float      Gap opening penalty
*  -extension          float      Gap extension penalty
  [-outfile]           outfile    Output file name

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1             integer    Start of each sequence to be used
   -send1               integer    End of each sequence to be used
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name

   "-outfile" associated qualifiers
   -odirectory2         string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence set USA</td>
<td>Readable set of sequences</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-type</td>
<td>Select type</td>
<td><table><tr><td>F</td> <td><i>(Frequency)</i></td></tr><tr><td>G</td> <td><i>(Gribskov)</i></td></tr><tr><td>H</td> <td><i>(Henikoff)</i></td></tr></table></td>
<td>F</td>
</tr>

<tr>
<td>-datafile</td>
<td>Scoring matrix</td>
<td>Comparison matrix file in EMBOSS data path</td>
<td>'Epprofile' for Gribskov type, or EBLOSUM62</td>
</tr>

<tr>
<td>-name</td>
<td>Enter a name for the profile</td>
<td>Any string is accepted</td>
<td>mymatrix</td>
</tr>

<tr>
<td>-threshold</td>
<td>Enter threshold reporting percentage</td>
<td>Integer from 1 to 100</td>
<td>75</td>
</tr>

<tr>
<td>-open</td>
<td>Gap opening penalty</td>
<td>Any numeric value</td>
<td>3.0</td>
</tr>

<tr>
<td>-extension</td>
<td>Gap extension penalty</td>
<td>Any numeric value</td>
<td>0.3</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.prophecy</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>

<b>prophecy</b> reads a protein or a nucleic sequence alignment USA.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: globins.msf</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
!!AA_MULTIPLE_ALIGNMENT 1.0

  ../data/globins.msf MSF:  164 Type: P 25/06/01 CompCheck: 4278 ..

  Name: HBB_HUMAN Len: 164  Check: 6914 Weight: 0.14
  Name: HBB_HORSE Len: 164  Check: 6007 Weight: 0.15
  Name: HBA_HUMAN Len: 164  Check: 3921 Weight: 0.15
  Name: HBA_HORSE Len: 164  Check: 4770 Weight: 0.19
  Name: MYG_PHYCA Len: 164  Check: 7930 Weight: 0.23
  Name: GLB5_PETMA Len: 164  Check: 1857 Weight: 0.21
  Name: LGB2_LUPLU Len: 164  Check: 2879 Weight: 0.10

//

           1                                               50
HBB_HUMAN  ~~~~~~~~VHLTPEEKSAVTALWGKVN.VDEVGGEALGR.LLVVYPWTQR
HBB_HORSE  ~~~~~~~~VQLSGEEKAAVLALWDKVN.EEEVGGEALGR.LLVVYPWTQR
HBA_HUMAN  ~~~~~~~~~~~~~~VLSPADKTNVKAA.WGKVGAHAGEYGAEALERMFLS
HBA_HORSE  ~~~~~~~~~~~~~~VLSAADKTNVKAA.WSKVGGHAGEYGAEALERMFLG
MYG_PHYCA  ~~~~~~~VLSEGEWQLVLHVWAKVEAD.VAGHGQDILIR.LFKSHPETLE
GLB5_PETMA PIVDTGSVAPLSAAEKTKIRSAWAPVYSTYETSGVDILVKFFTSTPAAQE
LGB2_LUPLU ~~~~~~~~GALTESQAALVKSSWEEFNANIPKHTHRFFILVLEIAPAAKD

           51                                             100
HBB_HUMAN  FFESFGDLSTPDAVMGNPKVKAHGKKVLGAFSDGLAHLDNLKGTFATLSE
HBB_HORSE  FFDSFGDLSNPGAVMGNPKVKAHGKKVLHSFGEGVHHLDNLKGTFAALSE
HBA_HUMAN  FPTTKTYFPHFDLSHGSAQVKGHGKKVADALTNAVAHVDDMPNALSALSD
HBA_HORSE  FPTTKTYFPHFDLSHGSAQVKAHGKKVGDALTLAVGHLDDLPGALSNLSD
MYG_PHYCA  KFDRFKHLKTEAEMKASEDLKKHGVTVLTALGAILKKKGHHEAELKPLAQ
GLB5_PETMA FFPKFKGLTTADQLKKSADVRWHAERIINAVNDAVASMDDTEKMSMKLRD
LGB2_LUPLU LFSFLKGTSEVPQNNPELQAHAGKVFKLVYEAAIQLQVTGVVVTDATLKN

           101                                            150
HBB_HUMAN  LHCDKLH..VDPENFRLLGNVLVCVLAHHFGKEFTPPVQAAYQKVVAGVA
HBB_HORSE  LHCDKLH..VDPENFRLLGNVLVVVLARHFGKDFTPELQASYQKVVAGVA
HBA_HUMAN  LHAHKLR..VDPVNFKLLSHCLLVTLAAHLPAEFTPAVHASLDKFLASVS
HBA_HORSE  LHAHKLR..VDPVNFKLLSHCLLSTLAVHLPNDFTPAVHASLDKFLSSVS
MYG_PHYCA  SHATKHK..IPIKYLEFISEAIIHVLHSRHPGDFGADAQGAMNKALELFR
GLB5_PETMA LSGKHAK..SFQVDPQYFKVLAAVIADTVAAGDAGFEKLMSMICILLRSA
LGB2_LUPLU LGSVHVSKGVADAHFPVVKEAILKTIKEVVGAKWSEELNSAWTIAYDELA

           151        164
HBB_HUMAN  NALAHKYH~~~~~~
HBB_HORSE  NALAHKYH~~~~~~
HBA_HUMAN  TVLTSKYR~~~~~~
HBA_HORSE  TVLTSKYR~~~~~~
MYG_PHYCA  KDIAAKYKELGYQG
GLB5_PETMA Y~~~~~~~~~~~~~
LGB2_LUPLU IVIKKEMNDAA~~~

</pre>
</td></tr></table><p>



<H2>
    Output file format
</H2>

The output is a profile file.

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: globins.prophecy </h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
# Pure Frequency Matrix
# Columns are amino acid counts A-&gt;Z
# Rows are alignment positions 1-&gt;n
Simple
Name		globins
Length		164
Maximum score	496
Thresh		75
Consensus	PIVDTGSVVALSEEEKSAVDAAWVKANAVAEVGGHALERGLLALEPATLEFFDSFKDLSTFDASHGSAQVKAHGKKVLDALGAAVAHLDDLEGTLAALSDLHADKLHKGVDPVNFKLLSEALLVTLAAHFGADFTPEVQASLDKALAGVANVLAHKYHDAAYQG
0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  0  0  0  
0  0  0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  
0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  1  0  0  0  0  0  
0  0  0  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  
0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  1  0  0  0  0  0  0  0  
0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  
0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  
0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  2  0  0  0  0  0  
1  0  0  0  0  0  1  0  0  0  0  1  0  0  0  0  0  0  0  0  0  2  0  0  0  0  0  
1  0  0  0  0  0  0  1  0  0  0  0  0  0  0  1  1  0  1  0  0  0  0  0  0  0  0  
0  0  0  0  1  0  0  0  0  0  0  4  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  
0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  0  0  0  2  2  0  0  0  0  0  0  0  
1  0  0  0  2  0  1  0  0  0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  0  0  0  
1  0  0  0  2  0  0  0  0  0  0  0  0  0  0  0  0  0  1  0  0  0  1  0  0  0  0  
0  0  0  0  3  0  0  0  0  0  0  0  0  0  0  0  2  0  0  0  0  2  0  0  0  0  0  
1  0  0  0  0  0  0  0  0  0  3  3  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  
2  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  3  1  0  1  0  0  0  0  0  
3  0  0  0  0  0  0  0  0  0  1  2  0  0  0  1  0  0  0  0  0  0  0  0  0  0  0  
2  0  0  0  0  0  0  1  1  0  0  0  0  0  0  0  0  0  0  0  0  3  0  0  0  0  0  
0  0  0  2  0  0  0  0  0  0  1  1  0  0  0  0  0  1  0  1  0  1  0  0  0  0  0  
2  0  0  0  0  0  0  0  0  0  2  0  0  0  0  0  0  0  2  0  0  0  1  0  0  0  0  
2  0  0  0  0  0  0  0  0  0  0  2  0  0  0  0  0  0  1  2  0  0  0  0  0  0  0  
0  0  0  0  0  0  0  0  0  0  1  0  0  2  0  0  0  0  0  0  0  0  4  0  0  0  0  
1  0  0  1  1  0  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  3  0  0  0  0  0  
0  0  0  0  2  0  0  0  0  0  4  0  0  0  0  1  0  0  0  0  0  0  0  0  0  0  0  
3  0  0  0  0  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  3  0  0  0  0  0  
2  0  0  1  0  0  0  0  0  0  0  0  0  3  0  0  0  0  0  0  0  0  0  0  1  0  0  
1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  
0  0  0  0  1  0  0  0  0  0  0  0  0  1  0  0  0  0  0  1  0  2  2  0  0  0  0  
1  0  0  1  1  0  1  0  1  0  0  0  0  0  0  0  0  0  1  0  0  0  0  0  1  0  0  
0  0  0  0  3  0  1  0  0  0  2  0  0  0  0  1  0  0  0  0  0  0  0  0  0  0  0  
0  0  0  0  0  0  0  1  0  0  1  0  0  0  0  0  0  0  0  1  0  4  0  0  0  0  0  
0  0  0  0  0  0  5  1  0  0  0  0  0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  
1  0  0  0  0  0  4  0  0  0  0  0  0  0  0  0  1  0  0  1  0  0  0  0  0  0  0  
0  0  0  1  2  0  0  3  0  0  0  0  0  0  0  0  0  0  0  0  0  1  0  0  0  0  0  
4  0  0  1  0  0  0  0  1  0  0  0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  0  
0  0  0  0  0  1  2  0  1  0  0  3  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  
0  0  0  0  2  1  2  0  1  0  0  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  
0  0  0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  3  0  0  0  1  0  0  2  0  0  
0  0  0  0  0  0  2  0  0  0  1  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  
2  0  0  0  0  1  0  0  0  0  0  3  0  0  0  0  0  0  0  0  0  1  0  0  0  0  0  


<font color=red>  [Part of this file has been deleted for brevity]</font>

0  0  0  1  0  0  0  1  0  0  0  0  0  4  0  0  0  0  0  0  0  0  0  0  1  0  0  
0  0  0  0  0  5  0  0  0  0  0  1  0  0  0  1  0  0  0  0  0  0  0  0  0  0  0  
0  0  0  0  1  0  0  0  0  0  2  0  0  0  0  1  1  2  0  0  0  0  0  0  0  0  0  
0  0  0  0  0  1  0  0  0  0  0  4  0  0  0  0  0  0  0  0  0  1  0  0  1  0  0  
0  0  0  0  0  1  0  0  1  0  0  4  0  0  0  0  0  0  0  0  0  1  0  0  0  0  0  
0  0  0  0  0  0  2  0  0  0  2  0  0  0  0  0  0  0  3  0  0  0  0  0  0  0  0  
0  0  0  0  2  0  0  2  0  0  0  0  0  2  0  0  0  0  0  0  0  1  0  0  0  0  0  
2  0  2  0  0  0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  0  2  0  0  0  0  0  
1  0  0  0  0  0  0  0  2  0  0  4  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  
1  0  0  0  0  0  0  0  1  0  0  3  0  0  0  0  0  0  0  0  0  2  0  0  0  0  0  
0  0  1  0  0  0  0  1  0  0  1  0  0  0  0  0  0  0  1  0  0  3  0  0  0  0  0  
0  0  0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  0  0  3  0  3  0  0  0  0  0  
1  0  0  0  0  0  0  0  1  0  0  5  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  
4  0  0  1  0  0  0  1  0  0  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  
1  0  0  0  1  0  0  1  0  0  0  0  0  0  0  0  0  1  1  1  0  1  0  0  0  0  0  
0  0  0  0  0  0  0  4  0  0  0  0  0  0  0  0  0  1  0  0  0  2  0  0  0  0  0  
1  0  0  0  0  2  0  1  0  0  0  2  0  0  0  0  0  0  0  0  0  1  0  0  0  0  0  
1  0  0  0  0  0  3  0  0  0  0  0  0  0  0  3  0  0  0  0  0  0  0  0  0  0  0  
2  0  0  0  0  0  2  0  0  0  2  0  0  1  0  0  0  0  0  0  0  0  0  0  0  0  0  
0  0  0  4  2  0  0  0  0  0  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  
1  0  0  0  0  5  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  1  0  0  0  0  
0  0  0  0  0  0  2  0  0  0  0  0  0  0  0  0  0  0  1  4  0  0  0  0  0  0  0  
1  0  0  0  1  1  0  0  0  0  0  0  0  0  0  4  0  0  0  0  0  0  0  0  0  0  0  
2  0  0  1  3  0  0  0  0  0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  0  0  0  
1  0  0  0  0  0  0  0  0  0  1  2  0  0  0  0  0  0  0  0  0  3  0  0  0  0  0  
0  0  0  0  0  0  0  2  0  0  0  1  0  1  0  0  3  0  0  0  0  0  0  0  0  0  0  
4  0  0  0  0  0  1  0  0  0  0  0  1  0  0  0  0  0  1  0  0  0  0  0  0  0  0  
3  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  4  0  0  0  0  0  0  0  0  
0  0  0  0  0  0  0  0  0  0  0  2  2  0  0  0  0  0  0  0  0  0  1  0  2  0  0  
0  0  0  2  0  0  0  0  1  0  0  0  0  1  0  0  2  0  0  1  0  0  0  0  0  0  0  
0  0  1  0  0  0  0  0  1  0  5  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  
2  0  0  0  0  2  0  0  1  0  0  0  0  0  0  0  0  0  0  0  0  2  0  0  0  0  0  
0  0  0  0  0  0  0  0  0  0  0  4  0  0  0  0  0  0  0  0  0  2  0  0  1  0  0  
3  0  0  1  1  0  0  0  0  0  0  1  0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  
0  0  0  0  1  0  2  0  0  0  0  1  0  0  0  0  0  1  2  0  0  0  0  0  0  0  0  
0  0  0  0  0  1  0  0  0  0  0  1  0  0  0  0  0  0  1  0  0  4  0  0  0  0  0  
4  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  1  2  0  0  0  0  0  0  0  0  
0  0  0  0  0  0  0  0  1  0  1  0  0  2  0  0  0  0  0  2  0  0  0  0  1  0  0  
2  0  0  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  3  0  0  0  0  0  
0  0  0  0  0  0  0  0  2  0  0  4  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  
3  0  0  0  0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  2  0  0  0  0  0  0  0  
1  0  0  0  0  0  0  2  0  0  1  0  0  0  0  0  0  0  2  0  0  0  0  0  0  0  0  
0  0  0  0  1  0  0  0  0  0  5  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  
0  0  0  0  0  0  0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  0  0  0  5  0  0  
0  0  0  0  0  0  0  2  0  0  1  0  0  1  0  0  0  2  0  0  0  0  0  0  0  0  0  
0  0  0  1  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  
1  0  0  0  0  0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  
1  0  0  0  0  0  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  
0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  1  0  0  
0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  0  0  
0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  
</pre>
</td></tr></table><p>

<a name="output.2"></a>
<h3>Output files for usage example 2</h3>
<p><h3>File: globins.prophecy </h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
# Gribskov Protein Profile
# Columns are amino acids A-&gt;Z
# Last column is indel penalty
# Rows are alignment positions 1-&gt;n
Gribskov
Name		globins
Matrix		pprofile
Length		164
Max_score	645.45
Threshold	75
Gap_open	3.00
Gap_extend	0.30
Consensus	PIVDTGSVVSLSEEELSAVDKAWVKANSVAEVGGHALERGLFASEPMTLEFFDTFKYLSTFDLSKGSADVKAHGKKVLDALGDAVAHLDDLEGTLAALSDLHAHKLKKGVDPVNFKLLSHCLLVVLASHLPGDFTPEVQASMDKFLASVATVLASKYRELGYQG
0.15 0.00 0.03 0.03 0.03 -0.18 0.09 0.06 -0.06 0.00 0.03 -0.09 -0.06 0.00 0.00 0.45 0.09 0.09 0.12 0.09 0.00 0.03 -0.24 0.00 -0.24 0.00 0.87 0.87
0.00 0.00 0.06 -0.06 -0.06 0.18 -0.09 -0.09 0.45 0.00 -0.06 0.24 0.18 -0.09 0.00 -0.06 -0.09 -0.09 -0.03 0.06 0.00 0.33 -0.15 0.00 0.03 0.00 0.87 0.87
0.06 0.00 0.06 -0.06 -0.06 0.06 0.06 -0.09 0.33 0.00 -0.06 0.24 0.18 -0.09 0.00 0.03 -0.06 -0.09 -0.03 0.06 0.00 0.45 -0.24 0.00 -0.03 0.00 0.87 0.87
0.09 0.00 -0.15 0.45 0.30 -0.30 0.18 0.12 -0.06 0.00 0.09 -0.15 -0.12 0.18 0.00 0.03 0.18 0.00 0.06 0.06 0.00 -0.06 -0.33 0.00 -0.15 0.00 0.87 0.87
0.12 0.00 0.06 0.06 0.06 -0.09 0.12 -0.03 0.06 0.00 0.06 -0.03 0.00 0.06 0.00 0.09 -0.03 -0.03 0.09 0.45 0.00 0.06 -0.18 0.00 -0.09 0.00 0.87 0.87
0.18 0.00 0.06 0.18 0.15 -0.18 0.45 -0.06 -0.09 0.00 -0.03 -0.15 -0.09 0.12 0.00 0.09 0.06 -0.09 0.18 0.12 0.00 0.06 -0.30 0.00 -0.18 0.00 0.87 0.87
0.12 0.00 0.18 0.06 0.06 -0.09 0.18 -0.06 -0.03 0.00 0.06 -0.12 -0.09 0.09 0.00 0.12 -0.03 0.03 0.45 0.09 0.00 -0.03 0.09 0.00 -0.12 0.00 0.87 0.87
0.13 0.00 0.13 -0.13 -0.13 0.13 0.13 -0.19 0.69 0.00 -0.13 0.50 0.38 -0.19 0.00 0.06 -0.13 -0.19 -0.06 0.13 0.00 0.94 -0.50 0.00 -0.06 0.00 0.87 0.87
0.59 0.00 -0.06 -0.07 -0.02 0.24 0.31 -0.25 0.68 0.00 -0.20 0.72 0.60 -0.14 0.00 0.14 -0.03 -0.39 0.03 0.23 0.00 0.97 -0.55 0.00 -0.12 0.00 0.87 0.87
0.52 0.00 0.12 0.35 0.35 -0.54 0.38 0.41 -0.22 0.00 0.20 -0.30 -0.22 0.31 0.00 0.76 0.53 0.27 0.61 0.20 0.00 -0.08 -0.38 0.00 -0.48 0.00 0.87 0.87
0.01 0.00 -0.88 -0.10 0.24 0.83 -0.26 -0.04 0.62 0.00 -0.16 1.19 0.96 -0.18 0.00 -0.22 0.11 -0.34 -0.28 -0.02 0.00 0.62 0.07 0.00 0.09 0.00 0.87 0.87
0.54 0.00 0.44 0.37 0.34 -0.45 0.94 -0.20 -0.08 0.00 0.14 -0.40 -0.25 0.35 0.00 0.41 -0.02 -0.08 1.07 0.80 0.00 0.08 -0.38 0.00 -0.51 0.00 0.87 0.87
0.82 0.00 -0.13 0.71 0.92 -0.68 0.80 0.16 -0.20 0.00 0.14 -0.34 -0.20 0.38 0.00 0.56 0.45 -0.09 0.42 0.36 0.00 0.03 -1.13 0.00 -0.61 0.00 0.87 0.87
0.37 0.00 -0.47 0.17 0.38 -0.05 0.14 0.07 -0.26 0.00 0.19 -0.05 -0.22 0.21 0.00 -0.01 0.13 0.35 0.52 0.05 0.00 -0.30 -0.16 0.00 0.01 0.00 0.87 0.87
0.41 0.00 -0.61 0.90 1.26 -0.71 0.55 0.42 0.25 0.00 0.31 0.13 0.15 0.40 0.00 0.26 1.04 0.04 0.05 0.19 0.00 0.49 -1.41 0.00 -0.69 0.00 4.55 4.55
0.13 0.00 -1.04 -0.15 0.01 0.48 -0.39 -0.11 0.51 0.00 0.83 0.99 1.12 -0.01 0.00 -0.10 0.23 0.20 -0.13 0.12 0.00 0.54 0.36 0.00 -0.23 0.00 4.55 4.55
1.00 0.00 0.64 0.24 0.24 -0.41 0.81 -0.30 0.35 0.00 0.13 -0.08 -0.01 0.24 0.00 0.58 -0.09 -0.17 1.23 0.86 0.00 0.56 -0.52 0.00 -0.50 0.00 4.55 4.55
1.09 0.00 -0.33 0.08 0.18 -0.09 0.21 -0.09 0.27 0.00 0.33 0.48 0.58 0.07 0.00 0.55 0.27 -0.09 0.23 0.35 0.00 0.48 -0.45 0.00 -0.42 0.00 4.55 4.55
0.81 0.00 0.28 0.11 0.11 0.02 0.25 0.19 0.96 0.00 -0.14 0.57 0.42 0.00 0.00 0.30 0.09 -0.24 0.04 0.33 0.00 1.16 -1.02 0.00 -0.07 0.00 4.55 4.55
0.18 0.00 -0.48 0.64 0.44 -0.46 0.23 0.20 0.36 0.00 0.51 0.16 0.35 0.23 0.00 0.18 0.36 0.36 0.10 0.44 0.00 0.49 -0.41 0.00 -0.54 0.00 4.55 4.55
0.54 0.00 -0.30 -0.00 -0.00 -0.24 0.14 -0.11 -0.31 0.00 0.85 -0.20 -0.13 0.31 0.00 0.17 0.07 0.74 1.03 0.20 0.00 -0.32 0.34 0.00 -0.23 0.00 4.55 4.55
1.15 0.00 0.04 0.11 0.19 -0.01 0.45 -0.22 0.41 0.00 0.00 0.45 0.45 0.10 0.00 0.39 0.02 -0.39 0.45 0.98 0.00 0.54 -0.54 0.00 -0.27 0.00 4.55 4.55
-0.59 0.00 -1.37 -0.55 -0.60 0.59 -0.70 0.19 -0.64 0.00 0.77 0.14 -0.34 0.60 0.00 -0.65 -0.10 1.43 0.47 -0.35 0.00 -0.90 1.17 0.00 0.70 0.00 4.55 4.55
0.84 0.00 0.10 0.51 0.46 -0.41 0.84 -0.17 0.76 0.00 -0.08 0.37 0.31 0.10 0.00 0.33 0.15 -0.39 0.23 0.43 0.00 1.25 -1.48 0.00 -0.47 0.00 4.55 4.55
0.29 0.00 -0.79 0.77 1.01 -1.00 0.24 0.34 -0.33 0.00 1.52 -0.50 0.03 0.60 0.00 0.59 0.73 0.81 0.39 0.36 0.00 -0.24 -0.67 0.00 -1.02 0.00 4.55 4.55
1.29 0.00 0.37 -0.04 0.02 -0.05 0.55 -0.31 0.87 0.00 -0.23 0.66 0.50 -0.12 0.00 0.39 -0.09 -0.53 0.21 0.43 0.00 1.26 -1.05 0.00 -0.13 0.00 4.55 4.55
0.85 0.00 0.11 0.82 0.60 -0.46 0.53 0.45 -0.20 0.00 0.14 -0.35 -0.33 1.10 0.00 0.04 0.34 -0.27 0.31 0.28 0.00 -0.17 -0.59 0.00 0.08 0.00 4.55 4.55
0.33 0.00 0.22 0.10 0.10 -0.16 0.27 -0.07 -0.03 0.00 0.06 -0.13 -0.09 0.12 0.00 0.19 -0.00 -0.01 0.51 0.15 0.00 -0.00 -0.02 0.00 -0.16 0.00 3.49 3.49
-0.07 0.00 -0.59 -0.28 -0.19 0.40 -0.10 -0.08 0.31 0.00 0.12 0.51 0.09 0.08 0.00 -0.22 -0.19 0.46 0.27 0.34 0.00 0.38 -0.15 0.00 0.27 0.00 4.55 4.55
0.76 0.00 0.40 0.62 0.60 -0.23 0.69 0.08 0.07 0.00 -0.05 -0.21 -0.21 0.39 0.00 0.11 0.11 -0.36 0.62 0.32 0.00 0.13 -0.59 0.00 -0.08 0.00 4.55 4.55
0.48 0.00 -0.64 1.07 1.40 -1.00 0.84 0.30 -0.37 0.00 0.92 -0.57 -0.17 0.68 0.00 0.43 0.73 0.33 0.49 0.41 0.00 -0.16 -1.18 0.00 -0.96 0.00 4.55 4.55
0.27 0.00 0.12 0.05 0.05 -0.03 0.22 0.21 0.92 0.00 0.13 0.58 0.47 0.01 0.00 0.26 0.04 -0.02 -0.04 0.63 0.00 1.28 -0.92 0.00 -0.17 0.00 4.55 4.55
0.84 0.00 0.41 0.85 0.73 -0.84 1.99 -0.09 -0.44 0.00 -0.05 -0.76 -0.50 0.65 0.00 0.52 0.30 -0.27 1.16 0.57 0.00 0.17 -1.15 0.00 -0.81 0.00 4.55 4.55
1.04 0.00 0.09 0.88 0.78 -1.00 1.73 -0.04 -0.37 0.00 0.06 -0.56 -0.30 0.60 0.00 0.54 0.72 -0.24 0.69 0.66 0.00 0.20 -1.41 0.00 -0.90 0.00 4.55 4.55
0.22 0.00 -0.42 1.10 1.14 -0.58 0.34 1.15 -0.01 0.00 0.23 -0.17 -0.22 0.63 0.00 0.23 0.76 0.22 -0.01 0.15 0.00 0.11 -1.12 0.00 -0.21 0.00 4.55 4.55
1.40 0.00 0.14 0.65 0.50 -0.62 0.58 0.00 0.39 0.00 0.14 -0.03 0.11 0.28 0.00 0.46 0.32 -0.15 0.40 0.47 0.00 0.44 -1.03 0.00 -0.47 0.00 4.55 4.55
0.15 0.00 -0.45 -0.28 -0.13 0.99 0.18 -0.35 0.98 0.00 -0.42 1.28 1.00 -0.26 0.00 -0.22 -0.18 -0.60 -0.08 0.14 0.00 1.05 -0.09 0.00 0.15 0.00 4.55 4.55


<font color=red>  [Part of this file has been deleted for brevity]</font>

0.16 0.00 -0.11 0.88 0.64 -0.34 0.31 0.88 -0.34 0.00 0.27 -0.44 -0.47 1.57 0.00 -0.20 0.43 -0.04 0.17 0.13 0.00 -0.41 -0.25 0.00 0.30 0.00 4.55 4.55
-0.40 0.00 -0.34 -1.18 -0.69 1.78 -0.70 -0.11 0.83 0.00 -0.69 1.65 0.86 -0.65 0.00 -0.27 -0.78 -0.56 -0.32 -0.26 0.00 0.50 1.18 0.00 1.21 0.00 4.55 4.55
0.11 0.00 -0.78 0.67 0.83 -1.02 0.09 0.60 -0.41 0.00 1.29 -0.48 0.09 0.52 0.00 0.51 1.05 1.17 0.23 0.13 0.00 -0.33 -0.04 0.00 -1.00 0.00 4.55 4.55
-0.32 0.00 -0.42 -0.96 -0.65 1.99 -0.80 -0.17 1.10 0.00 -0.68 1.95 1.30 -0.60 0.00 -0.69 -0.56 -0.75 -0.59 -0.25 0.00 0.97 1.06 0.00 1.13 0.00 4.55 4.55
-0.21 0.00 -0.66 -0.84 -0.54 1.76 -0.70 -0.35 1.55 0.00 -0.54 2.09 1.51 -0.65 0.00 -0.50 -0.46 -0.65 -0.50 -0.09 0.00 1.36 0.53 0.00 0.68 0.00 4.55 4.55
0.57 0.00 0.31 0.54 0.50 -0.76 1.07 -0.20 -0.29 0.00 0.79 -0.67 -0.28 0.59 0.00 0.49 0.18 0.31 1.56 0.50 0.00 -0.09 -0.13 0.00 -0.84 0.00 4.55 4.55
0.24 0.00 -0.40 0.85 1.05 -0.48 0.36 1.03 -0.03 0.00 0.30 -0.16 -0.18 1.01 0.00 0.17 0.68 0.19 0.09 0.19 0.00 0.09 -0.93 0.00 -0.16 0.00 4.55 4.55
0.91 0.00 0.71 -0.33 -0.32 0.16 0.31 -0.28 0.79 0.00 -0.46 0.35 0.32 -0.30 0.00 0.24 -0.31 -0.53 0.32 0.34 0.00 1.05 -1.14 0.00 0.39 0.00 4.55 4.55
0.36 0.00 -0.54 -0.45 -0.27 1.21 -0.41 -0.35 1.43 0.00 -0.36 1.70 1.36 -0.44 0.00 -0.21 -0.17 -0.59 -0.29 0.12 0.00 1.30 -0.03 0.00 0.23 0.00 4.55 4.55
0.47 0.00 -0.26 -0.37 -0.25 0.88 -0.15 -0.38 1.45 0.00 -0.34 1.51 1.20 -0.41 0.00 -0.06 -0.18 -0.56 -0.21 0.21 0.00 1.55 -0.42 0.00 0.09 0.00 4.55 4.55
0.28 0.00 0.49 -0.02 -0.04 -0.07 0.27 0.21 0.69 0.00 0.04 0.21 0.17 0.02 0.00 0.28 -0.04 0.03 0.42 0.26 0.00 0.98 -0.76 0.00 0.03 0.00 4.55 4.55
0.40 0.00 0.33 -0.08 -0.08 0.14 0.31 -0.38 1.39 0.00 -0.08 0.77 0.63 -0.19 0.00 0.20 -0.30 -0.38 0.08 1.15 0.00 1.57 -1.12 0.00 -0.23 0.00 4.55 4.55
0.33 0.00 -0.86 -0.55 -0.31 1.41 -0.48 -0.32 1.20 0.00 -0.40 1.93 1.56 -0.47 0.00 -0.25 -0.11 -0.62 -0.39 0.03 0.00 1.20 0.30 0.00 0.29 0.00 4.55 4.55
1.41 0.00 0.00 0.89 0.74 -0.87 0.64 0.54 -0.19 0.00 0.34 -0.35 -0.19 0.58 0.00 0.56 0.61 0.01 0.38 0.42 0.00 -0.01 -1.07 0.00 -0.41 0.00 4.55 4.55
0.59 0.00 0.21 0.36 0.43 -0.45 0.47 0.27 0.17 0.00 0.31 -0.13 0.02 0.31 0.00 0.47 0.22 0.28 0.65 0.68 0.00 0.32 -0.37 0.00 -0.45 0.00 4.55 4.55
-0.10 0.00 -0.10 0.27 0.27 -0.17 -0.19 1.38 0.12 0.00 0.26 0.04 0.06 0.35 0.00 0.32 0.58 0.81 -0.19 -0.03 0.00 0.30 -0.02 0.00 0.03 0.00 4.55 4.55
0.19 0.00 -0.34 -0.46 -0.20 1.05 -0.35 0.28 0.70 0.00 -0.39 1.24 0.78 -0.22 0.00 -0.16 -0.15 -0.37 -0.28 -0.06 0.00 0.65 0.35 0.00 0.68 0.00 4.55 4.55
1.19 0.00 0.28 0.51 0.45 -0.97 1.26 0.02 -0.33 0.00 0.03 -0.55 -0.33 0.28 0.00 1.54 0.42 -0.01 0.78 0.59 0.00 0.25 -1.45 0.00 -1.08 0.00 4.55 4.55
0.97 0.00 -0.10 0.77 0.68 -0.94 1.22 0.02 -0.35 0.00 0.67 -0.58 -0.19 0.90 0.00 0.41 0.47 0.06 0.68 0.53 0.00 0.03 -0.95 0.00 -0.76 0.00 4.55 4.55
0.46 0.00 -0.89 2.13 1.78 -1.45 0.86 0.63 -0.33 0.00 0.67 -0.72 -0.50 0.93 0.00 0.17 0.97 0.11 0.33 0.33 0.00 -0.33 -1.67 0.00 -0.85 0.00 4.55 4.55
-0.28 0.00 -0.20 -1.30 -0.80 1.86 -0.70 -0.17 0.67 0.00 -0.72 1.52 0.57 -0.60 0.00 -0.70 -0.99 -0.52 -0.21 -0.33 0.00 0.19 1.45 0.00 1.66 0.00 4.55 4.55
0.79 0.00 0.39 0.59 0.52 -0.69 1.39 -0.24 -0.02 0.00 0.15 -0.46 -0.23 0.47 0.00 0.52 0.02 -0.26 0.86 1.64 0.00 0.29 -1.13 0.00 -0.70 0.00 4.55 4.55
0.84 0.00 0.07 0.03 0.22 -0.34 0.36 0.17 -0.03 0.00 -0.05 0.01 -0.06 -0.01 0.00 1.35 0.18 0.02 0.43 0.34 0.00 0.19 -0.78 0.00 -0.50 0.00 4.55 4.55
1.12 0.00 -0.39 1.32 1.48 -1.09 0.88 0.39 -0.24 0.00 0.32 -0.47 -0.30 0.62 0.00 0.64 0.75 -0.09 0.47 0.45 0.00 -0.08 -1.63 0.00 -0.80 0.00 4.55 4.55
0.59 0.00 -0.23 -0.13 -0.06 0.22 0.13 -0.28 0.98 0.00 0.21 0.96 0.90 -0.16 0.00 0.16 0.01 -0.21 -0.02 0.29 0.00 1.32 -0.60 0.00 -0.24 0.00 4.55 4.55
0.10 0.00 -0.78 0.58 0.62 -0.35 -0.04 1.19 -0.17 0.00 0.31 0.22 0.17 0.63 0.00 0.23 1.43 0.43 -0.25 -0.12 0.00 -0.10 -0.31 0.00 -0.22 0.00 4.55 4.55
1.60 0.00 0.24 0.38 0.40 -0.54 1.03 -0.27 0.07 0.00 0.06 0.05 0.31 0.26 0.00 0.55 0.23 -0.29 0.68 0.53 0.00 0.41 -1.10 0.00 -0.55 0.00 4.55 4.55
1.41 0.00 0.80 0.40 0.40 -0.64 1.00 -0.27 -0.10 0.00 0.20 -0.47 -0.30 0.43 0.00 0.74 0.03 -0.10 1.77 0.57 0.00 0.03 -0.24 0.00 -0.60 0.00 4.55 4.55
-0.29 0.00 -0.52 -0.86 -0.64 1.61 -0.82 -0.18 0.74 0.00 -0.25 1.68 1.44 -0.47 0.00 -0.72 -0.37 -0.13 -0.51 -0.26 0.00 0.61 0.72 0.00 0.86 0.00 4.55 4.55
0.35 0.00 -0.50 1.14 0.87 -0.84 0.47 0.50 0.16 0.00 0.41 -0.19 -0.11 0.89 0.00 0.16 0.94 0.09 0.17 0.40 0.00 0.08 -1.08 0.00 -0.54 0.00 4.55 4.55
0.09 0.00 -0.26 0.19 0.16 -0.68 -0.11 0.05 0.03 0.00 1.63 -0.49 0.15 0.36 0.00 0.12 0.27 0.85 0.41 0.33 0.00 -0.03 -0.31 0.00 -0.42 0.00 4.55 4.55
0.55 0.00 0.24 -0.49 -0.29 0.76 -0.02 -0.31 1.20 0.00 -0.43 1.11 0.67 -0.36 0.00 -0.07 -0.47 -0.60 -0.03 0.19 0.00 1.14 -0.28 0.00 0.48 0.00 4.55 4.55
-0.07 0.00 -0.67 -0.71 -0.49 1.61 -0.56 -0.30 1.36 0.00 -0.50 2.05 1.57 -0.58 0.00 -0.41 -0.28 -0.66 -0.54 -0.07 0.00 1.50 0.38 0.00 0.51 0.00 4.55 4.55
1.16 0.00 -0.16 0.64 0.79 -0.38 0.64 0.01 0.12 0.00 0.11 0.11 0.16 0.34 0.00 0.38 0.35 -0.28 0.63 0.40 0.00 0.24 -0.79 0.00 -0.44 0.00 4.55 4.55
0.36 0.00 -0.06 0.32 0.42 -0.24 0.73 -0.04 -0.03 0.00 0.24 -0.07 0.16 0.28 0.00 0.32 0.21 0.24 0.90 0.28 0.00 0.18 0.13 0.00 -0.60 0.00 4.55 4.55
0.12 0.00 0.21 -0.52 -0.36 0.75 0.09 -0.39 1.27 0.00 -0.36 1.21 0.79 -0.40 0.00 -0.03 -0.49 -0.46 0.20 0.16 0.00 1.50 -0.16 0.00 0.26 0.00 4.55 4.55
1.38 0.00 0.45 0.35 0.35 -0.74 0.71 -0.02 -0.15 0.00 0.36 -0.41 -0.08 0.35 0.00 0.72 0.25 0.28 1.10 0.46 0.00 0.02 -0.11 0.00 -0.65 0.00 4.55 4.55
0.19 0.00 0.10 0.27 0.22 -0.07 0.10 0.24 0.15 0.00 0.55 -0.11 -0.00 0.78 0.00 -0.09 0.03 0.03 0.20 0.82 0.00 0.03 -0.12 0.00 0.08 0.00 4.55 4.55
0.85 0.00 0.09 0.49 0.33 -0.41 0.57 -0.10 0.63 0.00 -0.03 0.30 0.25 0.09 0.00 0.30 0.15 -0.31 0.17 0.36 0.00 0.96 -1.20 0.00 -0.35 0.00 0.92 0.92
-0.09 0.00 -0.63 -0.54 -0.36 1.36 -0.59 -0.32 1.43 0.00 -0.36 1.73 1.36 -0.50 0.00 -0.36 -0.23 -0.50 -0.41 0.00 0.00 1.24 0.21 0.00 0.32 0.00 0.92 0.92
1.31 0.00 0.23 0.36 0.36 -0.60 0.63 -0.11 0.07 0.00 0.31 -0.17 0.03 0.30 0.00 0.53 0.16 -0.16 0.47 1.05 0.00 0.22 -0.87 0.00 -0.45 0.00 0.92 0.92
0.65 0.00 0.26 0.40 0.40 -0.44 0.39 0.51 -0.20 0.00 0.35 -0.35 -0.24 0.48 0.00 0.46 0.32 0.27 0.81 0.26 0.00 -0.14 -0.14 0.00 -0.25 0.00 0.92 0.92
0.04 0.00 -0.82 0.51 0.58 -0.82 -0.05 0.18 -0.27 0.00 1.89 -0.41 0.22 0.56 0.00 0.14 0.58 0.98 0.27 0.27 0.00 -0.27 -0.03 0.00 -0.81 0.00 0.92 0.92
-0.37 0.00 1.14 -0.67 -0.64 1.67 -0.78 0.33 0.21 0.00 -0.71 0.54 0.09 -0.17 0.00 -1.01 -0.74 -0.71 -0.53 -0.37 0.00 -0.04 1.31 0.00 1.83 0.00 0.92 0.92
-0.16 0.00 -0.43 0.35 0.34 -0.55 -0.20 0.97 -0.38 0.00 0.98 -0.43 -0.00 0.60 0.00 0.26 0.63 1.21 0.07 0.00 0.00 -0.38 0.58 0.00 -0.38 0.00 0.92 0.92
0.14 0.00 -0.27 0.54 0.64 -0.34 0.25 0.19 -0.09 0.00 0.14 -0.17 -0.12 0.25 0.00 0.05 0.28 0.00 0.09 0.09 0.00 -0.09 -0.52 0.00 -0.24 0.00 0.92 0.92
0.18 0.00 -0.22 -0.12 -0.06 0.32 -0.08 -0.08 0.26 0.00 -0.10 0.48 0.39 -0.10 0.00 -0.03 -0.00 -0.17 -0.07 0.02 0.00 0.29 0.05 0.00 0.06 0.00 0.92 0.92
0.41 0.00 0.11 0.24 0.21 -0.27 0.58 -0.08 -0.10 0.00 -0.03 -0.18 -0.10 0.16 0.00 0.17 0.09 -0.14 0.25 0.19 0.00 0.09 -0.44 0.00 -0.24 0.00 0.92 0.92
-0.10 0.00 0.33 -0.16 -0.16 0.43 -0.20 0.10 0.03 0.00 -0.20 0.10 -0.03 -0.03 0.00 -0.26 -0.20 -0.20 -0.13 -0.10 0.00 -0.03 0.36 0.00 0.49 0.00 0.92 0.92
0.07 0.00 -0.20 0.20 0.20 -0.26 0.07 0.20 -0.10 0.00 0.13 -0.03 0.00 0.13 0.00 0.10 0.49 0.13 -0.03 -0.03 0.00 -0.07 -0.16 0.00 -0.20 0.00 0.92 0.92
0.20 0.00 0.07 0.20 0.16 -0.20 0.49 -0.07 -0.10 0.00 -0.03 -0.16 -0.10 0.13 0.00 0.10 0.07 -0.10 0.20 0.13 0.00 0.07 -0.33 0.00 -0.20 0.00 0.92 0.92
</pre>
</td></tr></table><p>


<p>

<h3>Simple frequency matrix</h3>

<p>

The columns represent amino acid counts for the amino acid residues from
A to Z.  The rows represent the alignment positions from 1->n.  The file
is a "Simple" frequency matrix of "Length" 20 amino acids.  The maximum
score this matrix can give is 496.  The "Threshold" value is an
instruction to the <b>profit</b> application to only report matches
above the given score. 

<p>

<h3>Gribskov profile</h3>

<p>

The columns represent the amino acids from A to Z, the rows denote the
alignment position from 1 to n.  The last column is the indel penalty. 
The "Name" is the name used by <b>prophet</b> to refer to the profile,
the "Matrix" is the name of the scoring matrix used (containing residue
substitution values).  "Length" is the length of the alignment. 
"Max_score" is the maximum score this profile can produce.  The
threshold is an instruction to <b>prophet</b> to only report hits equal
to or above the given value.  The gap opening and extension values are
used by <b>prophet</b> in the dynamic alignment (Smith Waterman
equivalent). 


<H2>
    Data files
</H2>


The two profile methods require a residue substitution scoring
matrix.

<p>

The Gribskov marices use the data file 'Epprofile' by default.  This is
derived from a PAM250 scoring matrix. 

<p>

The Henikoff matrices use the data file 'EBLOSUM62' by default.

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>


<H2>
    Notes
</H2>

None.


<H2>
    References
</H2>

<ol>

<li> Gribskov M, McLachlan AD, Eisenberg D. Proc Natl Acad Sci U S
A. 1987 Jul; 84(13): 4355-8.

<li> Henikoff S, Henikoff JG. J Mol Biol. 1994 Nov 4; 243(4): 574-8.

</ol>

<H2>
    Warnings
</H2>

None.


<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>


It always exits with status 0.

<H2>
    Known bugs
</H2>


None.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="profit.html">profit</a></td><td>Scan a sequence or database with a matrix or profile</td></tr>
<tr><td><a href="prophet.html">prophet</a></td><td>Gapped alignment for profiles</td></tr>
</table>


<H2>
    Author(s)
</H2>
Alan Bleasby (ableasby&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK




<H2>
    History
</H2>

1999 - Written Alan Bleasby.


<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.


<H2>
    Comments
</H2>
None


</BODY>
</HTML>

