
<HTML>

<HEAD>
  <TITLE>
  EMBOSS
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: PROTEIN COMPOSITION
</H2>
</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>

<!--END OF HEADER-->





<table border cellpadding=4 bgcolor="#FFFFF0">

<tr><th>Program name</th><th>Description</th></tr>

<tr>
<td><a href="backtranseq.html">backtranseq</a></td>
<td>
Back translate a protein sequence
</td>
</tr>

<tr>
<td><a href="charge.html">charge</a></td>
<td>
Protein charge plot
</td>
</tr>

<tr>
<td><a href="checktrans.html">checktrans</a></td>
<td>
Reports STOP codons and ORF statistics of a protein sequence
</td>
</tr>

<tr>
<td><a href="compseq.html">compseq</a></td>
<td>
Counts the composition of dimer/trimer/etc words in a sequence
</td>
</tr>

<tr>
<td><a href="emowse.html">emowse</a></td>
<td>
Protein identification by mass spectrometry
</td>
</tr>

<tr>
<td><a href="iep.html">iep</a></td>
<td>
Calculates the isoelectric point of a protein
</td>
</tr>

<tr>
<td><a href="mwfilter.html">mwfilter</a></td>
<td>
Filter noisy molwts from mass spec output
</td>
</tr>

<tr>
<td><a href="octanol.html">octanol</a></td>
<td>
Displays protein hydropathy
</td>
</tr>

<tr>
<td><a href="pepinfo.html">pepinfo</a></td>
<td>
Plots simple amino acid properties in parallel
</td>
</tr>

<tr>
<td><a href="pepstats.html">pepstats</a></td>
<td>
Protein statistics
</td>
</tr>

<tr>
<td><a href="pepwindow.html">pepwindow</a></td>
<td>
Displays protein hydropathy
</td>
</tr>

<tr>
<td><a href="pepwindowall.html">pepwindowall</a></td>
<td>
Displays protein hydropathy of a set of sequences
</td>
</tr>


</table>



</table>

</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>
