<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: psiblasts
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">
<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>
<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>
</td>
<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: psiblasts
</H2>
</font>
</td>
</tr>
</table>
</td>
</tr>
</table>
<p><hr><p>

<!--END OF HEADER-->

<H1>
Program psiblasts
</H1>
<H2>
    Function
</H2>
Runs PSI-BLAST given scopalign alignments

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
        Algorithms
  -->



This is part of Jon Ison's protein structure analysis package.

<p>

This package is still being developed.

<p>

Please ignore this program until further details can be documented.

<p>

All further queries should go to Jon Ison.  (<A
HREF="mailto:jison@hgmp.mrc.ac.uk">Jon Ison</A>)


 













<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->


Here is a sample session with psiblasts:

<pre>

% psiblasts

</pre>














<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-align]             string     Location of alignment files for input
  [-alignextn]         string     File extension of alignment files
  [-niter]             integer    Number of PSIBLAST iterations
  [-evalue]            float      Threshold E-value for inclusion in family
  [-maxhits]           integer    Maximum number of hits
  [-submatrix]         string     Residue substitution matrix
  [-families]          outfile    Name of families file for output
  [-logf]              outfile    Name of log file for the build

   Optional qualifiers: (none)
   Advanced qualifiers: (none)
   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-align]<br>(Parameter 1)</td>
<td>Location of alignment files for input</td>
<td>Any string is accepted</td>
<td>./</td>
</tr>

<tr>
<td>[-alignextn]<br>(Parameter 2)</td>
<td>File extension of alignment files</td>
<td>Any string is accepted</td>
<td>.align</td>
</tr>

<tr>
<td>[-niter]<br>(Parameter 3)</td>
<td>Number of PSIBLAST iterations</td>
<td>Any integer value</td>
<td>1</td>
</tr>

<tr>
<td>[-evalue]<br>(Parameter 4)</td>
<td>Threshold E-value for inclusion in family</td>
<td>Any integer value</td>
<td>0.001</td>
</tr>

<tr>
<td>[-maxhits]<br>(Parameter 5)</td>
<td>Maximum number of hits</td>
<td>Any integer value</td>
<td>1000</td>
</tr>

<tr>
<td>[-submatrix]<br>(Parameter 6)</td>
<td>Residue substitution matrix</td>
<td>Any string is accepted</td>
<td>./EBLOSUM62</td>
</tr>

<tr>
<td>[-families]<br>(Parameter 7)</td>
<td>Name of families file for output</td>
<td>Output file</td>
<td>scop.hits</td>
</tr>

<tr>
<td>[-logf]<br>(Parameter 8)</td>
<td>Name of log file for the build</td>
<td>Output file</td>
<td>psiblasts.log</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->















<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 

        If you wish to include the standard description of the avalable
	report formats, use:
<p>

The output is a standard EMBOSS report file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-rformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  The available format names are: embl,
genbank, gff, pir, swiss, trace, listfile, dbmotif, diffseq, excel,
feattable, motif, regions, seqtable, simple, srs, table, tagseq

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html</A>
for further information on report formats.

<p>

   -->


















<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.

	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

   -->















<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="contacts.html">contacts</a></td><td>Reads coordinate files and writes contact files</td></tr>
<tr><td><a href="dichet.html">dichet</a></td><td>Parse dictionary of heterogen groups</td></tr>
<tr><td><a href="interface.html">interface</a></td><td>Reads coordinate files and writes inter-chain contact files</td></tr>
<tr><td><a href="scopalign.html">scopalign</a></td><td>Generate alignments for SCOP families</td></tr>
<tr><td><a href="seqsort.html">seqsort</a></td><td>Removes ambiguities from a set of hits resulting from a database search</td></tr>
<tr><td><a href="siggen.html">siggen</a></td><td>Generates a sparse protein signature</td></tr>
<tr><td><a href="sigscan.html">sigscan</a></td><td>Scans a sparse protein signature against swissprot</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

This application was written by Jon Ison <a
href="mailto:jison@hgmp.mrc.ac.uk">(jison@hgmp.mrc.ac.uk)</a>

















<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (date) - author.














<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->








</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

