<HTML>

<HEAD>
  <TITLE>
  EMBOSS: rebaseextract
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
rebaseextract
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
    Function
</H2>
Process the REBASE database for use by restriction enzyme applications

<H2>
    Description
</H2>


<p><b>rebaseextract</b> processes the REBASE database for use by the EMBOSS restriction enzyme applications. It derives recognition site and cleavage information from the <tt>withrefm</tt> file of an REBASE distribution. It creates three files in the EMBOSS data subdirectory <tt>REBASE</tt>: a pattern file, a reference file and a supplier file. It will also (by default) produce an <tt>embossre.equ</tt> file of preferred isoschizomers using restriction enzyme prototypes in the "proto" file. This can be turned off by setting the <tt>-equivalences</tt> option to be false.</p>

<H2>
    Usage
</H2>


<b>Here is a sample session with rebaseextract</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>rebaseextract </b>
Process the REBASE database for use by restriction enzyme applications
REBASE database withrefm file: <b>withrefm</b>
REBASE database proto file: <b>proto</b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>


<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-infile]            infile     REBASE database withrefm file
  [-protofile]         infile     REBASE database proto file

   Additional (Optional) qualifiers:
   -[no]equivalences   boolean    [Y] This option calculates an embossre.equ
                                  file using restriction enzyme prototypes in
                                  the withrefm file.

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers: (none)
   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages

</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-infile]<br>(Parameter 1)</td>
<td>REBASE database withrefm file</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-protofile]<br>(Parameter 2)</td>
<td>REBASE database proto file</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]equivalences</td>
<td>This option calculates an embossre.equ file using restriction enzyme prototypes in the withrefm file.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<H2>
    Input file format
</H2>

The input file must be the "withrefm" file of a REBASE distribution.

<p>

For example, the withrefm file for REBASE version 005 is at: <a
href="ftp://ftp.neb.com/pub/rebase/withrefm.005">ftp://ftp.neb.com/pub/rebase/withrefm.005</a>

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: withrefm</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
 
REBASE version 106                                              withrefm.106
 
    =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
    REBASE, The Restriction Enzyme Database   http://rebase.neb.com
    Copyright (c)  Dr. Richard J. Roberts, 2001.   All rights reserved.
    =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
 
Rich Roberts                                                    May 31 2001
 

&lt;ENZYME NAME&gt;   Restriction enzyme name.
&lt;ISOSCHIZOMERS&gt; Other enzymes with this specificity.
&lt;RECOGNITION SEQUENCE&gt; 
                These are written from 5' to 3', only one strand being given.
                If the point of cleavage has been determined, the precise site
                is marked with ^.  For enzymes such as HgaI, MboII etc., which
                cleave away from their recognition sequence the cleavage sites
                are indicated in parentheses.  

                For example HgaI GACGC (5/10) indicates cleavage as follows:
                                5' GACGCNNNNN^      3'
                                3' CTGCGNNNNNNNNNN^ 5'

                In all cases the recognition sequences are oriented so that
                the cleavage sites lie on their 3' side.

                REBASE Recognition sequences representations use the standard 
                abbreviations (Eur. J. Biochem. 150: 1-5, 1985) to represent 
                ambiguity.
                                R = G or A
                                Y = C or T
                                M = A or C
                                K = G or T
                                S = G or C
                                W = A or T
                                B = not A (C or G or T)
                                D = not C (A or G or T)
                                H = not G (A or C or T)
                                V = not T (A or C or G)
                                N = A or C or G or T



                ENZYMES WITH UNUSUAL CLEAVAGE PROPERTIES:  

                Enzymes that cut on both sides of their recognition sequences,
                such as BcgI, Bsp24I, CjeI and CjePI, have 4 cleavage sites
                each instead of 2.



<font color=red>  [Part of this file has been deleted for brevity]</font>

&lt;5&gt;Klebsiella pneumoniae OK8
&lt;6&gt;ATCC 49790
&lt;7&gt;ABCDEFGHIJKLMNOQRSTU
&lt;8&gt;Kiss, A., Finta, C., Venetianer, P., (1991) Nucleic Acids Res., vol. 19, pp. 3460.
Smith, D.I., Blattner, F.R., Davies, J., (1976) Nucleic Acids Res., vol. 3, pp. 343-353.
Tomassini, J., Roychoudhury, R., Wu, R., Roberts, R.J., (1978) Nucleic Acids Res., vol. 5, pp. 4055-4064.

&lt;1&gt;Ksp632I
&lt;2&gt;Bco5I,Bco116I,BcoKI,BcoSI,BcrAI,BseZI,BsrEI,Bst6I,Bst158I,Bsu6I,Eam1104I,EarI,TdeII,Uba1192I,Uba1276I,VpaKutEI,VpaKutFI,VpaO5I
&lt;3&gt;CTCTTC(1/4)
&lt;4&gt;
&lt;5&gt;Kluyvera species 632
&lt;6&gt;DSM 4196
&lt;7&gt;M
&lt;8&gt;Bolton, B.J., Schmitz, G.G., Jarsch, M., Comer, M.J., Kessler, C., (1988) Gene, vol. 66, pp. 31-43.
Tsukahara, S., Yamakawa, H., Takai, K., Takaku, H., (1994) Nucleosides &amp; Nucleotides, vol. 13, pp. 1617-1626.

&lt;1&gt;MaeII
&lt;2&gt;HpyCH4IV,HpyF13III,HpyF35II,HpyF74II,TaiI,TscI,Tsp49I,TspIDSI,TspWAM8AI,TtmI
&lt;3&gt;A^CGT
&lt;4&gt;
&lt;5&gt;Methanococcus aeolicus PL-15/H
&lt;6&gt;K.O. Stetter
&lt;7&gt;M
&lt;8&gt;Schmid, K., Thomm, M., Laminet, A., Laue, F.G., Kessler, C., Stetter, K.O., Schmitt, R., (1984) Nucleic Acids Res., vol. 12, pp. 2619-2628.

&lt;1&gt;NotI
&lt;2&gt;CciNI,CspBI,MchAI
&lt;3&gt;GC^GGCCGC
&lt;4&gt;?(4)
&lt;5&gt;Nocardia otitidis-caviarum
&lt;6&gt;ATCC 14630
&lt;7&gt;ABCDEFGHJKLMNOQRSTU
&lt;8&gt;Borsetti, R., Wise, D., Qiang, B.-Q., Schildkraut, I., Unpublished observations.
Morgan, R.D., Unpublished observations.
Morgan, R.D., Benner, J.S., Claus, T.E., US Patent Office, 1994.
Qiang, B.-Q., Schildkraut, I., (1987) Methods Enzymol., vol. 155, pp. 15-21.

&lt;1&gt;TaqI
&lt;2&gt;CviSIII,EsaBC3I,HpyV,Hpy26II,HpyF14III,HpyF16I,HpyF23I,HpyF24I,HpyF26III,HpyF30I,HpyF35I,HpyF40II,HpyF42IV,HpyF45I,HpyF49I,HpyF52I,HpyF59III,HpyF62II,HpyF64I,HpyF65II,HpyF66IV,HpyF71I,HpyF73II,HpyJP26II,PpaAII,Taq20I,Tbr51I,TfiA3I,TfiTok4A2I,TfiTok6A1I,TflI,Tsc4aI,Tsp32I,Tsp32II,Tsp358I,Tsp505I,Tsp510I,TspAK13D21I,TspAK16D24I,TspNI,TspVi4AI,TspVil3I,Tth24I,TthHB8I,TthRQI
&lt;3&gt;T^CGA
&lt;4&gt;4(6)
&lt;5&gt;Thermus aquaticus YTI
&lt;6&gt;J.I. Harris
&lt;7&gt;ABCDEFGIJLMNOQRSTU
&lt;8&gt;Anton, B.P., Brooks, J.E., Unpublished observations.
Fomenkov, A., Xiao, J.-P., Dila, D., Raleigh, E., Xu, S.-Y., (1994) Nucleic Acids Res., vol. 22, pp. 2399-2403.
McClelland, M., (1981) Nucleic Acids Res., vol. 9, pp. 6795-6804.
Sato, S., Hutchison, C.A. III, Harris, J.I., (1977) Proc. Natl. Acad. Sci. U. S. A., vol. 74, pp. 542-546.
Zebala, J.A., (1993) Diss. Abstr., vol. 54, pp. 1394-1398.

</pre>
</td></tr></table><p>
<p><h3>File: proto</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
 
REBASE version 305                                              proto.305
 
    =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
    REBASE, The Restriction Enzyme Database   http://rebase.neb.com
    Copyright (c)  Dr. Richard J. Roberts, 2003.   All rights reserved.
    =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
 
Rich Roberts                                                    Apr 30 2003
 



	    TYPE II ENZYMES
	    ---------------

BseYI                          CCCAGC (-5/-1)
BsiYI                          CCNNNNN^NNGG
BsrI                           ACTGG (1/-1)
HaeIII                         GG^CC
HpaII                          C^CGG
Ksp632I                        CTCTTC (1/4)
MaeII                          A^CGT



	    TYPE I ENZYMES
	    ---------------

EcoAI                          GAGNNNNNNNGTCA
EcoBI                          TGANNNNNNNNTGCT
EcoDI                          TTANNNNNNNGTCY
EcoDR2                         TCANNNNNNGTCG
EcoDR3                         TCANNNNNNNATCG
EcoDXXI                        TCANNNNNNNRTTC
EcoEI                          GAGNNNNNNNATGC
EcoKI                          AACNNNNNNGTGC



	    TYPE III ENZYMES
	    ---------------

EcoPI                          AGACC
EcoP15I                        CAGCAG (25/27)
HinfIII                        CGAAT
StyLTI                         CAGAG

</pre>
</td></tr></table><p>

<H2>
    Output file format
</H2>



<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>Directory: REBASE</h3>
<p>This directory contains output files, for example embossre.enz embossre.equ embossre.ref and embossre.sup.
<p><h3>File: REBASE/embossre.enz</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
# REBASE enzyme patterns for EMBOSS
#
# Format:
# name&lt;ws&gt;pattern&lt;ws&gt;len&lt;ws&gt;ncuts&lt;ws&gt;blunt&lt;ws&gt;c1&lt;ws&gt;c2&lt;ws&gt;c3&lt;ws&gt;c4
#
# Where:
# name = name of enzyme
# pattern = recognition site
# len = length of pattern
# ncuts = number of cuts made by enzyme
#         Zero represents unknown
# blunt = true if blunt end cut, false if sticky
# c1 = First 5' cut
# c2 = First 3' cut
# c3 = Second 5' cut
# c4 = Second 3' cut
#
# Examples:
# AAC^TGG -&gt; 6 2 1 3 3 0 0
# A^ACTGG -&gt; 6 2 0 1 5 0 0
# AACTGG  -&gt; 6 0 0 0 0 0 0
# AACTGG(-5/-1) -&gt; 6 2 0 1 5 0 0
# (8/13)GACNNNNNNTCA(12/7) -&gt; 12 4 0 -9 -14 24 19
#
# i.e. cuts are always to the right of the given
# residue and sequences are always with reference to
# the 5' strand.
# Sequences are numbered ... -3 -2 -1 1 2 3 ... with
# the first residue of the pattern at base number 1.
#
AaeI	ggatcc	6	0	0	0	0	0	0
AciI	CCGC	4	2	0	1	3	0	0
AclI	AACGTT	6	2	0	2	4	0	0
BamHI	GGATCC	6	2	0	1	5	0	0
BceAI	ACGGC	5	2	0	17	19	0	0
Bsc4I	CCNNNNNNNGG	11	2	0	7	4	0	0
Bse1I	ACTGG	5	2	0	6	4	0	0
BseYI	CCCAGC	6	2	0	1	5	0	0
BshI	GGCC	4	2	1	2	2	0	0
BsiSI	CCGG	4	2	0	1	3	0	0
BsiYI	CCNNNNNNNGG	11	2	0	7	4	0	0
BssKI	CCNGG	5	2	0	-1	5	0	0
BsrI	ACTGG	5	2	0	6	4	0	0
Bsu6I	CTCTTC	6	2	0	7	10	0	0
ClaI	ATCGAT	6	2	0	2	4	0	0
EcoRI	GAATTC	6	2	0	1	5	0	0
EcoRII	CCWGG	5	2	0	-1	5	0	0
HaeIII	GGCC	4	2	1	2	2	0	0
HhaI	GCGC	4	2	0	3	1	0	0
Hin4I	GAYNNNNNVTC	11	4	0	-9	-14	24	19
Hin6I	GCGC	4	2	0	1	3	0	0
HinP1I	GCGC	4	2	0	1	3	0	0
HindI	cac	3	0	0	0	0	0	0
HindII	GTYRAC	6	2	1	3	3	0	0
HindIII	AAGCTT	6	2	0	1	5	0	0
HpaII	CCGG	4	2	0	1	3	0	0
HpyCH4IV	ACGT	4	2	0	1	3	0	0
HspAI	GCGC	4	2	0	1	3	0	0
KpnI	GGTACC	6	2	0	5	1	0	0
Ksp632I	CTCTTC	6	2	0	7	10	0	0
MaeII	ACGT	4	2	0	1	3	0	0
NotI	GCGGCCGC	8	2	0	2	6	0	0
TaqI	TCGA	4	2	0	1	3	0	0
</pre>
</td></tr></table><p>
<p><h3>File: REBASE/embossre.equ</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Bsc4I BsiYI
Bse1I BsrI
BshI HaeIII
BsiSI HpaII
Bsu6I Ksp632I
HpyCH4IV MaeII
</pre>
</td></tr></table><p>
<p><h3>File: REBASE/embossre.ref</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
# REBASE enzyme information for EMBOSS
#
# Format:
# Line 1: Name of Enzyme
# Line 2: Organism
# Line 3: Isoschizomers
# Line 4: Methylation
# Line 5: Source
# Line 6: Suppliers
# Line 7: Number of following references
# Lines 8..n: References
# // (end of entry marker)
#
AaeI
Acetobacter aceti sub. liquefaciens
BamHI,AacI,AcaII,AccEBI,AinII,AliI,Ali12257I,Ali12258I,ApaCI,AsiI,AspTII,Atu1II,BamFI,BamKI,BamNI,Bca1259I,Bce751I,Bco10278I,BnaI,BsaDI,Bsp30I,Bsp46I,Bsp90II,Bsp98I,Bsp130I,Bsp131I,Bsp144I,Bsp4009I,BspAAIII,BstI,Bst1126I,Bst2464I,Bst2902I,BstQI,Bsu90I,Bsu8565I,Bsu8646I,BsuB519I,BsuB763I,CelI,DdsI,GdoI,GinI,GoxI,GseIII,MleI,Mlu23I,NasBI,Nsp29132II,NspSAIV,OkrAI,Pac1110I,Pae177I,Pfl8I,Psp56I,RhsI,Rlu4I,RspLKII,SolI,SpvI,SurI,Uba19I,Uba31I,Uba38I,Uba51I,Uba88I,Uba1098I,Uba1163I,Uba1167I,Uba1172I,Uba1173I,Uba1205I,Uba1224I,Uba1242I,Uba1250I,Uba1258I,Uba1297I,Uba1302I,Uba1324I,Uba1325I,Uba1334I,Uba1339I,Uba1346I,Uba1383I,Uba1398I,Uba1402I,Uba1414I

M. Van Montagu

1
Seurinck, J., van Montagu, M., Unpublished observations.
//
AciI
Arthrobacter citreus

?(5),-2(5)
NEB 577
N
2
Lunnen, K.D., Heiter, D., Wilson, G.G., Unpublished observations.
Polisson, C., Morgan, R.D., (1990) Nucleic Acids Res., vol. 18, pp. 5911.
//
AclI
Acinetobacter calcoaceticus M4
Psp1406I
3(5)
S.K. Degtyarev
IN
2
Degtyarev, S.K., Abdurashitov, M.A., Kolyhalov, A.A., Rechkunova, N.I., (1992) Nucleic Acids Res., vol. 20, pp. 3787.
Lunnen, K.D., Wilson, G.G., Unpublished observations.
//
BamHI
Bacillus amyloliquefaciens H
AacI,AaeI,AcaII,AccEBI,AinII,AliI,Ali12257I,Ali12258I,ApaCI,AsiI,AspTII,Atu1II,BamFI,BamKI,BamNI,Bca1259I,Bce751I,Bco10278I,BnaI,BsaDI,Bsp30I,Bsp46I,Bsp90II,Bsp98I,Bsp130I,Bsp131I,Bsp144I,Bsp4009I,BspAAIII,BstI,Bst1126I,Bst2464I,Bst2902I,BstQI,Bsu90I,Bsu8565I,Bsu8646I,BsuB519I,BsuB763I,CelI,DdsI,GdoI,GinI,GoxI,GseIII,MleI,Mlu23I,NasBI,Nsp29132II,NspSAIV,OkrAI,Pac1110I,Pae177I,Pfl8I,Psp56I,RhsI,Rlu4I,RspLKII,SolI,SpvI,SurI,Uba19I,Uba31I,Uba38I,Uba51I,Uba88I,Uba1098I,Uba1163I,Uba1167I,Uba1172I,Uba1173I,Uba1205I,Uba1224I,Uba1242I,Uba1250I,Uba1258I,Uba1297I,Uba1302I,Uba1324I,Uba1325I,Uba1334I,Uba1339I,Uba1346I,Uba1383I,Uba1398I,Uba1402I,Uba1414I
5(4)
ATCC 49763
ABCDEFGHIJKLMNOQRSTUV
10
Brooks, J.E., Howard, K.A., US Patent Office, 1994.


<font color=red>  [Part of this file has been deleted for brevity]</font>

ATCC 49790
ABCDEFGHIJKLMNOQRSTU
3
Kiss, A., Finta, C., Venetianer, P., (1991) Nucleic Acids Res., vol. 19, pp. 3460.
Smith, D.I., Blattner, F.R., Davies, J., (1976) Nucleic Acids Res., vol. 3, pp. 343-353.
Tomassini, J., Roychoudhury, R., Wu, R., Roberts, R.J., (1978) Nucleic Acids Res., vol. 5, pp. 4055-4064.
//
Ksp632I
Kluyvera species 632
Bco5I,Bco116I,BcoKI,BcoSI,BcrAI,BseZI,BsrEI,Bst6I,Bst158I,Bsu6I,Eam1104I,EarI,TdeII,Uba1192I,Uba1276I,VpaKutEI,VpaKutFI,VpaO5I

DSM 4196
M
2
Bolton, B.J., Schmitz, G.G., Jarsch, M., Comer, M.J., Kessler, C., (1988) Gene, vol. 66, pp. 31-43.
Tsukahara, S., Yamakawa, H., Takai, K., Takaku, H., (1994) Nucleosides &amp; Nucleotides, vol. 13, pp. 1617-1626.
//
MaeII
Methanococcus aeolicus PL-15/H
HpyCH4IV,HpyF13III,HpyF35II,HpyF74II,TaiI,TscI,Tsp49I,TspIDSI,TspWAM8AI,TtmI

K.O. Stetter
M
1
Schmid, K., Thomm, M., Laminet, A., Laue, F.G., Kessler, C., Stetter, K.O., Schmitt, R., (1984) Nucleic Acids Res., vol. 12, pp. 2619-2628.
//
NotI
Nocardia otitidis-caviarum
CciNI,CspBI,MchAI
?(4)
ATCC 14630
ABCDEFGHJKLMNOQRSTU
4
Borsetti, R., Wise, D., Qiang, B.-Q., Schildkraut, I., Unpublished observations.
Morgan, R.D., Unpublished observations.
Morgan, R.D., Benner, J.S., Claus, T.E., US Patent Office, 1994.
Qiang, B.-Q., Schildkraut, I., (1987) Methods Enzymol., vol. 155, pp. 15-21.
//
TaqI
Thermus aquaticus YTI
CviSIII,EsaBC3I,HpyV,Hpy26II,HpyF14III,HpyF16I,HpyF23I,HpyF24I,HpyF26III,HpyF30I,HpyF35I,HpyF40II,HpyF42IV,HpyF45I,HpyF49I,HpyF52I,HpyF59III,HpyF62II,HpyF64I,HpyF65II,HpyF66IV,HpyF71I,HpyF73II,HpyJP26II,PpaAII,Taq20I,Tbr51I,TfiA3I,TfiTok4A2I,TfiTok6A1I,TflI,Tsc4aI,Tsp32I,Tsp32II,Tsp358I,Tsp505I,Tsp510I,TspAK13D21I,TspAK16D24I,TspNI,TspVi4AI,TspVil3I,Tth24I,TthHB8I,TthRQI
4(6)
J.I. Harris
ABCDEFGIJLMNOQRSTU
5
Anton, B.P., Brooks, J.E., Unpublished observations.
Fomenkov, A., Xiao, J.-P., Dila, D., Raleigh, E., Xu, S.-Y., (1994) Nucleic Acids Res., vol. 22, pp. 2399-2403.
McClelland, M., (1981) Nucleic Acids Res., vol. 9, pp. 6795-6804.
Sato, S., Hutchison, C.A. III, Harris, J.I., (1977) Proc. Natl. Acad. Sci. U. S. A., vol. 74, pp. 542-546.
Zebala, J.A., (1993) Diss. Abstr., vol. 54, pp. 1394-1398.
//
</pre>
</td></tr></table><p>
<p><h3>File: REBASE/embossre.sup</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
# REBASE Supplier information for EMBOSS
#
# Format:
# Code of Supplier&lt;ws&gt;Supplier name
#
A Amersham Pharmacia Biotech (1/01)
B Life Technologies Inc. (1/98)
C Minotech, Molecular Biology Products (12/00)
D HYBAID GmbH (12/00)
E Stratagene (11/00)
F Fermentas AB (5/01)
G Q-BIOgene (1/01)
H American Allied Biochemical, Inc. (10/98)
I SibEnzyme Ltd. (1/01)
J Nippon Gene Co., Ltd. (6/00)
K Takara Shuzo Co. Ltd. (2/01)
L Transgenomic Ltd. (1/01)
M Roche Molecular Biochemicals (1/01)
N New England BioLabs (12/00)
O Toyobo Biochemicals (11/98)
P Megabase Research Products (5/99)
Q CHIMERx (10/97)
R Promega Corporation (6/99)
S Sigma Chemical Corporation (11/98)
T Advanced Biotechnologies Ltd. (3/98)
U Bangalore Genei (2/01)
V MRC-Holland (3/01)
</pre>
</td></tr></table><p>

<p>

The output files are held in the REBASE subdirectory of the EMBOSS data
directory. There are three:


<ul>
	<li> embossre.enz		Enzyme pattern file
	<li> embossre.ref		Enzyme references
	<li> embossre.sup		Enzyme suppliers
</ul>

<p>

<b>rebaseextract</b> will also (by default) produce an 'embossre.equ'
file in the EMBOSS data directory.  This can be turned off by setting
the -equivalences option to be false.  This option calculates an
'embossre.equ' file using restriction enzyme prototypes in the
"withrefm" file.  The 'embossre.equ' file is a file of preferred
isoschizomers.  You may edit it to contain your available restriction
enzymes. 

<p>



<H2>
    Data files
</H2>


The "withrefm" file of an REBASE distribution is the input file for this
program.

<H2>
    Notes
</H2>

<p>The Restriction Enzyme database (REBASE) is a collection of information about restriction enzymes and related proteins. It contains published and unpublished references, recognition and cleavage sites, isoschizomers, commercial availability, methylation sensitivity, crystal and sequence data. DNA methyltransferases, homing endonucleases, nicking enzymes, specificity subunits and control proteins are also included. Most recently, putative DNA methyltransferases and restriction enzymes, as predicted from analysis of genomic sequences, are also listed.</p>

<p>The home page of REBASE is: <a href="http://rebase.neb.com/">http://rebase.neb.com/</a></p>

<p>The EMBOSS programs that find restriction cutting sites use the data files produced by this program and will not work without them. Running this program may be the job of your system manager.</p>

<p>The ready-made files produced by this program may already be available at the REBASE web site: <a href="http://rebase.neb.com/rebase/rebase.files.html">http://rebase.neb.com/rebase/rebase.files.html</a> or <a href="http://rebase.neb.com/rebase/rebase.f37.html">http://rebase.neb.com/rebase/rebase.f37.html</a> </p>

<p>You may edit the <tt>embossre.equ</tt> file it to contain details for your available restriction enzymes.</p>



<H2>
    References
</H2>


<ol>

<li>
Nucleic Acids Research 27: 312-313 (1999).

</ol>
<H2>
    Warnings
</H2>


The program will warn you if the input file is incorrectly formatted.

<H2>
    Diagnostic Error Messages
</H2>


<H2>
    Exit status
</H2>


It exits with status 0 unless an error is reported.

<H2>
    Known bugs
</H2>


<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="aaindexextract.html">aaindexextract</a></td>
<td>Extract amino acid property data from AAINDEX</td>
</tr>

<tr>
<td><a href="cutgextract.html">cutgextract</a></td>
<td>Extract codon usage tables from from CUTG database</td>
</tr>

<tr>
<td><a href="jaspextract.html">jaspextract</a></td>
<td>Extract data from JASPAR</td>
</tr>

<tr>
<td><a href="printsextract.html">printsextract</a></td>
<td>Extract data from PRINTS database for use by pscan</td>
</tr>

<tr>
<td><a href="prosextract.html">prosextract</a></td>
<td>Processes the PROSITE motif database for use by patmatmotifs</td>
</tr>

<tr>
<td><a href="tfextract.html">tfextract</a></td>
<td>Process TRANSFAC transcription factor database for use by tfscan</td>
</tr>

</table>


<H2>
    Author(s)
</H2>


Alan Bleasby (ajb&nbsp;&copy;&nbsp;ebi.ac.uk)
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK



<H2>
    History
</H2>

Completed 12th April 1999

<H2>
    Target users
</H2>
This program is intended to be used by administrators responsible
for software and database installation and maintenance. 


<H2>
    Comments
</H2>
None

</BODY>
</HTML>

