<HTML>

<HEAD>
  <TITLE>
  EMBOSS: rebaseextract
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">






<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

rebaseextract
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>






<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Extract data from REBASE


<H2>
    Description
</H2>


The Restriction Enzyme database (REBASE) is a collection of information
about restriction enzymes and related proteins.  It contains published
and unpublished references, recognition and cleavage sites,
isoschizomers, commercial availability, methylation sensitivity, crystal
and sequence data.  DNA methyltransferases, homing endonucleases,
nicking enzymes, specificity subunits and control proteins are also
included.  Most recently, putative DNA methyltransferases and
restriction enzymes, as predicted from analysis of genomic sequences,
are also listed. 

<p>

The home page of REBASE is:
<A HREF="http://rebase.neb.com/">http://rebase.neb.com/</A>

<p>

This program derives recognition site and cleavage information from the
"withrefm" file of an REBASE distribution.  It creates three files in
the EMBOSS data subdirectory REBASE.  A pattern file, a reference file
and a supplier file.  

<p>

It will also (by default) produce an 'embossre.equ' file.  This can be
turned off by setting the -equivalences option to be false.  This option
calculates an 'embossre.equ' file using restriction enzyme prototypes in
the "withrefm" file.  The 'embossre.equ' file is a file of preferred
isoschizomers.  You may edit it to contain your available restriction
enzymes. 

<p>

The EMBOSS programs that find restriction cutting sites use the data
files produced by this program and will not work without them. 

<p>

Running this program may be the job of your system manager. 

<H2>
    Usage
</H2>


<b>Here is a sample session with rebaseextract</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>rebaseextract </b>
Extract data from REBASE
Full pathname of WITHREFM file: <b>../../data/withrefm</b>
Full pathname of PROTO file: <b>../../data/proto</b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-infile]            infile     Full pathname of WITHREFM file
  [-protofile]         infile     Full pathname of PROTO file

   Additional (Optional) qualifiers:
   -[no]equivalences   boolean    This option calculates an embossre.equ file
                                  using restriction enzyme prototypes in the
                                  withrefm file.

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers: (none)
   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-infile]<br>(Parameter 1)</td>
<td>Full pathname of WITHREFM file</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-protofile]<br>(Parameter 2)</td>
<td>Full pathname of PROTO file</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]equivalences</td>
<td>This option calculates an embossre.equ file using restriction enzyme prototypes in the withrefm file.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>

The input file must be the "withrefm" file of a REBASE distribution.

<p>

For example, the withrefm file for REBASE version 005 is at: <a
href="ftp://ftp.neb.com/pub/rebase/withrefm.005">ftp://ftp.neb.com/pub/rebase/withrefm.005</a>

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: ../../data/withrefm</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
 
REBASE version 106                                              withrefm.106
 
    =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
    REBASE, The Restriction Enzyme Database   http://rebase.neb.com
    Copyright (c)  Dr. Richard J. Roberts, 2001.   All rights reserved.
    =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
 
Rich Roberts                                                    May 31 2001
 

&lt;ENZYME NAME&gt;   Restriction enzyme name.
&lt;ISOSCHIZOMERS&gt; Other enzymes with this specificity.
&lt;RECOGNITION SEQUENCE&gt; 
                These are written from 5' to 3', only one strand being given.
                If the point of cleavage has been determined, the precise site
                is marked with ^.  For enzymes such as HgaI, MboII etc., which
                cleave away from their recognition sequence the cleavage sites
                are indicated in parentheses.  

                For example HgaI GACGC (5/10) indicates cleavage as follows:
                                5' GACGCNNNNN^      3'
                                3' CTGCGNNNNNNNNNN^ 5'

                In all cases the recognition sequences are oriented so that
                the cleavage sites lie on their 3' side.

                REBASE Recognition sequences representations use the standard 
                abbreviations (Eur. J. Biochem. 150: 1-5, 1985) to represent 
                ambiguity.
                                R = G or A
                                Y = C or T
                                M = A or C
                                K = G or T
                                S = G or C
                                W = A or T
                                B = not A (C or G or T)
                                D = not C (A or G or T)
                                H = not G (A or C or T)
                                V = not T (A or C or G)
                                N = A or C or G or T



                ENZYMES WITH UNUSUAL CLEAVAGE PROPERTIES:  

                Enzymes that cut on both sides of their recognition sequences,
                such as BcgI, Bsp24I, CjeI and CjePI, have 4 cleavage sites
                each instead of 2.



<font color=red>  [Part of this file has been deleted for brevity]</font>

&lt;6&gt;S.A. Thompson
&lt;7&gt;N
&lt;8&gt;Morgan, R.D., Unpublished observations.
Morgan, R.D., Xu, Q., US Patent Office, 2001.
Xu, Q., Morgan, R., Blaser, M., Unpublished observations.

&lt;1&gt;HspAI
&lt;2&gt;HhaI,AspLEI,BcaI,BspLAI,BstHHI,CcoP95I,CfoI,Csp1470I,FnuDIII,Hin6I,Hin7I,HinGUI,HinP1I,HinS1I,HinS2I,Hpy99III,HpyF10I,HsoI,MnnIV,NgoEII,SciNI
&lt;3&gt;G^CGC
&lt;4&gt;
&lt;5&gt;Haemophilus species A
&lt;6&gt;S.K. Degtyarev
&lt;7&gt;I
&lt;8&gt;Rechkunova, N.I., Prikhod'ko, E.A., Shevchenko, A.V., Degtyarev, S.K., Unpublished observations.

&lt;1&gt;KpnI
&lt;2&gt;Acc65I,AhaB8I,Asp718I,BspJ106I,Eco149I,Esp19I,KpnK14I,MvsI,MvsAI,MvsBI,MvsCI,MvsDI,MvsEI,NmiI,Sau10I,SthI,SthAI,SthBI,SthCI,SthDI,SthEI,SthFI,SthGI,SthHI,SthJI,SthKI,SthLI,SthMI,SthNI,Uba76I,Uba85I,Uba86I,Uba87I,Uba1201I
&lt;3&gt;GGTAC^C
&lt;4&gt;4(6)
&lt;5&gt;Klebsiella pneumoniae OK8
&lt;6&gt;ATCC 49790
&lt;7&gt;ABCDEFGHIJKLMNOQRSTU
&lt;8&gt;Kiss, A., Finta, C., Venetianer, P., (1991) Nucleic Acids Res., vol. 19, pp. 3460.
Smith, D.I., Blattner, F.R., Davies, J., (1976) Nucleic Acids Res., vol. 3, pp. 343-353.
Tomassini, J., Roychoudhury, R., Wu, R., Roberts, R.J., (1978) Nucleic Acids Res., vol. 5, pp. 4055-4064.

&lt;1&gt;NotI
&lt;2&gt;CciNI,CspBI,MchAI
&lt;3&gt;GC^GGCCGC
&lt;4&gt;?(4)
&lt;5&gt;Nocardia otitidis-caviarum
&lt;6&gt;ATCC 14630
&lt;7&gt;ABCDEFGHJKLMNOQRSTU
&lt;8&gt;Borsetti, R., Wise, D., Qiang, B.-Q., Schildkraut, I., Unpublished observations.
Morgan, R.D., Unpublished observations.
Morgan, R.D., Benner, J.S., Claus, T.E., US Patent Office, 1994.
Qiang, B.-Q., Schildkraut, I., (1987) Methods Enzymol., vol. 155, pp. 15-21.

&lt;1&gt;TaqI
&lt;2&gt;CviSIII,EsaBC3I,HpyV,Hpy26II,HpyF14III,HpyF16I,HpyF23I,HpyF24I,HpyF26III,HpyF30I,HpyF35I,HpyF40II,HpyF42IV,HpyF45I,HpyF49I,HpyF52I,HpyF59III,HpyF62II,HpyF64I,HpyF65II,HpyF66IV,HpyF71I,HpyF73II,HpyJP26II,PpaAII,Taq20I,Tbr51I,TfiA3I,TfiTok4A2I,TfiTok6A1I,TflI,Tsc4aI,Tsp32I,Tsp32II,Tsp358I,Tsp505I,Tsp510I,TspAK13D21I,TspAK16D24I,TspNI,TspVi4AI,TspVil3I,Tth24I,TthHB8I,TthRQI
&lt;3&gt;T^CGA
&lt;4&gt;4(6)
&lt;5&gt;Thermus aquaticus YTI
&lt;6&gt;J.I. Harris
&lt;7&gt;ABCDEFGIJLMNOQRSTU
&lt;8&gt;Anton, B.P., Brooks, J.E., Unpublished observations.
Fomenkov, A., Xiao, J.-P., Dila, D., Raleigh, E., Xu, S.-Y., (1994) Nucleic Acids Res., vol. 22, pp. 2399-2403.
McClelland, M., (1981) Nucleic Acids Res., vol. 9, pp. 6795-6804.
Sato, S., Hutchison, C.A. III, Harris, J.I., (1977) Proc. Natl. Acad. Sci. U. S. A., vol. 74, pp. 542-546.
Zebala, J.A., (1993) Diss. Abstr., vol. 54, pp. 1394-1398.

</pre>
</td></tr></table><p>
<p><h3>File: ../../data/proto</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
 
REBASE version 305                                              proto.305
 
    =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
    REBASE, The Restriction Enzyme Database   http://rebase.neb.com
    Copyright (c)  Dr. Richard J. Roberts, 2003.   All rights reserved.
    =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
 
Rich Roberts                                                    Apr 30 2003
 



	    TYPE II ENZYMES
	    ---------------

BsiYI                          CCNNNNN^NNGG
BsrI                           ACTGG (1/-1)
HaeIII                         GG^CC
HpaII                          C^CGG
Ksp632I                        CTCTTC (1/4)
MaeII                          A^CGT



	    TYPE I ENZYMES
	    ---------------

EcoAI                          GAGNNNNNNNGTCA
EcoBI                          TGANNNNNNNNTGCT
EcoDI                          TTANNNNNNNGTCY
EcoDR2                         TCANNNNNNGTCG
EcoDR3                         TCANNNNNNNATCG
EcoDXXI                        TCANNNNNNNRTTC
EcoEI                          GAGNNNNNNNATGC
EcoKI                          AACNNNNNNGTGC



	    TYPE III ENZYMES
	    ---------------

EcoPI                          AGACC
EcoP15I                        CAGCAG (25/27)
HinfIII                        CGAAT
StyLTI                         CAGAG

</pre>
</td></tr></table><p>


<H2>
    Output file format
</H2>



<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: REBASE</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
</pre>
</td></tr></table><p>
<p><h3>File: embossre.equ</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Bsc4I BsiYI
Bse1I BsrI
BshI HaeIII
BsiSI HpaII
Bsu6I Ksp632I
HpyCH4IV MaeII
</pre>
</td></tr></table><p>


<p>

The output files are held in the REBASE subdirectory of the EMBOSS data
directory. There are three:


<ul>
	<li> embossre.enz		Enzyme pattern file
	<li> embossre.ref		Enzyme references
	<li> embossre.sup		Enzyme suppliers
</ul>

<p>

<b>rebaseextract</b> will also (by default) produce an 'embossre.equ'
file in the EMBOSS data directory.  This can be turned off by setting
the -equivalences option to be false.  This option calculates an
'embossre.equ' file using restriction enzyme prototypes in the
"withrefm" file.  The 'embossre.equ' file is a file of preferred
isoschizomers.  You may edit it to contain your available restriction
enzymes. 

<p>



<H2>
    Data files
</H2>


The "withrefm" file of an REBASE distribution is the input file for this
program.

<H2>
    Notes
</H2>


The home page of REBASE is:
<A HREF="http://rebase.neb.com/">http://rebase.neb.com/</A>

<p>
Running this program may be the job of your system manager. 

<p>
The ready-made files produced by this program may already be available
at the REBASE web site:
<a href="http://rebase.neb.com/rebase/rebase.files.html">http://rebase.neb.com/rebase/rebase.files.html</a>
or
<a href="http://rebase.neb.com/rebase/rebase.f37.html">http://rebase.neb.com/rebase/rebase.f37.html</a>

<H2>
    References
</H2>


<ol>

<li>
Nucleic Acids Research 27: 312-313 (1999).

</ol>
<H2>
    Warnings
</H2>


The program will warn you if the input file is incorrectly formatted.

<H2>
    Diagnostic Error Messages
</H2>


<H2>
    Exit status
</H2>


It exits with status 0 unless an error is reported.

<H2>
    Known bugs
</H2>


<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="aaindexextract.html">aaindexextract</a></td><td>Extract data from AAINDEX</td></tr>
<tr><td><a href="cutgextract.html">cutgextract</a></td><td>Extract data from CUTG</td></tr>
<tr><td><a href="printsextract.html">printsextract</a></td><td>Extract data from PRINTS</td></tr>
<tr><td><a href="prosextract.html">prosextract</a></td><td>Builds the PROSITE motif database for patmatmotifs to search</td></tr>
<tr><td><a href="tfextract.html">tfextract</a></td><td>Extract data from TRANSFAC</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


Alan Bleasby (ableasby&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK




<H2>
    History
</H2>


Completed 12th April 1999

<!--
<p>
[an error occurred while processing this directive]
-->

<H2>
    Target users
</H2>


This program is intended to be used by administrators responsible
for software and database installation and maintenance. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

