<HTML>
<HEAD>
  <TITLE>
  EMBOSS: recoder
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
recoder
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
    Function
</H2>
Remove restriction sites but maintain same translation
<H2>
    Description
</H2>

<b>recoder</b> scans a given nucleotide sequence for restriction
sites. It reports single base positions in the restriction pattern which
when mutated remove the restriction site whilst maintaining the same
translation (in frame 1 of the input sequence).

<p>

Several restriction enzymes can be specified or
alternatively all the enzymes in the REBASE database can be
investigated. To find out whether the single point mutations found by
'recoder', introduce new restriction sites, 'silent' should be run on the
original sequence. ('Silent' searches for silent point mutation sites
which maintain the same translation. 

<p>

The output for 'recoder' is similar to the format used by 'silent'.


<H2>
    Usage
</H2>
<b>Here is a sample session with recoder</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>recoder </b>
Remove restriction sites but maintain same translation
Input nucleotide sequence: <b>tembl:x65923</b>
Comma separated enzyme list [all]: <b>EcoRII</b>
Output report [x65923.recoder]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-sequence]          sequence   Nucleotide sequence filename and optional
                                  format, or reference (input USA)
   -enzymes            string     [all] Comma separated enzyme list (Any
                                  string is accepted)
  [-outfile]           report     [*.recoder] Output report file name

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers:
   -sshow              boolean    [N] Display untranslated sequence
   -tshow              boolean    [N] Display translated sequence

   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1            integer    Start of the sequence to be used
   -send1              integer    End of the sequence to be used
   -sreverse1          boolean    Reverse (if DNA)
   -sask1              boolean    Ask for begin/end/reverse
   -snucleotide1       boolean    Sequence is nucleotide
   -sprotein1          boolean    Sequence is protein
   -slower1            boolean    Make lower case
   -supper1            boolean    Make upper case
   -sformat1           string     Input sequence format
   -sdbname1           string     Database name
   -sid1               string     Entryname
   -ufo1               string     UFO features
   -fformat1           string     Features format
   -fopenfile1         string     Features file name

   "-outfile" associated qualifiers
   -rformat2           string     Report format
   -rname2             string     Base file name
   -rextension2        string     File name extension
   -rdirectory2        string     Output directory
   -raccshow2          boolean    Show accession number in the report
   -rdesshow2          boolean    Show description in the report
   -rscoreshow2        boolean    Show the score in the report
   -rusashow2          boolean    Show the full USA in the report
   -rmaxall2           integer    Maximum total hits to report
   -rmaxseq2           integer    Maximum hits to report for one sequence

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Nucleotide sequence filename and optional format, or reference (input USA)</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-enzymes</td>
<td>Comma separated enzyme list</td>
<td>Any string is accepted</td>
<td>all</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output report file name</td>
<td>Report output file</td>
<td><i>&lt;*&gt;</i>.recoder</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-sshow</td>
<td>Display untranslated sequence</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-tshow</td>
<td>Display translated sequence</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

</table>

<H2>
    Input file format
</H2>

<b>recoder</b> reads a normal nucleic acid sequence USA.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:x65923' is a sequence entry in the example nucleic acid database 'tembl'
<p>
<p><h3>Database entry: tembl:x65923</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   X65923; SV 1; linear; mRNA; STD; HUM; 518 BP.
XX
AC   X65923;
XX
DT   13-MAY-1992 (Rel. 31, Created)
DT   18-APR-2005 (Rel. 83, Last updated, Version 11)
XX
DE   H.sapiens fau mRNA
XX
KW   fau gene.
XX
OS   Homo sapiens (human)
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia;
OC   Eutheria; Euarchontoglires; Primates; Haplorrhini; Catarrhini; Hominidae;
OC   Homo.
XX
RN   [1]
RP   1-518
RA   Michiels L.M.R.;
RT   ;
RL   Submitted (29-APR-1992) to the EMBL/GenBank/DDBJ databases.
RL   L.M.R. Michiels, University of Antwerp, Dept of Biochemistry,
RL   Universiteisplein 1, 2610 Wilrijk, BELGIUM
XX
RN   [2]
RP   1-518
RX   PUBMED; 8395683.
RA   Michiels L., Van der Rauwelaert E., Van Hasselt F., Kas K., Merregaert J.;
RT   " fau cDNA encodes a ubiquitin-like-S30 fusion protein and is expressed as
RT   an antisense sequences in the Finkel-Biskis-Reilly murine sarcoma virus";
RL   Oncogene 8(9):2537-2546(1993).
XX
DR   H-InvDB; HIT000322806.
XX
FH   Key             Location/Qualifiers
FH
FT   source          1..518
FT                   /organism="Homo sapiens"
FT                   /chromosome="11q"
FT                   /map="13"
FT                   /mol_type="mRNA"
FT                   /clone_lib="cDNA"
FT                   /clone="pUIA 631"
FT                   /tissue_type="placenta"
FT                   /db_xref="taxon:9606"
FT   misc_feature    57..278
FT                   /note="ubiquitin like part"
FT   CDS             57..458
FT                   /gene="fau"
FT                   /db_xref="GDB:135476"
FT                   /db_xref="GOA:P35544"
FT                   /db_xref="GOA:P62861"
FT                   /db_xref="HGNC:3597"
FT                   /db_xref="UniProtKB/Swiss-Prot:P35544"
FT                   /db_xref="UniProtKB/Swiss-Prot:P62861"
FT                   /protein_id="CAA46716.1"
FT                   /translation="MQLFVRAQELHTFEVTGQETVAQIKAHVASLEGIAPEDQVVLLAG
FT                   APLEDEATLGQCGVEALTTLEVAGRMLGGKVHGSLARAGKVRGQTPKVAKQEKKKKKTG
FT                   RAKRRMQYNRRFVNVVPTFGKKKGPNANS"
FT   misc_feature    98..102
FT                   /note="nucleolar localization signal"
FT   misc_feature    279..458
FT                   /note="S30 part"
FT   polyA_signal    484..489
FT   polyA_site      509
XX
SQ   Sequence 518 BP; 125 A; 139 C; 148 G; 106 T; 0 other;
     ttcctctttc tcgactccat cttcgcggta gctgggaccg ccgttcagtc gccaatatgc        60
     agctctttgt ccgcgcccag gagctacaca ccttcgaggt gaccggccag gaaacggtcg       120
     cccagatcaa ggctcatgta gcctcactgg agggcattgc cccggaagat caagtcgtgc       180
     tcctggcagg cgcgcccctg gaggatgagg ccactctggg ccagtgcggg gtggaggccc       240
     tgactaccct ggaagtagca ggccgcatgc ttggaggtaa agttcatggt tccctggccc       300
     gtgctggaaa agtgagaggt cagactccta aggtggccaa acaggagaag aagaagaaga       360
     agacaggtcg ggctaagcgg cggatgcagt acaaccggcg ctttgtcaac gttgtgccca       420
     cctttggcaa gaagaagggc cccaatgcca actcttaagt cttttgtaat tctggctttc       480
     tctaataaaa aagccactta gttcagtcaa aaaaaaaa                               518
//
</pre>
</td></tr></table><p>




<H2>
    Output file format
</H2>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: x65923.recoder</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
########################################
# Program: recoder
# Rundate: Sun 15 Jul 2007 12:00:00
# Commandline: recoder
#    -sequence tembl:x65923
#    -enzymes EcoRII
# Report_format: table
# Report_file: x65923.recoder
########################################

#=======================================
#
# Sequence: X65923     from: 1   to: 518
# HitCount: 34
#
# KEY:
# Dir: Direction (Rev for reverse complement)
# EnzymeName: Enzyme name
# RS-Pattern: Restriction enzyme recognition site pattern
# Base-Posn: Position of base to be mutated
# AAs: Amino acid. Original sequence(.)After mutation
# Mutation: The base mutation to perform
# 
# Creating silent mutations
#
#=======================================

  Start     End Dir    EnzymeName RS-Pattern Base-Posn    AAs Mutation
     77      81 .      EcoRII     CCWGG             78    P.P     C-&gt;G
     77      81 .      EcoRII     CCWGG             78    P.P     C-&gt;A
     77      81 .      EcoRII     CCWGG             78    P.P     C-&gt;T
     77      81 .      EcoRII     CCWGG             79    R.R     A-&gt;C
     77      81 .      EcoRII     CCWGG             81    R.R     G-&gt;A
    107     111 .      EcoRII     CCWGG            108    A.A     C-&gt;G
    107     111 .      EcoRII     CCWGG            108    A.A     C-&gt;A
    107     111 .      EcoRII     CCWGG            108    A.A     C-&gt;T
    107     111 .      EcoRII     CCWGG            109    R.R     A-&gt;C
    107     111 .      EcoRII     CCWGG            111    R.R     G-&gt;A
    182     186 .      EcoRII     CCWGG            183    S.S     C-&gt;G
    182     186 .      EcoRII     CCWGG            183    S.S     C-&gt;A
    182     186 .      EcoRII     CCWGG            183    S.S     C-&gt;T
    197     201 .      EcoRII     CCWGG            198    P.P     C-&gt;G
    197     201 .      EcoRII     CCWGG            198    P.P     C-&gt;A
    197     201 .      EcoRII     CCWGG            198    P.P     C-&gt;T
    248     252 .      EcoRII     CCWGG            249    P.P     C-&gt;G
    248     252 .      EcoRII     CCWGG            249    P.P     C-&gt;A
    248     252 .      EcoRII     CCWGG            249    P.P     C-&gt;T
    293     297 .      EcoRII     CCWGG            294    P.P     C-&gt;G
    293     297 .      EcoRII     CCWGG            294    P.P     C-&gt;A
    293     297 .      EcoRII     CCWGG            294    P.P     C-&gt;T
     81      77 Rev    EcoRII     CCWGG             79    P.P     T-&gt;G
     81      77 Rev    EcoRII     CCWGG             79    P.P     T-&gt;C
    111     107 Rev    EcoRII     CCWGG            109    P.P     T-&gt;G
    111     107 Rev    EcoRII     CCWGG            109    P.P     T-&gt;C
    186     182 Rev    EcoRII     CCWGG            184    P.P     A-&gt;G
    186     182 Rev    EcoRII     CCWGG            184    P.P     A-&gt;C
    201     197 Rev    EcoRII     CCWGG            199    P.P     A-&gt;G
    201     197 Rev    EcoRII     CCWGG            199    P.P     A-&gt;C
    252     248 Rev    EcoRII     CCWGG            250    P.P     A-&gt;G
    252     248 Rev    EcoRII     CCWGG            250    P.P     A-&gt;C
    297     293 Rev    EcoRII     CCWGG            295    P.P     A-&gt;G
    297     293 Rev    EcoRII     CCWGG            295    P.P     A-&gt;C

#---------------------------------------
#---------------------------------------

#---------------------------------------
# Total_sequences: 1
# Total_hitcount: 34
#---------------------------------------
</pre>
</td></tr></table><p>



<H2>
    Data files
</H2>

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>
<p>
The EMBOSS REBASE restriction enzyme data files are stored iin
directory 'data/REBASE/*' under the EMBOSS installation directory.

<p>

These files must first be set up using the program <a
href="rebaseextract.html">'<b>rebaseextract</b>'</a>.  Running
'rebaseextract' may be the job of your system manager.

<p>
      
The data files are stored in the REBASE directory of the standard EMBOSS
data directory. The names are:

<ul>
<li> embossre.enz     Cleavage information
<li> embossre.ref     Reference/methylation information
<li> embossre.sup     Supplier information
</ul>
	 
The column information is described at the top of the data files
	 
<p>
	 
The reported enzyme from any one group of isoschizomers (the prototype)
is specified in the REBASE database and the information is held in the
data file 'embossre.equ'.  You may edit this file to set your own
preferred prototype, if you wish. 
	 
<p>
	 
The format of the file "embossre.equ" is
<br>
Enzyme-name Prototype-name
	 
<p>
	 
i.e.  two columns of enzyme names separated by a space.  The first name
of the pair of enzymes is the name that is not preferred and the second
is the preferred (prototype) name. 
	 
	          

<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with status 0.

<H2>
    Known bugs
</H2>

None.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr>
<td><a href="redata.html">redata</a></td>
<td>Search REBASE for enzyme name, references, suppliers etc</td>
</tr>

<tr>
<td><a href="remap.html">remap</a></td>
<td>Display sequence with restriction sites, translation etc</td>
</tr>

<tr>
<td><a href="restover.html">restover</a></td>
<td>Find restriction enzymes producing specific overhang</td>
</tr>

<tr>
<td><a href="restrict.html">restrict</a></td>
<td>Finds restriction enzyme cleavage sites</td>
</tr>

<tr>
<td><a href="showseq.html">showseq</a></td>
<td>Display a sequence with features, translation etc</td>
</tr>

<tr>
<td><a href="silent.html">silent</a></td>
<td>Silent mutation restriction enzyme scan</td>
</tr>

</table>
<p>

<b>silent</b> does the opposite to <b>recoder</b>.  <b>silent</b> finds
sites where a restriction enzyme site can be introduced without changing
the translation in frame 1 of the sequence.  <b>recoder</b> finds sites
where a restriction enzyme site can be removed without changing the
translation in frame 1 of the sequence. 

<H2>
    Author(s)
</H2>

Tim Carver (tcarver&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK

<H2>
    History
</H2>

Written (January 2001) - Tim Carver
<p>

Renamed from <b>recode</b> to <b>recoder</b> 16 May 2001 as the old name
clashed with a common UNIX print utility:
<br>
<A HREF="http://www.iro.umontreal.ca/contrib/recode/HTML/">http://www.iro.umontreal.ca/contrib/recode/HTML/</a>

<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.


<H2>
    Comments
</H2>
None

</BODY>
</HTML>

