<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: recode
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">
<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>
<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>
</td>
<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: recode
</H2>
</font>
</td>
</tr>
</table>
</td>
</tr>
</table>
<p><hr><p>

<!--END OF HEADER-->

<H1>
Program recode
</H1>
<H2>
    Function
</H2>
Remove restriction sites but maintain the same translation

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
        Algorithms
  -->

<b>recode</b> scans a given nucleotide sequence for restriction
sites. It reports single base positions in the restriction pattern which
when mutated remove the restriction site whilst maintaining the same
translation (in frame 1 of the input sequence).

<p>

Several restriction enzymes can be specified or
alternatively all the enzymes in the REBASE database can be
investigated. To find out whether the single point mutations found by
'recode', introduce new restriction sites, 'silent' should be run on the
original sequence. ('Silent' searches for silent point mutation sites
which maintain the same translation. 

<p>

The output for 'recode' is similar to the format used by 'silent'.















<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->


Here is a sample session with recode:

<pre>

% recode
Find and remove restriction sites but maintain the same translation
Input sequence: em:hsfau
Comma separated enzyme list [all]: EcoRII
Output file [hsfau.recode]:
</pre>










<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-seq]               sequence   Nucleic acid sequence
   -enzymes            string     Comma separated enzyme list
  [-outf]              outfile    Results file name

   Optional qualifiers: (none)
   Advanced qualifiers:
   -sshow              bool       Display untranslated sequence
   -tshow              bool       Display translated sequence


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-seq]<br>(Parameter 1)</td>
<td>Nucleic acid sequence</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-enzymes</td>
<td>Comma separated enzyme list</td>
<td>Any string is accepted</td>
<td>all</td>
</tr>

<tr>
<td>[-outf]<br>(Parameter 2)</td>
<td>Results file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.recode</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-sshow</td>
<td>Display untranslated sequence</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-tshow</td>
<td>Display translated sequence</td>
<td>Yes/No</td>
<td>No</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

Normal nucleic acid USA.










<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->


This is the out[ut from the above example:
<p>
<hr>
<pre>
Results for HSFAU:

KEY:
        Enzyme          Enzyme name
        RS-Pattern      Restriction enzyme recognition site pattern
        Match-Posn      Position of the first base of RS pattern in sequence
        AA              Amino acid. Original sequence(.)After mutation
        Base-Posn       Position of base to be mutated in sequence
        Mutation        The base mutation to perform

Creating silent mutations

Enzyme      RS-Pattern  Match-Posn   AA  Base-Posn Mutation
EcoRII      CCWGG          77        P.P    78       C->G  
EcoRII      CCWGG          77        P.P    78       C->A  
EcoRII      CCWGG          77        P.P    78       C->T  
EcoRII      CCWGG          77        R.R    79       A->C  
EcoRII      CCWGG          77        R.R    81       G->A  
EcoRII      CCWGG          107       A.A    108      C->G  
EcoRII      CCWGG          107       A.A    108      C->A  
EcoRII      CCWGG          107       A.A    108      C->T  
EcoRII      CCWGG          107       R.R    109      A->C  
EcoRII      CCWGG          107       R.R    111      G->A  
EcoRII      CCWGG          182       S.S    183      C->G  
EcoRII      CCWGG          182       S.S    183      C->A  
EcoRII      CCWGG          182       S.S    183      C->T  
EcoRII      CCWGG          197       P.P    198      C->G  
EcoRII      CCWGG          197       P.P    198      C->A  
EcoRII      CCWGG          197       P.P    198      C->T  
EcoRII      CCWGG          248       P.P    249      C->G  
EcoRII      CCWGG          248       P.P    249      C->A  
EcoRII      CCWGG          248       P.P    249      C->T  
EcoRII      CCWGG          293       P.P    294      C->G  
EcoRII      CCWGG          293       P.P    294      C->A  
EcoRII      CCWGG          293       P.P    294      C->T  



Results for reverse of HSFAU:

Creating silent mutations

Enzyme      RS-Pattern  Match-Posn   AA  Base-Posn Mutation
EcoRII      CCWGG          77        P.P    79       T->G  
EcoRII      CCWGG          77        P.P    79       T->C  
EcoRII      CCWGG          107       P.P    109      T->G  
EcoRII      CCWGG          107       P.P    109      T->C  
EcoRII      CCWGG          182       P.P    184      A->G  
EcoRII      CCWGG          182       P.P    184      A->C  
EcoRII      CCWGG          197       P.P    199      A->G  
EcoRII      CCWGG          197       P.P    199      A->C  
EcoRII      CCWGG          248       P.P    250      A->G  
EcoRII      CCWGG          248       P.P    250      A->C  
EcoRII      CCWGG          293       P.P    295      A->C  
EcoRII      CCWGG          293       P.P    295      A->G  

</pre>
<hr>
<p>


















<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.
	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

   -->

None.













<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="chaos.html">chaos</a></td><td>Create a chaos game representation plot for a sequence</td></tr>
<tr><td><a href="cpgplot.html">cpgplot</a></td><td>Plot CpG rich areas</td></tr>
<tr><td><a href="cpgreport.html">cpgreport</a></td><td>Reports all CpG rich regions</td></tr>
<tr><td><a href="diffseq.html">diffseq</a></td><td>Find differences (SNPs) between nearly identical sequences</td></tr>
<tr><td><a href="dotmatcher.html">dotmatcher</a></td><td>Displays a thresholded dotplot of two sequences</td></tr>
<tr><td><a href="dotpath.html">dotpath</a></td><td>Displays a non-overlapping wordmatch dotplot of two sequences</td></tr>
<tr><td><a href="dottup.html">dottup</a></td><td>Displays a wordmatch dotplot of two sequences</td></tr>
<tr><td><a href="einverted.html">einverted</a></td><td>Finds DNA inverted repeats</td></tr>
<tr><td><a href="equicktandem.html">equicktandem</a></td><td>Finds tandem repeats</td></tr>
<tr><td><a href="etandem.html">etandem</a></td><td>Looks for tandem repeats in a nucleotide sequence</td></tr>
<tr><td><a href="isochore.html">isochore</a></td><td>Plots isochores in large DNA sequences</td></tr>
<tr><td><a href="marscan.html">marscan</a></td><td>Finds MAR/SAR sites in nucleic sequences</td></tr>
<tr><td><a href="newcpgreport.html">newcpgreport</a></td><td>Report CpG rich areas</td></tr>
<tr><td><a href="newcpgseek.html">newcpgseek</a></td><td>Reports CpG rich regions</td></tr>
<tr><td><a href="palindrome.html">palindrome</a></td><td>Looks for inverted repeats in a nucleotide sequence</td></tr>
<tr><td><a href="polydot.html">polydot</a></td><td>Displays all-against-all dotplots of a set of sequences</td></tr>
<tr><td><a href="rebaseextract.html">rebaseextract</a></td><td>Extract data from REBASE</td></tr>
<tr><td><a href="redata.html">redata</a></td><td>Search REBASE for enzyme name, references, suppliers etc</td></tr>
<tr><td><a href="remap.html">remap</a></td><td>Display a sequence with restriction cut sites, translation etc</td></tr>
<tr><td><a href="restover.html">restover</a></td><td>Finds restriction enzymes that produce a specific overhang</td></tr>
<tr><td><a href="restrict.html">restrict</a></td><td>Finds restriction enzyme cleavage sites</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="silent.html">silent</a></td><td>Silent mutation restriction enzyme scan</td></tr>
<tr><td><a href="tfscan.html">tfscan</a></td><td>Scans DNA sequences for transcription factors</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->

<p>

<b>silent</b> does the opposite to <b>recode</b>.  <b>silent</b> finds
sites where a restriction enzyme site can be introduced without changing
the translation in frame 1 of the sequence.  <b>recode</b> finds sites
where a restriction enzyme site can be removed without changing the
translation in frame 1 of the sequence. 







<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

This application was written by Tim Carver <a
href="mailto:tcarver@hgmp.mrc.ac.uk">(tcarver@hgmp.mrc.ac.uk)</a>















<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (January 2001) - Tim Carver













<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->








</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

