<HTML>

<HEAD>
  <TITLE>
  EMBOSS: redata
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

redata
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<H2>
    Function
</H2>
Search REBASE for enzyme name, references, suppliers etc


<H2>
    Description
</H2>


The Restriction Enzyme database (REBASE) is a collection of information about
restriction enzymes and related proteins. It contains published and
unpublished references, recognition and cleavage sites, isoschizomers,
commercial availability, methylation sensitivity, crystal and sequence data.
DNA methyltransferases, homing endonucleases, nicking enzymes, specificity
subunits and control proteins are also included. Most recently, putative DNA
methyltransferases and restriction enzymes, as predicted from analysis of
genomic sequences, are also listed. 

<p>
The home page of REBASE is: <a
href="http://rebase.neb.com/">http://rebase.neb.com/</a>

<p>
This program searches the REBASE database for information on a specified
restriction enzyme.  

<p>
It outputs a report including the cut site, isoschizomers, references
and commercial suppliers of the enzyme. 

<H2>
    Usage
</H2>
<b>Here is a sample session with redata</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>redata </b>
Search REBASE for enzyme name, references, suppliers etc.
Restriction enzyme name [BamHI]: <b>BamHI</b>
Output file [outfile.redata]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.1">Go to the output files for this example</a><p><p>


<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-enzyme]            string     Enter the name of the restriction enzyme
                                  that you wish to get details of. The names
                                  often have a 'I' in them - this is a capital
                                  'i', not a '1' or an 'l'. The names are
                                  case-independent ('AaeI' is the same as
                                  'aaei')
  [-outfile]           outfile    Output file name

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers:
   -[no]isoschizomers  boolean    Show other enzymes with this specificity.
                                  (Isoschizomers)
   -[no]references     boolean    Show references
   -[no]suppliers      boolean    Show suppliers

   Associated qualifiers:

   "-outfile" associated qualifiers
   -odirectory2         string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-enzyme]<br>(Parameter 1)</td>
<td>Enter the name of the restriction enzyme that you wish to get details of. The names often have a 'I' in them - this is a capital 'i', not a '1' or an 'l'. The names are case-independent ('AaeI' is the same as 'aaei')</td>
<td>Any string is accepted</td>
<td>BamHI</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.redata</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]isoschizomers</td>
<td>Show other enzymes with this specificity. (Isoschizomers)</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]references</td>
<td>Show references</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]suppliers</td>
<td>Show suppliers</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

</table>



<H2>
    Input file format
</H2>




<H2>
    Output file format
</H2>



<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: outfile.redata</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
BamHI

Recognition site is GGATCC leaving sticky ends
  Cut positions 5':1 3':5
Organism: Bacillus amyloliquefaciens H
Methylated: 5(4)
Source: ATCC 49763

Isoschizomers:
   AacI        AaeI        AcaII       AccEBI      AinII       AliI        
   Ali12257I   Ali12258I   ApaCI       AsiI        AspTII      Atu1II      
   BamFI       BamKI       BamNI       Bca1259I    Bce751I     Bco10278I   
   BnaI        BsaDI       Bsp30I      Bsp46I      Bsp90II     Bsp98I      
   Bsp130I     Bsp131I     Bsp144I     Bsp4009I    BspAAIII    BstI        
   Bst1126I    Bst2464I    Bst2902I    BstQI       Bsu90I      Bsu8565I    
   Bsu8646I    BsuB519I    BsuB763I    CelI        DdsI        GdoI        
   GinI        GoxI        GseIII      MleI        Mlu23I      NasBI       
   Nsp29132II  NspSAIV     OkrAI       Pac1110I    Pae177I     Pfl8I       
   Psp56I      RhsI        Rlu4I       RspLKII     SolI        SpvI        
   SurI        Uba19I      Uba31I      Uba38I      Uba51I      Uba88I      
   Uba1098I    Uba1163I    Uba1167I    Uba1172I    Uba1173I    Uba1205I    
   Uba1224I    Uba1242I    Uba1250I    Uba1258I    Uba1297I    Uba1302I    
   Uba1324I    Uba1325I    Uba1334I    Uba1339I    Uba1346I    Uba1383I    
   Uba1398I    Uba1402I    Uba1414I    

Suppliers:
Amersham Pharmacia Biotech (1/01)
Life Technologies Inc. (1/98)
Minotech, Molecular Biology Products (12/00)
HYBAID GmbH (12/00)
Stratagene (11/00)
Fermentas AB (5/01)
Q-BIOgene (1/01)
American Allied Biochemical, Inc. (10/98)
SibEnzyme Ltd. (1/01)
Nippon Gene Co., Ltd. (6/00)
Takara Shuzo Co. Ltd. (2/01)
Transgenomic Ltd. (1/01)
Roche Molecular Biochemicals (1/01)
New England BioLabs (12/00)
Toyobo Biochemicals (11/98)
CHIMERx (10/97)
Promega Corporation (6/99)
Sigma Chemical Corporation (11/98)
Advanced Biotechnologies Ltd. (3/98)
Bangalore Genei (2/01)
MRC-Holland (3/01)

References:
Brooks, J.E., Howard, K.A., US Patent Office, 1994.
Brooks, J.E., Nathan, P.D., Landry, D., Sznyter, L.A., Waite-Rees, P., Ives, C.L., Moran, L.S., Slatko, B.E., Benner, J.S., (1991) Nucleic Acids Res., vol. 19, pp. 841-850.
Hattman, S., Keister, T., Gottehrer, A., (1978) J. Mol. Biol., vol. 124, pp. 701-711.
Lee, S.P., Porter, D., Chirikjian, J.G., Knutson, J.R., Han, M.K., (1994) Anal. Biochem., vol. 220, pp. 377-383.
Newman, M., Strzelecka, T., Dorner, L.F., Schildkraut, I., Aggarwal, A.K., (1994) Structure, vol. 2, pp. 439-452.
Roberts, R.J., Wilson, G.A., Young, F.E., (1977) Nature, vol. 265, pp. 82-84.
Roy, K.B., Vrushank, D., Jayaram, B., (1994) Anal. Biochem., vol. 220, pp. 160-164.
Strzelecka, T., Newman, M., Dorner, L.F., Knott, R., Schildkraut, I., Aggarwal, A.K., (1994) J. Mol. Biol., vol. 239, pp. 430-432.
Wilson, G.A., Young, F.E., (1975) J. Mol. Biol., vol. 97, pp. 123-125.
Xu, S.-Y., Fomenkov, A., (1994) Biotechniques, vol. 17, pp. 57.
</pre>
</td></tr></table><p>


<p>

This includes the full list of isoschizomers (enzymes with the same
target sequence), suppliers and references from the REBASE database. 

<H2>
    Data files
</H2>

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

<p>
The EMBOSS REBASE restriction enzyme data files are stored iin
directory 'data/REBASE/*' under the EMBOSS installation directory.

<p>

These files must first be set up using the program <a
href="rebaseextract.html">'<b>rebaseextract</b>'</a>.  Running
'rebaseextract' may be the job of your system manager.

<p>
      
The data files are stored in the REBASE directory of the standard EMBOSS
data directory. The names are:

<ul>
<li> embossre.enz     Cleavage information
<li> embossre.ref     Reference/methylation information
<li> embossre.sup     Supplier information
</ul>
	 
The column information is described at the top of the data files
	 
<p>
	 
The reported enzyme from any one group of isoschizomers (the prototype)
is specified in the REBASE database and the information is held in the
data file 'embossre.equ'.  You may edit this file to set your own
preferred prototype, if you wish. 
	 
<p>
	 
The format of the file "embossre.equ" is
<br>
Enzyme-name Prototype-name
	 
<p>
	 
i.e.  two columns of enzyme names separated by a space.  The first name
of the pair of enzymes is the name that is not preferred and the second
is the preferred (prototype) name. 
	 
	          


<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>


None.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="recoder.html">recoder</a></td><td>Remove restriction sites but maintain same translation</td></tr>
<tr><td><a href="remap.html">remap</a></td><td>Display sequence with restriction sites, translation etc</td></tr>
<tr><td><a href="restover.html">restover</a></td><td>Find restriction enzymes producing specific overhang</td></tr>
<tr><td><a href="restrict.html">restrict</a></td><td>Finds restriction enzyme cleavage sites</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="silent.html">silent</a></td><td>Silent mutation restriction enzyme scan</td></tr>
</table>


<H2>
    Author(s)
</H2>


Alan Bleasby (ajb&nbsp;&copy;&nbsp;ebi.ac.uk)
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK




<H2>
    History
</H2>

Written 1999 - Alan Bleasby


<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything, from naive users to embedded scripts.



<H2>
    Comments
</H2>
None



</BODY>
</HTML>

