<HTML>

<HEAD>
  <TITLE>
  EMBOSS: remap
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
remap
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>

Display restriction enzyme binding sites in a nucleotide sequence

<H2>
    Description
</H2>
<p><b>remap</b> scans one or more nucleotide sequences for recognition sites and/or cut sites for a supplied set of restriction enzymes.  One or more restriction enzymes can be specified or alternatively all the enzymes in the REBASE database can be investigated. The minimum length of a recognition site to be reported must be specified. It writes an output file showing the location of the cut sites and (optionally) the recognition sites. Sites on both strands are shown by default but there are many options to control exactly what sites are reported and the format of the output file. Optionally, the translated sequence is reported. Additionally, the output file lists enzymess that cut / do not cut the sequence and which match / do not matching certain specified criteria.</p>


   
<H2>
    Usage
</H2>
<b>Here is a sample session with remap</b>
<p>
This example uses only a small region of the input sequence to save space. This is run with a small test version of the restriction enzyme database and so you will probably see more enzymes when you run this. 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>remap -notran -sbeg 1 -send 60 </b>
Display restriction enzyme binding sites in a nucleotide sequence
Input nucleotide sequence(s): <b>tembl:j01636</b>
Comma separated enzyme list [all]: <b>taqi,bsu6i,acii,bsski</b>
Minimum recognition site length [4]: <b></b>
Output file [j01636.remap]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>
<p>
<b>Example 2</b>
<p>
This is an example where all enzymes in the REBASE database are used, (but only the prototypes of the isoschizomers are reported by default). This is run with a small test version of the restriction enzyme database and so you will probably see more enzymes when you run this. 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>remap -notran -sbeg 1 -send 60 </b>
Display restriction enzyme binding sites in a nucleotide sequence
Input nucleotide sequence(s): <b>tembl:j01636</b>
Comma separated enzyme list [all]: <b></b>
Minimum recognition site length [4]: <b></b>
Output file [j01636.remap]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.2">Go to the output files for this example</a><p><p>
<p>
<b>Example 3</b>
<p>
This is an example where all enzymes in the REBASE database are used but the -limit qualifier is not set so that all of the enzymes are displayed and not just only the prototypes of the isoschizomers. This is run with a small test version of the restriction enzyme database and so you will probably see more enzymes when you run this. 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>remap -notran -sbeg 1 -send 60 -nolimit </b>
Display restriction enzyme binding sites in a nucleotide sequence
Input nucleotide sequence(s): <b>tembl:j01636</b>
Comma separated enzyme list [all]: <b></b>
Minimum recognition site length [4]: <b></b>
Output file [j01636.remap]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.3">Go to the output files for this example</a><p><p>
<p>
<b>Example 4</b>
<p>
This shows the 'flat' format: This is run with a small test version of the restriction enzyme database and so you will probably see more enzymes when you run this. 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>remap -notran -sbeg 1 -send 60 -flat </b>
Display restriction enzyme binding sites in a nucleotide sequence
Input nucleotide sequence(s): <b>tembl:j01636</b>
Comma separated enzyme list [all]: <b></b>
Minimum recognition site length [4]: <b></b>
Output file [j01636.remap]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.4">Go to the output files for this example</a><p><p>

<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
Display restriction enzyme binding sites in a nucleotide sequence
Version: EMBOSS:6.2.0

   Standard (Mandatory) qualifiers:
  [-sequence]          seqall     Nucleotide sequence(s) filename and optional
                                  format, or reference (input USA)
   -enzymes            string     [all] The name 'all' reads in all enzyme
                                  names from the REBASE database. You can
                                  specify enzymes by giving their names with
                                  commas between then, such as:
                                  'HincII,hinfI,ppiI,hindiii'.
                                  The case of the names is not important. You
                                  can specify a file of enzyme names to read
                                  in by giving the name of the file holding
                                  the enzyme names with a '@' character in
                                  front of it, for example, '@enz.list'.
                                  Blank lines and lines starting with a hash
                                  character or '!' are ignored and all other
                                  lines are concatenated together with a comma
                                  character ',' and then treated as the list
                                  of enzymes to search for.
                                  An example of a file of enzyme names is:
                                  ! my enzymes
                                  HincII, ppiII
                                  ! other enzymes
                                  hindiii
                                  HinfI
                                  PpiI (Any string)
   -sitelen            integer    [4] This sets the minimum length of the
                                  restriction enzyme recognition site. Any
                                  enzymes with sites shorter than this will be
                                  ignored. (Integer from 2 to 20)
  [-outfile]           outfile    [*.remap] Output file name

   Additional (Optional) qualifiers:
   -mincuts            integer    [1] This sets the minimum number of cuts for
                                  any restriction enzyme that will be
                                  considered. Any enzymes that cut fewer times
                                  than this will be ignored. (Integer from 1
                                  to 1000)
   -maxcuts            integer    [2000000000] This sets the maximum number of
                                  cuts for any restriction enzyme that will
                                  be considered. Any enzymes that cut more
                                  times than this will be ignored. (Integer up
                                  to 2000000000)
   -single             boolean    [N] If this is set then this forces the
                                  values of the mincuts and maxcuts qualifiers
                                  to both be 1. Any other value you may have
                                  set them to will be ignored.
   -[no]blunt          boolean    [Y] This allows those enzymes which cut at
                                  the same position on the forward and reverse
                                  strands to be considered.
   -[no]sticky         boolean    [Y] This allows those enzymes which cut at
                                  different positions on the forward and
                                  reverse strands, leaving an overhang, to be
                                  considered.
   -[no]ambiguity      boolean    [Y] This allows those enzymes which have one
                                  or more 'N' ambiguity codes in their
                                  pattern to be considered
   -plasmid            boolean    [N] If this is set then this allows searches
                                  for restriction enzyme recognition site and
                                  cut postions that span the end of the
                                  sequence to be considered.
   -methylation        boolean    [N] If this is set then RE recognition sites
                                  will not match methylated bases.
   -[no]commercial     boolean    [Y] If this is set, then only those enzymes
                                  with a commercial supplier will be searched
                                  for. This qualifier is ignored if you have
                                  specified an explicit list of enzymes to
                                  search for, rather than searching through
                                  'all' the enzymes in the REBASE database. It
                                  is assumed that, if you are asking for an
                                  explicit enzyme, then you probably know
                                  where to get it from and so all enzymes
                                  names that you have asked to be searched
                                  for, and which cut, will be reported whether
                                  or not they have a commercial supplier.
   -table              menu       [0] Genetic code to use (Values: 0
                                  (Standard); 1 (Standard (with alternative
                                  initiation codons)); 2 (Vertebrate
                                  Mitochondrial); 3 (Yeast Mitochondrial); 4
                                  (Mold, Protozoan, Coelenterate Mitochondrial
                                  and Mycoplasma/Spiroplasma); 5
                                  (Invertebrate Mitochondrial); 6 (Ciliate
                                  Macronuclear and Dasycladacean); 9
                                  (Echinoderm Mitochondrial); 10 (Euplotid
                                  Nuclear); 11 (Bacterial); 12 (Alternative
                                  Yeast Nuclear); 13 (Ascidian Mitochondrial);
                                  14 (Flatworm Mitochondrial); 15
                                  (Blepharisma Macronuclear); 16
                                  (Chlorophycean Mitochondrial); 21 (Trematode
                                  Mitochondrial); 22 (Scenedesmus obliquus);
                                  23 (Thraustochytrium Mitochondrial))
   -frame              menu       [6] This allows you to specify the frames
                                  that are translated. If you are not
                                  displaying cut sites on the reverse sense,
                                  then the reverse sense translations will not
                                  be displayed even if you have requested
                                  frames 4, 5 or 6. By default, all six frames
                                  will be displayed. (Values: 1 (1); 2 (2); 3
                                  (3); F (Forward three frames); -1 (-1); -2
                                  (-2); -3 (-3); R (Reverse three frames); 6
                                  (All six frames))
   -[no]cutlist        boolean    [Y] This produces lists in the output of the
                                  enzymes that cut, those that cut but are
                                  excluded because that cut fewer times than
                                  mincut or more times than maxcut and those
                                  enzymes that do not cut.
   -flatreformat       boolean    [N] This changes the output format to one
                                  where the recognition site is indicated by a
                                  row of '===' characters and the cut site is
                                  pointed to by a '>' character in the
                                  forward sense, or a '<' in the reverse sense
                                  strand.
   -[no]limit          boolean    [Y] This limits the reporting of enzymes to
                                  just one enzyme from each group of
                                  isoschizomers. The enzyme chosen to
                                  represent an isoschizomer group is the
                                  prototype indicated in the data file
                                  'embossre.equ', which is created by the
                                  program 'rebaseextract'. If you prefer
                                  different prototypes to be used, make a copy
                                  of embossre.equ in your home directory and
                                  edit it. If this value is set to be false
                                  then all of the input enzymes will be
                                  reported. You might like to set this to
                                  false if you are supplying an explicit set
                                  of enzymes rather than searching 'all' of
                                  them.

   Advanced (Unprompted) qualifiers:
   -mfile              datafile   [Emethylsites.dat] Restriction enzyme
                                  methylation data file
   -[no]translation    boolean    [Y] This displays the 6-frame translations
                                  of the sequence in the output.
   -[no]reverse        boolean    [Y] This displays the cut sites and
                                  translation of the reverse sense.
   -orfminsize         integer    [If this value is left as 0 then all of the
                                  translation is shown.] This sets the minimum
                                  size of Open Reading Frames (ORFs) to
                                  display in the translations. All other
                                  translation regions are masked by changing
                                  the amino acids to '-' characters. (Integer
                                  0 or more)
   -uppercase          range      [If this is left blank, then the sequence
                                  case is left alone.] Regions to put in
                                  uppercase.
                                  If this is left blank, then the sequence
                                  case is left alone.
                                  A set of regions is specified by a set of
                                  pairs of positions.
                                  The positions are integers.
                                  They are separated by any non-digit,
                                  non-alpha character.
                                  Examples of region specifications are:
                                  24-45, 56-78
                                  1:45, 67=99;765..888
                                  1,5,8,10,23,45,57,99
   -highlight          range      [(full sequence)] Regions to colour if
                                  formatting for HTML.
                                  If this is left blank, then the sequence is
                                  left alone.
                                  A set of regions is specified by a set of
                                  pairs of positions.
                                  The positions are integers.
                                  They are followed by any valid HTML font
                                  colour.
                                  Examples of region specifications are:
                                  24-45 blue 56-78 orange
                                  1-100 green 120-156 red
                                  A file of ranges to colour (one range per
                                  line) can be specified as '@filename'.
   -threeletter        boolean    [N] Display protein sequences in
                                  three-letter code
   -number             boolean    [N] Number the sequences
   -width              integer    [60] Width of sequence to display (Integer 1
                                  or more)
   -length             integer    [0] Line length of page (0 for indefinite)
                                  (Integer 0 or more)
   -margin             integer    [10] Margin around sequence for numbering
                                  (Integer 0 or more)
   -[no]name           boolean    [Y] Set this to be false if you do not wish
                                  to display the ID name of the sequence
   -[no]description    boolean    [Y] Set this to be false if you do not wish
                                  to display the description of the sequence
   -offset             integer    [1] Offset to start numbering the sequence
                                  from (Any integer value)
   -html               boolean    [N] Use HTML formatting

   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1            integer    Start of each sequence to be used
   -send1              integer    End of each sequence to be used
   -sreverse1          boolean    Reverse (if DNA)
   -sask1              boolean    Ask for begin/end/reverse
   -snucleotide1       boolean    Sequence is nucleotide
   -sprotein1          boolean    Sequence is protein
   -slower1            boolean    Make lower case
   -supper1            boolean    Make upper case
   -sformat1           string     Input sequence format
   -sdbname1           string     Database name
   -sid1               string     Entryname
   -ufo1               string     UFO features
   -fformat1           string     Features format
   -fopenfile1         string     Features file name

   "-outfile" associated qualifiers
   -odirectory2        string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-sequence]<br>(Parameter 1)</td>
<td>seqall</td>
<td>Nucleotide sequence(s) filename and optional format, or reference (input USA)</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-enzymes</td>
<td>string</td>
<td>The name 'all' reads in all enzyme names from the REBASE database. You can specify enzymes by giving their names with commas between then, such as: 'HincII,hinfI,ppiI,hindiii'.
The case of the names is not important. You can specify a file of enzyme names to read in by giving the name of the file holding the enzyme names with a '@' character in front of it, for example, '@enz.list'.
Blank lines and lines starting with a hash character or '!' are ignored and all other lines are concatenated together with a comma character ',' and then treated as the list of enzymes to search for.
An example of a file of enzyme names is:
! my enzymes
HincII, ppiII
! other enzymes
hindiii
HinfI
PpiI</td>
<td>Any string</td>
<td>all</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-sitelen</td>
<td>integer</td>
<td>This sets the minimum length of the restriction enzyme recognition site. Any enzymes with sites shorter than this will be ignored.</td>
<td>Integer from 2 to 20</td>
<td>4</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-outfile]<br>(Parameter 2)</td>
<td>outfile</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;*&gt;</i>.remap</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>-mincuts</td>
<td>integer</td>
<td>This sets the minimum number of cuts for any restriction enzyme that will be considered. Any enzymes that cut fewer times than this will be ignored.</td>
<td>Integer from 1 to 1000</td>
<td>1</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-maxcuts</td>
<td>integer</td>
<td>This sets the maximum number of cuts for any restriction enzyme that will be considered. Any enzymes that cut more times than this will be ignored.</td>
<td>Integer up to 2000000000</td>
<td>2000000000</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-single</td>
<td>boolean</td>
<td>If this is set then this forces the values of the mincuts and maxcuts qualifiers to both be 1. Any other value you may have set them to will be ignored.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-[no]blunt</td>
<td>boolean</td>
<td>This allows those enzymes which cut at the same position on the forward and reverse strands to be considered.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-[no]sticky</td>
<td>boolean</td>
<td>This allows those enzymes which cut at different positions on the forward and reverse strands, leaving an overhang, to be considered.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-[no]ambiguity</td>
<td>boolean</td>
<td>This allows those enzymes which have one or more 'N' ambiguity codes in their pattern to be considered</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-plasmid</td>
<td>boolean</td>
<td>If this is set then this allows searches for restriction enzyme recognition site and cut postions that span the end of the sequence to be considered.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-methylation</td>
<td>boolean</td>
<td>If this is set then RE recognition sites will not match methylated bases.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-[no]commercial</td>
<td>boolean</td>
<td>If this is set, then only those enzymes with a commercial supplier will be searched for. This qualifier is ignored if you have specified an explicit list of enzymes to search for, rather than searching through 'all' the enzymes in the REBASE database. It is assumed that, if you are asking for an explicit enzyme, then you probably know where to get it from and so all enzymes names that you have asked to be searched for, and which cut, will be reported whether or not they have a commercial supplier.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-table</td>
<td>list</td>
<td>Genetic code to use</td>
<td><table><tr><td>0</td> <td><i>(Standard)</i></td></tr><tr><td>1</td> <td><i>(Standard (with alternative initiation codons))</i></td></tr><tr><td>2</td> <td><i>(Vertebrate Mitochondrial)</i></td></tr><tr><td>3</td> <td><i>(Yeast Mitochondrial)</i></td></tr><tr><td>4</td> <td><i>(Mold, Protozoan, Coelenterate Mitochondrial and Mycoplasma/Spiroplasma)</i></td></tr><tr><td>5</td> <td><i>(Invertebrate Mitochondrial)</i></td></tr><tr><td>6</td> <td><i>(Ciliate Macronuclear and Dasycladacean)</i></td></tr><tr><td>9</td> <td><i>(Echinoderm Mitochondrial)</i></td></tr><tr><td>10</td> <td><i>(Euplotid Nuclear)</i></td></tr><tr><td>11</td> <td><i>(Bacterial)</i></td></tr><tr><td>12</td> <td><i>(Alternative Yeast Nuclear)</i></td></tr><tr><td>13</td> <td><i>(Ascidian Mitochondrial)</i></td></tr><tr><td>14</td> <td><i>(Flatworm Mitochondrial)</i></td></tr><tr><td>15</td> <td><i>(Blepharisma Macronuclear)</i></td></tr><tr><td>16</td> <td><i>(Chlorophycean Mitochondrial)</i></td></tr><tr><td>21</td> <td><i>(Trematode Mitochondrial)</i></td></tr><tr><td>22</td> <td><i>(Scenedesmus obliquus)</i></td></tr><tr><td>23</td> <td><i>(Thraustochytrium Mitochondrial)</i></td></tr></table></td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-frame</td>
<td>list</td>
<td>This allows you to specify the frames that are translated. If you are not displaying cut sites on the reverse sense, then the reverse sense translations will not be displayed even if you have requested frames 4, 5 or 6. By default, all six frames will be displayed.</td>
<td><table><tr><td>1</td> <td><i>(1)</i></td></tr><tr><td>2</td> <td><i>(2)</i></td></tr><tr><td>3</td> <td><i>(3)</i></td></tr><tr><td>F</td> <td><i>(Forward three frames)</i></td></tr><tr><td>-1</td> <td><i>(-1)</i></td></tr><tr><td>-2</td> <td><i>(-2)</i></td></tr><tr><td>-3</td> <td><i>(-3)</i></td></tr><tr><td>R</td> <td><i>(Reverse three frames)</i></td></tr><tr><td>6</td> <td><i>(All six frames)</i></td></tr></table></td>
<td>6</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-[no]cutlist</td>
<td>boolean</td>
<td>This produces lists in the output of the enzymes that cut, those that cut but are excluded because that cut fewer times than mincut or more times than maxcut and those enzymes that do not cut.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-flatreformat</td>
<td>boolean</td>
<td>This changes the output format to one where the recognition site is indicated by a row of '===' characters and the cut site is pointed to by a '&gt;' character in the forward sense, or a '&lt;' in the reverse sense strand.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-[no]limit</td>
<td>boolean</td>
<td>This limits the reporting of enzymes to just one enzyme from each group of isoschizomers. The enzyme chosen to represent an isoschizomer group is the prototype indicated in the data file 'embossre.equ', which is created by the program 'rebaseextract'. If you prefer different prototypes to be used, make a copy of embossre.equ in your home directory and edit it. If this value is set to be false then all of the input enzymes will be reported. You might like to set this to false if you are supplying an explicit set of enzymes rather than searching 'all' of them.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>-mfile</td>
<td>datafile</td>
<td>Restriction enzyme methylation data file</td>
<td>Data file</td>
<td>Emethylsites.dat</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-[no]translation</td>
<td>boolean</td>
<td>This displays the 6-frame translations of the sequence in the output.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-[no]reverse</td>
<td>boolean</td>
<td>This displays the cut sites and translation of the reverse sense.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-orfminsize</td>
<td>integer</td>
<td>This sets the minimum size of Open Reading Frames (ORFs) to display in the translations. All other translation regions are masked by changing the amino acids to '-' characters.</td>
<td>Integer 0 or more</td>
<td>If this value is left as 0 then all of the translation is shown.</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-uppercase</td>
<td>range</td>
<td>Regions to put in uppercase.
If this is left blank, then the sequence case is left alone.
A set of regions is specified by a set of pairs of positions.
The positions are integers.
They are separated by any non-digit, non-alpha character.
Examples of region specifications are:
24-45, 56-78
1:45, 67=99;765..888
1,5,8,10,23,45,57,99</td>
<td>Sequence range</td>
<td>If this is left blank, then the sequence case is left alone.</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-highlight</td>
<td>range</td>
<td>Regions to colour if formatting for HTML.
If this is left blank, then the sequence is left alone.
A set of regions is specified by a set of pairs of positions.
The positions are integers.
They are followed by any valid HTML font colour.
Examples of region specifications are:
24-45 blue 56-78 orange
1-100 green 120-156 red
A file of ranges to colour (one range per line) can be specified as '@filename'.</td>
<td>Sequence range</td>
<td><i>full sequence</i></td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-threeletter</td>
<td>boolean</td>
<td>Display protein sequences in three-letter code</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-number</td>
<td>boolean</td>
<td>Number the sequences</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-width</td>
<td>integer</td>
<td>Width of sequence to display</td>
<td>Integer 1 or more</td>
<td>60</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-length</td>
<td>integer</td>
<td>Line length of page (0 for indefinite)</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-margin</td>
<td>integer</td>
<td>Margin around sequence for numbering</td>
<td>Integer 0 or more</td>
<td>10</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-[no]name</td>
<td>boolean</td>
<td>Set this to be false if you do not wish to display the ID name of the sequence</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-[no]description</td>
<td>boolean</td>
<td>Set this to be false if you do not wish to display the description of the sequence</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-offset</td>
<td>integer</td>
<td>Offset to start numbering the sequence from</td>
<td>Any integer value</td>
<td>1</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-html</td>
<td>boolean</td>
<td>Use HTML formatting</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-sequence" associated seqall qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sbegin1<br>-sbegin_sequence</td>
<td>integer</td>
<td>Start of each sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -send1<br>-send_sequence</td>
<td>integer</td>
<td>End of each sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sreverse1<br>-sreverse_sequence</td>
<td>boolean</td>
<td>Reverse (if DNA)</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sask1<br>-sask_sequence</td>
<td>boolean</td>
<td>Ask for begin/end/reverse</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -snucleotide1<br>-snucleotide_sequence</td>
<td>boolean</td>
<td>Sequence is nucleotide</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sprotein1<br>-sprotein_sequence</td>
<td>boolean</td>
<td>Sequence is protein</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -slower1<br>-slower_sequence</td>
<td>boolean</td>
<td>Make lower case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -supper1<br>-supper_sequence</td>
<td>boolean</td>
<td>Make upper case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sformat1<br>-sformat_sequence</td>
<td>string</td>
<td>Input sequence format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sdbname1<br>-sdbname_sequence</td>
<td>string</td>
<td>Database name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sid1<br>-sid_sequence</td>
<td>string</td>
<td>Entryname</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ufo1<br>-ufo_sequence</td>
<td>string</td>
<td>UFO features</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fformat1<br>-fformat_sequence</td>
<td>string</td>
<td>Features format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fopenfile1<br>-fopenfile_sequence</td>
<td>string</td>
<td>Features file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-outfile" associated outfile qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -odirectory2<br>-odirectory_outfile</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>


<H2>
    Input file format
</H2>



<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:j01636' is a sequence entry in the example nucleic acid database 'tembl'
<p>
<p><h3>Database entry: tembl:j01636</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   J01636; SV 1; linear; genomic DNA; STD; PRO; 7477 BP.
XX
AC   J01636; J01637; K01483; K01793;
XX
DT   30-NOV-1990 (Rel. 26, Created)
DT   09-SEP-2004 (Rel. 81, Last updated, Version 8)
XX
DE   E.coli lactose operon with lacI, lacZ, lacY and lacA genes.
XX
KW   acetyltransferase; beta-D-galactosidase; galactosidase; lac operon;
KW   lac repressor protein; lacA gene; lacI gene; lactose permease; lacY gene;
KW   lacZ gene; mutagenesis; palindrome; promoter region;
KW   thiogalactoside acetyltransferase.
XX
OS   Escherichia coli
OC   Bacteria; Proteobacteria; Gammaproteobacteria; Enterobacteriales;
OC   Enterobacteriaceae; Escherichia.
XX
RN   [1]
RP   1243-1266
RX   DOI; 10.1073/pnas.70.12.3581.
RX   PUBMED; 4587255.
RA   Gilbert W., Maxam A.;
RT   "The nucleotide sequence of the lac operator";
RL   Proc. Natl. Acad. Sci. U.S.A. 70(12):3581-3584(1973).
XX
RN   [2]
RP   1246-1308
RX   DOI; 10.1073/pnas.70.12.3585.
RX   PUBMED; 4587256.
RA   Maizels N.M.;
RT   "The nucleotide sequence of the lactose messenger ribonucleic acid
RT   transcribed from the UV5 promoter mutant of Escherichia coli";
RL   Proc. Natl. Acad. Sci. U.S.A. 70(12):3585-3589(1973).
XX
RN   [3]
RX   PUBMED; 4598642.
RA   Gilbert W., Maizels N., Maxam A.;
RT   "Sequences of controlling regions of the lactose operon";
RL   Cold Spring Harb. Symp. Quant. Biol. 38:845-855(1974).
XX
RN   [4]
RA   Gilbert W., Gralla J., Majors A.J., Maxam A.;
RT   "Lactose operator sequences and the action of lac repressor";
RL   (in) Sund H., Blauer G. (Eds.);
RL   PROTEIN-LIGAND INTERACTIONS:193-207;
RL   Walter de Gruyter, New York (1975)
XX
RN   [5]
RP   1146-1282


<font color=red>  [Part of this file has been deleted for brevity]</font>

     cgatttggct acatgacatc aaccatatca gcaaaagtga tacgggtatt atttttgccg      4560
     ctatttctct gttctcgcta ttattccaac cgctgtttgg tctgctttct gacaaactcg      4620
     ggctgcgcaa atacctgctg tggattatta ccggcatgtt agtgatgttt gcgccgttct      4680
     ttatttttat cttcgggcca ctgttacaat acaacatttt agtaggatcg attgttggtg      4740
     gtatttatct aggcttttgt tttaacgccg gtgcgccagc agtagaggca tttattgaga      4800
     aagtcagccg tcgcagtaat ttcgaatttg gtcgcgcgcg gatgtttggc tgtgttggct      4860
     gggcgctgtg tgcctcgatt gtcggcatca tgttcaccat caataatcag tttgttttct      4920
     ggctgggctc tggctgtgca ctcatcctcg ccgttttact ctttttcgcc aaaacggatg      4980
     cgccctcttc tgccacggtt gccaatgcgg taggtgccaa ccattcggca tttagcctta      5040
     agctggcact ggaactgttc agacagccaa aactgtggtt tttgtcactg tatgttattg      5100
     gcgtttcctg cacctacgat gtttttgacc aacagtttgc taatttcttt acttcgttct      5160
     ttgctaccgg tgaacagggt acgcgggtat ttggctacgt aacgacaatg ggcgaattac      5220
     ttaacgcctc gattatgttc tttgcgccac tgatcattaa tcgcatcggt gggaaaaacg      5280
     ccctgctgct ggctggcact attatgtctg tacgtattat tggctcatcg ttcgccacct      5340
     cagcgctgga agtggttatt ctgaaaacgc tgcatatgtt tgaagtaccg ttcctgctgg      5400
     tgggctgctt taaatatatt accagccagt ttgaagtgcg tttttcagcg acgatttatc      5460
     tggtctgttt ctgcttcttt aagcaactgg cgatgatttt tatgtctgta ctggcgggca      5520
     atatgtatga aagcatcggt ttccagggcg cttatctggt gctgggtctg gtggcgctgg      5580
     gcttcacctt aatttccgtg ttcacgctta gcggccccgg cccgctttcc ctgctgcgtc      5640
     gtcaggtgaa tgaagtcgct taagcaatca atgtcggatg cggcgcgacg cttatccgac      5700
     caacatatca taacggagtg atcgcattga acatgccaat gaccgaaaga ataagagcag      5760
     gcaagctatt taccgatatg tgcgaaggct taccggaaaa aagacttcgt gggaaaacgt      5820
     taatgtatga gtttaatcac tcgcatccat cagaagttga aaaaagagaa agcctgatta      5880
     aagaaatgtt tgccacggta ggggaaaacg cctgggtaga accgcctgtc tatttctctt      5940
     acggttccaa catccatata ggccgcaatt tttatgcaaa tttcaattta accattgtcg      6000
     atgactacac ggtaacaatc ggtgataacg tactgattgc acccaacgtt actctttccg      6060
     ttacgggaca ccctgtacac catgaattga gaaaaaacgg cgagatgtac tcttttccga      6120
     taacgattgg caataacgtc tggatcggaa gtcatgtggt tattaatcca ggcgtcacca      6180
     tcggggataa ttctgttatt ggcgcgggta gtatcgtcac aaaagacatt ccaccaaacg      6240
     tcgtggcggc tggcgttcct tgtcgggtta ttcgcgaaat aaacgaccgg gataagcact      6300
     attatttcaa agattataaa gttgaatcgt cagtttaaat tataaaaatt gcctgatacg      6360
     ctgcgcttat caggcctaca agttcagcga tctacattag ccgcatccgg catgaacaaa      6420
     gcgcaggaac aagcgtcgca tcatgcctct ttgacccaca gctgcggaaa acgtactggt      6480
     gcaaaacgca gggttatgat catcagccca acgacgcaca gcgcatgaaa tgcccagtcc      6540
     atcaggtaat tgccgctgat actacgcagc acgccagaaa accacggggc aagcccggcg      6600
     atgataaaac cgattccctg cataaacgcc accagcttgc cagcaatagc cggttgcaca      6660
     gagtgatcga gcgccagcag caaacagagc ggaaacgcgc cgcccagacc taacccacac      6720
     accatcgccc acaataccgg caattgcatc ggcagccaga taaagccgca gaaccccacc      6780
     agttgtaaca ccagcgccag cattaacagt ttgcgccgat cctgatggcg agccatagca      6840
     ggcatcagca aagctcctgc ggcttgccca agcgtcatca atgccagtaa ggaaccgctg      6900
     tactgcgcgc tggcaccaat ctcaatatag aaagcgggta accaggcaat caggctggcg      6960
     taaccgccgt taatcagacc gaagtaaaca cccagcgtcc acgcgcgggg agtgaatacc      7020
     acgcgaaccg gagtggttgt tgtcttgtgg gaagaggcga cctcgcgggc gctttgccac      7080
     caccaggcaa agagcgcaac aacggcaggc agcgccacca ggcgagtgtt tgataccagg      7140
     tttcgctatg ttgaactaac cagggcgtta tggcggcacc aagcccaccg ccgcccatca      7200
     gagccgcgga ccacagcccc atcaccagtg gcgtgcgctg ctgaaaccgc cgtttaatca      7260
     ccgaagcatc accgcctgaa tgatgccgat ccccacccca ccaagcagtg cgctgctaag      7320
     cagcagcgca ctttgcgggt aaagctcacg catcaatgca ccgacggcaa tcagcaacag      7380
     actgatggcg acactgcgac gttcgctgac atgctgatga agccagcttc cggccagcgc      7440
     cagcccgccc atggtaacca ccggcagagc ggtcgac                               7477
//
</pre>
</td></tr></table><p>

<p>

You can specifiy a file of ranges to display in uppercase by giving the
'-uppercase' qualifier the value '@' followed by the name of the file
containing the ranges.  (eg: '-upper @myfile'). 

<p>

The format of the range file is: 

<p>

<ul>
<li>      Comment lines start with '#' in the first column. 
<li>      Comment lines and blank lines are ignored. 
<li>      The line may start with white-space. 
<li>      There are two positive (integer) numbers per line separated by one or more
      space or TAB characters. 
<li>      The second number must be greater or equal to the first number. 
<li>      There can be optional text after the two numbers to annotate the line. 
<li>      White-space before or after the text is removed. 
</ul>

<p>

An example range file is: 

<pre>

# this is my set of ranges
12   23
 4   5       this is like 12-23, but smaller
67   10348   interesting region

</pre>

<p>

You can specifiy a file of ranges to highlight in a different colour
when outputting in HTML format (using the '-html' qualifier) by giving
the '-highlight' qualifier the value '@' followed by the name of the
file containing the ranges.  (eg: '-highlight @myfile'). 

<p>

The format of this file is very similar to the format of the above
uppercase range file, except that the text after the start and end
positions is used as the HTML colour name.  This colour name is used 'as
is' when specifying the colour in HTML in a '&lt;FONT COLOR=xxx&gt;'
construct, (where 'xxx' is the name of the colour). 

<p>

The standard names of HTML font colours are given in:

<br>

<A HREF="http://http://www.w3.org/TR/REC-html40/types.html">
http://http://www.w3.org/TR/REC-html40/types.html</A>
and
<A HREF="http://www.ausmall.com.au/freegraf/ncolour2.htm">
http://www.ausmall.com.au/freegraf/ncolour2.htm</A>
and
<A HREF="http://mindprod.com/htmlcolours.html">
http://mindprod.com/htmlcolours.html</A>
(amongst other places).

<p>

An example highlight range file is:

<p>

<hr>
<pre>
# this is my set of ranges
12   23		red
 4   5		darkturquoise
67   10348	#FFE4E1
</pre>
<hr>

<H2>
    Output file format
</H2>



<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: j01636.remap </h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
J01636
E.coli lactose operon with lacI, lacZ, lacY and lacA genes.

                                                        BssKI
                 TaqI                 AciI              Ksp632I
                 \                    \                 \
          gacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagt
                   10        20        30        40        50        60        
          ----:----|----:----|----:----|----:----|----:----|----:----|
          ctgtggtagcttaccgcgttttggaaagcgccataccgtactatcgcgggccttctctca
                   /                    /                  / /
                   TaqI                 AciI               | BssKI 
                                                           Ksp632I


# Enzymes that cut  Frequency	Isoschizomers
      AciI	    1	
     BssKI	    1	
   Ksp632I	    1	Bsu6I
      TaqI	    1	



# Enzymes which cut less frequently than the MINCUTS criterion
# Enzymes &lt; MINCUTS Frequency	Isoschizomers



# Enzymes which cut more frequently than the MAXCUTS criterion
# Enzymes &gt; MAXCUTS Frequency	Isoschizomers



# Enzymes that do not cut




# No. of cutting enzymes which do not match the
# SITELEN, BLUNT, STICKY, COMMERCIAL, AMBIGUOUS citeria

0

</pre>
</td></tr></table><p>

<a name="output.2"></a>
<h3>Output files for usage example 2</h3>
<p><h3>File: j01636.remap</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
J01636
E.coli lactose operon with lacI, lacZ, lacY and lacA genes.

                                                      Hin6I
                 TaqI                                 | HhaI
                 |  BsiYI                             | BssKI
                 |  |   Hin6I                         | Ksp632I
                 |  |   | HhaI        AciI            | | HpaII 
                 \  \   \ \           \               \ \ \
          gacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagt
                   10        20        30        40        50        60        
          ----:----|----:----|----:----|----:----|----:----|----:----|
          ctgtggtagcttaccgcgttttggaaagcgccataccgtactatcgcgggccttctctca
                 / /    / /             /             / /  ///
                 | TaqI | Hin6I         AciI          | |  ||BssKI 
                 BsiYI  HhaI                          | |  |HpaII  
                                                      | |  Ksp632I
                                                      | Hin6I
                                                      HhaI


# Enzymes that cut  Frequency	Isoschizomers
      AciI	    1	
     BsiYI	    1	Bsc4I
     BssKI	    1	
      HhaI	    2	
     Hin6I	    2	HinP1I,HspAI
     HpaII	    1	BsiSI
   Ksp632I	    1	Bsu6I
      TaqI	    1	



# Enzymes which cut less frequently than the MINCUTS criterion
# Enzymes &lt; MINCUTS Frequency	Isoschizomers



# Enzymes which cut more frequently than the MAXCUTS criterion
# Enzymes &gt; MAXCUTS Frequency	Isoschizomers



# Enzymes that do not cut

AclI      BamHI     BceAI     BseYI     BsrI      ClaI      EcoRI     EcoRII    
HaeIII    Hin4I     HindII    HindIII   KpnI      MaeII     NotI      


# No. of cutting enzymes which do not match the
# SITELEN, BLUNT, STICKY, COMMERCIAL, AMBIGUOUS citeria

0

</pre>
</td></tr></table><p>

<a name="output.3"></a>
<h3>Output files for usage example 3</h3>
<p><h3>File: j01636.remap</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
J01636
E.coli lactose operon with lacI, lacZ, lacY and lacA genes.

                                                      HspAI
                                                      Hin6I 
                 TaqI                                 HinP1I
                 |  BsiYI                             | HhaI 
                 |  Bsc4I                             | Bsu6I
                 |  |   HspAI                         | BssKI 
                 |  |   Hin6I                         | Ksp632I
                 |  |   HinP1I                        | | HpaII 
                 |  |   | HhaI        AciI            | | BsiSI 
                 \  \   \ \           \               \ \ \
          gacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagt
                   10        20        30        40        50        60        
          ----:----|----:----|----:----|----:----|----:----|----:----|
          ctgtggtagcttaccgcgttttggaaagcgccataccgtactatcgcgggccttctctca
                 / /    / /             /             / /  ///
                 | TaqI | HinP1I        AciI          | |  ||BssKI 
                 Bsc4I  | Hin6I                       | |  |BsiSI  
                 BsiYI  | HspAI                       | |  |HpaII  
                        HhaI                          | |  Ksp632I
                                                      | |  Bsu6I
                                                      | HinP1I
                                                      | Hin6I 
                                                      | HspAI
                                                      HhaI


# Enzymes that cut  Frequency
      AciI	    1	
     Bsc4I	    1	
     BsiSI	    1	
     BsiYI	    1	
     BssKI	    1	
     Bsu6I	    1	
      HhaI	    2	
     Hin6I	    2	
    HinP1I	    2	
     HpaII	    1	
     HspAI	    2	
   Ksp632I	    1	
      TaqI	    1	



# Enzymes which cut less frequently than the MINCUTS criterion
# Enzymes &lt; MINCUTS Frequency



# Enzymes which cut more frequently than the MAXCUTS criterion
# Enzymes &gt; MAXCUTS Frequency



# Enzymes that do not cut

AclI      BamHI     BceAI     Bse1I     BseYI     BshI      BsrI      ClaI      
EcoRI     EcoRII    HaeIII    Hin4I     HindII    HindIII   HpyCH4IV  KpnI      
MaeII     NotI      


# No. of cutting enzymes which do not match the
# SITELEN, BLUNT, STICKY, COMMERCIAL, AMBIGUOUS citeria

0

</pre>
</td></tr></table><p>

<a name="output.4"></a>
<h3>Output files for usage example 4</h3>
<p><h3>File: j01636.remap</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
J01636
E.coli lactose operon with lacI, lacZ, lacY and lacA genes.

                                                                  Ksp632I
                                                        &gt;.........====
                                                          HpaII
                                                          &gt;===
                        HhaI                             BssKI
                        ==&gt;=                            &gt;=====
                 TaqI   Hin6I                         HhaI
                 &gt;===   &gt;===                          ==&gt;=
              BsiYI                      AciI         Hin6I
              ======&gt;====             &gt;..====         &gt;===
          gacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagt
                   10        20        30        40        50        60        
          ----:----|----:----|----:----|----:----|----:----|----:----|
          ctgtggtagcttaccgcgttttggaaagcgccataccgtactatcgcgggccttctctca
              ====&lt;======                &lt;===         ===&lt;
              BsiYI                      AciI         Hin6I
                 ===&lt;   ===&lt;                          =&lt;==  &lt;.....====
                 TaqI   Hin6I                         HhaI        Ksp632I
                        =&lt;==                             =====&lt;
                        HhaI                             BssKI
                                                          ===&lt;
                                                          HpaII


# Enzymes that cut  Frequency	Isoschizomers
      AciI	    1	
     BsiYI	    1	Bsc4I
     BssKI	    1	
      HhaI	    2	
     Hin6I	    2	HinP1I,HspAI
     HpaII	    1	BsiSI
   Ksp632I	    1	Bsu6I
      TaqI	    1	



# Enzymes which cut less frequently than the MINCUTS criterion
# Enzymes &lt; MINCUTS Frequency	Isoschizomers



# Enzymes which cut more frequently than the MAXCUTS criterion
# Enzymes &gt; MAXCUTS Frequency	Isoschizomers



# Enzymes that do not cut

AclI      BamHI     BceAI     BseYI     BsrI      ClaI      EcoRI     EcoRII    
HaeIII    Hin4I     HindII    HindIII   KpnI      MaeII     NotI      


# No. of cutting enzymes which do not match the
# SITELEN, BLUNT, STICKY, COMMERCIAL, AMBIGUOUS citeria

0

</pre>
</td></tr></table><p>

<p>

The name of the sequence is displayed, followed by the description of
the sequence.

<p>

The formatted display of cut sites on the sequence follows, with the
six-frame translation below it.  The cut sites are indicated by a slash
character '\' that points to the poition between the nucleotides where
the cuts occur.  Cuts by many enzymes at the same position are indicated
by stacking the enzyme names on top of each other. 

<p>

At the end the section header 'Enzymes that cut' is displayed followed
by a list of the enzymes that cut the specified sequence and the number
of times that they cut.  For each enzyme that cuts, a list of
isoschizomers of that enzyme (sharing the same recognition site pattern
and cut sites) is given.  

<p>

This is followed by lists of the enzymes that do cut, but which cut less
often than the '-mincut' qualifier or more often than the '-maxcut'
qualifier. 

<p>

Any of the isoschizomers that are excluded from cutting, (either through
restrictions such as the permitted number of cuts, blunt cutters only,
single cutters only etc.  or because their name has not been given in
the input list of enzymes), will not be listed. 

<p>

Then a list is displayed of the enzymes whose names were input and which
match the other criteria ('-sitelen', '-blunt', '-sticky', '-ambiguity' or
'-commercial') but which do not cut. 

<p>

Finally the number of enzymes that were rejected from consideration
because they do not match the '-sitelen', '-blunt', '-sticky', '-ambiguity' or
'-commercial' criteria is displayed.

<p>

The '-flatreformat' qualifier changes the display to emphasise the
recognition site of the restriction enzyme, which is indicated by a row
of '=' characters.  The cut site if pointed to by a '&gt;' or '&lt;' character
and if the cut site is not within or imemdiately adjacent to the
recognition site, they are linked by a row of '.' characters.  

<p>

The name of the enzyme is displayed above (or below when the reverse
sense site if displayed) the recognition site.  The name of the enzyme
is also displayed above the cut site if this occurs on a different
display line to the recognition site (i.e.  if it wraps onto the next
line of sequence). 



<H2>
    Data files
</H2>

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

<p>

The EMBOSS REBASE restriction enzyme data files are stored in
directory 'data/REBASE/*' under the EMBOSS installation directory.

<p>

These files must first be set up using the program <a
href="rebaseextract.html">'<b>rebaseextract</b>'</a>.  Running
'rebaseextract' may be the job of your system manager.

<p>
      
The data files are stored in the REBASE directory of the standard EMBOSS
data directory. The names are:

<ul>
<li> embossre.enz     Cleavage information
<li> embossre.ref     Reference/methylation information
<li> embossre.sup     Supplier information
</ul>
	 
The column information is described at the top of the data files
	 
<p>
	 
The reported enzyme from any one group of isoschizomers (the prototype)
is specified in the REBASE database and the information is held in the
data file 'embossre.equ'.  You may edit this file to set your own
preferred prototype, if you wish. 
	 
<p>
	 
The format of the file "embossre.equ" is
<br>
Enzyme-name Prototype-name
	 
<p>
	 
i.e.  two columns of enzyme names separated by a space.  The first name
of the pair of enzymes is the name that is not preferred and the second
is the preferred (prototype) name. 
	 
	          

<H2>
    Notes
</H2>


<p>The Restriction Enzyme database (REBASE) is a collection of information about restriction enzymes and related proteins. It contains published and unpublished references, recognition and cleavage sites, isoschizomers, commercial availability, methylation sensitivity, crystal and sequence data. DNA methyltransferases, homing endonucleases, nicking enzymes, specificity subunits and control proteins are also included. Most recently, putative DNA methyltransferases and restriction enzymes, as predicted from analysis of genomic sequences, are also listed.</p>

<p>The home page of REBASE is: <a href="http://rebase.neb.com/">http://rebase.neb.com/</a></p>

<p>Where the translation is given in the output file, the genetic code and one or more frames for translation may be specified. The <tt>-no[reverse]</tt> option specifies whether the translation (and cut and recognition sites) are shown for the reverse sense strand.</p>

<p>By default, only one enzyme of any group of isoschizomers (enzymes that have the same recognition site and cut positions) is reported. This behaviour can be changed by specifying <tt>-nolimit</tt>, in which case all isoschizomers are reported.  The default behaviour uses the representative enzyme of an isoschizomer group (the prototype) which is specified in the EMBOSS data file <tt>embossre.equ</tt>.  This file is generated from the REBASE database by running <b>rebaseextract</b>. You may edit this file to set your own preferred prototype, if you wish.</p> 

<p>As well as the display of where enzymes cut in the sequence, <b>remap</b> displays:</p>

<ul>
<li>The list of enzymes that cut the sequence and match the required criteria.
<li>The list of enzymes that cut the sequence and fail the MINCUTS criteria.
<li>The list of enzymes that cut the sequence and fail the MAXCUTS criteria.
<li>The list of enzymes that do not cut the sequence but which match all the required criteria.
<li>The number of enzymes that cut the sequence and fail the SITELEN, BLUNT, STICKY, COMMERCIAL, AMBIGUOUS criteria.
</ul>



<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

<p><b>remap</b> uses the EMBOSS REBASE restriction enzyme data files stored in directory <tt>data/REBASE/*</tt> under the EMBOSS installation directory. These files must first be set up using the program <b>rebaseextract</b>. Running <b>rebaseextract</b> may be the job of your system manager.</p>


<H2>
    Diagnostic Error Messages
</H2>

None.


<H2>
    Exit status
</H2>

It always exits with status 0.

<H2>
    Known bugs
</H2>

None.


<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="abiview.html">abiview</a></td>
<td>Display the trace in an ABI sequencer file</td>
</tr>

<tr>
<td><a href="backtranambig.html">backtranambig</a></td>
<td>Back-translate a protein sequence to ambiguous nucleotide sequence</td>
</tr>

<tr>
<td><a href="backtranseq.html">backtranseq</a></td>
<td>Back-translate a protein sequence to a nucleotide sequence</td>
</tr>

<tr>
<td><a href="cirdna.html">cirdna</a></td>
<td>Draws circular maps of DNA constructs</td>
</tr>

<tr>
<td><a href="coderet.html">coderet</a></td>
<td>Extract CDS, mRNA and translations from feature tables</td>
</tr>

<tr>
<td><a href="lindna.html">lindna</a></td>
<td>Draws linear maps of DNA constructs</td>
</tr>

<tr>
<td><a href="pepnet.html">pepnet</a></td>
<td>Draw a helical net for a protein sequence</td>
</tr>

<tr>
<td><a href="pepwheel.html">pepwheel</a></td>
<td>Draw a helical wheel diagram for a protein sequence</td>
</tr>

<tr>
<td><a href="plotorf.html">plotorf</a></td>
<td>Plot potential open reading frames in a nucleotide sequence</td>
</tr>

<tr>
<td><a href="prettyplot.html">prettyplot</a></td>
<td>Draw a sequence alignment with pretty formatting</td>
</tr>

<tr>
<td><a href="prettyseq.html">prettyseq</a></td>
<td>Write a nucleotide sequence and its translation to file</td>
</tr>

<tr>
<td><a href="recoder.html">recoder</a></td>
<td>Find restriction sites to remove (mutate) with no translation change</td>
</tr>

<tr>
<td><a href="redata.html">redata</a></td>
<td>Retrieve information from REBASE restriction enzyme database</td>
</tr>

<tr>
<td><a href="restover.html">restover</a></td>
<td>Find restriction enzymes producing a specific overhang</td>
</tr>

<tr>
<td><a href="restrict.html">restrict</a></td>
<td>Report restriction enzyme cleavage sites in a nucleotide sequence</td>
</tr>

<tr>
<td><a href="seealso.html">seealso</a></td>
<td>Finds programs with similar function to a specified program</td>
</tr>

<tr>
<td><a href="showalign.html">showalign</a></td>
<td>Display a multiple sequence alignment in pretty format</td>
</tr>

<tr>
<td><a href="showdb.html">showdb</a></td>
<td>Displays information on configured databases</td>
</tr>

<tr>
<td><a href="showfeat.html">showfeat</a></td>
<td>Display features of a sequence in pretty format</td>
</tr>

<tr>
<td><a href="showorf.html">showorf</a></td>
<td>Display a nucleotide sequence and translation in pretty format</td>
</tr>

<tr>
<td><a href="showpep.html">showpep</a></td>
<td>Displays protein sequences with features in pretty format</td>
</tr>

<tr>
<td><a href="showseq.html">showseq</a></td>
<td>Displays sequences with features in pretty format</td>
</tr>

<tr>
<td><a href="silent.html">silent</a></td>
<td>Find restriction sites to insert (mutate) with no translation change</td>
</tr>

<tr>
<td><a href="sixpack.html">sixpack</a></td>
<td>Display a DNA sequence with 6-frame translation and ORFs</td>
</tr>

<tr>
<td><a href="textsearch.html">textsearch</a></td>
<td>Search the textual description of sequence(s)</td>
</tr>

<tr>
<td><a href="transeq.html">transeq</a></td>
<td>Translate nucleic acid sequences</td>
</tr>

</table>


<H2>
    Author(s)
</H2>


Gary Williams (gwilliam&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK



<H2>
    History
</H2>

Written Spring 2000
<p>

Changed 7 Dec 2000 - GWW - to declare isoschizomers that cut

<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.


<H2>
    Comments
</H2>
None

</BODY>
</HTML>

