<HTML>

<HEAD>
  <TITLE>
  EMBOSS: restrict
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">






<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

restrict
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>






<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Finds restriction enzyme cleavage sites


<H2>
    Description
</H2>


Restrict uses the REBASE database of restriction enzymes to predict
cut sites in a DNA sequence. The program allows you to select a range
of cuts, whether the DNA is circular, whether IUB ambiguity codes are
used, whether blunt or sticky ends or both are reported. You may also
force the reporting of single cleavage sites.

<H2>
    Usage
</H2>


<b>Here is a sample session with restrict</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>restrict </b>
Finds restriction enzyme cleavage sites
Input sequence(s): <b>tembl:hsfau</b>
Minimum recognition site length [4]: <b></b>
Comma separated enzyme list [all]: <b></b>
Output report [hsfau.restrict]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
   -sitelen            integer    Minimum recognition site length
   -enzymes            string     The name 'all' reads in all enzyme names
                                  from the REBASE database. You can specify
                                  enzymes by giving their names with commas
                                  between then, such as:
                                  'HincII,hinfI,ppiI,hindiii'.
                                  The case of the names is not important. You
                                  can specify a file of enzyme names to read
                                  in by giving the name of the file holding
                                  the enzyme names with a '@' character in
                                  front of it, for example, '@enz.list'.
                                  Blank lines and lines starting with a hash
                                  character or '!' are ignored and all other
                                  lines are concatenated together with a comma
                                  character ',' and then treated as the list
                                  of enzymes to search for.
                                  An example of a file of enzyme names is:
                                  ! my enzymes
                                  HincII, ppiII
                                  ! other enzymes
                                  hindiii
                                  HinfI
                                  PpiI
  [-outfile]           report     Output report file name

   Optional qualifiers: (none)
   Advanced qualifiers:
   -min                integer    Minimum cuts per RE
   -max                integer    Maximum cuts per RE
   -single             boolean    Force single site only cuts
   -[no]blunt          boolean    Allow blunt end cutters
   -[no]sticky         boolean    Allow sticky end cutters
   -[no]ambiguity      boolean    Allow ambiguous matches
   -plasmid            boolean    Allow circular DNA
   -[no]commercial     boolean    Only enzymes with suppliers
   -datafile           string     Alternative RE data file
   -[no]limit          boolean    Limits reports to one isoschizomer
   -preferred          boolean    Report preferred isoschizomers
   -alphabetic         boolean    Sort output alphabetically
   -fragments          boolean    Show fragment lengths
   -name               boolean    Show sequence name

   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-sitelen</td>
<td>Minimum recognition site length</td>
<td>Integer from 2 to 20</td>
<td>4</td>
</tr>

<tr>
<td>-enzymes</td>
<td>The name 'all' reads in all enzyme names from the REBASE database. You can specify enzymes by giving their names with commas between then, such as: 'HincII,hinfI,ppiI,hindiii'.
The case of the names is not important. You can specify a file of enzyme names to read in by giving the name of the file holding the enzyme names with a '@' character in front of it, for example, '@enz.list'.
Blank lines and lines starting with a hash character or '!' are ignored and all other lines are concatenated together with a comma character ',' and then treated as the list of enzymes to search for.
An example of a file of enzyme names is:
! my enzymes
HincII, ppiII
! other enzymes
hindiii
HinfI
PpiI</td>
<td>Any string is accepted</td>
<td>all</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output report file name</td>
<td>Report output file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-min</td>
<td>Minimum cuts per RE</td>
<td>Integer from 1 to 1000</td>
<td>1</td>
</tr>

<tr>
<td>-max</td>
<td>Maximum cuts per RE</td>
<td>Integer up to 2000000000</td>
<td>2000000000</td>
</tr>

<tr>
<td>-single</td>
<td>Force single site only cuts</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]blunt</td>
<td>Allow blunt end cutters</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]sticky</td>
<td>Allow sticky end cutters</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]ambiguity</td>
<td>Allow ambiguous matches</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-plasmid</td>
<td>Allow circular DNA</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]commercial</td>
<td>Only enzymes with suppliers</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-datafile</td>
<td>Alternative RE data file</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-[no]limit</td>
<td>Limits reports to one isoschizomer</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-preferred</td>
<td>Report preferred isoschizomers</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-alphabetic</td>
<td>Sort output alphabetically</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-fragments</td>
<td>Show fragment lengths</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-name</td>
<td>Show sequence name</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

</table>



<H2>
    Input file format
</H2>


<b>restrict</b> reads one or more DNA sequence USAs.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:hsfau' is a sequence entry in the example nucleic acid database 'tembl'
<p>
<p><h3>Database entry: tembl:hsfau</h3>
<table width="90%"><tr><td bgcolor="#FFDDFF">
<pre>
ID   HSFAU      standard; RNA; HUM; 518 BP.
XX
AC   X65923;
XX
SV   X65923.1
XX
DT   13-MAY-1992 (Rel. 31, Created)
DT   23-SEP-1993 (Rel. 37, Last updated, Version 10)
XX
DE   H.sapiens fau mRNA
XX
KW   fau gene.
XX
OS   Homo sapiens (human)
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia;
OC   Eutheria; Primates; Catarrhini; Hominidae; Homo.
XX
RN   [1]
RP   1-518
RA   Michiels L.M.R.;
RT   ;
RL   Submitted (29-APR-1992) to the EMBL/GenBank/DDBJ databases.
RL   L.M.R. Michiels, University of Antwerp, Dept of Biochemistry,
RL   Universiteisplein 1, 2610 Wilrijk, BELGIUM
XX
RN   [2]
RP   1-518
RX   MEDLINE; 93368957.
RA   Michiels L., Van der Rauwelaert E., Van Hasselt F., Kas K., Merregaert J.;
RT   " fau cDNA encodes a ubiquitin-like-S30 fusion protein and is expressed as
RT   an antisense sequences in the Finkel-Biskis-Reilly murine sarcoma virus";
RL   Oncogene 8:2537-2546(1993).
XX
DR   SWISS-PROT; P35544; UBIM_HUMAN.
DR   SWISS-PROT; Q05472; RS30_HUMAN.
XX
FH   Key             Location/Qualifiers
FH
FT   source          1..518
FT                   /chromosome="11q"
FT                   /db_xref="taxon:9606"
FT                   /organism="Homo sapiens"
FT                   /tissue_type="placenta"
FT                   /clone_lib="cDNA"
FT                   /clone="pUIA 631"
FT                   /map="13"
FT   misc_feature    57..278
FT                   /note="ubiquitin like part"
FT   CDS             57..458
FT                   /db_xref="SWISS-PROT:P35544"
FT                   /db_xref="SWISS-PROT:Q05472"
FT                   /gene="fau"
FT                   /protein_id="CAA46716.1"
FT                   /translation="MQLFVRAQELHTFEVTGQETVAQIKAHVASLEGIAPEDQVVLLAG
FT                   APLEDEATLGQCGVEALTTLEVAGRMLGGKVHGSLARAGKVRGQTPKVAKQEKKKKKTG
FT                   RAKRRMQYNRRFVNVVPTFGKKKGPNANS"
FT   misc_feature    98..102
FT                   /note="nucleolar localization signal"
FT   misc_feature    279..458
FT                   /note="S30 part"
FT   polyA_signal    484..489
FT   polyA_site      509
XX
SQ   Sequence 518 BP; 125 A; 139 C; 148 G; 106 T; 0 other;
     ttcctctttc tcgactccat cttcgcggta gctgggaccg ccgttcagtc gccaatatgc        60
     agctctttgt ccgcgcccag gagctacaca ccttcgaggt gaccggccag gaaacggtcg       120
     cccagatcaa ggctcatgta gcctcactgg agggcattgc cccggaagat caagtcgtgc       180
     tcctggcagg cgcgcccctg gaggatgagg ccactctggg ccagtgcggg gtggaggccc       240
     tgactaccct ggaagtagca ggccgcatgc ttggaggtaa agttcatggt tccctggccc       300
     gtgctggaaa agtgagaggt cagactccta aggtggccaa acaggagaag aagaagaaga       360
     agacaggtcg ggctaagcgg cggatgcagt acaaccggcg ctttgtcaac gttgtgccca       420
     cctttggcaa gaagaagggc cccaatgcca actcttaagt cttttgtaat tctggctttc       480
     tctaataaaa aagccactta gttcagtcaa aaaaaaaa                               518
//
</pre>
</td></tr></table><p>


<H2>
    Output file format
</H2>


<p>

The output is a standard EMBOSS report file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-rformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  The available format names are: embl,
genbank, gff, pir, swiss, trace, listfile, dbmotif, diffseq, excel,
feattable, motif, regions, seqtable, simple, srs, table, tagseq

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html</A>
for further information on report formats.

<p>


<p>

By default <b>restrict</b> writes a 'table' report file.

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: hsfau.restrict</h3>
<table width="90%"><tr><td bgcolor="#DDFFDD">
<pre>
########################################
# Program: restrict
# Rundate: Thu Nov 07 14:42:26 2002
# Report_format: table
# Report_file: hsfau.restrict
########################################

#=======================================
#
# Sequence: HSFAU     from: 1   to: 518
# HitCount: 53
#
# Minimum cuts per enzyme: 1
# Maximum cuts per enzyme: 2000000000
# Minimum length of recognition site: 4
# Blunt ends allowed
# Sticky ends allowed
# DNA is linear
# Ambiguities allowed
#
#=======================================

  Start     End   Score Enzyme_name Restriction_site 5prime 3prime 5primerev 3primerev
     11      14        0 TaqI        TCGA                 11     13         .         .
     28      31        0 AciI        CCGC                 25     27         .         .
     38      41        0 AciI        CCGC                 38     40         .         .
     44      48        0 BceAI       ACGGC                25     27         .         .
     71      74        0 AciI        CCGC                 71     73         .         .
     71      81        0 Bsc4I       CCNNNNNNNGG          77     74         .         .
     73      76        0 HhaI        GCGC                 75     73         .         .
     73      76        0 Hin6I       GCGC                 73     75         .         .
     77      81        0 BssKI       CCNGG                76     81         .         .
     77      81        0 EcoRII      CCWGG                76     81         .         .
     94      97        0 TaqI        TCGA                 94     96         .         .
    103     106        0 BsiSI       CCGG                103    105         .         .
    105     108        0 BshI        GGCC                106    106         .         .
    107     117        0 Bsc4I       CCNNNNNNNGG         113    110         .         .
    107     111        0 BssKI       CCNGG               106    111         .         .
    107     111        0 EcoRII      CCWGG               106    111         .         .
    122     132        0 Bsc4I       CCNNNNNNNGG         128    125         .         .
    125     135        0 Hin4I       GAYNNNNNVTC         116    111       148       143
    146     150        0 Bse1I       ACTGG               151    149         .         .
    161     165        0 BssKI       CCNGG               160    165         .         .
    162     165        0 BsiSI       CCGG                162    164         .         .
    182     186        0 BssKI       CCNGG               181    186         .         .
    182     186        0 EcoRII      CCWGG               181    186         .         .
    190     193        0 HhaI        GCGC                192    190         .         .
    190     193        0 Hin6I       GCGC                190    192         .         .
    192     195        0 Hin6I       GCGC                192    194         .         .
    192     195        0 HhaI        GCGC                194    192         .         .
    197     201        0 BssKI       CCNGG               196    201         .         .
    197     201        0 EcoRII      CCWGG               196    201         .         .
    209     212        0 BshI        GGCC                210    210         .         .
    219     222        0 BshI        GGCC                220    220         .         .
    221     231        0 Bsc4I       CCNNNNNNNGG         227    224         .         .
    225     229        0 Bse1I       ACTGG               221    219         .         .
    229     232        0 AciI        CCGC                226    228         .         .
    236     239        0 BshI        GGCC                237    237         .         .
    248     252        0 BssKI       CCNGG               247    252         .         .
    248     252        0 EcoRII      CCWGG               247    252         .         .
    261     264        0 BshI        GGCC                262    262         .         .
    263     266        0 AciI        CCGC                263    265         .         .
    293     297        0 BssKI       CCNGG               292    297         .         .
    293     297        0 EcoRII      CCWGG               292    297         .         .
    296     299        0 BshI        GGCC                297    297         .         .
    335     338        0 BshI        GGCC                336    336         .         .
    380     383        0 AciI        CCGC                377    379         .         .
    383     386        0 AciI        CCGC                380    382         .         .
    395     398        0 BsiSI       CCGG                395    397         .         .
    398     401        0 HhaI        GCGC                400    398         .         .
    398     401        0 Hin6I       GCGC                398    400         .         .
    405     410        0 HindII      GTYRAC              407    407         .         .
    408     413        0 AclI        AACGTT              409    411         .         .
    409     412        0 HpyCH4IV    ACGT                409    411         .         .
    417     427        0 Bsc4I       CCNNNNNNNGG         423    420         .         .
    438     441        0 BshI        GGCC                439    439         .         .

#---------------------------------------
#---------------------------------------
</pre>
</td></tr></table><p>


<p>

The output from restrict is a simple text one. The base number,
restriction enzyme name, recognition site and cut positions are
shown. Note that cuts are always to the right of the residue
shown and that 5' cuts are referred to by their associated 3'
number sequence.

<p>

The program reports enzymes that cut at two or four sites. The
program also reports isoschizomers and enzymes having the same
recognition sequence but different cut sites.


<H2>
    Data files
</H2>


The data files are stored in the REBASE directory of the standard EMBOSS
data directory. The names are:

<ul>
   <li> embossre.enz     Cleavage information
   <li> embossre.ref     Reference/methylation information
   <li> embossre.sup     Supplier information
</ul>

The column information is described at the top of the data files

<H2>
    Notes
</H2>


Output file size is related to the size of the recognition site and the
maximum number of allowed cutting positions. Setting the site length
to six and restricting the cuts to two is a common choice of parameters.
The size of the output can sometimes be reduced by specifying
the -noambiguity switch.
<p>

The data files must have been created before running this program. This is
done by running the <b>rebaseextract</b> program with the "withrefm" file from
an REBASE release. You may have to ask your system manager to do this.

<H2>
    References
</H2>


<ol>

<li>
Nucleic Acids Research 27: 312-313 (1999).

</ol>

<H2>
    Warnings
</H2>


The program will warn you if a protein sequence is given.

<H2>
    Diagnostic Error Messages
</H2>

None.


<H2>
    Exit status
</H2>


It exits with status 0 unless an error is reported.

<H2>
    Known bugs
</H2>


None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="recoder.html">recoder</a></td><td>Remove restriction sites but maintain the same translation</td></tr>
<tr><td><a href="redata.html">redata</a></td><td>Search REBASE for enzyme name, references, suppliers etc</td></tr>
<tr><td><a href="remap.html">remap</a></td><td>Display a sequence with restriction cut sites, translation etc</td></tr>
<tr><td><a href="restover.html">restover</a></td><td>Finds restriction enzymes that produce a specific overhang</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="silent.html">silent</a></td><td>Silent mutation restriction enzyme scan</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->




<H2>
    Author(s)
</H2>


This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>


Completed 16th April 1999

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>




</BODY>
</HTML>

