<HTML>

<HEAD>
  <TITLE>
  EMBOSS: restrict
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">
<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
restrict
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
Report restriction enzyme cleavage sites in a nucleotide sequence

<H2>
    Description
</H2>


<p><b>restrict</b> scans one or more nucleotide sequences for cut sites for a supplied set of restriction enzymes.  One or more restriction enzymes can be specified or alternatively all the enzymes in the REBASE database can be investigated.  The minimum length of a recognition site to be reported must be specified. It writes an output file showing the location of the cut sites. There are several options to control exactly what sites are reported and the format of the output file. Optionally, the fragment lengths of the forward sense strand produced by complete restriction by each restriction enzyme on its own, or by using all of the input enzymes together, may be reported. Results are added to the tail section of the report.</p>


<H2>
Algorithm
</H2>



<H2>
    Usage
</H2>
<b>Here is a sample session with restrict</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>restrict </b>
Report restriction enzyme cleavage sites in a nucleotide sequence
Input nucleotide sequence(s): <b>tembl:x65923</b>
Minimum recognition site length [4]: <b></b>
Comma separated enzyme list [all]: <b></b>
Output report [x65923.restrict]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>
<p>
<b>Example 2</b>
<p>
This gives the lengths of the restriction fragments produced by cutting with all of the specified enzymes. 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>restrict -fragments </b>
Report restriction enzyme cleavage sites in a nucleotide sequence
Input nucleotide sequence(s): <b>tembl:x65923</b>
Minimum recognition site length [4]: <b></b>
Comma separated enzyme list [all]: <b></b>
Output report [x65923.restrict]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.2">Go to the output files for this example</a><p><p>
<p>
<b>Example 3</b>
<p>
This gives the lengths of the restriction fragments created by cutting with just one of each of the specified enzymes in turn. 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>restrict -solofragment </b>
Report restriction enzyme cleavage sites in a nucleotide sequence
Input nucleotide sequence(s): <b>tembl:x65923</b>
Minimum recognition site length [4]: <b></b>
Comma separated enzyme list [all]: <b></b>
Output report [x65923.restrict]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.3">Go to the output files for this example</a><p><p>


<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
Report restriction enzyme cleavage sites in a nucleotide sequence
Version: EMBOSS:6.2.0

   Standard (Mandatory) qualifiers:
  [-sequence]          seqall     Nucleotide sequence(s) filename and optional
                                  format, or reference (input USA)
   -sitelen            integer    [4] This sets the minimum length of the
                                  restriction enzyme recognition site. Any
                                  enzymes with sites shorter than this will be
                                  ignored. (Integer from 2 to 20)
   -enzymes            string     [all] The name 'all' reads in all enzyme
                                  names from the REBASE database. You can
                                  specify enzymes by giving their names with
                                  commas between then, such as:
                                  'HincII,hinfI,ppiI,hindiii'.
                                  The case of the names is not important. You
                                  can specify a file of enzyme names to read
                                  in by giving the name of the file holding
                                  the enzyme names with a '@' character in
                                  front of it, for example, '@enz.list'.
                                  Blank lines and lines starting with a hash
                                  character or '!' are ignored and all other
                                  lines are concatenated together with a comma
                                  character ',' and then treated as the list
                                  of enzymes to search for.
                                  An example of a file of enzyme names is:
                                  ! my enzymes
                                  HincII, ppiII
                                  ! other enzymes
                                  hindiii
                                  HinfI
                                  PpiI (Any string)
  [-outfile]           report     [*.restrict] Output report file name

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers:
   -datafile           datafile   Restriction enzyme data file (optional)
   -mfile              datafile   [Emethylsites.dat] Restriction enzyme
                                  methylation data file (optional)
   -min                integer    [1] This sets the minimum number of cuts for
                                  any restriction enzyme that will be
                                  considered. Any enzymes that cut fewer times
                                  than this will be ignored. (Integer from 1
                                  to 1000)
   -max                integer    [2000000000] This sets the maximum number of
                                  cuts for any restriction enzyme that will
                                  be considered. Any enzymes that cut more
                                  times than this will be ignored. (Integer up
                                  to 2000000000)
   -solofragment       boolean    [N] This gives the fragment lengths of the
                                  forward sense strand produced by complete
                                  restriction by each restriction enzyme on
                                  its own. Results are added to the tail
                                  section of the report.
   -single             boolean    [N] If this is set then this forces the
                                  values of the mincuts and maxcuts qualifiers
                                  to both be 1. Any other value you may have
                                  set them to will be ignored.
   -[no]blunt          boolean    [Y] This allows those enzymes which cut at
                                  the same position on the forward and reverse
                                  strands to be considered.
   -[no]sticky         boolean    [Y] This allows those enzymes which cut at
                                  different positions on the forward and
                                  reverse strands, leaving an overhang, to be
                                  considered.
   -[no]ambiguity      boolean    [Y] This allows those enzymes which have one
                                  or more 'N' ambiguity codes in their
                                  pattern to be considered
   -plasmid            boolean    [N] If this is set then this allows searches
                                  for restriction enzyme recognition site and
                                  cut postions that span the end of the
                                  sequence to be considered.
   -methylation        boolean    [N] If this is set then RE recognition sites
                                  will not match methylated bases.
   -[no]commercial     boolean    [Y] If this is set, then only those enzymes
                                  with a commercial supplier will be searched
                                  for. This qualifier is ignored if you have
                                  specified an explicit list of enzymes to
                                  search for, rather than searching through
                                  'all' the enzymes in the REBASE database. It
                                  is assumed that, if you are asking for an
                                  explicit enzyme, then you probably know
                                  where to get it from and so all enzymes
                                  names that you have asked to be searched
                                  for, and which cut, will be reported whether
                                  or not they have a commercial supplier.
   -[no]limit          boolean    [Y] This limits the reporting of enzymes to
                                  just one enzyme from each group of
                                  isoschizomers. The enzyme chosen to
                                  represent an isoschizomer group is the
                                  prototype indicated in the data file
                                  'embossre.equ', which is created by the
                                  program 'rebaseextract'. If you prefer
                                  different prototypes to be used, make a copy
                                  of embossre.equ in your home directory and
                                  edit it. If this value is set to be false
                                  then all of the input enzymes will be
                                  reported. You might like to set this to
                                  false if you are supplying an explicit set
                                  of enzymes rather than searching 'all' of
                                  them.
   -alphabetic         boolean    [N] Sort output alphabetically
   -fragments          boolean    [N] This gives the fragment lengths of the
                                  forward sense strand produced by complete
                                  restriction using all of the input enzymes
                                  together. Results are added to the tail
                                  section of the report.
   -name               boolean    [N] Show sequence name

   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1            integer    Start of each sequence to be used
   -send1              integer    End of each sequence to be used
   -sreverse1          boolean    Reverse (if DNA)
   -sask1              boolean    Ask for begin/end/reverse
   -snucleotide1       boolean    Sequence is nucleotide
   -sprotein1          boolean    Sequence is protein
   -slower1            boolean    Make lower case
   -supper1            boolean    Make upper case
   -sformat1           string     Input sequence format
   -sdbname1           string     Database name
   -sid1               string     Entryname
   -ufo1               string     UFO features
   -fformat1           string     Features format
   -fopenfile1         string     Features file name

   "-outfile" associated qualifiers
   -rformat2           string     Report format
   -rname2             string     Base file name
   -rextension2        string     File name extension
   -rdirectory2        string     Output directory
   -raccshow2          boolean    Show accession number in the report
   -rdesshow2          boolean    Show description in the report
   -rscoreshow2        boolean    Show the score in the report
   -rstrandshow2       boolean    Show the nucleotide strand in the report
   -rusashow2          boolean    Show the full USA in the report
   -rmaxall2           integer    Maximum total hits to report
   -rmaxseq2           integer    Maximum hits to report for one sequence

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-sequence]<br>(Parameter 1)</td>
<td>seqall</td>
<td>Nucleotide sequence(s) filename and optional format, or reference (input USA)</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-sitelen</td>
<td>integer</td>
<td>This sets the minimum length of the restriction enzyme recognition site. Any enzymes with sites shorter than this will be ignored.</td>
<td>Integer from 2 to 20</td>
<td>4</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-enzymes</td>
<td>string</td>
<td>The name 'all' reads in all enzyme names from the REBASE database. You can specify enzymes by giving their names with commas between then, such as: 'HincII,hinfI,ppiI,hindiii'.
The case of the names is not important. You can specify a file of enzyme names to read in by giving the name of the file holding the enzyme names with a '@' character in front of it, for example, '@enz.list'.
Blank lines and lines starting with a hash character or '!' are ignored and all other lines are concatenated together with a comma character ',' and then treated as the list of enzymes to search for.
An example of a file of enzyme names is:
! my enzymes
HincII, ppiII
! other enzymes
hindiii
HinfI
PpiI</td>
<td>Any string</td>
<td>all</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-outfile]<br>(Parameter 2)</td>
<td>report</td>
<td>Output report file name</td>
<td>Report output file</td>
<td><i>&lt;*&gt;</i>.restrict</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>-datafile</td>
<td>datafile</td>
<td>Restriction enzyme data file (optional)</td>
<td>Data file</td>
<td><i>File in the data file path</i></td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-mfile</td>
<td>datafile</td>
<td>Restriction enzyme methylation data file (optional)</td>
<td>Data file</td>
<td>Emethylsites.dat</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-min</td>
<td>integer</td>
<td>This sets the minimum number of cuts for any restriction enzyme that will be considered. Any enzymes that cut fewer times than this will be ignored.</td>
<td>Integer from 1 to 1000</td>
<td>1</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-max</td>
<td>integer</td>
<td>This sets the maximum number of cuts for any restriction enzyme that will be considered. Any enzymes that cut more times than this will be ignored.</td>
<td>Integer up to 2000000000</td>
<td>2000000000</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-solofragment</td>
<td>boolean</td>
<td>This gives the fragment lengths of the forward sense strand produced by complete restriction by each restriction enzyme on its own. Results are added to the tail section of the report.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-single</td>
<td>boolean</td>
<td>If this is set then this forces the values of the mincuts and maxcuts qualifiers to both be 1. Any other value you may have set them to will be ignored.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-[no]blunt</td>
<td>boolean</td>
<td>This allows those enzymes which cut at the same position on the forward and reverse strands to be considered.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-[no]sticky</td>
<td>boolean</td>
<td>This allows those enzymes which cut at different positions on the forward and reverse strands, leaving an overhang, to be considered.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-[no]ambiguity</td>
<td>boolean</td>
<td>This allows those enzymes which have one or more 'N' ambiguity codes in their pattern to be considered</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-plasmid</td>
<td>boolean</td>
<td>If this is set then this allows searches for restriction enzyme recognition site and cut postions that span the end of the sequence to be considered.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-methylation</td>
<td>boolean</td>
<td>If this is set then RE recognition sites will not match methylated bases.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-[no]commercial</td>
<td>boolean</td>
<td>If this is set, then only those enzymes with a commercial supplier will be searched for. This qualifier is ignored if you have specified an explicit list of enzymes to search for, rather than searching through 'all' the enzymes in the REBASE database. It is assumed that, if you are asking for an explicit enzyme, then you probably know where to get it from and so all enzymes names that you have asked to be searched for, and which cut, will be reported whether or not they have a commercial supplier.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-[no]limit</td>
<td>boolean</td>
<td>This limits the reporting of enzymes to just one enzyme from each group of isoschizomers. The enzyme chosen to represent an isoschizomer group is the prototype indicated in the data file 'embossre.equ', which is created by the program 'rebaseextract'. If you prefer different prototypes to be used, make a copy of embossre.equ in your home directory and edit it. If this value is set to be false then all of the input enzymes will be reported. You might like to set this to false if you are supplying an explicit set of enzymes rather than searching 'all' of them.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-alphabetic</td>
<td>boolean</td>
<td>Sort output alphabetically</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-fragments</td>
<td>boolean</td>
<td>This gives the fragment lengths of the forward sense strand produced by complete restriction using all of the input enzymes together. Results are added to the tail section of the report.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-name</td>
<td>boolean</td>
<td>Show sequence name</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-sequence" associated seqall qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sbegin1<br>-sbegin_sequence</td>
<td>integer</td>
<td>Start of each sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -send1<br>-send_sequence</td>
<td>integer</td>
<td>End of each sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sreverse1<br>-sreverse_sequence</td>
<td>boolean</td>
<td>Reverse (if DNA)</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sask1<br>-sask_sequence</td>
<td>boolean</td>
<td>Ask for begin/end/reverse</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -snucleotide1<br>-snucleotide_sequence</td>
<td>boolean</td>
<td>Sequence is nucleotide</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sprotein1<br>-sprotein_sequence</td>
<td>boolean</td>
<td>Sequence is protein</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -slower1<br>-slower_sequence</td>
<td>boolean</td>
<td>Make lower case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -supper1<br>-supper_sequence</td>
<td>boolean</td>
<td>Make upper case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sformat1<br>-sformat_sequence</td>
<td>string</td>
<td>Input sequence format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sdbname1<br>-sdbname_sequence</td>
<td>string</td>
<td>Database name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sid1<br>-sid_sequence</td>
<td>string</td>
<td>Entryname</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ufo1<br>-ufo_sequence</td>
<td>string</td>
<td>UFO features</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fformat1<br>-fformat_sequence</td>
<td>string</td>
<td>Features format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fopenfile1<br>-fopenfile_sequence</td>
<td>string</td>
<td>Features file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-outfile" associated report qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rformat2<br>-rformat_outfile</td>
<td>string</td>
<td>Report format</td>
<td>Any string</td>
<td>table</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rname2<br>-rname_outfile</td>
<td>string</td>
<td>Base file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rextension2<br>-rextension_outfile</td>
<td>string</td>
<td>File name extension</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rdirectory2<br>-rdirectory_outfile</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -raccshow2<br>-raccshow_outfile</td>
<td>boolean</td>
<td>Show accession number in the report</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rdesshow2<br>-rdesshow_outfile</td>
<td>boolean</td>
<td>Show description in the report</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rscoreshow2<br>-rscoreshow_outfile</td>
<td>boolean</td>
<td>Show the score in the report</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rstrandshow2<br>-rstrandshow_outfile</td>
<td>boolean</td>
<td>Show the nucleotide strand in the report</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rusashow2<br>-rusashow_outfile</td>
<td>boolean</td>
<td>Show the full USA in the report</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rmaxall2<br>-rmaxall_outfile</td>
<td>integer</td>
<td>Maximum total hits to report</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rmaxseq2<br>-rmaxseq_outfile</td>
<td>integer</td>
<td>Maximum hits to report for one sequence</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>


<H2>
    Input file format
</H2>


<b>restrict</b> reads one or more DNA sequence USAs.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:x65923' is a sequence entry in the example nucleic acid database 'tembl'
<p>
<p><h3>Database entry: tembl:x65923</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   X65923; SV 1; linear; mRNA; STD; HUM; 518 BP.
XX
AC   X65923;
XX
DT   13-MAY-1992 (Rel. 31, Created)
DT   18-APR-2005 (Rel. 83, Last updated, Version 11)
XX
DE   H.sapiens fau mRNA
XX
KW   fau gene.
XX
OS   Homo sapiens (human)
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia;
OC   Eutheria; Euarchontoglires; Primates; Haplorrhini; Catarrhini; Hominidae;
OC   Homo.
XX
RN   [1]
RP   1-518
RA   Michiels L.M.R.;
RT   ;
RL   Submitted (29-APR-1992) to the EMBL/GenBank/DDBJ databases.
RL   L.M.R. Michiels, University of Antwerp, Dept of Biochemistry,
RL   Universiteisplein 1, 2610 Wilrijk, BELGIUM
XX
RN   [2]
RP   1-518
RX   PUBMED; 8395683.
RA   Michiels L., Van der Rauwelaert E., Van Hasselt F., Kas K., Merregaert J.;
RT   "fau cDNA encodes a ubiquitin-like-S30 fusion protein and is expressed as
RT   an antisense sequence in the Finkel-Biskis-Reilly murine sarcoma virus";
RL   Oncogene 8(9):2537-2546(1993).
XX
DR   H-InvDB; HIT000322806.
XX
FH   Key             Location/Qualifiers
FH
FT   source          1..518
FT                   /organism="Homo sapiens"
FT                   /chromosome="11q"
FT                   /map="13"
FT                   /mol_type="mRNA"
FT                   /clone_lib="cDNA"
FT                   /clone="pUIA 631"
FT                   /tissue_type="placenta"
FT                   /db_xref="taxon:9606"
FT   misc_feature    57..278
FT                   /note="ubiquitin like part"
FT   CDS             57..458
FT                   /gene="fau"
FT                   /db_xref="GDB:135476"
FT                   /db_xref="GOA:P62861"
FT                   /db_xref="HGNC:3597"
FT                   /db_xref="HSSP:1GJZ"
FT                   /db_xref="InterPro:IPR006846"
FT                   /db_xref="UniProtKB/Swiss-Prot:P35544"
FT                   /protein_id="CAA46716.1"
FT                   /translation="MQLFVRAQELHTFEVTGQETVAQIKAHVASLEGIAPEDQVVLLAG
FT                   APLEDEATLGQCGVEALTTLEVAGRMLGGKVHGSLARAGKVRGQTPKVAKQEKKKKKTG
FT                   RAKRRMQYNRRFVNVVPTFGKKKGPNANS"
FT   misc_feature    98..102
FT                   /note="nucleolar localization signal"
FT   misc_feature    279..458
FT                   /note="S30 part"
FT   polyA_signal    484..489
FT   polyA_site      509
XX
SQ   Sequence 518 BP; 125 A; 139 C; 148 G; 106 T; 0 other;
     ttcctctttc tcgactccat cttcgcggta gctgggaccg ccgttcagtc gccaatatgc        60
     agctctttgt ccgcgcccag gagctacaca ccttcgaggt gaccggccag gaaacggtcg       120
     cccagatcaa ggctcatgta gcctcactgg agggcattgc cccggaagat caagtcgtgc       180
     tcctggcagg cgcgcccctg gaggatgagg ccactctggg ccagtgcggg gtggaggccc       240
     tgactaccct ggaagtagca ggccgcatgc ttggaggtaa agttcatggt tccctggccc       300
     gtgctggaaa agtgagaggt cagactccta aggtggccaa acaggagaag aagaagaaga       360
     agacaggtcg ggctaagcgg cggatgcagt acaaccggcg ctttgtcaac gttgtgccca       420
     cctttggcaa gaagaagggc cccaatgcca actcttaagt cttttgtaat tctggctttc       480
     tctaataaaa aagccactta gttcagtcaa aaaaaaaa                               518
//
</pre>
</td></tr></table><p>

<H2>
    Output file format
</H2>


<p>

The output is a standard EMBOSS report file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-rformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  The available format names are: embl,
genbank, gff, pir, swiss, trace, listfile, dbmotif, diffseq, excel,
feattable, motif, regions, seqtable, simple, srs, table, tagseq

<p>

See:
<A href="http://emboss.sf.net/docs/themes/ReportFormats.html">
http://emboss.sf.net/docs/themes/ReportFormats.html</A>
for further information on report formats.

<p>

<p>

By default <b>restrict</b> writes a 'table' report file.

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: x65923.restrict</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
########################################
# Program: restrict
# Rundate: Fri 15 Jan 2010 12:00:00
# Commandline: restrict
#    -sequence tembl:x65923
# Report_format: table
# Report_file: x65923.restrict
########################################

#=======================================
#
# Sequence: X65923     from: 1   to: 518
# HitCount: 54
#
# Minimum cuts per enzyme: 1
# Maximum cuts per enzyme: 2000000000
# Minimum length of recognition site: 4
# Blunt ends allowed
# Sticky ends allowed
# DNA is linear
# Ambiguities allowed
#
#=======================================

  Start     End  Strand Enzyme_name Restriction_site 5prime 3prime 5primerev 3primerev
     11      14       + TaqI        TCGA                 11     13         .         .
     25      28       - AciI        CCGC                 25     27         .         .
     31      36       - BseYI       CCCAGC               31     35         .         .
     38      41       + AciI        CCGC                 38     40         .         .
     40      44       - BceAI       ACGGC                25     27         .         .
     71      81       + BsiYI       CCNNNNNNNGG          77     74         .         .
     71      74       + AciI        CCGC                 71     73         .         .
     73      76       + Hin6I       GCGC                 73     75         .         .
     73      76       + HhaI        GCGC                 75     73         .         .
     77      81       + EcoRII      CCWGG                76     81         .         .
     77      81       + BssKI       CCNGG                76     81         .         .
     94      97       + TaqI        TCGA                 94     96         .         .
    103     106       + HpaII       CCGG                103    105         .         .
    105     108       + HaeIII      GGCC                106    106         .         .
    107     111       + EcoRII      CCWGG               106    111         .         .
    107     111       + BssKI       CCNGG               106    111         .         .
    107     117       + BsiYI       CCNNNNNNNGG         113    110         .         .
    122     132       + BsiYI       CCNNNNNNNGG         128    125         .         .
    125     135       + Hin4I       GAYNNNNNVTC         116    111       148       143
    146     150       + BsrI        ACTGG               151    149         .         .
    161     165       + BssKI       CCNGG               160    165         .         .
    162     165       + HpaII       CCGG                162    164         .         .
    182     186       + EcoRII      CCWGG               181    186         .         .
    182     186       + BssKI       CCNGG               181    186         .         .
    190     193       + Hin6I       GCGC                190    192         .         .
    190     193       + HhaI        GCGC                192    190         .         .
    192     195       + Hin6I       GCGC                192    194         .         .
    192     195       + HhaI        GCGC                194    192         .         .
    197     201       + EcoRII      CCWGG               196    201         .         .
    197     201       + BssKI       CCNGG               196    201         .         .
    209     212       + HaeIII      GGCC                210    210         .         .
    219     222       + HaeIII      GGCC                220    220         .         .
    221     231       + BsiYI       CCNNNNNNNGG         227    224         .         .
    221     225       - BsrI        ACTGG               221    219         .         .
    226     229       - AciI        CCGC                226    228         .         .
    236     239       + HaeIII      GGCC                237    237         .         .
    248     252       + EcoRII      CCWGG               247    252         .         .
    248     252       + BssKI       CCNGG               247    252         .         .
    261     264       + HaeIII      GGCC                262    262         .         .
    263     266       + AciI        CCGC                263    265         .         .
    293     297       + EcoRII      CCWGG               292    297         .         .
    293     297       + BssKI       CCNGG               292    297         .         .
    296     299       + HaeIII      GGCC                297    297         .         .
    335     338       + HaeIII      GGCC                336    336         .         .
    377     380       - AciI        CCGC                377    379         .         .
    380     383       - AciI        CCGC                380    382         .         .
    395     398       + HpaII       CCGG                395    397         .         .
    398     401       + Hin6I       GCGC                398    400         .         .
    398     401       + HhaI        GCGC                400    398         .         .
    405     410       + HindII      GTYRAC              407    407         .         .
    408     413       + AclI        AACGTT              409    411         .         .
    409     412       + MaeII       ACGT                409    411         .         .
    417     427       + BsiYI       CCNNNNNNNGG         423    420         .         .
    438     441       + HaeIII      GGCC                439    439         .         .

#---------------------------------------
#---------------------------------------

#---------------------------------------
# Total_sequences: 1
# Total_length: 518
# Reported_sequences: 1
# Reported_hitcount: 54
#---------------------------------------
</pre>
</td></tr></table><p>

<a name="output.2"></a>
<h3>Output files for usage example 2</h3>
<p><h3>File: x65923.restrict</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
########################################
# Program: restrict
# Rundate: Fri 15 Jan 2010 12:00:00
# Commandline: restrict
#    -fragments
#    -sequence tembl:x65923
# Report_format: table
# Report_file: x65923.restrict
########################################

#=======================================
#
# Sequence: X65923     from: 1   to: 518
# HitCount: 54
#
# Minimum cuts per enzyme: 1
# Maximum cuts per enzyme: 2000000000
# Minimum length of recognition site: 4
# Blunt ends allowed
# Sticky ends allowed
# DNA is linear
# Ambiguities allowed
#
#=======================================

  Start     End  Strand Enzyme_name Restriction_site 5prime 3prime 5primerev 3primerev
     11      14       + TaqI        TCGA                 11     13         .         .
     25      28       - AciI        CCGC                 25     27         .         .
     31      36       - BseYI       CCCAGC               31     35         .         .
     38      41       + AciI        CCGC                 38     40         .         .
     40      44       - BceAI       ACGGC                25     27         .         .
     71      81       + BsiYI       CCNNNNNNNGG          77     74         .         .
     71      74       + AciI        CCGC                 71     73         .         .
     73      76       + Hin6I       GCGC                 73     75         .         .
     73      76       + HhaI        GCGC                 75     73         .         .
     77      81       + EcoRII      CCWGG                76     81         .         .
     77      81       + BssKI       CCNGG                76     81         .         .
     94      97       + TaqI        TCGA                 94     96         .         .
    103     106       + HpaII       CCGG                103    105         .         .
    105     108       + HaeIII      GGCC                106    106         .         .
    107     111       + EcoRII      CCWGG               106    111         .         .
    107     111       + BssKI       CCNGG               106    111         .         .
    107     117       + BsiYI       CCNNNNNNNGG         113    110         .         .
    122     132       + BsiYI       CCNNNNNNNGG         128    125         .         .
    125     135       + Hin4I       GAYNNNNNVTC         116    111       148       143
    146     150       + BsrI        ACTGG               151    149         .         .
    161     165       + BssKI       CCNGG               160    165         .         .
    162     165       + HpaII       CCGG                162    164         .         .
    182     186       + EcoRII      CCWGG               181    186         .         .
    182     186       + BssKI       CCNGG               181    186         .         .


<font color=red>  [Part of this file has been deleted for brevity]</font>

#     29
#     20
#     19
#     17
#     16
#     15
#     15
#     14
#     14
#     14
#     12
#     11
#     10
#     10
#     10
#     9
#     9
#     9
#     7
#     7
#     7
#     6
#     5
#     5
#     3
#     3
#     3
#     3
#     3
#     2
#     2
#     2
#     2
#     2
#     2
#     2
#     2
#     1
#     1
#     1
#     1
#     1
#
#---------------------------------------

#---------------------------------------
# Total_sequences: 1
# Total_length: 518
# Reported_sequences: 1
# Reported_hitcount: 54
#---------------------------------------
</pre>
</td></tr></table><p>

<a name="output.3"></a>
<h3>Output files for usage example 3</h3>
<p><h3>File: x65923.restrict</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
########################################
# Program: restrict
# Rundate: Fri 15 Jan 2010 12:00:00
# Commandline: restrict
#    -solofragment
#    -sequence tembl:x65923
# Report_format: table
# Report_file: x65923.restrict
########################################

#=======================================
#
# Sequence: X65923     from: 1   to: 518
# HitCount: 54
#
# Minimum cuts per enzyme: 1
# Maximum cuts per enzyme: 2000000000
# Minimum length of recognition site: 4
# Blunt ends allowed
# Sticky ends allowed
# DNA is linear
# Ambiguities allowed
#
#=======================================

  Start     End  Strand Enzyme_name Restriction_site 5prime 3prime 5primerev 3primerev
     11      14       + TaqI        TCGA                 11     13         .         .
     25      28       - AciI        CCGC                 25     27         .         .
     31      36       - BseYI       CCCAGC               31     35         .         .
     38      41       + AciI        CCGC                 38     40         .         .
     40      44       - BceAI       ACGGC                25     27         .         .
     71      81       + BsiYI       CCNNNNNNNGG          77     74         .         .
     71      74       + AciI        CCGC                 71     73         .         .
     73      76       + Hin6I       GCGC                 73     75         .         .
     73      76       + HhaI        GCGC                 75     73         .         .
     77      81       + EcoRII      CCWGG                76     81         .         .
     77      81       + BssKI       CCNGG                76     81         .         .
     94      97       + TaqI        TCGA                 94     96         .         .
    103     106       + HpaII       CCGG                103    105         .         .
    105     108       + HaeIII      GGCC                106    106         .         .
    107     111       + EcoRII      CCWGG               106    111         .         .
    107     111       + BssKI       CCNGG               106    111         .         .
    107     117       + BsiYI       CCNNNNNNNGG         113    110         .         .
    122     132       + BsiYI       CCNNNNNNNGG         128    125         .         .
    125     135       + Hin4I       GAYNNNNNVTC         116    111       148       143
    146     150       + BsrI        ACTGG               151    149         .         .
    161     165       + BssKI       CCNGG               160    165         .         .
    162     165       + HpaII       CCGG                162    164         .         .
    182     186       + EcoRII      CCWGG               181    186         .         .
    182     186       + BssKI       CCNGG               181    186         .         .


<font color=red>  [Part of this file has been deleted for brevity]</font>

# BssKI:
# [CCNGG]
# 	15	21	30	45	51	54
# 	76	226
# 
# EcoRII:
# [CCWGG]
# 	15	30	45	51	75	76
# 	226
# 
# HaeIII:
# [GGCC]
# 	10	17	25	35	39	79
# 	103	104	106
# 
# HhaI:
# [GCGC]
# 	2	75	117	118	206
# 
# Hin4I:
# [GAYNNNNNVTC]
# 	32	116	370
# 
# Hin6I:
# [GCGC]
# 	2	73	117	120	206
# 
# HindII:
# [GTYRAC]
# 	111	407
# 
# HpaII:
# [CCGG]
# 	59	103	123	233
# 
# MaeII:
# [ACGT]
# 	109	409
# 
# TaqI:
# [TCGA]
# 	11	83	424
#
#---------------------------------------

#---------------------------------------
# Total_sequences: 1
# Total_length: 518
# Reported_sequences: 1
# Reported_hitcount: 54
#---------------------------------------
</pre>
</td></tr></table><p>

<p>

The output from restrict is a simple text one. The base number,
restriction enzyme name, recognition site and cut positions are
shown. Note that cuts are always to the right of the residue
shown and that 5' cuts are referred to by their associated 3'
number sequence.

<p>

The program reports enzymes that cut at two or four sites. The
program also reports isoschizomers and enzymes having the same
recognition sequence but different cut sites.


<p>

When the "-fragments" or "-solofragments" qualifiers are given then the sizes of
the fragments produced by either all of the specified enzymes cutting, or
by each enzyme cutting individually, are given in the 'tail' section
at the end of the report file.

<H2>
    Data files
</H2>

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>
<p>
The EMBOSS REBASE restriction enzyme data files are stored in
directory 'data/REBASE/*' under the EMBOSS installation directory.

<p>

These files must first be set up using the program <a
href="rebaseextract.html">'<b>rebaseextract</b>'</a>.  Running
'rebaseextract' may be the job of your system manager.

<p>
      
The data files are stored in the REBASE directory of the standard EMBOSS
data directory. The names are:

<ul>
<li> embossre.enz     Cleavage information
<li> embossre.ref     Reference/methylation information
<li> embossre.sup     Supplier information
</ul>
	 
The column information is described at the top of the data files
	 
<p>
	 
The reported enzyme from any one group of isoschizomers (the prototype)
is specified in the REBASE database and the information is held in the
data file 'embossre.equ'.  You may edit this file to set your own
preferred prototype, if you wish. 
	 
<p>
	 
The format of the file "embossre.equ" is
<br>
Enzyme-name Prototype-name
	 
<p>
	 
i.e.  two columns of enzyme names separated by a space.  The first name
of the pair of enzymes is the name that is not preferred and the second
is the preferred (prototype) name. 
	 
	          



<H2>
    Notes
</H2>

<p>Several criteria may be set to control what sites are reported: <tt>-min</tt>, <tt>-max</tt>, <tt>-single</tt> (minimum or maximum number of cuts, or single site cuts only.  <tt>-blunt</tt> (enzymes which cut at the same position on the forward and reverse strands). <tt>-sticky</tt> (enzymes which cut at different positions on the forward and reverse strands, leaving an overhang). <tt>-ambiguity</tt> (enzymes which have one or more <tt>N</tt> ambiguity codes in their pattern).  <tt>-commercial</tt> (enzymes with a commercial supplier). <tt>-plasmid</tt> (allows searches for restriction enzyme recognition site and cut postions that span the end of the sequence).</p>

<p>By default, only one enzyme of any group of isoschizomers (enzymes that have the same recognition site and cut positions) is reported. This behaviour can be changed by specifying <tt>-nolimit</tt>, in which case all isoschizomers are reported.  The default behaviour uses the representative enzyme of an isoschizomer group (the prototype) which is specified in the EMBOSS data file <tt>embossre.equ</tt>.  This file is generated from the REBASE database by running <b>rebaseextract</b>. You may edit this file to set your own preferred prototype,if you wish.</p> 

<p>Output file size is related to the size of the recognition site and the maximum number of allowed cutting positions. Setting the site length to six and restricting the cuts to two is a common choice of parameters. The size of the output can sometimes be reduced by specifying the <tt>-noambiguity</tt> switch.</p>



<H2>
    References
</H2>


<ol>

<li>
Nucleic Acids Research 27: 312-313 (1999).

</ol>

<H2>
    Warnings
</H2>
<p><b>restrict</b> uses the EMBOSS REBASE restriction enzyme data files stored in directory <tt>data/REBASE/*</tt> under the EMBOSS installation directory. These files must first be set up using the program <b>rebaseextract</b>. Running <b>rebaseextract</b> may be the job of your system manager.</p>



<H2>
    Diagnostic Error Messages
</H2>

None.


<H2>
    Exit status
</H2>


It exits with status 0 unless an error is reported.

<H2>
    Known bugs
</H2>


None.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="recoder.html">recoder</a></td>
<td>Find restriction sites to remove (mutate) with no translation change</td>
</tr>

<tr>
<td><a href="redata.html">redata</a></td>
<td>Retrieve information from REBASE restriction enzyme database</td>
</tr>

<tr>
<td><a href="remap.html">remap</a></td>
<td>Display restriction enzyme binding sites in a nucleotide sequence</td>
</tr>

<tr>
<td><a href="restover.html">restover</a></td>
<td>Find restriction enzymes producing a specific overhang</td>
</tr>

<tr>
<td><a href="showseq.html">showseq</a></td>
<td>Displays sequences with features in pretty format</td>
</tr>

<tr>
<td><a href="silent.html">silent</a></td>
<td>Find restriction sites to insert (mutate) with no translation change</td>
</tr>

</table>


<H2>
    Author(s)
</H2>


Alan Bleasby (ajb&nbsp;&copy;&nbsp;ebi.ac.uk)
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK


<H2>
    History
</H2>

Completed 16th April 1999

<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.


<H2>
    Comments
</H2>
None




</BODY>
</HTML>

