<HTML>

<HEAD>
  <TITLE>
  EMBOSS: restrict
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">






<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

restrict
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>






<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Finds restriction enzyme cleavage sites


<H2>
    Description
</H2>


Restrict uses the REBASE database of restriction enzymes to predict
cut sites in a DNA sequence. The program allows you to select a range
of cuts, whether the DNA is circular, whether IUB ambiguity codes are
used, whether blunt or sticky ends or both are reported. You may also
force the reporting of single cleavage sites.

<H2>
    Usage
</H2>


Here is a sample session with restrict.

<pre>

% restrict
Finds restriction enzyme cleavage sites
Input sequence(s): embl:hsfau
Minimum recognition site length [4]: 
Comma separated enzyme list [all]: 
Output file [hsfau.restrict]: 

</pre>

<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
   -sitelen            integer    Minimum recognition site length
   -enzymes            string     The name 'all' reads in all enzyme names
                                  from the REBASE database. You can specify
                                  enzymes by giving their names with commas
                                  between then, such as:
                                  'HincII,hinfI,ppiI,hindiii'.
                                  The case of the names is not important. You
                                  can specify a file of enzyme names to read
                                  in by giving the name of the file holding
                                  the enzyme names with a '@' character in
                                  front of it, for example, '@enz.list'.
                                  Blank lines and lines starting with a hash
                                  character or '!' are ignored and all other
                                  lines are concatenated together with a comma
                                  character ',' and then treated as the list
                                  of enzymes to search for.
                                  An example of a file of enzyme names is:
                                  ! my enzymes
                                  HincII, ppiII
                                  ! other enzymes
                                  hindiii
                                  HinfI
                                  PpiI
  [-outfile]           report     Output report file name

   Optional qualifiers: (none)
   Advanced qualifiers:
   -min                integer    Minimum cuts per RE
   -max                integer    Maximum cuts per RE
   -single             boolean    Force single site only cuts
   -[no]blunt          boolean    Allow blunt end cutters
   -[no]sticky         boolean    Allow sticky end cutters
   -[no]ambiguity      boolean    Allow ambiguous matches
   -plasmid            boolean    Allow circular DNA
   -[no]commercial     boolean    Only enzymes with suppliers
   -datafile           string     Alternative RE data file
   -[no]limit          boolean    Limits reports to one isoschizomer
   -preferred          boolean    Report preferred isoschizomers
   -alphabetic         boolean    Sort output alphabetically
   -fragments          boolean    Show fragment lengths
   -name               boolean    Show sequence name

   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-sitelen</td>
<td>Minimum recognition site length</td>
<td>Integer from 2 to 20</td>
<td>4</td>
</tr>

<tr>
<td>-enzymes</td>
<td>The name 'all' reads in all enzyme names from the REBASE database. You can specify enzymes by giving their names with commas between then, such as: 'HincII,hinfI,ppiI,hindiii'.
The case of the names is not important. You can specify a file of enzyme names to read in by giving the name of the file holding the enzyme names with a '@' character in front of it, for example, '@enz.list'.
Blank lines and lines starting with a hash character or '!' are ignored and all other lines are concatenated together with a comma character ',' and then treated as the list of enzymes to search for.
An example of a file of enzyme names is:
! my enzymes
HincII, ppiII
! other enzymes
hindiii
HinfI
PpiI</td>
<td>Any string is accepted</td>
<td>all</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output report file name</td>
<td>Report file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-min</td>
<td>Minimum cuts per RE</td>
<td>Integer from 1 to 1000</td>
<td>1</td>
</tr>

<tr>
<td>-max</td>
<td>Maximum cuts per RE</td>
<td>Integer up to 2000000000</td>
<td>2000000000</td>
</tr>

<tr>
<td>-single</td>
<td>Force single site only cuts</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]blunt</td>
<td>Allow blunt end cutters</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]sticky</td>
<td>Allow sticky end cutters</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]ambiguity</td>
<td>Allow ambiguous matches</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-plasmid</td>
<td>Allow circular DNA</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]commercial</td>
<td>Only enzymes with suppliers</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-datafile</td>
<td>Alternative RE data file</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-[no]limit</td>
<td>Limits reports to one isoschizomer</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-preferred</td>
<td>Report preferred isoschizomers</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-alphabetic</td>
<td>Sort output alphabetically</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-fragments</td>
<td>Show fragment lengths</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-name</td>
<td>Show sequence name</td>
<td>Yes/No</td>
<td>No</td>
</tr>

</table>



<H2>
    Input file format
</H2>


The input sequence can be one or more DNA sequences.

<H2>
    Output file format
</H2>


<p>

The output is a standard EMBOSS report file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-rformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  The available format names are: embl,
genbank, gff, pir, swiss, trace, listfile, dbmotif, diffseq, excel,
feattable, motif, regions, seqtable, simple, srs, table, tagseq

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html</A>
for further information on report formats.

<p>


<p>

By default <b>restrict</b> writes a 'table' report file.

<p>

The output from restrict is a simple text one. The base number,
restriction enzyme name, recognition site and cut positions are
shown. Note that cuts are always to the right of the residue
shown and that 5' cuts are referred to by their associated 3'
number sequence.

<p>

The program reports enzymes that cut at two or four sites. The
program also reports isoschizomers and enzymes having the same
recognition sequence but different cut sites.

<p>

Here is part of a sample output:

<p>

<pre>
<hr>
########################################
# Program: restrict
# Rundate: Mon Feb 11 13:46:44 2002
# Report_file: hsfau.restrict
########################################

#=======================================
#
# Sequence: HSFAU     from: 1   to: 518
# HitCount: 214
#
# Minimum cuts per enzyme: 1
# Maximum cuts per enzyme: 2000000000
# Minimum length of recognition site: 4
# Blunt ends allowed
# Sticky ends allowed
# DNA is linear
# Ambiguities allowed
#
#
#=======================================

USA               Start     End   Score Enzyme_name Restriction_site 5prime 3prime 5primerev 3primerev
HSFAU                 3       6   0.000 MnlI        CCTC             13     12   .         .
HSFAU                 9      14   0.000 Hpy188III   TCNNGA           10     12   .         .
HSFAU                11      14   0.000 TaqI        TCGA             11     13   .         .
HSFAU                13      17   0.000 HinfI       GANTC            13     16   .         .
HSFAU                17      21   0.000 MlyI        GAGTC            7      7   .         .
HSFAU                17      21   0.000 PleI        GAGTC            7      8   .         .
HSFAU                24      27   0.000 AccII       CGCG             25     25   .         .
HSFAU                24      28   0.000 MboII       GAAGA            12     11   .         .
HSFAU                28      31   0.000 AciI        CCGC             25     27   .         .

etc.

HSFAU               437     443   0.000 DraII       RGGNCCY          438    441   .         .
HSFAU               438     442   0.000 AspS9I      GGNCC            438    441   .         .
HSFAU               438     443   0.000 BscBI       GGNNCC           440    440   .         .
HSFAU               438     441   0.000 BshFI       GGCC             439    439   .         .
HSFAU               438     441   0.000 CviJI       RGCY             439    439   .         .
HSFAU               454     459   0.000 AflII       CTTAAG           454    458   .         .
HSFAU               454     459   0.000 SmlI        CTYRAG           454    458   .         .
HSFAU               455     458   0.000 MseI        TTAA             455    457   .         .
HSFAU               468     471   0.000 Sse9I       AATT             467    471   .         .
HSFAU               474     477   0.000 CviJI       RGCY             475    475   .         .
HSFAU               492     495   0.000 CviJI       RGCY             493    493   .         .
HSFAU               497     501   0.000 BstDEI      CTNAG            497    500   .         .

#---------------------------------------
#---------------------------------------
<hr>
</pre>

<H2>
    Data files
</H2>


The data files are stored in the REBASE directory of the standard EMBOSS
data directory. The names are:

<ul>
   <li> embossre.enz     Cleavage information
   <li> embossre.ref     Reference/methylation information
   <li> embossre.sup     Supplier information
</ul>

The column information is described at the top of the data files

<H2>
    Notes
</H2>


Output file size is related to the size of the recognition site and the
maximum number of allowed cutting positions. Setting the site length
to six and restricting the cuts to two is a common choice of parameters.
The size of the output can sometimes be reduced by specifying
the -noambiguity switch.
<p>

The data files must have been created before running this program. This is
done by running the <b>rebaseextract</b> program with the "withrefm" file from
an REBASE release. You may have to ask your system manager to do this.

<H2>
    References
</H2>


<ol>

<li>
Nucleic Acids Research 27: 312-313 (1999).

</ol>

<H2>
    Warnings
</H2>


The program will warn you if a protein sequence is given.

<H2>
    Diagnostic Error Messages
</H2>


<H2>
    Exit status
</H2>


It exits with status 0 unless an error is reported.

<H2>
    Known bugs
</H2>


<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="recoder.html">recoder</a></td><td>Remove restriction sites but maintain the same translation</td></tr>
<tr><td><a href="redata.html">redata</a></td><td>Search REBASE for enzyme name, references, suppliers etc</td></tr>
<tr><td><a href="remap.html">remap</a></td><td>Display a sequence with restriction cut sites, translation etc</td></tr>
<tr><td><a href="restover.html">restover</a></td><td>Finds restriction enzymes that produce a specific overhang</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="silent.html">silent</a></td><td>Silent mutation restriction enzyme scan</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->




<H2>
    Author(s)
</H2>


This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>


Completed 16th April 1999

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>




</BODY>
</HTML>

