<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: restover
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">
<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>
<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>
</td>
<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: restover
</H2>
</font>
</td>
</tr>
</table>
</td>
</tr>
</table>
<p><hr><p>

<!--END OF HEADER-->

<H1>
Program restover
</H1>
<H2>
    Function
</H2>
Finds restriction enzymes that produce a specific overhang

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
        Algorithms
  -->

The Restriction Enzyme database (REBASE) is a collection of information
about restriction enzymes and related proteins.  It contains published
and unpublished references, recognition and cleavage sites,
isoschizomers, commercial availability, methylation sensitivity, crystal
and sequence data.  DNA methyltransferases, homing endonucleases,
nicking enzymes, specificity subunits and control proteins are also
included.  Most recently, putative DNA methyltransferases and
restriction enzymes, as predicted from analysis of genomic sequences,
are also listed. 

<p>

The home page of REBASE is: 
<A HREF="http://rebase.neb.com/">http://rebase.neb.com/</A>

<p>

<b>restover</b> takes a specified sequence and a short sequence of a
cut-site overhang and searches the REBASE database for matching enzymes
that create the desired overhang sequence when they cut the
input sequence. 












<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->


Here is a sample session with restover:

<pre>

% restover
Finds restriction enzymes that produce a specific overhang
Input sequence(s): em:hsfau
Overlap sequence: cg
Output file [hsfau.restover]: 

</pre>














<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
  [-seqcomp]           string     Overlap sequence
  [-outfile]           outfile    Output file name

   Optional qualifiers: (none)
   Advanced qualifiers:
   -min                integer    Minimum cuts per RE
   -max                integer    Maximum cuts per RE
   -single             bool       Force single site only cuts
   -threeprime         bool       3' overhang? (else 5') e.g. BamHI has CTAG
                                  as a 5' overhang, and ApaI has CCGG as 3'
                                  overhang.
   -[no]blunt          bool       Allow blunt end cutters
   -[no]sticky         bool       Allow sticky end cutters
   -[no]ambiguity      bool       Allow ambiguous matches
   -plasmid            bool       Allow circular DNA
   -[no]commercial     bool       Only enzymes with suppliers
   -datafile           string     Alternative RE data file
   -html               bool       Create HTML output
   -[no]limit          bool       Limits reports to one isoschizomer
   -preferred          bool       Report preferred isoschizomers
   -alphabetic         bool       Sort output alphabetically
   -fragments          bool       Show fragment lengths
   -name               bool       Show sequence name

   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-seqcomp]<br>(Parameter 2)</td>
<td>Overlap sequence</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 3)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.restover</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-min</td>
<td>Minimum cuts per RE</td>
<td>Integer from 1 to 1000</td>
<td>1</td>
</tr>

<tr>
<td>-max</td>
<td>Maximum cuts per RE</td>
<td>Integer up to 2000000000</td>
<td>2000000000</td>
</tr>

<tr>
<td>-single</td>
<td>Force single site only cuts</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-threeprime</td>
<td>3' overhang? (else 5') e.g. BamHI has CTAG as a 5' overhang, and ApaI has CCGG as 3' overhang.</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]blunt</td>
<td>Allow blunt end cutters</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]sticky</td>
<td>Allow sticky end cutters</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]ambiguity</td>
<td>Allow ambiguous matches</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-plasmid</td>
<td>Allow circular DNA</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]commercial</td>
<td>Only enzymes with suppliers</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-datafile</td>
<td>Alternative RE data file</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-html</td>
<td>Create HTML output</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]limit</td>
<td>Limits reports to one isoschizomer</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-preferred</td>
<td>Report preferred isoschizomers</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-alphabetic</td>
<td>Sort output alphabetically</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-fragments</td>
<td>Show fragment lengths</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-name</td>
<td>Show sequence name</td>
<td>Yes/No</td>
<td>No</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->


It reads in a normal DNA sequence USA.













<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->

The output file from the above example is:
<p>

<hr>
<pre>
# Restrict of HSFAU from 1 to 518
#
# Minimum cuts per enzyme: 1
# Maximum cuts per enzyme: 2000000000
# Minimum length of recognition site: 2
# Number of hits with any overlap: 208
# Base Number   Enzyme          Site            5'      3'      [5'     3']
        11      TaqI            TCGA            11      13
        28      AciI            CCGC            25      27
        38      AciI            CCGC            38      40
        71      AciI            CCGC            71      73
        73      Hin6I           GCGC            73      75
        94      TaqI            TCGA            94      96
        103     BsiSI           CCGG            103     105
        162     BsiSI           CCGG            162     164
        190     Hin6I           GCGC            190     192
        192     Hin6I           GCGC            192     194
        225     Bse1I           ACTGG           221     219
        229     AciI            CCGC            226     228
        263     AciI            CCGC            263     265
        380     AciI            CCGC            377     379
        383     AciI            CCGC            380     382
        395     BsiSI           CCGG            395     397
        398     Hin6I           GCGC            398     400
        408     AclI            AACGTT          409     411
        409     HpyCH4IV        ACGT            409     411
</pre>
<hr>

<p>

The output from <b>restover</b> is a simple text one.  The base number,
restriction enzyme name, recognition site and cut positions are shown. 
Note that cuts are always to the right of the residue shown and that 5'
cuts are referred to by their associated 3' number sequence.  The
program reports enzymes that cut at two or four sites.














<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.
	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

   -->

The data files are stored in the REBASE directory of the standard EMBOSS
data directory.  The names are:

<p>
<ul>
<li>      embossre.enz Cleavage information 
<li>      embossre.ref Reference/methylation information 
<li>      embossre.sup Supplier information 
</ul>
<p>
The column information is described at the top of the data files 













<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

The data files must have been created before running this program.  This
is done by running the <b>rebaseextract</b> program with the "withrefm" file
from an REBASE release.  You may have to ask your system manager to do
this. 








<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.






<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="recoder.html">recoder</a></td><td>Remove restriction sites but maintain the same translation</td></tr>
<tr><td><a href="redata.html">redata</a></td><td>Search REBASE for enzyme name, references, suppliers etc</td></tr>
<tr><td><a href="remap.html">remap</a></td><td>Display a sequence with restriction cut sites, translation etc</td></tr>
<tr><td><a href="restrict.html">restrict</a></td><td>Finds restriction enzyme cleavage sites</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="silent.html">silent</a></td><td>Silent mutation restriction enzyme scan</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

This application was written by Bernd Jagla
<A HREF="mailto:bernd@golgi.ski.mskcc.org">
bernd@golgi.ski.mskcc.org</A>
<br>

Cellular Biochemistry and Biophysics Program, Rockefeller
Research Laboratories, Memorial Sloan-Kettering Cancer Center, 1275 York
Avenue, Box 251,New York, NY 10021.
















<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (Jan 2001) - Bernd Jagla













<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->








</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

