<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: scope
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">
<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>
<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>
</td>
<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: scope
</H2>
</font>
</td>
</tr>
</table>
</td>
</tr>
</table>
<p><hr><p>

<!--END OF HEADER-->

<H1>
Program scope
</H1>
<H2>
    Function
</H2>
Convert raw scop classification file to embl-like format

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
        Algorithms
  -->

Nearly all proteins have structural similarities with other proteins
and, in some of these cases, share a common evolutionary origin.  A
knowledge of these relationships is crucial to our understanding of the
evolution of proteins and of development.  It will also play an
important role in the analysis of the sequence data that is being
produced by worldwide genome projects. 

<p>

The 
<A HREF="http://scop.mrc-lmb.cam.ac.uk/scop/">SCOP</A> 
database aims to provide a detailed and comprehensive
description of the structural and evolutionary relationships between all
proteins whose structure is known, including all entries in the Protein
Data Bank 
(<A HREF="http://pdb-browsers.ebi.ac.uk//">PDB</A>).  
<p>

<b>scope</b> reads the SCOP classification file available at
<A HREF="http://scop.mrc-lmb.cam.ac.uk/scop/search.cgi?dir=lin">
http://scop.mrc-lmb.cam.ac.uk/scop/search.cgi?dir=lin</A>
<p>

<b>scope</b> writes the SCOP classification to an EMBL-like format file.
<p>

No changes are made to the data other than changing the format in which it is held.
<p>

This EMBL-like format SCOP file is used by several other EMBOSS programs.
<p>

The reason why the SCOP database format is changed to an EMBL-like
format before being used used by other EMBOSS programs is that it is an easier
format to work with than the native SCOP database format. 












<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->


Here is a sample session with scope:
<p>

<pre>

% scope
Convert raw scop classification file to embl-like format
Name of scop file for input (raw format) [scop.orig]: /data/scop/scop.orig
Name of scop file for output (embl-like format) [Escop.dat]: Escop.test

</pre>














<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-infile]            infile     Name of scop file for input (raw format)
  [-outfile]           outfile    Name of scop file for output (embl-like
                                  format)

   Optional qualifiers: (none)
   Advanced qualifiers: (none)

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-infile]<br>(Parameter 1)</td>
<td>Name of scop file for input (raw format)</td>
<td>Input file</td>
<td>scop.orig</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Name of scop file for output (embl-like format)</td>
<td>Output file</td>
<td>Escop.dat</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->


The native format SCOP database input file is available at
<A HREF="http://scop.mrc-lmb.cam.ac.uk/scop/search.cgi?dir=lin">
http://scop.mrc-lmb.cam.ac.uk/scop/search.cgi?dir=lin</A>
<p>

The format of this file is explained at
<A HREF="http://scop.mrc-lmb.cam.ac.uk/scop/parindex.html">
http://scop.mrc-lmb.cam.ac.uk/scop/parindex.html</A>
<p>

The file given at this URL contains a single line for each domain in SCOP, 
including text describing the position of the domain in the SCOP hierarchy.
Note that other SCOP classification files, without this annotation, are 
available at
<A HREF="http://scop.mrc-lmb.cam.ac.uk/scop/parindex.html">
http://scop.mrc-lmb.cam.ac.uk/scop/parindex.html</A>








<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->



The output
records used to describe an entry are given below.  Records (4) to (8) 
are used to describe the position of the domain in the scop hierarchy.
<p>

<ol>

<LI>ID - Domain identifier code.  This is a 7-character code that uniquely
identifies the domain in scop.  It is identical to the first 7 characters 
of a line in the scop classification file.  The first character is always 
'D', the next four characters are the PDB identifier code, the fifth 
character is the PDB chain identifier to which the domain belongs (a '.' is
given in cases where the domain is composed of multiple chains, a '_' is 
given where a chain identifier was not specified in the PDB file) and the 
final character is the number of the domain in the chain (for chains 
comprising more than one domain) or '_' (the chain comprises a single 
domain only).

<LI>EN - PDB identifier code.  This is the 4-character PDB identifier code
of the PDB entry containing the domain.

<LI>OS - Source of the protein.  It is identical to the text given after 
'Species' in the scop classification file.

<LI>CL - Domain class.  It is identical to the text given after 'Class' in 
the scop classification file.

<LI>FO - Domain fold.  It is identical to the text given after 'Fold' in 
the scop classification file.

<LI>SF - Domain superfamily.  It is identical to the text given after 
'Superfamily' in the scop classification file.

<LI>FA - Domain family. It is identical to the text given after 'Family' in 
the scop classification file.

<LI>DO - Domain name. It is identical to the text given after 'Protein' in 
the scop classification file.

<LI>NC - Number of chains comprising the domain (usually 1).  If the number
of chains is greater than 1, then the domain entry will have a section  
containing a CN and a CH record (see below) for each chain.

<LI>CN - Chain number.  The number given in brackets after this record 
indicates the start of the data for the relevent chain.

<LI>CH - Domain definition.  The character given before CHAIN is the PDB 
chain identifier (a '.' is given in cases where a chain identifier was not 
specified in the scop classification file), the strings before START and 
END give the start and end positions respectively of the domain in the PDB 
file (a '.' is given in cases where a position was not specified).  Note 
that the start and end positions refer to residue numbering given in the 
original pdb file and therefore must be treated as strings.

</ol>

<p>

An example of an excerpt from an output file follows:

<p>

<hr>
<pre>
ID   D3SDHA_
XX
EN   3SDH
XX
OS   Ark clam (Scapharca inaequivalvis)
XX
CL   All alpha proteins
XX
FO   Globin-like
XX
SF   Globin-like
XX
FA   Globins
XX
DO   Hemoglobin I
XX
NC   1
XX
CN   [1]
XX
CH   a CHAIN; . START; . END;
//
ID   D3SDHB_
XX
EN   3SDH
XX
OS   Ark clam (Scapharca inaequivalvis)
XX
CL   All alpha proteins
XX
FO   Globin-like
XX
SF   Globin-like
XX
FA   Globins
XX
DO   Hemoglobin I
XX
NC   1
XX
CN   [1]
XX
CH   b CHAIN; . START; . END;
//
</pre>
<hr>





<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.
	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

   -->

None.













<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="domainer.html">domainer</a></td><td>Build domain coordinate files</td></tr>
<tr><td><a href="nrscope.html">nrscope</a></td><td>Converts redundant EMBL-format SCOP file to non-redundant one</td></tr>
<tr><td><a href="scopalign.html">scopalign</a></td><td>Generate alignments for SCOP families</td></tr>
<tr><td><a href="stamps.html">stamps</a></td><td>Generate alignments for SCOP families</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

This application was written by Jon Ison <a
href="mailto:jison@hgmp.mrc.ac.uk">(jison@hgmp.mrc.ac.uk)</a>
















<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (Jan 2001) - Jon Ison.















<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
This program is intended to be used by everyone and everything, from naive users to embedded scripts.

   -->

This program is intended to be run by EMBOSS site maintainers or those
responsible for setting up and maintaining protein 3D structural data
for use by others. 












<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->








</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

