<!-- START OF HEADER -->
<HTML><HEAD>
<TITLE> EMBASSY: SCOPPARSE documentation. </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

<H2> SCOPPARSE documentation
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<!-- END OF HEADER -->






<!-- CONTENTS
     This always includes the sections below.
     Other subsections can be added for individual applications.
-->
<H2>CONTENTS </H2>
<b> <a href="#1.0">1.0     SUMMARY                   </a></b><br>
<b> <a href="#2.0">2.0     INPUTS & OUTPUTS          </a></b><br>
<b> <a href="#3.0">3.0     INPUT FILE FORMAT         </a></b><br>
<b> <a href="#4.0">4.0     OUTPUT FILE FORMAT        </a></b><br>
<b> <a href="#5.0">5.0     DATA FILES                </a></b><br>
<b> <a href="#6.0">6.0     USAGE                     </a></b><br>    
<b> <a href="#7.0">7.0     KNOWN BUGS & WARNINGS     </a></b><br>    
<b> <a href="#8.0">8.0     NOTES                     </a></b><br>
<b> <a href="#9.0">9.0     DESCRIPTION               </a></b><br>
<b> <a href="#10.0">10.0   ALGORITHM                 </a></b><br>
<b> <a href="#11.0">11.0   RELATED APPLICATIONS      </a></b><br>
<b> <a href="#12.0">12.0   DIAGNOSTIC ERROR MESSAGES </a></b><br>
<b> <a href="#13.0">13.0   AUTHORS                   </a></b><br>
<b> <a href="#14.0">14.0   REFERENCES                </a></b><br>

<!-- SUMMARY
     Succint description of the application, particularly its inputs, outputs
     and what it does.  The same text is given at the top of the source (.c)
     file and in the <documentation> attribute of the <application definition>
     of the ACD file.
-->
<a name="1.0"></a>
<br><br><br><H2> 1.0   SUMMARY  </H2>

Generate DCF file from raw SCOP files




<!-- INPUTS & OUTPUTS
     Short summary of the application inputs and outputs in its different 
     modes of usage (if appropriate). More detail than the summary.
-->
<a name="2.0"></a>
<br><br><br><H2> 2.0   INPUTS & OUTPUTS          </H2>
SCOPPARSE parses the dir.cla.scop.txt and dir.des.scop.txt SCOP 
classification files, e.g. available at URLs:
<br> http://scop.mrc-lmb.cam.ac.uk/scop/parse/dir.cla.scop.txt_1.57
<br>    http://scop.mrc-lmb.cam.ac.uk/scop/parse/dir.des.scop.txt_1.57
<br>
<br>
The format of these files is explained at URL:
<br> http://scop.mrc-lmb.cam.ac.uk/scop/release-notes-1.55.html
<br>
<br>SCOPPARSE writes the classification to a DCF file (EMBL-like format).  
No changes are made to the data other than changing the 
format in which it is held.  The file does not include domain sequence 
information.  The input and output files are specified by the user.





<!-- INPUT FILE FORMAT 
     Description and example(s) of input file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.
     Use "<b>SCOPPARSE</b> reads any normal sequence USAs." if
     appropriate.
-->
<a name="3.0"></a>
<br><br><br><H2> 3.0   INPUT FILE FORMAT         </H2>
An excerpt from the dir.cla.scop.txt (Figure 1) and dir.des.scop.txt (Figure 2) SCOP 
input files is shown below.  The format of these files is explained
on the SCOP website:
<br>http://scop.mrc-lmb.cam.ac.uk/scop/parse/dir.cla.scop.txt_1.57


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: scop.cla.raw</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
# dir.cla.scop.txt 
# SCOP release 1.57 (January 2002)  [File format version 1.00]
# http://scop.mrc-lmb.cam.ac.uk/scop/
# Copyright (c) 1994-2002 the scop authors; see http://scop.mrc-lmb.cam.ac.uk/scop/lic/copy.html
d1cs4a_	1cs4	A:	d.58.29.1	39418	cl=53931,cf=54861,sf=55073,fa=55074,dm=55077,sp=55078,px=39418
d1ii7a_	1ii7	A:	d.159.1.4	62415	cl=53931,cf=56299,sf=56300,fa=64427,dm=64428,sp=64429,px=62415
</pre>
</td></tr></table><p>
<p><h3>File: scop.des.raw</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
# dir.des.scop.txt 
# SCOP release 1.57 (January 2002)  [File format version 1.00]
# http://scop.mrc-lmb.cam.ac.uk/scop/
# Copyright (c) 1994-2002 the scop authors; see http://scop.mrc-lmb.cam.ac.uk/scop/lic/copy.html
53931	cl	d	-	Alpha and beta proteins (a+b)
54861	cf	d.58	-	Ferredoxin-like
55073	sf	d.58.29	-	Adenylyl and guanylyl cyclase catalytic domain
55074	fa	d.58.29.1	-	Adenylyl and guanylyl cyclase catalytic domain
55077	dm	d.58.29.1	-	Adenylyl cyclase VC1, domain C1a
55078	sp	d.58.29.1	-	Dog (Canis familiaris)
39418	px	d.58.29.1	d1cs4a_	1cs4 A:
56299	cf	d.159	-	Metallo-dependent phosphatases
56300	sf	d.159.1	-	Metallo-dependent phosphatases
64427	fa	d.159.1.4	-	DNA double-strand break repair nuclease
64428	dm	d.159.1.4	-	Mre11
64429	sp	d.159.1.4	-	Archaeon Pyrococcus furiosus
62415	px	d.159.1.4	d1ii7a_	1ii7 A:
</pre>
</td></tr></table><p>


<!--
<br>
<br><b>Figure 1 Excerpt from dir.cla.scop.txt </b>
<table><td bgcolor="#CFCCFF">
<pre>
# dir.cla.scop.txt 
# SCOP release 1.57 (January 2002)  [File format version 1.00]
# http://scop.mrc-lmb.cam.ac.uk/scop/
# Copyright (c) 1994-2002 the SCOP authors; see http://scop.mrc-lmb.cam.ac.uk/scop/lic/copy.html
d1cs4a_	1cs4	A:	d.58.29.1	39418	cl=53931,cf=54861,sf=55073,fa=55074,dm=55077,sp=55078,px=39418
d1ii7a_	1ii7	A:	d.159.1.4	62415	cl=53931,cf=56299,sf=56300,fa=64427,dm=64428,sp=64429,px=62415
</pre>
</table>

<br>
<br><b>Figure 2 Excerpt from dir.des.scop.txt</b>
<table><td bgcolor="#CFCCFF">
<pre>
# dir.des.scop.txt 
# SCOP release 1.57 (January 2002)  [File format version 1.00]
# http://scop.mrc-lmb.cam.ac.uk/scop/
# Copyright (c) 1994-2002 the scop authors; see http://scop.mrc-lmb.cam.ac.uk/scop/lic/copy.html
53931	cl	d	-	Alpha and beta proteins (a+b)
54861	cf	d.58	-	Ferredoxin-like
55073	sf	d.58.29	-	Adenylyl and guanylyl cyclase catalytic domain
55074	fa	d.58.29.1	-	Adenylyl and guanylyl cyclase catalytic domain
55077	dm	d.58.29.1	-	Adenylyl cyclase VC1, domain C1a
55078	sp	d.58.29.1	-	Dog (Canis familiaris)
39418	px	d.58.29.1	d1cs4a_	1cs4 A:
56299	cf	d.159	-	Metallo-dependent phosphatases
56300	sf	d.159.1	-	Metallo-dependent phosphatases
64427	fa	d.159.1.4	-	DNA double-strand break repair nuclease
64428	dm	d.159.1.4	-	Mre11
64429	sp	d.159.1.4	-	Archaeon Pyrococcus furiosus
62415	px	d.159.1.4	d1ii7a_	1ii7 A:
</pre>
</table>

-->




<!-- OUTPUT FILE FORMAT 
     Description and example(s) of output file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.  
     If the standard description of the avalable report formats is required, 
     use:   #include file="inc/reportformats.ihtml"
     Use "Outputs a graph to the specified graphics device."
     or  "outputs a report format file. The default format is ..."
     if appropriate.
-->
<a name="4.0"></a>
<br><br><br><H2> 4.0   OUTPUT FILE FORMAT         </H2>
An example of the DCF output file is shown in Figure 3.
The records used to describe an entry are as follows.  Records (4) to (9) 
are used to describe the position of the domain in the SCOP hierarchy.
Various other <a href="#8.3"> ADDITIONAL RECORDS </a> may be present if the 
file is processed by other programs, e.g. DOMAINSEQS or DOMAINSSE.


<UL>
<LI>(1)  ID - Domain identifier code.  This is a 7-character code that uniquely
identifies the domain in SCOP.  It is identical to the first 7 characters 
of a line in the SCOP classification file.  The first character is always 
'D', the next four characters are the PDB identifier code, the fifth 
character is the PDB chain identifier to which the domain belongs (a '.' is
given in cases where the domain is composed of multiple chains, a '_' is 
given where a chain identifier was not specified in the PDB file) and the 
final character is the number of the domain in the chain (for chains 
comprising more than one domain) or '_' (the chain comprises a single 
domain only).
<LI>(2)  EN - PDB identifier code.  This is the 4-character PDB identifier code
of the PDB entry containing the domain.
<LI>(3) TY - domain type.  "CATH" or "SCOP" is given ("SCOP" for DCF files 
generated by using SCOPPARSE).
<LI>(4)  SI - SCOP Sunid's.  The integers preceeding the codes CL, FO, SF, FA,
DO, SO and DD are the SCOP sunids for Class, Fold, Superfamily, Family, 
Domain, Source and domain data respectively. These numbers uniquely 
identify the appropriate node in the SCOP parsable files.
<LI>(5)  CL - Domain class.  It is identical to the text given after 'Class' in 
the SCOP classification file.
<LI>(6)  FO - Domain fold.  It is identical to the text given after 'Fold' in 
the SCOP classification file.
<LI>(7)  SF - Domain superfamily.  It is identical to the text given after 
'Superfamily' in the SCOP classification file.
<LI>(8)  FA - Domain family. It is identical to the text given after 'Family' 
in the SCOP classification file.
<LI>(9)  DO - Domain name. It is identical to the text given after 'Protein' in 
the SCOP classification file.
<LI>(10) OS - Source of the protein.  It is identical to the text given after 
'Species' in the SCOP classification file.
<LI>(11) DS - Sequence of the domain according to the PDB file.  This sequence
is taken from the domain clean coordinate file generated by DOMAINER.  The DS    
record will only be present if the DCF file has been 
processed using DOMAINSEQS.
<LI>(12) NC - Number of chains comprising the domain, or number of segments 
from the same chain that the domain is comprised of.  NC is usually 1.  If
the number of chains is greater than 1, then the domain entry will have a 
section  containing a CN and a CH record (see below) for each chain.
<LI>(13) CN - Chain number.  The number given in brackets after this record 
indicates the start of the data for the relevent chain.
<LI>(14) CH - Domain definition.  The character given before CHAIN is the PDB 
chain identifier (a '.' is given in cases where a chain identifier was not 
specified in the DCF file), the strings before START and 
END give the start and end positions respectively of the domain in the PDB 
file (a '.' is given in cases where a position was not specified).  Note 
that the start and end positions refer to residue numbering given in the 
original PDB file and therefore must be treated as strings.
<LI>(15) XX - used for spacing.
<LI>(16) // - used to delimit records for a domain.
</UL>




<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: all.scop</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
ID   D1CS4A_
XX
EN   1CS4
XX
TY   SCOP
XX
SI   53931 CL; 54861 FO; 55073 SF; 55074 FA; 55077 DO; 55078 SO; 39418 DD;
XX
CL   Alpha and beta proteins (a+b)
XX
FO   Ferredoxin-like
XX
SF   Adenylyl and guanylyl cyclase catalytic domain
XX
FA   Adenylyl and guanylyl cyclase catalytic domain
XX
DO   Adenylyl cyclase VC1, domain C1a
XX
OS   Dog (Canis familiaris)
XX
NC   1
XX
CN   [1]
XX
CH   A CHAIN; . START; . END;
//
ID   D1II7A_
XX
EN   1II7
XX
TY   SCOP
XX
SI   53931 CL; 56299 FO; 56300 SF; 64427 FA; 64428 DO; 64429 SO; 62415 DD;
XX
CL   Alpha and beta proteins (a+b)
XX
FO   Metallo-dependent phosphatases
XX
SF   Metallo-dependent phosphatases
XX
FA   DNA double-strand break repair nuclease
XX
DO   Mre11
XX
OS   Archaeon Pyrococcus furiosus
XX
NC   1
XX
CN   [1]
XX
CH   A CHAIN; . START; . END;
//
</pre>
</td></tr></table><p>


<!--
<br><b>Figure 3 Excerpt from domain classification file</b>
<table><td bgcolor="#CFCCFF">
<pre>
ID   D3SDHA_
XX
EN   3SDH
XX
TY   SCOP
XX   
SI   46456 CL; 46457 FO; 46458 SF; 46463 FA; 46464 DO; 46465 SO; 14984 DD;
XX
CL   All alpha proteins
XX
FO   Globin-like
XX
SF   Globin-like
XX
FA   Globins
XX
DO   Hemoglobin I
XX
OS   Ark clam (Scapharca inaequivalvis)
XX
DS   SEQUENCE   146 AA;  15946 MW;  B7D52605 CRC32;
     PSVYDAAAQL TADVKKDLRD SWKVIGSDKK GNGVALMTTL FADNQETIGY FKRLGNVSQG
     MANDKLRGHS ITLMYALQNF IDQLDNPDDL VCVVEKFAVN HITRKISAAE FGKINGPIKK
     VLASKNFGDK YANAWAKLVA VVQAAL
XX
NC   1
XX
CN   [1]
XX
CH   A CHAIN; . START; . END;
//
ID   D3SDHB_
XX
EN   3SDH
XX
TY   SCOP
XX   
XX
SI   46456 CL; 46457 FO; 46458 SF; 46463 FA; 46464 DO; 46465 SO; 14985 DD;
XX
CL   All alpha proteins
XX
FO   Globin-like
XX
SF   Globin-like
XX
FA   Globins
XX
DO   Hemoglobin I
XX
OS   Ark clam (Scapharca inaequivalvis)
XX
DS   SEQUENCE   146 AA;  15946 MW;  B7D52605 CRC32;
     PSVYDAAAQL TADVKKDLRD SWKVIGSDKK GNGVALMTTL FADNQETIGY FKRLGNVSQG
     MANDKLRGHS ITLMYALQNF IDQLDNPDDL VCVVEKFAVN HITRKISAAE FGKINGPIKK
     VLASKNFGDK YANAWAKLVA VVQAAL
XX
NC   1
XX
CN   [1]
XX
CH   B CHAIN; . START; . END;
//
</pre>
</table>




-->



<!-- DATA FILES         
     Any data files used (e.g. translation table file, substitution matrix 
     etc.  This includes example data file formats if they are not obvious.
     For a standard description of what data files are and how embossdata can
     be used to inspect and retrieve them, use:
     #include file="inc/localfiles.ihtml" 
-->

<a name="5.0"></a>
<br><br><br><H2> 5.0   DATA FILES                </H2>
No data files are used.







<!-- USAGE
     Example usage, as run from the command-line.
     Many examples illustrating different behaviours is good.
-->
<a name="6.0"></a>
<br><br><br><H2> 6.0   USAGE                     </H2>
<pre>
   Standard (Mandatory) qualifiers:
  [-classfile]         infile     This option specifies the name of raw SCOP
                                  classification file dir.cla.scop.txt_X.XX
                                  (input). This is the raw SCOP classification
                                  file available at
                                  http://scop.mrc-lmb.cam.ac.uk/scop/parse/dir.cla.scop.txt_1.57.
  [-desinfile]         infile     This option specifies the name of raw SCOP
                                  description file dir.des.scop.txt_X.XX
                                  (input). This is the raw SCOP description
                                  file available at
                                  http://scop.mrc-lmb.cam.ac.uk/scop/parse/dir.des.scop.txt_1.57.
   -nosegments         boolean    This option specifies whether to omit
                                  domains comprising of more than one segment.
                                  This is necessary if a continuous residue
                                  sequence is required.
   -nomultichain       boolean    This option specifies whether to omit
                                  domains comprising segments from more than
                                  one chain. This is necessary if a continuous
                                  residue sequence is required.
  [-dcffile]           outfile    This option specifies the name of SCOP DCF
                                  file (domain classification file) (output).
                                  A 'domain classification file' contains
                                  classification and other data for domains
                                  from the SCOP or CATH databases. The file is
                                  generated by using DOMAINER and is in DCF
                                  format (EMBL-like). Domain sequence
                                  information can be added to the file by
                                  using DOMAINSEQS.

   Additional (Optional) qualifiers:
   -nominor            boolean    This option specifies whether to omit
                                  domains from minor classes (defined as
                                  anything not in class 'All alpha proteins',
                                  'All beta proteins', 'Alpha and beta
                                  proteins (a/b)' or 'Alpha and beta proteins
                                  (a+b)'). This is necessary or appropriate
                                  for many analyses.

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-dcffile" associated qualifiers
   -odirectory3        string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report deaths


</pre>

<H3> 6.1   COMMAND LINE ARGUMENTS </H3>
 
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-classfile]<br>(Parameter 1)</td>
<td>This option specifies the name of raw SCOP classification file dir.cla.scop.txt_X.XX (input). This is the raw SCOP classification file available at http://scop.mrc-lmb.cam.ac.uk/scop/parse/dir.cla.scop.txt_1.57.</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-desinfile]<br>(Parameter 2)</td>
<td>This option specifies the name of raw SCOP description file dir.des.scop.txt_X.XX (input). This is the raw SCOP description file available at http://scop.mrc-lmb.cam.ac.uk/scop/parse/dir.des.scop.txt_1.57.</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-nosegments</td>
<td>This option specifies whether to omit domains comprising of more than one segment. This is necessary if a continuous residue sequence is required.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-nomultichain</td>
<td>This option specifies whether to omit domains comprising segments from more than one chain. This is necessary if a continuous residue sequence is required.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>[-dcffile]<br>(Parameter 3)</td>
<td>This option specifies the name of SCOP DCF file (domain classification file) (output). A 'domain classification file' contains classification and other data for domains from the SCOP or CATH databases. The file is generated by using DOMAINER and is in DCF format (EMBL-like). Domain sequence information can be added to the file by using DOMAINSEQS.</td>
<td>Output file</td>
<td>test.scop</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-nominor</td>
<td>This option specifies whether to omit domains from minor classes (defined as anything not in class 'All alpha proteins', 'All beta proteins', 'Alpha and beta proteins (a/b)' or 'Alpha and beta proteins (a+b)'). This is necessary or appropriate for many analyses.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H3> 6.2 EXAMPLE SESSION </H3>
An example of interactive use of SCOPPARSE is shown below.

<b>Here is a sample session with scopparse</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>scopparse </b>
Generate DCF file from raw SCOP files.
Name of raw SCOP classification file dir.cla.scop.txt_X.XX (input).: <b>scop.cla.raw</b>
Name of raw SCOP description file dir.des.scop.txt_X.XX (input).: <b>scop.des.raw</b>
Omit domains comprising of more than one segment. [N]: <b>Y</b>
Omit domains comprising segments from more than one chain. [N]: <b>N</b>
Name of SCOP DCF file (domain classification file) (output). [test.scop]: <b>all.scop</b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>


<br>
<!--
<br>The raw SCOP classification files /test_data/scop.cla.raw and /test_data/scop.des.raw and a domain classification file in DCF (EMBL-like) format called /test_data/scopparse/all.scop was written. The output file does not contain domains that comprise segments one or more segments or segments from more than one chain. -->


<!-- KNOWN BUGS & WARNINGS
     Bugs that have not yet been fixed, easily missued features, problems
     and caveats etc. Potentially stupid things the program will let you do.
-->
<a name="7.0"></a>
<br><br><br><H2> 7.0   KNOWN BUGS & WARNINGS     </H2>

None.


<!-- NOTES
     Important general remarks, including:
       Restrictions.
       Interesting behaviour.
       Useful things you can do with this program.
       Future plans.
       etc.
-->
<a name="8.0"></a>
<br><br><br><H2> 8.0   NOTES                     </H2>
Some SCOP domains are comprised of more than one segments of polypeptide 
chain, these segments belonging to a single or more than one polypeptide 
chains.  It is debatable whether a domain (using the widely accepted 
definition) can truly consist of regions from more than polypeptide. 
Accordingly, SCOPPARSE gives the option of omitting from the output file 
domains that consist of more than one segment and domains that consist
of more than one segment where the segments are from different chains.
<br><br>
SCOP includes several minor classes which are not appropriate for some
anaylses.  Accordingly, SCOPPARSE gives the option to omit domains from 
minor classes.  This is defined as anything not in class 'All alpha 
proteins', 'All beta proteins', 'Alpha and beta proteins (a/b)' or 'Alpha 
and beta proteins (a+b)' 





<H3> 8.1   GLOSSARY OF FILE TYPES    </H2>
<table BORDER CELLSPACING=5 CELLPADDING=5 BGCOLOR="#f5f5ff" >
<tr>
<td><b>FILE TYPE</b></td>
<td><b>FORMAT</b></td>
<td><b>DESCRIPTION</b></td>
<td><b>CREATED BY <b></td>
<td><b>SEE ALSO</b></td>
</tr>
<tr>
<td><b>SCOP parsable files</b></td>
<td> SCOP format. </td>
<td> Raw SCOP classification data.  </td>
<td> Available from http://scop.mrc-lmb.cam.ac.uk/scop/parse/</a> </td>
<td> N.A.  </td>
</tr>
<tr>
<td><b> Domain classification file (for SCOP)</b></td>
<td> DCF format (EMBL-like format for domain classification data). </td>
<td> Classification and other data for domains from SCOP.  The file is in DCF format (EMBL-like). </td>
<td> <a href="scopparse.html">SCOPPARSE</a> </td>
<td> Domain sequence information can be added to the file by using DOMAINSEQS.  </td>
</tr>
</table>


<a name="8.3"></a>
<br><br><br><H2> 8.3   ADDITIONAL RECORDS                    </H2>
The following records for database sequence and secondary structure may be
present in a DCF file that has been processed by using DOMAINSEQS or DOMAINSSE.
<br>
<UL>
<LI>(1) AC - Accession number of the domain sequence.  This record will only 
be present if the DCF file has been processed using 
DOMAINSEQS and if an accession number for the PDB file corresponding to the 
domain is given in the swissprot:PDB equivalence file (generated by 
PDBTOSP) that DOMAINSEQS makes use of.
<LI>(2) SP - Swissprot code of the domain sequence.  This record will only 
be present if the domain classification file has been processed using 
DOMAINSEQS and if an swissprot code for the PDB file corresponding to the 
domain is given in the swissprot:PDB equivalence file (generated by 
PDBTOSP) that DOMAINSEQS makes use of.
<LI>(3) RA - Position of domain in swissprot sequence.  The integers 
preceeding START and END give the start and end points respectively of the
domain sequence relative to the full-length swissprot sequence.
<LI>(4) SQ - Sequence of the domain according to swissprot.  This sequence
is taken from the swissprot database.  The SQ record will only be 
present if the SCOP classification file has been processed using 
DOMAINSEQS and if an accession number for the PDB file corresponding to the 
domain is given in the swissprot:PDB equivalence file (generated by 
PDBTOSP) that DOMAINSEQS makes use of.
<LI>(5) SS - Secondary structure assignment (per residue). A simplified 3-state assignment:
H (helix), E (extended, beta strand) and L (loop, open coil) is used. 
<LI>(6) SE - Secondary structure element assignment (per secondary structure element).
The same 3-state assignment as for the SS record is used:
H (helix), E (extended, beta strand) and L (loop, open coil).
</UL>



<table><td bgcolor="#CFCCFF">
<pre>
XX
AC   P02213
XX
SP   GLB1_SCAIN
XX
RA   1 START; 146 END;
XX
SQ   SEQUENCE   146 AA;  15947 MW;  5868B4E5 CRC32;
     PSVYDAAAQL TADVKKDLRD SWKVIGSDKK GNGVALMTTL FADNQETIGY FKRLGDVSQG
     MANDKLRGHS ITLMYALQNF IDQLDNPDDL VCVVEKFAVN HITRKISAAE FGKINGPIKK
     VLASKNFGDK YANAWAKLVA VVQAAL
XX
AC   P02213
XX
SP   GLB1_SCAIN
XX
RA   1 START; 146 END; 
XX
SQ   SEQUENCE   146 AA;  15947 MW;  5868B4E5 CRC32;
     PSVYDAAAQL TADVKKDLRD SWKVIGSDKK GNGVALMTTL FADNQETIGY FKRLGDVSQG
     MANDKLRGHS ITLMYALQNF IDQLDNPDDL VCVVEKFAVN HITRKISAAE FGKINGPIKK
     VLASKNFGDK YANAWAKLVA VVQAAL
</pre>
</table>



None


<!-- DESCRIPTION
     A complete, non-technical, user-level description of the application.
-->
<a name="9.0"></a>
<br><br><br><H2> 9.0   DESCRIPTION               </H2>
The raw SCOP classification files are inconvenient for some uses because 
the text describing the domain classification is given in a different file
to the classification itself, the file formats are not easily extended and
differ from other related classifications such as CATH.
SCOPPARSE reads the raw SCOP classification files and writes a single file
in DCF (EMBL-like) format, which is an easier format to work with, is more 
human-readable and is more extensible than the native SCOP database format.





<!-- ALGORITHM
     A technical description of algorithmic aspects, describing exactly how
     the key aspects of the application work.
-->
<a name="10.0"></a>
<br><br><br><H2> 10.0  ALGORITHM                 </H2>
None.





<!-- RELATED APPLICATIONS
     Other applications that either generate the input, use the output or 
     are in some other way related to the application are described here. 
-->
<a name="11.0"></a>
<br><br><br><H2> 11.0  RELATED APPLICATIONS      </H2>

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="../../apps/aaindexextract.html">aaindexextract</a></td><td>Extract data from AAINDEX</td></tr>
<tr><td><a href="../domalign/allversusall.html">allversusall</a></td><td>Sequence similarity data from all-versus-all comparison</td></tr>
<tr><td><a href="cathparse.html">cathparse</a></td><td>Generates DCF file from raw CATH files</td></tr>
<tr><td><a href="../../apps/cutgextract.html">cutgextract</a></td><td>Extract data from CUTG</td></tr>
<tr><td><a href="../structure/domainer.html">domainer</a></td><td>Generates domain CCF files from protein CCF files</td></tr>
<tr><td><a href="domainnr.html">domainnr</a></td><td>Removes redundant domains from a DCF file</td></tr>
<tr><td><a href="domainseqs.html">domainseqs</a></td><td>Adds sequence records to a DCF file</td></tr>
<tr><td><a href="domainsse.html">domainsse</a></td><td>Add secondary structure records to a DCF file</td></tr>
<tr><td><a href="../structure/hetparse.html">hetparse</a></td><td>Converts heterogen group dictionary to EMBL-like format</td></tr>
<tr><td><a href="../structure/pdbparse.html">pdbparse</a></td><td>Parses PDB files and writes protein CCF files</td></tr>
<tr><td><a href="../structure/pdbplus.html">pdbplus</a></td><td>Add accessibility & secondary structure to a CCF file</td></tr>
<tr><td><a href="../structure/pdbtosp.html">pdbtosp</a></td><td>Convert swissprot:PDB codes file to EMBL-like format</td></tr>
<tr><td><a href="../../apps/printsextract.html">printsextract</a></td><td>Extract data from PRINTS</td></tr>
<tr><td><a href="../../apps/prosextract.html">prosextract</a></td><td>Build the PROSITE motif database for use by patmatmotifs</td></tr>
<tr><td><a href="../../apps/rebaseextract.html">rebaseextract</a></td><td>Extract data from REBASE</td></tr>
<tr><td><a href="../domsearch/seqnr.html">seqnr</a></td><td>Removes redundancy from DHF files</td></tr>
<tr><td><a href="../structure/sites.html">sites</a></td><td>Generate residue-ligand CON files from CCF files</td></tr>
<tr><td><a href="ssematch.html">ssematch</a></td><td>Search a DCF file for secondary structure matches</td></tr>
<tr><td><a href="../../apps/tfextract.html">tfextract</a></td><td>Extract data from TRANSFAC</td></tr>
</table>




<!-- DIAGNOSTIC ERROR MESSAGES
     Description of error messages or log file, if one is written.
-->
<a name="12.0"></a>
<br><br><br><H2> 12.0  DIAGNOSTIC ERROR MESSAGES </H2>
None.






<!-- AUTHORS
     The main author first then all others. 
-->
<a name="13.0"></a>
<br><br><br><H2> 13.0  AUTHORS                   </H2>

Alan Bleasby (ableasby@hgmp.mrc.ac.uk)

<br><br>Jon Ison <a href="mailto:jison@rfcgr.mrc.ac.uk">(jison@rfcgr.mrc.ac.uk)</a>
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK



<!-- REFERENCES
     Quote the paper where the application was first published, described, used etc. 
-->
<a name="14.0"></a>
<br><br><br><H2> 14.0  REFERENCES                </H2>

Please cite the authors and EMBOSS.
<br><br>
<i>Rice P, Longden I and Bleasby A (2000) "EMBOSS - The European
Molecular Biology Open Software Suite"  Trends in Genetics,
15:276-278.

<p>
See also <a href="http://emboss.sourceforge.net/">http://emboss.sourceforge.net/</a></i>



<H3>14.1 Other useful references  </H3>
1. Conte, L.L., Ailey, B., Hubbard, T.J. Brenner, S.E., Murzin, A.G. and Chothia, C. (2000) SCOP: a structural classification of proteins database. Nucleic Acids Res. 28, 257-259.


</BODY>
</HTML>
