<HTML>

<HEAD>
  <TITLE>
  EMBOSS: seealso
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
seealso
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
Finds programs with similar function to a specified program

<H2>
    Description
</H2>



<p><b>seealso</b> takes the name of an EMBOSS or EMBASSY program and returns a list of applications in the same functional group.  Application names and short descriptions are listed. Optionally, the output is written to an HTML table.</p>

<H2>
    Usage
</H2>
<b>Here is a sample session with seealso</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>seealso </b>
Finds programs with similar function to a specified program
Program to search for: <b>matcher</b>

SEE ALSO
seqmatchall  All-against-all word comparison of a sequence set
supermatcher Calculate approximate local pair-wise alignments of larger sequences
water        Smith-Waterman local alignment of sequences
wordfinder   Match large sequences against one or more other sequences
wordmatch    Finds regions of identity (exact matches) of two sequences


</pre></td></tr></table><p>
<p>
<p>


<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-search]            string     Enter the name of an EMBOSS program (Any
                                  string of at least 1 characters)

   Additional (Optional) qualifiers:
   -explode            boolean    [N] The groups that EMBOSS applications
                                  belong to have two forms, exploded and not
                                  exploded. The exploded group names are more
                                  numerous and often vaguely phrased than the
                                  non-exploded ones. The exploded names are
                                  formed from definitions of the group names
                                  that start like NAME1:NAME2 and which are
                                  then expanded into many combinations of the
                                  names as: 'NAME1', 'NAME2', 'NAME1 NAME2',
                                  NAME2 NAME1'. The non-expanded names are
                                  simply like: 'NAME1 NAME2'.
                                  Using expanded group names will find many
                                  more programs which share at least some of
                                  the expanded names than using the
                                  non-exploded names and so you will get more
                                  programs reported as sharing a similar
                                  function than you will if you specify that
                                  you wish to use non-exploded names
   -outfile            outfile    [stdout] Output file name
   -html               toggle     [N] If you are sending the output to a file,
                                  this will format it for displaying as a
                                  table in a WWW document.
   -groups             boolean    [N] If you use this option, then only the
                                  group names will output to the file

   Advanced (Unprompted) qualifiers:
   -[no]emboss         boolean    [Y] If you use this option then EMBOSS
                                  program documentation will be searched. If
                                  this option is set to be false, then only
                                  the EMBASSY programs will be searched (if
                                  the -embassy option is true). EMBASSY
                                  programs are not strictly part of EMBOSS,
                                  but use the same code libraries and share
                                  the same look and feel, but are generally
                                  developed by people who wish the programs to
                                  be outside of the GNU Public Licence scheme
   -[no]embassy        boolean    [Y] If you use this option then EMBASSY
                                  program documentation will be searched. If
                                  this option is set to be false, then only
                                  the EMBOSS programs will be searched (if the
                                  -emboss option is true). EMBASSY programs
                                  are not strictly part of EMBOSS, but use the
                                  same code libraries and share the same look
                                  and feel, but are generally developed by
                                  people who wish the programs to be outside
                                  of the GNU Public Licence scheme
   -showembassy        string     If you use this option then this EMBASSY
                                  package program documentation will be
                                  searched. EMBASSY programs are not strictly
                                  part of EMBOSS, but use the same code
                                  libraries and share the same look and feel,
                                  but are generally developed by people who
                                  wish the programs to be outside of the GNU
                                  Public Licence scheme (Any string is
                                  accepted)
   -colon              boolean    [N] The groups that EMBOSS applications
                                  belong to have up to two levels, for example
                                  the primary group 'ALIGNMENT' has several
                                  sub-groups, or second-level groups, e.g.:
                                  CONSENSUS, DIFFERENCES, DOT PLOTS, GLOBAL,
                                  LOCAL, MULTIPLE. To aid programs that parse
                                  the output of seealso that require the names
                                  of these subgroups, a colon ':' will be
                                  placed between the first and second level of
                                  the group name if this option is true.

   Associated qualifiers:

   "-outfile" associated qualifiers
   -odirectory         string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages

</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-search]<br>(Parameter 1)</td>
<td>Enter the name of an EMBOSS program</td>
<td>Any string of at least 1 characters</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-explode</td>
<td>The groups that EMBOSS applications belong to have two forms, exploded and not exploded. The exploded group names are more numerous and often vaguely phrased than the non-exploded ones. The exploded names are formed from definitions of the group names that start like NAME1:NAME2 and which are then expanded into many combinations of the names as: 'NAME1', 'NAME2', 'NAME1 NAME2', NAME2 NAME1'. The non-expanded names are simply like: 'NAME1 NAME2'.
Using expanded group names will find many more programs which share at least some of the expanded names than using the non-exploded names and so you will get more programs reported as sharing a similar function than you will if you specify that you wish to use non-exploded names</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-outfile</td>
<td>Output file name</td>
<td>Output file</td>
<td>stdout</td>
</tr>

<tr>
<td>-html</td>
<td>If you are sending the output to a file, this will format it for displaying as a table in a WWW document.</td>
<td>Toggle value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-groups</td>
<td>If you use this option, then only the group names will output to the file</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]emboss</td>
<td>If you use this option then EMBOSS program documentation will be searched. If this option is set to be false, then only the EMBASSY programs will be searched (if the -embassy option is true). EMBASSY programs are not strictly part of EMBOSS, but use the same code libraries and share the same look and feel, but are generally developed by people who wish the programs to be outside of the GNU Public Licence scheme</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]embassy</td>
<td>If you use this option then EMBASSY program documentation will be searched. If this option is set to be false, then only the EMBOSS programs will be searched (if the -emboss option is true). EMBASSY programs are not strictly part of EMBOSS, but use the same code libraries and share the same look and feel, but are generally developed by people who wish the programs to be outside of the GNU Public Licence scheme</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-showembassy</td>
<td>If you use this option then this EMBASSY package program documentation will be searched. EMBASSY programs are not strictly part of EMBOSS, but use the same code libraries and share the same look and feel, but are generally developed by people who wish the programs to be outside of the GNU Public Licence scheme</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-colon</td>
<td>The groups that EMBOSS applications belong to have up to two levels, for example the primary group 'ALIGNMENT' has several sub-groups, or second-level groups, e.g.: CONSENSUS, DIFFERENCES, DOT PLOTS, GLOBAL, LOCAL, MULTIPLE. To aid programs that parse the output of seealso that require the names of these subgroups, a colon ':' will be placed between the first and second level of the group name if this option is true.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

</table>


<H2>
    Input file format
</H2>


<H2>
    Output file format
</H2>



<p>

The result is header line 'SEE ALSO' then a simple list of program names
followed by the one-line description of the program. 

<p>

If the '-html' qualifier is used, then the result is an HTML-formatted
table, lacking the HTML document structures such as &lt;HTML&gt;,
&lt;BODY&gt;, &lt;/BODY&gt;, &lt;/HTML&gt; which the user is expected
to provide as the table will usually be included in the middle of a
larger HTML document.

<H2>
    Data files
</H2>


None.

<H2>
    Notes
</H2>


<p>The functional group of a program is set in its ACD file; this is an EMBOSS system file which specifies the application command-line interface such as required parameters, help on the parameters, one-line description of the program and so on.  Normally the names of the groups are fairly specific.  For example "NUCLEIC COMPOSITION" is a small group of applications for nucleotide sequence composition analysis.</p>

<p>The <tt>-explode</tt> qualifier will increase the number of groups that the program retrieves by splitting the group name at selected points to produce such groups as: "NUCLEIC", "NUCLEIC COMPOSITION" and "COMPOSITION". All programs with the exploded group names which also include "NUCLEIC" and "COMPOSITION" will now be reported. The result of this qualifier is thus a report of a larger number of programs with a more tenuous link to the specified program. </p>

<p>The groups that the program belongs to can be output by using the <tt>-groups</tt> qualifier. </p>

<H2>
    References
</H2>


None.

<H2>
    Warnings
</H2>

<p>The (optional) HTML-formatted table output lacks the HTML document structures such as <HTML>, <BODY>, </BODY>, </HTML>. It is expected the table will be included in the middle of a larger HTML document.</p>

<H2>
    Diagnostic Error Messages
</H2>


The error message:

<pre>

FATAL ERROR: No application specified.

</pre>

is output if a non-existent program name is specified.

<H2>
    Exit status
</H2>


The program exits with exist status '0' unless a non-existent program name
is specified, when the exit status is '1'.

<H2>
    Known bugs
</H2>


None.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="abiview.html">abiview</a></td>
<td>Display the trace in an ABI sequencer file</td>
</tr>

<tr>
<td><a href="cirdna.html">cirdna</a></td>
<td>Draws circular maps of DNA constructs</td>
</tr>

<tr>
<td><a href="infoalign.html">infoalign</a></td>
<td>Display basic information about a multiple sequence alignment</td>
</tr>

<tr>
<td><a href="infobase.html">infobase</a></td>
<td>Return information on a given nucleotide base</td>
</tr>

<tr>
<td><a href="inforesidue.html">inforesidue</a></td>
<td>Return information on a given amino acid residue</td>
</tr>

<tr>
<td><a href="infoseq.html">infoseq</a></td>
<td>Display basic information about sequences</td>
</tr>

<tr>
<td><a href="lindna.html">lindna</a></td>
<td>Draws linear maps of DNA constructs</td>
</tr>

<tr>
<td><a href="pepnet.html">pepnet</a></td>
<td>Draw a helical net for a protein sequence</td>
</tr>

<tr>
<td><a href="pepwheel.html">pepwheel</a></td>
<td>Draw a helical wheel diagram for a protein sequence</td>
</tr>

<tr>
<td><a href="prettyplot.html">prettyplot</a></td>
<td>Draw a sequence alignment with pretty formatting</td>
</tr>

<tr>
<td><a href="prettyseq.html">prettyseq</a></td>
<td>Write a nucleotide sequence and its translation to file</td>
</tr>

<tr>
<td><a href="remap.html">remap</a></td>
<td>Display restriction enzyme binding sites in a nucleotide sequence</td>
</tr>

<tr>
<td><a href="showalign.html">showalign</a></td>
<td>Display a multiple sequence alignment in pretty format</td>
</tr>

<tr>
<td><a href="showdb.html">showdb</a></td>
<td>Displays information on configured databases</td>
</tr>

<tr>
<td><a href="showfeat.html">showfeat</a></td>
<td>Display features of a sequence in pretty format</td>
</tr>

<tr>
<td><a href="showpep.html">showpep</a></td>
<td>Displays protein sequences with features in pretty format</td>
</tr>

<tr>
<td><a href="showseq.html">showseq</a></td>
<td>Displays sequences with features in pretty format</td>
</tr>

<tr>
<td><a href="sixpack.html">sixpack</a></td>
<td>Display a DNA sequence with 6-frame translation and ORFs</td>
</tr>

<tr>
<td><a href="textsearch.html">textsearch</a></td>
<td>Search the textual description of sequence(s)</td>
</tr>

<tr>
<td><a href="tfm.html">tfm</a></td>
<td>Displays full documentation for an application</td>
</tr>

<tr>
<td><a href="whichdb.html">whichdb</a></td>
<td>Search all sequence databases for an entry and retrieve it</td>
</tr>

<tr>
<td><a href="wossname.html">wossname</a></td>
<td>Finds programs by keywords in their short description</td>
</tr>

</table>


<H2>
    Author(s)
</H2>


Gary Williams (gwilliam&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK



<H2>
    History
</H2>

Written 16 June 2000 in response to a need to improve the automatic
construction of EMBOSS documentation web pages. 
<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.



<H2>
    Comments
</H2>
None



</BODY>
</HTML>

