<HTML>

<HEAD>
  <TITLE>
  EMBOSS: seqmatchall
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">






<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

seqmatchall
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>







<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Does an all-against-all comparison of a set of sequences


<H2>
    Description
</H2>


This takes a set of sequences and does an all-against-all pairwise
comparison of words (fragments of the sequences of a specified fixed
size) in the sequences, finding regions of identity between any two
sequences. 

<p>
The larger the specified word size, the faster the comparison will
proceed.  Regions whose stretches of identity are shorter than the word
size will be missed.  You should therefore choose a word size that is
small enough to find those regions of similarity you are interested in
within a reasonable time-frame. 

<H2>
    Usage
</H2>


<b>Here is a sample session with seqmatchall</b>
<p>
Here is an example using an increased word size to avoid accidental matches: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>seqmatchall </b>
Does an all-against-all comparison of a set of sequences
Input sequence set: <b>tembl:eclac*</b>
Word size [4]: <b>15</b>
Output file [eclac.seqmatchall]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>




<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-sequence]          seqset     Sequence set USA
   -wordsize           integer    Word size
  [-outfile]           outfile    Output file name

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1             integer    Start of each sequence to be used
   -send1               integer    End of each sequence to be used
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name

   "-outfile" associated qualifiers
   -odirectory2         string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence set USA</td>
<td>Readable set of sequences</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-wordsize</td>
<td>Word size</td>
<td>Integer 2 or more</td>
<td>4</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.seqmatchall</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>


<b>seqmatchall</b> reads a set of sequence USAs.  

<p>

The sequences must be either all protein or all nucleic acid.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:eclac*' is a sequence entry in the example nucleic acid database 'tembl'
<p>


<H2>
    Output file format
</H2>



<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: eclac.seqmatchall</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
1832  5646 7477 ECLAC 1 1832 ECLACA
1113  49 1161 ECLAC 1 1113 ECLACI
1500  4305 5804 ECLAC 1 1500 ECLACY
3078  1287 4364 ECLAC 1 3078 ECLACZ
159  1 159 ECLACA 1342 1500 ECLACY
60  1 60 ECLACY 3019 3078 ECLACZ
</pre>
</td></tr></table><p>



<p>

ECLAC (the complete E.coli lac operon) matches ECLACI ECLACZ ECLACY
and ECLACA (the individual genes), and there is a short overlap
between ECLACY and the flanking genes ECLACZ and ECLACA

<p>

The output is a list of regions of identity in pairs of sequences, each
consisting of one line with 7 columns of data separated by TABs or space
characters. 

<p>

The columns of data consist of:

<p>

<ul>
<li>The length of the region of identity.
<li>The start position in sequence 1.
<li>The end position in sequence 1.
<li>The name of sequence 1.
<li>The start position in sequence 2.
<li>The end position in sequence 2.
<li>The name of sequence 2.
</ul>

<H2>
    Data files
</H2>


None.

<H2>
    Notes
</H2>


The larger the word size, the faster the comparisons will proceed, but
regions of identitly smaller than the word size will not be reported.

<H2>
    References
</H2>


None.

<H2>
    Warnings
</H2>

None.


<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>


It exits with a status of 0.

<H2>
    Known bugs
</H2>


None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="matcher.html">matcher</a></td><td>Finds the best local alignments between two sequences</td></tr>
<tr><td><a href="supermatcher.html">supermatcher</a></td><td>Finds a match of a large sequence against one or more sequences</td></tr>
<tr><td><a href="water.html">water</a></td><td>Smith-Waterman local alignment</td></tr>
<tr><td><a href="wordmatch.html">wordmatch</a></td><td>Finds all exact matches of a given size between 2 sequences</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->



<P>
<a href="polydot.html">polydot</A> will give a graphical view of the same matches.

<H2>
    Author(s)
</H2>


Ian Longden (il&nbsp;&copy;&nbsp;sanger.ac.uk)
<br>
Sanger Institute, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.                      



<H2>
    History
</H2>


1999 - written.

<!--
<p>
[an error occurred while processing this directive]
-->

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>



</BODY>
</HTML>

