<HTML>
<HEAD>
  <TITLE>
  EMBOSS: seqretsplit
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">


<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

seqretsplit
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<H2>
    Function
</H2>
Reads and writes (returns) sequences in individual files

<H2>
    Description
</H2>

<b>seqretsplit</b> is exactly the same as the program <b>seqret</b>
except that when it writes out more than one sequence, it writes each
sequence to an individual file. 
<P>

Its main use is therefore to split a file containing multiple sequences
into many files, each containing one sequence. 
<p>

The names of the files it creates are derived from the ID name of the
sequence, followed by an extension denoting the format of the sequence. 
You have no control over the names of the files it writes out. 
<P>

For example, if the files <i>embl:hsfa11*</i> are read in and the
output is specified as <i>wibble.seq</i>, then the following files are
expected to be created:
<P>

<pre>
hsfa110.fasta
hsfa111.fasta
hsfa112.fasta
hsfa113.fasta
hsfa114.fasta
</pre>
<p>
(No file <i>wibble.seq</i> is created.)
<p>

Why would you want to split a multiple sequence file into many
individual files? 
<p>

EMBOSS programs can read in many sequences from one file where this is
sensible.  Sometimes EMBOSS programs can only read in one sequence at a
time because that is the sensible way to do things, but your sequence is
one sequence of many in a file.  You can specify that sequence using the
USA <i>filename:sequenceID</i>, but you may still feel more comfortable
splitting your sequences up into many files first.
<p>

Many non-EMBOSS programs will also have restrictions on whether they can
read in multiple sequence files or not.
<p>


<H2>
    Usage
</H2>
<b>Here is a sample session with seqretsplit</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>seqretsplit tembl:hsfa* </b>
Reads and writes (returns) sequences in individual files
Output sequence [hsfau.fasta]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>


<p>

The specification of the output file is not used in this case.
<p>

At some point this ought to change and you will not be prompted for the
output file at all. 

<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-sequence]          seqall     Sequence database USA
  [-outseq]            seqoutall  Output sequence(s) USA

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers:
   -feature            boolean    Use feature information
   -firstonly          boolean    Read one sequence and stop

   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1             integer    Start of each sequence to be used
   -send1               integer    End of each sequence to be used
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name

   "-outseq" associated qualifiers
   -osformat2           string     Output seq format
   -osextension2        string     File name extension
   -osname2             string     Base file name
   -osdirectory2        string     Output directory
   -osdbname2           string     Database name to add
   -ossingle2           boolean    Separate file for each entry
   -oufo2               string     UFO features
   -offormat2           string     Features format
   -ofname2             string     Features file name
   -ofdirectory2        string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outseq]<br>(Parameter 2)</td>
<td>Output sequence(s) USA</td>
<td>Writeable sequence(s)</td>
<td><i>&lt;sequence&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-feature</td>
<td>Use feature information</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-firstonly</td>
<td>Read one sequence and stop</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

</table>


<H2>
    Input file format
</H2>

<b>seqretsplit</b> reads a normal sequence USA.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:hsfa*' is a sequence entry in the example nucleic acid database 'tembl'
<p>



<H2>
    Output file format
</H2>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: hsfau.fasta</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt;HSFAU X65923.1 H.sapiens fau mRNA
ttcctctttctcgactccatcttcgcggtagctgggaccgccgttcagtcgccaatatgc
agctctttgtccgcgcccaggagctacacaccttcgaggtgaccggccaggaaacggtcg
cccagatcaaggctcatgtagcctcactggagggcattgccccggaagatcaagtcgtgc
tcctggcaggcgcgcccctggaggatgaggccactctgggccagtgcggggtggaggccc
tgactaccctggaagtagcaggccgcatgcttggaggtaaagttcatggttccctggccc
gtgctggaaaagtgagaggtcagactcctaaggtggccaaacaggagaagaagaagaaga
agacaggtcgggctaagcggcggatgcagtacaaccggcgctttgtcaacgttgtgccca
cctttggcaagaagaagggccccaatgccaactcttaagtcttttgtaattctggctttc
tctaataaaaaagccacttagttcagtcaaaaaaaaaa
</pre>
</td></tr></table><p>
<p><h3>File: hsfau1.fasta</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt;HSFAU1 X65921.1 H.sapiens fau 1 gene
ctaccattttccctctcgattctatatgtacactcgggacaagttctcctgatcgaaaac
ggcaaaactaaggccccaagtaggaatgccttagttttcggggttaacaatgattaacac
tgagcctcacacccacgcgatgccctcagctcctcgctcagcgctctcaccaacagccgt
agcccgcagccccgctggacaccggttctccatccccgcagcgtagcccggaacatggta
gctgccatctttacctgctacgccagccttctgtgcgcgcaactgtctggtcccgccccg
tcctgcgcgagctgctgcccaggcaggttcgccggtgcgagcgtaaaggggcggagctag
gactgccttgggcggtacaaatagcagggaaccgcgcggtcgctcagcagtgacgtgaca
cgcagcccacggtctgtactgacgcgccctcgcttcttcctctttctcgactccatcttc
gcggtagctgggaccgccgttcaggtaagaatggggccttggctggatccgaagggcttg
tagcaggttggctgcggggtcagaaggcgcggggggaaccgaagaacggggcctgctccg
tggccctgctccagtccctatccgaactccttgggaggcactggccttccgcacgtgagc
cgccgcgaccaccatcccgtcgcgatcgtttctggaccgctttccactcccaaatctcct
ttatcccagagcatttcttggcttctcttacaagccgtcttttctttactcagtcgccaa
tatgcagctctttgtccgcgcccaggagctacacaccttcgaggtgaccggccaggaaac
ggtcgcccagatcaaggtaaggctgcttggtgcgccctgggttccattttcttgtgctct
tcactctcgcggcccgagggaacgcttacgagccttatctttccctgtaggctcatgtag
cctcactggagggcattgccccggaagatcaagtcgtgctcctggcaggcgcgcccctgg
aggatgaggccactctgggccagtgcggggtggaggccctgactaccctggaagtagcag
gccgcatgcttggaggtgagtgagagaggaatgttctttgaagtaccggtaagcgtctag
tgagtgtggggtgcatagtcctgacagctgagtgtcacacctatggtaatagagtacttc
tcactgtcttcagttcagagtgattcttcctgtttacatccctcatgttgaacacagacg
tccatgggagactgagccagagtgtagttgtatttcagtcacatcacgagatcctagtct
ggttatcagcttccacactaaaaattaggtcagaccaggccccaaagtgctctataaatt
agaagctggaagatcctgaaatgaaacttaagatttcaaggtcaaatatctgcaactttg
ttctcattacctattgggcgcagcttctctttaaaggcttgaattgagaaaagaggggtt
ctgctgggtggcaccttcttgctcttacctgctggtgccttcctttcccactacaggtaa
agtccatggttccctggcccgtgctggaaaagtgagaggtcagactcctaaggtgagtga
gagtattagtggtcatggtgttaggactttttttcctttcacagctaaaccaagtccctg
ggctcttactcggtttgccttctccctccctggagatgagcctgagggaagggatgctag
gtgtggaagacaggaaccagggcctgattaaccttcccttctccaggtggccaaacagga
gaagaagaagaagaagacaggtcgggctaagcggcggatgcagtacaaccggcgctttgt
caacgttgtgcccacctttggcaagaagaagggccccaatgccaactcttaagtcttttg
taattctggctttctctaataaaaaagccacttagttcagtcatcgcattgtttcatctt
tacttgcaaggcctcagggagaggtgtgcttctcgg
</pre>
</td></tr></table><p>


<p>

One file for each input sequence is written out.

<p>

The names of the files it creates are derived from the ID name of the
sequence, followed by an extension denoting the format of the sequence. 
You have no control over the names of the files it writes out. 

<p>

For example, if the files <i>embl:hsfa11*</i> are read in and the
output is specified as <i>wibble.seq</i>, then the following files are
expected to be created:

<p>

<pre>
hsfa110.fasta
hsfa111.fasta
hsfa112.fasta
hsfa113.fasta
hsfa114.fasta
</pre>

<p>

(No file <i>wibble.seq</i> is created.)

<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with status 0.

<H2>
    Known bugs
</H2>

It shouldn't really prompt for the output filename.
<p>
This isa side effect of the waysequence output works in
EMBOSS. Writing multiple sequences to separate files (the -ossingle
qualifier) does this, and seqretsplit has set it automatically on.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="biosed.html">biosed</a></td><td>Replace or delete sequence sections</td></tr>
<tr><td><a href="cutseq.html">cutseq</a></td><td>Removes a specified section from a sequence</td></tr>
<tr><td><a href="degapseq.html">degapseq</a></td><td>Removes gap characters from sequences</td></tr>
<tr><td><a href="descseq.html">descseq</a></td><td>Alter the name or description of a sequence</td></tr>
<tr><td><a href="entret.html">entret</a></td><td>Reads and writes (returns) flatfile entries</td></tr>
<tr><td><a href="extractfeat.html">extractfeat</a></td><td>Extract features from a sequence</td></tr>
<tr><td><a href="extractseq.html">extractseq</a></td><td>Extract regions from a sequence</td></tr>
<tr><td><a href="listor.html">listor</a></td><td>Writes a list file of the logical OR of two sets of sequences</td></tr>
<tr><td><a href="maskfeat.html">maskfeat</a></td><td>Mask off features of a sequence</td></tr>
<tr><td><a href="maskseq.html">maskseq</a></td><td>Mask off regions of a sequence</td></tr>
<tr><td><a href="newseq.html">newseq</a></td><td>Type in a short new sequence</td></tr>
<tr><td><a href="noreturn.html">noreturn</a></td><td>Removes carriage return from ASCII files</td></tr>
<tr><td><a href="notseq.html">notseq</a></td><td>Excludes a set of sequences and writes out the remaining ones</td></tr>
<tr><td><a href="nthseq.html">nthseq</a></td><td>Writes one sequence from a multiple set of sequences</td></tr>
<tr><td><a href="pasteseq.html">pasteseq</a></td><td>Insert one sequence into another</td></tr>
<tr><td><a href="revseq.html">revseq</a></td><td>Reverse and complement a sequence</td></tr>
<tr><td><a href="seqret.html">seqret</a></td><td>Reads and writes (returns) sequences</td></tr>
<tr><td><a href="skipseq.html">skipseq</a></td><td>Reads and writes (returns) sequences, skipping the first few</td></tr>
<tr><td><a href="splitter.html">splitter</a></td><td>Split a sequence into (overlapping) smaller sequences</td></tr>
<tr><td><a href="trimest.html">trimest</a></td><td>Trim poly-A tails off EST sequences</td></tr>
<tr><td><a href="trimseq.html">trimseq</a></td><td>Trim ambiguous bits off the ends of sequences</td></tr>
<tr><td><a href="union.html">union</a></td><td>Reads sequence fragments and builds one sequence</td></tr>
<tr><td><a href="vectorstrip.html">vectorstrip</a></td><td>Strips out DNA between a pair of vector sequences</td></tr>
<tr><td><a href="yank.html">yank</a></td><td>Reads a sequence range, appends the full USA to a list file</td></tr>
</table>



<H2>
    Author(s)
</H2>

Peter Rice (pmr&nbsp;&copy;&nbsp;ebi.ac.uk)
<br>
Informatics Division, European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK


<H2>
    History
</H2>

Written (Jan 2000) - Peter Rice


<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.


<H2>
    Comments
</H2>
The specification of the output file is not used when the output file
names are generated automatically..
<p>

At some point this ought to change and you will not be prompted for the
output file at all. 




</BODY>
</HTML>

