<HTML>

<HEAD>
  <TITLE>
  EMBOSS: seqret
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: seqret
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program seqret
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Reads and writes (returns) sequences


<H2>
    Description
</H2>


The simplicity of the above description of this program greatly
understates the rich functionality of this program. 

<p>
Because EMBOSS programs can take a wide range of qualifiers that
slightly change the behaviour of the program when reading or writing a
sequence, this program can do many more things than simply "read and
write a sequence". 

<p>
seqret can read a sequence or many sequences from databases, files,
files of sequence names, the command-line or the output of other
programs and then can write them to files, the screen or pass them to
other programs.  Because it can read in a sequence from a database and
write it to a file, seqret is a program for extracting sequences from
databases.  Because it can write the sequence to the screen, seqret is a
program for displaying sequences. 

<p>
seqret can read sequences in any of a wide range of standard sequence
formats.  You can specify the input and output formats being used.  If
you don't specify the input format, seqret will try a set of possible
formats until it reads it in successfully.  Because you can specify the
output sequence format, seqret is a program to reformat a sequence. 

<p>
seqret can read in the reverse complement of a nucleic acid sequence. 
It therefore is a program for producing the reverse complement of a
sequence. 

<p>
seqret can read in a sequence whose begin and end positions you have
specified and write out that fragment.  It is therefore a utility for
doing simple extraction of a region of a sequence. 

<p>
seqret can change the case of the sequence being read in to upper or to
lower case.  It is therefore a simple sequence beautification utility. 

<p>
seqret can do any combination of the above functions.

<p>
The sequence input and output specification of this (and many other
EMBOSS programs) is described as being a Uniform Sequence Address. 

<p>
The Uniform Sequence Address, or USA, is a somewhat tongue-in-cheek
reference to a URL-style sequence naming used by all EMBOSS
applications. 

<p>
The USA is a very flexible way of specifying one or more sequences from
a variety of sources and includes sequence files, database queries and
external applications. 

<p>
The basic USA syntax is one of: 

<ul>
<li>"file" 
<li>"file:entry" 
<li>"format::file" 
<li>"format::file:entry" 
<li>"database:entry" 
<li>"database"
<li>"@file" 
</ul>

<p>
Note that ':' separates the name of a file containing many possible
entries from the specific name of a sequence entry in that file.  It
also separates the name of a database from an entry in that database

<p>
Note also that '::' separates the specified format of a file from the
name of the file.  Normally the format can be omitted, in which case the
program will attempt to identify the correct format when reading the
sequence in and will default to using FASTA format when writing the
sequence out. 

<p>
Valid names of the databases set up in your local implementation of
EMBOSS can be seen by using the program <a href="showdb.html">'showdb'.</a>

<p>
Database queries, and individual entries in files that have more than
one sequence entry, use wildcards of "?" for any character and "*" for
any string of characters.  There are some problems with the Unix shell
catching these characters so they do need to be hidden in quotes or
preceded by a backslash on the Unix command line, (for example
"embl:hs\*")

<p>
The output USA name 'stdout' is special.  It makes the output go to the
device 'standard output'.  This is the screen, by default. 

<p>

<h3>Example USAs</h3>

The following are valid USAs for sequences:

<p>
<TABLE border=3>
<TR>
<TH>USA</TH>
<TH>Description</TH>
</TR>

<TR>
<TD>xxx.seq</TD>
<TD>A sequence file "xxx.seq" in any format</TD>
</TR>

<TR>
<TD>fasta::xxx.seq</TD>
<TD>A sequence file "xxx.seq" in fasta format</TD>
</TR>

<TR>
<TD>gcg::egmsmg.gcg</TD>
<TD>A sequence file "egmsmg.gcg" in GCG 9 format</TD>
</TR>

<TR>
<TD>egmsmg.gcg -sformat=gcg</TD>
<TD>A sequence file "egmsmg.gcg" in GCG 9 format</TD>
</TR>

<TR>
<TD>embl::paamir.em</TD>
<TD>A sequence file "paamir.em" in EMBL format</TD>
</TR>

<TR>
<TD>embl:paamir</TD>
<TD>EMBL entry PAAMIR, using whatever access method is defined locally
for the EMBL database</TD>
</TR>

<TR>
<TD>embl:X13776</TD>
<TD>EMBL entry X13776, using whatever access method is defined locally
for the EMBL database and searching by accession number and entry name
(X13776 is the accession number in this case)</TD>
</TR>

<TR>
<TD>embl-acc:X13776</TD>
<TD>EMBL entry X13776, using whatever access method is defined locally
for the EMBL database and searching by accession number only
</TD>
</TR>

<TR>
<TD>embl-id:paamir</TD>
<TD>EMBL entry PAAMIR, using whatever access method is defined locally
for the EMBL database, and searching by ID only</TD>
</TR>

<TR>
<TD>embl:paami*</TD>
<TD>EMBL entries PAAMIB, PAAMIE and so on, usually in alphabetical order,
using whatever access method is defined locally for the EMBL database</TD>
</TR>

<TR>
<TD>embl or EMBL:*</TD>
<TD>All sequences in the EMBL database</TD>
</TR>

<TR>
<TD>@mylist</TD>
<TD>Reads file <i>mylist</i> and uses each line as a separate USA.  This
is standard VMS list file syntax, also used in SRS 4.0 but missing in
SRS 5.0.  The list file is a list of USAs (one per line).  List files
can contain references to other lists files or any other standard
USA.</TD>
</TR>

<TR>
<TD>list::mylist</TD>
<TD>Same as "@mylist" above</TD>
</TR>

<TR>
<TD>'getz -e [embl-id:paamir] |'</TD>
<TD>The pipe character "|" causes EMBOSS to fire up getz (SRS 5.1)
to extract entry PAAMIR from EMBL in EMBL format. Any application or
script which writes one or more sequences to stdout can be used in
this way.
</TR>

<TR>
<TD>asis::atacgcagttatctgaccat</TD>
<TD>So far the shortest USA we
could invent. In 'asis' format the name is the sequence so no file
needs to be opened. This is a special case.  It was intended as a
joke, but could be quite useful for generating command lines.</TD>
</TR>

</TABLE>

<h3>Input sequence formats</h3>

To date, the following sequence formats are accepted as
input.
<P>

By default, (i.e.  if no format is explicitly specified) EMBOSS tries
each format in turn until one succeeds.

<!--
INPUT FORMAT TABLE
LAST UPDATED: 23 OCT 2000 - GWW
-->

<p>
<table border=3>
<tr><th>Input Format</th><th>Comments</th></tr>



<tr><td>gcg</td><td>
GCG 9.x and 10.x format with the format and sequence type identified on the
first line of the file
</td></tr>

<tr><td>gcg8</td><td>
GCG 8.x format where anything up to the first line containing ".." is
considered as heading, and the remainder is sequence data.  This
format is complicated by the header appearing to be in other formats
such as EMBL, and by the possibility of reading a large amount of data
in the wrong format before discovering that there is no ".." line
because it is not GCG format after all.
</td></tr>

<tr><td>embl<br>em</td><td>
EMBL entry format, or at least a minimal subset of the fields. The <a
href="http://www.mrc-lmb.cam.ac.uk/pubseq/"> Staden package</a> and
others use EMBL or similar formats for sequence data.
</td></tr>

<tr><td>swiss<br>sw</td><td>   
SWISSPROT entry format, or at least a minimal subset of the fields.
</td></tr>

<tr><td>fasta<br>pearson</td><td>
FASTA format with an optional accession number after the sequence
identifier, eg:
<br>
<b>&gt;name description</b>
<br>
or   
<br>
<b>&gt;name accession description</b>
<br>
and with an optional database name in GCG style
fasta format included as part of the sequence identifier, eg:
<br>
<b>&gt;database:name accession description</b>
<br>
</td></tr>
   
<tr><td>ncbi</td><td>
FASTA format with optional accession number and database name in NCBI
style included as part of the sequence identifier.
eg
<br>
<b>&gt;database|accession|id description</b>
<br>
(and other variants on this theme!)
</td></tr>

<tr><td>genbank<br>gb</td><td>
GENBANK entry format, or at least a minimal subset of the fields.
</td></tr>

<tr><td>nbrf<br>pir</td><td>
NBRF (PIR) format, as used in the PIR database sequence files.
</td></tr>

<tr><td>codata</td><td>
CODATA format.
</td></tr>

<tr><td>strider</td><td>
DNA strider format
</td></tr>
   
<tr><td>clustal<br>aln</td><td>
ClustalW ALN (multiple alignment) format.
</td></tr>

<tr><td>phylip</td><td>
PHYLIP non-interleaved multiple alignment format.
</td></tr>

<tr><td>acedb</td><td>
ACeDB format
</td></tr>

<tr><td>msf</td><td>
<a href="http://www.gcg.com/">Wisconsin Package</a>
GCG's MSF multiple sequence format.
</td></tr>

<tr><td>hennig86</td><td>
Hennig86 format
</td></tr>

<tr><td>jackknifer</td><td>
Jackknifer format
</td></tr>

<tr><td>jackknifernon</td><td> 
Jackknifernon format
</td></tr>

<tr><td>nexus<br>paup</td><td>   
Nexus/PAUP format
</td></tr>

<tr><td>nexusnon<br>paupnon</td><td>
Nexusnon/PAUPnon format
</td></tr>

<tr><td>treecon</td><td>
Treecon format
</td></tr>

<tr><td>mega</td><td>
Mega format
</td></tr>

<tr><td>meganon</td><td>
Meganon format
</td></tr>

<tr><td>ig</td><td>
IntelliGenetics format.
</td></tr>

<tr><td>staden<br>experiment</td><td>
The experiment file format used by the "gap" program in the Staden
package, where the sequence identifier is optional and the remainer is
plain text.  Some alternative nucleotide ambiguity codes are used and
must be converted.
</td></tr>


<tr><td>unknown<br>text<br>plain</td><td>
Plain text.
This is the format with no format. The whole of the file is
read in as a sequence. 
No attempt is made to parse the file contents in any way.
Anything is acceptable in this format.
</td></tr>


<tr><td>raw</td><td>
Like unknown/text/plain format except that
it accepts only alphanumeric and whitespace
characters and rejects anything else.
</td></tr>


<tr><td>asis</td><td>
This is not so much a sequence format as a quick way of entering a
sequence on the command line, but it is included here for completeness.
Where a filename would normally be given, in <b>asis</b> format there is
the sequence itself.
An example would be:
<br>
asis::atacgcagttatctgaccat
<br>
In 'asis' format the name is the sequence so no file needs to be opened.
This is a special case.  It was intended as a joke, but could be quite
useful for generating command lines.
</td></tr>


</table>





<h3>Output sequence formats</h3>

To date, the following sequence formats are available as output.
<p>

Some sequence formats can hold multiple sequences in one file, these are
marked as <b>multiple</b> in the following table.  The details of how many
sequences are held in one file differs between formats, but they either
allow many sequences to be concatenated one after the other, or they
hold the sequences together in some sort of aligned set of sequences.
<P>

Other formats, such as GCG, plain and staden formats can only hold one
sequence per file, these are marked as <b>single</b>.  An attempt to
concatenate several sequences in one file leaves the results as a mess
that makes it impossible to decide where the sequences start and end or
what is annotation and what is sequence. 
<p>
   
These <b>single</b> formats therefore cause problems when there are
multiple sequences to write out because a single file containing
multiple sequences in that format is invalid.  When these formats are
specified for output, an EMBOSS program will allow you to write many
sequences to one file, but EMBOSS programs will not be able to reliably
read in the resulting mess. 
<p>

<B>N.B</B> This behaviour changed in EMBOSS version 1.7.0.  (31 Oct
2000) Previously, EMBOSS programs that were asked to write multiple
sequences in a <b>single</b> format whould ignore the requested output
file name and would write each sequence into a separate file whose name
was constructed from the sequence name and the name of the format.  This
resulted in ouput to files whose names could not be reliably controlled. 
A decision was taken that EMBOSS users were intelligent people who could
live with the consequences of their actions and who could learn not to
write out multiple sequences to a file in formats that could not cope
with multiple sequences. 
<P>

It you really wish to write multiple sequences out in formats that can
not cope with multiple sequences, you are advised to add the global
qualifier <b>-ossingle</b> on the command line.  This will force the
EMBOSS program to ignore the given output file name and will generate its
own file names.  One sequence will be written to each such file.  These file
names are made from the sequence ID name, with the name of the format as
the extension (e.g.  <b>hsfau.gcg</b>). 
<P>

This is not ideal.  Preferably, you should stay away from formats that
can't cope with multiple sequences in a file. 
<P>


<!--
OUTPUT FORMAT TABLE
LAST UPDATED: 30 OCT 2000 - GWW
-->

<p>
<table border=3>
<tr><th>Output Format</th><th>Single/<br>Multiple</th><th>Comments</th></tr>

<tr><td>gcg</td><td>single</td><td>
<a href="http://www.gcg.com/">Wisconsin Package</a> GCG 9.x and 10.x format
with the sequence type on the first line of the file.
</td></tr>

<tr><td>gcg8</td><td>single</td><td>
GCG 8.x format where anything up to the first line containing ".." is
considered as heading, and the remainder is sequence data.
</td></tr>

<tr><td>embl<br>em</td><td>multiple</td><td>
EMBL entry format with available fields filled in and others with no
infomation omitted. The EMBOSS command line allows missing data such
as accession numbers to be provided if they are not obtainable from
the input sequence.
</td></tr>

<tr><td>swiss<br>sw</td><td>multiple</td><td>
SwisProt entry format with available fields filled in and others with no
infomation omitted.  The EMBOSS command line allows missing data such as
accession numbers to be provided if they are not obtainable from the
input sequence. 
</td></tr>

<tr><td>fasta</td><td>multiple</td><td>
Standard Pearson FASTA format, but with the accession number included
after the identifier if available.
</td></tr>

<tr><td>pearson</td><td>multiple</td><td>
Simple Pearson FASTA format, an alias for "fasta" format.
</td></tr>

<tr><td>ncbi</td><td>multiple</td><td>
NCBI style FASTA format with the database name, entry name and
accession number separated by pipe ("|") characters.
</td></tr>

<tr><td>nbrf<br>pir</td><td>multiple</td><td>
NBRF (PIR) format, as used in the PIR database sequence files.
</td></tr>


<tr><td>genbank<br>gb</td><td>multiple</td><td>
GENBANK entry format with available fields filled in and others with no
infomation omitted. The EMBOSS command line allows missing data such
as accession numbers to be provided if they are not obtainable from
the input sequence.
</td></tr>


<tr><td>ig</td><td>multiple</td><td>
Intelligenetics format, as used by the
Intelligenetics package
</td></tr>

<tr><td>codata</td><td>multiple</td><td>
CODATA format.
</td></tr>

<tr><td>strider</td><td>multiple</td><td>
DNA strider format
</td></tr>

<tr><td>acedb</td><td>multiple</td><td>
ACeDB format
</td></tr>

<tr><td>staden<br>experiment</td><td>single</td><td>
The experiment file format used by the "gap" program in the 
<a href="http://www.mrc-lmb.cam.ac.uk/pubseq/">
Staden package</a>.
Some alternative nucleotide ambiguity codes are used and are converted.
</td></tr>

<tr><td>text<br>plain<br>raw</td><td>single</td><td>
Plain sequence, no annotation or heading.
</td></tr>

<tr><td>fitch</td><td>multiple</td><td>
Fitch format
</td></tr>

<tr><td>msf</td><td>multiple</td><td>
<a href="http://www.gcg.com/">Wisconsin Package</a>
GCG's MSF multiple sequence format.
</td></tr>

<tr><td>clustal<br>aln</td><td>multiple</td><td>
Clustal multiple sequence format.
</td></tr>

<tr><td>phylip</td><td>multiple</td><td>
PHYLIP non-interleaved format.
</td></tr>

<tr><td>phylip3</td><td>multiple</td><td>
PHYLIP interleaved format.
</td></tr>

<tr><td>asn1</td><td>multiple</td><td>
A subset of ASN.1 containing entry name, accession number, description
and sequence, similar to the current ASN.1 output of
<a href="http://iubio.bio.indiana.edu/soft/molbio/readseq/">
readseq</a> 
</td></tr>

<tr><td>hennig86</td><td>multiple</td><td>
Hennig86 format
</td></tr>

<tr><td>mega</td><td>multiple</td><td>
Mega format
</td></tr>

<tr><td>meganon</td><td>multiple</td><td>
Meganon format
</td></tr>

<tr><td>nexus<br>paup</td><td>multiple</td><td>
Nexus/PAUP format
</td></tr>

<tr><td>nexusnon<br>paupnon</td><td>multiple</td><td>
Nexusnon/PAUPnon format
</td></tr>

<tr><td>jackknifer</td><td>multiple</td><td>
Jackknifer format
</td></tr>

<tr><td>jackknifernon</td><td>multiple</td><td>
Jackknifernon format
</td></tr>

<tr><td>treecon</td><td>multiple</td><td>
Treecon format
</td></tr>


<tr><td>debug</td><td>multiple</td><td>
EMBOSS sequence object report for debugging showing all available fields.
Not all fields will contain data - this depends very much on the input
format used.
</td></tr>


</table>


<h3>Future directions</h3>

More formats, both for input and for output, can be easily added, so
suggestions are always welcome. 



<h3>Associated qualifiers</h3>

<p>
As noted previously there are many 'associated' qualifiers that alter
the behaviour of seqret when it reads in or writes out a sequence.  As
these are used in all EMBOSS programs that read in or write out
sequences, they are not reported by the '-help' qualifier.  They are
however reported by the pair of qualifiers: '-help -verbose':


<p>
Some of the more useful associated qualifiers are:

<table border=3>
<tr><th>Qualifier</th>		<th>Description</th></tr>

<tr><td>-sbegin</td><td>The first position to be used in the sequence</td></tr>
<tr><td>-send</td><td>The last position to be used in the sequence</td></tr>
<tr><td>-sreverse</td><td>Use the reverse complement of a nucleic acid sequence</td></tr>
<tr><td>-sask</td><td>Ask the user for begin/end/reverse information</td></tr>
<tr><td>-slower</td><td>Convert the sequence to lower case</td></tr>
<tr><td>-supper</td><td>Convert the sequence to upper case</td></tr>
<tr><td>-sformat</td><td>Specify the input sequence format</td></tr>
<tr><td>-osformat</td><td>Specify the output sequence format</td></tr>
<tr><td>-ossingle</td><td>Write each entry into a separate file</td></tr>
<tr><td>-auto</td><td>Turn off prompts and don't report the one-line description</td></tr>
<tr><td>-stdout</td><td>Write the results to 'standard output' (the screen)</td></tr>
<tr><td>-filter</td><td>Read input from another program, write to the screen</td></tr>
<tr><td>-options</td><td>Prompt for optional qualifiers</td></tr>
<tr><td>-help</td><td>Display a table of the command-line options</td></tr>
</table>

<p>
The set of associated qualifiers for sequences behave in different ways
depending on where they appear. 

<p>
If these qualifiers immediately follow a parameter they apply only to
that parameter and not to all cases.  If they occur before any
parameters, they apply to all following sequence parameters. 

<p>
If there are no two parameters of equal type, the order of parameters
and their qualifiers is irrelevant. 

<p>
Where a qualifier is defined more than once, for example "-sformat" for
2 input sequences to be aligned, the qualifier name can have a number to
indicate which sequence is meant.  "-sbegin2=25" will apply only to the
second sequence, no matter where it appears on the command line. 

<p>
The -sbegin and -send qualifiers take an integer number specifying the
position to begin or end reading a sequence.  If the number is positive,
the number is the position counting from the first base or residue of
the sequence.  If the number is negative the position is counted from
the end of the sequence, so position -1 is the last base or residue of
the sequence.  (If -sbegin 0 is used, it is assumed to be the same as
-sbegin 1 and -send 0 is the same as -send -1.)

<p>
The filter qualifier makes the program behave like a filter, reading its
(first) input 'file' from the standard input, and writing its (first)
output 'file' to the standard output.  The -filter qualifier will also
invoke the -auto qualifier, so the user is never prompted for any
missing values. 

<p>
Example:

<pre>

% cat sequence.seq | seqret -filter | lpr

</pre>

<p>
The example shows the application seqret being run with the -filter
qualifier.  The input file is 'piped' into the program using the unix
command cat and the output is 'piped' directly to the unix program lpr,
which will print it on the printer. 

<p>
When the -options qualifier is used and not all the parameters are given
on the command line, it will query the user for those parameters.  It
will not only query the user for the required parameters as it would do
without the -options qualifier, but it will also query the user for the
optional parameters.

<p>
When the -stdout qualifier is used, the user will still be prompted for
all the info that is required, but will write to standard output by
default.  The user will also still be prompted for an output filename,
in case the user wants to save the output to a file. 

<H2>
    Usage
</H2>


Here is a sample session with seqret.   It is used to extract an entry
from a database and write it to a file:

<pre>

% seqret
Input sequence: embl:paamir
Output sequence [paamir.fasta]: 

</pre>


<p>
Here seqret is used to display the contents of the sequence on the
screen:

<pre>

% seqret 
Reads and writes (returns) sequences
Input sequence: embl:paamir
Output sequence [paamir.fasta]: stdout

</pre>

<p>
Here it is used in three different ways to write the result to a file in
GCG format.  Once by using the qualifier '-osformat', once by using the
format in the output USA on the command line and once by specifying the
format in the USA at the prompt. 

<pre>

% seqret -osf gcg
Reads and writes (returns) sequences
Input sequence: embl:paamir
Output sequence [paamir.gcg]:

</pre>

<pre>

% seqret -outseq gcg::paamir.gcg
Reads and writes (returns) sequences
Input sequence: embl:paamir

</pre>

<pre>

% seqret
Reads and writes (returns) sequences
Input sequence: embl:paamir
Output sequence [paamir.fasta]: gcg::paamir.gcg

</pre>

<p>
Here seqret is used to produce the reverse-complement of a sequence:

<pre>

% seqret -srev
Reads and writes (returns) sequences
Input sequence: embl:paamir
Output sequence [paamir.fasta]: 

</pre>

<p>
Here seqret is used to extract the bases between the positions starting
at 5 and ending at 25:

<pre>

% seqret -sbegin 5 -send 25
Reads and writes (returns) sequences
Input sequence: embl:paamir
Output sequence [paamir.fasta]:

</pre>

<p>
Here seqret is used to extract the bases between the positions starting
at 5 and ending at 5 bases before the end of the sequence:

<pre>

% seqret -sbegin 5 -send -5
Reads and writes (returns) sequences
Input sequence: embl:paamir
Output sequence [paamir.fasta]:

</pre>

<p>
Here seqret is used to read all entries in the database 'tembl' that
start with 'hs' and writes them to a file:

<pre>

% seqret
Reads and writes (returns) sequences
Input sequence(s): embl:hs*
Output sequence [hs989235.fasta]:

</pre>

<p>
Here seqret is used to read all entries in the database 'tembl' that
start with 'hs' and writes them to a file. In this example the
specification is all done in the command line and to stop Unix getting
confused by the '*' character, it has to have a backslash ('\') before
it:

<pre>

% seqret embl:hs\*  hs989235.fasta
Reads and writes (returns) sequences

</pre>

<p>
Here seqret is used to read only the entry 'hsfau' from the file
'hs989235.fasta' which contains many entries:

<pre>

% seqret
Reads and writes (returns) sequences
Input sequence(s): hs989235.fasta:hsfau
Output sequence [hsfau.fasta]: 

</pre>

<p>
Here seqret is used to read all entries in the file 'hs989235.fasta', but
only writes the first one of these entries out to a file:

<pre>

% seqret -firstonly
Reads and writes (returns) sequences
Input sequence(s): hs989235.fasta
Output sequence [hs989235.fasta]: first.fasta

</pre>


<p>
Here seqret is used to display on the screen the short sequence
"actgatcgtg" in uppercase in EMBL format:

<pre>

% seqret -supper -osf embl asis::actgatcgtg stdout
Reads and writes (returns) sequences
ID              standard; DNA; UNC; 10 BP.
SQ   Sequence 10 BP; 2 A; 2 C; 3 G; 3 T; 0 other;
     ACTGATCGTG                                10
//

</pre>




<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
  [-outseq]            seqoutall  Output sequence(s) USA

   Optional qualifiers: (none)
   Advanced qualifiers:
   -firstonly          bool       Read one sequence and stop

   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outseq]<br>(Parameter 2)</td>
<td>Output sequence(s) USA</td>
<td>Writeable sequence(s)</td>
<td><i>&lt;sequence&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-firstonly</td>
<td>Read one sequence and stop</td>
<td>Yes/No</td>
<td>No</td>
</tr>

</table>



<H2>
    Input file format
</H2>


The USA of one or more sequences.


<H2>
    Output file format
</H2>


An output USA.

<p>
The output from seqret is one or more sequences, and by default will be
written in FASTA format. 

<p>
If the '-firstonly' qualifier is used then only the first sequence of the
input USA specification will be written out. 

<p>
In some cases the output filename will be the same as the input
filename, but as seqret reads only the first sequence before opening
the output file it may try to overwrite the input.  Note that this is
not true of seqretset which reads all sequences into memory at
startup, but which can need a large amount of memory for many sequences..

<H2>
    Data files
</H2>


None.

<H2>
    Notes
</H2>


This description of what you can do when reading or writing files is not
specific to the program seqret.  All EMBOSS programs that read or write
sequences can do the same. 

<p>
seqret is often one of the first programs taught in EMBOSS training
courses.  This is because it is versatile, it is extremely powerful for
its size (17 lines of code) it illustrates many aspects of EMBOSS
programs and it was one of the first EMBOSS programs to be written, so
it has a special place in the hearts of EMBOSS developers. 

<p>
The name 'seqret' derives both from its function ("sequence return") and
from the fact that immense amounts of functionality can come from so few
lines of source code - most of the work is done by the EMBOSS libraries
which the program calls and whose complexity is hidden, or "secret". 


<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>


None.

<H2>
    Diagnostic Error Messages
</H2>

None.


<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.


<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="cutseq.html">cutseq</a></td><td>Removes a specified section from a sequence</td></tr>
<tr><td><a href="degapseq.html">degapseq</a></td><td>Removes gap characters from sequences</td></tr>
<tr><td><a href="descseq.html">descseq</a></td><td>Alter the name or description of a sequence</td></tr>
<tr><td><a href="entret.html">entret</a></td><td>Reads and writes (returns) flatfile entries</td></tr>
<tr><td><a href="extractseq.html">extractseq</a></td><td>Extract regions from a sequence</td></tr>
<tr><td><a href="infoseq.html">infoseq</a></td><td>Displays some simple information about sequences</td></tr>
<tr><td><a href="listor.html">listor</a></td><td>Writes a list file of the logical OR of two sets of sequences</td></tr>
<tr><td><a href="maskfeat.html">maskfeat</a></td><td>Mask off features of a sequence</td></tr>
<tr><td><a href="maskseq.html">maskseq</a></td><td>Mask off regions of a sequence</td></tr>
<tr><td><a href="newseq.html">newseq</a></td><td>Type in a short new sequence</td></tr>
<tr><td><a href="noreturn.html">noreturn</a></td><td>Removes carriage return from ASCII files</td></tr>
<tr><td><a href="notseq.html">notseq</a></td><td>Excludes a set of sequences and writes out the remaining ones</td></tr>
<tr><td><a href="nthseq.html">nthseq</a></td><td>Writes one sequence from a multiple set of sequences</td></tr>
<tr><td><a href="pasteseq.html">pasteseq</a></td><td>Insert one sequence into another</td></tr>
<tr><td><a href="revseq.html">revseq</a></td><td>Reverse and complement a sequence</td></tr>
<tr><td><a href="seqretall.html">seqretall</a></td><td>Reads and writes (returns) a set of sequences one at a time</td></tr>
<tr><td><a href="seqretset.html">seqretset</a></td><td>Reads and writes (returns) a set of sequences all at once</td></tr>
<tr><td><a href="seqretsplit.html">seqretsplit</a></td><td>Reads and writes (returns) sequences in individual files</td></tr>
<tr><td><a href="splitter.html">splitter</a></td><td>Split a sequence into (overlapping) smaller sequences</td></tr>
<tr><td><a href="swissparse.html">swissparse</a></td><td>Retrieves sequences from swissprot using keyword search</td></tr>
<tr><td><a href="trimest.html">trimest</a></td><td>Trim poly-A tails off EST sequences</td></tr>
<tr><td><a href="trimseq.html">trimseq</a></td><td>Trim ambiguous bits off the ends of sequences</td></tr>
<tr><td><a href="vectorstrip.html">vectorstrip</a></td><td>Strips out DNA between a pair of vector sequences</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->




<p>
Valid names of the databases set up in your local implementation of
EMBOSS can be seen by using the program 'showdb'. 


<H2>
    Author(s)
</H2>


This application was written by Peter Rice <a
href="mailto:pmr@sanger.ac.uk">(pmr@sanger.ac.uk)</a> Informatics
Division, The Sanger Centre, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.

<H2>
    History
</H2>

1999 - Written by Peter Rice

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<h3>Fasta output format</h3>

<h4>Question</h4>
When i tried to convert the EMBL format file into fasta format using the
program "seqret", I found that the Access.no appears twice... 

<p>
<pre>
&gt;AF102796 AF102796 Homo sapiens alphaE-catenin (CTNNA1) gene, exon 11.
</pre>
<p>

<h4>Answer</h4>

"It is not a bug ... it is a feature"
<p>

There are many "FASTA formats". EMBOSS uses the format that ACEDB and the
Sanger Centre genome projects use. The first field after the ID is the
accession number, so that accession numbers can be kept when sequences are
converted to FASTA format, without using the NCBI format (with '|'
characters in the IDs).
<p>

Your EMBL format file has IDs that look like accession numbers, so
EMBOSS fills in the accession number for each sequence, and reports it
in the FASTA format. 



</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

