<HTML>
<HEAD>
  <TITLE>
  EMBOSS: seqretsplit
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">


<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
seqretsplit
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
    Function
</H2>
Reads sequences and writes them to individual files
<H2>
    Description
</H2>

<p><b>seqretsplit</b> is a variant of the standard program for reading and writing sequences, <b>seqret</b>.   It performs exactly the same function except that when it reads more than one sequence, it writes each sequence to an individual file. In all other respects, <b>skipseq</b> is the same as <b>seqret</b>.  Its main use is therefore to split a file containing multiple sequences into many files, each containing one sequence. There are many options built-in into EMBOSS for detailed specification of the input and output sequences, for example the sequence type and file format.  Optionally, feature information will be read and written.</p>


<H2>
    Usage
</H2>
<b>Here is a sample session with seqretsplit</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>seqretsplit tembl:m1190* </b>
Reads sequences and writes them to individual files
output sequence(s) [m11903.fasta]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>

<p>

The specification of the output file is not used in this case.
<p>

At some point this ought to change and you will not be prompted for the
output file at all. 

<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-sequence]          seqall     (Gapped) sequence(s) filename and optional
                                  format, or reference (input USA)
  [-outseq]            seqoutall  [<sequence>.<format>] Sequence set(s)
                                  filename and optional format (output USA)

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers:
   -feature            boolean    Use feature information
   -firstonly          boolean    Read one sequence and stop

   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1            integer    Start of each sequence to be used
   -send1              integer    End of each sequence to be used
   -sreverse1          boolean    Reverse (if DNA)
   -sask1              boolean    Ask for begin/end/reverse
   -snucleotide1       boolean    Sequence is nucleotide
   -sprotein1          boolean    Sequence is protein
   -slower1            boolean    Make lower case
   -supper1            boolean    Make upper case
   -sformat1           string     Input sequence format
   -sdbname1           string     Database name
   -sid1               string     Entryname
   -ufo1               string     UFO features
   -fformat1           string     Features format
   -fopenfile1         string     Features file name

   "-outseq" associated qualifiers
   -osformat2          string     Output seq format
   -osextension2       string     File name extension
   -osname2            string     Base file name
   -osdirectory2       string     Output directory
   -osdbname2          string     Database name to add
   -ossingle2          boolean    Separate file for each entry
   -oufo2              string     UFO features
   -offormat2          string     Features format
   -ofname2            string     Features file name
   -ofdirectory2       string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>(Gapped) sequence(s) filename and optional format, or reference (input USA)</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outseq]<br>(Parameter 2)</td>
<td>Sequence set(s) filename and optional format (output USA)</td>
<td>Writeable sequence(s)</td>
<td><i>&lt;*&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-feature</td>
<td>Use feature information</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-firstonly</td>
<td>Read one sequence and stop</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

</table>

<H2>
    Input file format
</H2>

<b>seqretsplit</b> reads a normal sequence USA.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:m1190*' is a sequence entry in the example nucleic acid database 'tembl'
<p>


<H2>
    Output file format
</H2>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: m11903.fasta</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt;M11903 M11903.1 Rattus norvegicus androgen-responsive protein precursor (Svf) gene, exons 1 and 1A, alternatively spliced.
cctttcaaatagaaactctcgtgaaggctgtctgagaacacaagctcaaggttgtgactg
atttcagtgatgccgtcttgaagagggataccgtgctagagaatgactcctgatcaaccc
tgaagacttctgcaagcccgaagtcgtgcttccccactctgaactgacatatgttcagga
agtagagacgtgcaccgttggatgttctcaaggtaaaaaggaagatttggaagaatgctc
tagtgttgttgccttggagaggaccagggaacagtacaagactcctactgagcagagaga
aaggagcctgacatttaccgataagaaaggtcatttgccttccaacctgtaggcaaggcc
agacaaggaaatatataaaggagaacctcagatcagctctcagtcaagacccttcctgac
aagatgagtcccaccgggttcttcctccttacggtgctccttgttctggtgacagaagca
gcctcgagggggccccgaggtgagtggcaattttgtgctatgggaaagatgtttgagaac
tatgttctcaaaagggagtctgcagaatgctgtgttcccagggcttctccatgaaggaaa
cttgagtcttttcaagctttaaccatagtcctactgtgagtctctgtgacttgacaagca
acattgctggtaaggagggctgagggggaatgcgggcaacggcctcgggtaacatcctca
ttgt
</pre>
</td></tr></table><p>
<p><h3>File: m11904.fasta</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt;M11904 M11904.1 Rattus norvegicus androgen-responsive protein precursor (Svf) gene, exon 2 and complete cds.
ggtatctccaaacacagcagctggctctcaacagagagtcctcatgcacaactaatccaa
gatacagaaagtggatatagagaatgagacattgttttctctcaacagaaaaattctcac
agtcagctgaagacccttatagtgaaaacatgaatctaaagattctggcgagcgggaggg
gatcaagttctacctttggggcattcagccgaagtgagaactctcggagtaacttcaaat
caaaaagtccaagcagtatcaccagggagaaagtgaatgaggaaagcaggagtgaaatga
gtagtaccagcagccattttggtctcaaaatgagaagatctcatggaggaggagaaatga
atccctttgaaaccaaagtaaagacccggatcactcgcaaataatgtgttccccggccaa
ctgaagacttgagcccaataggcaggtaagtgttatcaccaggtgagggcttacaaacta
ctcgtgcctaatccctaggccattgtaggattgtgcacgcagtaaagttgctataagggg
aggtatggaaacgacctacaaggcagacaaagatacgagctatactgtgt
</pre>
</td></tr></table><p>
<p><h3>File: m11905.fasta</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt;M11905 M11905.1 Rattus norvegicus androgen-responsive protein precursor (Svf) gene, exon 3.
ccgtgcaatctcttcctgtgtccacacagccctgttagaagcaactctctcgttctcaag
gccctacctgcaagaactacctttctcttcctccgcccaacaaaggaggaatgtttctgc
ttgtgacccaccagagatgaaatatagcagtgtcctgcagtaaaggggggccccagaggc
atgggacatacacgcattaatccctccacgtcttccctgtcctacctcacaggttgtcct
cgttccctgggtgtcactgaactaagagaagtctatgatgtcttcaggatgcaggatccc
acaggtgccccggaaatagtccgtgcttcttatttcctccttacacttgttttctttaag
attccggaacctgacaagattcaaatttaaccttttcaataaaaaagatactattctgca
tcattatctcctgaaatctcttgcttctgcagtacaggggctgggtgggattcctaaact
tgaccagttctgccgttaaaggaagatcccttctgtgccgtatcagagactatttccaga
ctctggataga
</pre>
</td></tr></table><p>

<p>

One file for each input sequence is written out.

<p>

The names of the files it creates are derived from the ID name of the
sequence, followed by an extension denoting the format of the sequence. 
You have no control over the names of the files it writes out. 

<p>

For example, if the files <i>embl:hsfa11*</i> are read in and the
output is specified as <i>wibble.seq</i>, then the following files are
expected to be created:

<p>

<pre>
hsfa110.fasta
hsfa111.fasta
hsfa112.fasta
hsfa113.fasta
hsfa114.fasta
</pre>

<p>

(No file <i>wibble.seq</i> is created.)

<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

<p>See the documentation for <b>seqret</b> to see the full range of things that you can do when reading and writing sequences.</p>
<p>Some non-EMBOSS programs will accept only single sequences.  In such cases <b>seqretsplit</b> is useful for splitting a multiple sequence file into many individual files.  Some EMBOSS programs will also read only a single sequence, which may, however, be one of many in a file. You can specify the sequence using the USA filename:sequenceID.  Nonetheless, some people feel more comfortable handling one sequence per file, so <b>seqretsplit</b> will be useful to them too.</p>

<p>One file for each input sequence is written.  The names of the files it creates are derived from the ID name of the sequence, followed by an extension denoting the format of the sequence. You have no control over the names of the files it writes out. For example, if the files <tt>embl:hsfa11*</tt> are read in and the output is specified as <tt>wibble.seq</tt>, then the following files are expected to be created:</p>

<pre>
hsfa110.fasta
hsfa111.fasta
hsfa112.fasta
hsfa113.fasta
hsfa114.fasta
</pre>

<p>(No file wibble.seq is created.)</p>

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with status 0.

<H2>
    Known bugs
</H2>

It shouldn't really prompt for the output filename.
<p>
This is a side effect of the way sequence output works in
EMBOSS. Writing multiple sequences to separate files (the -ossingle
qualifier) does this, and seqretsplit has set it automatically on.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="aligncopy.html">aligncopy</a></td>
<td>Reads and writes alignments</td>
</tr>

<tr>
<td><a href="aligncopypair.html">aligncopypair</a></td>
<td>Reads and writes pairs from alignments</td>
</tr>

<tr>
<td><a href="biosed.html">biosed</a></td>
<td>Replace or delete sequence sections</td>
</tr>

<tr>
<td><a href="codcopy.html">codcopy</a></td>
<td>Copy and reformat a codon usage table</td>
</tr>

<tr>
<td><a href="cutseq.html">cutseq</a></td>
<td>Removes a section from a sequence</td>
</tr>

<tr>
<td><a href="degapseq.html">degapseq</a></td>
<td>Removes non-alphabetic (e.g. gap) characters from sequences</td>
</tr>

<tr>
<td><a href="descseq.html">descseq</a></td>
<td>Alter the name or description of a sequence</td>
</tr>

<tr>
<td><a href="entret.html">entret</a></td>
<td>Retrieves sequence entries from flatfile databases and files</td>
</tr>

<tr>
<td><a href="extractalign.html">extractalign</a></td>
<td>Extract regions from a sequence alignment</td>
</tr>

<tr>
<td><a href="extractfeat.html">extractfeat</a></td>
<td>Extract features from sequence(s)</td>
</tr>

<tr>
<td><a href="extractseq.html">extractseq</a></td>
<td>Extract regions from a sequence</td>
</tr>

<tr>
<td><a href="featcopy.html">featcopy</a></td>
<td>Reads and writes a feature table</td>
</tr>

<tr>
<td><a href="featreport.html">featreport</a></td>
<td>Reads and writes a feature table</td>
</tr>

<tr>
<td><a href="listor.html">listor</a></td>
<td>Write a list file of the logical OR of two sets of sequences</td>
</tr>

<tr>
<td><a href="makenucseq.html">makenucseq</a></td>
<td>Create random nucleotide sequences</td>
</tr>

<tr>
<td><a href="makeprotseq.html">makeprotseq</a></td>
<td>Create random protein sequences</td>
</tr>

<tr>
<td><a href="maskambignuc.html">maskambignuc</a></td>
<td>Masks all ambiguity characters in nucleotide sequences with N</td>
</tr>

<tr>
<td><a href="maskambigprot.html">maskambigprot</a></td>
<td>Masks all ambiguity characters in protein sequences with X</td>
</tr>

<tr>
<td><a href="maskfeat.html">maskfeat</a></td>
<td>Write a sequence with masked features</td>
</tr>

<tr>
<td><a href="maskseq.html">maskseq</a></td>
<td>Write a sequence with masked regions</td>
</tr>

<tr>
<td><a href="newseq.html">newseq</a></td>
<td>Create a sequence file from a typed-in sequence</td>
</tr>

<tr>
<td><a href="nohtml.html">nohtml</a></td>
<td>Remove mark-up (e.g. HTML tags) from an ASCII text file</td>
</tr>

<tr>
<td><a href="noreturn.html">noreturn</a></td>
<td>Remove carriage return from ASCII files</td>
</tr>

<tr>
<td><a href="nospace.html">nospace</a></td>
<td>Remove all whitespace from an ASCII text file</td>
</tr>

<tr>
<td><a href="notab.html">notab</a></td>
<td>Replace tabs with spaces in an ASCII text file</td>
</tr>

<tr>
<td><a href="notseq.html">notseq</a></td>
<td>Write to file a subset of an input stream of sequences</td>
</tr>

<tr>
<td><a href="nthseq.html">nthseq</a></td>
<td>Write to file a single sequence from an input stream of sequences</td>
</tr>

<tr>
<td><a href="pasteseq.html">pasteseq</a></td>
<td>Insert one sequence into another</td>
</tr>

<tr>
<td><a href="revseq.html">revseq</a></td>
<td>Reverse and complement a nucleotide sequence</td>
</tr>

<tr>
<td><a href="seqret.html">seqret</a></td>
<td>Reads and writes (returns) sequences</td>
</tr>

<tr>
<td><a href="sizeseq.html">sizeseq</a></td>
<td>Sort sequences by size</td>
</tr>

<tr>
<td><a href="skipredundant.html">skipredundant</a></td>
<td>Remove redundant sequences from an input set</td>
</tr>

<tr>
<td><a href="skipseq.html">skipseq</a></td>
<td>Reads and writes (returns) sequences, skipping first few</td>
</tr>

<tr>
<td><a href="splitter.html">splitter</a></td>
<td>Split sequence(s) into smaller sequences</td>
</tr>

<tr>
<td><a href="trimest.html">trimest</a></td>
<td>Remove poly-A tails from nucleotide sequences</td>
</tr>

<tr>
<td><a href="trimseq.html">trimseq</a></td>
<td>Remove unwanted characters from start and end of sequence(s)</td>
</tr>

<tr>
<td><a href="trimspace.html">trimspace</a></td>
<td>Remove extra whitespace from an ASCII text file</td>
</tr>

<tr>
<td><a href="union.html">union</a></td>
<td>Concatenate multiple sequences into a single sequence</td>
</tr>

<tr>
<td><a href="vectorstrip.html">vectorstrip</a></td>
<td>Removes vectors from the ends of nucleotide sequence(s)</td>
</tr>

<tr>
<td><a href="yank.html">yank</a></td>
<td>Add a sequence reference (a full USA) to a list file</td>
</tr>

</table>


<H2>
    Author(s)
</H2>

Peter Rice (pmr&nbsp;&copy;&nbsp;ebi.ac.uk)
<br>
Informatics Division, European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK

<H2>
    History
</H2>

Written (Jan 2000) - Peter Rice

<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.

<H2>
    Comments
</H2>
The specification of the output file is not used when the output file
names are generated automatically..
<p>

At some point this ought to change and you will not be prompted for the
output file at all. 



</BODY>
</HTML>

