<!-- START OF HEADER -->
<HTML><HEAD>
<TITLE> EMBOSS: SEQSORT documentation. </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">
<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="../index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="../../emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

<H2> SEQSORT documentation </H2>
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>







<!-- CONTENTS
     This always includes the sections below.
     Other subsections can be added for individual applications.
-->
<br><H2>CONTENTS </H2>
<b> <a href="#1.0">1.0     SUMMARY                   </a></b><br>
<b> <a href="#2.0">2.0     INPUTS & OUTPUTS          </a></b><br>
<b> <a href="#3.0">3.0     INPUT FILE FORMAT         </a></b><br>
<b> <a href="#4.0">4.0     OUTPUT FILE FORMAT        </a></b><br>
<b> <a href="#5.0">5.0     DATA FILES                </a></b><br>
<b> <a href="#6.0">6.0     USAGE                     </a></b><br>    
<b> <a href="#7.0">7.0     KNOWN BUGS & WARNINGS     </a></b><br>    
<b> <a href="#8.0">8.0     NOTES                     </a></b><br>
<b> <a href="#9.0">9.0     DESCRIPTION               </a></b><br>
<b> <a href="#10.0">10.0   ALGORITHM                 </a></b><br>
<b> <a href="#11.0">11.0   RELATED APPLICATIONS      </a></b><br>
<b> <a href="#12.0">12.0   DIAGNOSTIC ERROR MESSAGES </a></b><br>
<b> <a href="#13.0">13.0   AUTHORS                   </a></b><br>
<b> <a href="#14.0">14.0   REFERENCES                </a></b><br>

<!-- SUMMARY
     Succint description of the application, particularly its inputs, outputs
     and what it does.  The same text is given at the top of the source (.c)
     file and in the <documentation> attribute of the <application definition>
     of the ACD file.
-->
<a name="1.0"></a>
<br><br><br><H2> 1.0   SUMMARY  </H2>

Reads DHF files (domain hits files) of database hits (sequences) and removes hits of ambiguous classification

<!-- INPUTS & OUTPUTS
     Short summary of the application inputs and outputs in its different 
     modes of usage (if appropriate). More detail than the summary.
-->
<a name="2.0"></a>
<br><br><br><H2> 2.0   INPUTS & OUTPUTS          </H2>

SEQSORT reads a directory of DHF files (domain hits files) where each
file contains hits to a single SCOP family, compares, processes and
collates the hits and writes a directory of DHF files which contain
only those hits that could be uniquely assigned to a SCOP family.
Optionally, two further files of hits are written: (i) a domain
families file, of ALL hits from the input files that could be uniquely
assigned to a SCOP family and (ii) a domain ambiguities file, of hits
from ALL the input files that are of ambiguous family assignment and
are assigned as relatives to a SCOP superfamily or fold instead.

<br>
The path for the domain hits files (input and output) and the
names of the output files are specified by the user.  The file
extension of the DHF files are set in the ACD file.





<!-- INPUT FILE FORMAT 
     Description and example(s) of input file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.

     Cannot use the test data files because they might be empty or need 
     hand-editing 

     Use "<b>SEQSORT</b> reads any normal sequence USAs." if
     appropriate.
-->
<a name="3.0"></a>
<br><br><br><H2> 3.0   INPUT FILE FORMAT         </H2>
The format of the domain hits file is described in <a href="seqsearch.html">SEQSEARCH documentation.</a>





<!-- OUTPUT FILE FORMAT 
     Description and example(s) of output file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.  
     If the standard description of the avalable report formats is required, 
     use:   #include file="inc/reportformats.ihtml"
     Use "Outputs a graph to the specified graphics device."
     or  "outputs a report format file. The default format is ..."
     if appropriate.
-->
<a name="4.0"></a>
<br><br><br><H2> 4.0   OUTPUT FILE FORMAT         </H2>
The format of the domain hits file is described in <a href="seqsearch.html">SEQSEARCH documentation.</a>

<br><br> The domain families file and domain ambiguities file also use the DHF format.  Whereas normally a DHF file contains hits for a single node from SCOP or CATH, the families and ambiguities files may contain domains from multiple different families (domain families file), or superfamilies or folds (ambiguities file).  Domains of the same node (e.g. family) will be grouped together in blocks, i.e. all hits for domain A, then all hits for domain B and so on (see Figure 1).


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: scop.fam</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
CL   Alpha and beta proteins (a+b)
XX
FO   Ferredoxin-like
XX
SF   Aspartate carbamoyltransferase, Regulatory-chain, N-terminal domain
XX
FA   Aspartate carbamoyltransferase, Regulatory-chain, N-terminal domain
XX
SI   54894
XX
NS   27
XX
NN   [1]
XX
TY   HIT
XX
SC   0.00
XX
AC   Q9YBD5
XX
RA   11 START; 105 END;
XX
SQ   SEQUENCE    95 AA;  10687 MW;  B8ACF34B CRC32;
     VRKIRSGVVI DHIPPGRAFT MLKALGLLPP RGYRWRIAVV INAESSKLGR KDILKIEGYK
     PRQRDLEVLG IIAPGATFNV IEDYKVVEKV KLKLP
XX
NN   [2]
XX
TY   HIT
XX
SC   0.00
XX
AC   Q9UX07
XX
RA   12 START; 104 END;
XX
SQ   SEQUENCE    93 AA;  10312 MW;  12DDDE7C CRC32;
     VSKIRNGTVI DHIPAGRALA VLRILGIRGS EGYRVALVMN VESKKIGRKD IVKIEDRVID
     EKEASLITLI APSATINIIR DYVVTEKRHL EVP
XX
NN   [3]
XX
TY   HIT
XX
SC   0.00
XX
AC   Q9KP65
XX
RA   9 START; 100 END;
XX


<font color=red>  [Part of this file has been deleted for brevity]</font>

     MMFHKIYIQK HDNVSILFAD IEGFTSLASQ CTAQELVMTL NELFARFDKL AAENHCLRIK
     ILGDCYYCVS GLPEARADHA HCCVEMGVDM IEAISLVREV TGVNVNMRVG IHSGRVHCGV
     LGLRKWQFDV WSNDVTLANH MEAGGRAGRI HITRATLQYL NGDYEVEPGR GGERNAYLKE
     QHIETFLIL
XX
NN   [111]
XX
TY   HIT
XX
SC   0.00
XX
AC   O30820
XX
RA   241 START; 400 END;
XX
SQ   SEQUENCE   160 AA;  17506 MW;  B1E1A024 CRC32;
     NIIADKYDEA SVLFADIVGF TERASSTAPA DLVRFLDRLY SAFDELVDQH GLEKIKVSGD
     SYMVVSGVPR PRPDHTQALA DFALDMTNVA AQLKDPRGNP VPLRVGLATG PVVAGVVGSR
     RFFYDVWGDA VNVASRMEST DSVGQIQVPD EVYERLKDDF
XX
NN   [112]
XX
TY   HIT
XX
SC   0.00
XX
AC   O19179
XX
RA   877 START; 1032 END;
XX
SQ   SEQUENCE   156 AA;  17212 MW;  675838EA CRC32;
     PEYFEEVTLY FSDIVGFTTI SAMSEPIEVV DLLNDLYTLF DAIIGSHDVY KVETIGDAYM
     VASGLPQRNG QRHAAEIANM ALDILSAVGS FRMRHMPEVP VRIRIGLHSG PCVAGVVGLT
     MPRYCLFGDT VNTASRMEST GLPYRIHVNM STVRIL
XX
NN   [113]
XX
TY   HIT
XX
SC   0.00
XX
AC   O02740
XX
RA   877 START; 1042 END;
XX
SQ   SEQUENCE   166 AA;  18151 MW;  E2BB2824 CRC32;
     PEGFDLVTLY FSDIVGFTTI SAMSEPIEVV DLLNDLYTLF DAIIGSHDVY KVETIGDAYM
     VASGLPKRNG MRHAAEIANM SLDILSSVGT FKMRHMPEVP VRIRIGLHSG PVVAGVVGLT
     MPRYCLFGDT VNTASRMEST GLPYRIHVSH STVTILRTLG EGYEVE
XX
//
</pre>
</td></tr></table><p>
<p><h3>File: scop.oth</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
</pre>
</td></tr></table><p>


<!--
<br><br><b>Figure 1   Excerpt from a domain families file (output)</b>
<table><td bgcolor="#CFCCFF">
<pre>
> P00478^.^8^99^54894^Alpha and beta proteins (a+b)^.^.^Ferredoxin-like^Aspartate carbamoyltransferase, Regulatory-chain, N-terminal domain^Aspartate carbamoyltransferase, Regulatory-chain, N-terminal domain^PSIBLAST^191.00^0.000e+00^0.000e+00
VEAIKRGTVIDHIPAQIGFKLLSLFKLTETDQRITIGLNLPSGEMGRKDLIKIENTFLSEDQVDQLALYAPQATVNRIDNYEVVGKSRPSLP
> O58452^.^6^99^54894^Alpha and beta proteins (a+b)^.^.^Ferredoxin-like^Aspartate carbamoyltransferase, Regulatory-chain, N-terminal domain^Aspartate carbamoyltransferase, Regulatory-chain, N-terminal domain^PSIBLAST^94.30^0.000e+00^6.000e-20
VSAIKEGTVIDHIPAGKGLKVIEILGLSKLSNGGSVLLAMNVPSKKLGRKDIVKVEGKFLSEEEVNKIALVAPTATVNIIRNYKVVEKFKVEVP
> O30129^.^6^98^54894^Alpha and beta proteins (a+b)^.^.^Ferredoxin-like^Aspartate carbamoyltransferase, Regulatory-chain, N-terminal domain^Aspartate carbamoyltransferase, Regulatory-chain, N-terminal domain^PSIBLAST^79.60^0.000e+00^2.000e-15
VSKIKEGTVIDHINAGKALLVLKILKIQPGTDLTVSMAMNVPSSKMGKKDIVKVEGMFIRDEELNKIALISPNATINLIRDYEIERKFKVSPP
> O26938^.^11^101^54894^Alpha and beta proteins (a+b)^.^.^Ferredoxin-like^Aspartate carbamoyltransferase, Regulatory-chain, N-terminal domain^Aspartate carbamoyltransferase, Regulatory-chain, N-terminal domain^PSIBLAST^81.90^0.000e+00^3.000e-16
VKPIKNGTVIDHITANRSLNVLNILGLPDGRSKVTVAMNMDSSQLGSKDIVKIENRELKPSEVDQIALIAPRATINIVRDYKIVEKAKVRL
> Q9WVI4^.^513^667^55074^Alpha and beta proteins (a+b)^.^.^Ferredoxin-like^Adenylyl and guanylyl cyclase catalytic domain^Adenylyl and guanylyl cyclase catalytic domain^PSIBLAST^113.00^0.000e+00^2.000e-25
RKFDDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDIYKVETIGDAYCVASGLHRKSLCHAKPIALMALKMMELSEEVLTPDGRPIQMRIGIHSGSVLAGVVGVRMPRYCLFGNNVTLASKFESGSHPRRINISPTTYQLLKRE
> Q9ERL9^.^475^626^55074^Alpha and beta proteins (a+b)^.^.^Ferredoxin-like^Adenylyl and guanylyl cyclase catalytic domain^Adenylyl and guanylyl cyclase catalytic domain^PSIBLAST^101.00^0.000e+00^9.000e-22
KKFSEVTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYKVETIGDAYCVAGGLHRESDTHAVQIALMALKMMELSNEVMSPHGEPIKMRIGLHSGSVFAGVVGVKMPRYCLFGNNVTLANKFESCSVPRKINVSPTTYRLL
</pre>
</table>


-->

<!-- DATA FILES         
     Any data files used (e.g. translation table file, substitution matrix 
     etc.  This includes example data file formats if they are not obvious.
     For a standard description of what data files are and how embossdata can
     be used to inspect and retrieve them, use:
     #include file="inc/localfiles.ihtml" 
-->

<a name="5.0"></a>
<br><br><br><H2> 5.0   DATA FILES                </H2>
None.







<!-- USAGE
     Example usage, as run from the command-line.
     Many examples illustrating different behaviours is good.
-->
<a name="6.0"></a>
<br><br><br><H2> 6.0   USAGE                     </H2>
<H3> 6.1   COMMAND LINE ARGUMENTS </H3>
 

<pre>
   Standard (Mandatory) qualifiers (* if not always prompted):
  [-dhfindir]          directory  This option specifies the location of DHF
                                  files (domain hits files) (input). A 'domain
                                  hits file' contains database hits
                                  (sequences) with domain classification
                                  information, in the DHF format (FASTA or
                                  EMBL-like). The hits are relatives to a SCOP
                                  or CATH family and are found from a search
                                  of a sequence database. Files containing
                                  hits retrieved by PSIBLAST are generated by
                                  using SEQSEARCH.
   -overlap            integer    This option specifies the number of
                                  overlapping residues required for merging of
                                  two hits. Each family is also processed so
                                  that ovlerapping hits (hits with identical
                                  accesssion number that overlap by at least a
                                  user-defined number of residues) are
                                  replaced by a hit that is produced from
                                  merging the two overlapping hits.
   -dofamilies         toggle     This option specifies to write a domain
                                  families file. If this option is set a
                                  domain families file is written.
   -doambiguities      toggle     This option specifies whether to write a
                                  domain ambiguities file. If this option is
                                  set a domain ambiguities file is written.
  [-dhfoutdir]         outdir     This option specifies the location of DHF
                                  files (domain hits files) (output). A
                                  'domain hits file' contains database hits
                                  (sequences) with domain classification
                                  information, in the DHF format (FASTA or
                                  EMBL-like). The hits are relatives to a SCOP
                                  or CATH family and are found from a search
                                  of a sequence database. Files containing
                                  hits retrieved by PSIBLAST are generated by
                                  using SEQSEARCH.
*  -hitsfile           outfile    This option specifies the name of domain
                                  families file (output). A 'domain families
                                  file' contains sequence relatives (hits) for
                                  each of a number of different SCOP or CATH
                                  families found from searching a sequence
                                  database, e.g. by using SEQSEARCH
                                  (psiblast). The file contains the collated
                                  search results for the indvidual families;
                                  only those hits of unambiguous family
                                  assignment are included. Hits of ambiguous
                                  family assignment are assigned as relatives
                                  to a SCOP or CATH superfamily or fold
                                  instead and are collated into a 'domain
                                  ambiguities file'. The domain families and
                                  ambiguities files are generated by using
                                  SEQSORT and use the same format as a DHF
                                  file (domain hits file).
*  -ambigfile          outfile    This option specifies the name of domain
                                  ambiguities file (output). A 'domain
                                  families file' contains sequence relatives
                                  (hits) for each of a number of different
                                  SCOP or CATH families found from searching a
                                  sequence database, e.g. by using SEQSEARCH
                                  (psiblast). The file contains the collated
                                  search results for the indvidual families;
                                  only those hits of unambiguous family
                                  assignment are included. Hits of ambiguous
                                  family assignment are assigned as relatives
                                  to a SCOP or CATH superfamily or fold
                                  instead and are collated into a 'domain
                                  ambiguities file'. The domain families and
                                  ambiguities files are generated by using
                                  SEQSORT and use the same format as a DHF
                                  file (domain hits file).

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-hitsfile" associated qualifiers
   -odirectory         string     Output directory

   "-ambigfile" associated qualifiers
   -odirectory         string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report deaths


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-dhfindir]<br>(Parameter 1)</td>
<td>This option specifies the location of DHF files (domain hits files) (input). A 'domain hits file' contains database hits (sequences) with domain classification information, in the DHF format (FASTA or EMBL-like). The hits are relatives to a SCOP or CATH family and are found from a search of a sequence database. Files containing hits retrieved by PSIBLAST are generated by using SEQSEARCH.</td>
<td>Directory</td>
<td>./</td>
</tr>

<tr>
<td>-overlap</td>
<td>This option specifies the number of overlapping residues required for merging of two hits. Each family is also processed so that ovlerapping hits (hits with identical accesssion number that overlap by at least a user-defined number of residues) are replaced by a hit that is produced from merging the two overlapping hits.</td>
<td>Any integer value</td>
<td>10</td>
</tr>

<tr>
<td>-dofamilies</td>
<td>This option specifies to write a domain families file. If this option is set a domain families file is written.</td>
<td>Toggle value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-doambiguities</td>
<td>This option specifies whether to write a domain ambiguities file. If this option is set a domain ambiguities file is written.</td>
<td>Toggle value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>[-dhfoutdir]<br>(Parameter 2)</td>
<td>This option specifies the location of DHF files (domain hits files) (output). A 'domain hits file' contains database hits (sequences) with domain classification information, in the DHF format (FASTA or EMBL-like). The hits are relatives to a SCOP or CATH family and are found from a search of a sequence database. Files containing hits retrieved by PSIBLAST are generated by using SEQSEARCH.</td>
<td>Output directory</td>
<td>./</td>
</tr>

<tr>
<td>-hitsfile</td>
<td>This option specifies the name of domain families file (output). A 'domain families file' contains sequence relatives (hits) for each of a number of different SCOP or CATH families found from searching a sequence database, e.g. by using SEQSEARCH (psiblast). The file contains the collated search results for the indvidual families; only those hits of unambiguous family assignment are included. Hits of ambiguous family assignment are assigned as relatives to a SCOP or CATH superfamily or fold instead and are collated into a 'domain ambiguities file'. The domain families and ambiguities files are generated by using SEQSORT and use the same format as a DHF file (domain hits file).</td>
<td>Output file</td>
<td>fam.dhf</td>
</tr>

<tr>
<td>-ambigfile</td>
<td>This option specifies the name of domain ambiguities file (output). A 'domain families file' contains sequence relatives (hits) for each of a number of different SCOP or CATH families found from searching a sequence database, e.g. by using SEQSEARCH (psiblast). The file contains the collated search results for the indvidual families; only those hits of unambiguous family assignment are included. Hits of ambiguous family assignment are assigned as relatives to a SCOP or CATH superfamily or fold instead and are collated into a 'domain ambiguities file'. The domain families and ambiguities files are generated by using SEQSORT and use the same format as a DHF file (domain hits file).</td>
<td>Output file</td>
<td>oth.dhf</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<H3> 6.2 EXAMPLE SESSION </H3>
An example of interactive use of SEQSORT is shown below.

<b>Here is a sample session with seqsort</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>seqsort </b>
Reads multiple files of hits and writes (i) a scop families file and (ii) a
scop ambiguities file.
Mode of operation
         1 : Process psiblast hits
         2 : Process swissparse hits (WARNING - untested)
         3 : Merge psiblast and swissparse hits (WARNING - untested)
Select mode [1]: <b></b>
Location of scop hits files (input) [./]: <b>structure</b>
Extension of scop hits files (input) [.hits]: <b></b>
Number of overlapping residues required for merging of two hits [10]: <b>20</b>
Name of scop families file (output) [fam.hits]: <b>scop.fam</b>
Name of ambiguities file (output) [oth.hits]: <b>scop.oth</b>
Warning: Empty list in embDmxScophitsToHitlist
Warning: Empty list in embDmxScophitsToHitlist
Warning: Empty list in embDmxScophitsToHitlist

</pre></td></tr></table><p>
<p>
<a href="#output.1">Go to the output files for this example</a><p><p>


All DHF files (of the file extension .dhf specicied in ACD) in
directory /test_data were processed by SEQSORT.  DHF output files were
written to /test_data/seqsort/seqsort_hits.  A shared segment of 10 or
more identical was required to class two sequences as overlapping.
Domain families and ambiguities files called fam.dhf and oth.dhf
respectively were written to /test_data/seqsort.

<!-- KNOWN BUGS & WARNINGS
     Bugs that have not yet been fixed, easily missued features, problems
     and caveats etc. Potentially stupid things the program will let you do.
-->
<a name="7.0"></a>
<br><br><br><H2> 7.0   KNOWN BUGS & WARNINGS     </H2>
None.





<!-- NOTES
     Important general remarks, including:
       Restrictions.
       Interesting behaviour.
       Useful things you can do with this program.
       Future plans.
       etc.
-->
<a name="8.0"></a>
<br><br><br><H2> 8.0   NOTES                     </H2>
None.


<H3> 8.1   GLOSSARY OF FILE TYPES    </H3>
<table BORDER CELLSPACING=5 CELLPADDING=5 BGCOLOR="#f5f5ff" >
<tr>
<td><b>FILE TYPE</b></td>
<td><b>FORMAT</b></td>
<td><b>DESCRIPTION</b></td>
<td><b>CREATED BY <b></td>
<td><b>SEE ALSO</b></td>
</tr>
<tr>
<td><b>Domain hits file </b></td>
<td> DHF format (FASTA-like with domain classification information). </td>
<td> Database hits (sequences) with domain classification information.  The hits are relatives to a SCOP or CATH family and are found from a search of a sequence database. </td>
<td> <a href="seqsearch.html">SEQSEARCH</a> (hits retrieved by PSIBLAST)</td>
<td> N.A. </td>
</tr>
<tr>
<td><b> Domain families & ambiguities file </b></td>
<td> Contains sequence relatives (hits) for each of a number of different SCOP or CATH families found from PSIBLAST searches of a sequence database.  The file contains the collated search results for the indvidual families; only those hits of unambiguous family assignment are included.  Hits of ambiguous family assignment are assigned as relatives to a SCOP or CATH superfamily or fold instead and are collated into a 'domain ambiguities file'. </td>
<td>  The domain families and ambiguities files are generated by using SEQSORT and use the same format as a DHF file (domain hits file). </td>
<td> N.A. </td>
</tr>
<tr>
<td><b>Domain validation file </b></td>
<td> Contains sequence relatives (hits) for each of a number of different SCOP or CATH families, superfamilies and folds.  The file contains the collated results from PSIBLAST searches of a sequence database for the indvidual families; hits of unambiguous family assignment are listed under their respective family, otherwise a hit is assigned as relatives to a superfamily or fold instead.   </td>
<td> The domain validation file is generated by using SEQNR and uses the same format as a DHF file (domain hits file). </td>
<td> N.A. </td>
</tr>
</table>




None




<!-- DESCRIPTION
     A complete, non-technical, user-level description of the application.
-->
<a name="9.0"></a>
<br><br><br><H2> 9.0   DESCRIPTION               </H2>

The results of multiple searches of a sequence database using an
homology search tool such as blast may contain overlapping or
identical hits, especially where the query sequences are related, for
instance, belong to different families but the same superfamily.  For
certain analyses it is desirable to assign a hit with confidence to a
unique family, or otherwise assign it as a member of a larger
superfamily or fold instead.  SEQSORT reads a directory of DHF files
(domain hits files) where each file containing hits to a different
SCOP family, compares, processes and collates the hits and writes a
directory of DHF files which contain only those hits that could be
uniquely assigned to a SCOP family.  Optionally, two further files are
written: (i) a domain families file, of hits (from ALL the input
files) that could be uniquely assigned to a SCOP family and (ii) a
domain ambiguities file, for hits (from ALL the input files) of
ambiguous family assignment which are assigned as relatives to a SCOP
superfamily or fold instead.


<!-- ALGORITHM
     A technical description of algorithmic aspects, describing exactly how
     the key aspects of the application work.
-->
<a name="10.0"></a>
<br><br><br><H2> 10.0  ALGORITHM                 </H2>

A rough outline of the algorithm follows; a better description will
appear in a publication in preparation.  Hits from searches for all
domain families are collated into a single list and the list sorted
according to family name.  The hits hits within each family are sorted
by accession number, then hits within a family and with identical
accession number are sorted by the start position of the hit relative
to the full length sequence in swissprot.  In each family identical
hits (i.e. those with identical accession number and the same start
and end points relative to the full-length sequence in swissprot) were
removed leaving only a single copy.  Each family is also processed so
that ovlerapping hits (hits with identical accesssion number that
overlap by at least a user-defined number of residues) are replaced by
a hit that is produced from merging the two overlapping hits.  If two
hits have the same accession number and overlap but are from searches
for different families, the hits are merged and the merged hit placed
into a new list for hits to superfamilies (if the two families
belonged to the same superfamily) or for hits to folds (if the two
families were in different superfamilies but the same fold).  In this
way hits that are unique to a particular family are identified, and
hits of ambiguous family assignment are assigned as belonging to a
superfamily or fold instead.





<!-- RELATED APPLICATIONS
     Other applications that either generate the input, use the output or 
     are in some other way related to the application are described here. 
     (Take this from "Sister applications" in the old documentation)
-->
<a name="11.0"></a>
<br><br><br><H2> 11.0  RELATED APPLICATIONS      </H2>

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="../construct/contactcount.html">contactcount</a></td><td>Counts specific versus non-specific contacts in a directory of cleaned protein chain contact files</td></tr>
<tr><td><a href="contacts.html">contacts</a></td><td>Reads CCF files (clean coordinate files) and writes CON files (contact files) of intra-chain residue-residue contact data</td></tr>
<tr><td><a href="domainalign.html">domainalign</a></td><td>Generates DAF files (domain alignment files) of structure-based sequence alignments for nodes in a DCF file (domain classification file)</td></tr>
<tr><td><a href="domainrep.html">domainrep</a></td><td>Reorder DCF file (domain classification file) so that the representative structure of each user-specified node is given first</td></tr>
<tr><td><a href="domainreso.html">domainreso</a></td><td>Removes low resolution domains from a DCF file (domain classification file)</td></tr>
<tr><td><a href="interface.html">interface</a></td><td>Reads CCF files (clean coordinate files) and writes CON files (contact files) of inter-chain residue-residue contact data</td></tr>
<tr><td><a href="libgen.html">libgen</a></td><td>Generates various types of discriminating elements for each alignment in a directory</td></tr>
<tr><td><a href="../../apps/psiphi.html">psiphi</a></td><td>Calculates phi and psi torsion angles from cleaned EMBOSS-style protein co-ordinate file</td></tr>
<tr><td><a href="rocon.html">rocon</a></td><td>Reads a DHF file (domain hits file) of hits (sequences of unknown structural classification) and a DHF file of validation sequences (known classification) and writes a 'hits file' for the hits, which are classified and rank-ordered on the basis of score</td></tr>
<tr><td><a href="rocplot.html">rocplot</a></td><td>Provides interpretation and graphical display of the performance of discriminating elements (e.g. profiles for protein families). rocplot reads file(s) of hits from discriminator-database search(es), performs ROC analysis on the hits, and writes graphs illustrating the diagnostic performance of the discriminating elements</td></tr>
<tr><td><a href="seqalign.html">seqalign</a></td><td>Reads a DAF file (domain alignment file) and a DHF file (domain hits file) and writes a DAF file extended with the hits</td></tr>
<tr><td><a href="seqfraggle.html">seqfraggle</a></td><td>Removes fragments from DHF files (domain hits files) or other files of sequences</td></tr>
<tr><td><a href="seqsearch.html">seqsearch</a></td><td>Generate database hits (sequences) for nodes in a DCF file (domain classification file) by using PSI-BLAST</td></tr>
<tr><td><a href="seqwords.html">seqwords</a></td><td>Generates DHF files (domain hits files) of database hits (sequences) for nodes in a DCF file (domain classification file) by keyword search of UniProt</td></tr>
<tr><td><a href="siggen.html">siggen</a></td><td>Generates a sparse protein signature from an alignment and residue contact data</td></tr>
<tr><td><a href="sigscan.html">sigscan</a></td><td>Generates a DHF file (domain hits file) of hits (sequences) from scanning a signature against a sequence database</td></tr>
</table>



<!-- DIAGNOSTIC ERROR MESSAGES
     Description of error messages or log file, if one is written.
-->
<a name="12.0"></a>
<br><br><br><H2> 12.0  DIAGNOSTIC ERROR MESSAGES </H2>
None. 






<!-- AUTHORS -->
<a name="13.0"><br><br><br><H2> 13.0  AUTHORS                   </H2></a>
Ranjeeva Ranasinghe <a href="mailto:rranasin@rfcgr.mrc.ac.uk">(rranasin@rfcgr.mrc.ac.uk)</a>

<br><br>
Jon Ison <a href="mailto:jison@rfcgr.mrc.ac.uk">(jison@rfcgr.mrc.ac.uk)</a>
<br>MRC Rosalind Franklin Centre for Genomics Research, Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK




<!-- REFERENCES
     Quote the paper where the application was first published, described, used etc. 
-->
<a name="14.0"></a>
<br><br><br><H2> 14.0  REFERENCES                </H2>

Please cite the authors and EMBOSS.
<br><br>
<i>Rice P, Longden I and Bleasby A (2000) "EMBOSS - The European
Molecular Biology Open Software Suite"  Trends in Genetics,
15:276-278.

<p>
See also <a href="http://emboss.sourceforge.net/">http://emboss.sourceforge.net/</a></i>



<H3>14.1 Other useful references  </H3>

<br>

</BODY>
</HTML>
