<!-- START OF HEADER -->
<HTML><HEAD>
<TITLE> EMBASSY: SEQWORDS documentation. </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

<H2> SEQWORDS documentation
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<!-- END OF HEADER -->




<!-- CONTENTS
     This always includes the sections below.
     Other subsections can be added for individual applications.
-->
<H2>CONTENTS </H2>
<b> <a href="#1.0">1.0     SUMMARY                   </a></b><br>
<b> <a href="#2.0">2.0     INPUTS & OUTPUTS          </a></b><br>
<b> <a href="#3.0">3.0     INPUT FILE FORMAT         </a></b><br>
<b> <a href="#4.0">4.0     OUTPUT FILE FORMAT        </a></b><br>
<b> <a href="#5.0">5.0     DATA FILES                </a></b><br>
<b> <a href="#6.0">6.0     USAGE                     </a></b><br>    
<b> <a href="#7.0">7.0     KNOWN BUGS & WARNINGS     </a></b><br>    
<b> <a href="#8.0">8.0     NOTES                     </a></b><br>
<b> <a href="#9.0">9.0     DESCRIPTION               </a></b><br>
<b> <a href="#10.0">10.0   ALGORITHM                 </a></b><br>
<b> <a href="#11.0">11.0   RELATED APPLICATIONS      </a></b><br>
<b> <a href="#12.0">12.0   DIAGNOSTIC ERROR MESSAGES </a></b><br>
<b> <a href="#13.0">13.0   AUTHORS                   </a></b><br>
<b> <a href="#14.0">14.0   REFERENCES                </a></b><br>


<!-- SUMMARY
     Succint description of the application, particularly its inputs, outputs
     and what it does.  The same text is given at the top of the source (.c)
     file and in the <documentation> attribute of the <application definition>
     of the ACD file.
-->
<a name="1.0"></a>
<br><br><br><H2> 1.0   SUMMARY  </H2>
Generates DHF files (domain hits files) of database hits (sequences) from Swissprot matching keywords from a keywords file.

Generates DHF files from keyword search of UniProt



<!-- INPUTS & OUTPUTS
     Short summary of the application inputs and outputs in its different 
     modes of usage (if appropriate). More detail than the summary.
-->
<a name="2.0"></a>
<br><br><br><H2> 2.0   INPUTS & OUTPUTS          </H2>
SEQWORDS searches a swissprot-format sequence database with keywords taken from a keywords file, and writes a DHF file (domain hits file) of sequences whose swissprot entries contains at least one of the keywords.  If an entry contains a keyword in a domain record of the feature table, then the sequence of the domain is written to the output file, otherwise the entire sequence is written.  The user specifies the name of the swissprot-format sequence database (input), keywords file (input) and DHF file (output).


<!-- INPUT FILE FORMAT 
     Description and example(s) of input file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.
     Use "<b>SEQWORDS</b> reads any normal sequence USAs." if
     appropriate.
-->



<a name="3.0"></a>
<br><br><br><H2> 3.0   INPUT FILE FORMAT         </H2>

The keywords file (below) contains lists of keywords specific to a number of SCOP or CATH nodes, e.g. families and superfamilies.  Each list of keywords is given after a block of SCOP or CATH classification records; for family-specific search terms, the block must contain a CL, FO, SF and an FA record (see below). For superfamily-specific terms, clearly only the CL, FO and SF should be specified. A single keyword must be given per line after the record 'TE'. Each block of SCOP or CATH classification records and search terms must be delimited by the record '//' (the file should also end with this record). 
<br>It is possible to provide search terms above the level of superfamily, for example, fold and class-specific search terms for SCOP by using the CL and FO records only as appropriate.  However, text searches of swissprot for members of scop folds and classes are unlikely to produce specific or meaningful results.


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: seqwords.terms</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
TY   SCOP
XX
CL   Alpha and beta proteins (a/b)
XX
FO   NAD(P)-binding Rossmann-fold domains
XX
SF   NAD(P)-binding Rossmann-fold domains
XX
FA   Lactate &amp; malate dehydrogenases, N-terminal domain
XX
TE   NAD(P)-binding Rossmann-fold
TE   Lactate &amp; malate dehydrogenases
TE   Lactate dehydrogenase
TE   Malate dehydrogenase
//
</pre>
</td></tr></table><p>
<p><h3>File: seqwords.seq</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   ACEA_ECOLI     STANDARD;      PRT;   434 AA.
AC   P05313;
DT   01-NOV-1988 (Rel. 09, Created)
DT   01-NOV-1988 (Rel. 09, Last sequence update)
DT   15-DEC-1998 (Rel. 37, Last annotation update)
DE   ISOCITRATE LYASE (EC 4.1.3.1) (ISOCITRASE) (ISOCITRATASE) (ICL).
GN   ACEA OR ICL.
OS   Escherichia coli.
OC   Bacteria; Proteobacteria; gamma subdivision; Enterobacteriaceae;
OC   Escherichia.
RN   [1]
RP   SEQUENCE FROM N.A.
RC   STRAIN=K12;
RX   MEDLINE; 89083515.
RA   Byrne C.R., Stokes H.W., Ward K.A.;
RT   "Nucleotide sequence of the aceB gene encoding malate synthase A in
RT   Escherichia coli.";
RL   Nucleic Acids Res. 16:10924-10924(1988).
RN   [2]
RP   SEQUENCE FROM N.A.
RC   STRAIN=K12;
RX   MEDLINE; 88262573.
RA   Rieul C., Bleicher F., Duclos B., Cortay J.-C., Cozzone A.J.;
RT   "Nucleotide sequence of the aceA gene coding for isocitrate lyase in
RT   Escherichia coli.";
RL   Nucleic Acids Res. 16:5689-5689(1988).
RN   [3]
RP   SEQUENCE FROM N.A.
RX   MEDLINE; 89008064.
RA   Matsuoka M., McFadden B.A.;
RT   "Isolation, hyperexpression, and sequencing of the aceA gene encoding
RT   isocitrate lyase in Escherichia coli.";
RL   J. Bacteriol. 170:4528-4536(1988).
RN   [4]
RP   SEQUENCE FROM N.A.
RC   STRAIN=K12 / MG1655;
RX   MEDLINE; 94089392.
RA   Blattner F.R., Burland V.D., Plunkett G. III, Sofia H.J.,
RA   Daniels D.L.;
RT   "Analysis of the Escherichia coli genome. IV. DNA sequence of the
RT   region from 89.2 to 92.8 minutes.";
RL   Nucleic Acids Res. 21:5408-5417(1993).
RN   [5]
RP   SEQUENCE OF 293-434 FROM N.A.
RX   MEDLINE; 88227861.
RA   Klumpp D.J., Plank D.W., Bowdin L.J., Stueland C.S., Chung T.,
RA   Laporte D.C.;
RT   "Nucleotide sequence of aceK, the gene encoding isocitrate
RT   dehydrogenase kinase/phosphatase.";
RL   J. Bacteriol. 170:2763-2769(1988).


<font color=red>  [Part of this file has been deleted for brevity]</font>

FT   CONFLICT     70     70       A -&gt; R (IN REF. 2).
FT   CONFLICT     80     80       A -&gt; R (IN REF. 1 AND 2).
FT   CONFLICT    116    116       I -&gt; N (IN REF. 2).
FT   CONFLICT    144    144       F -&gt; L (IN REF. 1).
FT   CONFLICT    305    312       LGEEFVNK -&gt; WAKSSLISN (IN REF. 2).
FT   CONFLICT    307    307       E -&gt; Q (IN REF. 1).
FT   STRAND        2      6
FT   TURN          7      9
FT   HELIX        11     23
FT   TURN         26     27
FT   STRAND       28     33
FT   TURN         37     38
FT   HELIX        39     47
FT   TURN         48     48
FT   STRAND       53     58
FT   HELIX        64     67
FT   TURN         68     69
FT   STRAND       72     75
FT   TURN         83     84
FT   HELIX        87    108
FT   TURN        110    111
FT   STRAND      113    116
FT   HELIX       121    134
FT   TURN        135    136
FT   TURN        140    141
FT   STRAND      143    145
FT   HELIX       148    162
FT   TURN        163    163
FT   HELIX       166    168
FT   STRAND      173    175
FT   TURN        179    181
FT   STRAND      182    184
FT   HELIX       186    188
FT   TURN        190    191
FT   HELIX       196    217
FT   TURN        218    219
FT   HELIX       225    242
FT   TURN        243    244
FT   STRAND      248    255
FT   STRAND      263    271
FT   TURN        272    273
FT   STRAND      274    278
FT   HELIX       286    311
SQ   SEQUENCE   312 AA;  32337 MW;  17741A3B5AD068BA CRC64;
     MKVAVLGAAG GIGQALALLL KTQLPSGSEL SLYDIAPVTP GVAVDLSHIP TAVKIKGFSG
     EDATPALEGA DVVLISAGVA RKPGMDRSDL FNVNAGIVKN LVQQVAKTCP KACIGIITNP
     VNTTVAIAAE VLKKAGVYDK NKLFGVTTLD IIRSNTFVAE LKGKQPGEVE VPVIGGHSGV
     TILPLLSQVP GVSFTEQEVA DLTKRIQNAG TEVVEAKAGG GSATLSMGQA AARFGLSLVR
     ALQGEQGVVE CAYVEGDGQY ARFFSQPLLL GKNGVEERKS IGTLSAFEQN ALEGMLDTLK
     KDIALGEEFV NK
//
</pre>
</td></tr></table><p>







<!-- OUTPUT FILE FORMAT 
     Description and example(s) of output file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.  
     If the standard description of the avalable report formats is required, 
     use:   #include file="inc/reportformats.ihtml"
     Use "Outputs a graph to the specified graphics device."
     or  "outputs a report format file. The default format is ..."
     if appropriate.
-->
<a name="4.0"></a>
<br><br><br><H2> 4.0   OUTPUT FILE FORMAT         </H2>

<b>DHF file (domain hits file)</b>
<br>The format of the DHF file (domain hits file) of hit sequences generated by SEQWORDS (Figure 1) is described fully in <a href="seqsearch.html">SEQSEARCH documentation</a> and only summarised here. The file contains two lines per hit, the first is a description of the hit in 16 text tokens delimited by '^'. The second line contains the protein sequence. The first 4 tokens refer to the hit (sequence) itself, the tokens are as follows:
<UL>
<LI>(i) Accession number
<LI>(ii)  Database code, 
<LI>(iii - iv) Start and end positions of the hit relative to the full length sequence. 
</UL>

<br>The next 9 tokens refer to the domain family, superfamily etc for which the terms were defined (in the keywords file) and are as follows: 
<UL>
<LI>(v)  Type of domain (one of 'SCOP' or 'CATH'), 
<LI>(vi)  SCOP or CATH domain identifier.  
<LI>(vii) SCOP or CATH node unique identifier, e.g. SCOP Family Sunid.
<LI>(viii) Domain class.  Textual description of the 'Class' (SCOP and CATH domains). 
<LI>(ix) Domain architecture. Textual description of the 'Architecture' (CATH only). 
<LI>(x) Domain topology. Textual description of the 'Topology' (CATH only).  
<LI>(xi) Domain fold. Textual description of the 'Fold' (SCOP domains only).  
<LI>(xii) Domain superfamily. Textual description of the 'Superfamily' (SCOP and CATH domains).  
<LI>(xiii) Domain family. Textual description of the 'Fold' (SCOP only). 
</UL>

<br>The next 4 tokens refer to the hit, specifically, information about the search result as follows: 
<UL>
<LI>(xiv) Model type.  The type of model that was used to generate the hit. For DHF files generated by using SEQWORDS a value of KEYWORD is given.  Several other values are possible, however, see <a href="seqsearch.html">SEQSEARCH documentation</a>.
<LI>(xv) SC - Score of hit from search algoritm  (not written by SEQWORDS).
<LI>(xvi) P-value of hit (not written by SEQWORDS).
<LI>(xvii) E-value of hit (not written by SEQWORDS).
</UL>



<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: seqwords.dhf</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt; Q60150^.^1^312^SCOP^.^0^Alpha and beta proteins (a/b)^.^.^NAD(P)-binding Rossmann-fold domains^NAD(P)-binding Rossmann-fold domains^Lactate &amp; malate dehydrogenases, N-terminal domain^KEYWORD^0.00^0.000e+00^0.000e+00
MKVAVLGAAGGIGQALALLLKTQLPSGSELSLYDIAPVTPGVAVDLSHIPTAVKIKGFSGEDATPALEGADVVLISAGVARKPGMDRSDLFNVNAGIVKNLVQQVAKTCPKACIGIITNPVNTTVAIAAEVLKKAGVYDKNKLFGVTTLDIIRSNTFVAELKGKQPGEVEVPVIGGHSGVTILPLLSQVPGVSFTEQEVADLTKRIQNAGTEVVEAKAGGGSATLSMGQAAARFGLSLVRALQGEQGVVECAYVEGDGQYARFFSQPLLLGKNGVEERKSIGTLSAFEQNALEGMLDTLKKDIALGEEFVNK
</pre>
</td></tr></table><p>




<!-- DATA FILES         
     Any data files used (e.g. translation table file, substitution matrix 
     etc.  This includes example data file formats if they are not obvious.
     For a standard description of what data files are and how embossdata can
     be used to inspect and retrieve them, use:
     #include file="inc/localfiles.ihtml" 
-->

<a name="5.0"></a>
<br><br><br><H2> 5.0   DATA FILES                </H2>
SEQWORDS does not use a data file.


<!-- USAGE
     Example usage, as run from the command-line.
     Many examples illustrating different behaviours is good.
-->
<a name="6.0"></a>
<br><br><br><H2> 6.0   USAGE                     </H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" >
<tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-keyfile]           infile     This option specifies the name of keywords
                                  file (input). This contains a list of
                                  keywords specific to a number of SCOP or
                                  CATH families and superfamilies used by
                                  SEQWORDS to search a sequence database.
  [-spfile]            infile     This option specifies the name of the
                                  sequence database (input) to search.
  [-outfile]           outfile    This option specifies the name of the DHF
                                  file (domain hits file) (output). A 'domain
                                  hits file' contains database hits
                                  (sequences) with domain classification
                                  information, in the DHF format (FASTA-like).
                                  The hits are relatives to a SCOP or CATH
                                  family (or other node in the structural
                                  hierarchies) and are found from a search of
                                  a sequence database. Files containing hits
                                  retrieved by PSIBLAST are generated by using
                                  SEQSEARCH, hits retrieved by a sparse
                                  protein signatare by using SIGSCAN or
                                  various types of HMM and profile by using
                                  LIBSCAN.

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-outfile" associated qualifiers
   -odirectory3        string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report deaths

</pre>
</pre>
</td></tr>
</table>


<H3> 6.1   COMMAND LINE ARGUMENTS </H3>
 
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-keyfile]<br>(Parameter 1)</td>
<td>This option specifies the name of keywords file (input). This contains a list of keywords specific to a number of SCOP or CATH families and superfamilies used by SEQWORDS to search a sequence database.</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-spfile]<br>(Parameter 2)</td>
<td>This option specifies the name of the sequence database (input) to search.</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 3)</td>
<td>This option specifies the name of the DHF file (domain hits file) (output). A 'domain hits file' contains database hits (sequences) with domain classification information, in the DHF format (FASTA-like). The hits are relatives to a SCOP or CATH family (or other node in the structural hierarchies) and are found from a search of a sequence database. Files containing hits retrieved by PSIBLAST are generated by using SEQSEARCH, hits retrieved by a sparse protein signatare by using SIGSCAN or various types of HMM and profile by using LIBSCAN.</td>
<td>Output file</td>
<td>test.hits</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<H3> 6.2 EXAMPLE SESSION </H3>
An example of interactive use of SEQWORDS is shown below.

<b>Here is a sample session with seqwords</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>seqwords </b>
Generates DHF files from keyword search of UniProt.
Name of keywords file (input): <b>seqwords.terms</b>
Name of sequence database (input): <b>seqwords.seq</b>
Name of DHF file (domain hits file) (output) [test.hits]: <b>seqwords.dhf</b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>


<!-- <br><br>
The keywords file test_data/seqwords.terms was read and used to search the swissprot database test_data/seqwords.seq.  The DHF file test_data/seqwords/seqwords.dhf containing the results of the search was written.
<br><br>


<!-- KNOWN BUGS & WARNINGS
     Bugs that have not yet been fixed, easily missued features, problems
     and caveats etc. Potentially stupid things the program will let you do.
-->
<a name="7.0"></a>
<br><br><br><H2> 7.0   KNOWN BUGS & WARNINGS     </H2>
SEQWORDS is <b>slow</b> - swissprot is read multiple times (once for each list of terms).  Changing it to do a single file read would require modifying the program to take an array of hitlist and terms structures. An easy-ish change.








<!-- NOTES
     Important general remarks, including:
       Restrictions.
       Interesting behaviour.
       Useful things you can do with this program.
       Future plans.
       etc.
-->
<a name="8.0"></a>
<br><br><br><H2> 8.0   NOTES                     </H2>



<H3> 8.1   GLOSSARY OF FILE TYPES    </H3>
<table BORDER CELLSPACING=5 CELLPADDING=5 BGCOLOR="#f5f5ff" >
<tr>
<td><b>FILE TYPE</b></td>
<td><b>FORMAT</b></td>
<td><b>DESCRIPTION</b></td>
<td><b>CREATED BY <b></td>
<td><b>SEE ALSO</b></td>
</tr>
<tr>
<td><b>Domain hits file </b></td>
<td> DHF format (FASTA-like). </td>
<td> Database hits (sequences) with domain classification information.  The hits are relatives to a SCOP or CATH family (or other node in the structural hierarchies) and are found from a search of a discriminating element (e.g. a protein signature, hidden Markov model, simple frequency matrix, Gribskov profile or Hennikoff profile) against a sequence database. </td>
<td> <a href="seqsearch.html">SEQSEARCH</a> (hits retrieved by PSIBLAST). <a href="sigscan.html">SIGSCAN</a> (hits retrieved by sparse protein signature). <a href="libscan.html">LIBSCAN</a> (hits retrieved by various types of HMM and profile). </td>
<td> N.A. </td>
</tr>
<tr>
<td><b>Keywords file </b></td>
<td> Text</td>
<td> Contains a list of keywords specific to a number of SCOP families and superfamilies used by SEQWORDS to search a sequence database. </td>
<td> N.A. </td>
<td> N.A. </td>
</tr>
</table>



None


<!-- DESCRIPTION
     A complete, non-technical, user-level description of the application.
-->
<a name="9.0"></a>
<br><br><br><H2> 9.0   DESCRIPTION               </H2>
Retrieval of domain sequence information from swissprot by mutliple keywords can be a time-consuming process.  SEQWORDS parses a file of keywords and the swissprot database and writes a file of sequences whose swissprot entries contains at least one of the keywords.  The domain sequence is taken if the keyword appears in the domain record of the feature table, otherwise the full-length sequence is taken.


<!-- ALGORITHM
     A technical description of algorithmic aspects, describing exactly how
     the key aspects of the application work.
-->
<a name="10.0"></a>
<br><br><br><H2> 10.0  ALGORITHM                 </H2>
None.



<!-- RELATED APPLICATIONS
     Other applications that either generate the input, use the output or 
     are in some other way related to the application are described here. 
-->

<a name="11.0"></a>
<br><br><br><H2> 11.0  RELATED APPLICATIONS      </H2>

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="../../apps/contactcount.html">contactcount</a></td><td>Count specific versus non-specific contacts</td></tr>
<tr><td><a href="../structure/contacts.html">contacts</a></td><td>Generate intra-chain CON files from CCF files</td></tr>
<tr><td><a href="../domalign/domainalign.html">domainalign</a></td><td>Generate alignments (DAF file) for nodes in a DCF file</td></tr>
<tr><td><a href="../domalign/domainrep.html">domainrep</a></td><td>Reorder DCF file to identify representative structures</td></tr>
<tr><td><a href="../domainatrix/domainreso.html">domainreso</a></td><td>Remove low resolution domains from a DCF file</td></tr>
<tr><td><a href="../structure/interface.html">interface</a></td><td>Generate inter-chain CON files from CCF files</td></tr>
<tr><td><a href="../signature/libgen.html">libgen</a></td><td>Generate discriminating elements from alignments</td></tr>
<tr><td><a href="../signature/matgen3d.html">matgen3d</a></td><td>Generate a 3D-1D scoring matrix from CCF files</td></tr>
<tr><td><a href="../../apps/psiphi.html">psiphi</a></td><td>Phi and psi torsion angles from protein coordinates</td></tr>
<tr><td><a href="../signature/rocon.html">rocon</a></td><td>Generates a hits file from comparing two DHF files</td></tr>
<tr><td><a href="../signature/rocplot.html">rocplot</a></td><td>Performs ROC analysis on hits files</td></tr>
<tr><td><a href="../reconstruct/scorecmapdir.html">scorecmapdir</a></td><td>Contact scores for cleaned protein chain contact files</td></tr>
<tr><td><a href="../domalign/seqalign.html">seqalign</a></td><td>Extend alignments (DAF file) with sequences (DHF file)</td></tr>
<tr><td><a href="seqfraggle.html">seqfraggle</a></td><td>Removes fragment sequences from DHF files</td></tr>
<tr><td><a href="seqsearch.html">seqsearch</a></td><td>Generate PSI-BLAST hits (DHF file) from a DAF file</td></tr>
<tr><td><a href="seqsort.html">seqsort</a></td><td>Remove ambiguous classified sequences from DHF files</td></tr>
<tr><td><a href="../signature/siggen.html">siggen</a></td><td>Generates a sparse protein signature from an alignment</td></tr>
<tr><td><a href="../signature/siggenlig.html">siggenlig</a></td><td>Generate ligand-binding signatures from a CON file</td></tr>
<tr><td><a href="../signature/sigscan.html">sigscan</a></td><td>Generate hits (DHF file) from a signature search</td></tr>
<tr><td><a href="../signature/sigscanlig.html">sigscanlig</a></td><td>Search ligand-signature library & write hits (LHF file)</td></tr>
</table>




<!-- DIAGNOSTIC ERROR MESSAGES
     Description of error messages or log file, if one is written.
-->
<a name="12.0"></a>
<br><br><br><H2> 12.0  DIAGNOSTIC ERROR MESSAGES </H2>
None.



<!-- AUTHORS
     The main author first then all others. 
-->
<a name="13.0"></a>
<br><br><br><H2> 13.0  AUTHORS                   </H2>
Jon Ison <a href="mailto:jison@rfcgr.mrc.ac.uk">(jison@rfcgr.mrc.ac.uk)</a>
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK




<!-- REFERENCES
     Quote the paper where the application was first published, described, used etc. 
-->
<a name="14.0"></a>
<br><br><br><H2> 14.0  REFERENCES                </H2>

Please cite the authors and EMBOSS.
<br><br>
<i>Rice P, Longden I and Bleasby A (2000) "EMBOSS - The European
Molecular Biology Open Software Suite"  Trends in Genetics,
15:276-278.

<p>
See also <a href="http://emboss.sourceforge.net/">http://emboss.sourceforge.net/</a></i>




<H3>14.1 Other useful references  </H3>


</BODY>
</HTML>



