<HTML>

<HEAD>
  <TITLE>
  EMBOSS: showseq
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: showseq
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program showseq
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Display a sequence with features, translation etc


<H2>
    Description
</H2>

<b>showseq</b> displays a protein or a nucleic acid sequence in a style
suitable for publication.
<p>

The output is sent to the screen by default for the user to view, but it
can write the results to a file.
<p>

The display style of the output is very flexible.  The user can select a
style from the pre-set choice of formats, or can design a style to suit
their purposes and aesthetic tastes.
<p>

The displayed sequence can be numbered either by numbering the start and
ending positions, or by placing a ruler with ticks above or below the
sequence.
<p>

The width of a line can be set.  The width of a margin around the
sequence reserved for numbering can be set.  The initial position to
start numbering from can be set.
<p>

The sequence can be translated, using the selectable genetic code
tables.  The translation can be done in one, three or six frames.  The
translation can be displayed in one-letter or three-letter amino acid
codes.  The translation can optionally be displayed only when it is in
open reading frames (ORFs) of a specified minimum size.  One or more
specified regions of the sequence can be individually translated.
<p>

Specified regions of the sequence can be displayed in uppercase to
highlight them.
<p>

The output can be formatted for HTML.
<p>

If the output is being formatted for HTML, then specified regions of the
sequence can be displayed in any valid HTML colours.
<p>

The Restriction Enzyme database (REBASE) is a collection of information
about restriction enzymes and related proteins.  It contains published
and unpublished references, recognition and cleavage sites,
isoschizomers, commercial availability, methylation sensitivity, crystal
and sequence data.  DNA methyltransferases, homing endonucleases,
nicking enzymes, specificity subunits and control proteins are also
included.  Most recently, putative DNA methyltransferases and
restriction enzymes, as predicted from analysis of genomic sequences,
are also listed.
<p>

The home page of REBASE is: 
<A HREF="http://rebase.neb.com/rebase/rebase.htm">
http://rebase.neb.com/rebase/rebase.html</A>
<p>

This program can use REBASE data to find the recognition sites and/or
cut sites of restriction enzymes in a nucleic acid sequence.
<p>

This program can display the cut sites on both strands.
<p>

One potentially very useful option is '-flatreformat' that displays not
only the cut sites which many other restriction cut-site programs will
show, but also shows the recognition site.
<p>

If the sequence is in EMBL, Genbank or SwissProt format, the feature table
of the sequence can be dispalyed with the sequence. GFF file features can also
be displayed if they are included on the command line with the <b>-ufo=file</b>
qualifier.
<p>


<H2>
    Usage
</H2>


Here is a sample session with <b>showseq</b>. By default the output appears
on standard output (the terminal) but can be saved to a file. We only look at a
small section of the sequence to save space.
<p>

<pre>
% showseq embl:eclac -sbeg 1 -send 100
Display a sequence with features, translation etc..
Output file [stdout]:
Things to display
         0 : Enter your own list of things to display
         1 : Sequence only
         2 : Default sequence with features
         3 : Pretty sequence
         4 : One frame translation
         5 : Three frame translations
         6 : Six frame translations
         7 : Restriction enzyme map
         8 : Baroque

Display format [2]:
ECLAC
E.coli lactose operon with lacI, lacZ, lacY and lacA genes.

                   10        20        30        40        50        60        
          ----:----|----:----|----:----|----:----|----:----|----:----|
          gacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagt
                         |
          variation note="c in wild-type; t in 'up' promoter mutant I-Q [11]"
                                                            |=========
          mRNA note="lacI (repressor) mRNA; preferred in vivo 3' end [12],[29]"

                   70        80        90        100       110       120       
          ----:----|----:----|----:----|----:----|----:----|----:----|
          caattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggt
          ============================================================
          mRNA note="lacI (repressor) mRNA; preferred in vivo 3' end [12],[29]"
                            |=========================================
          CDS codon_start="1" db_xref="SWISS-PROT:P03023" note="lac repressor p

</pre>
<p>    
        
Note that although we asked for the sequence display to end at position
'100', it has displayed the sequence up to the end of the line -
position '120'.  This is a feature of this program to make the display
of things like restriction enzyme cutting sites easier.
<p>
          
The standard list of output formats are only a small selection of the 
possible ways in which a sequence might be displayed.  Precise control
over the output format is acheived by selecting the qualifier '-format
0' (Option 0 in the list of things to display).  For example:
<p>

<pre>
% showseq embl:eclac -sbeg 1 -send 120
Display a sequence with features, translation etc..
Output file [stdout]:
Things to display
         0 : Enter your own list of things to display
         1 : Sequence only
         2 : Default sequence with features
         3 : Pretty sequence
         4 : One frame translation
         5 : Three frame translations
         6 : Six frame translations
         7 : Restriction enzyme map
         8 : Baroque
Display format [2]: 0
Specify your own things to display
         S : Sequence
         B : Blank line
         1 : Frame1 translation
         2 : Frame2 translation
         3 : Frame3 translation
        -1 : CompFrame1 translation
        -2 : CompFrame2 translation
        -3 : CompFrame3 translation
         T : Ticks line
         N : Number ticks line
         C : Complement sequence
         F : Features
         R : Restriction enzyme cut sites in forward sense
        -R : Restriction enzyme cut sites in reverse sense
         A : Annotation
Enter a list of things to display [B N T S A F]: b,s,t,c
ECLAC
E.coli lactose operon with lacI, lacZ, lacY and lacA genes.
          
          gacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagt
          ----:----|----:----|----:----|----:----|----:----|----:----|
          ctgtggtagcttaccgcgttttggaaagcgccataccgtactatcgcgggccttctctca

          caattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggt
          ----:----|----:----|----:----|----:----|----:----|----:----|
          gttaagtcccaccacttacactttggtcattgcaatatgctacagcgtctcatacggcca
</pre>
<p>

By choosing format '0' and then specifying that we want to display the
things: 'b,s,t,c', we will output the sequence in the following way:
<p>

For every new line that the sequence starts to write, the output display
will contain first a blank line ('b'), then the sequence itself ('s')
then a line of with ticks every 10 characters ('t') then the reverse
complement of the sequence ('c').  Subsequent lines of the sequence
output will repeat this format.
<p>
         
The 'thing' codes used in the list of standard formats are:
<p>
<pre>
Sequence only:                  S A
Default sequence:               B N T S A F
Pretty sequence:                B N T S A
One frame translation:          B N T S B 1 A F
Three frame translations:       B N T S B 1 2 3 A F
Six frame translations:         B N T S B 1 2 3 T -3 -2 -1 A F
Restriction enzyme map:         B R S N T C -R B 1 2 3 T -3 -2 -1 A
Baroque:                        B 1 2 3 N T R S T C -R T -3 -2 -1 A F
</pre>          
<p>          
        
The following are some examples of different formats:
<p>
<pre>
Just sequence:

% showseq embl:eclac stdout -sbeg 1 -send 120 -noname -nodesc -format 0 -thing S
Display a sequence with features, translation etc..
          gacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagt
          caattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggt
</pre>
<p>

Protein sequence displayed in three-letter codes.  (The codes are
displayed downwards, so the first code is 'Met'):
<p>
<pre>
% showseq sw:rs24_fugru stdout -three -format 2
RS24_FUGRU
40S RIBOSOMAL PROTEIN S24.   
         
                   10        20        30        40        50        60
          ----:----|----:----|----:----|----:----|----:----|----:----|
          MAATVTVATALPMTAALLGALGMVVAVLHPGLATVPLTGIAGLLALMTLTTPAVVPVPGP
          esshaharhryhehsreelryleaasaeirlylharyhllrlyelyeyyhhrsaahahlh
          tnprlrlgrgsetrnguungsntllplusoysarlosruegusuastrsrroplleleye

                   70        80        90        100       110       120
          ----:----|----:----|----:----|----:----|----:----|----:----|
          ATGPGGGLTTGPAMVTASLATALLAGPLHALAAHGLPGLLLTSALGALGALAAMLLVAGT
          rhlhlllyhhlhleayseesylyyslryirelrilehlyyyherylrylrysreyyarlh
          grneyyysrryeatlrpruprassnuossguagsyueusssrrgsngsugsngtsslgyr
        
                   130       140       150       160       170       180
          ----:----|----:----|----:----|----:----|----:----|----:----|
          LLASVGASLLLA 
          yylealleyyys
          ssarlyarsssp
</pre>
<p>
Number the sequence lines in the margin:
<p>
<pre>
% showseq embl:mmam stdout -format 1 -number
Display a sequence with features, translation etc..
Output file [stdout]:
MMAM      
Mus musculus (cell line C3H/F2-11) chromosome 12 anti-DNA antibody
heavy chain mRNA.
        1 gagnnccagctgcagcagtctggacctgagctggtaaagcctggggcttcagtgaagatg 60
       61 tcctgcaaggcttctggatacacattcactagctatgttatgcactgggtgaatcagaag 120
      121 cctgggcagggccttgagtggattggatatattaatccttacaatgatggtactaactac 180
      181 aatgagaagttcaaaggcaaggccacactgacttcagacaaatcctccagcacagcctac 240
      241 atggagttcagcagcctgacctctgaggactctgcggtctattactgtgcaagaaaaact 300
      301 tcctactatagtaacctatattactttgactactggggccaaggcaccactctcacagtc 360
      361 tcctca                                                       366
</pre>
<p>

Start the numbering at a specified value ('123' in this case):
<p>
<pre>
% showseq embl:mmam stdout -format 1 -number -offset 123
Display a sequence with features, translation etc..
MMAM
Mus musculus (cell line C3H/F2-11) chromosome 12 anti-DNA antibody 
heavy chain mRNA.
      123 gagnnccagctgcagcagtctggacctgagctggtaaagcctggggcttcagtgaagatg 182
      183 tcctgcaaggcttctggatacacattcactagctatgttatgcactgggtgaatcagaag 242
      243 cctgggcagggccttgagtggattggatatattaatccttacaatgatggtactaactac 302
      303 aatgagaagttcaaaggcaaggccacactgacttcagacaaatcctccagcacagcctac 362
      363 atggagttcagcagcctgacctctgaggactctgcggtctattactgtgcaagaaaaact 422
      423 tcctactatagtaacctatattactttgactactggggccaaggcaccactctcacagtc 482
      483 tcctca                                                       488
</pre>
<p>

Make selected regions uppercase.  (Use '-slower' to force the rest of
the sequence to be lowercase).
<p>
<pre>          
% showseq embl:mmam stdout -format 1 -slower -upper '25-45,101-203,333-362'
Display a sequence with features, translation etc..  
MMAM
Mus musculus (cell line C3H/F2-11) chromosome 12 anti-DNA antibody
heavy chain mRNA.
          gagnnccagctgcagcagtctggaCCTGAGCTGGTAAAGCCTGGGgcttcagtgaagatg
          tcctgcaaggcttctggatacacattcactagctatgttaTGCACTGGGTGAATCAGAAG
          CCTGGGCAGGGCCTTGAGTGGATTGGATATATTAATCCTTACAATGATGGTACTAACTAC
          AATGAGAAGTTCAAAGGCAAGGCcacactgacttcagacaaatcctccagcacagcctac
          atggagttcagcagcctgacctctgaggactctgcggtctattactgtgcaagaaaaact
          tcctactatagtaacctatattactttgactaCTGGGGCCAAGGCACCACTCTCACAGTC
          TCctca
</pre>
<p>
Translate selected regions:
<p>
<pre>
% showseq embl:mmam stdout -format 4 -send 120 -trans 25-49,66-76
Display a sequence with features, translation etc..
MMAM
Mus musculus (cell line C3H/F2-11) chromosome 12 anti-DNA antibody
heavy chain mRNA.
          
                   10        20        30        40        50        60
          ----:----|----:----|----:----|----:----|----:----|----:----|
          gagnnccagctgcagcagtctggacctgagctggtaaagcctggggcttcagtgaagatg

                                  P  E  L  V  K  P  G  A  S
          
                   70        80        90        100       110       120
          ----:----|----:----|----:----|----:----|----:----|----:----|
          tcctgcaaggcttctggatacacattcactagctatgttatgcactgggtgaatcagaag
        
                 R  L  L
</pre>
<p>          
          
Add your own annotation to the display
<p>
<pre>
% showseq embl:mmam stdout -format 2 -send 120 -annotation '13-26 binding site 15-15 SNP'
Display a sequence with features, translation etc..
MMAM
Mus musculus (cell line C3H/F2-11) chromosome 12 anti-DNA antibody
heavy chain mRNA.

                   10        20        30        40        50        60        
          ----:----|----:----|----:----|----:----|----:----|----:----|
          gagnnccagctgcagcagtctggacctgagctggtaaagcctggggcttcagtgaagatg
                      |------------|
                      binding site
                        |
                        SNP

                   70        80        90        100       110       120       
          ----:----|----:----|----:----|----:----|----:----|----:----|
          tcctgcaaggcttctggatacacattcactagctatgttatgcactgggtgaatcagaag


</pre>
<p>  


<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers (* if not always prompted):
  [-sequence]          seqall     Sequence database USA
   -format             menu       Display format
*  -things             menu       Specify a list of one or more code
                                  characters in the order in which you wish
                                  things to be displayed one above the other
                                  down the page. For example if you wish to
                                  see things displayed in the order: sequence,
                                  complement sequence, ticks line, frame 1
                                  translation, blank line; then you should
                                  enter 'S,C,T,1,B'.
  [-outfile]           outfile    If you enter the name of a file here then
                                  this program will write the sequence details
                                  into that file.

   Optional qualifiers:
   -translate          range      Regions to translate (if translating).
                                  If this is left blank the complete sequence
                                  is translated.
                                  A set of regions is specified by a set of
                                  pairs of positions.
                                  The positions are integers.
                                  They are separated by any non-digit,
                                  non-alpha character.
                                  Examples of region specifications are:
                                  24-45, 56-78
                                  1:45, 67=99;765..888
                                  1,5,8,10,23,45,57,99
   -uppercase          range      Regions to put in uppercase.
                                  If this is left blank, then the sequence
                                  case is left alone.
                                  A set of regions is specified by a set of
                                  pairs of positions.
                                  The positions are integers.
                                  They are separated by any non-digit,
                                  non-alpha character.
                                  Examples of region specifications are:
                                  24-45, 56-78
                                  1:45, 67=99;765..888
                                  1,5,8,10,23,45,57,99
   -highlight          range      Regions to colour if formatting for HTML.
                                  If this is left blank, then the sequence is
                                  left alone.
                                  A set of regions is specified by a set of
                                  pairs of positions.
                                  The positions are integers.
                                  They are followed by any valid HTML font
                                  colour.
                                  Examples of region specifications are:
                                  24-45 blue 56-78 orange
                                  1-100 green 120-156 red
                                  A file of ranges to colour (one range per
                                  line) can be specified as '@filename'.
   -annotation         range      Regions to annotate by marking.
                                  If this is left blank, then no annotation is
                                  added.
                                  A set of regions is specified by a set of
                                  pairs of positions followed by optional
                                  text.
                                  The positions are integers.
                                  They are followed by any text (but not
                                  digits when on the command-line).
                                  Examples of region specifications are:
                                  24-45 new domain 56-78 match to Mouse
                                  1-100 First part 120-156 oligo
                                  A file of ranges to annotate (one range per
                                  line) can be specified as '@filename'.
   -enzymes            string     The name 'all' reads in all enzyme names
                                  from the REBASE database. You can specify
                                  enzymes by giving their names with commas
                                  between then, such as:
                                  'HincII,hinfI,ppiI,hindiii'.
                                  The case of the names is not important. You
                                  can specify a file of enzyme names to read
                                  in by giving the name of the file holding
                                  the enzyme names with a '@' character in
                                  front of it, for example, '@enz.list'.
                                  Blank lines and lines starting with a hash
                                  character or '!' are ignored and all other
                                  lines are concatenated together with a comma
                                  character ',' and then treated as the list
                                  of enzymes to search for.
                                  An example of a file of enzyme names is:
                                  ! my enzymes
                                  HincII, ppiII
                                  ! other enzymes
                                  hindiii
                                  HinfI
                                  PpiI
   -table              menu       Code to use
   -matchsource        string     By default any feature source in the feature
                                  table is shown. You can set this to match
                                  any feature source you wish to show.
                                  The source name is usually either the name
                                  of the program that detected the feature or
                                  it is the feature table (eg: EMBL) that the
                                  feature came from.
                                  The source may be wildcarded by using '*'.
                                  If you wish to show more than one source,
                                  separate their names with the character '|',
                                  eg:
                                  gene* | embl
   -matchtype          string     By default any feature type in the feature
                                  table is shown. You can set this to match
                                  any feature type you wish to show.
                                  See http://www3.ebi.ac.uk/Services/WebFeat/
                                  for a list of the EMBL feature types and see
                                  Appendix A of the Swissprot user manual in
                                  http://www.expasy.ch/txt/userman.txt for a
                                  list of the Swissprot feature types.
                                  The type may be wildcarded by using '*'.
                                  If you wish to show more than one type,
                                  separate their names with the character '|',
                                  eg:
                                  *UTR | intron
   -matchsense         integer    By default any feature type in the feature
                                  table is shown. You can set this to match
                                  any feature sense you wish to show. 0 - any
                                  sense, 1 - forward sense, -1 - reverse sense
   -minscore           float      If this is greater than or equal to the
                                  maximum score, then any score is permitted
   -maxscore           float      If this is less than or equal to the maximum
                                  score, then any score is permitted
   -matchtag           string     Tags are the types of extra values that a
                                  feature may have. For example in the EMBL
                                  feature table, a 'CDS' type of feature may
                                  have the tags '/codon', '/codon_start',
                                  '/db_xref', '/EC_number', '/evidence',
                                  '/exception', '/function', '/gene',
                                  '/label', '/map', '/note', '/number',
                                  '/partial', '/product', '/protein_id',
                                  '/pseudo', '/standard_name', '/translation',
                                  '/transl_except', '/transl_table', or
                                  '/usedin'. Some of these tags also have
                                  values, for example '/gene' can have the
                                  value of the gene name.
                                  By default any feature tag in the feature
                                  table is shown. You can set this to match
                                  any feature tag you wish to show.
                                  The tag may be wildcarded by using '*'.
                                  If you wish to show more than one tag,
                                  separate their names with the character '|',
                                  eg:
                                  gene | label
   -matchvalue         string     Tag values are the values associated with a
                                  feature tag. Tags are the types of extra
                                  values that a feature may have. For example
                                  in the EMBL feature table, a 'CDS' type of
                                  feature may have the tags '/codon',
                                  '/codon_start', '/db_xref', '/EC_number',
                                  '/evidence', '/exception', '/function',
                                  '/gene', '/label', '/map', '/note',
                                  '/number', '/partial', '/product',
                                  '/protein_id', '/pseudo', '/standard_name',
                                  '/translation', '/transl_except',
                                  '/transl_table', or '/usedin'. Only some of
                                  these tags can have values, for example
                                  '/gene' can have the value of the gene name.
                                  By default any feature tag value in the
                                  feature table is shown. You can set this to
                                  match any feature tag valueyou wish to show.
                                  The tag value may be wildcarded by using
                                  '*'.
                                  If you wish to show more than one tag value,
                                  separate their names with the character
                                  '|', eg:
                                  pax* | 10

   Advanced qualifiers:
   -orfminsize         integer    Minimum size of Open Reading Frames (ORFs)
                                  to display in the translations.
   -flatreformat       bool       Display RE sites in flat format
   -mincuts            integer    Minimum cuts per RE
   -maxcuts            integer    Maximum cuts per RE
   -sitelen            integer    Minimum recognition site length
   -single             bool       Force single RE site only cuts
   -[no]blunt          bool       Allow blunt end RE cutters
   -[no]sticky         bool       Allow sticky end RE cutters
   -[no]ambiguity      bool       Allow ambiguous RE matches
   -plasmid            bool       Allow circular DNA
   -[no]commercial     bool       Only use restriction enzymes with suppliers
   -[no]limit          bool       Limits RE hits to one isoschizomer
   -preferred          bool       Report preferred isoschizomers
   -threeletter        bool       Display protein sequences in three-letter
                                  code
   -number             bool       Number the sequences
   -width              integer    Width of sequence to display
   -length             integer    Line length of page (0 for indefinite)
   -margin             integer    Margin around sequence for numbering
   -[no]name           bool       Set this to be false if you do not wish to
                                  display the ID name of the sequence
   -[no]description    bool       Set this to be false if you do not wish to
                                  display the description of the sequence
   -offset             integer    Offset to start numbering the sequence from
   -html               bool       Use HTML formatting

   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-format</td>
<td>Display format</td>
<td><table><tr><td>0</td> <td><i>(Enter your own list of things to display)</i></td></tr><tr><td>1</td> <td><i>(Sequence only)</i></td></tr><tr><td>2</td> <td><i>(Default sequence with features)</i></td></tr><tr><td>3</td> <td><i>(Pretty sequence)</i></td></tr><tr><td>4</td> <td><i>(One frame translation)</i></td></tr><tr><td>5</td> <td><i>(Three frame translations)</i></td></tr><tr><td>6</td> <td><i>(Six frame translations)</i></td></tr><tr><td>7</td> <td><i>(Restriction enzyme map)</i></td></tr><tr><td>8</td> <td><i>(Baroque)</i></td></tr></table></td>
<td>2</td>
</tr>

<tr>
<td>-things</td>
<td>Specify a list of one or more code characters in the order in which you wish things to be displayed one above the other down the page. For example if you wish to see things displayed in the order: sequence, complement sequence, ticks line, frame 1 translation, blank line; then you should enter 'S,C,T,1,B'.</td>
<td><table><tr><td>S</td> <td><i>(Sequence)</i></td></tr><tr><td>B</td> <td><i>(Blank line)</i></td></tr><tr><td>1</td> <td><i>(Frame1 translation)</i></td></tr><tr><td>2</td> <td><i>(Frame2 translation)</i></td></tr><tr><td>3</td> <td><i>(Frame3 translation)</i></td></tr><tr><td>-1</td> <td><i>(CompFrame1 translation)</i></td></tr><tr><td>-2</td> <td><i>(CompFrame2 translation)</i></td></tr><tr><td>-3</td> <td><i>(CompFrame3 translation)</i></td></tr><tr><td>T</td> <td><i>(Ticks line)</i></td></tr><tr><td>N</td> <td><i>(Number ticks line)</i></td></tr><tr><td>C</td> <td><i>(Complement sequence)</i></td></tr><tr><td>F</td> <td><i>(Features)</i></td></tr><tr><td>R</td> <td><i>(Restriction enzyme cut sites in forward sense)</i></td></tr><tr><td>-R</td> <td><i>(Restriction enzyme cut sites in reverse sense)</i></td></tr><tr><td>A</td> <td><i>(Annotation)</i></td></tr></table></td>
<td>B,N,T,S,A,F</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>If you enter the name of a file here then this program will write the sequence details into that file.</td>
<td>Output file</td>
<td>stdout</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-translate</td>
<td>Regions to translate (if translating).
If this is left blank the complete sequence is translated.
A set of regions is specified by a set of pairs of positions.
The positions are integers.
They are separated by any non-digit, non-alpha character.
Examples of region specifications are:
24-45, 56-78
1:45, 67=99;765..888
1,5,8,10,23,45,57,99</td>
<td>Sequence range</td>
<td>If this is left blank the complete sequence is translated.</td>
</tr>

<tr>
<td>-uppercase</td>
<td>Regions to put in uppercase.
If this is left blank, then the sequence case is left alone.
A set of regions is specified by a set of pairs of positions.
The positions are integers.
They are separated by any non-digit, non-alpha character.
Examples of region specifications are:
24-45, 56-78
1:45, 67=99;765..888
1,5,8,10,23,45,57,99</td>
<td>Sequence range</td>
<td>If this is left blank, then the sequence case is left alone.</td>
</tr>

<tr>
<td>-highlight</td>
<td>Regions to colour if formatting for HTML.
If this is left blank, then the sequence is left alone.
A set of regions is specified by a set of pairs of positions.
The positions are integers.
They are followed by any valid HTML font colour.
Examples of region specifications are:
24-45 blue 56-78 orange
1-100 green 120-156 red
A file of ranges to colour (one range per line) can be specified as '@filename'.</td>
<td>Sequence range</td>
<td><i>full sequence</i></td>
</tr>

<tr>
<td>-annotation</td>
<td>Regions to annotate by marking.
If this is left blank, then no annotation is added.
A set of regions is specified by a set of pairs of positions followed by optional text.
The positions are integers.
They are followed by any text (but not digits when on the command-line).
Examples of region specifications are:
24-45 new domain 56-78 match to Mouse
1-100 First part 120-156 oligo
A file of ranges to annotate (one range per line) can be specified as '@filename'.</td>
<td>Sequence range</td>
<td>If this is left blank, then no annotation is added.</td>
</tr>

<tr>
<td>-enzymes</td>
<td>The name 'all' reads in all enzyme names from the REBASE database. You can specify enzymes by giving their names with commas between then, such as: 'HincII,hinfI,ppiI,hindiii'.
The case of the names is not important. You can specify a file of enzyme names to read in by giving the name of the file holding the enzyme names with a '@' character in front of it, for example, '@enz.list'.
Blank lines and lines starting with a hash character or '!' are ignored and all other lines are concatenated together with a comma character ',' and then treated as the list of enzymes to search for.
An example of a file of enzyme names is:
! my enzymes
HincII, ppiII
! other enzymes
hindiii
HinfI
PpiI</td>
<td>Any string is accepted</td>
<td>all</td>
</tr>

<tr>
<td>-table</td>
<td>Code to use</td>
<td><table><tr><td>0</td> <td><i>(Standard)</i></td></tr><tr><td>1</td> <td><i>(Standard (with alternative initiation codons))</i></td></tr><tr><td>2</td> <td><i>(Vertebrate Mitochondrial)</i></td></tr><tr><td>3</td> <td><i>(Yeast Mitochondrial)</i></td></tr><tr><td>4</td> <td><i>(Mold, Protozoan, Coelenterate Mitochondrial and Mycoplasma/Spiroplasma)</i></td></tr><tr><td>5</td> <td><i>(Invertebrate Mitochondrial)</i></td></tr><tr><td>6</td> <td><i>(Ciliate Macronuclear and Dasycladacean)</i></td></tr><tr><td>9</td> <td><i>(Echinoderm Mitochondrial)</i></td></tr><tr><td>10</td> <td><i>(Euplotid Nuclear)</i></td></tr><tr><td>11</td> <td><i>(Bacterial)</i></td></tr><tr><td>12</td> <td><i>(Alternative Yeast Nuclear)</i></td></tr><tr><td>13</td> <td><i>(Ascidian Mitochondrial)</i></td></tr><tr><td>14</td> <td><i>(Flatworm Mitochondrial)</i></td></tr><tr><td>15</td> <td><i>(Blepharisma Macronuclear)</i></td></tr><tr><td>16</td> <td><i>(Chlorophycean Mitochondrial)</i></td></tr><tr><td>21</td> <td><i>(Trematode Mitochondrial)</i></td></tr><tr><td>22</td> <td><i>(Scenedesmus obliquus)</i></td></tr><tr><td>23</td> <td><i>(Thraustochytrium Mitochondrial)</i></td></tr></table></td>
<td>0</td>
</tr>

<tr>
<td>-matchsource</td>
<td>By default any feature source in the feature table is shown. You can set this to match any feature source you wish to show.
The source name is usually either the name of the program that detected the feature or it is the feature table (eg: EMBL) that the feature came from.
The source may be wildcarded by using '*'.
If you wish to show more than one source, separate their names with the character '|', eg:
gene* | embl</td>
<td>Any string is accepted</td>
<td>*</td>
</tr>

<tr>
<td>-matchtype</td>
<td>By default any feature type in the feature table is shown. You can set this to match any feature type you wish to show.
See http://www3.ebi.ac.uk/Services/WebFeat/ for a list of the EMBL feature types and see Appendix A of the Swissprot user manual in http://www.expasy.ch/txt/userman.txt for a list of the Swissprot feature types.
The type may be wildcarded by using '*'.
If you wish to show more than one type, separate their names with the character '|', eg:
*UTR | intron</td>
<td>Any string is accepted</td>
<td>*</td>
</tr>

<tr>
<td>-matchsense</td>
<td>By default any feature type in the feature table is shown. You can set this to match any feature sense you wish to show. 0 - any sense, 1 - forward sense, -1 - reverse sense</td>
<td>Any integer value</td>
<td>0 - any sense, 1 - forward sense, -1 - reverse sense</td>
</tr>

<tr>
<td>-minscore</td>
<td>If this is greater than or equal to the maximum score, then any score is permitted</td>
<td>Any integer value</td>
<td>0.0</td>
</tr>

<tr>
<td>-maxscore</td>
<td>If this is less than or equal to the maximum score, then any score is permitted</td>
<td>Any integer value</td>
<td>0.0</td>
</tr>

<tr>
<td>-matchtag</td>
<td>Tags are the types of extra values that a feature may have. For example in the EMBL feature table, a 'CDS' type of feature may have the tags '/codon', '/codon_start', '/db_xref', '/EC_number', '/evidence', '/exception', '/function', '/gene', '/label', '/map', '/note', '/number', '/partial', '/product', '/protein_id', '/pseudo', '/standard_name', '/translation', '/transl_except', '/transl_table', or '/usedin'. Some of these tags also have values, for example '/gene' can have the value of the gene name.
By default any feature tag in the feature table is shown. You can set this to match any feature tag you wish to show.
The tag may be wildcarded by using '*'.
If you wish to show more than one tag, separate their names with the character '|', eg:
gene | label</td>
<td>Any string is accepted</td>
<td>*</td>
</tr>

<tr>
<td>-matchvalue</td>
<td>Tag values are the values associated with a feature tag. Tags are the types of extra values that a feature may have. For example in the EMBL feature table, a 'CDS' type of feature may have the tags '/codon', '/codon_start', '/db_xref', '/EC_number', '/evidence', '/exception', '/function', '/gene', '/label', '/map', '/note', '/number', '/partial', '/product', '/protein_id', '/pseudo', '/standard_name', '/translation', '/transl_except', '/transl_table', or '/usedin'. Only some of these tags can have values, for example '/gene' can have the value of the gene name. By default any feature tag value in the feature table is shown. You can set this to match any feature tag valueyou wish to show.
The tag value may be wildcarded by using '*'.
If you wish to show more than one tag value, separate their names with the character '|', eg:
pax* | 10</td>
<td>Any string is accepted</td>
<td>*</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-orfminsize</td>
<td>Minimum size of Open Reading Frames (ORFs) to display in the translations.</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr>
<td>-flatreformat</td>
<td>Display RE sites in flat format</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-mincuts</td>
<td>Minimum cuts per RE</td>
<td>Integer from 1 to 1000</td>
<td>1</td>
</tr>

<tr>
<td>-maxcuts</td>
<td>Maximum cuts per RE</td>
<td>Integer up to 2000000000</td>
<td>2000000000</td>
</tr>

<tr>
<td>-sitelen</td>
<td>Minimum recognition site length</td>
<td>Integer from 2 to 20</td>
<td>4</td>
</tr>

<tr>
<td>-single</td>
<td>Force single RE site only cuts</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]blunt</td>
<td>Allow blunt end RE cutters</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]sticky</td>
<td>Allow sticky end RE cutters</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]ambiguity</td>
<td>Allow ambiguous RE matches</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-plasmid</td>
<td>Allow circular DNA</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]commercial</td>
<td>Only use restriction enzymes with suppliers</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]limit</td>
<td>Limits RE hits to one isoschizomer</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-preferred</td>
<td>Report preferred isoschizomers</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-threeletter</td>
<td>Display protein sequences in three-letter code</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-number</td>
<td>Number the sequences</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-width</td>
<td>Width of sequence to display</td>
<td>Integer 1 or more</td>
<td>60</td>
</tr>

<tr>
<td>-length</td>
<td>Line length of page (0 for indefinite)</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr>
<td>-margin</td>
<td>Margin around sequence for numbering</td>
<td>Integer 0 or more</td>
<td>10</td>
</tr>

<tr>
<td>-[no]name</td>
<td>Set this to be false if you do not wish to display the ID name of the sequence</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]description</td>
<td>Set this to be false if you do not wish to display the description of the sequence</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-offset</td>
<td>Offset to start numbering the sequence from</td>
<td>Any integer value</td>
<td>1</td>
</tr>

<tr>
<td>-html</td>
<td>Use HTML formatting</td>
<td>Yes/No</td>
<td>No</td>
</tr>

</table>



<H2>
    Input file format
</H2>

You can specifiy a file of ranges to display in uppercase by giving the
'-uppercase' qualifier the value '@' followed by the name of the file
containing the ranges.  (eg: '-upper @myfile').
<p>
          
The format of the range file is:
<p>
<ul>
      <li>Comment lines start with '#' in the first column.
      <li>Comment lines and blank lines are ignored.
      <li>The line may start with white-space.
      <li>There are two positive (integer) numbers per line separated by one
                or more space or
      <li>TAB characters.
      <li>The second number must be greater or equal to the first number.  
      <li>There can be optional text after the two numbers to annotate the line.
      <li>White-space before or after the text is removed.
</ul>
<p>
An example range file is:
<p>

<pre>          
# this is my set of ranges
12   23                           
 4   5       this is like 12-23, but smaller
67   10348   interesting region
</pre>
<p>

You can specifiy a file of ranges to highlight in a different colour
when outputting in HTML format (using the '-html' qualifier) by giving
the '-highlight' qualifier the value '@' followed by the name of the
file containing the ranges.  (eg: '-highlight @myfile').
<p>

The format of this file is very similar to the format of the above
uppercase range file, except that the text after the start and end
positions is used as the HTML colour name.  This colour name is used 'as
is' when specifying the colour in HTML in a '<FONT COLOR=xxx>'
construct, (where 'xxx' is the name of the colour).
<p>

The standard names of HTML font colours are given in:
<A HREF="http://www.iconbazaar.com/color_tables/named_colors.html">
http://www.iconbazaar.com/color_tables/named_colors.html</A>
<p>

An example highlight range file is:
<p>

<pre>          
# this is my set of ranges
12   23         red
 4   5          darkturquoise
67   10348      #FFE4E1
</pre>
<p>

You can specifiy a file of ranges to annotate
by giving
the '-annotate' qualifier the value '@' followed by the name of the
file containing the ranges.  (eg: '-annotate @myfile').
<p>
     
The format of this file is very similar to the format of the above
highlight range file, except that the text after the start and end
positions is used as the displayed text of the annotated region.
<p>

An example annotation range file is:
<p>
   
<pre>
# this is my set of ranges
12   23         exon 1
 4   5          CAP site
67   10348      exon 2
</pre>
<p>


You can specify a file of enzyme names to read in by giving the
'-enzymes' qualifier the name of the file holding the enzyme names with
a '@' character in front of it, for example, '@enz.list'.
<p>

Blank lines and lines starting with a '#' or '!' character are ignored
and all other lines are concatenated together with a comma character ','
and then treated as the list of enzymes to search for.
<p>
         
An example of a file of enzyme names is:
<p>

<pre>      
# my enzymes
HincII, ppiI
# other enzymes
hindiii
HinfI
</pre>
<p>            


<H2>
    Output file format
</H2>

Most of the variants of the output format have already been described in
the 'Description' and 'Usage' sections, but here is some more just to
fill out this section ;-)
<p>

The output format is extremely variable and under the control of the 
qualifiers used.
<p>
          
The sequence can be formatted for HTML display by using the '-html'
qualifier.  The top and tail html tags &lt;HEAD&gt;, &lt;BODY&gt; etc.  are not
included as it is expected that the output of this program will be
included in a more extensive HTML page and so these parts are left to
the user to provide.
<p>

The name of the sequence is displayed, followed by the description of
the sequence.  These can be turned off with the '-noname' and
'-nodescription' qualifiers.
<p>
      
Then the sequence is output, one line at a time.  Any associated
information to be displayed is also output above and below the sequence
line, as specified by the '-format' and or '-things' qualifiers.  (See
the 'Description' section for detals).
<p>
          
The margins around the sequence are specified by the use of the
'-margin' qaulifier and any numbering of the sequence and its
translations are placed in the margin.
<p>

A display of the restriction enzyme cut sites can be selected via
'-format 6' option or the '-format 0 -thing b,r,s,-r' style of options.
<p>

The option '-format 7' will produce a formatted display of cut sites on
the sequence, with the six-frame translation below it.  The cut sites
are indicated by a slash character '\' that points to the poition
between the nucleotides where the cuts occur.  Cuts by many enzymes at
the same position are indicated by stacking the enzyme names on top of
each other.
<p>

At the end the section header 'Enzymes that cut' is displayed followed
by a list of the enzymes that cut the specified sequence and the number
of times that they cut.
<p>

The '-flatreformat' qualifier changes the display to emphasise the
recognition site of the restriction enzyme, which is indicated by a row
of '=' characters.  The cut site if pointed to by a '>' or '<' character
and if the cut site is not within or imemdiately adjacent to the
recognition site, they are linked by a row or '.' characters.
<p>

The name of the enzyme is displayed above (or below when the reverse
sense site if displayed) the recognition site.  The name of the enzyme
is also displayed above the cut site if this occurs on a different
display line to the recognition site (i.e.  if it wraps onto the next
line of sequence).
<p>

An example of this display follows:
<p>

<pre>
% showseq embl:eclac stdout -send 60 -format 7 -enz TaqI,Hin6I,AciI,Hin6I,BssKI,Bsu6
Display a sequence with features, translation etc..
ECLAC
E.coli lactose operon with lacI, lacZ, lacY and lacA genes.
          
                                                      Hin6I
                 TaqI   Hin6I         AciI            | BssKI
                 \      \             \               \ \
          GACACCATCGAATGGCGCAAAACCTTTCGCGGTATGGCATGATAGCGCCCGGAAGAGAGT
                   10        20        30        40        50        60        
          ----:----|----:----|----:----|----:----|----:----|----:----|
          CTGTGGTAGCTTACCGCGTTTTGGAAAGCGCCATACCGTACTATCGCGGGCCTTCTCTCA
                   /      /             /               /    /
                   TaqI   Hin6I         AciI            |    BssKI
                                                        Hin6I

          D  T  I  E  W  R  K  T  F  R  G  M  A  *  *  R  P  E  E  S  
           T  P  S  N  G  A  K  P  F  A  V  W  H  D  S  A  R  K  R  V 
            H  H  R  M  A  Q  N  L  S  R  Y  G  M  I  A  P  G  R  E  S
          ----:----|----:----|----:----|----:----|----:----|----:----|
           S  V  M  S  H  R  L  V  K  R  P  I  A  H  Y  R  G  S  S  L 
          X  C  W  R  I  A  C  F  R  E  R  Y  P  M  I  A  G  P  L  S  
            V  G  D  F  P  A  F  G  K  A  T  H  C  S  L  A  R  F  L  T
</pre>
<p>




<H2>
    Data files
</H2>

EMBOSS data files are distributed with the application and stored in the
standard EMBOSS data directory, which is defined by EMBOSS environment
variable EMBOSS_DATA.
<p>
     
Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called ".embossdata".  Files
for all EMBOSS runs can be put in the user's home directory, or again in
a subdirectory called ".embossdata".
<p>
      
The directories are searched in the following order:
<p>
<pre>
      . (your current directory)
      .embossdata (under your current directory)
      ~/ (your home directory)
      ~/.embossdata
</pre>
<p>
The Genetic Code data files are based on the NCBI genetic code tables.
Their names and descriptions are:                       
<p>
<pre>

EGC.0      Standard (Differs from GC.1 in that it only has initiation site 'AUG')
EGC.1      Standard
EGC.2      Vertebrate Mitochondrial
EGC.3      Yeast Mitochondrial
EGC.4      Mold, Protozoan, Coelenterate Mitochondrial and Mycoplasma/Spiroplasma
EGC.5      Invertebrate Mitochondrial
EGC.6      Ciliate Macronuclear and Dasycladacean
EGC.9      Echinoderm Mitochondrial
EGC.10     Euplotid Nuclear
EGC.11     Bacterial
EGC.12     Alternative Yeast Nuclear
EGC.13     Ascidian Mitochondrial
EGC.14     Flatworm Mitochondrial
EGC.15     Blepharisma Macronuclear
</pre>
<p>
The format of these files is very simple.
<p>

It consists of several lines of optional comments, each starting with a
'#' character.
<p>

These are followed the line: 'Genetic Code [n]', where 'n' is the number
of the genetic code file.
<p>

This is followed by the description of the code and then by four lines
giving the IUPAC one-letter code of the translated amino acid, the start
codons (indicdated by an 'M') and the three bases of the codon, lined up
one on top of the other.
<p>
      
For example:
<p>

<pre>
# Genetic Code Table
#     
# Obtained from: http://www.ncbi.nlm.nih.gov/collab/FT/genetic_codes.html
# and: http://www3.ncbi.nlm.nih.gov/htbin-post/Taxonomy/wprintgc?mode=c
#
# Differs from Genetic Code [1] only in that the initiation sites have been
# changed to only 'AUG'

Genetic Code [0]
Standard

AAs  =   FFLLSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG
Starts = -----------------------------------M----------------------------
Base1  = TTTTTTTTTTTTTTTTCCCCCCCCCCCCCCCCAAAAAAAAAAAAAAAAGGGGGGGGGGGGGGGG
Base2  = TTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGG
Base3  = TCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAG
</pre>
<p>      
      
<b>showseq</b> uses the EMBOSS REBASE data files in 'data/REBASE/*' under the
EMBOSS installation directory.
<p>

These files must first be set up using the program '<b>rebaseextract</b>'.
Running 'rebaseextract' may be the job of your system manager.
<p>
      


<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with status 0.

<H2>
    Known bugs
</H2>

None known.
<p>

If you ask for the sequence display to end at position '100', with the
qualifier '-send 100', it will display the sequence up to the end of the
line - position '120'.  This is a feature of this program to make the
display of things like restriction enzyme cutting sites easier.
<p>

It is not a bug. Please don't report it.
<p>
      


<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="abiview.html">abiview</a></td><td>Reads ABI file and display the trace</td></tr>
<tr><td><a href="backtranseq.html">backtranseq</a></td><td>Back translate a protein sequence</td></tr>
<tr><td><a href="cirdna.html">cirdna</a></td><td>Draws circular maps of DNA constructs</td></tr>
<tr><td><a href="coderet.html">coderet</a></td><td>Extract CDS, mRNA and translations from feature tables</td></tr>
<tr><td><a href="lindna.html">lindna</a></td><td>Draws linear maps of DNA constructs</td></tr>
<tr><td><a href="pepnet.html">pepnet</a></td><td>Displays proteins as a helical net</td></tr>
<tr><td><a href="pepwheel.html">pepwheel</a></td><td>Shows protein sequences as helices</td></tr>
<tr><td><a href="plotorf.html">plotorf</a></td><td>Plot potential open reading frames</td></tr>
<tr><td><a href="prettyplot.html">prettyplot</a></td><td>Displays aligned sequences, with colouring and boxing</td></tr>
<tr><td><a href="prettyseq.html">prettyseq</a></td><td>Output sequence with translated ranges</td></tr>
<tr><td><a href="recoder.html">recoder</a></td><td>Remove restriction sites but maintain the same translation</td></tr>
<tr><td><a href="redata.html">redata</a></td><td>Search REBASE for enzyme name, references, suppliers etc</td></tr>
<tr><td><a href="remap.html">remap</a></td><td>Display a sequence with restriction cut sites, translation etc</td></tr>
<tr><td><a href="restover.html">restover</a></td><td>Finds restriction enzymes that produce a specific overhang</td></tr>
<tr><td><a href="restrict.html">restrict</a></td><td>Finds restriction enzyme cleavage sites</td></tr>
<tr><td><a href="seealso.html">seealso</a></td><td>Finds programs sharing group names</td></tr>
<tr><td><a href="showalign.html">showalign</a></td><td>Displays a multiple sequence alignment</td></tr>
<tr><td><a href="showdb.html">showdb</a></td><td>Displays information on the currently available databases</td></tr>
<tr><td><a href="showfeat.html">showfeat</a></td><td>Show features of a sequence</td></tr>
<tr><td><a href="showorf.html">showorf</a></td><td>Pretty output of DNA translations</td></tr>
<tr><td><a href="silent.html">silent</a></td><td>Silent mutation restriction enzyme scan</td></tr>
<tr><td><a href="textsearch.html">textsearch</a></td><td>Search sequence documentation text. SRS and Entrez are faster!</td></tr>
<tr><td><a href="transeq.html">transeq</a></td><td>Translate nucleic acid sequences</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Gary Williams <a
href="mailto:gwilliam@hgmp.mrc.ac.uk">(gwilliam@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>

Written 1999 - GWW
<br>
23 Aug 2000 - features display added - GWW
<br>
20 Nov 2001 - feature matches and annotation display added - GWW

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

