<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: showalign
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">
<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>
<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>
</td>
<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: showalign
</H2>
</font>
</td>
</tr>
</table>
</td>
</tr>
</table>
<p><hr><p>

<!--END OF HEADER-->

<H1>
Program showalign
</H1>
<H2>
    Function
</H2>
Display a multiple sequence alignment

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
        Algorithms
  -->

<b>showalign</b> displays an aligned set of protein or a nucleic acid
sequences in a style suitable for publication. 

<p>

The output is sent to the screen by default for the user to view, but it
can write the results to a file.

<p>


The output highlights various differences or similarities between each
of the sequences and a reference sequence by setting selected types of
matches to a reference sequence to be '.' characters.

<p>


The reference sequence can be either the calculated consensus sequence
(the default) or it can be one of the set of aligned sequences,
specified by either the ordinal number of that sequence in the input
file, or by its name. 

<p>

The output sequences can be displayed in either the input order (the
default) or they can be sorted in order of their similarity to the
reference sequence or sorted alphabetically by their names. 

<p>

By using the '-show' option, the displayed sequences can either be shown as:
<UL>
<LI>complete (-show=All), 
<LI>only identical matches between
the sequence and the reference sequence, all other positions being
replaced by '.' characters (-show=Identities)
<LI>non-identical matches, with identical matches being replaced by '.'
characters, similar matches are shown in lower case (-show=Non-identities)
<LI>similar matches, with non-similar matches being replaced by '.'  
characters, similar matches are shown in lower case (-show=Similarities)
<LI>dissimilar matches, with identical or similar matches being replaced by 
'.' characters (-show=Dissimilarities)
</UL>

<p>

A small table of the way these alignments are displayed illustrates this.

<br>

If we have a reference protein sequence of "III" and a sequence aligned
to this of "ILW", then we have an identical matching residue, then a
similar one, then a dissimilar one. 

<br>

The different methods of display would give the following:

<p>

<hr>
<pre>
Reference       III

All             ILw
Identical       I..
Non-id          .lW
Similar         Il.
Dissimilar      ..W
</pre>
<hr>

<p>

Changing the similar matches to lowercase can optionally be disable by
using the option -nosimilarcase.

<p>

The displayed sequence can be numbered by placing a ruler with ticks
above the sequence. 

<p>

The width of a line can be set.  The width of a margin to the left of
the sequences that shows the sequence names can be set. 

<p>

Specified regions of the sequence can be displayed in uppercase to
highlight them.

<p>

The output can be formatted for HTML.

<p>

If the output is being formatted for HTML, then specified regions of the
sequence can be displayed in any valid HTML colours.

<p>



<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->


Here is a sample session with showalign:

<pre>

% showalign
Display a multiple sequence alignment
Input sequence set: globin.seq
Output file [stdout]: 
                  10        20        30        40        50        60
          ----:----|----:----|----:----|----:----|----:----|----:----|
Consensus vhltxeeksavtXlwgkvnvdevggealgrllvvypwtqrffesfgdlssxdavmxnpkv
HBB_MACFU ....P...n...T.....................................P....G....
HBD_HUMAN ....P...t..NA........A............................P....G....
HBB_RABIT ...sS.......A.......e.............................An...N....
HBB_BOVIN -M..A...a...AF....K..........-...................tA....N....
HBB_LAMGL .n.sGd..n..HG..S..K..........-........r..........tA....N....
HBE_HUMAN ..F.A...a...S..S.m..e.A...................d...n...PS.ilG....
HBE_RABIT ..F.P...CIisKQ..q..i..T...................dn..n...SS.i.G....
Consensus vhltxeeksavtXlwgkvnvdevggealgrllvvypwtqrffesfgdlssxdavmxnpkv

                  70        80        90       100       110       120
          ----:----|----:----|----:----|----:----|----:----|----:----|
Consensus kahgkkvlxsfxdglXhldnlkgtfaklselhcdxxhvdpenfkllgnvlvivlaxhfgk
HBB_MACFU ........Ga.S...N..........q.......--...............C...H....
HBD_HUMAN ........Ga.S...A.........sq.......--.......r.......C...Rn...
HBB_RABIT ........Aa.Se..S..................KL.......r..........sH....
HBB_BOVIN ........D..Sn.mK...d......A.......KL...............v...Rn...
HBB_LAMGL ....S...N..G...S........y.........KL.......r.......v...R....
HBE_HUMAN ........T..G.AiKnm....PA..........--.............m..i..T....
HBE_RABIT ........T..G.AiKnm.....A..........--..............l....T....
Consensus kahgkkvlxsfxdglXhldnlkgtfaklselhcdxxhvdpenfkllgnvlvivlaxhfgk

                 130       140       150       160       170       180
          ----:----|----:----|----:----|----:----|----:----|----:----|
Consensus eftpqvqaayqkvvagvanalahkyhxxxx
HBB_MACFU ..........................
HBD_HUMAN .....m....................
HBB_RABIT ..........................
HBB_BOVIN ....Vl..Df.............r..
HBB_LAMGL ....Dl.................r..ACDH
HBE_HUMAN ....e....w..l.sA
HBE_RABIT ....e....w..l.s...I.......
Consensus eftpqvqaayqkvvagvanalahkyhxxxx
</pre>




<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqset     The sequence alignment to be displayed.
  [-outfile]           outfile    If you enter the name of a file here then
                                  this program will write the sequence details
                                  into that file.

   Optional qualifiers:
   -refseq             string     If you give the number in the alignment or
                                  the name of a sequence, it will be taken to
                                  be the reference sequence. The reference
                                  sequence is always show in full and is the
                                  one against which all the other sequences
                                  are compared. If this is set to 0 then the
                                  consensus sequence will be used as the
                                  reference sequence. By default the consensus
                                  sequence is used as the reference sequence.
   -[no]bottom         bool       If this is true then the refernce sequence
                                  is displayed at the bottom of the alignment
                                  as well as at the top.
   -show               list       What to show
   -order              list       Output order of the sequences
   -[no]similarcase    bool       If this is set True, then when -show is set
                                  to 'Similarities' or 'Non-identities' and a
                                  residue is similar but not identical to the
                                  reference sequence residue, it will be
                                  changed to lower-case. If -show is set to
                                  'All' then non-identical, non-similar
                                  residues will be changed to lower-case. If
                                  this is False then no change to the case of
                                  the residues is made on the basis of their
                                  similarity to the reference sequence.
   -matrix             matrix     Similarity scoring Matrix file
   -[no]consensus      bool       If this is true then the consensus line is
                                  displayed at the bottom.

   Advanced qualifiers:
   -uppercase          range      Regions to put in uppercase.
                                  If this is left blank, then the sequence
                                  case is left alone.
                                  A set of regions is specified by a set of
                                  pairs of positions.
                                  The positions are integers.
                                  They are separated by any non-digit,
                                  non-alpha character.
                                  Examples of region specifications are:
                                  24-45, 56-78
                                  1:45, 67=99;765..888
                                  1,5,8,10,23,45,57,99
   -[no]number         bool       If this option is true then a line giving
                                  the positions in the alignment is displayed
                                  every 10 characters above the alignment.
   -[no]ruler          bool       If this option is true then a ruler line
                                  marking every 5th and 10th character in the
                                  alignment is displayed.
   -width              integer    Width of sequence to display
   -margin             integer    This sets the length of the left-hand margin
                                  for sequence names. If the margin is set at
                                  0 then no margin and no names are
                                  displayed. If the margin is set to a value
                                  that is less than the length of a sequence
                                  name then the sequence name is displayed
                                  truncated to the length of the margin. If
                                  the margin is set to -1 then the minimum
                                  margin width that will allow all the
                                  sequence names to be displayed in full plus
                                  a space at the end of the name will
                                  automatically be selected.
   -[no]names          bool       Set this to be false if you do not wish to
                                  display the ID name of the sequences
   -html               bool       Use HTML formatting
   -highlight          range      Regions to colour if formatting for HTML.
                                  If this is left blank, then the sequence is
                                  left alone.
                                  A set of regions is specified by a set of
                                  pairs of positions.
                                  The positions are integers.
                                  They are followed by any valid HTML font
                                  colour.
                                  Examples of region specifications are:
                                  24-45 blue 56-78 orange
                                  1-100 green 120-156 red
                                  A file of ranges to colour (one range per
                                  line) can be specified as '@filename'.
   -plurality          float      Set a cut-off for the % of positive scoring
                                  matches below which there is no consensus.
                                  The default plurality is taken as 50% of the
                                  total weight of all the sequences in the
                                  alignment.
   -setcase            float      Sets the threshold for the scores of the
                                  positive matches above which the consensus
                                  is is upper-case and below which the
                                  consensus is in lower-case.
   -identity           float      Provides the facility of setting the
                                  required number of identities at a position
                                  for it to give a consensus. Therefore, if
                                  this is set to 100% only columns of
                                  identities contribute to the consensus.

   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>The sequence alignment to be displayed.</td>
<td>Readable sequences</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>If you enter the name of a file here then this program will write the sequence details into that file.</td>
<td>Output file</td>
<td>stdout</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-refseq</td>
<td>If you give the number in the alignment or the name of a sequence, it will be taken to be the reference sequence. The reference sequence is always show in full and is the one against which all the other sequences are compared. If this is set to 0 then the consensus sequence will be used as the reference sequence. By default the consensus sequence is used as the reference sequence.</td>
<td>Any string is accepted</td>
<td>0</td>
</tr>

<tr>
<td>-[no]bottom</td>
<td>If this is true then the refernce sequence is displayed at the bottom of the alignment as well as at the top.</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-show</td>
<td>What to show</td>
<td><table><tr><td>A</td> <td><i>(All of the sequences)</i></td></tr><tr><td>I</td> <td><i>(Identities between the sequences)</i></td></tr><tr><td>N</td> <td><i>(Non-identities between the sequences)</i></td></tr><tr><td>S</td> <td><i>(Similarities between the sequences)</i></td></tr><tr><td>D</td> <td><i>(Dissimilarities between the sequences)</i></td></tr></table></td>
<td>N</td>
</tr>

<tr>
<td>-order</td>
<td>Output order of the sequences</td>
<td><table><tr><td>I</td> <td><i>(Input order - no change)</i></td></tr><tr><td>A</td> <td><i>(Alphabetical order of the names)</i></td></tr><tr><td>S</td> <td><i>(Similarity to the reference sequence)</i></td></tr></table></td>
<td>I</td>
</tr>

<tr>
<td>-[no]similarcase</td>
<td>If this is set True, then when -show is set to 'Similarities' or 'Non-identities' and a residue is similar but not identical to the reference sequence residue, it will be changed to lower-case. If -show is set to 'All' then non-identical, non-similar residues will be changed to lower-case. If this is False then no change to the case of the residues is made on the basis of their similarity to the reference sequence.</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-matrix</td>
<td>Similarity scoring Matrix file</td>
<td>Comparison matrix file in EMBOSS data path</td>
<td>EBLOSUM62 for protein<br>EDNAFULL for DNA</td>
</tr>

<tr>
<td>-[no]consensus</td>
<td>If this is true then the consensus line is displayed at the bottom.</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-uppercase</td>
<td>Regions to put in uppercase.
If this is left blank, then the sequence case is left alone.
A set of regions is specified by a set of pairs of positions.
The positions are integers.
They are separated by any non-digit, non-alpha character.
Examples of region specifications are:
24-45, 56-78
1:45, 67=99;765..888
1,5,8,10,23,45,57,99</td>
<td>Sequence range</td>
<td>If this is left blank, then the sequence case is left alone.</td>
</tr>

<tr>
<td>-[no]number</td>
<td>If this option is true then a line giving the positions in the alignment is displayed every 10 characters above the alignment.</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]ruler</td>
<td>If this option is true then a ruler line marking every 5th and 10th character in the alignment is displayed.</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-width</td>
<td>Width of sequence to display</td>
<td>Integer 1 or more</td>
<td>60</td>
</tr>

<tr>
<td>-margin</td>
<td>This sets the length of the left-hand margin for sequence names. If the margin is set at 0 then no margin and no names are displayed. If the margin is set to a value that is less than the length of a sequence name then the sequence name is displayed truncated to the length of the margin. If the margin is set to -1 then the minimum margin width that will allow all the sequence names to be displayed in full plus a space at the end of the name will automatically be selected.</td>
<td>Integer -1 or more</td>
<td>-1</td>
</tr>

<tr>
<td>-[no]names</td>
<td>Set this to be false if you do not wish to display the ID name of the sequences</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-html</td>
<td>Use HTML formatting</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-highlight</td>
<td>Regions to colour if formatting for HTML.
If this is left blank, then the sequence is left alone.
A set of regions is specified by a set of pairs of positions.
The positions are integers.
They are followed by any valid HTML font colour.
Examples of region specifications are:
24-45 blue 56-78 orange
1-100 green 120-156 red
A file of ranges to colour (one range per line) can be specified as '@filename'.</td>
<td>Sequence range</td>
<td><i>full sequence</i></td>
</tr>

<tr>
<td>-plurality</td>
<td>Set a cut-off for the % of positive scoring matches below which there is no consensus. The default plurality is taken as 50% of the total weight of all the sequences in the alignment.</td>
<td>Number from 0.000 to 100.000</td>
<td>50.0</td>
</tr>

<tr>
<td>-setcase</td>
<td>Sets the threshold for the scores of the positive matches above which the consensus is is upper-case and below which the consensus is in lower-case.</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr>
<td>-identity</td>
<td>Provides the facility of setting the required number of identities at a position for it to give a consensus. Therefore, if this is set to 100% only columns of identities contribute to the consensus.</td>
<td>Number from 0.000 to 100.000</td>
<td>0.0</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>showalign</b> reads in a set of aligned protein or nucleic sequences. 

<p>

You can specifiy a file of ranges to display in uppercase by giving the
'-uppercase' qualifier the value '@' followed by the name of the file
containing the ranges.  (eg: '-upper @myfile').
<p>
          
The format of the range file is:
<p>
<ul>
      <li>Comment lines start with '#' in the first column.
      <li>Comment lines and blank lines are ignored.
      <li>The line may start with white-space.
      <li>There are two positive (integer) numbers per line separated by one
                or more space or
      <li>TAB characters.
      <li>The second number must be greater or equal to the first number.  
      <li>There can be optional text after the two numbers to annotate the line.
      <li>White-space before or after the text is removed.
</ul>
<p>
An example range file is:
<p>

<pre>          
# this is my set of ranges
12   23                           
 4   5       this is like 12-23, but smaller
67   10348   interesting region
</pre>
<p>

You can specifiy a file of ranges to highlight in a different colour
when outputting in HTML format (using the '-html' qualifier) by giving
the '-highlight' qualifier the value '@' followed by the name of the
file containing the ranges.  (eg: '-highlight @myfile').
<p>

The format of this file is very similar to the format of the above
uppercase range file, except that the text after the start and end
positions is used as the HTML colour name.  This colour name is used 'as
is' when specifying the colour in HTML in a '<FONT COLOR=xxx>'
construct, (where 'xxx' is the name of the colour).
<p>

The standard names of HTML font colours are given in:
<A HREF="http://www.iconbazaar.com/color_tables/named_colors.html">
http://www.iconbazaar.com/color_tables/named_colors.html</A>
<p>

An example highlight range file is:
<p>

<pre>          
# this is my set of ranges
12   23         red
 4   5          darkturquoise
67   10348      #FFE4E1
</pre>
<p>













<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->


<b>showalign</b> writes out a text file, optionally formatted for HTML. 

<p>

Some examples of different optional output styles follows:

<p>

<p>
<hr>
<pre>
% showalign g.seq stdout -show=n
Display a multiple sequence alignment
                  10        20        30        40        50        60
          ----:----|----:----|----:----|----:----|----:----|----:----|
Consensus vhltxeeksavtXlwgkvnvdevggealgrllvvypwtqrffesfgdlssxdavmxnpk
HBB_MACFU ....P...n...T.....................................P....G...
HBD_HUMAN ....P...t..NA........A............................P....G...
HBB_RABIT ...sS.......A.......e.............................An...N...
HBB_BOVIN -M..A...a...AF....K..........-...................tA....N...
HBB_LAMGL .n.sGd..n..HG..S..K..........-........r..........tA....N...
HBE_HUMAN ..F.A...a...S..S.m..e.A...................d...n...PS.ilG...
HBE_RABIT ..F.P...CIisKQ..q..i..T...................dn..n...SS.i.G...
Consensus vhltxeeksavtXlwgkvnvdevggealgrllvvypwtqrffesfgdlssxdavmxnpk
</pre>
<hr>
<p>


<p>
<hr>
<pre>
% showalign g.seq stdout -show=a
Display a multiple sequence alignment
                  10        20        30        40        50        60
          ----:----|----:----|----:----|----:----|----:----|----:----|
Consensus vhltxeeksavtXlwgkvnvdevggealgrllvvypwtqrffesfgdlssxdavmxnpk
HBB_MACFU VHLTpEEKNAVTtLWGKVNVDEVGGEALGRLLVVYPWTQRFFESFGDLSSpDAVMgNPK
HBD_HUMAN VHLTpEEKTAVnaLWGKVNVDaVGGEALGRLLVVYPWTQRFFESFGDLSSpDAVMgNPK
HBB_RABIT VHLSsEEKSAVTaLWGKVNVEEVGGEALGRLLVVYPWTQRFFESFGDLSSaNAVMnNPK
HBB_BOVIN -mLTaEEKAAVTafWGKVkVDEVGGEALG-LLVVYPWTQRFFESFGDLSTaDAVMnNPK
HBB_LAMGL VNLSgDEKNAVhgLWsKVkVDEVGGEALG-LLVVYPWTRRFFESFGDLSTaDAVMnNPK
HBE_HUMAN VHfTaEEKAAVTsLWsKMNVEEaGGEALGRLLVVYPWTQRFFDSFGNLSSpsAILgNPK
HBE_RABIT VHfTpEEKciISkqWGQVNIDEtGGEALGRLLVVYPWTQRFFDNFGNLSSssAIMgNPK
Consensus vhltxeeksavtXlwgkvnvdevggealgrllvvypwtqrffesfgdlssxdavmxnpk
</pre>
<hr>
<p>


<p>
<hr>
<pre>
% showalign g.seq stdout -show=i
Display a multiple sequence alignment
                  10        20        30        40        50        60
          ----:----|----:----|----:----|----:----|----:----|----:----|
Consensus vhltxeeksavtXlwgkvnvdevggealgrllvvypwtqrffesfgdlssxdavmxnpk
HBB_MACFU VHLT.EEK.AVT.LWGKVNVDEVGGEALGRLLVVYPWTQRFFESFGDLSS.DAVM.NPK
HBD_HUMAN VHLT.EEK.AV..LWGKVNVD.VGGEALGRLLVVYPWTQRFFESFGDLSS.DAVM.NPK
HBB_RABIT VHL..EEKSAVT.LWGKVNV.EVGGEALGRLLVVYPWTQRFFESFGDLSS..AVM.NPK
HBB_BOVIN -.LT.EEK.AVT..WGKV.VDEVGGEALG-LLVVYPWTQRFFESFGDLS..DAVM.NPK
HBB_LAMGL V.L...EK.AV..LW.KV.VDEVGGEALG-LLVVYPWT.RFFESFGDLS..DAVM.NPK
HBE_HUMAN VH.T.EEK.AVT.LW.K.NV.E.GGEALGRLLVVYPWTQRFF.SFG.LSS..A...NPK
HBE_RABIT VH.T.EEK......WG.VN.DE.GGEALGRLLVVYPWTQRFF..FG.LSS..A.M.NPK
Consensus vhltxeeksavtXlwgkvnvdevggealgrllvvypwtqrffesfgdlssxdavmxnpk
</pre>
<hr>
<p>


<p>
<hr>
<pre>
% showalign g.seq stdout -show=s
Display a multiple sequence alignment
                  10        20        30        40        50        60
          ----:----|----:----|----:----|----:----|----:----|----:----|
Consensus vhltxeeksavtXlwgkvnvdevggealgrllvvypwtqrffesfgdlssxdavmxnpk
HBB_MACFU VHLT.EEKnAVT.LWGKVNVDEVGGEALGRLLVVYPWTQRFFESFGDLSS.DAVM.NPK
HBD_HUMAN VHLT.EEKtAV..LWGKVNVD.VGGEALGRLLVVYPWTQRFFESFGDLSS.DAVM.NPK
HBB_RABIT VHLs.EEKSAVT.LWGKVNVeEVGGEALGRLLVVYPWTQRFFESFGDLSS.nAVM.NPK
HBB_BOVIN -.LT.EEKaAVT..WGKV.VDEVGGEALG-LLVVYPWTQRFFESFGDLSt.DAVM.NPK
HBB_LAMGL VnLs.dEKnAV..LW.KV.VDEVGGEALG-LLVVYPWTrRFFESFGDLSt.DAVM.NPK
HBE_HUMAN VH.T.EEKaAVT.LW.KmNVeE.GGEALGRLLVVYPWTQRFFdSFGnLSS..Ail.NPK
HBE_RABIT VH.T.EEK..is..WGqVNiDE.GGEALGRLLVVYPWTQRFFdnFGnLSS..AiM.NPK
Consensus vhltxeeksavtXlwgkvnvdevggealgrllvvypwtqrffesfgdlssxdavmxnpk
</pre>
<hr>
<p>


<p>
<hr>
<pre>
% showalign g.seq stdout -show=d
Display a multiple sequence alignment
                  10        20        30        40        50        60
          ----:----|----:----|----:----|----:----|----:----|----:----|
Consensus vhltxeeksavtXlwgkvnvdevggealgrllvvypwtqrffesfgdlssxdavmxnpk
HBB_MACFU ....P.......T.....................................P....G...
HBD_HUMAN ....P......NA........A............................P....G...
HBB_RABIT ....S.......A.....................................A....N...
HBB_BOVIN -M..A.......AF....K..........-....................A....N...
HBB_LAMGL ....G......HG..S..K..........-....................A....N...
HBE_HUMAN ..F.A.......S..S......A...........................PS...G...
HBE_RABIT ..F.P...CI..KQ........T...........................SS...G...
Consensus vhltxeeksavtXlwgkvnvdevggealgrllvvypwtqrffesfgdlssxdavmxnpk
</pre>
<hr>
<p>


<p>
<hr>
<pre>
% showalign g.seq stdout -ref=1
Display a multiple sequence alignment
                  10        20        30        40        50        60
          ----:----|----:----|----:----|----:----|----:----|----:----|
HBB_MACFU VHLTPEEKNAVTTLWGKVNVDEVGGEALGRLLVVYPWTQRFFESFGDLSSPDAVMGNPK
HBD_HUMAN ........T..NA........A.....................................
HBB_RABIT ...sS...s...A.......e.............................An...N...
HBB_BOVIN -M..A...A...AF....K..........-...................tA....N...
HBB_LAMGL .n.sGd.....HG..S..K..........-........r..........tA....N...
HBE_HUMAN ..F.A...A...s..S.m..e.A...................d...n....S.il....
HBE_RABIT ..F.....CIisKQ..q..i..T...................dn..n...SS.i.....
HBB_MACFU VHLTPEEKNAVTTLWGKVNVDEVGGEALGRLLVVYPWTQRFFESFGDLSSPDAVMGNPK
Consensus vhltxeeksavtXlwgkvnvdevggealgrllvvypwtqrffesfgdlssxdavmxnpk
</pre>
<hr>
<p>


<p>
<hr>
<pre>
Display a range of sequences in uppercase, everything else in lowercase
% showalign g.seq stdout -nocon -ref=1 -sl -upper 9-15 -nosimilarcaseDisplay a multiple sequence alignment
                  10        20        30        40        50        60
          ----:----|----:----|----:----|----:----|----:----|----:----|
HBB_MACFU vhltpeekNAVTTLWgkvnvdevggealgrllvvypwtqrffesfgdlsspdavmgnpk
HBD_HUMAN ........T..NA........a.....................................
HBB_RABIT ...ss...S...A.......e.............................an...n...
HBB_BOVIN -m..a...A...AF....k..........-...................ta....n...
HBB_LAMGL .n.sgd.....HG..s..k..........-........r..........ta....n...
HBE_HUMAN ..f.a...A...S..s.m..e.a...................d...n....s.il....
HBE_RABIT ..f.....CIISKQ..q..i..t...................dn..n...ss.i.....
HBB_MACFU vhltpeekNAVTTLWgkvnvdevggealgrllvvypwtqrffesfgdlsspdavmgnpk
</pre>
<hr>
<p>



<p>
<hr>
<pre>
Alphabetical order
% showalign g.seq stdout -order=a
Display a multiple sequence alignment
                  10        20        30        40        50        60
          ----:----|----:----|----:----|----:----|----:----|----:----|
Consensus vhltxeeksavtXlwgkvnvdevggealgrllvvypwtqrffesfgdlssxdavmxnpk
HBB_BOVIN -M..A...a...AF....K..........-...................tA....N...
HBB_LAMGL .n.sGd..n..HG..S..K..........-........r..........tA....N...
HBB_MACFU ....P...n...T.....................................P....G...
HBB_RABIT ...sS.......A.......e.............................An...N...
HBD_HUMAN ....P...t..NA........A............................P....G...
HBE_HUMAN ..F.A...a...S..S.m..e.A...................d...n...PS.ilG...
HBE_RABIT ..F.P...CIisKQ..q..i..T...................dn..n...SS.i.G...
Consensus vhltxeeksavtXlwgkvnvdevggealgrllvvypwtqrffesfgdlssxdavmxnpk
</pre>
<hr>
<p>


<p>
<hr>
<pre>
In order of similarity to the reference sequence
% showalign g.seq stdout -order=s
Display a multiple sequence alignment
                  10        20        30        40        50        60
          ----:----|----:----|----:----|----:----|----:----|----:----|
Consensus vhltxeeksavtXlwgkvnvdevggealgrllvvypwtqrffesfgdlssxdavmxnpk
HBB_MACFU ....P...n...T.....................................P....G...
HBB_RABIT ...sS.......A.......e.............................An...N...
HBD_HUMAN ....P...t..NA........A............................P....G...
HBB_BOVIN -M..A...a...AF....K..........-...................tA....N...
HBE_HUMAN ..F.A...a...S..S.m..e.A...................d...n...PS.ilG...
HBB_LAMGL .n.sGd..n..HG..S..K..........-........r..........tA....N...
HBE_RABIT ..F.P...CIisKQ..q..i..T...................dn..n...SS.i.G...
Consensus vhltxeeksavtXlwgkvnvdevggealgrllvvypwtqrffesfgdlssxdavmxnpk
</pre>
<hr>
<p>


<p>
<hr>
<pre>
Format for HTML and highlight some inetersting regions in different colours
% showalign g.seq stdout -html -high '4-13 green 43-43 red 51-56 blue' 
Display a multiple sequence alignment
                  10        20        30        40        50        60
          ----:----|----:----|----:----|----:----|----:----|----:----|
Consensus vhl<font color=green>txeeksavtX</font>lwgkvnvdevggealgrllvvypwtqrff<font color=red>e</font>sfgdlss<font color=blue>xdavmx</font>npk
HBB_MACFU ...<font color=green>.P...n...T</font>.............................<font color=red>.</font>.......<font color=blue>P....G</font>...
HBD_HUMAN ...<font color=green>.P...t..NA</font>........A....................<font color=red>.</font>.......<font color=blue>P....G</font>...
HBB_RABIT ...<font color=green>sS.......A</font>.......e.....................<font color=red>.</font>.......<font color=blue>An...N</font>...
HBB_BOVIN -M.<font color=green>.A...a...A</font>F....K..........-............<font color=red>.</font>......t<font color=blue>A....N</font>...
HBB_LAMGL .n.<font color=green>sGd..n..HG</font>..S..K..........-........r...<font color=red>.</font>......t<font color=blue>A....N</font>...
HBE_HUMAN ..F<font color=green>.A...a...S</font>..S.m..e.A...................<font color=red>d</font>...n...<font color=blue>PS.ilG</font>...
HBE_RABIT ..F<font color=green>.P...CIisK</font>Q..q..i..T...................<font color=red>d</font>n..n...<font color=blue>SS.i.G</font>...
Consensus vhl<font color=green>txeeksavtX</font>lwgkvnvdevggealgrllvvypwtqrff<font color=red>e</font>sfgdlss<font color=blue>xdavmx</font>npk
</pre>
<hr>
<p>



<p>
<hr>
<pre>
No consensus line at the bottom
No ruler line
No numbers line
Don't repeat the reference sequence at the bottom of the sequences
Use sequence 1 as the reference sequence
Display residues from position 10 to 30 only
% showalign g.seq stdout -nocon -norule -nonum -nobot -ref=1 -sb=10 -send=30
Display a multiple sequence alignment

HBB_MACFU AVTTLWGKVNVDEVGGEALGR
HBD_HUMAN ..NA........A........
HBB_RABIT ...A.......e.........
HBB_BOVIN ...AF....K..........-
HBB_LAMGL ..HG..S..K..........-
HBE_HUMAN ...s..S.m..e.A.......
HBE_RABIT IisKQ..q..i..T.......
</pre>
<hr>
<p>









<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.
	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
   -->

<b>showalign</b> reads in scoring matrices to determine the consesnus
sequence and to determine which matches are similar or not. 

<p>

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>















<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="abiview.html">abiview</a></td><td>Reads ABI file and display the trace</td></tr>
<tr><td><a href="cirdna.html">cirdna</a></td><td>Draws circular maps of DNA constructs</td></tr>
<tr><td><a href="emma.html">emma</a></td><td>Multiple alignment program - interface to ClustalW program</td></tr>
<tr><td><a href="infoalign.html">infoalign</a></td><td>Information on a multiple sequence alignment</td></tr>
<tr><td><a href="lindna.html">lindna</a></td><td>Draws linear maps of DNA constructs</td></tr>
<tr><td><a href="pepnet.html">pepnet</a></td><td>Displays proteins as a helical net</td></tr>
<tr><td><a href="pepwheel.html">pepwheel</a></td><td>Shows protein sequences as helices</td></tr>
<tr><td><a href="plotcon.html">plotcon</a></td><td>Plots the quality of conservation of a sequence alignment</td></tr>
<tr><td><a href="prettyplot.html">prettyplot</a></td><td>Displays aligned sequences, with colouring and boxing</td></tr>
<tr><td><a href="prettyseq.html">prettyseq</a></td><td>Output sequence with translated ranges</td></tr>
<tr><td><a href="remap.html">remap</a></td><td>Display a sequence with restriction cut sites, translation etc</td></tr>
<tr><td><a href="seealso.html">seealso</a></td><td>Finds programs sharing group names</td></tr>
<tr><td><a href="showdb.html">showdb</a></td><td>Displays information on the currently available databases</td></tr>
<tr><td><a href="showfeat.html">showfeat</a></td><td>Show features of a sequence</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="textsearch.html">textsearch</a></td><td>Search sequence documentation text. SRS and Entrez are faster!</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

This application was written by Gary Williams <a
href="mailto:gwilliam@hgmp.mrc.ac.uk">(gwilliam@hgmp.mrc.ac.uk)</a>

















<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (23 May 2001) - Gary Williams














<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->








</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

