<HTML>

<HEAD>
  <TITLE>
  EMBOSS: showdb
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">







<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

showdb
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>






<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Displays information on the currently available databases


<H2>
    Description
</H2>


This writes out a simple table displaying the names, contents and
available ways of accessing the sequence databases.

<p>
The available ways of accessing the databases are 'ID', 'Query' and
'All'.  These refer to the way that you can search the databases to
get entries from them, which is governed by the ways the database has
been set up and the way it is organised and indexed.

<p>
Different databases may have different access capabilities, depending
on how your local site is organised.

<p>
EMBOSS has been designed to be extremely flexible in its use of
sequence databases formats, so that it is easy to set EMBOSS up to use
your site's existing databases.  Sometimes this means that it is hard
to extract entries from some databases in particular ways.  For
example, a flat file database with no index is only useful for reading
all entries, while a database located in another site that is
available via the WWW may only provide single entries.

<p>
'ID' (also known as 'single entry') allows the programs to extract a
single explicitly named entry from the database, for example
embl:x13776

<p>
'Query' (also known as 'wild') indicates that programs can extract a
set of matching wildcard entry names (this may be slow for some
methods of access).  For example you can look at all of the human PAX
proteins in SWISS_PROT by: swissprot:pax*_human

<p>
'All' allows the programs to analyse all the entries in the database
sequentially.  For example this lets you look at all entries in the
database with the notation: embl:*

<p>
A database may have several different methods of access available.
Ideally all of the databases available on your site will be available
in all three ways, but this is not the best of all posssible worlds
and so you might like to check how you can access the databases by
running this program and having a look.

<H2>
    Usage
</H2>


<b>Here is a sample session with showdb</b>
<p>
Display information on the currently available databases: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>showdb </b>
Displays information on the currently available databases

# Name        Type ID  Qry All Comment
# ====        ==== ==  === === =======
enspep        P    OK  OK  OK  ENSEMBL pep
enspepp       P    OK  OK  OK  ENSEMBL pep predicted
enspeppredicted P    OK  OK  OK  ENSEMBL pep predicted
remtrembl     P    OK  OK  OK  REMTREMBL sequences
sptrembl      P    OK  OK  OK  SPTREMBL sequences
sw            P    OK  OK  OK  SWISSPROT sequences
swissprot     P    OK  OK  OK  SWISSPROT sequences
tpir          P    OK  OK  OK  PIR using NBRF access for 4 files
trembl        P    OK  OK  OK  TREMBL sequences
tsw           P    OK  OK  OK  Swissprot native format with EMBL CD-ROM index
tswnew        P    OK  OK  OK  Swissnew as 3 files in native format with EMBL CD-ROM index
twp           P    OK  OK  OK  EMBL new in native format with EMBL CD-ROM index
em            N    OK  OK  OK  EMBL sequences
embl          N    OK  OK  OK  EMBL sequences
emblnew       N    OK  OK  OK  New EMBL sequences
enscdna       N    OK  OK  OK  ENSEMBL cdna
ensdna        N    OK  OK  OK  ENSEMBL dna
ensdnam       N    OK  OK  OK  ENSEMBL dna masked
ensdnamasked  N    OK  OK  OK  ENSEMBL dna masked
ensembl       N    OK  OK  OK  ENSEMBL 
est           N    OK  OK  OK  EMBL EST sequences
fungal        N    OK  OK  OK  EMBL fungal sequences
gb            N    OK  OK  OK  Genbank updates
genbank       N    OK  OK  OK  Genbank
gss           N    OK  OK  OK  EMBL GSS sequences
hsunigene     N    OK  -   -   hsunigene 
htg           N    OK  OK  OK  EMBL HTG sequences
human         N    OK  OK  OK  EMBL human sequences
humanest      N    OK  OK  OK  EMBL human EST sequences
invertebrate  N    OK  OK  OK  EMBL invertebrate sequences
mammalian     N    OK  OK  OK  EMBL mammalian sequences
mouseest      N    OK  OK  OK  EMBL mouse EST sequences
newgb         N    OK  OK  OK  Genbank updates
newgenbank    N    OK  OK  OK  Genbank updates
organelle     N    OK  OK  OK  EMBL organelle sequences
otherest      N    OK  OK  OK  EMBL other EST sequences
othervertebrate N    OK  OK  OK  EMBL other vertebrate sequences
patent        N    OK  OK  OK  EMBL patented sequences
phage         N    OK  OK  OK  EMBL phage sequences
plant         N    OK  OK  OK  EMBL plant sequences
prokaryotic   N    OK  OK  OK  EMBL prokaryotic sequences
refseq        N    OK  OK  OK  REFSEQ sequences with features in /home/br2/data/Chris...
refseq_srs    N    OK  -   -   REFSEQ-ID sequences from SRS
rodent        N    OK  OK  OK  EMBL rodent sequences
sts           N    OK  OK  OK  EMBL STS sequences
synthetic     N    OK  OK  OK  EMBL synthetic sequences
tembl         N    OK  OK  OK  EMBL in native format with EMBL CD-ROM index
tgb           N    OK  -   -   Genbank IDs
tgenbank      N    OK  OK  OK  GenBank in native format with EMBL CD-ROM index
tigregad      N    OK  OK  OK  TIGR EGAD sequences
tigrmgi       N    OK  OK  OK  TIGR MGI sequences
unclassified  N    OK  OK  OK  EMBL unclassified sequences
vertebrate    N    OK  OK  OK  EMBL vertebrate sequences
viral         N    OK  OK  OK  EMBL viral sequences

</pre></td></tr></table><p>
<p>
<p>
<p>
<b>Example 2</b>
<p>
Write the results to a file: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>showdb -outfile showdb.out </b>
Displays information on the currently available databases

</pre></td></tr></table><p>
<p>
<a href="#output.2">Go to the output files for this example</a><p><p>
<p>
<b>Example 3</b>
<p>
Display information on one explicit database: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>showdb -database tsw </b>
Displays information on the currently available databases

# Name        Type ID  Qry All Comment
# ====        ==== ==  === === =======
tsw           P    OK  OK  OK  Swissprot native format with EMBL CD-ROM index

</pre></td></tr></table><p>
<p>
<a href="#input.3">Go to the input files for this example</a><br><p>
<p>
<b>Example 4</b>
<p>
Display information on the databases formatted in HTML: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>showdb -html </b>
Displays information on the currently available databases

<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Name</th><th>Type</th><th>ID</th><th>Qry</th><th>All</th><th>Comment</th></tr>
<tr><td>enspep</td><td>P</td><td>OK  </td><td>OK  </td><td>OK  </td><td>ENSEMBL pep</td></tr>
<tr><td>enspepp</td><td>P</td><td>OK  </td><td>OK  </td><td>OK  </td><td>ENSEMBL pep predicted</td></tr>
<tr><td>enspeppredicted</td><td>P</td><td>OK  </td><td>OK  </td><td>OK  </td><td>ENSEMBL pep predicted</td></tr>
<tr><td>remtrembl</td><td>P</td><td>OK  </td><td>OK  </td><td>OK  </td><td>REMTREMBL sequences</td></tr>
<tr><td>sptrembl</td><td>P</td><td>OK  </td><td>OK  </td><td>OK  </td><td>SPTREMBL sequences</td></tr>
<tr><td>sw</td><td>P</td><td>OK  </td><td>OK  </td><td>OK  </td><td>SWISSPROT sequences</td></tr>
<tr><td>swissprot</td><td>P</td><td>OK  </td><td>OK  </td><td>OK  </td><td>SWISSPROT sequences</td></tr>
<tr><td>tpir</td><td>P</td><td>OK  </td><td>OK  </td><td>OK  </td><td>PIR using NBRF access for 4 files</td></tr>
<tr><td>trembl</td><td>P</td><td>OK  </td><td>OK  </td><td>OK  </td><td>TREMBL sequences</td></tr>
<tr><td>tsw</td><td>P</td><td>OK  </td><td>OK  </td><td>OK  </td><td>Swissprot native format with EMBL CD-ROM index</td></tr>
<tr><td>tswnew</td><td>P</td><td>OK  </td><td>OK  </td><td>OK  </td><td>Swissnew as 3 files in native format with EMBL CD-ROM index</td></tr>
<tr><td>twp</td><td>P</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL new in native format with EMBL CD-ROM index</td></tr>
<tr><td>em</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL sequences</td></tr>
<tr><td>embl</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL sequences</td></tr>
<tr><td>emblnew</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>New EMBL sequences</td></tr>
<tr><td>enscdna</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>ENSEMBL cdna</td></tr>
<tr><td>ensdna</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>ENSEMBL dna</td></tr>
<tr><td>ensdnam</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>ENSEMBL dna masked</td></tr>
<tr><td>ensdnamasked</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>ENSEMBL dna masked</td></tr>
<tr><td>ensembl</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>ENSEMBL </td></tr>
<tr><td>est</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL EST sequences</td></tr>
<tr><td>fungal</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL fungal sequences</td></tr>
<tr><td>gb</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>Genbank updates</td></tr>
<tr><td>genbank</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>Genbank</td></tr>
<tr><td>gss</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL GSS sequences</td></tr>
<tr><td>hsunigene</td><td>N</td><td>OK  </td><td>-   </td><td>-   </td><td>hsunigene </td></tr>
<tr><td>htg</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL HTG sequences</td></tr>
<tr><td>human</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL human sequences</td></tr>
<tr><td>humanest</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL human EST sequences</td></tr>
<tr><td>invertebrate</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL invertebrate sequences</td></tr>
<tr><td>mammalian</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL mammalian sequences</td></tr>
<tr><td>mouseest</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL mouse EST sequences</td></tr>
<tr><td>newgb</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>Genbank updates</td></tr>
<tr><td>newgenbank</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>Genbank updates</td></tr>
<tr><td>organelle</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL organelle sequences</td></tr>
<tr><td>otherest</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL other EST sequences</td></tr>
<tr><td>othervertebrate</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL other vertebrate sequences</td></tr>
<tr><td>patent</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL patented sequences</td></tr>
<tr><td>phage</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL phage sequences</td></tr>
<tr><td>plant</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL plant sequences</td></tr>
<tr><td>prokaryotic</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL prokaryotic sequences</td></tr>
<tr><td>refseq</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>REFSEQ sequences with features in /home/br2/data/Chris...</td></tr>
<tr><td>refseq_srs</td><td>N</td><td>OK  </td><td>-   </td><td>-   </td><td>REFSEQ-ID sequences from SRS</td></tr>
<tr><td>rodent</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL rodent sequences</td></tr>
<tr><td>sts</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL STS sequences</td></tr>
<tr><td>synthetic</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL synthetic sequences</td></tr>
<tr><td>tembl</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL in native format with EMBL CD-ROM index</td></tr>
<tr><td>tgb</td><td>N</td><td>OK  </td><td>-   </td><td>-   </td><td>Genbank IDs</td></tr>
<tr><td>tgenbank</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>GenBank in native format with EMBL CD-ROM index</td></tr>
<tr><td>tigregad</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>TIGR EGAD sequences</td></tr>
<tr><td>tigrmgi</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>TIGR MGI sequences</td></tr>
<tr><td>unclassified</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL unclassified sequences</td></tr>
<tr><td>vertebrate</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL vertebrate sequences</td></tr>
<tr><td>viral</td><td>N</td><td>OK  </td><td>OK  </td><td>OK  </td><td>EMBL viral sequences</td></tr>
</table>

</pre></td></tr></table><p>
<p>
<p>
<p>
<b>Example 5</b>
<p>
Display protein databases only: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>showdb -nonucleic </b>
Displays information on the currently available databases

# Name        Type ID  Qry All Comment
# ====        ==== ==  === === =======
enspep        P    OK  OK  OK  ENSEMBL pep
enspepp       P    OK  OK  OK  ENSEMBL pep predicted
enspeppredicted P    OK  OK  OK  ENSEMBL pep predicted
remtrembl     P    OK  OK  OK  REMTREMBL sequences
sptrembl      P    OK  OK  OK  SPTREMBL sequences
sw            P    OK  OK  OK  SWISSPROT sequences
swissprot     P    OK  OK  OK  SWISSPROT sequences
tpir          P    OK  OK  OK  PIR using NBRF access for 4 files
trembl        P    OK  OK  OK  TREMBL sequences
tsw           P    OK  OK  OK  Swissprot native format with EMBL CD-ROM index
tswnew        P    OK  OK  OK  Swissnew as 3 files in native format with EMBL CD-ROM index
twp           P    OK  OK  OK  EMBL new in native format with EMBL CD-ROM index

</pre></td></tr></table><p>
<p>
<p>
<p>
<b>Example 6</b>
<p>
Display the information with no headings: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>showdb -noheading </b>
Displays information on the currently available databases

enspep        P    OK  OK  OK  ENSEMBL pep
enspepp       P    OK  OK  OK  ENSEMBL pep predicted
enspeppredicted P    OK  OK  OK  ENSEMBL pep predicted
remtrembl     P    OK  OK  OK  REMTREMBL sequences
sptrembl      P    OK  OK  OK  SPTREMBL sequences
sw            P    OK  OK  OK  SWISSPROT sequences
swissprot     P    OK  OK  OK  SWISSPROT sequences
tpir          P    OK  OK  OK  PIR using NBRF access for 4 files
trembl        P    OK  OK  OK  TREMBL sequences
tsw           P    OK  OK  OK  Swissprot native format with EMBL CD-ROM index
tswnew        P    OK  OK  OK  Swissnew as 3 files in native format with EMBL CD-ROM index
twp           P    OK  OK  OK  EMBL new in native format with EMBL CD-ROM index
em            N    OK  OK  OK  EMBL sequences
embl          N    OK  OK  OK  EMBL sequences
emblnew       N    OK  OK  OK  New EMBL sequences
enscdna       N    OK  OK  OK  ENSEMBL cdna
ensdna        N    OK  OK  OK  ENSEMBL dna
ensdnam       N    OK  OK  OK  ENSEMBL dna masked
ensdnamasked  N    OK  OK  OK  ENSEMBL dna masked
ensembl       N    OK  OK  OK  ENSEMBL 
est           N    OK  OK  OK  EMBL EST sequences
fungal        N    OK  OK  OK  EMBL fungal sequences
gb            N    OK  OK  OK  Genbank updates
genbank       N    OK  OK  OK  Genbank
gss           N    OK  OK  OK  EMBL GSS sequences
hsunigene     N    OK  -   -   hsunigene 
htg           N    OK  OK  OK  EMBL HTG sequences
human         N    OK  OK  OK  EMBL human sequences
humanest      N    OK  OK  OK  EMBL human EST sequences
invertebrate  N    OK  OK  OK  EMBL invertebrate sequences
mammalian     N    OK  OK  OK  EMBL mammalian sequences
mouseest      N    OK  OK  OK  EMBL mouse EST sequences
newgb         N    OK  OK  OK  Genbank updates
newgenbank    N    OK  OK  OK  Genbank updates
organelle     N    OK  OK  OK  EMBL organelle sequences
otherest      N    OK  OK  OK  EMBL other EST sequences
othervertebrate N    OK  OK  OK  EMBL other vertebrate sequences
patent        N    OK  OK  OK  EMBL patented sequences
phage         N    OK  OK  OK  EMBL phage sequences
plant         N    OK  OK  OK  EMBL plant sequences
prokaryotic   N    OK  OK  OK  EMBL prokaryotic sequences
refseq        N    OK  OK  OK  REFSEQ sequences with features in /home/br2/data/Chris...
refseq_srs    N    OK  -   -   REFSEQ-ID sequences from SRS
rodent        N    OK  OK  OK  EMBL rodent sequences
sts           N    OK  OK  OK  EMBL STS sequences
synthetic     N    OK  OK  OK  EMBL synthetic sequences
tembl         N    OK  OK  OK  EMBL in native format with EMBL CD-ROM index
tgb           N    OK  -   -   Genbank IDs
tgenbank      N    OK  OK  OK  GenBank in native format with EMBL CD-ROM index
tigregad      N    OK  OK  OK  TIGR EGAD sequences
tigrmgi       N    OK  OK  OK  TIGR MGI sequences
unclassified  N    OK  OK  OK  EMBL unclassified sequences
vertebrate    N    OK  OK  OK  EMBL vertebrate sequences
viral         N    OK  OK  OK  EMBL viral sequences

</pre></td></tr></table><p>
<p>
<p>
<p>
<b>Example 7</b>
<p>
Display just a list of the available database names: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>showdb -noheading -notype -noid -noquery -noall -nocomment -auto </b>

enspep        
enspepp       
enspeppredicted 
remtrembl     
sptrembl      
sw            
swissprot     
tpir          
trembl        
tsw           
tswnew        
twp           
em            
embl          
emblnew       
enscdna       
ensdna        
ensdnam       
ensdnamasked  
ensembl       
est           
fungal        
gb            
genbank       
gss           
hsunigene     
htg           
human         
humanest      
invertebrate  
mammalian     
mouseest      
newgb         
newgenbank    
organelle     
otherest      
othervertebrate 
patent        
phage         
plant         
prokaryotic   
refseq        
refseq_srs    
rodent        
sts           
synthetic     
tembl         
tgb           
tgenbank      
tigregad      
tigrmgi       
unclassified  
vertebrate    
viral         

</pre></td></tr></table><p>
<p>
<p>
<p>
<b>Example 8</b>
<p>
Display only the names and types: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>showdb -only -type </b>
Displays information on the currently available databases

enspep        P    
enspepp       P    
enspeppredicted P    
remtrembl     P    
sptrembl      P    
sw            P    
swissprot     P    
tpir          P    
trembl        P    
tsw           P    
tswnew        P    
twp           P    
em            N    
embl          N    
emblnew       N    
enscdna       N    
ensdna        N    
ensdnam       N    
ensdnamasked  N    
ensembl       N    
est           N    
fungal        N    
gb            N    
genbank       N    
gss           N    
hsunigene     N    
htg           N    
human         N    
humanest      N    
invertebrate  N    
mammalian     N    
mouseest      N    
newgb         N    
newgenbank    N    
organelle     N    
otherest      N    
othervertebrate N    
patent        N    
phage         N    
plant         N    
prokaryotic   N    
refseq        N    
refseq_srs    N    
rodent        N    
sts           N    
synthetic     N    
tembl         N    
tgb           N    
tgenbank      N    
tigregad      N    
tigrmgi       N    
unclassified  N    
vertebrate    N    
viral         N    

</pre></td></tr></table><p>
<p>
<p>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers: (none)
   Additional (Optional) qualifiers:
   -database           string     Name of a single database to give
                                  information on
   -html               boolean    Format output as an HTML table
   -[no]protein        boolean    Display protein databases
   -[no]nucleic        boolean    Display nucleic acid databases
   -fields             boolean    This displays the search fields that can be
                                  used on this database, other than the
                                  standard 'id' or 'acc' fields.
   -release            boolean    Display 'release' column
   -outfile            outfile    Output file name

   Advanced (Unprompted) qualifiers:
   -only               boolean    This is a way of shortening the command line
                                  if you only want a few things to be
                                  displayed. Instead of specifying:
                                  '-nohead -notype -noid -noquery -noall'
                                  to get only the comment output, you can
                                  specify
                                  '-only -comment'
   -heading            boolean    Display column headings
   -type               boolean    Display 'type' column
   -id                 boolean    Display 'id' column
   -query              boolean    Display 'qry' column
   -all                boolean    Display 'all' column
   -comment            boolean    Display 'comment' column

   Associated qualifiers:

   "-outfile" associated qualifiers
   -odirectory          string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-database</td>
<td>Name of a single database to give information on</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-html</td>
<td>Format output as an HTML table</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]protein</td>
<td>Display protein databases</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]nucleic</td>
<td>Display nucleic acid databases</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-fields</td>
<td>This displays the search fields that can be used on this database, other than the standard 'id' or 'acc' fields.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-release</td>
<td>Display 'release' column</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-outfile</td>
<td>Output file name</td>
<td>Output file</td>
<td>stdout</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-only</td>
<td>This is a way of shortening the command line if you only want a few things to be displayed. Instead of specifying:
'-nohead -notype -noid -noquery -noall'
to get only the comment output, you can specify
'-only -comment'</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-heading</td>
<td>Display column headings</td>
<td>Boolean value Yes/No</td>
<td>@(!$(only))</td>
</tr>

<tr>
<td>-type</td>
<td>Display 'type' column</td>
<td>Boolean value Yes/No</td>
<td>@(!$(only))</td>
</tr>

<tr>
<td>-id</td>
<td>Display 'id' column</td>
<td>Boolean value Yes/No</td>
<td>@(!$(only))</td>
</tr>

<tr>
<td>-query</td>
<td>Display 'qry' column</td>
<td>Boolean value Yes/No</td>
<td>@(!$(only))</td>
</tr>

<tr>
<td>-all</td>
<td>Display 'all' column</td>
<td>Boolean value Yes/No</td>
<td>@(!$(only))</td>
</tr>

<tr>
<td>-comment</td>
<td>Display 'comment' column</td>
<td>Boolean value Yes/No</td>
<td>@(!$(only))</td>
</tr>

</table>



<H2>
    Input file format
</H2>

    None.

<p>


<a name="input.3"></a>
<h3>Input files for usage example 3</h3>

'tsw' is a sequence entry in the example protein database 'tsw'
<p>


<H2>
    Output file format
</H2>



<a name="output.2"></a>
<h3>Output files for usage example 2</h3>
<p><h3>File: showdb.out</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
# Name        Type ID  Qry All Comment
# ====        ==== ==  === === =======
enspep        P    OK  OK  OK  ENSEMBL pep
enspepp       P    OK  OK  OK  ENSEMBL pep predicted
enspeppredicted P    OK  OK  OK  ENSEMBL pep predicted
remtrembl     P    OK  OK  OK  REMTREMBL sequences
sptrembl      P    OK  OK  OK  SPTREMBL sequences
sw            P    OK  OK  OK  SWISSPROT sequences
swissprot     P    OK  OK  OK  SWISSPROT sequences
tpir          P    OK  OK  OK  PIR using NBRF access for 4 files
trembl        P    OK  OK  OK  TREMBL sequences
tsw           P    OK  OK  OK  Swissprot native format with EMBL CD-ROM index
tswnew        P    OK  OK  OK  Swissnew as 3 files in native format with EMBL CD-ROM index
twp           P    OK  OK  OK  EMBL new in native format with EMBL CD-ROM index
em            N    OK  OK  OK  EMBL sequences
embl          N    OK  OK  OK  EMBL sequences
emblnew       N    OK  OK  OK  New EMBL sequences
enscdna       N    OK  OK  OK  ENSEMBL cdna
ensdna        N    OK  OK  OK  ENSEMBL dna
ensdnam       N    OK  OK  OK  ENSEMBL dna masked
ensdnamasked  N    OK  OK  OK  ENSEMBL dna masked
ensembl       N    OK  OK  OK  ENSEMBL 
est           N    OK  OK  OK  EMBL EST sequences
fungal        N    OK  OK  OK  EMBL fungal sequences
gb            N    OK  OK  OK  Genbank updates
genbank       N    OK  OK  OK  Genbank
gss           N    OK  OK  OK  EMBL GSS sequences
hsunigene     N    OK  -   -   hsunigene 
htg           N    OK  OK  OK  EMBL HTG sequences
human         N    OK  OK  OK  EMBL human sequences
humanest      N    OK  OK  OK  EMBL human EST sequences
invertebrate  N    OK  OK  OK  EMBL invertebrate sequences
mammalian     N    OK  OK  OK  EMBL mammalian sequences
mouseest      N    OK  OK  OK  EMBL mouse EST sequences
newgb         N    OK  OK  OK  Genbank updates
newgenbank    N    OK  OK  OK  Genbank updates
organelle     N    OK  OK  OK  EMBL organelle sequences
otherest      N    OK  OK  OK  EMBL other EST sequences
othervertebrate N    OK  OK  OK  EMBL other vertebrate sequences
patent        N    OK  OK  OK  EMBL patented sequences
phage         N    OK  OK  OK  EMBL phage sequences
plant         N    OK  OK  OK  EMBL plant sequences
prokaryotic   N    OK  OK  OK  EMBL prokaryotic sequences
refseq        N    OK  OK  OK  REFSEQ sequences with features in /home/br2/data/Chris...
refseq_srs    N    OK  -   -   REFSEQ-ID sequences from SRS
rodent        N    OK  OK  OK  EMBL rodent sequences
sts           N    OK  OK  OK  EMBL STS sequences
synthetic     N    OK  OK  OK  EMBL synthetic sequences
tembl         N    OK  OK  OK  EMBL in native format with EMBL CD-ROM index
tgb           N    OK  -   -   Genbank IDs
tgenbank      N    OK  OK  OK  GenBank in native format with EMBL CD-ROM index
tigregad      N    OK  OK  OK  TIGR EGAD sequences
tigrmgi       N    OK  OK  OK  TIGR MGI sequences
unclassified  N    OK  OK  OK  EMBL unclassified sequences
vertebrate    N    OK  OK  OK  EMBL vertebrate sequences
viral         N    OK  OK  OK  EMBL viral sequences
</pre>
</td></tr></table><p>


<p>

The output is a simple table.

<p>

Type 'P' indicates that this is a Protein database.

<p>

Type 'N' indicates that this is a Nucleic database.

<p>

'OK' under ID, Qry or All indicates that that access method can be
used on this database.  A '-' indicates that you cannot access this
database in that way.

<p>

Note that 'OK' does not mean that the database is working correctly.  It
simply means that <b>showdb</b> has read the database definition
correctly and that this method of access to the database should be
possible. 

<p>

If you are setting up a new database, then you should check that it
works correctly by extracting entries from it using <b>seqret</b>. 

<p>

When the -html qualifier is specified, then the output will be wrapped
in HTML tags, ready for inclusion in a Web page.  Note that tags such
as &lt;HTML&gt;, &lt;BODY&gt;, &lt;/BODY&gt; and &lt;/HTML&gt; are not
output by this program as the table of databases is expected to form
only part of the contents of a web page - the rest of the web page
must be supplied by the user.


<H2>
    Data files
</H2>


The databases are specified in the files <i>"emboss.defaults"</i> for
site wide definitions, and <i>"~/.embossrc"</i> for the user's own
settings.
	
<H2>
    Notes
</H2>


Note that 'OK' in the output does not mean that the database is working
correctly.  It simply means that <b>showdb</b> has read the database
definition correctly and that this method of access to the database
should be possible. 

<p>

If you are setting up a new database, then you should check that it
works correctly by extracting entries from it using <b>seqret</b>. 

<p>

You can set up your own private databases by editing a file called
'.embossrc' to contain database specifications. 

<p>

You can set up public databases by editing the file
'/usr/local/share/EMBOSS/emboss.default' (if you have the permission to
do this). 

<p>

The 'emboss.default' file will already have the definitions of the test
databases in: tsw, tembl, tpir, etc.  These are the databases that are
used in the examples shown in the documentation of the programs. 

<p>

For details of the database definitions, you should first read David's
Administration Guide:
<a href="http://www.uk.embnet.org/Software/EMBOSS/Doc/Admin_guide/adminguide/">
http://www.uk.embnet.org/Software/EMBOSS/Doc/Admin_guide/adminguide/
</a>
especially:
<a href="http://www.uk.embnet.org/Software/EMBOSS/Doc/Admin_guide/adminguide/node4.html">
http://www.uk.embnet.org/Software/EMBOSS/Doc/Admin_guide/adminguide/node4.html
</a>

<p>

See also:
<a href="http://www.uk.embnet.org/Software/EMBOSS/Usa/databases.html">
http://www.uk.embnet.org/Software/EMBOSS/Usa/databases.html
</a>
for a lot of detail of the syntax of database definition.

<p>

Just because showdb can display the database definitions, it does NOT
mean that the databases are set up correctly.  You must now test them
using <b>seqret</b>. 


<H2>
    References
</H2>


	None.

<H2>
    Warnings
</H2>


	None.

<H2>
    Diagnostic Error Messages
</H2>


"The database 'xyz' does not exist" You have supplied the name of a
database with the -database qualifier, but that database does not
exist as far as EMBOSS is concerned.

<H2>
    Exit status
</H2>


It always exits with status 0, unless the above diagnostic message is
displayed.

<H2>
    Known bugs
</H2>


        None noted.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="abiview.html">abiview</a></td><td>Reads ABI file and display the trace</td></tr>
<tr><td><a href="cirdna.html">cirdna</a></td><td>Draws circular maps of DNA constructs</td></tr>
<tr><td><a href="infoalign.html">infoalign</a></td><td>Information on a multiple sequence alignment</td></tr>
<tr><td><a href="infoseq.html">infoseq</a></td><td>Displays some simple information about sequences</td></tr>
<tr><td><a href="lindna.html">lindna</a></td><td>Draws linear maps of DNA constructs</td></tr>
<tr><td><a href="pepnet.html">pepnet</a></td><td>Displays proteins as a helical net</td></tr>
<tr><td><a href="pepwheel.html">pepwheel</a></td><td>Shows protein sequences as helices</td></tr>
<tr><td><a href="prettyplot.html">prettyplot</a></td><td>Displays aligned sequences, with colouring and boxing</td></tr>
<tr><td><a href="prettyseq.html">prettyseq</a></td><td>Output sequence with translated ranges</td></tr>
<tr><td><a href="remap.html">remap</a></td><td>Display a sequence with restriction cut sites, translation etc</td></tr>
<tr><td><a href="seealso.html">seealso</a></td><td>Finds programs sharing group names</td></tr>
<tr><td><a href="showalign.html">showalign</a></td><td>Displays a multiple sequence alignment</td></tr>
<tr><td><a href="showfeat.html">showfeat</a></td><td>Show features of a sequence</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="sixpack.html">sixpack</a></td><td>Display a DNA sequence with 6-frame translation and ORFs</td></tr>
<tr><td><a href="textsearch.html">textsearch</a></td><td>Search sequence documentation text. SRS and Entrez are faster!</td></tr>
<tr><td><a href="tfm.html">tfm</a></td><td>Displays a program's help documentation manual</td></tr>
<tr><td><a href="whichdb.html">whichdb</a></td><td>Search all databases for an entry</td></tr>
<tr><td><a href="wossname.html">wossname</a></td><td>Finds programs by keywords in their one-line documentation</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


Gary Williams (gwilliam&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK




<H2>
    History
</H2>


Completed 6th August 1999.

<!--
<p>
[an error occurred while processing this directive]
-->

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

