<HTML>

<HEAD>
  <TITLE>
  EMBOSS: showfeat
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: showfeat
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program showfeat
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Show features of a sequence


<H2>
    Description
</H2>



Showfeat reads a sequence and a feature table, and writes a text
representation of the features to standard output.

<H2>
    Usage
</H2>


Here is a sample session with showfeat. The feature table is specified
as a -ufo qualifier (uniform feature object), in this case a file
containing an EMBL feature table.

<pre>

% showfeat -ufo paamir.ft
Show features of a sequence.
Input sequence(s): embl:paamir
Output file [stdout]: 
PAAMIR
Pseudomonas aeruginosa amiC and amiR gene for aliphatic amidase regulation
|==========================================================| 2167
|----------------------------------------------------------&gt; source
&gt;                                                            promoter
|&gt;                                                           promoter
  &gt;                                                          RBS
  |-------------------------------&gt;                          CDS
                |&gt;                                           misc_feature
                        |------&gt;                             mutation
                                  |----------------&gt;         CDS
|                                                            misc_feature
                                  |                          conflict

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
  [-outfile]           outfile    If you enter the name of a file here then
                                  this program will write the feature details
                                  into that file.

   Optional qualifiers:
   -matchsource        string     By default any feature source in the feature
                                  table is shown. You can set this to match
                                  any feature source you wish to show.
                                  The source name is usually either the name
                                  of the program that detected the feature or
                                  it is the feature table (eg: EMBL) that the
                                  feature came from.
                                  The source may be wildcarded by using '*'.
                                  If you wish to show more than one source,
                                  separate their names with the character '|',
                                  eg:
                                  gene* | embl
   -matchtype          string     By default any feature type in the feature
                                  table is shown. You can set this to match
                                  any feature type you wish to show.
                                  See http://www3.ebi.ac.uk/Services/WebFeat/
                                  for a list of the EMBL feature types and see
                                  Appendix A of the Swissprot user manual in
                                  http://www.expasy.ch/txt/userman.txt for a
                                  list of the Swissprot feature types.
                                  The type may be wildcarded by using '*'.
                                  If you wish to show more than one type,
                                  separate their names with the character '|',
                                  eg:
                                  *UTR | intron
   -matchtag           string     Tags are the types of extra values that a
                                  feature may have. For example in the EMBL
                                  feature table, a 'CDS' type of feature may
                                  have the tags '/codon', '/codon_start',
                                  '/db_xref', '/EC_number', '/evidence',
                                  '/exception', '/function', '/gene',
                                  '/label', '/map', '/note', '/number',
                                  '/partial', '/product', '/protein_id',
                                  '/pseudo', '/standard_name', '/translation',
                                  '/transl_except', '/transl_table', or
                                  '/usedin'. Some of these tags also have
                                  values, for example '/gene' can have the
                                  value of the gene name.
                                  By default any feature tag in the feature
                                  table is shown. You can set this to match
                                  any feature tag you wish to show.
                                  The tag may be wildcarded by using '*'.
                                  If you wish to show more than one tag,
                                  separate their names with the character '|',
                                  eg:
                                  gene | label
   -matchvalue         string     Tag values are the values associated with a
                                  feature tag. Tags are the types of extra
                                  values that a feature may have. For example
                                  in the EMBL feature table, a 'CDS' type of
                                  feature may have the tags '/codon',
                                  '/codon_start', '/db_xref', '/EC_number',
                                  '/evidence', '/exception', '/function',
                                  '/gene', '/label', '/map', '/note',
                                  '/number', '/partial', '/product',
                                  '/protein_id', '/pseudo', '/standard_name',
                                  '/translation', '/transl_except',
                                  '/transl_table', or '/usedin'. Only some of
                                  these tags can have values, for example
                                  '/gene' can have the value of the gene name.
                                  By default any feature tag value in the
                                  feature table is shown. You can set this to
                                  match any feature tag valueyou wish to show.
                                  The tag value may be wildcarded by using
                                  '*'.
                                  If you wish to show more than one tag value,
                                  separate their names with the character
                                  '|', eg:
                                  pax* | 10
   -sort               list       Sort features by Type, Start or Source

   Advanced qualifiers:
   -html               bool       Use HTML formatting
   -[no]id             bool       Set this to be false if you do not wish to
                                  display the ID name of the sequence.
   -[no]description    bool       Set this to be false if you do not wish to
                                  display the description of the sequence.
   -[no]scale          bool       Set this to be false if you do not wish to
                                  display the scale line.
   -width              integer    You can expand (or contract) the width of
                                  the ASCII-character graphics display of the
                                  positions of the features using this value.
                                  For example, a width of 80 characters would
                                  cover a standard page width and a width a 10
                                  characters would be nearly unreadable.
                                  If the width is set to less than 4, the
                                  graphics lines and the scale line will not
                                  be displayed.
   -collapse           bool       If this is set, then features from the same
                                  source and of the same type and sense are
                                  all printed on the same line. For instance
                                  if there are several features from the EMBL
                                  feature table (ie. the same source) which
                                  are all of type 'exon' in the same sense,
                                  then they will all be displayed on the same
                                  line. This makes it hard to distinguish
                                  overlapping features.
                                  If this is set to false then each feature is
                                  displayed on a separate line making it
                                  easier to distinguish where features start
                                  and end.
   -[no]forward        bool       Set this to be false if you do not wish to
                                  display forward sense features.
   -[no]reverse        bool       Set this to be false if you do not wish to
                                  display reverse sense features.
   -[no]unknown        bool       Set this to be false if you do not wish to
                                  display unknown sense features. (ie.
                                  features with no directionality - all
                                  protein features are of this type and some
                                  nucleic features (for example, CG-rich
                                  regions)).
   -strand             bool       Set this if you wish to display the strand
                                  of the features. Protein features are always
                                  directionless (indicated by '0'), forward
                                  is indicated by '+' and reverse is '-'.
   -source             bool       Set this if you wish to display the source
                                  of the features.
                                  The source name is usually either the name
                                  of the program that detected the feature or
                                  it is the name of the feature table (eg:
                                  EMBL) that the feature came from.
   -position           bool       Set this if you wish to display the start
                                  and end position of the features. If several
                                  features are being displayed on the same
                                  line, then the start and end positions will
                                  be joined by a comma, for example:
                                  '189-189,225-225'.
   -[no]type           bool       Set this to be false if you do not wish to
                                  display the type of the features.
   -tags               bool       Set this to be false if you do not wish to
                                  display the tags and values of the features.
   -[no]values         bool       Set this to be false if you do not wish to
                                  display the tag values of the features. If
                                  this is set to be false, only the tag names
                                  will be displayed. If the tags are not
                                  displayed, then the values will not be
                                  displayed. The value of the 'translation'
                                  tag is never displayed as it is often
                                  extremely long.


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>If you enter the name of a file here then this program will write the feature details into that file.</td>
<td>Output file</td>
<td>stdout</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-matchsource</td>
<td>By default any feature source in the feature table is shown. You can set this to match any feature source you wish to show.
The source name is usually either the name of the program that detected the feature or it is the feature table (eg: EMBL) that the feature came from.
The source may be wildcarded by using '*'.
If you wish to show more than one source, separate their names with the character '|', eg:
gene* | embl</td>
<td>Any string is accepted</td>
<td>*</td>
</tr>

<tr>
<td>-matchtype</td>
<td>By default any feature type in the feature table is shown. You can set this to match any feature type you wish to show.
See http://www3.ebi.ac.uk/Services/WebFeat/ for a list of the EMBL feature types and see Appendix A of the Swissprot user manual in http://www.expasy.ch/txt/userman.txt for a list of the Swissprot feature types.
The type may be wildcarded by using '*'.
If you wish to show more than one type, separate their names with the character '|', eg:
*UTR | intron</td>
<td>Any string is accepted</td>
<td>*</td>
</tr>

<tr>
<td>-matchtag</td>
<td>Tags are the types of extra values that a feature may have. For example in the EMBL feature table, a 'CDS' type of feature may have the tags '/codon', '/codon_start', '/db_xref', '/EC_number', '/evidence', '/exception', '/function', '/gene', '/label', '/map', '/note', '/number', '/partial', '/product', '/protein_id', '/pseudo', '/standard_name', '/translation', '/transl_except', '/transl_table', or '/usedin'. Some of these tags also have values, for example '/gene' can have the value of the gene name.
By default any feature tag in the feature table is shown. You can set this to match any feature tag you wish to show.
The tag may be wildcarded by using '*'.
If you wish to show more than one tag, separate their names with the character '|', eg:
gene | label</td>
<td>Any string is accepted</td>
<td>*</td>
</tr>

<tr>
<td>-matchvalue</td>
<td>Tag values are the values associated with a feature tag. Tags are the types of extra values that a feature may have. For example in the EMBL feature table, a 'CDS' type of feature may have the tags '/codon', '/codon_start', '/db_xref', '/EC_number', '/evidence', '/exception', '/function', '/gene', '/label', '/map', '/note', '/number', '/partial', '/product', '/protein_id', '/pseudo', '/standard_name', '/translation', '/transl_except', '/transl_table', or '/usedin'. Only some of these tags can have values, for example '/gene' can have the value of the gene name. By default any feature tag value in the feature table is shown. You can set this to match any feature tag valueyou wish to show.
The tag value may be wildcarded by using '*'.
If you wish to show more than one tag value, separate their names with the character '|', eg:
pax* | 10</td>
<td>Any string is accepted</td>
<td>*</td>
</tr>

<tr>
<td>-sort</td>
<td>Sort features by Type, Start or Source</td>
<td><table><tr><td>source</td> <td><i>(Sort by Source)</i></td></tr><tr><td>start</td> <td><i>(Sort by Start position)</i></td></tr><tr><td>type</td> <td><i>(Sort by Type)</i></td></tr></table></td>
<td>start</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-html</td>
<td>Use HTML formatting</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]id</td>
<td>Set this to be false if you do not wish to display the ID name of the sequence.</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]description</td>
<td>Set this to be false if you do not wish to display the description of the sequence.</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]scale</td>
<td>Set this to be false if you do not wish to display the scale line.</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-width</td>
<td>You can expand (or contract) the width of the ASCII-character graphics display of the positions of the features using this value.
For example, a width of 80 characters would cover a standard page width and a width a 10 characters would be nearly unreadable.
If the width is set to less than 4, the graphics lines and the scale line will not be displayed.</td>
<td>Integer 0 or more</td>
<td>60</td>
</tr>

<tr>
<td>-collapse</td>
<td>If this is set, then features from the same source and of the same type and sense are all printed on the same line. For instance if there are several features from the EMBL feature table (ie. the same source) which are all of type 'exon' in the same sense, then they will all be displayed on the same line. This makes it hard to distinguish overlapping features.
If this is set to false then each feature is displayed on a separate line making it easier to distinguish where features start and end.</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]forward</td>
<td>Set this to be false if you do not wish to display forward sense features.</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]reverse</td>
<td>Set this to be false if you do not wish to display reverse sense features.</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]unknown</td>
<td>Set this to be false if you do not wish to display unknown sense features. (ie. features with no directionality - all protein features are of this type and some nucleic features (for example, CG-rich regions)).</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-strand</td>
<td>Set this if you wish to display the strand of the features. Protein features are always directionless (indicated by '0'), forward is indicated by '+' and reverse is '-'.</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-source</td>
<td>Set this if you wish to display the source of the features.
The source name is usually either the name of the program that detected the feature or it is the name of the feature table (eg: EMBL) that the feature came from.</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-position</td>
<td>Set this if you wish to display the start and end position of the features. If several features are being displayed on the same line, then the start and end positions will be joined by a comma, for example: '189-189,225-225'.</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]type</td>
<td>Set this to be false if you do not wish to display the type of the features.</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-tags</td>
<td>Set this to be false if you do not wish to display the tags and values of the features.</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]values</td>
<td>Set this to be false if you do not wish to display the tag values of the features. If this is set to be false, only the tag names will be displayed. If the tags are not displayed, then the values will not be displayed. The value of the 'translation' tag is never displayed as it is often extremely long.</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

</table>



<H2>
    Input file format
</H2>


The feature input is the feature table lines from the database entry
in EMBL format.

<p>
Here is the file:

<pre>

FH   Key             Location/Qualifiers
FH
FT   source          1..2167
FT                   /db_xref="taxon:287"
FT                   /organism="Pseudomonas aeruginosa"
FT                   /strain="PAC"
FT                   /isolate="PAC 1"
FT                   /map="38 min"
FT   CDS             1289..1879
FT                   /db_xref="SWISS-PROT:P10932"
FT                   /note="aliphatic amidase regulator, positive  regulator of
FT                   amiE"
FT                   /transl_table=11
FT                   /gene="amiR"
FT                   /protein_id="CAA32023.1"
FT                   /translation="MSANSLLGSLRELQVLVLNPPGEVSDALVLQLIRIGCSVRQCWPP
FT                   PEAFDVPVDVVFTSIFQNGHHDEIAALLAAGTPRTTLVALVEYESPAVLSQIIELECHG
FT                   VITQPLDAHRVLPVLVSARRISEEMAKLKQKTEQLQDRIAGQARINQAKVLLMQRHGWD
FT                   EREAHQHLSREAMKRREPILKIAQELLGNEPSA"
FT   CDS             135..1292
FT                   /db_xref="SWISS-PROT:P27017"
FT                   /note="negative regulator of amiR"
FT                   /transl_table=11
FT                   /gene="amiC"
FT                   /protein_id="CAA32024.1"
FT                   /translation="MGSHQERPLIGLLFSETGVTADIERSHAYGALLAVEQLNREGGVG
FT                   GRPIETLSQDPGGDPDRYRLCAEDFIRNRGVRFLVGCYMSHTRKAVMPVVERADALLCY
FT                   PTPYEGFEYSPNIVYGGPAPNQNSAPLAAYLIRHYGERVVFIGSDYIYPRESNHVMRHL
FT                   YRQHGGTVLEEIYIPLYPSDDDLQRAVERIYQARADVVFSTVVGTGTAELYRAIARRYG
FT                   DGRRPPIASLTTSEAEVAKMESDVAEGQVVVAPYFSSIDTPASRAFVQACHGFFPENAT
FT                   ITAWAEAAYWQTLLLGRAAQAAGNWRVEDVQRHLYDIDIDAPQGPVRVERQNNHSRLSS
FT                   RIAEIDARGVFQVRWQSPEPIRPDPYVVVHNLDDWSASMGGGPLP"
FT   promoter        8..24
FT                   /note="proposed rpoN-dependent promoter"
FT   promoter        65..81
FT                   /note="proposed rpoN-dependent promoter"
FT   RBS             121..126
FT                   /note="proposed Shine-Dalgarno sequence"
FT   mutation        912..1167
FT                   /note="ClaI fragment deleted in pSW36,  constitutive
FT                   phenotype"
FT                   /replace=""
FT                   /gene="amiC"
FT   misc_feature    1
FT                   /note="last base of an XhoI site"
FT   misc_feature    648..653
FT                   /note="end of 658bp XhoI fragment, deletion in  pSW3 causes
FT                   constitutive expression of amiE"
FT   conflict        1281
FT                   /replace="g"
FT                   /citation=[3]
</pre>

<H2>
    Output file format
</H2>


The output is a text representation of the feature table, and goes by
default to standard output (the terminal).

<p>

Here is the output from the example run:

<pre>

PAAMIR
Pseudomonas aeruginosa amiC and amiR gene for aliphatic amidase regulation
|==========================================================| 2167
|----------------------------------------------------------&gt; source
&gt;                                                            promoter
|&gt;                                                           promoter
  &gt;                                                          RBS
  |-------------------------------&gt;                          CDS
                |&gt;                                           misc_feature
                        |------&gt;                             mutation
                                  |----------------&gt;         CDS
|                                                            misc_feature
                                  |                          conflict

</pre>

<H2>
    Data files
</H2>


<H2>
    Notes
</H2>


<H2>
    References
</H2>


<H2>
    Warnings
</H2>


<H2>
    Diagnostic Error Messages
</H2>


<H2>
    Exit status
</H2>


<H2>
    Known bugs
</H2>


<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="cirdna.html">cirdna</a></td><td>Draws circular maps of DNA constructs</td></tr>
<tr><td><a href="lindna.html">lindna</a></td><td>Draws linear maps of DNA constructs</td></tr>
<tr><td><a href="remap.html">remap</a></td><td>Display a sequence with restriction cut sites, translation etc</td></tr>
<tr><td><a href="seqret.html">seqret</a></td><td>Reads and writes (returns) sequences</td></tr>
<tr><td><a href="seqretall.html">seqretall</a></td><td>Reads and writes (returns) a set of sequences one at a time</td></tr>
<tr><td><a href="seqretset.html">seqretset</a></td><td>Reads and writes (returns) a set of sequences all at once</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Gary Williams <a
href="mailto:gwilliam@hgmp.mrc.ac.uk">(gwilliam@hgmp.mrc.ac.uk)</a>


<H2>
    History
</H2>


<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

