<HTML>

<HEAD>
  <TITLE>
  EMBOSS: sigcleave
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">







<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

sigcleave
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>







<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Reports protein signal cleavage sites


<H2>
    Description
</H2>


SigCleave uses the method of von Heijne as modified by von Heijne in
his later book where treatment of positions -1 and -3 in the matrix is
slightly altered (see references).

<H2>
    Usage
</H2>


Here is a sample session with sigcleave.

<pre>

% sigcleave
Reports peptide signal cleavage sites
Input sequence: sw:ach2_drome
Output file [ach2_drome.out]: 
Minimum weight [3.5]: 

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
   -minweight          float      Minimum scoring weight value for the
                                  predicted cleavage site
  [-outfile]           report     Output report file name

   Optional qualifiers:
   -prokaryote         boolean    Specifies the sequence is prokaryotic and
                                  changes the default scoring data file name

   Advanced qualifiers:
   -pval               integer    Specifies the number of columns before the
                                  residue at the cleavage site in the weight
                                  matrix table
   -nval               integer    specifies the number of columns after the
                                  residue at the cleavage site in the weight
                                  matrix table

   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-minweight</td>
<td>Minimum scoring weight value for the predicted cleavage site</td>
<td>Number from 0.000 to 100.000</td>
<td>3.5</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output report file name</td>
<td>Report file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-prokaryote</td>
<td>Specifies the sequence is prokaryotic and changes the default scoring data file name</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-pval</td>
<td>Specifies the number of columns before the residue at the cleavage site in the weight matrix table</td>
<td>Integer from -13 to -1</td>
<td>-13</td>
</tr>

<tr>
<td>-nval</td>
<td>specifies the number of columns after the residue at the cleavage site in the weight matrix table</td>
<td>Integer 1 or more</td>
<td>Pval+15 (2)</td>
</tr>

</table>



<H2>
    Input file format
</H2>


The input sequence can be one or more protein sequences.

<H2>
    Output file format
</H2>


<p>

The output is a standard EMBOSS report file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-rformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  The available format names are: embl,
genbank, gff, pir, swiss, trace, listfile, dbmotif, diffseq, excel,
feattable, motif, regions, seqtable, simple, srs, table, tagseq

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html</A>
for further information on report formats.

<p>


<p>

By default <b>sigcleave</b> writes a 'motif' report file.

<p>

The output from the above example is:

<p>

<pre>
<hr>
########################################
# Program: sigcleave
# Rundate: Mon Feb 11 13:50:56 2002
# Report_file: ach2_drome.sig
########################################

#=======================================
#
# Sequence: ACH2_DROME     from: 1   to: 576
# HitCount: 9
#
# Reporting scores over 3.50
#
#=======================================

(1) Score 13.739 length 13 at residues 29->41
 Sequence: LLVLLLLCETVQA
           |           |
          29           41
 mature_peptide: NPDAKRLYDDLLSNYNRLIRPVSNNTDTVLVKLGLRLSQLIDLNLKDQIL

(2) Score 3.632 length 13 at residues 308->320
 Sequence: LLISEIIPSTSLA
           |           |
         308           320
 mature_peptide: LPLLGKYLLFTMLLVGLSVVITIIILNIHYRKPSTHKMRPWIRSFFIKRL

(3) Score 3.751 length 13 at residues 527->539
 Sequence: LFLWLFMIASLVG
           |           |
         527           539
 mature_peptide: TFVILGEAPSLYDDTKAIDVQLSDVAKQIYNLTEKKN

(4) Score 4.026 length 13 at residues 31->43
 Sequence: VLLLLCETVQANP
           |           |
          31           43
 mature_peptide: DAKRLYDDLLSNYNRLIRPVSNNTDTVLVKLGLRLSQLIDLNLKDQILTT

(5) Score 5.057 length 13 at residues 24->36
 Sequence: KPLCLLLVLLLLC
           |           |
          24           36
 mature_peptide: ETVQANPDAKRLYDDLLSNYNRLIRPVSNNTDTVLVKLGLRLSQLIDLNL

(6) Score 6.981 length 13 at residues 330->342
 Sequence: FTMLLVGLSVVIT
           |           |
         330           342
 mature_peptide: IIILNIHYRKPSTHKMRPWIRSFFIKRLPKLLLMRVPKDLLRDLAANKIN

(7) Score 7.360 length 13 at residues 528->540
 Sequence: FLWLFMIASLVGT
           |           |
         528           540
 mature_peptide: FVILGEAPSLYDDTKAIDVQLSDVAKQIYNLTEKKN

(8) Score 10.465 length 13 at residues 28->40
 Sequence: LLLVLLLLCETVQ
           |           |
          28           40
 mature_peptide: ANPDAKRLYDDLLSNYNRLIRPVSNNTDTVLVKLGLRLSQLIDLNLKDQI

(9) Score 12.135 length 13 at residues 26->38
 Sequence: LCLLLVLLLLCET
           |           |
          26           38
 mature_peptide: VQANPDAKRLYDDLLSNYNRLIRPVSNNTDTVLVKLGLRLSQLIDLNLKD


#---------------------------------------
#---------------------------------------       
<hr>
</pre>

<H2>
    Data files
</H2>


<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by EMBOSS environment variable EMBOSS_DATA.

<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>

The data file names are:

<ul>
   <li> Esig.euk          Eukaryotic signal data
   <li> Esig.pro          Prokaryotic signal data
</ul>

Here is the default file for eukaryotic signals:

<p>

<pre>
# Amino acid counts for 161 Eukaryotic Signal Peptides,
# from von Heijne (1986), Nucl. Acids. Res. 14:4683-4690
#
# The cleavage site is between +1 and -1
#
Sample: 161 aligned sequences
#
# R -13 -12 -11 -10  -9  -8  -7  -6  -5  -4  -3  -2  -1  +1  +2 Expect
# - --- --- --- --- --- --- --- --- --- --- --- --- --- --- --- ------
  A  16  13  14  15  20  18  18  17  25  15  47   6  80  18   6  14.5
  C   3   6   9   7   9  14   6   8   5   6  19   3   9   8   3   4.5
  D   0   0   0   0   0   0   0   0   5   3   0   5   0  10  11   8.9
  E   0   0   0   1   0   0   0   0   3   7   0   7   0  13  14  10.0
  F  13   9  11  11   6   7  18  13   4   5   0  13   0   6   4   5.6
  G   4   4   3   6   3  13   3   2  19  34   5   7  39  10   7  12.1
  H   0   0   0   0   0   1   1   0   5   0   0   6   0   4   2   3.4
  I  15  15   8   6  11   5   4   8   5   1  10   5   0   8   7   7.4
  K   0   0   0   1   0   0   1   0   0   4   0   2   0  11   9  11.3
  L  71  68  72  79  78  45  64  49  10  23   8  20   1   8   4  12.1
  M   0   3   7   4   1   6   2   2   0   0   0   1   0   1   2   2.7
  N   0   1   0   1   1   0   0   0   3   3   0  10   0   4   7   7.1
  P   2   0   2   0   0   4   1   8  20  14   0   1   3   0  22   7.4
  Q   0   0   0   1   0   6   1   0  10   8   0  18   3  19  10   6.3
  R   2   0   0   0   0   1   0   0   7   4   0  15   0  12   9   7.6
  S   9   3   8   6  13  10  15  16  26  11  23  17  20  15  10  11.4
  T   2  10   5   4   5  13   7   7  12   6  17   8   6   3  10   9.7
  V  20  25  15  18  13  15  11  27   0  12  32   3   0   8  17  11.1
  W   4   3   3   1   1   2   6   3   1   3   0   9   0   2   0   1.8
  Y   0   1   4   0   0   1   3   1   1   2   0   5   0   1   7   5.6
</pre>

<H2>
    Notes
</H2>


The value of minweight should be at least 3.5. At this level,
the method should correctly identify 95% of signal peptides, and
reject 95% of non-signal peptides. The cleavage site should be
correctly predicted in 75-80% of cases.
<p>

If you use matrix tables with a different number of residues before or
after the cleavage site, you must also set the advanced parameters nval
and pval. 


<H2>
    References
</H2>

<ol>

<li>
von Heijne, G. Nucleic Acids Res.: 14:4683 (1986)


<li>
von Heijne, G. "Sequence Analysis in Molecular Biology: Treasure Trove or
  Trivial Pursuit" (Acad. Press, (1987), 113-117)

</ol>

<H2>
    Warnings
</H2>


The program will warn you if a nucleic acid sequence is given or if the
data file is not mathematically accurate.

<H2>
    Diagnostic Error Messages
</H2>


<H2>
    Exit status
</H2>


It exits with status 0 unless an error is reported.

<H2>
    Known bugs
</H2>


None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="antigenic.html">antigenic</a></td><td>Finds antigenic sites in proteins</td></tr>
<tr><td><a href="digest.html">digest</a></td><td>Protein proteolytic enzyme or reagent cleavage digest</td></tr>
<tr><td><a href="fuzzpro.html">fuzzpro</a></td><td>Protein pattern search</td></tr>
<tr><td><a href="fuzztran.html">fuzztran</a></td><td>Protein pattern search after translation</td></tr>
<tr><td><a href="helixturnhelix.html">helixturnhelix</a></td><td>Report nucleic acid binding motifs</td></tr>
<tr><td><a href="oddcomp.html">oddcomp</a></td><td>Finds protein sequence regions with a biased composition</td></tr>
<tr><td><a href="patmatdb.html">patmatdb</a></td><td>Search a protein sequence with a motif</td></tr>
<tr><td><a href="patmatmotifs.html">patmatmotifs</a></td><td>Search a PROSITE motif database with a protein sequence</td></tr>
<tr><td><a href="pepcoil.html">pepcoil</a></td><td>Predicts coiled coil regions</td></tr>
<tr><td><a href="preg.html">preg</a></td><td>Regular expression search of a protein sequence</td></tr>
<tr><td><a href="pscan.html">pscan</a></td><td>Scans proteins using PRINTS</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>

<p>
Original program "SIGCLEAVE" by Peter Rice (EGCG 1989)

<H2>
    History
</H2>


Completed 10th March 1999

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

