<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: siggen
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">
<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>
<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>
</td>
<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: siggen
</H2>
</font>
</td>
</tr>
</table>
</td>
</tr>
</table>
<p><hr><p>

<!--END OF HEADER-->

<H1>
Program siggen
</H1>
<H2>
    Function
</H2>
Generates a sparse protein signature

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
        Algorithms
  -->

<b>siggen</b> parses a multiple structure alignment generated by the
EMBOSS application <b>scopalign</b> and corresponding files of residue contact
data generated by the EMBOSS application <b>contacts</b> and generates a
protein signature of a specified sparsity. 

<p>

Each position in the alignment is scored on the basis of a single or any
combination of up to 3 scoring schemes.  A signature of, for example,
10% sparsity would include data from the top 10% highest scoring
alignment positions. 

<p>

The resulting protein signature file is used by the application
<b>sigscan</b> to find examples of the signature in other proteins. 


<h3>Signatures</h3>


Signatures extend the comcept of the motif as a tool for characterizing
protein families.  They consist of a set of N key residue postitions
(A1, A2 ...An) preceeded by gaps (G) thus G1A1G2A2...GnAn.  Both a
residue and a gap can be variable.  A signature is matched to a protein
sequence and scored using a dynamic programming algorithm which permits
variability in gap distance and residue type.  Generating a signature
involves identifying residues associated with points of contact in
interactions between secondary structure alements.  A raw signature
consists of a set of positions with potential key structural roles
sampled from a sequence alignment constructed with reference to this
contact data.  Raw signatures are refined by samplinfg different
gap-residue pairs until the specificity of a signature for the family
cannot be further improved. 






<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->


Here is a sample session with siggen:

<pre>

% siggen
Generates a sparse protein signature
Location of alignment files for input [./]: ./jontest
Extension of alignment files for input [.align]:
Location of contact files for input [./]: ./jontest
Extension of contact files [.con]:
% sparsity of signature [10]:
Generate a randomized signature [N]:
Substitution matrix to be used [./EBLOSUM62]:
Score alignment on basis of residue conservation [Y]:
Score alignment on basis of number of contacts [Y]:
Score alignment on basis of conservation of contacts [Y]: N
Score alignment on a combined measure of number and conservation of contacts [N]:
Ignore alignment postitions with post_similar value of 0 [Y]:
Name of signature file for output [sig.sig]:
</pre>














<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-algpath]           string     Location of alignment files for input
  [-algextn]           string     Extension of alignment files for input
  [-conpath]           string     Location of contact files for input
  [-conextn]           string     Extension of contact files
  [-sparsity]          integer    % sparsity of signature
  [-randomize]         bool       Generate a randomized signature
  [-datafile]          matrixf    Substitution matrix to be used
  [-[no]scoreseq]      bool       Score alignment on basis of residue
                                  conservation
  [-[no]scorencon]     bool       Score alignment on basis of number of
                                  contacts
  [-[no]scoreccon]     bool       Score alignment on basis of conservation of
                                  contacts
  [-scoreboth]         bool       Score alignment on a combined measure of
                                  number and conservation of contacts
  [-[no]postsim]       bool       Ignore alignment postitions with
                                  post_similar value of 0
  [-sigfile]           outfile    Name of signature file for output

   Optional qualifiers: (none)
   Advanced qualifiers: (none)
   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-algpath]<br>(Parameter 1)</td>
<td>Location of alignment files for input</td>
<td>Any string is accepted</td>
<td>./</td>
</tr>

<tr>
<td>[-algextn]<br>(Parameter 2)</td>
<td>Extension of alignment files for input</td>
<td>Any string is accepted</td>
<td>.align</td>
</tr>

<tr>
<td>[-conpath]<br>(Parameter 3)</td>
<td>Location of contact files for input</td>
<td>Any string is accepted</td>
<td>./</td>
</tr>

<tr>
<td>[-conextn]<br>(Parameter 4)</td>
<td>Extension of contact files</td>
<td>Any string is accepted</td>
<td>.con</td>
</tr>

<tr>
<td>[-sparsity]<br>(Parameter 5)</td>
<td>% sparsity of signature</td>
<td>Any integer value</td>
<td>10</td>
</tr>

<tr>
<td>[-randomize]<br>(Parameter 6)</td>
<td>Generate a randomized signature</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>[-datafile]<br>(Parameter 7)</td>
<td>Substitution matrix to be used</td>
<td>Comparison matrix file in EMBOSS data path</td>
<td>./EBLOSUM62</td>
</tr>

<tr>
<td>[-[no]scoreseq]<br>(Parameter 8)</td>
<td>Score alignment on basis of residue conservation</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>[-[no]scorencon]<br>(Parameter 9)</td>
<td>Score alignment on basis of number of contacts</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>[-[no]scoreccon]<br>(Parameter 10)</td>
<td>Score alignment on basis of conservation of contacts</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>[-scoreboth]<br>(Parameter 11)</td>
<td>Score alignment on a combined measure of number and conservation of contacts</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>[-[no]postsim]<br>(Parameter 12)</td>
<td>Ignore alignment postitions with post_similar value of 0</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>[-sigfile]<br>(Parameter 13)</td>
<td>Name of signature file for output</td>
<td>Output file</td>
<td>sig.sig</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>siggen</b> reads in multiple structure alignment generated by the
EMBOSS application <b>scopalign</b> and corresponding files of residue
contact data generated by the EMBOSS application <b>contacts</b>. 













<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->

The output file (Figure 1) uses the following records.
The four SCOP classification records are taken from the alignment input file:

<p>

<ol>
<li>CL - Domain class.  It is identical to the text given after 'Class' in 
the scop classification file (see documentation for the EMBOSS application 
scope). 

<li>FO - Domain fold.  It is identical to the text given after 'Fold' in 
the scop classification file (see scope documentation). 

<li>SF - Domain superfamily.  It is identical to the text given after 
'Superfamily' in the scop classification file (see scope documentation). 

<li>FA - Domain family. It is identical to the text given after 'Family' in 
the scop classification file (see scope documentation). 

<li>NP - Number of signature positions.

<li>NN - Signature position number.  The number given in brackets after 
this record indicates the start of the data for the relevent signature positi
on.

<li>IN - Informative line about signature position. The number of different 
amino acid residues seen for this position is given after 'NRES', the number 
of different sizes of gap follows 'NGAP', and the window size after 'WSIZ'.
When a signature is aligned to a protein sequence, the permissible gaps between 
two signature positions is determined by the empirical gaps and the window 
size for the C-terminal position (see sigscan.c)
Two rows of data for the emprical residues and gaps are then given:

<li>AA - The identifier of a residue seen in this position and the frequency
of its occurence are delimited by ';'.

<li>GA - The size of a gap seen in this position and the frequency 
of its occurence are delimited by ';'.

<li>// - used to delimit data for each signature. The last line of a file 
always  contains '//' only.
</ol>

<p>

Example excerpt from an output signature file:

<p>
<hr>
<pre>
CL   All beta proteins
XX
FO   Lipocalins
XX
SF   Lipocalins
XX
FA   Fatty acid binding protein-like
XX
NP   2
XX
NN   [1] 
XX
IN   NRES 3 ; NGAP 2 ; WSIZ 2  
XX
AA   A ; 2
AA   V ; 1
AA   L ; 4
XX
GA   1 ; 5
GA   2 ; 2
XX
NN   [2] 
XX
IN   NRES 2 ; NGAP 2 ; WSIZ 5  
XX
AA   F ; 1
AA   Y ; 5
XX
GA   12 ; 3
GA   10 ; 2
XX
//
</pre>
<hr>
<p>


Important

<p>

<ol>

<li>In the case a signature file is generated by hand, it is essential
that the gap data given is listed in order of increasing gap size. 

<li>In the current implementation, window size records always have the
value of 0.  These should be changed manually unless a very rigid
pattern is required.  A future implementation will provide a range of
methods for generating values of window size depending upon the
alignment (window size is identified by the WSIZ record in the signature
output file). 

<li>Siggen presumes that standard SCOP domain identifiers are given in
the input alignment if the id is 7 characters long and the first
character is a 'd' or 'D'.  In this case the contact data for that chain
will be parsed.  Otherwise contact data for chain 1 will be parsed. 
</ol>











<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.
	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
   -->

<b>siggen</b> reads in a protein residue comparison matrix.  By default,
this is EBLOSUM62. 

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>














<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

Ison JC, Blades MJ, Bleasby AJ, Daniel SC, Parish JH "Key residues
approach to the definition of protein families and analysis of sparse
family signatures" (2000) PROTEINS: Structure, Function and Genetics
40:330-341








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="contacts.html">contacts</a></td><td>Reads coordinate files and writes contact files</td></tr>
<tr><td><a href="scopalign.html">scopalign</a></td><td>Generate alignments for SCOP families</td></tr>
<tr><td><a href="sigscan.html">sigscan</a></td><td>Scans a sparse protein signature against swissprot</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

This application was written by Jon Ison <a
href="mailto:jison@hgmp.mrc.ac.uk">(jison@hgmp.mrc.ac.uk)</a>

















<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (June 2001) - Jon Ison.













<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->








</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

