<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: sigscan
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">
<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>
<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>
</td>
<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: sigscan
</H2>
</font>
</td>
</tr>
</table>
</td>
</tr>
</table>
<p><hr><p>

<!--END OF HEADER-->

<H1>
Program sigscan
</H1>
<H2>
    Function
</H2>
Scans a sparse protein signature against swissprot

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
        Algorithms
  -->

<b>sigscan</b> scans a signature such as that generated by the EMBOSS
application <b>siggen</b> against a protein sequence database and generates
files of scored hits and corresponding alignments.  

<!-- NOT IMPLEMENTED YET
Optionally, <b>sigscan</b> can use a file of residue pair
substitution frequencies as generated by the EMBOSS application
<b>pairsub</b> during a signature:sequence match.  
-->

An (optionally grouped) scop families file can be provided in which case
a classification of hits is provided in the signature hits output file. 
See documentation for the EMBOSS application <b>psiblasts</b> for an
explanation of the scop families file and groups for information on how
to group it. 













<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->


Here is a sample session with sigscan:

<pre>

% sigscan

</pre>















<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sigin]             infile     Name of signature file for input
  [-database]          seqall     Name of sequence database to search
  [-targetf]           infile     Name of (optionally grouped) scop families
                                  file for input
  [-thresh]            integer    Minimum length (residues) of overlap
                                  required for two hits with the same code to
                                  be counted as the same hit.
  [-sub]               matrixf    Residue substitution matrix
  [-gapo]              float      The gap insertion penalty is the score taken
                                  away when a gap is created. The best value
                                  depends on the choice of comparison matrix.
                                  The default value assumes you are using the
                                  EBLOSUM62 matrix for protein sequences, and
                                  the EDNAMAT matrix for nucleotide sequences.
  [-gape]              float      The gap extension, penalty is added to the
                                  standard gap penalty for each base or
                                  residue in the gap. This is how long gaps
                                  are penalized. Usually you will expect a few
                                  long gaps rather than many short gaps, so
                                  the gap extension penalty should be lower
                                  than the gap penalty. An exception is where
                                  one or both sequences are single reads with
                                  possible sequencing errors in which case you
                                  would expect many single base gaps. You can
                                  get this result by setting the gap open
                                  penalty to zero (or very low) and using the
                                  gap extension penalty to control gap
                                  scoring.
   -nterm              list       Select number
  [-nhits]             integer    Number of hits to output
  [-hitsf]             outfile    Name of signature hits file for output
  [-alignf]            outfile    Name of signature alignments file for output

   Optional qualifiers: (none)
   Advanced qualifiers: (none)
   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sigin]<br>(Parameter 1)</td>
<td>Name of signature file for input</td>
<td>Input file</td>
<td>test.sig</td>
</tr>

<tr>
<td>[-database]<br>(Parameter 2)</td>
<td>Name of sequence database to search</td>
<td>Readable sequence(s)</td>
<td>./test.seq</td>
</tr>

<tr>
<td>[-targetf]<br>(Parameter 3)</td>
<td>Name of (optionally grouped) scop families file for input</td>
<td>Input file</td>
<td>test.fam</td>
</tr>

<tr>
<td>[-thresh]<br>(Parameter 4)</td>
<td>Minimum length (residues) of overlap required for two hits with the same code to be counted as the same hit.</td>
<td>Any integer value</td>
<td>20</td>
</tr>

<tr>
<td>[-sub]<br>(Parameter 5)</td>
<td>Residue substitution matrix</td>
<td>Comparison matrix file in EMBOSS data path</td>
<td>./EBLOSUM62</td>
</tr>

<tr>
<td>[-gapo]<br>(Parameter 6)</td>
<td>The gap insertion penalty is the score taken away when a gap is created. The best value depends on the choice of comparison matrix. The default value assumes you are using the EBLOSUM62 matrix for protein sequences, and the EDNAMAT matrix for nucleotide sequences.</td>
<td>Floating point number from 1.0 to 100.0</td>
<td>10.0 for any sequence</td>
</tr>

<tr>
<td>[-gape]<br>(Parameter 7)</td>
<td>The gap extension, penalty is added to the standard gap penalty for each base or residue in the gap. This is how long gaps are penalized. Usually you will expect a few long gaps rather than many short gaps, so the gap extension penalty should be lower than the gap penalty. An exception is where one or both sequences are single reads with possible sequencing errors in which case you would expect many single base gaps. You can get this result by setting the gap open penalty to zero (or very low) and using the gap extension penalty to control gap scoring.</td>
<td>Floating point number from 0.0 to 10.0</td>
<td>0.5 for any sequence</td>
</tr>

<tr>
<td>-nterm</td>
<td>Select number</td>
<td><table><tr><td>1</td> <td><i>(Align anywhere and allow only complete signature-sequence fit)</i></td></tr><tr><td>2</td> <td><i>(Align anywhere and allow partial signature-sequence fit)</i></td></tr><tr><td>3</td> <td><i>(Use empirical gaps only)</i></td></tr></table></td>
<td>1</td>
</tr>

<tr>
<td>[-nhits]<br>(Parameter 8)</td>
<td>Number of hits to output</td>
<td>Any integer value</td>
<td>100</td>
</tr>

<tr>
<td>[-hitsf]<br>(Parameter 9)</td>
<td>Name of signature hits file for output</td>
<td>Output file</td>
<td>test.hits</td>
</tr>

<tr>
<td>[-alignf]<br>(Parameter 10)</td>
<td>Name of signature alignments file for output</td>
<td>Output file</td>
<td>test.align</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

Excerpts from a signature hits (Figure 1) are shown. The records used 
are are as follows:

<p>

<ol>
<li>DE - bibliographic information. The text 'Results of signature search'
is always given.

<li>Four SCOP classification records are given:

<ul>
<li>CL - Domain class.  It is identical to the text given after 'Class' in 
the scop classification file (see documentation for the EMBOSS application 
scope).
<li>FO - Domain fold.  It is identical to the text given after 'Fold' in 
the scop classification file (see scope documentation).
<li>SF - Domain superfamily.  It is identical to the text given after 
'Superfamily' in the scop classification file (see scope documentation).
<li>FA - Domain family. It is identical to the text given after 'Family' in 
the scop classification file (see scope documentation).
</ul>

<li>HI - hit data.  The data are as follows (column numbers are given in 
parentheses). 
<ul>
<li>(i) HI is always given. 
<li>(ii) Rank-order of the hit. 
<li>(iii) Database identifier code. 
<li>(iv) The group number of the protein if a grouped scop families file was provided or '.' otherwise. 
<li>(v) The primary classification of the hit. For true hits (genuine relatives to the signature)
one of 'TRAIN', 'PSIBLAST' or 'OTHER' is given. Otherwise, 'CROSS', 'FALSE'
or 'UNKNOWN' is given ('.' is given if a scop families file was not 
provided). 
<li>(vi) The secondary classification of the hit, either 'FALSE', 
'TRUE' or 'UNKNOWN' ('.' is given if a scop families file was not 
provided). The classes of hits are defined below.  
<li>(vii) Score of sequence-signature match.  
<li>(viii) E-value (see below).
</ul>

<li>XX - used for spacing.
<li>// - The file ends with a line containing '//' only.

</ol>

<p>

Example excerpt from a signature hits file:

<p>
<hr>
<pre>
DE   Results of signature search
XX
CL   All alpha proteins
XX
FO   Globin-like
XX
SF   Globin-like
XX
FA   Globins
XX
XX
HI   1    1RBPDFG   1    TRUE     TRUE    234  0.0001 
HI   2    1GFT35J   3    TRUE     TRUE    234  0.0008 
HI   3    1KJUFGH   1    TRUE     TRUE    224  0.0108 
HI   4    1GYU15R   2    CLOSE    TRUE    220  0.1876 
HI   5    1LKI89O   2    CLOSE    TRUE    203  0.6787 
HI   6    1QRTY58   1    TRUE     TRUE    199  0.9978 
HI   7    2IOM78G   1    FALSE    FALSE   198  1.0844
HI   8    1SZR234   1    CLOSE    TRUE    198  1.4343 
HI   9    3PONI57   1    DISTANT  FALSE   197  2.8849 
HI  10    1PHDJBS   3    CLOSE    TRUE    190  2.9872
HI  11    1HIOHDW   1    UNKNOWN  UNKNOWN 160  5,8676 
HI  12    199976T   1    CLOSE    TRUE    140  8.8346 
XX
//
</pre>
<hr>
<p>













<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->


Excerpts from an alignment file are shown (Figure 2). The records used 
are are as follows:

<p>

(1) The DE, CL, FO, SF, FA, XX and // records have the same meaning as 
in the hits file (above).

<p>

(2) Other lines contain either a fragment of protein sequence preceeded
by an accession number, or a fragment of an alignment of a signature
to the protein sequence (signature positions are marked with a '*').
The two numbers on either side of the sequence are begin and end residue
numbers for that line.

<p>

Example excerpt from a signature alignment file

<p>

<hr>
<pre>
DE   Results of signature search
XX
CL   Alpha and beta proteins (a/b)
XX
FO   alpha/beta-Hydrolases
XX
SF   alpha/beta-Hydrolases
XX
FA   Acetylcholinesterase-like
XX
OPSD_HUMAN      1        MNGTEGPNFYVPFSNATGVVRSPFEYPQYYLAEPWQFSMLAAYMF 45    
SIGNATURE       -        ---------*------------*---------------*------   
OPSD_XENLA      1        MNGTEGPNFYVPMSNKTGVVRSPFDYPQYYLAEPWQYSALAAYMF 45    
SIGNATURE       -        --------*-------------*----------------*-----   
XX
OPSD_HUMAN      46       LLIVLGFPINFLTLYVTVQHKKLRTPLNYILLNLAVADLFMVLGG 90
SIGNATURE       -        --------------*--------------*------------*--       
OPSD_XENLA      46       LLILLGLPINFMTLFVTIQHKKLRTPLNYILLNLVFANHFMVLCG 90    
SIGNATURE       -        --------------*--------------*------------*--       
XX
OPSD_HUMAN      91       FTSTLYTSLHGYFVFGPTGCNLEGFFATLGGEIALWSLVVLAIER 135   
SIGNATURE       -        ---------*--*--------------------------**----       
OPSD_XENLA      91       FTVTMYTSMHGYFIFGPTGCYIEGFFATLGGEVALWSLVVLAVER 135   
SIGNATURE       -        ---------*----*-------------------------**---       
XX
//
</pre>
<hr>

<p>

Definition of classes of hit 

<p>

The primary classification is an objective definition of the hit and has 
one of the following values:

<p>

TRAIN - the sequence was included in the original alignment from which the 
signature was generated.

<p>

PSIBLAST - A protein which was detected by psiblast  (see psiblasts.c) to 
be a homologue to at least one of the proteins in the family from which 
the signature was derived. Such proteins are identified by the 'PSIBLAST' 
record in the scop families file.

<p>

OTHER - A true member of the family but not a homologue as detected by 
psi-blast. Such proteins may have been found from the literature and 
manually added to the scop families file or may have been detected by the 
EMBOSS program swissparse (see swissparse.c). They are identified in the 

<p>

SCOP families file by the 'OTHER' record.

<p>

CROSS - A protein which is homologous to a protein of the same fold,
but differnt family, of the proteins from which the signature was
derived.

<p>

FALSE - A homologue to a protein with a different fold to the family
of the signature.

<p>

UNKNOWN - The protein is not known to be CROSS, FALSE or a true hit 
(TRAIN, PSIBLAST or OTHER).

<p>

The secondary classification is provided for convenience and a value as 
follows:

<p>

Hits of TRAIN, PSIBLAST and OTHER classification are all listed as TRUE.

<p>

Hits of CROSS, FALSE or UNKNOWN objective classification are listed as CROSS,
FALSE or UNKNOWN respectively.

<p>

The subjective column allows for hand-annotation of the hits files so that 
proteins of UNKNOWN objective classification can re-classified by a human 
expert as TRUE, FALSE, CROSS or otherwise left as UNKNOWN for the purpose of 
generating signature performance plots with the EMBOSS application sigplot.

















<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.
	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

   -->

None.













<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

Important - <b>sigscan</b> presumes that SCOP family names are unique. If 
this were not the case, changes to ajXyzClassifyHits would have to
be made.

<p>

Important - In the case where a signature file is generated by hand, it
is essential that the gap data given is listed in order of increasing
gap size. 





<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="siggen.html">siggen</a></td><td>Generates a sparse protein signature</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

This application was written by Jon Ison <a
href="mailto:jison@hgmp.mrc.ac.uk">(jison@hgmp.mrc.ac.uk)</a>
















<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (July 2001) - Jon Ison














<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->








</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

