<!-- START OF HEADER -->
<HTML><HEAD>
<TITLE> EMBASSY: SIGSCANLIG documentation</TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

<H2> SIGSCANLIG documentation
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<!-- END OF HEADER -->


<!-- CONTENTS
     This always includes the sections below.
     Other subsections can be added for individual applications.
-->
<H2>CONTENTS </H2>
<b> <a href="#1.0">1.0     SUMMARY                   </a></b><br>
<b> <a href="#2.0">2.0     INPUTS & OUTPUTS          </a></b><br>
<b> <a href="#3.0">3.0     INPUT FILE FORMAT         </a></b><br>
<b> <a href="#4.0">4.0     OUTPUT FILE FORMAT        </a></b><br>
<b> <a href="#5.0">5.0     DATA FILES                </a></b><br>
<b> <a href="#6.0">6.0     USAGE                     </a></b><br>    
<b> <a href="#7.0">7.0     KNOWN BUGS & WARNINGS     </a></b><br>    
<b> <a href="#8.0">8.0     NOTES                     </a></b><br>
<b> <a href="#9.0">9.0   DESCRIPTION               </a></b><br>
<b> <a href="#10.0">10.0   ALGORITHM                 </a></b><br>
<b> <a href="#11.0">11.0   RELATED APPLICATIONS      </a></b><br>
<b> <a href="#12.0">12.0   DIAGNOSTIC ERROR MESSAGES </a></b><br>
<b> <a href="#13.0">13.0   AUTHORS                   </a></b><br>
<b> <a href="#14.0">14.0   REFERENCES                </a></b><br>

<!-- SUMMARY
     Succint description of the application, particularly its inputs, outputs
     and what it does.  The same text is given at the top of the source (.c)
     file and in the <documentation> attribute of the <application definition>
     of the ACD file.
-->
<a name="1.0"></a>
<br><br><br><H2> 1.0   SUMMARY  </H2>
Search ligand-signature library & write hits (LHF file)

<p>
Generates LHF files (ligand hits file) of hits from scanning sequences against a library of ligand-binding signatures.



<!-- INPUTS & OUTPUTS
     Short summary of the application inputs and outputs in its different 
     modes of usage (if appropriate). More detail than the summary.
-->
<a name="2.0"></a>
<br><br><br><H2> 2.0   INPUTS & OUTPUTS          </H2>
SIGSCANLIG reads one more sequences, e.g. from a sequence database, and scans each in turn against a library of ligand-binding signatures (directory of signature files).  The signatures may be of type 1D or 3D.  To perform the scan, the user provides a scoring matrix which is either a residue substitution matrix (1D signatures) or matrix of 3D:1D (environment:residue) scores (3D). Various other parameters including gap insertion and extension penalties and a score mode are also set. For each query sequence, 3 files are generated as follows: (i) An LHF file (ligand hits file) of hits to the signatures. (ii) A SAF file (signature alignment file) of the corresponding signature-sequence alignments.  (iii) A results file, containing a list of hits (ligands), rank-ordered on the basis of score.  The user specifies the query sequences and also the paths of the signature files (input), LHF, SAF and results files (output).  The file extensions are specified in the ACD file. 

 




<!-- INPUT FILE FORMAT 
     Description and example(s) of input file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.
     Use "<b>sigscanlig</b> reads any normal sequence USAs." if
     appropriate.
-->
<a name="3.0"></a>
<br><br><br><H2> 3.0   INPUT FILE FORMAT         </H2>

The format of the signature file is described in <a href="siggen.html">SIGGEN documentation</a> (1D signatures) and the <a href="siggenlig.html">SIGGENLIG documentation</a> (3D signatures).

<br>SIGSCANLIG can read protein sequences in any sequence format supported by EMBOSS.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: swtiny</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
&gt; Q9WVI4^.^516^664^.^55074^Alpha and beta proteins (a+b)^.^.^Ferredoxin-like^Adenylyl and guanylyl cyclase catalytic domain^Adenylyl and guanylyl cyclase catalytic domain^PSIBLAST^84.40^0.000e+00^2.000e-16
DDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDIYKVETIGDAYCVASG
LHRKSLCHAKPIALMALKMMELSEEVLTPDGRPIQMRIGIHSGSVLAGVVGVRMPRYCLF
GNNVTLASKFESGSHPRRINISPTTYQLL
&gt; Q9ERL9^.^480^631^.^55074^Alpha and beta proteins (a+b)^.^.^Ferredoxin-like^Adenylyl and guanylyl cyclase catalytic domain^Adenylyl and guanylyl cyclase catalytic domain^PSIBLAST^72.10^0.000e+00^1.000e-12
VTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYKVETIGDAYCVAGGLH
RESDTHAVQIALMALKMMELSNEVMSPHGEPIKMRIGLHSGSVFAGVVGVKMPRYCLFGN
NVTLANKFESCSVPRKINVSPTTYRLLKDCPG
&gt; Q9DGG6^.^380^560^.^55074^Alpha and beta proteins (a+b)^.^.^Ferredoxin-like^Adenylyl and guanylyl cyclase catalytic domain^Adenylyl and guanylyl cyclase catalytic domain^PSIBLAST^142.00^0.000e+00^8.000e-34
EQVSILFADIVGFTKMSANKSAHALVGLLNDLFGRFDRLCEDTKCEKISTLGDCYYCVAG
CPEPRADHAYCCIEMGLGMIKAIEQFCQEKKEMVNMRVGVHTGTVLCGILGMRRFKFDVW
SNDVNLANLMEQLGVAGKVHISEATAKYLDDRYEMEDGKVTERVGQSAVADQLKGLKTYL
I
&gt; Q99396^.^262^473^.^55074^Alpha and beta proteins (a+b)^.^.^Ferredoxin-like^Adenylyl and guanylyl cyclase catalytic domain^Adenylyl and guanylyl cyclase catalytic domain^PSIBLAST^192.00^0.000e+00^0.000e+00
KELADPVTLIFTDIESSTAQWATQPELMPDAVATHHSMVRSLIENYDCYEVKTVGDSFMI
ACKSPFAAVQLAQELQLRFLRLDWGTTVFDEFYREFEERHAEEGDGKYKPPTARLDPEVY
RQLWNGLRVRVGIHTGLCDIRYDEVTKGYDYYGQTANTAARTESVGNGGQVLMTCETYHS
LSTAERSQFDVTPLGGVPLRGVSEPVEVYQLN
</pre>
</td></tr></table><p>






<!-- OUTPUT FILE FORMAT 
     Description and example(s) of output file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.  
     If the standard description of the avalable report formats is required, 
     use:   #include file="inc/reportformats.ihtml"
     Use "Outputs a graph to the specified graphics device."
     or  "outputs a report format file. The default format is ..."
     if appropriate.
-->
<a name="4.0"></a>
<br><br><br><H2> 4.0   OUTPUT FILE FORMAT         </H2>




<b>LHF file (ligand hits file)</b>
<br>Each LHF file contains the hits from the search of a single query sequence against the library of signatures.  The file (Figure 1) contains two lines per hit. The first contains a description of the hit in 17 text tokens delimited by '^'.  The second line contains the protein sequence.  The tokens are as follows (a '.' is given where a token does not have a value).

<UL>
<LI>(i) Accession number.
<LI>(ii)  Database code.
<LI>(iii - iv) Start and end positions of the hit relative to the full length sequence. 
</UL>

<br>The next 9 tokens refer to the domain family, superfamily etc for which the signature was derived and are as follows: 
<UL>
<LI>(v)  'LIGAND' is always given.
<LI>(vi)  4-character PDB identifier code. 
<LI>(vii)  7-character domain identifier code from SCOP or CATH.
<LI>(viii) Ligand identifier code.  A 3-letter abbreviation of the full ligand name.

<LI>(ix) Site number.  This uniquely identifies a ligand-binding site in the CON file generated by running SITES.
<LI>(x) The total number of binding sites for a given ligand.
<LI>(xi) Patch number.  The sequential order (from the N-terminus) of the patch in the ligand-binding site from which the signature was derived. 
<LI>(xii) Number of patches. The total number of patches in this ligand-binding site.
<LI>(xiii) Signature data type, either '1D' or '3D', for sequence or structure-based signatures respectively. 
<LI>(xiv) Signature type, either 'P' for patch signatures or 'F' for full-length signatures.
</UL>

<br>The next 3 tokens refer to the hit, specifically, information about the search result as follows: 
<UL>
<LI>(xv) Score of hit from search algoritm.
<LI>(xvi) P-value of hit (not written by SIGSCANLIG).
<LI>(xvii) E-value of hit (not written by SIGSCANLIG).
</UL>


<br><br>
<b>SAF file (signature alignment file)</b>
<br>
The format of the SAF file (signature alignment file, Figure 2) is more or less identical to DAF (domain alignment file) described in the <a href="domainalign.html">DOMAINALIGN documentation</a>.  The file conforms to EMBOSS "simple" multiple sequence alignment format.  Each file contains the signature-sequence alignments corresponding to the search of a single query sequence against a library of signatures.  The SAF file generated by SIGSCANLIG contains some unique features which are summarised here. 

<br><br>
 All lines beginning with '#' are comments and the first line is always '# DE   Alignment of query sequence against library of signatures'. Each signature-sequence alignment is contained in one or more blocks in the file. These blocks are preceeded by a comment line containing information for the corresponding hit from the LHF (above). Each block contains contains the signature identifier code and accession number of the protein sequence, sequence positions and a section of the signature-sequence alignment.

<UL>
<li>The signature identifier code uniquely identifies the signature and is of the format <I>ligand identifer code</I>_<I>site number</I>_<I>patch number</I>.  The ligand identifier, site number and the patch number are defined above (a '0' will be given for patch number if the site was represented as a single 'full-length' signature rather than patches).

<LI>The sequence positions are the start and end residue positions of the appropriate section of sequence. The character '-' is used for gaps.  

<LI> The protein sequence is given after the accession number and the signature-sequence alignment is given as a markup line underneath the sequence (signature positions are marked with a '*').
</UL>


<br><br>
<b>Results file</b>
<BR>Each results file (Figure 3) contains the summarised results of the search of a single query sequence against the library of signatures.  The results file contains a list of ligands, rank-ordered on the basis of score (highest-scoring ligand is given first). There are four columns under the labels 'LIGID, 'PATCHES', 'SITES' and 'SCORE'.  These give the ligand identifier, the number of patches for that ligand binding site, the number of sites for the ligand and the score of the sequence-signature match.  In the example shown (Figure 3) there is a single ligand.  Typically, the library would represent all ligands represented in PDB. 

<br><br><b>Figure 3   Excerpt from a results file (output) </b>
<table><td bgcolor="#CFCCFF">
<pre>
LIGID     PATCHES   SITES     SCORE     
101       0         1         0.48      
</pre>
</table>

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>Directory: lhf</h3>
<p>This directory contains output files, for example Q99396^.^262^473^.lhf    Q9DGG6^.^380^560^.lhf    Q9ERL9^.^480^631^.lhf    and Q9WVI4^.^516^664^.lhf.
<p><h3>File: lhf/Q99396^.^262^473^.lhf   </h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt; Q99396^.^262^473^.^55074^Alpha^.^99^138^LIGAND^1ii7;^d1ii7a_^101^2^2^0^0^1DF^7.00^0.000e+00^0.000e+00
KELADPVTLIFTDIESSTAQWATQPELMPDAVATHHSMVRSLIENYDCYEVKTVGDSFMIACKSPFAAVQLAQELQLRFLRLDWGTTVFDEFYREFEERHAEEGDGKYKPPTARLDPEVYRQLWNGLRVRVGIHTGLCDIRYDEVTKGYDYYGQTANTAARTESVGNGGQVLMTCETYHSLSTAERSQFDVTPLGGVPLRGVSEPVEVYQLN
&gt; Q99396^.^262^473^.^55074^Alpha^.^97^138^LIGAND^1ii7;^d1ii7a_^MN^1^1^0^0^1DF^5.33^0.000e+00^0.000e+00
KELADPVTLIFTDIESSTAQWATQPELMPDAVATHHSMVRSLIENYDCYEVKTVGDSFMIACKSPFAAVQLAQELQLRFLRLDWGTTVFDEFYREFEERHAEEGDGKYKPPTARLDPEVYRQLWNGLRVRVGIHTGLCDIRYDEVTKGYDYYGQTANTAARTESVGNGGQVLMTCETYHSLSTAERSQFDVTPLGGVPLRGVSEPVEVYQLN
&gt; Q99396^.^262^473^.^55074^Alpha^.^130^210^LIGAND^2hhb;^.^PO4^1^1^0^0^1DF^4.00^0.000e+00^0.000e+00
KELADPVTLIFTDIESSTAQWATQPELMPDAVATHHSMVRSLIENYDCYEVKTVGDSFMIACKSPFAAVQLAQELQLRFLRLDWGTTVFDEFYREFEERHAEEGDGKYKPPTARLDPEVYRQLWNGLRVRVGIHTGLCDIRYDEVTKGYDYYGQTANTAARTESVGNGGQVLMTCETYHSLSTAERSQFDVTPLGGVPLRGVSEPVEVYQLN
&gt; Q99396^.^262^473^.^55074^Alpha^.^12^17^LIGAND^1cs4;^d1cs4a_^POP^1^1^0^0^1DF^3.00^0.000e+00^0.000e+00
KELADPVTLIFTDIESSTAQWATQPELMPDAVATHHSMVRSLIENYDCYEVKTVGDSFMIACKSPFAAVQLAQELQLRFLRLDWGTTVFDEFYREFEERHAEEGDGKYKPPTARLDPEVYRQLWNGLRVRVGIHTGLCDIRYDEVTKGYDYYGQTANTAARTESVGNGGQVLMTCETYHSLSTAERSQFDVTPLGGVPLRGVSEPVEVYQLN
&gt; Q99396^.^262^473^.^55074^Alpha^.^12^56^LIGAND^1cs4;^d1cs4a_^MG^1^1^0^0^1DF^2.00^0.000e+00^0.000e+00
KELADPVTLIFTDIESSTAQWATQPELMPDAVATHHSMVRSLIENYDCYEVKTVGDSFMIACKSPFAAVQLAQELQLRFLRLDWGTTVFDEFYREFEERHAEEGDGKYKPPTARLDPEVYRQLWNGLRVRVGIHTGLCDIRYDEVTKGYDYYGQTANTAARTESVGNGGQVLMTCETYHSLSTAERSQFDVTPLGGVPLRGVSEPVEVYQLN
&gt; Q99396^.^262^473^.^55074^Alpha^.^11^55^LIGAND^1cs4;^d1cs4a_^101^1^2^0^0^1DF^1.83^0.000e+00^0.000e+00
KELADPVTLIFTDIESSTAQWATQPELMPDAVATHHSMVRSLIENYDCYEVKTVGDSFMIACKSPFAAVQLAQELQLRFLRLDWGTTVFDEFYREFEERHAEEGDGKYKPPTARLDPEVYRQLWNGLRVRVGIHTGLCDIRYDEVTKGYDYYGQTANTAARTESVGNGGQVLMTCETYHSLSTAERSQFDVTPLGGVPLRGVSEPVEVYQLN
&gt; Q99396^.^262^473^.^55074^Alpha^.^16^126^LIGAND^2hhb;^.^HEM^4^4^0^0^1DF^1.00^0.000e+00^0.000e+00
KELADPVTLIFTDIESSTAQWATQPELMPDAVATHHSMVRSLIENYDCYEVKTVGDSFMIACKSPFAAVQLAQELQLRFLRLDWGTTVFDEFYREFEERHAEEGDGKYKPPTARLDPEVYRQLWNGLRVRVGIHTGLCDIRYDEVTKGYDYYGQTANTAARTESVGNGGQVLMTCETYHSLSTAERSQFDVTPLGGVPLRGVSEPVEVYQLN
&gt; Q99396^.^262^473^.^55074^Alpha^.^32^126^LIGAND^2hhb;^.^HEM^1^4^0^0^1DF^0.94^0.000e+00^0.000e+00
KELADPVTLIFTDIESSTAQWATQPELMPDAVATHHSMVRSLIENYDCYEVKTVGDSFMIACKSPFAAVQLAQELQLRFLRLDWGTTVFDEFYREFEERHAEEGDGKYKPPTARLDPEVYRQLWNGLRVRVGIHTGLCDIRYDEVTKGYDYYGQTANTAARTESVGNGGQVLMTCETYHSLSTAERSQFDVTPLGGVPLRGVSEPVEVYQLN
&gt; Q99396^.^262^473^.^55074^Alpha^.^22^126^LIGAND^2hhb;^.^HEM^3^4^0^0^1DF^0.74^0.000e+00^0.000e+00
KELADPVTLIFTDIESSTAQWATQPELMPDAVATHHSMVRSLIENYDCYEVKTVGDSFMIACKSPFAAVQLAQELQLRFLRLDWGTTVFDEFYREFEERHAEEGDGKYKPPTARLDPEVYRQLWNGLRVRVGIHTGLCDIRYDEVTKGYDYYGQTANTAARTESVGNGGQVLMTCETYHSLSTAERSQFDVTPLGGVPLRGVSEPVEVYQLN
&gt; Q99396^.^262^473^.^55074^Alpha^.^16^126^LIGAND^2hhb;^.^HEM^2^4^0^0^1DF^0.59^0.000e+00^0.000e+00
KELADPVTLIFTDIESSTAQWATQPELMPDAVATHHSMVRSLIENYDCYEVKTVGDSFMIACKSPFAAVQLAQELQLRFLRLDWGTTVFDEFYREFEERHAEEGDGKYKPPTARLDPEVYRQLWNGLRVRVGIHTGLCDIRYDEVTKGYDYYGQTANTAARTESVGNGGQVLMTCETYHSLSTAERSQFDVTPLGGVPLRGVSEPVEVYQLN
&gt; Q99396^.^262^473^.^55074^Alpha^.^0^0^LIGAND^1cs4;^d1cs4a_^FOK^1^1^0^0^1DF^-1000000.00^0.000e+00^0.000e+00
KELADPVTLIFTDIESSTAQWATQPELMPDAVATHHSMVRSLIENYDCYEVKTVGDSFMIACKSPFAAVQLAQELQLRFLRLDWGTTVFDEFYREFEERHAEEGDGKYKPPTARLDPEVYRQLWNGLRVRVGIHTGLCDIRYDEVTKGYDYYGQTANTAARTESVGNGGQVLMTCETYHSLSTAERSQFDVTPLGGVPLRGVSEPVEVYQLN
</pre>
</td></tr></table><p>
<p><h3>File: lhf/Q9DGG6^.^380^560^.lhf   </h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt; Q9DGG6^.^380^560^.^55074^Alpha^.^8^52^LIGAND^1cs4;^d1cs4a_^101^1^2^0^0^1DF^5.50^0.000e+00^0.000e+00
EQVSILFADIVGFTKMSANKSAHALVGLLNDLFGRFDRLCEDTKCEKISTLGDCYYCVAGCPEPRADHAYCCIEMGLGMIKAIEQFCQEKKEMVNMRVGVHTGTVLCGILGMRRFKFDVWSNDVNLANLMEQLGVAGKVHISEATAKYLDDRYEMEDGKVTERVGQSAVADQLKGLKTYLI
&gt; Q9DGG6^.^380^560^.^55074^Alpha^.^8^52^LIGAND^1cs4;^d1cs4a_^MG^1^1^0^0^1DF^5.50^0.000e+00^0.000e+00
EQVSILFADIVGFTKMSANKSAHALVGLLNDLFGRFDRLCEDTKCEKISTLGDCYYCVAGCPEPRADHAYCCIEMGLGMIKAIEQFCQEKKEMVNMRVGVHTGTVLCGILGMRRFKFDVWSNDVNLANLMEQLGVAGKVHISEATAKYLDDRYEMEDGKVTERVGQSAVADQLKGLKTYLI
&gt; Q9DGG6^.^380^560^.^55074^Alpha^.^8^13^LIGAND^1cs4;^d1cs4a_^POP^1^1^0^0^1DF^4.17^0.000e+00^0.000e+00
EQVSILFADIVGFTKMSANKSAHALVGLLNDLFGRFDRLCEDTKCEKISTLGDCYYCVAGCPEPRADHAYCCIEMGLGMIKAIEQFCQEKKEMVNMRVGVHTGTVLCGILGMRRFKFDVWSNDVNLANLMEQLGVAGKVHISEATAKYLDDRYEMEDGKVTERVGQSAVADQLKGLKTYLI
&gt; Q9DGG6^.^380^560^.^55074^Alpha^.^99^179^LIGAND^2hhb;^.^PO4^1^1^0^0^1DF^4.00^0.000e+00^0.000e+00
EQVSILFADIVGFTKMSANKSAHALVGLLNDLFGRFDRLCEDTKCEKISTLGDCYYCVAGCPEPRADHAYCCIEMGLGMIKAIEQFCQEKKEMVNMRVGVHTGTVLCGILGMRRFKFDVWSNDVNLANLMEQLGVAGKVHISEATAKYLDDRYEMEDGKVTERVGQSAVADQLKGLKTYLI
&gt; Q9DGG6^.^380^560^.^55074^Alpha^.^100^139^LIGAND^1ii7;^d1ii7a_^101^2^2^0^0^1DF^3.50^0.000e+00^0.000e+00
EQVSILFADIVGFTKMSANKSAHALVGLLNDLFGRFDRLCEDTKCEKISTLGDCYYCVAGCPEPRADHAYCCIEMGLGMIKAIEQFCQEKKEMVNMRVGVHTGTVLCGILGMRRFKFDVWSNDVNLANLMEQLGVAGKVHISEATAKYLDDRYEMEDGKVTERVGQSAVADQLKGLKTYLI
&gt; Q9DGG6^.^380^560^.^55074^Alpha^.^20^61^LIGAND^1ii7;^d1ii7a_^MN^1^1^0^0^1DF^2.33^0.000e+00^0.000e+00
EQVSILFADIVGFTKMSANKSAHALVGLLNDLFGRFDRLCEDTKCEKISTLGDCYYCVAGCPEPRADHAYCCIEMGLGMIKAIEQFCQEKKEMVNMRVGVHTGTVLCGILGMRRFKFDVWSNDVNLANLMEQLGVAGKVHISEATAKYLDDRYEMEDGKVTERVGQSAVADQLKGLKTYLI
&gt; Q9DGG6^.^380^560^.^55074^Alpha^.^74^178^LIGAND^2hhb;^.^HEM^3^4^0^0^1DF^1.00^0.000e+00^0.000e+00
EQVSILFADIVGFTKMSANKSAHALVGLLNDLFGRFDRLCEDTKCEKISTLGDCYYCVAGCPEPRADHAYCCIEMGLGMIKAIEQFCQEKKEMVNMRVGVHTGTVLCGILGMRRFKFDVWSNDVNLANLMEQLGVAGKVHISEATAKYLDDRYEMEDGKVTERVGQSAVADQLKGLKTYLI
&gt; Q9DGG6^.^380^560^.^55074^Alpha^.^84^178^LIGAND^2hhb;^.^HEM^1^4^0^0^1DF^0.82^0.000e+00^0.000e+00
EQVSILFADIVGFTKMSANKSAHALVGLLNDLFGRFDRLCEDTKCEKISTLGDCYYCVAGCPEPRADHAYCCIEMGLGMIKAIEQFCQEKKEMVNMRVGVHTGTVLCGILGMRRFKFDVWSNDVNLANLMEQLGVAGKVHISEATAKYLDDRYEMEDGKVTERVGQSAVADQLKGLKTYLI
&gt; Q9DGG6^.^380^560^.^55074^Alpha^.^68^178^LIGAND^2hhb;^.^HEM^2^4^0^0^1DF^0.65^0.000e+00^0.000e+00
EQVSILFADIVGFTKMSANKSAHALVGLLNDLFGRFDRLCEDTKCEKISTLGDCYYCVAGCPEPRADHAYCCIEMGLGMIKAIEQFCQEKKEMVNMRVGVHTGTVLCGILGMRRFKFDVWSNDVNLANLMEQLGVAGKVHISEATAKYLDDRYEMEDGKVTERVGQSAVADQLKGLKTYLI
&gt; Q9DGG6^.^380^560^.^55074^Alpha^.^68^178^LIGAND^2hhb;^.^HEM^4^4^0^0^1DF^0.47^0.000e+00^0.000e+00
EQVSILFADIVGFTKMSANKSAHALVGLLNDLFGRFDRLCEDTKCEKISTLGDCYYCVAGCPEPRADHAYCCIEMGLGMIKAIEQFCQEKKEMVNMRVGVHTGTVLCGILGMRRFKFDVWSNDVNLANLMEQLGVAGKVHISEATAKYLDDRYEMEDGKVTERVGQSAVADQLKGLKTYLI
&gt; Q9DGG6^.^380^560^.^55074^Alpha^.^0^0^LIGAND^1cs4;^d1cs4a_^FOK^1^1^0^0^1DF^-1000000.00^0.000e+00^0.000e+00
EQVSILFADIVGFTKMSANKSAHALVGLLNDLFGRFDRLCEDTKCEKISTLGDCYYCVAGCPEPRADHAYCCIEMGLGMIKAIEQFCQEKKEMVNMRVGVHTGTVLCGILGMRRFKFDVWSNDVNLANLMEQLGVAGKVHISEATAKYLDDRYEMEDGKVTERVGQSAVADQLKGLKTYLI
</pre>
</td></tr></table><p>
<p><h3>File: lhf/Q9ERL9^.^480^631^.lhf   </h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt; Q9ERL9^.^480^631^.^55074^Alpha^.^6^50^LIGAND^1cs4;^d1cs4a_^MG^1^1^0^0^1DF^5.50^0.000e+00^0.000e+00
VTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYKVETIGDAYCVAGGLHRESDTHAVQIALMALKMMELSNEVMSPHGEPIKMRIGLHSGSVFAGVVGVKMPRYCLFGNNVTLANKFESCSVPRKINVSPTTYRLLKDCPG
&gt; Q9ERL9^.^480^631^.^55074^Alpha^.^6^50^LIGAND^1cs4;^d1cs4a_^101^1^2^0^0^1DF^5.17^0.000e+00^0.000e+00
VTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYKVETIGDAYCVAGGLHRESDTHAVQIALMALKMMELSNEVMSPHGEPIKMRIGLHSGSVFAGVVGVKMPRYCLFGNNVTLANKFESCSVPRKINVSPTTYRLLKDCPG
&gt; Q9ERL9^.^480^631^.^55074^Alpha^.^98^137^LIGAND^1ii7;^d1ii7a_^101^2^2^0^0^1DF^4.50^0.000e+00^0.000e+00
VTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYKVETIGDAYCVAGGLHRESDTHAVQIALMALKMMELSNEVMSPHGEPIKMRIGLHSGSVFAGVVGVKMPRYCLFGNNVTLANKFESCSVPRKINVSPTTYRLLKDCPG
&gt; Q9ERL9^.^480^631^.^55074^Alpha^.^6^11^LIGAND^1cs4;^d1cs4a_^POP^1^1^0^0^1DF^4.17^0.000e+00^0.000e+00
VTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYKVETIGDAYCVAGGLHRESDTHAVQIALMALKMMELSNEVMSPHGEPIKMRIGLHSGSVFAGVVGVKMPRYCLFGNNVTLANKFESCSVPRKINVSPTTYRLLKDCPG
&gt; Q9ERL9^.^480^631^.^55074^Alpha^.^63^104^LIGAND^1ii7;^d1ii7a_^MN^1^1^0^0^1DF^4.00^0.000e+00^0.000e+00
VTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYKVETIGDAYCVAGGLHRESDTHAVQIALMALKMMELSNEVMSPHGEPIKMRIGLHSGSVFAGVVGVKMPRYCLFGNNVTLANKFESCSVPRKINVSPTTYRLLKDCPG
&gt; Q9ERL9^.^480^631^.^55074^Alpha^.^22^102^LIGAND^2hhb;^.^PO4^1^1^0^0^1DF^2.50^0.000e+00^0.000e+00
VTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYKVETIGDAYCVAGGLHRESDTHAVQIALMALKMMELSNEVMSPHGEPIKMRIGLHSGSVFAGVVGVKMPRYCLFGNNVTLANKFESCSVPRKINVSPTTYRLLKDCPG
&gt; Q9ERL9^.^480^631^.^55074^Alpha^.^4^114^LIGAND^2hhb;^.^HEM^4^4^0^0^1DF^0.60^0.000e+00^0.000e+00
VTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYKVETIGDAYCVAGGLHRESDTHAVQIALMALKMMELSNEVMSPHGEPIKMRIGLHSGSVFAGVVGVKMPRYCLFGNNVTLANKFESCSVPRKINVSPTTYRLLKDCPG
&gt; Q9ERL9^.^480^631^.^55074^Alpha^.^25^119^LIGAND^2hhb;^.^HEM^1^4^0^0^1DF^0.29^0.000e+00^0.000e+00
VTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYKVETIGDAYCVAGGLHRESDTHAVQIALMALKMMELSNEVMSPHGEPIKMRIGLHSGSVFAGVVGVKMPRYCLFGNNVTLANKFESCSVPRKINVSPTTYRLLKDCPG
&gt; Q9ERL9^.^480^631^.^55074^Alpha^.^12^122^LIGAND^2hhb;^.^HEM^2^4^0^0^1DF^0.24^0.000e+00^0.000e+00
VTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYKVETIGDAYCVAGGLHRESDTHAVQIALMALKMMELSNEVMSPHGEPIKMRIGLHSGSVFAGVVGVKMPRYCLFGNNVTLANKFESCSVPRKINVSPTTYRLLKDCPG
&gt; Q9ERL9^.^480^631^.^55074^Alpha^.^15^119^LIGAND^2hhb;^.^HEM^3^4^0^0^1DF^0.21^0.000e+00^0.000e+00
VTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYKVETIGDAYCVAGGLHRESDTHAVQIALMALKMMELSNEVMSPHGEPIKMRIGLHSGSVFAGVVGVKMPRYCLFGNNVTLANKFESCSVPRKINVSPTTYRLLKDCPG
&gt; Q9ERL9^.^480^631^.^55074^Alpha^.^0^0^LIGAND^1cs4;^d1cs4a_^FOK^1^1^0^0^1DF^-1000000.00^0.000e+00^0.000e+00
VTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYKVETIGDAYCVAGGLHRESDTHAVQIALMALKMMELSNEVMSPHGEPIKMRIGLHSGSVFAGVVGVKMPRYCLFGNNVTLANKFESCSVPRKINVSPTTYRLLKDCPG
</pre>
</td></tr></table><p>
<p><h3>File: lhf/Q9WVI4^.^516^664^.lhf</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt; Q9WVI4^.^516^664^.^55074^Alpha^.^8^52^LIGAND^1cs4;^d1cs4a_^MG^1^1^0^0^1DF^5.50^0.000e+00^0.000e+00
DDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDIYKVETIGDAYCVASGLHRKSLCHAKPIALMALKMMELSEEVLTPDGRPIQMRIGIHSGSVLAGVVGVRMPRYCLFGNNVTLASKFESGSHPRRINISPTTYQLL
&gt; Q9WVI4^.^516^664^.^55074^Alpha^.^8^52^LIGAND^1cs4;^d1cs4a_^101^1^2^0^0^1DF^5.17^0.000e+00^0.000e+00
DDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDIYKVETIGDAYCVASGLHRKSLCHAKPIALMALKMMELSEEVLTPDGRPIQMRIGIHSGSVLAGVVGVRMPRYCLFGNNVTLASKFESGSHPRRINISPTTYQLL
&gt; Q9WVI4^.^516^664^.^55074^Alpha^.^100^139^LIGAND^1ii7;^d1ii7a_^101^2^2^0^0^1DF^4.50^0.000e+00^0.000e+00
DDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDIYKVETIGDAYCVASGLHRKSLCHAKPIALMALKMMELSEEVLTPDGRPIQMRIGIHSGSVLAGVVGVRMPRYCLFGNNVTLASKFESGSHPRRINISPTTYQLL
&gt; Q9WVI4^.^516^664^.^55074^Alpha^.^8^13^LIGAND^1cs4;^d1cs4a_^POP^1^1^0^0^1DF^4.17^0.000e+00^0.000e+00
DDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDIYKVETIGDAYCVASGLHRKSLCHAKPIALMALKMMELSEEVLTPDGRPIQMRIGIHSGSVLAGVVGVRMPRYCLFGNNVTLASKFESGSHPRRINISPTTYQLL
&gt; Q9WVI4^.^516^664^.^55074^Alpha^.^25^105^LIGAND^2hhb;^.^PO4^1^1^0^0^1DF^3.50^0.000e+00^0.000e+00
DDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDIYKVETIGDAYCVASGLHRKSLCHAKPIALMALKMMELSEEVLTPDGRPIQMRIGIHSGSVLAGVVGVRMPRYCLFGNNVTLASKFESGSHPRRINISPTTYQLL
&gt; Q9WVI4^.^516^664^.^55074^Alpha^.^43^84^LIGAND^1ii7;^d1ii7a_^MN^1^1^0^0^1DF^3.33^0.000e+00^0.000e+00
DDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDIYKVETIGDAYCVASGLHRKSLCHAKPIALMALKMMELSEEVLTPDGRPIQMRIGIHSGSVLAGVVGVRMPRYCLFGNNVTLASKFESGSHPRRINISPTTYQLL
&gt; Q9WVI4^.^516^664^.^55074^Alpha^.^27^121^LIGAND^2hhb;^.^HEM^1^4^0^0^1DF^0.53^0.000e+00^0.000e+00
DDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDIYKVETIGDAYCVASGLHRKSLCHAKPIALMALKMMELSEEVLTPDGRPIQMRIGIHSGSVLAGVVGVRMPRYCLFGNNVTLASKFESGSHPRRINISPTTYQLL
&gt; Q9WVI4^.^516^664^.^55074^Alpha^.^17^121^LIGAND^2hhb;^.^HEM^3^4^0^0^1DF^0.42^0.000e+00^0.000e+00
DDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDIYKVETIGDAYCVASGLHRKSLCHAKPIALMALKMMELSEEVLTPDGRPIQMRIGIHSGSVLAGVVGVRMPRYCLFGNNVTLASKFESGSHPRRINISPTTYQLL
&gt; Q9WVI4^.^516^664^.^55074^Alpha^.^6^116^LIGAND^2hhb;^.^HEM^4^4^0^0^1DF^0.33^0.000e+00^0.000e+00
DDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDIYKVETIGDAYCVASGLHRKSLCHAKPIALMALKMMELSEEVLTPDGRPIQMRIGIHSGSVLAGVVGVRMPRYCLFGNNVTLASKFESGSHPRRINISPTTYQLL
&gt; Q9WVI4^.^516^664^.^55074^Alpha^.^14^124^LIGAND^2hhb;^.^HEM^2^4^0^0^1DF^0.18^0.000e+00^0.000e+00
DDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDIYKVETIGDAYCVASGLHRKSLCHAKPIALMALKMMELSEEVLTPDGRPIQMRIGIHSGSVLAGVVGVRMPRYCLFGNNVTLASKFESGSHPRRINISPTTYQLL
&gt; Q9WVI4^.^516^664^.^55074^Alpha^.^0^0^LIGAND^1cs4;^d1cs4a_^FOK^1^1^0^0^1DF^-1000000.00^0.000e+00^0.000e+00
DDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDIYKVETIGDAYCVASGLHRKSLCHAKPIALMALKMMELSEEVLTPDGRPIQMRIGIHSGSVLAGVVGVRMPRYCLFGNNVTLASKFESGSHPRRINISPTTYQLL
</pre>
</td></tr></table><p>
<p><h3>Directory: aln</h3>
<p>This directory contains output files, for example Q99396^.^262^473^.aln    Q9DGG6^.^380^560^.aln    Q9ERL9^.^480^631^.aln    and Q9WVI4^.^516^664^.aln.
<p><h3>File: aln/Q99396^.^262^473^.aln   </h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
# DE   Alignment of query sequence against library of signatures
# XX
# &gt; Q99396^.^262^473^.^55074^Alpha^.^99^138^LIGAND^1ii7;^d1ii7a_^101^2^2^0^0^1DF^7.00^0.000e+00^0.000e+00
# XX
Q99396^.^262^473^.^55074^Alpha1      KELADPVTLIFTDIESSTAQWATQPELMPDAVATHHSMVRSLIEN 45
101_2_0           -      ---------------------------------------------
Q99396^.^262^473^.^55074^Alpha46     YDCYEVKTVGDSFMIACKSPFAAVQLAQELQLRFLRLDWGTTVFD 90
101_2_0           -      ---------------------------------------------
Q99396^.^262^473^.^55074^Alpha91     EFYREFEERHAEEGDGKYKPPTARLDPEVYRQLWNGLRVRVGIHT 135
101_2_0           -      ---------*-----------------------------------
Q99396^.^262^473^.^55074^Alpha136    GLCDIRYDEVTKGYDYYGQTANTAARTESVGNGGQVLMTCETYHS 180
101_2_0           -      ---*-----------------------------------------
Q99396^.^262^473^.^55074^Alpha181    LSTAERSQFDVTPLGGVPLRGVSEPVEVYQLN              225
101_2_0           -      --------------------------------             
# XX
# &gt; Q99396^.^262^473^.^55074^Alpha^.^97^138^LIGAND^1ii7;^d1ii7a_^MN^1^1^0^0^1DF^5.33^0.000e+00^0.000e+00
# XX
Q99396^.^262^473^.^55074^Alpha1      KELADPVTLIFTDIESSTAQWATQPELMPDAVATHHSMVRSLIEN 45
MN_1_0            -      ---------------------------------------------
Q99396^.^262^473^.^55074^Alpha46     YDCYEVKTVGDSFMIACKSPFAAVQLAQELQLRFLRLDWGTTVFD 90
MN_1_0            -      ---------------------------------------------
Q99396^.^262^473^.^55074^Alpha91     EFYREFEERHAEEGDGKYKPPTARLDPEVYRQLWNGLRVRVGIHT 135
MN_1_0            -      -------*-*-----------------------------------
Q99396^.^262^473^.^55074^Alpha136    GLCDIRYDEVTKGYDYYGQTANTAARTESVGNGGQVLMTCETYHS 180
MN_1_0            -      ---*-----------------------------------------
Q99396^.^262^473^.^55074^Alpha181    LSTAERSQFDVTPLGGVPLRGVSEPVEVYQLN              225
MN_1_0            -      --------------------------------             
# XX
# &gt; Q99396^.^262^473^.^55074^Alpha^.^130^210^LIGAND^2hhb;^.^PO4^1^1^0^0^1DF^4.00^0.000e+00^0.000e+00
# XX
Q99396^.^262^473^.^55074^Alpha1      KELADPVTLIFTDIESSTAQWATQPELMPDAVATHHSMVRSLIEN 45
PO4_1_0           -      ---------------------------------------------
Q99396^.^262^473^.^55074^Alpha46     YDCYEVKTVGDSFMIACKSPFAAVQLAQELQLRFLRLDWGTTVFD 90
PO4_1_0           -      ---------------------------------------------
Q99396^.^262^473^.^55074^Alpha91     EFYREFEERHAEEGDGKYKPPTARLDPEVYRQLWNGLRVRVGIHT 135
PO4_1_0           -      ----------------------------------------*----
Q99396^.^262^473^.^55074^Alpha136    GLCDIRYDEVTKGYDYYGQTANTAARTESVGNGGQVLMTCETYHS 180
PO4_1_0           -      ---------------------------------------------
Q99396^.^262^473^.^55074^Alpha181    LSTAERSQFDVTPLGGVPLRGVSEPVEVYQLN              225
PO4_1_0           -      ------------------------------*-             
# XX
# &gt; Q99396^.^262^473^.^55074^Alpha^.^12^17^LIGAND^1cs4;^d1cs4a_^POP^1^1^0^0^1DF^3.00^0.000e+00^0.000e+00
# XX
Q99396^.^262^473^.^55074^Alpha1      KELADPVTLIFTDIESSTAQWATQPELMPDAVATHHSMVRSLIEN 45
POP_1_0           -      ------------******---------------------------
Q99396^.^262^473^.^55074^Alpha46     YDCYEVKTVGDSFMIACKSPFAAVQLAQELQLRFLRLDWGTTVFD 90
POP_1_0           -      ---------------------------------------------
Q99396^.^262^473^.^55074^Alpha91     EFYREFEERHAEEGDGKYKPPTARLDPEVYRQLWNGLRVRVGIHT 135
POP_1_0           -      ---------------------------------------------
Q99396^.^262^473^.^55074^Alpha136    GLCDIRYDEVTKGYDYYGQTANTAARTESVGNGGQVLMTCETYHS 180


<font color=red>  [Part of this file has been deleted for brevity]</font>

# &gt; Q99396^.^262^473^.^55074^Alpha^.^32^126^LIGAND^2hhb;^.^HEM^1^4^0^0^1DF^0.94^0.000e+00^0.000e+00
# XX
Q99396^.^262^473^.^55074^Alpha1      KELADPVTLIFTDIESSTAQWATQPELMPDAVATHHSMVRSLIEN 45
HEM_1_0           -      --------------------------------**-**--------
Q99396^.^262^473^.^55074^Alpha46     YDCYEVKTVGDSFMIACKSPFAAVQLAQELQLRFLRLDWGTTVFD 90
HEM_1_0           -      ---*--**--*-----------------*--**---*-*---**-
Q99396^.^262^473^.^55074^Alpha91     EFYREFEERHAEEGDGKYKPPTARLDPEVYRQLWNGLRVRVGIHT 135
HEM_1_0           -      -*----------------------------------*--------
Q99396^.^262^473^.^55074^Alpha136    GLCDIRYDEVTKGYDYYGQTANTAARTESVGNGGQVLMTCETYHS 180
HEM_1_0           -      ---------------------------------------------
Q99396^.^262^473^.^55074^Alpha181    LSTAERSQFDVTPLGGVPLRGVSEPVEVYQLN              225
HEM_1_0           -      --------------------------------             
# XX
# &gt; Q99396^.^262^473^.^55074^Alpha^.^22^126^LIGAND^2hhb;^.^HEM^3^4^0^0^1DF^0.74^0.000e+00^0.000e+00
# XX
Q99396^.^262^473^.^55074^Alpha1      KELADPVTLIFTDIESSTAQWATQPELMPDAVATHHSMVRSLIEN 45
HEM_3_0           -      ----------------------*------*--**-**--------
Q99396^.^262^473^.^55074^Alpha46     YDCYEVKTVGDSFMIACKSPFAAVQLAQELQLRFLRLDWGTTVFD 90
HEM_3_0           -      ---*--**--*-----------------*--**---*-*---**-
Q99396^.^262^473^.^55074^Alpha91     EFYREFEERHAEEGDGKYKPPTARLDPEVYRQLWNGLRVRVGIHT 135
HEM_3_0           -      -*----------------------------------*--------
Q99396^.^262^473^.^55074^Alpha136    GLCDIRYDEVTKGYDYYGQTANTAARTESVGNGGQVLMTCETYHS 180
HEM_3_0           -      ---------------------------------------------
Q99396^.^262^473^.^55074^Alpha181    LSTAERSQFDVTPLGGVPLRGVSEPVEVYQLN              225
HEM_3_0           -      --------------------------------             
# XX
# &gt; Q99396^.^262^473^.^55074^Alpha^.^16^126^LIGAND^2hhb;^.^HEM^2^4^0^0^1DF^0.59^0.000e+00^0.000e+00
# XX
Q99396^.^262^473^.^55074^Alpha1      KELADPVTLIFTDIESSTAQWATQPELMPDAVATHHSMVRSLIEN 45
HEM_2_0           -      ----------------*---------**-----------------
Q99396^.^262^473^.^55074^Alpha46     YDCYEVKTVGDSFMIACKSPFAAVQLAQELQLRFLRLDWGTTVFD 90
HEM_2_0           -      ---*--**--**----------------*--**---*-*---**-
Q99396^.^262^473^.^55074^Alpha91     EFYREFEERHAEEGDGKYKPPTARLDPEVYRQLWNGLRVRVGIHT 135
HEM_2_0           -      -*----------------------------------*--------
Q99396^.^262^473^.^55074^Alpha136    GLCDIRYDEVTKGYDYYGQTANTAARTESVGNGGQVLMTCETYHS 180
HEM_2_0           -      ---------------------------------------------
Q99396^.^262^473^.^55074^Alpha181    LSTAERSQFDVTPLGGVPLRGVSEPVEVYQLN              225
HEM_2_0           -      --------------------------------             
# XX
# &gt; Q99396^.^262^473^.^55074^Alpha^.^0^0^LIGAND^1cs4;^d1cs4a_^FOK^1^1^0^0^1DF^-1000000.00^0.000e+00^0.000e+00
# XX
Q99396^.^262^473^.^55074^Alpha1      KELADPVTLIFTDIESSTAQWATQPELMPDAVATHHSMVRSLIEN 45
FOK_1_0           -      *--------------------------------------------
Q99396^.^262^473^.^55074^Alpha46     YDCYEVKTVGDSFMIACKSPFAAVQLAQELQLRFLRLDWGTTVFD 90
FOK_1_0           -      ---------------------------------------------
Q99396^.^262^473^.^55074^Alpha91     EFYREFEERHAEEGDGKYKPPTARLDPEVYRQLWNGLRVRVGIHT 135
FOK_1_0           -      ---------------------------------------------
Q99396^.^262^473^.^55074^Alpha136    GLCDIRYDEVTKGYDYYGQTANTAARTESVGNGGQVLMTCETYHS 180
FOK_1_0           -      ---------------------------------------------
Q99396^.^262^473^.^55074^Alpha181    LSTAERSQFDVTPLGGVPLRGVSEPVEVYQLN              225
FOK_1_0           -      --------------------------------             
</pre>
</td></tr></table><p>
<p><h3>File: aln/Q9DGG6^.^380^560^.aln   </h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
# DE   Alignment of query sequence against library of signatures
# XX
# &gt; Q9DGG6^.^380^560^.^55074^Alpha^.^8^52^LIGAND^1cs4;^d1cs4a_^101^1^2^0^0^1DF^5.50^0.000e+00^0.000e+00
# XX
Q9DGG6^.^380^560^.^55074^Alpha1      EQVSILFADIVGFTKMSANKSAHALVGLLNDLFGRFDRLCEDTKC 45
101_1_0           -      --------*---**-------------------------------
Q9DGG6^.^380^560^.^55074^Alpha46     EKISTLGDCYYCVAGCPEPRADHAYCCIEMGLGMIKAIEQFCQEK 90
101_1_0           -      -----***-------------------------------------
Q9DGG6^.^380^560^.^55074^Alpha91     KEMVNMRVGVHTGTVLCGILGMRRFKFDVWSNDVNLANLMEQLGV 135
101_1_0           -      ---------------------------------------------
Q9DGG6^.^380^560^.^55074^Alpha136    AGKVHISEATAKYLDDRYEMEDGKVTERVGQSAVADQLKGLKTYL 180
101_1_0           -      ---------------------------------------------
Q9DGG6^.^380^560^.^55074^Alpha181    I                                             225
101_1_0           -      -                                            
# XX
# &gt; Q9DGG6^.^380^560^.^55074^Alpha^.^8^52^LIGAND^1cs4;^d1cs4a_^MG^1^1^0^0^1DF^5.50^0.000e+00^0.000e+00
# XX
Q9DGG6^.^380^560^.^55074^Alpha1      EQVSILFADIVGFTKMSANKSAHALVGLLNDLFGRFDRLCEDTKC 45
MG_1_0            -      --------**--*--------------------------------
Q9DGG6^.^380^560^.^55074^Alpha46     EKISTLGDCYYCVAGCPEPRADHAYCCIEMGLGMIKAIEQFCQEK 90
MG_1_0            -      -------*-------------------------------------
Q9DGG6^.^380^560^.^55074^Alpha91     KEMVNMRVGVHTGTVLCGILGMRRFKFDVWSNDVNLANLMEQLGV 135
MG_1_0            -      ---------------------------------------------
Q9DGG6^.^380^560^.^55074^Alpha136    AGKVHISEATAKYLDDRYEMEDGKVTERVGQSAVADQLKGLKTYL 180
MG_1_0            -      ---------------------------------------------
Q9DGG6^.^380^560^.^55074^Alpha181    I                                             225
MG_1_0            -      -                                            
# XX
# &gt; Q9DGG6^.^380^560^.^55074^Alpha^.^8^13^LIGAND^1cs4;^d1cs4a_^POP^1^1^0^0^1DF^4.17^0.000e+00^0.000e+00
# XX
Q9DGG6^.^380^560^.^55074^Alpha1      EQVSILFADIVGFTKMSANKSAHALVGLLNDLFGRFDRLCEDTKC 45
POP_1_0           -      --------******-------------------------------
Q9DGG6^.^380^560^.^55074^Alpha46     EKISTLGDCYYCVAGCPEPRADHAYCCIEMGLGMIKAIEQFCQEK 90
POP_1_0           -      ---------------------------------------------
Q9DGG6^.^380^560^.^55074^Alpha91     KEMVNMRVGVHTGTVLCGILGMRRFKFDVWSNDVNLANLMEQLGV 135
POP_1_0           -      ---------------------------------------------
Q9DGG6^.^380^560^.^55074^Alpha136    AGKVHISEATAKYLDDRYEMEDGKVTERVGQSAVADQLKGLKTYL 180
POP_1_0           -      ---------------------------------------------
Q9DGG6^.^380^560^.^55074^Alpha181    I                                             225
POP_1_0           -      -                                            
# XX
# &gt; Q9DGG6^.^380^560^.^55074^Alpha^.^99^179^LIGAND^2hhb;^.^PO4^1^1^0^0^1DF^4.00^0.000e+00^0.000e+00
# XX
Q9DGG6^.^380^560^.^55074^Alpha1      EQVSILFADIVGFTKMSANKSAHALVGLLNDLFGRFDRLCEDTKC 45
PO4_1_0           -      ---------------------------------------------
Q9DGG6^.^380^560^.^55074^Alpha46     EKISTLGDCYYCVAGCPEPRADHAYCCIEMGLGMIKAIEQFCQEK 90
PO4_1_0           -      ---------------------------------------------
Q9DGG6^.^380^560^.^55074^Alpha91     KEMVNMRVGVHTGTVLCGILGMRRFKFDVWSNDVNLANLMEQLGV 135
PO4_1_0           -      ---------*-----------------------------------
Q9DGG6^.^380^560^.^55074^Alpha136    AGKVHISEATAKYLDDRYEMEDGKVTERVGQSAVADQLKGLKTYL 180


<font color=red>  [Part of this file has been deleted for brevity]</font>

# &gt; Q9DGG6^.^380^560^.^55074^Alpha^.^84^178^LIGAND^2hhb;^.^HEM^1^4^0^0^1DF^0.82^0.000e+00^0.000e+00
# XX
Q9DGG6^.^380^560^.^55074^Alpha1      EQVSILFADIVGFTKMSANKSAHALVGLLNDLFGRFDRLCEDTKC 45
HEM_1_0           -      ---------------------------------------------
Q9DGG6^.^380^560^.^55074^Alpha46     EKISTLGDCYYCVAGCPEPRADHAYCCIEMGLGMIKAIEQFCQEK 90
HEM_1_0           -      ---------------------------------------**-**-
Q9DGG6^.^380^560^.^55074^Alpha91     KEMVNMRVGVHTGTVLCGILGMRRFKFDVWSNDVNLANLMEQLGV 135
HEM_1_0           -      ----------*--**--*-----------------*--**---*-
Q9DGG6^.^380^560^.^55074^Alpha136    AGKVHISEATAKYLDDRYEMEDGKVTERVGQSAVADQLKGLKTYL 180
HEM_1_0           -      *---**--*----------------------------------*-
Q9DGG6^.^380^560^.^55074^Alpha181    I                                             225
HEM_1_0           -      -                                            
# XX
# &gt; Q9DGG6^.^380^560^.^55074^Alpha^.^68^178^LIGAND^2hhb;^.^HEM^2^4^0^0^1DF^0.65^0.000e+00^0.000e+00
# XX
Q9DGG6^.^380^560^.^55074^Alpha1      EQVSILFADIVGFTKMSANKSAHALVGLLNDLFGRFDRLCEDTKC 45
HEM_2_0           -      ---------------------------------------------
Q9DGG6^.^380^560^.^55074^Alpha46     EKISTLGDCYYCVAGCPEPRADHAYCCIEMGLGMIKAIEQFCQEK 90
HEM_2_0           -      -----------------------*---------**----------
Q9DGG6^.^380^560^.^55074^Alpha91     KEMVNMRVGVHTGTVLCGILGMRRFKFDVWSNDVNLANLMEQLGV 135
HEM_2_0           -      ----------*--**--**----------------*--**---*-
Q9DGG6^.^380^560^.^55074^Alpha136    AGKVHISEATAKYLDDRYEMEDGKVTERVGQSAVADQLKGLKTYL 180
HEM_2_0           -      *---**--*----------------------------------*-
Q9DGG6^.^380^560^.^55074^Alpha181    I                                             225
HEM_2_0           -      -                                            
# XX
# &gt; Q9DGG6^.^380^560^.^55074^Alpha^.^68^178^LIGAND^2hhb;^.^HEM^4^4^0^0^1DF^0.47^0.000e+00^0.000e+00
# XX
Q9DGG6^.^380^560^.^55074^Alpha1      EQVSILFADIVGFTKMSANKSAHALVGLLNDLFGRFDRLCEDTKC 45
HEM_4_0           -      ---------------------------------------------
Q9DGG6^.^380^560^.^55074^Alpha46     EKISTLGDCYYCVAGCPEPRADHAYCCIEMGLGMIKAIEQFCQEK 90
HEM_4_0           -      -----------------------*---------**----------
Q9DGG6^.^380^560^.^55074^Alpha91     KEMVNMRVGVHTGTVLCGILGMRRFKFDVWSNDVNLANLMEQLGV 135
HEM_4_0           -      ----------*--**--*--------------------**---*-
Q9DGG6^.^380^560^.^55074^Alpha136    AGKVHISEATAKYLDDRYEMEDGKVTERVGQSAVADQLKGLKTYL 180
HEM_4_0           -      *---**--*----------------------------------*-
Q9DGG6^.^380^560^.^55074^Alpha181    I                                             225
HEM_4_0           -      -                                            
# XX
# &gt; Q9DGG6^.^380^560^.^55074^Alpha^.^0^0^LIGAND^1cs4;^d1cs4a_^FOK^1^1^0^0^1DF^-1000000.00^0.000e+00^0.000e+00
# XX
Q9DGG6^.^380^560^.^55074^Alpha1      EQVSILFADIVGFTKMSANKSAHALVGLLNDLFGRFDRLCEDTKC 45
FOK_1_0           -      *--------------------------------------------
Q9DGG6^.^380^560^.^55074^Alpha46     EKISTLGDCYYCVAGCPEPRADHAYCCIEMGLGMIKAIEQFCQEK 90
FOK_1_0           -      ---------------------------------------------
Q9DGG6^.^380^560^.^55074^Alpha91     KEMVNMRVGVHTGTVLCGILGMRRFKFDVWSNDVNLANLMEQLGV 135
FOK_1_0           -      ---------------------------------------------
Q9DGG6^.^380^560^.^55074^Alpha136    AGKVHISEATAKYLDDRYEMEDGKVTERVGQSAVADQLKGLKTYL 180
FOK_1_0           -      ---------------------------------------------
Q9DGG6^.^380^560^.^55074^Alpha181    I                                             225
FOK_1_0           -      -                                            
</pre>
</td></tr></table><p>
<p><h3>File: aln/Q9ERL9^.^480^631^.aln   </h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
# DE   Alignment of query sequence against library of signatures
# XX
# &gt; Q9ERL9^.^480^631^.^55074^Alpha^.^6^50^LIGAND^1cs4;^d1cs4a_^MG^1^1^0^0^1DF^5.50^0.000e+00^0.000e+00
# XX
Q9ERL9^.^480^631^.^55074^Alpha1      VTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYK 45
MG_1_0            -      ------**--*----------------------------------
Q9ERL9^.^480^631^.^55074^Alpha46     VETIGDAYCVAGGLHRESDTHAVQIALMALKMMELSNEVMSPHGE 90
MG_1_0            -      -----*---------------------------------------
Q9ERL9^.^480^631^.^55074^Alpha91     PIKMRIGLHSGSVFAGVVGVKMPRYCLFGNNVTLANKFESCSVPR 135
MG_1_0            -      ---------------------------------------------
Q9ERL9^.^480^631^.^55074^Alpha136    KINVSPTTYRLLKDCPG                             180
MG_1_0            -      -----------------                            
# XX
# &gt; Q9ERL9^.^480^631^.^55074^Alpha^.^6^50^LIGAND^1cs4;^d1cs4a_^101^1^2^0^0^1DF^5.17^0.000e+00^0.000e+00
# XX
Q9ERL9^.^480^631^.^55074^Alpha1      VTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYK 45
101_1_0           -      ------*---**---------------------------------
Q9ERL9^.^480^631^.^55074^Alpha46     VETIGDAYCVAGGLHRESDTHAVQIALMALKMMELSNEVMSPHGE 90
101_1_0           -      ---***---------------------------------------
Q9ERL9^.^480^631^.^55074^Alpha91     PIKMRIGLHSGSVFAGVVGVKMPRYCLFGNNVTLANKFESCSVPR 135
101_1_0           -      ---------------------------------------------
Q9ERL9^.^480^631^.^55074^Alpha136    KINVSPTTYRLLKDCPG                             180
101_1_0           -      -----------------                            
# XX
# &gt; Q9ERL9^.^480^631^.^55074^Alpha^.^98^137^LIGAND^1ii7;^d1ii7a_^101^2^2^0^0^1DF^4.50^0.000e+00^0.000e+00
# XX
Q9ERL9^.^480^631^.^55074^Alpha1      VTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYK 45
101_2_0           -      ---------------------------------------------
Q9ERL9^.^480^631^.^55074^Alpha46     VETIGDAYCVAGGLHRESDTHAVQIALMALKMMELSNEVMSPHGE 90
101_2_0           -      ---------------------------------------------
Q9ERL9^.^480^631^.^55074^Alpha91     PIKMRIGLHSGSVFAGVVGVKMPRYCLFGNNVTLANKFESCSVPR 135
101_2_0           -      --------*------------------------------------
Q9ERL9^.^480^631^.^55074^Alpha136    KINVSPTTYRLLKDCPG                             180
101_2_0           -      --*--------------                            
# XX
# &gt; Q9ERL9^.^480^631^.^55074^Alpha^.^6^11^LIGAND^1cs4;^d1cs4a_^POP^1^1^0^0^1DF^4.17^0.000e+00^0.000e+00
# XX
Q9ERL9^.^480^631^.^55074^Alpha1      VTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYK 45
POP_1_0           -      ------******---------------------------------
Q9ERL9^.^480^631^.^55074^Alpha46     VETIGDAYCVAGGLHRESDTHAVQIALMALKMMELSNEVMSPHGE 90
POP_1_0           -      ---------------------------------------------
Q9ERL9^.^480^631^.^55074^Alpha91     PIKMRIGLHSGSVFAGVVGVKMPRYCLFGNNVTLANKFESCSVPR 135
POP_1_0           -      ---------------------------------------------
Q9ERL9^.^480^631^.^55074^Alpha136    KINVSPTTYRLLKDCPG                             180
POP_1_0           -      -----------------                            
# XX
# &gt; Q9ERL9^.^480^631^.^55074^Alpha^.^63^104^LIGAND^1ii7;^d1ii7a_^MN^1^1^0^0^1DF^4.00^0.000e+00^0.000e+00
# XX
Q9ERL9^.^480^631^.^55074^Alpha1      VTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYK 45
MN_1_0            -      ---------------------------------------------


<font color=red>  [Part of this file has been deleted for brevity]</font>

HEM_4_0           -      ----*---------**--------------------*--**--*-
Q9ERL9^.^480^631^.^55074^Alpha46     VETIGDAYCVAGGLHRESDTHAVQIALMALKMMELSNEVMSPHGE 90
HEM_4_0           -      -------------------**---*-*---**--*----------
Q9ERL9^.^480^631^.^55074^Alpha91     PIKMRIGLHSGSVFAGVVGVKMPRYCLFGNNVTLANKFESCSVPR 135
HEM_4_0           -      ------------------------*--------------------
Q9ERL9^.^480^631^.^55074^Alpha136    KINVSPTTYRLLKDCPG                             180
HEM_4_0           -      -----------------                            
# XX
# &gt; Q9ERL9^.^480^631^.^55074^Alpha^.^25^119^LIGAND^2hhb;^.^HEM^1^4^0^0^1DF^0.29^0.000e+00^0.000e+00
# XX
Q9ERL9^.^480^631^.^55074^Alpha1      VTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYK 45
HEM_1_0           -      -------------------------**-**-----------*--*
Q9ERL9^.^480^631^.^55074^Alpha46     VETIGDAYCVAGGLHRESDTHAVQIALMALKMMELSNEVMSPHGE 90
HEM_1_0           -      *--*-----------------*--**---*-*---**--*-----
Q9ERL9^.^480^631^.^55074^Alpha91     PIKMRIGLHSGSVFAGVVGVKMPRYCLFGNNVTLANKFESCSVPR 135
HEM_1_0           -      -----------------------------*---------------
Q9ERL9^.^480^631^.^55074^Alpha136    KINVSPTTYRLLKDCPG                             180
HEM_1_0           -      -----------------                            
# XX
# &gt; Q9ERL9^.^480^631^.^55074^Alpha^.^12^122^LIGAND^2hhb;^.^HEM^2^4^0^0^1DF^0.24^0.000e+00^0.000e+00
# XX
Q9ERL9^.^480^631^.^55074^Alpha1      VTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYK 45
HEM_2_0           -      ------------*---------**--------------------*
Q9ERL9^.^480^631^.^55074^Alpha46     VETIGDAYCVAGGLHRESDTHAVQIALMALKMMELSNEVMSPHGE 90
HEM_2_0           -      --**--**----------------*--**---*-*---**--*--
Q9ERL9^.^480^631^.^55074^Alpha91     PIKMRIGLHSGSVFAGVVGVKMPRYCLFGNNVTLANKFESCSVPR 135
HEM_2_0           -      --------------------------------*------------
Q9ERL9^.^480^631^.^55074^Alpha136    KINVSPTTYRLLKDCPG                             180
HEM_2_0           -      -----------------                            
# XX
# &gt; Q9ERL9^.^480^631^.^55074^Alpha^.^15^119^LIGAND^2hhb;^.^HEM^3^4^0^0^1DF^0.21^0.000e+00^0.000e+00
# XX
Q9ERL9^.^480^631^.^55074^Alpha1      VTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYK 45
HEM_3_0           -      ---------------*------*--**-**-----------*--*
Q9ERL9^.^480^631^.^55074^Alpha46     VETIGDAYCVAGGLHRESDTHAVQIALMALKMMELSNEVMSPHGE 90
HEM_3_0           -      *--*-----------------*--**---*-*---**--*-----
Q9ERL9^.^480^631^.^55074^Alpha91     PIKMRIGLHSGSVFAGVVGVKMPRYCLFGNNVTLANKFESCSVPR 135
HEM_3_0           -      -----------------------------*---------------
Q9ERL9^.^480^631^.^55074^Alpha136    KINVSPTTYRLLKDCPG                             180
HEM_3_0           -      -----------------                            
# XX
# &gt; Q9ERL9^.^480^631^.^55074^Alpha^.^0^0^LIGAND^1cs4;^d1cs4a_^FOK^1^1^0^0^1DF^-1000000.00^0.000e+00^0.000e+00
# XX
Q9ERL9^.^480^631^.^55074^Alpha1      VTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYK 45
FOK_1_0           -      *--------------------------------------------
Q9ERL9^.^480^631^.^55074^Alpha46     VETIGDAYCVAGGLHRESDTHAVQIALMALKMMELSNEVMSPHGE 90
FOK_1_0           -      ---------------------------------------------
Q9ERL9^.^480^631^.^55074^Alpha91     PIKMRIGLHSGSVFAGVVGVKMPRYCLFGNNVTLANKFESCSVPR 135
FOK_1_0           -      ---------------------------------------------
Q9ERL9^.^480^631^.^55074^Alpha136    KINVSPTTYRLLKDCPG                             180
FOK_1_0           -      -----------------                            
</pre>
</td></tr></table><p>
<p><h3>File: aln/Q9WVI4^.^516^664^.aln</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
# DE   Alignment of query sequence against library of signatures
# XX
# &gt; Q9WVI4^.^516^664^.^55074^Alpha^.^8^52^LIGAND^1cs4;^d1cs4a_^MG^1^1^0^0^1DF^5.50^0.000e+00^0.000e+00
# XX
Q9WVI4^.^516^664^.^55074^Alpha1      DDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDI 45
MG_1_0            -      --------**--*--------------------------------
Q9WVI4^.^516^664^.^55074^Alpha46     YKVETIGDAYCVASGLHRKSLCHAKPIALMALKMMELSEEVLTPD 90
MG_1_0            -      -------*-------------------------------------
Q9WVI4^.^516^664^.^55074^Alpha91     GRPIQMRIGIHSGSVLAGVVGVRMPRYCLFGNNVTLASKFESGSH 135
MG_1_0            -      ---------------------------------------------
Q9WVI4^.^516^664^.^55074^Alpha136    PRRINISPTTYQLL                                180
MG_1_0            -      --------------                               
# XX
# &gt; Q9WVI4^.^516^664^.^55074^Alpha^.^8^52^LIGAND^1cs4;^d1cs4a_^101^1^2^0^0^1DF^5.17^0.000e+00^0.000e+00
# XX
Q9WVI4^.^516^664^.^55074^Alpha1      DDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDI 45
101_1_0           -      --------*---**-------------------------------
Q9WVI4^.^516^664^.^55074^Alpha46     YKVETIGDAYCVASGLHRKSLCHAKPIALMALKMMELSEEVLTPD 90
101_1_0           -      -----***-------------------------------------
Q9WVI4^.^516^664^.^55074^Alpha91     GRPIQMRIGIHSGSVLAGVVGVRMPRYCLFGNNVTLASKFESGSH 135
101_1_0           -      ---------------------------------------------
Q9WVI4^.^516^664^.^55074^Alpha136    PRRINISPTTYQLL                                180
101_1_0           -      --------------                               
# XX
# &gt; Q9WVI4^.^516^664^.^55074^Alpha^.^100^139^LIGAND^1ii7;^d1ii7a_^101^2^2^0^0^1DF^4.50^0.000e+00^0.000e+00
# XX
Q9WVI4^.^516^664^.^55074^Alpha1      DDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDI 45
101_2_0           -      ---------------------------------------------
Q9WVI4^.^516^664^.^55074^Alpha46     YKVETIGDAYCVASGLHRKSLCHAKPIALMALKMMELSEEVLTPD 90
101_2_0           -      ---------------------------------------------
Q9WVI4^.^516^664^.^55074^Alpha91     GRPIQMRIGIHSGSVLAGVVGVRMPRYCLFGNNVTLASKFESGSH 135
101_2_0           -      ----------*----------------------------------
Q9WVI4^.^516^664^.^55074^Alpha136    PRRINISPTTYQLL                                180
101_2_0           -      ----*---------                               
# XX
# &gt; Q9WVI4^.^516^664^.^55074^Alpha^.^8^13^LIGAND^1cs4;^d1cs4a_^POP^1^1^0^0^1DF^4.17^0.000e+00^0.000e+00
# XX
Q9WVI4^.^516^664^.^55074^Alpha1      DDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDI 45
POP_1_0           -      --------******-------------------------------
Q9WVI4^.^516^664^.^55074^Alpha46     YKVETIGDAYCVASGLHRKSLCHAKPIALMALKMMELSEEVLTPD 90
POP_1_0           -      ---------------------------------------------
Q9WVI4^.^516^664^.^55074^Alpha91     GRPIQMRIGIHSGSVLAGVVGVRMPRYCLFGNNVTLASKFESGSH 135
POP_1_0           -      ---------------------------------------------
Q9WVI4^.^516^664^.^55074^Alpha136    PRRINISPTTYQLL                                180
POP_1_0           -      --------------                               
# XX
# &gt; Q9WVI4^.^516^664^.^55074^Alpha^.^25^105^LIGAND^2hhb;^.^PO4^1^1^0^0^1DF^3.50^0.000e+00^0.000e+00
# XX
Q9WVI4^.^516^664^.^55074^Alpha1      DDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDI 45
PO4_1_0           -      -------------------------*-------------------


<font color=red>  [Part of this file has been deleted for brevity]</font>

HEM_1_0           -      ---------------------------**-**-----------*-
Q9WVI4^.^516^664^.^55074^Alpha46     YKVETIGDAYCVASGLHRKSLCHAKPIALMALKMMELSEEVLTPD 90
HEM_1_0           -      -**--*-----------------*--**---*-*---**--*---
Q9WVI4^.^516^664^.^55074^Alpha91     GRPIQMRIGIHSGSVLAGVVGVRMPRYCLFGNNVTLASKFESGSH 135
HEM_1_0           -      -------------------------------*-------------
Q9WVI4^.^516^664^.^55074^Alpha136    PRRINISPTTYQLL                                180
HEM_1_0           -      --------------                               
# XX
# &gt; Q9WVI4^.^516^664^.^55074^Alpha^.^17^121^LIGAND^2hhb;^.^HEM^3^4^0^0^1DF^0.42^0.000e+00^0.000e+00
# XX
Q9WVI4^.^516^664^.^55074^Alpha1      DDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDI 45
HEM_3_0           -      -----------------*------*--**-**-----------*-
Q9WVI4^.^516^664^.^55074^Alpha46     YKVETIGDAYCVASGLHRKSLCHAKPIALMALKMMELSEEVLTPD 90
HEM_3_0           -      -**--*-----------------*--**---*-*---**--*---
Q9WVI4^.^516^664^.^55074^Alpha91     GRPIQMRIGIHSGSVLAGVVGVRMPRYCLFGNNVTLASKFESGSH 135
HEM_3_0           -      -------------------------------*-------------
Q9WVI4^.^516^664^.^55074^Alpha136    PRRINISPTTYQLL                                180
HEM_3_0           -      --------------                               
# XX
# &gt; Q9WVI4^.^516^664^.^55074^Alpha^.^6^116^LIGAND^2hhb;^.^HEM^4^4^0^0^1DF^0.33^0.000e+00^0.000e+00
# XX
Q9WVI4^.^516^664^.^55074^Alpha1      DDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDI 45
HEM_4_0           -      ------*---------**--------------------*--**--
Q9WVI4^.^516^664^.^55074^Alpha46     YKVETIGDAYCVASGLHRKSLCHAKPIALMALKMMELSEEVLTPD 90
HEM_4_0           -      *--------------------**---*-*---**--*--------
Q9WVI4^.^516^664^.^55074^Alpha91     GRPIQMRIGIHSGSVLAGVVGVRMPRYCLFGNNVTLASKFESGSH 135
HEM_4_0           -      --------------------------*------------------
Q9WVI4^.^516^664^.^55074^Alpha136    PRRINISPTTYQLL                                180
HEM_4_0           -      --------------                               
# XX
# &gt; Q9WVI4^.^516^664^.^55074^Alpha^.^14^124^LIGAND^2hhb;^.^HEM^2^4^0^0^1DF^0.18^0.000e+00^0.000e+00
# XX
Q9WVI4^.^516^664^.^55074^Alpha1      DDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDI 45
HEM_2_0           -      --------------*---------**-------------------
Q9WVI4^.^516^664^.^55074^Alpha46     YKVETIGDAYCVASGLHRKSLCHAKPIALMALKMMELSEEVLTPD 90
HEM_2_0           -      -*--**--**----------------*--**---*-*---**--*
Q9WVI4^.^516^664^.^55074^Alpha91     GRPIQMRIGIHSGSVLAGVVGVRMPRYCLFGNNVTLASKFESGSH 135
HEM_2_0           -      ----------------------------------*----------
Q9WVI4^.^516^664^.^55074^Alpha136    PRRINISPTTYQLL                                180
HEM_2_0           -      --------------                               
# XX
# &gt; Q9WVI4^.^516^664^.^55074^Alpha^.^0^0^LIGAND^1cs4;^d1cs4a_^FOK^1^1^0^0^1DF^-1000000.00^0.000e+00^0.000e+00
# XX
Q9WVI4^.^516^664^.^55074^Alpha1      DDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDI 45
FOK_1_0           -      *--------------------------------------------
Q9WVI4^.^516^664^.^55074^Alpha46     YKVETIGDAYCVASGLHRKSLCHAKPIALMALKMMELSEEVLTPD 90
FOK_1_0           -      ---------------------------------------------
Q9WVI4^.^516^664^.^55074^Alpha91     GRPIQMRIGIHSGSVLAGVVGVRMPRYCLFGNNVTLASKFESGSH 135
FOK_1_0           -      ---------------------------------------------
Q9WVI4^.^516^664^.^55074^Alpha136    PRRINISPTTYQLL                                180
FOK_1_0           -      --------------                               
</pre>
</td></tr></table><p>
<p><h3>Directory: results</h3>
<p>This directory contains output files, for example Q99396^.^262^473^.results        Q9DGG6^.^380^560^.results        Q9ERL9^.^480^631^.results        and Q9WVI4^.^516^664^.results.
<p><h3>File: results/Q99396^.^262^473^.results       </h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
LIGID     PATCHES   SITES     SCORE     
MN        1         1         5.33      
101       2         2         4.42      
PO4       1         1         4.00      
POP       1         1         3.00      
MG        1         1         2.00      
HEM       4         4         0.82      
FOK       1         1         -1000000.00
</pre>
</td></tr></table><p>
<p><h3>File: results/Q9DGG6^.^380^560^.results       </h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
LIGID     PATCHES   SITES     SCORE     
MG        1         1         5.50      
101       2         2         4.50      
POP       1         1         4.17      
PO4       1         1         4.00      
MN        1         1         2.33      
HEM       4         4         0.73      
FOK       1         1         -1000000.00
</pre>
</td></tr></table><p>
<p><h3>File: results/Q9ERL9^.^480^631^.results       </h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
LIGID     PATCHES   SITES     SCORE     
MG        1         1         5.50      
101       2         2         4.83      
POP       1         1         4.17      
MN        1         1         4.00      
PO4       1         1         2.50      
HEM       4         4         0.33      
FOK       1         1         -1000000.00
</pre>
</td></tr></table><p>
<p><h3>File: results/Q9WVI4^.^516^664^.results</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
LIGID     PATCHES   SITES     SCORE     
MG        1         1         5.50      
101       2         2         4.83      
POP       1         1         4.17      
PO4       1         1         3.50      
MN        1         1         3.33      
HEM       4         4         0.37      
FOK       1         1         -1000000.00
</pre>
</td></tr></table><p>




<!-- DATA FILES         
     Any data files used (e.g. translation table file, substitution matrix 
     etc.  This includes example data file formats if they are not obvious.
     For a standard description of what data files are and how embossdata can
     be used to inspect and retrieve them, use:
     #include file="inc/localfiles.ihtml"
-->

<a name="5.0"></a>
<br><br><br><H2> 5.0   DATA FILES                </H2>
SIGSCANLIG requires a matrix of 3D:1D scores (residue:environment scoring matrix).




<!-- USAGE
     Example usage, as run from the command-line.
     Many examples illustrating different behaviours is good.
-->
<a name="6.0"></a>
<br><br><br><H2> 6.0   USAGE                     </H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" >
<tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-siginfiles]        dirlist    This option specifies the directory of
                                  signature files (input). A 'signature file'
                                  contains a sparse sequence signature
                                  suitable for use with the SIGSCAN and
                                  SIGSCANLIG programs. The files are generated
                                  by using SIGGEN and SIGGENLIG.
  [-dbsequences]       seqall     This option specifies the name of the
                                  database of query sequences.
   -sub                matrixf    This option specifies the scoring matrix to
                                  use, e.g. residue substitution matrix (1D
                                  signatures) or matrix of 3D:1D
                                  (environment:residue) scores (3D).
   -mode               menu       This option specifies the mode of scoring
                                  ligands. In 'Patch Score Mode' (mode 1) the
                                  score for a ligand is the mean of its patch
                                  scores. In 'Site Score Mode' the score for a
                                  ligand is the mean of its site scores. A
                                  site score is the mean of patch score for
                                  all patches for a given site.
   -gapo               float      This option specifies the gap insertion
                                  penalty. The gap insertion penalty is the
                                  score taken away when a gap is created. The
                                  best value depends on the choice of
                                  comparison matrix. The default value assumes
                                  you are using the EBLOSUM62 matrix for
                                  protein sequences, and the EDNAMAT matrix
                                  for nucleotide sequences.
   -gape               float      This option specifies the gap extension
                                  penalty. The gap extension penalty is added
                                  to the standard gap penalty for each base or
                                  residue in the gap. This is how long gaps
                                  are penalized. Usually you will expect a few
                                  long gaps rather than many short gaps, so
                                  the gap extension penalty should be lower
                                  than the gap penalty.
   -nterm              menu       This option specifies the N-terminal
                                  matching option. This determines how the
                                  first signature position is aligned to a
                                  sequence from the database.
  [-hitsdir]           outdir     This option specifies the directory of the
                                  LHF files (ligand hits files) (output). A
                                  'ligand hits file' contains database hits
                                  (sequences) with ligand classification
                                  information, in the LHF format (FASTA-like).
                                  The hits are putative ligand-binding
                                  sequences and are found from a search of a
                                  sequence database, in this case, by using
                                  SIGSCANLIG.
  [-aligndir]          outdir     This option specifies the name of the SAF
                                  (signature alignment file) (output).A
                                  'signature alignment file' contains one or
                                  more signnature-sequence alignments. The
                                  file is in DAF format (CLUSTAL-like) and is
                                  annotated with bibliographic information,
                                  either the domain family classification (for
                                  SIGSCAN output) or ligand classification
                                  (for SIGSCANLIG output). The files generated
                                  by SIGSCAN will contain a
                                  signature-sequence alignment for a single
                                  signature against a library of one or more
                                  sequences. The files generated by using
                                  SIGSCANLIG will contain a signature-sequence
                                  alignment for a single query sequence
                                  against a library of one or more signatures.
  [-resultsdir]        outdir     This option specifies the directory for the
                                  results files. A results file is generated
                                  for each query sequence and contains a list
                                  of hits (ligands), rank-ordered on the basis
                                  of score.

   Additional (Optional) qualifiers (* if not always prompted):
   -nhits              integer    This option specifies the maximum number of
                                  hits to output.
   -domax              toggle     This option specifies whether to consider a
                                  max. number of top-scoring hits only for
                                  each ligand.
*  -maxhits            integer    This option specifies the maximum number of
                                  top-scoring hits to consider for each
                                  ligand.

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-dbsequences" associated qualifiers
   -sbegin2            integer    Start of each sequence to be used
   -send2              integer    End of each sequence to be used
   -sreverse2          boolean    Reverse (if DNA)
   -sask2              boolean    Ask for begin/end/reverse
   -snucleotide2       boolean    Sequence is nucleotide
   -sprotein2          boolean    Sequence is protein
   -slower2            boolean    Make lower case
   -supper2            boolean    Make upper case
   -sformat2           string     Input sequence format
   -sdbname2           string     Database name
   -sid2               string     Entryname
   -ufo2               string     UFO features
   -fformat2           string     Features format
   -fopenfile2         string     Features file name

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report deaths

</pre>
</pre>
</td></tr>
</table>

<H3> 6.1   COMMAND LINE ARGUMENTS </H3>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-siginfiles]<br>(Parameter 1)</td>
<td>This option specifies the directory of signature files (input). A 'signature file' contains a sparse sequence signature suitable for use with the SIGSCAN and SIGSCANLIG programs. The files are generated by using SIGGEN and SIGGENLIG.</td>
<td>Directory with files</td>
<td>./</td>
</tr>

<tr>
<td>[-dbsequences]<br>(Parameter 2)</td>
<td>This option specifies the name of the database of query sequences.</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-sub</td>
<td>This option specifies the scoring matrix to use, e.g. residue substitution matrix (1D signatures) or matrix of 3D:1D (environment:residue) scores (3D).</td>
<td>Comparison matrix file in EMBOSS data path</td>
<td>EBLOSUM62</td>
</tr>

<tr>
<td>-mode</td>
<td>This option specifies the mode of scoring ligands. In 'Patch Score Mode' (mode 1) the score for a ligand is the mean of its patch scores. In 'Site Score Mode' the score for a ligand is the mean of its site scores. A site score is the mean of patch score for all patches for a given site.</td>
<td><table><tr><td>1</td> <td><i>(Patch score mode)</i></td></tr><tr><td>2</td> <td><i>(Site score mode)</i></td></tr></table></td>
<td>1</td>
</tr>

<tr>
<td>-gapo</td>
<td>This option specifies the gap insertion penalty. The gap insertion penalty is the score taken away when a gap is created. The best value depends on the choice of comparison matrix. The default value assumes you are using the EBLOSUM62 matrix for protein sequences, and the EDNAMAT matrix for nucleotide sequences.</td>
<td>Floating point number from 1.0 to 100.0</td>
<td>10.0 for any sequence</td>
</tr>

<tr>
<td>-gape</td>
<td>This option specifies the gap extension penalty. The gap extension penalty is added to the standard gap penalty for each base or residue in the gap. This is how long gaps are penalized. Usually you will expect a few long gaps rather than many short gaps, so the gap extension penalty should be lower than the gap penalty.</td>
<td>Floating point number from 0.0 to 10.0</td>
<td>0.5 for any sequence</td>
</tr>

<tr>
<td>-nterm</td>
<td>This option specifies the N-terminal matching option. This determines how the first signature position is aligned to a sequence from the database.</td>
<td><table><tr><td>1</td> <td><i>(Align anywhere and allow only complete signature-sequence fit)</i></td></tr><tr><td>2</td> <td><i>(Align anywhere and allow partial signature-sequence fit)</i></td></tr><tr><td>3</td> <td><i>(Use empirical gaps only)</i></td></tr></table></td>
<td>1</td>
</tr>

<tr>
<td>[-hitsdir]<br>(Parameter 3)</td>
<td>This option specifies the directory of the LHF files (ligand hits files) (output). A 'ligand hits file' contains database hits (sequences) with ligand classification information, in the LHF format (FASTA-like). The hits are putative ligand-binding sequences and are found from a search of a sequence database, in this case, by using SIGSCANLIG.</td>
<td>Output directory</td>
<td>./</td>
</tr>

<tr>
<td>[-aligndir]<br>(Parameter 4)</td>
<td>This option specifies the name of the SAF (signature alignment file) (output).A 'signature alignment file' contains one or more signnature-sequence alignments. The file is in DAF format (CLUSTAL-like) and is annotated with bibliographic information, either the domain family classification (for SIGSCAN output) or ligand classification (for SIGSCANLIG output). The files generated by SIGSCAN will contain a signature-sequence alignment for a single signature against a library of one or more sequences. The files generated by using SIGSCANLIG will contain a signature-sequence alignment for a single query sequence against a library of one or more signatures.</td>
<td>Output directory</td>
<td>./</td>
</tr>

<tr>
<td>[-resultsdir]<br>(Parameter 5)</td>
<td>This option specifies the directory for the results files. A results file is generated for each query sequence and contains a list of hits (ligands), rank-ordered on the basis of score.</td>
<td>Output directory</td>
<td>./</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-nhits</td>
<td>This option specifies the maximum number of hits to output.</td>
<td>Any integer value</td>
<td>100</td>
</tr>

<tr>
<td>-domax</td>
<td>This option specifies whether to consider a max. number of top-scoring hits only for each ligand.</td>
<td>Toggle value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-maxhits</td>
<td>This option specifies the maximum number of top-scoring hits to consider for each ligand.</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H3> 6.2 EXAMPLE SESSION </H3>
An example of interactive use of SIGSCANLIG is shown below.

<b>Here is a sample session with sigscanlig</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>sigscanlig </b>
Search ligand-signature library & write hits (LHF file).
Directory of signature files (input) [./]: <b>../siggenlig-keep</b>
Database of query sequences.: <b>swtiny</b>
Scoring matrix [EBLOSUM62]: <b></b>
Available modes
         1 : Patch score mode
         2 : Site score mode
Select mode of scoring ligands. [1]: <b>1</b>
Gap insertion penalty [10]: <b></b>
Gap extension penalty [0.5]: <b></b>
N-terminal matching options
         1 : Align anywhere and allow only complete signature-sequence fit
         2 : Align anywhere and allow partial signature-sequence fit
         3 : Use empirical gaps only
Select number [1]: <b></b>
Directory of LHF files (ligand hits file) (output) [./]: <b>lhf</b>
Directory of alignment files for signature-sequence alignments (output) [./]: <b>aln</b>
Directory of results file (output) [./]: <b>results</b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>




<!-- KNOWN BUGS & WARNINGS
     Bugs that have not yet been fixed, easily missued features, problems
     and caveats etc. Potentially stupid things the program will let you do.
-->
<a name="7.0"></a>
<br><br><br><H2> 7.0   KNOWN BUGS & WARNINGS     </H2>
None.



<!-- NOTES
     Important general remarks, including:
       Restrictions.
       Interesting behaviour.
       Useful things you can do with this program.
       Future plans.
       etc.
-->
<a name="8.0"></a>
<br><br><br><H2> 8.0   NOTES                     </H2>
SIGSCANLIG does not generate p-values or E-values and a value of '0' is given in the LHF files.

<H3> 8.1   GLOSSARY OF FILE TYPES    </H3>
<table BORDER CELLSPACING=5 CELLPADDING=5 BGCOLOR="#f5f5ff" >
<tr>
<td><b>FILE TYPE</b></td>
<td><b>FORMAT</b></td>
<td><b>DESCRIPTION</b></td>
<td><b>CREATED BY <b></td>
<td><b>SEE ALSO</b></td>
</tr>
<tr>
<td><b>Ligand hits file </b></td>
<td> LHF format (FASTA-like). </td>
<td> Database hits (sequences) with ligand classification information.  The hits are putative ligand-binding sequences and are found from a search of a discriminating element (e.g. a protein signature) against a sequence database. </td>
<td> <a href="sigscanlig.html">SIGSCANLIG</a> (hits retrieved by 1D and 3D signatures). </td>
<td> N.A. </td>
</tr>
<tr>
<td><b>Signature alignment file </b></td>
<td> SAF format (CLUSTAL-like). </td>
<td> One or more signnature-sequence alignments, annotated with bibliographic information, either the domain family classification (for SIGSCAN output) or ligand classification (for SIGSCANLIG output). </td>
<td> <a href="sigscan.html">SIGSCAN</a> (signature-sequence alignments for a single signature against a database of one or more sequences) or <a href="sigscanlig.html">SIGSCANLIG</a> (signature-sequence alignment for a single query sequence against a library of one or more signatures). </td>
<td> N.A. </td>
</tr>
<tr>
<td><b>Signature file </b></td>
<td> SIG format </td>
<td> Contains a sparse sequence signature suitable for use with the SIGSCANLIG program. Contains a sparse sequence signature. </td>
<td> <a href="siggen.html">SIGGEN</a>, <a href="siggenlig.html">SIGGENLIG</a>, <a href="libgen.html">LIBGEN</a> </td>
<td> The files are generated by using SIGGEN. </td>
</tr>
</table>


<!-- DESCRIPTION
     A complete, non-technical, user-level description of the application.
-->
<a name="9.0"></a>
<br><br><br><H2> 9.0   DESCRIPTION               </H2>

SIGSCANLIG is an application to screen a query protein sequence against a library of signatures for ligand-binding. 
See Blades et al., Ison et al. and Daniel et al. for a description of other types of protein signatures and their application.




<!-- ALGORITHM
     A technical description of algorithmic aspects, describing exactly how
     the key aspects of the application work.
-->
<a name="10.0"></a>
<br><br><br><H2> 10.0  ALGORITHM                 </H2>

The algorithm is based on an approach first described in Daniel et al (1999) 
that was applied to the definition of protein families (Ison et al, 2000)
and later to automatically-generated signatures (Blades et al, 2005).

<br><br><B>Description of scoring modes</B><br>
The user specifies the mode of scoring ligands, i.e. the way in which the scores in the results files are generated. In 'Patch Score Mode' (mode 1) the score for a ligand is the mean of its patch scores.  In 'Site Score Mode' the score for a ligand is the mean of its site scores.  A site score is the mean of patch score for all patches for a given site.

<br><br><B>Selection of gap penalties</B><br>
The user specifies the gap insertion and extension penalties. The gap insertion penalty is the score taken away when a gap is created. The gap extension penalty is added to the standard gap penalty for each base or residue in the gap. This is how long gaps are penalized.  Usually you will expect a few long gaps rather than many short gaps, so the gap extension penalty should be lower than the gap penalty. The best value of these penalties depends on the choice of comparison matrix. The default value assumes you are using the EBLOSUM62 matrix for 1D signature scans.  If you are performing 3D signature scans you should select a value appropriate to the scoring matrix used.






<!-- RELATED APPLICATIONS
     Other applications that either generate the input, use the output or 
     are in some other way related to the application are described here. 
-->
<a name="11.0"></a>
<br><br><br><H2> 11.0  RELATED APPLICATIONS      </H2>
<a name="11.0"></a>
<br><br><br><br><br><br><H2> 11.0  RELATED APPLICATIONS      </H2>

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="../../apps/contactcount.html">contactcount</a></td><td>Count specific versus non-specific contacts</td></tr>
<tr><td><a href="../structure/contacts.html">contacts</a></td><td>Generate intra-chain CON files from CCF files</td></tr>
<tr><td><a href="../domalign/domainalign.html">domainalign</a></td><td>Generate alignments (DAF file) for nodes in a DCF file</td></tr>
<tr><td><a href="../domalign/domainrep.html">domainrep</a></td><td>Reorder DCF file to identify representative structures</td></tr>
<tr><td><a href="../domainatrix/domainreso.html">domainreso</a></td><td>Remove low resolution domains from a DCF file</td></tr>
<tr><td><a href="../structure/interface.html">interface</a></td><td>Generate inter-chain CON files from CCF files</td></tr>
<tr><td><a href="libgen.html">libgen</a></td><td>Generate discriminating elements from alignments</td></tr>
<tr><td><a href="matgen3d.html">matgen3d</a></td><td>Generate a 3D-1D scoring matrix from CCF files</td></tr>
<tr><td><a href="../../apps/psiphi.html">psiphi</a></td><td>Phi and psi torsion angles from protein coordinates</td></tr>
<tr><td><a href="rocon.html">rocon</a></td><td>Generates a hits file from comparing two DHF files</td></tr>
<tr><td><a href="rocplot.html">rocplot</a></td><td>Performs ROC analysis on hits files</td></tr>
<tr><td><a href="../reconstruct/scorecmapdir.html">scorecmapdir</a></td><td>Contact scores for cleaned protein chain contact files</td></tr>
<tr><td><a href="../domalign/seqalign.html">seqalign</a></td><td>Extend alignments (DAF file) with sequences (DHF file)</td></tr>
<tr><td><a href="../domsearch/seqfraggle.html">seqfraggle</a></td><td>Removes fragment sequences from DHF files</td></tr>
<tr><td><a href="../domsearch/seqsearch.html">seqsearch</a></td><td>Generate PSI-BLAST hits (DHF file) from a DAF file</td></tr>
<tr><td><a href="../domsearch/seqsort.html">seqsort</a></td><td>Remove ambiguous classified sequences from DHF files</td></tr>
<tr><td><a href="../domsearch/seqwords.html">seqwords</a></td><td>Generates DHF files from keyword search of UniProt</td></tr>
<tr><td><a href="siggen.html">siggen</a></td><td>Generates a sparse protein signature from an alignment</td></tr>
<tr><td><a href="siggenlig.html">siggenlig</a></td><td>Generate ligand-binding signatures from a CON file</td></tr>
<tr><td><a href="sigscan.html">sigscan</a></td><td>Generate hits (DHF file) from a signature search</td></tr>
</table>








<!-- DIAGNOSTIC ERROR MESSAGES
     Description of error messages or log file, if one is written.
-->
<a name="12.0"></a>
<br><br><br><H2> 12.0  DIAGNOSTIC ERROR MESSAGES </H2>

None.







<!-- AUTHORS
     The main author first then all others. 
-->
<a name="13.0"></a>
<br><br><br><H2> 13.0  AUTHORS                   </H2>
Jon Ison <a href="mailto:jison@rfcgr.mrc.ac.uk">(jison@rfcgr.mrc.ac.uk)</a>
<br><br>

Waqas Awan <a href="mailto:wawan@rfcgr.mrc.ac.uk">(mblades@rfcgr.mrc.ac.uk)</a>
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK



<!-- REFERENCES
     Quote the paper where the application was first published, described, used etc. 
-->
<a name="14.0"></a>
<br><br><br><H2> 14.0  REFERENCES                </H2>
Automatic generation and evaluation of sparse protein signatures for families of protein structural domains.  Blades MJ, Ison JC, Ranasinghe R and Findlay, JBC. Protein Science (2005), 14:13-23.

<br><br>
A key residues approach to the definition of protein families and analysis of sparse family signatures.  JC Ison, AJ Bleasby, MJ Blades, SC Daniel, JH Parish, JBC Findlay.  PROTEINS: Structure, Function & Genetics.  2000, 40:330-341

<br><br>
Alignment of a sparse protein signature with protein sequences: application to fold prediction for three small globulins.  SC Daniel, JH Parish, JC Ison, MJ Blades & JBC Findlay.  FEBS Letters.  1999, 459:349-352.

Please cite the authors and EMBOSS.
<br><br>
<i>Rice P, Longden I and Bleasby A (2000) "EMBOSS - The European
Molecular Biology Open Software Suite"  Trends in Genetics,
15:276-278.

<p>
See also <a href="http://emboss.sourceforge.net/">http://emboss.sourceforge.net/</a></i>



<H3>14.1 Other useful references  </H3>
 

</BODY>
</HTML>
