<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: sirna
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

sirna
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<!--END OF HEADER-->






<H2>
    Function
</H2>
Finds siRNA duplexes in mRNA

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
  -->

RNA interference, or RNAi, is a phenomenon in which double stranded RNA
(dsRNA) effects silencing of the expression of genes that are highly
homologous to either of the RNA strands in the duplex.  Gene silencing
in RNAi results from the degradation of mRNA sequences, and the effect
has been used to determine the function of many genes in Drosophilia, C. 
elegans, and many plant species. 

<p>

The duration of knockdown by siRNA can typically last for 7-10 days, and
has been shown to transfer to daughter cells.  Of further note, siRNAs
are effective at quantities much lower than alternative gene silencing
methodologies, including antisense and ribozyme based strategies. 

<p>
                       
Due to various mechanisms of antiviral response to long dsRNA, RNAi at
first proved more difficult to establish in mammalian species.  Then,
Tuschl, Elbashir, and others discovered that RNAi can be elicited very
effectively by well-defined 21-base duplex RNAs.  When these small
interfering RNA, or siRNA, are added in duplex form with a transfection
agent to mammalian cell cultures, the 21-base-pair RNA acts in concert
with cellular components to silence the gene with sequence homology to
one of the siRNA sequences. 

Strategies for the design of effective siRNA sequences have been
recently documented, most notably by Sayda Elbashir, Thomas Tuschl, et
al. 

<p>

Their studies of mammalian RNAi suggest that the most efficient
gene-silencing effect is achieved using double-stranded siRNA having a
19-nucleotide complementary region and a 2-nucleotide 3' overhang at
each end.  Current models of the RNAi mechanism suggest that the
antisense siRNA strand recognizes the specific gene target. 

<p>

In gene-specific RNAi, the coding region (CDS) of the mRNA is usually
targeted.  The search for an appropriate target sequence should begin
50-100 nucleotides downstream of the start codon.  UTR-binding proteins
and/or translation initiation complexes may interfere with the binding
of the siRNP endonuclease complex.  Tuschl, Elbashir et al.  say that
they have successfully used siRNAs targetting the 3' UTR. 


To avoid interference
from mRNA regulatory proteins, sequences in the 5' untranslated region
or near the start codon should not be targeted. 

<p>

A set of rules for the design of siRNA has been suggested <A href
="http://www.mpibpc.gwdg.de/abteilungen/100/105/sirna.html
">http://www.mpibpc.gwdg.de/abteilungen/100/105/sirna.html</a> based on
the work of Tuschl, Elbashir et al. 

<p>

They suggest searching for 23-nt sequence motif AA(N19)TT (N, any
nucleotide) and select hits with approx.  50% G/C-content (30% to 70%
has also worked in for them).  If no suitable sequences are found, the
search is extended using the motif NA(N21).  The sequence of the sense
siRNA corresponds to (N19)TT or N21 (position 3 to 23 of the 23-nt
motif), respectively.  In the latter case, they convert the 3' end of
the sense siRNA to TT. 

<p>

The rationale for this sequence conversion is to generate a symmetric
duplex with respect to the sequence composition of the sense and
antisense 3' overhangs.  The antisense siRNA is synthesized as the
complement to position 1 to 21 of the 23-nt motif.  Because position 1
of the 23-nt motif is not recognized sequence-specifically by the
antisense siRNA, the 3'-most nucleotide residue of the antisense siRNA,
can be chosen deliberately.  However, the penultimate nucleotide of the
antisense siRNA (complementary to position 2 of the 23-nt motif) should
always be complementary to the targeted sequence.  For simplifying
chemical synthesis, they always use TT.  

<p>

More recently, they preferentially select siRNAs corresponding to the
target motif NAR(N17)YNN, where R is purine (A, G) and Y is pyrimidine
(C, U).  The respective 21-nt sense and antisense siRNAs therefore begin
with a purine nucleotide and can also be expressed from pol III
expression vectors without a change in targeting site; expression of
RNAs from pol III promoters is only efficient when the first transcribed
nucleotide is a purine. 

<p>

They always design siRNAs with symmetric 3' TT overhangs, believing that
symmetric 3' overhangs help to ensure that the siRNPs are formed with
approximately equal ratios of sense and antisense target RNA-cleaving
siRNPs Please note that the modification of the overhang of the sense
sequence of the siRNA duplex is not expected to affect targeted mRNA
recognition, as the antisense siRNA strand guides target recognition. 
In summary, no matter what you do to your overhangs, siRNAs should still
function to a reasonable extent.  However, using TT in the 3' overhang
will always help your RNA synthesis company to let you know when you
accidentally order a siRNA sequences 3' to 5' rather than in the
recommended format of 5' to 3'.  

<p>

<b>sirna</b> reports both the sense and antisense siRNAs as 5' to 3'. 

<p>

<a href="http://www.xeragon.com/siRNA_support.html">Xeragon.com</a> also
suggest that choosing a region of the mRNA with a GC content as
close as possible to 50% is a more important consideration than choosing
a target sequence that begins with AA.  They also suggest that a key
consideration in target selection is to avoid having more than three
guanosines in a row, since poly G sequences can hyperstack and form
agglomerates that potentially interfere with the siRNA silencing
mechanism.

<p>

siRNAs appear to effectively silence genes in more than 80% of cases. 
Current data indicate that there are regions of some mRNAs where gene
silencing does not work.  To help ensure that a given target gene is
silenced, it is advised that at least two target sequences as far apart
on the gene as possible be chosen. 

<p>

mRNA secondary structure does not appear to have a significant effect on
gene silencing. 

<p>

<b>sirna</b> looks at the feature table of the input mRNA sequence to
find the coding regions (CDS).  It will ignore the 5' UTR and the first
50 bases of the CDS.  It will assign a penalty of 2 points to any siRNA
in positions 51 to 100 in the CDS.  If there is no CDS in the feature
table, you can specify the CDS by using the command-line qualifier
'-sbegin' to indicate where the CDS should start.  If there is no CDS in
the feature table and you do not use the command-line qualifier
'-sbegin', then <b>sirna</b> will assume that the CDS region is not
known and will look for siRNAs in the whole of the sequence with no
penalties. 


<H2>
    Algorithm
</H2>

<pre>
for each input sequence:

    find the start position of the CDS in the feature table
    if there is no such CDS, take the -sbegin position as the CDS start

    for each 23 base window along the sequence:

        set the score for this window = 0
        if base 2 of the window is not 'a': ignore this window
        if the window is within 50 bases of the CDS start: ignore this window
	if the window is within 100 bases of the CDS: score = -2
	measure the %GC of the 20 bases from position 2 to 21 of the window
	for the following %GC values change the score:
		%GC <= 25% (<= 5 bases): ignore this window
		%GC 30% (6 bases): score + 0
		%GC 35% (7 bases): score + 2
		%GC 40% (8 bases): score + 4
		%GC 45% (9 bases): score + 5
		%GC 50% (10 bases): score + 6
		%GC 55% (11 bases): score + 5
		%GC 60% (12 bases): score + 4
		%GC 65% (13 bases): score + 2
		%GC 70% (14 bases): score + 0
		%GC >= 75% (>= 15 bases): ignore this window
	if the window starts with a 'AA': score + 3
	if the window does not start 'AA' and it is required: ignore this window
	if the window ends with a 'TT': score + 1
	if the window does not end 'TT' and it is required: ignore this window
	if 4 G's in a row are found: ignore this window
	if any 4 bases in a row are present and not required: ignore this window
	if PolIII probes are required and the window is not NARN(17)YNN: ignore this window
        if the score is > 0: store this window for output
	
    sort the windows found by their score
    output the 23-base windows to the sequence file
    take the window bases 2 to 21, add 'dTdT' output to the report file
    take the window bases 2 to 21, reverse complement, add 'dTdT' output to the report file
</pre>


<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

<b>Here is a sample session with sirna</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>sirna </b>
Finds siRNA duplexes in mRNA
Input sequence(s): <b>tembl:hsfau</b>
Output report [hsfau.sirna]: <b></b>
Output sequence [hsfau.fasta]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>




<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
  [-outfile]           report     The output is a table of the forward and
                                  reverse parts of the siRNA duplex. Both the
                                  forward and reverse sequences are written 5'
                                  to 3', ready to be ordered. The last two
                                  bases have been replaced by 'dTdT'. The
                                  starting position of the 23 base region and
                                  the %GC content is also given.
  [-outseq]            seqoutall  This is a file of the sequences of the 23
                                  base regions that the siRNAs are selected
                                  from. You may use it to do searches of mRNA
                                  databases (e.g. REFSEQ) to confirm that the
                                  probes are unique to the gene you wish to
                                  use it on.

   Optional qualifiers:
   -poliii             boolean    This option allows you to select only the 21
                                  base probes that start with a purine and so
                                  can be expressed from Pol III expression
                                  vectors. This is the NARN(17)YNN pattern
                                  that has been suggested by Tuschl et al.
   -aa                 boolean    This option allows you to select only those
                                  23 base regions that start with AA. If this
                                  option is not selected then regions that
                                  start with AA will be favoured by giving
                                  them a higher score, but regions that do not
                                  start with AA will also be reported.
   -tt                 boolean    This option allows you to select only those
                                  23 base regions that end with TT. If this
                                  option is not selected then regions that end
                                  with TT will be favoured by giving them a
                                  higher score, but regions that do not end
                                  with TT will also be reported.
   -[no]polybase       boolean    If this option is FALSE then only those 23
                                  base regions that have no repeat of 4 or
                                  more of any bases in a row will be reported.
                                  No regions will ever be reported that have
                                  4 or more G's in a row.

   Advanced qualifiers: (none)
   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>The output is a table of the forward and reverse parts of the siRNA duplex. Both the forward and reverse sequences are written 5' to 3', ready to be ordered. The last two bases have been replaced by 'dTdT'. The starting position of the 23 base region and the %GC content is also given.</td>
<td>Report output file</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>[-outseq]<br>(Parameter 3)</td>
<td>This is a file of the sequences of the 23 base regions that the siRNAs are selected from. You may use it to do searches of mRNA databases (e.g. REFSEQ) to confirm that the probes are unique to the gene you wish to use it on.</td>
<td>Writeable sequence(s)</td>
<td><i>&lt;sequence&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-poliii</td>
<td>This option allows you to select only the 21 base probes that start with a purine and so can be expressed from Pol III expression vectors. This is the NARN(17)YNN pattern that has been suggested by Tuschl et al.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-aa</td>
<td>This option allows you to select only those 23 base regions that start with AA. If this option is not selected then regions that start with AA will be favoured by giving them a higher score, but regions that do not start with AA will also be reported.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-tt</td>
<td>This option allows you to select only those 23 base regions that end with TT. If this option is not selected then regions that end with TT will be favoured by giving them a higher score, but regions that do not end with TT will also be reported.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]polybase</td>
<td>If this option is FALSE then only those 23 base regions that have no repeat of 4 or more of any bases in a row will be reported. No regions will ever be reported that have 4 or more G's in a row.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>sirna</b> reads any normal sequence USAs.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:hsfau' is a sequence entry in the example nucleic acid database 'tembl'
<p>
<p><h3>Database entry: tembl:hsfau</h3>
<table width="90%"><tr><td bgcolor="#FFDDFF">
<pre>
ID   HSFAU      standard; RNA; HUM; 518 BP.
XX
AC   X65923;
XX
SV   X65923.1
XX
DT   13-MAY-1992 (Rel. 31, Created)
DT   23-SEP-1993 (Rel. 37, Last updated, Version 10)
XX
DE   H.sapiens fau mRNA
XX
KW   fau gene.
XX
OS   Homo sapiens (human)
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia;
OC   Eutheria; Primates; Catarrhini; Hominidae; Homo.
XX
RN   [1]
RP   1-518
RA   Michiels L.M.R.;
RT   ;
RL   Submitted (29-APR-1992) to the EMBL/GenBank/DDBJ databases.
RL   L.M.R. Michiels, University of Antwerp, Dept of Biochemistry,
RL   Universiteisplein 1, 2610 Wilrijk, BELGIUM
XX
RN   [2]
RP   1-518
RX   MEDLINE; 93368957.
RA   Michiels L., Van der Rauwelaert E., Van Hasselt F., Kas K., Merregaert J.;
RT   " fau cDNA encodes a ubiquitin-like-S30 fusion protein and is expressed as
RT   an antisense sequences in the Finkel-Biskis-Reilly murine sarcoma virus";
RL   Oncogene 8:2537-2546(1993).
XX
DR   SWISS-PROT; P35544; UBIM_HUMAN.
DR   SWISS-PROT; Q05472; RS30_HUMAN.
XX
FH   Key             Location/Qualifiers
FH
FT   source          1..518
FT                   /chromosome="11q"
FT                   /db_xref="taxon:9606"
FT                   /organism="Homo sapiens"
FT                   /tissue_type="placenta"
FT                   /clone_lib="cDNA"
FT                   /clone="pUIA 631"
FT                   /map="13"
FT   misc_feature    57..278
FT                   /note="ubiquitin like part"
FT   CDS             57..458
FT                   /db_xref="SWISS-PROT:P35544"
FT                   /db_xref="SWISS-PROT:Q05472"
FT                   /gene="fau"
FT                   /protein_id="CAA46716.1"
FT                   /translation="MQLFVRAQELHTFEVTGQETVAQIKAHVASLEGIAPEDQVVLLAG
FT                   APLEDEATLGQCGVEALTTLEVAGRMLGGKVHGSLARAGKVRGQTPKVAKQEKKKKKTG
FT                   RAKRRMQYNRRFVNVVPTFGKKKGPNANS"
FT   misc_feature    98..102
FT                   /note="nucleolar localization signal"
FT   misc_feature    279..458
FT                   /note="S30 part"
FT   polyA_signal    484..489
FT   polyA_site      509
XX
SQ   Sequence 518 BP; 125 A; 139 C; 148 G; 106 T; 0 other;
     ttcctctttc tcgactccat cttcgcggta gctgggaccg ccgttcagtc gccaatatgc        60
     agctctttgt ccgcgcccag gagctacaca ccttcgaggt gaccggccag gaaacggtcg       120
     cccagatcaa ggctcatgta gcctcactgg agggcattgc cccggaagat caagtcgtgc       180
     tcctggcagg cgcgcccctg gaggatgagg ccactctggg ccagtgcggg gtggaggccc       240
     tgactaccct ggaagtagca ggccgcatgc ttggaggtaa agttcatggt tccctggccc       300
     gtgctggaaa agtgagaggt cagactccta aggtggccaa acaggagaag aagaagaaga       360
     agacaggtcg ggctaagcgg cggatgcagt acaaccggcg ctttgtcaac gttgtgccca       420
     cctttggcaa gaagaagggc cccaatgcca actcttaagt cttttgtaat tctggctttc       480
     tctaataaaa aagccactta gttcagtcaa aaaaaaaa                               518
//
</pre>
</td></tr></table><p>






<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 

        If you wish to include the standard description of the avalable
	report formats, use:
   -->
<p>

The output is a standard EMBOSS report file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-rformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  The available format names are: embl,
genbank, gff, pir, swiss, trace, listfile, dbmotif, diffseq, excel,
feattable, motif, regions, seqtable, simple, srs, table, tagseq

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html</A>
for further information on report formats.

<p>

<p>

<b>sirna</b> 
outputs a report format file. The default format is 'table'

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: hsfau.fasta</h3>
<table width="90%"><tr><td bgcolor="#DDFFDD">
<pre>
&gt;HSFAU_308 %GC 50.0 Score 9 H.sapiens fau mRNA
aaaagtgagaggtcagactccta
&gt;HSFAU_309 %GC 50.0 Score 9 H.sapiens fau mRNA
aaagtgagaggtcagactcctaa
&gt;HSFAU_310 %GC 50.0 Score 9 H.sapiens fau mRNA
aagtgagaggtcagactcctaag
&gt;HSFAU_351 %GC 50.0 Score 9 H.sapiens fau mRNA
aagaagaagaagacaggtcgggc
&gt;HSFAU_166 %GC 55.0 Score 8 H.sapiens fau mRNA
aagatcaagtcgtgctcctggca
&gt;HSFAU_279 %GC 55.0 Score 8 H.sapiens fau mRNA
aaagttcatggttccctggcccg
&gt;HSFAU_330 %GC 55.0 Score 8 H.sapiens fau mRNA
aaggtggccaaacaggagaagaa
&gt;HSFAU_354 %GC 55.0 Score 8 H.sapiens fau mRNA
aagaagaagacaggtcgggctaa
&gt;HSFAU_357 %GC 55.0 Score 8 H.sapiens fau mRNA
aagaagacaggtcgggctaagcg
&gt;HSFAU_393 %GC 55.0 Score 8 H.sapiens fau mRNA
aaccggcgctttgtcaacgttgt
&gt;HSFAU_253 %GC 60.0 Score 7 H.sapiens fau mRNA
aagtagcaggccgcatgcttgga
&gt;HSFAU_280 %GC 60.0 Score 7 H.sapiens fau mRNA
aagttcatggttccctggcccgt
&gt;HSFAU_339 %GC 40.0 Score 7 H.sapiens fau mRNA
aaacaggagaagaagaagaagaa
&gt;HSFAU_340 %GC 40.0 Score 7 H.sapiens fau mRNA
aacaggagaagaagaagaagaag
&gt;HSFAU_348 %GC 40.0 Score 7 H.sapiens fau mRNA
aagaagaagaagaagacaggtcg
&gt;HSFAU_375 %GC 60.0 Score 7 H.sapiens fau mRNA
aagcggcggatgcagtacaaccg
&gt;HSFAU_408 %GC 60.0 Score 7 H.sapiens fau mRNA
aacgttgtgcccacctttggcaa
&gt;HSFAU_429 %GC 60.0 Score 7 H.sapiens fau mRNA
aagaagaagggccccaatgccaa
&gt;HSFAU_432 %GC 60.0 Score 7 H.sapiens fau mRNA
aagaagggccccaatgccaactc
&gt;HSFAU_435 %GC 60.0 Score 7 H.sapiens fau mRNA
aagggccccaatgccaactctta
&gt;HSFAU_488 %GC 40.0 Score 7 H.sapiens fau mRNA
aaaaagccacttagttcagtcaa
&gt;HSFAU_489 %GC 40.0 Score 7 H.sapiens fau mRNA
aaaagccacttagttcagtcaaa
&gt;HSFAU_490 %GC 40.0 Score 7 H.sapiens fau mRNA
aaagccacttagttcagtcaaaa
&gt;HSFAU_491 %GC 40.0 Score 7 H.sapiens fau mRNA
aagccacttagttcagtcaaaaa
&gt;HSFAU_129 %GC 55.0 Score 6 H.sapiens fau mRNA
aaggctcatgtagcctcactgga


<font color=red>  [Part of this file has been deleted for brevity]</font>

gaggccctgactaccctggaagt
&gt;HSFAU_259 %GC 60.0 Score 4 H.sapiens fau mRNA
caggccgcatgcttggaggtaaa
&gt;HSFAU_266 %GC 40.0 Score 4 H.sapiens fau mRNA
catgcttggaggtaaagttcatg
&gt;HSFAU_342 %GC 40.0 Score 4 H.sapiens fau mRNA
caggagaagaagaagaagaagac
&gt;HSFAU_347 %GC 40.0 Score 4 H.sapiens fau mRNA
gaagaagaagaagaagacaggtc
&gt;HSFAU_359 %GC 60.0 Score 4 H.sapiens fau mRNA
gaagacaggtcgggctaagcggc
&gt;HSFAU_111 %GC 55.0 Score 3 H.sapiens fau mRNA
gaaacggtcgcccagatcaaggc
&gt;HSFAU_113 %GC 65.0 Score 3 H.sapiens fau mRNA
aacggtcgcccagatcaaggctc
&gt;HSFAU_172 %GC 70.0 Score 3 H.sapiens fau mRNA
aagtcgtgctcctggcaggcgcg
&gt;HSFAU_443 %GC 35.0 Score 3 H.sapiens fau mRNA
caatgccaactcttaagtctttt
&gt;HSFAU_456 %GC 35.0 Score 3 H.sapiens fau mRNA
taagtcttttgtaattctggctt
&gt;HSFAU_468 %GC 30.0 Score 3 H.sapiens fau mRNA
aattctggctttctctaataaaa
&gt;HSFAU_484 %GC 30.0 Score 3 H.sapiens fau mRNA
aataaaaaagccacttagttcag
&gt;HSFAU_108 %GC 60.0 Score 2 H.sapiens fau mRNA
caggaaacggtcgcccagatcaa
&gt;HSFAU_135 %GC 60.0 Score 2 H.sapiens fau mRNA
catgtagcctcactggagggcat
&gt;HSFAU_139 %GC 60.0 Score 2 H.sapiens fau mRNA
tagcctcactggagggcattgcc
&gt;HSFAU_150 %GC 60.0 Score 2 H.sapiens fau mRNA
gagggcattgccccggaagatca
&gt;HSFAU_171 %GC 65.0 Score 2 H.sapiens fau mRNA
caagtcgtgctcctggcaggcgc
&gt;HSFAU_201 %GC 65.0 Score 2 H.sapiens fau mRNA
gaggatgaggccactctgggcca
&gt;HSFAU_204 %GC 65.0 Score 2 H.sapiens fau mRNA
gatgaggccactctgggccagtg
&gt;HSFAU_245 %GC 65.0 Score 2 H.sapiens fau mRNA
taccctggaagtagcaggccgca
&gt;HSFAU_256 %GC 65.0 Score 2 H.sapiens fau mRNA
tagcaggccgcatgcttggaggt
&gt;HSFAU_285 %GC 65.0 Score 2 H.sapiens fau mRNA
catggttccctggcccgtgctgg
&gt;HSFAU_338 %GC 35.0 Score 2 H.sapiens fau mRNA
caaacaggagaagaagaagaaga
&gt;HSFAU_345 %GC 35.0 Score 2 H.sapiens fau mRNA
gagaagaagaagaagaagacagg
&gt;HSFAU_486 %GC 35.0 Score 2 H.sapiens fau mRNA
taaaaaagccacttagttcagtc
</pre>
</td></tr></table><p>
<p><h3>File: hsfau.sirna</h3>
<table width="90%"><tr><td bgcolor="#DDFFDD">
<pre>
########################################
# Program: sirna
# Rundate: Wed Dec 11 14:28:16 2002
# Report_format: table
# Report_file: hsfau.sirna
########################################

#=======================================
#
# Sequence: HSFAU     from: 1   to: 518
# HitCount: 85
#
# CDS region found in feature table starting at 57
#
#=======================================

  Start     End   Score    GC% Sense_siRNA Antisense_siRNA
    308     330   9.000   50.0 AAGUGAGAGGUCAGACUCCdTdT GGAGUCUGACCUCUCACUUdTdT
    309     331   9.000   50.0 AGUGAGAGGUCAGACUCCUdTdT AGGAGUCUGACCUCUCACUdTdT
    310     332   9.000   50.0 GUGAGAGGUCAGACUCCUAdTdT UAGGAGUCUGACCUCUCACdTdT
    351     373   9.000   50.0 GAAGAAGAAGACAGGUCGGdTdT CCGACCUGUCUUCUUCUUCdTdT
    166     188   8.000   55.0 GAUCAAGUCGUGCUCCUGGdTdT CCAGGAGCACGACUUGAUCdTdT
    279     301   8.000   55.0 AGUUCAUGGUUCCCUGGCCdTdT GGCCAGGGAACCAUGAACUdTdT
    330     352   8.000   55.0 GGUGGCCAAACAGGAGAAGdTdT CUUCUCCUGUUUGGCCACCdTdT
    354     376   8.000   55.0 GAAGAAGACAGGUCGGGCUdTdT AGCCCGACCUGUCUUCUUCdTdT
    357     379   8.000   55.0 GAAGACAGGUCGGGCUAAGdTdT CUUAGCCCGACCUGUCUUCdTdT
    393     415   8.000   55.0 CCGGCGCUUUGUCAACGUUdTdT AACGUUGACAAAGCGCCGGdTdT
    253     275   7.000   60.0 GUAGCAGGCCGCAUGCUUGdTdT CAAGCAUGCGGCCUGCUACdTdT
    280     302   7.000   60.0 GUUCAUGGUUCCCUGGCCCdTdT GGGCCAGGGAACCAUGAACdTdT
    339     361   7.000   40.0 ACAGGAGAAGAAGAAGAAGdTdT CUUCUUCUUCUUCUCCUGUdTdT
    340     362   7.000   40.0 CAGGAGAAGAAGAAGAAGAdTdT UCUUCUUCUUCUUCUCCUGdTdT
    348     370   7.000   40.0 GAAGAAGAAGAAGACAGGUdTdT ACCUGUCUUCUUCUUCUUCdTdT
    375     397   7.000   60.0 GCGGCGGAUGCAGUACAACdTdT GUUGUACUGCAUCCGCCGCdTdT
    408     430   7.000   60.0 CGUUGUGCCCACCUUUGGCdTdT GCCAAAGGUGGGCACAACGdTdT
    429     451   7.000   60.0 GAAGAAGGGCCCCAAUGCCdTdT GGCAUUGGGGCCCUUCUUCdTdT
    432     454   7.000   60.0 GAAGGGCCCCAAUGCCAACdTdT GUUGGCAUUGGGGCCCUUCdTdT
    435     457   7.000   60.0 GGGCCCCAAUGCCAACUCUdTdT AGAGUUGGCAUUGGGGCCCdTdT
    488     510   7.000   40.0 AAAGCCACUUAGUUCAGUCdTdT GACUGAACUAAGUGGCUUUdTdT
    489     511   7.000   40.0 AAGCCACUUAGUUCAGUCAdTdT UGACUGAACUAAGUGGCUUdTdT
    490     512   7.000   40.0 AGCCACUUAGUUCAGUCAAdTdT UUGACUGAACUAAGUGGCUdTdT
    491     513   7.000   40.0 GCCACUUAGUUCAGUCAAAdTdT UUUGACUGAACUAAGUGGCdTdT
    129     151   6.000   55.0 GGCUCAUGUAGCCUCACUGdTdT CAGUGAGGCUACAUGAGCCdTdT
    165     187   6.000   50.0 AGAUCAAGUCGUGCUCCUGdTdT CAGGAGCACGACUUGAUCUdTdT
    278     300   6.000   50.0 AAGUUCAUGGUUCCCUGGCdTdT GCCAGGGAACCAUGAACUUdTdT
    314     336   6.000   50.0 GAGGUCAGACUCCUAAGGUdTdT ACCUUAGGAGUCUGACCUCdTdT
    321     343   6.000   50.0 GACUCCUAAGGUGGCCAAAdTdT UUUGGCCACCUUAGGAGUCdTdT
    323     345   6.000   50.0 CUCCUAAGGUGGCCAAACAdTdT UGUUUGGCCACCUUAGGAGdTdT
    329     351   6.000   50.0 AGGUGGCCAAACAGGAGAAdTdT UUCUCCUGUUUGGCCACCUdTdT
    356     378   6.000   50.0 AGAAGACAGGUCGGGCUAAdTdT UUAGCCCGACCUGUCUUCUdTdT
    419     441   6.000   50.0 CCUUUGGCAAGAAGAAGGGdTdT CCCUUCUUCUUGCCAAAGGdTdT


<font color=red>  [Part of this file has been deleted for brevity]</font>

    252     274   5.000   55.0 AGUAGCAGGCCGCAUGCUUdTdT AAGCAUGCGGCCUGCUACUdTdT
    274     296   5.000   45.0 GGUAAAGUUCAUGGUUCCCdTdT GGGAACCAUGAACUUUACCdTdT
    307     329   5.000   45.0 AAAGUGAGAGGUCAGACUCdTdT GAGUCUGACCUCUCACUUUdTdT
    316     338   5.000   55.0 GGUCAGACUCCUAAGGUGGdTdT CCACCUUAGGAGUCUGACCdTdT
    350     372   5.000   45.0 AGAAGAAGAAGACAGGUCGdTdT CGACCUGUCUUCUUCUUCUdTdT
    353     375   5.000   55.0 AGAAGAAGACAGGUCGGGCdTdT GCCCGACCUGUCUUCUUCUdTdT
    360     382   5.000   65.0 GACAGGUCGGGCUAAGCGGdTdT CCGCUUAGCCCGACCUGUCdTdT
    374     396   5.000   55.0 AGCGGCGGAUGCAGUACAAdTdT UUGUACUGCAUCCGCCGCUdTdT
    383     405   5.000   55.0 UGCAGUACAACCGGCGCUUdTdT AAGCGCCGGUUGUACUGCAdTdT
    387     409   5.000   55.0 GUACAACCGGCGCUUUGUCdTdT GACAAAGCGCCGGUUGUACdTdT
    390     412   5.000   55.0 CAACCGGCGCUUUGUCAACdTdT GUUGACAAAGCGCCGGUUGdTdT
    392     414   5.000   55.0 ACCGGCGCUUUGUCAACGUdTdT ACGUUGACAAAGCGCCGGUdTdT
    407     429   5.000   55.0 ACGUUGUGCCCACCUUUGGdTdT CCAAAGGUGGGCACAACGUdTdT
    428     450   5.000   55.0 AGAAGAAGGGCCCCAAUGCdTdT GCAUUGGGGCCCUUCUUCUdTdT
    431     453   5.000   55.0 AGAAGGGCCCCAAUGCCAAdTdT UUGGCAUUGGGGCCCUUCUdTdT
    434     456   5.000   60.0 AGGGCCCCAAUGCCAACUCdTdT GAGUUGGCAUUGGGGCCCUdTdT
    444     466   5.000   35.0 UGCCAACUCUUAAGUCUUUdTdT AAAGACUUAAGAGUUGGCAdTdT
    487     509   5.000   35.0 AAAAGCCACUUAGUUCAGUdTdT ACUGAACUAAGUGGCUUUUdTdT
    123     145   4.000   50.0 GAUCAAGGCUCAUGUAGCCdTdT GGCUACAUGAGCCUUGAUCdTdT
    125     147   4.000   50.0 UCAAGGCUCAUGUAGCCUCdTdT GAGGCUACAUGAGCCUUGAdTdT
    128     150   4.000   50.0 AGGCUCAUGUAGCCUCACUdTdT AGUGAGGCUACAUGAGCCUdTdT
    155     177   4.000   50.0 UUGCCCCGGAAGAUCAAGUdTdT ACUUGAUCUUCCGGGGCAAdTdT
    234     256   4.000   60.0 GGCCCUGACUACCCUGGAAdTdT UUCCAGGGUAGUCAGGGCCdTdT
    259     281   4.000   60.0 GGCCGCAUGCUUGGAGGUAdTdT UACCUCCAAGCAUGCGGCCdTdT
    266     288   4.000   40.0 UGCUUGGAGGUAAAGUUCAdTdT UGAACUUUACCUCCAAGCAdTdT
    342     364   4.000   40.0 GGAGAAGAAGAAGAAGAAGdTdT CUUCUUCUUCUUCUUCUCCdTdT
    347     369   4.000   40.0 AGAAGAAGAAGAAGACAGGdTdT CCUGUCUUCUUCUUCUUCUdTdT
    359     381   4.000   60.0 AGACAGGUCGGGCUAAGCGdTdT CGCUUAGCCCGACCUGUCUdTdT
    111     133   3.000   55.0 AACGGUCGCCCAGAUCAAGdTdT CUUGAUCUGGGCGACCGUUdTdT
    113     135   3.000   65.0 CGGUCGCCCAGAUCAAGGCdTdT GCCUUGAUCUGGGCGACCGdTdT
    172     194   3.000   70.0 GUCGUGCUCCUGGCAGGCGdTdT CGCCUGCCAGGAGCACGACdTdT
    443     465   3.000   35.0 AUGCCAACUCUUAAGUCUUdTdT AAGACUUAAGAGUUGGCAUdTdT
    456     478   3.000   35.0 AGUCUUUUGUAAUUCUGGCdTdT GCCAGAAUUACAAAAGACUdTdT
    468     490   3.000   30.0 UUCUGGCUUUCUCUAAUAAdTdT UUAUUAGAGAAAGCCAGAAdTdT
    484     506   3.000   30.0 UAAAAAAGCCACUUAGUUCdTdT GAACUAAGUGGCUUUUUUAdTdT
    108     130   2.000   60.0 GGAAACGGUCGCCCAGAUCdTdT GAUCUGGGCGACCGUUUCCdTdT
    135     157   2.000   60.0 UGUAGCCUCACUGGAGGGCdTdT GCCCUCCAGUGAGGCUACAdTdT
    139     161   2.000   60.0 GCCUCACUGGAGGGCAUUGdTdT CAAUGCCCUCCAGUGAGGCdTdT
    150     172   2.000   60.0 GGGCAUUGCCCCGGAAGAUdTdT AUCUUCCGGGGCAAUGCCCdTdT
    171     193   2.000   65.0 AGUCGUGCUCCUGGCAGGCdTdT GCCUGCCAGGAGCACGACUdTdT
    201     223   2.000   65.0 GGAUGAGGCCACUCUGGGCdTdT GCCCAGAGUGGCCUCAUCCdTdT
    204     226   2.000   65.0 UGAGGCCACUCUGGGCCAGdTdT CUGGCCCAGAGUGGCCUCAdTdT
    245     267   2.000   65.0 CCCUGGAAGUAGCAGGCCGdTdT CGGCCUGCUACUUCCAGGGdTdT
    256     278   2.000   65.0 GCAGGCCGCAUGCUUGGAGdTdT CUCCAAGCAUGCGGCCUGCdTdT
    285     307   2.000   65.0 UGGUUCCCUGGCCCGUGCUdTdT AGCACGGGCCAGGGAACCAdTdT
    338     360   2.000   35.0 AACAGGAGAAGAAGAAGAAdTdT UUCUUCUUCUUCUCCUGUUdTdT
    345     367   2.000   35.0 GAAGAAGAAGAAGAAGACAdTdT UGUCUUCUUCUUCUUCUUCdTdT
    486     508   2.000   35.0 AAAAAGCCACUUAGUUCAGdTdT CUGAACUAAGUGGCUUUUUdTdT

#---------------------------------------
#---------------------------------------
</pre>
</td></tr></table><p>


<p>

The siRNAs are reported in order of best score first.

<p>

<b>sirna</b> reports both the sense and antisense siRNAs as 5' to 3'. 




<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.

	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

   -->

None.














<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

<ol>
<li>Elbashir, S. M., et al. (2001a). Duplexes of 21-nucleotide RNAs mediate RNA interference in
    mammalian cell culture. Nature 411: 494-498.
<li>
    Elbashir, S. M., W. Lendeckel and T. Tuschl (2001b). RNA interference is mediated by 21 and 22
    nt RNAs. Genes & Dev. 15: 188-200.
</ol>








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

It is assumed that the input sequence is mRNA.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="banana.html">banana</a></td><td>Bending and curvature plot in B-DNA</td></tr>
<tr><td><a href="btwisted.html">btwisted</a></td><td>Calculates the twisting in a B-DNA sequence</td></tr>
<tr><td><a href="chaos.html">chaos</a></td><td>Create a chaos game representation plot for a sequence</td></tr>
<tr><td><a href="compseq.html">compseq</a></td><td>Counts the composition of dimer/trimer/etc words in a sequence</td></tr>
<tr><td><a href="dan.html">dan</a></td><td>Calculates DNA RNA/DNA melting temperature</td></tr>
<tr><td><a href="freak.html">freak</a></td><td>Residue/base frequency table or plot</td></tr>
<tr><td><a href="isochore.html">isochore</a></td><td>Plots isochores in large DNA sequences</td></tr>
<tr><td><a href="wordcount.html">wordcount</a></td><td>Counts words of a specified size in a DNA sequence</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

This application was written by Gary Williams <a
href="mailto:gwilliam@hgmp.mrc.ac.uk">(gwilliam@hgmp.mrc.ac.uk)</a>
















<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (November 2002) - Gary Williams.













<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->








</BODY>
</HTML>

