<!-- START OF HEADER -->
<HTML><HEAD>
<TITLE> EMBOSS: SITES documentation. </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">
<H2> SITES documentation </H2>
<!-- END OF HEADER -->





<!-- CONTENTS
     This always includes the sections below.
     Other subsections can be added for individual applications.
-->
<br><H2>CONTENTS </H2>
<b> <a href="#1.0">1.0     SUMMARY                   </a></b><br>
<b> <a href="#2.0">2.0     INPUTS & OUTPUTS          </a></b><br>
<b> <a href="#3.0">3.0     INPUT FILE FORMAT         </a></b><br>
<b> <a href="#4.0">4.0     OUTPUT FILE FORMAT        </a></b><br>
<b> <a href="#5.0">5.0     DATA FILES                </a></b><br>
<b> <a href="#6.0">6.0     USAGE                     </a></b><br>    
<b> <a href="#7.0">7.0     KNOWN BUGS & WARNINGS     </a></b><br>    
<b> <a href="#8.0">8.0     NOTES                     </a></b><br>
<b> <a href="#9.0">9.0     DESCRIPTION               </a></b><br>
<b> <a href="#10.0">10.0   ALGORITHM                 </a></b><br>
<b> <a href="#11.0">11.0   RELATED APPLICATIONS      </a></b><br>
<b> <a href="#12.0">12.0   DIAGNOSTIC ERROR MESSAGES </a></b><br>
<b> <a href="#13.0">13.0   AUTHORS                   </a></b><br>
<b> <a href="#14.0">14.0   REFERENCES                </a></b><br>

<!-- SUMMARY
     Succint description of the application, particularly its inputs, outputs
     and what it does.  The same text is given at the top of the source (.c)
     file and in the <documentation> attribute of the <application definition>
     of the ACD file.
-->
<a name="1.0"></a>
<br><br><br><H2> 1.0   SUMMARY  </H2>

Reads CCF files (clean coordinate files) and writes CON files (contact files) of residue-ligand contact data for domains in a DCF file (domain classification file)

<!-- INPUTS & OUTPUTS
     Short summary of the application inputs and outputs in its different 
     modes of usage (if appropriate). More detail than the summary.
-->
<a name="2.0"></a>
<br><br><br><H2> 2.0   INPUTS & OUTPUTS          </H2>
SITES reads CCF files (clean coordinate file) and writes a CON files (contacts file) 
of residue-ligand contact data for domains in a DCF file (domain classification file).
The CON file contains contact data for all ligand-domain pairs (using domain definitions
from the DCF file) found in the CCF files. The input and output files are specified by 
the user (file extensions in the ACD file).  A log file is also written.


<!-- INPUT FILE FORMAT 
     Description and example(s) of input file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.

     Cannot use the test data files because they might be empty or need 
     hand-editing 

     Use "<b>SITES</b> reads any normal sequence USAs." if
     appropriate.
-->
<a name="3.0"></a>
<br><br><br><H2> 3.0   INPUT FILE FORMAT         </H2>

The format of the protein CCF file is described in the <a href="pdbparse.html">PDBPARSE documentation.</a>

<i>Under development, no example input file(s) yet</i>


<!-- OUTPUT FILE FORMAT 
     Description and example(s) of output file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.  
     If the standard description of the avalable report formats is required, 
     use:   #include file="inc/reportformats.ihtml"
     Use "Outputs a graph to the specified graphics device."
     or  "outputs a report format file. The default format is ..."
     if appropriate.
-->
<a name="4.0"></a>
<br><br><br><H2> 4.0   OUTPUT FILE FORMAT         </H2>
The CON format used for the contact files (Figure 1) is similar to EMBL format and is described in the <a href="contacts.html">CONTACTS documentation.</a>  A few of the records differ in the SITES output compared to the CONTACTS output, however, so for the sake of clarity all records are described below.

<UL>
<LI> XX - used for spacing and comments. The first line is bibliographic information and contains 
the text "Residue-ligand contact data (for domains)".
<LI> TY - type of contact.  For CON files generated by SITES, 'LIGAND' is always given.
<LI>  EX - experimental information. The value of the threshold contact 
distance is given as a floating point number after 'THRESH'. For CON files generated by 
SITES, a '.' is given after 'IGNORE', 'NMOD' and 'NCHA' (these records are used by the CONTACTS 
and INTERFACE applications and can be disregarded). 
<LI> NE - number of entries in the file. For CON files generated by SITES, this is the number 
of unique ligands:domain pairs.  
Following the NE record, the file has a section for each entry containing
records for entry number (EN), identifier codes (ID), ligand description (DE), 
polypeptide chain-specific data (CN), chain sequence information (S1) and 
number of contacts (NC), that together define the ligand:domain pair and its contacts. 
<LI> EN - entry number.  The number in brackets indicates the start of an entry (ligand:domain
pair).
<LI> ID - identifier codes: (1) PDB: 4-character PDB identifier code. (2) DOM:
7-character domain identifier code from SCOP or CATH. (3) LIG: Ligand identifier (an 
abbreviation of its full name).
<LI> DE - Full name of the ligand, see <a href="hetparse.html">HETPARSE</a> documentation. 
<LI> CN - polypeptide chain-specific data.  Tokens delimiting data items are as follow. 
 (1) MO: The model number (from the PDB file). '1' is always given for CON files generated
by using SITES (contacts were calculated from the coordinates for a single model from a domain CCF file).  (2) CN1: Chain number. 
'1' is always given (domains from a domain CCF file are always listed as from a single
chain only).  (3) CN2: 
Not used by SITES, a '.' is given.  (4) ID1: PDB chain identifier (a '.` given in cases where a 
chain identifier was not specified in the original PDB file or, for domain CCF files, the domain 
from SCOP or CATH is comprised of more than one chain). (5) ID2: Not used by SITES, a '.' is given. 
(6) NRES1: number of residues in chain. (7) NRES2: Not used by SITES, a '.' is given. 
<LI> S1 - polypeptide chain sequence for domain.  The number of residues is given 
before AA on the first line.  The sequece is given on subsequent lines. 
<LI> NC - number of contacts: (1) SM: Not used by SITES, a '.' is given. (2) LI: Number of 
residue-ligand contacts; between side-chain or main-chain atoms of an amino acid residue and a 
ligand.
<LI> LI - Line of residue-ligand contact data. The amino acid identifier and residue number are 
given. Residue numbers are taken from the CCF file and give a correct index into the sequence 
(i.e. they are not necessarily the same as the original PDB file).  This sequence is 
given in the CON file itself (S1 record).
<LI> // - delimiter for individual entries in the file and also given on
the last line of the file.
</UL> 

<i>Under development, no example output file(s) yet</i>



<!--
<br><br><b>Figure 1  Excerpt from a CON file (contact file) (output) </b>
<table><td bgcolor="#CFCCFF">
<pre>


XX   Residue-ligand contact data (for domains).
XX
TY   LIGAND
XX
EX   THRESH 1.0; IGNORE .; NMOD .; NCHA .;
XX
NE   6
XX
EN   [1]
XX
ID   PDB 1cs4; DOM d1cs4a_; LIG 101;
XX
DE   2'-DEOXY-ADENOSINE 3'-MONOPHOSPHATE
XX
CN   MO .; CN1 1; CN2 .; ID1 A; ID2 .; NRES1 52; NRES2 .
XX
S1   SEQUENCE    52 AA;   5817 MW;  47362A43 CRC32;
     ADIEGFTSLA SQCTAQELVM TLNELFARFD KLAAENHCLR IKILGDCYYC VS
XX
NC   SM .; LI 6
XX
LI   ASP 2
LI   PHE 6
LI   THR 7
LI   LEU 44
LI   GLY 45
LI   ASP 46
XX
//
EN   [2]
XX
ID   PDB 1ii7; DOM d1ii7a_; LIG 101;
XX
DE   2'-DEOXY-ADENOSINE 3'-MONOPHOSPHATE
XX
CN   MO .; CN1 1; CN2 .; ID1 A; ID2 .; NRES1 65; NRES2 .
XX
S1   SEQUENCE    65 AA;   7396 MW;  0CFB92A3 CRC32;
     MKFAHLADIH LGYEQFHKPQ REEEFAEAFK NALEIAVQEN VDFILIAGDL FHSSRPSPGT
     LKKAI
XX
NC   SM .; LI 2
XX
LI   HIS 10
LI   ASP 49
XX
//
EN   [3]
XX
ID   PDB 1cs4; DOM d1cs4a_; LIG FOK;
XX
DE   FORSKOLIN
XX
CN   MO .; CN1 1; CN2 .; ID1 A; ID2 .; NRES1 52; NRES2 .
XX
S1   SEQUENCE    52 AA;   5817 MW;  47362A43 CRC32;
     ADIEGFTSLA SQCTAQELVM TLNELFARFD KLAAENHCLR IKILGDCYYC VS
XX
NC   SM .; LI 1
XX
LI   LEU 44
XX
//
EN   [4]
XX
ID   PDB 1cs4; DOM d1cs4a_; LIG MG;
XX
DE   MAGNESIUM ION
XX
CN   MO .; CN1 1; CN2 .; ID1 A; ID2 .; NRES1 52; NRES2 .
XX
S1   SEQUENCE    52 AA;   5817 MW;  47362A43 CRC32;
     ADIEGFTSLA SQCTAQELVM TLNELFARFD KLAAENHCLR IKILGDCYYC VS
XX
NC   SM .; LI 4
XX
LI   ASP 2
LI   ILE 3
LI   PHE 6
LI   ASP 46
XX
//
EN   [5]
XX
ID   PDB 1ii7; DOM d1ii7a_; LIG MN;
XX
DE   MANGANESE (II) ION
XX
CN   MO .; CN1 1; CN2 .; ID1 A; ID2 .; NRES1 65; NRES2 .
XX
S1   SEQUENCE    65 AA;   7396 MW;  0CFB92A3 CRC32;
     MKFAHLADIH LGYEQFHKPQ REEEFAEAFK NALEIAVQEN VDFILIAGDL FHSSRPSPGT
     LKKAI
XX
NC   SM .; LI 3
XX
LI   ASP 8
LI   HIS 10
LI   ASP 49
XX
//
EN   [6]
XX
ID   PDB 1cs4; DOM d1cs4a_; LIG POP;
XX
DE   PYROPHOSPHATE 2-
XX
CN   MO .; CN1 1; CN2 .; ID1 A; ID2 .; NRES1 52; NRES2 .
XX
S1   SEQUENCE    52 AA;   5817 MW;  47362A43 CRC32;
     ADIEGFTSLA SQCTAQELVM TLNELFARFD KLAAENHCLR IKILGDCYYC VS
XX
NC   SM .; LI 6
XX
LI   ASP 2
LI   ILE 3
LI   GLU 4
LI   GLY 5
LI   PHE 6
LI   THR 7
XX
//

</pre>
</table>

-->


<!-- DATA FILES         
     Any data files used (e.g. translation table file, substitution matrix 
     etc.  This includes example data file formats if they are not obvious.
     For a standard description of what data files are and how embossdata can
     be used to inspect and retrieve them, use:
     #include file="inc/localfiles.ihtml" 
-->

<a name="5.0"></a>
<br><br><br><H2> 5.0   DATA FILES                </H2>
SITES uses a data file containing van der Waals radii for atoms in 
proteins (see <a href="contacts.html">CONTACTS documentation.</a>)
The file Evdw.dat is such a data file and is part of the EMBOSS distribution.

<br><br>SITES uses a data file containing a dictionary of heterogen groups in PDB.
This file may be generated by using <a href="hetparse.html">HETPARSE </a> and is part of the EMBOSS distribution.
The file Ehet.dat is such a data file and is part of the EMBOSS distribution.



<!-- USAGE
     Example usage, as run from the command-line.
     Many examples illustrating different behaviours is good.
-->
<a name="6.0"></a>
<br><br><br><H2> 6.0   USAGE                     </H2>
<H3> 6.1   COMMAND LINE ARGUMENTS </H3>

<pre>
   Standard (Mandatory) qualifiers:
  [-protpath]          dirlist    This option specifies the location of the
                                  protein CCF files (clean coordinate files)
                                  (input). A 'clean cordinate file' contains
                                  protein coordinate and derived data for a
                                  single PDB file ('protein clean coordinate
                                  file') or a single domain from SCOP or CATH
                                  ('domain clean coordinate file'), in CCF
                                  format (EMBL-like). The files, generated by
                                  using PDBPARSE (PDB files) or DOMAINER
                                  (domains), contain 'cleaned-up' data that is
                                  self-consistent and error-corrected.
                                  Records for residue solvent accessibility
                                  and secondary structure are added to the
                                  file by using PDBPLUS.
  [-domaindir]         directory  This option specifies the location of the
                                  domain CCF files (clean coordinate files)
                                  (input). A 'clean cordinate file' contains
                                  protein coordinate and derived data for a
                                  single PDB file ('protein clean coordinate
                                  file') or a single domain from SCOP or CATH
                                  ('domain clean coordinate file'), in CCF
                                  format (EMBL-like). The files, generated by
                                  using PDBPARSE (PDB files) or DOMAINER
                                  (domains), contain 'cleaned-up' data that is
                                  self-consistent and error-corrected.
                                  Records for residue solvent accessibility
                                  and secondary structure are added to the
                                  file by using PDBPLUS.
   -dcffile            infile     This option specifies the name of the DCF
                                  file (domain classification file) (input). A
                                  'domain classification file' contains
                                  classification and other data for domains
                                  from SCOP or CATH, in DCF format
                                  (EMBL-like). The files are generated by
                                  using SCOPPARSE and CATHPARSE. Domain
                                  sequence information can be added to the
                                  file by using DOMAINSEQS.
   -threshold          float      This option specifies the threshold contact
                                  distance.
  [-outfile]           outfile    This option specifies the name of the output
                                  file.
   -logfile            outfile    This option specifies the name of the log
                                  file.

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers:
   -dicfile            datafile   This option specifies the dictionary of
                                  heterogen groups in PDB. This file is
                                  generated by using HETPARSE and is part of
                                  the EMBOSS distribution.
   -vdwfile            datafile   This option specifies the name of the data
                                  file with van der Waals radii for atoms in
                                  amino acid residues. This file is part of
                                  the EMBOSS distribution.

   Associated qualifiers:

   "-outfile" associated qualifiers
   -odirectory3        string     Output directory

   "-logfile" associated qualifiers
   -odirectory         string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report deaths


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-protpath]<br>(Parameter 1)</td>
<td>This option specifies the location of the protein CCF files (clean coordinate files) (input). A 'clean cordinate file' contains protein coordinate and derived data for a single PDB file ('protein clean coordinate file') or a single domain from SCOP or CATH ('domain clean coordinate file'), in CCF format (EMBL-like). The files, generated by using PDBPARSE (PDB files) or DOMAINER (domains), contain 'cleaned-up' data that is self-consistent and error-corrected. Records for residue solvent accessibility and secondary structure are added to the file by using PDBPLUS.</td>
<td>Directory with files</td>
<td>./</td>
</tr>

<tr>
<td>[-domaindir]<br>(Parameter 2)</td>
<td>This option specifies the location of the domain CCF files (clean coordinate files) (input). A 'clean cordinate file' contains protein coordinate and derived data for a single PDB file ('protein clean coordinate file') or a single domain from SCOP or CATH ('domain clean coordinate file'), in CCF format (EMBL-like). The files, generated by using PDBPARSE (PDB files) or DOMAINER (domains), contain 'cleaned-up' data that is self-consistent and error-corrected. Records for residue solvent accessibility and secondary structure are added to the file by using PDBPLUS.</td>
<td>Directory</td>
<td>./</td>
</tr>

<tr>
<td>-dcffile</td>
<td>This option specifies the name of the DCF file (domain classification file) (input). A 'domain classification file' contains classification and other data for domains from SCOP or CATH, in DCF format (EMBL-like). The files are generated by using SCOPPARSE and CATHPARSE. Domain sequence information can be added to the file by using DOMAINSEQS.</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-threshold</td>
<td>This option specifies the threshold contact distance.</td>
<td>Any numeric value</td>
<td>1.0</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 3)</td>
<td>This option specifies the name of the output file.</td>
<td>Output file</td>
<td>SITES.con</td>
</tr>

<tr>
<td>-logfile</td>
<td>This option specifies the name of the log file.</td>
<td>Output file</td>
<td>SITES.log</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-dicfile</td>
<td>This option specifies the dictionary of heterogen groups in PDB. This file is generated by using HETPARSE and is part of the EMBOSS distribution.</td>
<td>Data file</td>
<td>Ehet.dat</td>
</tr>

<tr>
<td>-vdwfile</td>
<td>This option specifies the name of the data file with van der Waals radii for atoms in amino acid residues. This file is part of the EMBOSS distribution.</td>
<td>Data file</td>
<td>Evdw.dat</td>
</tr>

</table>


<H3> 6.2 EXAMPLE SESSION </H3>
An example of interactive use of SITES is shown below.

<i>Under development, no final example yet</i>


Protein CCF files were read from /testdata/sites/in and domain CCF
files from /testdata/sites.  Ligand definitions were taken from
Ehet.dat and domain definitions were taken from the DCF file
testdata/sites/in/sites.dcf.  A CON file of ligand-residue contacts
for all domains defined in the DCF file and present in the protein CCF
files, was written to testdata/sites/out/SITES.con. van der Waals
radii were taken from the emboss data file Evdw.dat and the default
contact distance of 1 Angstrom was used to determine contacts. The log
file testdata/sites/out/SITES.log was written.  The file extensions of
all input and output files were specified in the ACD file.


<!-- KNOWN BUGS & WARNINGS
     Bugs that have not yet been fixed, easily missued features, problems
     and caveats etc. Potentially stupid things the program will let you do.
-->
<a name="7.0"></a>
<br><br><br><H2> 7.0   KNOWN BUGS & WARNINGS     </H2>
None.




<!-- NOTES
     Important general remarks, including:
       Restrictions.
       Interesting behaviour.
       Useful things you can do with this program.
       Future plans.
       etc.
-->
<a name="8.0"></a>
<br><br><br><H2> 8.0   NOTES                     </H2>

<b> Types of contact </b>
<br><b>LI</b> records are used for contacts to
ligands (as defined above).  In CONTACTS and INTERFACE output, <b>SM</b> records are used for contacts between either either side-chain or main-chain atoms.  
In a future implementation, <b>SS</b> will be used for side-chain only contacts, <b>MM</b> will 
be used for main-chain only contacts, and there will probably be several other forms of contact too.




<H3> 8.1   GLOSSARY OF FILE TYPES    </H3>
<table BORDER CELLSPACING=5 CELLPADDING=5 BGCOLOR="#f5f5ff" >
<tr>
<td><b>FILE TYPE</b></td>
<td><b>FORMAT</b></td>
<td><b>DESCRIPTION</b></td>
<td><b>CREATED BY <b></td>
<td><b>SEE ALSO</b></td>
</tr>
<tr>
<td><b> Clean coordinate file (for protein) </b></td>
<td> CCF format (EMBL-like).</td>
<td> Protein coordinate and derived data for a single PDB file.  The data are 'cleaned-up': self-consistent and error-corrected.  </td>
<td> <a href="pdbparse.html">PDBPARSE</a> </td>
<td> Records for residue solvent accessibility and secondary structure are added to the file by using PDBPLUS. </td>
</tr>
<tr>
<td><b> Clean coordinate file (for domain)</b></td>
<td> CCF format (EMBL-like).</td>
<td> Protein coordinate and derived data for a single domain from SCOP or CATH. The data are 'cleaned-up': self-consistent and error-corrected. </td>
<td> <a href="domainer.html">DOMAINER</a></td>
<td> Records for residue solvent accessibility and secondary structure are added to the file by using PDBPLUS.</td>
</tr>
<tr>
<td><b> Contact file (intra-chain residue-residue contacts)</b></td>
<td> CON format (EMBL-like.) </td>
<td> Intra-chain residue-residue contact data for a protein or a domain from SCOP or CATH.  </td>
<td> <a href="contacts.html">CONTACTS</a> </td>
<td> N.A. </td>
</tr>
<tr>
<td><b> Contact file (inter-chain residue-residue contacts)</b></td>
<td> CON format (EMBL-like.) </td>
<td> Inter-chain residue-residue contact data for a protein or a domain from SCOP or CATH.  </td>
<td> <a href="interface.html">INTERFACE</a> </td>
<td> N.A. </td>
</tr>
<tr>
<td><b> Contact file (residue-ligand contacts)</b></td>
<td> CON format (EMBL-like.) </td>
<td> Residue-ligand contact data for a protein or a domain from SCOP or CATH. </td>
<td> <a href="sites.html">SITES</a> </td>
<td> N.A. </td>
</tr>
<tr>
<td><b>van der Waals radii  </b></td>
<td> A file of van der Waals radii for atoms in amino acid residues. Part of the emboss distribution.</td>
<td> N.A. </td>
<td> N.A. </td>
</tr>
<tr>
<td><b>Dictionary of heterogen groups  </b></td>
<td> A file of the dictionary of heterogen groups in PDB. </td>
<td> <a href="hetparse.html">HETPARSE</a> </td>
<td> N.A. </td>
</tr>
</table>



<!-- DESCRIPTION
     A complete, non-technical, user-level description of the application.
-->
<a name="9.0"></a>
<br><br><br><H2> 9.0   DESCRIPTION               </H2>
Knowledge of the physical contacts that amino acid residues make 
with protein ligands is required for several different analyses.  
SITES calculates residue-ligand contact data from protein CCF files 
(clean coordinate files) and organises the data according to domains
taken from a DCF file (domain classification file).




None




<!-- ALGORITHM
     A technical description of algorithmic aspects, describing exactly how
     the key aspects of the application work.
-->
<a name="10.0"></a>
<br><br><br><H2> 10.0  ALGORITHM                 </H2>
Contact between two residues is defined as when the van der Waals surface 
of any atom of the first residue comes within the threshold contact 
distance of the van der Waals surface of any atom of the second residue. 
The threshold contact distance is a user-defined distance with a default 
value of 1 Angstrom.





<!-- RELATED APPLICATIONS
     Other applications that either generate the input, use the output or 
     are in some other way related to the application are described here. 
-->
<a name="11.0"></a>
<br><br><br><H2> 11.0  RELATED APPLICATIONS      </H2>

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="../../apps/aaindexextract.html">aaindexextract</a></td><td>Extract data from AAINDEX</td></tr>
<tr><td><a href="allversusall.html">allversusall</a></td><td>Does an all-versus-all global alignment for each set of sequences in an input directory and writes files of sequence similarity values</td></tr>
<tr><td><a href="cathparse.html">cathparse</a></td><td>Reads raw CATH classification files and writes DCF file (domain classification file)</td></tr>
<tr><td><a href="../../apps/cutgextract.html">cutgextract</a></td><td>Extract data from CUTG</td></tr>
<tr><td><a href="domainer.html">domainer</a></td><td>Reads CCF files (clean coordinate files) for proteins and writes CCF files for domains, taken from a DCF file (domain classification file)</td></tr>
<tr><td><a href="domainnr.html">domainnr</a></td><td>Removes redundant domains from a DCF file (domain classification file). The file must contain domain sequence information, which can be added by using DOMAINSEQS</td></tr>
<tr><td><a href="domainseqs.html">domainseqs</a></td><td>Adds sequence records to a DCF file (domain classification file)</td></tr>
<tr><td><a href="domainsse.html">domainsse</a></td><td>Adds secondary structure records to a DCF file (domain classification file)</td></tr>
<tr><td><a href="hetparse.html">hetparse</a></td><td>Converts raw dictionary of heterogen groups to a file in EMBL-like format</td></tr>
<tr><td><a href="pdbparse.html">pdbparse</a></td><td>Parses PDB files and writes CCF files (clean coordinate files) for proteins</td></tr>
<tr><td><a href="pdbplus.html">pdbplus</a></td><td>Add residue solvent accessibility and secondary structure data to a CCF file (clean coordinate file) for a protein or domain</td></tr>
<tr><td><a href="pdbtosp.html">pdbtosp</a></td><td>Convert raw swissprot:PDB equivalence file to EMBL-like format</td></tr>
<tr><td><a href="../../apps/printsextract.html">printsextract</a></td><td>Extract data from PRINTS</td></tr>
<tr><td><a href="../../apps/prosextract.html">prosextract</a></td><td>Builds the PROSITE motif database for patmatmotifs to search</td></tr>
<tr><td><a href="../../apps/rebaseextract.html">rebaseextract</a></td><td>Extract data from REBASE</td></tr>
<tr><td><a href="scopparse.html">scopparse</a></td><td>Reads raw SCOP classification files and writes a DCF file (domain classification file)</td></tr>
<tr><td><a href="seqnr.html">seqnr</a></td><td>Removes redundancy from DHF files (domain hits files) or other files of sequences</td></tr>
<tr><td><a href="ssematch.html">ssematch</a></td><td>Searches a DCF file (domain classification file) for secondary structure matches</td></tr>
<tr><td><a href="../../apps/tfextract.html">tfextract</a></td><td>Extract data from TRANSFAC</td></tr>
</table>


<!-- DIAGNOSTIC ERROR MESSAGES
     Description of error messages or log file, if one is written.
-->
<a name="12.0"></a>
<br><br><br><H2> 12.0  DIAGNOSTIC ERROR MESSAGES </H2>


SITES generates a log file an excerpt of which is shown below. 
The file contains a line for each protein CCF that was read
containing diagnostic information is given  (in case of 
difficulty email Jon Ison, jison@hgmp.mrc.ac.uk).



<br><br><b>Figure 2   Excerpt from an INTERFACE log file </b>
<table><td bgcolor="#CFCCFF">
<pre>
Excerpt of log file
CCF: 000_testdata_new/sites/in/1cs4.ccf	HETS:YES	NHETS:7	SCOP:YES	NDOMS: 1
CCF: 000_testdata_new/sites/in/1ii7.ccf	HETS:YES	NHETS:5	SCOP:YES	NDOMS: 1
</pre>
</table>



<!-- AUTHORS -->
<a name="13.0"><br><br><br><H2> 13.0  AUTHORS                   </H2></a>
Waqas Awan <a href="mailto:wawan@rfcgr.mrc.ac.uk">(wawan@rfcgr.mrc.ac.uk)</a>
<br>
Jon Ison <a href="mailto:jison@rfcgr.mrc.ac.uk">(jison@rfcgr.mrc.ac.uk)</a>
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK




<!-- REFERENCES
     Quote the paper where the application was first published, described, used etc. 
-->
<a name="14.0"></a>
<br><br><br><H2> 14.0  REFERENCES                </H2>
Please cite the authors and EMBOSS.
<br><br>
<i>Rice P, Bleasby A.J.  2000.  EMBOSS - The European Molecular Biology Open Software Suite.  Trends in Genetics, 15:276-278.  See also http://www.uk.embnet.org/Software/EMBOSS</i>

<H3>14.1 Other useful references  </H3>

<br>

</BODY>
</HTML>
