<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: sixpack
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

sixpack
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<!--END OF HEADER-->






<H2>
    Function
</H2>
Display a DNA sequence with 6-frame translation and ORFs

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
  -->

<b>sixpack</b> takes a nucleic acid sequence and writes out the forward
and reverse senses of the sequence with the 3 forward and three reverse
translations in a pretty display format. 

<p>

It also writes a file containing the open reading frames that are larger
than the specified minimum size (default 1 base, showing all possible
open reading frames).  These open reading frames are written as protein
sequences in the default output sequence format. 

<p>

An open reading frame is defined in this program as any possible
translation between two STOP codons. 











<H2>
    Algorithm
</H2>

<!-- 
        Algorithms
	  Pseudocode/Vague description/Handwaving/web links to useful pages
  -->

The nucleic acid sequence is read in.
<br>
The required genetic code is read in from the ECG* data files.
<br>
The three forward and three reverse translations are created.
<br>
The name and description are written to the ouput display file.
<br>
Any required regions to be changed to upper case are changed.
<br>
Any required regions to be highlighted in HTML colour tags are changed.
<br>
The reverse sense sequence is placed below the forward sequence.
<br>
The forward translations are placed above the sequences.
<br>
The reverse translation are placed below the sequences.
<br>
The display is written out, split at the ends of lines.
<br>
Any ORFs that are longer than the specified minimum size are written to
the output sequence file. 









<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

<b>Here is a sample session with sixpack</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>sixpack </b>
Display a DNA sequence with 6-frame translation and ORFs
Input sequence: <b>tembl:hsfau</b>
Output file [hsfau.sixpack]: <b></b>
Output sequence [hsfau.fasta]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>




<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          sequence   Sequence USA
  [-outfile]           outfile    If you enter the name of a file here then
                                  this program will write the sequence details
                                  into that file.
   -outseq             seqoutall  If you enter the name of a file here then
                                  this program will write the ORFs into that
                                  file.

   Optional qualifiers:
   -table              menu       Genetics code used for the translation
   -[no]firstorf       boolean    Count the beginning of a sequence as a
                                  possible ORF, even if it's inferior to the
                                  minimal ORF size.
   -[no]lastorf        boolean    Count the end of a sequence as a possible
                                  ORF, even if it's not finishing with a STOP,
                                  or inferior to the minimal ORF size.
   -mstart             boolean    Displays only ORFs starting with an M.

   Advanced qualifiers:
   -[no]reverse        boolean    Display also the translation of the DNA
                                  sequence in the 3 reverse frames
   -orfminsize         integer    Minimum size of Open Reading Frames (ORFs)
                                  to display in the translations.
   -uppercase          range      Regions to put in uppercase.
                                  If this is left blank, then the sequence
                                  case is left alone.
                                  A set of regions is specified by a set of
                                  pairs of positions.
                                  The positions are integers.
                                  They are separated by any non-digit,
                                  non-alpha character.
                                  Examples of region specifications are:
                                  24-45, 56-78
                                  1:45, 67=99;765..888
                                  1,5,8,10,23,45,57,99
   -highlight          range      Regions to colour if formatting for HTML.
                                  If this is left blank, then the sequence is
                                  left alone.
                                  A set of regions is specified by a set of
                                  pairs of positions.
                                  The positions are integers.
                                  They are followed by any valid HTML font
                                  colour.
                                  Examples of region specifications are:
                                  24-45 blue 56-78 orange
                                  1-100 green 120-156 red
                                  A file of ranges to colour (one range per
                                  line) can be specifed as '@filename'.
   -[no]number         boolean    Number the sequence at the beginning and the
                                  end of each line.
   -width              integer    Number of nucleotides displayed on each line
   -length             integer    Line length of page (0 for indefinite)
   -margin             integer    Margin around sequence for numbering.
   -[no]name           boolean    Set this to be false if you do not wish to
                                  display the ID name of the sequence.
   -[no]description    boolean    Set this to be false if you do not wish to
                                  display the description of the sequence.
   -offset             integer    Number from which you want the DNA sequence
                                  to be numbered.
   -html               boolean    Use HTML formatting

   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>If you enter the name of a file here then this program will write the sequence details into that file.</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.sixpack</td>
</tr>

<tr>
<td>-outseq</td>
<td>If you enter the name of a file here then this program will write the ORFs into that file.</td>
<td>Writeable sequence(s)</td>
<td><i>&lt;sequence&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-table</td>
<td>Genetics code used for the translation</td>
<td><table><tr><td>0</td> <td><i>(Standard)</i></td></tr><tr><td>1</td> <td><i>(Standard (with alternative initiation codons))</i></td></tr><tr><td>2</td> <td><i>(Vertebrate Mitochondrial)</i></td></tr><tr><td>3</td> <td><i>(Yeast Mitochondrial)</i></td></tr><tr><td>4</td> <td><i>(Mold, Protozoan, Coelenterate Mitochondrial and Mycoplasma/Spiroplasma)</i></td></tr><tr><td>5</td> <td><i>(Invertebrate Mitochondrial)</i></td></tr><tr><td>6</td> <td><i>(Ciliate Macronuclear and Dasycladacean)</i></td></tr><tr><td>9</td> <td><i>(Echinoderm Mitochondrial)</i></td></tr><tr><td>10</td> <td><i>(Euplotid Nuclear)</i></td></tr><tr><td>11</td> <td><i>(Bacterial)</i></td></tr><tr><td>12</td> <td><i>(Alternative Yeast Nuclear)</i></td></tr><tr><td>13</td> <td><i>(Ascidian Mitochondrial)</i></td></tr><tr><td>14</td> <td><i>(Flatworm Mitochondrial)</i></td></tr><tr><td>15</td> <td><i>(Blepharisma Macronuclear)</i></td></tr><tr><td>16</td> <td><i>(Chlorophycean Mitochondrial)</i></td></tr><tr><td>21</td> <td><i>(Trematode Mitochondrial)</i></td></tr><tr><td>22</td> <td><i>(Scenedesmus obliquus)</i></td></tr><tr><td>23</td> <td><i>(Thraustochytrium Mitochondrial)</i></td></tr></table></td>
<td>0</td>
</tr>

<tr>
<td>-[no]firstorf</td>
<td>Count the beginning of a sequence as a possible ORF, even if it's inferior to the minimal ORF size.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]lastorf</td>
<td>Count the end of a sequence as a possible ORF, even if it's not finishing with a STOP, or inferior to the minimal ORF size.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-mstart</td>
<td>Displays only ORFs starting with an M.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]reverse</td>
<td>Display also the translation of the DNA sequence in the 3 reverse frames</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-orfminsize</td>
<td>Minimum size of Open Reading Frames (ORFs) to display in the translations.</td>
<td>Integer 1 or more</td>
<td>1</td>
</tr>

<tr>
<td>-uppercase</td>
<td>Regions to put in uppercase.
If this is left blank, then the sequence case is left alone.
A set of regions is specified by a set of pairs of positions.
The positions are integers.
They are separated by any non-digit, non-alpha character.
Examples of region specifications are:
24-45, 56-78
1:45, 67=99;765..888
1,5,8,10,23,45,57,99</td>
<td>Sequence range</td>
<td>If this is left blank, then the sequence case is left alone.</td>
</tr>

<tr>
<td>-highlight</td>
<td>Regions to colour if formatting for HTML.
If this is left blank, then the sequence is left alone.
A set of regions is specified by a set of pairs of positions.
The positions are integers.
They are followed by any valid HTML font colour.
Examples of region specifications are:
24-45 blue 56-78 orange
1-100 green 120-156 red
A file of ranges to colour (one range per line) can be specifed as '@filename'.</td>
<td>Sequence range</td>
<td><i>full sequence</i></td>
</tr>

<tr>
<td>-[no]number</td>
<td>Number the sequence at the beginning and the end of each line.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-width</td>
<td>Number of nucleotides displayed on each line</td>
<td>Integer 1 or more</td>
<td>60</td>
</tr>

<tr>
<td>-length</td>
<td>Line length of page (0 for indefinite)</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr>
<td>-margin</td>
<td>Margin around sequence for numbering.</td>
<td>Integer 0 or more</td>
<td>10</td>
</tr>

<tr>
<td>-[no]name</td>
<td>Set this to be false if you do not wish to display the ID name of the sequence.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]description</td>
<td>Set this to be false if you do not wish to display the description of the sequence.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-offset</td>
<td>Number from which you want the DNA sequence to be numbered.</td>
<td>Any integer value</td>
<td>1</td>
</tr>

<tr>
<td>-html</td>
<td>Use HTML formatting</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>sixpack</b> reads any normal sequence USAs.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:hsfau' is a sequence entry in the example nucleic acid database 'tembl'
<p>
<p><h3>Database entry: tembl:hsfau</h3>
<table width="90%"><tr><td bgcolor="#FFDDFF">
<pre>
ID   HSFAU      standard; RNA; HUM; 518 BP.
XX
AC   X65923;
XX
SV   X65923.1
XX
DT   13-MAY-1992 (Rel. 31, Created)
DT   23-SEP-1993 (Rel. 37, Last updated, Version 10)
XX
DE   H.sapiens fau mRNA
XX
KW   fau gene.
XX
OS   Homo sapiens (human)
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia;
OC   Eutheria; Primates; Catarrhini; Hominidae; Homo.
XX
RN   [1]
RP   1-518
RA   Michiels L.M.R.;
RT   ;
RL   Submitted (29-APR-1992) to the EMBL/GenBank/DDBJ databases.
RL   L.M.R. Michiels, University of Antwerp, Dept of Biochemistry,
RL   Universiteisplein 1, 2610 Wilrijk, BELGIUM
XX
RN   [2]
RP   1-518
RX   MEDLINE; 93368957.
RA   Michiels L., Van der Rauwelaert E., Van Hasselt F., Kas K., Merregaert J.;
RT   " fau cDNA encodes a ubiquitin-like-S30 fusion protein and is expressed as
RT   an antisense sequences in the Finkel-Biskis-Reilly murine sarcoma virus";
RL   Oncogene 8:2537-2546(1993).
XX
DR   SWISS-PROT; P35544; UBIM_HUMAN.
DR   SWISS-PROT; Q05472; RS30_HUMAN.
XX
FH   Key             Location/Qualifiers
FH
FT   source          1..518
FT                   /chromosome="11q"
FT                   /db_xref="taxon:9606"
FT                   /organism="Homo sapiens"
FT                   /tissue_type="placenta"
FT                   /clone_lib="cDNA"
FT                   /clone="pUIA 631"
FT                   /map="13"
FT   misc_feature    57..278
FT                   /note="ubiquitin like part"
FT   CDS             57..458
FT                   /db_xref="SWISS-PROT:P35544"
FT                   /db_xref="SWISS-PROT:Q05472"
FT                   /gene="fau"
FT                   /protein_id="CAA46716.1"
FT                   /translation="MQLFVRAQELHTFEVTGQETVAQIKAHVASLEGIAPEDQVVLLAG
FT                   APLEDEATLGQCGVEALTTLEVAGRMLGGKVHGSLARAGKVRGQTPKVAKQEKKKKKTG
FT                   RAKRRMQYNRRFVNVVPTFGKKKGPNANS"
FT   misc_feature    98..102
FT                   /note="nucleolar localization signal"
FT   misc_feature    279..458
FT                   /note="S30 part"
FT   polyA_signal    484..489
FT   polyA_site      509
XX
SQ   Sequence 518 BP; 125 A; 139 C; 148 G; 106 T; 0 other;
     ttcctctttc tcgactccat cttcgcggta gctgggaccg ccgttcagtc gccaatatgc        60
     agctctttgt ccgcgcccag gagctacaca ccttcgaggt gaccggccag gaaacggtcg       120
     cccagatcaa ggctcatgta gcctcactgg agggcattgc cccggaagat caagtcgtgc       180
     tcctggcagg cgcgcccctg gaggatgagg ccactctggg ccagtgcggg gtggaggccc       240
     tgactaccct ggaagtagca ggccgcatgc ttggaggtaa agttcatggt tccctggccc       300
     gtgctggaaa agtgagaggt cagactccta aggtggccaa acaggagaag aagaagaaga       360
     agacaggtcg ggctaagcgg cggatgcagt acaaccggcg ctttgtcaac gttgtgccca       420
     cctttggcaa gaagaagggc cccaatgcca actcttaagt cttttgtaat tctggctttc       480
     tctaataaaa aagccactta gttcagtcaa aaaaaaaa                               518
//
</pre>
</td></tr></table><p>






<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 

        If you wish to include the standard description of the avalable
	report formats, use:
<p>

The output is a standard EMBOSS report file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-rformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  The available format names are: embl,
genbank, gff, pir, swiss, trace, listfile, dbmotif, diffseq, excel,
feattable, motif, regions, seqtable, simple, srs, table, tagseq

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html</A>
for further information on report formats.

<p>

   -->


<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: hsfau.fasta</h3>
<table width="90%"><tr><td bgcolor="#DDFFDD">
<pre>
&gt;HSFAU_1_ORF1  Translation of HSFAU in frame 1, ORF 1, threshold 1, 33aa
FLFLDSIFAVAGTAVQSPICSSLSAPRSYTPSR
&gt;HSFAU_1_ORF2  Translation of HSFAU in frame 1, ORF 2, threshold 1, 12aa
PARKRSPRSRLM
&gt;HSFAU_1_ORF3  Translation of HSFAU in frame 1, ORF 3, threshold 1, 33aa
PHWRALPRKIKSCSWQARPWRMRPLWASAGWRP
&gt;HSFAU_1_ORF4  Translation of HSFAU in frame 1, ORF 4, threshold 1, 4aa
LPWK
&gt;HSFAU_1_ORF5  Translation of HSFAU in frame 1, ORF 5, threshold 1, 18aa
QAACLEVKFMVPWPVLEK
&gt;HSFAU_1_ORF6  Translation of HSFAU in frame 1, ORF 6, threshold 1, 61aa
EVRLLRWPNRRRRRRRQVGLSGGCSTTGALSTLCPPLARRRAPMPTLKSFVILAFSNKKA
T
&gt;HSFAU_1_ORF7  Translation of HSFAU in frame 1, ORF 7, threshold 1, 6aa
FSQKKX
&gt;HSFAU_2_ORF1  Translation of HSFAU in frame 2, ORF 1, threshold 1, 9aa
SSFSTPSSR
&gt;HSFAU_2_ORF2  Translation of HSFAU in frame 2, ORF 2, threshold 1, 58aa
LGPPFSRQYAALCPRPGATHLRGDRPGNGRPDQGSCSLTGGHCPGRSSRAPGRRAPGG
&gt;HSFAU_2_ORF3  Translation of HSFAU in frame 2, ORF 3, threshold 1, 23aa
GHSGPVRGGGPDYPGSSRPHAWR
&gt;HSFAU_2_ORF4  Translation of HSFAU in frame 2, ORF 4, threshold 1, 16aa
SSWFPGPCWKSERSDS
&gt;HSFAU_2_ORF5  Translation of HSFAU in frame 2, ORF 5, threshold 1, 14aa
GGQTGEEEEEDRSG
&gt;HSFAU_2_ORF6  Translation of HSFAU in frame 2, ORF 6, threshold 1, 30aa
AADAVQPALCQRCAHLWQEEGPQCQLLSLL
&gt;HSFAU_2_ORF7  Translation of HSFAU in frame 2, ORF 7, threshold 1, 17aa
FWLSLIKKPLSSVKKKX
&gt;HSFAU_3_ORF1  Translation of HSFAU in frame 3, ORF 1, threshold 1, 151aa
PLSRLHLRGSWDRRSVANMQLFVRAQELHTFEVTGQETVAQIKAHVASLEGIAPEDQVVL
LAGAPLEDEATLGQCGVEALTTLEVAGRMLGGKVHGSLARAGKVRGQTPKVAKQEKKKKK
TGRAKRRMQYNRRFVNVVPTFGKKKGPNANS
&gt;HSFAU_3_ORF2  Translation of HSFAU in frame 3, ORF 2, threshold 1, 8aa
VFCNSGFL
&gt;HSFAU_3_ORF3  Translation of HSFAU in frame 3, ORF 3, threshold 1, 10aa
KSHLVQSKKK
&gt;HSFAU_4_ORF1  Translation of HSFAU in frame 4, ORF 1, threshold 1, 36aa
FFLTELSGFFIRESQNYKRLKSWHWGPSSCQRWAQR
&gt;HSFAU_4_ORF2  Translation of HSFAU in frame 4, ORF 2, threshold 1, 10aa
QSAGCTASAA
&gt;HSFAU_4_ORF3  Translation of HSFAU in frame 4, ORF 3, threshold 1, 14aa
PDLSSSSSSPVWPP
&gt;HSFAU_4_ORF4  Translation of HSFAU in frame 4, ORF 4, threshold 1, 27aa
ESDLSLFQHGPGNHELYLQACGLLLPG
&gt;HSFAU_4_ORF5  Translation of HSFAU in frame 4, ORF 5, threshold 1, 38aa
SGPPPRTGPEWPHPPGARLPGARLDLPGQCPPVRLHEP
&gt;HSFAU_4_ORF6  Translation of HSFAU in frame 4, ORF 6, threshold 1, 42aa
SGRPFPGRSPRRCVAPGRGQRAAYWRLNGGPSYREDGVEKEE
&gt;HSFAU_5_ORF1  Translation of HSFAU in frame 5, ORF 1, threshold 1, 4aa
FFFD
&gt;HSFAU_5_ORF2  Translation of HSFAU in frame 5, ORF 2, threshold 1, 6aa
TKWLFY
&gt;HSFAU_5_ORF3  Translation of HSFAU in frame 5, ORF 3, threshold 1, 8aa
RKPELQKT
&gt;HSFAU_5_ORF4  Translation of HSFAU in frame 5, ORF 4, threshold 1, 44aa
ELALGPFFLPKVGTTLTKRRLYCIRRLARPVFFFFFSCLATLGV
&gt;HSFAU_5_ORF5  Translation of HSFAU in frame 5, ORF 5, threshold 1, 11aa
PLTFPARAREP
&gt;HSFAU_5_ORF6  Translation of HSFAU in frame 5, ORF 6, threshold 1, 37aa
TLPPSMRPATSRVVRASTPHWPRVASSSRGAPARSTT
&gt;HSFAU_5_ORF7  Translation of HSFAU in frame 5, ORF 7, threshold 1, 11aa
SSGAMPSSEAT
&gt;HSFAU_5_ORF8  Translation of HSFAU in frame 5, ORF 8, threshold 1, 45aa
ALIWATVSWPVTSKVCSSWARTKSCILATERRSQLPRRWSRERGX
&gt;HSFAU_6_ORF1  Translation of HSFAU in frame 6, ORF 1, threshold 1, 3aa
FFF
&gt;HSFAU_6_ORF2  Translation of HSFAU in frame 6, ORF 2, threshold 1, 2aa
LN
&gt;HSFAU_6_ORF3  Translation of HSFAU in frame 6, ORF 3, threshold 1, 117aa
VAFLLEKARITKDLRVGIGALLLAKGGHNVDKAPVVLHPPLSPTCLLLLLLLFGHLRSLT
SHFSSTGQGTMNFTSKHAACYFQGSQGLHPALAQSGLILQGRACQEHDLIFRGNALQ
&gt;HSFAU_6_ORF4  Translation of HSFAU in frame 6, ORF 4, threshold 1, 19aa
GYMSLDLGDRFLAGHLEGV
&gt;HSFAU_6_ORF5  Translation of HSFAU in frame 6, ORF 5, threshold 1, 12aa
LLGADKELHIGD
&gt;HSFAU_6_ORF6  Translation of HSFAU in frame 6, ORF 6, threshold 1, 15aa
TAVPATAKMESRKRX
</pre>
</td></tr></table><p>
<p><h3>File: hsfau.sixpack</h3>
<table width="90%"><tr><td bgcolor="#DDFFDD">
<pre>
HSFAU
H.sapiens fau mRNA


          F  L  F  L  D  S  I  F  A  V  A  G  T  A  V  Q  S  P  I  C     F1
           S  S  F  S  T  P  S  S  R  *  L  G  P  P  F  S  R  Q  Y  A    F2
            P  L  S  R  L  H  L  R  G  S  W  D  R  R  S  V  A  N  M  Q   F3
        1 ttcctctttctcgactccatcttcgcggtagctgggaccgccgttcagtcgccaatatgc 60
          ----:----|----:----|----:----|----:----|----:----|----:----|
        1 aaggagaaagagctgaggtagaagcgccatcgaccctggcggcaagtcagcggttatacg 60
           X  R  K  R  S  E  M  K  A  T  A  P  V  A  T  *  D  G  I  H    F6
          X  G  R  E  R  S  W  R  R  P  L  Q  S  R  R  E  T  A  L  I     F5
            E  E  K  E  V  G  D  E  R  Y  S  P  G  G  N  L  R  W  Y  A   F4


          S  S  L  S  A  P  R  S  Y  T  P  S  R  *  P  A  R  K  R  S     F1
           A  L  C  P  R  P  G  A  T  H  L  R  G  D  R  P  G  N  G  R    F2
            L  F  V  R  A  Q  E  L  H  T  F  E  V  T  G  Q  E  T  V  A   F3
       61 agctctttgtccgcgcccaggagctacacaccttcgaggtgaccggccaggaaacggtcg 120
          ----:----|----:----|----:----|----:----|----:----|----:----|
       61 tcgagaaacaggcgcgggtcctcgatgtgtggaagctccactggccggtcctttgccagc 120
           L  E  K  D  A  G  L  L  *  V  G  E  L  H  G  A  L  F  R  D    F6
          C  S  K  T  R  A  W  S  S  C  V  K  S  T  V  P  W  S  V  T     F5
            A  R  Q  G  R  G  P  A  V  C  R  R  P  S  R  G  P  F  P  R   F4


          P  R  S  R  L  M  *  P  H  W  R  A  L  P  R  K  I  K  S  C     F1
           P  D  Q  G  S  C  S  L  T  G  G  H  C  P  G  R  S  S  R  A    F2
            Q  I  K  A  H  V  A  S  L  E  G  I  A  P  E  D  Q  V  V  L   F3
      121 cccagatcaaggctcatgtagcctcactggagggcattgccccggaagatcaagtcgtgc 180
          ----:----|----:----|----:----|----:----|----:----|----:----|
      121 gggtctagttccgagtacatcggagtgacctcccgtaacggggccttctagttcagcacg 180
           G  L  D  L  S  M  Y  G  *  Q  L  A  N  G  R  F  I  L  D  H    F6
          A  W  I  L  A  *  T  A  E  S  S  P  M  A  G  S  S  *  T  T     F5
            G  S  *  P  E  H  L  R  V  P  P  C  Q  G  P  L  D  L  R  A   F4


          S  W  Q  A  R  P  W  R  M  R  P  L  W  A  S  A  G  W  R  P     F1
           P  G  R  R  A  P  G  G  *  G  H  S  G  P  V  R  G  G  G  P    F2
            L  A  G  A  P  L  E  D  E  A  T  L  G  Q  C  G  V  E  A  L   F3
      181 tcctggcaggcgcgcccctggaggatgaggccactctgggccagtgcggggtggaggccc 240
          ----:----|----:----|----:----|----:----|----:----|----:----|
      181 aggaccgtccgcgcggggacctcctactccggtgagacccggtcacgccccacctccggg 240
           E  Q  C  A  R  G  Q  L  I  L  G  S  Q  A  L  A  P  H  L  G    F6
          S  R  A  P  A  G  R  S  S  S  A  V  R  P  W  H  P  T  S  A     F5
            G  P  L  R  A  G  P  P  H  P  W  E  P  G  T  R  P  P  P  G   F4


          *  L  P  W  K  *  Q  A  A  C  L  E  V  K  F  M  V  P  W  P     F1
           D  Y  P  G  S  S  R  P  H  A  W  R  *  S  S  W  F  P  G  P    F2


<font color=red>  [Part of this file has been deleted for brevity]</font>

      301 cacgaccttttcactctccagtctgaggattccaccggtttgtcctcttcttcttcttct 360
           T  S  S  F  H  S  T  L  S  R  L  H  G  F  L  L  L  L  L  L    F6
          R  A  P  F  T  L  P  *  V  G  L  T  A  L  C  S  F  F  F  F     F5
            H  Q  F  L  S  L  D  S  E  *  P  P  W  V  P  S  S  S  S  S   F4


          R  Q  V  G  L  S  G  G  C  S  T  T  G  A  L  S  T  L  C  P     F1
           D  R  S  G  *  A  A  D  A  V  Q  P  A  L  C  Q  R  C  A  H    F2
            T  G  R  A  K  R  R  M  Q  Y  N  R  R  F  V  N  V  V  P  T   F3
      361 agacaggtcgggctaagcggcggatgcagtacaaccggcgctttgtcaacgttgtgccca 420
          ----:----|----:----|----:----|----:----|----:----|----:----|
      361 tctgtccagcccgattcgccgcctacgtcatgttggccgcgaaacagttgcaacacgggt 420
           L  C  T  P  S  L  P  P  H  L  V  V  P  A  K  D  V  N  H  G    F6
          F  V  P  R  A  L  R  R  I  C  Y  L  R  R  K  T  L  T  T  G     F5
            S  L  D  P  *  A  A  S  A  T  C  G  A  S  Q  *  R  Q  A  W   F4


          P  L  A  R  R  R  A  P  M  P  T  L  K  S  F  V  I  L  A  F     F1
           L  W  Q  E  E  G  P  Q  C  Q  L  L  S  L  L  *  F  W  L  S    F2
            F  G  K  K  K  G  P  N  A  N  S  *  V  F  C  N  S  G  F  L   F3
      421 cctttggcaagaagaagggccccaatgccaactcttaagtcttttgtaattctggctttc 480
          ----:----|----:----|----:----|----:----|----:----|----:----|
      421 ggaaaccgttcttcttcccggggttacggttgagaattcagaaaacattaagaccgaaag 480
           G  K  A  L  L  L  A  G  I  G  V  R  L  D  K  T  I  R  A  K    F6
          V  K  P  L  F  F  P  G  L  A  L  E  *  T  K  Q  L  E  P  K     F5
            R  Q  C  S  S  P  G  W  H  W  S  K  L  R  K  Y  N  Q  S  E   F4


          S  N  K  K  A  T  *  F  S  Q  K  K  X                          F1
           L  I  K  K  P  L  S  S  V  K  K  K  X                         F2
            *  *  K  S  H  L  V  Q  S  K  K  K                           F3
      481 tctaataaaaaagccacttagttcagtcaaaaaaaaaa                       518
          ----:----|----:----|----:----|----:----|----:----|----:----|
      481 agattattttttcggtgaatcaagtcagtttttttttt                       518
           E  L  L  F  A  V  *  N  L  *  F  F  F                         F6
          R  *  Y  F  L  W  K  T  *  D  F  F  F                          F5
            R  I  F  F  G  S  L  E  T  L  F  F                           F4

##############################
Minimum size of ORFs : 1

Total ORFs in frame 1 :     7
Total ORFs in frame 2 :     7
Total ORFs in frame 3 :     3
Total ORFs in frame 4 :     6
Total ORFs in frame 5 :     8
Total ORFs in frame 6 :     6

Total ORFs :    37
##############################

</pre>
</td></tr></table><p>






<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.

	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
   -->

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>


<p>
The Genetic Code data files are based on the NCBI genetic code tables.
Their names and descriptions are:

<dl>
<dt>EGC.0 </dt><dd>
      Standard (Differs from GC.1 in that it only has initiation site 'AUG')
<dt>EGC.1 </dt><dd>
      Standard
<dt>EGC.2 </dt><dd>
      Vertebrate Mitochodrial
<dt>EGC.3 </dt><dd>
      Yeast Mitochondrial
<dt>EGC.4 </dt><dd>
      Mold, Protozoan, Coelenterate Mitochondrial and Mycoplasma/Spiroplasma
<dt>EGC.5 </dt><dd>      
      Invertebrate Mitochondrial
<dt>EGC.6 </dt><dd>
      Ciliate Macronuclear and Dasycladacean
<dt>EGC.9 </dt><dd>
      Echinoderm Mitochondrial
<dt>EGC.10 </dt><dd>
      Euplotid Nuclear
<dt>EGC.11 </dt><dd>
      Bacterial
<dt>EGC.12 </dt><dd>
      Alternative Yeast Nuclear
<dt>EGC.13 </dt><dd>
      Ascidian Mitochondrial
<dt>EGC.14 </dt><dd>
      Flatworm Mitochondrial
<dt>EGC.15</dt><dd>
      Blepharisma Macronuclear
<dt>EGC.16</dt><dd>
      Chlorophycean Mitochondrial
<dt>EGC.21</dt><dd>  
      Trematode Mitochondrial
<dt>EGC.22</dt><dd>
      Scenedesmus obliquus
<dt>EGC.23</dt><dd>
      Thraustochytrium Mitochondrial
</dl>
<p>
The format of these files is very simple.

<p>
It consists of several lines of optional comments, each starting with
a '#' character.

<p>
These are followed the line: 'Genetic Code [n]', where 'n' is the
number of the genetic code file.

<p>
This is followed by the description of the code and then by four lines
giving the IUPAC one-letter code of the translated amino acid, the  
start codons (indicdated by an 'M') and the three bases of the codon,
lined up one on top of the other.

<p>
For example:

<pre>

------------------------------------------------------------------------------
# Genetic Code Table
#
# Obtained from: http://www.ncbi.nlm.nih.gov/collab/FT/genetic_codes.html
# and: http://www3.ncbi.nlm.nih.gov/htbin-post/Taxonomy/wprintgc?mode=c
#
# Differs from Genetic Code [1] only in that the initiation sites have been
# changed to only 'AUG'

Genetic Code [0]
Standard

AAs  =   FFLLSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG
Starts = -----------------------------------M----------------------------
Base1  = TTTTTTTTTTTTTTTTCCCCCCCCCCCCCCCCAAAAAAAAAAAAAAAAGGGGGGGGGGGGGGGG
Base2  = TTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGG
Base3  = TCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAG
------------------------------------------------------------------------------

</pre>
                     













<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="abiview.html">abiview</a></td><td>Reads ABI file and display the trace</td></tr>
<tr><td><a href="backtranseq.html">backtranseq</a></td><td>Back translate a protein sequence</td></tr>
<tr><td><a href="cirdna.html">cirdna</a></td><td>Draws circular maps of DNA constructs</td></tr>
<tr><td><a href="coderet.html">coderet</a></td><td>Extract CDS, mRNA and translations from feature tables</td></tr>
<tr><td><a href="getorf.html">getorf</a></td><td>Finds and extracts open reading frames (ORFs)</td></tr>
<tr><td><a href="lindna.html">lindna</a></td><td>Draws linear maps of DNA constructs</td></tr>
<tr><td><a href="marscan.html">marscan</a></td><td>Finds MAR/SAR sites in nucleic sequences</td></tr>
<tr><td><a href="pepnet.html">pepnet</a></td><td>Displays proteins as a helical net</td></tr>
<tr><td><a href="pepwheel.html">pepwheel</a></td><td>Shows protein sequences as helices</td></tr>
<tr><td><a href="plotorf.html">plotorf</a></td><td>Plot potential open reading frames</td></tr>
<tr><td><a href="prettyplot.html">prettyplot</a></td><td>Displays aligned sequences, with colouring and boxing</td></tr>
<tr><td><a href="prettyseq.html">prettyseq</a></td><td>Output sequence with translated ranges</td></tr>
<tr><td><a href="remap.html">remap</a></td><td>Display a sequence with restriction cut sites, translation etc</td></tr>
<tr><td><a href="seealso.html">seealso</a></td><td>Finds programs sharing group names</td></tr>
<tr><td><a href="showalign.html">showalign</a></td><td>Displays a multiple sequence alignment</td></tr>
<tr><td><a href="showdb.html">showdb</a></td><td>Displays information on the currently available databases</td></tr>
<tr><td><a href="showfeat.html">showfeat</a></td><td>Show features of a sequence</td></tr>
<tr><td><a href="showorf.html">showorf</a></td><td>Pretty output of DNA translations</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="textsearch.html">textsearch</a></td><td>Search sequence documentation text. SRS and Entrez are faster!</td></tr>
<tr><td><a href="transeq.html">transeq</a></td><td>Translate nucleic acid sequences</td></tr>
<tr><td><a href="wobble.html">wobble</a></td><td>Wobble base plot</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

This application was written by Thomas Laurent <a
href="mailto:thomas.laurent@uk.lionbioscience.com">(thomas.laurent@uk.lionbioscience.com)</a>


















<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (November 2002) - Thomas Laurent












<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












</BODY>
</HTML>

