<HTML>
<HEAD>
  <TITLE>
  EMBOSS: skipseq
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
skipseq
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
    Function
</H2>
Reads and writes (returns) sequences, skipping first few
<H2>
    Description
</H2>

<b>skipseq</b> skips the first few sequences in a multiple set of
sequences, and writes out the rest of them. 

<p>

<b>skipseq</b> is a variant of the standard program for reading and
writing sequences, <b>seqret</b>. 

<p>

<b>seqret</b> has an option to allow it to only read the first sequence
from a multiple set of sequences (-firstonly).  <b>seqret</b> cannot,
however, skip the first few sequences from a multiple set of sequence,
writing out the rest; this is what <b>skipseq</b> is for.

<p>


In all other respects, <b>skipseq</b> is the same as <b>seqret</b>. 

<H2>
    Usage
</H2>
<b>Here is a sample session with skipseq</b>
<p>
This does not skip any sequences. It is exactly equivalent to seqret: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>skipseq </b>
Reads and writes (returns) sequences, skipping first few
Input (gapped) sequence(s): <b>tembl:eclac*</b>
Number of sequences to skip at start [0]: <b></b>
output sequence(s) [eclac.fasta]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>
<p>
<b>Example 2</b>
<p>
This skips the first input sequence, writing out the others: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>skipseq -skip 1 </b>
Reads and writes (returns) sequences, skipping first few
Input (gapped) sequence(s): <b>tembl:eclac*</b>
output sequence(s) [eclac.fasta]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.2">Go to the output files for this example</a><p><p>

<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-sequence]          seqall     (Gapped) sequence(s) filename and optional
                                  format, or reference (input USA)
   -skip               integer    [0] Number of sequences to skip at start
                                  (Any integer value)
  [-outseq]            seqoutall  [<sequence>.<format>] Sequence set(s)
                                  filename and optional format (output USA)

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers:
   -feature            boolean    Use feature information

   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1            integer    Start of each sequence to be used
   -send1              integer    End of each sequence to be used
   -sreverse1          boolean    Reverse (if DNA)
   -sask1              boolean    Ask for begin/end/reverse
   -snucleotide1       boolean    Sequence is nucleotide
   -sprotein1          boolean    Sequence is protein
   -slower1            boolean    Make lower case
   -supper1            boolean    Make upper case
   -sformat1           string     Input sequence format
   -sdbname1           string     Database name
   -sid1               string     Entryname
   -ufo1               string     UFO features
   -fformat1           string     Features format
   -fopenfile1         string     Features file name

   "-outseq" associated qualifiers
   -osformat2          string     Output seq format
   -osextension2       string     File name extension
   -osname2            string     Base file name
   -osdirectory2       string     Output directory
   -osdbname2          string     Database name to add
   -ossingle2          boolean    Separate file for each entry
   -oufo2              string     UFO features
   -offormat2          string     Features format
   -ofname2            string     Features file name
   -ofdirectory2       string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>(Gapped) sequence(s) filename and optional format, or reference (input USA)</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-skip</td>
<td>Number of sequences to skip at start</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr>
<td>[-outseq]<br>(Parameter 2)</td>
<td>Sequence set(s) filename and optional format (output USA)</td>
<td>Writeable sequence(s)</td>
<td><i>&lt;*&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-feature</td>
<td>Use feature information</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

</table>

<H2>
    Input file format
</H2>
<b>skipseq</b> reads one or more sequence USAs.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:eclac*' is a sequence entry in the example nucleic acid database 'tembl'
<p>

<p>

See the documentation for 
<A href="seqret.html">seqret</a>
to see the full range of things that you can do when reading and writing
sequences. 

<H2>
    Output file format
</H2>

<b>skipseq</b> writes one or more sequences files, having optionally
skipped the first few sequences. 

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: eclac.fasta</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt;ECLAC J01636.1 E.coli lactose operon with lacI, lacZ, lacY and lacA genes.
gacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagt
caattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggt
gtctcttatcagaccgtttcccgcgtggtgaaccaggccagccacgtttctgcgaaaacg
cgggaaaaagtggaagcggcgatggcggagctgaattacattcccaaccgcgtggcacaa
caactggcgggcaaacagtcgttgctgattggcgttgccacctccagtctggccctgcac
gcgccgtcgcaaattgtcgcggcgattaaatctcgcgccgatcaactgggtgccagcgtg
gtggtgtcgatggtagaacgaagcggcgtcgaagcctgtaaagcggcggtgcacaatctt
ctcgcgcaacgcgtcagtgggctgatcattaactatccgctggatgaccaggatgccatt
gctgtggaagctgcctgcactaatgttccggcgttatttcttgatgtctctgaccagaca
cccatcaacagtattattttctcccatgaagacggtacgcgactgggcgtggagcatctg
gtcgcattgggtcaccagcaaatcgcgctgttagcgggcccattaagttctgtctcggcg
cgtctgcgtctggctggctggcataaatatctcactcgcaatcaaattcagccgatagcg
gaacgggaaggcgactggagtgccatgtccggttttcaacaaaccatgcaaatgctgaat
gagggcatcgttcccactgcgatgctggttgccaacgatcagatggcgctgggcgcaatg
cgcgccattaccgagtccgggctgcgcgttggtgcggatatctcggtagtgggatacgac
gataccgaagacagctcatgttatatcccgccgtcaaccaccatcaaacaggattttcgc
ctgctggggcaaaccagcgtggaccgcttgctgcaactctctcagggccaggcggtgaag
ggcaatcagctgttgcccgtctcactggtgaaaagaaaaaccaccctggcgcccaatacg
caaaccgcctctccccgcgcgttggccgattcattaatgcagctggcacgacaggtttcc
cgactggaaagcgggcagtgagcgcaacgcaattaatgtgagttagctcactcattaggc
accccaggctttacactttatgcttccggctcgtatgttgtgtggaattgtgagcggata
acaatttcacacaggaaacagctatgaccatgattacggattcactggccgtcgttttac
aacgtcgtgactgggaaaaccctggcgttacccaacttaatcgccttgcagcacatcccc
ctttcgccagctggcgtaatagcgaagaggcccgcaccgatcgcccttcccaacagttgc
gcagcctgaatggcgaatggcgctttgcctggtttccggcaccagaagcggtgccggaaa
gctggctggagtgcgatcttcctgaggccgatactgtcgtcgtcccctcaaactggcaga
tgcacggttacgatgcgcccatctacaccaacgtaacctatcccattacggtcaatccgc
cgtttgttcccacggagaatccgacgggttgttactcgctcacatttaatgttgatgaaa
gctggctacaggaaggccagacgcgaattatttttgatggcgttaactcggcgtttcatc
tgtggtgcaacgggcgctgggtcggttacggccaggacagtcgtttgccgtctgaatttg
acctgagcgcatttttacgcgccggagaaaaccgcctcgcggtgatggtgctgcgttgga
gtgacggcagttatctggaagatcaggatatgtggcggatgagcggcattttccgtgacg
tctcgttgctgcataaaccgactacacaaatcagcgatttccatgttgccactcgcttta
atgatgatttcagccgcgctgtactggaggctgaagttcagatgtgcggcgagttgcgtg
actacctacgggtaacagtttctttatggcagggtgaaacgcaggtcgccagcggcaccg
cgcctttcggcggtgaaattatcgatgagcgtggtggttatgccgatcgcgtcacactac
gtctgaacgtcgaaaacccgaaactgtggagcgccgaaatcccgaatctctatcgtgcgg
tggttgaactgcacaccgccgacggcacgctgattgaagcagaagcctgcgatgtcggtt
tccgcgaggtgcggattgaaaatggtctgctgctgctgaacggcaagccgttgctgattc
gaggcgttaaccgtcacgagcatcatcctctgcatggtcaggtcatggatgagcagacga
tggtgcaggatatcctgctgatgaagcagaacaactttaacgccgtgcgctgttcgcatt
atccgaaccatccgctgtggtacacgctgtgcgaccgctacggcctgtatgtggtggatg
aagccaatattgaaacccacggcatggtgccaatgaatcgtctgaccgatgatccgcgct
ggctaccggcgatgagcgaacgcgtaacgcgaatggtgcagcgcgatcgtaatcacccga
gtgtgatcatctggtcgctggggaatgaatcaggccacggcgctaatcacgacgcgctgt
atcgctggatcaaatctgtcgatccttcccgcccggtgcagtatgaaggcggcggagccg
acaccacggccaccgatattatttgcccgatgtacgcgcgcgtggatgaagaccagccct
tcccggctgtgccgaaatggtccatcaaaaaatggctttcgctacctggagagacgcgcc
cgctgatcctttgcgaatacgcccacgcgatgggtaacagtcttggcggtttcgctaaat


<font color=red>  [Part of this file has been deleted for brevity]</font>

gttacccaacttaatcgccttgcagcacatccccctttcgccagctggcgtaatagcgaa
gaggcccgcaccgatcgcccttcccaacagttgcgcagcctgaatggcgaatggcgcttt
gcctggtttccggcaccagaagcggtgccggaaagctggctggagtgcgatcttcctgag
gccgatactgtcgtcgtcccctcaaactggcagatgcacggttacgatgcgcccatctac
accaacgtaacctatcccattacggtcaatccgccgtttgttcccacggagaatccgacg
ggttgttactcgctcacatttaatgttgatgaaagctggctacaggaaggccagacgcga
attatttttgatggcgttaactcggcgtttcatctgtggtgcaacgggcgctgggtcggt
tacggccaggacagtcgtttgccgtctgaatttgacctgagcgcatttttacgcgccgga
gaaaaccgcctcgcggtgatggtgctgcgttggagtgacggcagttatctggaagatcag
gatatgtggcggatgagcggcattttccgtgacgtctcgttgctgcataaaccgactaca
caaatcagcgatttccatgttgccactcgctttaatgatgatttcagccgcgctgtactg
gaggctgaagttcagatgtgcggcgagttgcgtgactacctacgggtaacagtttcttta
tggcagggtgaaacgcaggtcgccagcggcaccgcgcctttcggcggtgaaattatcgat
gagcgtggtggttatgccgatcgcgtcacactacgtctgaacgtcgaaaacccgaaactg
tggagcgccgaaatcccgaatctctatcgtgcggtggttgaactgcacaccgccgacggc
acgctgattgaagcagaagcctgcgatgtcggtttccgcgaggtgcggattgaaaatggt
ctgctgctgctgaacggcaagccgttgctgattcgaggcgttaaccgtcacgagcatcat
cctctgcatggtcaggtcatggatgagcagacgatggtgcaggatatcctgctgatgaag
cagaacaactttaacgccgtgcgctgttcgcattatccgaaccatccgctgtggtacacg
ctgtgcgaccgctacggcctgtatgtggtggatgaagccaatattgaaacccacggcatg
gtgccaatgaatcgtctgaccgatgatccgcgctggctaccggcgatgagcgaacgcgta
acgcgaatggtgcagcgcgatcgtaatcacccgagtgtgatcatctggtcgctggggaat
gaatcaggccacggcgctaatcacgacgcgctgtatcgctggatcaaatctgtcgatcct
tcccgcccggtgcagtatgaaggcggcggagccgacaccacggccaccgatattatttgc
ccgatgtacgcgcgcgtggatgaagaccagcccttcccggctgtgccgaaatggtccatc
aaaaaatggctttcgctacctggagagacgcgcccgctgatcctttgcgaatacgcccac
gcgatgggtaacagtcttggcggtttcgctaaatactggcaggcgtttcgtcagtatccc
cgtttacagggcggcttcgtctgggactgggtggatcagtcgctgattaaatatgatgaa
aacggcaacccgtggtcggcttacggcggtgattttggcgatacgccgaacgatcgccag
ttctgtatgaacggtctggtctttgccgaccgcacgccgcatccagcgctgacggaagca
aaacaccagcagcagtttttccagttccgtttatccgggcaaaccatcgaagtgaccagc
gaatacctgttccgtcatagcgataacgagctcctgcactggatggtggcgctggatggt
aagccgctggcaagcggtgaagtgcctctggatgtcgctccacaaggtaaacagttgatt
gaactgcctgaactaccgcagccggagagcgccgggcaactctggctcacagtacgcgta
gtgcaaccgaacgcgaccgcatggtcagaagccgggcacatcagcgcctggcagcagtgg
cgtctggcggaaaacctcagtgtgacgctccccgccgcgtcccacgccatcccgcatctg
accaccagcgaaatggatttttgcatcgagctgggtaataagcgttggcaatttaaccgc
cagtcaggctttctttcacagatgtggattggcgataaaaaacaactgctgacgccgctg
cgcgatcagttcacccgtgcaccgctggataacgacattggcgtaagtgaagcgacccgc
attgaccctaacgcctgggtcgaacgctggaaggcggcgggccattaccaggccgaagca
gcgttgttgcagtgcacggcagatacacttgctgatgcggtgctgattacgaccgctcac
gcgtggcagcatcaggggaaaaccttatttatcagccggaaaacctaccggattgatggt
agtggtcaaatggcgattaccgttgatgttgaagtggcgagcgatacaccgcatccggcg
cggattggcctgaactgccagctggcgcaggtagcagagcgggtaaactggctcggatta
gggccgcaagaaaactatcccgaccgccttactgccgcctgttttgaccgctgggatctg
ccattgtcagacatgtataccccgtacgtcttcccgagcgaaaacggtctgcgctgcggg
acgcgcgaattgaattatggcccacaccagtggcgcggcgacttccagttcaacatcagc
cgctacagtcaacagcaactgatggaaaccagccatcgccatctgctgcacgcggaagaa
ggcacatggctgaatatcgacggtttccatatggggattggtggcgacgactcctggagc
ccgtcagtatcggcggaattccagctgagcgccggtcgctaccattaccagttggtctgg
tgtcaaaaataataataa
</pre>
</td></tr></table><p>

<a name="output.2"></a>
<h3>Output files for usage example 2</h3>
<p><h3>File: eclac.fasta</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt;ECLACA X51872.1 Escherichia coli lacA gene for thiogalactoside transacetylase
gtgaatgaagtcgcttaagcaatcaatgtcggatgcggcgcgacgcttatccgaccaaca
tatcataacggagtgatcgcattgaacatgccaatgaccgaaagaataagagcaggcaag
ctatttaccgatatgtgcgaaggcttaccggaaaaaagacttcgtgggaaaacgttaatg
tatgagtttaatcactcgcatccatcagaagttgaaaaaagagaaagcctgattaaagaa
atgtttgccacggtaggggaaaacgcctgggtagaaccgcctgtctatttctcttacggt
tccaacatccatataggccgcaatttttatgcaaatttcaatttaaccattgtcgatgac
tacacggtaacaatcggtgataacgtactgattgcacccaacgttactctttccgttacg
ggacaccctgtacaccatgaattgagaaaaaacggcgagatgtactcttttccgataacg
attggcaataacgtctggatcggaagtcatgtggttattaatccaggcgtcaccatcggg
gataattctgttattggcgcgggtagtatcgtcacaaaagacattccaccaaacgtcgtg
gcggctggcgttccttgtcgggttattcgcgaaataaacgaccgggataagcactattat
ttcaaagattataaagttgaatcgtcagtttaaattataaaaattgcctgatacgctgcg
cttatcaggcctacaagttcagcgatctacattagccgcatccggcatgaacaaagcgca
ggaacaagcgtcgcatcatgcctctttgacccacagctgcggaaaacgtactggtgcaaa
acgcagggttatgatcatcagcccaacgacgcacagcgcatgaaatgcccagtccatcag
gtaattgccgctgatactacgcagcacgccagaaaaccacggggcaagcccggcgatgat
aaaaccgattccctgcataaacgccaccagcttgccagcaatagccggttgcacagagtg
atcgagcgccagcagcaaacagagcggaaacgcgccgcccagacctaacccacacaccat
cgcccacaataccggcaattgcatcggcagccagataaagccgcagaaccccaccagttg
taacaccagcgccagcattaacagtttgcgccgatcctgatggcgagccatagcaggcat
cagcaaagctcctgcggcttgcccaagcgtcatcaatgccagtaaggaaccgctgtactg
cgcgctggcaccaatctcaatatagaaagcgggtaaccaggcaatcaggctggcgtaacc
gccgttaatcagaccgaagtaaacacccagcgtccacgcgcggggagtgaataccacgcg
aaccggagtggttgttgtcttgtgggaagaggcgacctcgcgggcgctttgccaccacca
ggcaaagagcgcaacaacggcaggcagcgccaccaggcgagtgtttgataccaggtttcg
ctatgttgaactaaccagggcgttatggcggcaccaagcccaccgccgcccatcagagcc
gcggaccacagccccatcaccagtggcgtgcgctgctgaaaccgccgtttaatcaccgaa
gcatcaccgcctgaatgatgccgatccccaccccaccaagcagtgcgctgctaagcagca
gcgcactttgcgggtaaagctcacgcatcaatgcaccgacggcaatcagcaacagactga
tggcgacactgcgacgttcgctgacatgctgatgaagccagcttccggccagcgccagcc
cgcccatggtaaccaccggcagagcggtcgac
&gt;ECLACI V00294.1 E. coli laci gene (codes for the lac repressor).
ccggaagagagtcaattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgca
gagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccagccacgtt
tctgcgaaaacgcgggaaaaagtggaagcggcgatggcggagctgaattacattcccaac
cgcgtggcacaacaactggcgggcaaacagtcgttgctgattggcgttgccacctccagt
ctggccctgcacgcgccgtcgcaaattgtcgcggcgattaaatctcgcgccgatcaactg
ggtgccagcgtggtggtgtcgatggtagaacgaagcggcgtcgaagcctgtaaagcggcg
gtgcacaatcttctcgcgcaacgcgtcagtgggctgatcattaactatccgctggatgac
caggatgccattgctgtggaagctgcctgcactaatgttccggcgttatttcttgatgtc
tctgaccagacacccatcaacagtattattttctcccatgaagacggtacgcgactgggc
gtggagcatctggtcgcattgggtcaccagcaaatcgcgctgttagcgggcccattaagt
tctgtctcggcgcgtctgcgtctggctggctggcataaatatctcactcgcaatcaaatt
cagccgatagcggaacgggaaggcgactggagtgccatgtccggttttcaacaaaccatg
caaatgctgaatgagggcatcgttcccactgcgatgctggttgccaacgatcagatggcg
ctgggcgcaatgcgcgccattaccgagtccgggctgcgcgttggtgcggatatctcggta
gtgggatacgacgataccgaagacagctcatgttatatcccgccgtcaaccaccatcaaa
caggattttcgcctgctggggcaaaccagcgtggaccgcttgctgcaactctctcagggc
caggcggtgaagggcaatcagctgttgcccgtctcactggtgaaaagaaaaaccaccctg


<font color=red>  [Part of this file has been deleted for brevity]</font>

gttacccaacttaatcgccttgcagcacatccccctttcgccagctggcgtaatagcgaa
gaggcccgcaccgatcgcccttcccaacagttgcgcagcctgaatggcgaatggcgcttt
gcctggtttccggcaccagaagcggtgccggaaagctggctggagtgcgatcttcctgag
gccgatactgtcgtcgtcccctcaaactggcagatgcacggttacgatgcgcccatctac
accaacgtaacctatcccattacggtcaatccgccgtttgttcccacggagaatccgacg
ggttgttactcgctcacatttaatgttgatgaaagctggctacaggaaggccagacgcga
attatttttgatggcgttaactcggcgtttcatctgtggtgcaacgggcgctgggtcggt
tacggccaggacagtcgtttgccgtctgaatttgacctgagcgcatttttacgcgccgga
gaaaaccgcctcgcggtgatggtgctgcgttggagtgacggcagttatctggaagatcag
gatatgtggcggatgagcggcattttccgtgacgtctcgttgctgcataaaccgactaca
caaatcagcgatttccatgttgccactcgctttaatgatgatttcagccgcgctgtactg
gaggctgaagttcagatgtgcggcgagttgcgtgactacctacgggtaacagtttcttta
tggcagggtgaaacgcaggtcgccagcggcaccgcgcctttcggcggtgaaattatcgat
gagcgtggtggttatgccgatcgcgtcacactacgtctgaacgtcgaaaacccgaaactg
tggagcgccgaaatcccgaatctctatcgtgcggtggttgaactgcacaccgccgacggc
acgctgattgaagcagaagcctgcgatgtcggtttccgcgaggtgcggattgaaaatggt
ctgctgctgctgaacggcaagccgttgctgattcgaggcgttaaccgtcacgagcatcat
cctctgcatggtcaggtcatggatgagcagacgatggtgcaggatatcctgctgatgaag
cagaacaactttaacgccgtgcgctgttcgcattatccgaaccatccgctgtggtacacg
ctgtgcgaccgctacggcctgtatgtggtggatgaagccaatattgaaacccacggcatg
gtgccaatgaatcgtctgaccgatgatccgcgctggctaccggcgatgagcgaacgcgta
acgcgaatggtgcagcgcgatcgtaatcacccgagtgtgatcatctggtcgctggggaat
gaatcaggccacggcgctaatcacgacgcgctgtatcgctggatcaaatctgtcgatcct
tcccgcccggtgcagtatgaaggcggcggagccgacaccacggccaccgatattatttgc
ccgatgtacgcgcgcgtggatgaagaccagcccttcccggctgtgccgaaatggtccatc
aaaaaatggctttcgctacctggagagacgcgcccgctgatcctttgcgaatacgcccac
gcgatgggtaacagtcttggcggtttcgctaaatactggcaggcgtttcgtcagtatccc
cgtttacagggcggcttcgtctgggactgggtggatcagtcgctgattaaatatgatgaa
aacggcaacccgtggtcggcttacggcggtgattttggcgatacgccgaacgatcgccag
ttctgtatgaacggtctggtctttgccgaccgcacgccgcatccagcgctgacggaagca
aaacaccagcagcagtttttccagttccgtttatccgggcaaaccatcgaagtgaccagc
gaatacctgttccgtcatagcgataacgagctcctgcactggatggtggcgctggatggt
aagccgctggcaagcggtgaagtgcctctggatgtcgctccacaaggtaaacagttgatt
gaactgcctgaactaccgcagccggagagcgccgggcaactctggctcacagtacgcgta
gtgcaaccgaacgcgaccgcatggtcagaagccgggcacatcagcgcctggcagcagtgg
cgtctggcggaaaacctcagtgtgacgctccccgccgcgtcccacgccatcccgcatctg
accaccagcgaaatggatttttgcatcgagctgggtaataagcgttggcaatttaaccgc
cagtcaggctttctttcacagatgtggattggcgataaaaaacaactgctgacgccgctg
cgcgatcagttcacccgtgcaccgctggataacgacattggcgtaagtgaagcgacccgc
attgaccctaacgcctgggtcgaacgctggaaggcggcgggccattaccaggccgaagca
gcgttgttgcagtgcacggcagatacacttgctgatgcggtgctgattacgaccgctcac
gcgtggcagcatcaggggaaaaccttatttatcagccggaaaacctaccggattgatggt
agtggtcaaatggcgattaccgttgatgttgaagtggcgagcgatacaccgcatccggcg
cggattggcctgaactgccagctggcgcaggtagcagagcgggtaaactggctcggatta
gggccgcaagaaaactatcccgaccgccttactgccgcctgttttgaccgctgggatctg
ccattgtcagacatgtataccccgtacgtcttcccgagcgaaaacggtctgcgctgcggg
acgcgcgaattgaattatggcccacaccagtggcgcggcgacttccagttcaacatcagc
cgctacagtcaacagcaactgatggaaaccagccatcgccatctgctgcacgcggaagaa
ggcacatggctgaatatcgacggtttccatatggggattggtggcgacgactcctggagc
ccgtcagtatcggcggaattccagctgagcgccggtcgctaccattaccagttggtctgg
tgtcaaaaataataataa
</pre>
</td></tr></table><p>

<p>


See the documentation for
<A href="seqret.html">seqret</a>
to see the full range of things that you can do when reading and writing
sequences.
         

<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

You may skip all of the input files.
<br>
If the -skip value is greater than or equal to the number of input
sequences, there is no warning: this is legal.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with status 0.

<H2>
    Known bugs
</H2>

None.


<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr>
<td><a href="biosed.html">biosed</a></td>
<td>Replace or delete sequence sections</td>
</tr>

<tr>
<td><a href="codcopy.html">codcopy</a></td>
<td>Reads and writes a codon usage table</td>
</tr>

<tr>
<td><a href="cutseq.html">cutseq</a></td>
<td>Removes a specified section from a sequence</td>
</tr>

<tr>
<td><a href="degapseq.html">degapseq</a></td>
<td>Removes gap characters from sequences</td>
</tr>

<tr>
<td><a href="descseq.html">descseq</a></td>
<td>Alter the name or description of a sequence</td>
</tr>

<tr>
<td><a href="entret.html">entret</a></td>
<td>Reads and writes (returns) flatfile entries</td>
</tr>

<tr>
<td><a href="extractalign.html">extractalign</a></td>
<td>Extract regions from a sequence alignment</td>
</tr>

<tr>
<td><a href="extractfeat.html">extractfeat</a></td>
<td>Extract features from a sequence</td>
</tr>

<tr>
<td><a href="extractseq.html">extractseq</a></td>
<td>Extract regions from a sequence</td>
</tr>

<tr>
<td><a href="listor.html">listor</a></td>
<td>Write a list file of the logical OR of two sets of sequences</td>
</tr>

<tr>
<td><a href="makenucseq.html">makenucseq</a></td>
<td>Creates random nucleotide sequences</td>
</tr>

<tr>
<td><a href="makeprotseq.html">makeprotseq</a></td>
<td>Creates random protein sequences</td>
</tr>

<tr>
<td><a href="maskfeat.html">maskfeat</a></td>
<td>Mask off features of a sequence</td>
</tr>

<tr>
<td><a href="maskseq.html">maskseq</a></td>
<td>Mask off regions of a sequence</td>
</tr>

<tr>
<td><a href="newseq.html">newseq</a></td>
<td>Type in a short new sequence</td>
</tr>

<tr>
<td><a href="noreturn.html">noreturn</a></td>
<td>Removes carriage return from ASCII files</td>
</tr>

<tr>
<td><a href="notseq.html">notseq</a></td>
<td>Exclude a set of sequences and write out the remaining ones</td>
</tr>

<tr>
<td><a href="nthseq.html">nthseq</a></td>
<td>Writes one sequence from a multiple set of sequences</td>
</tr>

<tr>
<td><a href="pasteseq.html">pasteseq</a></td>
<td>Insert one sequence into another</td>
</tr>

<tr>
<td><a href="revseq.html">revseq</a></td>
<td>Reverse and complement a sequence</td>
</tr>

<tr>
<td><a href="seqret.html">seqret</a></td>
<td>Reads and writes (returns) sequences</td>
</tr>

<tr>
<td><a href="seqretsplit.html">seqretsplit</a></td>
<td>Reads and writes (returns) sequences in individual files</td>
</tr>

<tr>
<td><a href="splitter.html">splitter</a></td>
<td>Split a sequence into (overlapping) smaller sequences</td>
</tr>

<tr>
<td><a href="trimest.html">trimest</a></td>
<td>Trim poly-A tails off EST sequences</td>
</tr>

<tr>
<td><a href="trimseq.html">trimseq</a></td>
<td>Trim ambiguous bits off the ends of sequences</td>
</tr>

<tr>
<td><a href="union.html">union</a></td>
<td>Reads sequence fragments and builds one sequence</td>
</tr>

<tr>
<td><a href="vectorstrip.html">vectorstrip</a></td>
<td>Strips out DNA between a pair of vector sequences</td>
</tr>

<tr>
<td><a href="yank.html">yank</a></td>
<td>Reads a sequence range, appends the full USA to a list file</td>
</tr>

</table>
<p>


<b>skipseq</b> is a variant of the standard program for reading and
writing sequences, <b>seqret</b>.
<H2>
    Author(s)
</H2>
Peter Rice (pmr&nbsp;&copy;&nbsp;ebi.ac.uk)
<br>
Informatics Division, European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK


<H2>
    History
</H2>


Written (25 June 2002) - Peter Rice

<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.

<H2>
    Comments
</H2>
None

</BODY>
</HTML>

