<HTML>

<HEAD>
  <TITLE>
  EMBOSS: Stretcher
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: Stretcher
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program Stretcher
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Finds the best global alignment between two sequences


<H2>
    Description
</H2>

<pre>

 calculates a global alignment of two sequences

 Please cite: Myers and Miller, CABIOS (1989)
 version 2.0u. Modified for EMBOSS May 1999 
</pre>

<H2>
    Usage
</H2>


 Here is a sample session with stretcher.

<pre>

 stretcher -sequencea s1 -sequenceb s2 


 hbahum                                              141  vs.
 hbbhum                                              146 
 scoring matrix: BLOSUM62, gap penalties: 12/2
 43.2% identity;         Global alignment score: 272

                 10        20        30        40         50        
 hbahum V-LSPADKTNVKAAWGKVGAHAGEYGAEALERMFLSFPTTKTYFPHF-DLSH-----GSA
        : :.: .:. : : ::::  .. : :.::: :... .: :. .:  : :::      :. 
 hbbhum VHLTPEEKSAVTALWGKV--NVDEVGGEALGRLLVVYPWTQRFFESFGDLSTPDAVMGNP
                10          20        30        40        50        

             60        70        80        90       100       110   
 hbahum QVKGHGKKVADALTNAVAHVDDMPNALSALSDLHAHKLRVDPVNFKLLSHCLLVTLAAHL
        .::.:::::  :.....::.:.. .....::.::. ::.::: ::.::.. :. .:: :.
 hbbhum KVKAHGKKVLGAFSDGLAHLDNLKGTFATLSELHCDKLHVDPENFRLLGNVLVCVLAHHF
        60        70        80        90       100       110        

            120       130       140 
 hbahum PAEFTPAVHASLDKFLASVSTVLTSKYR
          :::: :.:. .: .:.:...:. ::.
 hbbhum GKEFTPPVQAAYQKVVAGVANALAHKYH
       120       130       140      

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequencea]         sequence   Sequence USA
  [-sequenceb]         sequence   Sequence USA
  [-outfile]           outfile    Output file name

   Optional qualifiers:
   -datafile           matrix     Matrix file
   -gappenalty         integer    gappenalty
   -gaplength          integer    gap length penalty
   -markx              integer    alternative displays
   -length             integer    number per line

   Advanced qualifiers: (none)
   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequencea]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-sequenceb]<br>(Parameter 2)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 3)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.stretcher</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-datafile</td>
<td>Matrix file</td>
<td>Comparison matrix file in EMBOSS data path</td>
<td>EBLOSUM62 for protein<br>EDNAFULL for DNA</td>
</tr>

<tr>
<td>-gappenalty</td>
<td>gappenalty</td>
<td>Positive integer</td>
<td>12 for protein, 16 for nucleic</td>
</tr>

<tr>
<td>-gaplength</td>
<td>gap length penalty</td>
<td>Positive integer</td>
<td>2 for protein, 4 for nucleic</td>
</tr>

<tr>
<td>-markx</td>
<td>alternative displays</td>
<td>Positive integer up to 10</td>
<td>0</td>
</tr>

<tr>
<td>-length</td>
<td>number per line</td>
<td>Positive integer</td>
<td>60</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>


Any 2 sequence USAs of the same type (DNA or protein).

<H2>
    Output file format
</H2>


<H2>
    Data files
</H2>


For protein sequences EBLOSUM62 is used for the substitution
matrix. For nucleotide sequence, EDNAMAT is used. Others can be
specified.

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by EMBOSS environment variable EMBOSS_DATA.

<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>

<H2>
    Notes
</H2>


<H2>
    References
</H2>

<ol>

<li>
 Myers and Miller, CABIOS (1989)
</ol>

<H2>
    Warnings
</H2>


<H2>
    Diagnostic Error Messages
</H2>


<H2>
    Exit status
</H2>


<H2>
    Known bugs
</H2>


<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="est2genome.html">est2genome</a></td><td>Align EST and genomic DNA sequences</td></tr>
<tr><td><a href="needle.html">needle</a></td><td>Needleman-Wunsch global alignment</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was modified for inclusion in EMBOSS by Ian Longden <a
href="mailto:il@sanger.ac.uk">(il@sanger.ac.uk)</a> Informatics
Division, The Sanger Centre, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.

<H2>
    History
</H2>

<pre>
 Completed 13th May 1999.
</pre>

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

