<HTML>

<HEAD>
  <TITLE>
  EMBOSS: stretcher
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">







<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

stretcher
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>







<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Finds the best global alignment between two sequences


<H2>
    Description
</H2>

<pre>

 calculates a global alignment of two sequences

 Please cite: Myers and Miller, CABIOS (1989)
 version 2.0u. Modified for EMBOSS May 1999 
</pre>

<H2>
    Usage
</H2>


 Here is a sample session with stretcher.

<pre>

% stretcher tsw:hba_human tsw:hbb_human
Finds the best global alignment between two sequences
Output alignment [hba_human.stretcher]: 

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequencea]         sequence   Sequence USA
  [-sequenceb]         sequence   Sequence USA
  [-outfile]           align      Output alignment file name

   Optional qualifiers:
   -datafile           matrix     This is the scoring matrix file used when
                                  comparing sequences. By default it is the
                                  file 'EBLOSUM62' (for proteins) or the file
                                  'EDNAFULL' (for nucleic sequences). These
                                  files are found in the 'data' directory of
                                  the EMBOSS installation.
   -gappenalty         integer    Gap penalty
   -gaplength          integer    Gap length penalty

   Advanced qualifiers: (none)
   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequencea]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-sequenceb]<br>(Parameter 2)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 3)</td>
<td>Output alignment file name</td>
<td>Alignment file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-datafile</td>
<td>This is the scoring matrix file used when comparing sequences. By default it is the file 'EBLOSUM62' (for proteins) or the file 'EDNAFULL' (for nucleic sequences). These files are found in the 'data' directory of the EMBOSS installation.</td>
<td>Comparison matrix file in EMBOSS data path</td>
<td>EBLOSUM62 for protein<br>EDNAFULL for DNA</td>
</tr>

<tr>
<td>-gappenalty</td>
<td>Gap penalty</td>
<td>Positive integer</td>
<td>12 for protein, 16 for nucleic</td>
</tr>

<tr>
<td>-gaplength</td>
<td>Gap length penalty</td>
<td>Positive integer</td>
<td>2 for protein, 4 for nucleic</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>


Any 2 sequence USAs of the same type (DNA or protein).

<H2>
    Output file format
</H2>

<p>

The output is a standard EMBOSS alignment file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-aformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  Some of the alignment formats can cope
with an unlimited number of sequences, while others are only for pairs
of sequences.  

<p>

The available multiple alignment format names are: unknown, multiple,
simple, fasta, msf, trace, srs

<p>

The available pairwise alignment format names are: pair, markx0, markx1,
markx2, markx3, markx10, srspair, score

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/AlignFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/AlignFormats.html</A>
for further information on alignment formats.

<p>


<p>

The output from the example follows:

<p>

<pre>
########################################
# Program:  stretcher
# Rundate:  Mon May 20 16:25:11 2002
# Report_file: hba_human.stretcher
########################################
#=======================================
#
# Aligned_sequences: 2
# 1: HBA_HUMAN
# 2: HBB_HUMAN
# Matrix: EBLOSUM62
# Gap_penalty: 12
# Extend_penalty: 2
#
# Length: 148
# Identity:      64/148 (43.2%)
# Similarity:    89/148 (60.1%)
# Gaps:           9/148 ( 6.1%)
# Score: 272
#
#
#=======================================


                10        20        30        40
HBA_HU V-LSPADKTNVKAAWGKVGAHAGEYGAEALERMFLSFPTTKTYFPHF-DL
       : :.: .:. : : ::::  .. : :.::: :... .: :. .:  : ::
HBB_HU VHLTPEEKSAVTALWGKV--NVDEVGGEALGRLLVVYPWTQRFFESFGDL
               10          20        30        40

       50             60        70        80        90
HBA_HU SH-----GSAQVKGHGKKVADALTNAVAHVDDMPNALSALSDLHAHKLRV
       :      :. .::.:::::  :.....::.:.. .....::.::. ::.:
HBB_HU STPDAVMGNPKVKAHGKKVLGAFSDGLAHLDNLKGTFATLSELHCDKLHV
       50        60        70        80        90

           100       110       120       130       140
HBA_HU DPVNFKLLSHCLLVTLAAHLPAEFTPAVHASLDKFLASVSTVLTSKYR
       :: ::.::.. :. .:: :.  :::: :.:. .: .:.:...:. ::.
HBB_HU DPENFRLLGNVLVCVLAHHFGKEFTPPVQAAYQKVVAGVANALAHKYH
      100       110       120       130       140

#---------------------------------------
#---------------------------------------         
</pre>

<H2>
    Data files
</H2>


For protein sequences EBLOSUM62 is used for the substitution
matrix. For nucleotide sequence, EDNAMAT is used. Others can be
specified.

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by EMBOSS environment variable EMBOSS_DATA.

<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>

<H2>
    Notes
</H2>


None.

<H2>
    References
</H2>

<ol>

<li>
 Myers and Miller, CABIOS (1989)
</ol>

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It exits with a status of 0.

<H2>
    Known bugs
</H2>

None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="alignwrap.html">alignwrap</a></td><td>Aligns a set of sequences to a seed alignment</td></tr>
<tr><td><a href="est2genome.html">est2genome</a></td><td>Align EST and genomic DNA sequences</td></tr>
<tr><td><a href="needle.html">needle</a></td><td>Needleman-Wunsch global alignment</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was modified for inclusion in EMBOSS by Ian Longden <a
href="mailto:il@sanger.ac.uk">(il@sanger.ac.uk)</a> Informatics
Division, The Sanger Centre, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.

<H2>
    History
</H2>

<pre>
 Completed 13th May 1999.
</pre>

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

