<HTML>

<HEAD>
  <TITLE>
  EMBOSS: stssearch
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: stssearch
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program stssearch
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Searches a DNA database for matches with a set of STS primers


<H2>
    Description
</H2>


Searches a DNA sequence database with a set of STS primers and reports
expected matches and possible mismatches. A replacement for stssearch
in the EGCG package.
<P>

This reads in one or more sequences to be searched.  For each pair of
primers, this looks for exact matches between a the primers and the
query sequence in either orientation.
<p>

Any matches found will be reported. Only one primer need match.
<p>



<H2>
    Usage
</H2>


Here is a sample session with stssearch.

<pre>

% stssearch
Input sequence(s): embl:eclac*
Primer file: lac.primers
</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequences]         seqall     Sequence database USA
  [-primers]           infile     Primer file
  [-out]               outfile    Output file name

   Optional qualifiers: (none)
   Advanced qualifiers: (none)

</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequences]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-primers]<br>(Parameter 2)</td>
<td>Primer file</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-out]<br>(Parameter 3)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.stssearch</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>


Here is the primer input file for the example run:

<hr>
<pre>
PrimA ACCAGACACCCATCAACAG    TATTTATGCCAGCCAGCCAG
PrimB CGAAAGAATAAGAGCAGGCAAG GTAAGAGAAATAGACAGGCGG
PrimC CGTCAGTATCCCCGTTTACAG  TATCGCCAAAATCACCGCC
PrimD AATACGCAAACCGCCTCTCC   TTATCCGCTCACAATTCCACAC
PrimE AATACGCAAACCGCCTCTCC   CACAACCCGCTCACAATTCCA
</pre>
<hr>
<p>

It consists of three columns separated by tabs or spaces.
<P>

The first column is the name of the primer pair.
<br>
The second column is the sequence of the first primer.
<br>
The third column is the sequence of the second primer.


<H2>
    Output file format
</H2>


Here is the output from the example run:

<hr>
<pre>
ECLAC: PrimA PrimerA matched at 532
ECLAC: (rev) PrimA PrimerB matched at 689
ECLAC: PrimB PrimerA matched at 5743
ECLAC: (rev) PrimB PrimerB matched at 5942
ECLAC: PrimC PrimerA matched at 2954
ECLAC: (rev) PrimC PrimerB matched at 3069
ECLAC: PrimD PrimerA matched at 1074
ECLAC: (rev) PrimD PrimerB matched at 1261
ECLAC: PrimE PrimerA matched at 1074
ECLACA: PrimB PrimerA matched at 98
ECLACA: (rev) PrimB PrimerB matched at 297
ECLACI: PrimA PrimerA matched at 484
ECLACI: (rev) PrimA PrimerB matched at 641
ECLACI: PrimD PrimerA matched at 1026
ECLACI: PrimE PrimerA matched at 1026
ECLACY: PrimB PrimerA matched at 1439
ECLACZ: PrimC PrimerA matched at 1668
ECLACZ: (rev) PrimC PrimerB matched at 1783
</pre>
<hr>
<p>

This consists of one line per match. This consists of:
<p>
<ul>
<li>The name of the sequence that the match is found in, followed by a ':'.
<li>If the match is in the reverse sense, it has a '(rev)'.
<li>The name of the primer pair.
<li>'PrimerA' or 'PrimerB'.
<li>'matched at' the start of the match position.
</ul>

<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>


None.

<H2>
    References
</H2>

None.


<H2>
    Warnings
</H2>

None.


<H2>
    Diagnostic Error Messages
</H2>

None.


<H2>
    Exit status
</H2>

None.


<H2>
    Known bugs
</H2>

None.


<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="dotmatcher.html">dotmatcher</a></td><td>Displays a thresholded dotplot of two sequences</td></tr>
<tr><td><a href="polydot.html">polydot</a></td><td>Displays all-against-all dotplots of a set of sequences</td></tr>
<tr><td><a href="prima.html">prima</a></td><td>Selects primers for PCR and DNA amplification</td></tr>
<tr><td><a href="primersearch.html">primersearch</a></td><td>Searches DNA sequences for matches with primer pairs</td></tr>
<tr><td><a href="seqmatchall.html">seqmatchall</a></td><td>Does an all-against-all comparison of a set of sequences</td></tr>
<tr><td><a href="wordmatch.html">wordmatch</a></td><td>Finds all exact matches of a given size between 2 sequences</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->
<p>

If you want something that only reports matches of both primer pairs and
can find mismatches, use <A href="primersearch.html">primersearch</a>. 

<H2>
    Author(s)
</H2>


This application was written by Peter Rice <a
href="mailto:pmr@sanger.ac.uk">(pmr@sanger.ac.uk)</a> Informatics
Division, The Sanger Centre, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.

<H2>
    History
</H2>

Finished.


<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

