<HTML>
<HEAD>
  <TITLE>
  EMBOSS: textget
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
textget
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
Get text data entries
<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<b>textget</b> reads text data, usually from a named database, and
writes a text file containing the original text of the entry or
entries.


<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

Here is a sample session with <b>textget</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>textget "srs:unilib:99" </b>
Get text data entries
Text output file [outfile.text]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
Get text data entries
Version: EMBOSS:6.4.0.0

   Standard (Mandatory) qualifiers:
  [-text]              text       Text filename and optional format, or
                                  reference (input query)
  [-outfile]           outtext    (no help text) outtext value

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-text" associated qualifiers
   -iformat1           string     Input text format
   -idbname1           string     User-provided database name

   "-outfile" associated qualifiers
   -odirectory2        string     Output directory
   -oformat2           string     Text output format

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-text]<br>(Parameter 1)</td>
<td>text</td>
<td>Text filename and optional format, or reference (input query)</td>
<td>Text entries</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-outfile]<br>(Parameter 2)</td>
<td>outtext</td>
<td>(no help text) outtext value</td>
<td>Text entries</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-text" associated text qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -iformat1<br>-iformat_text</td>
<td>string</td>
<td>Input text format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -idbname1<br>-idbname_text</td>
<td>string</td>
<td>User-provided database name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-outfile" associated outtext qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -odirectory2<br>-odirectory_outfile</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -oformat2<br>-oformat_outfile</td>
<td>string</td>
<td>Text output format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<p>

The input is a standard EMBOSS text query.

<p>

The major text sources defined as standard in EMBOSS installations are
non-standard types on the srs server and databases with a "text" query
in the Data Resource Catalogue database DRCAT.

<p>

Data can also be read from text output in any supported text format
written by an EMBOSS or third-party application. The text format is
needed to identify the start and end of a single "entry" within a
larger text file.

<p>
The input format can be specified by using the command-line qualifier
<tt>-format xxx</tt>, where 'xxx' is replaced by the name of the
required format.  The available format names are: text, html, xml (uniprotxml),
obo, embl (swissprot)

<p>

See:
<A href="http://emboss.sf.net/docs/themes/TextFormats.html">
http://emboss.sf.net/docs/themes/TextFormats.html</A>
for further information on text formats.

<p>

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>Database entry: "srs:unilib:99"</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
&gt; 99
CGAP Lib id: 182
Unigene id: 729
dbEST lib id: 1452
Keyword: lymph node
Keyword: follicular lymphoma
Keyword: non-normalized
Keyword: bulk
Keyword: CGAP
Keyword: EST
Keyword: Stratagene non-normalized
Keyword: unknown developmental stage
Keyword: size fractionated
Keyword: directionally cloned
Keyword: phagemid
Keyword: oligo-dT primed
Keyword: EST
Keyword: CGAP
Lib Name: NCI_CGAP_Lym5
Organism: Homo sapiens
Organ: lymph node
Tissue_type: follicular lymphoma
Lab host: SOLR (Stratagene, kanamycin resistant)
Vector: pBluescript SK-
Vector type: phagemid (ampicillin resistant)
R. Site 1: EcoRI
R. Site 2: XhoI
Description: Cloned unidirectionally.  Primer: Oligo dT. Average insert size 1.2 kb. Non-amplified library. ~5' adaptor sequence: 5' GAATTCGGCACGAG 3' ~3' adaptor sequence: 5' CTCGAGTTTTTTTTTTTTTTTTTT 3'
Library treatment: non-normalized
Tissue description: Lymph node, follicular lymphoma, node 2
Tissue supplier: Mark Raffeld, M.D.
Sample type: Bulk
Image legend: 
Producer: Stratagene, Inc.
Clones generated to date: 2304
Sequences generated to date: 1293

</pre>
</td></tr></table><p>





<H2>
    Output file format
</H2>

<p>

The output is a standard text file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <tt>-oformat xxx</tt>, where 'xxx' is replaced by
the name of the required format.  The available format names are:
text, html, xml, json and list

<p>

See:
<A href="http://emboss.sf.net/docs/themes/TextFormats.html">
http://emboss.sf.net/docs/themes/TextFormats.html</A>
for further information on text formats.

<p>


<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: outfile.text</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt; 99
CGAP Lib id: 182
Unigene id: 729
dbEST lib id: 1452
Keyword: lymph node
Keyword: follicular lymphoma
Keyword: non-normalized
Keyword: bulk
Keyword: CGAP
Keyword: EST
Keyword: Stratagene non-normalized
Keyword: unknown developmental stage
Keyword: size fractionated
Keyword: directionally cloned
Keyword: phagemid
Keyword: oligo-dT primed
Keyword: EST
Keyword: CGAP
Lib Name: NCI_CGAP_Lym5
Organism: Homo sapiens
Organ: lymph node
Tissue_type: follicular lymphoma
Lab host: SOLR (Stratagene, kanamycin resistant)
Vector: pBluescript SK-
Vector type: phagemid (ampicillin resistant)
R. Site 1: EcoRI
R. Site 2: XhoI
Description: Cloned unidirectionally.  Primer: Oligo dT. Average insert size 1.2 kb. Non-amplified library. ~5' adaptor sequence: 5' GAATTCGGCACGAG 3' ~3' adaptor sequence: 5' CTCGAGTTTTTTTTTTTTTTTTTT 3'
Library treatment: non-normalized
Tissue description: Lymph node, follicular lymphoma, node 2
Tissue supplier: Mark Raffeld, M.D.
Sample type: Bulk
Image legend: 
Producer: Stratagene, Inc.
Clones generated to date: 2304
Sequences generated to date: 1293

</pre>
</td></tr></table><p>





<H2>
    Data files
</H2>

None.


<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
<ol>

<li>

</ol>

   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="drtext.html">drtext</a></td>
<td>Get data resource entries complete text</td>
</tr>

<tr>
<td><a href="entret.html">entret</a></td>
<td>Retrieves sequence entries from flatfile databases and files</td>
</tr>

<tr>
<td><a href="ontotext.html">ontotext</a></td>
<td>Get ontology term(s) original full text</td>
</tr>

<tr>
<td><a href="textsearch.html">textsearch</a></td>
<td>Search the textual description of sequence(s)</td>
</tr>

</table>
<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>

Peter Rice
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK

<p>
Please report all bugs to the EMBOSS bug team (emboss-bug&nbsp;&copy;&nbsp;emboss.open-bio.org) not to the original author.
<H2>
    History
</H2>
<!--
        Date written and what changes have been made go in this file.
   -->




<H2>
    Target users
</H2>
<!--
        For general users, use this text
   -->
This program is intended to be used by everyone and everything, from naive users to embedded scripts.

<H2>
    Comments
</H2>
<!--
        User/developer/other comments go in this file.
   -->
None


</BODY>
</HTML>

