<HTML>

<HEAD>
  <TITLE>
  EMBOSS: textsearch
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">







<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

textsearch
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>







<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Search sequence documentation text. SRS and Entrez are faster!


<H2>
    Description
</H2>


This is a small utility search for words in the description text of a
sequence and for each match list the sequence's name and/or
description.  NB. It only searches the description line of the
annotation, not the full annotation.

<H2>
    Usage
</H2>


<b>Here is a sample session with textsearch</b>
<p>
Search for 'lactose': 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>textsearch tsw:* 'lactose' </b>
Search sequence documentation text. SRS and Entrez are faster!
Output file [100k_rat.textsearch]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>
<p>
<b>Example 2</b>
<p>
Search for 'lactose' or 'permease' in E.coli proteins: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>textsearch tsw:*_ecoli 'lactose | permease' </b>
Search sequence documentation text. SRS and Entrez are faster!
Output file [laci_ecoli.textsearch]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.2">Go to the output files for this example</a><p><p>
<p>
<b>Example 3</b>
<p>
Output a search for 'lacz' formatted with HTML to a file: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>textsearch tembl:* 'lacz' -html -outfile embl.lacz.html </b>
Search sequence documentation text. SRS and Entrez are faster!

</pre></td></tr></table><p>
<p>
<a href="#input.3">Go to the input files for this example</a><br><a href="#output.3">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-sequence]          seqall     Sequence database USA
  [-pattern]           string     The search pattern is a regular expression.
                                  Use a | to indicate OR.
                                  For example:
                                  human|mouse
                                  will find text with either 'human' OR
                                  'mouse' in the text
  [-outfile]           outfile    Output file name

   Additional (Optional) qualifiers:
   -casesensitive      boolean    Do a case-sensitive search
   -html               boolean    Format output as an HTML table

   Advanced (Unprompted) qualifiers:
   -only               boolean    This is a way of shortening the command line
                                  if you only want a few things to be
                                  displayed. Instead of specifying:
                                  '-nohead -noname -nousa -noacc -nodesc'
                                  to get only the name output, you can specify
                                  '-only -name'
   -heading            boolean    Display column headings
   -usa                boolean    Display the USA of the sequence
   -accession          boolean    Display 'accession' column
   -name               boolean    Display 'name' column
   -description        boolean    Display 'description' column

   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1             integer    Start of each sequence to be used
   -send1               integer    End of each sequence to be used
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name

   "-outfile" associated qualifiers
   -odirectory3         string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-pattern]<br>(Parameter 2)</td>
<td>The search pattern is a regular expression. Use a | to indicate OR.
For example:
human|mouse
will find text with either 'human' OR 'mouse' in the text</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 3)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.textsearch</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-casesensitive</td>
<td>Do a case-sensitive search</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-html</td>
<td>Format output as an HTML table</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-only</td>
<td>This is a way of shortening the command line if you only want a few things to be displayed. Instead of specifying:
'-nohead -noname -nousa -noacc -nodesc'
to get only the name output, you can specify
'-only -name'</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-heading</td>
<td>Display column headings</td>
<td>Boolean value Yes/No</td>
<td>@(!$(only))</td>
</tr>

<tr>
<td>-usa</td>
<td>Display the USA of the sequence</td>
<td>Boolean value Yes/No</td>
<td>@(!$(only))</td>
</tr>

<tr>
<td>-accession</td>
<td>Display 'accession' column</td>
<td>Boolean value Yes/No</td>
<td>@(!$(only))</td>
</tr>

<tr>
<td>-name</td>
<td>Display 'name' column</td>
<td>Boolean value Yes/No</td>
<td>@(!$(only))</td>
</tr>

<tr>
<td>-description</td>
<td>Display 'description' column</td>
<td>Boolean value Yes/No</td>
<td>@(!$(only))</td>
</tr>

</table>



<H2>
    Input file format
</H2>


<b>textsearch</b> reads one or more normal sequence USAs.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tsw:*' is a sequence entry in the example protein database 'tsw'
<p>

<a name="input.3"></a>
<h3>Input files for usage example 3</h3>

'tembl:*' is a sequence entry in the example nucleic acid database 'tembl'
<p>


<H2>
    Output file format
</H2>



<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: 100k_rat.textsearch</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
# Search for: lactose
tsw-id:LACI_ECOLI LACI_ECOLI    P03023	LACTOSE OPERON REPRESSOR.
tsw-id:LACY_ECOLI LACY_ECOLI    P02920	LACTOSE PERMEASE (LACTOSE-PROTON SYMPORT).
</pre>
</td></tr></table><p>

<a name="output.2"></a>
<h3>Output files for usage example 2</h3>
<p><h3>File: laci_ecoli.textsearch</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
# Search for: lactose | permease
tsw-id:LACI_ECOLI LACI_ECOLI    P03023	LACTOSE OPERON REPRESSOR.
tsw-id:LACY_ECOLI LACY_ECOLI    P02920	LACTOSE PERMEASE (LACTOSE-PROTON SYMPORT).
</pre>
</td></tr></table><p>

<a name="output.3"></a>
<h3>Output files for usage example 3</h3>
<p><h3>File: embl.lacz.html</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr>Search for: lacz</tr>
<tr><td>tembl-id:ECLAC</td><td>ECLAC</td><td>J01636</td>	<td>E.coli lactose operon with lacI, lacZ, lacY and lacA genes.</td></tr>
<tr><td>tembl-id:ECLACZ</td><td>ECLACZ</td><td>V00296</td>	<td>E. coli gene lacZ coding for beta-galactosidase (EC 3.2.1.23).</td></tr>
</table>
</td></tr></table><p>



<p>

The first column in the name or ID of each sequence.  The remaining
text is the description line of the sequence.

<p>

When the -html qualifier is specified, then the output will be wrapped
in HTML tags, ready for inclusion in a Web page.  Note that tags such
as &lt;HTML&gt;, &lt;BODY&gt;, &lt;/BODY&gt; and &lt;/HTML&gt; are not
output by this program as the table of databases is expected to form
only part of the contents of a web page - the rest of the web page
must be supplier by the user.

<p>

The lines of out information are guaranteed not to have trailing
white-space at the end. So if '-nodesc' is used, there will not be any
whitespace after the ID name.

<H2>
    Data files
</H2>


None.

<H2>
    Notes
</H2>


This is a rather slow way to search for text in databases.  If you are
searching for text in public databases, you should consider using
either Entrez (<a
href="http://www.ncbi.nlm.nih.gov/Entrez/">http://www.ncbi.nlm.nih.gov/Entrez/</a>)
or SRS (<a
href="http://srs.rfcgr.mrc.ac.uk/">http://srs.rfcgr.mrc.ac.uk/</a> or <a
href="http://www.sanger.ac.uk/srs6/">http://www.sanger.ac.uk/srs6/</a>
etc.)
	
<H2>
    References
</H2>

None.


<H2>
    Warnings
</H2>

None.


<H2>
    Diagnostic Error Messages
</H2>

None.


<H2>
    Exit status
</H2>


It always exits with status 0

<H2>
    Known bugs
</H2>


None noted.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="abiview.html">abiview</a></td><td>Reads ABI file and display the trace</td></tr>
<tr><td><a href="cirdna.html">cirdna</a></td><td>Draws circular maps of DNA constructs</td></tr>
<tr><td><a href="infoalign.html">infoalign</a></td><td>Information on a multiple sequence alignment</td></tr>
<tr><td><a href="infoseq.html">infoseq</a></td><td>Displays some simple information about sequences</td></tr>
<tr><td><a href="lindna.html">lindna</a></td><td>Draws linear maps of DNA constructs</td></tr>
<tr><td><a href="pepnet.html">pepnet</a></td><td>Displays proteins as a helical net</td></tr>
<tr><td><a href="pepwheel.html">pepwheel</a></td><td>Shows protein sequences as helices</td></tr>
<tr><td><a href="prettyplot.html">prettyplot</a></td><td>Displays aligned sequences, with colouring and boxing</td></tr>
<tr><td><a href="prettyseq.html">prettyseq</a></td><td>Output sequence with translated ranges</td></tr>
<tr><td><a href="remap.html">remap</a></td><td>Display a sequence with restriction cut sites, translation etc</td></tr>
<tr><td><a href="seealso.html">seealso</a></td><td>Finds programs sharing group names</td></tr>
<tr><td><a href="showalign.html">showalign</a></td><td>Displays a multiple sequence alignment</td></tr>
<tr><td><a href="showdb.html">showdb</a></td><td>Displays information on the currently available databases</td></tr>
<tr><td><a href="showfeat.html">showfeat</a></td><td>Show features of a sequence</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="sixpack.html">sixpack</a></td><td>Display a DNA sequence with 6-frame translation and ORFs</td></tr>
<tr><td><a href="tfm.html">tfm</a></td><td>Displays a program's help documentation manual</td></tr>
<tr><td><a href="whichdb.html">whichdb</a></td><td>Search all databases for an entry</td></tr>
<tr><td><a href="wossname.html">wossname</a></td><td>Finds programs by keywords in their one-line documentation</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


Gary Williams (gwilliam&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK





<H2>
    History
</H2>

<!--
<p>
[an error occurred while processing this directive]
-->

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

