<HTML>

<HEAD>
  <TITLE>
  EMBOSS: tfextract
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: tfextract
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program tfextract
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Extract data from TRANSFAC


<H2>
    Description
</H2>

The
<A HREF="http://transfac.gbf.de/TRANSFAC/">TRANSFAC Database</A>
is a database of eukaryotic cis-acting regulatory DNA elements and
trans-acting factors.  It covers the whole range from yeast to human. 
<p>

TRANSFAC started 1988 with a printed compilation (Nucleic Acids Res. 
16: 1879-1902, 1988) and was transferred into computer-readable format
in 1990 (BioTechForum - Advances in Molecular Genetics (J.  Collins,
A.J.  Driesel, eds.) 4:95-108, 1991).  The basic structures of Table 1
and 2 of the compilation were taken as the core of the emergent
database.  The aim of the early compilation as well as of the TRANSFAC
database is
<p>

<ol>

<li> to guide through a meanwhile overwhelming amount of data in a field
which is connected to nearly all areas of modern molecular biology;

<li> to map the regulatory sites in the individual genes and,
ultimately, in the genome(s) as a whole;

<li> to develop a tool for the identification of regulatory elements in
newly unravelled genomic sequences;

<li> to provide a basis for a more comprehensive understanding of how
the genome governs transcriptional control. 

</ol>
<p>


The program <b>tfextract</b> extracts data from the TRANSFAC database
file <b>site.dat</b>.  This file contains information on individual
(putatively) regulatory protein binding sites.  About half of these
refer to sites within eukaryotic genes. Just under half of them
resulted from mutagenesis studies, in vitro selection procedures
starting from random oligonucleotide mixtures or from specific
theoretical considerations.  And finally, there are about 5% with
consensus binding sequences given in the IUPAC code, many of them being
taken from the compilation of Faisst and Meyer (Nucleic Acids Res. 
20:3-26, 1992).  A number of consensi have been generated by the
TRANSFAC team, generally derived from the profiles stored in the MATRIX
table. 
<p>

The data is split up by taxonomic groups:
<p>
<ul>
<li>Fungi
<li>Insects
<li>Plants
<li>Vertebrates
<li>Other
</ul>
<p>
and placed in individual files:
<p>
<ul>
<li>tffungi
<li>tfinsect
<li>tfplant
<li>tfvertebrate
<li>tfother
</ul>
<p>

These files are stored in the EMBOSS data directory, see <b>Data
Files</b> below. 

<H2>
    Usage
</H2>


Here is a sample session with tfextract.

<pre>

% tfextract
Extract data from TRANSFAC
Full pathname of transfac SITE.DAT: /data/transfac/site.dat

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-inf]               infile     Full pathname of transfac SITE.DAT

   Optional qualifiers: (none)
   Advanced qualifiers: (none)

</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-inf]<br>(Parameter 1)</td>
<td>Full pathname of transfac SITE.DAT</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>

It reads in the TRANSFAC file <b>site.dat</b> distributed from the TRANSFAC
site:
<p>
<A HREF="http://transfac.gbf.de/cgi-bin/download/download.pl">http://transfac.gbf.de/cgi-bin/download/download.pl</A>


<H2>
    Output file format
</H2>


The output from tfextract is a set of files in the emboss/data
directory containing reformatted sites from the transfac database.

<p>
These files are used by the tfscan program to search for TRANSFAC sites in
sequences.


<pre>

% ls -1s emboss/data/tf*
 18 emboss/data/tffungi
 17 emboss/data/tfinsect
 56 emboss/data/tfother
  4 emboss/data/tfplant
112 emboss/data/tfvertebrate

</pre>

<H2>
    Data files
</H2>

<!--
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.
        If you wish to include the standard description of what data
        files are and how to use embossdata to inspect and retrieve
        them, use:
   -->
<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>




<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

<ul>
<li>Nucleic Acids Res. 16: 1879-1902, 1988

<li>BioTechForum - Advances in Molecular Genetics (J.  Collins,A.J.  Driesel, eds.) 4:95-108, 1991

<li>Nucleic Acids Res. 20:3-26, 1992
</ul>

<H2>
    Warnings
</H2>

None.



<H2>
    Diagnostic Error Messages
</H2>

None.


<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>


None.

<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="domainer.html">domainer</a></td><td>Build domain coordinate files</td></tr>
<tr><td><a href="nrscope.html">nrscope</a></td><td>Converts redundant EMBL-format SCOP file to non-redundant one</td></tr>
<tr><td><a href="printsextract.html">printsextract</a></td><td>Extract data from PRINTS</td></tr>
<tr><td><a href="prosextract.html">prosextract</a></td><td>Builds the PROSITE motif database for patmatmotifs to search</td></tr>
<tr><td><a href="rebaseextract.html">rebaseextract</a></td><td>Extract data from REBASE</td></tr>
<tr><td><a href="scope.html">scope</a></td><td>Convert raw scop classification file to embl-like format</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>


Written Summer 2000 - Alan Bleasby.

<H2>
    Target users
</H2>


This program is intended to be used by administrators responsible
for software and database installation and maintenance. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

