<HTML>

<HEAD>
  <TITLE>
  EMBOSS: tfextract
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">







<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

tfextract
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>






<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Extract data from TRANSFAC


<H2>
    Description
</H2>

The
<A HREF="http://transfac.gbf.de/TRANSFAC/">TRANSFAC Database</A>
is a database of eukaryotic cis-acting regulatory DNA elements and
trans-acting factors.  It covers the whole range from yeast to human. 
<p>

TRANSFAC started 1988 with a printed compilation (Nucleic Acids Res. 
16: 1879-1902, 1988) and was transferred into computer-readable format
in 1990 (BioTechForum - Advances in Molecular Genetics (J.  Collins,
A.J.  Driesel, eds.) 4:95-108, 1991).  The basic structures of Table 1
and 2 of the compilation were taken as the core of the emergent
database.  The aim of the early compilation as well as of the TRANSFAC
database is
<p>

<ol>

<li> to guide through a meanwhile overwhelming amount of data in a field
which is connected to nearly all areas of modern molecular biology;

<li> to map the regulatory sites in the individual genes and,
ultimately, in the genome(s) as a whole;

<li> to develop a tool for the identification of regulatory elements in
newly unravelled genomic sequences;

<li> to provide a basis for a more comprehensive understanding of how
the genome governs transcriptional control. 

</ol>
<p>


The program <b>tfextract</b> extracts data from the TRANSFAC database
file <b>site.dat</b>.  This file contains information on individual
(putatively) regulatory protein binding sites.  About half of these
refer to sites within eukaryotic genes. Just under half of them
resulted from mutagenesis studies, in vitro selection procedures
starting from random oligonucleotide mixtures or from specific
theoretical considerations.  And finally, there are about 5% with
consensus binding sequences given in the IUPAC code, many of them being
taken from the compilation of Faisst and Meyer (Nucleic Acids Res. 
20:3-26, 1992).  A number of consensi have been generated by the
TRANSFAC team, generally derived from the profiles stored in the MATRIX
table. 
<p>

The data is split up by taxonomic groups:
<p>
<ul>
<li>Fungi
<li>Insects
<li>Plants
<li>Vertebrates
<li>Other
</ul>
<p>
and placed in individual files:
<p>
<ul>
<li>tffungi
<li>tfinsect
<li>tfplant
<li>tfvertebrate
<li>tfother
</ul>
<p>

These files are stored in the EMBOSS data directory, see <b>Data
Files</b> below. 

<H2>
    Usage
</H2>


<b>Here is a sample session with tfextract</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>tfextract </b>
Extract data from TRANSFAC
Full pathname of transfac SITE.DAT: <b>../../data/site.dat</b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-inf]               infile     Full pathname of transfac SITE.DAT

   Optional qualifiers: (none)
   Advanced qualifiers: (none)
   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-inf]<br>(Parameter 1)</td>
<td>Full pathname of transfac SITE.DAT</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>

It reads in the TRANSFAC file <b>site.dat</b> distributed from the TRANSFAC
site:
<p>
<A HREF="http://transfac.gbf.de/cgi-bin/download/download.pl">http://transfac.gbf.de/cgi-bin/download/download.pl</A>

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: ../../data/site.dat</h3>
<table width="90%"><tr><td bgcolor="#FFDDFF">
<pre>
AC  R00077
XX
ID  HS$ALBU_01
XX
DT  20.06.90 (created); ewi.
DT  24.08.95 (updated); hiwi.
XX
TY  D
XX
DE  albumin; Gene: G000188.
XX
SQ  tGGTTAGtaattactaa.
XX
SF  -363 
ST  -338 
XX
BF  T00368; HNF-1A;Quality: 1; Species: human, Homo sapiens.
BF  T00369; HNF-1;Quality: 1; Species: rat, Rattus norvegicus.
BF  T01950; HNF-1B;Quality: 1; Species: human, Homo sapiens.
BF  T01951; HNF-1C;Quality: 1; Species: human, Homo sapiens.
XX
OS  human, Homo sapiens
OC  eukaryota; animalia; metazoa; chordata; vertebrata; 
OC  tetrapoda; mammalia; eutheria; primates
XX
SO  0103; Hep3B
SO  0289; rl
XX
MM  gel retardation
MM  direct gel shift
MM  DNase I footprinting
MM  gel shift competition
MM  affinity chromatography
MM  methylation interference
XX
DR  EMBL: M13075; HSALBEX1(695:711).
XX
RN  [1]
RA  Frain M., Swart G., Monaci P., Nicosia A., Staempfli 
RA  S., Frank R., Cortese R.
RT  The liver-specific transcription factor LF-B1 contains 
RT  a highly diverged homeobox DNA binding domain
RL  Cell 59:145-157 (1989).
RN  [2]
RA  Frain M., Hardon E., Ciliberto G., Sala-Trepat J. M.
RT  Binding of a liver-specific factor to the human albumin 
RT  gene promoter and enhancer
RL  Mol. Cell. Biol. 10:991-999 (1990).
XX
//


<font color=red>  [Part of this file has been deleted for brevity]</font>

DR  EMBL: U11854; MM11854(1931:1941).
XX
RN  [1]
RA  Feinman R., Qiu W. Q., Pearse R. N., Nikolajczyk B. 
RA  S., Sen R., Sheffery M., Ravetch J. V.
RT  PU.1 and an HLH family member contribute to the myeloid-specific 
RT  transcription of the FcgammaRIIIA promoter
RL  EMBO J. 13:3852-3860 (1994).
XX
//
AC  R04413
XX
ID  MOUSE$FCGR3A_02
XX
DT  14.05.97 (created); ewi.
DT  14.05.97 (updated); ewi.
XX
TY  D
XX
DE  FcgammaRIIIA (low-affinity Fc receptor IIIA for IgG); Gene: G001014.
XX
SQ  TTCCTC.
XX
EL  MRR
XX
SF  -48 
ST  -43 
XX
BF  T00702; PU.1;Quality: 3; Species: mouse, Mus musculus.
XX
OS  mouse, Mus musculus
OC  eukaryota; animalia; metazoa; chordata; vertebrata; 
OC  tetrapoda; mammalia; eutheria; rodentia; myomorpha; muridae; murinae
XX
SO  0495; A20
SO  0848; RAW264.7
XX
MM  direct gel shift
MM  methylation interference
MM  supershift (antibody binding)
XX
DR  EMBL: U11854; MM11854(1971:1976).
XX
RN  [1]
RA  Feinman R., Qiu W. Q., Pearse R. N., Nikolajczyk B. 
RA  S., Sen R., Sheffery M., Ravetch J. V.
RT  PU.1 and an HLH family member contribute to the myeloid-specific 
RT  transcription of the FcgammaRIIIA promoter
RL  EMBO J. 13:3852-3860 (1994).
XX
//
</pre>
</td></tr></table><p>



<H2>
    Output file format
</H2>



<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: tffungi</h3>
<table width="90%"><tr><td bgcolor="#DDFFDD">
<pre>
</pre>
</td></tr></table><p>
<p><h3>File: tfinsect</h3>
<table width="90%"><tr><td bgcolor="#DDFFDD">
<pre>
</pre>
</td></tr></table><p>
<p><h3>File: tfother</h3>
<table width="90%"><tr><td bgcolor="#DDFFDD">
<pre>
</pre>
</td></tr></table><p>
<p><h3>File: tfplant</h3>
<table width="90%"><tr><td bgcolor="#DDFFDD">
<pre>
Y$ADH1_02            ACAATATGGACTTCCTCTTTTCTGG R04140
</pre>
</td></tr></table><p>
<p><h3>File: tfvertebrate</h3>
<table width="90%"><tr><td bgcolor="#DDFFDD">
<pre>
HS$ALBU_01           tGGTTAGtaattactaa R00077
HS$ALBU_02           TTGGCA R00078
HS$ALBU_03           TGGCA R00079
HS$ALBU_04           TTAATAAT R00080
HS$ALBU_05           TCTAGTTAATAATCTACAAT R00081
MOUSE$FCGR3A_01      GTCTGCTGACC R04412
MOUSE$FCGR3A_02      TTCCTC R04413
</pre>
</td></tr></table><p>


<p>

The output from tfextract is a set of files in the emboss/data
directory containing reformatted sites from the transfac database.

<p>

These files are used by the tfscan program to search for TRANSFAC sites in
sequences.



<H2>
    Data files
</H2>

<!--
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.
        If you wish to include the standard description of what data
        files are and how to use embossdata to inspect and retrieve
        them, use:
   -->
<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>




<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

<ul>
<li>Nucleic Acids Res. 16: 1879-1902, 1988

<li>BioTechForum - Advances in Molecular Genetics (J.  Collins,A.J.  Driesel, eds.) 4:95-108, 1991

<li>Nucleic Acids Res. 20:3-26, 1992
</ul>

<H2>
    Warnings
</H2>

None.



<H2>
    Diagnostic Error Messages
</H2>

None.


<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>


None.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="aaindexextract.html">aaindexextract</a></td><td>Extract data from AAINDEX</td></tr>
<tr><td><a href="cutgextract.html">cutgextract</a></td><td>Extract data from CUTG</td></tr>
<tr><td><a href="printsextract.html">printsextract</a></td><td>Extract data from PRINTS</td></tr>
<tr><td><a href="prosextract.html">prosextract</a></td><td>Builds the PROSITE motif database for patmatmotifs to search</td></tr>
<tr><td><a href="rebaseextract.html">rebaseextract</a></td><td>Extract data from REBASE</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>


Written Summer 2000 - Alan Bleasby.

<H2>
    Target users
</H2>


This program is intended to be used by administrators responsible
for software and database installation and maintenance. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

