<HTML>

<HEAD>
  <TITLE>
  EMBOSS: tfscan
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
tfscan
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
Identify transcription factor binding sites in DNA sequences

<H2>
    Description
</H2>

<p><b>tfscan</b> scans one or more DNA sequences for transcription
factor binding sites from the TRANSFAC database.  The taxonomic group
(Fungi, Insects, ,Plants, Vertebrates or Other) is specified.  Matches
are searched for using fast sequence word-matching, optionally
allowing mismatches.  Because the binding sites are so small, there
will be many spurious (false positive) matches. Optionally, the
minimum length of a match to be reported may be specified.</p>

<p>An output file is written with information on the matches,
including sequence ID and accession number, the start and end
positions of the match in an input sequence and the sequence of the
region where a match has been found. Binding factor information, where
available, is given at the end of the matches for each matching
entry.</p>





<H2>
    Usage
</H2>
<b>Here is a sample session with tfscan</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>tfscan </b>
Identify transcription factor binding sites in DNA sequences
Input nucleotide sequence(s): <b>tembl:k00650</b>
Transcription Factor Class
         F : fungi
         I : insect
         P : plant
         V : vertebrate
         O : other
         C : Custom
Select class [V]: <b>v</b>
Number of mismatches [0]: <b></b>
Output report [k00650.tfscan]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
Identify transcription factor binding sites in DNA sequences
Version: EMBOSS:6.3.0

   Standard (Mandatory) qualifiers (* if not always prompted):
  [-sequence]          seqall     Nucleotide sequence(s) filename and optional
                                  format, or reference (input USA)
   -menu               menu       [V] Select class (Values: F (fungi); I
                                  (insect); P (plant); V (vertebrate); O
                                  (other); C (Custom))
*  -custom             datafile   Transfac database data file (optional)
   -mismatch           integer    [0] Number of mismatches (Integer 0 or more)
  [-outfile]           report     [*.tfscan] Output report file name

   Additional (Optional) qualifiers:
   -minlength          integer    [1] Display matches equal to or above this
                                  length (Integer 1 or more)

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1            integer    Start of each sequence to be used
   -send1              integer    End of each sequence to be used
   -sreverse1          boolean    Reverse (if DNA)
   -sask1              boolean    Ask for begin/end/reverse
   -snucleotide1       boolean    Sequence is nucleotide
   -sprotein1          boolean    Sequence is protein
   -slower1            boolean    Make lower case
   -supper1            boolean    Make upper case
   -sformat1           string     Input sequence format
   -sdbname1           string     Database name
   -sid1               string     Entryname
   -ufo1               string     UFO features
   -fformat1           string     Features format
   -fopenfile1         string     Features file name

   "-outfile" associated qualifiers
   -rformat2           string     Report format
   -rname2             string     Base file name
   -rextension2        string     File name extension
   -rdirectory2        string     Output directory
   -raccshow2          boolean    Show accession number in the report
   -rdesshow2          boolean    Show description in the report
   -rscoreshow2        boolean    Show the score in the report
   -rstrandshow2       boolean    Show the nucleotide strand in the report
   -rusashow2          boolean    Show the full USA in the report
   -rmaxall2           integer    Maximum total hits to report
   -rmaxseq2           integer    Maximum hits to report for one sequence

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-sequence]<br>(Parameter 1)</td>
<td>seqall</td>
<td>Nucleotide sequence(s) filename and optional format, or reference (input USA)</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-menu</td>
<td>list</td>
<td>Select class</td>
<td><table><tr><td>F</td> <td><i>(fungi)</i></td></tr><tr><td>I</td> <td><i>(insect)</i></td></tr><tr><td>P</td> <td><i>(plant)</i></td></tr><tr><td>V</td> <td><i>(vertebrate)</i></td></tr><tr><td>O</td> <td><i>(other)</i></td></tr><tr><td>C</td> <td><i>(Custom)</i></td></tr></table></td>
<td>V</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-custom</td>
<td>datafile</td>
<td>Transfac database data file (optional)</td>
<td>Data file</td>
<td><i>File in the data file path</i></td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-mismatch</td>
<td>integer</td>
<td>Number of mismatches</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-outfile]<br>(Parameter 2)</td>
<td>report</td>
<td>Output report file name</td>
<td>Report output file</td>
<td><i>&lt;*&gt;</i>.tfscan</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>-minlength</td>
<td>integer</td>
<td>Display matches equal to or above this length</td>
<td>Integer 1 or more</td>
<td>1</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-sequence" associated seqall qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sbegin1<br>-sbegin_sequence</td>
<td>integer</td>
<td>Start of each sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -send1<br>-send_sequence</td>
<td>integer</td>
<td>End of each sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sreverse1<br>-sreverse_sequence</td>
<td>boolean</td>
<td>Reverse (if DNA)</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sask1<br>-sask_sequence</td>
<td>boolean</td>
<td>Ask for begin/end/reverse</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -snucleotide1<br>-snucleotide_sequence</td>
<td>boolean</td>
<td>Sequence is nucleotide</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sprotein1<br>-sprotein_sequence</td>
<td>boolean</td>
<td>Sequence is protein</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -slower1<br>-slower_sequence</td>
<td>boolean</td>
<td>Make lower case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -supper1<br>-supper_sequence</td>
<td>boolean</td>
<td>Make upper case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sformat1<br>-sformat_sequence</td>
<td>string</td>
<td>Input sequence format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sdbname1<br>-sdbname_sequence</td>
<td>string</td>
<td>Database name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sid1<br>-sid_sequence</td>
<td>string</td>
<td>Entryname</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ufo1<br>-ufo_sequence</td>
<td>string</td>
<td>UFO features</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fformat1<br>-fformat_sequence</td>
<td>string</td>
<td>Features format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fopenfile1<br>-fopenfile_sequence</td>
<td>string</td>
<td>Features file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-outfile" associated report qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rformat2<br>-rformat_outfile</td>
<td>string</td>
<td>Report format</td>
<td>Any string</td>
<td>seqtable</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rname2<br>-rname_outfile</td>
<td>string</td>
<td>Base file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rextension2<br>-rextension_outfile</td>
<td>string</td>
<td>File name extension</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rdirectory2<br>-rdirectory_outfile</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -raccshow2<br>-raccshow_outfile</td>
<td>boolean</td>
<td>Show accession number in the report</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rdesshow2<br>-rdesshow_outfile</td>
<td>boolean</td>
<td>Show description in the report</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rscoreshow2<br>-rscoreshow_outfile</td>
<td>boolean</td>
<td>Show the score in the report</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rstrandshow2<br>-rstrandshow_outfile</td>
<td>boolean</td>
<td>Show the nucleotide strand in the report</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rusashow2<br>-rusashow_outfile</td>
<td>boolean</td>
<td>Show the full USA in the report</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rmaxall2<br>-rmaxall_outfile</td>
<td>integer</td>
<td>Maximum total hits to report</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rmaxseq2<br>-rmaxseq_outfile</td>
<td>integer</td>
<td>Maximum hits to report for one sequence</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>


<H2>
    Input file format
</H2>

<b>tfscan</b> reads normal nucleic acid sequence USAs.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:k00650' is a sequence entry in the example nucleic acid database 'tembl'
<p>
<p><h3>Database entry: tembl:k00650</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   K00650; SV 1; linear; genomic DNA; STD; HUM; 6210 BP.
XX
AC   K00650; M16287;
XX
DT   26-JUL-1991 (Rel. 28, Created)
DT   14-NOV-2006 (Rel. 89, Last updated, Version 4)
XX
DE   Human fos proto-oncogene (c-fos), complete cds.
XX
KW   c-myc proto-oncogene; fos oncogene; proto-oncogene.
XX
OS   Homo sapiens (human)
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia;
OC   Eutheria; Euarchontoglires; Primates; Haplorrhini; Catarrhini; Hominidae;
OC   Homo.
XX
RN   [1]
RP   1-4165
RX   DOI; 10.1073/pnas.80.11.3183.
RX   PUBMED; 6574479.
RA   van Straaten F., Muller R., Curran T., Van Beveren C., Verma I.M.;
RT   "Complete nucleotide sequence of a human c-onc gene: deduced amino acid
RT   sequence of the human c-fos protein";
RL   Proc. Natl. Acad. Sci. U.S.A. 80(11):3183-3187(1983).
XX
RN   [2]
RX   DOI; 10.1016/0092-8674(85)90285-5.
RX   PUBMED; 2414012.
RA   Treisman R.;
RT   "Transient accumulation of c-fos RNA following serum stimulation requires a
RT   conserved 5' element and c-fos 3' sequences";
RL   Cell 42(3):889-902(1985).
XX
RN   [3]
RP   4166-6210
RX   PUBMED; 3555978.
RA   Verma I.M., Deschamps J., Van Beveren C., Sassone-Corsi P.;
RT   "Human fos gene";
RL   Cold Spring Harb. Symp. Quant. Biol. 51:0-0(0).
XX
DR   EPD; EP11145; HS_FOS.
XX
CC   [2]  sites; promoter region.
CC   C-fos is the human cellular homolog of the v-fos oncogene of
CC   Finkel-Biskis-Jinkins murine osteosarcoma virus (FBJ-MuSV).  [2] It
CC   was found that both human and murine c-fos genes contained an
CC   enhancer-like element in their 5' noncoding regions that was
CC   necessary for increased transcription following serum activation.
CC   The FBJ-MuSV v-fos oncogene contains a deletion relative to murine
CC   and human c-fos proto-oncogenes that causes complete divergence of


<font color=red>  [Part of this file has been deleted for brevity]</font>

     ccagctgtgc agctgcccac cgcaagggca gcagcagcaa tgagccttcc tctgactcgc      3300
     tcagctcacc cacgctgctg gccctgtgag ggggcaggga aggggaggca gccggcaccc      3360
     acaagtgcca ctgcccgagc tggtgcatta cagagaggag aaacacatct tccctagagg      3420
     gttcctgtag acctagggag gaccttatct gtgcgtgaaa cacaccaggc tgtgggcctc      3480
     aaggacttga aagcatccat gtgtggactc aagtccttac ctcttccgga gatgtagcaa      3540
     aacgcatgga gtgtgtattg ttcccagtga cacttcagag agctggtagt tagtagcatg      3600
     ttgagccagg cctgggtctg tgtctctttt ctctttctcc ttagtcttct catagcatta      3660
     actaatctat tgggttcatt attggaatta acctggtgct ggatattttc aaattgtatc      3720
     tagtgcagct gattttaaca ataactactg tgttcctggc aatagtgtgt tctgattaga      3780
     aatgaccaat attatactaa gaaaagatac gactttattt tctggtagat agaaataaat      3840
     agctatatcc atgtactgta gtttttcttc aacatcaatg ttcattgtaa tgttactgat      3900
     catgcattgt tgaggtggtc tgaatgttct gacattaaca gttttccatg aaaacgtttt      3960
     attgtgtttt taatttattt attaagatgg attctcagat atttatattt ttattttatt      4020
     tttttctacc ttgaggtctt ttgacatgtg gaaagtgaat ttgaatgaaa aatttaagca      4080
     ttgtttgctt attgttccaa gacattgtca ataaaagcat ttaagttgaa tgcgaccaac      4140
     cttgtgctct tttcattctg gaagtcttgt aagtttctga aaggtattat tggagaccag      4200
     tttgtcaaga agggtagctg ctggaggggg acacaccctc tgtctgatcc cttatcaaag      4260
     aggacaagga aactatagag ctgattttag aatattttac aaatacatgc cttccattgg      4320
     aatgctaaga ttttctactg cttctgggga cgggaaaccg ctgtgtaaca gcttttgtgg      4380
     gaatacattt tttctgtttc agtactcgca gggggaaata tttaaatttt gttgtgctaa      4440
     tattaaattc agatgttttg atcttaaagg aaccctttaa gcaaacagaa cctagctttg      4500
     tacagactat tttaactttt tattctcaca aaatcacgtg gagggttatt ctacttcaaa      4560
     gatgagcaaa ttgaagaatg gttagaataa acaactttct tgatattccg ttatcggcat      4620
     tagaatcttc ctgctcgtta tcgtatccag caggctgaac tgcctcttga tacttggtta      4680
     aaaaaaattt tcaggccggg cgcggtggcc catgcctgta atcctagcac tttgggaggc      4740
     cgaggcaggc ggatcacctg aggtcgggag ttcgagacca gcctgaccaa catggagaaa      4800
     ccccgtcttt actaaaaata caaaattagc ctggtgtggt ggtgcatgcc tgtaatccta      4860
     gctacttgag aggctgagac aggaaaatca cttgaactcg ggaggcggat gttgcagcga      4920
     actgagattg cgccattgca ctccagcctg ggcaacaaga ttgaaactct gtttaaaaaa      4980
     aaaagttttc actaatgtgt acattttttt gtactctttt attctcgaaa gggaaggagg      5040
     gctattgccc tatcccttat taataaatgc attgtggttt ctggtttctc taataccata      5100
     tgcccttcat tcagtttata gtgggcggaa gtgggggaga aaaagttgct cagaaatcaa      5160
     aagatatctc aaacagcaca aataatggct gatcgttctg caaacaaaaa gttacataat      5220
     agctcaagaa ggagaagtca acatgactct gaacaagctt taacttagaa actttatcat      5280
     cttaaggaag aacgtgacct ttgtccagga cgtctctggt aatggggcac ttacacacac      5340
     atgcacacgt acaaaccaca gggaaaggag accgcccttc tgcctctgct cgcgagtatc      5400
     acgcaggcac catgcactat gttttcacac acactgggtg gaagaagagc ttcagcgcca      5460
     gtcttctaat gctttggtga taatgaaaat cactgggtgc ttatggggtg tcatattcaa      5520
     tcgagttaaa agttttaatt caaaatgaca gttttactga ggttgatgtt ctcgtctatg      5580
     atatctctgc ccctcccata aaaatggaca tttaaaagca acttaccgct ctttagatca      5640
     ctcctatatc acacaccact tggggtgctg tttctgctag acttgtgatg acagtggcct      5700
     taggatccct gtttgctgtt caaagggcaa atattttata gcctttaaat atacctaaac      5760
     taaatacaga attaatataa ctaacaaaca cctggtctga aataacaagg tgatctaccc      5820
     tggaaggaac ccagctggtg ggccaggagc ggtggctcac acctgtaatt ccagcacttt      5880
     gggaggctga gacaggagga tcactggagt ccaggagttt gagaccagcc tgggcaacat      5940
     ggcaaaaccc agtgtgcttc tgttgtccca gctacactac tcaggaggct gaggcaggag      6000
     tatgacttga gcctgggagg gggaggttgc agagaactga tattgcacca ccactgcact      6060
     ccagcctggg tgacacagca aaaccctatc tcaaaaaaaa aaaaaaaaaa aaggaaccca      6120
     gctggttcct gtaggtgtgc aataataaca accagaggaa gaaaaggaag acgatttccc      6180
     agatgaagaa gggcagctgg accttcggac                                       6210
//
</pre>
</td></tr></table><p>

<H2>
    Output file format
</H2>



<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: k00650.tfscan</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
########################################
# Program: tfscan
# Rundate: Fri 15 Jan 2010 12:00:00
# Commandline: tfscan
#    -sequence tembl:k00650
#    -menu v
# Report_format: seqtable
# Report_file: k00650.tfscan
########################################

#=======================================
#
# Sequence: K00650     from: 1   to: 6210
# HitCount: 9
#=======================================

  Start     End  Strand Accession Factor                                                      Sequence
   3287    3292       + R04413    T00702; PU.1;Quality: 3; Species: mouse, Mus musculus.      ttcctc
   5940    5944       + R00079    T00599; NF-1/L;Quality: 6; Species: rat, Rattus norvegicus. tggca
   3757    3761       + R00079    T00599; NF-1/L;Quality: 6; Species: rat, Rattus norvegicus. tggca
   2776    2780       + R00079    T00599; NF-1/L;Quality: 6; Species: rat, Rattus norvegicus. tggca
   2418    2422       + R00079    T00599; NF-1/L;Quality: 6; Species: rat, Rattus norvegicus. tggca
   2010    2014       + R00079    T00599; NF-1/L;Quality: 6; Species: rat, Rattus norvegicus. tggca
   1676    1680       + R00079    T00599; NF-1/L;Quality: 6; Species: rat, Rattus norvegicus. tggca
   1356    1360       + R00079    T00599; NF-1/L;Quality: 6; Species: rat, Rattus norvegicus. tggca
   2009    2014       + R00078    T00599; NF-1/L;Quality: 6; Species: rat, Rattus norvegicus. ttggca

#---------------------------------------
#---------------------------------------

#---------------------------------------
# Reported_sequences: 1
# Reported_hitcount: 9
#---------------------------------------
</pre>
</td></tr></table><p>

<p>

The output consists of a title line then 5 columns separated by whitespace.

<p>

The first column is the identifier of the entry.

<p>

The second column is the Accession Number of the entry.

<p>

The third and fourth columns are the start and end positions of the
match in your input sequence.

<p>

The fifth column is the sequence of the region where a match has been found.

<p>

Binding factor information, where available, is given at the end
of the matches for each matching entry.



<H2>
    Data files
</H2>

<b>tfscan</b> reads the TRANSFAC SITE data held in the EMBOSS data files:
<p>
<ul>
<li>tffungi 
<li>tfinsect 
<li>tfplant 
<li>tfvertebrate 
<li>tfother
</ul>
<p>

Your EMBOSS administrator will have to run the EMBOSS program <b>tfextract</b>
in order to set these files up from the TRANSFAC distribution files.
<p>

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>


<H2>
    Notes
</H2>

<p>The TRANSFAC Database is a commercial database of eukaryotic cis-acting regulatory DNA elements and trans-acting factors. It covers the whole range from yeast to human.  The <tt>site.dat</tt> data file from TRANSFAC contains information on individual (putatively) regulatory protein binding sites. It has been divided into the following taxonomic groups.</p>
<ul>
<li>Fungi
<li>Insects
<li>Plants
<li>Vertebrates
<li>Other
</ul>
<p>An old public domain version of  TRANSFAC is available at: <a href="ftp://ftp.ebi.ac.uk/pub/databases/transfac/transfac32.tar.Z">ftp://ftp.ebi.ac.uk/pub/databases/transfac/transfac32.tar.Z</a></p>




<H2>
    References
</H2>

<ul>

<li>Nucleic Acids Res. 16: 1879-1902, 1988

<li>BioTechForum - Advances in Molecular Genetics
(J. Collins,A.J. Driesel, eds.) 4:95-108, 1991

<li>Nucleic Acids Res. 20:3-26, 1992 

</ul>


<H2>
    Warnings
</H2>

Your EMBOSS administrator will have to run the EMBOSS program <b>tfextract</b>
in order to set up the data files from the TRANSFAC distribution files.
<p>


<H2>
    Diagnostic Error Messages
</H2>

"EMBOSS An error in tfscan.c at line 82:
<br>
Either EMBOSS_DATA undefined or TFEXTRACT needs running"
<p>

This means that you should contact your EMBOSS administrator and ask
them to run the <b>tfextract</b> program to set up the TRANSFAC data for
EMBOSS. 
<p>

<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.


<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="jaspscan.html">jaspscan</a></td>
<td>Scans DNA sequences for transcription factors</td>
</tr>

</table>
<p>

Your EMBOSS administrator will have to run the EMBOSS program <b>tfextract</b>
in order to set up the data files from the TRANSFAC distribution files.
<p>


<H2>
    Author(s)
</H2>
Alan Bleasby 
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK

<p>
Please report all bugs to the EMBOSS bug team (emboss-bug&nbsp;&copy;&nbsp;emboss.open-bio.org) not to the original author.


<H2>
    History
</H2>

Written Summer 2000 - Alan Bleasby

<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.

<H2>
    Comments
</H2>
None

</BODY>
</HTML>

