<HTML>

<HEAD>
  <TITLE>
  EMBOSS: tfscan
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: tfscan
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program tfscan
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Scans DNA sequences for transcription factors


<H2>
    Description
</H2>

The
<A HREF="http://transfac.gbf.de/TRANSFAC/">TRANSFAC Database</A>
is a database of eukaryotic cis-acting regulatory DNA elements and
trans-acting factors.  It covers the whole range from yeast to human. 
<p>

The SITE data from TRANSFAC contains information on individual
(putatively) regulatory protein binding sites.  It has been divided into
the following taxonomic groups. 
<p>
<ul>
<li>Fungi
<li>Insects
<li>Plants
<li>Vertebrates
<li>Other
</ul>
<p>



The program <b>tfscan</b> takes a sequence and the name of one of these
taxonomic groups and does a fast match of the TRANSFAC sequences against
the input sequence (optionally allowing mismatches). 
<p>

The results is a list of the positions which match the binding sites in
the TRANSFAC SITE database. 
<p>

Because the binding sites are so small, there will be many spurious
(false positive) matches. 
<p>


<H2>
    Usage
</H2>


Here is a sample session with tfscan.

<pre>

% tfscan
Input sequence(s): embl:hsfos
Transcription Factor Class
         F : fungi
         I : insect
         P : plant
         V : vertebrate
         O : other
Select class [V]: v
Number of mismatches [0]: 
Output file [hsfos.tfscan]: 
</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
   -menu               list       Select class
   -mismatch           integer    Number of mismatches
  [-outfile]           outfile    Output file name

   Optional qualifiers: (none)
   Advanced qualifiers: (none)

</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-menu</td>
<td>Select class</td>
<td><table><tr><td>F</td> <td><i>(fungi)</i></td></tr><tr><td>I</td> <td><i>(insect)</i></td></tr><tr><td>P</td> <td><i>(plant)</i></td></tr><tr><td>V</td> <td><i>(vertebrate)</i></td></tr><tr><td>O</td> <td><i>(other)</i></td></tr></table></td>
<td>V</td>
</tr>

<tr>
<td>-mismatch</td>
<td>Number of mismatches</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.tfscan</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>

Normal nucleic acid sequence USA.

<H2>
    Output file format
</H2>


Here is part of the output from the example run. Some of the sites in
TRANSFAC are extremely short.
<p>
<hr>
<pre>
TFSCAN of HSFOS from 1 to 6210

HS$CFOS_20           R08485   384   396   agttcccgtcaat
DOG$ATP1A_01         R08484   3057  3063  gacatgg
HS$CEBPA_01          R08471   4535  4540  cacgtg
HS$GPB_05            R08210   3716  3721  gtatct
HS$GPB_02            R08207   5837  5842  ggtggg
HS$GPB_02            R08207   2399  2404  ggtggg
HS$GPB_02            R08207   2077  2082  ggtggg
MOUSE$TIMP1_02       R08168   2362  2368  caggaag
MOUSE$BMG_11         R08167   960   965   ggatag
RAT$RPK_02           R08166   6136  6140  tgtgc
RAT$RPK_02           R08166   5953  5957  tgtgc
RAT$RPK_02           R08166   4433  4437  tgtgc
RAT$RPK_02           R08166   4143  4147  tgtgc
RAT$RPK_02           R08166   3450  3454  tgtgc
RAT$RPK_02           R08166   3246  3250  tgtgc
RAT$RPK_02           R08166   3154  3158  tgtgc
RAT$RPK_02           R08166   1128  1132  tgtgc
HS$TIMP1_02          R08152   2361  2368  ccaggaag
HS$TIMP1_01          R08151   2642  2648  tgagtaa
HS$IL3_08            R05028   4376  4381  tgtggg
HS$IL3_08            R05028   3471  3476  tgtggg
HS$IL3_08            R05028   2584  2589  tgtggg
HS$IL3_08            R05028   2066  2071  tgtggg
HS$CATHD_01          R04883   1430  1435  ggcggg
HS$CATHD_01          R04883   1092  1097  ggcggg
HS$CATHD_01          R04883   569   574   ggcggg
RAT$IGFBP2_02        R04793   5123  5128  gggcgg
RAT$IGFBP2_02        R04793   1429  1434  gggcgg
RAT$IGFBP2_02        R04793   1091  1096  gggcgg
RAT$IGFBP2_02        R04793   607   612   gggcgg
RAT$IGFBP2_01        R04792   5123  5128  gggcgg
RAT$IGFBP2_01        R04792   1429  1434  gggcgg
RAT$IGFBP2_01        R04792   1091  1096  gggcgg
RAT$IGFBP2_01        R04792   607   612   gggcgg
HS$A14COL_01         R04791   5123  5128  gggcgg
HS$A14COL_01         R04791   1429  1434  gggcgg
HS$A14COL_01         R04791   1091  1096  gggcgg
HS$A14COL_01         R04791   607   612   gggcgg
HS$A24COL_03         R04790   5123  5128  gggcgg
etc......
</pre>
<hr>
<p>

The output consists of a title line then 5 columns separated by whitespace.
<p>

The first column is the identifier of the entry.
<p>

The second column is the Accession Number of the entry.
<p>

The third and fourth columns are the start and end positions of the match in your input sequence.
<p>

The fifth column is the sequence of the region where a match has been found.
<p>


For further details on an entry from the TRANSFAC database, see:
<br>
<A HREF="http://transfac.gbf.de/cgi-bin/qt/search.pl">http://transfac.gbf.de/cgi-bin/qt/search.pl</A>

<H2>
    Data files
</H2>

<b>tfscan</b> reads the TRANSFAC SITE data held in the EMBOSS data files:
<p>
<ul>
<li>tffungi 
<li>tfinsect 
<li>tfplant 
<li>tfvertebrate 
<li>tfother
</ul>
<p>

Your EMBOSS administrator will have to run the EMBOSS program <b>tfextract</b>
in order to set these files up from the TRANSFAC distribution files.
<p>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.
        If you wish to include the standard description of what data
        files are and how to use embossdata to inspect and retrieve
        them, use:
   -->
<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>



<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

<ul>
<li>Nucleic Acids Res. 16: 1879-1902, 1988
<li>BioTechForum - Advances in Molecular Genetics (J. Collins,A.J. Driesel, eds.) 4:95-108, 1991
<li>Nucleic Acids Res. 20:3-26, 1992 
</ul>


<H2>
    Warnings
</H2>

Your EMBOSS administrator will have to run the EMBOSS program <b>tfextract</b>
in order to set up the data files from the TRANSFAC distribution files.
<p>


<H2>
    Diagnostic Error Messages
</H2>

"EMBOSS An error in tfscan.c at line 82:
<br>
Either EMBOSS_DATA undefined or TFEXTRACT needs running"
<p>

This means that you should contact your EMBOSS administrator and ask
them to run the <b>tfextract</b> program to set up the TRANSFAC data for
EMBOSS. 
<p>

<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.


<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="chaos.html">chaos</a></td><td>Create a chaos game representation plot for a sequence</td></tr>
<tr><td><a href="cpgplot.html">cpgplot</a></td><td>Plot CpG rich areas</td></tr>
<tr><td><a href="cpgreport.html">cpgreport</a></td><td>Reports all CpG rich regions</td></tr>
<tr><td><a href="diffseq.html">diffseq</a></td><td>Find differences (SNPs) between nearly identical sequences</td></tr>
<tr><td><a href="dotmatcher.html">dotmatcher</a></td><td>Displays a thresholded dotplot of two sequences</td></tr>
<tr><td><a href="dotpath.html">dotpath</a></td><td>Displays a non-overlapping wordmatch dotplot of two sequences</td></tr>
<tr><td><a href="dottup.html">dottup</a></td><td>Displays a wordmatch dotplot of two sequences</td></tr>
<tr><td><a href="dreg.html">dreg</a></td><td>regular expression search of a nucleotide sequence</td></tr>
<tr><td><a href="einverted.html">einverted</a></td><td>Finds DNA inverted repeats</td></tr>
<tr><td><a href="equicktandem.html">equicktandem</a></td><td>Finds tandem repeats</td></tr>
<tr><td><a href="etandem.html">etandem</a></td><td>Looks for tandem repeats in a nucleotide sequence</td></tr>
<tr><td><a href="fuzznuc.html">fuzznuc</a></td><td>Nucleic acid pattern search</td></tr>
<tr><td><a href="fuzzpro.html">fuzzpro</a></td><td>Protein pattern search</td></tr>
<tr><td><a href="fuzztran.html">fuzztran</a></td><td>Protein pattern search after translation</td></tr>
<tr><td><a href="helixturnhelix.html">helixturnhelix</a></td><td>Report nucleic acid binding motifs</td></tr>
<tr><td><a href="isochore.html">isochore</a></td><td>Plots isochores in large DNA sequences</td></tr>
<tr><td><a href="newcpgreport.html">newcpgreport</a></td><td>Report CpG rich areas</td></tr>
<tr><td><a href="newcpgseek.html">newcpgseek</a></td><td>Reports CpG rich regions</td></tr>
<tr><td><a href="palindrome.html">palindrome</a></td><td>Looks for inverted repeats in a nucleotide sequence</td></tr>
<tr><td><a href="patmatdb.html">patmatdb</a></td><td>Search a protein sequence with a motif</td></tr>
<tr><td><a href="patmatmotifs.html">patmatmotifs</a></td><td>Search a PROSITE motif database with a protein sequence</td></tr>
<tr><td><a href="polydot.html">polydot</a></td><td>Displays all-against-all dotplots of a set of sequences</td></tr>
<tr><td><a href="preg.html">preg</a></td><td>regular expression search of a protein sequence</td></tr>
<tr><td><a href="printsextract.html">printsextract</a></td><td>Extract data from PRINTS</td></tr>
<tr><td><a href="prosextract.html">prosextract</a></td><td>Builds the PROSITE motif database for patmatmotifs to search</td></tr>
<tr><td><a href="pscan.html">pscan</a></td><td>Scans proteins using PRINTS</td></tr>
<tr><td><a href="recode.html">recode</a></td><td>Find and remove restriction sites but maintain the same translation</td></tr>
<tr><td><a href="redata.html">redata</a></td><td>Search REBASE for enzyme name, references, suppliers etc</td></tr>
<tr><td><a href="restrict.html">restrict</a></td><td>Finds restriction enzyme cleavage sites</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="silent.html">silent</a></td><td>Silent mutation restriction enzyme scan</td></tr>
<tr><td><a href="wobble.html">wobble</a></td><td>Wobble base plot</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->
<p>

Your EMBOSS administrator will have to run the EMBOSS program <b>tfextract</b>
in order to set up the data files from the TRANSFAC distribution files.
<p>


<H2>
    Author(s)
</H2>


This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>

Written Summer 2000 - Alan Bleasby

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

