<HTML>

<HEAD>
  <TITLE>
  EMBOSS: tfscan
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: tfscan
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program tfscan
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Scans DNA sequences for transcription factors


<H2>
    Description
</H2>


<H2>
    Usage
</H2>


Here is a sample session with tfscan.

<pre>

% tfscan
Input sequence(s): embl:hsfos
Transcription Factor Class
         F : fungi
         I : insect
         P : plant
         V : vertebrate
         O : other
Select class [V]: v
Number of mismatches [0]: 
Output file [hsfos.tfscan]: 

Scanning HSFOS...


</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
   -menu               list       Select class
   -mismatch           integer    Number of mismatches
  [-outfile]           outfile    Output file name

   Optional qualifiers: (none)
   Advanced qualifiers: (none)

</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-menu</td>
<td>Select class</td>
<td><table><tr><td>F</td> <td><i>(fungi)</i></td></tr><tr><td>I</td> <td><i>(insect)</i></td></tr><tr><td>P</td> <td><i>(plant)</i></td></tr><tr><td>V</td> <td><i>(vertebrate)</i></td></tr><tr><td>O</td> <td><i>(other)</i></td></tr></table></td>
<td>V</td>
</tr>

<tr>
<td>-mismatch</td>
<td>Number of mismatches</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.tfscan</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>


<H2>
    Output file format
</H2>


Here is part of the output from the example run. Some of the sites in
TRANSFAC are extremely short.

<pre>

TFSCAN of HSFOS from 1 to 6210

XENLA$VITB1_14       R04490   5200  5206  gcaaaca
XENLA$VITB1_14       R04490   4481  4487  gcaaaca

.................

HS$CFOS_19           R04047   412   438   caggatgtccatattaggacatctgcg
HS$CFOS_18           R04046   412   438   caggatgtccatattaggacatctgcg

.................

HS$CFOS_17           R03425   749   770   gagcaactgagaagccaagact

.................

HS$CFOS_16           R01889   420   429   ccatattagg

.................

HS$CFOS_15           R01640   412   419   caggatgt
MOUSE$CFOS_04        R00475   415   434   gatgtccatattaggacatc
MOUSE$CFOS_03        R00474   415   434   gatgtccatattaggacatc
MOUSE$CFOS_02        R00473   415   434   gatgtccatattaggacatc
MOUSE$CFOS_01        R00472   414   436   ggatgtccatattaggacatctg
HS$CFOS_14           R00471   669   676   gtgacgtt
HS$CFOS_13           R00470   649   656   gcgccacc
HS$CFOS_13           R00470   635   642   gcgccacc
HS$CFOS_12           R00469   1496  1502  tgcgtca
HS$CFOS_12           R00469   435   441   tgcgtca
HS$CFOS_11           R00468   1496  1502  tgcgtca
HS$CFOS_11           R00468   435   441   tgcgtca
HS$CFOS_10           R00467   415   434   gatgtccatattaggacatc
HS$CFOS_09           R00466   412   434   caggatgtccatattaggacatc
HS$CFOS_08           R00465   415   421   gatgtcc
HS$CFOS_07           R00464   415   434   gatgtccatattaggacatc
HS$CFOS_06           R00463   420   429   ccatattagg
HS$CFOS_05           R00462   415   434   gatgtccatattaggacatc
HS$CFOS_04           R00461   415   434   gatgtccatattaggacatc
HS$CFOS_03           R00460   401   440   ccccccttacacaggatgtccatattaggacatctgcg
tc
HS$CFOS_02           R00459   387   395   tcccgtcaa
HS$CFOS_01           R00458   383   397   cagttcccgtcaatc

</pre>

<H2>
    Data files
</H2>


<H2>
    Notes
</H2>


<H2>
    References
</H2>


<H2>
    Warnings
</H2>


<H2>
    Diagnostic Error Messages
</H2>


<H2>
    Exit status
</H2>


<H2>
    Known bugs
</H2>


<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="chaos.html">chaos</a></td><td>Create a chaos game representation plot for a sequence</td></tr>
<tr><td><a href="cpgplot.html">cpgplot</a></td><td>Plot CpG rich areas</td></tr>
<tr><td><a href="cpgreport.html">cpgreport</a></td><td>Reports all CpG rich regions</td></tr>
<tr><td><a href="diffseq.html">diffseq</a></td><td>Find differences (SNPs) between nearly identical sequences</td></tr>
<tr><td><a href="dotmatcher.html">dotmatcher</a></td><td>Displays a thresholded dotplot of two sequences</td></tr>
<tr><td><a href="dotpath.html">dotpath</a></td><td>Displays a non-overlapping wordmatch dotplot of two sequences</td></tr>
<tr><td><a href="dottup.html">dottup</a></td><td>Displays a wordmatch dotplot of two sequences</td></tr>
<tr><td><a href="dreg.html">dreg</a></td><td>regular expression search of a nucleotide sequence</td></tr>
<tr><td><a href="einverted.html">einverted</a></td><td>Finds DNA inverted repeats</td></tr>
<tr><td><a href="equicktandem.html">equicktandem</a></td><td>Finds tandem repeats</td></tr>
<tr><td><a href="etandem.html">etandem</a></td><td>Looks for tandem repeats in a nucleotide sequence</td></tr>
<tr><td><a href="fuzznuc.html">fuzznuc</a></td><td>Nucleic acid pattern search</td></tr>
<tr><td><a href="fuzzpro.html">fuzzpro</a></td><td>Protein pattern search</td></tr>
<tr><td><a href="fuzztran.html">fuzztran</a></td><td>Protein pattern search after translation</td></tr>
<tr><td><a href="helixturnhelix.html">helixturnhelix</a></td><td>Report nucleic acid binding motifs</td></tr>
<tr><td><a href="isochore.html">isochore</a></td><td>Plots isochores in large DNA sequences</td></tr>
<tr><td><a href="newcpgreport.html">newcpgreport</a></td><td>Report CpG rich areas</td></tr>
<tr><td><a href="newcpgseek.html">newcpgseek</a></td><td>Reports CpG rich regions</td></tr>
<tr><td><a href="palindrome.html">palindrome</a></td><td>Looks for inverted repeats in a nucleotide sequence</td></tr>
<tr><td><a href="patmatdb.html">patmatdb</a></td><td>Search a protein sequence with a motif</td></tr>
<tr><td><a href="patmatmotifs.html">patmatmotifs</a></td><td>Search a PROSITE motif database with a protein sequence</td></tr>
<tr><td><a href="polydot.html">polydot</a></td><td>Displays all-against-all dotplots of a set of sequences</td></tr>
<tr><td><a href="preg.html">preg</a></td><td>regular expression search of a protein sequence</td></tr>
<tr><td><a href="printsextract.html">printsextract</a></td><td>Extract data from PRINTS</td></tr>
<tr><td><a href="prosextract.html">prosextract</a></td><td>Builds the PROSITE motif database for patmatmotifs to search</td></tr>
<tr><td><a href="pscan.html">pscan</a></td><td>Scans proteins using PRINTS</td></tr>
<tr><td><a href="redata.html">redata</a></td><td>Search REBASE for enzyme name, references, suppliers etc</td></tr>
<tr><td><a href="restrict.html">restrict</a></td><td>Finds restriction enzyme cleavage sites</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
<tr><td><a href="silent.html">silent</a></td><td>Silent mutation restriction enzyme scan</td></tr>
<tr><td><a href="wobble.html">wobble</a></td><td>Wobble base plot</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>


<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

