<HTML>

<HEAD>
  <TITLE>
  EMBOSS: tmap
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
tmap
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
Predict and plot transmembrane segments in protein sequences

<H2>
    Description
</H2>

<p><b>tmap</b> predicts transmembrane segments for an aligned set of protein sequences, utilising the algorithm described in: "Persson, B. & Argos, P. (1994) Prediction of transmembrane segments in proteins utilising multiple sequence alignments J. Mol. Biol. 237, 182-192.".  A plot of the propensities to form the middle and the end of transmembrane regions is output. Bars are displayed in the plot above the regions predicted as being most likely to form transmembrane regions.  The transmembrane regions for the complete alignment are given first, followed by the predictions for each individual sequence in the alignment.</p>


<H2>
    Usage
</H2>
<b>Here is a sample session with tmap</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>tmap opsd.msf -out tmap.res -graph cps </b>
Predict and plot transmembrane segments in protein sequences

Created tmap.ps

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>


<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
Predict and plot transmembrane segments in protein sequences
Version: EMBOSS:6.3.0

   Standard (Mandatory) qualifiers:
  [-sequences]         seqset     File containing a sequence alignment
   -graph              xygraph    [$EMBOSS_GRAPHICS value, or x11] Graph type
                                  (ps, hpgl, hp7470, hp7580, meta, cps, x11,
                                  tek, tekt, none, data, xterm, png, gif, pdf,
                                  svg, pdf, svg)
  [-outfile]           report     [*.tmap] Output report file name

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-sequences" associated qualifiers
   -sbegin1            integer    Start of each sequence to be used
   -send1              integer    End of each sequence to be used
   -sreverse1          boolean    Reverse (if DNA)
   -sask1              boolean    Ask for begin/end/reverse
   -snucleotide1       boolean    Sequence is nucleotide
   -sprotein1          boolean    Sequence is protein
   -slower1            boolean    Make lower case
   -supper1            boolean    Make upper case
   -sformat1           string     Input sequence format
   -sdbname1           string     Database name
   -sid1               string     Entryname
   -ufo1               string     UFO features
   -fformat1           string     Features format
   -fopenfile1         string     Features file name

   "-graph" associated qualifiers
   -gprompt            boolean    Graph prompting
   -gdesc              string     Graph description
   -gtitle             string     Graph title
   -gsubtitle          string     Graph subtitle
   -gxtitle            string     Graph x axis title
   -gytitle            string     Graph y axis title
   -goutfile           string     Output file for non interactive displays
   -gdirectory         string     Output directory

   "-outfile" associated qualifiers
   -rformat2           string     Report format
   -rname2             string     Base file name
   -rextension2        string     File name extension
   -rdirectory2        string     Output directory
   -raccshow2          boolean    Show accession number in the report
   -rdesshow2          boolean    Show description in the report
   -rscoreshow2        boolean    Show the score in the report
   -rstrandshow2       boolean    Show the nucleotide strand in the report
   -rusashow2          boolean    Show the full USA in the report
   -rmaxall2           integer    Maximum total hits to report
   -rmaxseq2           integer    Maximum hits to report for one sequence

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-sequences]<br>(Parameter 1)</td>
<td>seqset</td>
<td>File containing a sequence alignment</td>
<td>Readable set of sequences</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-graph</td>
<td>xygraph</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including ps, hpgl, hp7470, hp7580, meta, cps, x11, tek, tekt, none, data, xterm, png, gif, pdf, svg, pdf, svg</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-outfile]<br>(Parameter 2)</td>
<td>report</td>
<td>Output report file name</td>
<td>Report output file</td>
<td><i>&lt;*&gt;</i>.tmap</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-sequences" associated seqset qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sbegin1<br>-sbegin_sequences</td>
<td>integer</td>
<td>Start of each sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -send1<br>-send_sequences</td>
<td>integer</td>
<td>End of each sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sreverse1<br>-sreverse_sequences</td>
<td>boolean</td>
<td>Reverse (if DNA)</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sask1<br>-sask_sequences</td>
<td>boolean</td>
<td>Ask for begin/end/reverse</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -snucleotide1<br>-snucleotide_sequences</td>
<td>boolean</td>
<td>Sequence is nucleotide</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sprotein1<br>-sprotein_sequences</td>
<td>boolean</td>
<td>Sequence is protein</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -slower1<br>-slower_sequences</td>
<td>boolean</td>
<td>Make lower case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -supper1<br>-supper_sequences</td>
<td>boolean</td>
<td>Make upper case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sformat1<br>-sformat_sequences</td>
<td>string</td>
<td>Input sequence format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sdbname1<br>-sdbname_sequences</td>
<td>string</td>
<td>Database name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sid1<br>-sid_sequences</td>
<td>string</td>
<td>Entryname</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ufo1<br>-ufo_sequences</td>
<td>string</td>
<td>UFO features</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fformat1<br>-fformat_sequences</td>
<td>string</td>
<td>Features format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fopenfile1<br>-fopenfile_sequences</td>
<td>string</td>
<td>Features file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-graph" associated xygraph qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -gprompt</td>
<td>boolean</td>
<td>Graph prompting</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -gdesc</td>
<td>string</td>
<td>Graph description</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -gtitle</td>
<td>string</td>
<td>Graph title</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -gsubtitle</td>
<td>string</td>
<td>Graph subtitle</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -gxtitle</td>
<td>string</td>
<td>Graph x axis title</td>
<td>Any string</td>
<td>Residue number</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -gytitle</td>
<td>string</td>
<td>Graph y axis title</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -goutfile</td>
<td>string</td>
<td>Output file for non interactive displays</td>
<td>Any string</td>
<td>tmap</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -gdirectory</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-outfile" associated report qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rformat2<br>-rformat_outfile</td>
<td>string</td>
<td>Report format</td>
<td>Any string</td>
<td>seqtable</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rname2<br>-rname_outfile</td>
<td>string</td>
<td>Base file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rextension2<br>-rextension_outfile</td>
<td>string</td>
<td>File name extension</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rdirectory2<br>-rdirectory_outfile</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -raccshow2<br>-raccshow_outfile</td>
<td>boolean</td>
<td>Show accession number in the report</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rdesshow2<br>-rdesshow_outfile</td>
<td>boolean</td>
<td>Show description in the report</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rscoreshow2<br>-rscoreshow_outfile</td>
<td>boolean</td>
<td>Show the score in the report</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rstrandshow2<br>-rstrandshow_outfile</td>
<td>boolean</td>
<td>Show the nucleotide strand in the report</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rusashow2<br>-rusashow_outfile</td>
<td>boolean</td>
<td>Show the full USA in the report</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rmaxall2<br>-rmaxall_outfile</td>
<td>integer</td>
<td>Maximum total hits to report</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -rmaxseq2<br>-rmaxseq_outfile</td>
<td>integer</td>
<td>Maximum hits to report for one sequence</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>


<H2>
    Input file format
</H2>


<b>tmap</b> reads a protein sequence USA for one or more aligned sequences.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: opsd.msf</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
!!AA_MULTIPLE_ALIGNMENT 1.0

  opsd.msf MSF:  354 Type: P 15/07/06 CompCheck: 5414 ..

  Name: OPSD_HUMAN Len: 354  Check: 2647 Weight: 50.00
  Name: OPSD_XENLA Len: 354  Check: 2767 Weight: 50.00

//

           1                                               50
OPSD_HUMAN MNGTEGPNFYVPFSNATGVVRSPFEYPQYYLAEPWQFSMLAAYMFLLIVL
OPSD_XENLA MNGTEGPNFYVPMSNKTGVVRSPFDYPQYYLAEPWQYSALAAYMFLLILL

           51                                             100
OPSD_HUMAN GFPINFLTLYVTVQHKKLRTPLNYILLNLAVADLFMVLGGFTSTLYTSLH
OPSD_XENLA GLPINFMTLFVTIQHKKLRTPLNYILLNLVFANHFMVLCGFTVTMYTSMH

           101                                            150
OPSD_HUMAN GYFVFGPTGCNLEGFFATLGGEIALWSLVVLAIERYVVVCKPMSNFRFGE
OPSD_XENLA GYFIFGPTGCYIEGFFATLGGEVALWSLVVLAVERYIVVCKPMANFRFGE

           151                                            200
OPSD_HUMAN NHAIMGVAFTWVMALACAAPPLAGWSRYIPEGLQCSCGIDYYTLKPEVNN
OPSD_XENLA NHAIMGVAFTWIMALSCAAPPLFGWSRYIPEGMQCSCGVDYYTLKPEVNN

           201                                            250
OPSD_HUMAN ESFVIYMFVVHFTIPMIIIFFCYGQLVFTVKEAAAQQQESATTQKAEKEV
OPSD_XENLA ESFVIYMFIVHFTIPLIVIFFCYGRLLCTVKEAAAQQQESLTTQKAEKEV

           251                                            300
OPSD_HUMAN TRMVIIMVIAFLICWVPYASVAFYIFTHQGSNFGPIFMTIPAFFAKSAAI
OPSD_XENLA TRMVVIMVVFFLICWVPYAYVAFYIFTHQGSNFGPVFMTVPAFFAKSSAI

           301                                            350
OPSD_HUMAN YNPVIYIMMNKQFRNCMLTTICCGKNPLGD.DEASATVSKTETSQVAPA~
OPSD_XENLA YNPVIYIVLNKQFRNCLITTLCCGKNPFGDEDGSSAATSKTEASSVSSSQ

           351
OPSD_HUMAN ~~~~
OPSD_XENLA VSPA

</pre>
</td></tr></table><p>


<H2>
    Output file format
</H2>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: tmap.res</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
########################################
# Program: tmap
# Rundate: Fri 15 Jan 2010 12:00:00
# Commandline: tmap
#    [-sequences] ../../data/opsd.msf
#    -outfile tmap.res
#    -graph cps
# Report_format: seqtable
# Report_file: tmap.res
########################################

#=======================================
#
# Sequence: Consensus     from: 1   to: 354
# HitCount: 7
#=======================================

  Start     End TransMem Sequence
     43      70        1 YMFLLIvLG-PINFlTLyVTvQHKKLRT
     71      98        2 PLNYILLNL--Ad-FMVL-GFT-TlYTS
    112     140        3 lEGFFATLGGEiALWSLVVLAiERYvVVC
    148     176        4 FGENHAIMGVAFTWvMALaCAAPPL-GWS
    201     229        5 ESFVIYMFvVHFTIPmIiIFFCYGqLv-T
    256     276        6 IMVi-FLICWVPYA-VAFYIF
    285     305        7 PiFMTiPAFFAKSaAIYNPVI

#---------------------------------------
#---------------------------------------
#=======================================
#
# Sequence: OPSD_HUMAN     from: 1   to: 354
# HitCount: 7
#=======================================

  Start     End TransMem Sequence
     43      70        1 YMFLLIVLGFPINFLTLYVTVQHKKLRT
     71      98        2 PLNYILLNLAVADLFMVLGGFTSTLYTS
    112     140        3 LEGFFATLGGEIALWSLVVLAIERYVVVC
    148     176        4 FGENHAIMGVAFTWVMALACAAPPLAGWS
    201     229        5 ESFVIYMFVVHFTIPMIIIFFCYGQLVFT
    256     276        6 IMVIAFLICWVPYASVAFYIF
    285     305        7 PIFMTIPAFFAKSAAIYNPVI

#---------------------------------------
#---------------------------------------
#=======================================
#
# Sequence: OPSD_XENLA     from: 1   to: 354
# HitCount: 7
#=======================================

  Start     End TransMem Sequence
     43      70        1 YMFLLILLGLPINFMTLFVTIQHKKLRT
     71      98        2 PLNYILLNLVFANHFMVLCGFTVTMYTS
    112     140        3 IEGFFATLGGEVALWSLVVLAVERYIVVC
    148     176        4 FGENHAIMGVAFTWIMALSCAAPPLFGWS
    201     229        5 ESFVIYMFIVHFTIPLIVIFFCYGRLLCT
    256     276        6 IMVVFFLICWVPYAYVAFYIF
    285     305        7 PVFMTVPAFFAKSSAIYNPVI

#---------------------------------------
#---------------------------------------

#---------------------------------------
# Total_sequences: 2
# Total_length: 702
# Reported_sequences: 3
# Reported_hitcount: 21
#---------------------------------------
</pre>
</td></tr></table><p>
<p><h3>Graphics File: tmap.ps</h3>
<p><img src="tmap.1.tmap.gif" alt="[tmap results]">

<p>

A plot of the propensities to form the middle (solid line) and the end
(dashed line) of transmembrane regions is output.

<p>

Bars are displayed in the plot above the regions predicted as being most
likely to form transmembrane regions. 

<p>

The text file (specified by the <b>-outfile</b> option) gives a summary
of these regions. 

<p>


The transmembrane regions for the complete alignment are given first,
followed by the predictions for each individual sequence in the
alignment.


<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

<p>Two sets of transmembrane propensity values are then used for the calculations: one for the middle, hydrophobic portion and one for the terminal regions of the transmembrane sequence spans. Average propensity values are calculated for each position along the alignment, with the contribution from each sequence weighted according to its dissimilarity relative to the other aligned sequences.</p>

<p>Eight-residue segments are considered as potential cores of transmembrane segments and elongated if thier middle propensity values are above a threshold. End propensity values are also considered as stop signals. Only helices with a length of 15 to 29 residues are allowed and corrections for strictly conserved charged residues are made.</p>

<p>The method is more successful than predictions based upon single sequences alone.</p>



<H2>
    References
</H2>

<ol>

<li>
 "Persson, B. & Argos, P. (1994) Prediction of transmembrane  
 segments in proteins utilsing multiple sequence alignments   
 J. Mol. Biol. 237, 182-192."                                 

</ol>

<H2>
    Warnings
</H2>

None.


<H2>
    Diagnostic Error Messages
</H2>

None.


<H2>
    Exit status
</H2>

0 if successful.

<H2>
    Known bugs
</H2>

None.


<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="garnier.html">garnier</a></td>
<td>Predicts protein secondary structure using GOR method</td>
</tr>

<tr>
<td><a href="helixturnhelix.html">helixturnhelix</a></td>
<td>Identify nucleic acid-binding motifs in protein sequences</td>
</tr>

<tr>
<td><a href="hmoment.html">hmoment</a></td>
<td>Calculate and plot hydrophobic moment for protein sequence(s)</td>
</tr>

<tr>
<td><a href="pepcoil.html">pepcoil</a></td>
<td>Predicts coiled coil regions in protein sequences</td>
</tr>

<tr>
<td><a href="pepnet.html">pepnet</a></td>
<td>Draw a helical net for a protein sequence</td>
</tr>

<tr>
<td><a href="pepwheel.html">pepwheel</a></td>
<td>Draw a helical wheel diagram for a protein sequence</td>
</tr>

</table>

<H2>
    Author(s)
</H2>


Original program by Bengt Persson and Patrick Argos.
<p>

This application was modified for inclusion in EMBOSS by 

Ian Longden formerly at:
<br>
Sanger Institute, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.                      

<p>
Please report all bugs to the EMBOSS bug team (emboss-bug&nbsp;&copy;&nbsp;emboss.open-bio.org) not to the original author.

<H2>
    History
</H2>

Completed 17th June 1999.

<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.



<H2>
    Comments
</H2>
The graph y axis is not labelled. It is some kind of score, but we
aer not sure exactly how to describe it.

</BODY>
</HTML>

