<HTML>

<HEAD>
  <TITLE>
  EMBOSS: tmap
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: tmap
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program tmap
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Displays membrane spanning regions


<H2>
    Description
</H2>

This program predicts transmembrane segments in proteins,    
utilising the algorithm described in:                        
"Persson, B. & Argos, P. (1994) Prediction of transmembrane  
segments in proteins utilising multiple sequence alignments   
J. Mol. Biol. 237, 182-192."                                 
<p>

<b>tmap</b> reads in one or more aligned protein sequences.
<p>

Two sets of propensity values are then used for the calculations: one
for the middle, hydrophobic portion and one for the terminal regions of
the transmembrane sequence spans.  Average propensity values are
calculated for each position along the alignment, with the contribution
from each sequence weighted according to its dissimilarity relative to
the other alignged sequences. 
<p>

Eight-residue segments are considered as potential cores of
transmembrane segments and elongated if thier middle propensity values
are above a threshold.  End propensity values are also considered as
stop signals.  Only helices with a length of 15 to 29 residues are
allowed and correctionbs for strictly conserved charged residues are
made. 
<p>

The method is more successful than predictions based upon single
sequences alone. 
<p>

The results are plotted on a graph and written to a text file.

<H2>
    Usage
</H2>


Here is a sample session with tmap.

<pre>

  % tmap sw:opsd_human -out tmap.res

</pre>
<a href="./tmap0.gif">click here for result</a>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-msf]               seqset     File containing a sequence alignment
   -graph              xygraph    Graph type

   Optional qualifiers:
   -outfile            outfile    Output file name

   Advanced qualifiers: (none)
   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-msf]<br>(Parameter 1)</td>
<td>File containing a sequence alignment</td>
<td>Readable sequences</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-graph</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including postscript, ps, hpgl, hp7470, hp7580, meta, colourps, cps, xwindows, x11, tektronics, tekt, tek4107t, tek, none, null, text, data, xterm, png</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-outfile</td>
<td>Output file name</td>
<td>Output file</td>
<td>tmap.res</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>


Any protein sequence USA for one or more aligned sequences.

<H2>
    Output file format
</H2>

A plot of the propensities to form the middle and the end of
transmembrane regions is output. 
<p>

Bars are displayed in the plot above the regions predicted as being most
likely to form transmembrane regions. 
<p>

The text file (specified by the <b>-output</b> option) gives a summary
of these regions. 
The text file produced by the above example is:
<p>

<hr>
<pre>
Program TMAP, version 46, to predict transmembrane segments from .msf file.
The program reads a multiple alignment file of the GCG multiple sequence format
and predicts membrane-spanning regions according to the algorithm in
Persson & Argos (1994), J. Mol. Biol. 237, 182-192.

RESULTS from program TMAP, edition 46'

Numbers give: a) number of transmembrane segment
              b) start of TM segment (alignment position / residue number)
              c) end of TM segment (alignment position / residue number)
              d) length of TM segment within parentheses

PREDICTED TRANSMEMBRANE SEGMENTS FOR ALIGNMENT 

  TM  1:   43 -   69  (27.0)
  TM  2:   73 -   97  (25.0)
  TM  3:  112 -  140  (29.0)
  TM  4:  148 -  176  (29.0)
  TM  5:  201 -  229  (29.0)
  TM  6:  255 -  275  (21.0)
  TM  7:  282 -  302  (21.0)


PREDICTED TRANSMEMBRANE SEGMENTS FOR PROTEIN OPSD_HUMAN

  TM  1:   43 -   69 (27)
  TM  2:   73 -   97 (25)
  TM  3:  112 -  140 (29)
  TM  4:  148 -  176 (29)
  TM  5:  201 -  229 (29)
  TM  6:  255 -  275 (21)
  TM  7:  282 -  302 (21)
</pre>
<hr>
<p>


The transmembrane regions for the complete alignment are given first,
followed by the predictions for each individual sequence in the
alignment.  (In the example above there is only 1 sequence in the
alignment.)
<p>

<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

None.


<H2>
    References
</H2>

 "Persson, B. & Argos, P. (1994) Prediction of transmembrane  
 segments in proteins utilsing multiple sequence alignments   
 J. Mol. Biol. 237, 182-192."                                 

<H2>
    Warnings
</H2>

None.


<H2>
    Diagnostic Error Messages
</H2>

None.


<H2>
    Exit status
</H2>

    0 if successful.

<H2>
    Known bugs
</H2>

None.


<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="garnier.html">garnier</a></td><td>Predicts protein secondary structure</td></tr>
<tr><td><a href="helixturnhelix.html">helixturnhelix</a></td><td>Report nucleic acid binding motifs</td></tr>
<tr><td><a href="hmoment.html">hmoment</a></td><td>Hydrophobic moment calculation</td></tr>
<tr><td><a href="pepcoil.html">pepcoil</a></td><td>Predicts coiled coil regions</td></tr>
<tr><td><a href="pepnet.html">pepnet</a></td><td>Displays proteins as a helical net</td></tr>
<tr><td><a href="pepwheel.html">pepwheel</a></td><td>Shows protein sequences as helices</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


Original program by Bengt Persson and Patrick Argos.
<p>

This application was modified for inclusion in EMBOSS by Ian Longden <a
href="mailto:il@sanger.ac.uk">(il@sanger.ac.uk)</a> Informatics
Division, The Sanger Centre, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.

<H2>
    History
</H2>

Completed 17th June 1999.
<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

