<HTML>

<HEAD>
  <TITLE>
  EMBOSS: transeq
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: transeq
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program transeq
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Translate nucleic acid sequences


<H2>
    Description
</H2>


	This translate nucleic acid sequences to the corresponding
	peptide sequence.  
<p>
	It can translate in any of the 3 forward or three reverse sense frames,
	or in all three forward or reverse frames, or in all six frames. 
<p>
	It can translate specified regions corersponding to the coding
	regions of your sequences. 
<p>
	It can translate using the standard ('Universal') genetic code
	and also with a selection of non-standard codes. 
<p>
	Termination (STOP) codons are translated as the character '*'.
<p>
	The output peptide sequence is always in the standard one-letter
	IUPAC code. 

<H2>
    Usage
</H2>


	To translate a sequence 'pop.seq' in the first frame (starting
	at the first base and proceeding to the end):
	<pre>% transeq pop.seq pop.pep</pre>
<p>
	To translate a sequence 'pop.seq' in the second frame:
	<pre>% transeq pop.seq pop.pep -frame=2</pre>
<p>
	To translate a sequence 'pop.seq' in the third frame in the
	reverse sense (starting at the last base and proceeding to the start):
	<pre>% transeq pop.seq pop.pep -frame=-1</pre>
<p>
	To translate a sequence 'pop.seq' in all three forward frames:
	<pre>% transeq pop.seq pop.pep -frame=F</pre>
<p>
	To translate a sequence 'pop.seq' in all three reverse frames:
	<pre>% transeq pop.seq pop.pep -frame=R</pre>
<p>
	To translate a sequence 'pop.seq' in all six forward and reverse
	frames:
	<pre>% transeq pop.seq pop.pep -frame=6</pre>
<p>
	To translate a specific set of regions corresponding to a known
	set of coding sequences:
	<pre>% transeq pop.seq pop.pep -reg=2-45,67-201,328-509</pre>
<p>
	To translate a sequence 'mito.seq' using the mammalian
	mitochondrion genetic code table:
	<pre>% transeq mito.seq mito.pep -table=2</pre>
<p>

<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
  [-outseq]            seqoutall  Output sequence(s) USA

   Optional qualifiers:
   -frame              list       Frame(s) to translate
   -table              list       Code to use
   -regions            range      Regions to translate.
                                  If this is left blank, then the complete
                                  sequence is translated.
                                  A set of regions is specified by a set of
                                  pairs of positions.
                                  The positions are integers.
                                  They are separated by any non-digit,
                                  non-alpha character.
                                  Examples of region specifications are:
                                  24-45, 56-78
                                  1:45, 67=99;765..888
                                  1,5,8,10,23,45,57,99
   -trim               bool       This removes all X and * characters from the
                                  right end of the translation. The trimming
                                  process starts at the end and continues
                                  until the next character is not a X or a *

   Advanced qualifiers: (none)

</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outseq]<br>(Parameter 2)</td>
<td>Output sequence(s) USA</td>
<td>Writeable sequence(s)</td>
<td><i>&lt;sequence&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-frame</td>
<td>Frame(s) to translate</td>
<td><table><tr><td>1</td> <td><i>(1)</i></td></tr><tr><td>2</td> <td><i>(2)</i></td></tr><tr><td>3</td> <td><i>(3)</i></td></tr><tr><td>F</td> <td><i>(Forward three frames)</i></td></tr><tr><td>-1</td> <td><i>(-1)</i></td></tr><tr><td>-2</td> <td><i>(-2)</i></td></tr><tr><td>-3</td> <td><i>(-3)</i></td></tr><tr><td>R</td> <td><i>(Reverse three frames)</i></td></tr><tr><td>6</td> <td><i>(All six frames)</i></td></tr></table></td>
<td>1</td>
</tr>

<tr>
<td>-table</td>
<td>Code to use</td>
<td><table><tr><td>0</td> <td><i>(Standard)</i></td></tr><tr><td>1</td> <td><i>(Standard (with alternative initiation codons))</i></td></tr><tr><td>2</td> <td><i>(Vertebrate Mitochondrial)</i></td></tr><tr><td>3</td> <td><i>(Yeast Mitochondrial)</i></td></tr><tr><td>4</td> <td><i>(Mold, Protozoan, Coelenterate Mitochondrial and Mycoplasma/Spiroplasma)</i></td></tr><tr><td>5</td> <td><i>(Invertebrate Mitochondrial)</i></td></tr><tr><td>6</td> <td><i>(Ciliate Macronuclear and Dasycladacean)</i></td></tr><tr><td>9</td> <td><i>(Echinoderm Mitochondrial)</i></td></tr><tr><td>10</td> <td><i>(Euplotid Nuclear)</i></td></tr><tr><td>11</td> <td><i>(Bacterial)</i></td></tr><tr><td>12</td> <td><i>(Alternative Yeast Nuclear)</i></td></tr><tr><td>13</td> <td><i>(Ascidian Mitochondrial)</i></td></tr><tr><td>14</td> <td><i>(Flatworm Mitochondrial)</i></td></tr><tr><td>15</td> <td><i>(Blepharisma Macronuclear)</i></td></tr><tr><td>16</td> <td><i>(Chlorophycean Mitochondrial)</i></td></tr><tr><td>21</td> <td><i>(Trematode Mitochondrial)</i></td></tr><tr><td>22</td> <td><i>(Scenedesmus obliquus)</i></td></tr><tr><td>23</td> <td><i>(Thraustochytrium Mitochondrial)</i></td></tr></table></td>
<td>0</td>
</tr>

<tr>
<td>-regions</td>
<td>Regions to translate.
If this is left blank, then the complete sequence is translated.
A set of regions is specified by a set of pairs of positions.
The positions are integers.
They are separated by any non-digit, non-alpha character.
Examples of region specifications are:
24-45, 56-78
1:45, 67=99;765..888
1,5,8,10,23,45,57,99</td>
<td>Sequence range</td>
<td>Whole sequence</td>
</tr>

<tr>
<td>-trim</td>
<td>This removes all X and * characters from the right end of the translation. The trimming process starts at the end and continues until the next character is not a X or a *</td>
<td>Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>


The input sequence can be one or more nucleic acid sequences.

<H2>
    Output file format
</H2>


One or more peptide sequences are written out.
<P>

The names of the resulting protein sequences are formed from the name of
the input nucleic acid sequence with '_' and the translation frame
appended to it.  Thus a nucleic acid sequence with the name 'XYZ'
franslated in all 6 frame would produce protein sequences with the
names: 'XYZ_1', 'XYZ_2', 'XYZ_3', 'XYZ_4', 'XYZ_5', 'XYZ_6'. 
<p>

For example, the result of the command
<br>
transeq em:hsfau -frame=6
<br>
would be:
<p>

<hr>
<pre>
>HSFAU_1 H.sapiens fau mRNA
FLFLDSIFAVAGTAVQSPICSSLSAPRSYTPSR*PARKRSPRSRLM*PHWRALPRKIKSC
SWQARPWRMRPLWASAGWRP*LPWK*QAACLEVKFMVPWPVLEK*EVRLLRWPNRRRRRR
RQVGLSGGCSTTGALSTLCPPLARRRAPMPTLKSFVILAFSNKKAT*FSQKKX
>HSFAU_2 H.sapiens fau mRNA
SSFSTPSSR*LGPPFSRQYAALCPRPGATHLRGDRPGNGRPDQGSCSLTGGHCPGRSSRA
PGRRAPGG*GHSGPVRGGGPDYPGSSRPHAWR*SSWFPGPCWKSERSDS*GGQTGEEEEE
DRSG*AADAVQPALCQRCAHLWQEEGPQCQLLSLL*FWLSLIKKPLSSVKKKX
>HSFAU_3 H.sapiens fau mRNA
PLSRLHLRGSWDRRSVANMQLFVRAQELHTFEVTGQETVAQIKAHVASLEGIAPEDQVVL
LAGAPLEDEATLGQCGVEALTTLEVAGRMLGGKVHGSLARAGKVRGQTPKVAKQEKKKKK
TGRAKRRMQYNRRFVNVVPTFGKKKGPNANS*VFCNSGFL**KSHLVQSKKK
>HSFAU_4 H.sapiens fau mRNA
FFF*LN*VAFLLEKARITKDLRVGIGALLLAKGGHNVDKAPVVLHPPLSPTCLLLLLLLF
GHLRSLTSHFSSTGQGTMNFTSKHAACYFQGSQGLHPALAQSGLILQGRACQEHDLIFRG
NALQ*GYMSLDLGDRFLAGHLEGV*LLGADKELHIGD*TAVPATAKMESRKRX
>HSFAU_5 H.sapiens fau mRNA
FFFD*TKWLFY*RKPELQKT*ELALGPFFLPKVGTTLTKRRLYCIRRLARPVFFFFFSCL
ATLGV*PLTFPARAREP*TLPPSMRPATSRVVRASTPHWPRVASSSRGAPARSTT*SSGA
MPSSEAT*ALIWATVSWPVTSKVCSSWARTKSCILATERRSQLPRRWSRERGX
>HSFAU_6 H.sapiens fau mRNA
FFLTELSGFFIRESQNYKRLKSWHWGPSSCQRWAQR*QSAGCTASAA*PDLSSSSSSPVW
PP*ESDLSLFQHGPGNHELYLQACGLLLPG*SGPPPRTGPEWPHPPGARLPGARLDLPGQ
CPPVRLHEP*SGRPFPGRSPRRCVAPGRGQRAAYWRLNGGPSYREDGVEKEE
</pre>
<hr>
<p>

If regions are specified, they are taken to be translated in frame 1 and
so the output name would be 'XYZ_1'. 
<p>



<H2>
    Data files
</H2>


<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by EMBOSS environment variable EMBOSS_DATA.

<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>
The Genetic Code data files are based on the NCBI genetic code tables. 
Their names and descriptions are:

<dl>
<dt>EGC.0 </dt><dd>
      Standard (Differs from GC.1 in that it only has initiation site 'AUG') 
<dt>EGC.1 </dt><dd>
      Standard 
<dt>EGC.2 </dt><dd>
      Vertebrate Mitochodrial
<dt>EGC.3 </dt><dd>
      Yeast Mitochondrial 
<dt>EGC.4 </dt><dd>
      Mold, Protozoan, Coelenterate Mitochondrial and Mycoplasma/Spiroplasma 
<dt>EGC.5 </dt><dd>
      Invertebrate Mitochondrial 
<dt>EGC.6 </dt><dd>
      Ciliate Macronuclear and Dasycladacean 
<dt>EGC.9 </dt><dd>
      Echinoderm Mitochondrial 
<dt>EGC.10 </dt><dd>
      Euplotid Nuclear 
<dt>EGC.11 </dt><dd>
      Bacterial 
<dt>EGC.12 </dt><dd>
      Alternative Yeast Nuclear 
<dt>EGC.13 </dt><dd>
      Ascidian Mitochondrial 
<dt>EGC.14 </dt><dd>
      Flatworm Mitochondrial 
<dt>EGC.15</dt><dd> 
      Blepharisma Macronuclear 
</dl>
<p>
The format of these files is very simple. 

<p>
It consists of several lines of optional comments, each starting with
a '#' character.

<p>
These are followed the line: 'Genetic Code [n]', where 'n' is the
number of the genetic code file.

<p>
This is followed by the description of the code and then by four lines
giving the IUPAC one-letter code of the translated amino acid, the
start codons (indicdated by an 'M') and the three bases of the codon,
lined up one on top of the other.

<p>
For example: 

<pre>

------------------------------------------------------------------------------
# Genetic Code Table
#
# Obtained from: http://www.ncbi.nlm.nih.gov/collab/FT/genetic_codes.html
# and: http://www3.ncbi.nlm.nih.gov/htbin-post/Taxonomy/wprintgc?mode=c
#
# Differs from Genetic Code [1] only in that the initiation sites have been
# changed to only 'AUG'

Genetic Code [0]
Standard
 
AAs  =   FFLLSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG
Starts = -----------------------------------M----------------------------
Base1  = TTTTTTTTTTTTTTTTCCCCCCCCCCCCCCCCAAAAAAAAAAAAAAAAGGGGGGGGGGGGGGGG
Base2  = TTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGG
Base3  = TCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAG
------------------------------------------------------------------------------

</pre>


<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.


<H2>
    Warnings
</H2>


	When translation using non-standard genetic code table, always
	check the table carefully for deviations from your particular
	organism's code.

<H2>
    Diagnostic Error Messages
</H2>


	Several warning messages about malformed region specifications:
<ul>
        <li>Non-digit found in region ...
        <li>Unpaired start of a region found in ...
        <li>Non-digit found in region ...
        <li>The start of a pair of region positions must be smaller than the
                end in ... 
</ul>
<H2>
    Exit status
</H2>


	It exits with status 0, unless a region is badly constructed.

<H2>
    Known bugs
</H2>


	Bugs noted but not yet fixed.
<p>
	None.

<!--
<H2>
    See also
</H2>
-->
<a name=See also><h2>See also</h2></a>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="backtranseq.html">backtranseq</a></td><td>Back translate a protein sequence</td></tr>
<tr><td><a href="cusp.html">cusp</a></td><td>Create a codon usage table</td></tr>
<tr><td><a href="getorf.html">getorf</a></td><td>Finds and extracts open reading frames (ORFs)</td></tr>
<tr><td><a href="plotorf.html">plotorf</a></td><td>Plot potential open reading frames</td></tr>
<tr><td><a href="prettyseq.html">prettyseq</a></td><td>Output sequence with translated ranges</td></tr>
<tr><td><a href="remap.html">remap</a></td><td>Display a sequence with restriction cut sites, translation etc</td></tr>
<tr><td><a href="showorf.html">showorf</a></td><td>Pretty output of DNA translations</td></tr>
<tr><td><a href="showseq.html">showseq</a></td><td>Display a sequence with features, translation etc</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Gary Williams <a
href="mailto:gwilliam@hgmp.mrc.ac.uk">(gwilliam@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>


	Completed 4 March 1999

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

