<HTML>

<HEAD>
  <TITLE>
  EMBOSS: trimseq
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">







<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

trimseq
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>







<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Trim ambiguous bits off the ends of sequences


<H2>
    Description
</H2>


This program is used to tidy up the ends of sequences, removing all
the bits that you would really rather were not published.

<p>
Specifically, it:
<ul>

<li>removes all gap characters from the ends.
<li>removes X's and N's (in nucleic sequences) from the ends.
<li>optionally removes *'s from the ends
<li>optionally removes IUPAC ambiguity codes from the ends
  (B and Z in proteins, M,R,W,S,Y,K,V,H,D and B in nucleic sequences)
</ul>
<p>
It then optionally trims off poor quality regions from the end, using
a threshold percentage of unwanted characters in a window which is moved
along the sequence from the ends. The unwanted characters which are used
are X's and N's (in nucleic sequences), optionally *'s, and optionally
IUPAC ambiguity codes.

<p>
The program stops trimming the ends when the percentage of
unwanted characters in the moving window drops below the
threshold percentage. 
	
<p>
Thus if the window size is set to 1 and the percentage threshold
is 100, no further poor quality regions will be removed.
If the window size is set to 5 and the percentage threshold is 40
then the sequence  AAGCTNNNNATT   will be trimmed to AAGCT,
while AAGCTNATT or AAGCTNNNNATTT will not be trimmed as less than
40% of the last 5 characters are N's.

<p>
After trimming these poor quality regions, it will again then
trim off any dangling gap characters from the ends .

<H2>
    Usage
</H2>


Tidy up the sequence ends, stopping at the first wanted code
<pre>
% trimseq xyz.seq xyz_clean.seq -window 1 -percent 100
</pre>

<p>
Tidy up the sequence ends, removing poor bits at the ends
<pre>
% trimseq xyz.seq xyz_clean.seq -window 5 -percent 40
</pre>

<p>
Tidy up the sequence ends, removing very poor bits at the ends
<pre>
% trimseq xyz.seq xyz_clean.seq -window 20 -percent 80
</pre>

<p>
Tidy up the sequence ends, removing even maginally poor bits at the ends
<pre>
% trimseq xyz.seq xyz_clean.seq -window 20 -percent 10
</pre>

<p>
Tidy up the sequence ends, removing poor bits including ambiguity codes
<pre>
% trimseq xyz.seq xyz_clean.seq -window 20 -percent 50 -strict
</pre>

<p>
Tidy up the sequence ends, removing asterisks from a protein end
<pre>
% trimseq xyz.seq xyz_clean.seq -window 1 -percent 100 -star
</pre>

<p>
Tidy up the sequence ends, removing poor bits at only the left end
<pre>
% trimseq xyz.seq xyz_clean.seq -window 20 -percent 50 -noright
</pre>


<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-sequence]          seqall     Sequence database USA
  [-outseq]            seqoutall  Output sequence(s) USA

   Additional (Optional) qualifiers:
   -window             integer    This determines the size of the region that
                                  is considered when deciding whether the
                                  percentage of ambiguity is greater than the
                                  threshold. A value of 5 means that a region
                                  of 5 letters in the sequence is shifted
                                  along the sequence from the ends and
                                  trimming is done only if there is a greater
                                  or equal percentage of ambiguity than the
                                  threshold percentage.
   -percent            float      This is the threshold of the percentage
                                  ambiguity in the window required in order to
                                  trim a sequence.
   -strict             boolean    In nucleic sequences, trim off not only N's
                                  and X's, but also the nucleotide IUPAC
                                  ambiguity codes M, R, W, S, Y, K, V, H, D
                                  and B. In protein sequences, trim off not
                                  only X's but also B and Z.
   -star               boolean    In protein sequences, trim off not only X's,
                                  but also the *'s

   Advanced (Unprompted) qualifiers:
   -[no]left           boolean    Trim at the start
   -[no]right          boolean    Trim at the end

   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1             integer    Start of each sequence to be used
   -send1               integer    End of each sequence to be used
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name

   "-outseq" associated qualifiers
   -osformat2           string     Output seq format
   -osextension2        string     File name extension
   -osname2             string     Base file name
   -osdirectory2        string     Output directory
   -osdbname2           string     Database name to add
   -ossingle2           boolean    Separate file for each entry
   -oufo2               string     UFO features
   -offormat2           string     Features format
   -ofname2             string     Features file name
   -ofdirectory2        string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outseq]<br>(Parameter 2)</td>
<td>Output sequence(s) USA</td>
<td>Writeable sequence(s)</td>
<td><i>&lt;sequence&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-window</td>
<td>This determines the size of the region that is considered when deciding whether the percentage of ambiguity is greater than the threshold. A value of 5 means that a region of 5 letters in the sequence is shifted along the sequence from the ends and trimming is done only if there is a greater or equal percentage of ambiguity than the threshold percentage.</td>
<td>Any integer value</td>
<td>1</td>
</tr>

<tr>
<td>-percent</td>
<td>This is the threshold of the percentage ambiguity in the window required in order to trim a sequence.</td>
<td>Any numeric value</td>
<td>100.0</td>
</tr>

<tr>
<td>-strict</td>
<td>In nucleic sequences, trim off not only N's and X's, but also the nucleotide IUPAC ambiguity codes M, R, W, S, Y, K, V, H, D and B. In protein sequences, trim off not only X's but also B and Z.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-star</td>
<td>In protein sequences, trim off not only X's, but also the *'s</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]left</td>
<td>Trim at the start</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]right</td>
<td>Trim at the end</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

</table>



<H2>
    Input file format
</H2>


Normal sequence.

<H2>
    Output file format
</H2>


Normal sequence file.

<H2>
    Data files
</H2>


None.
	
<H2>
    Notes
</H2>


If you use the '-star' qualifier and set the window size to greater
than 1, you may trim bits of sequence with internal *'s. This may
not be what you expected.

<H2>
    References
</H2>


None.

<H2>
    Warnings
</H2>


None.

<H2>
    Diagnostic Error Messages
</H2>


None.

<H2>
    Exit status
</H2>


It always exits with status 0.

<H2>
    Known bugs
</H2>


None noted.

<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="biosed.html">biosed</a></td><td>Replace or delete sequence sections</td></tr>
<tr><td><a href="cutseq.html">cutseq</a></td><td>Removes a specified section from a sequence</td></tr>
<tr><td><a href="degapseq.html">degapseq</a></td><td>Removes gap characters from sequences</td></tr>
<tr><td><a href="descseq.html">descseq</a></td><td>Alter the name or description of a sequence</td></tr>
<tr><td><a href="entret.html">entret</a></td><td>Reads and writes (returns) flatfile entries</td></tr>
<tr><td><a href="extractfeat.html">extractfeat</a></td><td>Extract features from a sequence</td></tr>
<tr><td><a href="extractseq.html">extractseq</a></td><td>Extract regions from a sequence</td></tr>
<tr><td><a href="listor.html">listor</a></td><td>Writes a list file of the logical OR of two sets of sequences</td></tr>
<tr><td><a href="maskfeat.html">maskfeat</a></td><td>Mask off features of a sequence</td></tr>
<tr><td><a href="maskseq.html">maskseq</a></td><td>Mask off regions of a sequence</td></tr>
<tr><td><a href="newseq.html">newseq</a></td><td>Type in a short new sequence</td></tr>
<tr><td><a href="noreturn.html">noreturn</a></td><td>Removes carriage return from ASCII files</td></tr>
<tr><td><a href="notseq.html">notseq</a></td><td>Excludes a set of sequences and writes out the remaining ones</td></tr>
<tr><td><a href="nthseq.html">nthseq</a></td><td>Writes one sequence from a multiple set of sequences</td></tr>
<tr><td><a href="pasteseq.html">pasteseq</a></td><td>Insert one sequence into another</td></tr>
<tr><td><a href="revseq.html">revseq</a></td><td>Reverse and complement a sequence</td></tr>
<tr><td><a href="seqret.html">seqret</a></td><td>Reads and writes (returns) sequences</td></tr>
<tr><td><a href="seqretsplit.html">seqretsplit</a></td><td>Reads and writes (returns) sequences in individual files</td></tr>
<tr><td><a href="skipseq.html">skipseq</a></td><td>Reads and writes (returns) sequences, skipping the first few</td></tr>
<tr><td><a href="splitter.html">splitter</a></td><td>Split a sequence into (overlapping) smaller sequences</td></tr>
<tr><td><a href="trimest.html">trimest</a></td><td>Trim poly-A tails off EST sequences</td></tr>
<tr><td><a href="union.html">union</a></td><td>Reads sequence fragments and builds one sequence</td></tr>
<tr><td><a href="vectorstrip.html">vectorstrip</a></td><td>Strips out DNA between a pair of vector sequences</td></tr>
<tr><td><a href="yank.html">yank</a></td><td>Reads a sequence range, appends the full USA to a list file</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->



<H2>
    Author(s)
</H2>


Gary Williams (gwilliam&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK




<H2>
    History
</H2>

<!--
<p>
[an error occurred while processing this directive]
-->

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

