<HTML>
<HEAD>
  <TITLE>
  EMBOSS: trimspace
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
trimspace
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
    Function
</H2>
Remove extra whitespace from an ASCII text file
<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

Replaces all white space with a single space. <b>nospace</b>
removes all whitespace from the input sequence.



<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

<b>Here is a sample session with trimspace</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>trimspace </b>
Remove extra whitespace from an ASCII text file
ASCII text file: <b>seqspace2.txt</b>
ASCII text output file [seqspace2.trimspace]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-infile]            infile     ASCII text file
  [-outfile]           outfile    [*.trimspace] ASCII text output file

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-outfile" associated qualifiers
   -odirectory2        string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-infile]<br>(Parameter 1)</td>
<td>ASCII text file</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>ASCII text output file</td>
<td>Output file</td>
<td><i>&lt;*&gt;</i>.trimspace</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>trimspace</b> reads any text file.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: seqspace2.txt</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
&gt;X13776   Pseudomonas aeruginosa    amiC and amiR gene
 ggtaccgctg gccgagcatc tgctcgatca ccaccagccg ggcgacggga
 actgcacgat ctacctggcg agcctggagc acgagcgggt tcgcttcgta
 cggcgctgag cgacagtcac aggagaggaa acggatggga tcgcaccagg
 agcggccgct gatcggcctg ctgttctccg aaaccggcgt caccgccgat
 atcgagcgct cgcacgcgta tggcgcattg ctcgcggtcg agcaactgaa
 ccgcgagggc ggcgtcggcg gtcgcccgat cgaaacgctg tcccaggacc
 ccggcggcga cccggaccgc tatcggctgt gcgccgagga cttcattcgc
 aaccgggggg tacggttcct cgtgggctgc tacatgtcgc acacgcgcaa
 ggcggtgatg ccggtggtcg agcgcgccga cgcgctgctc tgctacccga
 ccccctacga gggcttcgag tattcgccga acatcgtcta cggcggtccg
 gcgccgaacc agaacagtgc gccgctggcg gcgtacctga ttcgccacta
 cggcgagcgg gtggtgttca tcggctcgga ctacatctat ccgcgggaaa
 gcaaccatgt gatgcgccac ctgtatcgcc agcacggcgg cacggtgctc
 gaggaaatct acattccgct gtatccctcc gacgacgact tgcagcgcgc
 cgtcgagcgc atctaccagg cgcgcgccga cgtggtcttc tccaccgtgg
 tgggcaccgg caccgccgag ctgtatcgcg ccatcgcccg tcgctacggc
 gacggcaggc ggccgccgat cgccagcctg accaccagcg aggcggaggt
 ggcgaagatg gagagtgacg tggcagaggg gcaggtggtg gtcgcgcctt
 acttctccag catcgatacg cccgccagcc gggccttcgt ccaggcctgc
 catggtttct tcccggagaa cgcgaccatc accgcctggg ccgaggcggc
 ctactggcag accttgttgc tcggccgcgc cgcgcaggcc gcaggcaact
 ggcgggtgga agacgtgcag cggcacctgt acgacatcga catcgacgcg
 ccacaggggc cggtccgggt ggagcgccag aacaaccaca gccgcctgtc
 ttcgcgcatc gcggaaatcg atgcgcgcgg cgtgttccag gtccgctggc
 agtcgcccga accgattcgc cccgaccctt atgtcgtcgt gcataacctc
 gacgactggt ccgccagcat gggcggggga ccgctcccat gagcgccaac
 tcgctgctcg gcagcctgcg cgagttgcag gtgctggtcc tcaacccgcc
 gggggaggtc agcgacgccc tggtcttgca gctgatccgc atcggttgtt
 cggtgcgcca gtgctggccg ccgccggaag ccttcgacgt gccggtggac
 gtggtcttca ccagcatttt ccagaatggc caccacgacg agatcgctgc
 gctgctcgcc gccgggactc cgcgcactac cctggtggcg ctggtggagt
 acgaaagccc cgcggtgctc tcgcagatca tcgagctgga gtgccacggc
 gtgatcaccc agccgctcga tgcccaccgg gtgctgcctg tgctggtatc
 ggcgcggcgc atcagcgagg aaatggcgaa gctgaagcag aagaccgagc
 agctccagga ccgcatcgcc ggccaggccc ggatcaacca ggccaaggtg
 ttgctgatgc agcgccatgg ctgggacgag cgcgaggcgc accagcacct
 gtcgcgggaa gcgatgaagc ggcgcgagcc gatcctgaag atcgctcagg
 agttgctggg aaacgagccg tccgcctgag cgatccgggc cgaccagaac
 aataacaaga ggggtatcgt catcatgctg ggactggttc tgctgtacgt
 tggcgcggtg ctgtttctca atgccgtctg gttgctgggc aagatcagcg
 gtcgggaggt ggcggtgatc aacttcctgg tcggcgtgct gagcgcctgc
 gtcgcgttct acctgatctt ttccgcagca gccgggcagg gctcgctgaa
 ggccggagcg ctgaccctgc tattcgcttt tacctatctg tgggtggccg
 ccaaccagtt cctcgag

</pre>
</td></tr></table><p>





<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 

        If you wish to include the standard description of the avalable
	report formats, use:
        #include file="inc/reportformats.ihtml"
   -->

<b>trimspace</b> 
rewrites the input file.


<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: seqspace2.trimspace</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt;X13776 Pseudomonas aeruginosa amiC and amiR gene
ggtaccgctg gccgagcatc tgctcgatca ccaccagccg ggcgacggga
actgcacgat ctacctggcg agcctggagc acgagcgggt tcgcttcgta
cggcgctgag cgacagtcac aggagaggaa acggatggga tcgcaccagg
agcggccgct gatcggcctg ctgttctccg aaaccggcgt caccgccgat
atcgagcgct cgcacgcgta tggcgcattg ctcgcggtcg agcaactgaa
ccgcgagggc ggcgtcggcg gtcgcccgat cgaaacgctg tcccaggacc
ccggcggcga cccggaccgc tatcggctgt gcgccgagga cttcattcgc
aaccgggggg tacggttcct cgtgggctgc tacatgtcgc acacgcgcaa
ggcggtgatg ccggtggtcg agcgcgccga cgcgctgctc tgctacccga
ccccctacga gggcttcgag tattcgccga acatcgtcta cggcggtccg
gcgccgaacc agaacagtgc gccgctggcg gcgtacctga ttcgccacta
cggcgagcgg gtggtgttca tcggctcgga ctacatctat ccgcgggaaa
gcaaccatgt gatgcgccac ctgtatcgcc agcacggcgg cacggtgctc
gaggaaatct acattccgct gtatccctcc gacgacgact tgcagcgcgc
cgtcgagcgc atctaccagg cgcgcgccga cgtggtcttc tccaccgtgg
tgggcaccgg caccgccgag ctgtatcgcg ccatcgcccg tcgctacggc
gacggcaggc ggccgccgat cgccagcctg accaccagcg aggcggaggt
ggcgaagatg gagagtgacg tggcagaggg gcaggtggtg gtcgcgcctt
acttctccag catcgatacg cccgccagcc gggccttcgt ccaggcctgc
catggtttct tcccggagaa cgcgaccatc accgcctggg ccgaggcggc
ctactggcag accttgttgc tcggccgcgc cgcgcaggcc gcaggcaact
ggcgggtgga agacgtgcag cggcacctgt acgacatcga catcgacgcg
ccacaggggc cggtccgggt ggagcgccag aacaaccaca gccgcctgtc
ttcgcgcatc gcggaaatcg atgcgcgcgg cgtgttccag gtccgctggc
agtcgcccga accgattcgc cccgaccctt atgtcgtcgt gcataacctc
gacgactggt ccgccagcat gggcggggga ccgctcccat gagcgccaac
tcgctgctcg gcagcctgcg cgagttgcag gtgctggtcc tcaacccgcc
gggggaggtc agcgacgccc tggtcttgca gctgatccgc atcggttgtt
cggtgcgcca gtgctggccg ccgccggaag ccttcgacgt gccggtggac
gtggtcttca ccagcatttt ccagaatggc caccacgacg agatcgctgc
gctgctcgcc gccgggactc cgcgcactac cctggtggcg ctggtggagt
acgaaagccc cgcggtgctc tcgcagatca tcgagctgga gtgccacggc
gtgatcaccc agccgctcga tgcccaccgg gtgctgcctg tgctggtatc
ggcgcggcgc atcagcgagg aaatggcgaa gctgaagcag aagaccgagc
agctccagga ccgcatcgcc ggccaggccc ggatcaacca ggccaaggtg
ttgctgatgc agcgccatgg ctgggacgag cgcgaggcgc accagcacct
gtcgcgggaa gcgatgaagc ggcgcgagcc gatcctgaag atcgctcagg
agttgctggg aaacgagccg tccgcctgag cgatccgggc cgaccagaac
aataacaaga ggggtatcgt catcatgctg ggactggttc tgctgtacgt
tggcgcggtg ctgtttctca atgccgtctg gttgctgggc aagatcagcg
gtcgggaggt ggcggtgatc aacttcctgg tcggcgtgct gagcgcctgc
gtcgcgttct acctgatctt ttccgcagca gccgggcagg gctcgctgaa
ggccggagcg ctgaccctgc tattcgcttt tacctatctg tgggtggccg
ccaaccagtt cctcgag

</pre>
</td></tr></table><p>





<H2>
    Data files
</H2>

None.





<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
<ol>

<li>

</ol>

   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="aligncopy.html">aligncopy</a></td>
<td>Reads and writes alignments</td>
</tr>

<tr>
<td><a href="aligncopypair.html">aligncopypair</a></td>
<td>Reads and writes pairs from alignments</td>
</tr>

<tr>
<td><a href="biosed.html">biosed</a></td>
<td>Replace or delete sequence sections</td>
</tr>

<tr>
<td><a href="codcopy.html">codcopy</a></td>
<td>Copy and reformat a codon usage table</td>
</tr>

<tr>
<td><a href="cutseq.html">cutseq</a></td>
<td>Removes a section from a sequence</td>
</tr>

<tr>
<td><a href="degapseq.html">degapseq</a></td>
<td>Removes non-alphabetic (e.g. gap) characters from sequences</td>
</tr>

<tr>
<td><a href="descseq.html">descseq</a></td>
<td>Alter the name or description of a sequence</td>
</tr>

<tr>
<td><a href="entret.html">entret</a></td>
<td>Retrieves sequence entries from flatfile databases and files</td>
</tr>

<tr>
<td><a href="extractalign.html">extractalign</a></td>
<td>Extract regions from a sequence alignment</td>
</tr>

<tr>
<td><a href="extractfeat.html">extractfeat</a></td>
<td>Extract features from sequence(s)</td>
</tr>

<tr>
<td><a href="extractseq.html">extractseq</a></td>
<td>Extract regions from a sequence</td>
</tr>

<tr>
<td><a href="featcopy.html">featcopy</a></td>
<td>Reads and writes a feature table</td>
</tr>

<tr>
<td><a href="featreport.html">featreport</a></td>
<td>Reads and writes a feature table</td>
</tr>

<tr>
<td><a href="listor.html">listor</a></td>
<td>Write a list file of the logical OR of two sets of sequences</td>
</tr>

<tr>
<td><a href="makenucseq.html">makenucseq</a></td>
<td>Create random nucleotide sequences</td>
</tr>

<tr>
<td><a href="makeprotseq.html">makeprotseq</a></td>
<td>Create random protein sequences</td>
</tr>

<tr>
<td><a href="maskambignuc.html">maskambignuc</a></td>
<td>Masks all ambiguity characters in nucleotide sequences with N</td>
</tr>

<tr>
<td><a href="maskambigprot.html">maskambigprot</a></td>
<td>Masks all ambiguity characters in protein sequences with X</td>
</tr>

<tr>
<td><a href="maskfeat.html">maskfeat</a></td>
<td>Write a sequence with masked features</td>
</tr>

<tr>
<td><a href="maskseq.html">maskseq</a></td>
<td>Write a sequence with masked regions</td>
</tr>

<tr>
<td><a href="newseq.html">newseq</a></td>
<td>Create a sequence file from a typed-in sequence</td>
</tr>

<tr>
<td><a href="nohtml.html">nohtml</a></td>
<td>Remove mark-up (e.g. HTML tags) from an ASCII text file</td>
</tr>

<tr>
<td><a href="noreturn.html">noreturn</a></td>
<td>Remove carriage return from ASCII files</td>
</tr>

<tr>
<td><a href="nospace.html">nospace</a></td>
<td>Remove all whitespace from an ASCII text file</td>
</tr>

<tr>
<td><a href="notab.html">notab</a></td>
<td>Replace tabs with spaces in an ASCII text file</td>
</tr>

<tr>
<td><a href="notseq.html">notseq</a></td>
<td>Write to file a subset of an input stream of sequences</td>
</tr>

<tr>
<td><a href="nthseq.html">nthseq</a></td>
<td>Write to file a single sequence from an input stream of sequences</td>
</tr>

<tr>
<td><a href="pasteseq.html">pasteseq</a></td>
<td>Insert one sequence into another</td>
</tr>

<tr>
<td><a href="revseq.html">revseq</a></td>
<td>Reverse and complement a nucleotide sequence</td>
</tr>

<tr>
<td><a href="seqret.html">seqret</a></td>
<td>Reads and writes (returns) sequences</td>
</tr>

<tr>
<td><a href="seqretsplit.html">seqretsplit</a></td>
<td>Reads sequences and writes them to individual files</td>
</tr>

<tr>
<td><a href="sizeseq.html">sizeseq</a></td>
<td>Sort sequences by size</td>
</tr>

<tr>
<td><a href="skipredundant.html">skipredundant</a></td>
<td>Remove redundant sequences from an input set</td>
</tr>

<tr>
<td><a href="skipseq.html">skipseq</a></td>
<td>Reads and writes (returns) sequences, skipping first few</td>
</tr>

<tr>
<td><a href="splitter.html">splitter</a></td>
<td>Split sequence(s) into smaller sequences</td>
</tr>

<tr>
<td><a href="trimest.html">trimest</a></td>
<td>Remove poly-A tails from nucleotide sequences</td>
</tr>

<tr>
<td><a href="trimseq.html">trimseq</a></td>
<td>Remove unwanted characters from start and end of sequence(s)</td>
</tr>

<tr>
<td><a href="union.html">union</a></td>
<td>Concatenate multiple sequences into a single sequence</td>
</tr>

<tr>
<td><a href="vectorstrip.html">vectorstrip</a></td>
<td>Removes vectors from the ends of nucleotide sequence(s)</td>
</tr>

<tr>
<td><a href="yank.html">yank</a></td>
<td>Add a sequence reference (a full USA) to a list file</td>
</tr>

</table>
<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>

Jon Ison (jison&nbsp;&copy;&nbsp;ebi.ac.uk)
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK




<H2>
    History
</H2>
<!--
        Date written and what changes have been made go in this file.
   -->




<H2>
    Target users
</H2>
<!--
        For general users, use this text
   -->
This program is intended to be used by everyone and everything, from naive users to embedded scripts.

<H2>
    Comments
</H2>
<!--
        User/developer/other comments go in this file.
   -->
None


</BODY>
</HTML>

