<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: twofeat
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

twofeat
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<!--END OF HEADER-->






<H2>
    Function
</H2>
Finds neighbouring pairs of features in sequences

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
  -->

<b>twofeat</b> reads in the feature tables of sequences and reports
occurances of pairs of specified features. 

<p>

This is intended for use as a simple data-mining tool to enable you to
look for instances of pairs of features that occur near to each other in
the same sequence entry. 

<p>

For each of the pair of features, you can specify its type name, its
sense, its score and any tag/value pairs, amongst other things. 

<p>

You can then specify the type of relationship that the two features
should have.  You can specify the minimum and maximum distance between
them.  You can specify the type of overlap allowed: Any type of overlap
or no overlap is allowed, Overlap required, No overlaps are allowed,
Overlap required but one feature must not be completely within the
other, Feature A must be completely enclosed within feature B, Feature B
must be completely enclosed within feature A.  You can specify that the
distance should be measured from the nearest ends of the two features,
From the left ends, From right ends, From the furthest ends.  You can
specify that the features should be in any sense, in the same sense or
in opposite senses.  You can specify that the features should be in any
order, Feature A then feature B, Feature B then feature A. 

<p>

By default the resulting pairs of features found are then written to a
report file as a single feature from the first postion of the left-most
feature to the last position of the right-most feature.  You can modify
the output to report the pairs of features with no changes made to them. 







<H2>
    Algorithm
</H2>

<!-- 
        Algorithms
	  Pseudocode/Vague description/Handwaving/web links to useful pages
  -->

<pre>
For each sequence:
	identify the features that match the criteria for Feature A
	identify the features that match the criteria for Feature B
	compare all pairs of features
	if they satisfy the requirements output them to the report file
</pre>










<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

<b>Here is a sample session with twofeat</b>
<p>
Without the -twoout option the output report contains just one feature for every pair of features found. 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>twofeat tembl:hsfau </b>
Finds neighbouring pairs of features in sequences
Type of first feature [*]: <b>polyA_signal</b>
Type of second feature [*]: <b>polyA_site</b>
The minimum distance between the features [0]: <b>10</b>
The maximum distance between the features [0]: <b>50</b>
Output report [hsfau.twofeat]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>
<p>
<b>Example 2</b>
<p>
The -twoout option makes the output report give the pairs of features found. The ouput report format type has been changed to display the full original pairs of faetures in GFF format. 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>twofeat -twoout -rformat gff tembl:hsfau </b>
Finds neighbouring pairs of features in sequences
Type of first feature [*]: <b>polyA_signal</b>
Type of second feature [*]: <b>polyA_site</b>
The minimum distance between the features [0]: <b>10</b>
The maximum distance between the features [0]: <b>50</b>
Output report [hsfau.twofeat]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.2">Go to the output files for this example</a><p><p>




<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          seqall     Sequence database USA
   -atype              string     By default every feature in the feature
                                  table is allowed. You can set this to be any
                                  feature type you wish to allow.
                                  See http://www3.ebi.ac.uk/Services/WebFeat/
                                  for a list of the EMBL feature types and see
                                  Appendix A of the Swissprot user manual in
                                  http://www.expasy.ch/txt/userman.txt for a
                                  list of the Swissprot feature types.
                                  The type may be wildcarded by using '*'.
                                  If you wish to allow more than one type,
                                  separate their names with the character '|',
                                  eg:
                                  *UTR | intron
   -btype              string     By default every feature in the feature
                                  table is allowed. You can set this to be any
                                  feature type you wish to allow.
                                  See http://www3.ebi.ac.uk/Services/WebFeat/
                                  for a list of the EMBL feature types and see
                                  Appendix A of the Swissprot user manual in
                                  http://www.expasy.ch/txt/userman.txt for a
                                  list of the Swissprot feature types.
                                  The type may be wildcarded by using '*'.
                                  If you wish to allow more than one type,
                                  separate their names with the character '|',
                                  eg:
                                  *UTR | intron
   -minrange           integer    If this is greater or equal to 'maxrange',
                                  then no min or max range is specified
   -maxrange           integer    If this is less than or equal to 'minrange',
                                  then no min or max range is specified
  [-outfile]           report     File for output of neighbouring feature
                                  regions. This contains details of the pairs
                                  of features. By default, it is written in
                                  TABLE format.

   Optional qualifiers (* if not always prompted):
   -asource            string     By default any feature source in the feature
                                  table is allowed. You can set this to match
                                  any feature source you wish to allow.
                                  The source name is usually either the name
                                  of the program that detected the feature or
                                  it is the feature table (eg: EMBL) that the
                                  feature came from.
                                  The source may be wildcarded by using '*'.
                                  If you wish to allow more than one source,
                                  separate their names with the character '|',
                                  eg:
                                  gene* | embl
   -asense             menu       By default any feature sense is allowed. You
                                  can set this to match the required sense.
   -aminscore          float      If this is greater than or equal to the
                                  maximum score, then any score is allowed.
   -amaxscore          float      If this is less than or equal to the maximum
                                  score, then any score is permitted.
   -atag               string     Tags are the types of extra values that a
                                  feature may have. For example in the EMBL
                                  feature table, a 'CDS' type of feature may
                                  have the tags '/codon', '/codon_start',
                                  '/db_xref', '/EC_number', '/evidence',
                                  '/exception', '/function', '/gene',
                                  '/label', '/map', '/note', '/number',
                                  '/partial', '/product', '/protein_id',
                                  '/pseudo', '/standard_name', '/translation',
                                  '/transl_except', '/transl_table', or
                                  '/usedin'. Some of these tags also have
                                  values, for example '/gene' can have the
                                  value of the gene name.
                                  By default any feature tag in the feature
                                  table is allowed. You can set this to match
                                  any feature tag you wish to allow.
                                  The tag may be wildcarded by using '*'.
                                  If you wish to allow more than one tag,
                                  separate their names with the character '|',
                                  eg:
                                  gene | label
   -avalue             string     Tag values are the values associated with a
                                  feature tag. Tags are the types of extra
                                  values that a feature may have. For example
                                  in the EMBL feature table, a 'CDS' type of
                                  feature may have the tags '/codon',
                                  '/codon_start', '/db_xref', '/EC_number',
                                  '/evidence', '/exception', '/function',
                                  '/gene', '/label', '/map', '/note',
                                  '/number', '/partial', '/product',
                                  '/protein_id', '/pseudo', '/standard_name',
                                  '/translation', '/transl_except',
                                  '/transl_table', or '/usedin'. Only some of
                                  these tags can have values, for example
                                  '/gene' can have the value of the gene name.
                                  By default any feature tag value in the
                                  feature table is allowed. You can set this
                                  to match any feature tag value you wish to
                                  allow.
                                  The tag value may be wildcarded by using
                                  '*'.
                                  If you wish to allow more than one tag
                                  value, separate their names with the
                                  character '|', eg:
                                  pax* | 10
   -bsource            string     By default any feature source in the feature
                                  table is allowed. You can set this to match
                                  any feature source you wish to allow.
                                  The source name is usually either the name
                                  of the program that detected the feature or
                                  it is the feature table (eg: EMBL) that the
                                  feature came from.
                                  The source may be wildcarded by using '*'.
                                  If you wish to allow more than one source,
                                  separate their names with the character '|',
                                  eg:
                                  gene* | embl
   -bsense             menu       By default any feature sense is allowed. You
                                  can set this to match the required sense.
   -bminscore          float      If this is greater than or equal to the
                                  maximum score, then any score is allowed.
   -bmaxscore          float      If this is less than or equal to the maximum
                                  score, then any score is permitted.
   -btag               string     Tags are the types of extra values that a
                                  feature may have. For example in the EMBL
                                  feature table, a 'CDS' type of feature may
                                  have the tags '/codon', '/codon_start',
                                  '/db_xref', '/EC_number', '/evidence',
                                  '/exception', '/function', '/gene',
                                  '/label', '/map', '/note', '/number',
                                  '/partial', '/product', '/protein_id',
                                  '/pseudo', '/standard_name', '/translation',
                                  '/transl_except', '/transl_table', or
                                  '/usedin'. Some of these tags also have
                                  values, for example '/gene' can have the
                                  value of the gene name.
                                  By default any feature tag in the feature
                                  table is allowed. You can set this to match
                                  any feature tag you wish to allow.
                                  The tag may be wildcarded by using '*'.
                                  If you wish to allow more than one tag,
                                  separate their names with the character '|',
                                  eg:
                                  gene | label
   -bvalue             string     Tag values are the values associated with a
                                  feature tag. Tags are the types of extra
                                  values that a feature may have. For example
                                  in the EMBL feature table, a 'CDS' type of
                                  feature may have the tags '/codon',
                                  '/codon_start', '/db_xref', '/EC_number',
                                  '/evidence', '/exception', '/function',
                                  '/gene', '/label', '/map', '/note',
                                  '/number', '/partial', '/product',
                                  '/protein_id', '/pseudo', '/standard_name',
                                  '/translation', '/transl_except',
                                  '/transl_table', or '/usedin'. Only some of
                                  these tags can have values, for example
                                  '/gene' can have the value of the gene name.
                                  By default any feature tag value in the
                                  feature table is allowed. You can set this
                                  to match any feature tag value you wish to
                                  allow.
                                  The tag value may be wildcarded by using
                                  '*'.
                                  If you wish to allow more than one tag
                                  value, separate their names with the
                                  character '|', eg:
                                  pax* | 10
   -overlap            menu       This allows you to specify the allowed
                                  overlaps of the features A and B.
                                  You can allow any or no overlaps, specify
                                  that they must or must not overlap, that one
                                  must or must not be wholly enclosed within
                                  another feature.
   -rangetype          menu       This allows you to specify the positions
                                  from which the allowed minimum or maximum
                                  distance between the features is measured
   -sense              menu       This allows you to specify the required
                                  sense that the two features must be on. This
                                  is ignored (always 'Any') when looking at
                                  protein sequence features.
   -order              menu       This allows you to specify the required
                                  order of the two features. The order is
                                  measured from the start positions of the
                                  features. This criterion is always applied
                                  despite the specified overlap type required.
   -twoout             boolean    If you set this to be true, then the two
                                  features themselves will be written out. If
                                  it is left as false, then a single feature
                                  will be written out covering the two
                                  features you found.
*  -typeout            string     If you have specified that the pairs of
                                  features that are found should be reported
                                  as one feature in the ouput, then you can
                                  specify the 'type' name of the new feature
                                  here. By default every feature in the
                                  feature table is allowed. See
                                  http://www3.ebi.ac.uk/Services/WebFeat/ for
                                  a list of the EMBL feature types and see
                                  Appendix A of the Swissprot user manual in
                                  http://www.expasy.ch/txt/userman.txt for a
                                  list of the Swissprot feature types. If you
                                  specify an invalid feature type name, then
                                  the default name 'misc_feature' is used.

   Advanced qualifiers: (none)
   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-atype</td>
<td>By default every feature in the feature table is allowed. You can set this to be any feature type you wish to allow.
See http://www3.ebi.ac.uk/Services/WebFeat/ for a list of the EMBL feature types and see Appendix A of the Swissprot user manual in http://www.expasy.ch/txt/userman.txt for a list of the Swissprot feature types.
The type may be wildcarded by using '*'.
If you wish to allow more than one type, separate their names with the character '|', eg:
*UTR | intron</td>
<td>Any string is accepted</td>
<td>*</td>
</tr>

<tr>
<td>-btype</td>
<td>By default every feature in the feature table is allowed. You can set this to be any feature type you wish to allow.
See http://www3.ebi.ac.uk/Services/WebFeat/ for a list of the EMBL feature types and see Appendix A of the Swissprot user manual in http://www.expasy.ch/txt/userman.txt for a list of the Swissprot feature types.
The type may be wildcarded by using '*'.
If you wish to allow more than one type, separate their names with the character '|', eg:
*UTR | intron</td>
<td>Any string is accepted</td>
<td>*</td>
</tr>

<tr>
<td>-minrange</td>
<td>If this is greater or equal to 'maxrange', then no min or max range is specified</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr>
<td>-maxrange</td>
<td>If this is less than or equal to 'minrange', then no min or max range is specified</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>File for output of neighbouring feature regions. This contains details of the pairs of features. By default, it is written in TABLE format.</td>
<td>Report output file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-asource</td>
<td>By default any feature source in the feature table is allowed. You can set this to match any feature source you wish to allow.
The source name is usually either the name of the program that detected the feature or it is the feature table (eg: EMBL) that the feature came from.
The source may be wildcarded by using '*'.
If you wish to allow more than one source, separate their names with the character '|', eg:
gene* | embl</td>
<td>Any string is accepted</td>
<td>*</td>
</tr>

<tr>
<td>-asense</td>
<td>By default any feature sense is allowed. You can set this to match the required sense.</td>
<td><table><tr><td>0</td> <td><i>(Any sense)</i></td></tr><tr><td>+</td> <td><i>(Forward sense)</i></td></tr><tr><td>-</td> <td><i>(Reverse sense)</i></td></tr></table></td>
<td>0</td>
</tr>

<tr>
<td>-aminscore</td>
<td>If this is greater than or equal to the maximum score, then any score is allowed.</td>
<td>Any numeric value</td>
<td>0.0</td>
</tr>

<tr>
<td>-amaxscore</td>
<td>If this is less than or equal to the maximum score, then any score is permitted.</td>
<td>Any numeric value</td>
<td>0.0</td>
</tr>

<tr>
<td>-atag</td>
<td>Tags are the types of extra values that a feature may have. For example in the EMBL feature table, a 'CDS' type of feature may have the tags '/codon', '/codon_start', '/db_xref', '/EC_number', '/evidence', '/exception', '/function', '/gene', '/label', '/map', '/note', '/number', '/partial', '/product', '/protein_id', '/pseudo', '/standard_name', '/translation', '/transl_except', '/transl_table', or '/usedin'. Some of these tags also have values, for example '/gene' can have the value of the gene name.
By default any feature tag in the feature table is allowed. You can set this to match any feature tag you wish to allow.
The tag may be wildcarded by using '*'.
If you wish to allow more than one tag, separate their names with the character '|', eg:
gene | label</td>
<td>Any string is accepted</td>
<td>*</td>
</tr>

<tr>
<td>-avalue</td>
<td>Tag values are the values associated with a feature tag. Tags are the types of extra values that a feature may have. For example in the EMBL feature table, a 'CDS' type of feature may have the tags '/codon', '/codon_start', '/db_xref', '/EC_number', '/evidence', '/exception', '/function', '/gene', '/label', '/map', '/note', '/number', '/partial', '/product', '/protein_id', '/pseudo', '/standard_name', '/translation', '/transl_except', '/transl_table', or '/usedin'. Only some of these tags can have values, for example '/gene' can have the value of the gene name. By default any feature tag value in the feature table is allowed. You can set this to match any feature tag value you wish to allow.
The tag value may be wildcarded by using '*'.
If you wish to allow more than one tag value, separate their names with the character '|', eg:
pax* | 10</td>
<td>Any string is accepted</td>
<td>*</td>
</tr>

<tr>
<td>-bsource</td>
<td>By default any feature source in the feature table is allowed. You can set this to match any feature source you wish to allow.
The source name is usually either the name of the program that detected the feature or it is the feature table (eg: EMBL) that the feature came from.
The source may be wildcarded by using '*'.
If you wish to allow more than one source, separate their names with the character '|', eg:
gene* | embl</td>
<td>Any string is accepted</td>
<td>*</td>
</tr>

<tr>
<td>-bsense</td>
<td>By default any feature sense is allowed. You can set this to match the required sense.</td>
<td><table><tr><td>0</td> <td><i>(Any sense)</i></td></tr><tr><td>+</td> <td><i>(Forward sense)</i></td></tr><tr><td>-</td> <td><i>(Reverse sense)</i></td></tr></table></td>
<td>0</td>
</tr>

<tr>
<td>-bminscore</td>
<td>If this is greater than or equal to the maximum score, then any score is allowed.</td>
<td>Any numeric value</td>
<td>0.0</td>
</tr>

<tr>
<td>-bmaxscore</td>
<td>If this is less than or equal to the maximum score, then any score is permitted.</td>
<td>Any numeric value</td>
<td>0.0</td>
</tr>

<tr>
<td>-btag</td>
<td>Tags are the types of extra values that a feature may have. For example in the EMBL feature table, a 'CDS' type of feature may have the tags '/codon', '/codon_start', '/db_xref', '/EC_number', '/evidence', '/exception', '/function', '/gene', '/label', '/map', '/note', '/number', '/partial', '/product', '/protein_id', '/pseudo', '/standard_name', '/translation', '/transl_except', '/transl_table', or '/usedin'. Some of these tags also have values, for example '/gene' can have the value of the gene name.
By default any feature tag in the feature table is allowed. You can set this to match any feature tag you wish to allow.
The tag may be wildcarded by using '*'.
If you wish to allow more than one tag, separate their names with the character '|', eg:
gene | label</td>
<td>Any string is accepted</td>
<td>*</td>
</tr>

<tr>
<td>-bvalue</td>
<td>Tag values are the values associated with a feature tag. Tags are the types of extra values that a feature may have. For example in the EMBL feature table, a 'CDS' type of feature may have the tags '/codon', '/codon_start', '/db_xref', '/EC_number', '/evidence', '/exception', '/function', '/gene', '/label', '/map', '/note', '/number', '/partial', '/product', '/protein_id', '/pseudo', '/standard_name', '/translation', '/transl_except', '/transl_table', or '/usedin'. Only some of these tags can have values, for example '/gene' can have the value of the gene name. By default any feature tag value in the feature table is allowed. You can set this to match any feature tag value you wish to allow.
The tag value may be wildcarded by using '*'.
If you wish to allow more than one tag value, separate their names with the character '|', eg:
pax* | 10</td>
<td>Any string is accepted</td>
<td>*</td>
</tr>

<tr>
<td>-overlap</td>
<td>This allows you to specify the allowed overlaps of the features A and B.
You can allow any or no overlaps, specify that they must or must not overlap, that one must or must not be wholly enclosed within another feature.</td>
<td><table><tr><td>A</td> <td><i>(Any)</i></td></tr><tr><td>O</td> <td><i>(Overlap required)</i></td></tr><tr><td>NO</td> <td><i>(No overlaps are allowed)</i></td></tr><tr><td>NW</td> <td><i>(Overlap required but not within)</i></td></tr><tr><td>AW</td> <td><i>(A must be all within B)</i></td></tr><tr><td>BW</td> <td><i>(B must be all within A)</i></td></tr></table></td>
<td>A</td>
</tr>

<tr>
<td>-rangetype</td>
<td>This allows you to specify the positions from which the allowed minimum or maximum distance between the features is measured</td>
<td><table><tr><td>N</td> <td><i>(From nearest ends)</i></td></tr><tr><td>L</td> <td><i>(From left ends)</i></td></tr><tr><td>R</td> <td><i>(From right ends)</i></td></tr><tr><td>F</td> <td><i>(From furthest ends)</i></td></tr></table></td>
<td>N</td>
</tr>

<tr>
<td>-sense</td>
<td>This allows you to specify the required sense that the two features must be on. This is ignored (always 'Any') when looking at protein sequence features.</td>
<td><table><tr><td>A</td> <td><i>(Any sense)</i></td></tr><tr><td>S</td> <td><i>(Same sense)</i></td></tr><tr><td>O</td> <td><i>(Opposite sense)</i></td></tr></table></td>
<td>A</td>
</tr>

<tr>
<td>-order</td>
<td>This allows you to specify the required order of the two features. The order is measured from the start positions of the features. This criterion is always applied despite the specified overlap type required.</td>
<td><table><tr><td>A</td> <td><i>(Any)</i></td></tr><tr><td>AB</td> <td><i>(Feature A then feature B)</i></td></tr><tr><td>BA</td> <td><i>(Feature B then feature A)</i></td></tr></table></td>
<td>A</td>
</tr>

<tr>
<td>-twoout</td>
<td>If you set this to be true, then the two features themselves will be written out. If it is left as false, then a single feature will be written out covering the two features you found.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-typeout</td>
<td>If you have specified that the pairs of features that are found should be reported as one feature in the ouput, then you can specify the 'type' name of the new feature here. By default every feature in the feature table is allowed. See http://www3.ebi.ac.uk/Services/WebFeat/ for a list of the EMBL feature types and see Appendix A of the Swissprot user manual in http://www.expasy.ch/txt/userman.txt for a list of the Swissprot feature types. If you specify an invalid feature type name, then the default name 'misc_feature' is used.</td>
<td>Any string is accepted</td>
<td>misc_feature</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>twofeat</b> reads any normal sequence USAs.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tembl:hsfau' is a sequence entry in the example nucleic acid database 'tembl'
<p>
<p><h3>Database entry: tembl:hsfau</h3>
<table width="90%"><tr><td bgcolor="#FFDDFF">
<pre>
ID   HSFAU      standard; RNA; HUM; 518 BP.
XX
AC   X65923;
XX
SV   X65923.1
XX
DT   13-MAY-1992 (Rel. 31, Created)
DT   23-SEP-1993 (Rel. 37, Last updated, Version 10)
XX
DE   H.sapiens fau mRNA
XX
KW   fau gene.
XX
OS   Homo sapiens (human)
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia;
OC   Eutheria; Primates; Catarrhini; Hominidae; Homo.
XX
RN   [1]
RP   1-518
RA   Michiels L.M.R.;
RT   ;
RL   Submitted (29-APR-1992) to the EMBL/GenBank/DDBJ databases.
RL   L.M.R. Michiels, University of Antwerp, Dept of Biochemistry,
RL   Universiteisplein 1, 2610 Wilrijk, BELGIUM
XX
RN   [2]
RP   1-518
RX   MEDLINE; 93368957.
RA   Michiels L., Van der Rauwelaert E., Van Hasselt F., Kas K., Merregaert J.;
RT   " fau cDNA encodes a ubiquitin-like-S30 fusion protein and is expressed as
RT   an antisense sequences in the Finkel-Biskis-Reilly murine sarcoma virus";
RL   Oncogene 8:2537-2546(1993).
XX
DR   SWISS-PROT; P35544; UBIM_HUMAN.
DR   SWISS-PROT; Q05472; RS30_HUMAN.
XX
FH   Key             Location/Qualifiers
FH
FT   source          1..518
FT                   /chromosome="11q"
FT                   /db_xref="taxon:9606"
FT                   /organism="Homo sapiens"
FT                   /tissue_type="placenta"
FT                   /clone_lib="cDNA"
FT                   /clone="pUIA 631"
FT                   /map="13"
FT   misc_feature    57..278
FT                   /note="ubiquitin like part"
FT   CDS             57..458
FT                   /db_xref="SWISS-PROT:P35544"
FT                   /db_xref="SWISS-PROT:Q05472"
FT                   /gene="fau"
FT                   /protein_id="CAA46716.1"
FT                   /translation="MQLFVRAQELHTFEVTGQETVAQIKAHVASLEGIAPEDQVVLLAG
FT                   APLEDEATLGQCGVEALTTLEVAGRMLGGKVHGSLARAGKVRGQTPKVAKQEKKKKKTG
FT                   RAKRRMQYNRRFVNVVPTFGKKKGPNANS"
FT   misc_feature    98..102
FT                   /note="nucleolar localization signal"
FT   misc_feature    279..458
FT                   /note="S30 part"
FT   polyA_signal    484..489
FT   polyA_site      509
XX
SQ   Sequence 518 BP; 125 A; 139 C; 148 G; 106 T; 0 other;
     ttcctctttc tcgactccat cttcgcggta gctgggaccg ccgttcagtc gccaatatgc        60
     agctctttgt ccgcgcccag gagctacaca ccttcgaggt gaccggccag gaaacggtcg       120
     cccagatcaa ggctcatgta gcctcactgg agggcattgc cccggaagat caagtcgtgc       180
     tcctggcagg cgcgcccctg gaggatgagg ccactctggg ccagtgcggg gtggaggccc       240
     tgactaccct ggaagtagca ggccgcatgc ttggaggtaa agttcatggt tccctggccc       300
     gtgctggaaa agtgagaggt cagactccta aggtggccaa acaggagaag aagaagaaga       360
     agacaggtcg ggctaagcgg cggatgcagt acaaccggcg ctttgtcaac gttgtgccca       420
     cctttggcaa gaagaagggc cccaatgcca actcttaagt cttttgtaat tctggctttc       480
     tctaataaaa aagccactta gttcagtcaa aaaaaaaa                               518
//
</pre>
</td></tr></table><p>






<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 

        If you wish to include the standard description of the avalable
	report formats, use:
   -->

<p>

The output is a standard EMBOSS report file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-rformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  The available format names are: embl,
genbank, gff, pir, swiss, trace, listfile, dbmotif, diffseq, excel,
feattable, motif, regions, seqtable, simple, srs, table, tagseq

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html</A>
for further information on report formats.

<p>


<b>twofeat</b> 
outputs a report format file. The default format is <b>table</b>


<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: hsfau.twofeat</h3>
<table width="90%"><tr><td bgcolor="#DDFFDD">
<pre>
########################################
# Program: twofeat
# Rundate: Thu Jan 09 13:13:50 2003
# Report_format: table
# Report_file: hsfau.twofeat
########################################

#=======================================
#
# Sequence: HSFAU     from: 1   to: 518
# HitCount: 1
#=======================================

  Start     End   Score starta   enda startb   endb
    484     509   0.000    484    489    509    509

#---------------------------------------
#---------------------------------------
</pre>
</td></tr></table><p>

<a name="output.2"></a>
<h3>Output files for usage example 2</h3>
<p><h3>File: hsfau.twofeat</h3>
<table width="90%"><tr><td bgcolor="#DDFFDD">
<pre>
##gff-version 2.0
##date 2003-01-09
##Type DNA HSFAU
HSFAU	EMBL	polyA_signal	484	489	0.000	+	.	Sequence "HSFAU.6"
HSFAU	EMBL	polyA_site	509	509	0.000	+	.	Sequence "HSFAU.7" ; FeatFlags "0x40"
</pre>
</td></tr></table><p>






<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.

	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

   -->

None.













<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

It can't find features that are not in the input sequences.
It has no way of checking whether the input features are correct or not.
Remember this when you are searching public databases.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

There is a slight memory leak that must be fixed at some time.
This does no affect the results.







<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="coderet.html">coderet</a></td><td>Extract CDS, mRNA and translations from feature tables</td></tr>
<tr><td><a href="extractfeat.html">extractfeat</a></td><td>Extract features from a sequence</td></tr>
<tr><td><a href="maskfeat.html">maskfeat</a></td><td>Mask off features of a sequence</td></tr>
<tr><td><a href="showfeat.html">showfeat</a></td><td>Show features of a sequence</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

This application was written by Gary Williams <a
href="mailto:gwilliam@hgmp.mrc.ac.uk">(gwilliam@hgmp.mrc.ac.uk)</a>















<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

Written (2002) - Gary Williams












<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












</BODY>
</HTML>

