
<HTML>

<HEAD>
  <TITLE>
  EMBOSS
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">


<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

UTILS DATABASE CREATION
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<!--END OF HEADER-->





Database installation
<p>

<table border cellpadding=4 bgcolor="#FFFFF0">

<tr><th>Program name</th><th>Description</th></tr>

<tr>
<td><a href="aaindexextract.html">aaindexextract</a></td>
<td>
Extract data from AAINDEX
</td>
</tr>

<tr>
<td><a href="cutgextract.html">cutgextract</a></td>
<td>
Extract data from CUTG
</td>
</tr>

<tr>
<td><a href="printsextract.html">printsextract</a></td>
<td>
Extract data from PRINTS
</td>
</tr>

<tr>
<td><a href="prosextract.html">prosextract</a></td>
<td>
Builds the PROSITE motif database for patmatmotifs to search
</td>
</tr>

<tr>
<td><a href="rebaseextract.html">rebaseextract</a></td>
<td>
Extract data from REBASE
</td>
</tr>

<tr>
<td><a href="tfextract.html">tfextract</a></td>
<td>
Extract data from TRANSFAC
</td>
</tr>

<tr>
<td><a href="../embassy/domainatrix/allversusall.html">allversusall</a></td>
<td>
Does an all-versus-all global alignment for each set of sequences in an input directory and writes files of sequence similarity values
</td>
</tr>

<tr>
<td><a href="../embassy/domainatrix/cathparse.html">cathparse</a></td>
<td>
Reads raw CATH classification files and writes DCF file (domain classification file)
</td>
</tr>

<tr>
<td><a href="../embassy/domainatrix/domainer.html">domainer</a></td>
<td>
Reads CCF files (clean coordinate files) for proteins and writes CCF files for domains, taken from a DCF file (domain classification file)
</td>
</tr>

<tr>
<td><a href="../embassy/domainatrix/domainnr.html">domainnr</a></td>
<td>
Removes redundant domains from a DCF file (domain classification file). The file must contain domain sequence information, which can be added by using DOMAINSEQS
</td>
</tr>

<tr>
<td><a href="../embassy/domainatrix/domainseqs.html">domainseqs</a></td>
<td>
Adds sequence records to a DCF file (domain classification file)
</td>
</tr>

<tr>
<td><a href="../embassy/domainatrix/domainsse.html">domainsse</a></td>
<td>
Adds secondary structure records to a DCF file (domain classification file)
</td>
</tr>

<tr>
<td><a href="../embassy/domainatrix/hetparse.html">hetparse</a></td>
<td>
Converts raw dictionary of heterogen groups to a file in EMBL-like format
</td>
</tr>

<tr>
<td><a href="../embassy/domainatrix/pdbparse.html">pdbparse</a></td>
<td>
Parses PDB files and writes CCF files (clean coordinate files) for proteins
</td>
</tr>

<tr>
<td><a href="../embassy/domainatrix/pdbplus.html">pdbplus</a></td>
<td>
Add residue solvent accessibility and secondary structure data to a CCF file (clean coordinate file) for a protein or domain
</td>
</tr>

<tr>
<td><a href="../embassy/domainatrix/pdbtosp.html">pdbtosp</a></td>
<td>
Convert raw swissprot:PDB equivalence file to EMBL-like format
</td>
</tr>

<tr>
<td><a href="../embassy/domainatrix/scopparse.html">scopparse</a></td>
<td>
Reads raw SCOP classification files and writes a DCF file (domain classification file)
</td>
</tr>

<tr>
<td><a href="../embassy/domainatrix/seqnr.html">seqnr</a></td>
<td>
Removes redundancy from DHF files (domain hits files) or other files of sequences
</td>
</tr>

<tr>
<td><a href="../embassy/domainatrix/sites.html">sites</a></td>
<td>
Reads CCF files (clean coordinate files) and writes CON files (contact files) of residue-ligand contact data for domains in a DCF file (domain classification file)
</td>
</tr>

<tr>
<td><a href="../embassy/domainatrix/ssematch.html">ssematch</a></td>
<td>
Searches a DCF file (domain classification file) for secondary structure matches
</td>
</tr>


</table>



</BODY>
</HTML>
