<!--
Edit this file to replace 'vectorstrip' with the name of the program.
-->

<HTML>

<HEAD>
  <TITLE>
  EMBOSS
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: vectorstrip
</H2>
</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>

<!--END OF HEADER-->





<H1>
Program vectorstrip
</H1>
<H2>
    Function
</H2>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Strips out DNA between a pair of vector sequences


<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.

        This is what the user looks at to find out if this program will
        do what he wants. 

        Long description.

        Why it is useful.

        When it should/shouldn't be used.

        Algorithms

        Expanded description.

        eg: Reverse engineers Wheedle's 'talpsgob' program. This does a 
        splurge of the 5' end of the sequence followed by a durm.
        The net effect is a 5' end splurt.
        etc. for as long as is required (several pages is not unreasonable)
  -->

<b>vectorstrip</b> is intended to be useful for stripping vector sequence
from the ends of sequences of interest. For example, if a fragment has
been cloned into a vector and then sequenced, the sequence may contain
vector data eg from the cloning polylinker at the 5' and 3' ends of
the sequence.  <b>vectorstrip</b> will remove these contaminating regions and
output trimmed sequence ready for input into another application.

<p>

<b>vectorstrip</b> is suitable for use with low quality sequence data as it
can allow for mismatches between the sequence and the vector patterns
provided. You can specify the maximum level of mismatch
expected.
<p>
Vector data can either be provided in a file or interactively. If
presented in a file, <b>vectorstrip</b> will search all input sequences with
all vectors listed in that file. The intention is that the user can
maintain a single file for use with <b>vectorstrip</b>, containing all the
linker sequences commonly used in the laboratory.

<p>
The two patterns for each vector are searched separately against the
sequence. Once the search is completed, each of the hits of the 5'
sequence is paired with each of the hits of the 3' sequence and the
resulting subsequences are output. For example, if the 5' sequence
matches the sequence from (a) position 30-60, and(b)position 70-100,
and the 3' sequence matches from 150-175, then two subsequences will
be output: from 61-149, and from 101-149. The lower the quality of
the sequence, the more likely multiple hits become if nonzero
mismatches are accepted.

<p>
Default behaviour is to report only the best matches between the
vector patterns and the sequence. This means that if you
specify a maximum mismatch level of 10%, but the vector patterns
match the sequence with zero mismatches, the search will stop and the
program will output only these "best" matches. If there are no perfect
matches, the program will try searching again allowing 1 mismatch,
then 2, and so on until either the patterns match the sequence or the
maximum specified mismatch level is exceeded. You can tell <b>vectorstrip</b>
to show all possible matches up to your specified maximum level, as
illustrated in the examples below.

<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->
Here are several examples of running <b>vectorstrip</b>. The vectorfile and 
the sequence files used in these example are given <a href="#input">below</a>.
In each case, the same fragment has been cloned into the XhoI site of 
the polylinker of each vector. The cloned fragment is represented in 
lower case and the vector sequence in upper case so the sequence 
trimming can be readily seen.

<ol>
<li> Running <b>vectorstrip</b> on a list of sequences with default parameters
<pre>
unix % vectorstrip @seqs.list
Strips out DNA between a pair of vector sequences
Are your vector sequences in a file? [Y]: 
Name of vectorfile: vectors
Max allowed % mismatch [10]: 
Show only the best hits (minimise mismatches)? [Y]: 
Output file [pbluescript.vectorstrip]:stdout 
Output sequence [pbluescript.fasta]: 

Sequence: pBlueScript    Vector: pTYB1  No match


Sequence: pBlueScript    Vector: pBS_KS+
5' sequence matches:
        From 67 to 83 with 0 mismatches
3' sequence matches:
        From 205 to 219 with 0 mismatches
Sequences output to file:
        from 84 to 204
                tcgagagccgtattgcgatatagcgcacatgcgttggacacagatgagca
                cacagtgacatgagagacacagatatagagacagatagacgatagacaga
                cagcatatatagacagatagc
        sequence trimmed from 5' end:
                GGAAACAGCTAATGACCATGATTACGCCAAGCGCGCAATTAACCCTCACT
                AAAGGGAACAAAAGCTGGGTACCGGGCCCCCCC
        sequence trimmed from 3' end:
                TCGAGGTCGACGGTATCGATAAGCTTGATATCG


Sequence: pBlueScript    Vector: pLITMUS        No match

Sequence: litmus.seq     Vector: pTYB1  No match

Sequence: litmus.seq     Vector: pBS_KS+        No match


Sequence: litmus.seq     Vector: pLITMUS
5' sequence matches:
        From 43 to 61 with 0 mismatches
3' sequence matches:
        From 183 to 199 with 0 mismatches
Sequences output to file:
        from 62 to 182
                tcgagagccgtattgcgatatagcgcacatgcgttggacacagatgagca
                cacagtgacatgagagacacagatatagagacagatagacgatagacaga
                cagcatatatagacagatagc
        sequence trimmed from 5' end:
                TCTAGAACCGGTGACGTCTCCCATGGTGAAGCTTGGATCCACGATATCCT
                GCAGGAATTCC


Sequence: pTYB1.seq      Vector: pTYB1
5' sequence matches:
        From 40 to 58 with 0 mismatches
3' sequence matches:
        From 180 to 196 with 0 mismatches
Sequences output to file:
        from 59 to 179
                tcgagagccgtattgcgatatagcgcacatgcgttggacacagatgagca
                cacagtgacatgagagacacagatatagagacagatagacgatagacaga
                cagcatatatagacagatagc
        sequence trimmed from 5' end:
                CTTTAAGAAGGAGATATACATATGGCTAGCTCGCGAGTCGACGGCGGCCG
                CGAATTCC
        sequence trimmed from 3' end:
                TCGAGGGCTCTTCCTGCTTTGCCAAGGGTACCAATGTTTTAATGGCGGAT


Sequence: pTYB1.seq      Vector: pBS_KS+        No match

Sequence: pTYB1.seq      Vector: pLITMUS        No match


unix % more pbluescript.fasta 
>pBlueScript_from_84_to_204 KS+
tcgagagccgtattgcgatatagcgcacatgcgttggacacagatgagcacacagtgaca
tgagagacacagatatagagacagatagacgatagacagacagcatatatagacagatag
c
>litmus.seq_from_62_to_182
tcgagagccgtattgcgatatagcgcacatgcgttggacacagatgagcacacagtgaca
tgagagacacagatatagagacagatagacgatagacagacagcatatatagacagatag
c
>pTYB1.seq_from_59_to_179
tcgagagccgtattgcgatatagcgcacatgcgttggacacagatgagcacacagtgaca
tgagagacacagatatagagacagatagacgatagacagacagcatatatagacagatag
c
</pre>

<li> Running <b>vectorstrip</b> allowing maximum 30% mismatch, and asking only for best hits
<pre>
unix % vectorstrip litmus.seq
Strips out DNA between a pair of vector sequences
Are your vector sequences in a file? [Y]: 
Name of vectorfile: vectors
Max allowed % mismatch [10]: 30
Show only the best hits (minimise mismatches)? [Y]: 
Output file [litmus.vectorstrip]: stdout
Output sequence [litmus.fasta]: 

Sequence: litmus.seq     Vector: pTYB1  No match

Sequence: litmus.seq     Vector: pBS_KS+        No match


Sequence: litmus.seq     Vector: pLITMUS
5' sequence matches:
        From 43 to 61 with 0 mismatches
3' sequence matches:
        From 183 to 199 with 0 mismatches
Sequences output to file:
        from 62 to 182
                tcgagagccgtattgcgatatagcgcacatgcgttggacacagatgagca
                cacagtgacatgagagacacagatatagagacagatagacgatagacaga
                cagcatatatagacagatagc
        sequence trimmed from 5' end:
                TCTAGAACCGGTGACGTCTCCCATGGTGAAGCTTGGATCCACGATATCCT
                GCAGGAATTCC
        sequence trimmed from 3' end:
                TCGAGACCGTACGTGCGCGCGAATGCATCCAGATCTTCCCTCTAGTCAAG
                GCCTTAAGTGAGTCGTATTACGGA
</pre>

<li> Running <b>vectorstrip</b> allowing maximum 30% mismatch, and asking for all hits
<pre>
unix % vectorstrip litmus.seq
Strips out DNA between a pair of vector sequences
Are your vector sequences in a file? [Y]: 
Name of vectorfile: vectors
Max allowed % mismatch [10]: 30
Show only the best hits (minimise mismatches)? [Y]: N
Output file [litmus.vectorstrip]: stdout
Output sequence [litmus.fasta]: 

Sequence: litmus.seq     Vector: pTYB1  No match

Sequence: litmus.seq     Vector: pBS_KS+        No match


Sequence: litmus.seq     Vector: pLITMUS
5' sequence matches:
        From 43 to 61 with 0 mismatches
3' sequence matches:
        From 183 to 199 with 0 mismatches
        From 228 to 244 with 5 mismatches
Sequences output to file:
        from 62 to 182
                tcgagagccgtattgcgatatagcgcacatgcgttggacacagatgagca
                cacagtgacatgagagacacagatatagagacagatagacgatagacaga
                cagcatatatagacagatagc
        sequence trimmed from 5' end:
                TCTAGAACCGGTGACGTCTCCCATGGTGAAGCTTGGATCCACGATATCCT
                GCAGGAATTCC
        sequence trimmed from 3' end:
                TCGAGACCGTACGTGCGCGCGAATGCATCCAGATCTTCCCTCTAGTCAAG
                GCCTTAAGTGAGTCGTATTACGGA

        from 62 to 227
                tcgagagccgtattgcgatatagcgcacatgcgttggacacagatgagca
                cacagtgacatgagagacacagatatagagacagatagacgatagacaga
                cagcatatatagacagatagcTCGAGACCGTACGTGCGCGCGAATGCATC
                CAGATCTTCCCTCTAG
        sequence trimmed from 5' end:
                TCTAGAACCGGTGACGTCTCCCATGGTGAAGCTTGGATCCACGATATCCT
                GCAGGAATTCC
        sequence trimmed from 3' end:
                TCAAGGCCTTAAGTGAGTCGTATTACGGA
</pre>
<li> Running <b>vectorstrip</b> against a sequence containing Ns
<pre>
unix % vectorstrip pTYB1_N.seq
Strips out DNA between a pair of vector sequences
Are your vector sequences in a file? [Y]: 
Name of vectorfile: vectors
Max allowed % mismatch [10]: 30
Show only the best hits (minimise mismatches)? [Y]: 
Output file [ptyb1.vectorstrip]: stdout
Output sequence [ptyb1.fasta]: 


Sequence: pTYB1.seq      Vector: pTYB1
5' sequence matches:
        From 40 to 58 with 2 mismatches
3' sequence matches:
        From 180 to 196 with 2 mismatches
Sequences output to file:
        from 59 to 179
                tcnagagccgtatngcgatatngcgcacatgcgntggacacagangagca
                cacagtnacatgagagncacagatntagagacagatngacgataganaga
                cagcatanatagacanatagc
        sequence trimmed from 5' end:
                CTTTAAGNAGGAGANATACANATGGCNAGCTCGCGANTCGACGGCGGNCG
                CGAATNCC
        sequence trimmed from 3' end:
                TCGNGGGCTCTTCCNGCTTTGCCANGGGTACCAANGTTTTAATGGCNGAT


Sequence: pTYB1.seq      Vector: pBS_KS+        No match

Sequence: pTYB1.seq      Vector: pLITMUS        No match

</pre>
</ol>

<H2>
    Command line arguments
</H2>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers (* if not always prompted):
  [-sequence]          seqall     Sequence database USA
  [-[no]vectorfile]    bool       Are your vector sequences in a file?
* [-vectors]           infile     Name of vectorfile
   -mismatch           integer    Max allowed % mismatch
   -[no]besthits       bool       Show only the best hits (minimise
                                  mismatches)?
*  -linkera            string     5' sequence
*  -linkerb            string     3' sequence
  [-outf]              outfile    Output file name
  [-outseq]            seqoutall  Output sequence(s) USA

   Optional qualifiers: (none)
   Advanced qualifiers: (none)
   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-[no]vectorfile]<br>(Parameter 2)</td>
<td>Are your vector sequences in a file?</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>[-vectors]<br>(Parameter 3)</td>
<td>Name of vectorfile</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-mismatch</td>
<td>Max allowed % mismatch</td>
<td>Any integer value</td>
<td>10</td>
</tr>

<tr>
<td>-[no]besthits</td>
<td>Show only the best hits (minimise mismatches)?</td>
<td>Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-linkera</td>
<td>5' sequence</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-linkerb</td>
<td>3' sequence</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>[-outf]<br>(Parameter 4)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.vectorstrip</td>
</tr>

<tr>
<td>[-outseq]<br>(Parameter 5)</td>
<td>Output sequence(s) USA</td>
<td>Writeable sequence(s)</td>
<td><i>&lt;sequence&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>
<a name="input">
<!-- 
        This includes example input file formats.
        Except for sequences which can be described by 'sequence USA',
        this should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->
Each line should contain the name of the vector, the 5' pattern and 
the 3' pattern.<br>
Lines beginning with # are treated as comments and ignored.<br>
If only one vector sequence is given in the it will be assumed that 
this is the 5' pattern.<br>
If a vector name is given but no pattern data, the vector will not 
be used.
<p>
For example, the file used in the examples above is:<br>
<pre>
pTYB1	GACGGCGGCCGCGAATTCC	TCGAGGGCTCTTCCTGC
pBS_KS+	GGGTACCGGGCCCCCCC	TCGAGGTCGACGGTA		
pLITMUS	GATATCCTGCAGGAATTCC	TCGAGACCGTACGTGCG
</pre>
<p>
The sequence files used in the examples above are:
<ol>
<li> seqs.list<br>
<pre>
bluescript.seq
litmus.seq
pTYB1.seq
</pre>

<li>
bluescript.seq<br>
<pre>
> pBlueScript KS+
GGAAACAGCTAATGACCATGATTACGCCAAGCGCGCAATTAACCCTCACTAAAGGGAAC
AAAAGCTGGGTACCGGGCCCCCCCtcgagagccgtattgcgatatagcgcacatgcgtt
ggacacagatgagcacacagtgacatgagagacacagatatagagacagatagacgata
gacagacagcatatatagacagatagcTCGAGGTCGACGGTATCGATAAGCTTGATATCG
</pre>

<li>
litmus.seq<br>
<pre>
> litmus.seq 
TCTAGAACCGGTGACGTCTCCCATGGTGAAGCTTGGATCCACGATATCCTGCAGGAATT
CCtcgagagccgtattgcgatatagcgcacatgcgttggacacagatgagcacacagtg
acatgagagacacagatatagagacagatagacgatagacagacagcatatatagacag
atagcTCGAGACCGTACGTGCGCGCGAATGCATCCAGATCTTCCCTCTAGTCAAGGCCT
TAAGTGAGTCGTATTACGGA
</pre>

<li>
pTYB1.seq
<pre>
>pTYB1.seq
CTTTAAGAAGGAGATATACATATGGCTAGCTCGCGAGTCGACGGCGGCCGCGAATTCCtc
gagagccgtattgcgatatagcgcacatgcgttggacacagatgagcacacagtgacatg
agagacacagatatagagacagatagacgatagacagacagcatatatagacagatagcT
CGAGGGCTCTTCCTGCTTTGCCAAGGGTACCAATGTTTTAATGGCGGAT
</pre>

<li>
pTYB1_N.seq<br>
<pre>
>pTYB1.seq lower quality
CTTTAAGNAGGAGANATACANATGGCNAGCTCGCGANTCGACGGCGGNCGCGAATNCCtc
nagagccgtatngcgatatngcgcacatgcgntggacacagangagcacacagtnacatg
agagncacagatntagagacagatngacgataganagacagcatanatagacanatagcT
CGNGGGCTCTTCCNGCTTTGCCANGGGTACCAANGTTTTAATGGCNGAT

</pre>
</ol>

In each case the vector sequence is given in upper case and the 
cloned sequence in lower case.

<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats 
        Except for sequences which can be described by 'sequence USA',
        this should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->

Two types of output file are produced:
<ol>
<li>The sequence file(s) - contain the trimmed subsequence(s) produced by <b>vectorstrip</b> either all in one file, or in separate files if the command line flag -ossingle is used.
<li>Results summary file eg<br>
<pre>
Sequence: pTYB1.seq      Vector: pTYB1
5' sequence matches:
        From 40 to 58 with 2 mismatches
3' sequence matches:
        From 180 to 196 with 2 mismatches
Sequences output to file:
        from 59 to 179
                tcnagagccgtatngcgatatngcgcacatgcgntggacacagangagca
                cacagtnacatgagagncacagatntagagacagatngacgataganaga
                cagcatanatagacanatagc
        sequence trimmed from 5' end:
                CTTTAAGNAGGAGANATACANATGGCNAGCTCGCGANTCGACGGCGGNCG
                CGAATNCC
        sequence trimmed from 3' end:
                TCGNGGGCTCTTCCNGCTTTGCCANGGGTACCAANGTTTTAATGGCNGAT

</pre>
</ol>

<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.

        eg:
        data file format:

	blank lines and comment lines are ignored.
        comment lines start with '#'
        There are three columns separated by one or more space or TAB
	characters for X and Y values of durm, followed by energy
        all coefficients are given as floating point values
        An example is:

        # Durm coefficient enery file
        # X     Y       Energy
        0.1     0.2     0.331
        0.2     0.4     0.332
        0.3     0.6     0.333
        0.3     0.8     0.334



If you wish to include the standard description of what data files
are and how to use embossdata to inspect and retrieve them, use:

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>


   -->

None.

<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.

<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.


<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
        eg:
        It doesn't check to see if the sequence is a protein sequence or not.
   -->
None.


<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program.
        eg:

        "Threshold value greater than 1." - means that the threshold has
        been set to be greater than 1 but a burgle exception has
        occurred making this invalid - you should reduce the wobble
        parameter and try again. 

   -->
<ol>
<li><pre>No suitable vectors found - exiting</pre> indicates that the 5' and 3' patterns for the vectors were blank - usually this is as a result of an empty vectorfile.
<li><pre>Illegal pattern</pre> indicates that one of the vector patterns could not be compiled and therefore cannot be searched.
<li><pre>5' and 3' sequence matches are identical; inconclusive</pre> indicates that the 5' and 3' patterns provided were identical, and that they only match the sequence once. Thus the program cannot determine which part of the sequence is vector and which is insert.
</ol>

<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
        eg:
        "It always exits with status 0"
   -->
It always exits with status 0.
<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.



<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="biosed.html">biosed</a></td><td>Replace or delete sequence sections</td></tr>
<tr><td><a href="cutseq.html">cutseq</a></td><td>Removes a specified section from a sequence</td></tr>
<tr><td><a href="degapseq.html">degapseq</a></td><td>Removes gap characters from sequences</td></tr>
<tr><td><a href="descseq.html">descseq</a></td><td>Alter the name or description of a sequence</td></tr>
<tr><td><a href="entret.html">entret</a></td><td>Reads and writes (returns) flatfile entries</td></tr>
<tr><td><a href="extractfeat.html">extractfeat</a></td><td>Extract features from a sequence</td></tr>
<tr><td><a href="extractseq.html">extractseq</a></td><td>Extract regions from a sequence</td></tr>
<tr><td><a href="listor.html">listor</a></td><td>Writes a list file of the logical OR of two sets of sequences</td></tr>
<tr><td><a href="maskfeat.html">maskfeat</a></td><td>Mask off features of a sequence</td></tr>
<tr><td><a href="maskseq.html">maskseq</a></td><td>Mask off regions of a sequence</td></tr>
<tr><td><a href="newseq.html">newseq</a></td><td>Type in a short new sequence</td></tr>
<tr><td><a href="noreturn.html">noreturn</a></td><td>Removes carriage return from ASCII files</td></tr>
<tr><td><a href="notseq.html">notseq</a></td><td>Excludes a set of sequences and writes out the remaining ones</td></tr>
<tr><td><a href="nthseq.html">nthseq</a></td><td>Writes one sequence from a multiple set of sequences</td></tr>
<tr><td><a href="pasteseq.html">pasteseq</a></td><td>Insert one sequence into another</td></tr>
<tr><td><a href="revseq.html">revseq</a></td><td>Reverse and complement a sequence</td></tr>
<tr><td><a href="seqret.html">seqret</a></td><td>Reads and writes (returns) sequences</td></tr>
<tr><td><a href="seqretsplit.html">seqretsplit</a></td><td>Reads and writes (returns) sequences in individual files</td></tr>
<tr><td><a href="splitter.html">splitter</a></td><td>Split a sequence into (overlapping) smaller sequences</td></tr>
<tr><td><a href="swissparse.html">swissparse</a></td><td>Retrieves sequences from swissprot using keyword search</td></tr>
<tr><td><a href="trimest.html">trimest</a></td><td>Trim poly-A tails off EST sequences</td></tr>
<tr><td><a href="trimseq.html">trimseq</a></td><td>Trim ambiguous bits off the ends of sequences</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->




<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
e.g.

This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>
   -->
This application was written by Val Curwen <a
href="mailto:vcurwen@hgmp.mrc.ac.uk">(vcurwen@hgmp.mrc.ac.uk)</a>


<H2>
    History
</H2>

16 August 2000 - Val Curwen - Written.

<!--
        What changes have been made.
   -->


<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:

   -->
This program is intended to be used by everyone and everything, from naive users to embedded scripts.



<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->



</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

