<HTML>

<HEAD>
  <TITLE>
  EMBOSS
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">







<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

vectorstrip
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>







<!--END OF HEADER-->





<H1>
Program vectorstrip
</H1>
<H2>
    Function
</H2>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Strips out DNA between a pair of vector sequences


<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.

        This is what the user looks at to find out if this program will
        do what he wants. 

        Long description.

        Why it is useful.

        When it should/shouldn't be used.

        Algorithms

        Expanded description.

        eg: Reverse engineers Wheedle's 'talpsgob' program. This does a 
        splurge of the 5' end of the sequence followed by a durm.
        The net effect is a 5' end splurt.
        etc. for as long as is required (several pages is not unreasonable)
  -->

<b>vectorstrip</b> is intended to be useful for stripping vector sequence
from the ends of sequences of interest. For example, if a fragment has
been cloned into a vector and then sequenced, the sequence may contain
vector data eg from the cloning polylinker at the 5' and 3' ends of
the sequence.  <b>vectorstrip</b> will remove these contaminating regions and
output trimmed sequence ready for input into another application.

<p>

<b>vectorstrip</b> is suitable for use with low quality sequence data as it
can allow for mismatches between the sequence and the vector patterns
provided. You can specify the maximum level of mismatch
expected.
<p>
Vector data can either be provided in a file or interactively. If
presented in a file, <b>vectorstrip</b> will search all input sequences with
all vectors listed in that file. The intention is that the user can
maintain a single file for use with <b>vectorstrip</b>, containing all the
linker sequences commonly used in the laboratory.

<p>
The two patterns for each vector are searched separately against the
sequence. Once the search is completed, each of the hits of the 5'
sequence is paired with each of the hits of the 3' sequence and the
resulting subsequences are output. For example, if the 5' sequence
matches the sequence from (a) position 30-60, and(b)position 70-100,
and the 3' sequence matches from 150-175, then two subsequences will
be output: from 61-149, and from 101-149. The lower the quality of
the sequence, the more likely multiple hits become if nonzero
mismatches are accepted.

<p>
Default behaviour is to report only the best matches between the
vector patterns and the sequence. This means that if you
specify a maximum mismatch level of 10%, but the vector patterns
match the sequence with zero mismatches, the search will stop and the
program will output only these "best" matches. If there are no perfect
matches, the program will try searching again allowing 1 mismatch,
then 2, and so on until either the patterns match the sequence or the
maximum specified mismatch level is exceeded. You can tell <b>vectorstrip</b>
to show all possible matches up to your specified maximum level, as
illustrated in the examples below.

<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

<b>Here is a sample session with vectorstrip</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>vectorstrip @../../data/vecseqs.list </b>
Strips out DNA between a pair of vector sequences
Are your vector sequences in a file? [Y]: <b></b>
Name of vectorfile: <b>../../data/vectors</b>
Max allowed % mismatch [10]: <b></b>
Show only the best hits (minimise mismatches)? [Y]: <b></b>
Output file [pbluescript.vectorstrip]: <b>vector.strip</b>
Output sequence [pbluescript.fasta]: <b>vector.fasta</b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers (* if not always prompted):
  [-sequence]          seqall     Sequence database USA
  [-[no]vectorfile]    boolean    Are your vector sequences in a file?
* [-vectors]           infile     Name of vectorfile
   -mismatch           integer    Max allowed % mismatch
   -[no]besthits       boolean    Show only the best hits (minimise
                                  mismatches)?
*  -linkera            string     5' sequence
*  -linkerb            string     3' sequence
  [-outf]              outfile    Output file name
  [-outseq]            seqoutall  Output sequence(s) USA

   Optional qualifiers: (none)
   Advanced qualifiers: (none)
   Associated qualifiers:
  "-sequence" related qualifiers
   -sbegin1             integer    First base used
   -send1               integer    Last base used, def=seq length
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sopenfile1          string     Input filename
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name
  "-outf" related qualifiers
   -odirectory4         string     Output directory
  "-outseq" related qualifiers
   -osformat5           string     Output seq format
   -osextension5        string     File name extension
   -osname5             string     Base file name
   -osdirectory5        string     Output directory
   -osdbname5           string     Database name to add
   -ossingle5           boolean    Separate file for each entry
   -oufo5               string     UFO features
   -offormat5           string     Features format
   -ofname5             string     Features file name
   -ofdirectory5        string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for required and optional values
   -debug               boolean    Write debug output to program.dbg
   -acdlog              boolean    Write ACD processing log to program.acdlog
   -acdpretty           boolean    Rewrite ACD file as program.acdpretty
   -acdtable            boolean    Write HTML table of options
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-[no]vectorfile]<br>(Parameter 2)</td>
<td>Are your vector sequences in a file?</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>[-vectors]<br>(Parameter 3)</td>
<td>Name of vectorfile</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-mismatch</td>
<td>Max allowed % mismatch</td>
<td>Any integer value</td>
<td>10</td>
</tr>

<tr>
<td>-[no]besthits</td>
<td>Show only the best hits (minimise mismatches)?</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-linkera</td>
<td>5' sequence</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-linkerb</td>
<td>3' sequence</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>[-outf]<br>(Parameter 4)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.vectorstrip</td>
</tr>

<tr>
<td>[-outseq]<br>(Parameter 5)</td>
<td>Output sequence(s) USA</td>
<td>Writeable sequence(s)</td>
<td><i>&lt;sequence&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

</table>



<H2>
    Input file format
</H2>
<a name="input">
<!-- 
        This includes example input file formats.
        Except for sequences which can be described by 'sequence USA',
        this should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: ../../data/vecseqs.list</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
../../data/bluescript.seq
../../data/litmus.seq
../../data/pTYB1.seq
</pre>
</td></tr></table><p>
<p><h3>File: ../../data/vectors</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
# Example vector file for use by vectorstrip

# Vector	5'			3'

pTYB1   GACGGCGGCCGCGAATTCC     TCGAGGGCTCTTCCTGC
pBS_KS+ GGGTACCGGGCCCCCCC       TCGAGGTCGACGGTA         
pLITMUS GATATCCTGCAGGAATTCC     TCGAGACCGTACGTGCG
</pre>
</td></tr></table><p>


<p>

The same fragment has been cloned into the XhoI site of the polylinker
of each vector.  The cloned fragment is represented in lower case and
the vector sequence in upper case so the sequence trimming can be
readily seen. 

<p>


Each line of the vector file should contain the name of the vector, the
5' pattern and the 3' pattern. 

<br>

Lines beginning with # are treated as comments and ignored.

<br>

If only one vector sequence is given in the it will be assumed that 
this is the 5' pattern.

<br>

If a vector name is given but no pattern data, the vector will not 
be used.



<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats 
        Except for sequences which can be described by 'sequence USA',
        this should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 
   -->


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: vector.fasta</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt;pBlueScript_from_84_to_204 KS+
tcgagagccgtattgcgatatagcgcacatgcgttggacacagatgagcacacagtgaca
tgagagacacagatatagagacagatagacgatagacagacagcatatatagacagatag
c
&gt;litmus.seq_from_62_to_182
tcgagagccgtattgcgatatagcgcacatgcgttggacacagatgagcacacagtgaca
tgagagacacagatatagagacagatagacgatagacagacagcatatatagacagatag
c
&gt;pTYB1.seq_from_59_to_179
tcgagagccgtattgcgatatagcgcacatgcgttggacacagatgagcacacagtgaca
tgagagacacagatatagagacagatagacgatagacagacagcatatatagacagatag
c
</pre>
</td></tr></table><p>
<p><h3>File: vector.strip</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>

Sequence: pBlueScript 	 Vector: pTYB1	No match


Sequence: pBlueScript 	 Vector: pBS_KS+
5' sequence matches:
	From 67 to 83 with 0 mismatches
3' sequence matches:
	From 205 to 219 with 0 mismatches
Sequences output to file:
	from 84 to 204
		tcgagagccgtattgcgatatagcgcacatgcgttggacacagatgagca
		cacagtgacatgagagacacagatatagagacagatagacgatagacaga
		cagcatatatagacagatagc
	sequence trimmed from 5' end:
		GGAAACAGCTAATGACCATGATTACGCCAAGCGCGCAATTAACCCTCACT
		AAAGGGAACAAAAGCTGGGTACCGGGCCCCCCC
	sequence trimmed from 3' end:
		TCGAGGTCGACGGTATCGATAAGCTTGATATCG


Sequence: pBlueScript 	 Vector: pLITMUS	No match

Sequence: litmus.seq 	 Vector: pTYB1	No match

Sequence: litmus.seq 	 Vector: pBS_KS+	No match


Sequence: litmus.seq 	 Vector: pLITMUS
5' sequence matches:
	From 43 to 61 with 0 mismatches
3' sequence matches:
	From 183 to 199 with 0 mismatches
Sequences output to file:
	from 62 to 182
		tcgagagccgtattgcgatatagcgcacatgcgttggacacagatgagca
		cacagtgacatgagagacacagatatagagacagatagacgatagacaga
		cagcatatatagacagatagc
	sequence trimmed from 5' end:
		TCTAGAACCGGTGACGTCTCCCATGGTGAAGCTTGGATCCACGATATCCT
		GCAGGAATTCC
	sequence trimmed from 3' end:
		TCGAGACCGTACGTGCGCGCGAATGCATCCAGATCTTCCCTCTAGTCAAG
		GCCTTAAGTGAGTCGTATTACGGA



Sequence: pTYB1.seq 	 Vector: pTYB1
5' sequence matches:
	From 40 to 58 with 0 mismatches
3' sequence matches:
	From 180 to 196 with 0 mismatches
Sequences output to file:
	from 59 to 179
		tcgagagccgtattgcgatatagcgcacatgcgttggacacagatgagca
		cacagtgacatgagagacacagatatagagacagatagacgatagacaga
		cagcatatatagacagatagc
	sequence trimmed from 5' end:
		CTTTAAGAAGGAGATATACATATGGCTAGCTCGCGAGTCGACGGCGGCCG
		CGAATTCC
	sequence trimmed from 3' end:
		TCGAGGGCTCTTCCTGCTTTGCCAAGGGTACCAATGTTTTAATGGCGGAT


Sequence: pTYB1.seq 	 Vector: pBS_KS+	No match

Sequence: pTYB1.seq 	 Vector: pLITMUS	No match
</pre>
</td></tr></table><p>


<p>

Two types of output file are produced:
<ol>

<li>The sequence file(s) - contain the trimmed subsequence(s) produced
by <b>vectorstrip</b> either all in one file, or in separate files if
the command line flag <b>-ossingle</b> is used. 

<li>Results summary file 
</ol>

<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.

        eg:
        data file format:

	blank lines and comment lines are ignored.
        comment lines start with '#'
        There are three columns separated by one or more space or TAB
	characters for X and Y values of durm, followed by energy
        all coefficients are given as floating point values
        An example is:

        # Durm coefficient enery file
        # X     Y       Energy
        0.1     0.2     0.331
        0.2     0.4     0.332
        0.3     0.6     0.333
        0.3     0.8     0.334



If you wish to include the standard description of what data files
are and how to use embossdata to inspect and retrieve them, use:

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>


   -->

None.

<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.

<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.


<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
        eg:
        It doesn't check to see if the sequence is a protein sequence or not.
   -->
None.


<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program.
        eg:

        "Threshold value greater than 1." - means that the threshold has
        been set to be greater than 1 but a burgle exception has
        occurred making this invalid - you should reduce the wobble
        parameter and try again. 

   -->
<ol>
<li><pre>No suitable vectors found - exiting</pre> indicates that the 5' and 3' patterns for the vectors were blank - usually this is as a result of an empty vectorfile.
<li><pre>Illegal pattern</pre> indicates that one of the vector patterns could not be compiled and therefore cannot be searched.
<li><pre>5' and 3' sequence matches are identical; inconclusive</pre> indicates that the 5' and 3' patterns provided were identical, and that they only match the sequence once. Thus the program cannot determine which part of the sequence is vector and which is insert.
</ol>

<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
        eg:
        "It always exits with status 0"
   -->
It always exits with status 0.
<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.



<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="biosed.html">biosed</a></td><td>Replace or delete sequence sections</td></tr>
<tr><td><a href="cutseq.html">cutseq</a></td><td>Removes a specified section from a sequence</td></tr>
<tr><td><a href="degapseq.html">degapseq</a></td><td>Removes gap characters from sequences</td></tr>
<tr><td><a href="descseq.html">descseq</a></td><td>Alter the name or description of a sequence</td></tr>
<tr><td><a href="entret.html">entret</a></td><td>Reads and writes (returns) flatfile entries</td></tr>
<tr><td><a href="extractfeat.html">extractfeat</a></td><td>Extract features from a sequence</td></tr>
<tr><td><a href="extractseq.html">extractseq</a></td><td>Extract regions from a sequence</td></tr>
<tr><td><a href="listor.html">listor</a></td><td>Writes a list file of the logical OR of two sets of sequences</td></tr>
<tr><td><a href="maskfeat.html">maskfeat</a></td><td>Mask off features of a sequence</td></tr>
<tr><td><a href="maskseq.html">maskseq</a></td><td>Mask off regions of a sequence</td></tr>
<tr><td><a href="newseq.html">newseq</a></td><td>Type in a short new sequence</td></tr>
<tr><td><a href="noreturn.html">noreturn</a></td><td>Removes carriage return from ASCII files</td></tr>
<tr><td><a href="notseq.html">notseq</a></td><td>Excludes a set of sequences and writes out the remaining ones</td></tr>
<tr><td><a href="nthseq.html">nthseq</a></td><td>Writes one sequence from a multiple set of sequences</td></tr>
<tr><td><a href="pasteseq.html">pasteseq</a></td><td>Insert one sequence into another</td></tr>
<tr><td><a href="revseq.html">revseq</a></td><td>Reverse and complement a sequence</td></tr>
<tr><td><a href="seqret.html">seqret</a></td><td>Reads and writes (returns) sequences</td></tr>
<tr><td><a href="seqretsplit.html">seqretsplit</a></td><td>Reads and writes (returns) sequences in individual files</td></tr>
<tr><td><a href="skipseq.html">skipseq</a></td><td>Reads and writes (returns) sequences, skipping the first few</td></tr>
<tr><td><a href="splitter.html">splitter</a></td><td>Split a sequence into (overlapping) smaller sequences</td></tr>
<tr><td><a href="trimest.html">trimest</a></td><td>Trim poly-A tails off EST sequences</td></tr>
<tr><td><a href="trimseq.html">trimseq</a></td><td>Trim ambiguous bits off the ends of sequences</td></tr>
<tr><td><a href="union.html">union</a></td><td>Reads sequence fragments and builds one sequence</td></tr>
<tr><td><a href="yank.html">yank</a></td><td>Reads a sequence range, appends the full USA to a list file</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->




<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

Val Curwen (vac&nbsp;&copy;&nbsp;sanger.ac.uk)
<br>
Sanger Institute, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK. 




<H2>
    History
</H2>

16 August 2000 - Val Curwen - Written.

<!--
<p>
[an error occurred while processing this directive]
-->

<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:

   -->
This program is intended to be used by everyone and everything, from naive users to embedded scripts.



<H2>
    Comments
</H2>

<!--
	Any comments, requests etc. that people have about the program
   -->



</BODY>
</HTML>

