<HTML>

<HEAD>
  <TITLE>
  EMBOSS: water
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">







<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

water
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>







<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Smith-Waterman local alignment


<H2>
    Description
</H2>

<b>water</b> uses the Smith-Waterman algorithm (modified for speed
enhancments) to calculate the local alignment.
<P>

The Smith-Waterman algorithm is a member of the class of algorithms
that can calculate the best score and local alignment in the order of mn
steps, (where 'n' and 'm' are the lengths of the two sequences).  These
dynamic programming algorithms were first developed for protein sequence
comparison by Smith and Waterman, though similar methods were
independently devised during the late 1960's and early 1970's for use in
the fields of speech processing and computer science. 
<P>

A local alignment searches for regions of local similarity between two
sequences and need not include the entire length of the sequences. 
Local alignment methods are very useful for scanning databases or other
circumsatnces when you wish to find matches between small regions of
sequences, for example between protein domains. 
<P>

Dynamic programming methods ensure
the optimal local alignment by exploring all possible alignments and
choosing the best.  It does this by reading in a scoring matrix that
contains values for every possible residue or nucleotide match.  <b>water</b>
finds an alignment with the maximum possible score where the score of an
alignment is equal to the sum of the matches taken from the scoring
matrix. 
<P>

An important problem is the treatment of gaps, i.e., spaces inserted to
optimise the alignment score.  A penalty is subtracted from the score
for each gap opened (the 'gap open' penalty) and a penalty is subtracted
from the score for the total number of gap spaces multiplied by a cost
(the 'gap extension' penalty). 
<P>

Typically, the cost of extending a gap is set to be 5-10 times lower
than the cost for opening a gap. 
<P>

<H2>
    Usage
</H2>


<b>Here is a sample session with water</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>water tsw:hba_human tsw:hbb_human </b>
Smith-Waterman local alignment.
Gap opening penalty [10.0]: <b></b>
Gap extension penalty [0.5]: <b></b>
Output alignment [hba_human.water]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequencea]         sequence   Sequence USA
  [-seqall]            seqall     Sequence database USA
   -gapopen            float      The gap open penalty is the score taken away
                                  when a gap is created. The best value
                                  depends on the choice of comparison matrix.
                                  The default value assumes you are using the
                                  EBLOSUM62 matrix for protein sequences, and
                                  the EDNAFULL matrix for nucleotide
                                  sequences.
   -gapextend          float      The gap extension penalty is added to the
                                  standard gap penalty for each base or
                                  residue in the gap. This is how long gaps
                                  are penalized. Usually you will expect a few
                                  long gaps rather than many short gaps, so
                                  the gap extension penalty should be lower
                                  than the gap penalty. An exception is where
                                  one or both sequences are single reads with
                                  possible sequencing errors in which case you
                                  would expect many single base gaps. You can
                                  get this result by setting the gap open
                                  penalty to zero (or very low) and using the
                                  gap extension penalty to control gap
                                  scoring.
  [-outfile]           align      Output alignment file name

   Optional qualifiers:
   -datafile           matrixf    This is the scoring matrix file used when
                                  comparing sequences. By default it is the
                                  file 'EBLOSUM62' (for proteins) or the file
                                  'EDNAFULL' (for nucleic sequences). These
                                  files are found in the 'data' directory of
                                  the EMBOSS installation.

   Advanced qualifiers:
   -[no]brief          boolean    Brief identity and similarity

   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequencea]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-seqall]<br>(Parameter 2)</td>
<td>Sequence database USA</td>
<td>Readable sequence(s)</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-gapopen</td>
<td>The gap open penalty is the score taken away when a gap is created. The best value depends on the choice of comparison matrix. The default value assumes you are using the EBLOSUM62 matrix for protein sequences, and the EDNAFULL matrix for nucleotide sequences.</td>
<td>Number from 1.000 to 100.000</td>
<td>10.0 for any sequence</td>
</tr>

<tr>
<td>-gapextend</td>
<td>The gap extension penalty is added to the standard gap penalty for each base or residue in the gap. This is how long gaps are penalized. Usually you will expect a few long gaps rather than many short gaps, so the gap extension penalty should be lower than the gap penalty. An exception is where one or both sequences are single reads with possible sequencing errors in which case you would expect many single base gaps. You can get this result by setting the gap open penalty to zero (or very low) and using the gap extension penalty to control gap scoring.</td>
<td>Number from 0.100 to 10.000</td>
<td>0.5 for any sequence</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 3)</td>
<td>Output alignment file name</td>
<td>Alignment output file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-datafile</td>
<td>This is the scoring matrix file used when comparing sequences. By default it is the file 'EBLOSUM62' (for proteins) or the file 'EDNAFULL' (for nucleic sequences). These files are found in the 'data' directory of the EMBOSS installation.</td>
<td>Comparison matrix file in EMBOSS data path</td>
<td>EBLOSUM62 for protein<br>EDNAFULL for DNA</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]brief</td>
<td>Brief identity and similarity</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

</table>



<H2>
    Input file format
</H2>


<b>water</b> reads any two sequence USAs of the same type (DNA or protein).

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tsw:hba_human' is a sequence entry in the example protein database 'tsw'
<p>
<p><h3>Database entry: tsw:hba_human</h3>
<table width="90%"><tr><td bgcolor="#FFDDFF">
<pre>
ID   HBA_HUMAN      STANDARD;      PRT;   141 AA.
AC   P01922;
DT   21-JUL-1986 (Rel. 01, Created)
DT   21-JUL-1986 (Rel. 01, Last sequence update)
DT   15-JUL-1999 (Rel. 38, Last annotation update)
DE   HEMOGLOBIN ALPHA CHAIN.
GN   HBA1 AND HBA2.
OS   Homo sapiens (Human), Pan troglodytes (Chimpanzee), and
OS   Pan paniscus (Pygmy chimpanzee) (Bonobo).
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Mammalia;
OC   Eutheria; Primates; Catarrhini; Hominidae; Homo.
RN   [1]
RP   SEQUENCE FROM N.A. (ALPHA-1).
RX   MEDLINE; 81088339.
RA   MICHELSON A.M., ORKIN S.H.;
RT   "The 3' untranslated regions of the duplicated human alpha-globin
RT   genes are unexpectedly divergent.";
RL   Cell 22:371-377(1980).
RN   [2]
RP   SEQUENCE FROM N.A. (ALPHA-2).
RX   MEDLINE; 81175088.
RA   LIEBHABER S.A., GOOSSENS M.J., KAN Y.W.;
RT   "Cloning and complete nucleotide sequence of human 5'-alpha-globin
RT   gene.";
RL   Proc. Natl. Acad. Sci. U.S.A. 77:7054-7058(1980).
RN   [3]
RP   SEQUENCE FROM N.A. (ALPHA-2).
RX   MEDLINE; 80137531.
RA   WILSON J.T., WILSON L.B., REDDY V.B., CAVALLESCO C., GHOSH P.K.,
RA   DERIEL J.K., FORGET B.G., WEISSMAN S.M.;
RT   "Nucleotide sequence of the coding portion of human alpha globin
RT   messenger RNA.";
RL   J. Biol. Chem. 255:2807-2815(1980).
RN   [4]
RP   SEQUENCE FROM N.A. (ALPHA-1 AND ALPHA-2).
RA   FLINT J., HIGGS D.R.;
RL   Submitted (JAN-1997) to the EMBL/GenBank/DDBJ databases.
RN   [5]
RP   SEQUENCE.
RA   BRAUNITZER G., GEHRING-MULLER R., HILSCHMANN N., HILSE K., HOBOM G.,
RA   RUDLOFF V., WITTMANN-LIEBOLD B.;
RT   "The constitution of normal adult human haemoglobin.";
RL   Hoppe-Seyler's Z. Physiol. Chem. 325:283-286(1961).
RN   [6]
RP   SEQUENCE.
RA   HILL R.J., KONIGSBERG W.;
RT   "The structure of human hemoglobin: IV. The chymotryptic digestion of
RT   the alpha chain of human hemoglobin.";
RL   J. Biol. Chem. 237:3151-3156(1962).
RN   [7]


<font color=red>  [Part of this file has been deleted for brevity]</font>

FT                                /FTId=VAR_002841.
FT   VARIANT     130    130       A -&gt; D (IN YUDA; O2 AFFINITY DOWN).
FT                                /FTId=VAR_002842.
FT   VARIANT     131    131       S -&gt; P (IN QUESTEMBERT; HIGHLY UNSTABLE;
FT                                CAUSES ALPHA-THALASSEMIA).
FT                                /FTId=VAR_002843.
FT   VARIANT     133    133       S -&gt; R (IN VAL DE MARNE; O2 AFFINITY UP).
FT                                /FTId=VAR_002844.
FT   VARIANT     135    135       V -&gt; E (IN PAVIE).
FT                                /FTId=VAR_002845.
FT   VARIANT     136    136       L -&gt; M (IN CHICAGO).
FT                                /FTId=VAR_002846.
FT   VARIANT     136    136       L -&gt; P (IN BIBBA; UNSTABLE;
FT                                CAUSES ALPHA-THALASSEMIA).
FT                                /FTId=VAR_002847.
FT   VARIANT     138    138       S -&gt; P (IN ATTLEBORO; O2 AFFINITY UP).
FT                                /FTId=VAR_002848.
FT   VARIANT     139    139       K -&gt; E (IN HANAKAMI; O2 AFFINITY UP).
FT                                /FTId=VAR_002849.
FT   VARIANT     139    139       K -&gt; T (IN TOKONAME; O2 AFFINITY UP).
FT                                /FTId=VAR_002850.
FT   VARIANT     140    140       Y -&gt; H (IN ROUEN; O2 AFFINITY UP).
FT                                /FTId=VAR_002851.
FT   VARIANT     141    141       R -&gt; C (IN NUNOBIKI; O2 AFFINITY UP).
FT                                /FTId=VAR_002852.
FT   VARIANT     141    141       R -&gt; L (IN LEGNANO; O2 AFFINITY UP).
FT                                /FTId=VAR_002853.
FT   VARIANT     141    141       R -&gt; H (IN SURESNES; O2 AFFINITY UP).
FT                                /FTId=VAR_002854.
FT   VARIANT     141    141       R -&gt; P (IN SINGAPORE).
FT                                /FTId=VAR_002855.
FT   HELIX         4     35
FT   HELIX        37     42
FT   TURN         44     45
FT   TURN         50     51
FT   HELIX        53     71
FT   TURN         72     74
FT   HELIX        76     79
FT   TURN         80     80
FT   HELIX        81     89
FT   TURN         90     91
FT   TURN         95     95
FT   HELIX        96    112
FT   TURN        114    116
FT   HELIX       119    136
FT   TURN        137    139
SQ   SEQUENCE   141 AA;  15126 MW;  5EC7DB1E CRC32;
     VLSPADKTNV KAAWGKVGAH AGEYGAEALE RMFLSFPTTK TYFPHFDLSH GSAQVKGHGK
     KVADALTNAV AHVDDMPNAL SALSDLHAHK LRVDPVNFKL LSHCLLVTLA AHLPAEFTPA
     VHASLDKFLA SVSTVLTSKY R
//
</pre>
</td></tr></table><p>
<p><h3>Database entry: tsw:hbb_human</h3>
<table width="90%"><tr><td bgcolor="#FFDDFF">
<pre>
ID   HBB_HUMAN      STANDARD;      PRT;   146 AA.
AC   P02023;
DT   21-JUL-1986 (Rel. 01, Created)
DT   21-JUL-1986 (Rel. 01, Last sequence update)
DT   15-JUL-1999 (Rel. 38, Last annotation update)
DE   HEMOGLOBIN BETA CHAIN.
GN   HBB.
OS   Homo sapiens (Human), Pan troglodytes (Chimpanzee), and
OS   Pan paniscus (Pygmy chimpanzee) (Bonobo).
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Mammalia;
OC   Eutheria; Primates; Catarrhini; Hominidae; Homo.
RN   [1]
RP   SEQUENCE.
RC   SPECIES=HUMAN;
RA   BRAUNITZER G., GEHRING-MULLER R., HILSCHMANN N., HILSE K., HOBOM G.,
RA   RUDLOFF V., WITTMANN-LIEBOLD B.;
RT   "The constitution of normal adult human haemoglobin.";
RL   Hoppe-Seyler's Z. Physiol. Chem. 325:283-286(1961).
RN   [2]
RP   SEQUENCE FROM N.A.
RC   SPECIES=HUMAN;
RX   MEDLINE; 81064667.
RA   LAWN R.M., EFSTRATIADIS A., O'CONNELL C., MANIATIS T.;
RT   "The nucleotide sequence of the human beta-globin gene.";
RL   Cell 21:647-651(1980).
RN   [3]
RP   SEQUENCE OF 121-146 FROM N.A.
RC   SPECIES=HUMAN;
RX   MEDLINE; 85205333.
RA   LANG K.M., SPRITZ R.A.;
RT   "Cloning specific complete polyadenylylated 3'-terminal cDNA
RT   segments.";
RL   Gene 33:191-196(1985).
RN   [4]
RP   X-RAY CRYSTALLOGRAPHY (2.5 ANGSTROMS) OF DEOXYHEMOGLOBIN.
RC   SPECIES=HUMAN;
RX   MEDLINE; 76027820.
RA   FERMI G.;
RT   "Three-dimensional fourier synthesis of human deoxyhaemoglobin at
RT   2.5-A resolution: refinement of the atomic model.";
RL   J. Mol. Biol. 97:237-256(1975).
RN   [5]
RP   SEQUENCE.
RC   SPECIES=P.TROGLODYTES;
RX   MEDLINE; 66071496.
RA   RIFKIN D.B., KONIGSBERG W.;
RT   "The characterization of the tryptic peptides from the hemoglobin of
RT   the chimpanzee (Pan troglodytes).";
RL   Biochim. Biophys. Acta 104:457-461(1965).
RN   [6]


<font color=red>  [Part of this file has been deleted for brevity]</font>

FT   VARIANT     140    140       A -&gt; T (IN ST JACQUES: O2 AFFINITY UP).
FT                                /FTId=VAR_003081.
FT   VARIANT     140    140       A -&gt; V (IN PUTTELANGE; POLYCYTHEMIA;
FT                                O2 AFFINITY UP).
FT                                /FTId=VAR_003082.
FT   VARIANT     141    141       L -&gt; R (IN OLMSTED; UNSTABLE).
FT                                /FTId=VAR_003083.
FT   VARIANT     142    142       A -&gt; D (IN OHIO; O2 AFFINITY UP).
FT                                /FTId=VAR_003084.
FT   VARIANT     143    143       H -&gt; D (IN RANCHO MIRAGE).
FT                                /FTId=VAR_003085.
FT   VARIANT     143    143       H -&gt; Q (IN LITTLE ROCK; O2 AFFINITY UP).
FT                                /FTId=VAR_003086.
FT   VARIANT     143    143       H -&gt; P (IN SYRACUSE; O2 AFFINITY UP).
FT                                /FTId=VAR_003087.
FT   VARIANT     143    143       H -&gt; R (IN ABRUZZO; O2 AFFINITY UP).
FT                                /FTId=VAR_003088.
FT   VARIANT     144    144       K -&gt; E (IN MITO; O2 AFFINITY UP).
FT                                /FTId=VAR_003089.
FT   VARIANT     145    145       Y -&gt; C (IN RAINIER; O2 AFFINITY UP).
FT                                /FTId=VAR_003090.
FT   VARIANT     145    145       Y -&gt; H (IN BETHESDA; O2 AFFINITY UP).
FT                                /FTId=VAR_003091.
FT   VARIANT     146    146       H -&gt; D (IN HIROSHIMA; O2 AFFINITY UP).
FT                                /FTId=VAR_003092.
FT   VARIANT     146    146       H -&gt; L (IN COWTOWN; O2 AFFINITY UP).
FT                                /FTId=VAR_003093.
FT   VARIANT     146    146       H -&gt; P (IN YORK; O2 AFFINITY UP).
FT                                /FTId=VAR_003094.
FT   VARIANT     146    146       H -&gt; Q (IN KODAIRA; O2 AFFINITY UP).
FT                                /FTId=VAR_003095.
FT   HELIX         5     15
FT   TURN         16     17
FT   HELIX        20     34
FT   HELIX        36     41
FT   HELIX        43     45
FT   HELIX        51     55
FT   TURN         56     56
FT   HELIX        58     75
FT   TURN         76     77
FT   HELIX        78     94
FT   TURN         95     96
FT   TURN        100    100
FT   HELIX       101    121
FT   HELIX       124    142
FT   TURN        143    144
SQ   SEQUENCE   146 AA;  15867 MW;  EC9744C9 CRC32;
     VHLTPEEKSA VTALWGKVNV DEVGGEALGR LLVVYPWTQR FFESFGDLST PDAVMGNPKV
     KAHGKKVLGA FSDGLAHLDN LKGTFATLSE LHCDKLHVDP ENFRLLGNVL VCVLAHHFGK
     EFTPPVQAAY QKVVAGVANA LAHKYH
//
</pre>
</td></tr></table><p>


<H2>
    Output file format
</H2>

<p>

The output is a standard EMBOSS alignment file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-aformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  Some of the alignment formats can cope
with an unlimited number of sequences, while others are only for pairs
of sequences.  

<p>

The available multiple alignment format names are: unknown, multiple,
simple, fasta, msf, trace, srs

<p>

The available pairwise alignment format names are: pair, markx0, markx1,
markx2, markx3, markx10, srspair, score

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/AlignFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/AlignFormats.html</A>
for further information on alignment formats.

<p>


<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: hba_human.water</h3>
<table width="90%"><tr><td bgcolor="#DDFFDD">
<pre>
########################################
# Program:  water
# Rundate:  Thu Nov 07 14:48:47 2002
# Align_format: srspair
# Report_file: hba_human.water
########################################
#=======================================
#
# Aligned_sequences: 2
# 1: HBA_HUMAN
# 2: HBB_HUMAN
# Matrix: EBLOSUM62
# Gap_penalty: 10.0
# Extend_penalty: 0.5
#
# Length: 145
# Identity:      63/145 (43.4%)
# Similarity:    88/145 (60.7%)
# Gaps:           8/145 ( 5.5%)
# Score: 293.5
# 
#
#=======================================

HBA_HUMAN          2 LSPADKTNVKAAWGKVGAHAGEYGAEALERMFLSFPTTKTYFPHF-DLS-     49
                     |:|.:|:.|.|.||||  :..|.|.|||.|:.:.:|.|:.:|..| ||| 
HBB_HUMAN          3 LTPEEKSAVTALWGKV--NVDEVGGEALGRLLVVYPWTQRFFESFGDLST     50

HBA_HUMAN         50 ----HGSAQVKGHGKKVADALTNAVAHVDDMPNALSALSDLHAHKLRVDP     95
                         .|:.:||.|||||..|.::.:||:|::....:.||:||..||.|||
HBB_HUMAN         51 PDAVMGNPKVKAHGKKVLGAFSDGLAHLDNLKGTFATLSELHCDKLHVDP    100

HBA_HUMAN         96 VNFKLLSHCLLVTLAAHLPAEFTPAVHASLDKFLASVSTVLTSKY    140
                     .||:||.:.|:..||.|...||||.|.|:..|.:|.|:..|..||
HBB_HUMAN        101 ENFRLLGNVLVCVLAHHFGKEFTPPVQAAYQKVVAGVANALAHKY    145


#---------------------------------------
#---------------------------------------
</pre>
</td></tr></table><p>


<p>

The <b>Identity:</b> is the percentage of identical matches between the two
sequences over the reported aligned region (including any gaps in the length).

<p>

The <b>Similarity:</b> is the percentage of matches between the two
sequences over the reported aligned region (including any gaps in the length).


<H2>
    Data files
</H2>


For protein sequences EBLOSUM62 is used for the substitution
matrix. For nucleotide sequence, EDNAFULL is used. Others can be specified.

<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by EMBOSS environment variable EMBOSS_DATA.

<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>

<H2>
    Notes
</H2>

<b>water</b> is a true implementation of the Smith-Waterman algorithm
and so produces a full path matrix.  It therefore cannot be used with
genome sized sequences unless you have a <b>lot</b> of memory and a lot
of time. 

<H2>
    References
</H2>

<OL>
<LI>Smith TF, Waterman MS (1981) J. Mol. Biol 147(1);195-7
</OL>




<H2>
    Warnings
</H2>

Local alignment methods only report the best matching areas between two
sequences - there may be a large number of alternative local alignments
that do not score as highly.  If two proteins share more than one common
region, for example one has a single copy of a particular domain while
the other has two copies, it may be possible to "miss" the second and
subsequent alignments.  You will be able to see this situation if you
have done a dotplot and your local alignment does not show all the
features you expected to see. 
<P>

<b>water</b> is for aligning the best matching subsequences of two
sequences.  It does not necessarily align whole sequences against each
other; you should use <b>needle</b> if you wish to align closely related
sequences along their whole lengths. 
<P>

A true Smith Waterman implementation like <b>water</b> needs memory
proportional to the product of the sequence lengths.  For two sequences
of length 10,000,000 and 1,000 it therefore needs memory proportional to
10,000,000,000 characters.  Two arrays of this size are produced, one of
ints and one of floats so multiply that figure by 8 to get the memory
usage in bytes.  That doesn't include other overheads.  Therefore only
use water and needle for accurate alignment of reasonably short
sequences. 

<p>



If you run out of memory, try using <b>supermatcher</b> or
<b>matcher</b>.

<H2>
    Diagnostic Error Messages
</H2>

<PRE>
Uncaught exception
 Assertion failed
 raised at ajmem.c:xxx
</PRE>
<P>

Probably means you have run out of memory.  Try using
<b>supermatcher</b> or <b>matcher</b> if this happens. 

<H2>
    Exit status
</H2>

    0 if successful.

<H2>
    Known bugs
</H2>

None.


<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="matcher.html">matcher</a></td><td>Finds the best local alignments between two sequences</td></tr>
<tr><td><a href="seqmatchall.html">seqmatchall</a></td><td>Does an all-against-all comparison of a set of sequences</td></tr>
<tr><td><a href="supermatcher.html">supermatcher</a></td><td>Finds a match of a large sequence against one or more sequences</td></tr>
<tr><td><a href="wordmatch.html">wordmatch</a></td><td>Finds all exact matches of a given size between 2 sequences</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->

<p>

<b>matcher</b> is a local alignment program that gives as good an
alignment as <p>water</p> but it uses far less memory.  However,
<p>water</p> runs twice as fast as <b>matcher</b>. 

<p>

<b>supermatcher</b> is designed for local alignments of very large
sequences.  It gives good results as long as there is not significant
amounts of insertion or deletion in the alignment. 


<p>
<a href="supermatcher.html">supermatcher</a> Finds a match of a large sequence against one or more sequences
<a href="matcher.html">matcher</a> Finds the best local alignments between two sequences

<H2>
    Author(s)
</H2>


This application was written by Alan Bleasby <a
href="mailto:ableasby@hgmp.mrc.ac.uk">(ableasby@hgmp.mrc.ac.uk)</a>

<H2>
    History
</H2>

<pre>
Completed 7th July 1999.
Modified 27th July 1999 - tweaking scoring.
Modified 22 Oct 2000 - added ID and Similarity scores.
</pre>
<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

