<HTML>
<HEAD>
  <TITLE>
  EMBOSS: whichdb
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

whichdb
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<H2>
    Function
</H2>
Search all databases for an entry

<H2>
    Description
</H2>

<b>whichdb</b> searches all available EMBOSS databases for sequences
with a specified ID name or accession number.

<p>

EMBOSS does not allow database entries to be specified by just the ID
name or accession number.  You must specify both the name of the
database and the entry, for example 'embl:hsfau', not just 'hsfau'. 

<p>

If you do not know which database(s) contain the entry you are
interested in, <b>whichdb</b> will tell you.

<p>

<b>whichdb</b> can also, optionally, write out the sequences it finds to
files. 

<H3>IDs and Accessions</H3>

An entry in a database must have some way of being uniquely identified
in that database.  Most sequence databases have two such identifiers for
each sequence - an ID name and an Accession number. 

<p>

Why are there two such identifiers? The ID name was originally intended
to be a human-readable name that had some indication of the function of
its sequence.  In EMBL and GenBank the first two (or three) letters
indicated the species and the rest indicated the function, for example
'hsfau' is the 'Homo Sapiens FAU pseudogene'.  This naming scheme
started to be a problem when the number of entries added each day was so
vast that people could not make up the ID names fast enough.  Instead,
the Accession numbers were used as the ID name.  Therefore you will now
find ID names like 'AF061303', the same as the Accession number for that
sequence in EMBL. 

<p>

ID names are not guaranteed to remain the same between different
versions of a database (although in practice they usually do). 

<p>

Accession numbers are unique alphanumeric identifiers that are
guaranteed to remain with that sequence through the rest of the life of
the database.  If two sequences are merged into one, then the new
sequence will get a new Accession number and the Accession numbers of
the merged sequences will be retained as 'secondary' Accession numbers. 

<p>

EMBL, GenBank and SwissProt share an Accession numbering scheme - an
Accession number uniquely identifies a sequence within these three
databases. 


<H2>
    Usage
</H2>
<b>Here is a sample session with whichdb</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>whichdb -showall </b>
Search all databases for an entry
ID or Accession number: <b>hsfau1</b>
Output file [outfile.whichdb]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers (* if not always prompted):
  [-entry]             string     ID or Accession number
*  -outfile            outfile    Output file name

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers:
   -get                toggle     Retrieve sequences
   -showall            boolean    Show failed attempts

   Associated qualifiers:

   "-outfile" associated qualifiers
   -odirectory          string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-entry]<br>(Parameter 1)</td>
<td>ID or Accession number</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-outfile</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.whichdb</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-get</td>
<td>Retrieve sequences</td>
<td>Toggle value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-showall</td>
<td>Show failed attempts</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

</table>


<H2>
    Input file format
</H2>

None.

<p>



<H2>
    Output file format
</H2>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: outfile.whichdb</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
# Trying 'qangcgall:hsfau1'
# Failed 'qangcgall:hsfau1'
# Trying 'twp:hsfau1'
# Failed 'twp:hsfau1'
# Trying 'qapblast:hsfau1'
# Failed 'qapblast:hsfau1'
# Trying 'tsw:hsfau1'
# Failed 'tsw:hsfau1'
# Trying 'qapblastsplit:hsfau1'
# Failed 'qapblastsplit:hsfau1'
# Trying 'qangcginc:hsfau1'
# Failed 'qangcginc:hsfau1'
# Trying 'qangcgexc:hsfau1'
# Failed 'qangcgexc:hsfau1'
# Trying 'qasrsfasta:hsfau1'
qasrsfasta:hsfau1
# Trying 'qapir:hsfau1'
# Failed 'qapir:hsfau1'
# Trying 'tgenbank:hsfau1'
tgenbank:hsfau1
# Trying 'qanfasta:hsfau1'
# Failed 'qanfasta:hsfau1'
# Trying 'qanxflat:hsfau1'
# Failed 'qanxflat:hsfau1'
# Trying 'qanxgcg:hsfau1'
# Failed 'qanxgcg:hsfau1'
# Trying 'qapflat:hsfau1'
# Failed 'qapflat:hsfau1'
# Trying 'qapblastsplitinc:hsfau1'
# Failed 'qapblastsplitinc:hsfau1'
# Trying 'qapblastsplitexc:hsfau1'
# Failed 'qapblastsplitexc:hsfau1'
# Trying 'qapflatall:hsfau1'
# Failed 'qapflatall:hsfau1'
# Trying 'qanxflatall:hsfau1'
qanxflatall:hsfau1
# Trying 'qapflatinc:hsfau1'
# Failed 'qapflatinc:hsfau1'
# Trying 'qapflatexc:hsfau1'
# Failed 'qapflatexc:hsfau1'
# Trying 'qasrs:hsfau1'
qasrs:hsfau1
# Trying 'qapxflatinc:hsfau1'
# Failed 'qapxflatinc:hsfau1'
# Trying 'qapxflatexc:hsfau1'
# Failed 'qapxflatexc:hsfau1'
# Trying 'tpir:hsfau1'
# Failed 'tpir:hsfau1'
# Trying 'qasrswww:hsfau1'
qasrswww:hsfau1
# Trying 'qanxfasta:hsfau1'
# Failed 'qanxfasta:hsfau1'
# Trying 'qanxgcgall:hsfau1'
# Failed 'qanxgcgall:hsfau1'
# Trying 'qangcg:hsfau1'
# Failed 'qangcg:hsfau1'
# Trying 'qanfastaall:hsfau1'
# Failed 'qanfastaall:hsfau1'
# Trying 'qapirall:hsfau1'
# Failed 'qapirall:hsfau1'
# Trying 'qanxgcginc:hsfau1'
# Failed 'qanxgcginc:hsfau1'
# Trying 'qanxgcgexc:hsfau1'
# Failed 'qanxgcgexc:hsfau1'
# Trying 'qapirinc:hsfau1'
# Failed 'qapirinc:hsfau1'
# Trying 'qanflat:hsfau1'
# Failed 'qanflat:hsfau1'
# Trying 'qanxfastaall:hsfau1'
# Failed 'qanxfastaall:hsfau1'
# Trying 'qapfasta:hsfau1'
# Failed 'qapfasta:hsfau1'
# Trying 'qawxfasta:hsfau1'
# Failed 'qawxfasta:hsfau1'
# Trying 'qapirexc:hsfau1'
# Failed 'qapirexc:hsfau1'
# Trying 'tswnew:hsfau1'
# Failed 'tswnew:hsfau1'
# Trying 'qaxpirall:hsfau1'
# Failed 'qaxpirall:hsfau1'
# Trying 'qapxflat:hsfau1'
# Failed 'qapxflat:hsfau1'
# Trying 'qaxpirinc:hsfau1'
# Failed 'qaxpirinc:hsfau1'
# Trying 'qapxfasta:hsfau1'
# Failed 'qapxfasta:hsfau1'
# Trying 'tembl:hsfau1'
tembl:hsfau1
# Trying 'qaxpirexc:hsfau1'
# Failed 'qaxpirexc:hsfau1'
# Trying 'qapblastall:hsfau1'
# Failed 'qapblastall:hsfau1'
# Trying 'tgb:hsfau1'
# Failed 'tgb:hsfau1'
# Trying 'qaxpir:hsfau1'
# Failed 'qaxpir:hsfau1'
# Trying 'qawfasta:hsfau1'
# Failed 'qawfasta:hsfau1'
</pre>
</td></tr></table><p>


<p>

The reported database entries are written out to the specified output file.

<p>

Each database that is defined in the local implementation of EMBOSS
which contains an entry called 'hsfau' is reported.

<p>

Note that where a database is defined under several names ('embl',
'em'), each definition of that database containing the entry is
reported. 

<p>

When the '-get' option is used to output the sequences, they are written
to separate files.  The sequence format they are written in is always
'Fasta' format.  The normal command-line qualifiers for changing the
output formats etc.  will not work in this program. 

<p>

The names of the files that the sequences are written to is reported as
they are written, for example:

<p>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
Writing hsfau.embl
Writing hsfau.human
Writing hsfau.tembl
Writing hsfau.em
Writing hsfau.genbank
Writing hsfau.gb
</td></tr></table><p>



<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.


<H2>
    Exit status
</H2>

It always exits with status 0.

<H2>
    Known bugs
</H2>

None.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="infoalign.html">infoalign</a></td><td>Information on a multiple sequence alignment</td></tr>
<tr><td><a href="infoseq.html">infoseq</a></td><td>Displays some simple information about sequences</td></tr>
<tr><td><a href="seealso.html">seealso</a></td><td>Finds programs sharing group names</td></tr>
<tr><td><a href="showdb.html">showdb</a></td><td>Displays information on the currently available databases</td></tr>
<tr><td><a href="textsearch.html">textsearch</a></td><td>Search sequence documentation. Slow, use SRS and Entrez!</td></tr>
<tr><td><a href="tfm.html">tfm</a></td><td>Displays a program's help documentation manual</td></tr>
<tr><td><a href="wossname.html">wossname</a></td><td>Finds programs by keywords in their one-line documentation</td></tr>
</table>




<H2>
    Author(s)
</H2>
Alan Bleasby (ajb&nbsp;&copy;&nbsp;ebi.ac.uk)
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK



<H2>
    History
</H2>

Written (Jan 2002) - Alan Bleasby.



<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.


<H2>
    Comments
</H2>
None


</BODY>
</HTML>

