<HTML>

<HEAD>
  <TITLE>
  EMBOSS: wordmatch
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">







<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

wordmatch
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>







<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Finds all exact matches of a given size between 2 sequences


<H2>
    Description
</H2>

Finds all exact matches of a given minimum size between 2
    sequences displaying the start points in each sequence and the
    match length.
<p>

This program takes two sequences and finds regions where they are
identical.  These regions are reported in the output file (and
optionally) in GFF (Gene Feature Format) files. 
<p>

It will not find identical regions smaller than the specified <b>wordsize</b>.

<H2>
    Usage
</H2>


<b>Here is a sample session with wordmatch</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>wordmatch tsw:hba_human tsw:hbb_human </b>
Finds all exact matches of a given size between 2 sequences
Word size [4]: <b></b>
Output alignment [hba_human.wordmatch]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-asequence]         sequence   Sequence USA
  [-bsequence]         sequence   Sequence USA
   -wordsize           integer    Word size
  [-outfile]           align      Output alignment file name

   Optional qualifiers: (none)
   Advanced qualifiers:
   -afeatout           featout    File for output of normal tab delimited GFF
                                  features
   -bfeatout           featout    File for output of normal tab delimited GFF
                                  features

   Associated qualifiers:
  "-asequence" related qualifiers
   -sbegin1             integer    First base used
   -send1               integer    Last base used, def=seq length
   -sreverse1           boolean    Reverse (if DNA)
   -sask1               boolean    Ask for begin/end/reverse
   -snucleotide1        boolean    Sequence is nucleotide
   -sprotein1           boolean    Sequence is protein
   -slower1             boolean    Make lower case
   -supper1             boolean    Make upper case
   -sformat1            string     Input sequence format
   -sopenfile1          string     Input filename
   -sdbname1            string     Database name
   -sid1                string     Entryname
   -ufo1                string     UFO features
   -fformat1            string     Features format
   -fopenfile1          string     Features file name
  "-bsequence" related qualifiers
   -sbegin2             integer    First base used
   -send2               integer    Last base used, def=seq length
   -sreverse2           boolean    Reverse (if DNA)
   -sask2               boolean    Ask for begin/end/reverse
   -snucleotide2        boolean    Sequence is nucleotide
   -sprotein2           boolean    Sequence is protein
   -slower2             boolean    Make lower case
   -supper2             boolean    Make upper case
   -sformat2            string     Input sequence format
   -sopenfile2          string     Input filename
   -sdbname2            string     Database name
   -sid2                string     Entryname
   -ufo2                string     UFO features
   -fformat2            string     Features format
   -fopenfile2          string     Features file name
  "-outfile" related qualifiers
   -aformat3            string     Alignment format
   -aopenfile3          string     Alignment file name
   -aextension3         string     File name extension
   -adirectory3         string     Output directory
   -aname3              string     Base file name
   -awidth3             integer    Alignment width
   -aaccshow3           boolean    Show accession number in the header
   -adesshow3           boolean    Show description in the header
   -ausashow3           boolean    Show the full USA in the alignment
   -aglobal3            boolean    Show the full sequence in alignment
  "-afeatout" related qualifiers
   -offormat            string     Output feature format
   -ofopenfile          string     Features file name
   -ofextension         string     File name extension
   -ofdirectory         string     Output directory
   -ofname              string     Base file name
   -ofsingle            boolean    Separate file for each entry
  "-bfeatout" related qualifiers
   -offormat            string     Output feature format
   -ofopenfile          string     Features file name
   -ofextension         string     File name extension
   -ofdirectory         string     Output directory
   -ofname              string     Base file name
   -ofsingle            boolean    Separate file for each entry

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for required and optional values
   -debug               boolean    Write debug output to program.dbg
   -acdlog              boolean    Write ACD processing log to program.acdlog
   -acdpretty           boolean    Rewrite ACD file as program.acdpretty
   -acdtable            boolean    Write HTML table of options
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-asequence]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-bsequence]<br>(Parameter 2)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-wordsize</td>
<td>Word size</td>
<td>Integer 2 or more</td>
<td>4</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 3)</td>
<td>Output alignment file name</td>
<td>Alignment output file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-afeatout</td>
<td>File for output of normal tab delimited GFF features</td>
<td>Writeable feature table</td>
<td><i>unknown.gff</i></td>
</tr>

<tr>
<td>-bfeatout</td>
<td>File for output of normal tab delimited GFF features</td>
<td>Writeable feature table</td>
<td><i>unknown.gff</i></td>
</tr>

</table>



<H2>
    Input file format
</H2>


<b>wordmatch</b> reads any two sequence USAs of the same type (DNA or protein).

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>

'tsw:hba_human' is a sequence entry in the example protein database 'tsw'
<p>
<p><h3>Database entry: tsw:hba_human</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   HBA_HUMAN      STANDARD;      PRT;   141 AA.
AC   P01922;
DT   21-JUL-1986 (Rel. 01, Created)
DT   21-JUL-1986 (Rel. 01, Last sequence update)
DT   15-JUL-1999 (Rel. 38, Last annotation update)
DE   HEMOGLOBIN ALPHA CHAIN.
GN   HBA1 AND HBA2.
OS   Homo sapiens (Human), Pan troglodytes (Chimpanzee), and
OS   Pan paniscus (Pygmy chimpanzee) (Bonobo).
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Mammalia;
OC   Eutheria; Primates; Catarrhini; Hominidae; Homo.
RN   [1]
RP   SEQUENCE FROM N.A. (ALPHA-1).
RX   MEDLINE; 81088339.
RA   MICHELSON A.M., ORKIN S.H.;
RT   "The 3' untranslated regions of the duplicated human alpha-globin
RT   genes are unexpectedly divergent.";
RL   Cell 22:371-377(1980).
RN   [2]
RP   SEQUENCE FROM N.A. (ALPHA-2).
RX   MEDLINE; 81175088.
RA   LIEBHABER S.A., GOOSSENS M.J., KAN Y.W.;
RT   "Cloning and complete nucleotide sequence of human 5'-alpha-globin
RT   gene.";
RL   Proc. Natl. Acad. Sci. U.S.A. 77:7054-7058(1980).
RN   [3]
RP   SEQUENCE FROM N.A. (ALPHA-2).
RX   MEDLINE; 80137531.
RA   WILSON J.T., WILSON L.B., REDDY V.B., CAVALLESCO C., GHOSH P.K.,
RA   DERIEL J.K., FORGET B.G., WEISSMAN S.M.;
RT   "Nucleotide sequence of the coding portion of human alpha globin
RT   messenger RNA.";
RL   J. Biol. Chem. 255:2807-2815(1980).
RN   [4]
RP   SEQUENCE FROM N.A. (ALPHA-1 AND ALPHA-2).
RA   FLINT J., HIGGS D.R.;
RL   Submitted (JAN-1997) to the EMBL/GenBank/DDBJ databases.
RN   [5]
RP   SEQUENCE.
RA   BRAUNITZER G., GEHRING-MULLER R., HILSCHMANN N., HILSE K., HOBOM G.,
RA   RUDLOFF V., WITTMANN-LIEBOLD B.;
RT   "The constitution of normal adult human haemoglobin.";
RL   Hoppe-Seyler's Z. Physiol. Chem. 325:283-286(1961).
RN   [6]
RP   SEQUENCE.
RA   HILL R.J., KONIGSBERG W.;
RT   "The structure of human hemoglobin: IV. The chymotryptic digestion of
RT   the alpha chain of human hemoglobin.";
RL   J. Biol. Chem. 237:3151-3156(1962).
RN   [7]


<font color=red>  [Part of this file has been deleted for brevity]</font>

FT                                /FTId=VAR_002841.
FT   VARIANT     130    130       A -&gt; D (IN YUDA; O2 AFFINITY DOWN).
FT                                /FTId=VAR_002842.
FT   VARIANT     131    131       S -&gt; P (IN QUESTEMBERT; HIGHLY UNSTABLE;
FT                                CAUSES ALPHA-THALASSEMIA).
FT                                /FTId=VAR_002843.
FT   VARIANT     133    133       S -&gt; R (IN VAL DE MARNE; O2 AFFINITY UP).
FT                                /FTId=VAR_002844.
FT   VARIANT     135    135       V -&gt; E (IN PAVIE).
FT                                /FTId=VAR_002845.
FT   VARIANT     136    136       L -&gt; M (IN CHICAGO).
FT                                /FTId=VAR_002846.
FT   VARIANT     136    136       L -&gt; P (IN BIBBA; UNSTABLE;
FT                                CAUSES ALPHA-THALASSEMIA).
FT                                /FTId=VAR_002847.
FT   VARIANT     138    138       S -&gt; P (IN ATTLEBORO; O2 AFFINITY UP).
FT                                /FTId=VAR_002848.
FT   VARIANT     139    139       K -&gt; E (IN HANAKAMI; O2 AFFINITY UP).
FT                                /FTId=VAR_002849.
FT   VARIANT     139    139       K -&gt; T (IN TOKONAME; O2 AFFINITY UP).
FT                                /FTId=VAR_002850.
FT   VARIANT     140    140       Y -&gt; H (IN ROUEN; O2 AFFINITY UP).
FT                                /FTId=VAR_002851.
FT   VARIANT     141    141       R -&gt; C (IN NUNOBIKI; O2 AFFINITY UP).
FT                                /FTId=VAR_002852.
FT   VARIANT     141    141       R -&gt; L (IN LEGNANO; O2 AFFINITY UP).
FT                                /FTId=VAR_002853.
FT   VARIANT     141    141       R -&gt; H (IN SURESNES; O2 AFFINITY UP).
FT                                /FTId=VAR_002854.
FT   VARIANT     141    141       R -&gt; P (IN SINGAPORE).
FT                                /FTId=VAR_002855.
FT   HELIX         4     35
FT   HELIX        37     42
FT   TURN         44     45
FT   TURN         50     51
FT   HELIX        53     71
FT   TURN         72     74
FT   HELIX        76     79
FT   TURN         80     80
FT   HELIX        81     89
FT   TURN         90     91
FT   TURN         95     95
FT   HELIX        96    112
FT   TURN        114    116
FT   HELIX       119    136
FT   TURN        137    139
SQ   SEQUENCE   141 AA;  15126 MW;  5EC7DB1E CRC32;
     VLSPADKTNV KAAWGKVGAH AGEYGAEALE RMFLSFPTTK TYFPHFDLSH GSAQVKGHGK
     KVADALTNAV AHVDDMPNAL SALSDLHAHK LRVDPVNFKL LSHCLLVTLA AHLPAEFTPA
     VHASLDKFLA SVSTVLTSKY R
//
</pre>
</td></tr></table><p>
<p><h3>Database entry: tsw:hbb_human</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   HBB_HUMAN      STANDARD;      PRT;   146 AA.
AC   P02023;
DT   21-JUL-1986 (Rel. 01, Created)
DT   21-JUL-1986 (Rel. 01, Last sequence update)
DT   15-JUL-1999 (Rel. 38, Last annotation update)
DE   HEMOGLOBIN BETA CHAIN.
GN   HBB.
OS   Homo sapiens (Human), Pan troglodytes (Chimpanzee), and
OS   Pan paniscus (Pygmy chimpanzee) (Bonobo).
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Mammalia;
OC   Eutheria; Primates; Catarrhini; Hominidae; Homo.
RN   [1]
RP   SEQUENCE.
RC   SPECIES=HUMAN;
RA   BRAUNITZER G., GEHRING-MULLER R., HILSCHMANN N., HILSE K., HOBOM G.,
RA   RUDLOFF V., WITTMANN-LIEBOLD B.;
RT   "The constitution of normal adult human haemoglobin.";
RL   Hoppe-Seyler's Z. Physiol. Chem. 325:283-286(1961).
RN   [2]
RP   SEQUENCE FROM N.A.
RC   SPECIES=HUMAN;
RX   MEDLINE; 81064667.
RA   LAWN R.M., EFSTRATIADIS A., O'CONNELL C., MANIATIS T.;
RT   "The nucleotide sequence of the human beta-globin gene.";
RL   Cell 21:647-651(1980).
RN   [3]
RP   SEQUENCE OF 121-146 FROM N.A.
RC   SPECIES=HUMAN;
RX   MEDLINE; 85205333.
RA   LANG K.M., SPRITZ R.A.;
RT   "Cloning specific complete polyadenylylated 3'-terminal cDNA
RT   segments.";
RL   Gene 33:191-196(1985).
RN   [4]
RP   X-RAY CRYSTALLOGRAPHY (2.5 ANGSTROMS) OF DEOXYHEMOGLOBIN.
RC   SPECIES=HUMAN;
RX   MEDLINE; 76027820.
RA   FERMI G.;
RT   "Three-dimensional fourier synthesis of human deoxyhaemoglobin at
RT   2.5-A resolution: refinement of the atomic model.";
RL   J. Mol. Biol. 97:237-256(1975).
RN   [5]
RP   SEQUENCE.
RC   SPECIES=P.TROGLODYTES;
RX   MEDLINE; 66071496.
RA   RIFKIN D.B., KONIGSBERG W.;
RT   "The characterization of the tryptic peptides from the hemoglobin of
RT   the chimpanzee (Pan troglodytes).";
RL   Biochim. Biophys. Acta 104:457-461(1965).
RN   [6]


<font color=red>  [Part of this file has been deleted for brevity]</font>

FT   VARIANT     140    140       A -&gt; T (IN ST JACQUES: O2 AFFINITY UP).
FT                                /FTId=VAR_003081.
FT   VARIANT     140    140       A -&gt; V (IN PUTTELANGE; POLYCYTHEMIA;
FT                                O2 AFFINITY UP).
FT                                /FTId=VAR_003082.
FT   VARIANT     141    141       L -&gt; R (IN OLMSTED; UNSTABLE).
FT                                /FTId=VAR_003083.
FT   VARIANT     142    142       A -&gt; D (IN OHIO; O2 AFFINITY UP).
FT                                /FTId=VAR_003084.
FT   VARIANT     143    143       H -&gt; D (IN RANCHO MIRAGE).
FT                                /FTId=VAR_003085.
FT   VARIANT     143    143       H -&gt; Q (IN LITTLE ROCK; O2 AFFINITY UP).
FT                                /FTId=VAR_003086.
FT   VARIANT     143    143       H -&gt; P (IN SYRACUSE; O2 AFFINITY UP).
FT                                /FTId=VAR_003087.
FT   VARIANT     143    143       H -&gt; R (IN ABRUZZO; O2 AFFINITY UP).
FT                                /FTId=VAR_003088.
FT   VARIANT     144    144       K -&gt; E (IN MITO; O2 AFFINITY UP).
FT                                /FTId=VAR_003089.
FT   VARIANT     145    145       Y -&gt; C (IN RAINIER; O2 AFFINITY UP).
FT                                /FTId=VAR_003090.
FT   VARIANT     145    145       Y -&gt; H (IN BETHESDA; O2 AFFINITY UP).
FT                                /FTId=VAR_003091.
FT   VARIANT     146    146       H -&gt; D (IN HIROSHIMA; O2 AFFINITY UP).
FT                                /FTId=VAR_003092.
FT   VARIANT     146    146       H -&gt; L (IN COWTOWN; O2 AFFINITY UP).
FT                                /FTId=VAR_003093.
FT   VARIANT     146    146       H -&gt; P (IN YORK; O2 AFFINITY UP).
FT                                /FTId=VAR_003094.
FT   VARIANT     146    146       H -&gt; Q (IN KODAIRA; O2 AFFINITY UP).
FT                                /FTId=VAR_003095.
FT   HELIX         5     15
FT   TURN         16     17
FT   HELIX        20     34
FT   HELIX        36     41
FT   HELIX        43     45
FT   HELIX        51     55
FT   TURN         56     56
FT   HELIX        58     75
FT   TURN         76     77
FT   HELIX        78     94
FT   TURN         95     96
FT   TURN        100    100
FT   HELIX       101    121
FT   HELIX       124    142
FT   TURN        143    144
SQ   SEQUENCE   146 AA;  15867 MW;  EC9744C9 CRC32;
     VHLTPEEKSA VTALWGKVNV DEVGGEALGR LLVVYPWTQR FFESFGDLST PDAVMGNPKV
     KAHGKKVLGA FSDGLAHLDN LKGTFATLSE LHCDKLHVDP ENFRLLGNVL VCVLAHHFGK
     EFTPPVQAAY QKVVAGVANA LAHKYH
//
</pre>
</td></tr></table><p>



<H2>
    Output file format
</H2>

<p>

The output is a standard EMBOSS alignment file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-aformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  Some of the alignment formats can cope
with an unlimited number of sequences, while others are only for pairs
of sequences.  

<p>

The available multiple alignment format names are: unknown, multiple,
simple, fasta, msf, trace, srs

<p>

The available pairwise alignment format names are: pair, markx0, markx1,
markx2, markx3, markx10, srspair, score

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/AlignFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/AlignFormats.html</A>
for further information on alignment formats.

<p>


<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: HBA_HUMAN.gff</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
##gff-version 2.0
##date 2003-05-22
##Type Protein HBA_HUMAN
HBA_HUMAN	wordmatch	misc_feature	58	62	1.000	+	.	Sequence "HBA_HUMAN.1" ; note "HBB_HUMAN"
HBA_HUMAN	wordmatch	misc_feature	14	17	1.000	+	.	Sequence "HBA_HUMAN.2" ; note "HBB_HUMAN"
HBA_HUMAN	wordmatch	misc_feature	116	119	1.000	+	.	Sequence "HBA_HUMAN.3" ; note "HBB_HUMAN"
</pre>
</td></tr></table><p>
<p><h3>File: HBB_HUMAN.gff</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
##gff-version 2.0
##date 2003-05-22
##Type Protein HBB_HUMAN
HBB_HUMAN	wordmatch	misc_feature	63	67	1.000	+	.	Sequence "HBB_HUMAN.1" ; note "HBA_HUMAN"
HBB_HUMAN	wordmatch	misc_feature	15	18	1.000	+	.	Sequence "HBB_HUMAN.2" ; note "HBA_HUMAN"
HBB_HUMAN	wordmatch	misc_feature	121	124	1.000	+	.	Sequence "HBB_HUMAN.3" ; note "HBA_HUMAN"
</pre>
</td></tr></table><p>
<p><h3>File: hba_human.wordmatch</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
########################################
# Program:  wordmatch
# Rundate:  Thu May 22 11:00:13 2003
# Align_format: match
# Report_file: hba_human.wordmatch
########################################
#=======================================
#
# Aligned_sequences: 2
# 1: HBA_HUMAN
# 2: HBB_HUMAN
#=======================================

     5 HBA_HUMAN             58..62       HBB_HUMAN             63..67
     4 HBA_HUMAN             14..17       HBB_HUMAN             15..18
     4 HBA_HUMAN            116..119      HBB_HUMAN            121..124

#---------------------------------------
#---------------------------------------
</pre>
</td></tr></table><p>


<p>

The normal 'report' header is output.  It contains the details of the
program run and the input sequences. 

<p>

The data lines consist of five columns separated by spaces or TAB
characters.  Each line contains the information on one identical region. 
The first column is the length of the match.  The second column is the
name of the first sequence.  The third column is the start and end
position of the match.  The next two columns are the name and positions
of the second sequence. 


<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

None.


<H2>
    References
</H2>

None.


<H2>
    Warnings
</H2>

None.


<H2>
    Diagnostic Error Messages
</H2>

None.


<H2>
    Exit status
</H2>

    0 if successful.

<H2>
    Known bugs
</H2>

None.


<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="matcher.html">matcher</a></td><td>Finds the best local alignments between two sequences</td></tr>
<tr><td><a href="seqmatchall.html">seqmatchall</a></td><td>Does an all-against-all comparison of a set of sequences</td></tr>
<tr><td><a href="supermatcher.html">supermatcher</a></td><td>Finds a match of a large sequence against one or more sequences</td></tr>
<tr><td><a href="water.html">water</a></td><td>Smith-Waterman local alignment</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


Ian Longden (il&nbsp;&copy;&nbsp;sanger.ac.uk)
<br>
Sanger Institute, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.                      



<H2>
    History
</H2>

Completed 27th November 1998.

<!--
<p>
[an error occurred while processing this directive]
-->

<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

