<HTML>

<HEAD>
  <TITLE>
  EMBOSS: wordmatch
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center bgcolor="#000070" border=0 cellspacing=0 cellpadding=10>
<tr>
<td>

<a href="index.html" onmouseover="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="emboss_icon.gif" alt="" width=55 height=55>
</a>

</td>

<td align=middle valign=middle>
<font face="Arial,Helvetica" size=6 color="#ffffff">
<H2>
EMBOSS: wordmatch
</H2>


</font>
</td>

</tr>
</table>
</td>

</tr>
</table>

<p><hr><p>
<H1>
Program wordmatch
</H1>

<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Finds all exact matches of a given size between 2 sequences


<H2>
    Description
</H2>

Finds all exact matches of a given minimum size between 2
    sequences displaying the start points in each sequence and the
    match length.
<p>

This program takes two sequences and finds regions where they are
identical.  These regions are reported in the output file (and
optionally) in GFF (Gene Feature Format) files. 
<p>

It will not find identical regions smaller than the specified <b>wordsize</b>.

<H2>
    Usage
</H2>


Here is a sample session with wordmatch.

<pre>

% wordmatch sw:hba_human sw:hbb_human
Output file [hba_human.wordmatch]: 
Word size [4]: 

</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-asequence]         sequence   Sequence USA
  [-bsequence]         sequence   Sequence USA
   -wordsize           integer    Word size
  [-outfile]           outfile    Output file name

   Optional qualifiers: (none)
   Advanced qualifiers:
   -afeatout           featout    File for output of normal tab delimted gff's
   -bfeatout           featout    File for output of normal tab delimted gff's

   General qualifiers:
  -help                bool       report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-asequence]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-bsequence]<br>(Parameter 2)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-wordsize</td>
<td>Word size</td>
<td>Integer 2 or more</td>
<td>4</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 3)</td>
<td>Output file name</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.wordmatch</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-afeatout</td>
<td>File for output of normal tab delimted gff's</td>
<td>Writeable feature table</td>
<td><i>unknown.gff</i></td>
</tr>

<tr>
<td>-bfeatout</td>
<td>File for output of normal tab delimted gff's</td>
<td>Writeable feature table</td>
<td><i>unknown.gff</i></td>
</tr>

</table>



<H2>
    Input file format
</H2>


Any two sequence USAs of the same type (DNA or protein).

<H2>
    Output file format
</H2>


The file produced in the above example is:
<p>
<hr>
<pre>
FINALLY length = 3
 HBA_HUMAN  HBB_HUMAN Length
        58          63          5
        14          15          4
       116         121          4
</pre>
<hr>
<p>

The first line ('FINALLY...') gives the number of regions found.
<p>

The next line gives the headers for the subsequent columns of data.
This consists for the names of the two sequence and the word 'Length'.
<p>

Subsequent lines consist of three columns fo numbers separated by spaces
or TAB characters.  Each line contains the information on one identical
region.  The first column is the start position in the first sequence of
the identical region.  The second number is the start position in the
second sequence.  the third number is the length of the identical
region. 

<p>
If no regions are found, the output file is blank.

<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

None.


<H2>
    References
</H2>

None.


<H2>
    Warnings
</H2>

None.


<H2>
    Diagnostic Error Messages
</H2>

None.


<H2>
    Exit status
</H2>

    0 if successfull.

<H2>
    Known bugs
</H2>

None.


<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="matcher.html">matcher</a></td><td>Finds the best local alignments between two sequences</td></tr>
<tr><td><a href="seqmatchall.html">seqmatchall</a></td><td>Does an all-against-all comparison of a set of sequences</td></tr>
<tr><td><a href="supermatcher.html">supermatcher</a></td><td>Finds a match of a large sequence against one or more sequences</td></tr>
<tr><td><a href="water.html">water</a></td><td>Smith-Waterman local alignment</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Ian Longden <a
href="mailto:il@sanger.ac.uk">(il@sanger.ac.uk)</a> Informatics
Division, The Sanger Centre, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.

<H2>
    History
</H2>

Completed 27th November 1998.
<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</TD></TR></TABLE>
</CENTER>

</BODY>
</HTML>

