<HTML>

<HEAD>
  <TITLE>
  EMBOSS: wordmatch
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">







<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

wordmatch
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>







<H2>
    Function
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->
Finds all exact matches of a given size between 2 sequences


<H2>
    Description
</H2>

Finds all exact matches of a given minimum size between 2
    sequences displaying the start points in each sequence and the
    match length.
<p>

This program takes two sequences and finds regions where they are
identical.  These regions are reported in the output file (and
optionally) in GFF (Gene Feature Format) files. 
<p>

It will not find identical regions smaller than the specified <b>wordsize</b>.

<H2>
    Usage
</H2>


Here is a sample session with wordmatch.

<p>

<pre>
% wordmatch tsw:hba_human tsw:hbb_human
Finds all exact matches of a given size between 2 sequences
Word size [4]:
Output alignment [hba_human.wordmatch]: 
</pre>



<H2>
    Command line arguments
</H2>



<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-asequence]         sequence   Sequence USA
  [-bsequence]         sequence   Sequence USA
   -wordsize           integer    Word size
  [-outfile]           align      Output alignment file name

   Optional qualifiers: (none)
   Advanced qualifiers:
   -afeatout           featout    File for output of normal tab delimited GFF
                                  features
   -bfeatout           featout    File for output of normal tab delimited GFF
                                  features

   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>

<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-asequence]<br>(Parameter 1)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-bsequence]<br>(Parameter 2)</td>
<td>Sequence USA</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-wordsize</td>
<td>Word size</td>
<td>Integer 2 or more</td>
<td>4</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 3)</td>
<td>Output alignment file name</td>
<td>Alignment file</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-afeatout</td>
<td>File for output of normal tab delimited GFF features</td>
<td>Writeable feature table</td>
<td><i>unknown.gff</i></td>
</tr>

<tr>
<td>-bfeatout</td>
<td>File for output of normal tab delimited GFF features</td>
<td>Writeable feature table</td>
<td><i>unknown.gff</i></td>
</tr>

</table>



<H2>
    Input file format
</H2>


Any two sequence USAs of the same type (DNA or protein).

<H2>
    Output file format
</H2>

<p>

The output is a standard EMBOSS alignment file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-aformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  Some of the alignment formats can cope
with an unlimited number of sequences, while others are only for pairs
of sequences.  

<p>

The available multiple alignment format names are: unknown, multiple,
simple, fasta, msf, trace, srs

<p>

The available pairwise alignment format names are: pair, markx0, markx1,
markx2, markx3, markx10, srspair, score

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/AlignFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/AlignFormats.html</A>
for further information on alignment formats.

<p>


<p>

The file produced in the above example is:

<p>

<hr>
<pre>
########################################
# Program:  wordmatch
# Rundate:  Mon May 20 16:36:46 2002
# Report_file: hba_human.wordmatch
########################################
#=======================================
#
# Aligned_sequences: 2
# 1: HBA_HUMAN
# 2: HBB_HUMAN
#=======================================

     5 HBA_HUMAN             58..62       HBB_HUMAN             63..67
     4 HBA_HUMAN             14..17       HBB_HUMAN             15..18
     4 HBA_HUMAN            116..119      HBB_HUMAN            121..124

#---------------------------------------
#--------------------------------------- 
</pre>
<hr>
<p>

The normal 'report' header is output.  It contains the details of the
program run and the input sequences. 

<p>

The data lines consist of five columns separated by spaces or TAB
characters.  Each line contains the information on one identical region. 
The first column is the length of the match.  The second column is the
name of the first sequence.  The third column is the start and end
position of the match.  The next two columns are the name and positions
of the second sequence. 


<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

None.


<H2>
    References
</H2>

None.


<H2>
    Warnings
</H2>

None.


<H2>
    Diagnostic Error Messages
</H2>

None.


<H2>
    Exit status
</H2>

    0 if successful.

<H2>
    Known bugs
</H2>

None.


<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="matcher.html">matcher</a></td><td>Finds the best local alignments between two sequences</td></tr>
<tr><td><a href="seqmatchall.html">seqmatchall</a></td><td>Does an all-against-all comparison of a set of sequences</td></tr>
<tr><td><a href="supermatcher.html">supermatcher</a></td><td>Finds a match of a large sequence against one or more sequences</td></tr>
<tr><td><a href="water.html">water</a></td><td>Smith-Waterman local alignment</td></tr>
</table>

<!--
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find.
   -->


<H2>
    Author(s)
</H2>


This application was written by Ian Longden <a
href="mailto:il@sanger.ac.uk">(il@sanger.ac.uk)</a> Informatics
Division, The Sanger Centre, Wellcome Trust Genome Campus, Hinxton,
Cambridge, CB10 1SA, UK.

<H2>
    History
</H2>

Completed 27th November 1998.
<H2>
    Target users
</H2>


This program is intended to be used by everyone and everything,
from naive users to embedded scripts. 

<H2>
    Comments
</H2>


<pre>

</pre>


</BODY>
</HTML>

