<HTML>

<HEAD>
  <TITLE>
  EMBOSS: wossname
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

wossname
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<H2>
    Function
</H2>

Finds programs by keywords in their one-line documentation


<H2>
    Description
</H2>


This allows a user to search for keywords or parts of words in the
brief documentation (as displayed by a program when it first starts).
The program name and the brief description is output.  If no words to
search for are specified, then details of all the EMBOSS programs are
output.

<p>

The program has been written on the assumption that most people will
use it to quickly find the name of a program based on that program's
description, so the output goes to the screen by default.

<p>

This program may find some use in automatically generating lists of
EMBOSS programs and their groups for Web pages.

<H2>
    Usage
</H2>
<b>Here is a sample session with wossname</b>
<p>
Search for programs with 'restrict' in their description: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>wossname restrict </b>
Finds programs by keywords in their one-line documentation

SEARCH FOR 'RESTRICT'
erestml          Restriction site Maximum Likelihood method
frestboot        Bootstrapped restriction sites algorithm
frestdist        Computes distance matrix from restriction sites or fragments
frestml          Restriction site maximum Likelihood method
recoder          Remove restriction sites but maintain the same translation
remap            Display a sequence with restriction cut sites, translation etc
restover         Finds restriction enzymes that produce a specific overhang
restrict         Finds restriction enzyme cleavage sites
silent           Silent mutation restriction enzyme scan


</pre></td></tr></table><p>
<p>
<p>
<p>
<b>Example 2</b>
<p>
Display a listing of programs in their groups: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>wossname -search '' </b>
Finds programs by keywords in their one-line documentation

ACD
acdc             ACD compiler
acdpretty        ACD pretty printing utility
acdtable         Creates an HTML table from an ACD file
acdtrace         ACD compiler on-screen trace
acdvalid         ACD file validation

ALIGNMENT CONSENSUS
cons             Creates a consensus from multiple alignments
megamerger       Merge two large overlapping nucleic acid sequences
merger           Merge two overlapping nucleic acid sequences

ALIGNMENT DIFFERENCES
diffseq          Find differences between nearly identical sequences

ALIGNMENT DOT PLOTS
dotmatcher       Displays a thresholded dotplot of two sequences
dotpath          Displays a non-overlapping wordmatch dotplot of two sequences
dottup           Displays a wordmatch dotplot of two sequences
polydot          Displays all-against-all dotplots of a set of sequences

ALIGNMENT GLOBAL
alignrunner      Runs alignment program on all sequence pairs in a directory
contactalign     EMBOSS implementation of Damian Counsell's 2.5-D alignment algorithm
esim4            Align an mRNA to a genomic DNA sequence
est2genome       Align EST and genomic DNA sequences
nawalign         Damian Counsell's implementation for protein sequences of the Needleman and Wunsch alignment algorithm
nawalignrunner   Runs nawalign alignment program on all sequence pairs in a directory
needle           Needleman-Wunsch global alignment
needlerunner     Runs needle alignment program on all sequence pairs in a directory
scorer           Scores accuracy of protein-protein sequence alignment against gold standard structure-structure alignment
scorerrunner     Runs scorer to compare ordered pairs of substituted seqs in two directories and write the scores to a third
stretcher        Finds the best global alignment between two sequences
substitute       Substitutes matches from first (query) sequence of two aligned sequences in a trace into second (template sequence)
substituterunner Runs substitute on directory of traces and writes substituted sequences to another

ALIGNMENT LOCAL
matcher          Finds the best local alignments between two sequences
seqmatchall      Does an all-against-all comparison of a set of sequences
supermatcher     Finds a match of a large sequence against one or more sequences
water            Smith-Waterman local alignment
wordmatch        Finds all exact matches of a given size between 2 sequences

ALIGNMENT MULTIPLE
emma             Multiple alignment program - interface to ClustalW program
infoalign        Information on a multiple sequence alignment
mse              Multiple Sequence Editor
plotcon          Plots the quality of conservation of a sequence alignment
prettyplot       Displays aligned sequences, with colouring and boxing
showalign        Displays a multiple sequence alignment
tranalign        Align nucleic coding regions given the aligned proteins

DISPLAY
abiview          Reads ABI file and display the trace
cirdna           Draws circular maps of DNA constructs
lindna           Draws linear maps of DNA constructs
pepnet           Displays proteins as a helical net
pepwheel         Shows protein sequences as helices
prettyplot       Displays aligned sequences, with colouring and boxing
prettyseq        Output sequence with translated ranges
remap            Display a sequence with restriction cut sites, translation etc
seealso          Finds programs sharing group names
showalign        Displays a multiple sequence alignment
showdb           Displays information on the currently available databases
showfeat         Show features of a sequence
showseq          Display a sequence with features, translation etc
sixpack          Display a DNA sequence with 6-frame translation and ORFs
textsearch       Search sequence documentation text. SRS and Entrez are faster!

EDIT
biosed           Replace or delete sequence sections
cutseq           Removes a specified section from a sequence
degapseq         Removes gap characters from sequences
degapseqrunner   Runs degapseq program on all sequence pairs in a directory
descseq          Alter the name or description of a sequence
entret           Reads and writes (returns) flatfile entries
extractfeat      Extract features from a sequence
extractseq       Extract regions from a sequence
listor           Writes a list file of the logical OR of two sets of sequences
maskfeat         Mask off features of a sequence
maskseq          Mask off regions of a sequence
newseq           Type in a short new sequence
noreturn         Removes carriage return from ASCII files
notseq           Excludes a set of sequences and writes out the remaining ones
nthseq           Writes one sequence from a multiple set of sequences
pasteseq         Insert one sequence into another
revseq           Reverse and complement a sequence
seqret           Reads and writes (returns) sequences
seqretsplit      Reads and writes (returns) sequences in individual files
skipseq          Reads and writes (returns) sequences, skipping the first few
splitter         Split a sequence into (overlapping) smaller sequences
trimest          Trim poly-A tails off EST sequences
trimseq          Trim ambiguous bits off the ends of sequences
union            Reads sequence fragments and builds one sequence
vectorstrip      Strips out DNA between a pair of vector sequences
yank             Reads a sequence range, appends the full USA to a list file

ENZYME KINETICS
findkm           Find Km and Vmax for an enzyme reaction by a Hanes/Woolf plot

FEATURE TABLES
coderet          Extract CDS, mRNA and translations from feature tables
extractfeat      Extract features from a sequence
maskfeat         Mask off features of a sequence
showfeat         Show features of a sequence
twofeat          Finds neighbouring pairs of features in sequences

HMM
ealistat         Statistics for multiple alignment files
ehmmalign        Align sequences with an HMM
ehmmbuild        Build HMM
ehmmcalibrate    Calibrate a hidden Markov model
ehmmconvert      Convert between HMM formats
ehmmemit         Extract HMM sequences
ehmmfetch        Extract HMM from a database
ehmmindex        Index an HMM database
ehmmpfam         Align single sequence with an HMM
ehmmsearch       Search sequence database with an HMM

INFORMATION
infoalign        Information on a multiple sequence alignment
infoseq          Displays some simple information about sequences
seealso          Finds programs sharing group names
showdb           Displays information on the currently available databases
textsearch       Search sequence documentation text. SRS and Entrez are faster!
tfm              Displays a program's help documentation manual
whichdb          Search all databases for an entry
wossname         Finds programs by keywords in their one-line documentation

MENUS
emnu             Simple menu of EMBOSS applications

NUCLEIC 2D STRUCTURE
einverted        Finds DNA inverted repeats

NUCLEIC CODON USAGE
cai              CAI codon adaptation index
chips            Codon usage statistics
codcmp           Codon usage table comparison
cusp             Create a codon usage table
syco             Synonymous codon usage Gribskov statistic plot

NUCLEIC COMPOSITION
banana           Bending and curvature plot in B-DNA
btwisted         Calculates the twisting in a B-DNA sequence
chaos            Create a chaos game representation plot for a sequence
compseq          Counts the composition of dimer/trimer/etc words in a sequence
dan              Calculates DNA RNA/DNA melting temperature
freak            Residue/base frequency table or plot
isochore         Plots isochores in large DNA sequences
sirna            Finds siRNA duplexes in mRNA
wordcount        Counts words of a specified size in a DNA sequence

NUCLEIC CPG ISLANDS
cpgplot          Plot CpG rich areas
cpgreport        Reports all CpG rich regions
geecee           Calculates the fractional GC content of nucleic acid sequences
newcpgreport     Report CpG rich areas
newcpgseek       Reports CpG rich regions

NUCLEIC GENE FINDING
getorf           Finds and extracts open reading frames (ORFs)
marscan          Finds MAR/SAR sites in nucleic sequences
plotorf          Plot potential open reading frames
showorf          Pretty output of DNA translations
sixpack          Display a DNA sequence with 6-frame translation and ORFs
syco             Synonymous codon usage Gribskov statistic plot
tcode            Fickett TESTCODE statistic to identify protein-coding DNA
wobble           Wobble base plot

NUCLEIC MOTIFS
dreg             Regular expression search of a nucleotide sequence
fuzznuc          Nucleic acid pattern search
fuzztran         Protein pattern search after translation
marscan          Finds MAR/SAR sites in nucleic sequences

NUCLEIC MUTATION
msbar            Mutate sequence beyond all recognition
shuffleseq       Shuffles a set of sequences maintaining composition

NUCLEIC PRIMERS
eprimer3         Picks PCR primers and hybridization oligos
primersearch     Searches DNA sequences for matches with primer pairs
stssearch        Searches a DNA database for matches with a set of STS primers

NUCLEIC PROFILES
profit           Scan a sequence or database with a matrix or profile
prophecy         Creates matrices/profiles from multiple alignments
prophet          Gapped alignment for profiles

NUCLEIC REPEATS
einverted        Finds DNA inverted repeats
equicktandem     Finds tandem repeats
etandem          Looks for tandem repeats in a nucleotide sequence
palindrome       Looks for inverted repeats in a nucleotide sequence

NUCLEIC RESTRICTION
recoder          Remove restriction sites but maintain the same translation
redata           Search REBASE for enzyme name, references, suppliers etc
remap            Display a sequence with restriction cut sites, translation etc
restover         Finds restriction enzymes that produce a specific overhang
restrict         Finds restriction enzyme cleavage sites
showseq          Display a sequence with features, translation etc
silent           Silent mutation restriction enzyme scan

NUCLEIC TRANSCRIPTION
tfscan           Scans DNA sequences for transcription factors

NUCLEIC TRANSLATION
backtranseq      Back translate a protein sequence
coderet          Extract CDS, mRNA and translations from feature tables
plotorf          Plot potential open reading frames
prettyseq        Output sequence with translated ranges
remap            Display a sequence with restriction cut sites, translation etc
showorf          Pretty output of DNA translations
showseq          Display a sequence with features, translation etc
sixpack          Display a DNA sequence with 6-frame translation and ORFs
transeq          Translate nucleic acid sequences

PHYLOGENY CONSENSUS
econsense        Majority-rule and strict consensus tree
fconsense        Majority-rule and strict consensus tree
ftreedist        Distances between trees
ftreedistpair    Distances between two sets of trees

PHYLOGENY CONTINUOUS CHARACTERS
econtml          Continuous character Maximum Likelihood method
econtrast        Continuous character Contrasts
fcontrast        Continuous character Contrasts

PHYLOGENY DISCRETE CHARACTERS
eclique          Largest clique program
edollop          Dollo and polymorphism parsimony algorithm
edolpenny        Penny algorithm Dollo or polymorphism
efactor          Multistate to binary recoding program
emix             Mixed parsimony algorithm
epenny           Penny algorithm, branch-and-bound to find all most parsimonious trees
fclique          Largest clique program
fdollop          Dollo and polymorphism parsimony algorithm
fdolpenny        Penny algorithm Dollo or polymorphism
ffactor          Multistate to binary recoding program
fmix             Mixed parsimony algorithm
fmove            Interactive mixed method parsimony
fpars            Discrete character parsimony
fpenny           Penny algorithm, branch-and-bound to find all most parsimonious trees

PHYLOGENY DISTANCE MATRIX
distmat          Creates a distance matrix from multiple alignments
efitch           Fitch-Margoliash and Least-Squares Distance Methods
ekitsch          Fitch-Margoliash method with contemporary tips
eneighbor        Phylogenies from distance matrix by N-J or UPGMA method
ffitch           Fitch-Margoliash and Least-Squares Distance Methods
fkitsch          Fitch-Margoliash method with contemporary tips
fneighbor        Phylogenies from distance matrix by N-J or UPGMA method

PHYLOGENY GENE FREQUENCIES
egendist         Genetic Distance Matrix program
fcontml          Gene frequency and continuous character Maximum Likelihood method
fgendist         Compute genetic distances from gene frequencies

PHYLOGENY MOLECULAR SEQUENCE
ednacomp         DNA compatibility algorithm
ednadist         Nucleic acid sequence Distance Matrix program
ednainvar        Nucleic acid sequence Invariants method
ednaml           Estimates phylogenies from nucleic acid sequence Maximum Likelihood
ednamlk          Estimates phylogenies from nucleic acid sequence Maximum Likelihood with molecular clock
ednapars         DNA parsimony algorithm
ednapenny        Penny algorithm for DNA
eprotdist        Protein distance algorithm
eprotpars        Protein parsimony algorithm
erestml          Restriction site Maximum Likelihood method
eseqboot         Bootstrapped sequences algorithm
fdiscboot        Bootstrapped discrete sites algorithm
fdnacomp         DNA compatibility algorithm
fdnadist         Nucleic acid sequence Distance Matrix program
fdnainvar        Nucleic acid sequence Invariants method
fdnaml           Estimates nucleotide phylogeny by maximum likelihood
fdnamlk          Estimates nucleotide phylogeny by maximum likelihood
fdnamove         Interactive DNA parsimony
fdnapars         DNA parsimony algorithm
fdnapenny        Penny algorithm for DNA
fdolmove         Interactive Dollo or Polymorphism Parsimony
ffreqboot        Bootstrapped genetic frequencies algorithm
fproml           Protein phylogeny by maximum likelihood
fpromlk          Protein phylogeny by maximum likelihood
fprotdist        Protein distance algorithm
fprotpars        Protein pasimony algorithm
frestboot        Bootstrapped restriction sites algorithm
frestdist        Computes distance matrix from restriction sites or fragments
frestml          Restriction site maximum Likelihood method
fseqboot         Bootstrapped sequences algorithm
fseqbootall      Bootstrapped sequences algorithm

PHYLOGENY TREE DRAWING
fdrawgram        Plots a cladogram- or phenogram-like rooted tree diagram
fdrawtree        Plots an unrooted tree diagram
fretree          Interactive tree rearrangement

PROTEIN 2D STRUCTURE
garnier          Predicts protein secondary structure
helixturnhelix   Report nucleic acid binding motifs
hmoment          Hydrophobic moment calculation
pepcoil          Predicts coiled coil regions
pepnet           Displays proteins as a helical net
pepwheel         Shows protein sequences as helices
tmap             Displays membrane spanning regions
topo             Draws an image of a transmembrane protein

PROTEIN 3D STRUCTURE
contactcount     Counts specific versus non-specific contacts in a directory of cleaned protein chain contact files
contacts         Reads CCF files (clean coordinate files) and writes CON files (contact files) of intra-chain residue-residue contact data
domainalign      Generates DAF files (domain alignment files) of structure-based sequence alignments for nodes in a DCF file (domain classification file)
domainrep        Reorder DCF file (domain classification file) so that the representative structure of each user-specified node is given first
domainreso       Removes low resolution domains from a DCF file (domain classification file)
interface        Reads CCF files (clean coordinate files) and writes CON files (contact files) of inter-chain residue-residue contact data
libgen           Generates various types of discriminating elements for each alignment in a directory
psiphi           Calculates phi and psi torsion angles from cleaned EMBOSS-style protein co-ordinate file
rocon            Reads a DHF file (domain hits file) of hits (sequences of unknown structural classification) and a DHF file of validation sequences (known classification) and writes a 'hits file' for the hits, which are classified and rank-ordered on the basis of score
rocplot          Provides interpretation and graphical display of the performance of discriminating elements (e.g. profiles for protein families). rocplot reads file(s) of hits from discriminator-database search(es), performs ROC analysis on the hits, and writes graphs illustrating the diagnostic performance of the discriminating elements
seqalign         Reads a DAF file (domain alignment file) and a DHF file (domain hits file) and writes a DAF file extended with the hits
seqfraggle       Removes fragments from DHF files (domain hits files) or other files of sequences
seqsearch        Generate database hits (sequences) for nodes in a DCF file (domain classification file) by using PSI-BLAST
seqsort          Reads DHF files (domain hits files) of database hits (sequences) and removes hits of ambiguous classification
seqwords         Generates DHF files (domain hits files) of database hits (sequences) for nodes in a DCF file (domain classification file) by keyword search of UniProt
siggen           Generates a sparse protein signature from an alignment and residue contact data
sigscan          Generates a DHF file (domain hits file) of hits (sequences) from scanning a signature against a sequence database

PROTEIN COMPOSITION
backtranseq      Back translate a protein sequence
charge           Protein charge plot
checktrans       Reports STOP codons and ORF statistics of a protein
compseq          Counts the composition of dimer/trimer/etc words in a sequence
emowse           Protein identification by mass spectrometry
freak            Residue/base frequency table or plot
iep              Calculates the isoelectric point of a protein
mwcontam         Shows molwts that match across a set of files
mwfilter         Filter noisy molwts from mass spec output
octanol          Displays protein hydropathy
pepinfo          Plots simple amino acid properties in parallel
pepstats         Protein statistics
pepwindow        Displays protein hydropathy
pepwindowall     Displays protein hydropathy of a set of sequences

PROTEIN MOTIFS
antigenic        Finds antigenic sites in proteins
digest           Protein proteolytic enzyme or reagent cleavage digest
epestfind        Finds PEST motifs as potential proteolytic cleavage sites
fuzzpro          Protein pattern search
fuzztran         Protein pattern search after translation
helixturnhelix   Report nucleic acid binding motifs
meme             Motif detection
oddcomp          Finds protein sequence regions with a biased composition
patmatdb         Search a protein sequence with a motif
patmatmotifs     Search a PROSITE motif database with a protein sequence
pepcoil          Predicts coiled coil regions
preg             Regular expression search of a protein sequence
pscan            Scans proteins using PRINTS
sigcleave        Reports protein signal cleavage sites

PROTEIN MUTATION
msbar            Mutate sequence beyond all recognition
shuffleseq       Shuffles a set of sequences maintaining composition

PROTEIN PROFILES
profit           Scan a sequence or database with a matrix or profile
prophecy         Creates matrices/profiles from multiple alignments
prophet          Gapped alignment for profiles

TEST
crystalball      Answers every drug discovery question you have about this sequence
myseq            Demonstration of sequence reading

UTILS DATABASE CREATION
aaindexextract   Extract data from AAINDEX
allversusall     Does an all-versus-all global alignment for each set of sequences in an input directory and writes files of sequence similarity values
cathparse        Reads raw CATH classification files and writes DCF file (domain classification file)
cutgextract      Extract data from CUTG
domainer         Reads CCF files (clean coordinate files) for proteins and writes CCF files for domains, taken from a DCF file (domain classification file)
domainnr         Removes redundant domains from a DCF file (domain classification file). The file must contain domain sequence information, which can be added by using DOMAINSEQS
domainseqs       Adds sequence records to a DCF file (domain classification file)
domainsse        Adds secondary structure records to a DCF file (domain classification file)
hetparse         Converts raw dictionary of heterogen groups to a file in EMBL-like format
pdbparse         Parses PDB files and writes CCF files (clean coordinate files) for proteins
pdbplus          Add residue solvent accessibility and secondary structure data to a CCF file (clean coordinate file) for a protein or domain
pdbtosp          Convert raw swissprot:PDB equivalence file to EMBL-like format
printsextract    Extract data from PRINTS
prosextract      Builds the PROSITE motif database for patmatmotifs to search
rebaseextract    Extract data from REBASE
scopparse        Reads raw SCOP classification files and writes a DCF file (domain classification file)
seqnr            Removes redundancy from DHF files (domain hits files) or other files of sequences
sites            Reads CCF files (clean coordinate files) and writes CON files (contact files) of residue-ligand contact data for domains in a DCF file (domain classification file)
ssematch         Searches a DCF file (domain classification file) for secondary structure matches
tfextract        Extract data from TRANSFAC

UTILS DATABASE INDEXING
dbiblast         Index a BLAST database
dbifasta         Index a fasta database
dbiflat          Index a flat file database
dbigcg           Index a GCG formatted database

UTILS MISC
embossdata       Finds or fetches the data files read in by the EMBOSS programs
embossversion    Writes the current EMBOSS version number


</pre></td></tr></table><p>
<p>
<p>
<p>
<b>Example 3</b>
<p>
Display an alphabetic listing of all programs: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>wossname -search '' -alphabetic </b>
Finds programs by keywords in their one-line documentation

ALPHABETIC LIST OF PROGRAMS
aaindexextract   Extract data from AAINDEX
abiview          Reads ABI file and display the trace
acdc             ACD compiler
acdpretty        ACD pretty printing utility
acdtable         Creates an HTML table from an ACD file
acdtrace         ACD compiler on-screen trace
acdvalid         ACD file validation
alignrunner      Runs alignment program on all sequence pairs in a directory
allversusall     Does an all-versus-all global alignment for each set of sequences in an input directory and writes files of sequence similarity values
antigenic        Finds antigenic sites in proteins
backtranseq      Back translate a protein sequence
banana           Bending and curvature plot in B-DNA
biosed           Replace or delete sequence sections
btwisted         Calculates the twisting in a B-DNA sequence
cai              CAI codon adaptation index
cathparse        Reads raw CATH classification files and writes DCF file (domain classification file)
chaos            Create a chaos game representation plot for a sequence
charge           Protein charge plot
checktrans       Reports STOP codons and ORF statistics of a protein
chips            Codon usage statistics
cirdna           Draws circular maps of DNA constructs
codcmp           Codon usage table comparison
coderet          Extract CDS, mRNA and translations from feature tables
compseq          Counts the composition of dimer/trimer/etc words in a sequence
cons             Creates a consensus from multiple alignments
contactalign     EMBOSS implementation of Damian Counsell's 2.5-D alignment algorithm
contactcount     Counts specific versus non-specific contacts in a directory of cleaned protein chain contact files
contacts         Reads CCF files (clean coordinate files) and writes CON files (contact files) of intra-chain residue-residue contact data
cpgplot          Plot CpG rich areas
cpgreport        Reports all CpG rich regions
crystalball      Answers every drug discovery question you have about this sequence
cusp             Create a codon usage table
cutgextract      Extract data from CUTG
cutseq           Removes a specified section from a sequence
dan              Calculates DNA RNA/DNA melting temperature
dbiblast         Index a BLAST database
dbifasta         Index a fasta database
dbiflat          Index a flat file database
dbigcg           Index a GCG formatted database
degapseq         Removes gap characters from sequences
degapseqrunner   Runs degapseq program on all sequence pairs in a directory
descseq          Alter the name or description of a sequence
diffseq          Find differences between nearly identical sequences
digest           Protein proteolytic enzyme or reagent cleavage digest
distmat          Creates a distance matrix from multiple alignments
domainalign      Generates DAF files (domain alignment files) of structure-based sequence alignments for nodes in a DCF file (domain classification file)
domainer         Reads CCF files (clean coordinate files) for proteins and writes CCF files for domains, taken from a DCF file (domain classification file)
domainnr         Removes redundant domains from a DCF file (domain classification file). The file must contain domain sequence information, which can be added by using DOMAINSEQS
domainrep        Reorder DCF file (domain classification file) so that the representative structure of each user-specified node is given first
domainreso       Removes low resolution domains from a DCF file (domain classification file)
domainseqs       Adds sequence records to a DCF file (domain classification file)
domainsse        Adds secondary structure records to a DCF file (domain classification file)
dotmatcher       Displays a thresholded dotplot of two sequences
dotpath          Displays a non-overlapping wordmatch dotplot of two sequences
dottup           Displays a wordmatch dotplot of two sequences
dreg             Regular expression search of a nucleotide sequence
ealistat         Statistics for multiple alignment files
eclique          Largest clique program
econsense        Majority-rule and strict consensus tree
econtml          Continuous character Maximum Likelihood method
econtrast        Continuous character Contrasts
ednacomp         DNA compatibility algorithm
ednadist         Nucleic acid sequence Distance Matrix program
ednainvar        Nucleic acid sequence Invariants method
ednaml           Estimates phylogenies from nucleic acid sequence Maximum Likelihood
ednamlk          Estimates phylogenies from nucleic acid sequence Maximum Likelihood with molecular clock
ednapars         DNA parsimony algorithm
ednapenny        Penny algorithm for DNA
edollop          Dollo and polymorphism parsimony algorithm
edolpenny        Penny algorithm Dollo or polymorphism
efactor          Multistate to binary recoding program
efitch           Fitch-Margoliash and Least-Squares Distance Methods
egendist         Genetic Distance Matrix program
ehmmalign        Align sequences with an HMM
ehmmbuild        Build HMM
ehmmcalibrate    Calibrate a hidden Markov model
ehmmconvert      Convert between HMM formats
ehmmemit         Extract HMM sequences
ehmmfetch        Extract HMM from a database
ehmmindex        Index an HMM database
ehmmpfam         Align single sequence with an HMM
ehmmsearch       Search sequence database with an HMM
einverted        Finds DNA inverted repeats
ekitsch          Fitch-Margoliash method with contemporary tips
embossdata       Finds or fetches the data files read in by the EMBOSS programs
embossversion    Writes the current EMBOSS version number
emix             Mixed parsimony algorithm
emma             Multiple alignment program - interface to ClustalW program
emnu             Simple menu of EMBOSS applications
emowse           Protein identification by mass spectrometry
eneighbor        Phylogenies from distance matrix by N-J or UPGMA method
entret           Reads and writes (returns) flatfile entries
epenny           Penny algorithm, branch-and-bound to find all most parsimonious trees
epestfind        Finds PEST motifs as potential proteolytic cleavage sites
eprimer3         Picks PCR primers and hybridization oligos
eprotdist        Protein distance algorithm
eprotpars        Protein parsimony algorithm
equicktandem     Finds tandem repeats
erestml          Restriction site Maximum Likelihood method
eseqboot         Bootstrapped sequences algorithm
esim4            Align an mRNA to a genomic DNA sequence
est2genome       Align EST and genomic DNA sequences
etandem          Looks for tandem repeats in a nucleotide sequence
extractfeat      Extract features from a sequence
extractseq       Extract regions from a sequence
fclique          Largest clique program
fconsense        Majority-rule and strict consensus tree
fcontml          Gene frequency and continuous character Maximum Likelihood method
fcontrast        Continuous character Contrasts
fdiscboot        Bootstrapped discrete sites algorithm
fdnacomp         DNA compatibility algorithm
fdnadist         Nucleic acid sequence Distance Matrix program
fdnainvar        Nucleic acid sequence Invariants method
fdnaml           Estimates nucleotide phylogeny by maximum likelihood
fdnamlk          Estimates nucleotide phylogeny by maximum likelihood
fdnamove         Interactive DNA parsimony
fdnapars         DNA parsimony algorithm
fdnapenny        Penny algorithm for DNA
fdollop          Dollo and polymorphism parsimony algorithm
fdolmove         Interactive Dollo or Polymorphism Parsimony
fdolpenny        Penny algorithm Dollo or polymorphism
fdrawgram        Plots a cladogram- or phenogram-like rooted tree diagram
fdrawtree        Plots an unrooted tree diagram
ffactor          Multistate to binary recoding program
ffitch           Fitch-Margoliash and Least-Squares Distance Methods
ffreqboot        Bootstrapped genetic frequencies algorithm
fgendist         Compute genetic distances from gene frequencies
findkm           Find Km and Vmax for an enzyme reaction by a Hanes/Woolf plot
fkitsch          Fitch-Margoliash method with contemporary tips
fmix             Mixed parsimony algorithm
fmove            Interactive mixed method parsimony
fneighbor        Phylogenies from distance matrix by N-J or UPGMA method
fpars            Discrete character parsimony
fpenny           Penny algorithm, branch-and-bound to find all most parsimonious trees
fproml           Protein phylogeny by maximum likelihood
fpromlk          Protein phylogeny by maximum likelihood
fprotdist        Protein distance algorithm
fprotpars        Protein pasimony algorithm
freak            Residue/base frequency table or plot
frestboot        Bootstrapped restriction sites algorithm
frestdist        Computes distance matrix from restriction sites or fragments
frestml          Restriction site maximum Likelihood method
fretree          Interactive tree rearrangement
fseqboot         Bootstrapped sequences algorithm
fseqbootall      Bootstrapped sequences algorithm
ftreedist        Distances between trees
ftreedistpair    Distances between two sets of trees
fuzznuc          Nucleic acid pattern search
fuzzpro          Protein pattern search
fuzztran         Protein pattern search after translation
garnier          Predicts protein secondary structure
geecee           Calculates the fractional GC content of nucleic acid sequences
getorf           Finds and extracts open reading frames (ORFs)
helixturnhelix   Report nucleic acid binding motifs
hetparse         Converts raw dictionary of heterogen groups to a file in EMBL-like format
hmoment          Hydrophobic moment calculation
iep              Calculates the isoelectric point of a protein
infoalign        Information on a multiple sequence alignment
infoseq          Displays some simple information about sequences
interface        Reads CCF files (clean coordinate files) and writes CON files (contact files) of inter-chain residue-residue contact data
isochore         Plots isochores in large DNA sequences
libgen           Generates various types of discriminating elements for each alignment in a directory
lindna           Draws linear maps of DNA constructs
listor           Writes a list file of the logical OR of two sets of sequences
marscan          Finds MAR/SAR sites in nucleic sequences
maskfeat         Mask off features of a sequence
maskseq          Mask off regions of a sequence
matcher          Finds the best local alignments between two sequences
megamerger       Merge two large overlapping nucleic acid sequences
meme             Motif detection
merger           Merge two overlapping nucleic acid sequences
msbar            Mutate sequence beyond all recognition
mse              Multiple Sequence Editor
mwcontam         Shows molwts that match across a set of files
mwfilter         Filter noisy molwts from mass spec output
myseq            Demonstration of sequence reading
nawalign         Damian Counsell's implementation for protein sequences of the Needleman and Wunsch alignment algorithm
nawalignrunner   Runs nawalign alignment program on all sequence pairs in a directory
needle           Needleman-Wunsch global alignment
needlerunner     Runs needle alignment program on all sequence pairs in a directory
newcpgreport     Report CpG rich areas
newcpgseek       Reports CpG rich regions
newseq           Type in a short new sequence
noreturn         Removes carriage return from ASCII files
notseq           Excludes a set of sequences and writes out the remaining ones
nthseq           Writes one sequence from a multiple set of sequences
octanol          Displays protein hydropathy
oddcomp          Finds protein sequence regions with a biased composition
palindrome       Looks for inverted repeats in a nucleotide sequence
pasteseq         Insert one sequence into another
patmatdb         Search a protein sequence with a motif
patmatmotifs     Search a PROSITE motif database with a protein sequence
pdbparse         Parses PDB files and writes CCF files (clean coordinate files) for proteins
pdbplus          Add residue solvent accessibility and secondary structure data to a CCF file (clean coordinate file) for a protein or domain
pdbtosp          Convert raw swissprot:PDB equivalence file to EMBL-like format
pepcoil          Predicts coiled coil regions
pepinfo          Plots simple amino acid properties in parallel
pepnet           Displays proteins as a helical net
pepstats         Protein statistics
pepwheel         Shows protein sequences as helices
pepwindow        Displays protein hydropathy
pepwindowall     Displays protein hydropathy of a set of sequences
plotcon          Plots the quality of conservation of a sequence alignment
plotorf          Plot potential open reading frames
polydot          Displays all-against-all dotplots of a set of sequences
preg             Regular expression search of a protein sequence
prettyplot       Displays aligned sequences, with colouring and boxing
prettyseq        Output sequence with translated ranges
primersearch     Searches DNA sequences for matches with primer pairs
printsextract    Extract data from PRINTS
profit           Scan a sequence or database with a matrix or profile
prophecy         Creates matrices/profiles from multiple alignments
prophet          Gapped alignment for profiles
prosextract      Builds the PROSITE motif database for patmatmotifs to search
pscan            Scans proteins using PRINTS
psiphi           Calculates phi and psi torsion angles from cleaned EMBOSS-style protein co-ordinate file
rebaseextract    Extract data from REBASE
recoder          Remove restriction sites but maintain the same translation
redata           Search REBASE for enzyme name, references, suppliers etc
remap            Display a sequence with restriction cut sites, translation etc
restover         Finds restriction enzymes that produce a specific overhang
restrict         Finds restriction enzyme cleavage sites
revseq           Reverse and complement a sequence
rocon            Reads a DHF file (domain hits file) of hits (sequences of unknown structural classification) and a DHF file of validation sequences (known classification) and writes a 'hits file' for the hits, which are classified and rank-ordered on the basis of score
rocplot          Provides interpretation and graphical display of the performance of discriminating elements (e.g. profiles for protein families). rocplot reads file(s) of hits from discriminator-database search(es), performs ROC analysis on the hits, and writes graphs illustrating the diagnostic performance of the discriminating elements
scopparse        Reads raw SCOP classification files and writes a DCF file (domain classification file)
scorer           Scores accuracy of protein-protein sequence alignment against gold standard structure-structure alignment
scorerrunner     Runs scorer to compare ordered pairs of substituted seqs in two directories and write the scores to a third
seealso          Finds programs sharing group names
seqalign         Reads a DAF file (domain alignment file) and a DHF file (domain hits file) and writes a DAF file extended with the hits
seqfraggle       Removes fragments from DHF files (domain hits files) or other files of sequences
seqmatchall      Does an all-against-all comparison of a set of sequences
seqnr            Removes redundancy from DHF files (domain hits files) or other files of sequences
seqret           Reads and writes (returns) sequences
seqretsplit      Reads and writes (returns) sequences in individual files
seqsearch        Generate database hits (sequences) for nodes in a DCF file (domain classification file) by using PSI-BLAST
seqsort          Reads DHF files (domain hits files) of database hits (sequences) and removes hits of ambiguous classification
seqwords         Generates DHF files (domain hits files) of database hits (sequences) for nodes in a DCF file (domain classification file) by keyword search of UniProt
showalign        Displays a multiple sequence alignment
showdb           Displays information on the currently available databases
showfeat         Show features of a sequence
showorf          Pretty output of DNA translations
showseq          Display a sequence with features, translation etc
shuffleseq       Shuffles a set of sequences maintaining composition
sigcleave        Reports protein signal cleavage sites
siggen           Generates a sparse protein signature from an alignment and residue contact data
sigscan          Generates a DHF file (domain hits file) of hits (sequences) from scanning a signature against a sequence database
silent           Silent mutation restriction enzyme scan
sirna            Finds siRNA duplexes in mRNA
sites            Reads CCF files (clean coordinate files) and writes CON files (contact files) of residue-ligand contact data for domains in a DCF file (domain classification file)
sixpack          Display a DNA sequence with 6-frame translation and ORFs
skipseq          Reads and writes (returns) sequences, skipping the first few
splitter         Split a sequence into (overlapping) smaller sequences
ssematch         Searches a DCF file (domain classification file) for secondary structure matches
stretcher        Finds the best global alignment between two sequences
stssearch        Searches a DNA database for matches with a set of STS primers
substitute       Substitutes matches from first (query) sequence of two aligned sequences in a trace into second (template sequence)
substituterunner Runs substitute on directory of traces and writes substituted sequences to another
supermatcher     Finds a match of a large sequence against one or more sequences
syco             Synonymous codon usage Gribskov statistic plot
tcode            Fickett TESTCODE statistic to identify protein-coding DNA
textsearch       Search sequence documentation text. SRS and Entrez are faster!
tfextract        Extract data from TRANSFAC
tfm              Displays a program's help documentation manual
tfscan           Scans DNA sequences for transcription factors
tmap             Displays membrane spanning regions
topo             Draws an image of a transmembrane protein
tranalign        Align nucleic coding regions given the aligned proteins
transeq          Translate nucleic acid sequences
trimest          Trim poly-A tails off EST sequences
trimseq          Trim ambiguous bits off the ends of sequences
twofeat          Finds neighbouring pairs of features in sequences
union            Reads sequence fragments and builds one sequence
vectorstrip      Strips out DNA between a pair of vector sequences
water            Smith-Waterman local alignment
whichdb          Search all databases for an entry
wobble           Wobble base plot
wordcount        Counts words of a specified size in a DNA sequence
wordmatch        Finds all exact matches of a given size between 2 sequences
wossname         Finds programs by keywords in their one-line documentation
yank             Reads a sequence range, appends the full USA to a list file


</pre></td></tr></table><p>
<p>
<p>
<p>
<b>Example 4</b>
<p>
Display only the groups that the programs can belong to: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>wossname -search '' -groups </b>
Finds programs by keywords in their one-line documentation

ACD
ALIGNMENT CONSENSUS
ALIGNMENT DIFFERENCES
ALIGNMENT DOT PLOTS
ALIGNMENT GLOBAL
ALIGNMENT LOCAL
ALIGNMENT MULTIPLE
DISPLAY
EDIT
ENZYME KINETICS
FEATURE TABLES
HMM
INFORMATION
MENUS
NUCLEIC 2D STRUCTURE
NUCLEIC CODON USAGE
NUCLEIC COMPOSITION
NUCLEIC CPG ISLANDS
NUCLEIC GENE FINDING
NUCLEIC MOTIFS
NUCLEIC MUTATION
NUCLEIC PRIMERS
NUCLEIC PROFILES
NUCLEIC REPEATS
NUCLEIC RESTRICTION
NUCLEIC TRANSCRIPTION
NUCLEIC TRANSLATION
PHYLOGENY CONSENSUS
PHYLOGENY CONTINUOUS CHARACTERS
PHYLOGENY DISCRETE CHARACTERS
PHYLOGENY DISTANCE MATRIX
PHYLOGENY GENE FREQUENCIES
PHYLOGENY MOLECULAR SEQUENCE
PHYLOGENY TREE DRAWING
PROTEIN 2D STRUCTURE
PROTEIN 3D STRUCTURE
PROTEIN COMPOSITION
PROTEIN MOTIFS
PROTEIN MUTATION
PROTEIN PROFILES
TEST
UTILS DATABASE CREATION
UTILS DATABASE INDEXING
UTILS MISC

</pre></td></tr></table><p>
<p>
<p>
<p>
<b>Example 5</b>
<p>
Output html tags around the list of program groups: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>wossname -groups -html -prelink '#' </b>
Finds programs by keywords in their one-line documentation
Keyword to search for, or blank to list all programs: <b></b>

<ul>
<li><a href="#ACD">ACD</a></li>
<li><a href="#ALIGNMENT CONSENSUS">ALIGNMENT CONSENSUS</a></li>
<li><a href="#ALIGNMENT DIFFERENCES">ALIGNMENT DIFFERENCES</a></li>
<li><a href="#ALIGNMENT DOT PLOTS">ALIGNMENT DOT PLOTS</a></li>
<li><a href="#ALIGNMENT GLOBAL">ALIGNMENT GLOBAL</a></li>
<li><a href="#ALIGNMENT LOCAL">ALIGNMENT LOCAL</a></li>
<li><a href="#ALIGNMENT MULTIPLE">ALIGNMENT MULTIPLE</a></li>
<li><a href="#DISPLAY">DISPLAY</a></li>
<li><a href="#EDIT">EDIT</a></li>
<li><a href="#ENZYME KINETICS">ENZYME KINETICS</a></li>
<li><a href="#FEATURE TABLES">FEATURE TABLES</a></li>
<li><a href="#HMM">HMM</a></li>
<li><a href="#INFORMATION">INFORMATION</a></li>
<li><a href="#MENUS">MENUS</a></li>
<li><a href="#NUCLEIC 2D STRUCTURE">NUCLEIC 2D STRUCTURE</a></li>
<li><a href="#NUCLEIC CODON USAGE">NUCLEIC CODON USAGE</a></li>
<li><a href="#NUCLEIC COMPOSITION">NUCLEIC COMPOSITION</a></li>
<li><a href="#NUCLEIC CPG ISLANDS">NUCLEIC CPG ISLANDS</a></li>
<li><a href="#NUCLEIC GENE FINDING">NUCLEIC GENE FINDING</a></li>
<li><a href="#NUCLEIC MOTIFS">NUCLEIC MOTIFS</a></li>
<li><a href="#NUCLEIC MUTATION">NUCLEIC MUTATION</a></li>
<li><a href="#NUCLEIC PRIMERS">NUCLEIC PRIMERS</a></li>
<li><a href="#NUCLEIC PROFILES">NUCLEIC PROFILES</a></li>
<li><a href="#NUCLEIC REPEATS">NUCLEIC REPEATS</a></li>
<li><a href="#NUCLEIC RESTRICTION">NUCLEIC RESTRICTION</a></li>
<li><a href="#NUCLEIC TRANSCRIPTION">NUCLEIC TRANSCRIPTION</a></li>
<li><a href="#NUCLEIC TRANSLATION">NUCLEIC TRANSLATION</a></li>
<li><a href="#PHYLOGENY CONSENSUS">PHYLOGENY CONSENSUS</a></li>
<li><a href="#PHYLOGENY CONTINUOUS CHARACTERS">PHYLOGENY CONTINUOUS CHARACTERS</a></li>
<li><a href="#PHYLOGENY DISCRETE CHARACTERS">PHYLOGENY DISCRETE CHARACTERS</a></li>
<li><a href="#PHYLOGENY DISTANCE MATRIX">PHYLOGENY DISTANCE MATRIX</a></li>
<li><a href="#PHYLOGENY GENE FREQUENCIES">PHYLOGENY GENE FREQUENCIES</a></li>
<li><a href="#PHYLOGENY MOLECULAR SEQUENCE">PHYLOGENY MOLECULAR SEQUENCE</a></li>
<li><a href="#PHYLOGENY TREE DRAWING">PHYLOGENY TREE DRAWING</a></li>
<li><a href="#PROTEIN 2D STRUCTURE">PROTEIN 2D STRUCTURE</a></li>
<li><a href="#PROTEIN 3D STRUCTURE">PROTEIN 3D STRUCTURE</a></li>
<li><a href="#PROTEIN COMPOSITION">PROTEIN COMPOSITION</a></li>
<li><a href="#PROTEIN MOTIFS">PROTEIN MOTIFS</a></li>
<li><a href="#PROTEIN MUTATION">PROTEIN MUTATION</a></li>
<li><a href="#PROTEIN PROFILES">PROTEIN PROFILES</a></li>
<li><a href="#TEST">TEST</a></li>
<li><a href="#UTILS DATABASE CREATION">UTILS DATABASE CREATION</a></li>
<li><a href="#UTILS DATABASE INDEXING">UTILS DATABASE INDEXING</a></li>
<li><a href="#UTILS MISC">UTILS MISC</a></li>
</ul>

</pre></td></tr></table><p>
<p>
<p>
<p>
<b>Example 6</b>
<p>
Output html tags around the list of programs: 
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>wossname '' -html -prelink 'http://emboss.sf.net/apps/' -postlink '.html' -out wossname.html </b>
Finds programs by keywords in their one-line documentation

</pre></td></tr></table><p>
<p>
<a href="#output.6">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-search]            string     Enter a word or words here and a
                                  case-independent search for it will be made
                                  in the one-line documentation of all of the
                                  EMBOSS programs. If no keyword is specified,
                                  all programs will be listed.

   Additional (Optional) qualifiers (* if not always prompted):
   -explode            boolean    The groups that EMBOSS applications belong
                                  to have two forms, exploded and not
                                  exploded. The exploded group names are more
                                  numerous and often vaguely phrased than the
                                  non-exploded ones. The exploded names are
                                  formed from definitions of the group names
                                  that start like NAME1:NAME2 and which are
                                  then expanded into many combinations of the
                                  names as: 'NAME1', 'NAME2', 'NAME1 NAME2',
                                  NAME2 NAME1'. The non-expanded names are
                                  simply like: 'NAME1 NAME2'.
   -outfile            outfile    Output file name
   -html               toggle     If you are sending the output to a file,
                                  this will format it for displaying as a
                                  table in a WWW document.
*  -prelink            string     If you are outputting to a file in HTML
                                  format, you can make the program names into
                                  hyper-links by setting this to be the first
                                  half of the URL.
                                  For example, if you wish the program name to
                                  be a hyperlink to the URL:
                                  http://www.uk.embnet.org/Software/EMBOSS/Apps/programname.html
                                  then if you enter
                                  http://www.uk.embnet.org/Software/EMBOSS/Apps/
                                  here, then the first half of the URL before
                                  the programname will be set.
*  -postlink           string     If you are outputting to a file in HTML
                                  format, you can make the program names into
                                  hyper-links by setting this to be the second
                                  half of the URL.
                                  For example, if you wish the program name to
                                  be a hyperlink to the URL:
                                  http://www.uk.embnet.org/Software/EMBOSS/Apps/programname.html
                                  then if you enter '.html' here, the second
                                  half of the URL after the programname will
                                  be set.
   -groups             boolean    If you use this option, then only the group
                                  names will output to the file
   -alphabetic         boolean    If you use this option, then you will get a
                                  single list of the program names and
                                  descriptions instead of the programs being
                                  listed in their functional groups.

   Advanced (Unprompted) qualifiers:
   -[no]emboss         boolean    If you use this option then EMBOSS program
                                  documentation will be searched. If this
                                  option is set to be false, then only the
                                  EMBASSY programs will be searched (if the
                                  -embassy option is true). EMBASSY programs
                                  are not strictly part of EMBOSS, but use the
                                  same code libraries and share the same look
                                  and feel, but are generally developed by
                                  people who wish the programs to be outside
                                  of the GNU Public Licence scheme
   -[no]embassy        boolean    If you use this option then EMBASSY program
                                  documentation will be searched. If this
                                  option is set to be false, then only the
                                  EMBOSS programs will be searched (if the
                                  -emboss option is true). EMBASSY programs
                                  are not strictly part of EMBOSS, but use the
                                  same code libraries and share the same look
                                  and feel, but are generally developed by
                                  people who wish the programs to be outside
                                  of the GNU Public Licence scheme
   -showembassy        string     If you use this option then this EMBASSY
                                  package program documentation will be
                                  searched. EMBASSY programs are not strictly
                                  part of EMBOSS, but use the same code
                                  libraries and share the same look and feel,
                                  but are generally developed by people who
                                  wish the programs to be outside of the GNU
                                  Public Licence scheme
   -colon              boolean    The groups that EMBOSS applications belong
                                  to up to two levels, for example the primary
                                  group 'ALIGNMENT' has several sub-groups,
                                  or second-level groups, e.g.: CONSENSUS,
                                  DIFFERENCES, DOT PLOTS, GLOBAL, LOCAL,
                                  MULTIPLE. To aid programs that parse the
                                  output of wossname that require the names of
                                  these subgroups, a colon ':' will be placed
                                  between the first and second level of the
                                  group name if this option is true. Note:
                                  This does not apply if the group names have
                                  been exploded with the 'explode' option.
   -gui                boolean    This option is intended to help those who
                                  are designing Graphical User Interfaces to
                                  the EMBOSS applications. Some EMBOSS
                                  programs are inappropriate for running in a
                                  GUI, these include other menu programs and
                                  interactive editors. This option allows you
                                  to only report those programs that can be
                                  run from a GUI

   Associated qualifiers:

   "-outfile" associated qualifiers
   -odirectory          string     Output directory

   General qualifiers:
   -auto                boolean    Turn off prompts
   -stdout              boolean    Write standard output
   -filter              boolean    Read standard input, write standard output
   -options             boolean    Prompt for standard and additional values
   -debug               boolean    Write debug output to program.dbg
   -verbose             boolean    Report some/full command line options
   -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning             boolean    Report warnings
   -error               boolean    Report errors
   -fatal               boolean    Report fatal errors
   -die                 boolean    Report deaths


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-search]<br>(Parameter 1)</td>
<td>Enter a word or words here and a case-independent search for it will be made in the one-line documentation of all of the EMBOSS programs. If no keyword is specified, all programs will be listed.</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-explode</td>
<td>The groups that EMBOSS applications belong to have two forms, exploded and not exploded. The exploded group names are more numerous and often vaguely phrased than the non-exploded ones. The exploded names are formed from definitions of the group names that start like NAME1:NAME2 and which are then expanded into many combinations of the names as: 'NAME1', 'NAME2', 'NAME1 NAME2', NAME2 NAME1'. The non-expanded names are simply like: 'NAME1 NAME2'.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-outfile</td>
<td>Output file name</td>
<td>Output file</td>
<td>stdout</td>
</tr>

<tr>
<td>-html</td>
<td>If you are sending the output to a file, this will format it for displaying as a table in a WWW document.</td>
<td>Toggle value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-prelink</td>
<td>If you are outputting to a file in HTML format, you can make the program names into hyper-links by setting this to be the first half of the URL.
For example, if you wish the program name to be a hyperlink to the URL:
http://www.uk.embnet.org/Software/EMBOSS/Apps/programname.html
then if you enter
http://www.uk.embnet.org/Software/EMBOSS/Apps/ here, then the first half of the URL before the programname will be set.</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-postlink</td>
<td>If you are outputting to a file in HTML format, you can make the program names into hyper-links by setting this to be the second half of the URL.
For example, if you wish the program name to be a hyperlink to the URL:
http://www.uk.embnet.org/Software/EMBOSS/Apps/programname.html
then if you enter '.html' here, the second half of the URL after the programname will be set.</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-groups</td>
<td>If you use this option, then only the group names will output to the file</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-alphabetic</td>
<td>If you use this option, then you will get a single list of the program names and descriptions instead of the programs being listed in their functional groups.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]emboss</td>
<td>If you use this option then EMBOSS program documentation will be searched. If this option is set to be false, then only the EMBASSY programs will be searched (if the -embassy option is true). EMBASSY programs are not strictly part of EMBOSS, but use the same code libraries and share the same look and feel, but are generally developed by people who wish the programs to be outside of the GNU Public Licence scheme</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]embassy</td>
<td>If you use this option then EMBASSY program documentation will be searched. If this option is set to be false, then only the EMBOSS programs will be searched (if the -emboss option is true). EMBASSY programs are not strictly part of EMBOSS, but use the same code libraries and share the same look and feel, but are generally developed by people who wish the programs to be outside of the GNU Public Licence scheme</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-showembassy</td>
<td>If you use this option then this EMBASSY package program documentation will be searched. EMBASSY programs are not strictly part of EMBOSS, but use the same code libraries and share the same look and feel, but are generally developed by people who wish the programs to be outside of the GNU Public Licence scheme</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-colon</td>
<td>The groups that EMBOSS applications belong to up to two levels, for example the primary group 'ALIGNMENT' has several sub-groups, or second-level groups, e.g.: CONSENSUS, DIFFERENCES, DOT PLOTS, GLOBAL, LOCAL, MULTIPLE. To aid programs that parse the output of wossname that require the names of these subgroups, a colon ':' will be placed between the first and second level of the group name if this option is true. Note: This does not apply if the group names have been exploded with the 'explode' option.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-gui</td>
<td>This option is intended to help those who are designing Graphical User Interfaces to the EMBOSS applications. Some EMBOSS programs are inappropriate for running in a GUI, these include other menu programs and interactive editors. This option allows you to only report those programs that can be run from a GUI</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

</table>



<H2>
    Input file format
</H2>




<p>

<b>wossname</b> reads the brief descriptions and the program groups from
the ACD files.  (ACD files control the parameters that the programs
expect - don't worry about them unless you are writing EMBOSS
programs).

<p>

The format of ACD files is documented elsewhere.

<H2>
    Output file format
</H2>


The results are written out as a simple list preceded by the group
that the programs belong to or the word that is being searched for. 

<p>


<a name="output.6"></a>
<h3>Output files for usage example 6</h3>
<p><h3>File: wossname.html</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<h2><a name="ACD">ACD</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="http://emboss.sf.net/apps/acdc.html">acdc</a></td><td>ACD compiler</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/acdpretty.html">acdpretty</a></td><td>ACD pretty printing utility</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/acdtable.html">acdtable</a></td><td>Creates an HTML table from an ACD file</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/acdtrace.html">acdtrace</a></td><td>ACD compiler on-screen trace</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/acdvalid.html">acdvalid</a></td><td>ACD file validation</td></tr>
</table>
<h2><a name="ALIGNMENT CONSENSUS">ALIGNMENT CONSENSUS</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="http://emboss.sf.net/apps/cons.html">cons</a></td><td>Creates a consensus from multiple alignments</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/megamerger.html">megamerger</a></td><td>Merge two large overlapping nucleic acid sequences</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/merger.html">merger</a></td><td>Merge two overlapping nucleic acid sequences</td></tr>
</table>
<h2><a name="ALIGNMENT DIFFERENCES">ALIGNMENT DIFFERENCES</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="http://emboss.sf.net/apps/diffseq.html">diffseq</a></td><td>Find differences between nearly identical sequences</td></tr>
</table>
<h2><a name="ALIGNMENT DOT PLOTS">ALIGNMENT DOT PLOTS</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="http://emboss.sf.net/apps/dotmatcher.html">dotmatcher</a></td><td>Displays a thresholded dotplot of two sequences</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/dotpath.html">dotpath</a></td><td>Displays a non-overlapping wordmatch dotplot of two sequences</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/dottup.html">dottup</a></td><td>Displays a wordmatch dotplot of two sequences</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/polydot.html">polydot</a></td><td>Displays all-against-all dotplots of a set of sequences</td></tr>
</table>
<h2><a name="ALIGNMENT GLOBAL">ALIGNMENT GLOBAL</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="http://emboss.sf.net/apps/alignrunner.html">alignrunner</a></td><td>Runs alignment program on all sequence pairs in a directory</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/contactalign.html">contactalign</a></td><td>EMBOSS implementation of Damian Counsell's 2.5-D alignment algorithm</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/esim4.html">esim4</a></td><td>Align an mRNA to a genomic DNA sequence</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/est2genome.html">est2genome</a></td><td>Align EST and genomic DNA sequences</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/nawalign.html">nawalign</a></td><td>Damian Counsell's implementation for protein sequences of the Needleman and Wunsch alignment algorithm</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/nawalignrunner.html">nawalignrunner</a></td><td>Runs nawalign alignment program on all sequence pairs in a directory</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/needle.html">needle</a></td><td>Needleman-Wunsch global alignment</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/needlerunner.html">needlerunner</a></td><td>Runs needle alignment program on all sequence pairs in a directory</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/scorer.html">scorer</a></td><td>Scores accuracy of protein-protein sequence alignment against gold standard structure-structure alignment</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/scorerrunner.html">scorerrunner</a></td><td>Runs scorer to compare ordered pairs of substituted seqs in two directories and write the scores to a third</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/stretcher.html">stretcher</a></td><td>Finds the best global alignment between two sequences</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/substitute.html">substitute</a></td><td>Substitutes matches from first (query) sequence of two aligned sequences in a trace into second (template sequence)</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/substituterunner.html">substituterunner</a></td><td>Runs substitute on directory of traces and writes substituted sequences to another</td></tr>
</table>
<h2><a name="ALIGNMENT LOCAL">ALIGNMENT LOCAL</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="http://emboss.sf.net/apps/matcher.html">matcher</a></td><td>Finds the best local alignments between two sequences</td></tr>


<font color=red>  [Part of this file has been deleted for brevity]</font>

<h2><a name="PROTEIN PROFILES">PROTEIN PROFILES</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="http://emboss.sf.net/apps/profit.html">profit</a></td><td>Scan a sequence or database with a matrix or profile</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/prophecy.html">prophecy</a></td><td>Creates matrices/profiles from multiple alignments</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/prophet.html">prophet</a></td><td>Gapped alignment for profiles</td></tr>
</table>
<h2><a name="TEST">TEST</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="http://emboss.sf.net/apps/crystalball.html">crystalball</a></td><td>Answers every drug discovery question you have about this sequence</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/myseq.html">myseq</a></td><td>Demonstration of sequence reading</td></tr>
</table>
<h2><a name="UTILS DATABASE CREATION">UTILS DATABASE CREATION</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="http://emboss.sf.net/apps/aaindexextract.html">aaindexextract</a></td><td>Extract data from AAINDEX</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/allversusall.html">allversusall</a></td><td>Does an all-versus-all global alignment for each set of sequences in an input directory and writes files of sequence similarity values</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/cathparse.html">cathparse</a></td><td>Reads raw CATH classification files and writes DCF file (domain classification file)</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/cutgextract.html">cutgextract</a></td><td>Extract data from CUTG</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/domainer.html">domainer</a></td><td>Reads CCF files (clean coordinate files) for proteins and writes CCF files for domains, taken from a DCF file (domain classification file)</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/domainnr.html">domainnr</a></td><td>Removes redundant domains from a DCF file (domain classification file). The file must contain domain sequence information, which can be added by using DOMAINSEQS</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/domainseqs.html">domainseqs</a></td><td>Adds sequence records to a DCF file (domain classification file)</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/domainsse.html">domainsse</a></td><td>Adds secondary structure records to a DCF file (domain classification file)</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/hetparse.html">hetparse</a></td><td>Converts raw dictionary of heterogen groups to a file in EMBL-like format</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/pdbparse.html">pdbparse</a></td><td>Parses PDB files and writes CCF files (clean coordinate files) for proteins</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/pdbplus.html">pdbplus</a></td><td>Add residue solvent accessibility and secondary structure data to a CCF file (clean coordinate file) for a protein or domain</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/pdbtosp.html">pdbtosp</a></td><td>Convert raw swissprot:PDB equivalence file to EMBL-like format</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/printsextract.html">printsextract</a></td><td>Extract data from PRINTS</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/prosextract.html">prosextract</a></td><td>Builds the PROSITE motif database for patmatmotifs to search</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/rebaseextract.html">rebaseextract</a></td><td>Extract data from REBASE</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/scopparse.html">scopparse</a></td><td>Reads raw SCOP classification files and writes a DCF file (domain classification file)</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/seqnr.html">seqnr</a></td><td>Removes redundancy from DHF files (domain hits files) or other files of sequences</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/sites.html">sites</a></td><td>Reads CCF files (clean coordinate files) and writes CON files (contact files) of residue-ligand contact data for domains in a DCF file (domain classification file)</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/ssematch.html">ssematch</a></td><td>Searches a DCF file (domain classification file) for secondary structure matches</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/tfextract.html">tfextract</a></td><td>Extract data from TRANSFAC</td></tr>
</table>
<h2><a name="UTILS DATABASE INDEXING">UTILS DATABASE INDEXING</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="http://emboss.sf.net/apps/dbiblast.html">dbiblast</a></td><td>Index a BLAST database</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/dbifasta.html">dbifasta</a></td><td>Index a fasta database</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/dbiflat.html">dbiflat</a></td><td>Index a flat file database</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/dbigcg.html">dbigcg</a></td><td>Index a GCG formatted database</td></tr>
</table>
<h2><a name="UTILS MISC">UTILS MISC</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="http://emboss.sf.net/apps/embossdata.html">embossdata</a></td><td>Finds or fetches the data files read in by the EMBOSS programs</td></tr>
<tr><td><a href="http://emboss.sf.net/apps/embossversion.html">embossversion</a></td><td>Writes the current EMBOSS version number</td></tr>
</table>
</td></tr></table><p>



<p>

If the -html, -prelink and -postlink qualifiers are specified, then
the output will be wrapped in HTML tags, ready for inclusion in a Web
page.  Note that tags such as &lt;HTML&gt;, &lt;BODY&gt;,
&lt;/BODY&gt; and &lt;/HTML&gt; are not output by this program as the
list of programs is expected to form only part of the contents of a
web page - the rest of the web page must be supplied by the user.

<H2>
    Data files
</H2>


As noted above, the program ACD files are searched.
The average user will not need to worry about this.


<H2>
    Notes
</H2>


If a program's ACD file does not assign the program to a group,
it will be automatically assigned to the group 'ASSORTED'.

<p>

A program can belong to more than one group.

<p>

The name of this program comes from its use as a favorite word of the
author Terry Pratchett.  It is used especially by the characters of the
parrot in <b>Eric</b> and the dog <i>Gaspode</i> in <b>Moving
Pictures</b> and <b>Men at Arms</b>. 

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>


None.

<H2>
    Diagnostic Error Messages
</H2>


None.

<H2>
    Exit status
</H2>


It always exits with status 0

<H2>
    Known bugs
</H2>


None.

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="infoalign.html">infoalign</a></td><td>Information on a multiple sequence alignment</td></tr>
<tr><td><a href="infoseq.html">infoseq</a></td><td>Displays some simple information about sequences</td></tr>
<tr><td><a href="seealso.html">seealso</a></td><td>Finds programs sharing group names</td></tr>
<tr><td><a href="showdb.html">showdb</a></td><td>Displays information on the currently available databases</td></tr>
<tr><td><a href="textsearch.html">textsearch</a></td><td>Search sequence documentation text. SRS and Entrez are faster!</td></tr>
<tr><td><a href="tfm.html">tfm</a></td><td>Displays a program's help documentation manual</td></tr>
<tr><td><a href="whichdb.html">whichdb</a></td><td>Search all databases for an entry</td></tr>
</table>


<H2>
    Author(s)
</H2>
Gary Williams (gwilliam&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK



<H2>
    History
</H2>



<H2>
    Target users
</H2>
This program is intended to be used by everyone and everything, from naive users to embedded scripts.




<H2>
    Comments
</H2>
None


</BODY>
</HTML>

