<HTML>
<HEAD>
  <TITLE>
  EMBOSS: xmltext
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="/images/emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
xmltext
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
Get XML document original full text
<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
  -->

<b>xmltext</b> reads XML data and rewrites it as the original entry text.

<H2>
    Algorithm
</H2>

<!-- 
        Algorithms
	  Pseudocode/Vague description/Handwaving/web links to useful pages
  -->

The XML data is parsed using a DOM parser.

<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

Here is a sample session with <b>xmltext</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>xmltext dbfetch:embl:x13776 </b>
Get XML document original full text
Full text output file [x13776.xmltext]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
Get XML document original full text
Version: EMBOSS:6.6.0.0

   Standard (Mandatory) qualifiers:
  [-xml]               xml        Xml filename and optional format, or
                                  reference (input query)
  [-outfile]           outfile    [*.xmltext] Full text output file

   Additional (Optional) qualifiers:
   -subclasses         boolean    [N] Extend the query matches to include all
                                  terms which are specialisations
                                  (sub-classes) of the matched terms.
   -obsolete           boolean    [N] The default behaviour is to not use or
                                  return obsolete terms. This option if set
                                  will include all terms.

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-xml" associated qualifiers
   -iformat1           string     Input xml format
   -iquery1            string     Input query fields or ID list
   -ioffset1           integer    Input start position offset
   -idbname1           string     User-provided database name

   "-outfile" associated qualifiers
   -odirectory2        string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-xml]<br>(Parameter 1)</td>
<td>xml</td>
<td>Xml filename and optional format, or reference (input query)</td>
<td>Xml</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-outfile]<br>(Parameter 2)</td>
<td>outfile</td>
<td>Full text output file</td>
<td>Output file</td>
<td><i>&lt;*&gt;</i>.xmltext</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>-subclasses</td>
<td>boolean</td>
<td>Extend the query matches to include all terms which are specialisations (sub-classes) of the matched terms.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-obsolete</td>
<td>boolean</td>
<td>The default behaviour is to not use or return obsolete terms. This option if set will include all terms.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td align="left" colspan=5>"-xml" associated xml qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -iformat1<br>-iformat_xml</td>
<td>string</td>
<td>Input xml format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -iquery1<br>-iquery_xml</td>
<td>string</td>
<td>Input query fields or ID list</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ioffset1<br>-ioffset_xml</td>
<td>integer</td>
<td>Input start position offset</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -idbname1<br>-idbname_xml</td>
<td>string</td>
<td>User-provided database name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td align="left" colspan=5>"-outfile" associated outfile qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -odirectory2<br>-odirectory_outfile</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>xmltext</b> reads XML data.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>Database entry: dbfetch:embl:x13776</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   X13776; SV 1; linear; genomic DNA; STD; PRO; 2167 BP.
XX
AC   X13776; M43175;
XX
DT   19-APR-1989 (Rel. 19, Created)
DT   14-NOV-2006 (Rel. 89, Last updated, Version 24)
XX
DE   Pseudomonas aeruginosa amiC and amiR gene for aliphatic amidase regulation
XX
KW   aliphatic amidase regulator; amiC gene; amiR gene.
XX
OS   Pseudomonas aeruginosa
OC   Bacteria; Proteobacteria; Gammaproteobacteria; Pseudomonadales;
OC   Pseudomonadaceae; Pseudomonas.
XX
RN   [1]
RP   1167-2167
RA   Rice P.M.;
RT   ;
RL   Submitted (16-DEC-1988) to the INSDC.
RL   Rice P.M., EMBL, Postfach 10-2209, Meyerhofstrasse 1, 6900 Heidelberg, FRG.
XX
RN   [2]
RP   1167-2167
RX   DOI; 10.1016/0014-5793(89)80249-2.
RX   PUBMED; 2495988.
RA   Lowe N., Rice P.M., Drew R.E.;
RT   "Nucleotide sequence of the aliphatic amidase regulator gene (amiR) of
RT   Pseudomonas aeruginosa";
RL   FEBS Lett. 246(1-2):39-43(1989).
XX
RN   [3]
RP   1-1292
RX   PUBMED; 1907262.
RA   Wilson S., Drew R.;
RT   "Cloning and DNA sequence of amiC, a new gene regulating expression of the
RT   Pseudomonas aeruginosa aliphatic amidase, and purification of the amiC
RT   product";
RL   J. Bacteriol. 173(16):4914-4921(1991).
XX
RN   [4]
RP   1-2167
RA   Rice P.M.;
RT   ;
RL   Submitted (04-SEP-1991) to the INSDC.
RL   Rice P.M., EMBL, Postfach 10-2209, Meyerhofstrasse 1, 6900 Heidelberg, FRG.
XX
DR   EuropePMC; PMC208179; 1907262.
DR   EuropePMC; PMC395554; 7813419.
DR   GOA; Q51417.


<font color=red>  [Part of this file has been deleted for brevity]</font>

FT                   /db_xref="InterPro:IPR005561"
FT                   /db_xref="InterPro:IPR008327"
FT                   /db_xref="InterPro:IPR011006"
FT                   /db_xref="InterPro:IPR011991"
FT                   /db_xref="PDB:1QO0"
FT                   /db_xref="UniProtKB/Swiss-Prot:P10932"
FT                   /protein_id="CAA32023.1"
FT                   /translation="MSANSLLGSLRELQVLVLNPPGEVSDALVLQLIRIGCSVRQCWPP
FT                   PEAFDVPVDVVFTSIFQNGHHDEIAALLAAGTPRTTLVALVEYESPAVLSQIIELECHG
FT                   VITQPLDAHRVLPVLVSARRISEEMAKLKQKTEQLQDRIAGQARINQAKVLLMQRHGWD
FT                   EREAHQHLSREAMKRREPILKIAQELLGNEPSA"
XX
SQ   Sequence 2167 BP; 363 A; 712 C; 730 G; 362 T; 0 other;
     ggtaccgctg gccgagcatc tgctcgatca ccaccagccg ggcgacggga actgcacgat        60
     ctacctggcg agcctggagc acgagcgggt tcgcttcgta cggcgctgag cgacagtcac       120
     aggagaggaa acggatggga tcgcaccagg agcggccgct gatcggcctg ctgttctccg       180
     aaaccggcgt caccgccgat atcgagcgct cgcacgcgta tggcgcattg ctcgcggtcg       240
     agcaactgaa ccgcgagggc ggcgtcggcg gtcgcccgat cgaaacgctg tcccaggacc       300
     ccggcggcga cccggaccgc tatcggctgt gcgccgagga cttcattcgc aaccgggggg       360
     tacggttcct cgtgggctgc tacatgtcgc acacgcgcaa ggcggtgatg ccggtggtcg       420
     agcgcgccga cgcgctgctc tgctacccga ccccctacga gggcttcgag tattcgccga       480
     acatcgtcta cggcggtccg gcgccgaacc agaacagtgc gccgctggcg gcgtacctga       540
     ttcgccacta cggcgagcgg gtggtgttca tcggctcgga ctacatctat ccgcgggaaa       600
     gcaaccatgt gatgcgccac ctgtatcgcc agcacggcgg cacggtgctc gaggaaatct       660
     acattccgct gtatccctcc gacgacgact tgcagcgcgc cgtcgagcgc atctaccagg       720
     cgcgcgccga cgtggtcttc tccaccgtgg tgggcaccgg caccgccgag ctgtatcgcg       780
     ccatcgcccg tcgctacggc gacggcaggc ggccgccgat cgccagcctg accaccagcg       840
     aggcggaggt ggcgaagatg gagagtgacg tggcagaggg gcaggtggtg gtcgcgcctt       900
     acttctccag catcgatacg cccgccagcc gggccttcgt ccaggcctgc catggtttct       960
     tcccggagaa cgcgaccatc accgcctggg ccgaggcggc ctactggcag accttgttgc      1020
     tcggccgcgc cgcgcaggcc gcaggcaact ggcgggtgga agacgtgcag cggcacctgt      1080
     acgacatcga catcgacgcg ccacaggggc cggtccgggt ggagcgccag aacaaccaca      1140
     gccgcctgtc ttcgcgcatc gcggaaatcg atgcgcgcgg cgtgttccag gtccgctggc      1200
     agtcgcccga accgattcgc cccgaccctt atgtcgtcgt gcataacctc gacgactggt      1260
     ccgccagcat gggcggggga ccgctcccat gagcgccaac tcgctgctcg gcagcctgcg      1320
     cgagttgcag gtgctggtcc tcaacccgcc gggggaggtc agcgacgccc tggtcttgca      1380
     gctgatccgc atcggttgtt cggtgcgcca gtgctggccg ccgccggaag ccttcgacgt      1440
     gccggtggac gtggtcttca ccagcatttt ccagaatggc caccacgacg agatcgctgc      1500
     gctgctcgcc gccgggactc cgcgcactac cctggtggcg ctggtggagt acgaaagccc      1560
     cgcggtgctc tcgcagatca tcgagctgga gtgccacggc gtgatcaccc agccgctcga      1620
     tgcccaccgg gtgctgcctg tgctggtatc ggcgcggcgc atcagcgagg aaatggcgaa      1680
     gctgaagcag aagaccgagc agctccagga ccgcatcgcc ggccaggccc ggatcaacca      1740
     ggccaaggtg ttgctgatgc agcgccatgg ctgggacgag cgcgaggcgc accagcacct      1800
     gtcgcgggaa gcgatgaagc ggcgcgagcc gatcctgaag atcgctcagg agttgctggg      1860
     aaacgagccg tccgcctgag cgatccgggc cgaccagaac aataacaaga ggggtatcgt      1920
     catcatgctg ggactggttc tgctgtacgt tggcgcggtg ctgtttctca atgccgtctg      1980
     gttgctgggc aagatcagcg gtcgggaggt ggcggtgatc aacttcctgg tcggcgtgct      2040
     gagcgcctgc gtcgcgttct acctgatctt ttccgcagca gccgggcagg gctcgctgaa      2100
     ggccggagcg ctgaccctgc tattcgcttt tacctatctg tgggtggccg ccaaccagtt      2160
     cctcgag                                                                2167
//
</pre>
</td></tr></table><p>





<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 

        If you wish to include the standard description of the avalable
	report formats, use:
        #include file="inc/reportformats.ihtml"
   -->

<b>xmltext</b> writes the text of the original entry as a text file.

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: x13776.xmltext</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&lt;?xml version="1.0" encoding="UTF-8"?&gt;
&lt;ROOT request="X13776&amp;amp;display=xml&amp;amp;limit=200"&gt;
&lt;entry accession="X13776" version="1" entryVersion="24" dataClass="STD" taxonomicDivision="PRO" moleculeType="genomic DNA" sequenceLength="2167" topology="linear" firstPublic="1989-04-19" firstPublicRelease="19" lastUpdated="2006-11-14" lastUpdatedRelease="89"&gt;
	&lt;secondaryAccession&gt;M43175&lt;/secondaryAccession&gt;
	&lt;description&gt;Pseudomonas aeruginosa amiC and amiR gene for aliphatic amidase regulation&lt;/description&gt;
	&lt;comment&gt;
Data kindly reviewed (04-SEP-1991) by Rice P. M.
	&lt;/comment&gt;
	&lt;keyword&gt;aliphatic amidase regulator&lt;/keyword&gt;
	&lt;keyword&gt;amiC gene&lt;/keyword&gt;
	&lt;keyword&gt;amiR gene&lt;/keyword&gt;
	&lt;reference type="submission" number="1" location="1167-2167"&gt;
		&lt;author&gt;Rice P.M.&lt;/author&gt;
		&lt;submissionDate&gt;1988-12-16&lt;/submissionDate&gt;
		&lt;referenceLocation&gt;
Submitted (16-DEC-1988) to the INSDC.
Rice P.M., EMBL, Postfach 10-2209, Meyerhofstrasse 1, 6900 Heidelberg, FRG.
		&lt;/referenceLocation&gt;
	&lt;/reference&gt;
	&lt;reference type="article" number="2" location="1167-2167"&gt;
		&lt;title&gt;Nucleotide sequence of the aliphatic amidase regulator gene (amiR) of Pseudomonas aeruginosa&lt;/title&gt;
		&lt;author&gt;Lowe N.&lt;/author&gt;
		&lt;author&gt;Rice P.M.&lt;/author&gt;
		&lt;author&gt;Drew R.E.&lt;/author&gt;
		&lt;journal&gt;FEBS Lett.&lt;/journal&gt;
		&lt;year&gt;1989&lt;/year&gt;
		&lt;volume&gt;246&lt;/volume&gt;
		&lt;issue&gt;1-2&lt;/issue&gt;
		&lt;firstPage&gt;39&lt;/firstPage&gt;
		&lt;lastPage&gt;43&lt;/lastPage&gt;
		&lt;referenceLocation&gt;
FEBS Lett. 246(1-2):39-43(1989).
		&lt;/referenceLocation&gt;
		&lt;xref db="DOI" id="10.1016/0014-5793(89)80249-2"/&gt;
		&lt;xref db="PUBMED" id="2495988"/&gt;
	&lt;/reference&gt;
	&lt;reference type="article" number="3" location="1-1292"&gt;
		&lt;title&gt;Cloning and DNA sequence of amiC, a new gene regulating expression of the Pseudomonas aeruginosa aliphatic amidase, and purification of the amiC product&lt;/title&gt;
		&lt;author&gt;Wilson S.&lt;/author&gt;
		&lt;author&gt;Drew R.&lt;/author&gt;
		&lt;journal&gt;J. Bacteriol.&lt;/journal&gt;
		&lt;year&gt;1991&lt;/year&gt;
		&lt;volume&gt;173&lt;/volume&gt;
		&lt;issue&gt;16&lt;/issue&gt;
		&lt;firstPage&gt;4914&lt;/firstPage&gt;
		&lt;lastPage&gt;4921&lt;/lastPage&gt;
		&lt;referenceLocation&gt;
J. Bacteriol. 173(16):4914-4921(1991).
		&lt;/referenceLocation&gt;
		&lt;xref db="PUBMED" id="1907262"/&gt;


<font color=red>  [Part of this file has been deleted for brevity]</font>

		&lt;/qualifier&gt;
		&lt;qualifier name="translation"&gt;
			&lt;value&gt;
MSANSLLGSLRELQVLVLNPPGEVSDALVLQLIRIGCSVRQCWPPPEAFDVPVDVVFTSIFQNGHHDEIAALLAAGTPRT
TLVALVEYESPAVLSQIIELECHGVITQPLDAHRVLPVLVSARRISEEMAKLKQKTEQLQDRIAGQARINQAKVLLMQRH
GWDEREAHQHLSREAMKRREPILKIAQELLGNEPSA
			&lt;/value&gt;
		&lt;/qualifier&gt;
	&lt;/feature&gt;
	&lt;sequence&gt;
ggtaccgctggccgagcatctgctcgatcaccaccagccgggcgacgggaactgcacgat
ctacctggcgagcctggagcacgagcgggttcgcttcgtacggcgctgagcgacagtcac
aggagaggaaacggatgggatcgcaccaggagcggccgctgatcggcctgctgttctccg
aaaccggcgtcaccgccgatatcgagcgctcgcacgcgtatggcgcattgctcgcggtcg
agcaactgaaccgcgagggcggcgtcggcggtcgcccgatcgaaacgctgtcccaggacc
ccggcggcgacccggaccgctatcggctgtgcgccgaggacttcattcgcaaccgggggg
tacggttcctcgtgggctgctacatgtcgcacacgcgcaaggcggtgatgccggtggtcg
agcgcgccgacgcgctgctctgctacccgaccccctacgagggcttcgagtattcgccga
acatcgtctacggcggtccggcgccgaaccagaacagtgcgccgctggcggcgtacctga
ttcgccactacggcgagcgggtggtgttcatcggctcggactacatctatccgcgggaaa
gcaaccatgtgatgcgccacctgtatcgccagcacggcggcacggtgctcgaggaaatct
acattccgctgtatccctccgacgacgacttgcagcgcgccgtcgagcgcatctaccagg
cgcgcgccgacgtggtcttctccaccgtggtgggcaccggcaccgccgagctgtatcgcg
ccatcgcccgtcgctacggcgacggcaggcggccgccgatcgccagcctgaccaccagcg
aggcggaggtggcgaagatggagagtgacgtggcagaggggcaggtggtggtcgcgcctt
acttctccagcatcgatacgcccgccagccgggccttcgtccaggcctgccatggtttct
tcccggagaacgcgaccatcaccgcctgggccgaggcggcctactggcagaccttgttgc
tcggccgcgccgcgcaggccgcaggcaactggcgggtggaagacgtgcagcggcacctgt
acgacatcgacatcgacgcgccacaggggccggtccgggtggagcgccagaacaaccaca
gccgcctgtcttcgcgcatcgcggaaatcgatgcgcgcggcgtgttccaggtccgctggc
agtcgcccgaaccgattcgccccgacccttatgtcgtcgtgcataacctcgacgactggt
ccgccagcatgggcgggggaccgctcccatgagcgccaactcgctgctcggcagcctgcg
cgagttgcaggtgctggtcctcaacccgccgggggaggtcagcgacgccctggtcttgca
gctgatccgcatcggttgttcggtgcgccagtgctggccgccgccggaagccttcgacgt
gccggtggacgtggtcttcaccagcattttccagaatggccaccacgacgagatcgctgc
gctgctcgccgccgggactccgcgcactaccctggtggcgctggtggagtacgaaagccc
cgcggtgctctcgcagatcatcgagctggagtgccacggcgtgatcacccagccgctcga
tgcccaccgggtgctgcctgtgctggtatcggcgcggcgcatcagcgaggaaatggcgaa
gctgaagcagaagaccgagcagctccaggaccgcatcgccggccaggcccggatcaacca
ggccaaggtgttgctgatgcagcgccatggctgggacgagcgcgaggcgcaccagcacct
gtcgcgggaagcgatgaagcggcgcgagccgatcctgaagatcgctcaggagttgctggg
aaacgagccgtccgcctgagcgatccgggccgaccagaacaataacaagaggggtatcgt
catcatgctgggactggttctgctgtacgttggcgcggtgctgtttctcaatgccgtctg
gttgctgggcaagatcagcggtcgggaggtggcggtgatcaacttcctggtcggcgtgct
gagcgcctgcgtcgcgttctacctgatcttttccgcagcagccgggcagggctcgctgaa
ggccggagcgctgaccctgctattcgcttttacctatctgtgggtggccgccaaccagtt
cctcgag
	&lt;/sequence&gt;
&lt;/entry&gt;
&lt;/ROOT&gt;

</pre>
</td></tr></table><p>





<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.

	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:

        #include file="inc/localfiles.ihtml"
   -->

None.

<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
<ol>

<li>

</ol>

   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="drtext.html">drtext</a></td>
<td>Get data resource entries complete text</td>
</tr>

<tr>
<td><a href="entret.html">entret</a></td>
<td>Retrieve sequence entries from flatfile databases and files</td>
</tr>

<tr>
<td><a href="ontotext.html">ontotext</a></td>
<td>Get ontology term(s) original full text</td>
</tr>

<tr>
<td><a href="textget.html">textget</a></td>
<td>Get text data entries</td>
</tr>

<tr>
<td><a href="textsearch.html">textsearch</a></td>
<td>Search the textual description of sequence(s)</td>
</tr>

</table>
<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>

Peter Rice
<br>
European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK

<p>
Please report all bugs to the EMBOSS bug team (emboss-bug&nbsp;&copy;&nbsp;emboss.open-bio.org) not to the original author.

<H2>
    History
</H2>
<!--
        Date written and what changes have been made go in this file.
   -->




<H2>
    Target users
</H2>
<!--
        For general users, use this text
   -->
This program is intended to be used by everyone and everything, from naive users to embedded scripts.

<H2>
    Comments
</H2>
<!--
        User/developer/other comments go in this file.
   -->
None


</BODY>
</HTML>

