<HTML>

<HEAD>
  <TITLE>
  EMBOSS: cpgplot
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<!--#include file="header1.inc" -->
cpgplot
<!--#include file="header2.inc" -->

<H2>
    Function
</H2>
<!--#include file="inc/cpgplot.ione" -->

<H2>
    Description
</H2>

CpG refers to a C nucleotide immediately followed by a G.  The 'p' in
'CpG' refers to the phosphate group linking the two bases. 

<p>
Detection of regions of genomic sequences that are rich in the CpG
pattern is important because such regions are resistant to methylation
and tend to be associated with genes which are frequently switched on. 
Regions rich in the CpG pattern are known as CpG islands.  

<p>
It has been estimated that about half of all mammalian genes have a
CpG-rich region around their 5' end.  It is said that all mammalian
house-keeping genes have a CpG island!

<p>
Non-mammalian vertebrates have some CpG islands that are associated with
genes, but the association gets equivocal in the farther taxonomic
groups. 

<p>
Finding a CpG island upstream of predicted exons or genes is good
contributory evidence. 

<p>

By default, this program defines a CpG island as a region where, over an
average of 10 windows, the calculated % composition is over 50% and the
calculated Obs/Exp (i.e.  Observed/Expected) ratio is over 0.6 and the
conditions hold for a minimum of 200 bases.  These conditions can be
modified by setting the values of the appropriate parameters. 

<p>


The Observed number of CpG patterns in a window is simply the count of
the number of times a 'C' is found followed immediately by a 'G'. 

<p>

The Expected number of CpG patterns is calculated for each window as the
number of CpG dinucleotides you would expect to see in that window based
on the frequency of C's and G's in that window.  Thus, the Expected
frequency of CpG's in a window is calculated as the number of 'C's in
the window multiplied by the number of 'G's in the window, divided by
the window length. 

<p>

<pre>
	Expected = (number of C's  * number of G's) / window length


</pre>

<p>

This program reads in one or more sequences and calculates the Obs/Exp
ratio, the percentage CpG over a window which is moved along the
sequence.  These calculated values can be plotted, together with the
regions which match this program's definition of a CpG island. 

<H2>
    Usage
</H2>
<!--#include file="inc/cpgplot.usage" -->

<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
<!--#include file="inc/cpgplot.ihelp" -->
</pre>
</td></tr></table>
<P>
<!--#include file="inc/cpgplot.itable" -->

<H2>
    Input file format
</H2>

Any DNA sequence USA.

<p>
<!--#include file="inc/cpgplot.input" -->

<H2>
    Output file format
</H2>
<!--#include file="inc/cpgplot.output" -->


<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

The original program was described in:

<ol>

<li>
Larsen F, Gundersen G, Lopez R, Prydz H
"CpG islands as gene markers in the human genome."
Genomics 1992 Aug;13(4):1095-107

</ol>

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

    0 if successful.

<H2>
    Known bugs
</H2>

None.

<!--#include file="inc/cpgplot.isee" -->


<p>

As there is no official definition of what is a cpg island is, and worst
where they begin and end, we have to live with 2 definitions and thus
two methods.  These are:

<p>

1.  <b>newcpgseek</b> and <b>cpgreport</b> - both declare a putative
island if the score is higher than a threshold (17 at the moment).  They
now also displaying the actual CpG count, the % CG and the
observed/expected ration in the region where the score is above the
threshold.  This scoring method based on sum/frequencies overpredicts
islands but finds the smaller ones around primary exons.
<b>newcpgseek</b> uses the same method as <b>cpgreport</b> but the
output is different and more readable.

<p>

2.  <b>newcpgreport</b> and <b>cpgplot</b> use a sliding window within
which the Obs/Exp ratio of CpG is calculated.  The important thing to
note in this method is that an island, in order to be reported, is
defined as a region that satisfies the following contraints:

<p>
<pre>
   Obs/Exp ratio > 0.6
   % C + % G > 50%
   Length > 200.
</pre>
<p>

<p>

For all practical purposes you should probably use newcpgreport.  It is
actually used to produce the human cpgisland database you can find on
the EBI's ftp server as well as on the EBI's SRS server.

<p>

<b>geecee</b> measures CG content in the entire input sequence and is
not to be used to detect CpG islands.  It can be usefull for detecting
sequences that MIGHT contain an island.
        

<H2>
    Author(s)
</H2>
<!--#include file="inc/ableasby.address" -->

<H2>
    History
</H2>
<!--#include file="inc/cpgplot.history" -->

<H2>
    Target users
</H2>
<!--#include file="inc/target.itxt" -->

<H2>
    Comments
</H2>
<!--#include file="inc/cpgplot.comment" -->

</BODY>
</HTML>

