<HTML>

<HEAD>
  <TITLE>
  EMBOSS
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<!--#include file="header1.inc" -->
diffseq
<!--#include file="header2.inc" -->

<H2>
    Function
</H2>
<!--#include file="inc/diffseq.ione" -->

<H2>
    Description
</H2>

<b>diffseq</b> takes two overlapping, nearly identical sequences and
reports the differences between them, together with any features that
overlap with these regions. GFF files of the differences in each
sequence are also produced.

<p>

<b>diffseq</b> finds the region of overlap of the input sequences and
then reports differences within this region, like a local alignment.

<p>

The start and end positions of the overlap are reported.

<p>

<b>diffseq</b> should be of value when looking for SNPs, differences
between strains of an organism and anything else that requires the
differences between sequences to be highlighted.

<p>  

The sequences can be very long. The program does a match of all
sequence words of size 10 (by default). It then reduces this to the
minimum set of overlapping matches by sorting the matches in order of
size (largest size first) and then for each such match it removes any
smaller matches that overlap. The result is a set of the longest
ungapped alignments between the two sequences that do not overlap with
each other. The mismatched regions between these matches are reported.

<p>   

It should be possible to find differences between sequences that are Mega-bases
long.


<H2>
    Usage
</H2>
<!--#include file="inc/diffseq.usage" -->

<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
<!--#include file="inc/diffseq.ihelp" -->
</pre>
</td></tr></table>
<P>
<!--#include file="inc/diffseq.itable" -->

<H2>
    Input file format
</H2>
This program reads in two nucleic acid sequence USAs or two protein
sequence USAs. 

<p>
<!--#include file="inc/diffseq.input" -->


<H2>
    Output file format
</H2>

<!--#include file="inc/reportformats.ihtml" -->
<p>

By default <b>diffseq</b> writes a 'diffseq' report file.

<p>
<!--#include file="inc/diffseq.output" -->

<p>

The first line is the title giving the names of the sequences used.
<p>
The next two non-blank lines state the positions in each sequence
where the detected overlap between them starts.
<p>
There then follows a set of reports of the mismatches between the sequences.
<br>
Each report consists of 4 or more lines.
<br>
<ul>

<li>The first line has the name of the first sequence followed by the
start and end positions of the mismatched region in that sequence,
followed by the length of the mismatched region. If the mismatched
region is of zero length in this sequence, then only the position of
the last matching base before the mismatch is given.</li>

<li>If a feature of the first sequence overlaps with this mismatch
region, then one or more lines starting with 'Feature:' comes next
with the type, position and tag field of the feature.</li>

<li>Next is a line starting "Sequence:" giving the sequence of the
mismatch in the first sequence.</li> </ul>

<p>

This is followed by the equivalent information for the second
sequence, but in the reverse order, namely 'Sequence:' line,
'Feature:' lines and line giving the position of the mismatch in the
second sequence.
<p>

At the end of the report are two non-blank lines giving the positions in each
sequence where the detected
overlap between them ends.
<p>

The last three lines of the report gives the counts of SNPs (defined
as a change of one nucleotide to one other nucleotide, no deletions or
insertions are counted, no multi-base changes are counted).
<p>

If the input sequences are nucleic acid, The counts of transitions
(Pyrimide to Pyrimidine or Purine to Purine) and transversions
(Pyrimidine to Purine) are also given. 

<p>

It should be noted that not all features are reported.

<p>

The 'source' feature found in all EMBL/Genbank feature table entries is
not reported as this covers all of the sequence and so overlaps with any
difference found in that sequence and so is uninformative and
irritating.  It has therefore been removed from the output report. 

<p>

The translation information of CDS features is often extremely long and
does not add useful information to the report.  It has therefore been
removed from the output report. 

<H2>
    Data files
</H2>

None

<H2>
    Notes
</H2>

It should be noted that not all features are reported.

<p>

The 'source' feature found in all EMBL/Genbank feature table entries is
not reported as this covers all of the sequence and so overlaps with any
difference found in that sequence and so is uninformative and
irritating.  It has therefore been removed from the output report. 

<p>

The translation information of CDS features is often extremely long and
does not add useful information to the report.  It has therefore been
removed from the output report. 

<p>

If you run out of memory, use a larger word size.

<p>

Using a larger word size increases the length between mismatches that
will be reported as one event.  Thus a word size of 50 will report two
single-base differences that are with 50 bases of each other as one mismatch. 

   
<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with status 0.

<H2>
    Known bugs
</H2>

None.

<!--#include file="inc/diffseq.isee" -->

<H2>
    Author(s)
</H2>

<!--#include file="inc/gwilliam.address" -->

<H2>
    History
</H2>
<!--#include file="inc/diffseq.history" -->

<H2>
    Target users
</H2>
<!--#include file="inc/target.itxt" -->

<H2>
    Comments
</H2>
<!--#include file="inc/diffseq.comment" -->


</BODY>
</HTML>

