<HTML>

<HEAD>
  <TITLE>
  EMBOSS: newcpgseek
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<!--#include file="header1.inc" -->
newcpgseek
<!--#include file="header2.inc" -->

<H2>
    Function
</H2>
<!--#include file="inc/newcpgseek.ione" -->


<H2>
    Description
</H2>


<b>newcpgseek</b> reports CpG rich regions of a sequence as candidate
CpG islands. 

<p>

CpG refers to a C nucleotide immediately followed by a G.  The 'p' in
'CpG' refers to the phosphate group linking the two bases. 

<p>
Detection of regions of genomic sequences that are rich in the CpG
pattern is important because such regions are resistant to methylation
and tend to be associated with genes which are frequently switched on. 
Regions rich in the CpG pattern are known as CpG islands.  

<p>
It has been estimated that about half of all mammalian genes have a
CpG-rich region around their 5' end.  It is said that all mammalian
house-keeping genes have a CpG island!

<p>
Non-mammalian vertebrates have some CpG islands that are associated with
genes, but the association gets equivocal in the farther taxonomic
groups. 

<p>
Finding a CpG island upstream of predicted exons or genes is good
contributory evidence. 

<p>

CpG islands are usually defined as "length over 200bp with %GC over
50% and obs/ervedexpexted CpG more than 0.6".  However this program
uses a running sum rather than a window to produce a score: if there
is not a CpG at position i, then decrement runSum counter, but if CpG
then runSum += CPGSCORE.  Spans above the threshold are searched for
recursively.  If the score is higher than a threshold (17 at the
moment) then a putative island is declared.

<p>

This program reads in one or more sequences and finds regions where
there is a high absolute frequency of CpG dimers as well as a high
proportion of CpG compared to GpC. 



<H2>
    Usage
</H2>
<!--#include file="inc/newcpgseek.usage" -->


<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
<!--#include file="inc/newcpgseek.ihelp" -->
</pre>
</td></tr></table>
<P>

<!--#include file="inc/newcpgseek.itable" -->


<H2>
    Input file format
</H2>

<b>newcpgseek</b> reads a nucleic acid sequence USA.
	
<p>

<!--#include file="inc/newcpgseek.input" -->

<H2>
    Output file format
</H2>


<!--#include file="inc/newcpgseek.output" -->

<H2>
    Data files
</H2>

None.

<H2>
    Notes
</H2>

None.

<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>

None.

<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.

<!--#include file="inc/newcpgseek.isee" -->

<p>

As there is no official definition of what is a cpg island is, and worst
where they begin and end, we have to live with 2 definitions and thus
two methods.  These are:

<p>

1.  <b>newcpgseek</b> and <b>cpgreport</b> - both declare a putative
island if the score is higher than a threshold (17 at the moment).  They
now also displaying the actual CpG count, the % CG and the
observed/expected ration in the region where the score is above the
threshold.  This scoring method based on sum/frequencies overpredicts
islands but finds the smaller ones around primary exons.
<b>newcpgseek</b> uses the same method as <b>cpgreport</b> but the
output is different and more readable.

<p>

2.  <b>newcpgreport</b> and <b>cpgplot</b> use a sliding window within
which the Obs/Exp ratio of CpG is calculated.  The important thing to
note in this method is that an island, in order to be reported, is
defined as a region that satisfies the following contraints:

<p>
<pre>
   Obs/Exp ratio > 0.6
   % C + % G > 50%
   Length > 200.
</pre>
<p>

<p>

For all practical purposes you should probably use newcpgreport.  It is
actually used to produce the human cpgisland database you can find on
the EBI's ftp server as well as on the EBI's SRS server.

<p>

<b>geecee</b> measures CG content in the entire input sequence and is
not to be used to detect CpG islands.  It can be usefull for detecting
sequences that MIGHT contain an island.
         


<H2>
    Author(s)
</H2>


<!--#include file="inc/rlopez.address" -->


<H2>
    History
</H2>
<!--#include file="inc/newcpgseek.history" -->

<H2>
    Target users
</H2>
<!--#include file="inc/target.itxt" -->


<H2>
    Comments
</H2>
<!--#include file="inc/newcpgseek.comment" -->

</BODY>
</HTML>

