<HTML>

<HEAD>
  <TITLE>
  EMBOSS: Applications not found in GCG
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="http://emboss.sourceforge.net/index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true">
<img border=0 src="http://emboss.sourceforge.net/emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6"> 
EMBOSS: Applications not found in GCG
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



This is a rough list of the <a
href="http://emboss.sourceforge.net/apps/index.html">EMBOSS
applications</a> that have functions which are not covered by any GCG
program. 

<p>

Many EMBOSS programs have been excluded from this list because they do
most of what a GCG programs does, even though they have some extra
functionality that GCG doesn't have.  For example, '<a
href="http://emboss.sourceforge.net/apps/showseq.html">showseq</a>' has been excluded because most of what
it does is very similar to the GCG program 'publish', even though it can
mark the sequence with annotations from the feature table on the
sequence display which the GCG program cannot. 

<p>

It is therefore worth investigating the corresponding EMBOSS program
carefully to see what extra things it can do that the GCG program cannot. 

<p>

Note that all of the EMBOSS programs can read in sequences that are in
any of the known, standard sequence formats.  The EMBOSS program will
test the sequence it reads in to determine what the format is, so you
can use sequences in the formats GCG, MFS, GenBank, EMBL, FASTA, etc,
etc. 

<p>

By default, EMBOSS programs write the sequences in fasta format, but you
can explicitly tell an EMBOSS program to write the sequence in a
specified format, or you can set the default format for all EMBOSS
programs to use when writing out. 

<p>

<a href="http://emboss.sourceforge.net/apps/index.html">The
full list of EMBOSS applications is here</a> <br> <a
href="http://emboss.sourceforge.net/apps/groups.html"> and
the applications grouped by function is here</a>.

<p>

The EMBOSS developers are always interested in hearing your requirements
and suggestions for improvements.  They can be contacted via the <a
href="http://emboss.sourceforge.net/support/">mailing
lists</a>. 

<p>

Note also that the sources of the EMBOSS programs are available for your
local experts to inspect and change to meet your local requirements. 

<p>

<table border cellpadding=4 bgcolor="#FFFFF0">

<tr>
<th>Program name</th>
<th>Description</th>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/abiview.html">abiview</a></td>
<td>Reads ABI file and display the trace</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/antigenic.html">antigenic</a></td>
<td>Finds antigenic sites in proteins</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/banana.html">banana</a></td>
<td>Bending and curvature plot in B-DNA</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/biosed.html">biosed</a></td>
<td>Replace or delete sequence sections</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/btwisted.html">btwisted</a></td>
<td>Calculates the twisting in a B-DNA sequence</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/cai.html">cai</a></td>
<td>CAI codon adaptation index</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/chaos.html">chaos</a></td>
<td>Create a chaos game representation plot for a sequence</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/charge.html">charge</a></td>
<td>Protein charge plot</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/checktrans.html">checktrans</a></td>
<td>Reports STOP codons and ORF statistics of a protein</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/chips.html">chips</a></td>
<td>Codon usage statistics</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/coderet.html">coderet</a></td>
<td>Extract CDS, mRNA and translations from feature tables</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/cpgplot.html">cpgplot</a></td>
<td>Plot CpG rich areas</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/cpgreport.html">cpgreport</a></td>
<td>Reports all CpG rich regions</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/cutseq.html">cutseq</a></td>
<td>Removes a specified section from a sequence</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/dbiblast.html">dbiblast</a></td>
<td>Index a BLAST database</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/dbifasta.html">dbifasta</a></td>
<td>Index a fasta database</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/dbiflat.html">dbiflat</a></td>
<td>Index a flat file database</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/degapseq.html">degapseq</a></td>
<td>Removes gap characters from sequences </td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/descseq.html">descseq</a></td>
<td>Alter the name or description of a sequence</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/diffseq.html">diffseq</a></td>
<td>Find differences between nearly identical sequences</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/dotpath.html">dotpath</a></td>
<td>Displays a non-overlapping wordmatch dotplot of two sequences</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/dreg.html">dreg</a></td>
<td>regular expression search of a nucleotide sequence</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/embossversion.html">embossversion</a></td>
<td>Writes the current EMBOSS version number</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/emowse.html">emowse</a></td>
<td>Protein identification by mass spectrometry</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/entret.html">entret</a></td>
<td>Reads and writes (returns) flatfile entries</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/est2genome.html">est2genome</a></td>
<td>Align EST and genomic DNA sequences</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/extractfeat.html">extractfeat</a></td>
<td>Extract features from a sequence</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/extractseq.html">extractseq</a></td>
<td>Extract regions from a sequence</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/findkm.html">findkm</a></td>
<td>Find Km and Vmax for an enzyme reaction by a Hanes/Woolf plot</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/fuzztran.html">fuzztran</a></td>
<td>Protein pattern search after translation </td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/garnier.html">garnier</a></td>
<td>Predicts protein secondary structure </td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/geecee.html">geecee</a></td>
<td>Calculates the fractional GC content of nucleic acid sequences </td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/getorf.html">getorf</a></td>
<td>Finds and extracts open reading frames (ORFs) </td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/infoalign.html">infoalign</a></td>
<td>Information on a multiple sequence alignment</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/infoseq.html">infoseq</a></td>
<td>Displays some simple information about sequences</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/isochore.html">isochore</a></td>
<td>Plots isochores in large DNA sequences </td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/listor.html">listor</a></td>
<td>Writes a list file of the logical OR of two sets of sequences</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/marscan.html">marscan</a></td>
<td>Finds MAR/SAR sites in nucleic sequences</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/maskfeat.html">maskfeat</a></td>
<td>Mask off features of a sequence</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/maskseq.html">maskseq</a></td>
<td>Mask off regions of a sequence</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/megamerger.html">megamerger</a></td>
<td>Merge two large overlapping nucleic acid sequences</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/merger.html">merger</a></td>
<td>Merge two overlapping nucleic acid sequences</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/mwcontam.html">mwcontam</a></td>
<td>Shows molwts that match across a set of files</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/mwfilter.html">mwfilter</a></td>
<td>Filter noisy molwts from mass spec output</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/newcpgreport.html">newcpgreport</a></td>
<td>Report CpG rich areas</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/newcpgseek.html">newcpgseek</a></td>
<td>Reports CpG rich regions</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/notseq.html">notseq</a></td>
<td>Excludes a set of sequences and writes out the remaining ones</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/nthseq.html">nthseq</a></td>
<td>Writes one sequence from a multiple set of sequences</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/octanol.html">octanol</a></td>
<td>Displays protein hydropathy</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/oddcomp.html">oddcomp</a></td>
<td>Finds protein sequence regions with a biased composition</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/pasteseq.html">pasteseq</a></td>
<td>Insert one sequence into another</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/pepinfo.html">pepinfo</a></td>
<td>Plots simple amino acid properties in parallel </td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/pepnet.html">pepnet</a></td>
<td>Displays proteins as a helical net </td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/pepstats.html">pepstats</a></td>
<td>Protein statistics </td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/pepwindow.html">pepwindow</a></td>
<td>Displays protein hydropathy </td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/pepwindowall.html">pepwindowall</a></td>
<td>Displays protein hydropathy of a set of sequences </td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/pestfind.html">pestfind</a></td>
<td>Finds PEST motifs as potential proteolytic cleavage sites </td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/polydot.html">polydot</a></td>
<td>Displays all-against-all dotplots of a set of sequences</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/preg.html">preg</a></td>
<td>Regular expression search of a protein sequence</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/pscan.html">pscan</a></td>
<td>Scans proteins using PRINTS </td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/recoder.html">recoder</a></td>
<td>Remove restriction sites but maintain the same translation</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/redata.html">redata</a></td>
<td>Search REBASE for enzyme name, references, suppliers etc</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/restover.html">restover</a></td>
<td>Finds restriction enzymes that produce a specific overhang </td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/seealso.html">seealso</a></td>
<td>Finds programs sharing group names</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/seqmatchall.html">seqmatchall</a></td>
<td>Does an all-against-all comparison of a set of sequences</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/seqretsplit.html">seqretsplit</a></td>
<td>Reads and writes (returns) sequences in individual files</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/showdb.html">showdb</a></td>
<td>Displays information on the currently available databases</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/showfeat.html">showfeat</a></td>
<td>Show features of a sequence</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/sigcleave.html">sigcleave</a></td>
<td>Reports protein signal cleavage sites </td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/silent.html">silent</a></td>
<td>Silent mutation restriction enzyme scan </td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/sirna.html">sirna</a></td>
<td>Finds siRNA duplexes in mRNA</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/skipseq.html">skipseq</a></td>
<td>Reads and writes (returns) sequences, skipping the first few</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/splitter.html">splitter</a></td>
<td>Split a sequence into (overlapping) smaller sequences </td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/stssearch.html">stssearch</a></td>
<td>Searches a DNA database for matches with a set of STS primers </td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/supermatcher.html">supermatcher</a></td>
<td>Finds a match of a large sequence against one or more sequences </td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/tmap.html">tmap</a></td>
<td>Displays membrane spanning regions </td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/tranalign.html">tranalign</a></td>
<td>Align nucleic coding regions given the aligned proteins </td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/trimest.html">trimest</a></td>
<td>Trim poly-A tails off EST sequences</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/trimseq.html">trimseq</a></td>
<td>Trim ambiguous bits off the ends of sequences</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/twofeat.html">twofeat</a></td>
<td>Finds neighbouring pairs of features in sequences</td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/vectorstrip.html">vectorstrip</a></td>
<td>Strips out DNA between a pair of vector sequences </td>
</tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/wossname.html">wossname</a></td>
<td>Finds programs by keywords in their one-line documentation</td>
</tr>



</table>




</BODY>
</HTML>
