<HTML>

<HEAD>
  <TITLE>
  EMBOSS: supermatcher
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<!--#include file="header1.inc" -->
supermatcher
<!--#include file="header2.inc" -->

<H2>
    Function
</H2>
<!--#include file="inc/supermatcher.ione" -->


<H2>
    Description
</H2>
<p><b>supermatcher</b> calculates an approximate alignment between all the sequences in a first set of sequences and all those from a second stream, typically a database. The alignments are written to a standard alignment file.  A combination of a word-match and Smith-Waterman local alignment (dynamic programming) algorithms are used.  The alignments will be less acurate than an optimal alignment generated by full dynamic programming, but the program will run faster and use less memory, which means it is suitable for use with larger sequences.</p>

<H2>
    Usage
</H2>
<!--#include file="inc/supermatcher.usage" -->


<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
<!--#include file="inc/supermatcher.ihelp" -->
</pre>
</td></tr></table>
<P>

<!--#include file="inc/supermatcher.itable" -->


<H2>
    Input file format
</H2>

<b>supermatcher</b> reads two sequence USAs of the same type.  They must
both be protein or both be nucleic acid sequences. 

<p>

<!--#include file="inc/supermatcher.input" -->

<H2>
    Output file format
</H2>

<!--#include file="inc/alignformats.ihtml" -->

<p>

The output alignment is in <b>simple</b> format by default.

<p>

<!--#include file="inc/supermatcher.output" -->

<p>

The file 'supermatcher.error' will contain any errors that occured
during the program.  This may be that wordmatch could not find any
matches hence no suitable start point is found for the smith-waterman
calculation. 


<H2>
    Data files
</H2>

For protein sequences EBLOSUM62 is used for the substitution
    matrix. For nucleotide sequence, EDNAMAT is used. Others can be specified.

<p>
<!--#include file="inc/localfiles.ihtml" -->

<H2>
    Notes
</H2>
<p><b>supermatcher</b> generates approximate local alignments for large sequences. The alignments are approximate because as a first step, all the sequence word matches between two sequences are found.  By identifying the highest scoring, non-overlapping matches a set of approximate local alignments are calculated for two sequences.  These give the centre points for more acurate Smith-Waterman type alignments in a region of width specified by the user. The use of Smith-Waterman in narrow regions means the alignment overall will be rough, but due to the memory saving much larger sequences can be aligned.</p>

<p>For the Swmith-Waterman alignment, the gap open and extension penalties and substition matrix may be specified, the later by default is EBLOSUM62 for protein sequences and EDNAMAT for nucleotide sequence.</p>

<p>For the word-match alignment, the word length may be specified. The time required for alignment depends very much on word size. A small word size (e.g. 4) may take a very long time even for short sequences. Much larger word sizes (e.g. 30) will give a very quick result. The default of 16 should give reasonably fast alignments.</p>




<H2>
    References
</H2>

None.

<H2>
    Warnings
</H2>
<p><b>supermatcher</b> performs a Smith & Waterman alignment (albeit in a narrow best-matching regions identified by simple word-match) and therefore can use huge amounts of memory if the sequences are large.  The longer the sequences and the wider the specified alignment width, the more memory will be used.  If the program terminates due to lack of memory you can try running the UNIX command <tt>limit</tt> to see if your stack or memory usage have been limited and if so, run <tt>unlimit</tt>, (e.g.: <tt>% unlimit stacksize</tt>).</p>

<p><b>supermatcher</b> has two sequence inputs. The first (<tt>asequence</tt>) is a database or large file which is read one sequence at a time. The second (<tt>bsequence</tt>) is a "sequence set" which is loaded into memory. If one of the inputs has more number of sequences than the other one, specifying the file with many number of sequences first would be useful to decrease the amount of memory used while supermatcher running.</p>

<p>Because the alignment is made within a narrow area each side of the 'best' diagonal identified by word-matching, if there are sufficient indels between the two sequences, then the path of the Smith & Waterman alignment can wander outside of this area. Making the width larger can avoid this problem, but you then use more memory.</p>


<H2>
    Diagnostic Error Messages
</H2>

None.

<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>

None.

<!--#include file="inc/supermatcher.isee" -->

<H2>
    Author(s)
</H2>


<!--#include file="inc/ilongden.address" -->


<H2>
    History
</H2>
<!--#include file="inc/supermatcher.history" -->

<H2>
    Target users
</H2>
<!--#include file="inc/target.itxt" -->



<H2>
    Comments
</H2>
<!--#include file="inc/supermatcher.comment" -->

</BODY>
</HTML>

