<HTML>

<HEAD>
  <TITLE>
  EMBOSS: tfextract
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<!--#include file="header1.inc" -->
tfextract
<!--#include file="header2.inc" -->

<H2>
    Function
</H2>

<!--#include file="inc/tfextract.ione" -->


<H2>
    Description
</H2>

<p><b>tfextract</b> extracts data from the TRANSFAC transcription factor database file site.dat (available from <a href="ftp://ftp.ebi.ac.uk/pub/databases/transfac/">ftp://ftp.ebi.ac.uk/pub/databases/transfac/</a>) for other EMBOSS programs, such as <b>tfscan</b>, that use these data. The data is split up by taxonomic groups and placed in individual files that are stored in the EMBOSS data directory. </p>



<H2>
    Usage
</H2>
<!--#include file="inc/tfextract.usage" -->


<H2>
    Command line arguments
</H2>
<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
<!--#include file="inc/tfextract.ihelp" -->
</pre>
</td></tr></table>
<P>
<!--#include file="inc/tfextract.itable" -->


<H2>
    Input file format
</H2>

It reads in the TRANSFAC file <b>site.dat</b> available from:
<p>
<!--
<A HREF="http://transfac.gbf.de/cgi-bin/download/download.pl">http://transfac.gbf.de/cgi-bin/download/download.pl</A>
-->
<!--
<A HREF="ftp://transfac.gbf.de/pub/transfac/ascii/old/tf_3.2/transfac32.tar.Z">
ftp://transfac.gbf.de/pub/transfac/ascii/old/tf_3.2/transfac32.tar.Z
-->


<A HREF="ftp://ftp.ebi.ac.uk/pub/databases/transfac/">
ftp://ftp.ebi.ac.uk/pub/databases/transfac/
</A>

<p>

<!--#include file="inc/tfextract.input" -->


<H2>
    Output file format
</H2>


<!--#include file="inc/tfextract.output" -->

<p>

The output from tfextract is a set of files in the emboss/data
directory containing reformatted sites from the transfac database.

<p>

These files are used by the tfscan program to search for TRANSFAC sites in
sequences.



<H2>
    Data files
</H2>

<!--#include file="inc/localfiles.ihtml" -->



<H2>
    Notes
</H2>

<p>The TRANSFAC Database is a commercial database of eukaryotic cis-acting regulatory DNA elements and trans-acting factors. It covers the whole range from yeast to human. An old public domain version is available at: <a href="ftp://ftp.ebi.ac.uk/pub/databases/transfac/transfac32.tar.Z">ftp://ftp.ebi.ac.uk/pub/databases/transfac/transfac32.tar.Z</a></p>

<p>TRANSFAC started in 1988 with a printed compilation (Nucleic Acids Res. 16: 1879-1902, 1988) and was transferred into computer-readable format in 1990 (BioTechForum - Advances in Molecular Genetics (J. Collins, A.J. Driesel, eds.) 4:95-108, 1991). The basic structures of Table 1 and 2 of the compilation were taken as the core of the emergent database. The aim of the early compilation as well as of the TRANSFAC database is:

   1. to guide through a meanwhile overwhelming amount of data in a field which is connected to nearly all areas of modern molecular biology;
   2. to map the regulatory sites in the individual genes and, ultimately, in the genome(s) as a whole;
   3. to develop a tool for the identification of regulatory elements in newly unravelled genomic sequences;
   4. to provide a basis for a more comprehensive understanding of how the genome governs transcriptional control. 
</p>

<p>The program <b>tfextract</b> extracts data from the TRANSFAC database file <tt>site.dat</tt>. This file contains information on individual (putatively) regulatory protein binding sites. About half of these refer to sites within eukaryotic genes. Just under half of them resulted from mutagenesis studies, in vitro selection procedures starting from random oligonucleotide mixtures or from specific theoretical considerations. And finally, there are about 5% with consensus binding sequences given in the IUPAC code, many of them being taken from the compilation of Faisst and Meyer (Nucleic Acids Res. 20:3-26, 1992). A number of consensi have been generated by the TRANSFAC team, generally derived from the profiles stored in the MATRIX table.</p>


<p>The data is split up by taxonomic groups:<p>
<ul>
<li>Fungi
<li>Insects
<li>Plants
<li>Vertebrates
<li>Other
</ul>
<p>and placed in individual files:</p>
<ul>
<li>tffungi
<li>tfinsect
<li>tfplant
<li>tfvertebrate
<li>tfother
</ul>
<p>These files are stored in the EMBOSS data directory, see Data Files below.</p>




<H2>
    References
</H2>

<ul>

<li>Nucleic Acids Res. 16: 1879-1902, 1988

<li>BioTechForum - Advances in Molecular Genetics (J.  Collins,A.J.
Driesel, eds.) 4:95-108, 1991

<li>Nucleic Acids Res. 20:3-26, 1992

</ul>

<H2>
    Warnings
</H2>

None.



<H2>
    Diagnostic Error Messages
</H2>

None.


<H2>
    Exit status
</H2>

It always exits with a status of 0.

<H2>
    Known bugs
</H2>


None.

<!--#include file="inc/tfextract.isee" -->

<H2>
    Author(s)
</H2>
<!--#include file="inc/ableasby.address" -->

<H2>
    History
</H2>
<!--#include file="inc/tfextract.history" -->

<H2>
    Target users
</H2>
<!--#include file="inc/targetadmin.itxt" -->

<H2>
    Comments
</H2>
<!--#include file="inc/tfextract.comment" -->


</BODY>
</HTML>

