<HTML>

<HEAD>
  <TITLE>
  EMBOSS: trimseq
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<!--#include file="header1.inc" -->
trimseq
<!--#include file="header2.inc" -->

<H2>
    Function
</H2>
<!--#include file="inc/trimseq.ione" -->


<H2>
    Description
</H2>


<p><b>trimseq</b> reads one or more sequences and writes the same sequences out, but removing any regions at the start and / or end that contain unwanted characters. These include gap characters (where occur the sequences have been aligned), <tt>X</tt>'s and <tt>N</tt>'s (in nucleotide sequences), <tt>*</tt>'s (optionally) and (optionally) IUPAC ambiguity codes.</p>




<H2>
    Usage
</H2>
<!--#include file="inc/trimseq.usage" -->

<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
<!--#include file="inc/trimseq.ihelp" -->
</pre>
</td></tr></table>
<P>
<!--#include file="inc/trimseq.itable" -->


<H2>
    Input file format
</H2>
Normal sequence.

<p>
<!--#include file="inc/trimseq.input" -->

<H2>
    Output file format
</H2>


Normal sequence file.
<p>
<!--#include file="inc/trimseq.output" -->

<H2>
    Data files
</H2>

None.
	
<H2>
    Notes
</H2>

<p><b>trimseq</b> will remove from a terminal region:
<ul>
<li> all gap characters 
<li>      X's and N's (in nucleic sequences) 
<li>      *'s (optional)
<li>      IUPAC ambiguity codes: B and Z in proteins, M,R,W,S,Y,K,V,H,D and B in nucleic sequences (optional)</p>
</ul>

<p>Rather than removing individual characters, it removes an entire segment, using a threshold percentage of unwanted characters in a window of a specified size which is moved along the sequence from the ends. The program stops trimming when the percentage of unwanted characters in the moving window drops below the specified threshold percentage.</p>

<p>Thus if the window size is set to 1 and the percentage threshold is 100, no further poor quality regions will be removed. If the window size is set to 5 and the percentage threshold is 40 then the sequence <tt>AAGCTNNNNATT</tt> will be trimmed to <tt>AAGCT</tt>, while <tt>AAGCTNATT</tt> or <tt>AAGCTNNNNATTT</tt> will not be trimmed as less than 40% of the last 5 characters are N's.</p>

<p>After trimming these poor quality regions, it will again then trim off any dangling gap characters from the ends.</p>





<H2>
    References
</H2>


None.

<H2>
    Warnings
</H2>


None.

<H2>
    Diagnostic Error Messages
</H2>


None.

<H2>
    Exit status
</H2>


It always exits with status 0.

<H2>
    Known bugs
</H2>


None noted.

<!--#include file="inc/trimseq.isee" -->

<H2>
    Author(s)
</H2>
<!--#include file="inc/gwilliam.address" -->


<H2>
    History
</H2>
<!--#include file="inc/trimseq.history" -->

<H2>
    Target users
</H2>
<!--#include file="inc/target.itxt" -->


<H2>
    Comments
</H2>
<!--#include file="inc/trimseq.comment" -->

</BODY>
</HTML>

