<html>
<head>
   <meta name="keywords" content="java, biology, bioinformatics, EMBOSS, GUI, Jemboss, computational biology, sequence analysis, interface">
   <title>Jemboss File Manager</title>
</head>
<body text="#000000" bgcolor="#FFFFFF" link="#0000EE" vlink="#551A8B" alink="#FF0000">

<center><img SRC="../images/Jemboss_logo_large.gif" height=110 width=240><br>&nbsp;

</center>


<p>
<h1>Jemboss Alignment Viewer</h1>

This is under development at the
<a href="http://www.hgmp.mrc.ac.uk/Software/EMBOSS/Jemboss/">
HGMP-RC</a>.

Let us know what features
you would like to see this viewer software provide.

<ul>
<li><b><a href="#Load">Loading Sequences</a></b>
<li><b><a href="#Edit">Editing Functions</a></b>
<li><b><a href="#Lock">Locking Sequences</a></b>
<li><b><a href="#Trim">Trim Sequences</a></b>
<li><b><a href="#Colour">Colour Schemes</a></b>
<li><b><a href="#Score">Scoring Matrix</a></b>
<li><b><a href="#Cons">Consensus Sequence</a></b>
<li><b><a href="#Plotcon">Consensus Plot</a></b>
<li><b><a href="#ID">Identity Table</a></b>
</ul>

<a NAME="Load"><h2>Loading Sequences</h2></a>
Supported sequence file formats are FASTA and MSF (so far).
Load the aligned sequences in via the "File" menu and "Open".

<p>
When loaded the alignment appears in the order the sequences
appear in the sequence file. The name is on the right hand
side with the length of each sequence. At the top of the editor
is the relative sequence numbering. A tooltip popup when the
mouse arrow is placed over a site. The tooltip gives the name
of the sequence, the residue and the position of that site.

<a NAME="Edit"><h2>Editing Functions</h2></a>
Gaps can be introduced into a sequence by clicking on a 
site and dragging the mouse to the right. Gaps can similarly
be removed by dragging to the left. Sequences can be deleted
from the the viewer by right mouse clicking on the sequence
and selecting "Delete" from the pop up menu.

<a NAME="Lock">
<p>Sequences can be locked by selecting the name on the right
hand side of the alignment and clicking the 'Lock' button.
When gaps are then introduced into one of these sequences
the gap will also appear in those sequence in this group of
locked sequences. The lock can be switched off with the
'Unlock' button or 'Unlock All Sequences' in the 'Edit' menu.

<a NAME="Trim">
<p>Also under the 'Edit' menu there is an utility to 
trim the ends of sequences. This can be used to tidy up
the termini of an alignment.
 
<a NAME="Colour"><h2>Colour Schemes</h2></a>
There are various colour schemes for the residues these
can be found under the View menu. The residues
can be coloured by there property or using one of the other
common colour schemes (e.g. the <a href="#Taylor">Taylor method</a>). 
These colour schemes can be edited using the "Colour Display"
window and right clicking on a colour to bring up a palette to
select from.

<p>
Alternatively sequences can be coloured based on their
identities and their positive matches. The positive matches
are defined by the scoring matrix selected. This method is based
on the EMBOSS program prettyplot colour scheme.

<a NAME="Score"><h2>Scoring Matrix</h2></a>
The are used to calculate scores in the consensus. A different
scoring matrix can be selected from the "View" menu and clicking
on "Matrix Display". The current scoring matrix being used
is reported at the bottom of the alignment viewer.

<a NAME="Cons"><h2>Consensus Sequence</h2></a>
This uses the same algorithm as used by 
<a href="http://www.hgmp.mrc.ac.uk/Software/EMBOSS/Apps/cons.html">
cons</a> in EMBOSS. The consensus when calculated appears below the 
other sequences in the viewer.

<a NAME="Plotcon"><h2>Consensus Plot</h2></a>
Gives a graphical representation of the consensus. This is
based on
<a href="http://www.hgmp.mrc.ac.uk/Software/EMBOSS/Apps/plotcon.html">
plotcon</a> in EMBOSS.

<a Name="ID"><h2>Identity Table</h2></a>
This can be calculated and gives the percentage of identical 
matches between each pair of sequences. 

<p>
<a NAME="Taylor">W.R.Taylor</a></h2>
Protein residue colour references:
Protein Eng. vol.10 no. 7 pp743-746, 1997


<p>
<hr><i>tcarver@hgmp.mrc.ac.uk</i>
</body>
</html>


